// react-loosely-lazy/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "GlobalReactLooselyLazyProfiler", {
  enumerable: true,
  get: function () {
    return _profiler.GlobalReactLooselyLazyProfiler;
  }
});
Object.defineProperty(exports, "LazySuspense", {
  enumerable: true,
  get: function () {
    return _suspense.LazySuspense;
  }
});
Object.defineProperty(exports, "LazyWait", {
  enumerable: true,
  get: function () {
    return _lazyWait.LazyWait;
  }
});
Object.defineProperty(exports, "MODE", {
  enumerable: true,
  get: function () {
    return _config.MODE;
  }
});
Object.defineProperty(exports, "PRIORITY", {
  enumerable: true,
  get: function () {
    return _lazy.PRIORITY;
  }
});
Object.defineProperty(exports, "ProfilerContext", {
  enumerable: true,
  get: function () {
    return _profiler.ProfilerContext;
  }
});
Object.defineProperty(exports, "WaitContext", {
  enumerable: true,
  get: function () {
    return _lazyWait.WaitContext;
  }
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function () {
    return _init.LooselyLazy;
  }
});
Object.defineProperty(exports, "isLoaderError", {
  enumerable: true,
  get: function () {
    return _lazy.isLoaderError;
  }
});
Object.defineProperty(exports, "lazy", {
  enumerable: true,
  get: function () {
    return _lazy.lazy;
  }
});
Object.defineProperty(exports, "lazyAfterPaint", {
  enumerable: true,
  get: function () {
    return _lazy.lazyAfterPaint;
  }
});
Object.defineProperty(exports, "lazyForPaint", {
  enumerable: true,
  get: function () {
    return _lazy.lazyForPaint;
  }
});
Object.defineProperty(exports, "useLazyPhase", {
  enumerable: true,
  get: function () {
    return _phase.useLazyPhase;
  }
});
var _config = require("./config");
var _init = require("./init");
var _lazy = require("./lazy");
var _lazyWait = require("./lazy-wait");
var _suspense = require("./suspense");
var _phase = require("./phase");
var _profiler = require("./profiler");
