// @atlaskit/jql-editor/dist/cjs/ui/jql-editor-controls-content/syntax-help/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SyntaxHelp = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _tooltip = _interopRequireDefault(require("@atlaskit/tooltip"));
var _accessibility = require("../../../accessibility");
var _analytics = require("../../../analytics");
var _constants = require("../../../common/constants");
var _styled = require("../../../common/styled");
var _state = require("../../../state");
var _tooltipTag = require("../../tooltip-tag");
var _baseSyntaxHelp = require("../base-syntax-help");
var _messages = require("./messages");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
// eslint-disable-next-line @atlassian/tangerine/import/no-parent-imports

// eslint-disable-next-line @atlassian/tangerine/import/no-parent-imports

var SyntaxHelp = exports.SyntaxHelp = function SyntaxHelp() {
  var _useIntl = (0, _state.useIntl)(),
    _useIntl2 = (0, _slicedToArray2.default)(_useIntl, 1),
    intl = _useIntl2[0];
  var label = intl.formatMessage(_messages.messages.syntaxHelpTooltip);
  var _useScopedId = (0, _state.useScopedId)(_constants.SYNTAX_HELP_DESCRIPTION_ID),
    _useScopedId2 = (0, _slicedToArray2.default)(_useScopedId, 1),
    descriptionId = _useScopedId2[0];
  var _useOnSyntaxHelp = (0, _state.useOnSyntaxHelp)(),
    _useOnSyntaxHelp2 = (0, _slicedToArray2.default)(_useOnSyntaxHelp, 2),
    onSyntaxHelp = _useOnSyntaxHelp2[0],
    createAndFireAnalyticsEvent = _useOnSyntaxHelp2[1].createAndFireAnalyticsEvent;
  var onClick = (0, _react.useCallback)(function (e) {
    createAndFireAnalyticsEvent({
      action: _analytics.Action.CLICKED,
      actionSubject: _analytics.ActionSubject.BUTTON,
      actionSubjectId: _analytics.ActionSubjectId.EDITOR_HELP,
      eventType: _analytics.EventType.UI
    });
    var handled = onSyntaxHelp && onSyntaxHelp(e);
    // Prevent default behaviour when the event is handled on the consumer side.
    if (handled) {
      e.preventDefault();
    }
  }, [createAndFireAnalyticsEvent, onSyntaxHelp]);
  return /*#__PURE__*/_react.default.createElement(_tooltip.default, {
    position: 'bottom',
    content: /*#__PURE__*/_react.default.createElement(_styled.TooltipContent, null, label)
    // @ts-ignore: [PIT-1685] Fails in post-office due to backwards incompatibility issue with React 18
    ,
    tag: _tooltipTag.TooltipTag,
    isScreenReaderAnnouncementDisabled: true
  }, /*#__PURE__*/_react.default.createElement(_baseSyntaxHelp.BaseSyntaxHelp, {
    describedby: descriptionId,
    label: label,
    onClick: onClick
  }), /*#__PURE__*/_react.default.createElement(_accessibility.ScreenReaderText, {
    id: descriptionId
  }, intl.formatMessage(_messages.messages.syntaxHelpDescription)));
};
