// @atlaskit/breadcrumbs/dist/cjs/components/ellipsis-item.js
/* ellipsis-item.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./ellipsis-item.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _standardButton = _interopRequireDefault(require("@atlaskit/button/standard-button"));
var _noop = _interopRequireDefault(require("@atlaskit/ds-lib/noop"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
var noop = _noop.default;
var itemWrapperStyles = null;
var staticItemStyles = null;
var EllipsisItem = /*#__PURE__*/(0, _react.memo)(function (_ref) {
  var label = _ref.label,
    _ref$onClick = _ref.onClick,
    onClick = _ref$onClick === void 0 ? noop : _ref$onClick,
    testId = _ref.testId;
  return /*#__PURE__*/React.createElement("li", {
    className: (0, _runtime.ax)(["_1e0c1txw _vchhusvi _p12f1osq _4t3i59ms _2lx2vrvc _6rthze3t _1pfhze3t _12l2ze3t _ahbqze3t _85i5ze3t _1q51ze3t _y4tize3t _bozgze3t _5nddv77o _1j1pu2gc _1dsou2gc _toskidpf _15621n76 _191d1h6o"]),
    style: {
      "--_1fdrbou": (0, _runtime.ix)("".concat(24 / 14, "em"))
    }
  }, /*#__PURE__*/React.createElement(_standardButton.default, {
    appearance: "subtle-link",
    "aria-label": label,
    onClick: onClick,
    spacing: "none",
    testId: testId,
    className: (0, _runtime.ax)(["_1rjc1raa _k48p13dq _vwz41txs"]),
    style: {
      "--_ymfj9u": (0, _runtime.ix)("var(--ds-font-weight-regular, 400)".concat(" !important")),
      "--_lugd7d": (0, _runtime.ix)("var(--ds-space-025, 2px)".concat(" !important"))
    }
  }, "\u2026"));
});

// eslint-disable-next-line @repo/internal/react/require-jsdoc
var _default = exports.default = EllipsisItem;
