// @atlaskit/editor-plugin-mentions/dist/es2019/nodeviews/mention.js
import React from 'react';
import { useSharedPluginState } from '@atlaskit/editor-common/hooks';
import { WithProviders } from '@atlaskit/editor-common/provider-factory';
import { Mention } from '../ui/Mention';
export const MentionNodeView = props => {
  const {
    providerFactory,
    pluginInjectionApi
  } = props;
  const {
    id,
    text,
    accessLevel,
    localId
  } = props.node.attrs;
  const {
    mentionState
  } = useSharedPluginState(pluginInjectionApi, ['mention']);
  const mentionProvider = mentionState === null || mentionState === void 0 ? void 0 : mentionState.mentionProvider;
  const renderAssistiveTextWithProviders = providers => {
    var _props$options;
    const mentionProviderPromise = mentionProvider ? Promise.resolve(mentionProvider) : providers.mentionProvider;
    const profilecardProvider = (_props$options = props.options) === null || _props$options === void 0 ? void 0 : _props$options.profilecardProvider;
    return /*#__PURE__*/React.createElement(Mention, {
      id: id,
      text: text,
      accessLevel: accessLevel,
      mentionProvider: mentionProviderPromise,
      profilecardProvider: profilecardProvider,
      localId: localId
    });
  };
  return /*#__PURE__*/React.createElement(WithProviders, {
    providers: ['mentionProvider', 'profilecardProvider'],
    providerFactory: providerFactory,
    renderNode: renderAssistiveTextWithProviders
  });
};
