// @atlaskit/editor-plugin-mentions/dist/es2019/nodeviews/mentionNodeSpec.js
import { mention } from '@atlaskit/adf-schema';
import { convertToInlineCss } from '@atlaskit/editor-common/lazy-node-view';
import { N30A } from '@atlaskit/theme/colors';
import { editorExperiment } from '@atlaskit/tmp-editor-statsig/experiments';
const isSSR = Boolean(process.env.REACT_SSR);

/**
 * Wrapper for ADF mention node spec to augment toDOM implementation
 * with fallback UI for lazy node view rendering / window virtualization
 * @nodeSpecException:toDOM patch
 * @returns
 */
export const mentionNodeSpec = () => {
  if (isSSR || editorExperiment('platform_editor_inline_node_virtualization', 'off')) {
    return mention;
  }
  return {
    ...mention,
    toDOM: node => {
      // packages/elements/mention/src/components/Mention/index.tsx
      const mentionAttrs = {
        'aria-busy': 'true',
        contenteditable: 'false',
        'data-access-level': '',
        'data-mention-id': node.attrs.id,
        'data-prosemirror-content-type': 'node',
        'data-prosemirror-node-inline': 'true',
        'data-prosemirror-node-name': 'mention',
        style: convertToInlineCss({
          display: 'inline',
          border: `1px solid transparent`,
          background: `var(--ds-background-neutral, ${N30A})`,
          color: "var(--ds-text-subtle, #44546F)",
          borderRadius: '20px',
          cursor: 'pointer',
          padding: '0 0.3em 2px 0.23em',
          wordBreak: 'break-word'
        })
      };
      return ['span', mentionAttrs, node.attrs.text];
    }
  };
};
