// @atlaskit/menu/dist/cjs/menu-item/skeleton-item.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("@emotion/react");
var _deprecationWarning = require("@atlaskit/ds-lib/deprecation-warning");
var _noop = _interopRequireDefault(require("@atlaskit/ds-lib/noop"));
var _colors = require("@atlaskit/theme/colors");
var _constants = require("@atlaskit/theme/constants");
var _skeletonShimmer = _interopRequireDefault(require("../internal/components/skeleton-shimmer"));
/**
 * @jsxRuntime classic
 * @jsx jsx
 */

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var gridSize = (0, _constants.gridSize)();
var borderRadius = (0, _constants.borderRadius)();
var itemElemSpacing = gridSize * 1.5;
var itemExpectedElemSize = gridSize * 3;
var itemMinHeight = gridSize * 5;
var skeletonItemElemSize = gridSize * 2.5;
var itemElemSkeletonOffset = (itemExpectedElemSize - skeletonItemElemSize) / 2;
var skeletonTextBorderRadius = 100;
var skeletonContentHeight = 9;
var skeletonColor = "var(--ds-skeleton, ".concat(_colors.N20A, ")");
var skeletonStyles = (0, _react.css)({
  display: 'flex',
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  minHeight: itemMinHeight,
  padding: "0 ".concat("var(--ds-space-250, 20px)"),
  alignItems: 'center',
  pointerEvents: 'none',
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
  '::after': {
    height: skeletonContentHeight,
    backgroundColor: skeletonColor,
    borderRadius: skeletonTextBorderRadius,
    content: '""',
    // This is a little bespoke but we need to push everything down 1px
    // because the skeleton content should align to the bottom of the text.
    // Confirm VR test failures before accepting a change.
    marginBlockStart: 1
  }
});
var defaultWidthStyles = (0, _react.css)({
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
  ':nth-of-type(1n)::after': {
    flexBasis: '70%'
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
  ':nth-of-type(2n)::after': {
    flexBasis: '50%'
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
  ':nth-of-type(3n)::after': {
    flexBasis: '60%'
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
  ':nth-of-type(4n)::after': {
    flexBasis: '90%'
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
  ':nth-of-type(5n)::after': {
    flexBasis: '35%'
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
  ':nth-of-type(6n)::after': {
    flexBasis: '77%'
  }
});
var customWidthStyles = (0, _react.css)({
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
  '::after': {
    flexBasis: 'var(--width)'
  }
});
var beforeElementStyles = (0, _react.css)({
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
  '::before': {
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    width: skeletonItemElemSize,
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    height: skeletonItemElemSize,
    flexShrink: 0,
    backgroundColor: skeletonColor,
    content: '""',
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    marginInlineEnd: itemElemSpacing + itemElemSkeletonOffset,
    marginInlineStart: "var(--ds-space-025, 2px)"
  }
});
var avatarStyles = (0, _react.css)({
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
  '::before': {
    borderRadius: '100%'
  }
});
var iconStyles = (0, _react.css)({
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
  '::before': {
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    borderRadius: borderRadius
  }
});

/**
 * __Skeleton item__
 *
 * A skeleton item is used in place of an item when its contents it not ready.
 *
 * - [Examples](https://atlaskit.atlassian.com/packages/design-system/menu/docs/skeleton-item)
 * - [Code](https://atlaskit.atlassian.com/packages/design-system/menu)
 */
var SkeletonItem = function SkeletonItem(_ref) {
  var hasAvatar = _ref.hasAvatar,
    hasIcon = _ref.hasIcon,
    _ref$isShimmering = _ref.isShimmering,
    isShimmering = _ref$isShimmering === void 0 ? false : _ref$isShimmering,
    testId = _ref.testId,
    width = _ref.width,
    _ref$cssFn = _ref.cssFn,
    cssFn = _ref$cssFn === void 0 ? _noop.default : _ref$cssFn;
  (0, _deprecationWarning.propDeprecationWarning)("@atlaskit/menu" || '', 'cssFn', cssFn !== _noop.default, '' // TODO: Create DAC post when primitives/xcss are available as alternatives
  );

  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  var UNSAFE_overrides = (0, _react.css)(cssFn());
  return (0, _react.jsx)(_skeletonShimmer.default, {
    isShimmering: isShimmering
  }, function (_ref2) {
    var className = _ref2.className;
    return (0, _react.jsx)("div", {
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
      className: className
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
      ,
      style: {
        '--width': width
        // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
      },
      css: [skeletonStyles, (hasAvatar || hasIcon) && beforeElementStyles, hasAvatar && avatarStyles, hasIcon && iconStyles, width ? customWidthStyles : defaultWidthStyles,
      // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage
      UNSAFE_overrides],
      "data-testid": testId
    });
  });
};
var _default = exports.default = SkeletonItem;
