// @atlaskit/jql-autocomplete/dist/cjs/jql-autocomplete/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.JQLAutocomplete = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _antlr4ts = require("antlr4ts");
var _jqlParser = require("@atlaskit/jql-parser");
var _baseAutocomplete = require("../base-autocomplete");
var _constants = require("./constants");
var _defaults = require("./defaults");
var _ruleContextVisitor = require("./rule-context-visitor");
var _util = require("./util");
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var JQLAutocomplete = exports.JQLAutocomplete = /*#__PURE__*/function (_BaseAutocomplete) {
  (0, _inherits2.default)(JQLAutocomplete, _BaseAutocomplete);
  var _super = _createSuper(JQLAutocomplete);
  /**
   * Compute autocomplete suggestions for the given JQLParser and parse tree. Parse tree is used
   * when determining the context for rule suggestions. When `tree` is undefined (e.g. if parsing
   * failed due to a syntax error) then correct suggestions will still be returned, but contextual
   * information like field/operator will not be included.
   */
  function JQLAutocomplete(parser) {
    var _this;
    var ignoredTokens = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : _defaults.defaultIgnoredTokens;
    var preferredRules = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : _defaults.defaultPreferredRules;
    var delimiterTokens = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : _defaults.defaultDelimiterTokens;
    var tree = arguments.length > 4 ? arguments[4] : undefined;
    (0, _classCallCheck2.default)(this, JQLAutocomplete);
    _this = _super.call(this, parser, ignoredTokens, preferredRules, delimiterTokens);
    _this.tree = tree;
    return _this;
  }

  /**
   * Parse the provided text and return a new {@link JQLAutocomplete} object (**does** include
   * {@link JQLRuleContext} data for rule suggestions).
   *
   * @param text Text to compute suggestions for.
   * @param ignoredTokens Tokens which we do not want to return as suggestions. When not provided, {@link defaultIgnoredTokens} will be used.
   * @param preferredRules Rules we want to return as suggestions. When not provided, {@link defaultPreferredRules} will be used.
   * @param delimiterTokens Tokens to use as delimiters. When not provided, {@link defaultDelimiterTokens} will be used.
   */
  (0, _createClass2.default)(JQLAutocomplete, [{
    key: "assignRuleContextData",
    value:
    /**
     * Walk the parse tree and mutate the provided rules object to add context data for relevant rules.
     *
     * @override
     */
    function assignRuleContextData(maybeCaretToken, rules) {
      var _this2 = this;
      if (this.tree === undefined) {
        return rules;
      }
      rules.forEach(function (_ref, ruleNumber) {
        var _ref2 = (0, _slicedToArray2.default)(_ref, 2),
          ruleSuggestion = _ref2[0],
          ruleList = _ref2[1];
        if (_constants.rulesWithContext.includes(ruleNumber)) {
          var _this2$tree$accept, _this2$tree;
          var ruleContextVisitor = new _ruleContextVisitor.RuleContextVisitor(ruleSuggestion, ruleList, ruleNumber, _this2.getTokenStream(), maybeCaretToken);
          ruleSuggestion.context = (_this2$tree$accept = (_this2$tree = _this2.tree) === null || _this2$tree === void 0 ? void 0 : _this2$tree.accept(ruleContextVisitor)) !== null && _this2$tree$accept !== void 0 ? _this2$tree$accept : null;
        }
      });
      return rules;
    }
  }, {
    key: "getJQLSuggestionsForCaretPosition",
    value: function getJQLSuggestionsForCaretPosition(caretSelectionRange) {
      return this.mapToJQLSuggestions(this.getSuggestionsForCaretPosition(caretSelectionRange));
    }
  }, {
    key: "getJQLSuggestionsForTokenIndex",
    value: function getJQLSuggestionsForTokenIndex(tokenIndex) {
      return this.mapToJQLSuggestions(this.getSuggestionsForTokenIndex(tokenIndex));
    }

    /**
     * Map the provided rule suggestions keyed by parser rule index into {@link JQLRuleSuggestions}
     * using a consumer-friendly rule name.
     *
     * Here we maintain a fixed subset of JQL parse rules which we return as rule suggestions. We map
     * the parse rule index into a predefined string which consumers should use when interpreting
     * rules.
     *
     * Why do we only support a subset of rules? Because otherwise we'd have 2 options:
     * 1. Return all rules keyed by rule index. This means consumers need to import the JQLParser
     * module to interpret the rule and will become more tightly coupled to ANTLR generated
     * dependencies.
     * 2. Return all rules keyed by the rule display name. This will make implementation more fragile
     * as any change to rule names within our grammar has the potential to break all consumers.
     *
     * By only supporting a subset of rules, we can gradually evolve the list as new use cases emerge
     * and ensure that consumers have a stable API they can build upon.
     *
     * @override
     */
  }, {
    key: "mapToJQLSuggestions",
    value: function mapToJQLSuggestions(_ref3) {
      var tokens = _ref3.tokens,
        rules = _ref3.rules;
      var ruleSuggestions = {};
      rules.forEach(function (_ref4, key) {
        var _ref5 = (0, _slicedToArray2.default)(_ref4, 2),
          ruleSuggestion = _ref5[0],
          ruleList = _ref5[1];
        switch (key) {
          case _jqlParser.JQLParser.RULE_jqlField:
            ruleSuggestions.field = ruleSuggestion;
            break;
          case _jqlParser.JQLParser.RULE_jqlCustomField:
            ruleSuggestions.customField = ruleSuggestion;
            break;
          case _jqlParser.JQLParser.RULE_jqlEqualsOperator:
          case _jqlParser.JQLParser.RULE_jqlLikeOperator:
          case _jqlParser.JQLParser.RULE_jqlComparisonOperator:
          case _jqlParser.JQLParser.RULE_jqlInOperator:
          case _jqlParser.JQLParser.RULE_jqlIsOperator:
          case _jqlParser.JQLParser.RULE_jqlWasOperator:
          case _jqlParser.JQLParser.RULE_jqlWasInOperator:
          case _jqlParser.JQLParser.RULE_jqlChangedOperator:
            ruleSuggestions.operator = ruleSuggestion;
            break;
          case _jqlParser.JQLParser.RULE_jqlValue:
            // Disable autocomplete for predicate operands.
            // To be removed when we build proper autocomplete support.
            if (!(0, _util.isPredicateOperand)(ruleList)) {
              ruleSuggestions.value = ruleSuggestion;
            }
            break;
          case _jqlParser.JQLParser.RULE_jqlListStart:
            // Disable autocomplete for predicate operands.
            // To be removed when we build proper autocomplete support.
            if (!(0, _util.isPredicateOperand)(ruleList)) {
              ruleSuggestions.list = ruleSuggestion;
            }
            break;
          case _jqlParser.JQLParser.RULE_jqlFunction:
            // Disable autocomplete for predicate operands.
            // To be removed when we build proper autocomplete support.
            if (!(0, _util.isPredicateOperand)(ruleList)) {
              ruleSuggestions.function = ruleSuggestion;
            }
            break;
          case _jqlParser.JQLParser.RULE_jqlFieldProperty:
            ruleSuggestions.fieldProperty = ruleSuggestion;
            break;
          case _jqlParser.JQLParser.RULE_jqlArgument:
            for (var i = ruleList.length - 1; i >= 0; i--) {
              // The argument rule is shared between functions and field properties. We inspect the
              // rule list to determine which rule our suggestion is scoped to.
              if (ruleList[i] === _jqlParser.JQLParser.RULE_jqlFunction) {
                ruleSuggestions.functionArgument = ruleSuggestion;
                break;
              }
              if (ruleList[i] === _jqlParser.JQLParser.RULE_jqlPropertyArgument) {
                ruleSuggestions.fieldPropertyArgument = ruleSuggestion;
                break;
              }
              if (ruleList[i] === _jqlParser.JQLParser.RULE_jqlFieldProperty) {
                ruleSuggestions.fieldPropertyId = ruleSuggestion;
                break;
              }
            }
            break;
          default:
        }
      });
      return {
        tokens: tokens,
        rules: ruleSuggestions
      };
    }
  }, {
    key: "isCaretAtUnclosedString",
    value: function isCaretAtUnclosedString(caretToken) {
      return _constants.unclosedStringTokens.includes(caretToken.type);
    }

    /**
     *  This method is being overridden for a specific scenario where we encounter string inside an unclosed single quote or double quote.
     *  If this method returns null then the Position calculation will be further processed by getReplacePosition in the base-autocomplete/
     *
     *  @override
     *
     */
  }, {
    key: "overrideReplacePosition",
    value: function overrideReplacePosition(replaceStartToken, maybeCaretToken, caretSelectionRange) {
      if (maybeCaretToken && this.isCaretAtUnclosedString(maybeCaretToken)) {
        var _caretSelectionRange = (0, _slicedToArray2.default)(caretSelectionRange, 2),
          start = _caretSelectionRange[0],
          stop = _caretSelectionRange[1];

        // If the replaceStartToken is not defined (i.e. cursor at beginning of string) or the caret is
        // positioned at a delimiter/hidden token then we want to append at selection start.
        // Otherwise we want our replacement to start from the beginning of specified token.
        var startPosition = replaceStartToken === undefined || this.isCaretAtDelimiterOrHiddenToken(replaceStartToken) ? start : Math.min(replaceStartToken.startIndex, start);
        return [startPosition, stop];
      }
      return null;
    }
  }], [{
    key: "fromText",
    value: function fromText(text, ignoredTokens, preferredRules, delimiterTokens) {
      var charStream = _antlr4ts.CharStreams.fromString(text);
      var lexer = new _jqlParser.JQLLexer(charStream);
      var tokenStream = new _antlr4ts.CommonTokenStream(lexer);
      var parser = new _jqlParser.JQLParser(tokenStream);
      parser.removeErrorListeners();
      return new JQLAutocomplete(parser, ignoredTokens, preferredRules, delimiterTokens, parser.jqlQuery());
    }

    /**
     * Return a new {@link JQLAutocomplete} object using the provided parser instance (**does not**
     * include {@link JQLRuleContext} data for rule suggestions).
     *
     * @param parser Parser instance to compute suggestions for.
     * @param ignoredTokens Tokens which we do not want to return as suggestions. When not provided, {@link defaultIgnoredTokens} will be used.
     * @param preferredRules Rules we want to return as suggestions. When not provided, {@link defaultPreferredRules} will be used.
     * @param delimiterTokens Tokens to use as delimiters. When not provided, {@link defaultDelimiterTokens} will be used.
     */
  }, {
    key: "fromParser",
    value: function fromParser(parser, ignoredTokens, preferredRules, delimiterTokens) {
      return new JQLAutocomplete(parser, ignoredTokens, preferredRules, delimiterTokens);
    }
  }]);
  return JQLAutocomplete;
}(_baseAutocomplete.BaseAutocomplete);
