// @atlaskit/smart-card/dist/es2019/view/FlexibleCard/components/blocks/block/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
/* eslint-disable @atlaskit/ui-styling-standard/no-nested-selectors */
import "./index.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { useEffect } from 'react';
import { fg } from '@atlaskit/platform-feature-flags';
import { SmartLinkDirection, SmartLinkSize } from '../../../../../constants';
import { renderChildren } from '../utils';

// TODO: Remove on fg cleanup: platform-linking-visual-refresh-v1
const BaseBlockStylesOld = null;
const baseBlockStyles = null;
const highlightRemoveStyles = null;

// TODO: Remove on fg cleanup: platform-linking-visual-refresh-v1
const gapsStylesOld = {
  xlarge: "_zulp1k92 _1ljq1k92",
  large: "_zulp1j6v _1ljq1j6v",
  medium: "_zulp1kwk _1ljq1kwk",
  small: "_zulp6mu8 _1ljq6mu8"
};
const gapStyles = {
  xlarge: "_zulpv47k _1ljqv47k",
  large: "_zulppxbi _1ljqpxbi",
  medium: "_zulpu2gc _1ljqu2gc",
  small: "_zulp1b66 _1ljq1b66"
};
const directionStyles = {
  vertical: "_2lx21bp4 _4cvr1y6m",
  horizontal: "_2lx2vrvc _4cvr1h6o"
};

/**
 * A block represents a collection of elements and actions that are arranged
 * in a row. All elements and actions should be contained within a Block.
 * @internal
 * @param {ActionGroupProps} ActionGroupProps
 * @see Action
 */
const Block = ({
  children,
  direction = SmartLinkDirection.Horizontal,
  size = SmartLinkSize.Medium,
  testId = 'smart-block',
  className,
  blockRef,
  onRender,
  onTransitionEnd,
  style
}) => {
  useEffect(() => {
    onRender && onRender();
  }, [onRender]);
  return /*#__PURE__*/React.createElement("div", {
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop
    className: ax([!fg('platform-linking-visual-refresh-v1') && "_1reo15vq _18m915vq _4cvr1h6o _vwz41j6v _1e0c1txw _1ul9idpf _1bah1y6m _fiawglyw _1ouwidpf _f6ju1ns9 _1uzp1mp2 _ytq1bjq5 _a7e8uxlv", fg('platform-linking-visual-refresh-v1') && "_1reo15vq _18m915vq _4cvr1h6o _1e0c1txw _1ul9idpf _1bah1y6m _fiawglyw _1uzp1mp2 _a6bxmuej _e4i5bjq5 _1gu21oju _5b7tuxlv _17ae1oju", "_12ji1snc _1qu2agmp _12y310yn _12ji1kw7 _syaz1kw7 _18ql1j28 _1ifmglyw _181uglyw _1rkwglyw _te17glyw _uiztglyw", directionStyles[direction], !fg('platform-linking-visual-refresh-v1') && gapsStylesOld[size], fg('platform-linking-visual-refresh-v1') && gapStyles[size], className]),
    "data-smart-block": true,
    "data-testid": testId,
    onTransitionEnd: onTransitionEnd,
    ref: blockRef
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
    ,
    style: style
  }, renderChildren(children, size));
};
export default Block;
