// react-loosely-lazy/dist/esm/lazy/index.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
import { getAssetUrlsFromId } from '@react-loosely-lazy/manifest';
import { getConfig } from '../config';
import { PHASE } from '../constants';
import { hash, displayNameFromId, isNodeEnvironment } from '../utils';
import { PRIORITY, RETRY_DELAY, RETRY_FACTOR } from './constants';
import { createComponentClient } from './components/client';
import { createComponentServer } from './components/server';
import { createDeferred } from './deferred';
import { preloadAsset } from './preload';
import { retry } from './retry';
export { PRIORITY };
function lazyProxy(_loader, {
  defer = PHASE.PAINT,
  moduleId = '',
  ssr = true
} = {}) {
  const isServer = isNodeEnvironment();
  const dataLazyId = hash(moduleId);
  const LazyInternal = isServer ? createComponentServer({
    dataLazyId,
    defer,
    loader: _loader,
    moduleId,
    ssr
  }) : createComponentClient({
    dataLazyId,
    defer,
    // We separate the preload from the retryable request, as we do not want the preload to contribute to retry
    // attempts, when the the loader fallback is used
    deferred: createDeferred({
      loader: () => {
        const {
          retry: maxAttempts
        } = getConfig();
        return retry(_loader, {
          delay: RETRY_DELAY,
          factor: RETRY_FACTOR,
          maxAttempts
        });
      },
      preload: _loader
    }),
    moduleId
  });
  LazyInternal.displayName = `Lazy(${displayNameFromId(moduleId)})`;

  /**
   * Allows getting module chunks urls
   */
  const getAssetUrls = () => {
    const {
      manifest
    } = getConfig();
    return getAssetUrlsFromId(manifest, moduleId);
  };

  /**
   * Allows imperatively preload/ prefetch the module chunk asset
   */
  const preload = priority => {
    const p = priority != null ? priority : defer === PHASE.PAINT ? PRIORITY.HIGH : PRIORITY.LOW;
    return preloadAsset({
      loader: _loader,
      moduleId,
      priority: p
    });
  };
  return Object.assign(LazyInternal, {
    getAssetUrls,
    preload
  });
}
export const DEFAULT_OPTIONS = {
  lazyForPaint: {
    ssr: true,
    defer: PHASE.PAINT
  },
  lazyAfterPaint: {
    ssr: true,
    defer: PHASE.AFTER_PAINT
  },
  lazy: {
    ssr: false,
    defer: PHASE.LAZY
  }
};
export function lazyForPaint(loader, opts) {
  return lazyProxy(loader, _extends({}, DEFAULT_OPTIONS.lazyForPaint, opts || {}));
}
export function lazyAfterPaint(loader, opts) {
  return lazyProxy(loader, _extends({}, DEFAULT_OPTIONS.lazyAfterPaint, opts || {}));
}
export function lazy(loader, opts) {
  return lazyProxy(loader, _extends({}, DEFAULT_OPTIONS.lazy, opts || {}));
}
export { isLoaderError } from './errors';
