// @atlaskit/media-viewer/dist/cjs/viewers/doc/pdfRendererWrapper-emotion.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PDFRendererWrapper = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = require("@emotion/react");
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-global-styles, @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var pdfViewerClassName = 'pdfViewer';

// Styles are partially copied from https://github.com/mozilla/pdfjs-dist/blob/v2.9.359/web/pdf_viewer.css
/* eslint-disable @atlaskit/design-system/ensure-design-token-usage */
var globalStyles = (0, _react.css)((0, _defineProperty2.default)({}, ".".concat(pdfViewerClassName), {
  marginTop: "var(--ds-space-800, 64px)",
  marginBottom: "var(--ds-space-800, 64px)",
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '.page': {
    margin: "1px auto ".concat("var(--ds-space-negative-100, -8px)", " auto"),
    border: '9px solid transparent',
    position: 'relative'
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '.canvasWrapper': {
    overflow: 'hidden'
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '.textLayer': (0, _defineProperty2.default)({
    position: 'absolute',
    left: 0,
    top: 0,
    right: 0,
    bottom: 0,
    overflow: 'hidden',
    opacity: 0.2
  }, 'lineHeight', 1),
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '.textLayer span, .textLayer br': {
    color: 'transparent',
    position: 'absolute',
    whiteSpace: 'pre',
    cursor: 'text',
    transformOrigin: '0% 0%'
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors, @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
  '.textLayer ::-moz-selection': {
    background: 'rgba(0, 0, 255, 1)'
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '.textLayer ::selection': {
    background: 'rgba(0, 0, 255, 1)'
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '.annotationLayer section': {
    position: 'absolute',
    textAlign: 'initial'
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '.annotationLayer .linkAnnotation > a, .annotationLayer .buttonWidgetAnnotation.pushButton > a': {
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    height: '100%'
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '.annotationLayer .linkAnnotation > a:hover, .annotationLayer .buttonWidgetAnnotation.pushButton > a:hover': {
    opacity: 0.2,
    background: 'rgba(255, 255, 0, 1)',
    boxShadow: '0 2px 10px rgba(255, 255, 0, 1)'
  }
}));
var PDFRendererWrapper = exports.PDFRendererWrapper = function PDFRendererWrapper(_ref) {
  var children = _ref.children;
  return (0, _react.jsx)("div", {
    css: globalStyles
  }, children);
};
