// @atlaskit/editor-common/dist/cjs/ui/Mention/mention-with-providers.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MentionWithProvidersOld = exports.MentionWithProviders = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _element = require("@atlaskit/mention/element");
var _mentionWithProfilecard = _interopRequireDefault(require("./mention-with-profilecard"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var GENERIC_USER_IDS = ['HipChat', 'all', 'here'];
var noop = function noop() {};
var MentionWithProviders = exports.MentionWithProviders = function MentionWithProviders(_ref) {
  var accessLevel = _ref.accessLevel,
    eventHandlers = _ref.eventHandlers,
    id = _ref.id,
    mentionProvider = _ref.mentionProvider,
    profilecardProviderResolver = _ref.profilecardProvider,
    text = _ref.text,
    localId = _ref.localId;
  var _useState = (0, _react.useState)(null),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    profilecardProvider = _useState2[0],
    setProfilecardProvider = _useState2[1];
  var mountedRef = (0, _react.useRef)(true);
  (0, _react.useLayoutEffect)(function () {
    mountedRef.current = true;
    return function () {
      mountedRef.current = false;
    };
  }, []);
  (0, _react.useLayoutEffect)(function () {
    // We are not using async/await here to avoid having an intermediate Promise
    // introduced by the transpiler.
    // This will allow consumer to use a SynchronousPromise.resolve and avoid useless
    // rerendering
    profilecardProviderResolver === null || profilecardProviderResolver === void 0 || profilecardProviderResolver.then(function (result) {
      if (mountedRef.current) {
        setProfilecardProvider(result);
      }
    }).catch(function () {
      if (mountedRef.current) {
        setProfilecardProvider(null);
      }
    });
  }, [profilecardProviderResolver]);
  var MentionComponent = profilecardProvider && profilecardProviderResolver && GENERIC_USER_IDS.indexOf(id) === -1 ? _mentionWithProfilecard.default : _element.ResourcedMention;
  return /*#__PURE__*/_react.default.createElement(MentionComponent, {
    id: id,
    text: text,
    accessLevel: accessLevel,
    localId: localId,
    mentionProvider: mentionProvider
    // Ignored via go/ees005
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    ,
    profilecardProvider: profilecardProvider,
    onClick: eventHandlers === null || eventHandlers === void 0 ? void 0 : eventHandlers.onClick,
    onMouseEnter: eventHandlers === null || eventHandlers === void 0 ? void 0 : eventHandlers.onMouseEnter,
    onMouseLeave: eventHandlers === null || eventHandlers === void 0 ? void 0 : eventHandlers.onMouseLeave
  });
};

// eslint-disable-next-line @repo/internal/react/no-class-components
var MentionWithProvidersOld = exports.MentionWithProvidersOld = /*#__PURE__*/function (_PureComponent) {
  function MentionWithProvidersOld() {
    var _this;
    (0, _classCallCheck2.default)(this, MentionWithProvidersOld);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, MentionWithProvidersOld, [].concat(args));
    (0, _defineProperty2.default)(_this, "state", {
      profilecardProvider: null
    });
    return _this;
  }
  (0, _inherits2.default)(MentionWithProvidersOld, _PureComponent);
  return (0, _createClass2.default)(MentionWithProvidersOld, [{
    key: "UNSAFE_componentWillMount",
    value:
    // Ignored via go/ees005
    // eslint-disable-next-line react/no-unsafe
    function UNSAFE_componentWillMount() {
      this.updateProfilecardProvider(this.props);
    }

    // Ignored via go/ees005
    // eslint-disable-next-line react/no-unsafe
  }, {
    key: "UNSAFE_componentWillReceiveProps",
    value: function UNSAFE_componentWillReceiveProps(nextProps) {
      if (nextProps.profilecardProvider !== this.props.profilecardProvider) {
        this.updateProfilecardProvider(nextProps);
      }
    }
  }, {
    key: "updateProfilecardProvider",
    value: function updateProfilecardProvider(props) {
      var _this2 = this;
      // We are not using async/await here to avoid having an intermediate Promise
      // introduced by the transpiler.
      // This will allow consumer to use a SynchronousPromise.resolve and avoid useless
      // rerendering
      if (props.profilecardProvider) {
        props.profilecardProvider.then(function (profilecardProvider) {
          _this2.setState({
            profilecardProvider: profilecardProvider
          });
        }).catch(function () {
          _this2.setState({
            profilecardProvider: null
          });
        });
      } else {
        this.setState({
          profilecardProvider: null
        });
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
        accessLevel = _this$props.accessLevel,
        eventHandlers = _this$props.eventHandlers,
        id = _this$props.id,
        mentionProvider = _this$props.mentionProvider,
        text = _this$props.text,
        localId = _this$props.localId;
      var profilecardProvider = this.state.profilecardProvider;

      // Ignored via go/ees005
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      var actionHandlers = {};
      ['onClick', 'onMouseEnter', 'onMouseLeave'].forEach(function (handler) {
        actionHandlers[handler] = eventHandlers && eventHandlers[handler] || noop;
      });
      var MentionComponent = profilecardProvider && GENERIC_USER_IDS.indexOf(id) === -1 ? _mentionWithProfilecard.default : _element.ResourcedMention;
      return /*#__PURE__*/_react.default.createElement(MentionComponent, (0, _extends2.default)({
        id: id,
        text: text,
        accessLevel: accessLevel,
        localId: localId,
        mentionProvider: mentionProvider
        // Ignored via go/ees005
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        ,
        profilecardProvider: profilecardProvider
        // Ignored via go/ees005
        // eslint-disable-next-line react/jsx-props-no-spreading
      }, actionHandlers));
    }
  }]);
}(_react.PureComponent);
