// @atlaskit/editor-core/dist/cjs/ui/EditorContext/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useEditorContext = exports.default = exports.LegacyEditorContext = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _UNSAFE_do_not_use_editor_context = require("@atlaskit/editor-common/UNSAFE_do_not_use_editor_context");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _actions = _interopRequireDefault(require("../../actions"));
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var useEditorContext = exports.useEditorContext = function useEditorContext() {
  return _react.default.useContext(_UNSAFE_do_not_use_editor_context.EditorContext);
};

// Ignored via go/ees005
// eslint-disable-next-line @repo/internal/react/no-class-components, react/prefer-stateless-function
var LegacyEditorContext = exports.LegacyEditorContext = /*#__PURE__*/function (_React$Component) {
  function LegacyEditorContext(props) {
    (0, _classCallCheck2.default)(this, LegacyEditorContext);
    return _callSuper(this, LegacyEditorContext, [props]);
  }
  (0, _inherits2.default)(LegacyEditorContext, _React$Component);
  return (0, _createClass2.default)(LegacyEditorContext, [{
    key: "render",
    value: function render() {
      if ((0, _platformFeatureFlags.fg)('platform_editor_react18_phase2_v2') || (0, _platformFeatureFlags.fg)('platform_editor_react18_phase2_v2_extended')) {
        // Ignored via go/ees005
        // eslint-disable-next-line react/jsx-props-no-spreading
        return /*#__PURE__*/_react.default.createElement(LegacyEditorContextNew, this.props, this.props.children);
      }

      // Ignored via go/ees005
      // eslint-disable-next-line react/jsx-props-no-spreading
      return /*#__PURE__*/_react.default.createElement(LegacyEditorContextOld, this.props, this.props.children);
    }
  }]);
}(_react.default.Component);
function LegacyEditorContextNew(_ref) {
  var children = _ref.children,
    editorActions = _ref.editorActions;
  return /*#__PURE__*/_react.default.createElement(_UNSAFE_do_not_use_editor_context.EditorContext.Provider, {
    value: {
      editorActions: editorActions !== null && editorActions !== void 0 ? editorActions : new _actions.default()
    }
  }, children);
}

// Ignored via go/ees005
// eslint-disable-next-line @repo/internal/react/no-class-components
var LegacyEditorContextOld = exports.default = /*#__PURE__*/function (_React$Component2) {
  function LegacyEditorContextOld(props) {
    var _this;
    (0, _classCallCheck2.default)(this, LegacyEditorContextOld);
    _this = _callSuper(this, LegacyEditorContextOld, [props]);
    _this.editorActions = props.editorActions || new _actions.default();
    return _this;
  }
  (0, _inherits2.default)(LegacyEditorContextOld, _React$Component2);
  return (0, _createClass2.default)(LegacyEditorContextOld, [{
    key: "getChildContext",
    value: function getChildContext() {
      return {
        editorActions: this.editorActions
      };
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement(_UNSAFE_do_not_use_editor_context.EditorContext.Provider, {
        value: this.getChildContext()
      }, this.props.children);
    }
  }]);
}(_react.default.Component);
(0, _defineProperty2.default)(LegacyEditorContextOld, "childContextTypes", {
  editorActions: _propTypes.default.object
});
