// @atlaskit/editor-plugin-table/dist/cjs/ui/TableFloatingColumnControls/ColumnDropTargets/index.js
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ColumnDropTargets = void 0;
var _react = _interopRequireWildcard(require("react"));
var _types = require("../../../types");
var _ColumnDropTarget = require("./ColumnDropTarget");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var ColumnDropTargets = exports.ColumnDropTargets = function ColumnDropTargets(_ref) {
  var tableRef = _ref.tableRef,
    tableHeight = _ref.tableHeight,
    localId = _ref.localId,
    colWidths = _ref.colWidths,
    isHeaderSticky = _ref.isHeaderSticky,
    getScrollOffset = _ref.getScrollOffset;
  var dropTargetRef = (0, _react.useRef)(null);
  if (!tableRef) {
    return null;
  }
  if (isHeaderSticky && dropTargetRef.current) {
    var _getScrollOffset;
    dropTargetRef.current.style.marginLeft = "-".concat((_getScrollOffset = getScrollOffset === null || getScrollOffset === void 0 ? void 0 : getScrollOffset()) !== null && _getScrollOffset !== void 0 ? _getScrollOffset : 0, "px");
  }
  return /*#__PURE__*/_react.default.createElement("div", {
    ref: dropTargetRef
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    ,
    className: _types.TableCssClassName.DRAG_COLUMN_DROP_TARGET_CONTROLS,
    contentEditable: false
  }, /*#__PURE__*/_react.default.createElement("div", {
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    className: _types.TableCssClassName.DRAG_COLUMN_CONTROLS_INNER,
    "data-testid": "table-floating-column-controls-drop-targets"
  }, colWidths === null || colWidths === void 0 ? void 0 : colWidths.map(function (width, index) {
    return /*#__PURE__*/_react.default.createElement(_ColumnDropTarget.ColumnDropTarget
    // Ignored via go/ees005
    // eslint-disable-next-line react/no-array-index-key
    , {
      key: index,
      index: index,
      localId: localId,
      width: width,
      height: tableHeight,
      marginTop: 0
    });
  })));
};
