// @atlaskit/react-select/dist/cjs/components/containers.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.valueContainerCSS = exports.indicatorsContainerCSS = exports.containerCSS = exports.ValueContainer = exports.SelectContainer = exports.IndicatorsContainer = void 0;
var _react = _interopRequireDefault(require("react"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _containers = require("../compiled/components/containers");
var _containers2 = require("../emotion/components/containers");
/* eslint-disable @repo/internal/react/no-unsafe-spread-props */

// ==============================
// Root Container
// ==============================

var containerCSS = exports.containerCSS = function containerCSS(props) {
  return (0, _platformFeatureFlags.fg)('compiled-react-select') ? (0, _containers.containerCSS)() : (0, _containers2.containerCSS)(props);
};

// eslint-disable-next-line @repo/internal/react/require-jsdoc
var SelectContainer = exports.SelectContainer = function SelectContainer(props) {
  return (0, _platformFeatureFlags.fg)('compiled-react-select') ? /*#__PURE__*/_react.default.createElement(_containers.SelectContainer, props) : /*#__PURE__*/_react.default.createElement(_containers2.SelectContainer, props);
};

// ==============================
// Value Container
// ==============================

var valueContainerCSS = exports.valueContainerCSS = function valueContainerCSS(props) {
  return (0, _platformFeatureFlags.fg)('compiled-react-select') ? (0, _containers.valueContainerCSS)() : (0, _containers2.valueContainerCSS)(props);
};

// eslint-disable-next-line @repo/internal/react/require-jsdoc
var ValueContainer = exports.ValueContainer = function ValueContainer(props) {
  return (0, _platformFeatureFlags.fg)('compiled-react-select') ? /*#__PURE__*/_react.default.createElement(_containers.ValueContainer, props) : /*#__PURE__*/_react.default.createElement(_containers2.ValueContainer, props);
};

// ==============================
// Indicator Container
// ==============================

var indicatorsContainerCSS = exports.indicatorsContainerCSS = function indicatorsContainerCSS() {
  return (0, _platformFeatureFlags.fg)('compiled-react-select') ? (0, _containers.indicatorsContainerCSS)() : (0, _containers2.indicatorsContainerCSS)();
};

// eslint-disable-next-line @repo/internal/react/require-jsdoc
var IndicatorsContainer = exports.IndicatorsContainer = function IndicatorsContainer(props) {
  return (0, _platformFeatureFlags.fg)('compiled-react-select') ? /*#__PURE__*/_react.default.createElement(_containers.IndicatorsContainer, props) : /*#__PURE__*/_react.default.createElement(_containers2.IndicatorsContainer, props);
};
