// @atlaskit/react-select/dist/cjs/components/indicators.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.loadingIndicatorCSS = exports.dropdownIndicatorCSS = exports.clearIndicatorCSS = exports.LoadingIndicator = exports.DropdownIndicator = exports.ClearIndicator = void 0;
var _react = _interopRequireDefault(require("react"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _indicators = require("../compiled/components/indicators");
var _indicators2 = require("../emotion/components/indicators");
/* eslint-disable @repo/internal/react/no-unsafe-spread-props */

// ==============================
// Dropdown & Clear Buttons
// ==============================

var dropdownIndicatorCSS = exports.dropdownIndicatorCSS = function dropdownIndicatorCSS(props) {
  return (0, _platformFeatureFlags.fg)('compiled-react-select') ? (0, _indicators.dropdownIndicatorCSS)() : (0, _indicators2.dropdownIndicatorCSS)(props);
};

// eslint-disable-next-line @repo/internal/react/require-jsdoc
var DropdownIndicator = exports.DropdownIndicator = function DropdownIndicator(props) {
  return (0, _platformFeatureFlags.fg)('compiled-react-select') ? /*#__PURE__*/_react.default.createElement(_indicators.DropdownIndicator, props) : /*#__PURE__*/_react.default.createElement(_indicators2.DropdownIndicator, props);
};
var clearIndicatorCSS = exports.clearIndicatorCSS = function clearIndicatorCSS(props) {
  return (0, _platformFeatureFlags.fg)('compiled-react-select') ? (0, _indicators.clearIndicatorCSS)() : (0, _indicators2.clearIndicatorCSS)(props);
};

// eslint-disable-next-line @repo/internal/react/require-jsdoc
var ClearIndicator = exports.ClearIndicator = function ClearIndicator(props) {
  return (0, _platformFeatureFlags.fg)('compiled-react-select') ? /*#__PURE__*/_react.default.createElement(_indicators.ClearIndicator, props) : /*#__PURE__*/_react.default.createElement(_indicators2.ClearIndicator, props);
};

// ==============================
// Loading
// ==============================

var loadingIndicatorCSS = exports.loadingIndicatorCSS = function loadingIndicatorCSS(props) {
  return (0, _platformFeatureFlags.fg)('compiled-react-select') ? (0, _indicators.loadingIndicatorCSS)() : (0, _indicators2.loadingIndicatorCSS)(props);
};
// eslint-disable-next-line @repo/internal/react/require-jsdoc
var LoadingIndicator = exports.LoadingIndicator = function LoadingIndicator(props) {
  return (0, _platformFeatureFlags.fg)('compiled-react-select') ? /*#__PURE__*/_react.default.createElement(_indicators.LoadingIndicator, props) : /*#__PURE__*/_react.default.createElement(_indicators2.LoadingIndicator, props);
};
