// @atlaskit/media-file-preview/dist/cjs/errors.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isLocalPreviewError = exports.ensureMediaFilePreviewError = exports.SsrPreviewError = exports.RemotePreviewError = exports.MediaFilePreviewError = exports.LocalPreviewError = exports.ImageLoadError = void 0;
exports.isMediaFilePreviewError = isMediaFilePreviewError;
exports.isUnsupportedLocalPreviewError = exports.isRemotePreviewError = void 0;
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _wrapNativeSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/wrapNativeSuper"));
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
/**
 * Primary reason is logged through Data Portal.
 * Make sure all the values are whitelisted in Measure -> Event Regitry -> "mediaCardRender failed" event
 */
var MediaFilePreviewError = exports.MediaFilePreviewError = /*#__PURE__*/function (_Error) {
  function MediaFilePreviewError(primaryReason, secondaryError) {
    var _this;
    (0, _classCallCheck2.default)(this, MediaFilePreviewError);
    _this = _callSuper(this, MediaFilePreviewError, [primaryReason]);
    // https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-2.html#support-for-newtarget
    _this.primaryReason = primaryReason;
    _this.secondaryError = secondaryError;
    Object.setPrototypeOf(_this, (this instanceof MediaFilePreviewError ? this.constructor : void 0).prototype);

    // https://v8.dev/docs/stack-trace-api
    if ('captureStackTrace' in Error) {
      Error.captureStackTrace(_this, this instanceof MediaFilePreviewError ? this.constructor : void 0);
    }
    return _this;
  }
  (0, _inherits2.default)(MediaFilePreviewError, _Error);
  return (0, _createClass2.default)(MediaFilePreviewError);
}( /*#__PURE__*/(0, _wrapNativeSuper2.default)(Error));
var LocalPreviewError = exports.LocalPreviewError = /*#__PURE__*/function (_MediaFilePreviewErro) {
  function LocalPreviewError(primaryReason, secondaryError) {
    var _this2;
    (0, _classCallCheck2.default)(this, LocalPreviewError);
    _this2 = _callSuper(this, LocalPreviewError, [primaryReason, secondaryError]);
    _this2.primaryReason = primaryReason;
    _this2.secondaryError = secondaryError;
    return _this2;
  }
  (0, _inherits2.default)(LocalPreviewError, _MediaFilePreviewErro);
  return (0, _createClass2.default)(LocalPreviewError);
}(MediaFilePreviewError);
var RemotePreviewError = exports.RemotePreviewError = /*#__PURE__*/function (_MediaFilePreviewErro2) {
  function RemotePreviewError(primaryReason, secondaryError) {
    var _this3;
    (0, _classCallCheck2.default)(this, RemotePreviewError);
    _this3 = _callSuper(this, RemotePreviewError, [primaryReason, secondaryError]);
    _this3.primaryReason = primaryReason;
    _this3.secondaryError = secondaryError;
    return _this3;
  }
  (0, _inherits2.default)(RemotePreviewError, _MediaFilePreviewErro2);
  return (0, _createClass2.default)(RemotePreviewError);
}(MediaFilePreviewError);
var SsrPreviewError = exports.SsrPreviewError = /*#__PURE__*/function (_MediaFilePreviewErro3) {
  function SsrPreviewError(primaryReason, secondaryError) {
    var _this4;
    (0, _classCallCheck2.default)(this, SsrPreviewError);
    _this4 = _callSuper(this, SsrPreviewError, [primaryReason, secondaryError]);
    _this4.primaryReason = primaryReason;
    _this4.secondaryError = secondaryError;
    return _this4;
  }
  (0, _inherits2.default)(SsrPreviewError, _MediaFilePreviewErro3);
  return (0, _createClass2.default)(SsrPreviewError);
}(MediaFilePreviewError);
var getImageLoadPrimaryReason = function getImageLoadPrimaryReason(source) {
  switch (source) {
    case 'cache-remote':
      return 'cache-remote-uri';
    case 'cache-local':
      return 'cache-local-uri';
    case 'external':
      return 'external-uri';
    case 'local':
      return 'local-uri';
    case 'remote':
      return 'remote-uri';
    case 'ssr-client':
      return 'ssr-client-uri';
    case 'ssr-server':
    case 'ssr-data':
      return 'ssr-server-uri';
    // This fail reason will come from a bug, most likely.
    default:
      return "unknown-uri";
  }
};
var ImageLoadError = exports.ImageLoadError = /*#__PURE__*/function (_MediaFilePreviewErro4) {
  function ImageLoadError(source) {
    (0, _classCallCheck2.default)(this, ImageLoadError);
    return _callSuper(this, ImageLoadError, [getImageLoadPrimaryReason(source)]);
  }
  (0, _inherits2.default)(ImageLoadError, _MediaFilePreviewErro4);
  return (0, _createClass2.default)(ImageLoadError);
}(MediaFilePreviewError);
function isMediaFilePreviewError(err) {
  return err instanceof MediaFilePreviewError;
}
var isLocalPreviewError = exports.isLocalPreviewError = function isLocalPreviewError(err) {
  return err instanceof LocalPreviewError;
};
var isRemotePreviewError = exports.isRemotePreviewError = function isRemotePreviewError(err) {
  return err instanceof RemotePreviewError;
};
var isUnsupportedLocalPreviewError = exports.isUnsupportedLocalPreviewError = function isUnsupportedLocalPreviewError(err) {
  return isMediaFilePreviewError(err) && err.primaryReason === 'local-preview-unsupported';
};

// In a try/catch statement, the error caught is the type of unknown.
// We can use this helper to ensure that the error handled is the type of MediaFilePreviewError if unsure
// If updatePrimaryReason is true, if it's a MediaFilePreviewError already, it will update it's primary reason
var ensureMediaFilePreviewError = exports.ensureMediaFilePreviewError = function ensureMediaFilePreviewError(primaryReason, error, updatePrimaryReason) {
  if (isMediaFilePreviewError(error)) {
    if (updatePrimaryReason && error.primaryReason !== primaryReason) {
      return new MediaFilePreviewError(primaryReason, error.secondaryError);
    }
    return error;
  }
  return new MediaFilePreviewError(primaryReason, error);
};
