// @atlaskit/link-picker/dist/es2019/common/analytics/experiences.js
import { ConcurrentExperience, ExperiencePerformanceTypes, ExperienceTypes } from '@atlaskit/ufo';
export const ufoExperience = {
  mounted: 'component-mounted'
};
const ufoExperiences = {
  [ufoExperience.mounted]: new ConcurrentExperience(ufoExperience.mounted, {
    type: ExperienceTypes.Load,
    performanceType: ExperiencePerformanceTypes.PageSegmentLoad,
    platform: {
      component: 'link-picker'
    }
  })
};
export const startUfoExperience = (experienceName, id, metadata) => {
  const experience = ufoExperiences[experienceName].getInstance(id);
  experience.start();
  if (metadata) {
    experience.addMetadata(metadata);
  }
};
export const succeedUfoExperience = (experienceName, id, metadata) => {
  const experience = ufoExperiences[experienceName].getInstance(id);
  experience.success({
    metadata
  });
};
export const failUfoExperience = (experienceName, id, metadata) => {
  const experience = ufoExperiences[experienceName].getInstance(id);
  experience.failure({
    metadata
  });
};
export const abortUfoExperience = (experienceName, id, metadata) => {
  const experience = ufoExperiences[experienceName].getInstance(id);
  experience.abort({
    metadata
  });
};
