// @atlaskit/smart-card/dist/es2019/view/FlexibleCard/components/container/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
/* eslint-disable @atlaskit/design-system/consistent-css-prop-usage */
/* eslint-disable @atlaskit/ui-styling-standard/no-unsafe-values */
import "./index.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
import React, { useContext } from 'react';
import { di } from 'react-magnetic-di';
import { fg } from '@atlaskit/platform-feature-flags';
import { N40 } from '@atlaskit/theme/colors';
import { MediaPlacement, SmartLinkSize, SmartLinkTheme } from '../../../../constants';
import { FlexibleUiContext } from '../../../../state/flexible-ui-context';
import { isFlexUiPreviewPresent } from '../../../../state/flexible-ui-context/utils';
import { isFlexibleUiBlock, isFlexibleUiPreviewBlock, isFlexibleUiTitleBlock, isStyleCacheProvider } from '../../../../utils/flexible';
import { TitleBlock } from '../blocks';
import HoverCardControl from './hover-card-control';
import LayeredLink from './layered-link';
export const getChildrenOptions = (children, context) => {
  let options = {};
  if (isFlexUiPreviewPresent(context)) {
    React.Children.map(children, child => {
      if ( /*#__PURE__*/React.isValidElement(child)) {
        if (isFlexibleUiPreviewBlock(child)) {
          const {
            placement
          } = child.props;
          if (placement === MediaPlacement.Left) {
            options.previewOnLeft = true;
          }
          if (placement === MediaPlacement.Right) {
            options.previewOnRight = true;
          }
        }
      }
    });
  }
  return options;
};
const renderChildren = (children, containerSize, containerTheme, status, retry, onClick, removeBlockRestriction) => {
  return React.Children.map(children, child => {
    if ( /*#__PURE__*/React.isValidElement(child) && isFlexibleUiBlock(child)) {
      if (isFlexibleUiTitleBlock(child)) {
        if (isStyleCacheProvider(child)) {
          return updateChildrenTitleBlock(child, {
            onClick,
            retry,
            containerSize,
            status,
            theme: containerTheme
          });
        } else {
          const {
            size: blockSize
          } = child.props;
          const size = blockSize || containerSize;
          return /*#__PURE__*/React.cloneElement(child, {
            // @ts-expect-error
            onClick,
            retry,
            size,
            status,
            theme: containerTheme
          });
        }
      }
      const {
        size: blockSize
      } = child.props;
      const size = blockSize || containerSize;
      if (isStyleCacheProvider(child)) {
        return updateChildrenBlock(child, size, status);
      }

      // @ts-expect-error
      return /*#__PURE__*/React.cloneElement(child, {
        size,
        status
      });
    }
    if (fg('platform-linking-flexible-card-openness')) {
      if (removeBlockRestriction) {
        return child;
      }
    }
  });
};
const updateChildrenBlock = (node, size, status) => {
  const updatedChildren = React.Children.map(node.props.children, child => {
    if ( /*#__PURE__*/React.isValidElement(child) && isFlexibleUiBlock(child)) {
      // @ts-expect-error
      return /*#__PURE__*/React.cloneElement(child, {
        size,
        status
      });
    }
    return child;
  });
  return /*#__PURE__*/React.cloneElement(node, {
    children: updatedChildren
  });
};
const updateChildrenTitleBlock = (node, {
  containerSize,
  ...props
}) => {
  const updatedChildren = React.Children.map(node.props.children, child => {
    if ( /*#__PURE__*/React.isValidElement(child) && isFlexibleUiTitleBlock(child)) {
      const {
        size: blockSize
      } = child.props;
      const size = blockSize || containerSize;
      return /*#__PURE__*/React.cloneElement(child, {
        ...props,
        size
      });
    } else {
      return child;
    }
  });
  return /*#__PURE__*/React.cloneElement(node, {
    children: updatedChildren
  });
};

/**
 * Note: This function is only necessary for CompiledCSS within Jest tests due to the way it handles Styles.
 * CompiledCSS will inject a StyleCacheProvider around the component tree, which
 * causes the children to be wrapped in a StyleCacheProvider as well. This function recursively
 * searches for the first valid TitleBlock within the children of the StyleCacheProvider.
 */
export const getFlexibleUiBlock = node => {
  if (! /*#__PURE__*/React.isValidElement(node)) {
    return undefined;
  }
  if (node.type === TitleBlock) {
    return node;
  }
  if (isStyleCacheProvider(node)) {
    // Component wrapped with compiled at runtime, check for children
    let isChildrenValid;
    React.Children.map(node.props.children, child => {
      var _child$type;
      if (typeof child.type !== 'string' && ((_child$type = child.type) === null || _child$type === void 0 ? void 0 : _child$type.name) !== 'Style') {
        isChildrenValid = getFlexibleUiBlock(child);
      }
    });
    return isChildrenValid;
  }
  return undefined;
};
const getTitleBlockProps = children => {
  const block = React.Children.toArray(children).map(child => getFlexibleUiBlock(child)).filter(x => x !== undefined).at(0);
  if ( /*#__PURE__*/React.isValidElement(block)) {
    return block.props;
  }
};
const getLayeredLink = (testId, context, children, onClick) => {
  const {
    title,
    url = ''
  } = context || {};
  const {
    anchorTarget: target,
    text
  } = getTitleBlockProps(children) || {};
  return /*#__PURE__*/React.createElement(LayeredLink, {
    onClick: onClick,
    target: target,
    testId: testId,
    text: text || title,
    url: url
  });
};
const baseStyleCommon = null;

// TODO: Remove on fg cleanup: platform-linking-visual-refresh-v1
const backgroundStyleOld = null;
const backgroundStyles = null;

// TODO: Remove on fg cleanup: platform-linking-visual-refresh-v1
const elevationStylesOld = null;
const elevationStyles = null;
const clickableContainerStyles = null;

// TODO: Remove on fg cleanup: platform-linking-visual-refresh-v1
const gapStyleMapOld = {
  xlarge: "_zulp1jvo",
  large: "_zulph3jc",
  medium: "_zulp1qos",
  small: "_zulpfzsm"
};
const gapStyleMap = {
  xlarge: "_zulpr2fv",
  large: "_zulpzu5n",
  medium: "_zulp1bpf",
  small: "_zulpswc3"
};

/**
 * Get container padding based on smart link size
 * Equivalent version for DS primitives space token is getPrimitivesPaddingSpaceBySize()
 * at view/FlexibleCard/components/utils.tsx
 */
const getPadding = size => {
  if (fg('platform-linking-visual-refresh-v1')) {
    switch (size) {
      case SmartLinkSize.XLarge:
        return "var(--ds-space-300, 24px)";
      case SmartLinkSize.Large:
        return "var(--ds-space-250, 20px)";
      case SmartLinkSize.Medium:
        return "var(--ds-space-200, 16px)";
      case SmartLinkSize.Small:
      default:
        return "var(--ds-space-100, 8px)";
    }
  }
  switch (size) {
    case SmartLinkSize.XLarge:
      return '1.5rem';
    case SmartLinkSize.Large:
      return '1.25rem';
    case SmartLinkSize.Medium:
      return '1rem';
    case SmartLinkSize.Small:
    default:
      return '.5rem';
  }
};
const getGap = size => {
  if (fg('platform-linking-visual-refresh-v1')) {
    switch (size) {
      case SmartLinkSize.XLarge:
        return "var(--ds-space-250, 20px)";
      case SmartLinkSize.Large:
        return "var(--ds-space-200, 16px)";
      case SmartLinkSize.Medium:
        return "var(--ds-space-100, 8px)";
      case SmartLinkSize.Small:
      default:
        return "var(--ds-space-050, 4px)";
    }
  }
  switch (size) {
    case SmartLinkSize.XLarge:
      return '1.25rem';
    case SmartLinkSize.Large:
      return '1rem';
    case SmartLinkSize.Medium:
      return '.5rem';
    case SmartLinkSize.Small:
    default:
      return '.25rem';
  }
};

// TODO: Remove on fg cleanup: platform-linking-visual-refresh-v1
const previewOnLeftStyleMapOld = {
  xlarge: "_19bv15sf _y5vk1k92",
  large: "_19bvf557 _y5vk1j6v",
  medium: "_19bvqox2 _y5vk1kwk",
  small: "_19bv1ltm _y5vk6mu8"
};
const previewOnLeftStyleMap = {
  xlarge: "_19bver0k _y5vk1ejb",
  large: "_19bv16ru _y5vkv47k",
  medium: "_19bv1g7w _y5vk1eak",
  small: "_19bvlbes _y5vku2gc"
};

// TODO: Remove on fg cleanup: platform-linking-visual-refresh-v1
const previewOnRightStyleMapOld = {
  xlarge: "_u5f315sf _1yob1k92",
  large: "_u5f3f557 _1yob1j6v",
  medium: "_u5f3qox2 _1yob1kwk",
  small: "_u5f31ltm _1yob6mu8"
};
const previewOnRightStyleMap = {
  xlarge: "_u5f3er0k _1yob1ejb",
  large: "_u5f316ru _1yobv47k",
  medium: "_u5f31g7w _1yob1eak",
  small: "_u5f3lbes _1yobu2gc"
};

/**
 * A container is a hidden component that build the Flexible Smart Link.
 * All the Flexible UI components are wrapped inside the container.
 * It inherits the ui props from Card component and applies the custom styling
 * accordingly.
 * @internal
 * @see Block
 */
const Container = ({
  children,
  clickableContainer = false,
  hideBackground = false,
  hideElevation = false,
  hidePadding = false,
  onClick,
  retry,
  showHoverPreview = false,
  hoverPreviewOptions,
  actionOptions,
  removeBlockRestriction = false,
  size = SmartLinkSize.Medium,
  status,
  testId = 'smart-links-container',
  theme = SmartLinkTheme.Link
}) => {
  const padding = hidePadding ? fg('platform-linking-visual-refresh-v1') ? '0px' : '0rem' : getPadding(size);
  const gap = getGap(size);
  const context = useContext(FlexibleUiContext);
  const {
    previewOnLeft,
    previewOnRight
  } = getChildrenOptions(children, context);
  const canShowHoverPreview = showHoverPreview && status === 'resolved';
  // `retry` object contains action that can be performed on
  // unresolved link (unauthorized, forbidden, not found, etc.)
  const canShowAuthTooltip = showHoverPreview && status === 'unauthorized' && retry !== undefined;
  const containerPaddingStyles = null;
  const container = fg('platform-linking-visual-refresh-v1') ? /*#__PURE__*/React.createElement("div", {
    "data-smart-link-container": true,
    "data-testid": testId,
    className: ax(["_1e0c1txw _2lx21bp4 _1ul9idpf _1reo15vq _kqswh2mm _1pcmkb7n _1v7y1i6y _ufn21i6y", gapStyleMap[size], !hideBackground && "_bfhkhp5a", "_1yt418y6 _jb7v18y6 _y5vk167t _1yob1wow _r37x1r5k", previewOnLeft && previewOnLeftStyleMap[size], previewOnRight && previewOnRightStyleMap[size], !hideElevation && "_19itzgxb _2rkob7nn _19pkv77o _2hwxv77o _otyrv77o _18u0v77o", clickableContainer && "_nqwih2mm _18ywh2mm _1nq3h2mm _1t4ckb7n _1nxdkb7n _xnbykb7n"]),
    style: {
      "--_1xumd0e": ix(padding),
      "--_73mooq": ix(previewOnLeft ? gap : padding),
      "--_12k13bg": ix(previewOnRight ? gap : padding)
    }
  }, clickableContainer ? getLayeredLink(testId, context, children, onClick) : null, renderChildren(children, size, theme, status, retry, onClick, removeBlockRestriction)) : /*#__PURE__*/React.createElement("div", {
    "data-smart-link-container": true,
    "data-testid": testId,
    className: ax(["_1e0c1txw _2lx21bp4 _1ul9idpf _1reo15vq _kqswh2mm _1pcmkb7n _1v7y1i6y _ufn21i6y", gapStyleMapOld[size], !hideBackground && "_bfhkhp5a", "_1yt418y6 _jb7v18y6 _y5vk167t _1yob1wow _r37x1r5k", previewOnLeft && previewOnLeftStyleMapOld[size], previewOnRight && previewOnRightStyleMapOld[size], !hideElevation && "_19ith6cr _2rkopd34 _19pkv77o _2hwxv77o _otyrv77o _18u0v77o", clickableContainer && "_nqwih2mm _18ywh2mm _1nq3h2mm _1t4ckb7n _1nxdkb7n _xnbykb7n"]),
    style: {
      "--_1xumd0e": ix(padding),
      "--_73mooq": ix(previewOnLeft ? gap : padding),
      "--_12k13bg": ix(previewOnRight ? gap : padding)
    }
  }, clickableContainer ? getLayeredLink(testId, context, children, onClick) : null, renderChildren(children, size, theme, status, retry, onClick, removeBlockRestriction));
  if (context !== null && context !== void 0 && context.url && (canShowHoverPreview || canShowAuthTooltip)) {
    return /*#__PURE__*/React.createElement(HoverCardControl, {
      isHoverPreview: canShowHoverPreview,
      isAuthTooltip: canShowAuthTooltip,
      actionOptions: actionOptions,
      testId: testId,
      url: context.url,
      hoverPreviewOptions: hoverPreviewOptions
    }, container);
  }
  return container;
};
export default Container;
