// @atlaskit/teams-public/dist/cjs/controllers/hooks/use-product-permission/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useProductPermissions = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = require("react");
var _main = require("../../product-permission/main");
var useProductPermissions = exports.useProductPermissions = function useProductPermissions(_ref) {
  var userId = _ref.userId,
    cloudId = _ref.cloudId,
    permissionIds = _ref.permissionIds;
  var _ref2 = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {
      enabled: true
    },
    enabled = _ref2.enabled;
  var _useProductPermission = (0, _main.useProductPermissionsStore)(),
    _useProductPermission2 = (0, _slicedToArray2.default)(_useProductPermission, 2),
    state = _useProductPermission2[0],
    getPermissions = _useProductPermission2[1].getPermissions;
  var fetchData = (0, _react.useCallback)( /*#__PURE__*/(0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee() {
    return _regenerator.default.wrap(function _callee$(_context) {
      while (1) switch (_context.prev = _context.next) {
        case 0:
          getPermissions({
            cloudId: cloudId,
            userId: userId,
            enabled: enabled,
            permissionIds: permissionIds || ['manage', 'write']
          });
        case 1:
        case "end":
          return _context.stop();
      }
    }, _callee);
  })), [cloudId, enabled, getPermissions, userId, permissionIds]);
  (0, _react.useEffect)(function () {
    fetchData();
  }, [fetchData]);
  return {
    loading: state.isLoading,
    data: state.permissions,
    error: state.error
  };
};
