// @atlaskit/link-picker/dist/cjs/ui/link-picker/search-results/link-search-list/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.testIds = exports.messages = exports.LinkSearchList = void 0;
require("./index.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _reactIntlNext = require("react-intl-next");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _spinner = _interopRequireDefault(require("@atlaskit/spinner"));
var _visuallyHidden = _interopRequireDefault(require("@atlaskit/visually-hidden"));
var _minHeightContainer = require("../../../../common/ui/min-height-container");
var _handleNavKeyDown = require("../../../../common/utils/handleNavKeyDown");
var _linkSearchNoResults = require("./link-search-no-results");
var _listItem = require("./list-item");
var _useTrackResultsShown = require("./use-track-results-shown");
var _excluded = ["onChange", "onSelect", "onKeyDown", "items", "activeIndex", "selectedIndex", "isLoading", "ariaControls", "ariaLabelledBy", "ariaReadOnly", "role", "id", "hasSearchTerm", "activePlugin", "adaptiveHeight", "className", "tabPanelId"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var styles = {
  emptyStateNoResultsWrapper: "_1tkepxbi"
};
var listContainerStyles = null;
var spinnerContainerStyles = null;
var listStyles = null;
var baseListTitleStyles = null;
var oldListTitleStyles = null;
var newListTitleStyles = null;
var messages = exports.messages = (0, _reactIntlNext.defineMessages)({
  titleRecentlyViewed: {
    id: 'fabric.linkPicker.listTitle.recentlyViewed',
    defaultMessage: 'Recently Viewed',
    description: 'Describes type of items shown in the list for screen-reader users'
  },
  titleResults: {
    id: 'fabric.linkPicker.listTitle.results',
    defaultMessage: 'Results',
    description: 'Describes type of items shown in the list for screen-reader users'
  },
  searchLinkResults: {
    id: 'fabric.linkPicker.hyperlink.searchLinkResults',
    defaultMessage: '{count, plural, =0 {no results} one {# result} other {# results}} found',
    description: 'Announce search results for screen-reader users.'
  }
});
var testIds = exports.testIds = _objectSpread(_objectSpread(_objectSpread({}, _linkSearchNoResults.testIds), _listItem.testIds), {}, {
  resultListTitle: 'link-picker-list-title',
  searchResultList: 'link-search-list',
  searchResultLoadingIndicator: 'link-picker.results-loading-indicator',
  tabsLoadingIndicator: 'link-picker.tabs-loading-indicator'
});
var LinkSearchList = exports.LinkSearchList = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var _activePlugin$uiOptio;
  var onChange = _ref.onChange,
    onSelect = _ref.onSelect,
    onKeyDown = _ref.onKeyDown,
    items = _ref.items,
    activeIndex = _ref.activeIndex,
    selectedIndex = _ref.selectedIndex,
    isLoading = _ref.isLoading,
    ariaControls = _ref.ariaControls,
    ariaLabelledBy = _ref.ariaLabelledBy,
    ariaReadOnly = _ref.ariaReadOnly,
    role = _ref.role,
    id = _ref.id,
    hasSearchTerm = _ref.hasSearchTerm,
    activePlugin = _ref.activePlugin,
    adaptiveHeight = _ref.adaptiveHeight,
    className = _ref.className,
    tabPanelId = _ref.tabPanelId,
    restProps = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  var itemsContent;
  var loadingContent;
  var linkListTitle = hasSearchTerm ? messages.titleResults : messages.titleRecentlyViewed;
  (0, _useTrackResultsShown.useTrackResultsShown)(isLoading, items, hasSearchTerm);
  var itemRefs = (0, _react.useRef)({});
  var itemRefCallback = (0, _react.useCallback)(function (el, id) {
    if (el === null) {
      delete itemRefs.current[id];
    } else {
      itemRefs.current[id] = el;
    }
  }, []);
  var getTabIndex = (0, _react.useCallback)(function (index) {
    if (selectedIndex > -1) {
      return selectedIndex === index ? 0 : -1;
    }
    if (index === 0) {
      return 0;
    }
    return -1;
  }, [selectedIndex]);
  var handleKeyDown = (0, _react.useCallback)(function (event) {
    var updatedIndex = activeIndex;
    if (onKeyDown) {
      onKeyDown(event);
    }
    if (!(items !== null && items !== void 0 && items.length)) {
      return;
    }
    updatedIndex = (0, _handleNavKeyDown.handleNavKeyDown)(event, items.length, activeIndex);
    var item = items === null || items === void 0 ? void 0 : items[updatedIndex];
    if (['ArrowDown', 'ArrowUp', 'Home', 'End'].includes(event.key) && item) {
      if (itemRefs.current) {
        var _itemRefs$current$ite;
        (_itemRefs$current$ite = itemRefs.current[item.objectId]) === null || _itemRefs$current$ite === void 0 || _itemRefs$current$ite.focus();
      }
    }
  }, [activeIndex, items, onKeyDown]);
  if ((items === null || items === void 0 ? void 0 : items.length) === 0) {
    if (!hasSearchTerm) {
      var _activePlugin$emptySt;
      var emptyState = activePlugin === null || activePlugin === void 0 || (_activePlugin$emptySt = activePlugin.emptyStateNoResults) === null || _activePlugin$emptySt === void 0 ? void 0 : _activePlugin$emptySt.call(activePlugin);
      if (emptyState) {
        return /*#__PURE__*/React.createElement("div", {
          id: tabPanelId,
          className: (0, _runtime.ax)([styles.emptyStateNoResultsWrapper])
        }, emptyState);
      }
    }
    return /*#__PURE__*/React.createElement("div", {
      id: tabPanelId
    }, /*#__PURE__*/React.createElement(_linkSearchNoResults.NoResults, null));
  }
  var listItemNameMaxLines = activePlugin === null || activePlugin === void 0 || (_activePlugin$uiOptio = activePlugin.uiOptions) === null || _activePlugin$uiOptio === void 0 ? void 0 : _activePlugin$uiOptio.listItemNameMaxLines;
  if (items && items.length > 0) {
    itemsContent = /*#__PURE__*/React.createElement(_react.Fragment, null, /*#__PURE__*/React.createElement("div", {
      id: testIds.resultListTitle,
      "data-testid": testIds.resultListTitle,
      className: (0, _runtime.ax)(["_11c81o8v _k48pmoej _otyr1b66", (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v1') ? "_syaz1gjq" : "_syaz131l"])
    }, /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, linkListTitle)), /*#__PURE__*/React.createElement(_visuallyHidden.default, {
      id: "fabric.smartcard.linkpicker.suggested.results"
    }, hasSearchTerm && /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, (0, _extends2.default)({}, messages.searchLinkResults, {
      values: {
        count: items.length
      },
      "aria-live": "polite",
      "aria-atomic": "true"
    }))), /*#__PURE__*/React.createElement("ul", {
      id: id,
      role: role,
      "aria-controls": "fabric.smartcard.linkpicker.suggested.results",
      "aria-labelledby": testIds.resultListTitle,
      "aria-readonly": ariaReadOnly,
      "data-testid": testIds.searchResultList,
      className: (0, _runtime.ax)(["_qtt8glyw _ca0qze3t _u5f3ze3t _n3tdze3t _19bvze3t _19pkze3t _otyrze3t _18u01mjz _2hwxum63"])
    }, items.map(function (item, index) {
      return /*#__PURE__*/React.createElement(_listItem.LinkSearchListItem, {
        id: "".concat(testIds.searchResultItem, "-").concat(index),
        role: role && 'option',
        item: item,
        selected: selectedIndex === index,
        active: activeIndex === index,
        onFocus: function onFocus() {
          return onChange(item.objectId);
        },
        onKeyDown: handleKeyDown,
        onSelect: onSelect,
        key: item.objectId,
        tabIndex: getTabIndex(index),
        ref: function ref(el) {
          return itemRefCallback(el, item.objectId);
        },
        nameMaxLines: listItemNameMaxLines
      });
    })));
  }
  if (isLoading) {
    loadingContent = /*#__PURE__*/React.createElement(_minHeightContainer.MinHeightContainer, {
      minHeight: adaptiveHeight ? '80px' : '50px',
      className: (0, _runtime.ax)(["_16jlkb7n _2lx21bp4 _4cvr1h6o"])
    }, /*#__PURE__*/React.createElement(_spinner.default, {
      testId: testIds.searchResultLoadingIndicator,
      interactionName: "link-picker-search-list-loading",
      size: "medium"
    }));
  }
  return /*#__PURE__*/React.createElement("div", (0, _extends2.default)({
    id: tabPanelId,
    ref: ref,
    className: (0, _runtime.ax)(["_1bsb1osq _ca0qidpf _1tkeys9h _19pkpxbi _otyrpxbi _16jlkb7n _1e0c1txw _2lx21bp4", className])
  }, restProps), itemsContent, loadingContent);
});
