// @atlaskit/editor-common/dist/cjs/ui-color/ColorPalette/Palettes/highlightColorPalette.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.highlightColorPalette = exports.REMOVE_HIGHLIGHT_COLOR = exports.EditorDiagonalLineIcon = void 0;
var _react = require("@emotion/react");
var _adfSchema = require("@atlaskit/adf-schema");
var _icon = _interopRequireDefault(require("@atlaskit/icon"));
var _tokens = require("@atlaskit/tokens");
var _getColorMessage = _interopRequireDefault(require("./getColorMessage"));
var _paletteMessages = _interopRequireDefault(require("./paletteMessages"));
var _textColorPalette = require("./textColorPalette");
/**
 * @jsxRuntime classic
 * @jsx jsx
 */

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var REMOVE_HIGHLIGHT_COLOR = exports.REMOVE_HIGHLIGHT_COLOR = '#00000000';
var DiagonalLineGlyph = function DiagonalLineGlyph() {
  return (0, _react.jsx)("svg", {
    width: "24",
    height: "24",
    viewBox: "0 0 24 24",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, (0, _react.jsx)("rect", {
    width: "24",
    height: "24",
    fill: "white",
    fillOpacity: "0.01"
  }), (0, _react.jsx)("path", {
    fillRule: "evenodd",
    clipRule: "evenodd",
    d: "M17.2038 7.70388C17.5955 8.09565 17.5941 8.73128 17.2005 9.12127L9.1213 17.1277C8.73027 17.5152 8.09958 17.5138 7.71031 17.1245C7.31854 16.7328 7.31997 16.0971 7.71351 15.7071L15.7928 7.70069C16.1838 7.31318 16.8145 7.31461 17.2038 7.70388Z",
    fill: "currentColor",
    fillOpacity: "0.5"
  }));
};
var EditorDiagonalLineIcon = exports.EditorDiagonalLineIcon = function EditorDiagonalLineIcon() {
  var _useThemeObserver = (0, _tokens.useThemeObserver)(),
    colorMode = _useThemeObserver.colorMode;
  var primaryColor = colorMode === 'dark' ? "var(--ds-background-accent-gray-bolder, #626F86)" : "var(--ds-background-accent-gray-subtle, #8590A2)";
  return (0, _react.jsx)(_icon.default, {
    glyph: DiagonalLineGlyph,
    label: "",
    primaryColor: primaryColor
  });
};
var highlightColorPalette = exports.highlightColorPalette = [{
  value: REMOVE_HIGHLIGHT_COLOR,
  label: 'No color',
  // Mostly informative, only used for analytics
  border: "var(--ds-border, #091E4224)",
  message: (0, _getColorMessage.default)(_paletteMessages.default, 'no-color'),
  decorator: (0, _react.jsx)(EditorDiagonalLineIcon, null)
}];
_adfSchema.backgroundColorPalette.forEach(function (label, color) {
  highlightColorPalette.push((0, _textColorPalette.mapPaletteColor)(label, color));
});
