// @atlaskit/editor-plugin-media/dist/cjs/nodeviews/toDOM-fixes/mediaInline.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mediaInlineSpecWithFixedToDOM = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _adfSchema = require("@atlaskit/adf-schema");
var _experiments = require("@atlaskit/tmp-editor-statsig/experiments");
var _toDOMAttrs = require("./toDOMAttrs");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var skeletonStyling = "background: ".concat("var(--ds-background-neutral, #091E420F)", ";");
// Matches media
var fallbackAspectRatio = 1.25;

// @nodeSpecException:toDOM patch
var mediaInlineSpecWithFixedToDOM = exports.mediaInlineSpecWithFixedToDOM = function mediaInlineSpecWithFixedToDOM() {
  if ((0, _experiments.editorExperiment)('platform_editor_exp_lazy_node_views', false)) {
    return _adfSchema.mediaInline;
  }
  return _objectSpread(_objectSpread({}, _adfSchema.mediaInline), {}, {
    toDOM: function toDOM(node) {
      var dataAttrs = (0, _toDOMAttrs.getMediaAttrs)('mediaInline', node);
      if (node.attrs.type === 'image') {
        var aspectRatio = node.attrs.width && node.attrs.height ? node.attrs.width / node.attrs.height : fallbackAspectRatio;
        return ['span', _objectSpread(_objectSpread({}, dataAttrs), {}, {
          style: 'display: inline-block; transform: translateY(6px)',
          class: 'media-inline-image-wrapper'
        }), ['span', {
          height: '100%',
          width: '100%',
          // Transparent image workaround to control styling
          src: 'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7',
          style: "display: inline-block; aspect-ratio: ".concat(aspectRatio, "; height: 100%; width: 100%; border-radius: ", "var(--ds-border-radius, 3px)", "; ").concat(skeletonStyling)
        }]];
      } else {
        return ['span', _objectSpread(_objectSpread({}, dataAttrs), {}, {
          style: "display: inline-block; transform: translateY(6px); ".concat(skeletonStyling),
          class: 'media-inline-image-wrapper'
        }), ['span', {
          style: "display: inline-block; height: 100%; width: 100%; aspect-ratio: ".concat(fallbackAspectRatio, ";")
        }]];
      }
    }
  });
};
