// @atlaskit/profilecard/dist/cjs/components/team-profile-card/team-connections/main.js
/* main.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TeamConnections = void 0;
require("./main.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _react = _interopRequireWildcard(require("react"));
var _analyticsNext = require("@atlaskit/analytics-next");
var _menu = require("@atlaskit/menu");
var _compiled = require("@atlaskit/primitives/compiled");
var _teamsPublic = require("@atlaskit/teams-public");
var _analytics = require("../../../util/analytics");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var styles = {
  containerWrapperStyles: "_1e0c1txw _4cvr1h6o",
  containerIconStyles: "_2rkoiti9 _4t3i1tcg _1bsb1tcg",
  containerTypeIconButtonStyles: "_18u01wug _4t3i7vkz _1bsb7vkz"
};
var TeamConnections = exports.TeamConnections = function TeamConnections(_ref) {
  var containerType = _ref.containerType,
    title = _ref.title,
    containerIcon = _ref.containerIcon,
    link = _ref.link;
  var _getContainerProperti = (0, _teamsPublic.getContainerProperties)(containerType, 'medium'),
    description = _getContainerProperti.description,
    icon = _getContainerProperti.icon,
    containerTypeText = _getContainerProperti.containerTypeText;
  var _useAnalyticsEvents = (0, _analyticsNext.useAnalyticsEvents)(),
    createAnalyticsEvent = _useAnalyticsEvents.createAnalyticsEvent;
  var onClick = (0, _react.useCallback)(function () {
    (0, _analytics.fireEvent)(createAnalyticsEvent, {
      action: 'clicked',
      actionSubject: 'teamConnectionItem',
      actionSubjectId: 'teamProfileCard',
      attributes: {
        container: containerType
      }
    });
  }, [containerType, createAnalyticsEvent]);
  return /*#__PURE__*/_react.default.createElement(_menu.LinkItem, {
    href: link,
    onClick: onClick
  }, /*#__PURE__*/_react.default.createElement(_compiled.Inline, {
    space: "space.100",
    xcss: styles.containerWrapperStyles
  }, /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    as: "img",
    src: containerIcon,
    alt: "",
    testId: "linked-container-icon",
    xcss: styles.containerIconStyles
  }), /*#__PURE__*/_react.default.createElement(_compiled.Stack, null, /*#__PURE__*/_react.default.createElement(_compiled.Text, {
    maxLines: 1,
    color: "color.text"
  }, title), /*#__PURE__*/_react.default.createElement(_compiled.Inline, {
    space: "space.050"
  }, /*#__PURE__*/_react.default.createElement(_compiled.Text, {
    size: "small",
    color: "color.text.subtlest"
  }, description), /*#__PURE__*/_react.default.createElement(_compiled.Text, {
    size: "small",
    color: "color.text.subtlest"
  }, containerTypeText))), /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    backgroundColor: 'color.background.neutral.subtle',
    xcss: styles.containerTypeIconButtonStyles,
    testId: "container-type-icon"
  }, icon)));
};
