// @atlaskit/smart-card/dist/esm/view/FlexibleCard/components/actions/automation-action/automation-manual-triggers/manual-triggers-modal/sub-components/error-state/main.js
/* main.tsx generated by @compiled/babel-plugin v0.36.1 */
import "./main.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { defineMessages, useIntl } from 'react-intl-next';
import { Box, Stack } from '@atlaskit/primitives/compiled';
import ErrorIcon from './error-icon';
var styles = {
  errorState: "_y3gn1h6o _19pk1jfw",
  imageContainer: "_1e0c1ule _p12f7iqz _c71lplmo _otyr1ejb _19pk1b66",
  description: "_p12f1gwv"
};
var i18n = defineMessages({
  errorDescription: {
    id: 'automation-menu.modal.error.description',
    defaultMessage: 'This is taking awhile to load. Try refreshing the page.',
    description: "Description for the error section shown when rules can't be fetched"
  },
  errorIconLabel: {
    id: 'automation-menu.modal.error.image.alt',
    defaultMessage: 'Error icon',
    description: 'Alternative text for the image that renders when there is an error.'
  }
});
export var AutomationModalErrorState = function AutomationModalErrorState() {
  var _useIntl = useIntl(),
    formatMessage = _useIntl.formatMessage;
  return /*#__PURE__*/React.createElement(Stack, {
    xcss: styles.errorState,
    alignInline: "center",
    alignBlock: "center"
  }, /*#__PURE__*/React.createElement(Box, {
    xcss: styles.imageContainer
  }, /*#__PURE__*/React.createElement(ErrorIcon, null)), /*#__PURE__*/React.createElement(Box, {
    xcss: styles.description
  }, formatMessage(i18n.errorDescription)));
};
