// @atlaskit/editor-plugin-annotation/dist/cjs/pm-plugins/inline-comment.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.inlineCommentPlugin = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _adfSchema = require("@atlaskit/adf-schema");
var _analytics = require("@atlaskit/editor-common/analytics");
var _safePlugin = require("@atlaskit/editor-common/safe-plugin");
var _view = require("@atlaskit/editor-prosemirror/view");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _editorCommands = require("../editor-commands");
var _nodeviews = require("../nodeviews");
var _annotationManagerHooks = require("./annotation-manager-hooks");
var _pluginFactory = require("./plugin-factory");
var _utils = require("./utils");
var fetchProviderStates = /*#__PURE__*/function () {
  var _ref = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(provider, annotationIds) {
    var data, result;
    return _regenerator.default.wrap(function _callee$(_context) {
      while (1) switch (_context.prev = _context.next) {
        case 0:
          if (!((!provider || !provider.getState) && (0, _platformFeatureFlags.fg)('use_comments_data_annotation_updater'))) {
            _context.next = 2;
            break;
          }
          return _context.abrupt("return", {});
        case 2:
          _context.next = 4;
          return provider.getState(annotationIds);
        case 4:
          data = _context.sent;
          result = {};
          data.forEach(function (annotation) {
            if (annotation.annotationType === _adfSchema.AnnotationTypes.INLINE_COMMENT) {
              result[annotation.id] = annotation.state.resolved;
            }
          });
          return _context.abrupt("return", result);
        case 8:
        case "end":
          return _context.stop();
      }
    }, _callee);
  }));
  return function fetchProviderStates(_x, _x2) {
    return _ref.apply(this, arguments);
  };
}();

// fetchState is unable to return a command as it's runs async and may dispatch at a later time
// Requires `editorView` instead of the decomposition as the async means state may end up stale
var fetchState = /*#__PURE__*/function () {
  var _ref2 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee2(provider, annotationIds, editorView, editorAnalyticsAPI) {
    var inlineCommentStates;
    return _regenerator.default.wrap(function _callee2$(_context2) {
      while (1) switch (_context2.prev = _context2.next) {
        case 0:
          if (!((!annotationIds || !annotationIds.length) && !(0, _platformFeatureFlags.fg)('use_comments_data_annotation_updater'))) {
            _context2.next = 2;
            break;
          }
          return _context2.abrupt("return");
        case 2:
          _context2.next = 4;
          return fetchProviderStates(provider, annotationIds);
        case 4:
          inlineCommentStates = _context2.sent;
          if (!(Object.keys(inlineCommentStates).length === 0 && (0, _platformFeatureFlags.fg)('use_comments_data_annotation_updater'))) {
            _context2.next = 7;
            break;
          }
          return _context2.abrupt("return");
        case 7:
          if (editorView.dispatch) {
            (0, _editorCommands.updateInlineCommentResolvedState)(editorAnalyticsAPI)(inlineCommentStates)(editorView.state, editorView.dispatch);
          }
        case 8:
        case "end":
          return _context2.stop();
      }
    }, _callee2);
  }));
  return function fetchState(_x3, _x4, _x5, _x6) {
    return _ref2.apply(this, arguments);
  };
}();
var initialState = function initialState() {
  var disallowOnWhitespace = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;
  var featureFlagsPluginState = arguments.length > 1 ? arguments[1] : undefined;
  return {
    annotations: {},
    selectedAnnotations: [],
    hoveredAnnotations: [],
    mouseData: {
      isSelecting: false
    },
    disallowOnWhitespace: disallowOnWhitespace,
    isInlineCommentViewClosed: false,
    isVisible: true,
    skipSelectionHandling: false,
    featureFlagsPluginState: featureFlagsPluginState,
    isDrafting: false
  };
};
var hideToolbar = function hideToolbar(state, dispatch) {
  return function () {
    (0, _editorCommands.updateMouseState)({
      isSelecting: true
    })(state, dispatch);
  };
};

// Subscribe to updates from consumer
var onResolve = function onResolve(editorAnalyticsAPI) {
  return function (state, dispatch) {
    return function (annotationId) {
      (0, _editorCommands.updateInlineCommentResolvedState)(editorAnalyticsAPI)((0, _defineProperty2.default)({}, annotationId, true), _analytics.RESOLVE_METHOD.CONSUMER)(state, dispatch);
    };
  };
};
var onUnResolve = function onUnResolve(editorAnalyticsAPI) {
  return function (state, dispatch) {
    return function (annotationId) {
      (0, _editorCommands.updateInlineCommentResolvedState)(editorAnalyticsAPI)((0, _defineProperty2.default)({}, annotationId, false))(state, dispatch);
    };
  };
};
var onMouseUp = function onMouseUp(state, dispatch) {
  return function (e) {
    var _ref3 = (0, _utils.getPluginState)(state) || {},
      mouseData = _ref3.mouseData;
    if (mouseData !== null && mouseData !== void 0 && mouseData.isSelecting) {
      (0, _editorCommands.updateMouseState)({
        isSelecting: false
      })(state, dispatch);
    }
  };
};
var onSetVisibility = function onSetVisibility(view) {
  return function (isVisible) {
    var state = view.state,
      dispatch = view.dispatch;
    (0, _editorCommands.setInlineCommentsVisibility)(isVisible)(state, dispatch);
    if (isVisible) {
      // PM retains focus when we click away from the editor.
      // This will restore the visual aspect of the selection,
      // otherwise it will seem a floating toolbar will appear
      // for no reason.
      view.focus();
    }
  };
};
var inlineCommentPlugin = exports.inlineCommentPlugin = function inlineCommentPlugin(options) {
  var provider = options.provider,
    featureFlagsPluginState = options.featureFlagsPluginState,
    annotationManager = options.annotationManager;
  return new _safePlugin.SafePlugin({
    key: _utils.inlineCommentPluginKey,
    state: (0, _pluginFactory.createPluginState)(options.dispatch, initialState(provider.disallowOnWhitespace, featureFlagsPluginState)),
    view: function view(editorView) {
      if (annotationManager && (0, _platformFeatureFlags.fg)('platform_editor_comments_api_manager')) {
        annotationManager.hook('allowAnnotation', (0, _annotationManagerHooks.allowAnnotation)(editorView, options));
        annotationManager.hook('startDraft', (0, _annotationManagerHooks.startDraft)(editorView, options));
        annotationManager.hook('clearDraft', (0, _annotationManagerHooks.clearDraft)(editorView, options));
        annotationManager.hook('applyDraft', (0, _annotationManagerHooks.applyDraft)(editorView, options));
        annotationManager.hook('getDraft', (0, _annotationManagerHooks.getDraft)(editorView, options));
      }
      // Get initial state
      // Need to pass `editorView` to mitigate editor state going stale
      fetchState(provider, (0, _utils.getAllAnnotations)(editorView.state.doc), editorView, options.editorAnalyticsAPI);
      var resolve = function resolve(annotationId) {
        return onResolve(options.editorAnalyticsAPI)(editorView.state, editorView.dispatch)(annotationId);
      };
      var unResolve = function unResolve(annotationId) {
        return onUnResolve(options.editorAnalyticsAPI)(editorView.state, editorView.dispatch)(annotationId);
      };
      var mouseUp = function mouseUp(event) {
        return onMouseUp(editorView.state, editorView.dispatch)(event);
      };
      var setVisibility = function setVisibility(isVisible) {
        return onSetVisibility(editorView)(isVisible);
      };
      var setSelectedAnnotationFn = function setSelectedAnnotationFn(annotationId) {
        var pluginState = (0, _utils.getPluginState)(editorView.state);
        if ((0, _platformFeatureFlags.fg)('platform_editor_listen_for_focussed_query_param')) {
          // When feature flag is true, only close if no annotationId
          if (!annotationId) {
            (0, _editorCommands.closeComponent)()(editorView.state, editorView.dispatch);
          } else {
            (0, _editorCommands.setSelectedAnnotation)(annotationId)(editorView.state, editorView.dispatch);
          }
        } else {
          // When feature flag is false, close if:
          // 1. No annotationId OR
          // 2. View is closed and annotation clicks are enabled
          var shouldClose = !annotationId || (pluginState === null || pluginState === void 0 ? void 0 : pluginState.isInlineCommentViewClosed) && (0, _platformFeatureFlags.fg)('platform_editor_listen_for_annotation_clicks');
          if (shouldClose) {
            (0, _editorCommands.closeComponent)()(editorView.state, editorView.dispatch);
          } else {
            (0, _editorCommands.setSelectedAnnotation)(annotationId)(editorView.state, editorView.dispatch);
          }
        }
      };
      var setHoveredAnnotationFn = function setHoveredAnnotationFn(annotationId) {
        if (!annotationId) {
          (0, _editorCommands.closeComponent)()(editorView.state, editorView.dispatch);
        } else {
          (0, _editorCommands.setHoveredAnnotation)(annotationId)(editorView.state, editorView.dispatch);
        }
      };
      var removeHoveredannotationFn = function removeHoveredannotationFn() {
        (0, _editorCommands.setHoveredAnnotation)('')(editorView.state, editorView.dispatch);
      };
      var closeInlineCommentFn = function closeInlineCommentFn() {
        (0, _editorCommands.closeComponent)()(editorView.state, editorView.dispatch);
      };
      var updateSubscriber = provider.updateSubscriber;
      if (updateSubscriber) {
        updateSubscriber.on('resolve', resolve).on('delete', resolve).on('unresolve', unResolve).on('create', unResolve).on('setvisibility', setVisibility).on('setselectedannotation', setSelectedAnnotationFn).on('sethoveredannotation', setHoveredAnnotationFn).on('removehoveredannotation', removeHoveredannotationFn).on('closeinlinecomment', closeInlineCommentFn);
      }

      // eslint-disable-next-line @repo/internal/dom-events/no-unsafe-event-listeners
      editorView.root.addEventListener('mouseup', mouseUp);
      return {
        update: function update(view, _prevState) {
          if ((0, _platformFeatureFlags.fg)('confluence_comments_select_comment_experience')) {
            var _prevSelectedAnnotati;
            var _ref4 = (0, _utils.getPluginState)(view.state) || {},
              selectedAnnotations = _ref4.selectedAnnotations;
            var _ref5 = (0, _utils.getPluginState)(_prevState) || {},
              prevSelectedAnnotations = _ref5.selectedAnnotations;

            // If the new state has a selected annotation, and it's different from the previous one
            // then we mark the select annotation experience as complete
            if (selectedAnnotations && selectedAnnotations.length !== 0 && selectedAnnotations[0].id !== (prevSelectedAnnotations === null || prevSelectedAnnotations === void 0 || (_prevSelectedAnnotati = prevSelectedAnnotations[0]) === null || _prevSelectedAnnotati === void 0 ? void 0 : _prevSelectedAnnotati.id)) {
              var _options$selectCommen;
              (_options$selectCommen = options.selectCommentExperience) === null || _options$selectCommen === void 0 || _options$selectCommen.selectAnnotation.complete(selectedAnnotations[0].id);
            }
          }
          var _ref6 = (0, _utils.getPluginState)(view.state) || {},
            dirtyAnnotations = _ref6.dirtyAnnotations;
          if (!dirtyAnnotations) {
            return;
          }
          (0, _editorCommands.clearDirtyMark)()(view.state, view.dispatch);
          fetchState(provider, (0, _utils.getAllAnnotations)(view.state.doc), view, options.editorAnalyticsAPI);
        },
        destroy: function destroy() {
          // eslint-disable-next-line @repo/internal/dom-events/no-unsafe-event-listeners
          editorView.root.removeEventListener('mouseup', mouseUp);
          if (updateSubscriber) {
            updateSubscriber.off('resolve', resolve).off('delete', resolve).off('unresolve', unResolve).off('create', unResolve).off('setvisibility', setVisibility).off('setselectedannotation', setSelectedAnnotationFn).off('sethoveredannotation', setHoveredAnnotationFn).off('removehoveredannotation', removeHoveredannotationFn).off('closeinlinecomment', closeInlineCommentFn);
          }
        }
      };
    },
    props: {
      handleDOMEvents: {
        mousedown: function mousedown(view) {
          var pluginState = (0, _utils.getPluginState)(view.state);
          if (!(pluginState !== null && pluginState !== void 0 && pluginState.mouseData.isSelecting)) {
            hideToolbar(view.state, view.dispatch)();
          }
          return false;
        },
        dragstart: function dragstart(view, event) {
          // Mouseup won't be triggered after dropping
          // Hence, update the mouse data to cancel selecting when drag starts
          return onMouseUp(view.state, view.dispatch)(event);
        },
        click: function click(view, event) {
          var _event$target$closest, _pluginState$selected;
          if (!(0, _platformFeatureFlags.fg)('platform_editor_listen_for_annotation_clicks')) {
            return false;
          }
          if (!(event.target instanceof HTMLElement)) {
            return false;
          }

          // Find the nearest ancestor (or self) with the data-id attribute
          var annotationId = (_event$target$closest = event.target.closest('[data-id]')) === null || _event$target$closest === void 0 ? void 0 : _event$target$closest.getAttribute('data-id');
          if (!annotationId) {
            return false;
          }
          var pluginState = (0, _utils.getPluginState)(view.state);
          var isSelected = pluginState === null || pluginState === void 0 || (_pluginState$selected = pluginState.selectedAnnotations) === null || _pluginState$selected === void 0 ? void 0 : _pluginState$selected.some(function (selectedAnnotation) {
            return selectedAnnotation.id === annotationId;
          });
          // If the annotation is selected and the inline comment view is open, do nothing
          // as the user is already in the comment view.
          if (isSelected && !(pluginState !== null && pluginState !== void 0 && pluginState.isInlineCommentViewClosed)) {
            return false;
          }
          var _ref7 = pluginState || {},
            annotations = _ref7.annotations;
          var isUnresolved = annotations && annotations[annotationId] === false;
          if (!isUnresolved) {
            return false;
          }
          (0, _editorCommands.setSelectedAnnotation)(annotationId)(view.state, view.dispatch);
          return true;
        }
      },
      decorations: function decorations(state) {
        // highlight comments, depending on state
        var _ref8 = (0, _utils.getPluginState)(state) || {},
          draftDecorationSet = _ref8.draftDecorationSet,
          annotations = _ref8.annotations,
          selectedAnnotations = _ref8.selectedAnnotations,
          isVisible = _ref8.isVisible,
          isInlineCommentViewClosed = _ref8.isInlineCommentViewClosed,
          hoveredAnnotations = _ref8.hoveredAnnotations;
        var decorations = draftDecorationSet !== null && draftDecorationSet !== void 0 ? draftDecorationSet : _view.DecorationSet.empty;
        var focusDecorations = [];
        state.doc.descendants(function (node, pos) {
          var _provider$supportedBl;
          // Inline comment on mediaInline is not supported as part of comments on media project
          // Thus, we skip the decoration for mediaInline node
          if (node.type.name === 'mediaInline') {
            return false;
          }
          var isSupportedBlockNode = node.isBlock && ((_provider$supportedBl = provider.supportedBlockNodes) === null || _provider$supportedBl === void 0 ? void 0 : _provider$supportedBl.includes(node.type.name));
          node.marks.filter(function (mark) {
            return mark.type === state.schema.marks.annotation;
          }).forEach(function (mark) {
            if (isVisible) {
              var isUnresolved = !!annotations && annotations[mark.attrs.id] === false;
              var isSelected = !isInlineCommentViewClosed && !!(selectedAnnotations !== null && selectedAnnotations !== void 0 && selectedAnnotations.some(function (selectedAnnotation) {
                return selectedAnnotation.id === mark.attrs.id;
              }));
              var isHovered = !isInlineCommentViewClosed && !!(hoveredAnnotations !== null && hoveredAnnotations !== void 0 && hoveredAnnotations.some(function (hoveredAnnotation) {
                return hoveredAnnotation.id === mark.attrs.id;
              }));
              if (isSupportedBlockNode) {
                focusDecorations.push(_view.Decoration.node(pos, pos + node.nodeSize, {
                  class: "".concat((0, _nodeviews.getBlockAnnotationViewClassname)(isUnresolved, isSelected), " ").concat(isUnresolved)
                }, {
                  key: _utils.decorationKey.block
                }));
              } else {
                if ((0, _platformFeatureFlags.fg)('editor_inline_comments_on_inline_nodes')) {
                  if (node.isText) {
                    focusDecorations.push(_view.Decoration.inline(pos, pos + node.nodeSize, {
                      class: "".concat((0, _nodeviews.getAnnotationViewClassname)(isUnresolved, isSelected, isHovered), " ").concat(isUnresolved),
                      nodeName: 'span'
                    }));
                  } else {
                    focusDecorations.push(_view.Decoration.node(pos, pos + node.nodeSize, {
                      class: "".concat((0, _nodeviews.getAnnotationViewClassname)(isUnresolved, isSelected, isHovered), " ").concat(isUnresolved)
                    }, {
                      key: _utils.decorationKey.block
                    }));
                  }
                } else {
                  focusDecorations.push(_view.Decoration.inline(pos, pos + node.nodeSize, {
                    class: "".concat((0, _nodeviews.getAnnotationViewClassname)(isUnresolved, isSelected, isHovered), " ").concat(isUnresolved),
                    nodeName: 'span'
                  }));
                }
              }
            }
          });
        });
        decorations = decorations.add(state.doc, focusDecorations);
        return decorations;
      }
    }
  });
};
