// @atlaskit/editor-plugin-annotation/dist/cjs/pm-plugins/plugin-factory.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.shouldClearBookMarkCheck = exports.createPluginState = exports.createCommand = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _utils = require("@atlaskit/editor-common/utils");
var _state = require("@atlaskit/editor-prosemirror/state");
var _view = require("@atlaskit/editor-prosemirror/view");
var _reducer = _interopRequireDefault(require("./reducer"));
var _utils2 = require("./utils");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var handleDocChanged = function handleDocChanged(tr, prevPluginState) {
  if (!tr.getMeta('replaceDocument')) {
    return getSelectionChangedHandler(false)(tr, prevPluginState);
  }
  return _objectSpread(_objectSpread({}, prevPluginState), {}, {
    dirtyAnnotations: true
  });
};

/**
 * We clear bookmark on the following conditions:
 * 1. if current selection is an empty selection, or
 * 2. if the current selection and bookmark selection are different
 */
var shouldClearBookMarkCheck = exports.shouldClearBookMarkCheck = function shouldClearBookMarkCheck(tr, editorState, bookmark) {
  if (editorState.selection.empty || !bookmark) {
    return true;
  } else if (editorState.selection instanceof _state.NodeSelection) {
    var bookmarkSelection = bookmark === null || bookmark === void 0 ? void 0 : bookmark.resolve(tr.doc);
    if (bookmarkSelection instanceof _state.NodeSelection) {
      var selectionNode = editorState.selection.node;
      var bookmarkNode = bookmarkSelection.node;

      /**
       * Currently, after updating the alt text of a mediaSingle node,
       * the selection moves to the media node.
       * (then will append a transaction to its parent node)
       */
      if (selectionNode.type.name === 'media' && bookmarkNode.type.name === 'mediaSingle') {
        var _bookmarkNode$firstCh;
        return !((_bookmarkNode$firstCh = bookmarkNode.firstChild) !== null && _bookmarkNode$firstCh !== void 0 && _bookmarkNode$firstCh.eq(selectionNode));
      } else {
        return !bookmarkNode.eq(selectionNode);
      }
    }
  }

  // by default we discard bookmark
  return true;
};
var getSelectionChangedHandler = function getSelectionChangedHandler(reopenCommentView) {
  return function (tr, pluginState) {
    if (pluginState.skipSelectionHandling) {
      return _objectSpread(_objectSpread({}, pluginState), {}, {
        skipSelectionHandling: false
      }, reopenCommentView && {
        isInlineCommentViewClosed: false
      });
    }
    if (
    // If pluginState.selectedAnnotations is annotations of block node, i.e. when a new comment is created,
    // we keep it as it is so that we can show comment view component with the newly created comment
    (0, _utils2.isBlockNodeAnnotationsSelected)(tr.selection, pluginState.selectedAnnotations)) {
      return _objectSpread(_objectSpread({}, pluginState), reopenCommentView && {
        isInlineCommentViewClosed: false
      });
    }
    var selectedAnnotations = (0, _utils2.findAnnotationsInSelection)(tr.selection, tr.doc);
    if (selectedAnnotations.length === 0) {
      return _objectSpread(_objectSpread({}, pluginState), {}, {
        selectedAnnotations: selectedAnnotations,
        isInlineCommentViewClosed: true,
        selectAnnotationMethod: undefined
      });
    }
    if ((0, _utils2.isSelectedAnnotationsChanged)(selectedAnnotations, pluginState.selectedAnnotations)) {
      return _objectSpread(_objectSpread({}, pluginState), {}, {
        selectedAnnotations: selectedAnnotations,
        selectAnnotationMethod: undefined
      }, reopenCommentView && {
        isInlineCommentViewClosed: false
      });
    }
    return _objectSpread(_objectSpread(_objectSpread({}, pluginState), reopenCommentView && {
      isInlineCommentViewClosed: false
    }), {}, {
      selectAnnotationMethod: undefined
    });
  };
};
var _pluginFactory = (0, _utils.pluginFactory)(_utils2.inlineCommentPluginKey, _reducer.default, {
    onSelectionChanged: getSelectionChangedHandler(true),
    onDocChanged: handleDocChanged,
    mapping: function mapping(tr, pluginState, editorState) {
      var draftDecorationSet = pluginState.draftDecorationSet,
        bookmark = pluginState.bookmark;
      var mappedDecorationSet = _view.DecorationSet.empty,
        mappedBookmark;
      var hasMappedDecorations = false;
      if (draftDecorationSet) {
        mappedDecorationSet = draftDecorationSet.map(tr.mapping, tr.doc);
      }
      hasMappedDecorations = mappedDecorationSet.find(undefined, undefined, function (spec) {
        return Object.values(_utils2.decorationKey).includes(spec.key);
      }).length > 0;

      // When changes to decoration target make decoration invalid (e.g. delete text, add mark to node),
      // we need to reset bookmark to hide create component and to avoid invalid draft being published
      // We only perform this change when document selection has changed.
      if (!hasMappedDecorations && shouldClearBookMarkCheck(tr, editorState, bookmark)) {
        return _objectSpread(_objectSpread({}, pluginState), {}, {
          draftDecorationSet: mappedDecorationSet,
          bookmark: undefined
        });
      }
      if (bookmark) {
        mappedBookmark = bookmark.map(tr.mapping);
      }

      // return same pluginState if mappings did not change
      if (mappedBookmark === bookmark && mappedDecorationSet === draftDecorationSet) {
        return pluginState;
      }
      return _objectSpread(_objectSpread({}, pluginState), {}, {
        draftDecorationSet: mappedDecorationSet,
        bookmark: mappedBookmark
      });
    }
  }),
  createPluginState = exports.createPluginState = _pluginFactory.createPluginState,
  createCommand = exports.createCommand = _pluginFactory.createCommand;
