// @atlaskit/editor-core/dist/cjs/ui/WithHelpTrigger/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _analytics = require("@atlaskit/editor-common/analytics");
var _eventDispatcher = require("@atlaskit/editor-common/event-dispatcher");
var _analytics2 = require("@atlaskit/editor-common/utils/analytics");
var _helpDialog = require("@atlaskit/editor-plugins/help-dialog");
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
// Ignored via go/ees005
// eslint-disable-next-line @repo/internal/react/no-class-components
var WithHelpTrigger = exports.default = /*#__PURE__*/function (_React$Component) {
  function WithHelpTrigger() {
    var _this;
    (0, _classCallCheck2.default)(this, WithHelpTrigger);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, WithHelpTrigger, [].concat(args));
    (0, _defineProperty2.default)(_this, "openHelp", function () {
      // Ignored via go/ees005
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      var _ref = _this.context,
        editorActions = _ref.editorActions;

      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      var dispatch = (0, _eventDispatcher.createDispatch)(editorActions.eventDispatcher);
      dispatch(_analytics2.analyticsEventKey, {
        payload: {
          action: _analytics.ACTION.CLICKED,
          actionSubject: _analytics.ACTION_SUBJECT.BUTTON,
          actionSubjectId: _analytics.ACTION_SUBJECT_ID.BUTTON_HELP,
          attributes: {
            inputMethod: _analytics.INPUT_METHOD.TOOLBAR
          },
          eventType: _analytics.EVENT_TYPE.UI
        }
      });
      var editorView = editorActions._privateGetEditorView();
      if (editorView) {
        (0, _helpDialog.deprecatedOpenHelpCommand)(editorView.state.tr, editorView.dispatch);
      }
    });
    return _this;
  }
  (0, _inherits2.default)(WithHelpTrigger, _React$Component);
  return (0, _createClass2.default)(WithHelpTrigger, [{
    key: "render",
    value: function render() {
      return this.props.render(this.openHelp);
    }
  }]);
}(_react.default.Component);
(0, _defineProperty2.default)(WithHelpTrigger, "contextTypes", {
  editorActions: _propTypes.default.object.isRequired
});
