// @atlaskit/jql-editor/dist/cjs/ui/jql-editor-controls-content/expand-toggle/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ExpandToggle = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _tooltip = _interopRequireDefault(require("@atlaskit/tooltip"));
var _analytics = require("../../../analytics");
var _constants = require("../../../common/constants");
var _styled = require("../../../common/styled");
var _useEditorTheme = require("../../../hooks/use-editor-theme");
var _state = require("../../../state");
var _tooltipTag = require("../../tooltip-tag");
var _baseExpandToggle = require("../base-expand-toggle");
var _messages = require("./messages");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
// eslint-disable-next-line @atlassian/tangerine/import/no-parent-imports

// eslint-disable-next-line @atlassian/tangerine/import/no-parent-imports

var ExpandToggle = exports.ExpandToggle = function ExpandToggle() {
  var _useEditorThemeContex = (0, _useEditorTheme.useEditorThemeContext)(),
    expanded = _useEditorThemeContex.expanded,
    toggleExpanded = _useEditorThemeContex.toggleExpanded;
  var _useStoreActions = (0, _state.useStoreActions)(),
    _useStoreActions2 = (0, _slicedToArray2.default)(_useStoreActions, 2),
    createAndFireAnalyticsEvent = _useStoreActions2[1].createAndFireAnalyticsEvent;
  var _useScopedId = (0, _state.useScopedId)(_constants.JQL_EDITOR_INPUT_ID),
    _useScopedId2 = (0, _slicedToArray2.default)(_useScopedId, 1),
    editorId = _useScopedId2[0];
  var _useIntl = (0, _state.useIntl)(),
    _useIntl2 = (0, _slicedToArray2.default)(_useIntl, 1),
    intl = _useIntl2[0];
  var message;
  if (expanded) {
    message = intl.formatMessage(_messages.messages.collapseTooltip);
  } else {
    message = intl.formatMessage(_messages.messages.expandTooltip);
  }
  var onClick = (0, _react.useCallback)(function () {
    createAndFireAnalyticsEvent({
      action: _analytics.Action.CLICKED,
      actionSubject: _analytics.ActionSubject.BUTTON,
      actionSubjectId: _analytics.ActionSubjectId.EDITOR_EXPAND,
      eventType: _analytics.EventType.UI,
      attributes: {
        expanded: !expanded
      }
    });
    toggleExpanded();
  }, [expanded, createAndFireAnalyticsEvent, toggleExpanded]);
  return /*#__PURE__*/_react.default.createElement(_tooltip.default, {
    position: 'bottom',
    content: /*#__PURE__*/_react.default.createElement(_styled.TooltipContent, null, message)
    // @ts-ignore: [PIT-1685] Fails in post-office due to backwards incompatibility issue with React 18
    ,
    tag: _tooltipTag.TooltipTag
  }, /*#__PURE__*/_react.default.createElement(_baseExpandToggle.BaseExpandToggle, {
    expanded: expanded,
    editorId: editorId,
    label: intl.formatMessage(_messages.messages.buttonLabel),
    onClick: onClick
  }));
};
