// @atlaskit/drawer/dist/cjs/components/primitives/sidebar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = require("@emotion/react");
var _colors = require("@atlaskit/theme/colors");
var _excluded = ["cssFn"];
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
var defaultStyle = {
  alignItems: 'center',
  boxSizing: 'border-box',
  color: "var(--ds-text-subtle, ".concat(_colors.N500, ")"),
  display: 'flex',
  flexShrink: 0,
  flexDirection: 'column',
  height: '100vh',
  paddingBottom: "var(--ds-space-200, 16px)",
  paddingTop: "var(--ds-space-300, 24px)",
  width: "var(--ds-space-800, 64px)"
};
var sidebarCSS = function sidebarCSS() {
  return defaultStyle;
};
var Sidebar = function Sidebar(_ref) {
  var cssFn = _ref.cssFn,
    props = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @repo/internal/react/no-unsafe-spread-props, @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  return (0, _react.jsx)("div", (0, _extends2.default)({
    css: (0, _react.css)(cssFn(defaultStyle))
  }, props));
};
var sidebarDefaults = {
  component: Sidebar,
  cssFn: sidebarCSS
};
var _default = exports.default = sidebarDefaults;
