// @atlaskit/user-picker/dist/es2019/components/ExternalUserOption/ExternalAvatarItemOption.js
/**
 * @jsxRuntime classic
 * @jsx jsx
 */

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import { css, jsx } from '@emotion/react';
import { B400 } from '@atlaskit/theme/colors';
const outerWrapper = isDisabled => css({
  alignItems: 'center',
  boxSizing: 'border-box',
  display: 'flex',
  // eslint-disable-next-line @atlaskit/design-system/use-tokens-typography
  lineHeight: 1,
  outline: 'none',
  margin: 0,
  width: '100%',
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  cursor: isDisabled ? 'not-allowed' : 'pointer',
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  opacity: isDisabled ? "var(--ds-opacity-disabled, 0.4)" : undefined
});
const detailsWrapper = css({
  display: 'flex',
  maxWidth: '100%',
  minWidth: 0,
  flex: '1 1 100%',
  // eslint-disable-next-line @atlaskit/design-system/use-tokens-typography
  lineHeight: 1.4,
  paddingLeft: "var(--ds-space-100, 8px)",
  alignItems: 'center'
});
const textSection = css({
  width: 'calc(100% - 32px)',
  flex: 'auto'
});
const getTextStyle = isSecondary => {
  const secondaryCssArgs = isSecondary ? {
    color: `var(--ds-text-selected, ${B400})`,
    font: "var(--ds-font-body-small, normal 400 11px/16px ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)"
  } : {};
  return css({
    display: 'flex',
    maxWidth: '100%',
    margin: 0,
    color: `var(--ds-text-selected, ${B400})`,
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    ...{
      secondaryCssArgs
    },
    whiteSpace: 'nowrap',
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
    '> span': {
      maxWidth: 'inherit'
    }
  });
};
export const ExternalAvatarItemOption = ({
  avatar,
  isDisabled,
  primaryText,
  secondaryText,
  sourcesInfoTooltip
}) =>
// eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage -- Ignored via go/DSP-18766
jsx("div", {
  css: outerWrapper(isDisabled)
}, avatar, jsx("div", {
  css: detailsWrapper
}, jsx("div", {
  css: textSection
}, jsx("div", null, jsx("div", {
  css: getTextStyle()
}, primaryText)), secondaryText && jsx("div", null, jsx("div", {
  css: getTextStyle(true)
}, secondaryText))), jsx("div", null, sourcesInfoTooltip)));
