// @atlaskit/editor-core/dist/cjs/CollapsedEditor/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireDefault(require("react"));
var _ChromeCollapsed = _interopRequireDefault(require("../ui/ChromeCollapsed"));
var _excluded = ["Editor"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); } // Ignored via go/ees005
// eslint-disable-next-line import/no-namespace
// eslint-disable-next-line @typescript-eslint/no-explicit-any
// Ignored via go/ees005
// eslint-disable-next-line @repo/internal/react/no-class-components
var CollapsedEditor = exports.default = /*#__PURE__*/function (_React$Component) {
  function CollapsedEditor() {
    var _this;
    (0, _classCallCheck2.default)(this, CollapsedEditor);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, CollapsedEditor, [].concat(args));
    (0, _defineProperty2.default)(_this, "state", {
      editorModules: CollapsedEditor.editorModules
    });
    return _this;
  }
  (0, _inherits2.default)(CollapsedEditor, _React$Component);
  return (0, _createClass2.default)(CollapsedEditor, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      if (!this.state.editorModules) {
        this.loadEditorModules();
      }
    }
  }, {
    key: "loadEditorModules",
    value: function loadEditorModules() {
      var _this2 = this;
      Promise.resolve().then(function () {
        return _interopRequireWildcard(require( /* webpackChunkName:"@atlaskit-internal_editor-core-async" */'../'));
      }).then(function (modules) {
        CollapsedEditor.editorModules = modules;
        _this2.setState({
          editorModules: modules
        });
      });
    }
  }, {
    key: "render",
    value: function render() {
      if (!this.props.isExpanded) {
        return /*#__PURE__*/_react.default.createElement(_ChromeCollapsed.default, {
          onFocus: this.props.onClickToExpand,
          text: this.props.placeholder
        });
      }
      if (!this.state.editorModules) {
        // Ignored via go/ees007
        // eslint-disable-next-line @atlaskit/editor/enforce-todo-comment-format
        // TODO: Proper loading state
        return /*#__PURE__*/_react.default.createElement(_ChromeCollapsed.default, {
          text: "Loading..."
        });
      }
      var _this$state$editorMod = this.state.editorModules,
        Editor = _this$state$editorMod.Editor,
        rest = (0, _objectWithoutProperties2.default)(_this$state$editorMod, _excluded);
      return this.props.renderEditor(Editor, rest);
    }
  }]);
}(_react.default.Component);
