// @atlaskit/media-viewer/dist/esm/styleWrappers.js
import _extends from "@babel/runtime/helpers/extends";
import React, { forwardRef } from 'react';
import { fg } from '@atlaskit/platform-feature-flags';
import { Blanket as CompiledBlanket, HeaderWrapper as CompiledHeaderWrapper, ListWrapper as CompiledListWrapper, ArrowsWrapper as CompiledArrowsWrapper, CloseButtonWrapper as CompiledCloseButtonWrapper, ContentWrapper as CompiledContentWrapper, ZoomWrapper as CompiledZoomWrapper, ZoomCenterControls as CompiledZoomCenterControls, ZoomRightControls as CompiledZoomRightControls, ZoomLevelIndicator as CompiledZoomLevelIndicator, HDIconGroupWrapper as CompiledHDIconGroupWrapper, ErrorMessageWrapper as CompiledErrorMessageWrapper, ErrorImage as CompiledErrorImage, Video as CompiledVideo, PDFWrapper as CompiledPDFWrapper, Arrow as CompiledArrow, LeftWrapper as CompiledLeftWrapper, RightWrapper as CompiledRightWrapper, Header as CompiledHeader, LeftHeader as CompiledLeftHeader, ImageWrapper as CompiledImageWrapper, BaselineExtend as CompiledBaselineExtend, Img as CompiledImg, MedatadataTextWrapper as CompiledMedatadataTextWrapper, MetadataWrapper as CompiledMetadataWrapper, MetadataFileName as CompiledMetadataFileName, MetadataSubText as CompiledMetadataSubText, MetadataIconWrapper as CompiledMetadataIconWrapper, RightHeader as CompiledRightHeader, CustomAudioPlayerWrapper as CompiledCustomAudioPlayerWrapper, AudioPlayer as CompiledAudioPlayer, Audio as CompiledAudio, AudioCover as CompiledAudioCover, DefaultCoverWrapper as CompiledDefaultCoverWrapper, DownloadButtonWrapper as CompiledDownloadButtonWrapper, CustomVideoPlayerWrapper as CompiledCustomVideoPlayerWrapper, SidebarWrapper as CompiledSidebarWrapper, SpinnerWrapper as CompiledSpinnerWrapper, FormattedMessageWrapper as CompiledFormattedMessageWrapper } from './styleWrappers-compiled';
import { Blanket as EmotionBlanket, HeaderWrapper as EmotionHeaderWrapper, ListWrapper as EmotionListWrapper, ArrowsWrapper as EmotionArrowsWrapper, CloseButtonWrapper as EmotionCloseButtonWrapper, ContentWrapper as EmotionContentWrapper, ZoomWrapper as EmotionZoomWrapper, ZoomCenterControls as EmotionZoomCenterControls, ZoomRightControls as EmotionZoomRightControls, ZoomLevelIndicator as EmotionZoomLevelIndicator, HDIconGroupWrapper as EmotionHDIconGroupWrapper, ErrorMessageWrapper as EmotionErrorMessageWrapper, ErrorImage as EmotionErrorImage, Video as EmotionVideo, PDFWrapper as EmotionPDFWrapper, Arrow as EmotionArrow, LeftWrapper as EmotionLeftWrapper, RightWrapper as EmotionRightWrapper, Header as EmotionHeader, LeftHeader as EmotionLeftHeader, ImageWrapper as EmotionImageWrapper, BaselineExtend as EmotionBaselineExtend, Img as EmotionImg, MedatadataTextWrapper as EmotionMedatadataTextWrapper, MetadataWrapper as EmotionMetadataWrapper, MetadataFileName as EmotionMetadataFileName, MetadataSubText as EmotionMetadataSubText, MetadataIconWrapper as EmotionMetadataIconWrapper, RightHeader as EmotionRightHeader, CustomAudioPlayerWrapper as EmotionCustomAudioPlayerWrapper, AudioPlayer as EmotionAudioPlayer, Audio as EmotionAudio, AudioCover as EmotionAudioCover, DefaultCoverWrapper as EmotionDefaultCoverWrapper, DownloadButtonWrapper as EmotionDownloadButtonWrapper, CustomVideoPlayerWrapper as EmotionCustomVideoPlayerWrapper, SidebarWrapper as EmotionSidebarWrapper, SpinnerWrapper as EmotionSpinnerWrapper, FormattedMessageWrapper as EmotionFormattedMessageWrapper } from './styleWrappers-emotion';
// We are keeping this data-testid since JIRA is still using it in their codebase to perform checks. Before removing this, we need to ensure this 'media-viewer-popup' test id is not being used anywhere else in other codebases
export var Blanket = function Blanket(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledBlanket, props) : /*#__PURE__*/React.createElement(EmotionBlanket, props);
};
export var HeaderWrapper = function HeaderWrapper(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledHeaderWrapper, props) : /*#__PURE__*/React.createElement(EmotionHeaderWrapper, props);
};
HeaderWrapper.displayName = 'HeaderWrapper';
export var ListWrapper = function ListWrapper(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledListWrapper, props) : /*#__PURE__*/React.createElement(EmotionListWrapper, props);
};
ListWrapper.displayName = 'ListWrapper';
export var ArrowsWrapper = function ArrowsWrapper(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledArrowsWrapper, props) : /*#__PURE__*/React.createElement(EmotionArrowsWrapper, props);
};
export var CloseButtonWrapper = function CloseButtonWrapper(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledCloseButtonWrapper, props) : /*#__PURE__*/React.createElement(EmotionCloseButtonWrapper, props);
};
export var ContentWrapper = function ContentWrapper(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledContentWrapper, props) : /*#__PURE__*/React.createElement(EmotionContentWrapper, props);
};
export var ZoomWrapper = function ZoomWrapper(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledZoomWrapper, props) : /*#__PURE__*/React.createElement(EmotionZoomWrapper, props);
};
export var ZoomCenterControls = function ZoomCenterControls(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledZoomCenterControls, props) : /*#__PURE__*/React.createElement(EmotionZoomCenterControls, props);
};
export var ZoomRightControls = function ZoomRightControls(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledZoomRightControls, props) : /*#__PURE__*/React.createElement(EmotionZoomRightControls, props);
};
export var ZoomLevelIndicator = function ZoomLevelIndicator(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledZoomLevelIndicator, _extends({}, props, {
    "data-testid": "zoom-level-indicator"
  })) : /*#__PURE__*/React.createElement(EmotionZoomLevelIndicator, _extends({}, props, {
    "data-testid": "zoom-level-indicator"
  }));
};
export var HDIconGroupWrapper = function HDIconGroupWrapper(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledHDIconGroupWrapper, props) : /*#__PURE__*/React.createElement(EmotionHDIconGroupWrapper, props);
};
export var ErrorMessageWrapper = function ErrorMessageWrapper(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledErrorMessageWrapper, props) : /*#__PURE__*/React.createElement(EmotionErrorMessageWrapper, props);
};
export var ErrorImage = function ErrorImage(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledErrorImage, props) : /*#__PURE__*/React.createElement(EmotionErrorImage, props);
};
export var Video = function Video(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledVideo, props) : /*#__PURE__*/React.createElement(EmotionVideo, props);
};
export var PDFWrapper = /*#__PURE__*/forwardRef(function (props, ref) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledPDFWrapper, _extends({
    ref: ref
  }, props)) : /*#__PURE__*/React.createElement(EmotionPDFWrapper, _extends({
    ref: ref
  }, props));
});
export var Arrow = function Arrow(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledArrow, props) : /*#__PURE__*/React.createElement(EmotionArrow, props);
};
export var LeftWrapper = function LeftWrapper(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledLeftWrapper, props) : /*#__PURE__*/React.createElement(EmotionLeftWrapper, props);
};
export var RightWrapper = function RightWrapper(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledRightWrapper, props) : /*#__PURE__*/React.createElement(EmotionRightWrapper, props);
};

// header.tsx

export var Header = function Header(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledHeader, props) : /*#__PURE__*/React.createElement(EmotionHeader, props);
};
export var LeftHeader = function LeftHeader(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledLeftHeader, props) : /*#__PURE__*/React.createElement(EmotionLeftHeader, props);
};
export var ImageWrapper = /*#__PURE__*/forwardRef(function (props, ref) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledImageWrapper, _extends({
    ref: ref
  }, props)) : /*#__PURE__*/React.createElement(EmotionImageWrapper, _extends({
    ref: ref
  }, props));
});
export var BaselineExtend = function BaselineExtend() {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledBaselineExtend, null) : /*#__PURE__*/React.createElement(EmotionBaselineExtend, null);
};
export var Img = function Img(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledImg, props) : /*#__PURE__*/React.createElement(EmotionImg, props);
};
export var MedatadataTextWrapper = function MedatadataTextWrapper(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledMedatadataTextWrapper, props) : /*#__PURE__*/React.createElement(EmotionMedatadataTextWrapper, props);
};
export var MetadataWrapper = function MetadataWrapper(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledMetadataWrapper, props) : /*#__PURE__*/React.createElement(EmotionMetadataWrapper, props);
};
export var MetadataFileName = function MetadataFileName(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledMetadataFileName, props) : /*#__PURE__*/React.createElement(EmotionMetadataFileName, props);
};
export var MetadataSubText = function MetadataSubText(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledMetadataSubText, props) : /*#__PURE__*/React.createElement(EmotionMetadataSubText, props);
};
export var MetadataIconWrapper = function MetadataIconWrapper(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledMetadataIconWrapper, props) : /*#__PURE__*/React.createElement(EmotionMetadataIconWrapper, props);
};
export var RightHeader = function RightHeader(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledRightHeader, props) : /*#__PURE__*/React.createElement(EmotionRightHeader, props);
};
export var CustomAudioPlayerWrapper = function CustomAudioPlayerWrapper(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledCustomAudioPlayerWrapper, props) : /*#__PURE__*/React.createElement(EmotionCustomAudioPlayerWrapper, props);
};
export var AudioPlayer = function AudioPlayer(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledAudioPlayer, props) : /*#__PURE__*/React.createElement(EmotionAudioPlayer, props);
};
AudioPlayer.displayName = 'AudioPlayer';
export var Audio = /*#__PURE__*/forwardRef(function (props, ref) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledAudio, _extends({
    ref: ref
  }, props)) : /*#__PURE__*/React.createElement(EmotionAudio, _extends({
    ref: ref
  }, props));
});
export var AudioCover = function AudioCover(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledAudioCover, props) : /*#__PURE__*/React.createElement(EmotionAudioCover, props);
};
export var DefaultCoverWrapper = function DefaultCoverWrapper(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledDefaultCoverWrapper, props) : /*#__PURE__*/React.createElement(EmotionDefaultCoverWrapper, props);
};
export var DownloadButtonWrapper = function DownloadButtonWrapper(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledDownloadButtonWrapper, props) : /*#__PURE__*/React.createElement(EmotionDownloadButtonWrapper, props);
};
export var CustomVideoPlayerWrapper = function CustomVideoPlayerWrapper(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledCustomVideoPlayerWrapper, props) : /*#__PURE__*/React.createElement(EmotionCustomVideoPlayerWrapper, props);
};
export var SidebarWrapper = function SidebarWrapper(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledSidebarWrapper, props) : /*#__PURE__*/React.createElement(EmotionSidebarWrapper, props);
};
export var SpinnerWrapper = function SpinnerWrapper(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledSpinnerWrapper, props) : /*#__PURE__*/React.createElement(EmotionSpinnerWrapper, props);
};
export var FormattedMessageWrapper = function FormattedMessageWrapper(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledFormattedMessageWrapper, props) : /*#__PURE__*/React.createElement(EmotionFormattedMessageWrapper, props);
};
