// @atlaskit/media-ui/dist/es2019/MediaInlineCard/ErroredView/index.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import React from 'react';
import { R300 } from '@atlaskit/theme/colors';
import WarningIcon from '@atlaskit/icon/utility/migration/warning';
import { Frame } from '../Frame';
import { IconAndTitleLayout } from '../IconAndTitleLayout';
import { AKIconWrapper } from '../Icon';
import { NoLinkAppearance } from '../styled';
export class MediaInlineCardErroredView extends React.Component {
  constructor(...args) {
    super(...args);
    _defineProperty(this, "renderMessage", () => {
      const {
        message
      } = this.props;
      const errorMessage = /*#__PURE__*/React.createElement(NoLinkAppearance, null, message);
      return /*#__PURE__*/React.createElement(React.Fragment, null, errorMessage);
    });
  }
  render() {
    const {
      onClick,
      isSelected,
      innerRef,
      testId = 'media-inline-card-errored-view',
      icon
    } = this.props;
    return /*#__PURE__*/React.createElement(Frame, {
      testId: testId,
      innerRef: innerRef,
      onClick: onClick,
      isSelected: isSelected
    }, /*#__PURE__*/React.createElement(IconAndTitleLayout, {
      icon: icon || /*#__PURE__*/React.createElement(AKIconWrapper, null, /*#__PURE__*/React.createElement(WarningIcon, {
        label: "error",
        LEGACY_size: "small",
        color: `var(--ds-icon-danger, ${R300})`
      })),
      title: this.renderMessage()
    }));
  }
}
