// @atlaskit/media-card/dist/esm/utils/analytics.js
import { __assign, __awaiter, __generator } from "tslib";
import { isPreviewableType, } from '@atlaskit/media-client';
import { version as packageVersion, name as packageName, } from '../version.json';
import { createAndFireEvent, } from '@atlaskit/analytics-next';
import { ANALYTICS_MEDIA_CHANNEL } from '../root/media-card-analytics-error-boundary';
export function getBaseAnalyticsContext(componentName) {
    if (componentName === void 0) { componentName = 'mediaCard'; }
    /*
      This Context provides data needed to build packageHierarchy in Atlaskit Analytics Listener and Media Analytics Listener
    */
    return {
        packageVersion: packageVersion,
        packageName: packageName,
        componentName: componentName,
        component: componentName,
    };
}
export var getFileAttributes = function (metadata, fileStatus) { return ({
    fileSource: 'mediaCard',
    fileMediatype: metadata && metadata.mediaType,
    fileMimetype: metadata && metadata.mimeType,
    fileId: metadata && metadata.id,
    fileSize: metadata && metadata.size,
    fileStatus: fileStatus,
}); };
export function getMediaCardAnalyticsContext(metadata, fileStatus) {
    return {
        fileAttributes: getFileAttributes(metadata, fileStatus && fileStatus.status),
    };
}
export function createAndFireCustomMediaEvent(payload, createAnalyticsEvent) {
    if (createAnalyticsEvent) {
        var event_1 = createAnalyticsEvent(payload);
        event_1.fire(ANALYTICS_MEDIA_CHANNEL);
    }
}
export var createAndFireMediaEvent = function (payload) {
    return createAndFireEvent(ANALYTICS_MEDIA_CHANNEL)(payload);
};
export var getAnalyticsStatusFromCardStatus = function (cardStatus) {
    switch (cardStatus) {
        case 'error':
        case 'failed-processing':
            return 'failed';
        default:
            return;
    }
};
export var getAnalyticsErrorStateAttributes = function (previewable, hasMinimalData, fileState, error) {
    var unknownError = 'unknown error';
    var errorMessage = error instanceof Error ? error.message : error;
    var errorMessageInFileState = fileState && 'message' in fileState && fileState.message;
    if (!fileState && !errorMessage) {
        return {};
    }
    if (!fileState) {
        return {
            failReason: 'media-client-error',
            error: errorMessage,
        };
    }
    if (!previewable) {
        if (hasMinimalData) {
            return {};
        }
        if (!errorMessageInFileState) {
            return {
                failReason: 'file-status-error',
                error: 'Does not have minimal metadata (filename and filesize) OR metadata/media-type is undefined',
            };
        }
    }
    if (fileState && ['error', 'failed-processing'].includes(fileState.status)) {
        return {
            failReason: 'file-status-error',
            error: errorMessageInFileState || unknownError,
        };
    }
    return {};
};
export var getCopiedFileAnalyticsPayload = function (identifier) { return __awaiter(void 0, void 0, void 0, function () {
    var _a, _b;
    return __generator(this, function (_c) {
        switch (_c.label) {
            case 0:
                _a = {
                    eventType: 'ui',
                    action: 'copied',
                    actionSubject: 'file'
                };
                if (!(identifier.mediaItemType === 'file')) return [3 /*break*/, 2];
                return [4 /*yield*/, identifier.id];
            case 1:
                _b = _c.sent();
                return [3 /*break*/, 3];
            case 2:
                _b = identifier.mediaItemType;
                _c.label = 3;
            case 3: return [2 /*return*/, (_a.actionSubjectId = _b,
                    _a)];
        }
    });
}); };
export var getMediaCardCommencedAnalyticsPayload = function (actionSubjectId) {
    return {
        eventType: 'operational',
        action: 'commenced',
        actionSubject: 'mediaCardRender',
        actionSubjectId: actionSubjectId,
    };
};
export var getAnalyticsStatus = function (previewable, hasMinimalData, status) {
    if (!previewable && hasMinimalData) {
        return;
    }
    return getAnalyticsStatusFromCardStatus(status);
};
export var hasFilenameAndFilesize = function (metadata) {
    return !!metadata && !!metadata.name && !!metadata.size;
};
export var fileIsPreviewable = function (metadata) {
    return !!metadata && !!metadata.mediaType && isPreviewableType(metadata.mediaType);
};
export var getLoadingStatusAnalyticsPayload = function (action, actionSubjectId, fileAttributes, errorState) {
    var payload = {
        eventType: 'operational',
        action: action,
        actionSubject: 'mediaCardRender',
        actionSubjectId: actionSubjectId,
    };
    if (!payload.attributes)
        payload.attributes = {};
    if (fileAttributes) {
        payload.attributes = __assign(__assign({}, payload.attributes), { fileAttributes: fileAttributes });
    }
    if (errorState) {
        payload.attributes.failReason = errorState.failReason;
        payload.attributes.error = errorState.error;
    }
    return payload;
};
