// @atlaskit/side-navigation/dist/cjs/components/NavigationContent/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./index.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _react = _interopRequireWildcard(require("react"));
var _useScrollbarWidth = _interopRequireDefault(require("@atlaskit/ds-lib/use-scrollbar-width"));
var _context = require("../NestableNavigationContent/context");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var outerContainerCSS = null;
var innerContainerCSS = {
  basic: "_1reo1wug _18m91wug _1e0c1txw _vchhusvi _1bsb1osq _kqswh2mm _2lx21bp4 _14mjgqwt _1ywu1ule _1qdgyh40 _18poh2mm _1bqqcs5v _tua5idpf _1hfk1q2g _aetrb3bt _3mxo1wug",
  topScrollIndicator: "_rfx3gqwt _cfu11ule _kfgtyh40 _1cs8stnw _1g0vcs5v _1rus1q2g _1kt9b3bt _1enwidpf _z5wtidpf"
};
var containerCSS = {
  basic: "_kqswh2mm _12l2u2gc _ahbqu2gc _1pfhidpf _1p3h12x7 _1l6bpxbi _auo4rdoj _hp2110yn",
  marginBlockStart: "_1pfhv77o"
};

/**
 * __Navigation content__
 *
 * A navigation content is used as the container for navigation items.
 *
 * - [Examples](https://atlassian.design/components/side-navigation/examples#content)
 * - [Code](https://atlassian.design/components/side-navigation/code)
 */
var NavigationContent = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var showTopScrollIndicator = props.showTopScrollIndicator,
    children = props.children,
    testId = props.testId;
  var _useShouldNestedEleme = (0, _context.useShouldNestedElementRender)(),
    shouldRender = _useShouldNestedEleme.shouldRender;
  var scrollbar = (0, _useScrollbarWidth.default)();
  if (!shouldRender) {
    return children;
  }
  var typedRef = ref;
  return /*#__PURE__*/_react.default.createElement("div", {
    ref: typedRef,
    "data-testid": testId,
    className: (0, _runtime.ax)(["_1reo15vq _18m915vq _1e0c1txw _4t3i1osq _kqswh2mm _rfx3gqwt _cfu11ule _kfgtyh40 _1cs8stnw _1g0vkb7n _1rus1x3l _1kt9b3bt _1enwftgi _z5wtu2gc _14mjgqwt _1ywu1ule _1qdgyh40 _18postnw _1bqqkb7n _tua5idpf _1hfk1x3l _aetrb3bt _1gufidpf _1czdftgi _g0nfu2gc"]),
    style: {
      "--_1uyno6z": (0, _runtime.ix)("var(--ds-menu-seperator-color, ".concat("var(--ds-border, #091E4224)", ")"))
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    ref: scrollbar.ref,
    className: (0, _runtime.ax)([innerContainerCSS.basic, !showTopScrollIndicator && innerContainerCSS.topScrollIndicator])
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _runtime.ax)([containerCSS.basic, showTopScrollIndicator && containerCSS.marginBlockStart])
  }, children)));
});
var _default = exports.default = NavigationContent;
