// @atlaskit/editor-plugin-placeholder-text/dist/cjs/ui/FloatingToolbar/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = FloatingToolbar;
var _react = _interopRequireDefault(require("react"));
var _react2 = require("@emotion/react");
var _ui = require("@atlaskit/editor-common/ui");
var _styles = require("./styles");
/* eslint-disable @atlaskit/design-system/prefer-primitives */
/**
 * @jsxRuntime classic
 * @jsx jsx
 */

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

function FloatingToolbar(_ref) {
  var children = _ref.children,
    target = _ref.target,
    offset = _ref.offset,
    fitWidth = _ref.fitWidth,
    _ref$fitHeight = _ref.fitHeight,
    fitHeight = _ref$fitHeight === void 0 ? 40 : _ref$fitHeight,
    onPositionCalculated = _ref.onPositionCalculated,
    popupsMountPoint = _ref.popupsMountPoint,
    popupsBoundariesElement = _ref.popupsBoundariesElement,
    className = _ref.className,
    absoluteOffset = _ref.absoluteOffset,
    alignX = _ref.alignX,
    alignY = _ref.alignY,
    zIndex = _ref.zIndex;
  if (!target) {
    return null;
  }
  return (0, _react2.jsx)(_ui.Popup, {
    absoluteOffset: absoluteOffset,
    alignX: alignX,
    alignY: alignY,
    target: target,
    zIndex: zIndex,
    mountTo: popupsMountPoint,
    boundariesElement: popupsBoundariesElement,
    offset: offset,
    fitWidth: fitWidth,
    fitHeight: fitHeight,
    onPositionCalculated: onPositionCalculated
  }, (0, _react2.jsx)("div", {
    // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
    css: (0, _styles.container)(fitHeight),
    "data-testid": "popup-container"
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    ,
    className: className
  }, children));
}
