// @atlaskit/editor-plugin-placeholder-text/dist/cjs/ui/FloatingToolbar/styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.container = void 0;
var _react = require("@emotion/react");
var _colors = require("@atlaskit/theme/colors");
var _constants = require("@atlaskit/theme/constants");
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var container = exports.container = function container(height) {
  return (0, _react.css)({
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    borderRadius: "".concat((0, _constants.borderRadius)(), "px"),
    boxShadow: "var(--ds-shadow-overlay, ".concat("0 12px 24px -6px ".concat(_colors.N50A, ", 0 0 1px ").concat(_colors.N60A), ")"),
    display: 'flex',
    alignItems: 'center',
    boxSizing: 'border-box',
    padding: "var(--ds-space-050, 4px)".concat(" ", "var(--ds-space-100, 8px)"),
    backgroundColor: "var(--ds-background-input, ".concat(_colors.N0, ")"),
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    height: height ? "".concat(height, "px") : undefined
  });
};
