// @atlaskit/media-card/dist/es2019/utils/cardActions/cardActionsDropdownMenu.js
import React from 'react';
import { Component } from 'react';
import MoreIcon from '@atlaskit/icon/glyph/more';
import DropdownMenu, { DropdownItemGroup, DropdownItem, } from '@atlaskit/dropdown-menu';
import { CardActionButton } from './styled';
import { withAnalyticsEvents, } from '@atlaskit/analytics-next';
import { createAndFireMediaEvent } from '../analytics';
const CardActionButtonWithProps = (props) => (React.createElement(CardActionButton, Object.assign({}, props)));
const CardActionButtonWithAnalytics = withAnalyticsEvents({
    onClick: createAndFireMediaEvent({
        eventType: 'ui',
        action: 'clicked',
        actionSubject: 'button',
        actionSubjectId: 'mediaCardDropDownMenu',
    }),
})(CardActionButtonWithProps);
const DropdownItemWithProps = (props) => (React.createElement(DropdownItem, Object.assign({ "data-testid": "media-card-actions-menu-item" }, props)));
const createDropdownItemWithAnalytics = (action, index) => {
    const { label, handler } = action;
    const DropdownItemWithAnalytics = withAnalyticsEvents({
        onClick: createAndFireMediaEvent({
            eventType: 'ui',
            action: 'clicked',
            actionSubject: 'button',
            actionSubjectId: 'mediaCardDropDownMenuItem',
            attributes: {
                label,
            },
        }),
    })(DropdownItemWithProps);
    return (React.createElement(DropdownItemWithAnalytics, { key: index, onClick: handler }, label));
};
export class CardActionsDropdownMenu extends Component {
    render() {
        const { actions, triggerColor, onOpenChange } = this.props;
        if (actions.length > 0) {
            return (React.createElement(DropdownMenu, { "data-testid": "media-card-actions-menu", onOpenChange: onOpenChange, trigger: React.createElement(CardActionButtonWithAnalytics, { style: { color: triggerColor } },
                    React.createElement(MoreIcon, { label: "more" })) },
                React.createElement(DropdownItemGroup, null, actions.map(createDropdownItemWithAnalytics))));
        }
        else {
            return null;
        }
    }
}
