// @atlaskit/editor-plugin-media/dist/cjs/pm-plugins/commands.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.trackMediaPaste = exports.showMediaViewer = exports.hideMediaViewer = void 0;
var _actions = require("../pm-plugins/actions");
var _pluginKey = require("../pm-plugins/plugin-key");
var _mediaCommon = require("../pm-plugins/utils/media-common");
var showMediaViewer = exports.showMediaViewer = function showMediaViewer(media) {
  return function (_ref) {
    var tr = _ref.tr;
    tr.setMeta(_pluginKey.stateKey, {
      type: _actions.ACTIONS.SHOW_MEDIA_VIEWER,
      mediaViewerSelectedMedia: media,
      isMediaViewerVisible: true
    });
    return tr;
  };
};
var hideMediaViewer = exports.hideMediaViewer = function hideMediaViewer(_ref2) {
  var tr = _ref2.tr;
  tr.setMeta(_pluginKey.stateKey, {
    type: _actions.ACTIONS.HIDE_MEDIA_VIEWER,
    mediaViewerSelectedMedia: null,
    isMediaViewerVisible: false
  });
  return tr;
};
var trackMediaPaste = exports.trackMediaPaste = function trackMediaPaste(attrs) {
  return function (_ref3) {
    var tr = _ref3.tr;
    var identifier = (0, _mediaCommon.getIdentifier)(attrs);
    tr.setMeta(_pluginKey.stateKey, {
      type: _actions.ACTIONS.TRACK_MEDIA_PASTE,
      identifier: identifier
    });
    return tr;
  };
};
