// @atlaskit/link-datasource/dist/cjs/ui/issue-like-table/render-type/link/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.LINK_TYPE_TEST_ID = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _smartCard = require("@atlaskit/smart-card");
var _hoverCard = require("@atlaskit/smart-card/hover-card");
var _linkUrl = _interopRequireDefault(require("@atlaskit/smart-card/link-url"));
var _colors = require("@atlaskit/theme/colors");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var linkStyles = {
  key: {
    fontWeight: "var(--ds-font-weight-semibold, 600)",
    color: "var(--ds-text-subtlest, ".concat(_colors.N300, ")"),
    marginTop: "var(--ds-space-250, 20px)"
  },
  default: {}
};
var LINK_TYPE_TEST_ID = exports.LINK_TYPE_TEST_ID = 'link-datasource-render-type--link';
var LinkRenderType = function LinkRenderType(_ref) {
  var style = _ref.style,
    url = _ref.url,
    text = _ref.text,
    _ref$testId = _ref.testId,
    testId = _ref$testId === void 0 ? LINK_TYPE_TEST_ID : _ref$testId;
  var linkStyle = (0, _react.useMemo)(function () {
    return (style === null || style === void 0 ? void 0 : style.appearance) && linkStyles[style.appearance] || {};
  }, [style]);
  var anchor = (0, _react.useMemo)(function () {
    return /*#__PURE__*/_react.default.createElement(_hoverCard.HoverCard, {
      url: url
    }, /*#__PURE__*/_react.default.createElement(_linkUrl.default, (0, _extends2.default)({
      href: url
      // NOTE: This will no longer apply styles to `@atlaskit/link` when platform_editor_hyperlink_underline is enabled.
      // Wrap `@atlaskit/link` in a Text component to provide font styles to Link
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
    }, (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv') ? {} : {
      style: _objectSpread({}, linkStyle)
    }, {
      "data-testid": testId,
      target: "_blank"
    }), text || url));
  }, [linkStyle, url, text, testId]);
  var SmartCard = function SmartCard() {
    var handleClick = function handleClick(e) {
      e.preventDefault();
      window.open(url, '_blank', 'noopener, noreferrer');
    };
    return /*#__PURE__*/_react.default.createElement(_smartCard.Card, {
      appearance: "inline",
      onClick: handleClick,
      url: url,
      testId: testId,
      fallbackComponent: function fallbackComponent() {
        return anchor;
      }
    });
  };

  // url can be undefined before data is fetched whilst adding new link column to display
  if (!url) {
    return null;
  }
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, text ? anchor : /*#__PURE__*/_react.default.createElement(SmartCard, null), /*#__PURE__*/_react.default.createElement("br", null));
};
var _default = exports.default = /*#__PURE__*/_react.default.memo(LinkRenderType);
