// @atlaskit/user-picker/dist/esm/components/TeamOption/main.js
import _extends from "@babel/runtime/helpers/extends";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
function _callSuper(t, o, e) { return o = _getPrototypeOf(o), _possibleConstructorReturn(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], _getPrototypeOf(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
import { B400, N800, N200 } from '@atlaskit/theme/colors';
import { fg } from '@atlaskit/platform-feature-flags';
import { Inline } from '@atlaskit/primitives/compiled';
import { VerifiedTeamIcon } from '@atlaskit/people-teams-ui-public/verified-team-icon';
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import { jsx } from '@emotion/react';
import React from 'react';
import { FormattedMessage } from 'react-intl-next';
import { AvatarItemOption, textWrapper } from '../AvatarItemOption';
import { HighlightText } from '../HighlightText';
import { messages } from '../i18n';
import { SizeableAvatar } from '../SizeableAvatar';
export var TeamOption = /*#__PURE__*/function (_React$PureComponent) {
  function TeamOption() {
    var _this;
    _classCallCheck(this, TeamOption);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, TeamOption, [].concat(args));
    _defineProperty(_this, "getPrimaryText", function () {
      var _this$props$team = _this.props.team,
        name = _this$props$team.name,
        highlight = _this$props$team.highlight,
        verified = _this$props$team.verified;
      return [jsx("span", {
        key: "name"
        // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
        ,
        css: textWrapper(_this.props.isSelected ? "var(--ds-text-selected, ".concat(B400, ")") : "var(--ds-text, ".concat(N800, ")"))
      }, fg('verified-team-in-user-picker') ? jsx(Inline, {
        alignBlock: "center"
      }, jsx(HighlightText, {
        highlights: highlight && highlight.name
      }, name), verified && jsx(VerifiedTeamIcon, null)) : jsx(HighlightText, {
        highlights: highlight && highlight.name
      }, name))];
    });
    _defineProperty(_this, "renderByline", function () {
      var _this$props = _this.props,
        isSelected = _this$props.isSelected,
        _this$props$team2 = _this$props.team,
        memberCount = _this$props$team2.memberCount,
        includesYou = _this$props$team2.includesYou;

      // if Member count is missing, do not show the byline, regardless of the availability of includesYou
      if (memberCount === null || typeof memberCount === 'undefined') {
        return undefined;
      } else {
        if (includesYou === true) {
          if (memberCount > 50) {
            return _this.getBylineComponent(isSelected, jsx(FormattedMessage, messages.plus50MembersWithYou));
          } else {
            return _this.getBylineComponent(isSelected, jsx(FormattedMessage, _extends({}, messages.memberCountWithYou, {
              values: {
                count: memberCount
              }
            })));
          }
        } else {
          if (memberCount > 50) {
            return _this.getBylineComponent(isSelected, jsx(FormattedMessage, messages.plus50MembersWithoutYou));
          } else {
            return _this.getBylineComponent(isSelected, jsx(FormattedMessage, _extends({}, messages.memberCountWithoutYou, {
              values: {
                count: memberCount
              }
            })));
          }
        }
      }
    });
    _defineProperty(_this, "getBylineComponent", function (isSelected, message) {
      return jsx("span", {
        // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
        css: textWrapper(isSelected ? "var(--ds-text-selected, ".concat(B400, ")") : "var(--ds-text-subtlest, ".concat(N200, ")")),
        "data-testid": "user-picker-team-secondary-text"
      }, message);
    });
    _defineProperty(_this, "renderAvatar", function () {
      var avatarUrl = _this.props.team.avatarUrl;
      return jsx(SizeableAvatar, {
        appearance: "big",
        src: avatarUrl,
        type: "team"
      });
    });
    _defineProperty(_this, "getLozengeProps", function () {
      return typeof _this.props.team.lozenge === 'string' ? {
        text: _this.props.team.lozenge
      } : _this.props.team.lozenge;
    });
    _defineProperty(_this, "renderCustomByLine", function () {
      var _this$props$team3;
      if (!((_this$props$team3 = _this.props.team) !== null && _this$props$team3 !== void 0 && _this$props$team3.byline)) {
        return undefined;
      }
      return jsx("span", {
        // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
        css: textWrapper(_this.props.isSelected ? "var(--ds-text-selected, ".concat(B400, ")") : "var(--ds-text-subtlest, ".concat(N200, ")"))
      }, _this.props.team.byline);
    });
    return _this;
  }
  _inherits(TeamOption, _React$PureComponent);
  return _createClass(TeamOption, [{
    key: "render",
    value: function render() {
      return jsx(AvatarItemOption, {
        avatar: this.renderAvatar(),
        isDisabled: this.props.team.isDisabled,
        lozenge: this.getLozengeProps(),
        primaryText: this.getPrimaryText(),
        secondaryText: this.renderCustomByLine() || this.renderByline()
      });
    }
  }]);
}(React.PureComponent);
