// @atlaskit/editor-common/dist/cjs/media-inline/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DEFAULT_IMAGE_HEIGHT", {
  enumerable: true,
  get: function get() {
    return _constants.DEFAULT_IMAGE_HEIGHT;
  }
});
Object.defineProperty(exports, "DEFAULT_IMAGE_WIDTH", {
  enumerable: true,
  get: function get() {
    return _constants.DEFAULT_IMAGE_WIDTH;
  }
});
Object.defineProperty(exports, "DEFAULT_INLINE_IMAGE_ASPECT_RATIO", {
  enumerable: true,
  get: function get() {
    return _constants.DEFAULT_INLINE_IMAGE_ASPECT_RATIO;
  }
});
Object.defineProperty(exports, "INLINE_IMAGE_WRAPPER_CLASS_NAME", {
  enumerable: true,
  get: function get() {
    return _styles.INLINE_IMAGE_WRAPPER_CLASS_NAME;
  }
});
Object.defineProperty(exports, "InlineImageWrapper", {
  enumerable: true,
  get: function get() {
    return _inlineImageWrapper.InlineImageWrapper;
  }
});
Object.defineProperty(exports, "MediaInlineImageCard", {
  enumerable: true,
  get: function get() {
    return _mediaInlineImageCard.MediaInlineImageCard;
  }
});
Object.defineProperty(exports, "mediaInlineImageStyles", {
  enumerable: true,
  get: function get() {
    return _styles.mediaInlineImageStyles;
  }
});
Object.defineProperty(exports, "mediaInlineImagesEnabled", {
  enumerable: true,
  get: function get() {
    return _utils.mediaInlineImagesEnabled;
  }
});
var _mediaInlineImageCard = require("./media-inline-image-card");
var _inlineImageWrapper = require("./inline-image-wrapper");
var _constants = require("./constants");
var _styles = require("./styles");
var _utils = require("./utils");
