// @atlaskit/media-card/dist/cjs/card/card.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CardWithPerformanceObserver = exports.CardBase = exports.Card = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _mediaClient = require("@atlaskit/media-client");
var _mediaCommon = require("@atlaskit/media-common");
var _react = _interopRequireWildcard(require("react"));
var _reactIntlNext = require("react-intl-next");
var _externalImageCard = require("./externalImageCard");
var _fileCard = require("./fileCard");
var _mediaPerformanceObserver = require("../utils/mediaPerformanceObserver/mediaPerformanceObserver");
var _analyticsNext = require("@atlaskit/analytics-next");
var _label = _interopRequireDefault(require("@atlaskit/react-ufo/label"));
var _excluded = ["identifier"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var packageName = "@atlaskit/media-card";
var packageVersion = "79.2.0";
var CardBase = exports.CardBase = function CardBase(_ref) {
  var identifier = _ref.identifier,
    otherProps = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  var innerContent = (0, _mediaClient.isFileIdentifier)(identifier) ? /*#__PURE__*/_react.default.createElement(_label.default, {
    name: "media-card-file-card"
  }, /*#__PURE__*/_react.default.createElement(_fileCard.FileCard, (0, _extends2.default)({}, otherProps, {
    identifier: identifier,
    key: identifier.id
  }))) : /*#__PURE__*/_react.default.createElement(_label.default, {
    name: "media-card-external-image"
  }, /*#__PURE__*/_react.default.createElement(_externalImageCard.ExternalImageCard, (0, _extends2.default)({}, otherProps, {
    identifier: identifier,
    key: identifier.dataURI
  })));
  return otherProps.intl ? innerContent : /*#__PURE__*/_react.default.createElement(_reactIntlNext.IntlProvider, {
    locale: "en"
  }, innerContent);
};
var CardWithPerformanceObserver = exports.CardWithPerformanceObserver = function CardWithPerformanceObserver(props) {
  var _useAnalyticsEvents = (0, _analyticsNext.useAnalyticsEvents)(),
    createAnalyticsEvent = _useAnalyticsEvents.createAnalyticsEvent;
  (0, _react.useEffect)(function () {
    (0, _mediaPerformanceObserver.startResourceObserver)();
  }, []);
  (0, _react.useEffect)(function () {
    (0, _mediaPerformanceObserver.setAnalyticsContext)(createAnalyticsEvent);
  }, [createAnalyticsEvent]);
  return /*#__PURE__*/_react.default.createElement(CardBase, props);
};
var Card = exports.Card = (0, _mediaCommon.withMediaAnalyticsContext)({
  packageVersion: packageVersion,
  packageName: packageName,
  componentName: 'mediaCard',
  component: 'mediaCard'
})((0, _reactIntlNext.injectIntl)(CardWithPerformanceObserver, {
  enforceContext: false
}));
