// @atlaskit/media-card/dist/cjs/card/cardLoader.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _reactLoadable = _interopRequireDefault(require("react-loadable"));
var _cardLoading = require("../utils/lightCards/cardLoading");
var _segment = _interopRequireDefault(require("@atlaskit/react-ufo/segment"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var MediaCardContext = /*#__PURE__*/_react.default.createContext({});
var CardLoadingWithContext = function CardLoadingWithContext() {
  var props = (0, _react.useContext)(MediaCardContext);
  return /*#__PURE__*/_react.default.createElement(_cardLoading.CardLoading, (0, _extends2.default)({}, props, {
    interactionName: "media-card-async-loading"
  }));
};
var MediaCardWithMediaClientProvider = (0, _reactLoadable.default)({
  loader: function loader() {
    return Promise.resolve().then(function () {
      return _interopRequireWildcard(require( /* webpackChunkName: "@atlaskit-internal_media-card-with-media-client" */'./cardWithMediaClient'));
    }).then(function (mod) {
      return mod.CardWithMediaClient;
    });
  },
  loading: function loading() {
    return /*#__PURE__*/_react.default.createElement(CardLoadingWithContext, null);
  }
});
var CardLoader = function CardLoader(props) {
  return /*#__PURE__*/_react.default.createElement(_segment.default, {
    name: "media-card",
    mode: "list"
  }, /*#__PURE__*/_react.default.createElement(MediaCardContext.Provider, {
    value: props
  }, /*#__PURE__*/_react.default.createElement(MediaCardWithMediaClientProvider, props)));
};
CardLoader.preload = function () {
  return MediaCardWithMediaClientProvider.preload();
};
var _default = exports.default = CardLoader;
