// @atlaskit/media-card/dist/cjs/card/cardView.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CardViewBase = exports.CardView = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _mediaClient = require("@atlaskit/media-client");
var _analyticsNext = require("@atlaskit/analytics-next");
var _mimeTypeIcon = require("@atlaskit/media-ui/mime-type-icon");
var _spinner = _interopRequireDefault(require("@atlaskit/spinner"));
var _tooltip = _interopRequireDefault(require("@atlaskit/tooltip"));
var _useCallbackRef = require("use-callback-ref");
var _mediaUi = require("@atlaskit/media-ui");
var _analytics = require("../utils/analytics");
var _imageRenderer = require("./ui/imageRenderer");
var _titleBox = require("./ui/titleBox/titleBox");
var _failedTitleBox = require("./ui/titleBox/failedTitleBox");
var _progressBar = require("./ui/progressBar/progressBar");
var _playButton = require("./ui/playButton/playButton");
var _tickBox = require("./ui/tickBox/tickBox");
var _blanket = require("./ui/blanket/blanket");
var _actionsBar = require("./ui/actionsBar/actionsBar");
var _iconWrapper = require("./ui/iconWrapper/iconWrapper");
var _iconMessage2 = require("./ui/iconMessage");
var _errors = require("../errors");
var _wrapper = require("./ui/wrapper");
var _classnames = require("./classnames");
var _useBreakpoint = require("./useBreakpoint");
var _openMediaViewerButton = _interopRequireDefault(require("./ui/openMediaViewerButton/openMediaViewerButton"));
var _useCurrentValueRef = require("../utils/useCurrentValueRef");
var _svgView = require("./svgView");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var CardViewBase = exports.CardViewBase = function CardViewBase(_ref) {
  var identifier = _ref.identifier,
    _ref$innerRef = _ref.innerRef,
    innerRef = _ref$innerRef === void 0 ? null : _ref$innerRef,
    onImageLoad = _ref.onImageLoad,
    onImageError = _ref.onImageError,
    dimensions = _ref.dimensions,
    onClick = _ref.onClick,
    onMouseEnter = _ref.onMouseEnter,
    testId = _ref.testId,
    metadata = _ref.metadata,
    status = _ref.status,
    selected = _ref.selected,
    selectable = _ref.selectable,
    cardPreview = _ref.cardPreview,
    mediaCardCursor = _ref.mediaCardCursor,
    shouldHideTooltip = _ref.shouldHideTooltip,
    progress = _ref.progress,
    alt = _ref.alt,
    resizeMode = _ref.resizeMode,
    onDisplayImage = _ref.onDisplayImage,
    nativeLazyLoad = _ref.nativeLazyLoad,
    forceSyncDisplay = _ref.forceSyncDisplay,
    actions = _ref.actions,
    disableOverlay = _ref.disableOverlay,
    titleBoxBgColor = _ref.titleBoxBgColor,
    titleBoxIcon = _ref.titleBoxIcon,
    error = _ref.error,
    disableAnimation = _ref.disableAnimation,
    _ref$openMediaViewerB = _ref.openMediaViewerButtonRef,
    openMediaViewerButtonRef = _ref$openMediaViewerB === void 0 ? null : _ref$openMediaViewerB,
    shouldOpenMediaViewer = _ref.shouldOpenMediaViewer,
    overriddenCreationDate = _ref.overriddenCreationDate,
    onSvgError = _ref.onSvgError,
    onSvgLoad = _ref.onSvgLoad;
  var _useState = (0, _react.useState)(false),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    didSvgRender = _useState2[0],
    setDidSvgRender = _useState2[1];
  var _useState3 = (0, _react.useState)(false),
    _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
    didImageRender = _useState4[0],
    setDidImageRender = _useState4[1];
  var divRef = (0, _react.useRef)(null);
  var prevCardPreviewRef = (0, _react.useRef)();
  var breakpoint = (0, _useBreakpoint.useBreakpoint)(dimensions === null || dimensions === void 0 ? void 0 : dimensions.width, divRef);
  (0, _react.useEffect)(function () {
    // We should only switch didImageRender to false when cardPreview goes undefined, not when it is changed. as this method could be triggered after onImageLoad callback, falling on a race condition
    if (prevCardPreviewRef.current && !cardPreview) {
      setDidImageRender(false);
    }
    prevCardPreviewRef.current = cardPreview;
  }, [cardPreview]);
  var handleOnImageLoad = function handleOnImageLoad(prevCardPreview) {
    if (prevCardPreview.dataURI !== (cardPreview === null || cardPreview === void 0 ? void 0 : cardPreview.dataURI)) {
      return;
    }
    /*
        We render the icon & icon message always, even if there is cardPreview available.
        If the image fails to load/render, the icon will remain, i.e. the user won't see a change until the root card decides to chage status to error.
        If the image renders successfully, we switch this variable to hide the icon & icon message behind the thumbnail in case the image has transparency.
        It is less likely that root component replaces a suceeded cardPreview for a failed one than the opposite case. Therefore we prefer to hide the icon instead show when the image fails, for a smoother transition
      */
    setDidImageRender(true);
    onImageLoad === null || onImageLoad === void 0 || onImageLoad(cardPreview);
  };
  var handleOnImageError = function handleOnImageError(prevCardPreview) {
    if (prevCardPreview.dataURI !== (cardPreview === null || cardPreview === void 0 ? void 0 : cardPreview.dataURI)) {
      return;
    }
    setDidImageRender(false);
    onImageError === null || onImageError === void 0 || onImageError(cardPreview);
  };
  var shouldRenderPlayButton = function shouldRenderPlayButton() {
    var _ref2 = metadata || {},
      mediaType = _ref2.mediaType;
    if (mediaType !== 'video' || !cardPreview) {
      return false;
    }
    return true;
  };
  var mergedRef = (0, _useCallbackRef.useMergeRefs)([divRef, innerRef]);
  var getRenderConfigByStatus = function getRenderConfigByStatus() {
    var _ref3 = metadata || {},
      name = _ref3.name,
      mediaType = _ref3.mediaType,
      mimeType = _ref3.mimeType;
    var isZeroSize = metadata && metadata.size === 0;
    var defaultConfig = {
      renderTypeIcon: !didImageRender && !didSvgRender,
      renderImageRenderer: !didSvgRender,
      renderSvgView: mimeType === 'image/svg+xml',
      renderPlayButton: !!cardPreview && mediaType === 'video',
      renderBlanket: !disableOverlay,
      renderTitleBox: !disableOverlay,
      renderTickBox: !disableOverlay && !!selectable
    };
    var loadingConfig = _objectSpread(_objectSpread({}, defaultConfig), {}, {
      renderPlayButton: false,
      renderTypeIcon: false,
      renderSpinner: !didImageRender && !didSvgRender
    });
    switch (status) {
      case 'uploading':
        return _objectSpread(_objectSpread({}, defaultConfig), {}, {
          renderBlanket: !disableOverlay || mediaType !== 'video',
          isFixedBlanket: true,
          renderProgressBar: true
        });
      case 'processing':
        return _objectSpread(_objectSpread({}, defaultConfig), {}, {
          iconMessage: !didImageRender && !isZeroSize ? /*#__PURE__*/_react.default.createElement(_iconMessage2.CreatingPreview, {
            disableAnimation: disableAnimation
          }) : undefined
        });
      case 'complete':
        return defaultConfig;
      case 'error':
      case 'failed-processing':
        if (status === 'failed-processing' && mimeType === 'image/svg+xml') {
          return loadingConfig;
        }
        var baseErrorConfig = _objectSpread(_objectSpread({}, defaultConfig), {}, {
          renderTypeIcon: true,
          renderImageRenderer: false,
          renderSvgView: false,
          renderTitleBox: false,
          renderPlayButton: false
        });
        var _iconMessage;
        var _customTitleMessage;
        if ((0, _errors.isUploadError)(error)) {
          _iconMessage = /*#__PURE__*/_react.default.createElement(_iconMessage2.FailedToUpload, null);
          _customTitleMessage = _mediaUi.messages.failed_to_upload;
        } else if (!metadata) {
          _iconMessage = /*#__PURE__*/_react.default.createElement(_iconMessage2.FailedToLoad, null);
        } else {
          _iconMessage = /*#__PURE__*/_react.default.createElement(_iconMessage2.PreviewUnavailable, null);
        }
        if (!disableOverlay) {
          var _renderFailedTitleBox = !name || !!_customTitleMessage;
          return _objectSpread(_objectSpread({}, baseErrorConfig), {}, {
            renderTitleBox: !!name && !_customTitleMessage,
            renderFailedTitleBox: _renderFailedTitleBox,
            iconMessage: !_renderFailedTitleBox ? _iconMessage : undefined,
            customTitleMessage: _customTitleMessage
          });
        }
        return _objectSpread(_objectSpread({}, baseErrorConfig), {}, {
          iconMessage: _iconMessage
        });
      case 'loading-preview':
      case 'loading':
      default:
        return loadingConfig;
    }
  };
  var _getRenderConfigBySta = getRenderConfigByStatus(),
    renderTypeIcon = _getRenderConfigBySta.renderTypeIcon,
    iconMessage = _getRenderConfigBySta.iconMessage,
    renderImageRenderer = _getRenderConfigBySta.renderImageRenderer,
    renderSvgView = _getRenderConfigBySta.renderSvgView,
    renderSpinner = _getRenderConfigBySta.renderSpinner,
    renderPlayButton = _getRenderConfigBySta.renderPlayButton,
    renderBlanket = _getRenderConfigBySta.renderBlanket,
    renderProgressBar = _getRenderConfigBySta.renderProgressBar,
    renderTitleBox = _getRenderConfigBySta.renderTitleBox,
    renderFailedTitleBox = _getRenderConfigBySta.renderFailedTitleBox,
    renderTickBox = _getRenderConfigBySta.renderTickBox,
    isFixedBlanket = _getRenderConfigBySta.isFixedBlanket,
    customTitleMessage = _getRenderConfigBySta.customTitleMessage;
  var shouldDisplayBackground = !cardPreview || !disableOverlay || status === 'error' || status === 'failed-processing';
  var isPlayButtonClickable = shouldRenderPlayButton() && !!disableOverlay;
  var isTickBoxSelectable = !disableOverlay && !!selectable && !selected;
  // Disable tooltip for Media Single
  var shouldDisplayTooltip = !disableOverlay && !shouldHideTooltip;
  var _ref4 = metadata || {},
    mediaType = _ref4.mediaType,
    mimeType = _ref4.mimeType,
    name = _ref4.name,
    createdAt = _ref4.createdAt;
  var isTitleBoxVisible = renderTitleBox && name;
  var hasVisibleTitleBox = !!(isTitleBoxVisible || renderFailedTitleBox);
  var metadataRef = (0, _useCurrentValueRef.useCurrentValueRef)(metadata);
  var actionsWithDetails = (0, _react.useMemo)(function () {
    if (!actions) {
      return [];
    }
    return actions.map(function (action) {
      return _objectSpread(_objectSpread({}, action), {}, {
        handler: function handler() {
          if (!metadataRef.current) {
            action.handler();
          } else {
            action.handler({
              type: 'file',
              details: metadataRef.current
            });
          }
        }
      });
    });
  }, [actions, metadataRef]);
  var onSvgLoadBase = function onSvgLoadBase() {
    setDidSvgRender(true);
    onSvgLoad === null || onSvgLoad === void 0 || onSvgLoad();
  };
  var contents = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_wrapper.ImageContainer, {
    centerElements: didSvgRender && !(0, _platformFeatureFlags.fg)('platform_media_card_image_render') // The whole centerElements styles can go after removing the flag
    ,
    testId: _classnames.fileCardImageViewSelector,
    mediaName: name,
    status: status,
    progress: progress,
    selected: selected,
    source: cardPreview === null || cardPreview === void 0 ? void 0 : cardPreview.source
  }, renderTypeIcon && /*#__PURE__*/_react.default.createElement(_iconWrapper.IconWrapper, {
    breakpoint: breakpoint,
    hasTitleBox: hasVisibleTitleBox
  }, /*#__PURE__*/_react.default.createElement(_mimeTypeIcon.MimeTypeIcon, {
    testId: "media-card-file-type-icon",
    mediaType: mediaType,
    mimeType: mimeType,
    name: name
  }), iconMessage), renderSpinner && /*#__PURE__*/_react.default.createElement(_iconWrapper.IconWrapper, {
    breakpoint: breakpoint,
    hasTitleBox: hasVisibleTitleBox
  }, /*#__PURE__*/_react.default.createElement(_spinner.default, {
    testId: "media-card-loading",
    interactionName: "media-card-loading"
  })), renderSvgView && identifier && (0, _mediaClient.isFileIdentifier)(identifier) && /*#__PURE__*/_react.default.createElement(_svgView.SvgView, {
    identifier: identifier,
    resizeMode: resizeMode || 'crop',
    onError: onSvgError,
    onLoad: onSvgLoadBase,
    wrapperRef: divRef
  }), renderImageRenderer && identifier && /*#__PURE__*/_react.default.createElement(_imageRenderer.ImageRenderer, {
    cardPreview: cardPreview,
    mediaType: (metadata === null || metadata === void 0 ? void 0 : metadata.mediaType) || 'unknown',
    alt: alt !== null && alt !== void 0 ? alt : name,
    resizeMode: resizeMode,
    onDisplayImage: onDisplayImage,
    onImageLoad: handleOnImageLoad,
    onImageError: handleOnImageError,
    nativeLazyLoad: nativeLazyLoad,
    forceSyncDisplay: forceSyncDisplay,
    identifier: identifier,
    wrapperRef: divRef
  }), renderPlayButton && /*#__PURE__*/_react.default.createElement(_iconWrapper.IconWrapper, {
    breakpoint: breakpoint,
    hasTitleBox: hasVisibleTitleBox
  }, /*#__PURE__*/_react.default.createElement(_playButton.PlayButton, null)), renderBlanket && /*#__PURE__*/_react.default.createElement(_blanket.Blanket, {
    isFixed: isFixedBlanket
  }), renderTitleBox && /*#__PURE__*/_react.default.createElement(_titleBox.TitleBox, {
    name: name,
    createdAt: overriddenCreationDate !== null && overriddenCreationDate !== void 0 ? overriddenCreationDate : createdAt,
    breakpoint: breakpoint,
    titleBoxIcon: titleBoxIcon,
    titleBoxBgColor: titleBoxBgColor,
    hidden: !isTitleBoxVisible
  }), renderFailedTitleBox && /*#__PURE__*/_react.default.createElement(_failedTitleBox.FailedTitleBox, {
    breakpoint: breakpoint,
    customMessage: customTitleMessage
  }), renderProgressBar && /*#__PURE__*/_react.default.createElement(_progressBar.ProgressBar, {
    progress: progress,
    breakpoint: breakpoint,
    positionBottom: !hasVisibleTitleBox
  }), renderTickBox && /*#__PURE__*/_react.default.createElement(_tickBox.TickBox, {
    selected: selected
  })), disableOverlay || !actions || actions.length === 0 ? null : /*#__PURE__*/_react.default.createElement(_actionsBar.ActionsBar, {
    filename: name,
    actions: actionsWithDetails
  }));
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, shouldOpenMediaViewer && /*#__PURE__*/_react.default.createElement(_openMediaViewerButton.default, {
    fileName: name !== null && name !== void 0 ? name : '',
    innerRef: openMediaViewerButtonRef,
    onClick: onClick
  }), /*#__PURE__*/_react.default.createElement(_wrapper.Wrapper, {
    testId: testId || 'media-card-view',
    dimensions: dimensions,
    onClick: onClick,
    onMouseEnter: onMouseEnter,
    innerRef: mergedRef,
    breakpoint: breakpoint,
    mediaCardCursor: mediaCardCursor,
    disableOverlay: !!disableOverlay,
    selected: !!selected,
    displayBackground: shouldDisplayBackground,
    isPlayButtonClickable: isPlayButtonClickable,
    isTickBoxSelectable: isTickBoxSelectable,
    shouldDisplayTooltip: shouldDisplayTooltip
  }, shouldDisplayTooltip ? /*#__PURE__*/_react.default.createElement(_tooltip.default, {
    content: name,
    position: "bottom",
    tag: "div"
  }, contents) : contents));
};
var CardView = exports.CardView = (0, _analyticsNext.withAnalyticsEvents)({
  onClick: (0, _analytics.createAndFireMediaCardEvent)({
    eventType: 'ui',
    action: 'clicked',
    actionSubject: 'mediaCard',
    attributes: {}
  })
})(CardViewBase);
