// @atlaskit/media-card/dist/cjs/card/inlinePlayerWrapper-emotion.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InlinePlayerWrapper = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = require("@emotion/react");
var _inlinePlayerWrapperStyles = require("./inlinePlayerWrapperStyles");
var _vcMedia = require("@atlaskit/react-ufo/vc-media");
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var InlinePlayerWrapper = exports.InlinePlayerWrapper = function InlinePlayerWrapper(props) {
  var testId = props.testId,
    selected = props.selected,
    dimensions = props.dimensions,
    onClick = props.onClick,
    innerRef = props.innerRef;
  return (
    // eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions
    (0, _react.jsx)("div", (0, _extends2.default)({
      id: "inlinePlayerWrapper",
      "data-testid": testId
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
      ,
      className: _inlinePlayerWrapperStyles.inlinePlayerClassName
      // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
      ,
      css: (0, _inlinePlayerWrapperStyles.inlinePlayerWrapperStyles)({
        selected: selected,
        dimensions: dimensions
      }),
      onClick: onClick,
      ref: innerRef
    }, _vcMedia.VcMediaWrapperProps), props.children)
  );
};
