// @atlaskit/media-card/dist/cjs/card/media-card-analytics-error-boundary.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireDefault(require("react"));
var _mediaCommon = require("@atlaskit/media-common");
var _unhandledErrorCard = require("./ui/unhandledErrorCard");
var _analyticsNext = require("@atlaskit/analytics-next");
var _analytics = require("../utils/analytics");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var WrappedMediaCardAnalyticsErrorBoundary = /*#__PURE__*/function (_React$Component) {
  function WrappedMediaCardAnalyticsErrorBoundary(props) {
    var _this;
    (0, _classCallCheck2.default)(this, WrappedMediaCardAnalyticsErrorBoundary);
    _this = _callSuper(this, WrappedMediaCardAnalyticsErrorBoundary, [props]);
    (0, _defineProperty2.default)(_this, "fireOperationalEvent", function (error, info) {
      var _window;
      var _this$props = _this.props,
        _this$props$data = _this$props.data,
        data = _this$props$data === void 0 ? {} : _this$props$data,
        createAnalyticsEvent = _this$props.createAnalyticsEvent;
      var payload = {
        eventType: 'operational',
        action: 'failed',
        actionSubject: 'mediaCardRender',
        attributes: _objectSpread({
          browserInfo: (_window = window) !== null && _window !== void 0 && (_window = _window.navigator) !== null && _window !== void 0 && _window.userAgent ? window.navigator.userAgent : 'unknown',
          error: error,
          info: info,
          failReason: 'unexpected-error'
        }, data)
      };
      (0, _analytics.fireMediaCardEvent)(payload, createAnalyticsEvent);
    });
    (0, _defineProperty2.default)(_this, "handleOnClick", function (event) {
      try {
        var _this$props$onClick, _this$props2;
        (_this$props$onClick = (_this$props2 = _this.props).onClick) === null || _this$props$onClick === void 0 || _this$props$onClick.call(_this$props2, {
          event: event
        });
      } catch (e) {}
    });
    _this.state = {
      hasError: false
    };
    return _this;
  }
  (0, _inherits2.default)(WrappedMediaCardAnalyticsErrorBoundary, _React$Component);
  return (0, _createClass2.default)(WrappedMediaCardAnalyticsErrorBoundary, [{
    key: "componentDidCatch",
    value: function componentDidCatch(error, info) {
      try {
        this.fireOperationalEvent(error, info);
      } catch (e) {}
      this.setState({
        hasError: true
      });
    }
  }, {
    key: "render",
    value: function render() {
      var hasError = this.state.hasError;
      var _this$props3 = this.props,
        dimensions = _this$props3.dimensions,
        children = _this$props3.children;
      if (hasError) {
        return /*#__PURE__*/_react.default.createElement(_unhandledErrorCard.UnhandledErrorCard, {
          dimensions: dimensions,
          onClick: this.handleOnClick
        });
      }
      return children;
    }
  }]);
}(_react.default.Component);
(0, _defineProperty2.default)(WrappedMediaCardAnalyticsErrorBoundary, "displayName", 'MediaCardAnalyticsErrorBoundary');
var packageName = "@atlaskit/media-card";
var packageVersion = "79.2.0";

// @ts-ignore: [PIT-1685] Fails in post-office due to backwards incompatibility issue with React 18
var MediaCardAnalyticsErrorBoundary = (0, _mediaCommon.withMediaAnalyticsContext)({
  packageVersion: packageVersion,
  packageName: packageName,
  componentName: 'mediaCard',
  component: 'mediaCard'
})((0, _analyticsNext.withAnalyticsEvents)()(WrappedMediaCardAnalyticsErrorBoundary));
var _default = exports.default = MediaCardAnalyticsErrorBoundary;
