// @atlaskit/editor-wikimarkup-transformer/dist/cjs/encoder/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.encode = encode;
var _blockquote = require("./nodes/blockquote");
var _bulletList = require("./nodes/bullet-list");
var _taskList = require("./nodes/taskList");
var _decisionList = require("./nodes/decisionList");
var _codeBlock = require("./nodes/code-block");
var _doc = require("./nodes/doc");
var _expand = require("./nodes/expand");
var _heading = require("./nodes/heading");
var _mediaGroup = require("./nodes/media-group");
var _mediaInline = require("./nodes/media-inline");
var _orderedList = require("./nodes/ordered-list");
var _panel = require("./nodes/panel");
var _paragraph = require("./nodes/paragraph");
var _rule = require("./nodes/rule");
var _table = require("./nodes/table");
var _unknown = require("./nodes/unknown");
var _blockCard = require("./nodes/block-card");
var _embedCard = require("./nodes/embed-card");
var _caption = require("./nodes/caption");
// Ignored via go/ees005
// eslint-disable-next-line @typescript-eslint/no-explicit-any

var nodeEncoderMapping = {
  blockquote: _blockquote.blockquote,
  bulletList: _bulletList.bulletList,
  taskList: _taskList.taskList,
  decisionList: _decisionList.decisionList,
  codeBlock: _codeBlock.codeBlock,
  doc: _doc.doc,
  heading: _heading.heading,
  caption: _caption.caption,
  mediaGroup: _mediaGroup.mediaGroup,
  mediaInline: _mediaInline.mediaInline,
  mediaSingle: _mediaGroup.mediaGroup,
  orderedList: _orderedList.orderedList,
  panel: _panel.panel,
  paragraph: _paragraph.paragraph,
  rule: _rule.rule,
  table: _table.table,
  blockCard: _blockCard.blockCard,
  embedCard: _embedCard.embedCard,
  expand: _expand.expand
};
function encode(node, context) {
  var encoder = nodeEncoderMapping[node.type.name];
  try {
    if (encoder) {
      return encoder(node, {
        context: context
      });
    }
    return (0, _unknown.unknown)(node);
  } catch (err) {
    return (0, _unknown.unknown)(node);
  }
}
