// @atlaskit/media-ui/dist/cjs/MediaInlineCard/Frame/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Frame = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireDefault(require("react"));
var _styled = require("./styled");
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var Frame = exports.Frame = /*#__PURE__*/function (_React$Component) {
  function Frame() {
    var _this;
    (0, _classCallCheck2.default)(this, Frame);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, Frame, [].concat(args));
    (0, _defineProperty2.default)(_this, "handleClick", function (event) {
      var onClick = _this.props.onClick;
      if (onClick) {
        event.preventDefault();
        event.stopPropagation();
        onClick(event);
      }
    });
    (0, _defineProperty2.default)(_this, "handleKeyPress", function (event) {
      if (event.key !== ' ' && event.key !== 'Enter') {
        return;
      }
      var onClick = _this.props.onClick;
      if (onClick) {
        event.preventDefault();
        event.stopPropagation();
        onClick(event);
      }
    });
    return _this;
  }
  (0, _inherits2.default)(Frame, _React$Component);
  return (0, _createClass2.default)(Frame, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
        isSelected = _this$props.isSelected,
        children = _this$props.children,
        onClick = _this$props.onClick,
        innerRef = _this$props.innerRef,
        testId = _this$props.testId;
      var isInteractive = Boolean(onClick);
      return /*#__PURE__*/_react.default.createElement(_styled.Wrapper, {
        ref: innerRef,
        isSelected: isSelected,
        tabIndex: isInteractive ? 0 : undefined,
        role: isInteractive ? 'button' : undefined,
        onClick: this.handleClick,
        onKeyPress: this.handleKeyPress,
        "data-testid": testId
      }, children);
    }
  }]);
}(_react.default.Component);
