// @atlaskit/people-teams-ui-public/dist/esm/common/utils/analytics/analytics-subcontext/store.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
import { useEffect } from 'react';
import { createHook, createStore } from 'react-sweet-state';
var initialState = {
  eventAttributes: {}
};
var actions = {
  setEventAttributes: function setEventAttributes(eventAttributes) {
    return function (_ref) {
      var setState = _ref.setState,
        getState = _ref.getState;
      var existingEventAttributes = getState().eventAttributes;
      var newEventAttributes = _objectSpread(_objectSpread({}, existingEventAttributes), eventAttributes);
      Object.keys(newEventAttributes).forEach(function (keyString) {
        var key = keyString;
        if (newEventAttributes[key] === undefined) {
          delete newEventAttributes[key];
        }
      });
      setState({
        eventAttributes: newEventAttributes
      });
    };
  }
};
var AnalyticsSubcontextStore = createStore({
  initialState: initialState,
  actions: actions
});
export var usePeopleTeamsAnalyticsSubcontext = createHook(AnalyticsSubcontextStore);

/**
 * Inject an attribute into all events using `usePeopleTeamAnalyticsEvents`
 * It sets the attribute when the hook is called and clears it when the component is unmounted
 */
export var useInjectedEventAttribute = function useInjectedEventAttribute(key, value) {
  var setEventAttributes = usePeopleTeamsAnalyticsSubcontext()[1].setEventAttributes;
  useEffect(function () {
    setEventAttributes(_defineProperty({}, key, value));
    return function () {
      return setEventAttributes(_defineProperty({}, key, undefined));
    };
  }, [key, value, setEventAttributes]);
};
