// @atlaskit/smart-card/dist/cjs/view/FlexibleCard/components/actions/delete-action/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _reactIntlNext = require("react-intl-next");
var _closeCross = _interopRequireDefault(require("@atlaskit/icon/core/migration/close--cross"));
var _messages = require("../../../../../messages");
var _action = _interopRequireDefault(require("../action"));
var DeleteAction = function DeleteAction(props) {
  return /*#__PURE__*/_react.default.createElement(_action.default, (0, _extends2.default)({
    content: /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, _messages.messages.delete),
    icon: /*#__PURE__*/_react.default.createElement(_closeCross.default, {
      color: "currentColor",
      spacing: "spacious",
      label: "Delete"
    }),
    testId: "smart-action-delete-action",
    tooltipMessage: /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, _messages.messages.delete)
  }, props));
};
var _default = exports.default = DeleteAction;
