// @atlaskit/editor-common/dist/cjs/ui/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Announcer", {
  enumerable: true,
  get: function get() {
    return _announcer.default;
  }
});
Object.defineProperty(exports, "AssistiveText", {
  enumerable: true,
  get: function get() {
    return _AssistiveText.AssistiveText;
  }
});
Object.defineProperty(exports, "BaseTheme", {
  enumerable: true,
  get: function get() {
    return _BaseTheme.BaseTheme;
  }
});
Object.defineProperty(exports, "Caption", {
  enumerable: true,
  get: function get() {
    return _Caption.default;
  }
});
Object.defineProperty(exports, "ContextPanel", {
  enumerable: true,
  get: function get() {
    return _context.ContextPanel;
  }
});
Object.defineProperty(exports, "ContextPanelConsumer", {
  enumerable: true,
  get: function get() {
    return _context.ContextPanelConsumer;
  }
});
Object.defineProperty(exports, "ContextPanelProvider", {
  enumerable: true,
  get: function get() {
    return _context.ContextPanelProvider;
  }
});
Object.defineProperty(exports, "ContextPanelWidthProvider", {
  enumerable: true,
  get: function get() {
    return _context.ContextPanelWidthProvider;
  }
});
Object.defineProperty(exports, "DropList", {
  enumerable: true,
  get: function get() {
    return _DropList.default;
  }
});
Object.defineProperty(exports, "EDIT_AREA_ID", {
  enumerable: true,
  get: function get() {
    return _Toolbar.EDIT_AREA_ID;
  }
});
Object.defineProperty(exports, "ErrorMessage", {
  enumerable: true,
  get: function get() {
    return _Messages.ErrorMessage;
  }
});
Object.defineProperty(exports, "ExpandIconWrapper", {
  enumerable: true,
  get: function get() {
    return _Expand.ExpandIconWrapper;
  }
});
Object.defineProperty(exports, "ExpandLayoutWrapperWithRef", {
  enumerable: true,
  get: function get() {
    return _Expand.ExpandLayoutWrapperWithRef;
  }
});
Object.defineProperty(exports, "FloatingToolbarButton", {
  enumerable: true,
  get: function get() {
    return _Button.default;
  }
});
Object.defineProperty(exports, "FloatingToolbarSeparator", {
  enumerable: true,
  get: function get() {
    return _Separator.default;
  }
});
Object.defineProperty(exports, "HelperMessage", {
  enumerable: true,
  get: function get() {
    return _Messages.HelperMessage;
  }
});
Object.defineProperty(exports, "IntlErrorBoundary", {
  enumerable: true,
  get: function get() {
    return _IntlErrorBoundary.IntlErrorBoundary;
  }
});
Object.defineProperty(exports, "IntlProviderIfMissingWrapper", {
  enumerable: true,
  get: function get() {
    return _IntlProviderIfMissingWrapper.default;
  }
});
Object.defineProperty(exports, "LINKPICKER_HEIGHT_IN_PX", {
  enumerable: true,
  get: function get() {
    return _const.LINKPICKER_HEIGHT_IN_PX;
  }
});
Object.defineProperty(exports, "MediaBorderGapFiller", {
  enumerable: true,
  get: function get() {
    return _styled.MediaBorderGapFiller;
  }
});
Object.defineProperty(exports, "MediaSingle", {
  enumerable: true,
  get: function get() {
    return _MediaSingle.default;
  }
});
Object.defineProperty(exports, "MediaSingleDimensionHelper", {
  enumerable: true,
  get: function get() {
    return _styled.MediaSingleDimensionHelper;
  }
});
Object.defineProperty(exports, "PanelTextInput", {
  enumerable: true,
  get: function get() {
    return _PanelTextInput.default;
  }
});
Object.defineProperty(exports, "PlainOutsideClickTargetRefContext", {
  enumerable: true,
  get: function get() {
    return _withOuterListeners.PlainOutsideClickTargetRefContext;
  }
});
Object.defineProperty(exports, "Popup", {
  enumerable: true,
  get: function get() {
    return _Popup.default;
  }
});
Object.defineProperty(exports, "REACT_INTL_ERROR_MESSAGE", {
  enumerable: true,
  get: function get() {
    return _IntlErrorBoundary.REACT_INTL_ERROR_MESSAGE;
  }
});
Object.defineProperty(exports, "RECENT_SEARCH_HEIGHT_IN_PX", {
  enumerable: true,
  get: function get() {
    return _const.RECENT_SEARCH_HEIGHT_IN_PX;
  }
});
Object.defineProperty(exports, "RECENT_SEARCH_WIDTH_IN_PX", {
  enumerable: true,
  get: function get() {
    return _const.RECENT_SEARCH_WIDTH_IN_PX;
  }
});
Object.defineProperty(exports, "RECENT_SEARCH_WIDTH_WITHOUT_ITEMS_IN_PX", {
  enumerable: true,
  get: function get() {
    return _const.RECENT_SEARCH_WIDTH_WITHOUT_ITEMS_IN_PX;
  }
});
Object.defineProperty(exports, "Resizer", {
  enumerable: true,
  get: function get() {
    return _ResizerLegacy.default;
  }
});
Object.defineProperty(exports, "ShadowObserver", {
  enumerable: true,
  get: function get() {
    return _shadowObserver.ShadowObserver;
  }
});
Object.defineProperty(exports, "Shortcut", {
  enumerable: true,
  get: function get() {
    return _Shortcut.Shortcut;
  }
});
Object.defineProperty(exports, "SmallerEditIcon", {
  enumerable: true,
  get: function get() {
    return _SmallerEditIcon.SmallerEditIcon;
  }
});
Object.defineProperty(exports, "TableSelectorPopup", {
  enumerable: true,
  get: function get() {
    return _TableSelector.default;
  }
});
Object.defineProperty(exports, "ToolbarButtonGroup", {
  enumerable: true,
  get: function get() {
    return _ButtonGroup.ToolbarButtonGroup;
  }
});
Object.defineProperty(exports, "ToolbarDropdownTriggerWrapper", {
  enumerable: true,
  get: function get() {
    return _DropdownTriggerWrapper.ToolbarDropdownTriggerWrapper;
  }
});
Object.defineProperty(exports, "ToolbarDropdownWrapper", {
  enumerable: true,
  get: function get() {
    return _DropdownWrapper.ToolbarDropdownWrapper;
  }
});
Object.defineProperty(exports, "ToolbarExpandIcon", {
  enumerable: true,
  get: function get() {
    return _ExpandIcon.ToolbarExpandIcon;
  }
});
Object.defineProperty(exports, "ToolbarSeparator", {
  enumerable: true,
  get: function get() {
    return _Separator2.ToolbarSeparator;
  }
});
Object.defineProperty(exports, "UnsupportedBlock", {
  enumerable: true,
  get: function get() {
    return _UnsupportedBlock.default;
  }
});
Object.defineProperty(exports, "UnsupportedInline", {
  enumerable: true,
  get: function get() {
    return _UnsupportedInline.default;
  }
});
Object.defineProperty(exports, "ValidMessage", {
  enumerable: true,
  get: function get() {
    return _Messages.ValidMessage;
  }
});
Object.defineProperty(exports, "WidthConsumer", {
  enumerable: true,
  get: function get() {
    return _WidthProvider.WidthConsumer;
  }
});
Object.defineProperty(exports, "WidthContext", {
  enumerable: true,
  get: function get() {
    return _WidthProvider.WidthContext;
  }
});
Object.defineProperty(exports, "WidthProvider", {
  enumerable: true,
  get: function get() {
    return _WidthProvider.WidthProvider;
  }
});
Object.defineProperty(exports, "WithCreateAnalyticsEvent", {
  enumerable: true,
  get: function get() {
    return _WithCreateAnalyticsEvent.WithCreateAnalyticsEvent;
  }
});
Object.defineProperty(exports, "calcColumnsFromPx", {
  enumerable: true,
  get: function get() {
    return _grid.calcColumnsFromPx;
  }
});
Object.defineProperty(exports, "calcMediaPxWidth", {
  enumerable: true,
  get: function get() {
    return _grid.calcMediaPxWidth;
  }
});
Object.defineProperty(exports, "calcPctFromPx", {
  enumerable: true,
  get: function get() {
    return _grid.calcPctFromPx;
  }
});
Object.defineProperty(exports, "calcPxFromColumns", {
  enumerable: true,
  get: function get() {
    return _grid.calcPxFromColumns;
  }
});
Object.defineProperty(exports, "calcPxFromPct", {
  enumerable: true,
  get: function get() {
    return _grid.calcPxFromPct;
  }
});
Object.defineProperty(exports, "clearNextSiblingBlockMarkMarginTopStyle", {
  enumerable: true,
  get: function get() {
    return _clearNextSiblingMarginTop.clearNextSiblingBlockMarkMarginTopStyle;
  }
});
Object.defineProperty(exports, "clearNextSiblingMarginTopStyle", {
  enumerable: true,
  get: function get() {
    return _clearNextSiblingMarginTop.clearNextSiblingMarginTopStyle;
  }
});
Object.defineProperty(exports, "createWidthContext", {
  enumerable: true,
  get: function get() {
    return _WidthProvider.createWidthContext;
  }
});
Object.defineProperty(exports, "expandLayoutWrapperStyle", {
  enumerable: true,
  get: function get() {
    return _Expand.expandLayoutWrapperStyle;
  }
});
Object.defineProperty(exports, "expandMessages", {
  enumerable: true,
  get: function get() {
    return _Expand.messages;
  }
});
Object.defineProperty(exports, "findOverflowScrollParent", {
  enumerable: true,
  get: function get() {
    return _utils.findOverflowScrollParent;
  }
});
Object.defineProperty(exports, "getBreakpoint", {
  enumerable: true,
  get: function get() {
    return _WidthProvider.getBreakpoint;
  }
});
Object.defineProperty(exports, "handleSides", {
  enumerable: true,
  get: function get() {
    return _utils2.handleSides;
  }
});
Object.defineProperty(exports, "imageAlignmentMap", {
  enumerable: true,
  get: function get() {
    return _utils2.imageAlignmentMap;
  }
});
Object.defineProperty(exports, "layoutSupportsWidth", {
  enumerable: true,
  get: function get() {
    return _grid.layoutSupportsWidth;
  }
});
Object.defineProperty(exports, "mapBreakpointToLayoutMaxWidth", {
  enumerable: true,
  get: function get() {
    return _BaseTheme.mapBreakpointToLayoutMaxWidth;
  }
});
Object.defineProperty(exports, "mediaLinkStyle", {
  enumerable: true,
  get: function get() {
    return _link.mediaLinkStyle;
  }
});
Object.defineProperty(exports, "overflowShadow", {
  enumerable: true,
  get: function get() {
    return _OverflowShadow.default;
  }
});
Object.defineProperty(exports, "panelTextInput", {
  enumerable: true,
  get: function get() {
    return _styles.panelTextInput;
  }
});
Object.defineProperty(exports, "shadowClassNames", {
  enumerable: true,
  get: function get() {
    return _OverflowShadow.shadowClassNames;
  }
});
Object.defineProperty(exports, "shadowObserverClassNames", {
  enumerable: true,
  get: function get() {
    return _shadowObserver.shadowObserverClassNames;
  }
});
Object.defineProperty(exports, "sharedExpandStyles", {
  enumerable: true,
  get: function get() {
    return _sharedStyles.sharedExpandStyles;
  }
});
Object.defineProperty(exports, "sharedMultiBodiedExtensionStyles", {
  enumerable: true,
  get: function get() {
    return _MultiBodiedExtension.sharedMultiBodiedExtensionStyles;
  }
});
Object.defineProperty(exports, "snapTo", {
  enumerable: true,
  get: function get() {
    return _utils2.snapTo;
  }
});
Object.defineProperty(exports, "snapToGrid", {
  enumerable: true,
  get: function get() {
    return _grid.snapToGrid;
  }
});
Object.defineProperty(exports, "withOuterListeners", {
  enumerable: true,
  get: function get() {
    return _withOuterListeners.default;
  }
});
Object.defineProperty(exports, "wrappedLayouts", {
  enumerable: true,
  get: function get() {
    return _grid.wrappedLayouts;
  }
});
Object.defineProperty(exports, "wrapperStyle", {
  enumerable: true,
  get: function get() {
    return _styled2.wrapperStyle;
  }
});
var _AssistiveText = require("./AssistiveText/AssistiveText");
var _Caption = _interopRequireDefault(require("./Caption"));
var _MediaSingle = _interopRequireDefault(require("./MediaSingle"));
var _styled = require("./MediaSingle/styled");
var _grid = require("./MediaSingle/grid");
var _link = require("./MediaSingle/link");
var _Popup = _interopRequireDefault(require("./Popup"));
var _utils = require("./Popup/utils");
var _UnsupportedBlock = _interopRequireDefault(require("./UnsupportedBlock"));
var _UnsupportedInline = _interopRequireDefault(require("./UnsupportedInline"));
var _BaseTheme = require("./BaseTheme");
var _withOuterListeners = _interopRequireWildcard(require("./with-outer-listeners"));
var _WidthProvider = require("./WidthProvider");
var _OverflowShadow = _interopRequireWildcard(require("./OverflowShadow"));
var _shadowObserver = require("./OverflowShadow/shadowObserver");
var _WithCreateAnalyticsEvent = require("./WithCreateAnalyticsEvent");
var _Expand = require("./Expand");
var _sharedStyles = require("./Expand/sharedStyles");
var _Messages = require("./Messages");
var _clearNextSiblingMarginTop = require("./clear-next-sibling-margin-top");
var _IntlErrorBoundary = require("./IntlErrorBoundary");
var _IntlProviderIfMissingWrapper = _interopRequireDefault(require("./IntlProviderIfMissingWrapper"));
var _Button = _interopRequireDefault(require("./FloatingToolbar/Button"));
var _Separator = _interopRequireDefault(require("./FloatingToolbar/Separator"));
var _SmallerEditIcon = require("./FloatingToolbar/SmallerEditIcon");
var _const = require("./LinkSearch/const");
var _context = require("./ContextPanel/context");
var _ResizerLegacy = _interopRequireDefault(require("./ResizerLegacy"));
var _utils2 = require("./ResizerLegacy/utils");
var _styled2 = require("./ResizerLegacy/styled");
var _styles = require("./PanelTextInput/styles");
var _PanelTextInput = _interopRequireDefault(require("./PanelTextInput"));
var _announcer = _interopRequireDefault(require("./Announcer/announcer"));
var _Shortcut = require("./Shortcut");
var _Toolbar = require("./Toolbar");
var _ButtonGroup = require("./Toolbar/ButtonGroup");
var _DropdownTriggerWrapper = require("./Toolbar/DropdownTriggerWrapper");
var _DropdownWrapper = require("./Toolbar/DropdownWrapper");
var _ExpandIcon = require("./Toolbar/ExpandIcon");
var _Separator2 = require("./Toolbar/Separator");
var _DropList = _interopRequireDefault(require("./DropList"));
var _MultiBodiedExtension = require("./MultiBodiedExtension");
var _TableSelector = _interopRequireDefault(require("./TableSelector"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
