// @atlaskit/editor-common/dist/es2019/utils/index.js
export { getExtensionLozengeData } from './macro';
export { getMarksByOrder, getValidContent, getValidDocument, getValidMark, getValidNode, getValidUnknownNode, isSameMark, isSubSupType, markOrder, } from './validator';
export { default as browser } from './browser';
export { default as ErrorReporter, } from './error-reporter';
export { isPastDate, timestampToIsoFormat, timestampToString, timestampToTaskContext, timestampToUTCDate, todayTimestampInUTC, } from './date';
export { withImageLoader, } from './imageLoader';
export { absoluteBreakoutWidth, calcBreakoutWidth, calcWideWidth, breakoutConsts, } from './breakout';
export { default as ADFTraversor } from './traversor';
export { getAnalyticsAppearance } from './analytics';
export { measureRender } from './performance/measure-render';
export { startMeasure, stopMeasure, clearMeasure } from './performance/measure';
export { isPerformanceAPIAvailable, isPerformanceObserverAvailable, } from './performance/is-performance-api-available';
export { getResponseEndTime } from './performance/navigation';
export { getExtensionRenderer } from './extension-handler';
export { hasMergedCell, calcTableColumnWidths, convertProsemirrorTableNodeToArrayOfRows, } from './table';
export { createCompareNodes } from './compareNodes';
export { compose } from './compose';
export { ZERO_WIDTH_SPACE } from './whitespace';
