// @atlaskit/profilecard/dist/cjs/mocks/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "getMockProfileClient", {
  enumerable: true,
  get: function get() {
    return _mockProfileClient.default;
  }
});
Object.defineProperty(exports, "getMockTeamClient", {
  enumerable: true,
  get: function get() {
    return _mockTeamClient.default;
  }
});
Object.defineProperty(exports, "profiles", {
  enumerable: true,
  get: function get() {
    return _profileData.default;
  }
});
Object.defineProperty(exports, "simpleMockAgentClient", {
  enumerable: true,
  get: function get() {
    return _simpleMockClients.simpleMockAgentClient;
  }
});
Object.defineProperty(exports, "simpleMockTeamClient", {
  enumerable: true,
  get: function get() {
    return _simpleMockClients.simpleMockTeamClient;
  }
});
Object.defineProperty(exports, "simpleMockUserClient", {
  enumerable: true,
  get: function get() {
    return _simpleMockClients.simpleMockUserClient;
  }
});
Object.defineProperty(exports, "simpleProfileClient", {
  enumerable: true,
  get: function get() {
    return _simpleMockClients.simpleProfileClient;
  }
});
var _mockProfileClient = _interopRequireDefault(require("./mock-profile-client"));
var _mockTeamClient = _interopRequireDefault(require("./mock-team-client"));
var _profileData = _interopRequireDefault(require("./profile-data"));
var _simpleMockClients = require("./simple-mock-clients");
