// @atlaskit/profilecard/dist/cjs/mocks/mock-team-client.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = getMockTeamClient;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _errorUtils = require("../client/errorUtils");
var _TeamProfileCardClient = _interopRequireDefault(require("../client/TeamProfileCardClient"));
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function getMockTeamClient(data) {
  return /*#__PURE__*/function (_TeamProfileCardClien) {
    function MockTeamClient() {
      (0, _classCallCheck2.default)(this, MockTeamClient);
      return _callSuper(this, MockTeamClient, arguments);
    }
    (0, _inherits2.default)(MockTeamClient, _TeamProfileCardClien);
    return (0, _createClass2.default)(MockTeamClient, [{
      key: "makeRequest",
      value: function makeRequest(teamId) {
        var _data$error, _data$error2, _data$error3, _data$error4, _data$error5;
        var errorResponse = {
          reason: (_data$error = data.error) === null || _data$error === void 0 ? void 0 : _data$error.reason,
          code: (_data$error2 = data.error) === null || _data$error2 === void 0 ? void 0 : _data$error2.code,
          source: (_data$error3 = data.error) === null || _data$error3 === void 0 ? void 0 : _data$error3.source,
          message: (_data$error4 = data.error) === null || _data$error4 === void 0 ? void 0 : _data$error4.message,
          traceId: (_data$error5 = data.error) === null || _data$error5 === void 0 ? void 0 : _data$error5.traceId
        };
        if (!data.timeout) {
          if (data.error && Math.random() < data.errorRate) {
            try {
              (0, _errorUtils.handleAGGErrors)(data.error, data.traceId);
            } catch (e) {
              return Promise.reject(e);
            }
          }
          return Promise.resolve(data.team);
        }
        return new Promise(function (resolve, reject) {
          setTimeout(function () {
            if (data.error && Math.random() < data.errorRate) {
              reject(errorResponse);
            }
            return resolve(data.team);
          }, data.timeout);
        });
      }
    }]);
  }(_TeamProfileCardClient.default);
}
