// @atlaskit/profilecard/dist/cjs/mocks/mock-profile-client.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = getMockProfileClient;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _TeamCentralCardClient = _interopRequireDefault(require("../client/TeamCentralCardClient"));
var _UserProfileCardClient = _interopRequireDefault(require("../client/UserProfileCardClient"));
var _profileData = _interopRequireDefault(require("./profile-data"));
var _reportingLinesData = require("./reporting-lines-data");
var _util = require("./util");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function getMockProfileClient(BaseProfileClient, modifyResponse) {
  var MockUserClient = /*#__PURE__*/function (_UserProfileCardClien) {
    function MockUserClient() {
      (0, _classCallCheck2.default)(this, MockUserClient);
      return _callSuper(this, MockUserClient, arguments);
    }
    (0, _inherits2.default)(MockUserClient, _UserProfileCardClien);
    return (0, _createClass2.default)(MockUserClient, [{
      key: "makeRequest",
      value: function makeRequest(cloudId, userId) {
        var timeout = (0, _util.random)(1500) + 500;
        var matchError = userId.match(/^error:([0-9a-zA-Z\-]+)$/);
        var error = matchError && matchError[1];
        return new Promise(function (resolve, reject) {
          setTimeout(function () {
            if (error) {
              return reject({
                reason: error
              });
            }
            var userIdAsIndex = parseInt(userId, 10);
            var profile = _profileData.default[userIdAsIndex] || _profileData.default[0];
            if (!profile) {
              return reject({
                reason: 'default'
              });
            }
            var weekday = (0, _util.getWeekday)();
            var data = _objectSpread({}, profile);
            data.remoteTimeString = (0, _util.getTimeString)();
            data.remoteWeekdayIndex = weekday.index;
            data.remoteWeekdayString = weekday.string;
            return resolve(modifyResponse(data));
          }, timeout);
        });
      }
    }]);
  }(_UserProfileCardClient.default);
  var MockTeamCentralClient = /*#__PURE__*/function (_TeamCentralCardClien) {
    function MockTeamCentralClient() {
      (0, _classCallCheck2.default)(this, MockTeamCentralClient);
      return _callSuper(this, MockTeamCentralClient, arguments);
    }
    (0, _inherits2.default)(MockTeamCentralClient, _TeamCentralCardClien);
    return (0, _createClass2.default)(MockTeamCentralClient, [{
      key: "createOrgContainsAnyWorkspacePromise",
      value: function createOrgContainsAnyWorkspacePromise(config) {
        return Promise.resolve(true);
      }
    }, {
      key: "makeRequest",
      value: function makeRequest(userId) {
        var timeout = (0, _util.random)(1500) + 500;
        var matchError = userId.match(/^error:([0-9a-zA-Z\-]+)$/);
        var error = matchError && matchError[1];
        return new Promise(function (resolve, reject) {
          setTimeout(function () {
            if (error) {
              return reject({
                reason: error
              });
            }
            return resolve(_reportingLinesData.reportingLinesData);
          }, timeout);
        });
      }
    }]);
  }(_TeamCentralCardClient.default);
  return /*#__PURE__*/function (_BaseProfileClient) {
    function MockProfileClient(options) {
      var clients = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      (0, _classCallCheck2.default)(this, MockProfileClient);
      return _callSuper(this, MockProfileClient, [options, _objectSpread({
        userClient: new MockUserClient(options),
        teamCentralClient: new MockTeamCentralClient(_objectSpread(_objectSpread({}, options), {}, {
          gatewayGraphqlUrl: 'defaultGatewayGraphqlUrl'
        }))
      }, clients)]);
    }
    (0, _inherits2.default)(MockProfileClient, _BaseProfileClient);
    return (0, _createClass2.default)(MockProfileClient);
  }(BaseProfileClient);
}
