// @atlaskit/profilecard/dist/cjs/mocks/simple-mock-clients.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.simpleProfileClient = exports.simpleMockUserClient = exports.simpleMockTeamClient = exports.simpleMockAgentClient = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _ProfileCardClient = _interopRequireDefault(require("../client/ProfileCardClient"));
var _RovoAgentCardClient2 = _interopRequireDefault(require("../client/RovoAgentCardClient"));
var _TeamProfileCardClient = _interopRequireDefault(require("../client/TeamProfileCardClient"));
var _UserProfileCardClient = _interopRequireDefault(require("../client/UserProfileCardClient"));
var _agentData = _interopRequireDefault(require("./agent-data"));
var _profileData = _interopRequireDefault(require("./profile-data"));
var _teamData = _interopRequireDefault(require("./team-data"));
var _util = require("./util");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var SimpleMockTeamClient = /*#__PURE__*/function (_TeamProfileCardClien) {
  function SimpleMockTeamClient() {
    (0, _classCallCheck2.default)(this, SimpleMockTeamClient);
    return _callSuper(this, SimpleMockTeamClient, arguments);
  }
  (0, _inherits2.default)(SimpleMockTeamClient, _TeamProfileCardClien);
  return (0, _createClass2.default)(SimpleMockTeamClient, [{
    key: "makeRequest",
    value: function makeRequest(teamId) {
      var simpleMockTeam = (0, _teamData.default)({
        members: 5
      });
      return Promise.resolve(_objectSpread({
        teamId: teamId
      }, simpleMockTeam));
    }
  }]);
}(_TeamProfileCardClient.default);
var SimpleMockUserClient = /*#__PURE__*/function (_UserProfileCardClien) {
  function SimpleMockUserClient() {
    (0, _classCallCheck2.default)(this, SimpleMockUserClient);
    return _callSuper(this, SimpleMockUserClient, arguments);
  }
  (0, _inherits2.default)(SimpleMockUserClient, _UserProfileCardClien);
  return (0, _createClass2.default)(SimpleMockUserClient, [{
    key: "makeRequest",
    value: function makeRequest(cloudId, userId) {
      var profile = _profileData.default[0];
      var weekday = (0, _util.getWeekday)();
      var data = _objectSpread({}, profile);
      data.remoteTimeString = (0, _util.getTimeString)();
      data.remoteWeekdayIndex = weekday.index;
      data.remoteWeekdayString = weekday.string;
      return Promise.resolve(data);
    }
  }]);
}(_UserProfileCardClient.default);
var SimpleMockAgentClient = /*#__PURE__*/function (_RovoAgentCardClient) {
  function SimpleMockAgentClient() {
    (0, _classCallCheck2.default)(this, SimpleMockAgentClient);
    return _callSuper(this, SimpleMockAgentClient, arguments);
  }
  (0, _inherits2.default)(SimpleMockAgentClient, _RovoAgentCardClient);
  return (0, _createClass2.default)(SimpleMockAgentClient, [{
    key: "makeRequest",
    value: function makeRequest(id, cloudId) {
      return Promise.resolve(_agentData.default);
    }
  }]);
}(_RovoAgentCardClient2.default);
var args = {
  cacheSize: 10,
  maxCacheAge: 0,
  url: '/graphql/directory',
  gatewayGraphqlUrl: '/gateway/api/graphql'
};
var simpleMockUserClient = exports.simpleMockUserClient = new SimpleMockUserClient(args);
var simpleMockTeamClient = exports.simpleMockTeamClient = new SimpleMockTeamClient(args);
var simpleMockAgentClient = exports.simpleMockAgentClient = new SimpleMockAgentClient(args);
var simpleProfileClient = exports.simpleProfileClient = new _ProfileCardClient.default(args, {
  userClient: simpleMockUserClient,
  teamClient: simpleMockTeamClient,
  rovoAgentClient: simpleMockAgentClient
});
