// @atlaskit/media-card/dist/es2019/card/ui/titleBox/titleBoxComponents-emotion.js
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import { jsx } from '@emotion/react';
import { fg } from '@atlaskit/platform-feature-flags';
import { errorMessageWrapperStyles, newTitleBoxIconStyles, titleBoxFooterStyles, titleBoxHeaderStyles, titleBoxIconStyles, titleBoxWrapperStyles } from './styles';
export const TitleBoxWrapper = props => {
  const {
    breakpoint,
    titleBoxBgColor,
    hidden
  } = props;
  return jsx("div", {
    id: "titleBoxWrapper",
    "data-testid": "media-title-box"
    // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
    ,
    css: titleBoxWrapperStyles({
      breakpoint: breakpoint,
      display: hidden ? 'none' : 'flex',
      titleBoxBgColor: titleBoxBgColor
    })
  }, props.children);
};
export const TitleBoxHeader = props => {
  const {
    hasIconOverlap
  } = props;
  return jsx("div", {
    id: "titleBoxHeader",
    "data-testid": "title-box-header"
    // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
    ,
    css: titleBoxHeaderStyles({
      hasIconOverlap
    })
  }, props.children);
};
export const TitleBoxFooter = props => {
  const {
    hasIconOverlap
  } = props;
  return jsx("div", {
    id: "titleBoxFooter",
    "data-testid": "title-box-footer"
    // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
    ,
    css: titleBoxFooterStyles({
      hasIconOverlap
    })
  }, props.children);
};
export const TitleBoxIcon = props => {
  return jsx("div", {
    id: "titleBoxIcon",
    "data-testid": "title-box-icon",
    css: fg('platform-visual-refresh-icons') ?
    // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
    newTitleBoxIconStyles :
    // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
    titleBoxIconStyles
  }, props.children);
};
export const ErrorMessageWrapper = props => {
  // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
  return jsx("div", {
    css: errorMessageWrapperStyles
  }, props.children);
};
