// @atlaskit/user-picker/dist/cjs/components/AddOptionAvatar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AddOptionAvatar = void 0;
var _email = _interopRequireDefault(require("@atlaskit/icon/core/migration/email"));
var _colors = require("@atlaskit/theme/colors");
var _react = require("@emotion/react");
var _react2 = _interopRequireDefault(require("react"));
/**
 * @jsxRuntime classic
 * @jsx jsx
 */

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var getEmailAvatarWrapperStyle = function getEmailAvatarWrapperStyle(isLozenge) {
  var padding = isLozenge ? "var(--ds-space-0, 0px)" : "var(--ds-space-050, 4px)";
  return (0, _react.css)({
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    padding: padding,
    backgroundColor: "var(--ds-background-neutral, ".concat(_colors.N40, ")"),
    borderRadius: '50%',
    display: 'flex',
    alignItems: 'center',
    marginLeft: "var(--ds-space-025, 2px)"
  });
};
var AddOptionAvatar = exports.AddOptionAvatar = function AddOptionAvatar(_ref) {
  var isLozenge = _ref.isLozenge,
    label = _ref.label;
  return (
    // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage -- Ignored via go/DSP-18766
    (0, _react.jsx)("div", {
      css: getEmailAvatarWrapperStyle(isLozenge)
    }, (0, _react.jsx)(_email.default, {
      testId: "add-option-avatar-email-icon",
      label: label || '',
      LEGACY_size: isLozenge ? 'small' : 'medium',
      LEGACY_margin: "0 0 0 -2px",
      color: "var(--ds-text-subtle, ".concat(_colors.N500, ")")
    }))
  );
};
