// @atlaskit/user-picker/dist/cjs/components/components.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getPopupComponents = exports.getComponents = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _memoizeOne = _interopRequireDefault(require("memoize-one"));
var _ClearIndicator = require("./ClearIndicator");
var _MultiValue = require("./MultiValue");
var _MultiValueContainer = require("./MultiValueContainer");
var _Option = require("./Option");
var _SingleValue = require("./SingleValue");
var _Input = require("./Input");
var _SingleValueContainer = require("./SingleValueContainer");
var _PopupInput = require("./PopupInput");
var _PopupControl = require("./PopupControl");
var _Menu = require("./Menu");
var _Control = _interopRequireDefault(require("./Control"));
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
/**
 * Memoize getComponents to avoid rerenders.
 */
var getComponents = exports.getComponents = (0, _memoizeOne.default)(function (multi, anchor, showClearIndicator) {
  if (anchor) {
    return {
      Control: anchor,
      Option: _Option.Option
    };
  } else {
    return {
      MultiValue: _MultiValue.MultiValue,
      DropdownIndicator: null,
      SingleValue: _SingleValue.SingleValue,
      ClearIndicator: showClearIndicator || !multi ? _ClearIndicator.ClearIndicator : null,
      Option: _Option.Option,
      ValueContainer: multi ? _MultiValueContainer.MultiValueContainer : _SingleValueContainer.SingleValueContainer,
      Input: _Input.Input,
      Menu: _Menu.Menu,
      Control: _Control.default
    };
  }
});
var getPopupComponents = exports.getPopupComponents = (0, _memoizeOne.default)(function (hasPopupTitle) {
  var baseProps = {
    DropdownIndicator: null,
    SingleValue: _SingleValue.SingleValue,
    ClearIndicator: _ClearIndicator.ClearIndicator,
    Option: _Option.Option,
    ValueContainer: _SingleValueContainer.SingleValueContainer,
    Input: _PopupInput.PopupInput
  };
  if (hasPopupTitle) {
    return _objectSpread(_objectSpread({}, baseProps), {}, {
      Control: _PopupControl.PopupControl
    });
  }
  return baseProps;
});
