// @atlaskit/user-picker/dist/cjs/components/creatable.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getCreatableProps = void 0;
var _memoizeOne = _interopRequireDefault(require("memoize-one"));
var _types = require("../types");
var _emailValidation = require("./emailValidation");
var _utils = require("./utils");
var validOption = ['VALID', 'POTENTIAL'];
var isValidNewOption = function isValidNewOption() {
  var isValidEmail = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _emailValidation.isValidEmail;
  return function (inputValue) {
    return inputValue && validOption.includes(isValidEmail(inputValue));
  };
};
var getNewOptionData = function getNewOptionData(inputValue) {
  return {
    label: inputValue,
    value: inputValue,
    data: {
      id: inputValue,
      name: inputValue,
      type: _types.EmailType
    }
  };
};
var formatCreateLabel = function formatCreateLabel(inputText) {
  if (inputText) {
    return inputText.trim();
  }
  return '';
};
var isOptionDisabled = function isOptionDisabled() {
  var isValidEmail = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _emailValidation.isValidEmail;
  return function (option) {
    if ((0, _utils.isEmail)(option.data)) {
      return isValidEmail(option.data.id) !== 'VALID';
    }
    return !!option.isDisabled;
  };
};
function creatableProps(isValidEmail) {
  return {
    allowCreateWhileLoading: true,
    createOptionPosition: 'first',
    isValidNewOption: isValidNewOption(isValidEmail),
    getNewOptionData: getNewOptionData,
    formatCreateLabel: formatCreateLabel,
    isOptionDisabled: isOptionDisabled(isValidEmail)
  };
}

// ignoring the 'clear' memoizeOne function object property
var getCreatableProps = exports.getCreatableProps = (0, _memoizeOne.default)(creatableProps);
