// @atlaskit/user-picker/dist/cjs/components/Input.js
/* Input.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Input = void 0;
require("./Input.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireDefault(require("react"));
var _select = require("@atlaskit/select");
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var inputStyles = {
  root: "_5ce6131l _1ir3kb7n _h5gq131l"
};
var Input = exports.Input = /*#__PURE__*/function (_React$Component) {
  function Input() {
    var _this;
    (0, _classCallCheck2.default)(this, Input);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, Input, [].concat(args));
    // onKeyPress is used instead as
    // react-select is using onKeyDown for capturing keyboard input
    (0, _defineProperty2.default)(_this, "handleKeyPress", function (e) {
      var _this$props$selectPro;
      if (e.key === 'Enter') {
        e.preventDefault();
      }
      //@ts-ignore react-select unsupported props
      if ((_this$props$selectPro = _this.props.selectProps) !== null && _this$props$selectPro !== void 0 && _this$props$selectPro.disableInput) {
        e.preventDefault();
      }
    });
    return _this;
  }
  (0, _inherits2.default)(Input, _React$Component);
  return (0, _createClass2.default)(Input, [{
    key: "ariaLabelledBy",
    get:
    /**
     * Overrides the default behaviour of react-select lib
     *
     * The getter is a workaround for original behaviour of the react-select lib.
     * Placeholder and Input are linked not via label, aria-label or aria-labeledby, but through aria-describedby.
     * Basically in the getter we reassign Placeholder ID from aria-describedby to aria-labelledby
     * {@link https://github.com/JedWatson/react-select/issues/5651#issue-1731353197 GitHub}
     */
    function get() {
      var _this$props$ariaLabe;
      return (_this$props$ariaLabe = this.props['aria-labelledby']) !== null && _this$props$ariaLabe !== void 0 ? _this$props$ariaLabe : this.props['aria-describedby'];
    }
  }, {
    key: "ariaDescribedBy",
    get: function get() {
      var _this$props$selectPro2, _this$props$selectPro3;
      return (_this$props$selectPro2 = (_this$props$selectPro3 = this.props.selectProps) === null || _this$props$selectPro3 === void 0 ? void 0 : _this$props$selectPro3['aria-describedby']) !== null && _this$props$selectPro2 !== void 0 ? _this$props$selectPro2 : this.props['aria-describedby'];
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement(_select.components.Input, (0, _extends2.default)({}, this.props, {
        "aria-labelledby": this.ariaLabelledBy,
        "aria-describedby": this.ariaDescribedBy,
        innerRef: this.props.innerRef,
        onKeyPress: this.handleKeyPress,
        xcss: inputStyles.root
      }));
    }
  }]);
}(_react.default.Component);
