// @atlaskit/user-picker/dist/cjs/components/MultiValueContainer.js
/* MultiValueContainer.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MultiValueContainer = void 0;
require("./MultiValueContainer.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _select = require("@atlaskit/select");
var _react = _interopRequireWildcard(require("react"));
var _reactDom = _interopRequireDefault(require("react-dom"));
var _reactIntlNext = require("react-intl-next");
var _i18n = require("./i18n");
var _utils = require("./utils");
var _ValueContainerWrapper = _interopRequireDefault(require("./ValueContainerWrapper"));
var _excluded = ["children", "innerProps"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var valueContainerStyles = {
  root: "_yv0e5bpd _ca0q12x7 _n3td12x7 _19bv12x7 _1reo15vq _18m91wug _1fjgglyw _c71l1osq _1suq1j28 _1ke8idpf"
};
var MultiValueContainer = exports.MultiValueContainer = /*#__PURE__*/function (_React$PureComponent) {
  function MultiValueContainer(props) {
    var _this;
    (0, _classCallCheck2.default)(this, MultiValueContainer);
    _this = _callSuper(this, MultiValueContainer, [props]);
    (0, _defineProperty2.default)(_this, "timeoutId", null);
    (0, _defineProperty2.default)(_this, "scrollToBottom", function () {
      _this.timeoutId = window.setTimeout(function () {
        var current = _this.valueContainerInnerProps.ref.current;
        if (current !== null) {
          var container = _reactDom.default.findDOMNode(current);
          if (container instanceof HTMLDivElement) {
            container.scrollTop = container.scrollHeight;
          }
        }
        _this.timeoutId = null;
      });
    });
    (0, _defineProperty2.default)(_this, "showPlaceholder", function () {
      var value = _this.props.selectProps.value;
      return value && value.length > 0;
    });
    (0, _defineProperty2.default)(_this, "addPlaceholder", function (placeholder) {
      var children = _react.default.Children.map(_this.props.children, function (child) {
        return (0, _utils.isChildInput)(child) && _this.showPlaceholder() ? /*#__PURE__*/_react.default.cloneElement(child, {
          placeholder: placeholder
        }) : child;
      });
      return /*#__PURE__*/_react.default.createElement(_react.Fragment, null, children);
    });
    (0, _defineProperty2.default)(_this, "renderChildren", function () {
      var _this$props$selectPro = _this.props.selectProps,
        addMoreMessage = _this$props$selectPro.addMoreMessage,
        isDisabled = _this$props$selectPro.isDisabled;
      // Do not render "Add more..." message if picker is disabled
      if (isDisabled) {
        return _this.props.children;
      }
      if (addMoreMessage === undefined) {
        return /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, _i18n.messages.addMore, function (addMore) {
          var _join, _addMoreMessages;
          var addMoreMessages = addMore;
          if (addMore && typeof addMore === 'string') {
            addMoreMessages = [addMore];
          }
          var placeholder = (_join = (_addMoreMessages = addMoreMessages) === null || _addMoreMessages === void 0 ? void 0 : _addMoreMessages.join('')) !== null && _join !== void 0 ? _join : '';
          return _this.addPlaceholder(placeholder);
        });
      }
      return _this.addPlaceholder(addMoreMessage);
    });
    //@ts-ignore react-select unsupported props
    (0, _defineProperty2.default)(_this, "onValueContainerClick", _this.props.selectProps.onValueContainerClick);
    _this.state = {
      valueSize: 0,
      previousValueSize: 0
    };
    _this.valueContainerInnerProps = {
      ref: /*#__PURE__*/_react.default.createRef()
    };
    return _this;
  }
  (0, _inherits2.default)(MultiValueContainer, _React$PureComponent);
  return (0, _createClass2.default)(MultiValueContainer, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      var _this$state = this.state,
        previousValueSize = _this$state.previousValueSize,
        valueSize = _this$state.valueSize;
      //@ts-ignore react-select unsupported props
      var isFocused = this.props.selectProps.isFocused;
      if (valueSize > previousValueSize && isFocused) {
        if (this.timeoutId) {
          window.clearTimeout(this.timeoutId);
          this.timeoutId = null;
        }
        this.scrollToBottom();
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (this.timeoutId) {
        window.clearTimeout(this.timeoutId);
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
        children = _this$props.children,
        innerProps = _this$props.innerProps,
        valueContainerProps = (0, _objectWithoutProperties2.default)(_this$props, _excluded);
      var props = _objectSpread(_objectSpread({}, valueContainerProps), {}, {
        innerProps: this.valueContainerInnerProps
      });
      return /*#__PURE__*/_react.default.createElement(_ValueContainerWrapper.default, {
        isEnabled: this.onValueContainerClick,
        onMouseDown: this.onValueContainerClick
      }, /*#__PURE__*/_react.default.createElement(_select.components.ValueContainer, (0, _extends2.default)({}, props, {
        xcss: valueContainerStyles.root
      }), this.renderChildren()));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, prevState) {
      return {
        valueSize: nextProps.getValue ? nextProps.getValue().length : 0,
        previousValueSize: prevState.valueSize
      };
    }
  }]);
}(_react.default.PureComponent);
