// @atlaskit/user-picker/dist/cjs/components/popup.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getPopupProps = void 0;
var _memoizeOne = _interopRequireDefault(require("memoize-one"));
var getPopupProps = exports.getPopupProps = (0, _memoizeOne.default)(function (width, target, onFlip, boundariesElement, offset, placement, rootBoundary, shouldFlip, popupTitle, strategy) {
  return {
    searchThreshold: -1,
    controlShouldRenderValue: true,
    minMenuWidth: width,
    maxMenuWidth: width,
    autoFocus: false,
    target: target,
    popupTitle: popupTitle,
    popperProps: {
      placement: placement || 'auto',
      strategy: strategy || 'fixed',
      modifiers: [{
        name: 'offset',
        options: {
          offset: offset
        }
      }, {
        name: 'handleFlipStyle',
        enabled: true,
        order: 910,
        fn: function fn(data) {
          return onFlip(data);
        }
      }, {
        name: 'preventOverflow',
        options: {
          rootBoundary: rootBoundary,
          boundary: boundariesElement
        }
      }, {
        name: 'flip',
        enabled: shouldFlip
      }]
    }
  };
});
