// @atlaskit/user-picker/dist/cjs/components/SingleValue.js
/* SingleValue.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SingleValue = void 0;
require("./SingleValue.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _select = require("@atlaskit/select");
var _SizeableAvatar = require("./SizeableAvatar");
var _utils = require("./utils");
var _verifiedTeamIcon = require("@atlaskit/people-teams-ui-public/verified-team-icon");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _compiled = require("@atlaskit/primitives/compiled");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var styles = {
  avatarItem: "_1reo15vq _18m915vq _4cvr1h6o _1ul953f4",
  avatarItemTextWrapper: "_16jlkb7n _1o9zkb7n _i0dl1osq _1ul9idpf _p12f1osq _bozgu2gc _vwz419ii",
  avatarItemText: "_syaz1fxt _1reo15vq _1bto1l2s _o5721q9c"
};
var ElementAfter = function ElementAfter(props) {
  var data = props.data.data;
  if ((0, _utils.isTeam)(data) && data.verified) {
    return /*#__PURE__*/React.createElement(_verifiedTeamIcon.VerifiedTeamIcon, null);
  }
  return null;
};
var SingleValue = exports.SingleValue = function SingleValue(props) {
  var _props$data = props.data,
    label = _props$data.label,
    data = _props$data.data,
    _props$selectProps = props.selectProps,
    appearance = _props$selectProps.appearance,
    isFocused = _props$selectProps.isFocused;
  return !isFocused ? /*#__PURE__*/React.createElement(_select.components.SingleValue, props, /*#__PURE__*/React.createElement(_compiled.Flex, {
    xcss: styles.avatarItem
  }, /*#__PURE__*/React.createElement(_SizeableAvatar.SizeableAvatar, {
    src: (0, _utils.getAvatarUrl)(data),
    appearance: appearance,
    type: (0, _utils.isTeam)(data) && (0, _platformFeatureFlags.fg)('verified-team-in-user-picker') ? 'team' : 'person'
  }), /*#__PURE__*/React.createElement(_compiled.Box, {
    xcss: styles.avatarItem
  }, /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)([styles.avatarItemTextWrapper])
  }, /*#__PURE__*/React.createElement(_compiled.Box, {
    xcss: styles.avatarItemText
  }, (0, _platformFeatureFlags.fg)('verified-team-in-user-picker') ? /*#__PURE__*/React.createElement(_compiled.Inline, {
    alignBlock: "center"
  }, label, /*#__PURE__*/React.createElement(ElementAfter, props)) : label))))) : null;
};
