// @atlaskit/user-picker/dist/cjs/components/UserOption.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UserOption = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _colors = require("@atlaskit/theme/colors");
var _react = require("@emotion/react");
var _react2 = _interopRequireDefault(require("react"));
var _AvatarItemOption = require("./AvatarItemOption");
var _HighlightText = require("./HighlightText");
var _SizeableAvatar = require("./SizeableAvatar");
var _utils = require("./utils");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); } /**
 * @jsxRuntime classic
 * @jsx jsx
 */ // eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
var UserOption = exports.UserOption = /*#__PURE__*/function (_React$PureComponent) {
  function UserOption() {
    var _this;
    (0, _classCallCheck2.default)(this, UserOption);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, UserOption, [].concat(args));
    (0, _defineProperty2.default)(_this, "getPrimaryText", function () {
      var _this$props$user = _this.props.user,
        name = _this$props$user.name,
        publicName = _this$props$user.publicName,
        highlight = _this$props$user.highlight;
      var result = [(0, _react.jsx)("span", {
        key: "name"
        // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
        ,
        css: (0, _AvatarItemOption.textWrapper)(_this.props.isSelected ? "var(--ds-text-selected, ".concat(_colors.B400, ")") : "var(--ds-text, ".concat(_colors.N800, ")"))
      }, (0, _react.jsx)(_HighlightText.HighlightText, {
        highlights: highlight && highlight.name
      }, name))];
      if ((0, _utils.hasValue)(publicName) && name.trim() !== publicName.trim()) {
        result.push((0, _react.jsx)(_react2.default.Fragment, {
          key: "publicName"
        }, ' ', (0, _react.jsx)("span", {
          // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
          css: (0, _AvatarItemOption.textWrapper)(_this.props.isSelected ? "var(--ds-text-selected, ".concat(_colors.B400, ")") : "var(--ds-text-subtlest, ".concat(_colors.N200, ")"))
        }, "(", (0, _react.jsx)(_HighlightText.HighlightText, {
          highlights: highlight && highlight.publicName
        }, publicName), ")")));
      }
      return result;
    });
    (0, _defineProperty2.default)(_this, "renderSecondaryText", function () {
      return _this.props.user.byline ? (0, _react.jsx)("span", {
        // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
        css: (0, _AvatarItemOption.textWrapper)(_this.props.isSelected ? "var(--ds-text-selected, ".concat(_colors.B400, ")") : "var(--ds-text-subtlest, ".concat(_colors.N200, ")"))
      }, _this.props.user.byline) : undefined;
    });
    (0, _defineProperty2.default)(_this, "renderAvatar", function () {
      var _this$props = _this.props,
        avatarUrl = _this$props.user.avatarUrl,
        status = _this$props.status;
      return (0, _react.jsx)(_SizeableAvatar.SizeableAvatar, {
        appearance:
        // eslint-disable-next-line @atlaskit/platform/no-preconditioning, @atlaskit/platform/ensure-feature-flag-prefix
        (0, _platformFeatureFlags.fg)('platform-component-visual-refresh') && (0, _platformFeatureFlags.fg)('visual-refresh-user-picker') ? 'medium' : 'big',
        src: avatarUrl,
        presence: status
      });
    });
    (0, _defineProperty2.default)(_this, "getLozengeProps", function () {
      return typeof _this.props.user.lozenge === 'string' ? {
        text: _this.props.user.lozenge
      } : _this.props.user.lozenge;
    });
    return _this;
  }
  (0, _inherits2.default)(UserOption, _React$PureComponent);
  return (0, _createClass2.default)(UserOption, [{
    key: "render",
    value: function render() {
      return (0, _react.jsx)(_AvatarItemOption.AvatarItemOption, {
        avatar: this.renderAvatar(),
        lozenge: this.getLozengeProps(),
        isDisabled: this.props.user.isDisabled,
        primaryText: this.getPrimaryText(),
        secondaryText: this.renderSecondaryText()
      });
    }
  }]);
}(_react2.default.PureComponent);
