// @atlaskit/user-picker/dist/cjs/components/UserPicker.js
/* UserPicker.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UserPickerWithoutAnalytics = exports.UserPicker = void 0;
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _analyticsNext = require("@atlaskit/analytics-next");
var _select = _interopRequireWildcard(require("@atlaskit/select"));
var _ufo = require("@atlaskit/ufo");
var _react = _interopRequireDefault(require("react"));
var _BaseUserPicker = require("./BaseUserPicker");
var _styles = require("./styles");
var _components = require("./components");
var _creatable = require("./creatable");
var _creatableEmailSuggestion = require("./creatableEmailSuggestion");
var _MessagesIntlProvider = _interopRequireDefault(require("./MessagesIntlProvider"));
var _UserSourceProvider = require("../clients/UserSourceProvider");
var _ufoExperiences = require("../util/ufoExperiences");
var _uuid = require("uuid");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var UserPickerWithoutAnalytics = exports.UserPickerWithoutAnalytics = /*#__PURE__*/function (_React$Component) {
  function UserPickerWithoutAnalytics(props) {
    var _this;
    (0, _classCallCheck2.default)(this, UserPickerWithoutAnalytics);
    _this = _callSuper(this, UserPickerWithoutAnalytics, [props]);
    _this.ufoId = (0, _uuid.v4)();
    var experienceForId = _ufoExperiences.userPickerRenderedUfoExperience.getInstance(_this.ufoId);
    if (![_ufo.UFOExperienceState.IN_PROGRESS.id, _ufo.UFOExperienceState.STARTED.id].includes(experienceForId.state.id)) {
      experienceForId.start();
    }
    return _this;
  }
  (0, _inherits2.default)(UserPickerWithoutAnalytics, _React$Component);
  return (0, _createClass2.default)(UserPickerWithoutAnalytics, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var experienceForId = _ufoExperiences.userPickerRenderedUfoExperience.getInstance(this.ufoId);

      // Send UFO success if the experience is still in progress i.e. hasn't failed
      if ([_ufo.UFOExperienceState.IN_PROGRESS.id, _ufo.UFOExperienceState.STARTED.id].includes(experienceForId.state.id)) {
        experienceForId.success();
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
        emailLabel = _this$props.emailLabel,
        allowEmail = _this$props.allowEmail,
        suggestEmailsForDomain = _this$props.suggestEmailsForDomain,
        isMulti = _this$props.isMulti,
        isValidEmail = _this$props.isValidEmail,
        anchor = _this$props.anchor,
        menuPortalTarget = _this$props.menuPortalTarget,
        menuPosition = _this$props.menuPosition,
        menuShouldBlockScroll = _this$props.menuShouldBlockScroll,
        captureMenuScroll = _this$props.captureMenuScroll,
        closeMenuOnScroll = _this$props.closeMenuOnScroll,
        loadUserSource = _this$props.loadUserSource,
        _this$props$required = _this$props.required,
        required = _this$props$required === void 0 ? false : _this$props$required,
        _this$props$showClear = _this$props.showClearIndicator,
        showClearIndicator = _this$props$showClear === void 0 ? false : _this$props$showClear;
      var width = this.props.width;
      var SelectComponent = allowEmail ? _select.CreatableSelect : _select.default;
      var creatableProps = suggestEmailsForDomain ? (0, _creatableEmailSuggestion.getCreatableSuggestedEmailProps)(suggestEmailsForDomain, isValidEmail) : (0, _creatable.getCreatableProps)(isValidEmail);
      var defaultPickerProps = {
        closeMenuOnScroll: closeMenuOnScroll,
        menuPortalTarget: menuPortalTarget,
        menuPosition: menuPosition,
        menuShouldBlockScroll: menuShouldBlockScroll,
        captureMenuScroll: captureMenuScroll,
        required: required
      };
      var pickerProps = allowEmail ? _objectSpread(_objectSpread(_objectSpread({}, defaultPickerProps), creatableProps), {}, {
        emailLabel: emailLabel
      }) : _objectSpread({}, defaultPickerProps);
      return /*#__PURE__*/_react.default.createElement(_ufoExperiences.UfoErrorBoundary, {
        id: this.ufoId
      }, /*#__PURE__*/_react.default.createElement(_MessagesIntlProvider.default, null, /*#__PURE__*/_react.default.createElement(_UserSourceProvider.ExusUserSourceProvider, {
        fetchUserSource: loadUserSource
      }, /*#__PURE__*/_react.default.createElement(_BaseUserPicker.BaseUserPickerWithoutAnalytics, (0, _extends2.default)({}, this.props, {
        width: width,
        SelectComponent: SelectComponent
        // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
        ,
        styles: (0, _styles.getStyles)(width, isMulti, this.props.appearance === 'compact', this.props.styles, this.props.isInvalid,
        // eslint-disable-next-line @atlaskit/platform/no-preconditioning, @atlaskit/platform/ensure-feature-flag-prefix
        (0, _platformFeatureFlags.fg)('platform-component-visual-refresh') && (0, _platformFeatureFlags.fg)('visual-refresh-user-picker')),
        components: (0, _components.getComponents)(isMulti, anchor, showClearIndicator && (0, _platformFeatureFlags.fg)('user-picker-show-clear-indicator')),
        pickerProps: pickerProps
      })))));
    }
  }]);
}(_react.default.Component);
(0, _defineProperty2.default)(UserPickerWithoutAnalytics, "defaultProps", {
  width: 350,
  isMulti: false
});
var UserPicker = exports.UserPicker = (0, _analyticsNext.withAnalyticsEvents)()(UserPickerWithoutAnalytics);
