// @atlaskit/table/dist/cjs/ui/tr.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TR = void 0;
var _react = require("@emotion/react");
var _focusRing = _interopRequireDefault(require("@atlaskit/focus-ring"));
/**
 * @jsxRuntime classic
 * @jsx jsx
 */

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var baseStyles = (0, _react.css)({
  height: '3rem',
  position: 'relative',
  border: 'none',
  borderImageWidth: 0,
  borderSpacing: 0
});
var selectedStyles = (0, _react.css)({
  backgroundColor: "var(--ds-background-selected, #DEEBFF88)",
  '&:hover': {
    backgroundColor: "var(--ds-background-selected-hovered, #DEEBFF)" // B50
  }
});
var subitemStyles = (0, _react.css)({
  backgroundColor: "var(--ds-background-neutral, #091E420F)"
});
var bodyRowStyles = (0, _react.css)({
  borderBlockEnd: "1px solid ".concat("var(--ds-border, #eee)"),
  '&:hover': {
    backgroundColor: "var(--ds-background-neutral-subtle-hovered, #f8f8f8)"
  },
  '&:focus-visible::after': {
    boxShadow: 'none'
  }
});
/**
 * __Row__
 *
 * A row primitive.
 *
 * - [Examples](https://atlassian.design/components/table/examples)
 */
var TR = exports.TR = function TR(_ref) {
  var children = _ref.children,
    testId = _ref.testId,
    isSelected = _ref.isSelected,
    _ref$isBodyRow = _ref.isBodyRow,
    isBodyRow = _ref$isBodyRow === void 0 ? true : _ref$isBodyRow,
    isSubitem = _ref.isSubitem;
  return (0, _react.jsx)(_focusRing.default, {
    isInset: true
  }, (0, _react.jsx)("tr", {
    tabIndex: -1,
    "aria-selected": isSelected,
    "data-testid": testId,
    css: [baseStyles, isBodyRow && bodyRowStyles, isSelected && selectedStyles, isSubitem && subitemStyles]
  }, children));
};
