// @atlaskit/jql-editor/dist/cjs/ui/jql-editor-controls-content/base-search/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BaseSearch = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _button = require("@atlaskit/button");
var _new = require("@atlaskit/button/new");
var _searchEditorSearch = _interopRequireDefault(require("@atlaskit/icon/core/migration/search--editor-search"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _primitives = require("@atlaskit/primitives");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var style = {
  // Fixes an issue where loading button makes the editor flicker with a scrollbar
  overflow: 'hidden'
};
// Fixes icon margin issus after new icon migration
var iconStyle = (0, _primitives.xcss)({
  margin: 'space.050',
  display: 'flex'
});

// Fixes an issue where loading button makes the editor flicker with a scrollbar
var buttonContainerStyle = (0, _primitives.xcss)({
  overflow: 'hidden'
});
var BaseSearch = exports.BaseSearch = function BaseSearch(_ref) {
  var isDisabled = _ref.isDisabled,
    isSearching = _ref.isSearching,
    label = _ref.label,
    onSearch = _ref.onSearch;
  // Prevent click events being repeatedly fired if the Enter key is held down.
  var preventRepeatClick = (0, _react.useCallback)(function (e) {
    if (e.key === 'Enter' && e.repeat) {
      e.preventDefault();
    }
  }, []);
  return (0, _platformFeatureFlags.fg)('platform-component-visual-refresh') ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_primitives.Box, {
    xcss: buttonContainerStyle
  }, /*#__PURE__*/_react.default.createElement(_new.IconButton, (0, _extends2.default)({
    label: label,
    isDisabled: isDisabled,
    testId: "jql-editor-search",
    appearance: "default",
    spacing: "compact",
    onClick: onSearch,
    onKeyDown: preventRepeatClick,
    isLoading: isSearching,
    icon: _searchEditorSearch.default
  }, (0, _platformFeatureFlags.fg)('add_nin_press_interactions') && {
    interactionName: 'jql-editor-base-search-button'
  })))) : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_button.LoadingButton, (0, _extends2.default)({
    "aria-label": label,
    isDisabled: isDisabled,
    testId: "jql-editor-search"
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
    ,
    style: style,
    appearance: 'primary',
    spacing: 'none',
    onClick: onSearch,
    onKeyDown: preventRepeatClick,
    isLoading: isSearching,
    iconBefore: /*#__PURE__*/_react.default.createElement(_primitives.Box, {
      xcss: iconStyle
    }, /*#__PURE__*/_react.default.createElement(_searchEditorSearch.default, {
      color: "currentColor",
      label: '',
      LEGACY_size: 'medium',
      LEGACY_margin: "-4px"
    }))
  }, (0, _platformFeatureFlags.fg)('add_nin_press_interactions') && {
    interactionName: 'jql-editor-base-search-button'
  })));
};
