// @atlaskit/jql-ast/dist/cjs/printers/print-ast-to-doc.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.printAstToDoc = exports.AstToDocVisitor = void 0;
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _constants = require("../constants");
var _jastVisitors = require("../jast-visitors");
var _utils = require("./utils");
function _createSuper(t) { var r = _isNativeReflectConstruct(); return function () { var e, o = (0, _getPrototypeOf2.default)(t); if (r) { var s = (0, _getPrototypeOf2.default)(this).constructor; e = Reflect.construct(o, arguments, s); } else e = o.apply(this, arguments); return (0, _possibleConstructorReturn2.default)(this, e); }; }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var formatOperator = function formatOperator(operator, operatorCase) {
  switch (operatorCase) {
    case 'lower':
      return operator.toLowerCase();
    case 'upper':
      return operator.toUpperCase();
    case 'preserve':
      return operator;
  }
};
var AstToDocVisitor = exports.AstToDocVisitor = /*#__PURE__*/function (_AbstractJastVisitor) {
  (0, _inherits2.default)(AstToDocVisitor, _AbstractJastVisitor);
  var _super = _createSuper(AstToDocVisitor);
  function AstToDocVisitor() {
    var _this;
    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
      operatorCase = _ref.operatorCase;
    (0, _classCallCheck2.default)(this, AstToDocVisitor);
    _this = _super.call(this);
    _this.operatorCase = operatorCase || 'preserve';
    return _this;
  }
  (0, _createClass2.default)(AstToDocVisitor, [{
    key: "visitArgument",
    value: function visitArgument(argument) {
      return argument.text;
    }

    /**
     * Visit each of the provided nodes and form a new array with each `Doc` joined by the provided separator.
     */
  }, {
    key: "joinNodes",
    value: function joinNodes(nodes, separator) {
      var _this2 = this;
      var docs = [];
      nodes.forEach(function (node) {
        if (separator !== undefined && docs.length > 0) {
          docs.push(separator);
        }
        docs.push(node.accept(_this2));
      });
      return docs;
    }
  }, {
    key: "visitCompoundClause",
    value: function visitCompoundClause(compoundClause) {
      var _this3 = this;
      var operatorDoc = compoundClause.operator.accept(this);
      var clauseDocs = [];
      compoundClause.clauses.forEach(function (clause) {
        if (clauseDocs.length > 0) {
          clauseDocs.push((0, _utils.ifBreak)((0, _utils.newLine)(), ' '), operatorDoc, ' ');
        }

        // Wrap nested compound clauses in parentheses
        if (clause.clauseType === _constants.CLAUSE_TYPE_COMPOUND) {
          clauseDocs.push('(', clause.accept(_this3), ')');
        } else {
          clauseDocs.push(clause.accept(_this3));
        }
      });
      return (0, _utils.group)(clauseDocs);
    }
  }, {
    key: "visitCompoundOperator",
    value: function visitCompoundOperator(compoundOperator) {
      return formatOperator(compoundOperator.value, this.operatorCase);
    }
  }, {
    key: "visitField",
    value: function visitField(field) {
      var propertyDocs = field.properties ? this.joinNodes(field.properties) : [];
      return (0, _utils.group)([field.text].concat((0, _toConsumableArray2.default)(propertyDocs)));
    }
  }, {
    key: "visitFunction",
    value: function visitFunction(functionString) {
      return functionString.text;
    }
  }, {
    key: "visitFunctionOperand",
    value: function visitFunctionOperand(functionOperand) {
      var argumentDocs = this.joinNodes(functionOperand.arguments, ', ');
      return (0, _utils.group)([functionOperand.function.accept(this), '('].concat((0, _toConsumableArray2.default)(argumentDocs), [')']));
    }
  }, {
    key: "visitKeywordOperand",
    value: function visitKeywordOperand(keywordOperand) {
      return keywordOperand.value;
    }
  }, {
    key: "visitListOperand",
    value: function visitListOperand(listOperand) {
      var operandDocs = this.joinNodes(listOperand.values, ', ');
      return (0, _utils.group)(['('].concat((0, _toConsumableArray2.default)(operandDocs), [')']));
    }
  }, {
    key: "visitOperator",
    value: function visitOperator(operator) {
      return formatOperator(operator.text, this.operatorCase);
    }
  }, {
    key: "visitOrderBy",
    value: function visitOrderBy(orderBy) {
      var fieldDocs = this.joinNodes(orderBy.fields, ', ');
      return (0, _utils.group)([orderBy.operator.accept(this), ' '].concat((0, _toConsumableArray2.default)(fieldDocs)));
    }
  }, {
    key: "visitOrderByDirection",
    value: function visitOrderByDirection(orderByDirection) {
      return orderByDirection.value;
    }
  }, {
    key: "visitOrderByField",
    value: function visitOrderByField(orderByField) {
      var docs = [orderByField.field.accept(this)];
      if (orderByField.direction) {
        docs.push(' ', orderByField.direction.accept(this));
      }
      return (0, _utils.group)(docs);
    }
  }, {
    key: "visitOrderByOperator",
    value: function visitOrderByOperator(orderByOperator) {
      return formatOperator(orderByOperator.value, this.operatorCase);
    }
  }, {
    key: "visitPredicate",
    value: function visitPredicate(predicate) {
      var docs = [predicate.operator.accept(this)];
      if (predicate.operand) {
        docs.push(' ', predicate.operand.accept(this));
      }
      return (0, _utils.group)(docs);
    }
  }, {
    key: "visitPredicateOperator",
    value: function visitPredicateOperator(predicateOperator) {
      return formatOperator(predicateOperator.text, this.operatorCase);
    }
  }, {
    key: "visitProperty",
    value: function visitProperty(property) {
      var docs = [];
      if (property.key) {
        docs.push('[', property.key.accept(this), ']');
      }
      if (property.path) {
        var pathDocs = this.joinNodes(property.path, ' ');
        docs.push.apply(docs, (0, _toConsumableArray2.default)(pathDocs));
      }
      return (0, _utils.group)(docs);
    }
  }, {
    key: "visitQuery",
    value: function visitQuery(query) {
      if (query.where && query.orderBy) {
        return (0, _utils.group)([query.where.accept(this), (0, _utils.ifBreak)((0, _utils.newLine)(), ' '), query.orderBy.accept(this)]);
      } else if (query.where) {
        return query.where.accept(this);
      } else if (query.orderBy) {
        return query.orderBy.accept(this);
      }
      return '';
    }
  }, {
    key: "visitTerminalClause",
    value: function visitTerminalClause(terminalClause) {
      var _this4 = this;
      var docs = [terminalClause.field.accept(this)];
      if (terminalClause.operator) {
        docs.push(' ', terminalClause.operator.accept(this));
      }
      if (terminalClause.operand) {
        docs.push(' ', terminalClause.operand.accept(this));
      }
      terminalClause.predicates.forEach(function (predicate) {
        docs.push(' ', predicate.accept(_this4));
      });
      return (0, _utils.group)(docs);
    }
  }, {
    key: "visitValueOperand",
    value: function visitValueOperand(valueOperand) {
      return valueOperand.text;
    }
  }, {
    key: "visitNotClause",
    value: function visitNotClause(notClause) {
      var clauseDocs = [];
      // Wrap nested compound clauses in parentheses
      if (notClause.clause.clauseType === _constants.CLAUSE_TYPE_COMPOUND) {
        clauseDocs.push('(', notClause.clause.accept(this), ')');
      } else {
        clauseDocs.push(notClause.clause.accept(this));
      }
      return (0, _utils.group)([notClause.operator.accept(this), ' '].concat(clauseDocs));
    }
  }, {
    key: "visitNotClauseOperator",
    value: function visitNotClauseOperator(notClauseOperator) {
      return notClauseOperator.value;
    }
  }, {
    key: "visitChildren",
    value: function visitChildren() {
      throw new Error('Required visit method not implemented for node');
    }
  }, {
    key: "defaultResult",
    value: function defaultResult() {
      return '';
    }
  }]);
  return AstToDocVisitor;
}(_jastVisitors.AbstractJastVisitor);
var printAstToDoc = exports.printAstToDoc = function printAstToDoc(jast, options) {
  var astToDocVisitor = new AstToDocVisitor(options);
  return jast.query ? jast.query.accept(astToDocVisitor) : '';
};
