// @atlaskit/editor-plugin-highlight/dist/cjs/ui/PrimaryToolbarHighlightColor.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PrimaryToolbarHighlightColorWithIntl = void 0;
var _react = require("react");
var _react2 = require("@emotion/react");
var _reactIntlNext = require("react-intl-next");
var _analytics = require("@atlaskit/editor-common/analytics");
var _hooks = require("@atlaskit/editor-common/hooks");
var _icons = require("@atlaskit/editor-common/icons");
var _keymaps = require("@atlaskit/editor-common/keymaps");
var _messages = require("@atlaskit/editor-common/messages");
var _styles = require("@atlaskit/editor-common/styles");
var _uiMenu = require("@atlaskit/editor-common/ui-menu");
var _editorPalette = require("@atlaskit/editor-palette");
var _highlight = _interopRequireDefault(require("@atlaskit/icon/core/highlight"));
var _chevronDown = _interopRequireDefault(require("@atlaskit/icon/utility/migration/chevron-down"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _compiled = require("@atlaskit/primitives/compiled");
var _palette = require("../editor-commands/palette");
var _EditorHighlightIcon = require("./shared/EditorHighlightIcon");
var _PaletteDropdown = require("./shared/PaletteDropdown");
var _useDropdownEvents2 = require("./shared/useDropdownEvents");
/**
 * @jsxRuntime classic
 * @jsx jsx
 */

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var PrimaryToolbarHighlightColor = function PrimaryToolbarHighlightColor(_ref) {
  var popupsMountPoint = _ref.popupsMountPoint,
    popupsBoundariesElement = _ref.popupsBoundariesElement,
    popupsScrollableElement = _ref.popupsScrollableElement,
    isToolbarReducedSpacing = _ref.isToolbarReducedSpacing,
    disabled = _ref.disabled,
    pluginInjectionApi = _ref.pluginInjectionApi,
    formatMessage = _ref.intl.formatMessage,
    editorView = _ref.editorView;
  var toolbarItemRef = (0, _react.useRef)(null);
  var _useSharedPluginState = (0, _hooks.useSharedPluginState)(pluginInjectionApi, ['highlight']),
    highlightState = _useSharedPluginState.highlightState;
  var setIsDropdownOpen = function setIsDropdownOpen(isOpen) {
    if (!(highlightState !== null && highlightState !== void 0 && highlightState.disabled)) {
      var state = editorView.state,
        dispatch = editorView.dispatch;
      // Ignored via go/ees005
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      (0, _palette.setPalette)(pluginInjectionApi)({
        isPaletteOpen: isOpen,
        inputMethod: _analytics.INPUT_METHOD.TOOLBAR
      })(state, dispatch);
    }
  };
  var isDropdownOpen = !!(highlightState !== null && highlightState !== void 0 && highlightState.isPaletteOpen);
  var _useDropdownEvents = (0, _useDropdownEvents2.useDropdownEvents)({
      toolbarItemRef: toolbarItemRef,
      setIsDropdownOpen: setIsDropdownOpen,
      isDropdownOpen: isDropdownOpen,
      pluginInjectionApi: pluginInjectionApi
    }),
    handleClick = _useDropdownEvents.handleClick,
    handleKeyDown = _useDropdownEvents.handleKeyDown,
    handleClickOutside = _useDropdownEvents.handleClickOutside,
    handleEscapeKeydown = _useDropdownEvents.handleEscapeKeydown,
    handleColorChange = _useDropdownEvents.handleColorChange,
    isOpenedByKeyboard = _useDropdownEvents.isOpenedByKeyboard;

  // Don't render the toolbar option while the plugin is initialising
  if (!highlightState) {
    return null;
  }
  var toolbarButtonLabel = formatMessage(_messages.highlightMessages.highlight);

  // Get the design token for the  active color (if it exists) to modify the toolbar
  // icon, but show the nice rainbow if none is selected
  var activeColorToken = highlightState.activeColor === null ? null : (0, _editorPalette.hexToEditorTextBackgroundPaletteColor)(highlightState.activeColor);
  return (0, _react2.jsx)(_compiled.Flex, {
    alignItems: "center"
  }, (0, _react2.jsx)(_PaletteDropdown.PaletteDropdown, {
    popupsMountPoint: popupsMountPoint,
    popupsBoundariesElement: popupsBoundariesElement,
    popupsScrollableElement: popupsScrollableElement,
    isOpen: isDropdownOpen && !highlightState.disabled,
    activeColor: highlightState.activeColor,
    trigger: (0, _react2.jsx)(_uiMenu.ToolbarButton
    // eslint-disable-next-line @atlaskit/design-system/no-unsafe-style-overrides, @atlaskit/ui-styling-standard/no-imported-style-values
    , {
      css: _styles.disableBlueBorderStyles,
      buttonId: _uiMenu.TOOLBAR_BUTTON.BACKGROUND_COLOR,
      spacing: isToolbarReducedSpacing ? 'none' : 'default',
      disabled: disabled || highlightState.disabled,
      selected: isDropdownOpen,
      "aria-label": (0, _keymaps.tooltip)(_keymaps.toggleHighlightPalette, toolbarButtonLabel),
      "aria-keyshortcuts": (0, _keymaps.getAriaKeyshortcuts)(_keymaps.toggleHighlightPalette),
      "aria-expanded": isDropdownOpen,
      "aria-haspopup": true,
      title: (0, _keymaps.tooltip)(_keymaps.toggleHighlightPalette, toolbarButtonLabel),
      onClick: handleClick,
      onKeyDown: handleKeyDown,
      ref: toolbarItemRef,
      iconBefore:
      // eslint-disable-next-line @atlaskit/platform/ensure-feature-flag-prefix
      (0, _platformFeatureFlags.fg)('platform-visual-refresh-icons') ?
      // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values
      (0, _react2.jsx)("div", {
        css: _styles.triggerWrapperStylesWithPadding
      }, (0, _react2.jsx)(_icons.DynamicStrokeIconDecoration, {
        selectedColor: activeColorToken,
        disabled: highlightState.disabled,
        icon: (0, _react2.jsx)(_highlight.default, {
          label: "",
          color: "currentColor",
          spacing: "spacious"
        })
      }),
      //eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/design-system/consistent-css-prop-usage -- Ignored via go/DSP-18766
      (0, _react2.jsx)("span", {
        css: _styles.expandIconContainerStyle
      }, (0, _react2.jsx)(_chevronDown.default, {
        label: "",
        color: "currentColor"
      }))) : (0, _react2.jsx)(_compiled.Flex, null, (0, _react2.jsx)(_EditorHighlightIcon.EditorHighlightIcon, {
        selectedColor: activeColorToken,
        disabled: highlightState.disabled
      }), (0, _react2.jsx)("span", {
        // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
        css: _styles.expandIconWrapperStyle
      }, (0, _react2.jsx)(_chevronDown.default, {
        label: ""
      })))
    }),
    onColorChange: function onColorChange(color) {
      return handleColorChange({
        color: color,
        inputMethod: _analytics.INPUT_METHOD.TOOLBAR
      });
    },
    isOpenedByKeyboard: isOpenedByKeyboard,
    handleClickOutside: handleClickOutside,
    handleEscapeKeydown: handleEscapeKeydown
  }));
};
var PrimaryToolbarHighlightColorWithIntl = exports.PrimaryToolbarHighlightColorWithIntl = (0, _reactIntlNext.injectIntl)(PrimaryToolbarHighlightColor);
