// @atlaskit/link-datasource/dist/cjs/ui/table-footer/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TableFooter = void 0;
require("./index.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _reactIntlNext = require("react-intl-next");
var _button = _interopRequireDefault(require("@atlaskit/button"));
var _new = require("@atlaskit/button/new");
var _refresh = _interopRequireDefault(require("@atlaskit/icon/core/migration/refresh"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _compiled = require("@atlaskit/primitives/compiled");
var _colors = require("@atlaskit/theme/colors");
var _assetsModal = require("../assets-modal");
var _searchCount = _interopRequireWildcard(require("../common/modal/search-count"));
var _messages = require("./messages");
var _poweredByJsmAssets = require("./powered-by-jsm-assets");
var _providerLink = require("./provider-link");
var _syncInfo = require("./sync-info");
var _excluded = ["as", "style"],
  _excluded2 = ["as", "style"],
  _excluded3 = ["as", "style"],
  _excluded4 = ["as", "style"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var styles = {
  footer: "_11c81o8v _2rko1kw7 _x3dozgxb _syaz131l _ca0q1ejb _u5f3pxbi _n3td1ejb _19bvpxbi _vchhusvi _13liidpf _qrwqidpf _bfhkhp5a",
  separator: "_syazmuej"
};

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled -- To migrate as part of go/ui-styling-standard
var FooterWrapper = (0, _react.forwardRef)(function (_ref, __cmplr) {
  var _ref$as = _ref.as,
    C = _ref$as === void 0 ? "div" : _ref$as,
    __cmpls = _ref.style,
    __cmplp = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  return /*#__PURE__*/React.createElement(C, (0, _extends2.default)({}, __cmplp, {
    style: __cmpls,
    ref: __cmplr,
    className: (0, _runtime.ax)(["_2rko1kw7 _x3do17u4 _ca0qze3t _u5f3pxbi _n3tdze3t _19bvpxbi _vchhusvi _13liidpf _qrwqidpf _bfhk1j9a", __cmplp.className])
  }));
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled -- To migrate as part of go/ui-styling-standard
if (process.env.NODE_ENV !== 'production') {
  FooterWrapper.displayName = 'FooterWrapper';
}
var TopBorderWrapper = (0, _react.forwardRef)(function (_ref2, __cmplr) {
  var _ref2$as = _ref2.as,
    C = _ref2$as === void 0 ? "div" : _ref2$as,
    __cmpls = _ref2.style,
    __cmplp = (0, _objectWithoutProperties2.default)(_ref2, _excluded2);
  return /*#__PURE__*/React.createElement(C, (0, _extends2.default)({}, __cmplp, {
    style: __cmpls,
    ref: __cmplr,
    className: (0, _runtime.ax)(["_1e0c1txw _vchhusvi _1bah1yb4 _ca0qv47k _u5f3ze3t _n3tdv47k _19bvze3t", __cmplp.className])
  }));
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled -- To migrate as part of go/ui-styling-standard
if (process.env.NODE_ENV !== 'production') {
  TopBorderWrapper.displayName = 'TopBorderWrapper';
}
var SyncWrapper = (0, _react.forwardRef)(function (_ref3, __cmplr) {
  var _ref3$as = _ref3.as,
    C = _ref3$as === void 0 ? "div" : _ref3$as,
    __cmpls = _ref3.style,
    __cmplp = (0, _objectWithoutProperties2.default)(_ref3, _excluded3);
  return /*#__PURE__*/React.createElement(C, (0, _extends2.default)({}, __cmplp, {
    style: __cmpls,
    ref: __cmplr,
    className: (0, _runtime.ax)(["_1e0c1txw _4cvr1h6o _syaz1dyx", __cmplp.className])
  }));
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled -- To migrate as part of go/ui-styling-standard
if (process.env.NODE_ENV !== 'production') {
  SyncWrapper.displayName = 'SyncWrapper';
}
var SyncTextWrapper = (0, _react.forwardRef)(function (_ref4, __cmplr) {
  var _ref4$as = _ref4.as,
    C = _ref4$as === void 0 ? "div" : _ref4$as,
    __cmpls = _ref4.style,
    __cmplp = (0, _objectWithoutProperties2.default)(_ref4, _excluded4);
  return /*#__PURE__*/React.createElement(C, (0, _extends2.default)({}, __cmplp, {
    style: __cmpls,
    ref: __cmplr,
    className: (0, _runtime.ax)(["_11c8dcr7 _2hwx12x7", __cmplp.className])
  }));
});
if (process.env.NODE_ENV !== 'production') {
  SyncTextWrapper.displayName = 'SyncTextWrapper';
}
var TableFooter = exports.TableFooter = function TableFooter(_ref5) {
  var datasourceId = _ref5.datasourceId,
    itemCount = _ref5.itemCount,
    onRefresh = _ref5.onRefresh,
    isLoading = _ref5.isLoading,
    url = _ref5.url;
  var intl = (0, _reactIntlNext.useIntl)();
  var _useState = (0, _react.useState)(new Date()),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    lastSyncTime = _useState2[0],
    setLastSyncTime = _useState2[1];
  var showItemCount = itemCount && itemCount > 0 ? true : itemCount === 0 ? !isLoading : false;
  (0, _react.useEffect)(function () {
    if (isLoading) {
      setLastSyncTime(new Date());
    }
  }, [isLoading]);
  if ((0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv')) {
    return onRefresh || showItemCount ? /*#__PURE__*/React.createElement("div", {
      "data-testid": "table-footer",
      className: (0, _runtime.ax)([styles.footer])
    }, /*#__PURE__*/React.createElement(_compiled.Inline, {
      alignBlock: "center",
      alignInline: "end",
      grow: "hug",
      spread: "space-between"
    }, /*#__PURE__*/React.createElement(_compiled.Box, null, /*#__PURE__*/React.createElement(_providerLink.ProviderLink, {
      datasourceId: datasourceId
    })), /*#__PURE__*/React.createElement(_compiled.Inline, {
      alignBlock: "center",
      space: "space.100",
      separator: /*#__PURE__*/React.createElement("div", {
        className: (0, _runtime.ax)([styles.separator])
      }, "\u2022")
    }, onRefresh && /*#__PURE__*/React.createElement(_compiled.Inline, {
      alignBlock: "center",
      space: "space.050"
    }, /*#__PURE__*/React.createElement(_new.IconButton, {
      appearance: "subtle",
      icon: function icon(iconProps) {
        return /*#__PURE__*/React.createElement(_refresh.default, (0, _extends2.default)({}, iconProps, {
          color: "var(--ds-text-subtlest, #626F86)"
        }));
      },
      isDisabled: isLoading,
      label: intl.formatMessage(_messages.footerMessages.refreshLabel),
      onClick: onRefresh,
      spacing: "compact",
      testId: "refresh-button"
    }), /*#__PURE__*/React.createElement(_compiled.Box, {
      testId: "sync-text"
    }, isLoading ? /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, _messages.footerMessages.loadingText) : /*#__PURE__*/React.createElement(_syncInfo.SyncInfo, {
      lastSyncTime: lastSyncTime
    }))), showItemCount && /*#__PURE__*/React.createElement(_compiled.Flex, null, datasourceId === _assetsModal.ASSETS_LIST_OF_LINKS_DATASOURCE_ID ? /*#__PURE__*/React.createElement(_searchCount.AssetsItemCount, {
      searchCount: itemCount,
      url: url,
      testId: "item-count"
    }) : /*#__PURE__*/React.createElement(_searchCount.default, {
      searchCount: itemCount,
      url: url,
      prefixTextType: "item",
      testId: "item-count"
    }))))) : null;
  }

  // If only one of the two is passed in, still show the other one (Note: We keep the div encapsulating the one not shown to
  // ensure correct positioning since 'justify-content: space-between' is used).
  return onRefresh || showItemCount ? /*#__PURE__*/React.createElement(FooterWrapper, {
    "data-testid": "table-footer"
  }, /*#__PURE__*/React.createElement(TopBorderWrapper, null, showItemCount && /*#__PURE__*/React.createElement(_compiled.Flex, null, datasourceId === _assetsModal.ASSETS_LIST_OF_LINKS_DATASOURCE_ID ? /*#__PURE__*/React.createElement(_searchCount.AssetsItemCount, {
    searchCount: itemCount,
    url: url,
    testId: "item-count"
  }) : /*#__PURE__*/React.createElement(_searchCount.default, {
    searchCount: itemCount,
    url: url,
    prefixTextType: "item",
    testId: "item-count"
  })), datasourceId === _assetsModal.ASSETS_LIST_OF_LINKS_DATASOURCE_ID ? /*#__PURE__*/React.createElement(_poweredByJsmAssets.PoweredByJSMAssets, {
    text: intl.formatMessage((0, _platformFeatureFlags.fg)('assets_as_an_app_v2') ? _messages.footerMessages.poweredByAssets : _messages.footerMessages.powerByJSM)
  }) : null, /*#__PURE__*/React.createElement(SyncWrapper, null, onRefresh && /*#__PURE__*/React.createElement(_react.Fragment, null, /*#__PURE__*/React.createElement(SyncTextWrapper, {
    "data-testid": "sync-text"
  }, isLoading ? /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, _messages.footerMessages.loadingText) : /*#__PURE__*/React.createElement(_syncInfo.SyncInfo, {
    lastSyncTime: lastSyncTime
  })), /*#__PURE__*/React.createElement(_button.default, {
    onClick: onRefresh,
    appearance: "subtle",
    iconBefore: /*#__PURE__*/React.createElement(_refresh.default, {
      label: intl.formatMessage(_messages.footerMessages.refreshLabel),
      color: "currentColor",
      spacing: "spacious"
    }),
    isDisabled: isLoading,
    testId: "refresh-button"
  }))))) : null;
};
