// @atlaskit/jql-editor/dist/cjs/plugins/autocomplete/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultAutocompleteProvider = exports.TAB_KEY = exports.JQLAutocompletePluginKey = exports.HOME_KEY = exports.ESCAPE_KEY = exports.ENTER_KEY = exports.END_KEY = exports.CMD_ARROW_UP_KEY = exports.CMD_ARROW_DOWN_KEY = exports.AUTOCOMPLETE_PLUGIN_NAME = exports.ARROW_UP_KEY = exports.ARROW_DOWN_KEY = void 0;
var _empty = require("rxjs/observable/empty");
var _state = require("@atlaskit/editor-prosemirror/state");
var AUTOCOMPLETE_PLUGIN_NAME = exports.AUTOCOMPLETE_PLUGIN_NAME = 'jql-autocomplete-plugin';
var ARROW_UP_KEY = exports.ARROW_UP_KEY = 'ArrowUp';
var ARROW_DOWN_KEY = exports.ARROW_DOWN_KEY = 'ArrowDown';
var CMD_ARROW_UP_KEY = exports.CMD_ARROW_UP_KEY = 'Cmd-ArrowUp';
var CMD_ARROW_DOWN_KEY = exports.CMD_ARROW_DOWN_KEY = 'Cmd-ArrowDown';
var HOME_KEY = exports.HOME_KEY = 'Home'; // === Fn-Left in Mac
var END_KEY = exports.END_KEY = 'End'; // === Fn-Right in Mac
var ENTER_KEY = exports.ENTER_KEY = 'Enter';
var ESCAPE_KEY = exports.ESCAPE_KEY = 'Escape';
var TAB_KEY = exports.TAB_KEY = 'Tab';

// No-op autocomplete provider used when creating the initial state
var defaultAutocompleteProvider = exports.defaultAutocompleteProvider = {
  onFields: function onFields() {
    return (0, _empty.empty)();
  },
  onOperators: function onOperators() {
    return (0, _empty.empty)();
  },
  onValues: function onValues() {
    return (0, _empty.empty)();
  },
  onFunctions: function onFunctions() {
    return (0, _empty.empty)();
  }
};
var JQLAutocompletePluginKey = exports.JQLAutocompletePluginKey = new _state.PluginKey(AUTOCOMPLETE_PLUGIN_NAME);
