// @atlaskit/editor-common/dist/cjs/styles/shared/embedCard.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.embedSpacingStyles = exports.embedCardStyles = void 0;
var _react = require("@emotion/react");
var _editorSharedStyles = require("@atlaskit/editor-shared-styles");
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var embedCardStyles = exports.embedCardStyles = function embedCardStyles() {
  return (0, _react.css)({
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
    '.ProseMirror': {
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
      ".embedCardView-content-wrap[layout^='wrap-']": {
        maxWidth: '100%',
        position: 'relative',
        // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values, @atlaskit/ui-styling-standard/no-imported-style-values
        zIndex: _editorSharedStyles.akEditorWrappedNodeZIndex
      },
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
      ".embedCardView-content-wrap[layout='wrap-left']": {
        float: 'left'
      },
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
      ".embedCardView-content-wrap[layout='wrap-right']": {
        float: 'right'
      },
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
      ".embedCardView-content-wrap[layout='wrap-right'] + .embedCardView-content-wrap[layout='wrap-left']": {
        clear: 'both'
      }
    }
  });
};

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var embedSpacingStyles = exports.embedSpacingStyles = (0, _react.css)({
  margin: "0 ".concat("var(--ds-space-150, 12px)")
});
