// @atlaskit/editor-common/dist/cjs/styles/shared/expand.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.expandIconWrapperStyle = exports.expandIconContainerStyle = exports.expandClassNames = void 0;
var _react = require("@emotion/react");
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var expandIconWrapperStyle = exports.expandIconWrapperStyle = (0, _react.css)({
  marginLeft: "var(--ds-space-negative-100, -8px)"
});

/**
 * @private
 * @deprecated use `import { ToolbarExpandIcon } from '@atlaskit/editor-common/ui';` instead
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var expandIconContainerStyle = exports.expandIconContainerStyle = (0, _react.css)({
  display: 'flex',
  alignItems: 'center'
});
var prefix = 'ak-editor-expand';
var expandClassNames = exports.expandClassNames = {
  prefix: prefix,
  expanded: "".concat(prefix, "__expanded"),
  titleContainer: "".concat(prefix, "__title-container"),
  inputContainer: "".concat(prefix, "__input-container"),
  iconContainer: "".concat(prefix, "__icon-container"),
  icon: "".concat(prefix, "__icon"),
  titleInput: "".concat(prefix, "__title-input"),
  content: "".concat(prefix, "__content"),
  type: function type(_type) {
    return "".concat(prefix, "__type-").concat(_type);
  }
};
