// @atlaskit/smart-card/dist/cjs/extractors/inline/extractPropsFromSourceCodePullRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var _16_1 = tslib_1.__importDefault(require("@atlaskit/icon-object/glyph/pull-request/16"));
var extractPropsFromObject_1 = require("./extractPropsFromObject");
var extractPropsFromSourceCodeCommon_1 = require("./extractPropsFromSourceCodeCommon");
exports.pullRequestStateToLozengeMap = {
    open: 'inprogress',
    merged: 'success',
    declined: 'removed',
    closed: 'success',
};
var buildInlineSourceCodePullRequestTag = function (json) {
    if (json['atlassian:state']) {
        var state = json['atlassian:state'].toLowerCase();
        return {
            lozenge: {
                appearance: exports.pullRequestStateToLozengeMap[state] || 'default',
                text: state,
            },
        };
    }
    return {};
};
exports.buildIcon = function (json) {
    var name = json.name;
    return { icon: react_1.default.createElement(_16_1.default, { label: name }) };
};
exports.extractInlineViewPropsFromSourceCodePullRequest = function (json) {
    var props = extractPropsFromObject_1.extractInlineViewPropsFromObject(json);
    return tslib_1.__assign(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({}, props), exports.buildIcon(json)), extractPropsFromSourceCodeCommon_1.buildName(props, json)), buildInlineSourceCodePullRequestTag(json));
};
