// @atlaskit/jql-editor/dist/es2019/plugins/autocomplete/components/autocomplete/index.js
import React, { useCallback } from 'react';
import { EventType } from '@atlaskit/jql-editor-common';
import { Action, ActionSubject } from '../../../../analytics';
import { useAutocompleteLoading, useAutocompleteOptions, useStoreActions } from '../../../../state';
import AutocompleteDropdown from '../autocomplete-dropdown';
const Autocomplete = ({
  keymap,
  onClick
}) => {
  const [options] = useAutocompleteOptions();
  const [loading] = useAutocompleteLoading();
  const [, {
    createAndFireAnalyticsEvent
  }] = useStoreActions();
  const handleClick = useCallback((option, analyticsAttributes) => {
    onClick(option);
    createAndFireAnalyticsEvent({
      action: Action.SELECTED,
      actionSubject: ActionSubject.AUTOCOMPLETE_OPTION,
      eventType: EventType.TRACK,
      attributes: {
        ...analyticsAttributes
      }
    });
  }, [createAndFireAnalyticsEvent, onClick]);
  return /*#__PURE__*/React.createElement(AutocompleteDropdown, {
    options: options,
    loading: loading,
    keymap: keymap,
    onClick: handleClick
  });
};
export default Autocomplete;
