// @atlaskit/editor-plugins/dist/cjs/block-type/ui/consts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "BLOCK_QUOTE", {
  enumerable: true,
  get: function get() {
    return _consts.BLOCK_QUOTE;
  }
});
Object.defineProperty(exports, "CODE_BLOCK", {
  enumerable: true,
  get: function get() {
    return _consts.CODE_BLOCK;
  }
});
Object.defineProperty(exports, "HEADING_1", {
  enumerable: true,
  get: function get() {
    return _consts.HEADING_1;
  }
});
Object.defineProperty(exports, "HEADING_2", {
  enumerable: true,
  get: function get() {
    return _consts.HEADING_2;
  }
});
Object.defineProperty(exports, "HEADING_3", {
  enumerable: true,
  get: function get() {
    return _consts.HEADING_3;
  }
});
Object.defineProperty(exports, "HEADING_4", {
  enumerable: true,
  get: function get() {
    return _consts.HEADING_4;
  }
});
Object.defineProperty(exports, "HEADING_5", {
  enumerable: true,
  get: function get() {
    return _consts.HEADING_5;
  }
});
Object.defineProperty(exports, "HEADING_6", {
  enumerable: true,
  get: function get() {
    return _consts.HEADING_6;
  }
});
Object.defineProperty(exports, "NORMAL_TEXT", {
  enumerable: true,
  get: function get() {
    return _consts.NORMAL_TEXT;
  }
});
Object.defineProperty(exports, "PANEL", {
  enumerable: true,
  get: function get() {
    return _consts.PANEL;
  }
});
var _consts = require("@atlaskit/editor-plugin-block-type/consts");
