// @atlaskit/jql-ast/dist/cjs/visitors/terminal-clause.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TerminalClauseVisitor = void 0;
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _creators = require("../creators");
var _utils = require("../utils");
var _common = require("./common");
var _field = require("./field");
var _operand = require("./operand");
var _operator = require("./operator");
var _predicate = require("./predicate");
function _createSuper(t) { var r = _isNativeReflectConstruct(); return function () { var e, o = (0, _getPrototypeOf2.default)(t); if (r) { var s = (0, _getPrototypeOf2.default)(this).constructor; e = Reflect.construct(o, arguments, s); } else e = o.apply(this, arguments); return (0, _possibleConstructorReturn2.default)(this, e); }; }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var TerminalClauseVisitor = exports.TerminalClauseVisitor = /*#__PURE__*/function (_JastBuildingVisitor) {
  (0, _inherits2.default)(TerminalClauseVisitor, _JastBuildingVisitor);
  var _super = _createSuper(TerminalClauseVisitor);
  function TerminalClauseVisitor() {
    var _this;
    (0, _classCallCheck2.default)(this, TerminalClauseVisitor);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "fieldVisitor", new _field.FieldVisitor(_this.tokens));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "terminalClauseRhsVisitor", new TerminalClauseRhsVisitor(_this.tokens));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "visitJqlTerminalClause", function (ctx) {
      var field = ctx.jqlField().accept(_this.fieldVisitor);
      var _ctx$jqlTerminalClaus = ctx.jqlTerminalClauseRhs().accept(_this.terminalClauseRhsVisitor),
        operator = _ctx$jqlTerminalClaus.operator,
        operand = _ctx$jqlTerminalClaus.operand,
        predicates = _ctx$jqlTerminalClaus.predicates;
      var position = (0, _common.getPositionFromContext)(ctx);
      return _creators.internalCreators.terminalClause(field, operator, operand, predicates, position);
    });
    return _this;
  }
  return (0, _createClass2.default)(TerminalClauseVisitor);
}(_common.JastBuildingVisitor);
var TerminalClauseRhsVisitor = /*#__PURE__*/function (_JastBuildingVisitor2) {
  (0, _inherits2.default)(TerminalClauseRhsVisitor, _JastBuildingVisitor2);
  var _super2 = _createSuper(TerminalClauseRhsVisitor);
  function TerminalClauseRhsVisitor() {
    var _this2;
    (0, _classCallCheck2.default)(this, TerminalClauseRhsVisitor);
    for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      args[_key2] = arguments[_key2];
    }
    _this2 = _super2.call.apply(_super2, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this2), "operatorVisitor", new _operator.OperatorVisitor(_this2.tokens));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this2), "terminalClauseOperandVisitor", new TerminalClauseOperandVisitor(_this2.tokens));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this2), "predicateVisitor", new _predicate.PredicateVisitor(_this2.tokens));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this2), "visitJqlEqualsClause", function (ctx) {
      var operator = ctx.jqlEqualsOperator().accept(_this2.operatorVisitor);
      var operand = ctx.accept(_this2.terminalClauseOperandVisitor);
      return {
        operator: operator,
        operand: operand,
        predicates: []
      };
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this2), "visitJqlLikeClause", function (ctx) {
      var operator = ctx.jqlLikeOperator().accept(_this2.operatorVisitor);
      var operand = ctx.accept(_this2.terminalClauseOperandVisitor);
      return {
        operator: operator,
        operand: operand,
        predicates: []
      };
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this2), "visitJqlComparisonClause", function (ctx) {
      var operator = ctx.jqlComparisonOperator().accept(_this2.operatorVisitor);
      var operand = ctx.accept(_this2.terminalClauseOperandVisitor);
      return {
        operator: operator,
        operand: operand,
        predicates: []
      };
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this2), "visitJqlInClause", function (ctx) {
      var operator = ctx.jqlInOperator().accept(_this2.operatorVisitor);
      var operand = ctx.accept(_this2.terminalClauseOperandVisitor);
      return {
        operator: operator,
        operand: operand,
        predicates: []
      };
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this2), "visitJqlIsClause", function (ctx) {
      var operator = ctx.jqlIsOperator().accept(_this2.operatorVisitor);
      var operand = ctx.accept(_this2.terminalClauseOperandVisitor);
      return {
        operator: operator,
        operand: operand,
        predicates: []
      };
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this2), "visitJqlWasClause", function (ctx) {
      var operator = ctx.jqlWasOperator().accept(_this2.operatorVisitor);
      var operand = ctx.accept(_this2.terminalClauseOperandVisitor);
      var predicates = ctx.jqlWasPredicate().map(function (predicate) {
        return predicate.accept(_this2.predicateVisitor);
      }).filter(_utils.notUndefined);
      return {
        operator: operator,
        operand: operand,
        predicates: predicates
      };
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this2), "visitJqlWasInClause", function (ctx) {
      var operator = ctx.jqlWasInOperator().accept(_this2.operatorVisitor);
      var operand = ctx.accept(_this2.terminalClauseOperandVisitor);
      var predicates = ctx.jqlWasPredicate().map(function (predicate) {
        return predicate.accept(_this2.predicateVisitor);
      }).filter(_utils.notUndefined);
      return {
        operator: operator,
        operand: operand,
        predicates: predicates
      };
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this2), "visitJqlChangedClause", function (ctx) {
      var operator = ctx.jqlChangedOperator().accept(_this2.operatorVisitor);
      var predicates = ctx.jqlChangedPredicate().map(function (predicate) {
        return predicate.accept(_this2.predicateVisitor);
      }).filter(_utils.notUndefined);
      return {
        operator: operator,
        operand: undefined,
        predicates: predicates
      };
    });
    // Recover from clause type ambiguities, e.g. "issuetype was " (which can be a WAS or WAS IN clause)
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this2), "visitChildren", function (node) {
      var operator = node.accept(_this2.operatorVisitor);
      return {
        operator: operator,
        operand: undefined,
        predicates: []
      };
    });
    return _this2;
  }
  return (0, _createClass2.default)(TerminalClauseRhsVisitor);
}(_common.JastBuildingVisitor);
var TerminalClauseOperandVisitor = /*#__PURE__*/function (_JastBuildingVisitor3) {
  (0, _inherits2.default)(TerminalClauseOperandVisitor, _JastBuildingVisitor3);
  var _super3 = _createSuper(TerminalClauseOperandVisitor);
  function TerminalClauseOperandVisitor() {
    var _this3;
    (0, _classCallCheck2.default)(this, TerminalClauseOperandVisitor);
    for (var _len3 = arguments.length, args = new Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
      args[_key3] = arguments[_key3];
    }
    _this3 = _super3.call.apply(_super3, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this3), "operandVisitor", new _operand.OperandVisitor(_this3.tokens));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this3), "visitJqlEqualsClause", function (ctx) {
      return _this3.getOperandForContexts([ctx.jqlEmpty(), ctx.jqlValue(), ctx.jqlFunction()]);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this3), "visitJqlLikeClause", function (ctx) {
      return _this3.getOperandForContexts([ctx.jqlEmpty(), ctx.jqlValue(), ctx.jqlFunction()]);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this3), "visitJqlComparisonClause", function (ctx) {
      return _this3.getOperandForContexts([ctx.jqlValue(), ctx.jqlFunction()]);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this3), "visitJqlInClause", function (ctx) {
      return _this3.getOperandForContexts([ctx.jqlList(), ctx.jqlFunction()]);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this3), "visitJqlIsClause", function (ctx) {
      return _this3.getOperandForContexts([ctx.jqlEmpty()]);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this3), "visitJqlWasClause", function (ctx) {
      return _this3.getOperandForContexts([ctx.jqlEmpty(), ctx.jqlValue(), ctx.jqlFunction()]);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this3), "visitJqlWasInClause", function (ctx) {
      return _this3.getOperandForContexts([ctx.jqlList(), ctx.jqlFunction()]);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this3), "getOperandForContexts", function (contexts) {
      var operandContext = contexts.find(_utils.notUndefined);
      return operandContext === null || operandContext === void 0 ? void 0 : operandContext.accept(_this3.operandVisitor);
    });
    return _this3;
  }
  return (0, _createClass2.default)(TerminalClauseOperandVisitor);
}(_common.JastBuildingVisitor);
