// @atlaskit/profilecard/dist/esm/components/team-profile-card/team-containers-skeleton/linked-container-card-skeleton/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import "./index.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
import React from 'react';
import { Box, Inline, Stack } from '@atlaskit/primitives/compiled';
var styles = {
  containerWrapperSkeleton: "_2rkoiti9 _1h6dz9xs _12y3e4h9 _12jimuej _1qu2nqa1 _syaz1fxt _ca0qu2gc _u5f3u2gc _n3tdu2gc _19bvu2gc",
  containerIconSkeleton: "_2rkoiti9 _4t3i1tcg _1bsb1tcg _ca0qu2gc _u5f3u2gc _n3tdu2gc _19bvu2gc",
  firstLineSkeleton: "_2rko1q77 _1rjc1b66",
  secondLineSkeleton: "_2rko1q77 _1rjc1b66 _1bsbqkoa",
  containerTypeIconSkeleton: "_2rkoiti9 _4t3i7vkz _1bsb7vkz _ca0qu2gc _u5f3u2gc _n3tdu2gc _19bvu2gc"
};
export var LinkedContainerCardSkeleton = function LinkedContainerCardSkeleton() {
  return /*#__PURE__*/React.createElement(Inline, {
    space: "space.100",
    xcss: styles.containerWrapperSkeleton
  }, /*#__PURE__*/React.createElement(Box, {
    xcss: styles.containerIconSkeleton,
    backgroundColor: "color.background.neutral"
  }), /*#__PURE__*/React.createElement(Stack, {
    grow: "fill",
    space: "space.050"
  }, /*#__PURE__*/React.createElement(Box, {
    xcss: styles.firstLineSkeleton,
    backgroundColor: "color.background.neutral"
  }), /*#__PURE__*/React.createElement(Box, {
    xcss: styles.secondLineSkeleton,
    backgroundColor: "color.background.neutral"
  })), /*#__PURE__*/React.createElement(Box, {
    xcss: styles.containerTypeIconSkeleton,
    backgroundColor: "color.background.neutral"
  }));
};
