// @atlaskit/link-datasource/dist/es2019/ui/assets-modal/search-container/aql-search-input/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./index.compiled.css";
import { forwardRef } from 'react';
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { Fragment } from 'react';
import { useIntl } from 'react-intl-next';
import { LoadingButton } from '@atlaskit/button';
import { ErrorMessage, Field } from '@atlaskit/form';
import CrossCircleIcon from '@atlaskit/icon/core/migration/cross-circle';
import QuestionCircleIcon from '@atlaskit/icon/core/migration/question-circle';
import SearchIcon from '@atlaskit/icon/core/migration/search--editor-search';
import CheckCircleIcon from '@atlaskit/icon/core/migration/success--check-circle';
import Spinner from '@atlaskit/spinner';
import Textfield from '@atlaskit/textfield';
import { G300, N500, R400 } from '@atlaskit/theme/colors';
import Tooltip from '@atlaskit/tooltip';
import { useValidateAqlText } from '../../../../hooks/useValidateAqlText';
import { aqlKey } from '../../../../types/assets/types';
import { searchInputMessages } from './messages';

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled
const FieldContainer = forwardRef(({
  as: C = "div",
  style: __cmpls,
  ...__cmplp
}, __cmplr) => {
  return /*#__PURE__*/React.createElement(C, _extends({}, __cmplp, {
    style: __cmpls,
    ref: __cmplr,
    className: ax(["_16jlkb7n _1o9zkb7n _i0dlf1ug _19pkx0bf", __cmplp.className])
  }));
});
if (process.env.NODE_ENV !== 'production') {
  FieldContainer.displayName = 'FieldContainer';
}
const buttonBaseStyles = null;
const AQLSupportDocumentLink = 'https://support.atlassian.com/jira-service-management-cloud/docs/use-assets-query-language-aql/';
const searchButtonStyles = null;
const renderValidatorIcon = lastValidationResult => {
  if (lastValidationResult.type === 'loading') {
    return /*#__PURE__*/React.createElement(Spinner, {
      size: "medium",
      testId: "assets-datasource-modal--aql-validating"
    });
  }
  if (lastValidationResult.type === 'invalid') {
    return /*#__PURE__*/React.createElement(CrossCircleIcon, {
      label: "label",
      color: `var(--ds-icon-danger, ${R400})`,
      LEGACY_size: "medium",
      testId: "assets-datasource-modal--aql-invalid",
      spacing: "spacious"
    });
  }
  if (lastValidationResult.type === 'valid') {
    return /*#__PURE__*/React.createElement(CheckCircleIcon, {
      label: "label",
      color: `var(--ds-icon-success, ${G300})`,
      LEGACY_size: "medium",
      testId: "assets-datasource-modal--aql-valid",
      spacing: "spacious"
    });
  }
  return /*#__PURE__*/React.createElement(SearchIcon, {
    label: "label",
    LEGACY_size: "medium",
    testId: "assets-datasource-modal--aql-idle",
    color: "currentColor",
    spacing: "spacious"
  });
};
export const AqlSearchInput = ({
  value,
  workspaceId,
  testId = 'assets-datasource-modal--aql-search-input',
  isSearching
}) => {
  const {
    formatMessage
  } = useIntl();
  const {
    debouncedValidation,
    lastValidationResult
  } = useValidateAqlText(workspaceId, value);
  return /*#__PURE__*/React.createElement(FieldContainer, null, /*#__PURE__*/React.createElement(Field, {
    name: aqlKey,
    defaultValue: value,
    validate: debouncedValidation
  }, ({
    fieldProps
  }) => /*#__PURE__*/React.createElement(Fragment, null, /*#__PURE__*/React.createElement(Textfield, _extends({}, fieldProps, {
    elemBeforeInput: /*#__PURE__*/React.createElement("span", {
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
      style: {
        paddingLeft: "var(--ds-space-075, 6px)",
        width: 24,
        display: 'flex'
      }
    }, renderValidatorIcon(lastValidationResult)),
    elemAfterInput: /*#__PURE__*/React.createElement(Fragment, null, /*#__PURE__*/React.createElement(Tooltip, {
      content: formatMessage(searchInputMessages.helpTooltipText),
      position: "bottom"
    }, /*#__PURE__*/React.createElement("a", {
      href: AQLSupportDocumentLink,
      target: "_blank",
      className: ax(["_1e0c1txw _4t3i1osq _kqswh2mm _4cvr1h6o _1bah1h6o _2lx21bp4 _2hwx1f3t"])
    }, /*#__PURE__*/React.createElement(QuestionCircleIcon, {
      label: "label",
      color: `var(--ds-icon, ${N500})`,
      LEGACY_size: "medium",
      testId: "assets-datasource-modal-help",
      spacing: "spacious"
    }))), /*#__PURE__*/React.createElement(LoadingButton, {
      appearance: "primary",
      iconBefore: /*#__PURE__*/React.createElement(SearchIcon, {
        label: formatMessage(searchInputMessages.placeholder),
        LEGACY_size: "medium",
        color: "currentColor",
        spacing: "spacious"
      }),
      isLoading: isSearching,
      spacing: "none",
      testId: "assets-datasource-modal--aql-search-button",
      type: "submit",
      isDisabled: lastValidationResult.type !== 'valid',
      className: ax(["_2hwx12x7"])
    })),
    placeholder: formatMessage(searchInputMessages.placeholder),
    testId: testId
  })), lastValidationResult.type === 'invalid' && lastValidationResult.error && /*#__PURE__*/React.createElement(ErrorMessage, null, lastValidationResult.error))));
};
