// @atlaskit/editor-plugin-extension/dist/esm/ui/ConfigPanel/NestedForms/RemovableField.js
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
import React from 'react';

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import { css, jsx } from '@emotion/react';
import { injectIntl } from 'react-intl-next';
import { configPanelMessages as messages } from '@atlaskit/editor-common/extensions';
import CrossCircleIcon from '@atlaskit/icon/core/migration/cross-circle';
import { N80, R300 } from '@atlaskit/theme/colors';
import Tooltip from '@atlaskit/tooltip';
var removableFieldWrapperStyles = css({
  position: 'relative',
  marginBottom: 0
});
var wrapperWithMarginBottomStyles = css({
  marginBottom: "var(--ds-space-200, 16px)"
});
var removeButtonWrapperStyles = css({
  position: 'absolute',
  right: 0,
  top: 0,
  cursor: 'pointer',
  color: "var(--ds-icon-subtle, ".concat(N80, ")"),
  '&:hover': {
    color: "var(--ds-icon-danger, ".concat(R300, ")")
  }
});
var RemovableField = function RemovableField(_ref) {
  var _children$props$field;
  var name = _ref.name,
    canRemoveField = _ref.canRemoveField,
    onClickRemove = _ref.onClickRemove,
    children = _ref.children,
    intl = _ref.intl,
    className = _ref.className;
  var onClickCallback = React.useCallback(function () {
    return onClickRemove && onClickRemove(name);
  }, [name, onClickRemove]);
  var hasMarginBottom = ((_children$props$field = children.props.field) === null || _children$props$field === void 0 ? void 0 : _children$props$field.type) !== 'expand';
  return jsx("div", {
    css: [removableFieldWrapperStyles, hasMarginBottom && wrapperWithMarginBottomStyles]
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    ,
    className: className
  }, children, canRemoveField &&
  // eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions
  jsx("div", {
    css: removeButtonWrapperStyles,
    "data-testid": "remove-field-".concat(name),
    onClick: onClickCallback
  }, jsx(Tooltip, {
    content: intl.formatMessage(messages.removeField),
    position: "left"
  }, jsx(CrossCircleIcon, {
    LEGACY_size: "small",
    spacing: "none",
    label: intl.formatMessage(messages.removeField)
  }))));
};
export default injectIntl(RemovableField);
