// @atlaskit/smart-card/dist/cjs/view/RelatedLinksModal/components/related-link-item/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./index.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _constants = require("../../../../constants");
var _index = require("../../../../index");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var hoverStyleOld = null;
var hoverStyle = null;
var baseStyle = null;
var selectedStyle = null;
var relatedLinkItemStyles = null;
var RelatedLinkItem = function RelatedLinkItem(_ref) {
  var url = _ref.url,
    testId = _ref.testId,
    isSelected = _ref.isSelected,
    onFocus = _ref.onFocus;
  var subtitle = [{
    name: _index.ElementName.Provider,
    hideIcon: true
  }];
  var ui = {
    hideElevation: true,
    hidePadding: true,
    hideBackground: true,
    theme: _index.SmartLinkTheme.Black,
    clickableContainer: true,
    size: (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v2') ? _index.SmartLinkSize.Medium : _index.SmartLinkSize.Large
  };
  return /*#__PURE__*/React.createElement("div", {
    "data-testId": testId,
    onFocus: onFocus,
    className: (0, _runtime.ax)([(0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v2') && "_195g10v4 _18zr1ejb", (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v2') ? "_irr3166n _1di61dty" : "_8vu4iti9 _1mxox0bf _10swu2gc _irr31d5g", isSelected && (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v2') && "_bfhkfg4m _16qsd1ld _irr3i1yw _1di619ru"]),
    style: {
      "--_g94mg3": (0, _runtime.ix)("inset 2px 0 0 0 ".concat("var(--ds-border-selected, #0C66E4)"))
    }
  }, /*#__PURE__*/React.createElement(_index.Card, {
    appearance: "block",
    ui: ui,
    url: url
  }, /*#__PURE__*/React.createElement(_index.TitleBlock, {
    maxLines: 1,
    hideTitleTooltip: true,
    position: (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v2') ? _index.SmartLinkPosition.Top : _index.SmartLinkPosition.Center,
    subtitle: subtitle,
    anchorTarget: "_blank",
    size: ui.size,
    theme: ui.theme,
    status: _constants.SmartLinkStatus.Resolved,
    className: (0, _runtime.ax)(["_zulputpp _11c81o8v _ca0qu2gc _n3tdu2gc"])
  })));
};
var _default = exports.default = RelatedLinkItem;
