// @atlaskit/link-datasource/dist/esm/ui/common/initial-state-view/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import "./index.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { FormattedMessage, useIntl } from 'react-intl-next';
import Link from '@atlaskit/link';
import Lozenge from '@atlaskit/lozenge';
import { fg } from '@atlaskit/platform-feature-flags';
import { N300 } from '@atlaskit/theme/colors';
import { fontFallback } from '@atlaskit/theme/typography';
import { initialStateViewMessages } from './messages';
var initialStateViewContainerStyles = null;
var svgAndTextsWrapperStyles = null;
var betaTagStyles = null;
var searchTitleStylesOld = null;
var searchTitleStyles = null;
var mainTextStylesOld = null;
var mainTextStyles = null;
var learnMoreLinkStyles = null;
export var InitialStateView = function InitialStateView(_ref) {
  var icon = _ref.icon,
    _ref$showBeta = _ref.showBeta,
    showBeta = _ref$showBeta === void 0 ? false : _ref$showBeta,
    title = _ref.title,
    description = _ref.description,
    learnMoreLink = _ref.learnMoreLink;
  var _useIntl = useIntl(),
    formatMessage = _useIntl.formatMessage;
  return /*#__PURE__*/React.createElement("div", {
    "data-testid": "datasource-modal--initial-state-view",
    className: ax(["_1e0c1txw _1bah1h6o _4t3i1osq"])
  }, /*#__PURE__*/React.createElement("div", {
    className: ax(["_y3gn1h6o _1wpz1h6o _ca0qoahv _n3tdoahv"])
  }, icon, /*#__PURE__*/React.createElement("div", {
    className: ax([fg('platform-linking-visual-refresh-sllv') ? "_11c8lodh _zulpu2gc _syaz1fxt _ca0q1ejb _n3tdpxbi _1e0c1txw _1bah1h6o _4cvr1h6o" : "_zulpu2gc _syazg0pi _ca0qpxbi _n3tdu2gc _1e0c1txw _1bah1h6o _4cvr1h6o _wtt4ykg7 _pfzw85h9 _zzcc1umn _x3fx1hgc _v2gl1yp8 _1y0bdipb _z7oy1jq3"]),
    style: {
      "--_1dzgjqk": ix("var(--ds-text-subtlest, ".concat(N300, ")"))
    }
  }, showBeta && /*#__PURE__*/React.createElement("div", {
    className: ax(["_1e0c1txw"])
  }, /*#__PURE__*/React.createElement(Lozenge, {
    appearance: "new"
  }, /*#__PURE__*/React.createElement(FormattedMessage, initialStateViewMessages.beta))), formatMessage(title)), /*#__PURE__*/React.createElement("div", {
    className: ax([fg('platform-linking-visual-refresh-sllv') ? "_syaz1gjq" : "_syazg0pi"]),
    style: {
      "--_1dzgjqk": ix("var(--ds-text-subtlest, ".concat(N300, ")"))
    }
  }, formatMessage(description)), learnMoreLink ? fg('platform-linking-visual-refresh-sllv') ? /*#__PURE__*/React.createElement(Link, {
    href: learnMoreLink.href,
    target: "_blank"
  }, formatMessage(learnMoreLink.text)) : /*#__PURE__*/React.createElement("a", {
    href: learnMoreLink.href,
    target: "_blank",
    className: ax(["_ca0qpxbi _1e0c1o8l"])
  }, formatMessage(learnMoreLink.text)) : null));
};
