// @atlaskit/media-client/dist/esm/utils/with-media-client-hoc.js
import { __assign, __extends, __rest } from "tslib";
import React from 'react';
import { MediaClient } from '../client/media-client';
var mediaClientsMap = new Map();
export var getMediaClient = function (mediaClientConfig) {
    var mediaClient = mediaClientsMap.get(mediaClientConfig);
    if (!mediaClient) {
        mediaClient = new MediaClient(mediaClientConfig);
        mediaClientsMap.set(mediaClientConfig, mediaClient);
    }
    return mediaClient;
};
var createEmptyMediaClient = function () {
    var emptyConfig = {
        authProvider: function () {
            return Promise.resolve({
                clientId: '',
                token: '',
                baseUrl: '',
            });
        },
    };
    return new MediaClient(emptyConfig);
};
export var withMediaClient = function (Component) {
    return /** @class */ (function (_super) {
        __extends(class_1, _super);
        function class_1() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        class_1.prototype.render = function () {
            // TODO MPT-315: clean up after we move mediaClientConfig into FileIdentifier
            var _a = this.props, mediaClientConfig = _a.mediaClientConfig, otherProps = __rest(_a, ["mediaClientConfig"]);
            var mediaClient = !mediaClientConfig
                ? createEmptyMediaClient()
                : getMediaClient(mediaClientConfig);
            return React.createElement(Component, __assign({}, otherProps, { mediaClient: mediaClient }));
        };
        return class_1;
    }(React.Component));
};
