// @atlaskit/smart-card/dist/es2019/view/BlockCard/views/unresolved-view/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./index.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { fg } from '@atlaskit/platform-feature-flags';
import FlexibleCard from '../../../FlexibleCard';
import { CustomBlock, InternalFooterBlock, PreviewBlock, TitleBlock } from '../../../FlexibleCard/components/blocks';
import { FlexibleCardUiOptions, PreviewBlockOptions, titleBlockOptions } from '../utils';
const customBlockStylesOld = null;
const customBlockStyles = null;
const titleBlockCssOld = null;
const titleBlockCss = null;
const footerBlockCssOld = null;
const footerBlockCss = null;
const UnresolvedView = ({
  actions,
  cardState,
  children,
  onAuthorize,
  onClick,
  onError,
  showPreview = false,
  testId,
  title,
  url
}) => /*#__PURE__*/React.createElement(FlexibleCard, {
  appearance: "block",
  cardState: cardState,
  onAuthorize: onAuthorize,
  onClick: onClick,
  onError: onError,
  origin: "smartLinkCard",
  testId: testId,
  ui: FlexibleCardUiOptions,
  url: url
}, /*#__PURE__*/React.createElement(TitleBlock, _extends({}, titleBlockOptions, {
  hideIcon: !!title,
  text: title,
  status: cardState.status,
  className: ax([fg('platform-linking-visual-refresh-v1') ? "_zulpu2gc _13mh1pd9 _2yhz14q2" : "_zulp1f3t _13mh1pd9"])
})), /*#__PURE__*/React.createElement(CustomBlock, {
  status: cardState.status,
  className: ax([fg('platform-linking-visual-refresh-v1') ? "_zulp1b66 _1e0c1txw _4cvr1y6m _1wpz1fhb _7yjtidpf _ngc01fxt" : "_4cvr1y6m"])
}, children), showPreview && /*#__PURE__*/React.createElement(PreviewBlock, _extends({}, PreviewBlockOptions, {
  status: cardState.status
})), /*#__PURE__*/React.createElement(InternalFooterBlock, {
  actions: actions,
  testId: "smart-block-card-footer",
  status: cardState.status,
  className: ax([fg('platform-linking-visual-refresh-v1') ? "_4t3i1k8s _1e0c1txw _1bah1yb4 _4cvresu3 _1wpz1fhb" : "_4t3i1k8s _g0eioxwj _1qryoxwj _sxbroxwj _1gyioxwj"])
}));
export default UnresolvedView;
