// @atlaskit/mention/dist/cjs/components/MentionListError/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _react = _interopRequireDefault(require("react"));
var _heading = _interopRequireDefault(require("@atlaskit/heading"));
var _primitives = require("@atlaskit/primitives");
var _i18n = require("../../util/i18n");
var _GenericErrorIllustration = require("./GenericErrorIllustration");
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var advisedActionMessages = {
  '401': _i18n.LoginAgain,
  '403': _i18n.DifferentText,
  default: _i18n.DefaultAdvisedAction
};
var MentionListError = exports.default = /*#__PURE__*/function (_React$PureComponent) {
  function MentionListError() {
    (0, _classCallCheck2.default)(this, MentionListError);
    return _callSuper(this, MentionListError, arguments);
  }
  (0, _inherits2.default)(MentionListError, _React$PureComponent);
  return (0, _createClass2.default)(MentionListError, [{
    key: "render",
    value: function render() {
      var error = this.props.error;
      var ErrorMessage = MentionListError.getAdvisedActionMessage(error);
      return /*#__PURE__*/_react.default.createElement(_primitives.Stack, {
        space: "space.400",
        alignBlock: "center",
        alignInline: "center",
        xcss: mentionListErrorStyles
      }, /*#__PURE__*/_react.default.createElement(_GenericErrorIllustration.GenericErrorIllustration, null), /*#__PURE__*/_react.default.createElement(_primitives.Stack, {
        space: "space.100"
      }, /*#__PURE__*/_react.default.createElement(_heading.default, {
        size: "xsmall",
        as: "div"
      }, /*#__PURE__*/_react.default.createElement(_i18n.DefaultHeadline, null)), /*#__PURE__*/_react.default.createElement(_primitives.Text, {
        as: "p",
        color: "color.text.subtle"
      }, /*#__PURE__*/_react.default.createElement(ErrorMessage, null))));
    }
  }], [{
    key: "getAdvisedActionMessage",
    value:
    /**
     * Translate the supplied Error into a message suitable for display in the MentionList.
     *
     * @param error the error to be displayed
     */
    function getAdvisedActionMessage(error) {
      if (error && error.hasOwnProperty('statusCode')) {
        var httpError = error;
        return advisedActionMessages[httpError.statusCode.toString()] || advisedActionMessages.default;
      }
      return advisedActionMessages.default;
    }
  }]);
}(_react.default.PureComponent);
var mentionListErrorStyles = (0, _primitives.xcss)({
  paddingBlock: 'space.500',
  textAlign: 'center'
});
