// @atlaskit/editor-common/dist/cjs/styles/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AnnotationSharedCSSByState", {
  enumerable: true,
  get: function get() {
    return _annotation.AnnotationSharedCSSByState;
  }
});
Object.defineProperty(exports, "AnnotationSharedClassNames", {
  enumerable: true,
  get: function get() {
    return _annotation.AnnotationSharedClassNames;
  }
});
Object.defineProperty(exports, "BODIED_EXT_MBE_MARGIN_TOP", {
  enumerable: true,
  get: function get() {
    return _extension.BODIED_EXT_MBE_MARGIN_TOP;
  }
});
Object.defineProperty(exports, "BODIED_EXT_PADDING", {
  enumerable: true,
  get: function get() {
    return _extension.BODIED_EXT_PADDING;
  }
});
Object.defineProperty(exports, "BlockAnnotationSharedClassNames", {
  enumerable: true,
  get: function get() {
    return _annotation.BlockAnnotationSharedClassNames;
  }
});
Object.defineProperty(exports, "BreakoutCssClassName", {
  enumerable: true,
  get: function get() {
    return _breakout.BreakoutCssClassName;
  }
});
Object.defineProperty(exports, "CodeBlockSharedCssClassName", {
  enumerable: true,
  get: function get() {
    return _codeBlock.CodeBlockSharedCssClassName;
  }
});
Object.defineProperty(exports, "DATASOURCE_INNER_CONTAINER_CLASSNAME", {
  enumerable: true,
  get: function get() {
    return _smartCard2.DATASOURCE_INNER_CONTAINER_CLASSNAME;
  }
});
Object.defineProperty(exports, "DateSharedCssClassName", {
  enumerable: true,
  get: function get() {
    return _date.DateSharedCssClassName;
  }
});
Object.defineProperty(exports, "DropdownMenuSharedCssClassName", {
  enumerable: true,
  get: function get() {
    return _dropdownMenu.DropdownMenuSharedCssClassName;
  }
});
Object.defineProperty(exports, "EXTENSION_PADDING", {
  enumerable: true,
  get: function get() {
    return _extension.EXTENSION_PADDING;
  }
});
Object.defineProperty(exports, "FLOATING_TOOLBAR_LINKPICKER_CLASSNAME", {
  enumerable: true,
  get: function get() {
    return _smartCard2.FLOATING_TOOLBAR_LINKPICKER_CLASSNAME;
  }
});
Object.defineProperty(exports, "GRID_GUTTER", {
  enumerable: true,
  get: function get() {
    return _grid.GRID_GUTTER;
  }
});
Object.defineProperty(exports, "LAYOUT_COLUMN_PADDING", {
  enumerable: true,
  get: function get() {
    return _layout.LAYOUT_COLUMN_PADDING;
  }
});
Object.defineProperty(exports, "LAYOUT_SECTION_MARGIN", {
  enumerable: true,
  get: function get() {
    return _layout.LAYOUT_SECTION_MARGIN;
  }
});
Object.defineProperty(exports, "MediaSharedClassNames", {
  enumerable: true,
  get: function get() {
    return _media.ClassNames;
  }
});
Object.defineProperty(exports, "SmartCardSharedCssClassName", {
  enumerable: true,
  get: function get() {
    return _smartCard.SmartCardSharedCssClassName;
  }
});
Object.defineProperty(exports, "StatusSharedCssClassName", {
  enumerable: true,
  get: function get() {
    return _status.StatusSharedCssClassName;
  }
});
Object.defineProperty(exports, "TableSharedCssClassName", {
  enumerable: true,
  get: function get() {
    return _table.TableSharedCssClassName;
  }
});
Object.defineProperty(exports, "TaskDecisionSharedCssClassName", {
  enumerable: true,
  get: function get() {
    return _taskDecision.TaskDecisionSharedCssClassName;
  }
});
Object.defineProperty(exports, "UnsupportedSharedCssClassName", {
  enumerable: true,
  get: function get() {
    return _unsupportedContent.UnsupportedSharedCssClassName;
  }
});
Object.defineProperty(exports, "annotationSharedStyles", {
  enumerable: true,
  get: function get() {
    return _annotation.annotationSharedStyles;
  }
});
Object.defineProperty(exports, "backgroundColorStyles", {
  enumerable: true,
  get: function get() {
    return _backgroundColor.backgroundColorStyles;
  }
});
Object.defineProperty(exports, "blockMarksSharedStyles", {
  enumerable: true,
  get: function get() {
    return _blockMarks.blockMarksSharedStyles;
  }
});
Object.defineProperty(exports, "blockquoteSharedStyles", {
  enumerable: true,
  get: function get() {
    return _blockquote.blockquoteSharedStyles;
  }
});
Object.defineProperty(exports, "buttonGroupStyle", {
  enumerable: true,
  get: function get() {
    return _plugins.buttonGroupStyle;
  }
});
Object.defineProperty(exports, "buttonGroupStyleBeforeVisualRefresh", {
  enumerable: true,
  get: function get() {
    return _plugins.buttonGroupStyleBeforeVisualRefresh;
  }
});
Object.defineProperty(exports, "calcTableWidth", {
  enumerable: true,
  get: function get() {
    return _table.calcTableWidth;
  }
});
Object.defineProperty(exports, "codeBlockInListSafariFix", {
  enumerable: true,
  get: function get() {
    return _codeBlock.codeBlockInListSafariFix;
  }
});
Object.defineProperty(exports, "codeBlockSharedStyles", {
  enumerable: true,
  get: function get() {
    return _codeBlock.codeBlockSharedStyles;
  }
});
Object.defineProperty(exports, "codeMarkSharedStyles", {
  enumerable: true,
  get: function get() {
    return _codeMark.codeMarkSharedStyles;
  }
});
Object.defineProperty(exports, "columnLayoutResponsiveRendererStyles", {
  enumerable: true,
  get: function get() {
    return _columnLayout.columnLayoutResponsiveRendererStyles;
  }
});
Object.defineProperty(exports, "columnLayoutResponsiveSharedStyle", {
  enumerable: true,
  get: function get() {
    return _columnLayout.columnLayoutResponsiveSharedStyle;
  }
});
Object.defineProperty(exports, "columnLayoutSharedStyle", {
  enumerable: true,
  get: function get() {
    return _columnLayout.columnLayoutSharedStyle;
  }
});
Object.defineProperty(exports, "dateSharedStyle", {
  enumerable: true,
  get: function get() {
    return _date.dateSharedStyle;
  }
});
Object.defineProperty(exports, "disableBlueBorderStyles", {
  enumerable: true,
  get: function get() {
    return _plugins.disableBlueBorderStyles;
  }
});
Object.defineProperty(exports, "embedCardStyles", {
  enumerable: true,
  get: function get() {
    return _embedCard.embedCardStyles;
  }
});
Object.defineProperty(exports, "embedSpacingStyles", {
  enumerable: true,
  get: function get() {
    return _embedCard.embedSpacingStyles;
  }
});
Object.defineProperty(exports, "expandClassNames", {
  enumerable: true,
  get: function get() {
    return _expand.expandClassNames;
  }
});
Object.defineProperty(exports, "expandIconContainerStyle", {
  enumerable: true,
  get: function get() {
    return _expand.expandIconContainerStyle;
  }
});
Object.defineProperty(exports, "expandIconWrapperStyle", {
  enumerable: true,
  get: function get() {
    return _expand.expandIconWrapperStyle;
  }
});
Object.defineProperty(exports, "getOrderedListInlineStyles", {
  enumerable: true,
  get: function get() {
    return _lists.getOrderedListInlineStyles;
  }
});
Object.defineProperty(exports, "getSmartCardSharedStyles", {
  enumerable: true,
  get: function get() {
    return _smartCard.getSmartCardSharedStyles;
  }
});
Object.defineProperty(exports, "gridStyles", {
  enumerable: true,
  get: function get() {
    return _grid.gridStyles;
  }
});
Object.defineProperty(exports, "headingsSharedStyles", {
  enumerable: true,
  get: function get() {
    return _headings.headingsSharedStyles;
  }
});
Object.defineProperty(exports, "indentationSharedStyles", {
  enumerable: true,
  get: function get() {
    return _indentation.indentationSharedStyles;
  }
});
Object.defineProperty(exports, "linkSharedStyle", {
  enumerable: true,
  get: function get() {
    return _link.linkSharedStyle;
  }
});
Object.defineProperty(exports, "listItemCounterPadding", {
  enumerable: true,
  get: function get() {
    return _lists.listItemCounterPadding;
  }
});
Object.defineProperty(exports, "listsSharedStyles", {
  enumerable: true,
  get: function get() {
    return _lists.listsSharedStyles;
  }
});
Object.defineProperty(exports, "mediaSingleSharedStyle", {
  enumerable: true,
  get: function get() {
    return _mediaSingle.mediaSingleSharedStyle;
  }
});
Object.defineProperty(exports, "mediaSingleSharedStyleNew", {
  enumerable: true,
  get: function get() {
    return _mediaSingle.mediaSingleSharedStyleNew;
  }
});
Object.defineProperty(exports, "paragraphSharedStyles", {
  enumerable: true,
  get: function get() {
    return _paragraph.paragraphSharedStyles;
  }
});
Object.defineProperty(exports, "resizerHandleClassName", {
  enumerable: true,
  get: function get() {
    return _resizer.resizerHandleClassName;
  }
});
Object.defineProperty(exports, "resizerHandleThumbClassName", {
  enumerable: true,
  get: function get() {
    return _resizer.resizerHandleThumbClassName;
  }
});
Object.defineProperty(exports, "resizerHandleTrackClassName", {
  enumerable: true,
  get: function get() {
    return _resizer.resizerHandleTrackClassName;
  }
});
Object.defineProperty(exports, "resizerStyles", {
  enumerable: true,
  get: function get() {
    return _resizer.resizerStyles;
  }
});
Object.defineProperty(exports, "richMediaClassName", {
  enumerable: true,
  get: function get() {
    return _mediaSingle.richMediaClassName;
  }
});
Object.defineProperty(exports, "ruleSharedStyles", {
  enumerable: true,
  get: function get() {
    return _rule.ruleSharedStyles;
  }
});
Object.defineProperty(exports, "separatorStyles", {
  enumerable: true,
  get: function get() {
    return _plugins.separatorStyles;
  }
});
Object.defineProperty(exports, "shadowSharedStyle", {
  enumerable: true,
  get: function get() {
    return _shadow.shadowSharedStyle;
  }
});
Object.defineProperty(exports, "sharedExpandStyles", {
  enumerable: true,
  get: function get() {
    return _sharedStyles.sharedExpandStyles;
  }
});
Object.defineProperty(exports, "smartCardSharedStyles", {
  enumerable: true,
  get: function get() {
    return _smartCard.smartCardSharedStyles;
  }
});
Object.defineProperty(exports, "smartCardStyles", {
  enumerable: true,
  get: function get() {
    return _smartCard2.smartCardStyles;
  }
});
Object.defineProperty(exports, "tableCellBorderWidth", {
  enumerable: true,
  get: function get() {
    return _table.tableCellBorderWidth;
  }
});
Object.defineProperty(exports, "tableCellMinWidth", {
  enumerable: true,
  get: function get() {
    return _table.tableCellMinWidth;
  }
});
Object.defineProperty(exports, "tableCellPadding", {
  enumerable: true,
  get: function get() {
    return _table.tableCellPadding;
  }
});
Object.defineProperty(exports, "tableControlsSpacing", {
  enumerable: true,
  get: function get() {
    return _table.tableControlsSpacing;
  }
});
Object.defineProperty(exports, "tableMarginBottom", {
  enumerable: true,
  get: function get() {
    return _table.tableMarginBottom;
  }
});
Object.defineProperty(exports, "tableMarginSides", {
  enumerable: true,
  get: function get() {
    return _table.tableMarginSides;
  }
});
Object.defineProperty(exports, "tableMarginTop", {
  enumerable: true,
  get: function get() {
    return _table.tableMarginTop;
  }
});
Object.defineProperty(exports, "tableMarginTopWithControl", {
  enumerable: true,
  get: function get() {
    return _table.tableMarginTopWithControl;
  }
});
Object.defineProperty(exports, "tableNewColumnMinWidth", {
  enumerable: true,
  get: function get() {
    return _table.tableNewColumnMinWidth;
  }
});
Object.defineProperty(exports, "tablePadding", {
  enumerable: true,
  get: function get() {
    return _table.tablePadding;
  }
});
Object.defineProperty(exports, "tableResizeHandleWidth", {
  enumerable: true,
  get: function get() {
    return _table.tableResizeHandleWidth;
  }
});
Object.defineProperty(exports, "tableSharedStyle", {
  enumerable: true,
  get: function get() {
    return _table.tableSharedStyle;
  }
});
Object.defineProperty(exports, "tasksAndDecisionsStyles", {
  enumerable: true,
  get: function get() {
    return _taskDecision.tasksAndDecisionsStyles;
  }
});
Object.defineProperty(exports, "textColorStyles", {
  enumerable: true,
  get: function get() {
    return _textColor.textColorStyles;
  }
});
Object.defineProperty(exports, "triggerWrapperStyles", {
  enumerable: true,
  get: function get() {
    return _plugins.triggerWrapperStyles;
  }
});
Object.defineProperty(exports, "triggerWrapperStylesWithPadding", {
  enumerable: true,
  get: function get() {
    return _plugins.triggerWrapperStylesWithPadding;
  }
});
Object.defineProperty(exports, "unsupportedStyles", {
  enumerable: true,
  get: function get() {
    return _unsupportedContent.unsupportedStyles;
  }
});
Object.defineProperty(exports, "whitespaceSharedStyles", {
  enumerable: true,
  get: function get() {
    return _whitespace.whitespaceSharedStyles;
  }
});
Object.defineProperty(exports, "wrapperStyle", {
  enumerable: true,
  get: function get() {
    return _plugins.wrapperStyle;
  }
});
var _textColor = require("./shared/text-color");
var _backgroundColor = require("./shared/background-color");
var _table = require("./shared/table");
var _annotation = require("./shared/annotation");
var _columnLayout = require("./shared/column-layout");
var _mediaSingle = require("./shared/media-single");
var _blockquote = require("./shared/blockquote");
var _headings = require("./shared/headings");
var _rule = require("./shared/rule");
var _whitespace = require("./shared/whitespace");
var _paragraph = require("./shared/paragraph");
var _link = require("./shared/link");
var _lists = require("./shared/lists");
var _indentation = require("./shared/indentation");
var _blockMarks = require("./shared/block-marks");
var _codeMark = require("./shared/code-mark");
var _shadow = require("./shared/shadow");
var _date = require("./shared/date");
var _taskDecision = require("./shared/task-decision");
var _status = require("./shared/status");
var _smartCard = require("./shared/smart-card");
var _dropdownMenu = require("./shared/dropdown-menu");
var _codeBlock = require("./shared/code-block");
var _layout = require("./shared/layout");
var _extension = require("./shared/extension");
var _resizer = require("./shared/resizer");
var _grid = require("./shared/grid");
var _smartCard2 = require("./shared/smartCard");
var _embedCard = require("./shared/embedCard");
var _plugins = require("./shared/plugins");
var _unsupportedContent = require("./shared/unsupported-content");
var _expand = require("./shared/expand");
var _sharedStyles = require("../ui/Expand/sharedStyles");
var _media = require("./shared/media");
var _breakout = require("./shared/breakout");
