// @atlaskit/editor-common/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("./utils");
exports.ADFTraversor = utils_1.ADFTraversor;
exports.ErrorReporter = utils_1.ErrorReporter;
exports.ZERO_WIDTH_SPACE = utils_1.ZERO_WIDTH_SPACE;
exports.absoluteBreakoutWidth = utils_1.absoluteBreakoutWidth;
exports.browser = utils_1.browser;
exports.calcBreakoutWidth = utils_1.calcBreakoutWidth;
exports.breakoutConsts = utils_1.breakoutConsts;
exports.calcTableColumnWidths = utils_1.calcTableColumnWidths;
exports.calcWideWidth = utils_1.calcWideWidth;
exports.clearMeasure = utils_1.clearMeasure;
exports.compose = utils_1.compose;
exports.convertProsemirrorTableNodeToArrayOfRows = utils_1.convertProsemirrorTableNodeToArrayOfRows;
exports.createCompareNodes = utils_1.createCompareNodes;
exports.getAnalyticsAppearance = utils_1.getAnalyticsAppearance;
exports.getExtensionLozengeData = utils_1.getExtensionLozengeData;
exports.getExtensionRenderer = utils_1.getExtensionRenderer;
exports.getMarksByOrder = utils_1.getMarksByOrder;
exports.getResponseEndTime = utils_1.getResponseEndTime;
exports.getValidContent = utils_1.getValidContent;
exports.getValidDocument = utils_1.getValidDocument;
exports.getValidMark = utils_1.getValidMark;
exports.getValidNode = utils_1.getValidNode;
exports.getValidUnknownNode = utils_1.getValidUnknownNode;
exports.hasMergedCell = utils_1.hasMergedCell;
exports.isPastDate = utils_1.isPastDate;
exports.isPerformanceAPIAvailable = utils_1.isPerformanceAPIAvailable;
exports.isPerformanceObserverAvailable = utils_1.isPerformanceObserverAvailable;
exports.isSameMark = utils_1.isSameMark;
exports.isSubSupType = utils_1.isSubSupType;
exports.markOrder = utils_1.markOrder;
exports.measureRender = utils_1.measureRender;
exports.startMeasure = utils_1.startMeasure;
exports.stopMeasure = utils_1.stopMeasure;
exports.timestampToIsoFormat = utils_1.timestampToIsoFormat;
exports.timestampToString = utils_1.timestampToString;
exports.timestampToTaskContext = utils_1.timestampToTaskContext;
exports.timestampToUTCDate = utils_1.timestampToUTCDate;
exports.todayTimestampInUTC = utils_1.todayTimestampInUTC;
exports.withImageLoader = utils_1.withImageLoader;
var types_1 = require("./types");
exports.SortOrder = types_1.SortOrder;
exports.AnnotationUpdateEmitter = types_1.AnnotationUpdateEmitter;
exports.AnnotationUpdateEvent = types_1.AnnotationUpdateEvent;
var extensions_1 = require("./extensions");
exports.DefaultExtensionProvider = extensions_1.DefaultExtensionProvider;
exports.combineExtensionProviders = extensions_1.combineExtensionProviders;
exports.getExtensionKeyAndNodeKey = extensions_1.getExtensionKeyAndNodeKey;
exports.getExtensionModuleNode = extensions_1.getExtensionModuleNode;
exports.getItemsFromModule = extensions_1.getItemsFromModule;
exports.getNodeRenderer = extensions_1.getNodeRenderer;
exports.resolveImport = extensions_1.resolveImport;
var provider_factory_1 = require("./provider-factory");
exports.ProviderFactory = provider_factory_1.ProviderFactory;
exports.WithProviders = provider_factory_1.WithProviders;
var provider_helpers_1 = require("./provider-helpers");
exports.combineProviders = provider_helpers_1.combineProviders;
var styles_1 = require("./styles");
exports.TableSharedCssClassName = styles_1.TableSharedCssClassName;
exports.akEditorActiveBackground = styles_1.akEditorActiveBackground;
exports.akEditorActiveForeground = styles_1.akEditorActiveForeground;
exports.akEditorBlockquoteBorderColor = styles_1.akEditorBlockquoteBorderColor;
exports.akEditorBreakoutPadding = styles_1.akEditorBreakoutPadding;
exports.akEditorCodeBackground = styles_1.akEditorCodeBackground;
exports.akEditorCodeBlockPadding = styles_1.akEditorCodeBlockPadding;
exports.akEditorCodeFontFamily = styles_1.akEditorCodeFontFamily;
exports.akEditorCodeInlinePadding = styles_1.akEditorCodeInlinePadding;
exports.akEditorDefaultLayoutWidth = styles_1.akEditorDefaultLayoutWidth;
exports.akEditorDeleteBackground = styles_1.akEditorDeleteBackground;
exports.akEditorDeleteBackgroundShaded = styles_1.akEditorDeleteBackgroundShaded;
exports.akEditorDeleteBackgroundWithOpacity = styles_1.akEditorDeleteBackgroundWithOpacity;
exports.akEditorDeleteBorder = styles_1.akEditorDeleteBorder;
exports.akEditorDeleteIconColor = styles_1.akEditorDeleteIconColor;
exports.akEditorDropdownActiveBackground = styles_1.akEditorDropdownActiveBackground;
exports.akEditorFloatingDialogZIndex = styles_1.akEditorFloatingDialogZIndex;
exports.akEditorFloatingOverlapPanelZIndex = styles_1.akEditorFloatingOverlapPanelZIndex;
exports.akEditorFloatingPanelZIndex = styles_1.akEditorFloatingPanelZIndex;
exports.akEditorFocus = styles_1.akEditorFocus;
exports.akEditorFullPageMaxWidth = styles_1.akEditorFullPageMaxWidth;
exports.akEditorFullPageDefaultFontSize = styles_1.akEditorFullPageDefaultFontSize;
exports.akEditorFullWidthLayoutWidth = styles_1.akEditorFullWidthLayoutWidth;
exports.akEditorGridLineZIndex = styles_1.akEditorGridLineZIndex;
exports.akEditorGutterPadding = styles_1.akEditorGutterPadding;
exports.akEditorInactiveForeground = styles_1.akEditorInactiveForeground;
exports.akEditorMediaResizeHandlerPadding = styles_1.akEditorMediaResizeHandlerPadding;
exports.akEditorMediaResizeHandlerPaddingWide = styles_1.akEditorMediaResizeHandlerPaddingWide;
exports.akEditorMentionSelected = styles_1.akEditorMentionSelected;
exports.akEditorMenuZIndex = styles_1.akEditorMenuZIndex;
exports.akEditorMobileBreakoutPoint = styles_1.akEditorMobileBreakoutPoint;
exports.akEditorPopupBackground = styles_1.akEditorPopupBackground;
exports.akEditorPopupText = styles_1.akEditorPopupText;
exports.akEditorPrimaryButton = styles_1.akEditorPrimaryButton;
exports.akEditorSelectedBorder = styles_1.akEditorSelectedBorder;
exports.akEditorSelectedBorderBoldSize = styles_1.akEditorSelectedBorderBoldSize;
exports.akEditorSelectedBorderSize = styles_1.akEditorSelectedBorderSize;
exports.akEditorSelectedBlanketColor = styles_1.akEditorSelectedBlanketColor;
exports.akEditorSelectedBgColor = styles_1.akEditorSelectedBgColor;
exports.akEditorSelectedBlanketOpacity = styles_1.akEditorSelectedBlanketOpacity;
exports.akEditorSelectedIconColor = styles_1.akEditorSelectedIconColor;
exports.akEditorSmallZIndex = styles_1.akEditorSmallZIndex;
exports.akEditorShadowZIndex = styles_1.akEditorShadowZIndex;
exports.akEditorSubtleAccent = styles_1.akEditorSubtleAccent;
exports.akEditorSwoopCubicBezier = styles_1.akEditorSwoopCubicBezier;
exports.akEditorTableBorder = styles_1.akEditorTableBorder;
exports.akEditorTableBorderDark = styles_1.akEditorTableBorderDark;
exports.akEditorTableBorderDelete = styles_1.akEditorTableBorderDelete;
exports.akEditorTableBorderRadius = styles_1.akEditorTableBorderRadius;
exports.akEditorTableBorderSelected = styles_1.akEditorTableBorderSelected;
exports.akEditorTableCellBackgroundOpacity = styles_1.akEditorTableCellBackgroundOpacity;
exports.akEditorTableCellDelete = styles_1.akEditorTableCellDelete;
exports.akEditorTableCellMinWidth = styles_1.akEditorTableCellMinWidth;
exports.akEditorTableCellSelected = styles_1.akEditorTableCellSelected;
exports.akEditorTableFloatingControls = styles_1.akEditorTableFloatingControls;
exports.akEditorTableLegacyCellMinWidth = styles_1.akEditorTableLegacyCellMinWidth;
exports.akEditorTableNumberColumnWidth = styles_1.akEditorTableNumberColumnWidth;
exports.akEditorTableToolbar = styles_1.akEditorTableToolbar;
exports.akEditorTableToolbarDark = styles_1.akEditorTableToolbarDark;
exports.akEditorTableToolbarDelete = styles_1.akEditorTableToolbarDelete;
exports.akEditorTableToolbarSelected = styles_1.akEditorTableToolbarSelected;
exports.akEditorTableToolbarSize = styles_1.akEditorTableToolbarSize;
exports.akEditorUnitZIndex = styles_1.akEditorUnitZIndex;
exports.akEditorWideLayoutWidth = styles_1.akEditorWideLayoutWidth;
exports.akLayoutGutterOffset = styles_1.akLayoutGutterOffset;
exports.akRichMediaResizeZIndex = styles_1.akRichMediaResizeZIndex;
exports.akEditorStickyHeaderZIndex = styles_1.akEditorStickyHeaderZIndex;
exports.getAkEditorFullPageMaxWidth = styles_1.getAkEditorFullPageMaxWidth;
exports.blockMarksSharedStyles = styles_1.blockMarksSharedStyles;
exports.blockNodesVerticalMargin = styles_1.blockNodesVerticalMargin;
exports.blockquoteSharedStyles = styles_1.blockquoteSharedStyles;
exports.breakoutWideScaleRatio = styles_1.breakoutWideScaleRatio;
exports.calcTableWidth = styles_1.calcTableWidth;
exports.codeMarkSharedStyles = styles_1.codeMarkSharedStyles;
exports.columnLayoutSharedStyle = styles_1.columnLayoutSharedStyle;
exports.dateSharedStyle = styles_1.dateSharedStyle;
exports.DateSharedCssClassName = styles_1.DateSharedCssClassName;
exports.annotationSharedStyles = styles_1.annotationSharedStyles;
exports.AnnotationSharedCSSByState = styles_1.AnnotationSharedCSSByState;
exports.AnnotationSharedClassNames = styles_1.AnnotationSharedClassNames;
exports.editorFontSize = styles_1.editorFontSize;
exports.gridMediumMaxWidth = styles_1.gridMediumMaxWidth;
exports.headingsSharedStyles = styles_1.headingsSharedStyles;
exports.indentationSharedStyles = styles_1.indentationSharedStyles;
exports.inlineNodeSharedStyle = styles_1.inlineNodeSharedStyle;
exports.linkSharedStyle = styles_1.linkSharedStyle;
exports.listsSharedStyles = styles_1.listsSharedStyles;
exports.richMediaClassName = styles_1.richMediaClassName;
exports.mediaSingleSharedStyle = styles_1.mediaSingleSharedStyle;
exports.panelSharedStyles = styles_1.panelSharedStyles;
exports.PanelSharedCssClassName = styles_1.PanelSharedCssClassName;
exports.PanelSharedSelectors = styles_1.PanelSharedSelectors;
exports.paragraphSharedStyles = styles_1.paragraphSharedStyles;
exports.relativeSize = styles_1.relativeSize;
exports.ruleSharedStyles = styles_1.ruleSharedStyles;
exports.shadowSharedStyle = styles_1.shadowSharedStyle;
exports.tableCellBorderWidth = styles_1.tableCellBorderWidth;
exports.tableCellMinWidth = styles_1.tableCellMinWidth;
exports.tableCellPadding = styles_1.tableCellPadding;
exports.tableMarginBottom = styles_1.tableMarginBottom;
exports.tableMarginSides = styles_1.tableMarginSides;
exports.tableMarginTop = styles_1.tableMarginTop;
exports.tableNewColumnMinWidth = styles_1.tableNewColumnMinWidth;
exports.tableResizeHandleWidth = styles_1.tableResizeHandleWidth;
exports.tableSharedStyle = styles_1.tableSharedStyle;
exports.tasksAndDecisionsStyles = styles_1.tasksAndDecisionsStyles;
exports.TaskDecisionSharedCssClassName = styles_1.TaskDecisionSharedCssClassName;
exports.whitespaceSharedStyles = styles_1.whitespaceSharedStyles;
exports.akEditorSelectedBorderStyles = styles_1.akEditorSelectedBorderStyles;
exports.MentionSharedCssClassName = styles_1.MentionSharedCssClassName;
exports.EmojiSharedCssClassName = styles_1.EmojiSharedCssClassName;
exports.StatusSharedCssClassName = styles_1.StatusSharedCssClassName;
exports.smartCardSharedStyles = styles_1.smartCardSharedStyles;
exports.SmartCardSharedCssClassName = styles_1.SmartCardSharedCssClassName;
exports.DEFAULT_EMBED_CARD_WIDTH = styles_1.DEFAULT_EMBED_CARD_WIDTH;
exports.DEFAULT_EMBED_CARD_HEIGHT = styles_1.DEFAULT_EMBED_CARD_HEIGHT;
exports.akEditorLineHeight = styles_1.akEditorLineHeight;
var ui_1 = require("./ui");
exports.BaseTheme = ui_1.BaseTheme;
exports.ClearNextSiblingMarginTop = ui_1.ClearNextSiblingMarginTop;
exports.DEFAULT_IMAGE_HEIGHT = ui_1.DEFAULT_IMAGE_HEIGHT;
exports.DEFAULT_IMAGE_WIDTH = ui_1.DEFAULT_IMAGE_WIDTH;
exports.Emoji = ui_1.Emoji;
exports.ErrorMessage = ui_1.ErrorMessage;
exports.ExpandIconWrapper = ui_1.ExpandIconWrapper;
exports.ExpandLayoutWrapper = ui_1.ExpandLayoutWrapper;
exports.HelperMessage = ui_1.HelperMessage;
exports.MediaSingle = ui_1.MediaSingle;
exports.MediaSingleDimensionHelper = ui_1.MediaSingleDimensionHelper;
exports.Mention = ui_1.Mention;
exports.Popup = ui_1.Popup;
exports.UnsupportedBlock = ui_1.UnsupportedBlock;
exports.UnsupportedInline = ui_1.UnsupportedInline;
exports.ValidMessage = ui_1.ValidMessage;
exports.WidthConsumer = ui_1.WidthConsumer;
exports.WidthProvider = ui_1.WidthProvider;
exports.WithCreateAnalyticsEvent = ui_1.WithCreateAnalyticsEvent;
exports.calcColumnsFromPx = ui_1.calcColumnsFromPx;
exports.calcPctFromPx = ui_1.calcPctFromPx;
exports.calcPxFromColumns = ui_1.calcPxFromColumns;
exports.calcPxFromPct = ui_1.calcPxFromPct;
exports.expandMessages = ui_1.expandMessages;
exports.findOverflowScrollParent = ui_1.findOverflowScrollParent;
exports.getBreakpoint = ui_1.getBreakpoint;
exports.layoutSupportsWidth = ui_1.layoutSupportsWidth;
exports.mapBreakpointToLayoutMaxWidth = ui_1.mapBreakpointToLayoutMaxWidth;
exports.overflowShadow = ui_1.overflowShadow;
exports.shadowClassNames = ui_1.shadowClassNames;
exports.sharedExpandStyles = ui_1.sharedExpandStyles;
exports.snapToGrid = ui_1.snapToGrid;
exports.withOuterListeners = ui_1.withOuterListeners;
exports.MediaLink = ui_1.MediaLink;
exports.MediaLinkWrapper = ui_1.MediaLinkWrapper;
exports.wrappedLayouts = ui_1.wrappedLayouts;
exports.shouldAddDefaultWrappedWidth = ui_1.shouldAddDefaultWrappedWidth;
var link_1 = require("./messages/link");
exports.linkMessages = link_1.linkMessages;
