// @atlaskit/onboarding/dist/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Modal", {
  enumerable: true,
  get: function get() {
    return _components.Modal;
  }
});
Object.defineProperty(exports, "ModalTransition", {
  enumerable: true,
  get: function get() {
    return _modalTransition.default;
  }
});
Object.defineProperty(exports, "Spotlight", {
  enumerable: true,
  get: function get() {
    return _components.Spotlight;
  }
});
Object.defineProperty(exports, "SpotlightCard", {
  enumerable: true,
  get: function get() {
    return _components.SpotlightCard;
  }
});
Object.defineProperty(exports, "SpotlightManager", {
  enumerable: true,
  get: function get() {
    return _components.SpotlightManager;
  }
});
Object.defineProperty(exports, "SpotlightPulse", {
  enumerable: true,
  get: function get() {
    return _target.Pulse;
  }
});
Object.defineProperty(exports, "SpotlightTarget", {
  enumerable: true,
  get: function get() {
    return _components.SpotlightTarget;
  }
});
Object.defineProperty(exports, "SpotlightTransition", {
  enumerable: true,
  get: function get() {
    return _components.SpotlightTransition;
  }
});
Object.defineProperty(exports, "modalButtonTheme", {
  enumerable: true,
  get: function get() {
    return _components.modalButtonTheme;
  }
});
Object.defineProperty(exports, "spotlightButtonTheme", {
  enumerable: true,
  get: function get() {
    return _components.spotlightButtonTheme;
  }
});
Object.defineProperty(exports, "useSpotlight", {
  enumerable: true,
  get: function get() {
    return _components.useSpotlight;
  }
});
var _modalTransition = _interopRequireDefault(require("@atlaskit/modal-dialog/modal-transition"));
var _components = require("./components");
var _target = require("./styled/target");
