// @atlaskit/smart-card/dist/cjs/view/FlexibleCard/components/blocks/action-block/action-footer/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ActionFooter = void 0;
require("./index.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _error = _interopRequireDefault(require("@atlaskit/icon/utility/migration/error"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _compiled = require("@atlaskit/primitives/compiled");
var _motionWrapper = _interopRequireDefault(require("../../../common/motion-wrapper"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var containerStyles = null;
var errorStyles = null;
var styles = {
  containerStyles: "_kkk2n7od _2rkoglpi _1rjc12x7 _18zru2gc _19pku2gc _2hwxu2gc _otyrze3t _18u0u2gc _1bsb1osq _fiawglyw",
  errorContentStyles: "_bozgv77o",
  titleStyles: "_11c8dcr7 _syaz1gjq"
};
var getIcon = function getIcon(appearance) {
  switch (appearance) {
    case 'error':
      return /*#__PURE__*/React.createElement(_error.default, {
        color: "var(--ds-icon-danger, #C9372C)",
        spacing: "compact",
        label: "Error",
        LEGACY_size: "small",
        testId: "action-error-icon",
        LEGACY_margin: "0 ".concat("var(--ds-space-negative-025, -2px)")
      });
    default:
      return null;
  }
};
var ActionFooter = exports.ActionFooter = function ActionFooter(_ref) {
  var message = _ref.message,
    testId = _ref.testId;
  if (!message) {
    return null;
  }
  if ((0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v1')) {
    return /*#__PURE__*/React.createElement(_compiled.Box, {
      backgroundColor: message.appearance === 'error' ? 'color.background.danger' : undefined,
      testId: "".concat(testId, "-footer"),
      style: {
        width: "calc(100% - var(--container-padding-left) - var(--container-padding-right))"
      },
      xcss: styles.containerStyles
    }, /*#__PURE__*/React.createElement(_motionWrapper.default, {
      isFadeIn: true,
      show: true,
      showTransition: true
    }, /*#__PURE__*/React.createElement(_compiled.Inline, {
      alignBlock: "start",
      grow: "fill",
      space: "space.075",
      xcss: styles.errorContentStyles
    }, message.icon || getIcon(message.appearance), /*#__PURE__*/React.createElement(_compiled.Box, {
      xcss: styles.titleStyles
    }, message.title))));
  }
  return /*#__PURE__*/React.createElement("div", {
    "data-testId": "".concat(testId, "-footer"),
    className: (0, _runtime.ax)(["_kkk2n7od _2rkoglpi _1yt412x7 _18zru2gc _19pku2gc _2hwxu2gc _otyrze3t _18u0u2gc _vwz4gktf _1bsb1osq _fiawglyw", message.appearance === 'error' && "_bfhk1gly"])
  }, /*#__PURE__*/React.createElement(_motionWrapper.default, {
    isFadeIn: true,
    show: true,
    showTransition: true
  }, /*#__PURE__*/React.createElement(_compiled.Inline, {
    alignBlock: "start",
    grow: "fill",
    space: "space.075",
    xcss: styles.errorContentStyles
  }, message.icon || getIcon(message.appearance), /*#__PURE__*/React.createElement(_compiled.Box, {
    xcss: styles.titleStyles
  }, message.title))));
};
