// @atlaskit/link-datasource/dist/esm/ui/common/modal/count-view-smart-link/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import "./index.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
import React from 'react';
import { FormattedMessage } from 'react-intl-next';
import { Box } from '@atlaskit/primitives/compiled';
import LinkRenderType from '../../../issue-like-table/render-type/link';
var styles = {
  placeholderSmartLinkStyles: "_2rko16jk _syaz1e6v _ca0qze3t _n3tdze3t _19bvv77o _u5f3v77o _16qszucj",
  smartLinkContainerStyles: "_19bvv77o"
};
export var SmartCardPlaceholder = function SmartCardPlaceholder(_ref) {
  var placeholderText = _ref.placeholderText;
  return /*#__PURE__*/React.createElement(Box, {
    xcss: styles.smartLinkContainerStyles
  }, /*#__PURE__*/React.createElement(Box, {
    as: "span",
    testId: "datasource-modal--smart-card-placeholder",
    xcss: styles.placeholderSmartLinkStyles
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
    ,
    style: {
      backgroundColor: "var(--ds-surface-raised, #FFFFFF)"
    }
  }, /*#__PURE__*/React.createElement(FormattedMessage, placeholderText)));
};
export var SmartLink = function SmartLink(_ref2) {
  var url = _ref2.url;
  return /*#__PURE__*/React.createElement(Box, {
    xcss: styles.smartLinkContainerStyles
  }, /*#__PURE__*/React.createElement(LinkRenderType, {
    url: url
  }));
};
