// @atlaskit/jql-editor/dist/cjs/plugins/autocomplete/components/autocomplete-dropdown/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OptionList = exports.AutocompleteOptionsContainer = exports.AutocompleteLoadingFooter = exports.AutocompleteContainer = void 0;
var _react = require("@emotion/react");
var _styled = _interopRequireDefault(require("@emotion/styled"));
var _colors = require("@atlaskit/theme/colors");
var _constants = require("@atlaskit/theme/constants");
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var AutocompleteContainer = exports.AutocompleteContainer = _styled.default.div({
  position: 'absolute',
  backgroundColor: "var(--ds-surface-overlay, white)",
  borderRadius: "var(--ds-border-radius-100, 3px)",
  willChange: 'top, left',
  zIndex: _constants.layers.dialog(),
  boxShadow: "var(--ds-shadow-overlay, ".concat("0 4px 8px -2px ".concat(_colors.N50A, ", 0 0 1px ").concat(_colors.N60A), ")"),
  padding: "var(--ds-space-075, 6px)".concat(" ", "var(--ds-space-0, 0)"),
  minWidth: '200px',
  maxWidth: '400px',
  '&:focus': {
    outline: 'none'
  },
  marginLeft: "var(--ds-space-negative-100, -8px)",
  marginTop: "var(--ds-space-200, 16px)"
},
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-dynamic-styles, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
function (_ref) {
  var isOpen = _ref.isOpen;
  return isOpen ? (0, _react.css)({
    visibility: 'visible'
  }) : (0, _react.css)({
    visibility: 'hidden'
  });
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var AutocompleteOptionsContainer = exports.AutocompleteOptionsContainer = _styled.default.div({
  maxHeight: '288px',
  overflow: 'auto'
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var OptionList = exports.OptionList = _styled.default.ul({
  listStyle: 'none',
  margin: "var(--ds-space-0, 0)",
  padding: "var(--ds-space-0, 0)"
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var AutocompleteLoadingFooter = exports.AutocompleteLoadingFooter = _styled.default.div({
  display: 'flex',
  justifyContent: 'center',
  color: "var(--ds-text-subtlest, ".concat(_colors.N200, ")"),
  fontStyle: 'italic',
  padding: "var(--ds-space-150, 12px)",
  textAlign: 'center'
},
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-dynamic-styles -- Ignored via go/DSP-18766
function (_ref2) {
  var hasOptions = _ref2.hasOptions;
  return hasOptions && {
    borderTop: "solid 1px ".concat("var(--ds-border, ".concat(_colors.N40, ")")),
    marginTop: "var(--ds-space-075, 6px)",
    paddingTop: "var(--ds-space-250, 20px)"
  };
});
