// @atlaskit/smart-card/dist/cjs/view/FlexibleCard/components/actions/automation-action/automation-manual-triggers/manual-triggers-modal/sub-components/error-state/main.js
/* main.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AutomationModalErrorState = void 0;
require("./main.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _reactIntlNext = require("react-intl-next");
var _compiled = require("@atlaskit/primitives/compiled");
var _errorIcon = _interopRequireDefault(require("./error-icon"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var styles = {
  errorState: "_y3gn1h6o _19pk1jfw",
  imageContainer: "_1e0c1ule _p12f7iqz _c71lplmo _otyr1ejb _19pk1b66",
  description: "_p12f1gwv"
};
var i18n = (0, _reactIntlNext.defineMessages)({
  errorDescription: {
    id: 'automation-menu.modal.error.description',
    defaultMessage: 'This is taking awhile to load. Try refreshing the page.',
    description: "Description for the error section shown when rules can't be fetched"
  },
  errorIconLabel: {
    id: 'automation-menu.modal.error.image.alt',
    defaultMessage: 'Error icon',
    description: 'Alternative text for the image that renders when there is an error.'
  }
});
var AutomationModalErrorState = exports.AutomationModalErrorState = function AutomationModalErrorState() {
  var _useIntl = (0, _reactIntlNext.useIntl)(),
    formatMessage = _useIntl.formatMessage;
  return /*#__PURE__*/React.createElement(_compiled.Stack, {
    xcss: styles.errorState,
    alignInline: "center",
    alignBlock: "center"
  }, /*#__PURE__*/React.createElement(_compiled.Box, {
    xcss: styles.imageContainer
  }, /*#__PURE__*/React.createElement(_errorIcon.default, null)), /*#__PURE__*/React.createElement(_compiled.Box, {
    xcss: styles.description
  }, formatMessage(i18n.errorDescription)));
};
