// @atlaskit/drawer/dist/cjs/compiled/blanket.js
/* blanket.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./blanket.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _react = _interopRequireDefault(require("react"));
var _blanket = _interopRequireDefault(require("@atlaskit/blanket"));
var _motion = require("@atlaskit/motion");
var styles = {
  blanket: "_kqswh2mm"
};

/**
 * A wrapper around `@atlaskit/blanket` that adds a fade in/out transition.
 */
var Blanket = function Blanket(_ref) {
  var isOpen = _ref.isOpen,
    onBlanketClicked = _ref.onBlanketClicked,
    testId = _ref.testId;
  return /*#__PURE__*/_react.default.createElement(_motion.ExitingPersistence, {
    appear: true
  }, isOpen && /*#__PURE__*/_react.default.createElement(_motion.FadeIn
  /*
   * The fade out animation uses half the passed duration so it evens out.
   */, {
    duration: "large"
    /**
     * We don't expose this on `FadeIn` but it does get passed down.
     * TODO: figure out how we want to handle this...
     */
    // @ts-ignore
    ,
    animationTimingFunction: "ease-out"
  }, function (_ref2) {
    var className = _ref2.className;
    return /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _runtime.ax)([styles.blanket, className])
    }, /*#__PURE__*/_react.default.createElement(_blanket.default, {
      isTinted: true,
      onBlanketClicked: onBlanketClicked,
      testId: testId && testId
    }));
  }));
};
var _default = exports.default = Blanket;
