// @atlaskit/teams-public/dist/es2019/ui/team-containers/team-containers-skeleton/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import "./index.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
import React from 'react';
import { Box, Grid, Stack } from '@atlaskit/primitives/compiled';
import { LinkedContainerCardSkeleton } from './linked-container-card-skeleton';
const styles = {
  containers: "_yv0ewxkt",
  showMoreButtonSkeleton: "_2rko1q77 _1rjc1b66 _1bsbvbgk"
};
export const TeamContainersSkeleton = ({
  numberOfContainers
}) => {
  return /*#__PURE__*/React.createElement(Stack, {
    space: "space.200",
    testId: "team-containers-skeleton"
  }, /*#__PURE__*/React.createElement(Grid, {
    xcss: styles.containers,
    gap: "space.100",
    autoFlow: "row"
  }, [...Array(numberOfContainers)].map((_, index) => /*#__PURE__*/React.createElement(LinkedContainerCardSkeleton, {
    key: index
  }))), /*#__PURE__*/React.createElement(Box, {
    backgroundColor: "color.background.neutral",
    xcss: styles.showMoreButtonSkeleton
  }));
};
