// @atlaskit/smart-card/dist/cjs/view/FlexibleCard/components/actions/automation-action/automation-manual-triggers/manual-triggers-modal/sub-components/header/main.js
/* main.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AutomationModalHeader = void 0;
require("./main.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _reactIntlNext = require("react-intl-next");
var _heading = _interopRequireDefault(require("@atlaskit/heading"));
var _manualTriggerIcon = _interopRequireDefault(require("@atlaskit/legacy-custom-icons/manual-trigger-icon"));
var _modalDialog = require("@atlaskit/modal-dialog");
var _compiled = require("@atlaskit/primitives/compiled");
var _colors = require("@atlaskit/theme/colors");
var _menuContext = require("../../menu-context");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var styles = {
  iconStyle: "_2rkoglpi _18zr1b66 _1rjcv77o _2hwxutpp",
  modalDescriptionStyle: "_19pkutpp"
};
var i18n = (0, _reactIntlNext.defineMessages)({
  modalHeaderIconLabel: {
    id: 'automation-menu.modal.header.icon.label',
    defaultMessage: 'Automation modal header icon',
    description: 'A label for the icon in the header of the automation modal.'
  }
});
var AutomationModalHeader = exports.AutomationModalHeader = function AutomationModalHeader(_ref) {
  var modalTitle = _ref.modalTitle,
    modalDescription = _ref.modalDescription;
  var _useIntl = (0, _reactIntlNext.useIntl)(),
    formatMessage = _useIntl.formatMessage;
  var _useAutomationMenu = (0, _menuContext.useAutomationMenu)(),
    initialised = _useAutomationMenu.initialised,
    rules = _useAutomationMenu.rules;
  var _useModal = (0, _modalDialog.useModal)(),
    titleId = _useModal.titleId;
  var showDescription = initialised && rules.length > 0 && !!modalDescription;
  return /*#__PURE__*/React.createElement(_modalDialog.ModalHeader, null, /*#__PURE__*/React.createElement(_compiled.Stack, null, /*#__PURE__*/React.createElement(_compiled.Inline, {
    alignBlock: "center"
  }, /*#__PURE__*/React.createElement(_compiled.Box, {
    xcss: styles.iconStyle
    // eslint-disable-next-line @atlaskit/design-system/ensure-design-token-usage
    ,
    backgroundColor: "color.background.accent.green.subtlest"
  }, /*#__PURE__*/React.createElement(_manualTriggerIcon.default, {
    size: "small",
    label: formatMessage(i18n.modalHeaderIconLabel),
    primaryColor: "var(--ds-icon-accent-green, ".concat(_colors.G50, ")")
  })), /*#__PURE__*/React.createElement(_heading.default, {
    size: "medium",
    id: titleId
  }, modalTitle)), showDescription && /*#__PURE__*/React.createElement(_compiled.Box, {
    xcss: styles.modalDescriptionStyle
  }, modalDescription)));
};
