// @atlaskit/media-card/dist/cjs/utils/analytics/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "createAndFireMediaCardEvent", {
  enumerable: true,
  get: function get() {
    return _analytics.createAndFireMediaCardEvent;
  }
});
Object.defineProperty(exports, "extractErrorInfo", {
  enumerable: true,
  get: function get() {
    return _analytics.extractErrorInfo;
  }
});
Object.defineProperty(exports, "fireMediaCardEvent", {
  enumerable: true,
  get: function get() {
    return _analytics.fireMediaCardEvent;
  }
});
Object.defineProperty(exports, "getCacheHitEventPayload", {
  enumerable: true,
  get: function get() {
    return _analytics.getCacheHitEventPayload;
  }
});
Object.defineProperty(exports, "getDownloadFailedEventPayload", {
  enumerable: true,
  get: function get() {
    return _analytics.getDownloadFailedEventPayload;
  }
});
Object.defineProperty(exports, "getDownloadSucceededEventPayload", {
  enumerable: true,
  get: function get() {
    return _analytics.getDownloadSucceededEventPayload;
  }
});
Object.defineProperty(exports, "getErrorEventPayload", {
  enumerable: true,
  get: function get() {
    return _analytics.getErrorEventPayload;
  }
});
Object.defineProperty(exports, "getErrorTraceContext", {
  enumerable: true,
  get: function get() {
    return _analytics.getErrorTraceContext;
  }
});
Object.defineProperty(exports, "getFileAttributes", {
  enumerable: true,
  get: function get() {
    return _analytics.getFileAttributes;
  }
});
Object.defineProperty(exports, "getRemoteSuccessEventPayload", {
  enumerable: true,
  get: function get() {
    return _analytics.getRemoteSuccessEventPayload;
  }
});
Object.defineProperty(exports, "getRenderErrorErrorDetail", {
  enumerable: true,
  get: function get() {
    return _analytics.getRenderErrorErrorDetail;
  }
});
Object.defineProperty(exports, "getRenderErrorErrorReason", {
  enumerable: true,
  get: function get() {
    return _analytics.getRenderErrorErrorReason;
  }
});
Object.defineProperty(exports, "getRenderErrorEventPayload", {
  enumerable: true,
  get: function get() {
    return _analytics.getRenderErrorEventPayload;
  }
});
Object.defineProperty(exports, "getRenderErrorFailReason", {
  enumerable: true,
  get: function get() {
    return _analytics.getRenderErrorFailReason;
  }
});
Object.defineProperty(exports, "getRenderErrorRequestMetadata", {
  enumerable: true,
  get: function get() {
    return _analytics.getRenderErrorRequestMetadata;
  }
});
Object.defineProperty(exports, "getRenderFailedExternalUriPayload", {
  enumerable: true,
  get: function get() {
    return _analytics.getRenderFailedExternalUriPayload;
  }
});
Object.defineProperty(exports, "getRenderFailedFileStatusPayload", {
  enumerable: true,
  get: function get() {
    return _analytics.getRenderFailedFileStatusPayload;
  }
});
Object.defineProperty(exports, "getRenderSucceededEventPayload", {
  enumerable: true,
  get: function get() {
    return _analytics.getRenderSucceededEventPayload;
  }
});
var _analytics = require("./analytics");
