// @atlaskit/editor-plugin-avatar-group/dist/cjs/ui/AvatarGroupPluginWrapper.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _react2 = require("@emotion/react");
var _reactIntlNext = require("react-intl-next");
var _analytics = require("@atlaskit/editor-common/analytics");
var _messages = require("@atlaskit/editor-common/messages");
var _avatarsWithPluginState = _interopRequireDefault(require("./avatars-with-plugin-state"));
/**
 * @jsxRuntime classic
 * @jsx jsx
 */

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

// eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage
var toolbarButtonWrapper = (0, _react2.css)({
  display: 'flex',
  justifyContent: 'flex-end',
  flexGrow: 0,
  alignItems: 'center',
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '& > div': {
    marginRight: 0
  }
});

// eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
var toolbarButtonWrapperFullWidth = (0, _react2.css)(toolbarButtonWrapper, {
  flexGrow: 1
});
var AvatarGroupPluginWrapper = function AvatarGroupPluginWrapper(props) {
  var dispatchAnalyticsEvent = props.dispatchAnalyticsEvent,
    featureFlags = props.featureFlags;
  var intl = (0, _reactIntlNext.useIntl)();
  (0, _react.useEffect)(function () {
    if (!dispatchAnalyticsEvent) {
      return;
    }
    dispatchAnalyticsEvent({
      action: _analytics.ACTION.VIEWED,
      actionSubject: _analytics.ACTION_SUBJECT.BUTTON,
      actionSubjectId: _analytics.ACTION_SUBJECT_ID.AVATAR_GROUP_PLUGIN,
      eventType: _analytics.EVENT_TYPE.UI
    });
  }, [dispatchAnalyticsEvent]);
  return (0, _react2.jsx)("div", {
    "aria-label": intl.formatMessage(_messages.avatarGroupMessages.editors),
    "data-testid": 'avatar-group-in-plugin',
    css: props.takeFullWidth ? toolbarButtonWrapperFullWidth : toolbarButtonWrapper
  }, (0, _react2.jsx)(_avatarsWithPluginState.default, {
    editorView: props.editorView,
    eventDispatcher: props.eventDispatcher,
    inviteToEditComponent: props.collabEdit && props.collabEdit.inviteToEditComponent,
    inviteToEditHandler: props.collabEdit && props.collabEdit.inviteToEditHandler,
    isInviteToEditButtonSelected: props.collabEdit && props.collabEdit.isInviteToEditButtonSelected,
    featureFlags: featureFlags,
    editorAnalyticsAPI: props.editorAnalyticsAPI,
    editorAPI: props.editorAPI
  }));
};
var _default = exports.default = AvatarGroupPluginWrapper;
