// @atlaskit/editor-plugin-avatar-group/dist/cjs/ui/avatars-with-plugin-state.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactIntlNext = require("react-intl-next");
var _hooks = require("@atlaskit/editor-common/hooks");
var _messages = _interopRequireDefault(require("@atlaskit/editor-common/messages"));
var _avatars = require("./avatars");
var _inviteToEdit = require("./invite-to-edit");
var AvatarsWithPluginState = function AvatarsWithPluginState(props) {
  var title = props.intl.formatMessage(_messages.default.inviteToEditButtonTitle);
  var selected = props.isInviteToEditButtonSelected,
    onClick = props.inviteToEditHandler,
    Component = props.inviteToEditComponent,
    editorView = props.editorView,
    featureFlags = props.featureFlags,
    editorAnalyticsAPI = props.editorAnalyticsAPI,
    editorAPI = props.editorAPI;
  var _useSharedPluginState = (0, _hooks.useSharedPluginState)(editorAPI, ['collabEdit']),
    collabEditState = _useSharedPluginState.collabEditState;
  if (!collabEditState) {
    return null;
  }
  return /*#__PURE__*/_react.default.createElement(_avatars.Avatars, {
    sessionId: collabEditState.sessionId,
    participants: collabEditState.activeParticipants,
    editorView: editorView,
    featureFlags: featureFlags,
    editorAnalyticsAPI: editorAnalyticsAPI,
    editorAPI: editorAPI
  }, /*#__PURE__*/_react.default.createElement(_inviteToEdit.InviteToEditButton, {
    title: title,
    selected: selected,
    onClick: onClick,
    Component: Component
  }));
};
var _default = exports.default = (0, _reactIntlNext.injectIntl)(AvatarsWithPluginState);
