// @atlaskit/editor-plugin-avatar-group/dist/cjs/ui/invite-to-edit.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InviteToEditButton = void 0;
var _react = _interopRequireWildcard(require("react"));
var _react2 = require("@emotion/react");
var _uiMenu = require("@atlaskit/editor-common/ui-menu");
var _addEditorAdd = _interopRequireDefault(require("@atlaskit/icon/core/migration/add--editor-add"));
var _styles = require("./styles");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * @jsxRuntime classic
 * @jsx jsx
 */

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var ID = function ID(props) {
  return (0, _react2.jsx)(_react.Fragment, null, props.children);
};
var InviteToEditButton = exports.InviteToEditButton = function InviteToEditButton(props) {
  var Component = props.Component,
    onClick = props.onClick,
    selected = props.selected,
    title = props.title;
  var iconBefore = _react.default.useMemo(function () {
    return (0, _react2.jsx)(_addEditorAdd.default, {
      label: title
    });
  }, [title]);
  if (!Component && !onClick) {
    return null;
  }
  var Wrapper = Component ? Component : ID;
  return (
    // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
    (0, _react2.jsx)("div", {
      css: _styles.inviteTeamWrapperStyles
    }, (0, _react2.jsx)(Wrapper, null, (0, _react2.jsx)(_uiMenu.ToolbarButton
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    , {
      className: "invite-to-edit",
      onClick: onClick,
      selected: selected,
      title: title,
      titlePosition: "bottom",
      iconBefore: iconBefore
    })))
  );
};
