// @atlaskit/editor-plugin-avatar-group/dist/cjs/ui/styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.inviteTeamWrapperStyles = exports.badge = exports.avatarContainerStyles = void 0;
var _react = require("@emotion/react");
var _editorSharedStyles = require("@atlaskit/editor-shared-styles");
var _colors = require("@atlaskit/theme/colors");
/* eslint-disable @repo/internal/styles/no-exported-styles, @atlaskit/design-system/no-nested-styles */
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var inviteTeamWrapperStyles = exports.inviteTeamWrapperStyles = (0, _react.css)({
  background: "var(--ds-background-neutral, ".concat(_colors.N20, ")"),
  borderRadius: '50%',
  minWidth: "var(--ds-space-400, 32px)",
  marginLeft: "var(--ds-space-negative-050, -4px)"
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var avatarContainerStyles = exports.avatarContainerStyles = (0, _react.css)({
  marginRight: "var(--ds-space-100, 8px)",
  display: 'flex',
  alignItems: 'center',
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors, @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
  '&& > ul': {
    listStyleType: 'none'
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors, @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
  'div:last-child button.invite-to-edit': {
    borderRadius: '50%',
    height: '32px',
    width: '32px',
    padding: "var(--ds-space-025, 2px)"
  }
});
var badge = exports.badge = function badge(backgroundColor, textColor) {
  return (0, _react.css)({
    display: 'block',
    position: 'absolute',
    right: "var(--ds-space-025, 2px)",
    bottom: "var(--ds-space-025, 2px)",
    width: "var(--ds-space-150, 12px)",
    height: "var(--ds-space-150, 12px)",
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    zIndex: _editorSharedStyles.akEditorSmallZIndex,
    borderRadius: '3px',
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    background: backgroundColor,
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    color: textColor,
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    fontSize: (0, _editorSharedStyles.relativeFontSizeToBase16)(9),
    // eslint-disable-next-line @atlaskit/design-system/use-tokens-typography
    lineHeight: 0,
    paddingTop: "var(--ds-space-075, 6px)",
    textAlign: 'center',
    boxShadow: "0 0 1px ".concat("var(--ds-border-inverse, #fff)"),
    boxSizing: 'border-box'
  });
};
