// @atlaskit/link-picker/dist/cjs/ui/link-picker/search-results/link-search-error/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.testIds = exports.messages = exports.LinkSearchError = exports.CONTACT_SUPPORT_LINK_FEDRAMP = exports.CONTACT_SUPPORT_LINK = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _reactIntlNext = require("react-intl-next");
var _atlassianContext = require("@atlaskit/atlassian-context");
var _button = _interopRequireDefault(require("@atlaskit/button"));
var _new = _interopRequireDefault(require("@atlaskit/button/new"));
var _link = _interopRequireDefault(require("@atlaskit/link"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _genericErrorSvg = require("../../../../common/generic-error-svg");
var _genericErrorSvgV = require("../../../../common/generic-error-svg-v2");
var _emptyState = require("../../../../common/ui/empty-state");
var CONTACT_SUPPORT_LINK = exports.CONTACT_SUPPORT_LINK = 'https://support.atlassian.com/contact/';
var CONTACT_SUPPORT_LINK_FEDRAMP = exports.CONTACT_SUPPORT_LINK_FEDRAMP = 'https://gcs.atlassian-us-gov-mod.net/servicedesk/customer/portals';
var messages = exports.messages = (0, _reactIntlNext.defineMessages)({
  searchErrorHeader: {
    id: 'fabric.linkPicker.search.error.heading',
    defaultMessage: 'We’re having trouble loading data.',
    description: 'Heading message shown when a search throws an error'
  },
  searchErrorDescription: {
    id: 'fabric.linkPicker.search.error.description',
    defaultMessage: 'Refresh the page, or contact <a>Atlassian Support</a> if this keeps happening.',
    description: 'Describes possible actions when search throws an error'
  },
  searchErrorAction: {
    id: 'fabric.linkPicker.search.error.retry',
    defaultMessage: 'Refresh',
    description: 'Describe the action user can take to retry the search'
  }
});
var testIds = exports.testIds = {
  searchError: 'link-search-error'
};
var LinkSearchError = exports.LinkSearchError = function LinkSearchError(_ref) {
  var onRetry = _ref.onRetry;
  var intl = (0, _reactIntlNext.useIntl)();
  var ButtonComponent = (0, _platformFeatureFlags.fg)('platform-link-picker-remove-legacy-button') ? _new.default : _button.default;
  return /*#__PURE__*/_react.default.createElement(_emptyState.EmptyState, {
    testId: testIds.searchError,
    header: intl.formatMessage(messages.searchErrorHeader),
    description: /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, (0, _extends2.default)({}, messages.searchErrorDescription, {
      values: {
        a: function a(label) {
          return (0, _platformFeatureFlags.fg)('platform-link-picker-remove-legacy-button') ? /*#__PURE__*/_react.default.createElement(_link.default, {
            href: (0, _atlassianContext.isFedRamp)() ? CONTACT_SUPPORT_LINK_FEDRAMP : CONTACT_SUPPORT_LINK,
            target: "_blank",
            rel: "noopener noreferrer"
          }, label) : /*#__PURE__*/_react.default.createElement(_button.default, {
            appearance: "link",
            spacing: "none",
            href: (0, _atlassianContext.isFedRamp)() ? CONTACT_SUPPORT_LINK_FEDRAMP : CONTACT_SUPPORT_LINK,
            target: "_blank",
            rel: "noopener noreferrer"
          }, label);
        }
      }
    })),
    action: onRetry && (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-link-picker') ? /*#__PURE__*/_react.default.createElement(ButtonComponent, {
      appearance: "primary",
      onClick: onRetry
    }, /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, messages.searchErrorAction)) : null,
    renderImage: (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-link-picker') ? function () {
      return /*#__PURE__*/_react.default.createElement(_genericErrorSvgV.GenericErrorSVGV2, null);
    } : function () {
      return /*#__PURE__*/_react.default.createElement(_genericErrorSvg.GenericErrorSVG, null);
    }
  });
};
