// @atlaskit/jql-editor/dist/cjs/plugins/autocomplete/components/autocomplete-option/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TooltipContent = exports.OptionName = exports.OptionListItem = exports.OptionHighlight = exports.FieldTypeIcon = exports.FieldType = exports.DeprecatedOptionContainer = void 0;
var _styled = _interopRequireDefault(require("@emotion/styled"));
var _colors = require("@atlaskit/theme/colors");
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var TooltipContent = exports.TooltipContent = _styled.default.div({
  fontFamily: "var(--ds-font-family-body, ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)"
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var OptionListItem = exports.OptionListItem = _styled.default.li({
  cursor: 'pointer',
  padding: "var(--ds-space-075, 6px)".concat(" ", "var(--ds-space-100, 8px)"),
  fontFamily: "var(--ds-font-family-code, ui-monospace, Menlo, \"Segoe UI Mono\", \"Ubuntu Mono\", monospace)",
  // eslint-disable-next-line @atlaskit/design-system/use-tokens-typography
  lineHeight: '24px'
},
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-dynamic-styles -- Ignored via go/DSP-18766
function (_ref) {
  var isSelected = _ref.isSelected;
  return isSelected && {
    background: "var(--ds-background-neutral-subtle-hovered, ".concat(_colors.N20, ")")
  };
},
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-dynamic-styles -- Ignored via go/DSP-18766
function (_ref2) {
  var isDeprecated = _ref2.isDeprecated;
  return isDeprecated && {
    cursor: 'default',
    color: "var(--ds-text-disabled, ".concat(_colors.N50, ")")
  };
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var OptionName = exports.OptionName = _styled.default.div({
  color: "var(--ds-text, ".concat(_colors.N800, ")"),
  overflow: 'hidden',
  whiteSpace: 'nowrap',
  textOverflow: 'ellipsis',
  // Added so that overflowed option names do not squish the deprecated info icon
  flex: 1
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var DeprecatedOptionContainer = exports.DeprecatedOptionContainer = _styled.default.div({
  color: "var(--ds-text-disabled, ".concat(_colors.N50, ")"),
  display: 'flex',
  justifyContent: 'space-between',
  opacity: 0.6
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var OptionHighlight = exports.OptionHighlight = _styled.default.span({
  fontWeight: "var(--ds-font-weight-bold, 700)"
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var FieldType = exports.FieldType = _styled.default.div({
  display: 'flex',
  alignItems: 'center',
  marginTop: "var(--ds-space-negative-025, -2px)",
  color: "var(--ds-text-subtlest, ".concat(_colors.N100, ")")
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var FieldTypeIcon = exports.FieldTypeIcon = _styled.default.span({
  display: 'flex',
  marginRight: "var(--ds-space-050, 4px)"
});
