// @atlaskit/editor-common/dist/cjs/type-ahead/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TypeAheadAvailableNodes = exports.SelectItemMode = void 0;
Object.defineProperty(exports, "typeAheadListMessages", {
  enumerable: true,
  get: function get() {
    return _messages.typeAheadListMessages;
  }
});
var _messages = require("./messages");
// Disable no-re-export rule for entry point files
/* eslint-disable @atlaskit/editor/no-re-export */
var TypeAheadAvailableNodes = exports.TypeAheadAvailableNodes = /*#__PURE__*/function (TypeAheadAvailableNodes) {
  TypeAheadAvailableNodes["EMOJI"] = "emojiTypeAhead";
  TypeAheadAvailableNodes["MENTION"] = "mentionTypeAhead";
  TypeAheadAvailableNodes["QUICK_INSERT"] = "quickInsertTypeAhead";
  return TypeAheadAvailableNodes;
}({});
var SelectItemMode = exports.SelectItemMode = /*#__PURE__*/function (SelectItemMode) {
  SelectItemMode["SHIFT_ENTER"] = "shift-enter";
  SelectItemMode["ENTER"] = "enter";
  SelectItemMode["SPACE"] = "space";
  SelectItemMode["SELECTED"] = "selected";
  SelectItemMode["TAB"] = "tab";
  return SelectItemMode;
}({});
