// @atlaskit/media-filmstrip/dist/esm/filmstripView/wrappers.js
import _extends from "@babel/runtime/helpers/extends";
import React, { forwardRef } from 'react';
import { fg } from '@atlaskit/platform-feature-flags';
import { ArrowLeftWrapper as EmotionArrowLeftWrapper, ArrowRightWrapper as EmotionArrowRightWrapper, ShadowRight as EmotionShadowRight, LeftArrow as EmotionLeftArrow, RightArrow as EmotionRightArrow, FilmStripViewWrapper as EmotionFilmStripViewWrapper, FilmStripListWrapper as EmotionFilmStripListWrapper, FilmStripList as EmotionFilmStripList, FilmStripListItem as EmotionFilmStripListItem } from './wrappers-emotion';
import { ArrowLeftWrapper as CompiledArrowLeftWrapper, ArrowRightWrapper as CompiledArrowRightWrapper, ShadowRight as CompiledShadowRight, LeftArrow as CompiledLeftArrow, RightArrow as CompiledRightArrow, FilmStripViewWrapper as CompiledFilmStripViewWrapper, FilmStripListWrapper as CompiledFilmStripListWrapper, FilmStripList as CompiledFilmStripList, FilmStripListItem as CompiledFilmStripListItem } from './wrappers-compiled';
export var ArrowLeftWrapper = function ArrowLeftWrapper(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledArrowLeftWrapper, props) : /*#__PURE__*/React.createElement(EmotionArrowLeftWrapper, props);
};
export var ShadowRight = function ShadowRight(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledShadowRight, props) : /*#__PURE__*/React.createElement(EmotionShadowRight, props);
};
export var ArrowRightWrapper = function ArrowRightWrapper(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledArrowRightWrapper, props) : /*#__PURE__*/React.createElement(EmotionArrowRightWrapper, props);
};
export var LeftArrow = function LeftArrow(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledLeftArrow, props) : /*#__PURE__*/React.createElement(EmotionLeftArrow, props);
};
export var RightArrow = function RightArrow(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledRightArrow, props) : /*#__PURE__*/React.createElement(EmotionRightArrow, props);
};
export var FilmStripViewWrapper = function FilmStripViewWrapper(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledFilmStripViewWrapper, props) : /*#__PURE__*/React.createElement(EmotionFilmStripViewWrapper, props);
};
export var FilmStripListWrapper = /*#__PURE__*/forwardRef(function (props, ref) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledFilmStripListWrapper, _extends({}, props, {
    ref: ref
  })) : /*#__PURE__*/React.createElement(EmotionFilmStripListWrapper, _extends({}, props, {
    ref: ref
  }));
});
export var FilmStripList = /*#__PURE__*/forwardRef(function (props, ref) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledFilmStripList, _extends({}, props, {
    ref: ref
  })) : /*#__PURE__*/React.createElement(EmotionFilmStripList, _extends({}, props, {
    ref: ref
  }));
});
export var FilmStripListItem = function FilmStripListItem(props) {
  return fg('platform_media_compiled') ? /*#__PURE__*/React.createElement(CompiledFilmStripListItem, props) : /*#__PURE__*/React.createElement(EmotionFilmStripListItem, props);
};
