// @atlaskit/media-file-preview/dist/cjs/getPreview/objectURLCache.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createObjectURLCache = exports.PREVIEW_CACHE_LRU_SIZE = exports.ObjectURLCache = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _eventemitter = require("eventemitter2");
var _lru_map = require("lru_map");
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function _superPropGet(t, o, e, r) { var p = (0, _get2.default)((0, _getPrototypeOf2.default)(1 & r ? t.prototype : t), o, e); return 2 & r && "function" == typeof p ? function (t) { return p.apply(e, t); } : p; }
var PREVIEW_CACHE_LRU_SIZE = exports.PREVIEW_CACHE_LRU_SIZE = 50;
var ExtendedLRUCache = /*#__PURE__*/function (_LRUMap) {
  function ExtendedLRUCache(limit) {
    var _this;
    (0, _classCallCheck2.default)(this, ExtendedLRUCache);
    _this = _callSuper(this, ExtendedLRUCache, [limit]);
    _this.eventEmitter = new _eventemitter.EventEmitter2();
    return _this;
  }
  (0, _inherits2.default)(ExtendedLRUCache, _LRUMap);
  return (0, _createClass2.default)(ExtendedLRUCache, [{
    key: "shift",
    value: function shift() {
      var entry = _superPropGet(ExtendedLRUCache, "shift", this, 3)([]);
      this.eventEmitter.emit('shift', entry);
      return entry;
    }
  }, {
    key: "on",
    value: function on(event, callback) {
      this.eventEmitter.on(event, callback);
    }
  }]);
}(_lru_map.LRUMap);
var ObjectURLCache = exports.ObjectURLCache = /*#__PURE__*/function () {
  function ObjectURLCache(size) {
    (0, _classCallCheck2.default)(this, ObjectURLCache);
    this.cache = new ExtendedLRUCache(size);
    this.cache.on('shift', function (entry) {
      if (entry && entry[1].dataURI) {
        URL.revokeObjectURL(entry[1].dataURI);
      }
    });
  }
  return (0, _createClass2.default)(ObjectURLCache, [{
    key: "has",
    value: function has(key) {
      return !!this.cache.find(key);
    }
  }, {
    key: "get",
    value: function get(key) {
      return this.cache.get(key);
    }
  }, {
    key: "set",
    value: function set(key, value) {
      this.cache.set(key, value);
    }
  }, {
    key: "remove",
    value: function remove(key) {
      var removed = this.cache.delete(key);
      removed && URL.revokeObjectURL(removed.dataURI);
    }
  }, {
    key: "clear",
    value: function clear() {
      this.cache.clear();
    }
  }]);
}();
var createObjectURLCache = exports.createObjectURLCache = function createObjectURLCache() {
  return new ObjectURLCache(PREVIEW_CACHE_LRU_SIZE);
};
