// @atlaskit/media-ui/dist/cjs/imageMetaData/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SupportedImageMetaTag = exports.ImageType = exports.ExifOrientation = void 0;
var ImageType = exports.ImageType = /*#__PURE__*/function (ImageType) {
  ImageType["JPEG"] = "image/jpeg";
  ImageType["PNG"] = "image/png";
  return ImageType;
}({});
var SupportedImageMetaTag = exports.SupportedImageMetaTag = /*#__PURE__*/function (SupportedImageMetaTag) {
  SupportedImageMetaTag["XResolution"] = "XResolution";
  SupportedImageMetaTag["YResolution"] = "YResolution";
  SupportedImageMetaTag["Orientation"] = "Orientation";
  return SupportedImageMetaTag;
}({});
// http://sylvana.net/jpegcrop/exif_orientation.html
var ExifOrientation = exports.ExifOrientation = {
  'top-left': 1,
  // none
  'top-right': 2,
  // flip horizontal
  'bottom-right': 3,
  // rotate 180
  'bottom-left': 4,
  // flip vertical
  'left-top': 5,
  // transpose
  'right-top': 6,
  // rotate 90
  'right-bottom': 7,
  // transverse
  'left-bottom': 8 // rotate 270
};
