// @atlaskit/jql-editor/dist/cjs/ui/jql-editor-footer-content/jql-messages/warnings/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useFormattedWarningMessage = exports.WarningMessages = void 0;
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _reactMagneticDi = require("react-magnetic-di");
var _form = require("@atlaskit/form");
var _warning = _interopRequireDefault(require("@atlaskit/icon/glyph/warning"));
var _warning2 = _interopRequireDefault(require("@atlaskit/icon/utility/warning"));
var _compiled = require("@atlaskit/primitives/compiled");
var colors = _interopRequireWildcard(require("@atlaskit/theme/colors"));
var _useEditorViewHasWarnings = require("../../../../hooks/use-editor-view-has-warnings");
var _state = require("../../../../state");
var _format = require("../format");
var _messages = require("./messages");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var softDeprecateEpicParentDocsLink = 'https://support.atlassian.com/jira-software-cloud/docs/upcoming-changes-epic-link-replaced-with-parent/';
var getEpicLinkDeprecationTerm = function getEpicLinkDeprecationTerm(hydratedDeprecations) {
  var epicLinkWarning = hydratedDeprecations.filter(function (warning) {
    return warning.deprecatedSearcherKey === 'com.pyxis.greenhopper.jira:gh-epic-link-searcher';
  });
  return epicLinkWarning.length > 0 ? epicLinkWarning[0].id : null;
};
var getParentLinkDeprecationTerm = function getParentLinkDeprecationTerm(hydratedDeprecations) {
  var parentLinkWarning = hydratedDeprecations.filter(function (warning) {
    return warning.deprecatedSearcherKey === 'com.atlassian.jpo:jpo-custom-field-parent-searcher';
  });
  return parentLinkWarning.length > 0 ? parentLinkWarning[0].id : null;
};
var useFormattedWarningMessage = exports.useFormattedWarningMessage = function useFormattedWarningMessage() {
  var _useHydratedDeprecati = (0, _state.useHydratedDeprecations)(),
    _useHydratedDeprecati2 = (0, _slicedToArray2.default)(_useHydratedDeprecati, 1),
    hydratedDeprecations = _useHydratedDeprecati2[0];
  var _useExternalMessages = (0, _state.useExternalMessages)(),
    _useExternalMessages2 = (0, _slicedToArray2.default)(_useExternalMessages, 1),
    externalWarnings = _useExternalMessages2[0].warnings;
  var hasWarnings = (0, _useEditorViewHasWarnings.useEditorViewHasWarnings)();
  var _useIntl = (0, _state.useIntl)(),
    _useIntl2 = (0, _slicedToArray2.default)(_useIntl, 1),
    formatMessage = _useIntl2[0].formatMessage;
  var epicLinkDeprecationTerm = getEpicLinkDeprecationTerm(hydratedDeprecations);
  var parentLinkDeprecationTerm = getParentLinkDeprecationTerm(hydratedDeprecations);
  var formatWarnings = (0, _react.useCallback)(function (message) {
    var combinedMessages = [{
      type: 'warning',
      message: message
    }].concat((0, _toConsumableArray2.default)(externalWarnings));
    return /*#__PURE__*/_react.default.createElement(_format.FormatMessages, {
      messages: combinedMessages
    });
  }, [externalWarnings]);
  if (!hasWarnings) {
    return null;
  } else if (!hydratedDeprecations.length && externalWarnings.length) {
    return /*#__PURE__*/_react.default.createElement(_format.FormatMessages, {
      messages: externalWarnings
    });
  } else if (epicLinkDeprecationTerm == null && parentLinkDeprecationTerm == null) {
    var _hydratedDeprecations;
    return formatWarnings(formatMessage(_messages.messages.defaultWarning, {
      deprecatedField: (_hydratedDeprecations = hydratedDeprecations[0]) === null || _hydratedDeprecations === void 0 ? void 0 : _hydratedDeprecations.id
    }));
  } else if (epicLinkDeprecationTerm && parentLinkDeprecationTerm) {
    return formatWarnings(formatMessage(_messages.messages.deprecatedBothParentReplacementMessage, {
      link: function link(chunks) {
        return /*#__PURE__*/_react.default.createElement("a", {
          href: softDeprecateEpicParentDocsLink,
          target: "_blank",
          rel: "noopener noreferrer",
          onClick: function onClick(e) {
            return e.stopPropagation();
          }
        }, chunks);
      },
      receivedFirst: epicLinkDeprecationTerm,
      receivedSecond: parentLinkDeprecationTerm,
      parentReplacement: 'Parent'
    }));
  } else if (epicLinkDeprecationTerm || parentLinkDeprecationTerm) {
    return formatWarnings(formatMessage(_messages.messages.deprecatedParentReplacementMessage, {
      link: function link(chunks) {
        return /*#__PURE__*/_react.default.createElement("a", {
          href: softDeprecateEpicParentDocsLink,
          target: "_blank",
          rel: "noopener noreferrer",
          onClick: function onClick(e) {
            return e.stopPropagation();
          }
        }, chunks);
      },
      deprecatedField: epicLinkDeprecationTerm ? epicLinkDeprecationTerm : parentLinkDeprecationTerm,
      parentReplacement: 'Parent'
    }));
  }
};
var WarningMessages = exports.WarningMessages = function WarningMessages() {
  var warningMessage = useFormattedWarningMessage();
  return warningMessage != null ? /*#__PURE__*/_react.default.createElement(_format.MessageContainer, null, /*#__PURE__*/_react.default.createElement(_form.HelperMessage, {
    testId: "jql-editor-warning-message"
  }, /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    as: "span",
    paddingInlineEnd: "space.050"
  }, /*#__PURE__*/_react.default.createElement(_warning2.default, {
    label: "",
    color: "var(--ds-icon-warning, ".concat(colors.Y400, ")"),
    LEGACY_size: "small",
    LEGACY_fallbackIcon: _warning.default,
    LEGACY_margin: "0 -4px 0 0"
  })), warningMessage)) : null;
};
