// @atlaskit/teams-public/dist/esm/ui/team-containers/add-container-card/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import _extends from "@babel/runtime/helpers/extends";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import "./index.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
import React, { useState } from 'react';
import { IconButton } from '@atlaskit/button/new';
import AddIcon from '@atlaskit/icon/utility/add';
import { fg } from '@atlaskit/platform-feature-flags';
import { Box, Flex, Inline, Stack, Text } from '@atlaskit/primitives/compiled';
import { getContainerProperties } from '../../../common/utils/get-container-properties';
var styles = {
  card: "_4cvr1h6o _1bsb1osq",
  container: "_2rkoop52 _ca0qutpp _u5f3utpp _n3tdutpp _19bvutpp _d0altlke",
  iconWrapper: "_2rkoiti9 _12y3e4h9 _12jimuej _1qu2gq9o _syaz131l _4cvxmuej _ksodnqa1"
};
var AddContainerCardWrapper = function AddContainerCardWrapper(_ref) {
  var children = _ref.children,
    onClick = _ref.onClick;
  var _useState = useState(false),
    _useState2 = _slicedToArray(_useState, 2),
    hovered = _useState2[0],
    setHovered = _useState2[1];
  var handleMouseEnter = function handleMouseEnter() {
    return setHovered(true);
  };
  var handleMouseLeave = function handleMouseLeave() {
    return setHovered(false);
  };
  return /*#__PURE__*/React.createElement(Box, {
    backgroundColor: hovered ? 'elevation.surface.hovered' : 'elevation.surface.sunken',
    xcss: styles.container,
    onMouseEnter: handleMouseEnter,
    onMouseLeave: handleMouseLeave,
    onClick: onClick
  }, children);
};
export var AddContainerCard = function AddContainerCard(_ref2) {
  var containerType = _ref2.containerType,
    onAddAContainerClick = _ref2.onAddAContainerClick;
  var _getContainerProperti = getContainerProperties(containerType),
    description = _getContainerProperti.description,
    icon = _getContainerProperti.icon,
    title = _getContainerProperti.title;
  return /*#__PURE__*/React.createElement(AddContainerCardWrapper, {
    onClick: onAddAContainerClick
  }, /*#__PURE__*/React.createElement(Inline, {
    space: "space.100",
    xcss: styles.card
  }, /*#__PURE__*/React.createElement(Box, {
    xcss: styles.iconWrapper
  }, /*#__PURE__*/React.createElement(IconButton, {
    label: "Add a container",
    appearance: "subtle",
    icon: AddIcon,
    testId: "add-icon",
    onClick: function onClick(e) {
      onAddAContainerClick(e);
      e.stopPropagation();
    }
  })), /*#__PURE__*/React.createElement(Stack, null, /*#__PURE__*/React.createElement(Text, {
    maxLines: 1,
    weight: "medium",
    color: "color.text"
  }, title), /*#__PURE__*/React.createElement(Flex, _extends({
    gap: "space.050"
  }, fg('enable_card_alignment_fix') ? {
    alignItems: 'center'
  } : {}), icon, /*#__PURE__*/React.createElement(Text, {
    size: "small",
    color: "color.text.subtle"
  }, description)))));
};
