// @atlaskit/link-picker/dist/cjs/common/ui/empty-state/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EmptyState = void 0;
require("./index.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _heading = _interopRequireDefault(require("@atlaskit/heading"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _compiled = require("@atlaskit/primitives/compiled");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var styles = {
  container: "_1pfhoahv _6rthoahv _y3gn1h6o",
  containerV2: "_ca0q1ejb _u5f3oahv _19bvoahv _n3td1jfw _y3gn1h6o"
};
var EmptyState = exports.EmptyState = function EmptyState(_ref) {
  var testId = _ref.testId,
    header = _ref.header,
    description = _ref.description,
    action = _ref.action,
    renderImage = _ref.renderImage;
  if ((0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-link-picker')) {
    return /*#__PURE__*/React.createElement(_compiled.Flex, {
      xcss: styles.containerV2,
      testId: testId,
      direction: "column",
      alignItems: "center",
      gap: "space.200"
    }, renderImage === null || renderImage === void 0 ? void 0 : renderImage(), /*#__PURE__*/React.createElement(_heading.default, {
      size: "small",
      as: "h2"
    }, header), description && /*#__PURE__*/React.createElement(_compiled.Text, {
      as: "p",
      color: "color.text"
    }, description), action && action);
  }
  return /*#__PURE__*/React.createElement(_compiled.Flex, {
    xcss: styles.container,
    testId: testId,
    direction: "column",
    alignItems: "center",
    gap: "space.300"
  }, renderImage === null || renderImage === void 0 ? void 0 : renderImage(), /*#__PURE__*/React.createElement(_compiled.Flex, {
    direction: "column",
    alignItems: "center",
    gap: "space.200"
  }, /*#__PURE__*/React.createElement(_heading.default, {
    size: "medium",
    as: "h2"
  }, header), description && /*#__PURE__*/React.createElement(_compiled.Text, {
    as: "p",
    color: "color.text"
  }, description)));
};
