// @atlaskit/editor-plugin-highlight/dist/esm/highlightPlugin.js
import React from 'react';
import { backgroundColor } from '@atlaskit/adf-schema';
import { editorExperiment } from '@atlaskit/tmp-editor-statsig/experiments';
import { changeColor } from './editor-commands/change-color';
import { keymapPlugin } from './pm-plugins/keymap';
import { createPlugin, highlightPluginKey } from './pm-plugins/main';
import { FloatingToolbarHighlightColorWithIntl as FloatingToolbarHighlightColor } from './ui/FloatingToolbarHighlightColor';
import { PrimaryToolbarHighlightColorWithIntl as PrimaryToolbarHighlightColor } from './ui/PrimaryToolbarHighlightColor';
export var highlightPlugin = function highlightPlugin(_ref) {
  var _api$analytics, _api$primaryToolbar;
  var api = _ref.api,
    options = _ref.config;
  var editorAnalyticsAPI = api === null || api === void 0 || (_api$analytics = api.analytics) === null || _api$analytics === void 0 ? void 0 : _api$analytics.actions;
  var primaryToolbarComponent = function primaryToolbarComponent(_ref2) {
    var popupsMountPoint = _ref2.popupsMountPoint,
      popupsBoundariesElement = _ref2.popupsBoundariesElement,
      popupsScrollableElement = _ref2.popupsScrollableElement,
      disabled = _ref2.disabled,
      isToolbarReducedSpacing = _ref2.isToolbarReducedSpacing,
      editorView = _ref2.editorView;
    return /*#__PURE__*/React.createElement(PrimaryToolbarHighlightColor, {
      popupsMountPoint: popupsMountPoint,
      popupsBoundariesElement: popupsBoundariesElement,
      popupsScrollableElement: popupsScrollableElement,
      disabled: disabled,
      isToolbarReducedSpacing: isToolbarReducedSpacing,
      pluginInjectionApi: api,
      editorView: editorView
    });
  };
  api === null || api === void 0 || (_api$primaryToolbar = api.primaryToolbar) === null || _api$primaryToolbar === void 0 || _api$primaryToolbar.actions.registerComponent({
    name: 'highlight',
    component: primaryToolbarComponent
  });
  return {
    name: 'highlight',
    marks: function marks() {
      return [{
        name: 'backgroundColor',
        mark: backgroundColor
      }];
    },
    commands: {
      changeColor: changeColor(editorAnalyticsAPI)
    },
    pmPlugins: function pmPlugins() {
      return [{
        name: 'highlight',
        plugin: function plugin() {
          return createPlugin({
            api: api
          });
        }
      }, {
        name: 'highlightKeymap',
        plugin: function plugin() {
          return keymapPlugin({
            api: api
          });
        }
      }];
    },
    getSharedState: function getSharedState(editorState) {
      if (!editorState) {
        return;
      }
      return highlightPluginKey.getState(editorState);
    },
    pluginsOptions: {
      selectionToolbar: function selectionToolbar() {
        var _api$selectionToolbar;
        if ((api === null || api === void 0 || (_api$selectionToolbar = api.selectionToolbar) === null || _api$selectionToolbar === void 0 || (_api$selectionToolbar = _api$selectionToolbar.sharedState) === null || _api$selectionToolbar === void 0 || (_api$selectionToolbar = _api$selectionToolbar.currentState()) === null || _api$selectionToolbar === void 0 ? void 0 : _api$selectionToolbar.toolbarDocking) === 'none' && editorExperiment('platform_editor_controls', 'variant1', {
          exposure: true
        })) {
          var toolbarCustom = {
            type: 'custom',
            render: function render(_view, _idx, dispatchAnalyticsEvent) {
              return /*#__PURE__*/React.createElement(FloatingToolbarHighlightColor, {
                dispatchAnalyticsEvent: dispatchAnalyticsEvent,
                pluginInjectionApi: api
              });
            },
            fallback: []
          };
          var rank = editorExperiment('platform_editor_controls', 'variant1') ? 5 : -9;
          return {
            rank: rank,
            isToolbarAbove: true,
            items: [toolbarCustom],
            pluginName: 'highlight'
          };
        } else {
          return undefined;
        }
      }
    },
    primaryToolbarComponent: !(api !== null && api !== void 0 && api.primaryToolbar) ? primaryToolbarComponent : undefined
  };
};
