// @atlaskit/editor-plugin-tasks-and-decisions/dist/cjs/nodeviews/task-node-view.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.taskView = void 0;
var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));
var _lazyNodeView = require("@atlaskit/editor-common/lazy-node-view");
var _experiments = require("@atlaskit/tmp-editor-statsig/experiments");
var _taskItem = require("./taskItem");
var _TaskItemNodeView = require("./TaskItemNodeView");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != (0, _typeof2.default)(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var taskView = exports.taskView = function taskView(portalProviderAPI, eventDispatcher, providerFactory, api, intl, placeholder) {
  if ((0, _experiments.editorExperiment)('platform_editor_vanilla_dom', true, {
    exposure: true
  })) {
    return function (node, view, getPos) {
      return new _TaskItemNodeView.TaskItemNodeView(node, view, getPos, {
        placeholder: placeholder,
        api: api,
        intl: intl
      });
    };
  }
  if ((0, _experiments.editorExperiment)('platform_editor_exp_lazy_node_views', false)) {
    return (0, _taskItem.taskItemNodeViewFactory)(portalProviderAPI, eventDispatcher, providerFactory, api, intl, placeholder);
  }
  return (0, _lazyNodeView.withLazyLoading)({
    nodeName: 'taskItem',
    getNodeViewOptions: function getNodeViewOptions() {},
    loader: function loader() {
      var result = Promise.resolve().then(function () {
        return _interopRequireWildcard(require( /* webpackChunkName: "@atlaskit-internal_editor-plugin-tasks-and-decisions_task-item-nodeview" */
        './taskItem'));
      }).then(function (_ref) {
        var taskItemNodeViewFactory = _ref.taskItemNodeViewFactory;
        return function (node, view, getPos) {
          return taskItemNodeViewFactory(portalProviderAPI, eventDispatcher, providerFactory, api, intl, placeholder)(node, view, getPos);
        };
      });
      return result;
    }
  });
};
