// @atlaskit/editor-plugin-tasks-and-decisions/dist/cjs/nodeviews/taskItem.js
/* taskItem.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.taskItemNodeViewFactory = taskItemNodeViewFactory;
require("./taskItem.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _reactIntlNext = require("react-intl-next");
var _steps = require("@atlaskit/adf-schema/steps");
var _analyticsNext = require("@atlaskit/analytics-next");
var _analytics = require("@atlaskit/editor-common/analytics");
var _messages = require("@atlaskit/editor-common/messages");
var _reactNodeView = _interopRequireDefault(require("@atlaskit/editor-common/react-node-view"));
var _useSharedPluginStateSelector = require("@atlaskit/editor-common/use-shared-plugin-state-selector");
var _heading = _interopRequireDefault(require("@atlaskit/heading"));
var _checkMarkEditorDone = _interopRequireDefault(require("@atlaskit/icon/core/migration/check-mark--editor-done"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _popup = _interopRequireDefault(require("@atlaskit/popup"));
var _compiled = require("@atlaskit/primitives/compiled");
var _Task2 = _interopRequireDefault(require("../ui/Task"));
var _useShowPlaceholder = require("./hooks/use-show-placeholder");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function _superPropGet(t, o, e, r) { var p = (0, _get2.default)((0, _getPrototypeOf2.default)(1 & r ? t.prototype : t), o, e); return 2 & r && "function" == typeof p ? function (t) { return p.apply(e, t); } : p; } // Ignored via go/ees005
// eslint-disable-next-line import/no-named-as-default
// Ignored via go/ees005
// eslint-disable-next-line import/no-named-as-default
var TRYING_REQUEST_TIMEOUT = 3000;
var wrapperStyles = null;
var wrapperBoxStyles = null;
var dotStyles = null;

// eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage -- This rule thinks this isn't a `css()` call due to the name mapping
var dotStylesUnbounded = null;
var pressableStyles = {
  pressable: "_ca0qze3t _n3tdze3t _19bvze3t _u5f3ze3t _1wybdlk8 _syaz1e6v _bfhksm61 _9oik1r31 _1bnx8stv _jf4cnqa1 _9h8h16c2"
};
var RequestedMessage = function RequestedMessage() {
  var _useIntl = (0, _reactIntlNext.useIntl)(),
    formatMessage = _useIntl.formatMessage;
  return /*#__PURE__*/React.createElement(React.Fragment, null, formatMessage(_messages.tasksAndDecisionsMessages.requestToEdit), /*#__PURE__*/React.createElement(_checkMarkEditorDone.default, {
    label: "requested-to-edit",
    color: "var(--ds-icon-disabled, #091E424F)"
  }));
};
var RequestToEditButton = function RequestToEditButton(_ref) {
  var onClick = _ref.onClick;
  var _useIntl2 = (0, _reactIntlNext.useIntl)(),
    formatMessage = _useIntl2.formatMessage;
  return /*#__PURE__*/React.createElement(_compiled.Box, null, /*#__PURE__*/React.createElement(_compiled.Pressable, {
    onClick: onClick,
    xcss: pressableStyles.pressable
  }, formatMessage(_messages.tasksAndDecisionsMessages.requestToEdit)));
};
var anaylyticsEventPayload = function anaylyticsEventPayload(action) {
  return {
    action: action,
    actionSubject: _analytics.ACTION_SUBJECT.REQUEST_TO_EDIT_POP_UP,
    eventType: _analytics.EVENT_TYPE.UI,
    attributes: {
      platform: _analytics.PLATFORMS.WEB,
      mode: _analytics.MODE.EDITOR
    }
  };
};
var TaskItemWrapper = function TaskItemWrapper(_ref2) {
  var localId = _ref2.localId,
    forwardRef = _ref2.forwardRef,
    isDone = _ref2.isDone,
    onChange = _ref2.onChange,
    placeholder = _ref2.placeholder,
    providerFactory = _ref2.providerFactory,
    isContentNodeEmpty = _ref2.isContentNodeEmpty,
    api = _ref2.api,
    getPos = _ref2.getPos,
    editorView = _ref2.editorView;
  var hasRequestedEditPermission = (0, _useSharedPluginStateSelector.useSharedPluginStateSelector)(api, 'taskDecision.hasRequestedEditPermission');
  var focusedTaskItemLocalId = (0, _useSharedPluginStateSelector.useSharedPluginStateSelector)(api, 'taskDecision.focusedTaskItemLocalId');
  var isFocused = Boolean(focusedTaskItemLocalId === localId);
  var _useState = (0, _react.useState)(false),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    isOpen = _useState2[0],
    setIsOpen = _useState2[1];
  var _useState3 = (0, _react.useState)(hasRequestedEditPermission),
    _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
    requested = _useState4[0],
    setRequested = _useState4[1];
  var _useState5 = (0, _react.useState)(false),
    _useState6 = (0, _slicedToArray2.default)(_useState5, 2),
    tryingRequest = _useState6[0],
    setTryingRequest = _useState6[1];
  var _useIntl3 = (0, _reactIntlNext.useIntl)(),
    formatMessage = _useIntl3.formatMessage;
  (0, _react.useEffect)(function () {
    if ((0, _platformFeatureFlags.fg)('editor_request_to_edit_task')) {
      setRequested(hasRequestedEditPermission);
    }
  }, [hasRequestedEditPermission]);
  (0, _react.useEffect)(function () {
    if (!tryingRequest && (0, _platformFeatureFlags.fg)('editor_request_to_edit_task')) {
      var timout = setTimeout(function () {
        setTryingRequest(false);
      }, TRYING_REQUEST_TIMEOUT);
      return function () {
        return clearTimeout(timout);
      };
    }
  }, [tryingRequest]);
  var showPlaceholder = (0, _useShowPlaceholder.useShowPlaceholder)({
    editorView: editorView,
    isContentNodeEmpty: isContentNodeEmpty,
    getPos: getPos,
    api: api
  });
  var onHandleEdit = function onHandleEdit(editorAnalyticsAPI) {
    if ((0, _platformFeatureFlags.fg)('editor_request_to_edit_task')) {
      var _api$taskDecision;
      setTryingRequest(true);
      var tr = editorView.state.tr;
      var nodePos = getPos();
      if (typeof nodePos !== 'number') {
        return;
      }
      tr.setMeta('scrollIntoView', false);
      if (!(api !== null && api !== void 0 && (_api$taskDecision = api.taskDecision) !== null && _api$taskDecision !== void 0 && (_api$taskDecision = _api$taskDecision.sharedState.currentState()) !== null && _api$taskDecision !== void 0 && _api$taskDecision.hasEditPermission)) {
        var _api$taskDecision2;
        var requestToEdit = api === null || api === void 0 || (_api$taskDecision2 = api.taskDecision) === null || _api$taskDecision2 === void 0 || (_api$taskDecision2 = _api$taskDecision2.sharedState.currentState()) === null || _api$taskDecision2 === void 0 ? void 0 : _api$taskDecision2.requestToEditContent;
        if (requestToEdit) {
          requestToEdit();
        }
      }
      editorAnalyticsAPI === null || editorAnalyticsAPI === void 0 || editorAnalyticsAPI.attachAnalyticsEvent(anaylyticsEventPayload(_analytics.ACTION.REQUEST_TO_EDIT))(tr);
      editorView.dispatch(tr);
    }
  };
  var onHandleDismiss = function onHandleDismiss(editorAnalyticsAPI) {
    editorAnalyticsAPI === null || editorAnalyticsAPI === void 0 || editorAnalyticsAPI.fireAnalyticsEvent(anaylyticsEventPayload(_analytics.ACTION.DISMISSED));
    setIsOpen(false);
  };
  var onHandleClick = function onHandleClick() {
    if ((0, _platformFeatureFlags.fg)('editor_request_to_edit_task')) {
      setIsOpen(true);
    }
  };
  if (!(0, _platformFeatureFlags.fg)('editor_request_to_edit_task')) {
    return /*#__PURE__*/React.createElement(_Task2.default, {
      taskId: localId,
      contentRef: forwardRef,
      isDone: isDone,
      onChange: onChange,
      isFocused: isFocused,
      showPlaceholder: showPlaceholder,
      placeholder: placeholder,
      providers: providerFactory,
      api: api
    });
  }
  return /*#__PURE__*/React.createElement(_popup.default, {
    isOpen: isOpen,
    onClose: function onClose() {
      var _api$analytics;
      return onHandleDismiss(api === null || api === void 0 || (_api$analytics = api.analytics) === null || _api$analytics === void 0 ? void 0 : _api$analytics.actions);
    },
    content: function content() {
      var _api$editorViewMode;
      return /*#__PURE__*/React.createElement("div", {
        className: (0, _runtime.ax)(["_1e0c1txw _2lx21bp4 _p12f1qwj _ca0qpxbi _u5f31ejb _n3tdpxbi _19bv1ejb"])
      }, /*#__PURE__*/React.createElement(_compiled.Stack, {
        space: "space.150"
      }, /*#__PURE__*/React.createElement(_heading.default, {
        size: "xsmall"
      }, formatMessage(_messages.tasksAndDecisionsMessages.editAccessTitle)), /*#__PURE__*/React.createElement("div", null, formatMessage(_messages.tasksAndDecisionsMessages.requestToEditDescription)), /*#__PURE__*/React.createElement("div", {
        className: (0, _runtime.ax)(["_zulp1b66 _1e0c1txw _4cvr1h6o _syaz1lh4"])
      }, tryingRequest || requested ? /*#__PURE__*/React.createElement(RequestedMessage, null) : /*#__PURE__*/React.createElement(RequestToEditButton, {
        onClick: (api === null || api === void 0 || (_api$editorViewMode = api.editorViewMode) === null || _api$editorViewMode === void 0 || (_api$editorViewMode = _api$editorViewMode.sharedState.currentState()) === null || _api$editorViewMode === void 0 ? void 0 : _api$editorViewMode.mode) === 'view' ? function () {
          var _api$analytics2;
          return onHandleEdit(api === null || api === void 0 || (_api$analytics2 = api.analytics) === null || _api$analytics2 === void 0 ? void 0 : _api$analytics2.actions);
        } : undefined
      }), /*#__PURE__*/React.createElement("div", {
        className: (0, _runtime.ax)(["_19pku2gc _otyru2gc _18u0u2gc _2hwxu2gc _1e0c1o8l _1bsbyh40 _4t3iyh40 _bfhkc8cv", "_2rko1ssb"])
      }), /*#__PURE__*/React.createElement(_compiled.Box, null, /*#__PURE__*/React.createElement(_compiled.Pressable, {
        onClick: function onClick() {
          var _api$analytics3;
          return onHandleDismiss(api === null || api === void 0 || (_api$analytics3 = api.analytics) === null || _api$analytics3 === void 0 ? void 0 : _api$analytics3.actions);
        },
        xcss: pressableStyles.pressable
      }, formatMessage(_messages.tasksAndDecisionsMessages.dismiss))))));
    },
    trigger: function trigger(triggerProps) {
      var _api$taskDecision3;
      return /*#__PURE__*/React.createElement(_Task2.default, {
        taskId: localId,
        contentRef: forwardRef,
        inputRef: triggerProps.ref,
        isDone: isDone,
        onChange: onChange,
        onClick: onHandleClick,
        isFocused: isFocused,
        showPlaceholder: showPlaceholder,
        placeholder: placeholder,
        providers: providerFactory,
        disableOnChange: !(api !== null && api !== void 0 && (_api$taskDecision3 = api.taskDecision) !== null && _api$taskDecision3 !== void 0 && (_api$taskDecision3 = _api$taskDecision3.sharedState.currentState()) !== null && _api$taskDecision3 !== void 0 && _api$taskDecision3.hasEditPermission),
        api: api
      });
    },
    placement: 'bottom-start'
  });
};
var Task = /*#__PURE__*/function (_ReactNodeView) {
  function Task() {
    var _this;
    (0, _classCallCheck2.default)(this, Task);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, Task, [].concat(args));
    (0, _defineProperty2.default)(_this, "handleOnChange", function (taskId, isChecked) {
      var _this$api;
      var tr = _this.view.state.tr;
      var nodePos = _this.getPos();
      if (typeof nodePos !== 'number') {
        return false;
      }

      // SetAttrsStep should be used to prevent task updates from being dropped when mapping task ticks
      // from a previous version of the document, such as a published page.
      tr.step(new _steps.SetAttrsStep(nodePos, {
        state: isChecked ? 'DONE' : 'TODO',
        localId: taskId
      }));
      tr.setMeta('scrollIntoView', false);

      /**
       * This is a test implementation to call the request to edit mutation
       * from within editor when toggling a task where a user has no edit access.
       *
       * This will eventially be handled by https://product-fabric.atlassian.net/browse/ED-24773
       * to connect up the correct user action
       */
      if (!((_this$api = _this.api) !== null && _this$api !== void 0 && (_this$api = _this$api.taskDecision) !== null && _this$api !== void 0 && (_this$api = _this$api.sharedState.currentState()) !== null && _this$api !== void 0 && _this$api.hasEditPermission) && (0, _platformFeatureFlags.fg)('editor_request_to_edit_task')) {
        var _this$api2;
        var requestToEdit = (_this$api2 = _this.api) === null || _this$api2 === void 0 || (_this$api2 = _this$api2.taskDecision) === null || _this$api2 === void 0 || (_this$api2 = _this$api2.sharedState.currentState()) === null || _this$api2 === void 0 ? void 0 : _this$api2.requestToEditContent;
        if (requestToEdit) {
          requestToEdit();
        }
      }
      _this.view.dispatch(tr);
    });
    /**
     * Dynamically generates analytics data relating to the parent list.
     *
     * Required to be dynamic, as list (in prosemirror model) may have
     * changed (e.g. item movements, or additional items in list).
     * This node view will have not rerendered for those changes, so
     * cannot render the position and listSize into the
     * AnalyticsContext at initial render time.
     */
    (0, _defineProperty2.default)(_this, "addListAnalyticsData", function (event) {
      try {
        var nodePos = _this.getPos();
        if (typeof nodePos !== 'number') {
          return false;
        }
        var resolvedPos = _this.view.state.doc.resolve(nodePos);
        var position = resolvedPos.index();
        var listSize = resolvedPos.parent.childCount;
        var listLocalId = resolvedPos.parent.attrs.localId;
        event.update(function (payload) {
          var _this$api3;
          var _payload$attributes = payload.attributes,
            attributes = _payload$attributes === void 0 ? {} : _payload$attributes,
            actionSubject = payload.actionSubject;
          if (actionSubject !== 'action') {
            // Not action related, ignore
            return payload;
          }
          return _objectSpread(_objectSpread({}, payload), {}, {
            attributes: _objectSpread(_objectSpread({}, attributes), {}, {
              position: position,
              listSize: listSize,
              listLocalId: listLocalId
            }, (0, _platformFeatureFlags.fg)('editor_request_to_edit_task') && {
              hasEditPermission: (_this$api3 = _this.api) === null || _this$api3 === void 0 || (_this$api3 = _this$api3.taskDecision) === null || _this$api3 === void 0 || (_this$api3 = _this$api3.sharedState.currentState()) === null || _this$api3 === void 0 ? void 0 : _this$api3.hasEditPermission
            })
          });
        });
      } catch (e) {
        // This can occur if pos is NaN (seen it in some test cases)
        // Act defensively here, and lose some analytics data rather than
        // cause any user facing error.
      }
    });
    return _this;
  }
  (0, _inherits2.default)(Task, _ReactNodeView);
  return (0, _createClass2.default)(Task, [{
    key: "initWithAPI",
    value: function initWithAPI(api) {
      this.api = api;
      this.init();
      return this;
    }
  }, {
    key: "isContentEmpty",
    value: function isContentEmpty(node) {
      return node.content.childCount === 0;
    }
  }, {
    key: "createDomRef",
    value: function createDomRef() {
      var domRef = document.createElement('div');
      domRef.style.listStyleType = 'none';
      return domRef;
    }
  }, {
    key: "getContentDOM",
    value: function getContentDOM() {
      var dom = document.createElement('div');
      // setting a className prevents PM/Chrome mutation observer from
      // incorrectly deleting nodes
      dom.className = 'task-item';
      return {
        dom: dom
      };
    }
  }, {
    key: "render",
    value: function render(props, forwardRef) {
      var _this$node$attrs = this.node.attrs,
        localId = _this$node$attrs.localId,
        state = _this$node$attrs.state;
      var isContentNodeEmpty = this.isContentEmpty(this.node);
      return /*#__PURE__*/React.createElement(_analyticsNext.AnalyticsListener, {
        channel: "fabric-elements",
        onEvent: this.addListAnalyticsData
      }, /*#__PURE__*/React.createElement(TaskItemWrapper, {
        localId: localId,
        forwardRef: forwardRef,
        isDone: state === 'DONE',
        onChange: this.handleOnChange,
        isContentNodeEmpty: isContentNodeEmpty,
        placeholder: props.placeholder,
        providerFactory: props.providerFactory
        // The getPosHandler type is wrong, there is no `boolean` in the real implementation
        // @ts-expect-error 2322: Type 'getPosHandler' is not assignable to type '() => number | undefined'.
        ,
        getPos: this.getPos,
        editorView: this.view,
        api: this.api
      }));
    }
  }, {
    key: "viewShouldUpdate",
    value: function viewShouldUpdate(nextNode) {
      return this.isContentEmpty(this.node) && !this.isContentEmpty(nextNode) || this.isContentEmpty(nextNode) && !this.isContentEmpty(this.node);
    }
  }, {
    key: "update",
    value: function update(node, decorations) {
      return _superPropGet(Task, "update", this, 3)([node, decorations, undefined, function (currentNode, newNode) {
        return !!(currentNode.attrs.state === newNode.attrs.state);
      }]);
    }
  }, {
    key: "ignoreMutation",
    value: function ignoreMutation(mutation) {
      if (!this.contentDOM) {
        return true;
      }
      return !this.contentDOM.contains(mutation.target) && mutation.type !== 'selection';
    }
  }]);
}(_reactNodeView.default);
function taskItemNodeViewFactory(portalProviderAPI, eventDispatcher, providerFactory, api, intl, placeholder) {
  return function (node, view, getPos) {
    return new Task(node, view, getPos, portalProviderAPI, eventDispatcher, {
      placeholder: placeholder,
      providerFactory: providerFactory
    }).initWithAPI(api);
  };
}
