// @atlaskit/profilecard/dist/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AgentProfileCard", {
  enumerable: true,
  get: function get() {
    return _AgentProfileCard.default;
  }
});
Object.defineProperty(exports, "AgentProfileCardResourced", {
  enumerable: true,
  get: function get() {
    return _AgentProfileCardResourced.AgentProfileCardResourced;
  }
});
Object.defineProperty(exports, "AgentProfileCardTrigger", {
  enumerable: true,
  get: function get() {
    return _AgentProfileCardTrigger.AgentProfileCardTrigger;
  }
});
Object.defineProperty(exports, "DELAY_MS_HIDE", {
  enumerable: true,
  get: function get() {
    return _config.DELAY_MS_HIDE;
  }
});
Object.defineProperty(exports, "DELAY_MS_SHOW", {
  enumerable: true,
  get: function get() {
    return _config.DELAY_MS_SHOW;
  }
});
Object.defineProperty(exports, "ProfileCard", {
  enumerable: true,
  get: function get() {
    return _ProfileCard.default;
  }
});
Object.defineProperty(exports, "ProfileCardTrigger", {
  enumerable: true,
  get: function get() {
    return _ProfileCardTrigger.default;
  }
});
Object.defineProperty(exports, "ProfileClient", {
  enumerable: true,
  get: function get() {
    return _ProfileCardClient.default;
  }
});
Object.defineProperty(exports, "TeamProfileCard", {
  enumerable: true,
  get: function get() {
    return _TeamProfileCard.default;
  }
});
Object.defineProperty(exports, "TeamProfileCardTrigger", {
  enumerable: true,
  get: function get() {
    return _TeamProfileCardTrigger.default;
  }
});
Object.defineProperty(exports, "TeamProfileCardWithContainer", {
  enumerable: true,
  get: function get() {
    return _teamProfileCard.TeamProfileCard;
  }
});
Object.defineProperty(exports, "TeamProfileClient", {
  enumerable: true,
  get: function get() {
    return _TeamProfileCardClient.default;
  }
});
Object.defineProperty(exports, "UserProfileClient", {
  enumerable: true,
  get: function get() {
    return _UserProfileCardClient.default;
  }
});
exports.default = void 0;
Object.defineProperty(exports, "modifyResponse", {
  enumerable: true,
  get: function get() {
    return _UserProfileCardClient.modifyResponse;
  }
});
Object.defineProperty(exports, "withOuterListeners", {
  enumerable: true,
  get: function get() {
    return _withOuterListeners.default;
  }
});
var _ProfileCardClient = _interopRequireDefault(require("./client/ProfileCardClient"));
var _TeamProfileCardClient = _interopRequireDefault(require("./client/TeamProfileCardClient"));
var _UserProfileCardClient = _interopRequireWildcard(require("./client/UserProfileCardClient"));
var _AgentProfileCard = _interopRequireDefault(require("./components/Agent/AgentProfileCard"));
var _teamProfileCard = require("./components/team-profile-card");
var _TeamProfileCardTrigger = _interopRequireDefault(require("./components/Team/TeamProfileCardTrigger"));
var _ProfileCard = _interopRequireDefault(require("./components/User/ProfileCard"));
var _ProfileCardResourced = _interopRequireDefault(require("./components/User/ProfileCardResourced"));
var _ProfileCardTrigger = _interopRequireDefault(require("./components/User/ProfileCardTrigger"));
var _config = require("./util/config");
var _withOuterListeners = _interopRequireDefault(require("./util/withOuterListeners"));
var _TeamProfileCard = _interopRequireDefault(require("./components/Team/TeamProfileCard"));
var _AgentProfileCardTrigger = require("./components/Agent/AgentProfileCardTrigger");
var _AgentProfileCardResourced = require("./components/Agent/AgentProfileCardResourced");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
// Avoid exporting new components in this file as they will affect the bundle size of all consumers
// importing directly from @atlaskit/profilecard.
// Note: when generating and updating Flow types in Jira, `TeamProfileClient` and `TeamProfileCardClient` are the same type
// Note: when generating and updating Flow types in Jira, `UserProfileClient` and `UserProfileCardClient` are the same type
// Legacy Profile Card in Jira still needs to use `withOuterListeners`
// We cannot export this component inside `./components/Team/index.ts` because it breaks the lazyload
// so we have to export this here. `TeamProfileCard` helps to build legacy team profile card inside Jira FE.
var _default = exports.default = _ProfileCardResourced.default;
