// @atlaskit/editor-plugin-selection-extension/dist/es2019/pm-plugins/main.js
import { SafePlugin } from '@atlaskit/editor-common/safe-plugin';
import { PluginKey } from '@atlaskit/editor-prosemirror/state';
import { SelectionExtensionActionTypes } from '../types';
export const selectionExtensionPluginKey = new PluginKey('selectionExtensionPlugin');
export const createPlugin = () => {
  return new SafePlugin({
    key: selectionExtensionPluginKey,
    state: {
      init: () => {
        return {
          activeExtension: undefined
        };
      },
      apply: (tr, pluginState) => {
        const meta = tr.getMeta(selectionExtensionPluginKey);
        switch (meta === null || meta === void 0 ? void 0 : meta.type) {
          case SelectionExtensionActionTypes.SET_ACTIVE_EXTENSION:
            return {
              ...pluginState,
              activeExtension: meta.extension
            };
          case SelectionExtensionActionTypes.CLEAR_ACTIVE_EXTENSION:
            return {
              ...pluginState,
              activeExtension: undefined
            };
        }
        return pluginState;
      }
    }
  });
};
