// @atlaskit/editor-common/dist/cjs/ui-color/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Color", {
  enumerable: true,
  get: function get() {
    return _Color.default;
  }
});
Object.defineProperty(exports, "ColorPalette", {
  enumerable: true,
  get: function get() {
    return _ColorPalette.default;
  }
});
Object.defineProperty(exports, "DEFAULT_BORDER_COLOR", {
  enumerable: true,
  get: function get() {
    return _common.DEFAULT_BORDER_COLOR;
  }
});
Object.defineProperty(exports, "DEFAULT_COLOR_PICKER_COLUMNS", {
  enumerable: true,
  get: function get() {
    return _utils.DEFAULT_COLOR_PICKER_COLUMNS;
  }
});
Object.defineProperty(exports, "REMOVE_HIGHLIGHT_COLOR", {
  enumerable: true,
  get: function get() {
    return _highlightColorPalette.REMOVE_HIGHLIGHT_COLOR;
  }
});
Object.defineProperty(exports, "backgroundPaletteTooltipMessages", {
  enumerable: true,
  get: function get() {
    return _paletteMessagesTokenModeNames.backgroundPaletteTooltipMessages;
  }
});
Object.defineProperty(exports, "borderColorPalette", {
  enumerable: true,
  get: function get() {
    return _borderColorPalette.default;
  }
});
Object.defineProperty(exports, "borderPaletteTooltipMessages", {
  enumerable: true,
  get: function get() {
    return _paletteMessagesTokenModeNames.borderPaletteTooltipMessages;
  }
});
Object.defineProperty(exports, "cellBackgroundColorPalette", {
  enumerable: true,
  get: function get() {
    return _cellBackgroundColorPalette.default;
  }
});
Object.defineProperty(exports, "chartsColorPaletteTooltipMessages", {
  enumerable: true,
  get: function get() {
    return _paletteMessagesTokenModeNames.chartsColorPaletteTooltipMessages;
  }
});
Object.defineProperty(exports, "colorPaletteMessages", {
  enumerable: true,
  get: function get() {
    return _paletteMessages.default;
  }
});
Object.defineProperty(exports, "getColorsPerRowFromPalette", {
  enumerable: true,
  get: function get() {
    return _utils.getColorsPerRowFromPalette;
  }
});
Object.defineProperty(exports, "getSelectedRowAndColumn", {
  enumerable: true,
  get: function get() {
    return _utils.getSelectedRowAndColumn;
  }
});
Object.defineProperty(exports, "getSelectedRowAndColumnFromPalette", {
  enumerable: true,
  get: function get() {
    return _utils.getSelectedRowAndColumnFromPalette;
  }
});
Object.defineProperty(exports, "highlightColorPalette", {
  enumerable: true,
  get: function get() {
    return _highlightColorPalette.highlightColorPalette;
  }
});
Object.defineProperty(exports, "panelBackgroundPalette", {
  enumerable: true,
  get: function get() {
    return _panelBackgroundPalette.panelBackgroundPalette;
  }
});
Object.defineProperty(exports, "panelDarkModeBackgroundPalette", {
  enumerable: true,
  get: function get() {
    return _panelBackgroundPalette.panelDarkModeBackgroundPalette;
  }
});
Object.defineProperty(exports, "textColorPalette", {
  enumerable: true,
  get: function get() {
    return _textColorPalette.textColorPalette;
  }
});
Object.defineProperty(exports, "textPaletteTooltipMessages", {
  enumerable: true,
  get: function get() {
    return _paletteMessagesTokenModeNames.textPaletteTooltipMessages;
  }
});
var _ColorPalette = _interopRequireDefault(require("./ColorPalette"));
var _Color = _interopRequireDefault(require("./ColorPalette/Color"));
var _utils = require("./ColorPalette/utils");
var _cellBackgroundColorPalette = _interopRequireDefault(require("./ColorPalette/Palettes/cellBackgroundColorPalette"));
var _paletteMessages = _interopRequireDefault(require("./ColorPalette/Palettes/paletteMessages"));
var _panelBackgroundPalette = require("./ColorPalette/Palettes/panelBackgroundPalette");
var _textColorPalette = require("./ColorPalette/Palettes/textColorPalette");
var _highlightColorPalette = require("./ColorPalette/Palettes/highlightColorPalette");
var _paletteMessagesTokenModeNames = require("./ColorPalette/Palettes/paletteMessagesTokenModeNames");
var _common = require("./ColorPalette/Palettes/common");
var _borderColorPalette = _interopRequireDefault(require("./ColorPalette/Palettes/borderColorPalette"));
