// @atlaskit/smart-card/dist/cjs/view/RelatedLinksModal/components/related-links-list/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./index.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _reactIntlNext = require("react-intl-next");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _compiled = require("@atlaskit/primitives/compiled");
var _messages = require("../../../../messages");
var _relatedLinkItem = _interopRequireDefault(require("../related-link-item"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var styles = {
  sectionTitleOld: "_11c81vhk _1q511b66",
  sectionTitle: "_11c81vhk _syaz1gjq _1q511b66",
  boxStyle: "_11c81o8v _1rjcu2gc"
};
var RelatedLinksList = function RelatedLinksList(_ref) {
  var urls = _ref.urls,
    title = _ref.title,
    testId = _ref.testId,
    selected = _ref.selected,
    handleSelectedUpdate = _ref.handleSelectedUpdate;
  var TitleWrapper = function TitleWrapper(_ref2) {
    var children = _ref2.children;
    return (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v2') ? /*#__PURE__*/React.createElement(_compiled.Box, {
      xcss: styles.sectionTitle
    }, children) : /*#__PURE__*/React.createElement(_compiled.Box, {
      xcss: styles.sectionTitleOld
    }, children);
  };
  return /*#__PURE__*/React.createElement(_compiled.Stack, {
    testId: testId
  }, /*#__PURE__*/React.createElement(TitleWrapper, null, /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, title)), urls.length > 0 && /*#__PURE__*/React.createElement(_compiled.Box, null, urls.map(function (url, idx) {
    return /*#__PURE__*/React.createElement(_compiled.Stack, {
      key: "".concat(idx, "-").concat(url)
    }, /*#__PURE__*/React.createElement(_relatedLinkItem.default, {
      url: url,
      testId: "".concat(testId, "-item-").concat(idx),
      isSelected: selected === "".concat(idx, "-").concat(url),
      onFocus: function onFocus() {
        return handleSelectedUpdate && handleSelectedUpdate("".concat(idx, "-").concat(url));
      }
    }));
  })), urls.length === 0 && /*#__PURE__*/React.createElement(_compiled.Box, {
    xcss: styles.boxStyle
  }, /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, _messages.messages.related_links_not_found)));
};
var _default = exports.default = RelatedLinksList;
