// @atlaskit/react-select/dist/es2019/components/indicators.js
/* eslint-disable @repo/internal/react/no-unsafe-spread-props */
import React from 'react';
import { fg } from '@atlaskit/platform-feature-flags';
import { ClearIndicator as CompiledClearIndicator, clearIndicatorCSS as compiledClearIndicatorCSS, DropdownIndicator as CompiledDropdownIndicator, dropdownIndicatorCSS as compiledDropdownIndicatorCSS, LoadingIndicator as CompiledLoadingIndicator, loadingIndicatorCSS as compiledLoadingIndicatorCSS } from '../compiled/components/indicators';
import { ClearIndicator as EmotionClearIndicator, clearIndicatorCSS as emotionClearIndicatorCSS, DropdownIndicator as EmotionDropdownIndicator, dropdownIndicatorCSS as emotionDropdownIndicatorCSS, LoadingIndicator as EmotionLoadingIndicator, loadingIndicatorCSS as emotionLoadingIndicatorCSS } from '../emotion/components/indicators';

// ==============================
// Dropdown & Clear Buttons
// ==============================

export const dropdownIndicatorCSS = props => fg('compiled-react-select') ? compiledDropdownIndicatorCSS() : emotionDropdownIndicatorCSS(props);

// eslint-disable-next-line @repo/internal/react/require-jsdoc
export const DropdownIndicator = props => fg('compiled-react-select') ? /*#__PURE__*/React.createElement(CompiledDropdownIndicator, props) : /*#__PURE__*/React.createElement(EmotionDropdownIndicator, props);
export const clearIndicatorCSS = props => fg('compiled-react-select') ? compiledClearIndicatorCSS() : emotionClearIndicatorCSS(props);

// eslint-disable-next-line @repo/internal/react/require-jsdoc
export const ClearIndicator = props => fg('compiled-react-select') ? /*#__PURE__*/React.createElement(CompiledClearIndicator, props) : /*#__PURE__*/React.createElement(EmotionClearIndicator, props);

// ==============================
// Loading
// ==============================

export const loadingIndicatorCSS = props => fg('compiled-react-select') ? compiledLoadingIndicatorCSS() : emotionLoadingIndicatorCSS(props);
// eslint-disable-next-line @repo/internal/react/require-jsdoc
export const LoadingIndicator = props => fg('compiled-react-select') ? /*#__PURE__*/React.createElement(CompiledLoadingIndicator, props) : /*#__PURE__*/React.createElement(EmotionLoadingIndicator, props);
