// @atlaskit/editor-common/dist/cjs/link/LinkSearch/ToolbarComponents.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.inputWrapper = exports.containerWithProvider = exports.container = void 0;
var _react = require("@emotion/react");
var _ui = require("../../ui");
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var inputWrapper = exports.inputWrapper = (0, _react.css)({
  display: 'flex',
  // eslint-disable-next-line @atlaskit/design-system/use-tokens-typography
  lineHeight: 0,
  padding: "var(--ds-space-075, 6px)".concat(" 0"),
  alignItems: 'center'
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var container = exports.container = (0, _react.css)({
  display: 'flex',
  flexDirection: 'column',
  overflow: 'auto',
  padding: 0,
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  width: "".concat(_ui.RECENT_SEARCH_WIDTH_WITHOUT_ITEMS_IN_PX, "px"),
  // eslint-disable-next-line @atlaskit/design-system/use-tokens-typography
  lineHeight: 'initial'
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var containerWithProvider = exports.containerWithProvider = (0, _react.css)({
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  width: "".concat(_ui.RECENT_SEARCH_WIDTH_IN_PX, "px")
});
