// @atlaskit/frontend-utilities/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "FailedFetchError", {
  enumerable: true,
  get: function get() {
    return _retryOperation.FailedFetchError;
  }
});
Object.defineProperty(exports, "STORAGE_MOCK", {
  enumerable: true,
  get: function get() {
    return _localStorage.STORAGE_MOCK;
  }
});
Object.defineProperty(exports, "StorageClient", {
  enumerable: true,
  get: function get() {
    return _storageClient.StorageClient;
  }
});
Object.defineProperty(exports, "convertToError", {
  enumerable: true,
  get: function get() {
    return _convertToError.convertToError;
  }
});
Object.defineProperty(exports, "mockWindowStorage", {
  enumerable: true,
  get: function get() {
    return _localStorage.mockWindowStorage;
  }
});
Object.defineProperty(exports, "retryOnException", {
  enumerable: true,
  get: function get() {
    return _retryOperation.retryOnException;
  }
});
Object.defineProperty(exports, "simpleHash", {
  enumerable: true,
  get: function get() {
    return _simpleHash.simpleHash;
  }
});
Object.defineProperty(exports, "useInterval", {
  enumerable: true,
  get: function get() {
    return _hooks.useInterval;
  }
});
Object.defineProperty(exports, "useLocalStorage", {
  enumerable: true,
  get: function get() {
    return _hooks.useLocalStorage;
  }
});
Object.defineProperty(exports, "useLocalStorageRecord", {
  enumerable: true,
  get: function get() {
    return _hooks.useLocalStorageRecord;
  }
});
Object.defineProperty(exports, "usePrevious", {
  enumerable: true,
  get: function get() {
    return _hooks.usePrevious;
  }
});
var _convertToError = require("./error-handling/convert-to-error");
var _retryOperation = require("./network/retry-operation");
var _localStorage = require("./storage/local-storage");
var _storageClient = require("./storage/storage-client");
var _hooks = require("./hooks");
var _simpleHash = require("./utils/simple-hash");
