// @atlaskit/media-card/dist/cjs/card/ui/unhandledErrorCard/unhandledErrorCard-emotion.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UnhandledErrorCard = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _react = require("react");
var _react2 = require("@emotion/react");
var _warningEditorWarning = _interopRequireDefault(require("@atlaskit/icon/core/migration/warning--editor-warning"));
var _colors = require("@atlaskit/theme/colors");
var _mediaUi = require("@atlaskit/media-ui");
var _utils = require("../../../utils");
var _contentLoadingErrorMessage = require("./contentLoadingErrorMessage");
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); } /**
 * @jsxRuntime classic
 * @jsx jsx
 */ // eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
var isPercentage = function isPercentage(value) {
  return /^\d+(\.\d+)?%$/.test(value);
};
var shouldShowText = function shouldShowText(_ref) {
  var width = _ref.width,
    height = _ref.height;
  if (isPercentage(width) || isPercentage(height)) {
    return false;
  }
  return parseInt(width, 10) >= 240 && parseInt(height, 10) >= 90;
};
var normalizeDimension = function normalizeDimension(value, defaultValue) {
  var sValue = value.toString();
  return isPercentage(sValue) ? sValue : isNaN(parseInt(sValue, 10)) ? defaultValue + 'px' : parseInt(sValue, 10) + 'px';
};
var getConvertedDimension = function getConvertedDimension(dimensions) {
  var _dimensions$width = dimensions.width,
    width = _dimensions$width === void 0 ? _utils.defaultImageCardDimensions.width : _dimensions$width,
    _dimensions$height = dimensions.height,
    height = _dimensions$height === void 0 ? _utils.defaultImageCardDimensions.height : _dimensions$height;
  return {
    width: normalizeDimension(width, _utils.defaultImageCardDimensions.width),
    height: normalizeDimension(height, _utils.defaultImageCardDimensions.height)
  };
};
var wrapperStyles = function wrapperStyles() {
  var dimensions = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _utils.defaultImageCardDimensions;
  try {
    return (0, _react2.css)(
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    _mediaUi.center,
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    _mediaUi.borderRadius, {
      background: "var(--ds-background-neutral, ".concat(_colors.N20, ")"),
      color: "var(--ds-text-subtle, ".concat(_colors.N800, ")"),
      maxHeight: '100%',
      maxWidth: '100%'
    },
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    getConvertedDimension(dimensions), {
      display: 'flex',
      flexDirection: 'column',
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
      p: {
        font: "var(--ds-font-body, normal 400 14px/20px ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
        textAlign: 'center'
      }
    });
  } catch (e) {
    return null;
  }
};
var UnhandledErrorCard = exports.UnhandledErrorCard = /*#__PURE__*/function (_Component) {
  function UnhandledErrorCard() {
    (0, _classCallCheck2.default)(this, UnhandledErrorCard);
    return _callSuper(this, UnhandledErrorCard, arguments);
  }
  (0, _inherits2.default)(UnhandledErrorCard, _Component);
  return (0, _createClass2.default)(UnhandledErrorCard, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
        dimensions = _this$props.dimensions,
        onClick = _this$props.onClick;
      var isErrorHidden = !shouldShowText(getConvertedDimension(dimensions !== null && dimensions !== void 0 ? dimensions : _utils.defaultImageCardDimensions));
      return (
        // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions -- Ignored via go/DSP-18766
        (0, _react2.jsx)("div", {
          css: wrapperStyles(dimensions),
          onClick: onClick
        }, (0, _react2.jsx)(_warningEditorWarning.default, {
          label: "Error",
          color: "var(--ds-icon-warning, ".concat(_colors.Y500, ")"),
          LEGACY_size: "medium",
          spacing: "spacious"
        }), (0, _react2.jsx)(_contentLoadingErrorMessage.ContentLoadingErrorMessage, {
          isHidden: isErrorHidden
        }))
      );
    }
  }]);
}(_react.Component);
