// @atlaskit/editor-plugin-mentions/dist/cjs/ui/InviteItem/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.INVITE_ITEM_DESCRIPTION = void 0;
require("./index.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _reactIntlNext = require("react-intl-next");
var _messages = require("@atlaskit/editor-common/messages");
var _add = _interopRequireDefault(require("@atlaskit/icon/core/migration/add"));
var _colors = require("@atlaskit/theme/colors");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var mentionItemStyle = null;
var mentionItemSelectedStyle = null;
var rowStyle = null;
var avatarStyle = null;
var nameSectionStyle = null;
var capitalizedStyle = null;
var INVITE_ITEM_DESCRIPTION = exports.INVITE_ITEM_DESCRIPTION = {
  id: 'invite-teammate'
};

// Ignored via go/ees005
// eslint-disable-next-line @typescript-eslint/no-explicit-any
var leftClick = function leftClick(event) {
  return event.button === 0 && !event.altKey && !event.ctrlKey && !event.metaKey && !event.shiftKey;
};
var InviteItem = function InviteItem(_ref) {
  var productName = _ref.productName,
    onMount = _ref.onMount,
    onMouseEnter = _ref.onMouseEnter,
    onSelection = _ref.onSelection,
    selected = _ref.selected,
    userRole = _ref.userRole,
    intl = _ref.intl;
  var onSelected = (0, _react.useCallback)(
  // Ignored via go/ees005
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  function (event) {
    if (leftClick(event) && onSelection) {
      event.preventDefault();
      onSelection(INVITE_ITEM_DESCRIPTION, event);
    }
  }, [onSelection]);
  var onItemMouseEnter = (0, _react.useCallback)(
  // Ignored via go/ees005
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  function (event) {
    if (onMouseEnter) {
      onMouseEnter(INVITE_ITEM_DESCRIPTION, event);
    }
  }, [onMouseEnter]);
  (0, _react.useEffect)(function () {
    if (onMount) {
      onMount();
    }
  }, [onMount]);
  return /*#__PURE__*/_react.default.createElement("div", {
    onMouseDown: onSelected,
    onMouseEnter: onItemMouseEnter,
    "data-id": INVITE_ITEM_DESCRIPTION.id,
    className: (0, _runtime.ax)(["_1reo15vq _18m915vq _bfhk1j28 _1e0c1ule _2mzuglyw _80omtlke", selected && "_bfhk2tgk"]),
    style: {
      "--_1y60f1n": (0, _runtime.ix)("var(--ds-background-neutral-subtle-hovered, ".concat(_colors.N30, ")"))
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _runtime.ax)(["_1reo15vq _18m915vq _4cvr1h6o _1e0c1txw _2lx2vrvc _1n261g80 _ca0q12x7 _n3td12x7 _19bvdlk8 _u5f3dlk8 _1bto1l2s _s7n4nkob"])
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: (0, _runtime.ax)(["_16jlidpf _1o9zkb7n _i0dl1wug _1e0c1txw _4cvr1h6o _1bah1h6o _kqswh2mm _1bsb14no _4t3i14no"])
  }, /*#__PURE__*/_react.default.createElement(_add.default, {
    label: intl.formatMessage(_messages.mentionMessages.mentionsAddLabel),
    color: "var(--ds-icon-subtle, ".concat(_colors.N300, ")")
  })), /*#__PURE__*/_react.default.createElement("div", {
    "data-testid": "name-section",
    className: (0, _runtime.ax)(["_16jlkb7n _1o9zkb7n _i0dlf1ug _1ul9idpf _18u0dlk8 _syaz1vvm"]),
    style: {
      "--_bbz764": (0, _runtime.ix)("var(--ds-text-subtle, ".concat(_colors.N300, ")"))
    }
  }, /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage
  // Ignored via go/ees005
  // eslint-disable-next-line react/jsx-props-no-spreading
  , (0, _extends2.default)({}, _messages.mentionMessages.inviteItemTitle, {
    values: {
      userRole: userRole || 'basic',
      productName: /*#__PURE__*/_react.default.createElement("span", {
        "data-testid": "capitalized-message",
        className: (0, _runtime.ax)(["_1p1d1dk0"])
      }, productName)
    }
  })))));
};
var _default = exports.default = (0, _reactIntlNext.injectIntl)(InviteItem);
