// @atlaskit/link-datasource/dist/cjs/ui/table-footer/sync-info/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SyncInfo = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _reactIntlNext = require("react-intl-next");
var _messages = require("./messages");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var SECONDS_IN_MIN = 60;
var SECONDS_IN_HR = SECONDS_IN_MIN * 60;
var SECONDS_IN_DAY = SECONDS_IN_HR * 24;
var WithUpdatedLabel = function WithUpdatedLabel(formattedDate) {
  var i18nMessage = formattedDate ? _messages.messages.overOneMinuteText : _messages.messages.underOneMinuteText;
  return /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, (0, _extends2.default)({}, i18nMessage, {
    values: {
      date: formattedDate
    }
  }));
};
var SyncInfo = exports.SyncInfo = function SyncInfo(_ref) {
  var lastSyncTime = _ref.lastSyncTime;
  var calculateTimeDiff = (0, _react.useCallback)(function () {
    return Math.floor((Date.now() - lastSyncTime.getTime()) / 1000);
  }, [lastSyncTime]);
  var _useState = (0, _react.useState)(calculateTimeDiff()),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    secondsSinceUpdate = _useState2[0],
    setSecondsSinceUpdate = _useState2[1];
  var totalDays = Math.floor(secondsSinceUpdate / SECONDS_IN_DAY);
  var totalHours = Math.floor(secondsSinceUpdate / SECONDS_IN_HR);
  var totalMinutes = Math.floor(secondsSinceUpdate / SECONDS_IN_MIN);
  (0, _react.useEffect)(function () {
    setSecondsSinceUpdate(calculateTimeDiff());
    var interval = setInterval(function () {
      return setSecondsSinceUpdate(calculateTimeDiff());
    }, 1000);
    return function () {
      return clearInterval(interval);
    };
  }, [lastSyncTime, calculateTimeDiff]);
  if (totalMinutes >= 1 && totalMinutes < 60) {
    return /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedRelativeTime, {
      value: -totalMinutes,
      style: "long",
      unit: "minute"
    }, WithUpdatedLabel);
  }
  if (totalHours >= 1 && totalHours < 24) {
    return /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedRelativeTime, {
      value: -totalHours,
      style: "long",
      unit: "hour"
    }, WithUpdatedLabel);
  }
  if (totalDays >= 1 && totalDays < 8) {
    return /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedRelativeTime, {
      value: -totalDays,
      style: "long",
      unit: "day"
    }, WithUpdatedLabel);
  }
  if (totalDays >= 8) {
    return /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedDate, {
      year: "numeric",
      month: "short",
      day: "2-digit",
      value: lastSyncTime
    }, WithUpdatedLabel);
  }
  return WithUpdatedLabel();
};
