// @atlaskit/media-viewer/dist/esm/newgen/media-viewer.js
import { __extends } from "tslib";
import React from 'react';
import { IntlProvider, intlShape } from 'react-intl';
import { Shortcut } from '@atlaskit/media-ui';
import { withAnalyticsEvents, } from '@atlaskit/analytics-next';
import { mediaViewerModalEvent } from './analytics/media-viewer';
import { closedEvent } from './analytics/closed';
import { channel } from './analytics/index';
import { List } from './list';
import { Collection } from './collection';
import { Content } from './content';
import { Blanket, SidebarWrapper } from './styled';
import { start } from 'perf-marks';
var MediaViewerComponent = /** @class */ (function (_super) {
    __extends(MediaViewerComponent, _super);
    function MediaViewerComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            isSidebarVisible: false,
        };
        _this.fireAnalytics = function (payload) {
            var createAnalyticsEvent = _this.props.createAnalyticsEvent;
            if (createAnalyticsEvent) {
                var ev = createAnalyticsEvent(payload);
                ev.fire(channel);
            }
        };
        _this.onShortcutClosed = function () {
            _this.sendClosedEvent('escKey');
            var onClose = _this.props.onClose;
            if (onClose) {
                onClose();
            }
        };
        _this.onContentClose = function (_e, analyticsEvent) {
            var onClose = _this.props.onClose;
            if (analyticsEvent &&
                analyticsEvent.payload &&
                analyticsEvent.payload.actionSubject === 'button') {
                _this.sendClosedEvent('button');
            }
            if (onClose) {
                onClose();
            }
        };
        _this.toggleSidebar = function () {
            _this.setState({
                isSidebarVisible: !_this.state.isSidebarVisible,
            });
        };
        _this.renderSidebar = function () {
            var extensions = _this.props.extensions;
            var _a = _this.state, isSidebarVisible = _a.isSidebarVisible, selectedIdentifier = _a.selectedIdentifier;
            var sidebardSelectedIdentifier = selectedIdentifier || _this.defaultSelectedItem;
            if (sidebardSelectedIdentifier &&
                isSidebarVisible &&
                extensions &&
                extensions.sidebar) {
                return (React.createElement(SidebarWrapper, { "data-testid": "media-viewer-sidebar-content" }, extensions.sidebar.renderer(sidebardSelectedIdentifier, {
                    close: _this.toggleSidebar,
                })));
            }
        };
        _this.onNavigationChange = function (selectedIdentifier) {
            _this.setState({ selectedIdentifier: selectedIdentifier });
        };
        return _this;
    }
    MediaViewerComponent.prototype.UNSAFE_componentWillMount = function () {
        this.fireAnalytics(mediaViewerModalEvent());
        start('MediaViewer.SessionDuration');
    };
    MediaViewerComponent.prototype.sendClosedEvent = function (input) {
        this.fireAnalytics(closedEvent(input));
    };
    Object.defineProperty(MediaViewerComponent.prototype, "defaultSelectedItem", {
        get: function () {
            var _a = this.props, itemSource = _a.itemSource, selectedItem = _a.selectedItem;
            if (itemSource.kind === 'COLLECTION') {
                return selectedItem;
            }
            var items = itemSource.items;
            var firstItem = items[0];
            return selectedItem || firstItem;
        },
        enumerable: true,
        configurable: true
    });
    MediaViewerComponent.prototype.render = function () {
        var content = (React.createElement(Blanket, { "data-testid": "media-viewer-popup" },
            React.createElement(Shortcut, { keyCode: 27, handler: this.onShortcutClosed }),
            React.createElement(Content, { onClose: this.onContentClose }, this.renderContent()),
            this.renderSidebar()));
        return this.context.intl ? (content) : (React.createElement(IntlProvider, { locale: "en" }, content));
    };
    MediaViewerComponent.prototype.renderContent = function () {
        var _a = this.props, mediaClient = _a.mediaClient, onClose = _a.onClose, itemSource = _a.itemSource, extensions = _a.extensions, contextId = _a.contextId;
        var isSidebarVisible = this.state.isSidebarVisible;
        if (itemSource.kind === 'COLLECTION') {
            return (React.createElement(Collection, { pageSize: itemSource.pageSize, defaultSelectedItem: this.defaultSelectedItem, collectionName: itemSource.collectionName, mediaClient: mediaClient, onClose: onClose, extensions: extensions, onNavigationChange: this.onNavigationChange, onSidebarButtonClick: this.toggleSidebar }));
        }
        else if (itemSource.kind === 'ARRAY') {
            var items = itemSource.items;
            return (React.createElement(List, { defaultSelectedItem: this.defaultSelectedItem || items[0], items: items, mediaClient: mediaClient, onClose: onClose, extensions: extensions, onNavigationChange: this.onNavigationChange, onSidebarButtonClick: this.toggleSidebar, isSidebarVisible: isSidebarVisible, contextId: contextId }));
        }
        else {
            return null;
        }
    };
    MediaViewerComponent.contextTypes = {
        intl: intlShape,
    };
    return MediaViewerComponent;
}(React.Component));
export { MediaViewerComponent };
export var MediaViewer = withAnalyticsEvents()(MediaViewerComponent);
