// @atlaskit/give-kudos/dist/cjs/ui/GiveKudosLauncher/main.js
/* main.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./main.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _reactIntlNext = require("react-intl-next");
var _analyticsNext = require("@atlaskit/analytics-next");
var _new = require("@atlaskit/button/new");
var _standardButton = _interopRequireDefault(require("@atlaskit/button/standard-button"));
var _compiled = require("@atlaskit/drawer/compiled");
var _linkExternal = _interopRequireDefault(require("@atlaskit/icon/core/link-external"));
var _arrowLeft = _interopRequireDefault(require("@atlaskit/icon/core/migration/arrow-left"));
var _intlMessagesProvider = require("@atlaskit/intl-messages-provider");
var _link = _interopRequireDefault(require("@atlaskit/link"));
var _modalDialog = _interopRequireWildcard(require("@atlaskit/modal-dialog"));
var _portal = _interopRequireDefault(require("@atlaskit/portal"));
var _primitives = require("@atlaskit/primitives");
var _constants = require("@atlaskit/theme/constants");
var _fetchMessagesForLocale = require("../../common/utils/fetch-messages-for-locale");
var _en = _interopRequireDefault(require("../../i18n/en"));
var _messages = _interopRequireDefault(require("../../messages"));
var _types = require("../../types");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var styles = {
  drawerCloseButtonContainer: "_kqswstnw _154ipxbi _1ltvpxbi",
  iframe: "_19itglyw"
};
var ANALYTICS_CHANNEL = 'atlas';
var GiveKudosLauncher = function GiveKudosLauncher(props) {
  var _props$recipient4;
  var _useState = (0, _react.useState)(false),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    isCloseConfirmModalOpen = _useState2[0],
    setIsCloseConfirmModalOpen = _useState2[1];
  var _useState3 = (0, _react.useState)(false),
    _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
    isDirty = _useState4[0],
    setIsDirty = _useState4[1];
  var iframeEl = (0, _react.useRef)(null);
  var messageListenerEventHandler = (0, _react.useRef)(function (e) {});
  var unloadEventHandler = (0, _react.useRef)(function (e) {});
  var intl = (0, _reactIntlNext.useIntl)();
  var _useAnalyticsEvents = (0, _analyticsNext.useAnalyticsEvents)(),
    createAnalyticsEvent = _useAnalyticsEvents.createAnalyticsEvent;
  var addFlag = props.addFlag,
    teamCentralBaseUrl = props.teamCentralBaseUrl,
    analyticsSource = props.analyticsSource,
    onClose = props.onClose,
    testId = props.testId,
    onCreateKudosSuccess = props.onCreateKudosSuccess,
    isActionsEnabled = props.isActionsEnabled;
  var shouldBlockTransition = (0, _react.useCallback)(function (e) {
    e.preventDefault();
    e.returnValue = intl.formatMessage(_messages.default.unsavedKudosWarning);
  }, [intl]);
  var sendAnalytic = (0, _react.useCallback)(function (action, options) {
    var _props$recipient, _props$recipient2, _props$recipient3;
    var analyticsEvent = createAnalyticsEvent({
      action: action,
      actionSubject: 'createKudos',
      attributes: _objectSpread(_objectSpread({}, options), {}, {
        analyticsSource: analyticsSource,
        teamId: ((_props$recipient = props.recipient) === null || _props$recipient === void 0 ? void 0 : _props$recipient.type) === _types.KudosType.TEAM ? props.recipient.recipientId : undefined,
        KudosType: (_props$recipient2 = props.recipient) === null || _props$recipient2 === void 0 ? void 0 : _props$recipient2.type,
        recipientId: (_props$recipient3 = props.recipient) === null || _props$recipient3 === void 0 ? void 0 : _props$recipient3.recipientId
      })
    });
    analyticsEvent.fire(ANALYTICS_CHANNEL);
  }, [analyticsSource, createAnalyticsEvent, props.recipient]);
  var closeDrawer = (0, _react.useCallback)(function () {
    setIsDirty(false);
    setIsCloseConfirmModalOpen(false);
    onClose();
  }, [onClose]);
  var closeWarningModal = function closeWarningModal() {
    setIsCloseConfirmModalOpen(false);
  };
  var createFlagWithJsonStringifiedInput = (0, _react.useCallback)(function (flagEvent) {
    var handleCreateOrFail = function handleCreateOrFail(addFlagConfig) {
      closeDrawer();
      addFlag && addFlag(addFlagConfig);
      onCreateKudosSuccess && addFlagConfig.type === 'success' && onCreateKudosSuccess(flagEvent);
    };
    switch (flagEvent.eventType) {
      case _types.FlagEventType.KUDOS_CREATED:
        handleCreateOrFail({
          title: /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, _messages.default.kudosCreatedFlag),
          id: "kudosCreatedFlag-".concat(flagEvent.kudosUuid),
          description: /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, (0, _extends2.default)({}, _messages.default.kudosCreatedDescriptionFlag, {
            values: {
              a: function a(s) {
                return /*#__PURE__*/React.createElement(_link.default, {
                  href: "".concat(teamCentralBaseUrl, "/people/kudos/").concat(flagEvent.kudosUuid)
                }, s);
              }
            }
          })),
          actions: isActionsEnabled ? [{
            content: /*#__PURE__*/React.createElement(_primitives.Inline, {
              space: "space.050",
              alignBlock: "center"
            }, /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, _messages.default.kudosCreatedActionFlag), /*#__PURE__*/React.createElement(_linkExternal.default, {
              label: "",
              color: "currentColor"
            })),
            href: "".concat(teamCentralBaseUrl, "/people/kudos/").concat(flagEvent.kudosUuid),
            target: '_blank',
            onClick: function onClick() {
              return undefined;
            }
          }] : undefined,
          type: 'success'
        });
        break;
      case _types.FlagEventType.KUDOS_FAILED:
        handleCreateOrFail({
          title: /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, _messages.default.kudosCreationFailedFlag),
          id: "jiraKudosCreationFailedFlag-".concat(flagEvent.kudosUuid),
          description: /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, _messages.default.kudosCreationFailedDescriptionFlag),
          type: 'error'
        });
        break;
      case _types.FlagEventType.JIRA_KUDOS_CREATED:
        handleCreateOrFail({
          title: /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, _messages.default.JiraKudosCreatedFlag),
          id: "kudosCreatedFlag-".concat(flagEvent.kudosUuid),
          description: /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, _messages.default.JiraKudosCreatedDescriptionFlag),
          type: 'success',
          actions: [{
            content: 'Track gift request',
            href: flagEvent.jiraKudosUrl
          }, {
            content: 'View kudos',
            href: "".concat(teamCentralBaseUrl, "/people/kudos/").concat(flagEvent.kudosUuid)
          }]
        });
        break;
      case _types.FlagEventType.JIRA_KUDOS_FAILED:
        handleCreateOrFail({
          title: /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, _messages.default.JiraKudosCreationFailedFlag),
          id: "jiraKudosCreationFailedFlag-".concat(flagEvent.kudosUuid),
          description: /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, (0, _extends2.default)({}, _messages.default.JiraKudosCreationFailedDescriptionFlag, {
            values: {
              a: function a(s) {
                var _flagEvent$jiraKudosF;
                return /*#__PURE__*/React.createElement(_link.default, {
                  href: (_flagEvent$jiraKudosF = flagEvent.jiraKudosFormUrl) !== null && _flagEvent$jiraKudosF !== void 0 ? _flagEvent$jiraKudosF : ''
                }, s);
              }
            }
          })),
          type: 'error',
          actions: [{
            content: 'Visit go/kudos',
            href: flagEvent.jiraKudosFormUrl
          }, {
            content: 'View kudos',
            href: "".concat(teamCentralBaseUrl, "/people/kudos/").concat(flagEvent.kudosUuid)
          }]
        });
        break;
      case _types.FlagEventType.DIRTY:
        setIsDirty(true);
        break;
      case _types.FlagEventType.CLOSE:
        closeDrawer();
        break;
      default:
        // Not a known FlagEventType
        return;
    }
  }, [addFlag, closeDrawer, teamCentralBaseUrl, onCreateKudosSuccess, isActionsEnabled]);
  var messageListener = (0, _react.useCallback)(function (event) {
    if (!props.isOpen) {
      return;
    }
    if (event.data === 'dirty') {
      setIsDirty(true);
    } else if (event.data === 'close') {
      closeDrawer();
    } else {
      try {
        var eventData = JSON.parse(event.data);
        if (eventData.eventType && (0, _types.isFlagEventTypeValue)(eventData.eventType)) {
          createFlagWithJsonStringifiedInput(eventData);
        }
      } catch (e) {
        // Swallow any errors
      }
    }
  }, [props.isOpen, closeDrawer, createFlagWithJsonStringifiedInput]);
  (0, _react.useEffect)(function () {
    window.removeEventListener('message', messageListenerEventHandler.current);
    messageListenerEventHandler.current = messageListener;
    window.addEventListener('message', messageListenerEventHandler.current, false);
    return function () {
      window.removeEventListener('message', messageListenerEventHandler.current);
    };
  }, [messageListener]);
  (0, _react.useEffect)(function () {
    window.removeEventListener('beforeunload', unloadEventHandler.current);
    if (isDirty) {
      unloadEventHandler.current = shouldBlockTransition;
      window.addEventListener('beforeunload', unloadEventHandler.current, false);
    }
    return function () {
      window.removeEventListener('beforeunload', unloadEventHandler.current);
    };
  }, [isDirty, shouldBlockTransition]);
  var sendCancelAnalytic = function sendCancelAnalytic() {
    sendAnalytic('cancelled', {});
  };
  var handleCloseDrawerClickedFunc = function handleCloseDrawerClickedFunc() {
    if (!isDirty) {
      sendCancelAnalytic();
      closeDrawer();
      return;
    }
    setIsCloseConfirmModalOpen(true);
  };
  var handleCloseDrawerClickedFuncRef = (0, _react.useRef)(handleCloseDrawerClickedFunc);
  (0, _react.useEffect)(function () {
    handleCloseDrawerClickedFuncRef.current = handleCloseDrawerClickedFunc;
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isDirty]);
  var handleCloseDrawerClicked = function handleCloseDrawerClicked() {
    handleCloseDrawerClickedFuncRef.current();
  };
  var recipientParam = props.recipient ? "&type=".concat(props.recipient.type, "&recipientId=").concat(props.recipient.recipientId) : '';
  var giveKudosUrl = "".concat(props.teamCentralBaseUrl, "/give-kudos?cloudId=").concat(props.cloudId).concat(recipientParam, "&unsavedMessage=").concat(intl.formatMessage(_messages.default.unsavedKudosWarning));
  var renderDrawer = (0, _react.useMemo)(function () {
    if (props.isOpen) {
      sendAnalytic('opened', {});
    }
    return /*#__PURE__*/React.createElement(_compiled.Drawer, {
      width: "full",
      isOpen: props.isOpen,
      zIndex: _constants.layers.modal(),
      onClose: handleCloseDrawerClicked
    }, /*#__PURE__*/React.createElement("div", {
      className: (0, _runtime.ax)([styles.drawerCloseButtonContainer])
    }, /*#__PURE__*/React.createElement(_new.IconButton, {
      onClick: handleCloseDrawerClicked,
      icon: _arrowLeft.default,
      label: intl.formatMessage(_messages.default.closeDrawerButtonLabel),
      shape: "circle",
      appearance: "subtle"
    })), /*#__PURE__*/React.createElement("iframe", {
      src: giveKudosUrl,
      ref: iframeEl,
      width: "100%",
      height: "100%",
      frameBorder: "0",
      allow: "camera;microphone",
      className: (0, _runtime.ax)([styles.iframe])
    }));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [(_props$recipient4 = props.recipient) === null || _props$recipient4 === void 0 ? void 0 : _props$recipient4.recipientId, props.isOpen]);
  return /*#__PURE__*/React.createElement(_portal.default, {
    zIndex: _constants.layers.modal()
  }, /*#__PURE__*/React.createElement("div", {
    "data-testid": testId
  }, /*#__PURE__*/React.createElement(_modalDialog.ModalTransition, null, isCloseConfirmModalOpen && /*#__PURE__*/React.createElement(_modalDialog.default, {
    onClose: closeWarningModal,
    width: "small"
  }, /*#__PURE__*/React.createElement(_modalDialog.ModalHeader, null, /*#__PURE__*/React.createElement(_modalDialog.ModalTitle, null, /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, _messages.default.confirmCloseTitle))), /*#__PURE__*/React.createElement(_modalDialog.ModalBody, null, /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, _messages.default.unsavedKudosWarning)), /*#__PURE__*/React.createElement(_modalDialog.ModalFooter, null, /*#__PURE__*/React.createElement(_standardButton.default, {
    appearance: "subtle",
    onClick: closeWarningModal
  }, /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, _messages.default.unsavedKudosWarningCancelButton)), /*#__PURE__*/React.createElement(_standardButton.default, {
    appearance: "primary",
    onClick: function onClick() {
      sendCancelAnalytic();
      closeDrawer();
    }
  }, /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, _messages.default.unsavedKudosWarningCloseButton))))), renderDrawer));
};
var ComposedGiveKudosLauncher = function ComposedGiveKudosLauncher(props) {
  return /*#__PURE__*/React.createElement(_intlMessagesProvider.IntlMessagesProvider, {
    loaderFn: _fetchMessagesForLocale.fetchMessagesForLocale,
    defaultMessages: _en.default
  }, /*#__PURE__*/React.createElement(GiveKudosLauncher, props));
};
var _default = exports.default = ComposedGiveKudosLauncher;
