// @atlaskit/teams-public/dist/cjs/ui/team-containers/team-containers-skeleton/linked-container-card-skeleton/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LinkedContainerCardSkeleton = void 0;
require("./index.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _react = _interopRequireDefault(require("react"));
var _compiled = require("@atlaskit/primitives/compiled");
var styles = {
  container: "_2rkoiti9 _1h6dz9xs _12y3e4h9 _12jimuej _1qu2nqa1 _syaz1fxt _ca0qpxbi _u5f3pxbi _n3tdpxbi _19bvpxbi",
  avatarSkeleton: "_2rkoglpi _ca0qpxbi _u5f3pxbi _n3tdpxbi _19bvpxbi",
  firstLineSkeleton: "_2rko1q77 _1rjc1b66",
  secondLineSkeleton: "_2rko1q77 _1rjc1b66 _1bsbqkoa"
};
var LinkedContainerCardSkeleton = exports.LinkedContainerCardSkeleton = function LinkedContainerCardSkeleton() {
  return /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    xcss: styles.container
  }, /*#__PURE__*/_react.default.createElement(_compiled.Inline, {
    space: "space.100",
    alignBlock: "center"
  }, /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    backgroundColor: "color.background.neutral",
    xcss: styles.avatarSkeleton
  }), /*#__PURE__*/_react.default.createElement(_compiled.Stack, {
    space: "space.100",
    grow: "fill"
  }, /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    backgroundColor: "color.background.neutral",
    xcss: styles.firstLineSkeleton
  }), /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    backgroundColor: "color.background.neutral",
    xcss: styles.secondLineSkeleton
  }))));
};
