// @atlaskit/smart-card/dist/cjs/view/HoverCard/utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getMetadata = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _linkExtractors = require("@atlaskit/link-extractors");
var _constants = require("../../constants");
var _utils = require("../../extractors/flexible/utils");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var getSimulatedBetterMetadata = function getSimulatedBetterMetadata(extensionKey, data) {
  var _data$Type$includes, _data$Type;
  var types = data ? (0, _linkExtractors.extractType)(data) : undefined;
  var defaultMetadata = {
    primary: [_constants.ElementName.AuthorGroup, _constants.ElementName.ModifiedOn]
  };
  switch (extensionKey) {
    case 'bitbucket-object-provider':
    case 'native-bitbucket-object-provider':
      if (types !== null && types !== void 0 && types.includes('atlassian:SourceCodePullRequest')) {
        return {
          primary: [_constants.ElementName.AuthorGroup, _constants.ElementName.ModifiedOn, _constants.ElementName.SubscriberCount, _constants.ElementName.State],
          subtitle: [_constants.ElementName.SourceBranch, _constants.ElementName.TargetBranch]
        };
      }
      if (types !== null && types !== void 0 && types.includes('schema:DigitalDocument')) {
        return {
          primary: [_constants.ElementName.LatestCommit, _constants.ElementName.CollaboratorGroup, _constants.ElementName.ModifiedOn]
        };
      }
      return {
        primary: [_constants.ElementName.AuthorGroup, _constants.ElementName.ModifiedOn]
      };
    case 'confluence-object-provider':
      var primaryAttribution = data && (0, _utils.extractOwnedBy)(data) ? _constants.ElementName.OwnedByGroup : _constants.ElementName.AuthorGroup;
      return {
        primary: [primaryAttribution, _constants.ElementName.ModifiedOn],
        tertiary: [_constants.ElementName.ReactCount, _constants.ElementName.CommentCount, _constants.ElementName.ViewCount]
      };
    case 'jira-object-provider':
      var isJiraTask = (_data$Type$includes = data === null || data === void 0 || (_data$Type = data['@type']) === null || _data$Type === void 0 ? void 0 : _data$Type.includes('atlassian:Task')) !== null && _data$Type$includes !== void 0 ? _data$Type$includes : false;
      return _objectSpread(_objectSpread({}, defaultMetadata), isJiraTask && {
        primary: [_constants.ElementName.AssignedToGroup, _constants.ElementName.State, _constants.ElementName.StoryPoints, _constants.ElementName.Priority]
      });
    case 'trello-object-provider':
      return {
        primary: [_constants.ElementName.CollaboratorGroup, _constants.ElementName.State, _constants.ElementName.DueOn],
        secondary: [_constants.ElementName.ReactCount, _constants.ElementName.CommentCount, _constants.ElementName.AttachmentCount, _constants.ElementName.ChecklistProgress],
        subtitle: [_constants.ElementName.Location]
      };
    case 'watermelon-object-provider':
      if (types !== null && types !== void 0 && types.includes('atlassian:Project')) {
        return {
          primary: [_constants.ElementName.AuthorGroup, _constants.ElementName.ModifiedOn, _constants.ElementName.State, _constants.ElementName.DueOn]
        };
      }
      return {
        primary: [_constants.ElementName.AuthorGroup, _constants.ElementName.State, _constants.ElementName.DueOn]
      };
    case 'slack-object-provider':
      return {
        primary: [_constants.ElementName.AuthorGroup, _constants.ElementName.SentOn],
        tertiary: [_constants.ElementName.ReactCount, _constants.ElementName.CommentCount]
      };
    case 'google-object-provider':
    case 'figma-object-provider':
      return {
        primary: [_constants.ElementName.AuthorGroup, _constants.ElementName.ModifiedOn]
      };
    default:
      return defaultMetadata;
  }
};
var AvatarGroupsWithNamePrefix = [_constants.ElementName.AssignedToGroup, _constants.ElementName.OwnedByGroup, _constants.ElementName.AuthorGroup];
var toElementItem = function toElementItem(name) {
  var showNamePrefix = AvatarGroupsWithNamePrefix.indexOf(name) !== -1 ? true : undefined;
  var testId = "".concat(name.toLowerCase(), "-metadata-element");
  return {
    name: name,
    showNamePrefix: showNamePrefix,
    testId: testId
  };
};
var toElementItems = function toElementItems(elementNames) {
  if (!(elementNames !== null && elementNames !== void 0 && elementNames.length)) {
    return;
  }
  return elementNames.filter(function (name) {
    return name in _constants.ElementName;
  }).map(toElementItem);
};
var getMetadata = exports.getMetadata = function getMetadata(extensionKey, data) {
  var metadata = getSimulatedBetterMetadata(extensionKey, data);
  var primary = [].concat(metadata.primary, metadata.tertiary);
  return {
    subtitle: toElementItems(metadata.subtitle),
    primary: toElementItems(primary),
    secondary: toElementItems(metadata.secondary)
  };
};
