// @atlaskit/react-select/dist/es2019/emotion/components/group.js
import _extends from "@babel/runtime/helpers/extends";
/**
 * @jsxRuntime classic
 * @jsx jsx
 */

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled
import { jsx } from '@emotion/react';
import { cleanCommonProps, getStyleProps } from '../../utils';
export const groupCSS = ({}) => ({
  paddingBottom: "var(--ds-space-100, 8px)",
  paddingTop: "var(--ds-space-100, 8px)"
});
const Group = props => {
  const {
    children,
    cx,
    getStyles,
    getClassNames,
    Heading,
    headingProps,
    innerProps,
    label,
    selectProps
  } = props;
  return jsx("div", _extends({}, getStyleProps(props, 'group', {
    group: true
  }), innerProps), label && jsx(Heading, _extends({}, headingProps, {
    selectProps: selectProps,
    getStyles: getStyles,
    getClassNames: getClassNames,
    cx: cx
  }), label), jsx("div", null, children));
};
export const groupHeadingCSS = ({}) => ({
  label: 'group',
  cursor: 'default',
  display: 'block',
  marginBottom: '0.25em',
  paddingLeft: "var(--ds-space-150, 12px)",
  paddingRight: "var(--ds-space-150, 12px)",
  font: "var(--ds-font-body-small, normal 400 11px/16px ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  color: "var(--ds-text-subtle, #44546F)",
  fontWeight: "var(--ds-font-weight-bold, 700)",
  textTransform: 'none'
});

// eslint-disable-next-line @repo/internal/react/require-jsdoc
export const GroupHeading = props => {
  const {
    data,
    ...innerProps
  } = cleanCommonProps(props);
  return jsx("div", _extends({}, getStyleProps(props, 'groupHeading', {
    'group-heading': true
  }), innerProps));
};

// eslint-disable-next-line @repo/internal/react/require-jsdoc
export default Group;
