// @atlaskit/smart-card/dist/es2019/view/FlexibleCard/components/elements/link/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./index.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
/* eslint-disable @compiled/shorthand-property-sorting */
/* eslint-disable @atlaskit/design-system/use-tokens-typography */

import React, { useMemo } from 'react';
import Tooltip from '@atlaskit/tooltip';
import { SmartLinkSize, SmartLinkTheme } from '../../../../../constants';
import { useMouseDownEvent } from '../../../../../state/analytics/useLinkClicked';
import { hasWhiteSpace } from '../../utils';
const DEFAULT_MAX_LINES = 2;
const MAXIMUM_MAX_LINES = 2;
const MINIMUM_MAX_LINES = 1;
const containerStyles = null;
const getMaxLines = maxLines => {
  if (maxLines > MAXIMUM_MAX_LINES) {
    return DEFAULT_MAX_LINES;
  }
  if (maxLines < MINIMUM_MAX_LINES) {
    return MINIMUM_MAX_LINES;
  }
  return maxLines;
};
const withTooltip = (trigger, content, testId) => /*#__PURE__*/React.createElement(Tooltip, {
  content: content,
  testId: `${testId}-tooltip`,
  tag: "span"
}, trigger);
const linkStyleSizeMap = {
  xlarge: "_11c8lodh _k48pi7a9 _vwz41k8s",
  large: "_11c82smr _k48pi7a9 _vwz41j6v",
  medium: "_11c82smr _k48pi7a9 _vwz41j6v",
  small: "_11c8dcr7 _k48pi7a9 _vwz41j6v"
};
const baseAnchorStyle = null;
const workBreakStyleMap = {
  true: "_1nmz1hna",
  false: "_1nmz9jpi"
};
const anchorConstantsStyles = null;
const anchorLineClampMap = {
  1: "_1yyjkb7n",
  2: "_1yyjcs5v"
};
const anchorLinkLineHeight1Map = {
  xlarge: "_102k1k8s",
  large: "_102k1j6v",
  medium: "_102k1j6v",
  small: "_102k1j6v"
};
const anchorLinkLineHeight2Map = {
  xlarge: "_102k1wto",
  large: "_102kviql",
  medium: "_102kviql",
  small: "_102kviql"
};
const themeStyleMap = {
  grey: "_fjv61r31 _93u91r31 _g2lk1r31 _pnzf1r31 _hlwq8stv _1e2m8stv _1l4b8stv _ue8d8stv _yni7nqa1 _1bianqa1 _93jhnqa1 _64n6nqa1 _1ray131l _dlb7131l _rnxt131l _1h4d131l _10uedcr7 _ye10131l",
  black: "_syaz1gjq _9oik1r31 _1bnx8stv _jf4cnqa1 _30l31gjq _9h8h1fxt",
  link: "_syaz14q2 _9oik1r31 _1bnx8stv _jf4cnqa1 _30l314q2 _9h8h16c2"
};

/**
 * A base element that represent an anchor.
 * @internal
 * @param {LinkProps} LinkProps - The props necessary for the Link element.
 * @see LinkIcon
 */
const Link = ({
  hideTooltip,
  maxLines = DEFAULT_MAX_LINES,
  name,
  className,
  size = SmartLinkSize.Medium,
  testId = 'smart-element-link',
  text,
  theme = SmartLinkTheme.Link,
  url,
  onClick,
  target = '_blank'
}) => {
  const onMouseDown = useMouseDownEvent();
  const hasSpace = useMemo(() => text ? hasWhiteSpace(text) : false, [text]);
  const anchor = /*#__PURE__*/React.createElement("a", _extends({
    "data-smart-element": name,
    "data-smart-element-link": true,
    "data-testid": testId,
    onClick: onClick,
    onMouseDown: onMouseDown,
    href: url
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop
    ,
    className: ax(["_16jlkb7n _1o9zkb7n _i0dl1wug", linkStyleSizeMap[size], anchorLineClampMap[getMaxLines(maxLines)], "_1reo15vq _18m915vq _1e0ccj1k _1bto1l2s _sudp1e54", workBreakStyleMap[hasSpace ? 'true' : 'false'], getMaxLines(maxLines) === 1 && anchorLinkLineHeight1Map[size], getMaxLines(maxLines) === 2 && anchorLinkLineHeight2Map[size], themeStyleMap[theme], className])
    // We do not want set the target if it is the default value of '_self'. This prevents link
    // click issues in Confluence and Trello which rely on it not being set unless necessary.
  }, target !== '_self' && {
    target
  }), text);
  return /*#__PURE__*/React.createElement("span", {
    className: ax(["_16jlkb7n _1o9zkb7n _i0dl1wug"])
  }, hideTooltip || text === undefined ? anchor : withTooltip(anchor, text, testId));
};
export default Link;
