// @atlaskit/editor-plugin-table/dist/cjs/ui/FloatingDeleteButton/DeleteButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactIntlNext = require("react-intl-next");
var _types = require("../../types");
var DeleteButton = function DeleteButton(_ref) {
  var style = _ref.style,
    onClick = _ref.onClick,
    onMouseEnter = _ref.onMouseEnter,
    onMouseLeave = _ref.onMouseLeave,
    removeLabel = _ref.removeLabel,
    formatMessage = _ref.intl.formatMessage;
  return (
    /*#__PURE__*/
    // eslint-disable-next-line jsx-a11y/no-static-element-interactions
    _react.default.createElement("div", {
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
      className: _types.TableCssClassName.CONTROLS_DELETE_BUTTON_WRAP
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
      ,
      style: style,
      onMouseEnter: onMouseEnter,
      onMouseLeave: onMouseLeave
    }, /*#__PURE__*/_react.default.createElement("button", {
      type: "button",
      "aria-label": formatMessage(removeLabel, {
        0: 1
      })
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
      ,
      className: _types.TableCssClassName.CONTROLS_DELETE_BUTTON,
      onMouseDown: onClick,
      onMouseMove: function onMouseMove(e) {
        return e.preventDefault();
      }
    }, /*#__PURE__*/_react.default.createElement("svg", {
      className: _types.TableCssClassName.CONTROLS_BUTTON_ICON
    }, /*#__PURE__*/_react.default.createElement("path", {
      d: "M12.242 10.828L9.414 8l2.828-2.829a.999.999 0 1 0-1.414-1.414L8 6.587l-2.829-2.83a1 1 0 0 0-1.414 1.414l2.83 2.83-2.83 2.827a1 1 0 0 0 1.414 1.414l2.83-2.828 2.827 2.828a.999.999 0 1 0 1.414-1.414",
      fill: "currentColor",
      fillRule: "evenodd"
    }))))
  );
};
var _default = exports.default = (0, _reactIntlNext.injectIntl)(DeleteButton);
