// @atlaskit/editor-common/dist/cjs/extensions.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DefaultExtensionProvider", {
  enumerable: true,
  get: function get() {
    return _defaultExtensionProvider.default;
  }
});
Object.defineProperty(exports, "buildMenuItem", {
  enumerable: true,
  get: function get() {
    return _moduleHelpers.buildMenuItem;
  }
});
Object.defineProperty(exports, "combineExtensionProviders", {
  enumerable: true,
  get: function get() {
    return _combineExtensionProviders.default;
  }
});
Object.defineProperty(exports, "configPanelMessages", {
  enumerable: true,
  get: function get() {
    return _messages.configPanelMessages;
  }
});
Object.defineProperty(exports, "createAutoConverterRunner", {
  enumerable: true,
  get: function get() {
    return _moduleHelpers.createAutoConverterRunner;
  }
});
Object.defineProperty(exports, "getContextualToolbarItemsFromModule", {
  enumerable: true,
  get: function get() {
    return _moduleHelpers.getContextualToolbarItemsFromModule;
  }
});
Object.defineProperty(exports, "getCustomFieldResolver", {
  enumerable: true,
  get: function get() {
    return _extensionFieldsHelpers.getCustomFieldResolver;
  }
});
Object.defineProperty(exports, "getExtensionAutoConvertersFromProvider", {
  enumerable: true,
  get: function get() {
    return _moduleHelpers.getExtensionAutoConvertersFromProvider;
  }
});
Object.defineProperty(exports, "getExtensionKeyAndNodeKey", {
  enumerable: true,
  get: function get() {
    return _manifestHelpers.getExtensionKeyAndNodeKey;
  }
});
Object.defineProperty(exports, "getExtensionModuleNode", {
  enumerable: true,
  get: function get() {
    return _extensionHandlers.getExtensionModuleNode;
  }
});
Object.defineProperty(exports, "getExtensionModuleNodePrivateProps", {
  enumerable: true,
  get: function get() {
    return _extensionHandlers.getExtensionModuleNodePrivateProps;
  }
});
Object.defineProperty(exports, "getFieldDeserializer", {
  enumerable: true,
  get: function get() {
    return _extensionFieldsHelpers.getFieldDeserializer;
  }
});
Object.defineProperty(exports, "getFieldSerializer", {
  enumerable: true,
  get: function get() {
    return _extensionFieldsHelpers.getFieldSerializer;
  }
});
Object.defineProperty(exports, "getNodeRenderer", {
  enumerable: true,
  get: function get() {
    return _extensionHandlers.getNodeRenderer;
  }
});
Object.defineProperty(exports, "getQuickInsertItemsFromModule", {
  enumerable: true,
  get: function get() {
    return _moduleHelpers.getQuickInsertItemsFromModule;
  }
});
Object.defineProperty(exports, "getUserFieldContextProvider", {
  enumerable: true,
  get: function get() {
    return _extensionFieldsHelpers.getUserFieldContextProvider;
  }
});
Object.defineProperty(exports, "isDateRange", {
  enumerable: true,
  get: function get() {
    return _fieldDefinitions.isDateRange;
  }
});
Object.defineProperty(exports, "isExpand", {
  enumerable: true,
  get: function get() {
    return _fieldDefinitions.isExpand;
  }
});
Object.defineProperty(exports, "isFieldset", {
  enumerable: true,
  get: function get() {
    return _fieldDefinitions.isFieldset;
  }
});
Object.defineProperty(exports, "isTabField", {
  enumerable: true,
  get: function get() {
    return _fieldDefinitions.isTabField;
  }
});
Object.defineProperty(exports, "isTabGroup", {
  enumerable: true,
  get: function get() {
    return _fieldDefinitions.isTabGroup;
  }
});
Object.defineProperty(exports, "messages", {
  enumerable: true,
  get: function get() {
    return _messages.messages;
  }
});
Object.defineProperty(exports, "resolveImport", {
  enumerable: true,
  get: function get() {
    return _manifestHelpers.resolveImport;
  }
});
var _fieldDefinitions = require("./extensions/types/field-definitions");
var _moduleHelpers = require("./extensions/module-helpers");
var _defaultExtensionProvider = _interopRequireDefault(require("./extensions/default-extension-provider"));
var _combineExtensionProviders = _interopRequireDefault(require("./extensions/combine-extension-providers"));
var _manifestHelpers = require("./extensions/manifest-helpers");
var _extensionHandlers = require("./extensions/extension-handlers");
var _extensionFieldsHelpers = require("./extensions/extension-fields-helpers");
var _messages = require("./extensions/messages");
