// @atlaskit/editor-common/dist/cjs/panel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "PanelSharedCssClassName", {
  enumerable: true,
  get: function get() {
    return _panel.PanelSharedCssClassName;
  }
});
Object.defineProperty(exports, "PanelSharedSelectors", {
  enumerable: true,
  get: function get() {
    return _panel.PanelSharedSelectors;
  }
});
Object.defineProperty(exports, "darkPanelColors", {
  enumerable: true,
  get: function get() {
    return _panel.darkPanelColors;
  }
});
Object.defineProperty(exports, "getPanelTypeBackground", {
  enumerable: true,
  get: function get() {
    return _panel.getPanelTypeBackground;
  }
});
Object.defineProperty(exports, "getPanelTypeBackgroundNoTokens", {
  enumerable: true,
  get: function get() {
    return _panel.getPanelTypeBackgroundNoTokens;
  }
});
Object.defineProperty(exports, "panelSharedStyles", {
  enumerable: true,
  get: function get() {
    return _panel.panelSharedStyles;
  }
});
Object.defineProperty(exports, "panelSharedStylesWithoutPrefix", {
  enumerable: true,
  get: function get() {
    return _panel.panelSharedStylesWithoutPrefix;
  }
});
var _panel = require("./styles/shared/panel");
