// @atlaskit/editor-common/dist/cjs/validator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ADFStages", {
  enumerable: true,
  get: function get() {
    return _validator.ADFStages;
  }
});
Object.defineProperty(exports, "getMarksByOrder", {
  enumerable: true,
  get: function get() {
    return _validator.getMarksByOrder;
  }
});
Object.defineProperty(exports, "getValidContent", {
  enumerable: true,
  get: function get() {
    return _validator.getValidContent;
  }
});
Object.defineProperty(exports, "getValidDocument", {
  enumerable: true,
  get: function get() {
    return _validator.getValidDocument;
  }
});
Object.defineProperty(exports, "getValidMark", {
  enumerable: true,
  get: function get() {
    return _validator.getValidMark;
  }
});
Object.defineProperty(exports, "getValidNode", {
  enumerable: true,
  get: function get() {
    return _validator.getValidNode;
  }
});
Object.defineProperty(exports, "getValidUnknownNode", {
  enumerable: true,
  get: function get() {
    return _validator.getValidUnknownNode;
  }
});
Object.defineProperty(exports, "isSameMark", {
  enumerable: true,
  get: function get() {
    return _validator.isSameMark;
  }
});
Object.defineProperty(exports, "isSubSupType", {
  enumerable: true,
  get: function get() {
    return _validator.isSubSupType;
  }
});
Object.defineProperty(exports, "markOrder", {
  enumerable: true,
  get: function get() {
    return _validator.markOrder;
  }
});
var _validator = require("./utils/validator");
