// @atlaskit/editor-plugin-placeholder-text/dist/es2019/ui/PlaceholderFloatingToolbar/index.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import React from 'react';
import { injectIntl } from 'react-intl-next';
import { placeholderTextMessages as messages } from '@atlaskit/editor-common/messages';
import { PanelTextInput } from '@atlaskit/editor-common/ui';
import FloatingToolbar from '../FloatingToolbar';
import { getNearestNonTextNode, getOffsetParent, handlePositionCalculatedWith } from '../FloatingToolbar/utils';
// eslint-disable-next-line @repo/internal/react/no-class-components
class PlaceholderFloatingToolbar extends React.Component {
  constructor(...args) {
    super(...args);
    _defineProperty(this, "handleSubmit", value => {
      if (value) {
        this.props.insertPlaceholder(value);
        this.props.setFocusInEditor();
      } else {
        this.props.hidePlaceholderFloatingToolbar();
      }
    });
    _defineProperty(this, "handleBlur", () => {
      this.props.hidePlaceholderFloatingToolbar();
    });
  }
  render() {
    const {
      getNodeFromPos,
      showInsertPanelAt,
      editorViewDOM,
      popupsMountPoint,
      getFixedCoordinatesFromPos,
      popupsBoundariesElement,
      intl: {
        formatMessage
      }
    } = this.props;
    const target = getNodeFromPos(showInsertPanelAt);
    const offsetParent = getOffsetParent(editorViewDOM, popupsMountPoint);
    const getFixedCoordinates = () => getFixedCoordinatesFromPos(showInsertPanelAt);
    const handlePositionCalculated = handlePositionCalculatedWith(offsetParent, target, getFixedCoordinates);
    return /*#__PURE__*/React.createElement(FloatingToolbar
    // Ignored via go/ees005
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    , {
      target: getNearestNonTextNode(target),
      onPositionCalculated: handlePositionCalculated,
      popupsMountPoint: popupsMountPoint,
      popupsBoundariesElement: popupsBoundariesElement,
      fitHeight: 32,
      offset: [0, 12]
    }, /*#__PURE__*/React.createElement(PanelTextInput, {
      placeholder: formatMessage(messages.placeholderTextPlaceholder),
      onSubmit: this.handleSubmit,
      onBlur: this.handleBlur,
      autoFocus: true,
      width: 300
    }));
  }
}
export default injectIntl(PlaceholderFloatingToolbar);
