// @atlaskit/editor-plugin-table/dist/cjs/ui/ColumnResizeWidget/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ColumnResizeWidget = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = require("@emotion/react");
var _reactIntlNext = require("react-intl-next");
var _keymaps = require("@atlaskit/editor-common/keymaps");
var _messages = require("@atlaskit/editor-common/messages");
var _tooltip = _interopRequireDefault(require("@atlaskit/tooltip"));
var _types = require("../../types");
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var ColumnResizeWidget = exports.ColumnResizeWidget = function ColumnResizeWidget(_ref) {
  var startIndex = _ref.startIndex,
    endIndex = _ref.endIndex,
    includeTooltip = _ref.includeTooltip;
  var _useIntl = (0, _reactIntlNext.useIntl)(),
    formatMessage = _useIntl.formatMessage;
  if (!includeTooltip) {
    return (0, _react.jsx)("div", {
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
      className: _types.TableCssClassName.RESIZE_HANDLE_DECORATION,
      "data-start-index": startIndex,
      "data-end-index": endIndex
    });
  }
  return (0, _react.jsx)(_tooltip.default, {
    content: (0, _react.jsx)(_keymaps.ToolTipContent, {
      description: formatMessage(_messages.tableMessages.adjustColumns),
      keymap: _keymaps.startColumnResizing
    }),
    hideTooltipOnClick: true,
    hideTooltipOnMouseDown: true,
    position: "mouse",
    mousePosition: "auto-start"
  }, function (tooltipProps) {
    return (0, _react.jsx)("div", (0, _extends2.default)({
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
      className: _types.TableCssClassName.RESIZE_HANDLE_DECORATION,
      "data-start-index": startIndex,
      "data-end-index": endIndex
      // Ignored via go/ees005
      // eslint-disable-next-line react/jsx-props-no-spreading
    }, tooltipProps));
  });
};
