// @atlaskit/mention/dist/es2019/components/Mention/styles.js
import { colors, themed } from '@atlaskit/theme';
import styled from 'styled-components';
import { MentionType } from '../../types';
const mentionStyle = {
    [MentionType.SELF]: {
        background: themed({ light: colors.B400, dark: colors.B200 }),
        border: 'transparent',
        text: themed({ light: colors.N20, dark: colors.DN30 }),
    },
    [MentionType.RESTRICTED]: {
        background: 'transparent',
        border: themed({ light: colors.N500, dark: colors.DN80 }),
        text: themed({ light: colors.N500, dark: colors.DN100 }),
    },
    [MentionType.DEFAULT]: {
        background: themed({ light: colors.N30A, dark: colors.DN80 }),
        border: 'transparent',
        text: themed({ light: colors.N500, dark: colors.DN800 }),
    },
};
const getStyle = (props, property) => {
    const obj = mentionStyle[props.mentionType][property];
    // themed() returns a function
    return typeof obj === 'string' ? obj : obj(props);
};
export const MentionStyle = styled.span `
  ${(props) => `
  display: inline;
  background: ${getStyle(props, 'background')};
  border: 1px solid ${getStyle(props, 'border')};
  border-radius: 20px;
  color: ${getStyle(props, 'text')};
  cursor: pointer;
  padding: 0 0.3em 2px 0.23em;
  line-height: 1.714;
  font-size: 1em;
  font-weight: normal;
  word-break: break-word;
`};
`;
