// @atlaskit/inline-edit/dist/cjs/internal/buttons.js
/* buttons.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./buttons.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _new = require("@atlaskit/button/new");
var _cross = _interopRequireDefault(require("@atlaskit/icon/glyph/cross"));
var _cross2 = _interopRequireDefault(require("@atlaskit/icon/utility/cross"));
var _checkMarkCheck = _interopRequireDefault(require("@atlaskit/icon/utility/migration/check-mark--check"));
var _colors = require("@atlaskit/theme/colors");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/* eslint-disable @atlaskit/design-system/no-nested-styles */
/* eslint-disable @atlaskit/ui-styling-standard/no-nested-selectors */
var buttonsContainerStyles = null;
var buttonWrapperBaseStyles = null;
var Buttons = function Buttons(_ref) {
  var confirmButtonLabel = _ref.confirmButtonLabel,
    cancelButtonLabel = _ref.cancelButtonLabel,
    onMouseDown = _ref.onMouseDown,
    onCancelClick = _ref.onCancelClick,
    testId = _ref.testId;
  return /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)(["_1e0c1txw _kqswstnw _1o9zidpf _152t1osq _rjxpidpf _1pfh12x7"])
  }, /*#__PURE__*/React.createElement("div", {
    tabIndex: -1,
    className: (0, _runtime.ax)(["_2rko1sit _11c82smr _vchhusvi _1bsbxy5q _1pby1kze _bfhk13b4 _16qs1a23 _1mpy1b66 _1nlx13nm _l2uv11wk _7mfvdecw _szlb17km"]),
    style: {
      "--_1f6rtr1": (0, _runtime.ix)("var(--ds-surface-overlay, ".concat(_colors.N0, ")")),
      "--_3t1odq": (0, _runtime.ix)("var(--ds-shadow-overlay, ".concat("0 4px 8px -2px ".concat(_colors.N50A, ", 0 0 1px ").concat(_colors.N60A), ")")),
      "--_12yx4z3": (0, _runtime.ix)("var(--ds-surface-overlay, ".concat(_colors.N20A, ")")),
      "--_qfu7kd": (0, _runtime.ix)("var(--ds-surface-overlay-hovered, ".concat(_colors.N30A, ")")),
      "--_1lrmnoy": (0, _runtime.ix)("var(--ds-text, ".concat(_colors.B400, ")"))
    }
  }, /*#__PURE__*/React.createElement(_new.IconButton, {
    type: "submit",
    icon: function icon(iconProps) {
      return /*#__PURE__*/React.createElement(_checkMarkCheck.default, (0, _extends2.default)({}, iconProps, {
        LEGACY_size: "small"
      }));
    },
    onMouseDown: onMouseDown,
    label: confirmButtonLabel,
    testId: testId && "".concat(testId, "--confirm")
  })), /*#__PURE__*/React.createElement("div", {
    tabIndex: -1,
    className: (0, _runtime.ax)(["_2rko1sit _11c82smr _vchhusvi _1bsbxy5q _1pby1kze _bfhk13b4 _16qs1a23 _1mpy1b66 _1nlx13nm _l2uv11wk _7mfvdecw _szlb17km"]),
    style: {
      "--_1f6rtr1": (0, _runtime.ix)("var(--ds-surface-overlay, ".concat(_colors.N0, ")")),
      "--_3t1odq": (0, _runtime.ix)("var(--ds-shadow-overlay, ".concat("0 4px 8px -2px ".concat(_colors.N50A, ", 0 0 1px ").concat(_colors.N60A), ")")),
      "--_12yx4z3": (0, _runtime.ix)("var(--ds-surface-overlay, ".concat(_colors.N20A, ")")),
      "--_qfu7kd": (0, _runtime.ix)("var(--ds-surface-overlay-hovered, ".concat(_colors.N30A, ")")),
      "--_1lrmnoy": (0, _runtime.ix)("var(--ds-text, ".concat(_colors.B400, ")"))
    }
  }, /*#__PURE__*/React.createElement(_new.IconButton, {
    icon: function icon(iconProps) {
      return /*#__PURE__*/React.createElement(_cross2.default, (0, _extends2.default)({}, iconProps, {
        LEGACY_size: "small",
        LEGACY_fallbackIcon: _cross.default
      }));
    },
    label: cancelButtonLabel,
    onClick: onCancelClick,
    onMouseDown: onMouseDown,
    testId: testId && "".concat(testId, "--cancel")
  })));
};

// eslint-disable-next-line @repo/internal/react/require-jsdoc
var _default = exports.default = Buttons;
