// @atlaskit/link-picker/dist/es2019/ui/link-picker/search-results/link-search-list/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./index.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { forwardRef, Fragment, useCallback, useRef } from 'react';
import { defineMessages, FormattedMessage } from 'react-intl-next';
import { fg } from '@atlaskit/platform-feature-flags';
import Spinner from '@atlaskit/spinner';
import VisuallyHidden from '@atlaskit/visually-hidden';
import { MinHeightContainer } from '../../../../common/ui/min-height-container';
import { handleNavKeyDown } from '../../../../common/utils/handleNavKeyDown';
import { NoResults, testIds as noResultsTestIds } from './link-search-no-results';
import { LinkSearchListItem, testIds as searchResultItemTestIds } from './list-item';
import { useTrackResultsShown } from './use-track-results-shown';
const styles = {
  emptyStateNoResultsWrapper: "_1tkepxbi"
};
const listContainerStyles = null;
const spinnerContainerStyles = null;
const listStyles = null;
const baseListTitleStyles = null;
const oldListTitleStyles = null;
const newListTitleStyles = null;
export const messages = defineMessages({
  titleRecentlyViewed: {
    id: 'fabric.linkPicker.listTitle.recentlyViewed',
    defaultMessage: 'Recently Viewed',
    description: 'Describes type of items shown in the list for screen-reader users'
  },
  titleResults: {
    id: 'fabric.linkPicker.listTitle.results',
    defaultMessage: 'Results',
    description: 'Describes type of items shown in the list for screen-reader users'
  },
  searchLinkResults: {
    id: 'fabric.linkPicker.hyperlink.searchLinkResults',
    defaultMessage: '{count, plural, =0 {no results} one {# result} other {# results}} found',
    description: 'Announce search results for screen-reader users.'
  }
});
export const testIds = {
  ...noResultsTestIds,
  ...searchResultItemTestIds,
  resultListTitle: 'link-picker-list-title',
  searchResultList: 'link-search-list',
  searchResultLoadingIndicator: 'link-picker.results-loading-indicator',
  tabsLoadingIndicator: 'link-picker.tabs-loading-indicator'
};
export const LinkSearchList = /*#__PURE__*/forwardRef(({
  onChange,
  onSelect,
  onKeyDown,
  items,
  activeIndex,
  selectedIndex,
  isLoading,
  ariaControls,
  ariaLabelledBy,
  ariaReadOnly,
  role,
  id,
  hasSearchTerm,
  activePlugin,
  adaptiveHeight,
  className,
  tabPanelId,
  ...restProps
}, ref) => {
  var _activePlugin$uiOptio;
  let itemsContent;
  let loadingContent;
  const linkListTitle = hasSearchTerm ? messages.titleResults : messages.titleRecentlyViewed;
  useTrackResultsShown(isLoading, items, hasSearchTerm);
  const itemRefs = useRef({});
  const itemRefCallback = useCallback((el, id) => {
    if (el === null) {
      delete itemRefs.current[id];
    } else {
      itemRefs.current[id] = el;
    }
  }, []);
  const getTabIndex = useCallback(index => {
    if (selectedIndex > -1) {
      return selectedIndex === index ? 0 : -1;
    }
    if (index === 0) {
      return 0;
    }
    return -1;
  }, [selectedIndex]);
  const handleKeyDown = useCallback(event => {
    let updatedIndex = activeIndex;
    if (onKeyDown) {
      onKeyDown(event);
    }
    if (!(items !== null && items !== void 0 && items.length)) {
      return;
    }
    updatedIndex = handleNavKeyDown(event, items.length, activeIndex);
    const item = items === null || items === void 0 ? void 0 : items[updatedIndex];
    if (['ArrowDown', 'ArrowUp', 'Home', 'End'].includes(event.key) && item) {
      if (itemRefs.current) {
        var _itemRefs$current$ite;
        (_itemRefs$current$ite = itemRefs.current[item.objectId]) === null || _itemRefs$current$ite === void 0 ? void 0 : _itemRefs$current$ite.focus();
      }
    }
  }, [activeIndex, items, onKeyDown]);
  if ((items === null || items === void 0 ? void 0 : items.length) === 0) {
    if (!hasSearchTerm) {
      var _activePlugin$emptySt;
      const emptyState = activePlugin === null || activePlugin === void 0 ? void 0 : (_activePlugin$emptySt = activePlugin.emptyStateNoResults) === null || _activePlugin$emptySt === void 0 ? void 0 : _activePlugin$emptySt.call(activePlugin);
      if (emptyState) {
        return /*#__PURE__*/React.createElement("div", {
          id: tabPanelId,
          className: ax([styles.emptyStateNoResultsWrapper])
        }, emptyState);
      }
    }
    return /*#__PURE__*/React.createElement("div", {
      id: tabPanelId
    }, /*#__PURE__*/React.createElement(NoResults, null));
  }
  const listItemNameMaxLines = activePlugin === null || activePlugin === void 0 ? void 0 : (_activePlugin$uiOptio = activePlugin.uiOptions) === null || _activePlugin$uiOptio === void 0 ? void 0 : _activePlugin$uiOptio.listItemNameMaxLines;
  if (items && items.length > 0) {
    itemsContent = /*#__PURE__*/React.createElement(Fragment, null, /*#__PURE__*/React.createElement("div", {
      id: testIds.resultListTitle,
      "data-testid": testIds.resultListTitle,
      className: ax(["_11c81o8v _k48pmoej _otyr1b66", fg('platform-linking-visual-refresh-v1') ? "_syaz1gjq" : "_syaz131l"])
    }, /*#__PURE__*/React.createElement(FormattedMessage, linkListTitle)), /*#__PURE__*/React.createElement(VisuallyHidden, {
      id: "fabric.smartcard.linkpicker.suggested.results"
    }, hasSearchTerm && /*#__PURE__*/React.createElement(FormattedMessage, _extends({}, messages.searchLinkResults, {
      values: {
        count: items.length
      },
      "aria-live": "polite",
      "aria-atomic": "true"
    }))), /*#__PURE__*/React.createElement("ul", {
      id: id,
      role: role,
      "aria-controls": "fabric.smartcard.linkpicker.suggested.results",
      "aria-labelledby": testIds.resultListTitle,
      "aria-readonly": ariaReadOnly,
      "data-testid": testIds.searchResultList,
      className: ax(["_qtt8glyw _ca0qze3t _u5f3ze3t _n3tdze3t _19bvze3t _19pkze3t _otyrze3t _18u01mjz _2hwxum63"])
    }, items.map((item, index) => /*#__PURE__*/React.createElement(LinkSearchListItem, {
      id: `${testIds.searchResultItem}-${index}`,
      role: role && 'option',
      item: item,
      selected: selectedIndex === index,
      active: activeIndex === index,
      onFocus: () => onChange(item.objectId),
      onKeyDown: handleKeyDown,
      onSelect: onSelect,
      key: item.objectId,
      tabIndex: getTabIndex(index),
      ref: el => itemRefCallback(el, item.objectId),
      nameMaxLines: listItemNameMaxLines
    }))));
  }
  if (isLoading) {
    loadingContent = /*#__PURE__*/React.createElement(MinHeightContainer, {
      minHeight: adaptiveHeight ? '80px' : '50px',
      className: ax(["_16jlkb7n _2lx21bp4 _4cvr1h6o"])
    }, /*#__PURE__*/React.createElement(Spinner, {
      testId: testIds.searchResultLoadingIndicator,
      interactionName: "link-picker-search-list-loading",
      size: "medium"
    }));
  }
  return /*#__PURE__*/React.createElement("div", _extends({
    id: tabPanelId,
    ref: ref,
    className: ax(["_1bsb1osq _ca0qidpf _1tkeys9h _19pkpxbi _otyrpxbi _16jlkb7n _1e0c1txw _2lx21bp4", className])
  }, restProps), itemsContent, loadingContent);
});
