// @atlaskit/side-navigation/dist/es2019/components/Header/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./index.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
import React, { forwardRef } from 'react';
import { N500 } from '@atlaskit/theme/colors';
import { CustomItem } from '../Item';
const styles = {
  header: "_vrjt140y _1cmd10s3 _1cg71wug _11kjqtfy _jl2n73ad _1et6qtfy _1v6j10s3"
};

/**
 * __Container__
 *
 * A container for Header and Footer that safely handles props to the child component
 */
export const Container = ({
  children,
  'data-testid': testId,
  ...props
}) => {
  // https://stackoverflow.com/a/39333479
  const safeProps = (({
    className,
    onClick,
    onMouseDown,
    onDragStart,
    draggable,
    ref,
    tabIndex,
    disabled
  }) => ({
    className,
    onClick,
    onMouseDown,
    onDragStart,
    draggable,
    ref,
    tabIndex,
    disabled
  }))(props);
  return (
    /*#__PURE__*/
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
    React.createElement("div", _extends({
      "data-testid": testId,
      style: {
        position: 'relative'
      }
    }, safeProps), children)
  );
};
/**
 * __Header__
 *
 * - [Examples](https://atlassian.design/components/side-navigation/examples#header-and-footer)
 * - [Code](https://atlassian.design/components/side-navigation/code)
 */
const Header = /*#__PURE__*/forwardRef((props, ref) => {
  return /*#__PURE__*/React.createElement(CustomItem, _extends({}, props, {
    ref: ref,
    component: props.component || Container,
    // eslint-disable-next-line @repo/internal/react/no-unsafe-overrides
    overrides: {
      Title: {
        // eslint-disable-next-line @atlaskit/design-system/use-heading
        render: (_, {
          children,
          ...props
        }) => /*#__PURE__*/React.createElement("h2", props, children)
      }
    },
    className: ax([styles.header])
  }));
});
export default Header;
