// @atlaskit/editor-common/dist/cjs/types/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AnnotationUpdateEmitter", {
  enumerable: true,
  get: function get() {
    return _emitter.AnnotationUpdateEmitter;
  }
});
Object.defineProperty(exports, "AnnotationUpdateEvent", {
  enumerable: true,
  get: function get() {
    return _emitter.AnnotationUpdateEvent;
  }
});
exports.SortOrder = void 0;
Object.defineProperty(exports, "ToolbarSize", {
  enumerable: true,
  get: function get() {
    return _toolbar.ToolbarSize;
  }
});
Object.defineProperty(exports, "ToolbarWidths", {
  enumerable: true,
  get: function get() {
    return _toolbar.ToolbarWidths;
  }
});
Object.defineProperty(exports, "ToolbarWidthsFullPage", {
  enumerable: true,
  get: function get() {
    return _toolbar.ToolbarWidthsFullPage;
  }
});
Object.defineProperty(exports, "ToolbarWidthsFullPageNext", {
  enumerable: true,
  get: function get() {
    return _toolbar.ToolbarWidthsFullPageNext;
  }
});
Object.defineProperty(exports, "ToolbarWidthsNext", {
  enumerable: true,
  get: function get() {
    return _toolbar.ToolbarWidthsNext;
  }
});
var _emitter = require("./annotation/emitter");
var _toolbar = require("./toolbar");
// Disable no-re-export rule for entry point files
/* eslint-disable @atlaskit/editor/no-re-export */
var SortOrder = exports.SortOrder = /*#__PURE__*/function (SortOrder) {
  SortOrder["ASC"] = "asc";
  SortOrder["DESC"] = "desc";
  SortOrder["NO_ORDER"] = "no_order";
  return SortOrder;
}({}); // eslint-disable-next-line @repo/internal/deprecations/deprecation-ticket-required -- Ignored via go/ED-25883
/**
 *  @deprecated
 */
// eslint-disable-next-line @repo/internal/deprecations/deprecation-ticket-required -- Ignored via go/ED-25883
/**
 * @deprecated
 *
 * Please use ExtractPresetAPI from "@atlaskit/editor-common/preset" instead.
 */
// Ignored via go/ees005
// eslint-disable-next-line @typescript-eslint/no-explicit-any
