// @atlaskit/editor-plugin-tasks-and-decisions/dist/cjs/ui/Task/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TaskItem = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _reactIntlNext = require("react-intl-next");
var _messages = require("@atlaskit/editor-common/messages");
var _providerFactory2 = require("@atlaskit/editor-common/provider-factory");
var _useSharedPluginStateSelector = require("@atlaskit/editor-common/use-shared-plugin-state-selector");
var _taskItemWithProviders = _interopRequireDefault(require("./task-item-with-providers"));
var _excluded = ["providers", "intl", "taskDecisionProvider", "api", "placeholder"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
// eslint-disable-next-line @repo/internal/react/no-class-components
var TaskItem = exports.TaskItem = /*#__PURE__*/function (_PureComponent) {
  function TaskItem(props) {
    var _this;
    (0, _classCallCheck2.default)(this, TaskItem);
    _this = _callSuper(this, TaskItem, [props]);
    (0, _defineProperty2.default)(_this, "renderWithProvider", function (providers) {
      var _this$props = _this.props,
        _providerFactory = _this$props.providers,
        formatMessage = _this$props.intl.formatMessage,
        taskDecisionProvider = _this$props.taskDecisionProvider,
        api = _this$props.api,
        placeholder = _this$props.placeholder,
        otherProps = (0, _objectWithoutProperties2.default)(_this$props, _excluded);
      var contextIdentifierProvider = providers.contextIdentifierProvider;
      return /*#__PURE__*/_react.default.createElement(_taskItemWithProviders.default
      // Ignored via go/ees005
      // eslint-disable-next-line react/jsx-props-no-spreading
      , (0, _extends2.default)({}, otherProps, {
        placeholder: placeholder !== null && placeholder !== void 0 ? placeholder : formatMessage(_messages.tasksAndDecisionsMessages.taskPlaceholder),
        taskDecisionProvider: taskDecisionProvider ? Promise.resolve(taskDecisionProvider) : undefined,
        contextIdentifierProvider: contextIdentifierProvider
      }));
    });
    _this.providerFactory = props.providers || new _providerFactory2.ProviderFactory();
    return _this;
  }
  (0, _inherits2.default)(TaskItem, _PureComponent);
  return (0, _createClass2.default)(TaskItem, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (!this.props.providers) {
        // new ProviderFactory is created if no `providers` has been set
        // in this case when component is unmounted it's safe to destroy this providerFactory
        this.providerFactory.destroy();
      }
    }
  }, {
    key: "render",
    value: function render() {
      var providers = ['contextIdentifierProvider'];
      return /*#__PURE__*/_react.default.createElement(_providerFactory2.WithProviders, {
        providers: providers,
        providerFactory: this.providerFactory,
        renderNode: this.renderWithProvider
      });
    }
  }]);
}(_react.PureComponent);
(0, _defineProperty2.default)(TaskItem, "displayName", 'TaskItem');
var TaskItemWrapper = function TaskItemWrapper(props) {
  var provider = (0, _useSharedPluginStateSelector.useSharedPluginStateSelector)(props.api, 'taskDecision.taskDecisionProvider');
  // Ignored via go/ees005
  // eslint-disable-next-line react/jsx-props-no-spreading
  return /*#__PURE__*/_react.default.createElement(TaskItem, (0, _extends2.default)({
    taskDecisionProvider: provider
  }, props));
};
var _default = exports.default = (0, _reactIntlNext.injectIntl)(TaskItemWrapper);
