// @atlaskit/media-ui/dist/cjs/formatDate.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.partsFormatter = exports.formatterOptions = exports.formatDate = void 0;
var _locale = require("@atlaskit/locale");
var formatterOptions = exports.formatterOptions = {
  day: '2-digit',
  month: 'short',
  year: 'numeric',
  hour: '2-digit',
  minute: '2-digit',
  hour12: true
};
var partsFormatter = exports.partsFormatter = function partsFormatter(_ref) {
  var day = _ref.day,
    month = _ref.month,
    year = _ref.year,
    hour = _ref.hour,
    minute = _ref.minute,
    _ref$dayPeriod = _ref.dayPeriod,
    dayPeriod = _ref$dayPeriod === void 0 ? '' : _ref$dayPeriod;
  var formattedDayPeriod = dayPeriod.replace(/\./g, '').replace(/\s/g, '');
  return "".concat(day, " ").concat(month, " ").concat(year, ", ").concat(hour, ":").concat(minute, " ").concat(formattedDayPeriod);
};
var formatDate = exports.formatDate = function formatDate(timestamp) {
  var locale = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'en';
  var l10n = (0, _locale.createLocalizationProvider)(locale, formatterOptions);
  return partsFormatter(l10n.formatToParts(timestamp));
};
