// @atlaskit/media-ui/dist/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Bounds", {
  enumerable: true,
  get: function get() {
    return _camera.Bounds;
  }
});
Object.defineProperty(exports, "Camera", {
  enumerable: true,
  get: function get() {
    return _camera.Camera;
  }
});
Object.defineProperty(exports, "CustomMediaPlayer", {
  enumerable: true,
  get: function get() {
    return _index2.CustomMediaPlayer;
  }
});
Object.defineProperty(exports, "Ellipsify", {
  enumerable: true,
  get: function get() {
    return _ellipsify.Ellipsify;
  }
});
Object.defineProperty(exports, "ExifOrientation", {
  enumerable: true,
  get: function get() {
    return _imageMetaData.ExifOrientation;
  }
});
Object.defineProperty(exports, "InactivityDetector", {
  enumerable: true,
  get: function get() {
    return _inactivityDetector.InactivityDetector;
  }
});
Object.defineProperty(exports, "MediaButton", {
  enumerable: true,
  get: function get() {
    return _MediaButton.default;
  }
});
Object.defineProperty(exports, "MediaImage", {
  enumerable: true,
  get: function get() {
    return _mediaImage.MediaImage;
  }
});
Object.defineProperty(exports, "MediaInlineCardErroredView", {
  enumerable: true,
  get: function get() {
    return _MediaInlineCard.MediaInlineCardErroredView;
  }
});
Object.defineProperty(exports, "MediaInlineCardLoadedView", {
  enumerable: true,
  get: function get() {
    return _MediaInlineCard.MediaInlineCardLoadedView;
  }
});
Object.defineProperty(exports, "MediaInlineCardLoadingView", {
  enumerable: true,
  get: function get() {
    return _MediaInlineCard.MediaInlineCardLoadingView;
  }
});
Object.defineProperty(exports, "MimeTypeIcon", {
  enumerable: true,
  get: function get() {
    return _mimeTypeIcon.MimeTypeIcon;
  }
});
Object.defineProperty(exports, "ModalSpinner", {
  enumerable: true,
  get: function get() {
    return _modalSpinner.default;
  }
});
Object.defineProperty(exports, "Rectangle", {
  enumerable: true,
  get: function get() {
    return _camera.Rectangle;
  }
});
Object.defineProperty(exports, "Shortcut", {
  enumerable: true,
  get: function get() {
    return _shortcut.Shortcut;
  }
});
Object.defineProperty(exports, "TimeRange", {
  enumerable: true,
  get: function get() {
    return _timeRange.TimeRange;
  }
});
Object.defineProperty(exports, "Truncate", {
  enumerable: true,
  get: function get() {
    return _truncateText.Truncate;
  }
});
Object.defineProperty(exports, "Vector2", {
  enumerable: true,
  get: function get() {
    return _camera.Vector2;
  }
});
Object.defineProperty(exports, "absolute", {
  enumerable: true,
  get: function get() {
    return _mixins.absolute;
  }
});
Object.defineProperty(exports, "borderRadius", {
  enumerable: true,
  get: function get() {
    return _mixins.borderRadius;
  }
});
Object.defineProperty(exports, "borderRadiusBottom", {
  enumerable: true,
  get: function get() {
    return _mixins.borderRadiusBottom;
  }
});
Object.defineProperty(exports, "center", {
  enumerable: true,
  get: function get() {
    return _mixins.center;
  }
});
Object.defineProperty(exports, "dataURItoFile", {
  enumerable: true,
  get: function get() {
    return _util.dataURItoFile;
  }
});
Object.defineProperty(exports, "easeInOutCubic", {
  enumerable: true,
  get: function get() {
    return _mixins.easeInOutCubic;
  }
});
Object.defineProperty(exports, "ellipsis", {
  enumerable: true,
  get: function get() {
    return _mixins.ellipsis;
  }
});
Object.defineProperty(exports, "errorIcon", {
  enumerable: true,
  get: function get() {
    return _errorIcon.errorIcon;
  }
});
Object.defineProperty(exports, "fileToArrayBuffer", {
  enumerable: true,
  get: function get() {
    return _util.fileToArrayBuffer;
  }
});
Object.defineProperty(exports, "fileToDataURI", {
  enumerable: true,
  get: function get() {
    return _util.fileToDataURI;
  }
});
Object.defineProperty(exports, "findParentByClassname", {
  enumerable: true,
  get: function get() {
    return _util.findParentByClassname;
  }
});
Object.defineProperty(exports, "formatDate", {
  enumerable: true,
  get: function get() {
    return _formatDate.formatDate;
  }
});
Object.defineProperty(exports, "formatDuration", {
  enumerable: true,
  get: function get() {
    return _formatDuration.formatDuration;
  }
});
Object.defineProperty(exports, "getCssFromImageOrientation", {
  enumerable: true,
  get: function get() {
    return _imageMetaData.getCssFromImageOrientation;
  }
});
Object.defineProperty(exports, "getFileInfo", {
  enumerable: true,
  get: function get() {
    return _util.getFileInfo;
  }
});
Object.defineProperty(exports, "getFileInfoFromSrc", {
  enumerable: true,
  get: function get() {
    return _util.getFileInfoFromSrc;
  }
});
Object.defineProperty(exports, "getImageInfo", {
  enumerable: true,
  get: function get() {
    return _imageMetaData.getImageInfo;
  }
});
Object.defineProperty(exports, "getMetaTagNumericValue", {
  enumerable: true,
  get: function get() {
    return _imageMetaData.getMetaTagNumericValue;
  }
});
Object.defineProperty(exports, "getMimeIcon", {
  enumerable: true,
  get: function get() {
    return _util.getMimeIcon;
  }
});
Object.defineProperty(exports, "getOrientation", {
  enumerable: true,
  get: function get() {
    return _imageMetaData.getOrientation;
  }
});
Object.defineProperty(exports, "getScaleFactor", {
  enumerable: true,
  get: function get() {
    return _imageMetaData.getScaleFactor;
  }
});
Object.defineProperty(exports, "getScaleFactorFromFile", {
  enumerable: true,
  get: function get() {
    return _imageMetaData.getScaleFactorFromFile;
  }
});
Object.defineProperty(exports, "hideControlsClassName", {
  enumerable: true,
  get: function get() {
    return _classNames.hideControlsClassName;
  }
});
Object.defineProperty(exports, "isIntersectionObserverSupported", {
  enumerable: true,
  get: function get() {
    return _intersectionObserver.isIntersectionObserverSupported;
  }
});
Object.defineProperty(exports, "isRotated", {
  enumerable: true,
  get: function get() {
    return _imageMetaData.isRotated;
  }
});
Object.defineProperty(exports, "keyCodes", {
  enumerable: true,
  get: function get() {
    return _shortcut.keyCodes;
  }
});
Object.defineProperty(exports, "languages", {
  enumerable: true,
  get: function get() {
    return _languages.default;
  }
});
Object.defineProperty(exports, "loadImage", {
  enumerable: true,
  get: function get() {
    return _util.loadImage;
  }
});
exports.locales = void 0;
Object.defineProperty(exports, "messages", {
  enumerable: true,
  get: function get() {
    return _messages.messages;
  }
});
Object.defineProperty(exports, "readImageMetaData", {
  enumerable: true,
  get: function get() {
    return _imageMetaData.readImageMetaData;
  }
});
Object.defineProperty(exports, "size", {
  enumerable: true,
  get: function get() {
    return _mixins.size;
  }
});
Object.defineProperty(exports, "toHumanReadableMediaSize", {
  enumerable: true,
  get: function get() {
    return _humanReadableSize.toHumanReadableMediaSize;
  }
});
var untypedLocales = _interopRequireWildcard(require("./i18n/index"));
var _MediaInlineCard = require("./MediaInlineCard");
var _ellipsify = require("./ellipsify");
var _modalSpinner = _interopRequireDefault(require("./modalSpinner"));
var _humanReadableSize = require("./humanReadableSize");
var _mixins = require("./mixins");
var _camera = require("./camera");
var _messages = require("./messages");
var _languages = _interopRequireDefault(require("./i18n/languages"));
var _imageMetaData = require("./imageMetaData");
var _util = require("./util");
var _mimeTypeIcon = require("./mime-type-icon");
var _index2 = require("./customMediaPlayer/index");
var _timeRange = require("./customMediaPlayer/timeRange");
var _classNames = require("./classNames");
var _shortcut = require("./shortcut");
var _formatDuration = require("./formatDuration");
var _MediaButton = _interopRequireDefault(require("./MediaButton"));
var _mediaImage = require("./mediaImage");
var _inactivityDetector = require("./inactivityDetector/inactivityDetector");
var _intersectionObserver = require("./intersectionObserver");
var _truncateText = require("./truncateText");
var _errorIcon = require("./errorIcon");
var _formatDate = require("./formatDate");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var locales = exports.locales = untypedLocales;
