// @atlaskit/media-ui/dist/cjs/mixins.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.size = exports.ellipsis = exports.easeInOutCubic = exports.center = exports.borderRadiusBottom = exports.borderRadius = exports.absolute = void 0;
var _constants = require("@atlaskit/theme/constants");
var ellipsis = exports.ellipsis = function ellipsis() {
  var maxWidth = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '100%';
  var unit = typeof maxWidth === 'number' ? 'px' : '';
  return "\n    max-width: ".concat(maxWidth).concat(unit, ";\n    overflow: hidden;\n    text-overflow: ellipsis;\n    white-space: nowrap;\n  ");
};
var size = exports.size = function size() {
  var value = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '100%';
  var unit = typeof value === 'number' ? 'px' : '';
  return "\n    width: ".concat(value).concat(unit, ";\n    height: ").concat(value).concat(unit, ";\n  ");
};
var center = exports.center = "\n  display: flex;\n  align-items: center;\n  justify-content: center;\n";
var absolute = exports.absolute = function absolute() {
  var top = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0;
  var left = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
  return "\n  position: absolute;\n  top: ".concat(top, "px;\n  left: ").concat(left, "px;\n");
};
var borderRadius = exports.borderRadius = "\n  border-radius: ".concat((0, _constants.borderRadius)(), "px;\n");
var borderRadiusBottom = exports.borderRadiusBottom = "\n  border-bottom-left-radius: ".concat((0, _constants.borderRadius)(), "px;\n  border-bottom-right-radius: ").concat((0, _constants.borderRadius)(), "px;\n");
var easeInOutCubic = exports.easeInOutCubic = 'cubic-bezier(0.645, 0.045, 0.355, 1)';
