// @atlaskit/media-ui/dist/cjs/shortcut.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.keyCodes = exports.Shortcut = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = require("react");
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var keyCodes = exports.keyCodes = {
  space: 'Space',
  m: 'KeyM',
  rightArrow: 'ArrowRight',
  leftArrow: 'ArrowLeft'
};
var Shortcut = exports.Shortcut = /*#__PURE__*/function (_Component) {
  function Shortcut() {
    var _this;
    (0, _classCallCheck2.default)(this, Shortcut);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, Shortcut, [].concat(args));
    (0, _defineProperty2.default)(_this, "eventType", 'keydown');
    (0, _defineProperty2.default)(_this, "keyHandler", function (e) {
      var _this$props = _this.props,
        handler = _this$props.handler,
        code = _this$props.code;
      if (code !== undefined && e.code === code) {
        handler();
      }
    });
    (0, _defineProperty2.default)(_this, "init", function () {
      var _this$props$eventType = _this.props.eventType,
        eventType = _this$props$eventType === void 0 ? 'keydown' : _this$props$eventType;
      _this.eventType = eventType;
      document.addEventListener(_this.eventType, _this.keyHandler);
    });
    (0, _defineProperty2.default)(_this, "release", function () {
      document.removeEventListener(_this.eventType, _this.keyHandler);
    });
    return _this;
  }
  (0, _inherits2.default)(Shortcut, _Component);
  return (0, _createClass2.default)(Shortcut, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.init();
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.release();
    }
  }, {
    key: "render",
    value: function render() {
      return null;
    }
  }]);
}(_react.Component);
