// @atlaskit/media-ui/dist/cjs/truncateText-emotion.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TruncateRight = exports.TruncateLeft = exports.Truncate = void 0;
var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));
var _react = _interopRequireDefault(require("react"));
var _styled = _interopRequireDefault(require("@emotion/styled"));
var _truncateTextCompiled = require("./truncateText-compiled");
var _templateObject, _templateObject2; // eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
var truncateCommonStyles = "\n  display: inline-block;\n  vertical-align: bottom;\n  white-space: nowrap;\n  overflow: hidden;\n";

// eslint-disable-next-line @atlaskit/design-system/no-styled-tagged-template-expression, @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var TruncateLeft = exports.TruncateLeft = _styled.default.span(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2.default)(["\n\t", "\n\tmax-width: calc(100% - ", "em);\n\tmin-width: ", ";\n\ttext-overflow: ellipsis;\n"])), truncateCommonStyles, function (_ref) {
  var fontSizePX = _ref.fontSizePX,
    endFixedChars = _ref.endFixedChars;
  return fontFaceScaleFactor(fontSizePX) * endFixedChars + 1;
}, function (_ref2) {
  var fontSizePX = _ref2.fontSizePX,
    startFixedChars = _ref2.startFixedChars;
  return fontFaceScaleFactor(fontSizePX) * startFixedChars;
});

// eslint-disable-next-line @atlaskit/design-system/no-styled-tagged-template-expression, @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var TruncateRight = exports.TruncateRight = _styled.default.span(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2.default)(["\n\t", "\n\tmax-width: calc(100% - ", ");\n\tposition: relative;\n"])), truncateCommonStyles, function (_ref3) {
  var fontSizePX = _ref3.fontSizePX,
    startFixedChars = _ref3.startFixedChars;
  return fontFaceScaleFactor(fontSizePX) * startFixedChars;
});
var fontFaceScaleFactor = function fontFaceScaleFactor(fontSizePX) {
  return fontSizePX / 11 * 0.46;
}; /* factor for fontSize of 11px */

var placeholder = ' ';
var Truncate = exports.Truncate = function Truncate(_ref4) {
  var text = _ref4.text,
    _ref4$fontSizePX = _ref4.fontSizePX,
    fontSizePX = _ref4$fontSizePX === void 0 ? 11 : _ref4$fontSizePX,
    _ref4$startFixedChars = _ref4.startFixedChars,
    startFixedChars = _ref4$startFixedChars === void 0 ? 4 : _ref4$startFixedChars,
    _ref4$endFixedChars = _ref4.endFixedChars,
    endFixedChars = _ref4$endFixedChars === void 0 ? 7 : _ref4$endFixedChars;
  var _calculateTruncation = (0, _truncateTextCompiled.calculateTruncation)(text, startFixedChars, endFixedChars),
    leftStr = _calculateTruncation.left,
    rightStr = _calculateTruncation.right;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(TruncateLeft, {
    fontSizePX: fontSizePX,
    startFixedChars: startFixedChars,
    endFixedChars: endFixedChars,
    "data-testid": "truncate-left"
  }, leftStr), /*#__PURE__*/_react.default.createElement(TruncateRight, {
    fontSizePX: fontSizePX,
    startFixedChars: startFixedChars,
    endFixedChars: endFixedChars,
    "data-testid": "truncate-right"
  }, leftStr === rightStr ? placeholder : rightStr));
};
