// @atlaskit/media-ui/dist/cjs/truncateText-compiled.js
/* truncateText-compiled.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.calculateTruncation = exports.TruncateRight = exports.TruncateLeft = exports.Truncate = void 0;
require("./truncateText-compiled.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireWildcard(require("react"));
var _excluded = ["fontSizePX", "startFixedChars", "endFixedChars", "children"],
  _excluded2 = ["fontSizePX", "startFixedChars", "endFixedChars", "children"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var calculateTruncation = exports.calculateTruncation = function calculateTruncation(text, startFixedChars, endFixedChars) {
  var minAllowedLength = endFixedChars + startFixedChars;
  if (text.length <= minAllowedLength) {
    // if no truncation return same value for left and right
    return {
      left: text,
      right: text
    };
  }
  var splitAt = text.length - endFixedChars;
  var left = text.substr(0, splitAt);
  var right = text.substr(splitAt);
  return {
    left: left,
    right: right
  };
};
var truncateCommonStyles = null;
var truncateLeftStyles = null;
var TruncateLeft = exports.TruncateLeft = function TruncateLeft(_ref) {
  var fontSizePX = _ref.fontSizePX,
    startFixedChars = _ref.startFixedChars,
    endFixedChars = _ref.endFixedChars,
    children = _ref.children,
    props = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  return /*#__PURE__*/_react.default.createElement("span", (0, _extends2.default)({
    style: {
      maxWidth: "calc(\n\t\t100% -\n\t\t\t".concat(fontFaceScaleFactor(fontSizePX) * endFixedChars + 1, "em\n\t)"),
      minWidth: fontFaceScaleFactor(fontSizePX) * startFixedChars
    },
    "data-testid": "truncate-left"
  }, props, {
    className: (0, _runtime.ax)(["_1reo15vq _18m915vq _1e0c1o8l _s7n4yfq0 _o5721q9c", "_1bto1l2s"])
  }), children);
};
var truncateRightStyles = null;
var TruncateRight = exports.TruncateRight = function TruncateRight(_ref2) {
  var fontSizePX = _ref2.fontSizePX,
    startFixedChars = _ref2.startFixedChars,
    endFixedChars = _ref2.endFixedChars,
    children = _ref2.children,
    props = (0, _objectWithoutProperties2.default)(_ref2, _excluded2);
  return /*#__PURE__*/_react.default.createElement("span", (0, _extends2.default)({
    style: {
      maxWidth: "calc(\n\t\t\t\t100% -\n\t\t\t\t\t".concat(function (_ref3) {
        var fontSizePX = _ref3.fontSizePX,
          startFixedChars = _ref3.startFixedChars;
        return fontFaceScaleFactor(fontSizePX) * startFixedChars;
      }, "\n\t\t\t)")
    },
    "data-testid": "truncate-right"
  }, props, {
    className: (0, _runtime.ax)(["_1reo15vq _18m915vq _1e0c1o8l _s7n4yfq0 _o5721q9c", "_kqswh2mm"])
  }), children);
};
var fontFaceScaleFactor = function fontFaceScaleFactor(fontSizePX) {
  return fontSizePX / 11 * 0.46;
}; /* factor for fontSize of 11px */

var placeholder = ' ';
var Truncate = exports.Truncate = function Truncate(_ref4) {
  var text = _ref4.text,
    _ref4$fontSizePX = _ref4.fontSizePX,
    fontSizePX = _ref4$fontSizePX === void 0 ? 11 : _ref4$fontSizePX,
    _ref4$startFixedChars = _ref4.startFixedChars,
    startFixedChars = _ref4$startFixedChars === void 0 ? 4 : _ref4$startFixedChars,
    _ref4$endFixedChars = _ref4.endFixedChars,
    endFixedChars = _ref4$endFixedChars === void 0 ? 7 : _ref4$endFixedChars;
  var _calculateTruncation = calculateTruncation(text, startFixedChars, endFixedChars),
    leftStr = _calculateTruncation.left,
    rightStr = _calculateTruncation.right;
  return /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(TruncateLeft, {
    fontSizePX: fontSizePX,
    startFixedChars: startFixedChars,
    endFixedChars: endFixedChars
  }, leftStr), /*#__PURE__*/_react.default.createElement(TruncateRight, {
    fontSizePX: fontSizePX,
    startFixedChars: startFixedChars,
    endFixedChars: endFixedChars
  }, leftStr === rightStr ? placeholder : rightStr));
};
