// @atlaskit/media-ui/dist/cjs/util.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.dataURItoFile = dataURItoFile;
exports.fileToArrayBuffer = fileToArrayBuffer;
exports.fileToDataURI = fileToDataURI;
exports.findParentByClassname = void 0;
exports.getFileInfo = getFileInfo;
exports.getFileInfoFromSrc = getFileInfoFromSrc;
exports.getMimeIcon = getMimeIcon;
exports.loadImage = loadImage;
exports.readImageNaturalOrientationFromDOM = readImageNaturalOrientationFromDOM;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _ = _interopRequireDefault(require("@atlaskit/icon-file-type/glyph/pdf-document/24"));
var _2 = _interopRequireDefault(require("@atlaskit/icon-file-type/glyph/gif/24"));
var _3 = _interopRequireDefault(require("@atlaskit/icon-file-type/glyph/powerpoint-presentation/24"));
var _4 = _interopRequireDefault(require("@atlaskit/icon-file-type/glyph/word-document/24"));
var _5 = _interopRequireDefault(require("@atlaskit/icon-file-type/glyph/sketch/24"));
var _6 = _interopRequireDefault(require("@atlaskit/icon-file-type/glyph/figma/24"));
var _7 = _interopRequireDefault(require("@atlaskit/icon-file-type/glyph/executable/24"));
var _8 = _interopRequireDefault(require("@atlaskit/icon-file-type/glyph/google-doc/24"));
var _9 = _interopRequireDefault(require("@atlaskit/icon-file-type/glyph/google-form/24"));
var _10 = _interopRequireDefault(require("@atlaskit/icon-file-type/glyph/google-sheet/24"));
var _11 = _interopRequireDefault(require("@atlaskit/icon-file-type/glyph/google-slide/24"));
var _12 = _interopRequireDefault(require("@atlaskit/icon-file-type/glyph/excel-spreadsheet/24"));
var _13 = _interopRequireDefault(require("@atlaskit/icon-file-type/glyph/spreadsheet/24"));
var _14 = _interopRequireDefault(require("@atlaskit/icon-file-type/glyph/presentation/24"));
var _15 = _interopRequireDefault(require("@atlaskit/icon-file-type/glyph/source-code/24"));
var _codeViewer = require("./codeViewer");
function dataURItoFile(dataURI) {
  var filename = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'untitled';
  if (dataURI.length === 0) {
    throw new Error('dataURI not found');
  }

  // convert base64/URLEncoded data component to raw binary data held in a string
  var dataURIParts = dataURI.split(',');
  var byteString = dataURIParts[0].indexOf('base64') >= 0 ? atob(dataURIParts[1]) : decodeURIComponent(dataURIParts[1]);

  // separate out the mime component
  var mimeString;
  try {
    mimeString = dataURI.split(',')[0].split(':')[1].split(';')[0];
  } catch (e) {
    // https://stackoverflow.com/questions/1176022/unknown-file-type-mime
    mimeString = 'application/octet-stream';
  }

  // write the bytes of the string to a typed array
  var intArray = new Uint8Array(byteString.length);
  for (var i = 0; i < byteString.length; i++) {
    intArray[i] = byteString.charCodeAt(i);
  }
  var blob = new Blob([intArray], {
    type: mimeString
  });
  try {
    return new File([blob], filename, {
      type: mimeString
    });
  } catch (e) {
    // IE11 does not allow the File constructor (yay!)
    // we get around this by decorating the blob instance with File properties
    // effectively casting up from Blob to File.
    var ie11File = blob;
    var date = new Date();
    ie11File.lastModified = date.getTime();
    ie11File.lastModifiedDate = date;
    ie11File.name = filename;
    ie11File.webkitRelativePath = '';
    return ie11File;
  }
}
function fileToDataURI(blob) {
  return new Promise(function (resolve, reject) {
    var reader = new FileReader();
    reader.addEventListener('load', function () {
      var result = reader.result;
      if (typeof result === 'string') {
        resolve(result);
      } else if (result === null) {
        reject();
      }
    });
    reader.addEventListener('error', reject);
    reader.readAsDataURL(blob);
  });
}
function getFileInfo(_x, _x2) {
  return _getFileInfo.apply(this, arguments);
}
function _getFileInfo() {
  _getFileInfo = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(file, src) {
    return _regenerator.default.wrap(function _callee$(_context) {
      while (1) switch (_context.prev = _context.next) {
        case 0:
          _context.t0 = file;
          _context.t1 = src;
          if (_context.t1) {
            _context.next = 6;
            break;
          }
          _context.next = 5;
          return fileToDataURI(file);
        case 5:
          _context.t1 = _context.sent;
        case 6:
          _context.t2 = _context.t1;
          return _context.abrupt("return", {
            file: _context.t0,
            src: _context.t2
          });
        case 8:
        case "end":
          return _context.stop();
      }
    }, _callee);
  }));
  return _getFileInfo.apply(this, arguments);
}
function getFileInfoFromSrc(_x3, _x4) {
  return _getFileInfoFromSrc.apply(this, arguments);
}
function _getFileInfoFromSrc() {
  _getFileInfoFromSrc = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee2(src, file) {
    return _regenerator.default.wrap(function _callee2$(_context2) {
      while (1) switch (_context2.prev = _context2.next) {
        case 0:
          _context2.t0 = file;
          if (_context2.t0) {
            _context2.next = 5;
            break;
          }
          _context2.next = 4;
          return dataURItoFile(src);
        case 4:
          _context2.t0 = _context2.sent;
        case 5:
          _context2.t1 = _context2.t0;
          _context2.t2 = src;
          return _context2.abrupt("return", {
            file: _context2.t1,
            src: _context2.t2
          });
        case 8:
        case "end":
          return _context2.stop();
      }
    }, _callee2);
  }));
  return _getFileInfoFromSrc.apply(this, arguments);
}
function fileToArrayBuffer(file) {
  return new Promise(function (resolve, reject) {
    var reader = new FileReader();
    reader.addEventListener('load', function () {
      var array = new Uint8Array(reader.result);
      resolve(array);
    });
    reader.addEventListener('error', reject);
    reader.readAsArrayBuffer(file);
  });
}
function loadImage(src) {
  return new Promise(function (resolve, reject) {
    var img = new Image();
    img.src = src;
    img.onload = function () {
      resolve(img);
    };
    img.onerror = reject;
  });
}
function readImageNaturalOrientationFromDOM(img) {
  img.style.position = 'absolute';
  img.style.visibility = 'hidden';
  document.body.appendChild(img);
  var _img$getBoundingClien = img.getBoundingClientRect(),
    width = _img$getBoundingClien.width,
    height = _img$getBoundingClien.height;
  document.body.removeChild(img);
  return {
    width: width,
    height: height
  };
}
var findParentByClassname = exports.findParentByClassname = function findParentByClassname(element, className) {
  var maxParentElement = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : document.body;
  if (element.classList.contains(className)) {
    return element;
  }
  var currentElement = element;
  while (currentElement.parentElement !== maxParentElement) {
    if (currentElement.parentElement) {
      currentElement = currentElement.parentElement;
      if (currentElement.classList.contains(className)) {
        return currentElement;
      }
    } else {
      return undefined;
    }
  }
  return undefined;
};
var mimeTypes = [{
  label: 'pdf',
  mimeTypes: ['application/pdf'],
  icon: _.default
}, {
  label: 'google-form',
  mimeTypes: ['application/vnd.google-apps.form'],
  icon: _9.default
}, {
  label: 'google-slides',
  mimeTypes: ['application/vnd.google-apps.presentation'],
  icon: _11.default
}, {
  label: 'google-form',
  mimeTypes: ['application/vnd.google-apps.form'],
  icon: _9.default
}, {
  label: 'google-sheets',
  mimeTypes: ['application/vnd.google-apps.spreadsheet'],
  icon: _10.default
}, {
  label: 'google-docs',
  mimeTypes: ['application/vnd.google-apps.document', 'application/vnd.google-apps.kix'],
  icon: _8.default
}, {
  label: 'microsoft-word',
  mimeTypes: ['application/msword', 'application/vnd.openxmlformats-officedocument.word'],
  icon: _4.default
}, {
  label: 'presentation',
  mimeTypes: ['application/x-iwork-keynote-sffkey', 'application/vnd.oasis.opendocument.presentation'],
  icon: _14.default
}, {
  label: 'powerpoint-presentation',
  mimeTypes: ['application/vnd.openxmlformats-officedocument.presentationml.presentation', 'application/vnd.ms-powerpoint'],
  icon: _3.default
}, {
  label: 'giphy',
  mimeTypes: ['image/gif'],
  icon: _2.default
}, {
  label: 'spreadsheet',
  mimeTypes: ['text/csv'],
  icon: _13.default
}, {
  label: 'excel-spreadsheet',
  mimeTypes: ['application/vnd.openxmlformats-officedocument.spreadsheetml.sheet', 'application/vnd.ms-excel', 'application/x-iwork-keynote-sffnumbers'],
  icon: _12.default
}];

/*
 * Returns a label and icon
 */
function getMimeIcon(mimeType, fileName) {
  // based on the mimeType, determine the corresponding icon and label
  var iconInfo = mimeTypes.find(function (file) {
    return file.mimeTypes.indexOf(mimeType) > -1;
  });

  //return the appropriate icon and its label if we have it
  if (iconInfo) {
    return iconInfo;
  }
  if ((0, _codeViewer.isCodeViewerItem)(fileName, mimeType)) {
    return {
      label: 'source-code',
      icon: _15.default
    };
  }

  // we are not able to determine what icon to render based on the mimeType
  // hence we render the icon based on the filename
  if (fileName.match(/.*\.sketch$/)) {
    return {
      label: 'sketch',
      icon: _5.default
    };
  }
  if (fileName.match(/.*\.fig$/)) {
    return {
      label: 'figma',
      icon: _6.default
    };
  }
  if (fileName.match(/.*\.exe$/) || fileName.match(/.*\.dmg$/)) {
    return {
      label: 'executable',
      icon: _7.default
    };
  }

  // cannot find a corresponding mimeType icon.
  return undefined;
}
