// @atlaskit/editor-plugin-card/dist/cjs/ui/editDatasourceAction.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.editDatasource = void 0;
var _analytics = require("@atlaskit/editor-common/analytics");
var _utils = require("@atlaskit/editor-common/utils");
var _actions = require("../pm-plugins/actions");
var editDatasource = exports.editDatasource = function editDatasource(datasourceId, editorAnalyticsApi, appearance, extensionKey) {
  return function (state, dispatch) {
    var datasourceType = (0, _utils.getDatasourceType)(datasourceId);
    if (dispatch && datasourceType) {
      var tr = state.tr;
      (0, _actions.showDatasourceModal)(datasourceType)(tr);
      editorAnalyticsApi === null || editorAnalyticsApi === void 0 || editorAnalyticsApi.attachAnalyticsEvent({
        action: _analytics.ACTION.CLICKED,
        actionSubject: _analytics.ACTION_SUBJECT.BUTTON,
        actionSubjectId: _analytics.ACTION_SUBJECT_ID.EDIT_DATASOURCE,
        eventType: _analytics.EVENT_TYPE.UI,
        attributes: {
          extensionKey: extensionKey,
          appearance: appearance
        }
      })(tr);
      dispatch(tr);
      return true;
    }
    return false;
  };
};
