// @atlaskit/editor-plugin-card/dist/cjs/ui/local-storage.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.markLocalStorageKeyDiscovered = exports.isLocalStorageKeyDiscovered = exports.ONE_DAY_IN_MILLISECONDS = exports.LOCAL_STORAGE_DISCOVERY_KEY_TOOLBAR = exports.LOCAL_STORAGE_DISCOVERY_KEY_SMART_LINK = exports.LOCAL_STORAGE_DISCOVERED_KEY = exports.LOCAL_STORAGE_CLIENT_KEY = void 0;
var _frontendUtilities = require("@atlaskit/frontend-utilities");
var LOCAL_STORAGE_CLIENT_KEY = exports.LOCAL_STORAGE_CLIENT_KEY = '@atlaskit/editor-plugin-card';
var LOCAL_STORAGE_DISCOVERED_KEY = exports.LOCAL_STORAGE_DISCOVERED_KEY = 'discovered';
var LOCAL_STORAGE_DISCOVERY_KEY_SMART_LINK = exports.LOCAL_STORAGE_DISCOVERY_KEY_SMART_LINK = 'smart-link-upgrade-pulse';
var LOCAL_STORAGE_DISCOVERY_KEY_TOOLBAR = exports.LOCAL_STORAGE_DISCOVERY_KEY_TOOLBAR = 'toolbar-upgrade-pulse';
var ONE_DAY_IN_MILLISECONDS = exports.ONE_DAY_IN_MILLISECONDS = 86400000;
var storageClient = new _frontendUtilities.StorageClient(LOCAL_STORAGE_CLIENT_KEY);
var isLocalStorageKeyDiscovered = exports.isLocalStorageKeyDiscovered = function isLocalStorageKeyDiscovered(key) {
  try {
    var localStorageValue = storageClient.getItem(key);
    return !!localStorageValue && localStorageValue === LOCAL_STORAGE_DISCOVERED_KEY;
  } catch (_unused) {
    // If localStorage is not available, don't show feature discovery component. Treat it as 'discovered'.
    return true;
  }
};
var markLocalStorageKeyDiscovered = exports.markLocalStorageKeyDiscovered = function markLocalStorageKeyDiscovered(key, expiration) {
  try {
    storageClient.setItemWithExpiry(key, LOCAL_STORAGE_DISCOVERED_KEY, expiration);
  } catch (_unused2) {
    // silent error
  }
};
