// @atlaskit/editor-plugin-card/dist/cjs/ui/useFetchDatasourceDataInfo.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useFetchDatasourceDataInfo = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = require("react");
var _linkClientExtension = require("@atlaskit/link-client-extension");
/**
 * @jsxRuntime classic
 * @jsx jsx
 */

var useFetchDatasourceDataInfo = exports.useFetchDatasourceDataInfo = function useFetchDatasourceDataInfo(_ref) {
  var datasourceId = _ref.datasourceId,
    parameters = _ref.parameters,
    visibleColumnKeys = _ref.visibleColumnKeys;
  var _useState = (0, _react.useState)(undefined),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    extensionKey = _useState2[0],
    setExtensionKey = _useState2[1];
  var _useDatasourceClientE = (0, _linkClientExtension.useDatasourceClientExtension)(),
    getDatasourceData = _useDatasourceClientE.getDatasourceData;
  (0, _react.useEffect)(function () {
    var fetchDatasource = /*#__PURE__*/function () {
      var _ref2 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee() {
        var datasourceDataRequest, _yield$getDatasourceD, meta;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              _context.prev = 0;
              if (!(!datasourceId || !parameters || !visibleColumnKeys)) {
                _context.next = 3;
                break;
              }
              return _context.abrupt("return");
            case 3:
              datasourceDataRequest = {
                parameters: parameters,
                pageSize: _linkClientExtension.DEFAULT_GET_DATASOURCE_DATA_PAGE_SIZE,
                pageCursor: undefined,
                fields: visibleColumnKeys,
                includeSchema: true
              };
              _context.next = 6;
              return getDatasourceData(datasourceId, datasourceDataRequest, false);
            case 6:
              _yield$getDatasourceD = _context.sent;
              meta = _yield$getDatasourceD.meta;
              setExtensionKey(meta.extensionKey);
              _context.next = 15;
              break;
            case 11:
              _context.prev = 11;
              _context.t0 = _context["catch"](0);
              // eslint-disable-next-line no-console
              console.error(_context.t0);
              setExtensionKey(undefined);
            case 15:
            case "end":
              return _context.stop();
          }
        }, _callee, null, [[0, 11]]);
      }));
      return function fetchDatasource() {
        return _ref2.apply(this, arguments);
      };
    }();
    void fetchDatasource();
  }, [getDatasourceData, visibleColumnKeys, parameters, datasourceId]);
  return {
    extensionKey: extensionKey
  };
};
