// @atlaskit/editor-common/dist/cjs/media/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "altTextMessages", {
  enumerable: true,
  get: function get() {
    return _altText.altTextMessages;
  }
});
Object.defineProperty(exports, "captionMessages", {
  enumerable: true,
  get: function get() {
    return _caption.captionMessages;
  }
});
Object.defineProperty(exports, "commentMessages", {
  enumerable: true,
  get: function get() {
    return _comments.commentMessages;
  }
});
Object.defineProperty(exports, "dropPlaceholderMessages", {
  enumerable: true,
  get: function get() {
    return _dropPlaceholder.dropPlaceholderMessages;
  }
});
Object.defineProperty(exports, "externalMediaMessages", {
  enumerable: true,
  get: function get() {
    return _externalMedia.externalMediaMessages;
  }
});
Object.defineProperty(exports, "imageBorderMessages", {
  enumerable: true,
  get: function get() {
    return _imageBorder.imageBorderMessages;
  }
});
Object.defineProperty(exports, "mediaLinkToolbarMessages", {
  enumerable: true,
  get: function get() {
    return _mediaLinkingToolbar.mediaLinkToolbarMessages;
  }
});
Object.defineProperty(exports, "mediaResizeAnnouncerMessMessages", {
  enumerable: true,
  get: function get() {
    return _mediaResizeAnnouncerMess.mediaResizeAnnouncerMessMessages;
  }
});
Object.defineProperty(exports, "nodeViewsMessages", {
  enumerable: true,
  get: function get() {
    return _nodeviews.nodeViewsMessages;
  }
});
Object.defineProperty(exports, "pixelEntryMessages", {
  enumerable: true,
  get: function get() {
    return _pixelEntry.pixelEntryMessages;
  }
});
Object.defineProperty(exports, "toolbarMediaMessages", {
  enumerable: true,
  get: function get() {
    return _toolbarMedia.toolbarMediaMessages;
  }
});
var _mediaLinkingToolbar = require("./messages/mediaLinkingToolbar");
var _imageBorder = require("./messages/imageBorder");
var _nodeviews = require("./messages/nodeviews");
var _pixelEntry = require("./messages/pixelEntry");
var _toolbarMedia = require("./messages/toolbarMedia");
var _caption = require("./messages/caption");
var _dropPlaceholder = require("./messages/dropPlaceholder");
var _altText = require("./messages/altText");
var _mediaResizeAnnouncerMess = require("./messages/mediaResizeAnnouncerMess");
var _comments = require("./messages/comments");
var _externalMedia = require("./messages/externalMedia");
