// @atlaskit/editor-plugin-placeholder-text/dist/es2019/ui/FloatingToolbar/index.js
/* eslint-disable @atlaskit/design-system/prefer-primitives */
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
import React from 'react';

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import { jsx } from '@emotion/react';
import { Popup } from '@atlaskit/editor-common/ui';
import { container } from './styles';
export default function FloatingToolbar({
  children,
  target,
  offset,
  fitWidth,
  fitHeight = 40,
  onPositionCalculated,
  popupsMountPoint,
  popupsBoundariesElement,
  className,
  absoluteOffset,
  alignX,
  alignY,
  zIndex
}) {
  if (!target) {
    return null;
  }
  return jsx(Popup, {
    absoluteOffset: absoluteOffset,
    alignX: alignX,
    alignY: alignY,
    target: target,
    zIndex: zIndex,
    mountTo: popupsMountPoint,
    boundariesElement: popupsBoundariesElement,
    offset: offset,
    fitWidth: fitWidth,
    fitHeight: fitHeight,
    onPositionCalculated: onPositionCalculated
  }, jsx("div", {
    // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
    css: container(fitHeight),
    "data-testid": "popup-container"
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    ,
    className: className
  }, children));
}
