// @atlaskit/smart-card/dist/cjs/classNames.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.blockCardForbiddenViewLinkClassName = exports.blockCardForbiddenViewClassName = exports.blockCardErroredViewClassName = exports.blockCardContentHeaderNameClassName = exports.blockCardContentHeaderClassName = exports.blockCardContentClassName = void 0;
Object.defineProperty(exports, "blockCardIconImageClassName", {
  enumerable: true,
  get: function get() {
    return _Icon.blockCardIconImageClassName;
  }
});
exports.metadataListClassName = exports.contentFooterClassName = exports.blockCardResolvingViewClassName = exports.blockCardResolvedViewClassName = exports.blockCardResolvedViewByClassName = exports.blockCardNotFoundViewClassName = void 0;
var _Icon = require("./view/common/Icon");
var contentFooterClassName = exports.contentFooterClassName = 'smart-link-content-footer';
var metadataListClassName = exports.metadataListClassName = 'smart-link-metadata-list';
var blockCardResolvingViewClassName = exports.blockCardResolvingViewClassName = 'block-card-resolving-view';
var blockCardResolvedViewClassName = exports.blockCardResolvedViewClassName = 'block-card-resolved-view';
var blockCardResolvedViewByClassName = exports.blockCardResolvedViewByClassName = 'block-card-resolved-view-by';
var blockCardForbiddenViewClassName = exports.blockCardForbiddenViewClassName = 'block-card-forbidden-view';
var blockCardForbiddenViewLinkClassName = exports.blockCardForbiddenViewLinkClassName = 'block-card-forbidden-view-link';
var blockCardContentClassName = exports.blockCardContentClassName = 'block-card-content';
var blockCardContentHeaderClassName = exports.blockCardContentHeaderClassName = 'block-card-content-header';
var blockCardContentHeaderNameClassName = exports.blockCardContentHeaderNameClassName = 'block-card-content-header-name';
var blockCardNotFoundViewClassName = exports.blockCardNotFoundViewClassName = 'block-card-not-found-view';
var blockCardErroredViewClassName = exports.blockCardErroredViewClassName = 'block-card-errored-view';
