// @atlaskit/smart-card/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ActionName", {
  enumerable: true,
  get: function get() {
    return _constants.ActionName;
  }
});
Object.defineProperty(exports, "AssignedToElement", {
  enumerable: true,
  get: function get() {
    return _external.AssignedToElement;
  }
});
Object.defineProperty(exports, "AssignedToGroupElement", {
  enumerable: true,
  get: function get() {
    return _external.AssignedToGroupElement;
  }
});
Object.defineProperty(exports, "AttachmentCountElement", {
  enumerable: true,
  get: function get() {
    return _external.AttachmentCountElement;
  }
});
Object.defineProperty(exports, "AuthorGroupElement", {
  enumerable: true,
  get: function get() {
    return _external.AuthorGroupElement;
  }
});
Object.defineProperty(exports, "Card", {
  enumerable: true,
  get: function get() {
    return _Card.Card;
  }
});
Object.defineProperty(exports, "CardAction", {
  enumerable: true,
  get: function get() {
    return _types.CardAction;
  }
});
Object.defineProperty(exports, "ChecklistProgressElement", {
  enumerable: true,
  get: function get() {
    return _external.ChecklistProgressElement;
  }
});
Object.defineProperty(exports, "CollaboratorGroupElement", {
  enumerable: true,
  get: function get() {
    return _external.CollaboratorGroupElement;
  }
});
Object.defineProperty(exports, "CommentCountElement", {
  enumerable: true,
  get: function get() {
    return _external.CommentCountElement;
  }
});
Object.defineProperty(exports, "CopyLinkAction", {
  enumerable: true,
  get: function get() {
    return _external.CopyLinkAction;
  }
});
Object.defineProperty(exports, "CreatedByElement", {
  enumerable: true,
  get: function get() {
    return _external.CreatedByElement;
  }
});
Object.defineProperty(exports, "CreatedOnElement", {
  enumerable: true,
  get: function get() {
    return _external.CreatedOnElement;
  }
});
Object.defineProperty(exports, "CustomAction", {
  enumerable: true,
  get: function get() {
    return _external.CustomAction;
  }
});
Object.defineProperty(exports, "CustomBlock", {
  enumerable: true,
  get: function get() {
    return _blocks.CustomBlock;
  }
});
Object.defineProperty(exports, "DownloadAction", {
  enumerable: true,
  get: function get() {
    return _external.DownloadAction;
  }
});
Object.defineProperty(exports, "DueOnElement", {
  enumerable: true,
  get: function get() {
    return _external.DueOnElement;
  }
});
Object.defineProperty(exports, "EditorCardProvider", {
  enumerable: true,
  get: function get() {
    return _linkProvider.EditorCardProvider;
  }
});
Object.defineProperty(exports, "ElementName", {
  enumerable: true,
  get: function get() {
    return _constants.ElementName;
  }
});
Object.defineProperty(exports, "EmbedResizeMessageListener", {
  enumerable: true,
  get: function get() {
    return _EmbedResizeMessageListener.EmbedResizeMessageListener;
  }
});
Object.defineProperty(exports, "ExpandedFrame", {
  enumerable: true,
  get: function get() {
    return _ExpandedFrame.ExpandedFrame;
  }
});
Object.defineProperty(exports, "FollowAction", {
  enumerable: true,
  get: function get() {
    return _external.FollowAction;
  }
});
Object.defineProperty(exports, "FooterBlock", {
  enumerable: true,
  get: function get() {
    return _blocks.FooterBlock;
  }
});
Object.defineProperty(exports, "LatestCommitElement", {
  enumerable: true,
  get: function get() {
    return _external.LatestCommitElement;
  }
});
Object.defineProperty(exports, "LinkIconElement", {
  enumerable: true,
  get: function get() {
    return _external.LinkIconElement;
  }
});
Object.defineProperty(exports, "LocationElement", {
  enumerable: true,
  get: function get() {
    return _external.LocationElement;
  }
});
Object.defineProperty(exports, "MediaPlacement", {
  enumerable: true,
  get: function get() {
    return _constants.MediaPlacement;
  }
});
Object.defineProperty(exports, "MetadataBlock", {
  enumerable: true,
  get: function get() {
    return _blocks.MetadataBlock;
  }
});
Object.defineProperty(exports, "ModifiedByElement", {
  enumerable: true,
  get: function get() {
    return _external.ModifiedByElement;
  }
});
Object.defineProperty(exports, "ModifiedOnElement", {
  enumerable: true,
  get: function get() {
    return _external.ModifiedOnElement;
  }
});
Object.defineProperty(exports, "OwnedByElement", {
  enumerable: true,
  get: function get() {
    return _external.OwnedByElement;
  }
});
Object.defineProperty(exports, "OwnedByGroupElement", {
  enumerable: true,
  get: function get() {
    return _external.OwnedByGroupElement;
  }
});
Object.defineProperty(exports, "PreviewAction", {
  enumerable: true,
  get: function get() {
    return _external.PreviewAction;
  }
});
Object.defineProperty(exports, "PreviewBlock", {
  enumerable: true,
  get: function get() {
    return _blocks.PreviewBlock;
  }
});
Object.defineProperty(exports, "PreviewElement", {
  enumerable: true,
  get: function get() {
    return _external.PreviewElement;
  }
});
Object.defineProperty(exports, "PriorityElement", {
  enumerable: true,
  get: function get() {
    return _external.PriorityElement;
  }
});
Object.defineProperty(exports, "ProgrammingLanguageElement", {
  enumerable: true,
  get: function get() {
    return _external.ProgrammingLanguageElement;
  }
});
Object.defineProperty(exports, "ProviderElement", {
  enumerable: true,
  get: function get() {
    return _external.ProviderElement;
  }
});
Object.defineProperty(exports, "ReactCountElement", {
  enumerable: true,
  get: function get() {
    return _external.ReactCountElement;
  }
});
Object.defineProperty(exports, "ReadTimeElement", {
  enumerable: true,
  get: function get() {
    return _external.ReadTimeElement;
  }
});
Object.defineProperty(exports, "SentOnElement", {
  enumerable: true,
  get: function get() {
    return _external.SentOnElement;
  }
});
Object.defineProperty(exports, "SmartLinkDirection", {
  enumerable: true,
  get: function get() {
    return _constants.SmartLinkDirection;
  }
});
Object.defineProperty(exports, "SmartLinkEvents", {
  enumerable: true,
  get: function get() {
    return _analytics.SmartLinkEvents;
  }
});
Object.defineProperty(exports, "SmartLinkPosition", {
  enumerable: true,
  get: function get() {
    return _constants.SmartLinkPosition;
  }
});
Object.defineProperty(exports, "SmartLinkSize", {
  enumerable: true,
  get: function get() {
    return _constants.SmartLinkSize;
  }
});
Object.defineProperty(exports, "SmartLinkTheme", {
  enumerable: true,
  get: function get() {
    return _constants.SmartLinkTheme;
  }
});
Object.defineProperty(exports, "SnippetBlock", {
  enumerable: true,
  get: function get() {
    return _blocks.SnippetBlock;
  }
});
Object.defineProperty(exports, "SnippetElement", {
  enumerable: true,
  get: function get() {
    return _external.SnippetElement;
  }
});
Object.defineProperty(exports, "SourceBranchElement", {
  enumerable: true,
  get: function get() {
    return _external.SourceBranchElement;
  }
});
Object.defineProperty(exports, "StateElement", {
  enumerable: true,
  get: function get() {
    return _external.StateElement;
  }
});
Object.defineProperty(exports, "StoryPointsElement", {
  enumerable: true,
  get: function get() {
    return _external.StoryPointsElement;
  }
});
Object.defineProperty(exports, "SubTasksProgressElement", {
  enumerable: true,
  get: function get() {
    return _external.SubTasksProgressElement;
  }
});
Object.defineProperty(exports, "SubscriberCountElement", {
  enumerable: true,
  get: function get() {
    return _external.SubscriberCountElement;
  }
});
Object.defineProperty(exports, "TargetBranchElement", {
  enumerable: true,
  get: function get() {
    return _external.TargetBranchElement;
  }
});
Object.defineProperty(exports, "TitleBlock", {
  enumerable: true,
  get: function get() {
    return _blocks.TitleBlock;
  }
});
Object.defineProperty(exports, "TitleElement", {
  enumerable: true,
  get: function get() {
    return _external.TitleElement;
  }
});
Object.defineProperty(exports, "ViewCountElement", {
  enumerable: true,
  get: function get() {
    return _external.ViewCountElement;
  }
});
Object.defineProperty(exports, "VoteCountElement", {
  enumerable: true,
  get: function get() {
    return _external.VoteCountElement;
  }
});
Object.defineProperty(exports, "blockCardContentClassName", {
  enumerable: true,
  get: function get() {
    return _classNames.blockCardContentClassName;
  }
});
Object.defineProperty(exports, "blockCardContentHeaderClassName", {
  enumerable: true,
  get: function get() {
    return _classNames.blockCardContentHeaderClassName;
  }
});
Object.defineProperty(exports, "blockCardContentHeaderNameClassName", {
  enumerable: true,
  get: function get() {
    return _classNames.blockCardContentHeaderNameClassName;
  }
});
Object.defineProperty(exports, "blockCardErroredViewClassName", {
  enumerable: true,
  get: function get() {
    return _classNames.blockCardErroredViewClassName;
  }
});
Object.defineProperty(exports, "blockCardForbiddenViewClassName", {
  enumerable: true,
  get: function get() {
    return _classNames.blockCardForbiddenViewClassName;
  }
});
Object.defineProperty(exports, "blockCardForbiddenViewLinkClassName", {
  enumerable: true,
  get: function get() {
    return _classNames.blockCardForbiddenViewLinkClassName;
  }
});
Object.defineProperty(exports, "blockCardIconImageClassName", {
  enumerable: true,
  get: function get() {
    return _classNames.blockCardIconImageClassName;
  }
});
Object.defineProperty(exports, "blockCardNotFoundViewClassName", {
  enumerable: true,
  get: function get() {
    return _classNames.blockCardNotFoundViewClassName;
  }
});
Object.defineProperty(exports, "blockCardResolvedViewByClassName", {
  enumerable: true,
  get: function get() {
    return _classNames.blockCardResolvedViewByClassName;
  }
});
Object.defineProperty(exports, "blockCardResolvedViewClassName", {
  enumerable: true,
  get: function get() {
    return _classNames.blockCardResolvedViewClassName;
  }
});
Object.defineProperty(exports, "blockCardResolvingViewClassName", {
  enumerable: true,
  get: function get() {
    return _classNames.blockCardResolvingViewClassName;
  }
});
Object.defineProperty(exports, "contentFooterClassName", {
  enumerable: true,
  get: function get() {
    return _classNames.contentFooterClassName;
  }
});
Object.defineProperty(exports, "editorCardProvider", {
  enumerable: true,
  get: function get() {
    return _linkProvider.editorCardProvider;
  }
});
Object.defineProperty(exports, "embedHeaderHeight", {
  enumerable: true,
  get: function get() {
    return _styled.embedHeaderHeight;
  }
});
Object.defineProperty(exports, "loadingPlaceholderClassName", {
  enumerable: true,
  get: function get() {
    return _LazyFallback.loadingPlaceholderClassName;
  }
});
Object.defineProperty(exports, "metadataListClassName", {
  enumerable: true,
  get: function get() {
    return _classNames.metadataListClassName;
  }
});
Object.defineProperty(exports, "useSmartLinkEvents", {
  enumerable: true,
  get: function get() {
    return _useSmartLinkEvents.useSmartLinkEvents;
  }
});
var _Card = require("./view/Card");
var _types = require("./view/Card/types");
var _EmbedResizeMessageListener = require("./view/EmbedCard/EmbedResizeMessageListener");
var _ExpandedFrame = require("./view/EmbedCard/components/ExpandedFrame");
var _styled = require("./view/EmbedCard/components/styled");
var _analytics = require("./utils/analytics/analytics");
var _useSmartLinkEvents = require("./view/SmartLinkEvents/useSmartLinkEvents");
var _classNames = require("./classNames");
var _LazyFallback = require("./view/CardWithUrl/component-lazy/LazyFallback");
var _constants = require("./constants");
var _blocks = require("./view/FlexibleCard/components/blocks");
var _external = require("./view/FlexibleCard/external");
var _linkProvider = require("@atlaskit/link-provider");
