// @atlaskit/link-datasource/dist/cjs/ui/table-footer/powered-by-jsm-assets/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PoweredByJSMAssets = void 0;
require("./index.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _react = _interopRequireDefault(require("react"));
var _logo = require("@atlaskit/logo");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _compiled = require("@atlaskit/primitives/compiled");
var _assets = require("@atlaskit/temp-nav-app-icons/assets");
var _analytics = require("../../../analytics");
var styles = {
  jsmTextStyles: "_syaz1gjq _18u012x7 _30l316c2",
  jsmContainerStyles: "_1e0c1txw _2lx21sbv"
};
var PoweredByJSMAssets = exports.PoweredByJSMAssets = function PoweredByJSMAssets(props) {
  var ASSETS_LINK = '/jira/servicedesk/assets';
  var ASSETS_APP_LINK = '/jira/assets';
  var _useDatasourceAnalyti = (0, _analytics.useDatasourceAnalyticsEvents)(),
    fireEvent = _useDatasourceAnalyti.fireEvent;
  return /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    xcss: styles.jsmContainerStyles,
    padding: "space.150"
  }, /*#__PURE__*/_react.default.createElement("a", {
    "data-testid": 'powered-by-jsm-assets-link',
    onClick: function onClick() {
      return fireEvent('ui.link.clicked.poweredBy', {
        componentHierarchy: 'datasourceTable',
        extensionKey: 'jsm-cmdb-gateway'
      });
    },
    href: (0, _platformFeatureFlags.fg)('assets_as_an_app_v2') ? ASSETS_APP_LINK : ASSETS_LINK,
    rel: "noreferrer",
    target: "_blank",
    style: {
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
      display: 'inline-flex',
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
      alignItems: 'center'
    }
  }, (0, _platformFeatureFlags.fg)('assets_as_an_app_v2') ? /*#__PURE__*/_react.default.createElement(_assets.AssetsIcon, {
    size: "24"
  }) : /*#__PURE__*/_react.default.createElement(_logo.JiraServiceManagementIcon, {
    size: "xsmall",
    appearance: "brand",
    label: props.text
  }), /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    xcss: styles.jsmTextStyles
    // eslint-disable-next-line @atlaskit/design-system/use-tokens-typography, @atlaskit/ui-styling-standard/enforce-style-prop
    ,
    style: {
      fontSize: 'small'
    }
  }, props.text)));
};
