// @atlaskit/editor-plugin-block-controls/dist/cjs/ui/block-menu-items.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.menuItemsCallback = exports.getBlockMenuItems = void 0;
var _react = require("@emotion/react");
var _keymaps = require("@atlaskit/editor-common/keymaps");
var _messages = require("@atlaskit/editor-common/messages");
var _shortcut = require("@atlaskit/editor-shared-styles/shortcut");
var _arrowDown = _interopRequireDefault(require("@atlaskit/icon/core/arrow-down"));
var _arrowLeft = _interopRequireDefault(require("@atlaskit/icon/core/arrow-left"));
var _arrowRight = _interopRequireDefault(require("@atlaskit/icon/core/arrow-right"));
var _arrowUp = _interopRequireDefault(require("@atlaskit/icon/core/arrow-up"));
var _moveNode = require("../editor-commands/move-node");
var _consts = require("../pm-plugins/utils/consts");
/* eslint-disable @atlaskit/design-system/prefer-primitives */
/**
 * @jsxRuntime classic
 * @jsx jsx
 */

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var getBlockMenuItems = exports.getBlockMenuItems = function getBlockMenuItems(formatMessage) {
  return [{
    items: [{
      content: formatMessage(_messages.blockControlsMessages.moveUp),
      elemBefore: (0, _react.jsx)(_arrowUp.default, {
        label: ""
      }),
      value: {
        name: 'moveUp'
      },
      key: 'move_up',
      // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values
      elemAfter: (0, _react.jsx)("div", {
        css: _shortcut.shortcutStyle
      }, (0, _keymaps.tooltip)(_keymaps.dragToMoveUp))
    }, {
      content: formatMessage(_messages.blockControlsMessages.moveDown),
      elemBefore: (0, _react.jsx)(_arrowDown.default, {
        label: ""
      }),
      value: {
        name: 'moveDown'
      },
      key: 'move_down',
      // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values
      elemAfter: (0, _react.jsx)("div", {
        css: _shortcut.shortcutStyle
      }, (0, _keymaps.tooltip)(_keymaps.dragToMoveDown))
    }, {
      content: formatMessage(_messages.blockControlsMessages.moveLeft),
      elemBefore: (0, _react.jsx)(_arrowLeft.default, {
        label: ""
      }),
      value: {
        name: 'moveLeft'
      },
      key: 'move_left',
      // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values
      elemAfter: (0, _react.jsx)("div", {
        css: _shortcut.shortcutStyle
      }, (0, _keymaps.tooltip)(_keymaps.dragToMoveLeft))
    }, {
      content: formatMessage(_messages.blockControlsMessages.moveRight),
      elemBefore: (0, _react.jsx)(_arrowRight.default, {
        label: ""
      }),
      value: {
        name: 'moveRight'
      },
      key: 'move_right',
      // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values
      elemAfter: (0, _react.jsx)("div", {
        css: _shortcut.shortcutStyle
      }, (0, _keymaps.tooltip)(_keymaps.dragToMoveRight))
    }]
  }];
};
var menuItemsCallback = exports.menuItemsCallback = {
  moveUp: function moveUp(api, formatMessage) {
    return (0, _moveNode.moveNodeViaShortcut)(api, _consts.DIRECTION.UP, formatMessage);
  },
  moveDown: function moveDown(api, formatMessage) {
    return (0, _moveNode.moveNodeViaShortcut)(api, _consts.DIRECTION.DOWN, formatMessage);
  }
};
