// @atlaskit/media-svg/dist/cjs/media-svg/loading.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Loading = void 0;
var _react = require("@emotion/react");
var _spinner = _interopRequireDefault(require("@atlaskit/spinner"));
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var loadingStyles = (0, _react.css)({
  display: 'inline-flex',
  alignItems: 'center',
  justifyContent: 'center',
  overflow: 'hidden'
});
var Loading = exports.Loading = function Loading(_ref) {
  var _ref$dimensions = _ref.dimensions,
    _ref$dimensions2 = _ref$dimensions === void 0 ? {} : _ref$dimensions,
    width = _ref$dimensions2.width,
    height = _ref$dimensions2.height;
  return (0, _react.jsx)("span", {
    "data-testid": 'media-svg-loading',
    role: "status",
    css: loadingStyles
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
    ,
    style: {
      width: width,
      height: height
    }
  }, (0, _react.jsx)(_spinner.default, null));
};
