// @atlaskit/media-svg/dist/cjs/media-svg/media-svg.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _react2 = require("@emotion/react");
var _reactErrorBoundary = require("react-error-boundary");
var _errors = require("./errors");
var _loading = require("./loading");
var _svgRenderer = require("./svgRenderer");
var _useResolveSvg2 = require("./useResolveSvg");
/**
 * @jsxRuntime classic
 * @jsx jsx
 */

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var MediaSvg = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var testId = _ref.testId,
    identifier = _ref.identifier,
    dimensions = _ref.dimensions,
    onError = _ref.onError,
    alt = _ref.alt,
    onLoad = _ref.onLoad,
    onMouseDown = _ref.onMouseDown,
    style = _ref.style;
  var _useResolveSvg = (0, _useResolveSvg2.useResolveSvg)(identifier, onError),
    svgUrl = _useResolveSvg.svgUrl,
    source = _useResolveSvg.source;
  return (0, _react2.jsx)(_reactErrorBoundary.ErrorBoundary, {
    onError: (0, _errors.createUnexpectedErrorCallback)(onError),
    fallback: null
  }, svgUrl && source ? (0, _react2.jsx)(_svgRenderer.SvgRenderer, {
    identifier: identifier,
    testId: testId,
    svgUrl: svgUrl,
    source: source,
    dimensions: dimensions,
    onError: onError,
    alt: alt,
    onLoad: onLoad,
    onMouseDown: onMouseDown
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
    ,
    style: style,
    ref: ref
  }) : (0, _react2.jsx)(_loading.Loading, {
    dimensions: dimensions
  }));
});
var _default = exports.default = MediaSvg;
