// @atlaskit/media-svg/dist/cjs/media-svg/media-svg-compiled.js
/* media-svg-compiled.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _reactErrorBoundary = require("react-error-boundary");
var _errors = require("./errors");
var _loadingCompiled = require("./loading-compiled");
var _svgRendererCompiled = require("./svgRenderer-compiled");
var _useResolveSvg2 = require("./useResolveSvg");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var MediaSvg = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var testId = _ref.testId,
    identifier = _ref.identifier,
    dimensions = _ref.dimensions,
    onError = _ref.onError,
    alt = _ref.alt,
    onLoad = _ref.onLoad,
    onMouseDown = _ref.onMouseDown,
    style = _ref.style;
  var _useResolveSvg = (0, _useResolveSvg2.useResolveSvg)(identifier, onError),
    svgUrl = _useResolveSvg.svgUrl,
    source = _useResolveSvg.source;
  return /*#__PURE__*/React.createElement(_reactErrorBoundary.ErrorBoundary, {
    onError: (0, _errors.createUnexpectedErrorCallback)(onError),
    fallback: null
  }, svgUrl && source ? /*#__PURE__*/React.createElement(_svgRendererCompiled.SvgRenderer, {
    identifier: identifier,
    testId: testId,
    svgUrl: svgUrl,
    source: source,
    dimensions: dimensions,
    onError: onError,
    alt: alt,
    onLoad: onLoad,
    onMouseDown: onMouseDown
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
    ,
    style: style,
    ref: ref
  }) : /*#__PURE__*/React.createElement(_loadingCompiled.Loading, {
    dimensions: dimensions
  }));
});
var _default = exports.default = MediaSvg;
