// @atlaskit/atlassian-navigation/dist/es2019/skeleton-help-button.js
import React from 'react';
import QuestionCircleIcon from '@atlaskit/icon/core/migration/question-circle';
import Nav4QuestionCircleIcon from '@atlaskit/icon/core/question-circle';
import { SkeletonIconButton } from './components/SkeletonIconButton';
import { useTheme } from './theme';
/**
 * __Skeleton notification button__
 *
 * Skeleton buttons are lightweight HTML button elements with CSS that represent
 * their heavier interactive counterparts, for use when elements of the
 * navigation are loaded dynamically. This one represents the Help button.
 *
 * - [Examples](https://atlassian.design/components/atlassian-navigation/examples#skeleton-button)
 * - [Code](https://atlassian.design/components/atlassian-navigation/code)
 */
export const SkeletonHelpButton = ({
  label = ''
}) => {
  const {
    mode: {
      navigation
    }
  } = useTheme();
  return /*#__PURE__*/React.createElement(SkeletonIconButton, null, /*#__PURE__*/React.createElement(QuestionCircleIcon, {
    label: label,
    color: "currentColor",
    LEGACY_secondaryColor: navigation.backgroundColor
  }));
};

/**
 * __Nav 4 skeleton help button__
 *
 * A nav 4 skeleton help button {description}.
 */
export const Nav4SkeletonHelpButton = ({
  label = ''
}) => /*#__PURE__*/React.createElement(SkeletonIconButton, null, /*#__PURE__*/React.createElement(Nav4QuestionCircleIcon, {
  label: label,
  spacing: "spacious",
  color: "var(--ds-icon, #44546F)"
}));
