// @atlaskit/smart-card/dist/cjs/view/FlexibleCard/components/elements/lozenge/lozenge-action/lozenge-action-error/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./index.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _reactIntlNext = require("react-intl-next");
var _dropdownMenu = require("@atlaskit/dropdown-menu");
var _error = _interopRequireDefault(require("@atlaskit/icon/core/migration/error"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _colors = require("@atlaskit/theme/colors");
var _useAnalyticsEvents2 = require("../../../../../../../common/analytics/generated/use-analytics-events");
var _messages = require("../../../../../../../messages");
var _useInvokeClientAction = _interopRequireDefault(require("../../../../../../../state/hooks/use-invoke-client-action"));
var _utils = require("../../../../utils");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var MAX_LINE_NUMBER = 8;
var contentStyles = null;
var linkStyles = null;
var textStylesBase = null;
var dropdownItemGroupStyles = null;
var LozengeActionError = function LozengeActionError(_ref) {
  var errorMessage = _ref.errorMessage,
    testId = _ref.testId,
    _ref$maxLineNumber = _ref.maxLineNumber,
    maxLineNumber = _ref$maxLineNumber === void 0 ? MAX_LINE_NUMBER : _ref$maxLineNumber,
    invokePreviewAction = _ref.invokePreviewAction;
  var _useAnalyticsEvents = (0, _useAnalyticsEvents2.useAnalyticsEvents)(),
    fireEvent = _useAnalyticsEvents.fireEvent;
  var invoke = (0, _useInvokeClientAction.default)({});
  var isPreviewAvailable = invokePreviewAction !== undefined;
  var handlePreviewOpen = (0, _react.useCallback)(function () {
    if (isPreviewAvailable) {
      fireEvent('ui.button.clicked.smartLinkStatusOpenPreview', {});
      if (invokePreviewAction) {
        invoke(invokePreviewAction);
      }
    }
  }, [isPreviewAvailable, invoke, invokePreviewAction, fireEvent]);
  var dynamicCss = null;
  var content = (0, _react.useMemo)(function () {
    return /*#__PURE__*/React.createElement(_react.Fragment, null, /*#__PURE__*/React.createElement("div", {
      className: (0, _runtime.ax)(["_zulpu2gc _11c81ixg _1reo15vq _18m915vq _1e0c1txw _1ul9idpf _2lx2vrvc _19pkv77o _4cvr1y6m _pqme2smr _102e2smr _ngc01o8f _1t4p1o8f"]),
      style: {
        "--_ggxzyp": (0, _runtime.ix)("var(--ds-text, ".concat(_colors.N800, ")"))
      }
    }, /*#__PURE__*/React.createElement(_error.default, {
      testId: "".concat(testId, "-icon"),
      LEGACY_size: "medium",
      color: "var(--ds-icon-danger, ".concat(_colors.R500, ")"),
      LEGACY_secondaryColor: "var(--ds-background-danger, ".concat(_colors.R50, ")"),
      label: 'error',
      spacing: "spacious"
    }), /*#__PURE__*/React.createElement("span", {
      "data-testid": "".concat(testId, "-error-message"),
      className: (0, _runtime.ax)(["_11c81ixg _1reo15vq _18m915vq _o5724jg8 _1e0ccj1k _1bto1l2s _1nmz1hna _sudp1e54", "_1yyj1n6p _102k1hw0"]),
      style: {
        "--_t4prqe": (0, _runtime.ix)(maxLineNumber),
        "--_hm3wc5": (0, _runtime.ix)("calc(".concat(maxLineNumber, " * 1rem)"))
      }
    }, typeof errorMessage === 'string' ? errorMessage : (0, _utils.getFormattedMessage)(errorMessage))), isPreviewAvailable ? /*#__PURE__*/React.createElement("div", {
      className: (0, _runtime.ax)(["_zulpu2gc _1reo15vq _18m915vq _11c82smr _1e0c1txw _1ul9idpf _2lx2vrvc _4cvr1h6o _80omtlke _19pku2gc _18u0xy5q _otyrv77o"])
    }, /*#__PURE__*/React.createElement("a", {
      target: "_blank",
      "data-testid": "".concat(testId, "-open-embed"),
      onClick: handlePreviewOpen
    }, (0, _platformFeatureFlags.fg)('confluence-issue-terminology-refresh') ? /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, _messages.messages.open_issue_in_jiraIssueTermRefresh) : /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, _messages.messages.open_issue_in_jira))) : null);
  }, [errorMessage, handlePreviewOpen, isPreviewAvailable, testId, dynamicCss]);
  return /*#__PURE__*/React.createElement("span", {
    "data-testid": "".concat(testId, "-error-item-group"),
    className: (0, _runtime.ax)(["_14hp1wdt _6pup1kw7 _1eo073ad"])
  }, /*#__PURE__*/React.createElement(_dropdownMenu.DropdownItemGroup, null, /*#__PURE__*/React.createElement(_dropdownMenu.DropdownItem, {
    testId: "".concat(testId, "-error")
  }, content)));
};
var _default = exports.default = LozengeActionError;
