// @atlaskit/editor-plugin-block-type/dist/cjs/pm-plugins/ui/ToolbarBlockType/styled.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.wrapperSmallStyle = exports.keyboardShortcutSelect = exports.keyboardShortcut = exports.floatingToolbarWrapperStyle = exports.expandIconWrapperStyle = exports.blockTypeMenuItemStyle = void 0;
var _react = require("@emotion/react");
var _styles = require("@atlaskit/editor-common/styles");
var _shortcut = require("@atlaskit/editor-shared-styles/shortcut");
var _colors = require("@atlaskit/theme/colors");
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var blockTypeMenuItemStyle = exports.blockTypeMenuItemStyle = function blockTypeMenuItemStyle(tagName, selected, typographyTheme) {
  // TEMP FIX: See https://product-fabric.atlassian.net/browse/ED-13878
  var selectedStyle = selected ? "".concat(tagName, " { color: ", "var(--ds-text, white)", " !important; }") : '';
  return function () {
    return (0, _react.css)(
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    tagName === 'blockquote' ? _styles.blockquoteSharedStyles : (0, _styles.headingsSharedStyles)(typographyTheme), {
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
      '>': {
        // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
        'h1, h2, h3, h4, h5, h6': {
          marginTop: 0
        },
        // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
        blockquote: {
          paddingTop: 0,
          paddingBottom: 0,
          marginTop: 0
        }
      }
    },
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    selectedStyle);
  };
};

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles, @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
var keyboardShortcut = exports.keyboardShortcut = (0, _react.css)(_shortcut.shortcutStyle, {
  marginLeft: "var(--ds-space-200, 16px)",
  color: "var(--ds-icon, ".concat(_colors.N400, ")")
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var keyboardShortcutSelect = exports.keyboardShortcutSelect = (0, _react.css)({
  color: "var(--ds-icon, ".concat(_colors.N400, ")")
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var wrapperSmallStyle = exports.wrapperSmallStyle = (0, _react.css)({
  marginLeft: "var(--ds-space-050, 4px)",
  minWidth: '40px'
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var expandIconWrapperStyle = exports.expandIconWrapperStyle = (0, _react.css)({
  marginLeft: "var(--ds-space-negative-100, -8px)"
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles
var floatingToolbarWrapperStyle = exports.floatingToolbarWrapperStyle = (0, _react.css)({
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors
  "[data-role='droplistContent']": {
    maxHeight: '90vh'
  }
});
