// @atlaskit/editor-plugin-floating-toolbar/dist/es2019/ui/Input.js
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
import React, { useState } from 'react';

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import { jsx } from '@emotion/react';
import { panelTextInput } from '@atlaskit/editor-common/ui';
export const Input = props => {
  const {
    defaultValue,
    onBlur,
    onSubmit,
    placeholder
  } = props;
  const [text, setText] = useState(defaultValue || '');
  const handleChange = e => {
    setText(e.target.value);
  };
  const handleSubmit = e => {
    e.preventDefault();
    onSubmit && onSubmit(text);
  };
  const handleBlur = e => {
    e.preventDefault();
    onBlur && onBlur(text);
  };
  return jsx("form", {
    onSubmit: handleSubmit
  }, jsx("input", {
    // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
    css: panelTextInput,
    value: text,
    onChange: handleChange,
    placeholder: placeholder,
    onBlur: handleBlur
  }));
};
