// @atlaskit/editor-plugin-connectivity/dist/cjs/pm-plugins/main.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.key = exports.createPlugin = void 0;
var _safePlugin = require("@atlaskit/editor-common/safe-plugin");
var _state = require("@atlaskit/editor-prosemirror/state");
var key = exports.key = new _state.PluginKey('offlineMode');
var createPlugin = exports.createPlugin = function createPlugin() {
  return new _safePlugin.SafePlugin({
    key: key,
    state: {
      init: function init() {
        return {
          browserState: 'online',
          externalState: undefined
        };
      },
      apply: function apply(tr, pluginState) {
        var meta = tr.getMeta(key);
        if (meta) {
          var externalState = meta.externalState,
            browserState = meta.browserState;
          return {
            browserState: browserState !== null && browserState !== void 0 ? browserState : pluginState === null || pluginState === void 0 ? void 0 : pluginState.browserState,
            externalState: externalState === null ? undefined : externalState !== null && externalState !== void 0 ? externalState : pluginState === null || pluginState === void 0 ? void 0 : pluginState.externalState
          };
        }
        return pluginState;
      }
    }
  });
};
