// @atlaskit/jql-editor/dist/es2019/plugins/autocomplete/components/autocomplete-option/styled.js
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import styled from '@emotion/styled';
import { N100, N20, N50, N800 } from '@atlaskit/theme/colors';
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
export const TooltipContent = styled.div({
  fontFamily: "var(--ds-font-family-body, ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)"
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
export const OptionListItem = styled.li({
  cursor: 'pointer',
  padding: `${"var(--ds-space-075, 6px)"} ${"var(--ds-space-100, 8px)"}`,
  fontFamily: "var(--ds-font-family-code, ui-monospace, Menlo, \"Segoe UI Mono\", \"Ubuntu Mono\", monospace)",
  // eslint-disable-next-line @atlaskit/design-system/use-tokens-typography
  lineHeight: '24px'
},
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-dynamic-styles -- Ignored via go/DSP-18766
({
  isSelected
}) => isSelected && {
  background: `var(--ds-background-neutral-subtle-hovered, ${N20})`
},
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-dynamic-styles -- Ignored via go/DSP-18766
({
  isDeprecated
}) => isDeprecated && {
  cursor: 'default',
  color: `var(--ds-text-disabled, ${N50})`
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
export const OptionName = styled.div({
  color: `var(--ds-text, ${N800})`,
  overflow: 'hidden',
  whiteSpace: 'nowrap',
  textOverflow: 'ellipsis',
  // Added so that overflowed option names do not squish the deprecated info icon
  flex: 1
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
export const DeprecatedOptionContainer = styled.div({
  color: `var(--ds-text-disabled, ${N50})`,
  display: 'flex',
  justifyContent: 'space-between',
  opacity: 0.6
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
export const OptionHighlight = styled.span({
  fontWeight: "var(--ds-font-weight-bold, 700)"
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
export const FieldType = styled.div({
  display: 'flex',
  alignItems: 'center',
  marginTop: "var(--ds-space-negative-025, -2px)",
  color: `var(--ds-text-subtlest, ${N100})`
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
export const FieldTypeIcon = styled.span({
  display: 'flex',
  marginRight: "var(--ds-space-050, 4px)"
});
