// @atlaskit/rovo-agent-components/dist/cjs/ui/agent-dropdown-menu/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChatToAgentButton = exports.AgentDropdownMenu = void 0;
require("./index.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _reactIntlNext = require("react-intl-next");
var _new = _interopRequireWildcard(require("@atlaskit/button/new"));
var _dropdownMenu = _interopRequireWildcard(require("@atlaskit/dropdown-menu"));
var _showMoreHorizontalMore = _interopRequireDefault(require("@atlaskit/icon/core/migration/show-more-horizontal--more"));
var _compiled = require("@atlaskit/primitives/compiled");
var _chatIcon = require("../../common/ui/chat-icon");
var _messages = _interopRequireDefault(require("./messages"));
var _excluded = ["triggerRef"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var styles = {
  chatToAgentButtonContainer: "_1bsb1osq",
  chatToAgentButtonWrapper: "_1e0c1txw _1bah1h6o _k48p1wq8",
  chatPillButtonInline: "_18zrv77o",
  chatPillText: "_1nmz1hna _y3gn1e5h _o5721jtm",
  chatPillIconWrapper: "_1ul9gktf _4t3igktf"
};
var ChatToAgentButton = exports.ChatToAgentButton = function ChatToAgentButton(_ref) {
  var onClick = _ref.onClick;
  var _useIntl = (0, _reactIntlNext.useIntl)(),
    formatMessage = _useIntl.formatMessage;
  return /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    xcss: styles.chatToAgentButtonContainer
  }, /*#__PURE__*/_react.default.createElement(_new.default, {
    testId: "view-agent-modal-chat-to-agent-button",
    shouldFitContainer: true,
    onClick: onClick
  }, /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    xcss: styles.chatToAgentButtonWrapper
  }, /*#__PURE__*/_react.default.createElement(_compiled.Inline, {
    space: "space.050",
    xcss: styles.chatPillButtonInline
  }, /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    xcss: styles.chatPillIconWrapper
  }, /*#__PURE__*/_react.default.createElement(_chatIcon.ChatPillIcon, null)), /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    xcss: styles.chatPillText
  }, formatMessage(_messages.default.chatToAgentButton))))));
};
var AgentDropdownMenu = exports.AgentDropdownMenu = function AgentDropdownMenu(_ref2) {
  var isAgentCreatedByUser = _ref2.isAgentCreatedByUser,
    isAutodevTemplateAgent = _ref2.isAutodevTemplateAgent,
    onEditAgent = _ref2.onEditAgent,
    onCopyAgent = _ref2.onCopyAgent,
    onDuplicateAgent = _ref2.onDuplicateAgent,
    onDeleteAgent = _ref2.onDeleteAgent,
    onViewAgentFullProfileClick = _ref2.onViewAgentFullProfileClick,
    isForgeAgent = _ref2.isForgeAgent,
    _ref2$showViewAgentOp = _ref2.showViewAgentOption,
    showViewAgentOption = _ref2$showViewAgentOp === void 0 ? false : _ref2$showViewAgentOp,
    onViewAgentClick = _ref2.onViewAgentClick,
    onDropdownTriggerClick = _ref2.onDropdownTriggerClick,
    spacing = _ref2.spacing,
    appearance = _ref2.appearance,
    dropdownMenuTestId = _ref2.dropdownMenuTestId,
    doesAgentHaveIdentityAccountId = _ref2.doesAgentHaveIdentityAccountId,
    loadAgentPermissions = _ref2.loadAgentPermissions,
    loadPermissionsOnMount = _ref2.loadPermissionsOnMount,
    shouldTriggerStopPropagation = _ref2.shouldTriggerStopPropagation;
  var _useState = (0, _react.useState)(false),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    isLoading = _useState2[0],
    setIsLoading = _useState2[1];
  var _useIntl2 = (0, _reactIntlNext.useIntl)(),
    formatMessage = _useIntl2.formatMessage;
  var _useState3 = (0, _react.useState)(false),
    _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
    hasBeenCopied = _useState4[0],
    setHasBeenCopied = _useState4[1];
  var _useState5 = (0, _react.useState)(false),
    _useState6 = (0, _slicedToArray2.default)(_useState5, 2),
    isOpen = _useState6[0],
    setIsOpen = _useState6[1];
  var _useState7 = (0, _react.useState)(),
    _useState8 = (0, _slicedToArray2.default)(_useState7, 2),
    permissions = _useState8[0],
    setPermissions = _useState8[1];
  (0, _react.useEffect)(function () {
    var fetchData = /*#__PURE__*/function () {
      var _ref3 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee() {
        var canEditDelete, _yield$loadAgentPermi, isEditEnabled, isDeleteEnabled;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              if (loadAgentPermissions) {
                _context.next = 4;
                break;
              }
              canEditDelete = isAgentCreatedByUser && !isForgeAgent;
              setPermissions({
                isEditEnabled: canEditDelete,
                isDeleteEnabled: canEditDelete
              });
              return _context.abrupt("return");
            case 4:
              setIsLoading(true);
              _context.next = 7;
              return loadAgentPermissions();
            case 7:
              _yield$loadAgentPermi = _context.sent;
              isEditEnabled = _yield$loadAgentPermi.isEditEnabled;
              isDeleteEnabled = _yield$loadAgentPermi.isDeleteEnabled;
              setIsLoading(false);
              setPermissions({
                isEditEnabled: isEditEnabled,
                isDeleteEnabled: isDeleteEnabled
              });
            case 12:
            case "end":
              return _context.stop();
          }
        }, _callee);
      }));
      return function fetchData() {
        return _ref3.apply(this, arguments);
      };
    }();

    // Only load once
    if (!permissions && (isOpen || loadPermissionsOnMount)) {
      fetchData();
    }
  }, [isAgentCreatedByUser, isForgeAgent, isOpen, loadAgentPermissions, loadPermissionsOnMount, permissions]);
  (0, _react.useEffect)(function () {
    if (!isOpen) {
      setHasBeenCopied(false);
    }
  }, [isOpen]);
  var renderEditDelete = function renderEditDelete() {
    if (!(permissions !== null && permissions !== void 0 && permissions.isEditEnabled) && !(permissions !== null && permissions !== void 0 && permissions.isDeleteEnabled)) {
      return null;
    }
    return /*#__PURE__*/_react.default.createElement(_dropdownMenu.DropdownItemGroup, {
      hasSeparator: true
    }, permissions.isEditEnabled && /*#__PURE__*/_react.default.createElement(_dropdownMenu.DropdownItem, {
      onClick: onEditAgent
    }, formatMessage(_messages.default.editAgent)), permissions.isDeleteEnabled && /*#__PURE__*/_react.default.createElement(_dropdownMenu.DropdownItem, {
      onClick: onDeleteAgent
    }, formatMessage(_messages.default.deleteAgent)));
  };
  return /*#__PURE__*/_react.default.createElement(_dropdownMenu.default, {
    isLoading: isLoading,
    isOpen: isOpen,
    testId: dropdownMenuTestId,
    trigger: function trigger(_ref4) {
      var triggerRef = _ref4.triggerRef,
        props = (0, _objectWithoutProperties2.default)(_ref4, _excluded);
      return /*#__PURE__*/_react.default.createElement(_new.IconButton, (0, _extends2.default)({}, props, {
        icon: _showMoreHorizontalMore.default,
        label: formatMessage(_messages.default.moreActionsLabel),
        ref: triggerRef,
        spacing: spacing,
        appearance: appearance,
        onClick: function onClick(e, analyticsEvent) {
          var _props$onClick;
          if (onDropdownTriggerClick) {
            onDropdownTriggerClick(e, analyticsEvent);
          }
          if (shouldTriggerStopPropagation) {
            e.stopPropagation();
          }
          (_props$onClick = props.onClick) === null || _props$onClick === void 0 || _props$onClick.call(props, e);
        }
      }));
    },
    onOpenChange: function onOpenChange(args) {
      setIsOpen(args.isOpen);
      if (!args.isOpen) {
        setHasBeenCopied(false);
      }
    }
  }, /*#__PURE__*/_react.default.createElement(_dropdownMenu.DropdownItemGroup, null, showViewAgentOption && /*#__PURE__*/_react.default.createElement(_dropdownMenu.DropdownItem, {
    onClick: onViewAgentClick
  }, formatMessage(_messages.default.viewAgent)), doesAgentHaveIdentityAccountId && onViewAgentFullProfileClick && /*#__PURE__*/_react.default.createElement(_dropdownMenu.DropdownItem, {
    onClick: onViewAgentFullProfileClick
  }, formatMessage(_messages.default.viewAgentFullProfile)), !isForgeAgent && /*#__PURE__*/_react.default.createElement(_dropdownMenu.DropdownItem, {
    onClick: onDuplicateAgent
  }, isAutodevTemplateAgent ? formatMessage(_messages.default.useTemplateButton) : formatMessage(_messages.default.duplicateAgent)), /*#__PURE__*/_react.default.createElement(_dropdownMenu.DropdownItem, {
    onClick: function onClick(e) {
      e.stopPropagation();
      setHasBeenCopied(true);
      onCopyAgent === null || onCopyAgent === void 0 || onCopyAgent(e);
    }
  }, formatMessage(hasBeenCopied ? _messages.default.linkedCopiedToProfile : _messages.default.copyLinkToProfile))), renderEditDelete());
};
