// @atlaskit/outbound-auth-flow-client/dist/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AuthError", {
  enumerable: true,
  get: function get() {
    return _error.AuthError;
  }
});
exports.auth = auth;
var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));
var _error = require("./error");
var _types = require("./types");
function auth(startUrl, windowFeatures) {
  return new Promise(function (resolve, reject) {
    var authWindow = null;
    var authWindowInterval;
    var handleAuthWindowMessage = function handleAuthWindowMessage(event) {
      if (event.source !== authWindow) {
        return;
      }
      var data = event.data;
      if ((0, _typeof2.default)(data) !== 'object') {
        return;
      }
      switch (data.type) {
        case 'outbound-auth:success':
          finish();
          resolve();
          break;
        case 'outbound-auth:failure':
          finish();
          var errorType = data.errorType.toLowerCase();
          if ((0, _types.isOfTypeAuthError)(errorType)) {
            reject(new _error.AuthError(data.message, errorType));
          } else {
            reject(new _error.AuthError(data.message));
          }
          break;
      }
    };
    var handleAuthWindowInterval = function handleAuthWindowInterval() {
      if (authWindow && authWindow.closed) {
        finish();
        reject(new _error.AuthError('The auth window was closed', 'auth_window_closed'));
      }
    };
    var start = function start() {
      window.addEventListener('message', handleAuthWindowMessage);
      authWindow = window.open(startUrl, startUrl, windowFeatures);
      authWindowInterval = window.setInterval(handleAuthWindowInterval, 500);
    };
    var finish = function finish() {
      clearInterval(authWindowInterval);
      window.removeEventListener('message', handleAuthWindowMessage);
      if (authWindow) {
        authWindow.close();
        authWindow = null;
      }
    };
    start();
  });
}
