// @atlaskit/media-card/dist/cjs/card/ui/wrapper/imageContainer-emotion.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ImageContainer = void 0;
var _react = require("@emotion/react");
var _classnames = require("../../classnames");
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var imageContainerStyles = (0, _react.css)({
  display: 'flex',
  position: 'relative',
  maxWidth: '100%',
  width: '100%',
  height: '100%',
  maxHeight: '100%',
  overflow: 'hidden',
  borderRadius: "var(--ds-border-radius, 3px)"
});
var imageContainerCenterStyles = (0, _react.css)({
  flexDirection: 'column',
  justifyContent: 'center',
  alignItems: 'center'
});
var ImageContainer = exports.ImageContainer = function ImageContainer(_ref) {
  var children = _ref.children,
    mediaName = _ref.mediaName,
    status = _ref.status,
    progress = _ref.progress,
    selected = _ref.selected,
    source = _ref.source,
    centerElements = _ref.centerElements;
  return (0, _react.jsx)("div", {
    css: [imageContainerStyles, centerElements && imageContainerCenterStyles],
    "data-testid": _classnames.fileCardImageViewSelector
    /**
     * This wrapper MUST add the classname in order to allow the editor to prevent bubbling up the click event.
     * See the method isInteractiveElement in source platform/packages/editor/renderer/src/ui/Renderer/click-to-edit.ts
     * Also, many other consumer tests rely on this selector.
     */
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    ,
    className: _classnames.fileCardImageViewSelector,
    "data-test-media-name": mediaName,
    "data-test-status": status,
    "data-test-progress": progress,
    "data-test-selected": selected,
    "data-test-source": source
  }, children);
};
