// @atlaskit/atlassian-navigation/dist/cjs/theme/theme-helpers.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.hexToRGBA = exports.getContrastColor = exports.getBoxShadow = exports.generateTextColor = exports.convertHexShorthand = void 0;
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _chromatism = _interopRequireDefault(require("chromatism"));
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var hexToRGBA = exports.hexToRGBA = function hexToRGBA(hex) {
  var opacity = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 1;
  var rgba = _objectSpread(_objectSpread({}, _chromatism.default.convert(hex).rgb), {
    a: opacity
  });
  return "rgba(".concat(Object.values(rgba).join(', '), ")");
};
var getBoxShadow = exports.getBoxShadow = function getBoxShadow(color) {
  return "0 0 0 2px ".concat(color);
};
var hex = '[a-z0-9]';
var shortHandHexColorPattern = new RegExp("#(".concat(hex, ")(").concat(hex, ")(").concat(hex, ")"), 'i');
var repeat = function repeat(str, times) {
  return (0, _toConsumableArray2.default)(Array(times)).map(function () {
    return str;
  }).join('');
};
var isShortHexColor = function isShortHexColor(color) {
  return color && color.length === 4;
};
var completeHexColor = function completeHexColor(colors) {
  var hex = colors.map(function (color) {
    return repeat(color, 2);
  }).join('');
  return "#".concat(hex);
};
var completeTriplet = function completeTriplet(colors) {
  return "#".concat(repeat(colors[1], 6));
};
var isTriplet = function isTriplet(colors) {
  return colors[0] === colors[1] && colors[1] === colors[2];
};
var convertHexShorthand = exports.convertHexShorthand = function convertHexShorthand(color) {
  if (isShortHexColor(color)) {
    // when color = '#ccc', matches is structured as ['#ccc', 'c', 'c', 'c', ...]
    var matches = color.match(shortHandHexColorPattern);
    if (matches) {
      var colors = matches.slice(1, 4);
      if (isTriplet(colors)) {
        return completeTriplet(colors);
      } else {
        return completeHexColor(colors);
      }
    }
  }

  // return the color as is when it's not hex color shorthand
  return color;
};
var generateTextColor = exports.generateTextColor = function generateTextColor(color) {
  var converted = convertHexShorthand(color);
  return _chromatism.default.contrastRatio(converted).hex;
};
var getContrastColor = exports.getContrastColor = function getContrastColor(contrastValue, opacityValue, color) {
  return hexToRGBA(_chromatism.default.contrast(contrastValue, color).hex, opacityValue);
};
