// @atlaskit/smart-card/dist/es2019/view/EmbedCard/views/ErroredView.js
/* ErroredView.tsx generated by @compiled/babel-plugin v0.36.1 */
import "./ErroredView.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { FormattedMessage } from 'react-intl-next';
import Button from '@atlaskit/button';
import ErrorIcon from '@atlaskit/icon/core/migration/error';
import { Box, Inline } from '@atlaskit/primitives/compiled';
import { R300 } from '@atlaskit/theme/colors';
import { messages } from '../../../messages';
import { Frame } from '../components/ErrorFrame';
const styles = {
  boxStyles: "_19bv1b66 _u5f31b66"
};
export const EmbedCardErroredView = ({
  onRetry,
  isSelected = false,
  testId = 'embed-card-errored-view',
  inheritDimensions
}) => /*#__PURE__*/React.createElement(Frame, {
  inheritDimensions: inheritDimensions,
  compact: true,
  isSelected: isSelected,
  testId: testId
}, /*#__PURE__*/React.createElement(ErrorIcon, {
  LEGACY_size: "small",
  color: `var(--ds-icon-danger, ${R300})`,
  label: "error-icon"
}), /*#__PURE__*/React.createElement(Box, {
  xcss: styles.boxStyles
}, /*#__PURE__*/React.createElement(Inline, null, /*#__PURE__*/React.createElement(FormattedMessage, messages.could_not_load_link))), /*#__PURE__*/React.createElement(Button, {
  testId: "err-view-retry",
  appearance: "link",
  spacing: "none",
  onClick: onRetry
}, /*#__PURE__*/React.createElement(FormattedMessage, messages.try_again)));
