// @atlaskit/media-ui/dist/cjs/BlockCard/components/Frame.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
/** @jsx jsx */
var core_1 = require("@emotion/core");
var colors_1 = require("@atlaskit/theme/colors");
var utils_1 = require("../utils");
exports.Frame = function (props) {
    if (props === void 0) { props = {
        isSelected: false,
        isHoverable: false,
    }; }
    return props.compact ? core_1.jsx(exports.CompactFrame, tslib_1.__assign({}, props)) : core_1.jsx(exports.ExpandedFrame, tslib_1.__assign({}, props));
};
var sharedFrameStyles = {
    maxWidth: utils_1.gs(95),
    width: '100%',
    display: 'flex',
    backgroundColor: 'white',
};
exports.ExpandedFrame = function (_a) {
    var children = _a.children, isSelected = _a.isSelected, isHoverable = _a.isHoverable, testId = _a.testId;
    return (core_1.jsx("div", { css: utils_1.mq(tslib_1.__assign(tslib_1.__assign({}, sharedFrameStyles), { '&:hover': isHoverable
                ? {
                    backgroundColor: colors_1.N20A,
                    cursor: 'pointer',
                }
                : undefined, minHeight: [utils_1.gs(21), utils_1.gs(15)], borderRadius: isSelected ? utils_1.br() : utils_1.br(0.5), border: "2px solid " + (isSelected ? colors_1.B200 : 'transparent'), justifyContent: 'space-between', overflow: 'hidden', boxShadow: "0 1px 1px " + colors_1.N50A + ", 0 0 1px 0 " + colors_1.N60A + ";" })), "data-testid": testId }, children));
};
exports.CompactFrame = function (_a) {
    var children = _a.children, isHoverable = _a.isHoverable, isSelected = _a.isSelected, testId = _a.testId, inheritDimensions = _a.inheritDimensions;
    return (core_1.jsx("div", { css: utils_1.mq(tslib_1.__assign(tslib_1.__assign({}, sharedFrameStyles), { '&:hover': isHoverable
                ? {
                    backgroundColor: colors_1.N30A,
                }
                : undefined, borderRadius: isSelected ? utils_1.br() : utils_1.br(0.5), border: isSelected ? "2px solid " + colors_1.B200 : '', justifyContent: 'center', alignItems: 'center', height: inheritDimensions ? '100%' : utils_1.gs(5), backgroundColor: colors_1.N20A, width: ['calc(100% - 16px)', '100%'], padding: ["0px " + utils_1.gs(1), '0'] })), "data-testid": testId }, children));
};
