// @atlaskit/editor-plugin-table/dist/cjs/ui/TableFloatingControls/FloatingControlsWithSelection.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FloatingControlsWithSelection = void 0;
var _react = _interopRequireDefault(require("react"));
var _hooks = require("@atlaskit/editor-common/hooks");
var _ClassicCornerControls = require("./CornerControls/ClassicCornerControls");
var _ClassicControls = require("./RowControls/ClassicControls");
var FloatingControlsWithSelection = exports.FloatingControlsWithSelection = function FloatingControlsWithSelection(_ref) {
  var editorView = _ref.editorView,
    tableRef = _ref.tableRef,
    isInDanger = _ref.isInDanger,
    isResizing = _ref.isResizing,
    isHeaderRowEnabled = _ref.isHeaderRowEnabled,
    isHeaderColumnEnabled = _ref.isHeaderColumnEnabled,
    hoveredRows = _ref.hoveredRows,
    stickyTop = _ref.stickyTop,
    hoverRows = _ref.hoverRows,
    selectRow = _ref.selectRow,
    tableActive = _ref.tableActive,
    api = _ref.api;
  var _useSharedPluginState = (0, _hooks.useSharedPluginState)(api, ['selection']),
    selectionState = _useSharedPluginState.selectionState;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_ClassicCornerControls.CornerControls, {
    editorView: editorView,
    tableRef: tableRef,
    isInDanger: isInDanger,
    isResizing: isResizing,
    isHeaderRowEnabled: isHeaderRowEnabled,
    isHeaderColumnEnabled: isHeaderColumnEnabled,
    hoveredRows: hoveredRows,
    stickyTop: tableActive ? stickyTop : undefined
  }), /*#__PURE__*/_react.default.createElement(_ClassicControls.RowControls, {
    selection: selectionState === null || selectionState === void 0 ? void 0 : selectionState.selection,
    editorView: editorView,
    tableRef: tableRef,
    hoverRows: hoverRows,
    hoveredRows: hoveredRows,
    isInDanger: isInDanger,
    isResizing: isResizing,
    selectRow: selectRow,
    stickyTop: tableActive ? stickyTop : undefined
  }));
};
