// @atlaskit/editor-common/dist/cjs/resizer/ResizerBreakoutModeLabel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ResizerBreakoutModeLabel = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactIntlNext = require("react-intl-next");
var _primitives = require("@atlaskit/primitives");
var _messages = require("../messages");
var fullWidthLabelWrapperStyles = (0, _primitives.xcss)({
  height: "var(--ds-space-400, 32px)",
  display: 'flex',
  backgroundColor: 'elevation.surface.overlay',
  borderRadius: 'border.radius',
  boxShadow: 'elevation.shadow.overlay',
  boxSizing: 'border-box',
  alignItems: 'center'
});
var fullWidthLabelStyles = (0, _primitives.xcss)({
  marginLeft: 'space.100',
  marginRight: 'space.100',
  paddingLeft: 'space.075',
  paddingRight: 'space.075',
  paddingTop: 'space.050',
  paddingBottom: 'space.050'
});
var ResizerBreakoutModeLabel = exports.ResizerBreakoutModeLabel = function ResizerBreakoutModeLabel(_ref) {
  var breaktoutLayout = _ref.layout;
  var _useIntl = (0, _reactIntlNext.useIntl)(),
    formatMessage = _useIntl.formatMessage;
  var message = _react.default.useMemo(function () {
    switch (breaktoutLayout) {
      case 'full-width':
        return formatMessage(_messages.breakoutMessages.fullWidthLabel);
      case 'wide':
        return formatMessage(_messages.breakoutMessages.wideWidthLabel);
      default:
        return null;
    }
  }, [breaktoutLayout, formatMessage]);
  return message && /*#__PURE__*/_react.default.createElement(_primitives.Box, {
    xcss: fullWidthLabelWrapperStyles
  }, /*#__PURE__*/_react.default.createElement(_primitives.Inline, {
    xcss: fullWidthLabelStyles
  }, message));
};
