// @atlaskit/link-client-extension/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DEFAULT_GET_DATASOURCE_DATA_PAGE_SIZE", {
  enumerable: true,
  get: function get() {
    return _useDataSourceClientExtension.DEFAULT_GET_DATASOURCE_DATA_PAGE_SIZE;
  }
});
Object.defineProperty(exports, "mockActionsDiscoveryEmptyResponse", {
  enumerable: true,
  get: function get() {
    return _mocks.mockActionsDiscoveryEmptyResponse;
  }
});
Object.defineProperty(exports, "mockActionsDiscoveryResponse", {
  enumerable: true,
  get: function get() {
    return _mocks.mockActionsDiscoveryResponse;
  }
});
Object.defineProperty(exports, "mockDatasourceDataNoActionsResponse", {
  enumerable: true,
  get: function get() {
    return _mocks.mockDatasourceDataNoActionsResponse;
  }
});
Object.defineProperty(exports, "mockDatasourceDataResponse", {
  enumerable: true,
  get: function get() {
    return _mocks.mockDatasourceDataResponse;
  }
});
Object.defineProperty(exports, "mockDatasourceDataResponseWithSchema", {
  enumerable: true,
  get: function get() {
    return _mocks.mockDatasourceDataResponseWithSchema;
  }
});
Object.defineProperty(exports, "mockDatasourceDetailsResponse", {
  enumerable: true,
  get: function get() {
    return _mocks.mockDatasourceDetailsResponse;
  }
});
Object.defineProperty(exports, "mockDatasourceResponse", {
  enumerable: true,
  get: function get() {
    return _mocks.mockDatasourceDetailsResponse;
  }
});
Object.defineProperty(exports, "useDatasourceClientExtension", {
  enumerable: true,
  get: function get() {
    return _useDataSourceClientExtension.useDatasourceClientExtension;
  }
});
Object.defineProperty(exports, "useSmartLinkClientExtension", {
  enumerable: true,
  get: function get() {
    return _useSmartLinkClientExtension.useSmartLinkClientExtension;
  }
});
var _useSmartLinkClientExtension = require("./services/use-smart-link-client-extension");
var _useDataSourceClientExtension = require("./services/use-data-source-client-extension");
var _mocks = require("./services/use-data-source-client-extension/mocks");
