// @atlaskit/editor-plugin-undo-redo/dist/cjs/ui/ToolbarUndoRedo/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ToolbarUndoRedo = void 0;
var _react = require("@emotion/react");
var _reactIntlNext = require("react-intl-next");
var _hooks = require("@atlaskit/editor-common/hooks");
var _keymaps = require("@atlaskit/editor-common/keymaps");
var _messages = require("@atlaskit/editor-common/messages");
var _styles = require("@atlaskit/editor-common/styles");
var _uiMenu = require("@atlaskit/editor-common/ui-menu");
var _redo = _interopRequireDefault(require("@atlaskit/icon/core/migration/redo"));
var _undo = _interopRequireDefault(require("@atlaskit/icon/core/migration/undo"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _commands = require("../../pm-plugins/commands");
var _utils = require("../../pm-plugins/utils");
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var ToolbarUndoRedo = exports.ToolbarUndoRedo = function ToolbarUndoRedo(_ref) {
  var disabled = _ref.disabled,
    isReducedSpacing = _ref.isReducedSpacing,
    editorView = _ref.editorView,
    api = _ref.api,
    formatMessage = _ref.intl.formatMessage;
  var _useSharedPluginState = (0, _hooks.useSharedPluginState)(api, ['history']),
    historyState = _useSharedPluginState.historyState;
  var handleUndo = function handleUndo() {
    (0, _utils.forceFocus)(editorView, api)(_commands.undoFromToolbar);
  };
  var handleRedo = function handleRedo() {
    (0, _utils.forceFocus)(editorView, api)(_commands.redoFromToolbar);
  };
  var labelUndo = formatMessage(_messages.undoRedoMessages.undo);
  var labelRedo = formatMessage(_messages.undoRedoMessages.redo);
  var _ref2 = historyState !== null && historyState !== void 0 ? historyState : {},
    canUndo = _ref2.canUndo,
    canRedo = _ref2.canRedo;
  var redoKeymap = (0, _platformFeatureFlags.fg)('platform_editor_cmd_y_mac_redo_shortcut') ? _keymaps.redoAlt : _keymaps.redo;
  return (
    // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
    (0, _react.jsx)("span", {
      css:
      // eslint-disable-next-line @atlaskit/platform/ensure-feature-flag-registration, @atlaskit/platform/ensure-feature-flag-prefix
      (0, _platformFeatureFlags.fg)('platform-visual-refresh-icons') ?
      // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values
      _styles.buttonGroupStyle :
      // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values
      _styles.buttonGroupStyleBeforeVisualRefresh
    }, (0, _react.jsx)(_uiMenu.ToolbarButton, {
      buttonId: _uiMenu.TOOLBAR_BUTTON.UNDO,
      spacing: isReducedSpacing ? 'none' : 'default',
      onClick: handleUndo,
      disabled: !canUndo || disabled,
      "aria-label": (0, _keymaps.tooltip)(_keymaps.undo, labelUndo),
      "aria-keyshortcuts": (0, _keymaps.getAriaKeyshortcuts)(_keymaps.undo),
      title: (0, _react.jsx)(_keymaps.ToolTipContent, {
        description: labelUndo,
        keymap: _keymaps.undo
      }),
      iconBefore: (0, _react.jsx)(_undo.default, {
        label: "",
        color: "currentColor",
        spacing: "spacious"
      }),
      testId: "ak-editor-toolbar-button-undo"
    }), (0, _react.jsx)(_uiMenu.ToolbarButton, {
      spacing: isReducedSpacing ? 'none' : 'default',
      buttonId: _uiMenu.TOOLBAR_BUTTON.REDO,
      onClick: handleRedo,
      disabled: !canRedo || disabled,
      title: (0, _react.jsx)(_keymaps.ToolTipContent, {
        description: labelRedo,
        keymap: redoKeymap
      }),
      iconBefore: (0, _react.jsx)(_redo.default, {
        label: "",
        color: "currentColor",
        spacing: "spacious"
      }),
      testId: "ak-editor-toolbar-button-redo",
      "aria-label": (0, _keymaps.tooltip)(redoKeymap, labelRedo),
      "aria-keyshortcuts": (0, _keymaps.getAriaKeyshortcuts)(redoKeymap)
    }), !(api !== null && api !== void 0 && api.primaryToolbar) && /* eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage */
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/design-system/consistent-css-prop-usage -- Ignored via go/DSP-18766
    (0, _react.jsx)("span", {
      css: _styles.separatorStyles
    }))
  );
};
var _default = exports.default = (0, _reactIntlNext.injectIntl)(ToolbarUndoRedo);
