// @atlaskit/media-ui/dist/cjs/customMediaPlayer/analytics/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "createAndFireMediaCustomMediaPlayerEvent", {
  enumerable: true,
  get: function get() {
    return _analytics.createAndFireMediaCustomMediaPlayerEvent;
  }
});
Object.defineProperty(exports, "createCustomMediaPlayerScreenEvent", {
  enumerable: true,
  get: function get() {
    return _customMediaPlayer.createCustomMediaPlayerScreenEvent;
  }
});
Object.defineProperty(exports, "createFirstPlayedTrackEvent", {
  enumerable: true,
  get: function get() {
    return _playCount.createFirstPlayedTrackEvent;
  }
});
Object.defineProperty(exports, "createMediaButtonClickedEvent", {
  enumerable: true,
  get: function get() {
    return _mediaButtonClicked.createMediaButtonClickedEvent;
  }
});
Object.defineProperty(exports, "createMediaShortcutPressedEvent", {
  enumerable: true,
  get: function get() {
    return _shortcutPressed.createMediaShortcutPressedEvent;
  }
});
Object.defineProperty(exports, "createPlayPauseBlanketClickedEvent", {
  enumerable: true,
  get: function get() {
    return _playPauseBlanketClicked.createPlayPauseBlanketClickedEvent;
  }
});
Object.defineProperty(exports, "createPlaybackSpeedChangedEvent", {
  enumerable: true,
  get: function get() {
    return _playbackSpeedChanged.createPlaybackSpeedChangedEvent;
  }
});
Object.defineProperty(exports, "createPlayedTrackEvent", {
  enumerable: true,
  get: function get() {
    return _played.createPlayedTrackEvent;
  }
});
Object.defineProperty(exports, "createTimeRangeNavigatedEvent", {
  enumerable: true,
  get: function get() {
    return _timeRangeNavigated.createTimeRangeNavigatedEvent;
  }
});
Object.defineProperty(exports, "fireAnalyticsEvent", {
  enumerable: true,
  get: function get() {
    return _analytics.fireAnalyticsEvent;
  }
});
var _customMediaPlayer = require("./events/screen/customMediaPlayer");
var _mediaButtonClicked = require("./events/ui/mediaButtonClicked");
var _playbackSpeedChanged = require("./events/ui/playbackSpeedChanged");
var _playPauseBlanketClicked = require("./events/ui/playPauseBlanketClicked");
var _shortcutPressed = require("./events/ui/shortcutPressed");
var _timeRangeNavigated = require("./events/ui/timeRangeNavigated");
var _playCount = require("./events/track/playCount");
var _played = require("./events/track/played");
var _analytics = require("./utils/analytics");
