// @atlaskit/link-datasource/dist/cjs/ui/issue-like-table/render-type/date-time/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.DATETIME_TYPE_TEST_ID = void 0;
exports.getFormattedDate = getFormattedDate;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireDefault(require("react"));
var _reactIntlNext = require("react-intl-next");
var _compiled = require("@atlaskit/primitives/compiled");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var DATETIME_TYPE_TEST_ID = exports.DATETIME_TYPE_TEST_ID = 'link-datasource-render-type--datetime';
var dateOptions = {
  month: 'short',
  day: 'numeric',
  year: 'numeric'
};
var timeOptions = {
  hour12: false,
  hour: '2-digit',
  minute: '2-digit'
};
function getFormattedDate(value) {
  var display = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'datetime';
  var formatDate = arguments.length > 2 ? arguments[2] : undefined;
  /* In some cases we get a value of `2023-12-20` which when parsed by JS assumes meantime timezone, causing the date
     to be one day off in some timezones. We want it to display the date without converting timezones and a solution
    is to replace the hyphens with slashes. So it should be 20th Dec regardless of the timezone in this case.
     See https://stackoverflow.com/a/31732581
    */
  var dateValue = /^\d{4}-\d{2}-\d{2}$/.exec(value) ? value.replace(/-/g, '/') : value;
  var date = new Date(dateValue);
  if (!value || isNaN(date.getTime())) {
    return '';
  }
  var options = {
    date: dateOptions,
    time: timeOptions,
    datetime: _objectSpread(_objectSpread({}, dateOptions), timeOptions)
  };
  return formatDate(date, options[display] || options['date']);
}
var DateTimeRenderType = function DateTimeRenderType(_ref) {
  var value = _ref.value,
    _ref$testId = _ref.testId,
    testId = _ref$testId === void 0 ? DATETIME_TYPE_TEST_ID : _ref$testId,
    _ref$display = _ref.display,
    display = _ref$display === void 0 ? 'datetime' : _ref$display;
  var intl = (0, _reactIntlNext.useIntl)();
  var formattedString = getFormattedDate(value, display, intl.formatDate);
  if (formattedString === '') {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null);
  }
  return /*#__PURE__*/_react.default.createElement(_compiled.Text, {
    testId: testId
  }, formattedString);
};
var _default = exports.default = DateTimeRenderType;
