// @atlaskit/link-datasource/dist/cjs/ui/issue-like-table/edit-type/text/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _layering = require("@atlaskit/layering");
var _textfield = _interopRequireDefault(require("@atlaskit/textfield"));
var _ufoExperiences = require("../../../../analytics/ufoExperiences");
var _datasourceExperienceId = require("../../../../contexts/datasource-experience-id");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var TextEditType = function TextEditType(props) {
  var _currentValue$values$, _currentValue$values;
  var experienceId = (0, _datasourceExperienceId.useDatasourceExperienceId)();
  var currentValue = props.currentValue,
    labelId = props.labelId;
  (0, _react.useEffect)(function () {
    if (experienceId) {
      (0, _ufoExperiences.succeedUfoExperience)({
        name: 'inline-edit-rendered'
      }, experienceId);
    }
  }, [experienceId]);
  return /*#__PURE__*/_react.default.createElement(_layering.Layering, {
    isDisabled: false
  }, /*#__PURE__*/_react.default.createElement(_textfield.default, (0, _extends2.default)({}, props, {
    autoFocus: true,
    isCompact: true,
    testId: "inline-edit-text",
    style: {
      // We need 8px left padding to match read only version, but there is already 1px of border
      padding: "var(--ds-space-100, 8px)".concat(" calc(", "var(--ds-space-100, 8px)", " - 1px)")
    },
    value: (_currentValue$values$ = currentValue === null || currentValue === void 0 || (_currentValue$values = currentValue.values) === null || _currentValue$values === void 0 ? void 0 : _currentValue$values[0]) !== null && _currentValue$values$ !== void 0 ? _currentValue$values$ : '',
    "aria-labelledby": labelId,
    onChange: function onChange(e) {
      return props.setEditValues({
        type: 'string',
        values: [e.currentTarget.value]
      });
    }
  })));
};
var _default = exports.default = TextEditType;
