// @atlaskit/atlassian-navigation/dist/cjs/skeleton-switcher-button.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SkeletonSwitcherButton = exports.Nav4SkeletonSwitcherButton = void 0;
var _react = require("@emotion/react");
var _appSwitcher = _interopRequireDefault(require("@atlaskit/icon/core/app-switcher"));
var _appSwitcher2 = _interopRequireDefault(require("@atlaskit/icon/core/migration/app-switcher"));
var _SkeletonIconButton = require("./components/SkeletonIconButton");
var _theme = require("./theme");
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var skeletonSwitcherButtonStyles = (0, _react.css)({
  margin: 0,
  // eslint-disable-next-line @atlaskit/design-system/no-physical-properties
  marginRight: "var(--ds-space-050, 4px)",
  padding: "var(--ds-space-050, 4px)".concat(" ", "var(--ds-space-075, 6px)"),
  border: 0,
  borderRadius: "var(--ds-border-radius-circle, 100%)",
  pointerEvents: 'none',
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
  ':focus, :active, :hover': {
    appearance: 'none',
    border: 0,
    outline: 0
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
  '&:only-of-type': {
    // eslint-disable-next-line @atlaskit/design-system/no-physical-properties
    marginRight: 0
  },
  // eslint-disable-next-line @atlaskit/design-system/no-nested-styles, @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '& > span': {
    // eslint-disable-next-line @atlaskit/design-system/use-tokens-typography
    lineHeight: 'normal'
  },
  // eslint-disable-next-line @atlaskit/design-system/no-nested-styles, @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '& > img': {
    width: 24,
    height: 24,
    borderRadius: "var(--ds-border-radius-circle, 100%)",
    verticalAlign: 'middle'
  }
});

/**
 * __Skeleton switcher button__
 *
 * Skeleton buttons are lightweight HTML button elements with CSS that represent
 * their heavier interactive counterparts, for use when elements of the
 * navigation are loaded dynamically. This one represents the AppSwitcher button.
 *
 * - [Examples](https://atlassian.design/components/atlassian-navigation/examples#skeleton-button)
 * - [Code](https://atlassian.design/components/atlassian-navigation/code)
 */
var SkeletonSwitcherButton = exports.SkeletonSwitcherButton = function SkeletonSwitcherButton(_ref) {
  var _ref$label = _ref.label,
    label = _ref$label === void 0 ? '' : _ref$label,
    testId = _ref.testId;
  var theme = (0, _theme.useTheme)();
  return (
    // eslint-disable-next-line @atlaskit/design-system/no-html-button
    (0, _react.jsx)("button", {
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
      style: theme.mode.iconButton.default,
      "data-testid": testId,
      css: skeletonSwitcherButtonStyles,
      type: "button"
    }, (0, _react.jsx)(_appSwitcher2.default, {
      color: "currentColor",
      spacing: "spacious",
      label: label
    }))
  );
};

/**
 * __Nav 4 skeleton switcher button__
 *
 * A nav 4 skeleton switcher button
 */
var Nav4SkeletonSwitcherButton = exports.Nav4SkeletonSwitcherButton = function Nav4SkeletonSwitcherButton(_ref2) {
  var _ref2$label = _ref2.label,
    label = _ref2$label === void 0 ? '' : _ref2$label,
    testId = _ref2.testId;
  return (0, _react.jsx)(_SkeletonIconButton.SkeletonIconButton, null, (0, _react.jsx)(_appSwitcher.default, {
    label: label,
    spacing: "spacious",
    color: "var(--ds-icon, #44546F)"
  }));
};
