// @atlaskit/media-editor/dist/es2019/react/editorView/toolbar/buttons/toolButton.js
import React from 'react';
import { Component } from 'react';
import Button from '@atlaskit/button';
import { N0 } from '@atlaskit/theme/colors';
import ArrowIcon from '@atlaskit/icon/glyph/media-services/arrow';
import BrushIcon from '@atlaskit/icon/glyph/media-services/brush';
import LineIcon from '@atlaskit/icon/glyph/media-services/line';
import BlurIcon from '@atlaskit/icon/glyph/media-services/blur';
import OvalIcon from '@atlaskit/icon/glyph/media-services/oval';
import RectIcon from '@atlaskit/icon/glyph/media-services/rectangle';
import TextIcon from '@atlaskit/icon/glyph/media-services/text';
import { ButtonIconWrapper } from './styles';
export const toolIcons = {
    line: LineIcon,
    blur: BlurIcon,
    arrow: ArrowIcon,
    brush: BrushIcon,
    oval: OvalIcon,
    rectangle: RectIcon,
    text: TextIcon,
};
export class ToolButton extends Component {
    render() {
        const { tool, activeTool, onToolClick } = this.props;
        const Icon = toolIcons[tool];
        const isActive = tool === activeTool;
        const iconPrimaryColor = isActive ? N0 : undefined;
        const onClick = () => {
            onToolClick(tool);
        };
        const iconBefore = (React.createElement(ButtonIconWrapper, null,
            React.createElement(Icon, { label: tool, size: "medium", primaryColor: iconPrimaryColor })));
        return (React.createElement(Button, { iconBefore: iconBefore, appearance: "subtle", onClick: onClick, isSelected: isActive }));
    }
}
