// @atlaskit/editor-core/dist/es2019/ui/EditorContext/index.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import React from 'react';
import PropTypes from 'prop-types';
import { EditorContext } from '@atlaskit/editor-common/UNSAFE_do_not_use_editor_context';
import { fg } from '@atlaskit/platform-feature-flags';
import EditorActions from '../../actions';
export const useEditorContext = () => React.useContext(EditorContext);

// Ignored via go/ees005
// eslint-disable-next-line @repo/internal/react/no-class-components, react/prefer-stateless-function
export class LegacyEditorContext extends React.Component {
  constructor(props) {
    super(props);
  }
  render() {
    if (fg('platform_editor_react18_phase2_v2') || fg('platform_editor_react18_phase2_v2_extended')) {
      // Ignored via go/ees005
      // eslint-disable-next-line react/jsx-props-no-spreading
      return /*#__PURE__*/React.createElement(LegacyEditorContextNew, this.props, this.props.children);
    }

    // Ignored via go/ees005
    // eslint-disable-next-line react/jsx-props-no-spreading
    return /*#__PURE__*/React.createElement(LegacyEditorContextOld, this.props, this.props.children);
  }
}
function LegacyEditorContextNew({
  children,
  editorActions
}) {
  return /*#__PURE__*/React.createElement(EditorContext.Provider, {
    value: {
      editorActions: editorActions !== null && editorActions !== void 0 ? editorActions : new EditorActions()
    }
  }, children);
}

// Ignored via go/ees005
// eslint-disable-next-line @repo/internal/react/no-class-components
export default class LegacyEditorContextOld extends React.Component {
  constructor(props) {
    super(props);
    this.editorActions = props.editorActions || new EditorActions();
  }
  getChildContext() {
    return {
      editorActions: this.editorActions
    };
  }
  render() {
    return /*#__PURE__*/React.createElement(EditorContext.Provider, {
      value: this.getChildContext()
    }, this.props.children);
  }
}
_defineProperty(LegacyEditorContextOld, "childContextTypes", {
  editorActions: PropTypes.object
});
