// @atlaskit/link-picker/dist/cjs/ui/link-picker/form-footer/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.testIds = exports.messages = exports.FormFooter = void 0;
require("./index.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _reactIntlNext = require("react-intl-next");
var _button = _interopRequireWildcard(require("@atlaskit/button"));
var _new = _interopRequireDefault(require("@atlaskit/button/new"));
var _add = _interopRequireDefault(require("@atlaskit/icon/glyph/editor/add"));
var _add2 = _interopRequireDefault(require("@atlaskit/icon/utility/add"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _visuallyHidden = _interopRequireDefault(require("@atlaskit/visually-hidden"));
var _errors = require("../../../common/utils/errors");
var _linkPickerSubmitButton = require("./link-picker-submit-button");
var _excluded = ["isLoading", "isSubmitting", "error", "url", "queryState", "items", "isEditing", "onCancel", "action", "customSubmitButtonLabel", "submitMessageId", "hideSubmitButton", "className"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var formFooterStyles = null;
var formFooterV2Styles = null;
var formFooterActionStyles = null;
var messages = exports.messages = (0, _reactIntlNext.defineMessages)({
  cancelButton: {
    id: 'fabric.linkPicker.button.cancel',
    defaultMessage: 'Cancel',
    description: 'Button to cancel and dismiss the link picker'
  },
  submittingStatusMessage: {
    id: 'fabric.linkPicker.status.submitting',
    defaultMessage: 'Submitting',
    description: 'Accessibility text to indicate the form has been submitted, and submission is in-progress'
  }
});
var testIds = exports.testIds = {
  insertButton: 'link-picker-insert-button',
  cancelButton: 'link-picker-cancel-button',
  actionButton: 'link-picker-action-button',
  submitStatusA11yIndicator: 'link-picker-submit-status-a11y-indicator'
};
var FormFooter = exports.FormFooter = /*#__PURE__*/(0, _react.memo)(function (_ref) {
  var isLoading = _ref.isLoading,
    _ref$isSubmitting = _ref.isSubmitting,
    isSubmitting = _ref$isSubmitting === void 0 ? false : _ref$isSubmitting,
    error = _ref.error,
    url = _ref.url,
    queryState = _ref.queryState,
    items = _ref.items,
    isEditing = _ref.isEditing,
    onCancel = _ref.onCancel,
    action = _ref.action,
    customSubmitButtonLabel = _ref.customSubmitButtonLabel,
    submitMessageId = _ref.submitMessageId,
    hideSubmitButton = _ref.hideSubmitButton,
    className = _ref.className,
    restProps = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  var intl = (0, _reactIntlNext.useIntl)();
  if (error && error instanceof _errors.UnauthenticatedError) {
    return null;
  }
  var createButton = function createButton(pluginAction) {
    return (0, _platformFeatureFlags.fg)('platform-link-picker-remove-legacy-button') ? /*#__PURE__*/React.createElement(_new.default, {
      testId: testIds.actionButton,
      onClick: pluginAction.callback,
      appearance: "default",
      iconBefore: function iconBefore() {
        return /*#__PURE__*/React.createElement(_add2.default, {
          label: "",
          LEGACY_size: "medium",
          LEGACY_fallbackIcon: _add.default,
          color: "currentColor"
        });
      },
      isDisabled: isSubmitting,
      "aria-labelledby": isSubmitting ? submitMessageId : undefined
    }, typeof pluginAction.label === 'string' ? pluginAction.label : intl.formatMessage(pluginAction.label)) : /*#__PURE__*/React.createElement(_button.default, {
      testId: testIds.actionButton,
      onClick: pluginAction.callback,
      appearance: "default",
      iconBefore: /*#__PURE__*/React.createElement(_add2.default, {
        label: "",
        LEGACY_size: "medium",
        LEGACY_fallbackIcon: _add.default,
        color: "currentColor"
      }),
      isDisabled: isSubmitting,
      "aria-labelledby": isSubmitting ? submitMessageId : undefined
    }, typeof pluginAction.label === 'string' ? pluginAction.label : intl.formatMessage(pluginAction.label));
  };
  var ButtonComponent = (0, _platformFeatureFlags.fg)('platform-link-picker-remove-legacy-button') ? _new.default : _button.default;
  return /*#__PURE__*/React.createElement("footer", (0, _extends2.default)({
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop
    className: (0, _runtime.ax)([(0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-link-picker') ? "_1e0c1txw _1bahesu3 _ca0qpxbi" : "_1e0c1txw _1bahesu3", className])
  }, restProps), isSubmitting && /*#__PURE__*/React.createElement(_visuallyHidden.default, {
    role: "status",
    id: submitMessageId,
    testId: testIds.submitStatusA11yIndicator
  }, intl.formatMessage(messages.submittingStatusMessage)), action && /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)(["_2hwx1wug"])
  }, createButton(action)), /*#__PURE__*/React.createElement(_button.ButtonGroup, null, onCancel && /*#__PURE__*/React.createElement(ButtonComponent, {
    appearance: "subtle",
    onClick: onCancel,
    testId: testIds.cancelButton,
    isDisabled: isSubmitting,
    "aria-labelledby": isSubmitting ? submitMessageId : undefined
  }, intl.formatMessage(messages.cancelButton)), !hideSubmitButton && /*#__PURE__*/React.createElement(_linkPickerSubmitButton.LinkPickerSubmitButton, {
    isEditing: isEditing,
    isLoading: isLoading,
    isSubmitting: isSubmitting,
    customSubmitButtonLabel: customSubmitButtonLabel,
    error: error,
    items: items,
    queryState: queryState,
    submitMessageId: submitMessageId,
    testId: testIds.insertButton,
    url: url
  })));
});
