// @atlaskit/editor-plugin-table/dist/cjs/pm-plugins/plugin-factory.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getPluginState = exports.createPluginState = exports.createCommand = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _utils = require("@atlaskit/editor-common/utils");
var _handlers = require("./handlers");
var _pluginKey = require("./plugin-key");
var _reducer = _interopRequireDefault(require("./reducer"));
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _pluginFactory = (0, _utils.pluginFactory)(_pluginKey.pluginKey, _reducer.default, {
    mapping: function mapping(tr, pluginState) {
      if (tr.docChanged) {
        var updatedTargetCell = {};
        if (pluginState.targetCellPosition) {
          var _tr$mapping$mapResult = tr.mapping.mapResult(pluginState.targetCellPosition),
            pos = _tr$mapping$mapResult.pos,
            deleted = _tr$mapping$mapResult.deleted;
          updatedTargetCell = {
            targetCellPosition: deleted ? undefined : pos
          };
        }
        var updatedTablePos = {};
        if (pluginState.tablePos) {
          var _tr$mapping$mapResult2 = tr.mapping.mapResult(pluginState.tablePos, -1),
            _pos = _tr$mapping$mapResult2.pos,
            _deleted = _tr$mapping$mapResult2.deleted;
          updatedTablePos = {
            tablePos: _deleted ? undefined : _pos
          };
        }
        return _objectSpread(_objectSpread(_objectSpread({}, pluginState), updatedTargetCell), updatedTablePos);
      }
      return pluginState;
    },
    onDocChanged: _handlers.handleDocOrSelectionChanged,
    onSelectionChanged: _handlers.handleDocOrSelectionChanged
  }),
  createPluginState = exports.createPluginState = _pluginFactory.createPluginState,
  createCommand = exports.createCommand = _pluginFactory.createCommand,
  getPluginState = exports.getPluginState = _pluginFactory.getPluginState;
