// @atlaskit/editor-plugin-media/dist/cjs/ui/toolbar/layout-group.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LayoutGroup = void 0;
var _react = require("@emotion/react");
var _button = require("@atlaskit/button");
var _ui = require("@atlaskit/editor-common/ui");
var _primitives = require("@atlaskit/primitives");
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var containerStyles = (0, _primitives.xcss)({
  marginLeft: 'space.100'
});
var LayoutGroup = exports.LayoutGroup = function LayoutGroup(_ref) {
  var layoutButtons = _ref.layoutButtons,
    dispatchCommand = _ref.dispatchCommand,
    hide = _ref.hide;
  return (0, _react.jsx)(_primitives.Box, {
    xcss: containerStyles
  }, (0, _react.jsx)(_button.ButtonGroup, null, layoutButtons.map(function (item, idx) {
    switch (item.type) {
      case 'separator':
        // Ignored via go/ees005
        // eslint-disable-next-line react/no-array-index-key
        return (0, _react.jsx)(_ui.FloatingToolbarSeparator, {
          key: idx
        });
      case 'button':
        // Ignored via go/ees005
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        var ButtonIcon = item.icon;
        return (0, _react.jsx)(_ui.FloatingToolbarButton
        // Ignored via go/ees005
        // eslint-disable-next-line react/no-array-index-key
        , {
          key: idx,
          icon: item.icon ? (0, _react.jsx)(ButtonIcon, {
            label: item.title
          }) : undefined,
          title: item.title,
          selected: item.selected,
          disabled: item.disabled,
          onClick: function onClick() {
            dispatchCommand(item.onClick);
            hide();
          }
        });
    }
  })));
};
