// @atlaskit/editor-plugin-media/dist/cjs/ui/toolbar/pixel-resizing.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getResizeDropdownOption = exports.getPixelResizingToolbar = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireDefault(require("react"));
var _media = require("@atlaskit/editor-common/media");
var _utils = require("@atlaskit/editor-prosemirror/utils");
var _growHorizontal = _interopRequireDefault(require("@atlaskit/icon/core/grow-horizontal"));
var _commands = require("../../pm-plugins/pixel-resizing/commands");
var _ui = require("../../pm-plugins/pixel-resizing/ui");
var _constants = require("../../pm-plugins/pixel-resizing/ui/constants");
var _utils2 = require("./utils");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var getPixelResizingToolbar = exports.getPixelResizingToolbar = function getPixelResizingToolbar(toolbarBaseConfig, _ref) {
  var pluginInjectionApi = _ref.pluginInjectionApi,
    intl = _ref.intl,
    pluginState = _ref.pluginState,
    hoverDecoration = _ref.hoverDecoration,
    isEditorFullWidthEnabled = _ref.isEditorFullWidthEnabled;
  return _objectSpread(_objectSpread({}, toolbarBaseConfig), {}, {
    width: _constants.PIXEL_RESIZING_TOOLBAR_WIDTH,
    scrollable: true,
    items: [{
      type: 'custom',
      fallback: [],
      render: function render(editorView) {
        if (!editorView) {
          return null;
        }
        var selectedMediaSingleNode = (0, _utils2.getSelectedMediaSingle)(editorView.state);
        if (!editorView || !selectedMediaSingleNode) {
          return null;
        }
        return /*#__PURE__*/_react.default.createElement(_ui.PixelEntry, {
          editorView: editorView,
          intl: intl,
          selectedMediaSingleNode: selectedMediaSingleNode,
          pluginInjectionApi: pluginInjectionApi,
          pluginState: pluginState,
          hoverDecoration: hoverDecoration,
          isEditorFullWidthEnabled: isEditorFullWidthEnabled
        });
      }
    }]
  });
};
var getResizeDropdownOption = exports.getResizeDropdownOption = function getResizeDropdownOption(mediaOptions, state, formatMessage, selectedNodeType) {
  if ((selectedNodeType === null || selectedNodeType === void 0 ? void 0 : selectedNodeType.name) !== 'mediaSingle') {
    return [];
  }
  var allowResizing = mediaOptions.allowResizing,
    allowResizingInTables = mediaOptions.allowResizingInTables,
    allowAdvancedToolBarOptions = mediaOptions.allowAdvancedToolBarOptions,
    allowPixelResizing = mediaOptions.allowPixelResizing;
  var isWithinTable = (0, _utils.hasParentNodeOfType)(state.schema.nodes.table)(state.selection);
  if (allowAdvancedToolBarOptions && allowResizing && (!isWithinTable || allowResizingInTables === true) && allowPixelResizing) {
    return [{
      title: formatMessage(_media.pixelEntryMessages.resizeOption),
      onClick: (0, _commands.openPixelEditor)(),
      icon: /*#__PURE__*/_react.default.createElement(_growHorizontal.default, {
        label: ""
      }),
      testId: 'media-pixel-resizing-dropdown-option'
    }];
  }
  return [];
};
