// @atlaskit/smart-card/dist/cjs/view/EmbedCard/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EmbedCard = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _analyticsNext = require("@atlaskit/analytics-next");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _extractAccessContext = require("../../extractors/common/context/extractAccessContext");
var _embed = require("../../extractors/embed");
var _inline = require("../../extractors/inline");
var _helpers = require("../../state/helpers");
var _useControlDataExportConfig = require("../../state/hooks/use-control-data-export-config");
var _jsonld = require("../../utils/jsonld");
var _shouldDataExport = require("../../utils/should-data-export");
var _ResolvedView = _interopRequireDefault(require("../BlockCard/views/ResolvedView"));
var _ResolvedView2 = require("../InlineCard/ResolvedView");
var _ErroredView = require("./views/ErroredView");
var _forbiddenView = _interopRequireDefault(require("./views/forbidden-view"));
var _notFoundView = _interopRequireDefault(require("./views/not-found-view"));
var _ResolvedView3 = require("./views/ResolvedView");
var _unauthorizedView = _interopRequireDefault(require("./views/unauthorized-view"));
var EmbedCard = exports.EmbedCard = /*#__PURE__*/_react.default.forwardRef(function (_ref, iframeRef) {
  var _details$meta, _forbiddenViewProps$c, _forbiddenViewProps$c2, _notFoundViewProps$co, _notFoundViewProps$co2;
  var url = _ref.url,
    cardState = _ref.cardState,
    handleAuthorize = _ref.handleAuthorize,
    handleErrorRetry = _ref.handleErrorRetry,
    handleFrameClick = _ref.handleFrameClick,
    isSelected = _ref.isSelected,
    frameStyle = _ref.frameStyle,
    platform = _ref.platform,
    onResolve = _ref.onResolve,
    onError = _ref.onError,
    testId = _ref.testId,
    inheritDimensions = _ref.inheritDimensions,
    onIframeDwell = _ref.onIframeDwell,
    onIframeFocus = _ref.onIframeFocus,
    iframeUrlType = _ref.iframeUrlType,
    actionOptions = _ref.actionOptions,
    renderers = _ref.renderers;
  var _useAnalyticsEvents = (0, _analyticsNext.useAnalyticsEvents)(),
    createAnalyticsEvent = _useAnalyticsEvents.createAnalyticsEvent;
  var status = cardState.status,
    details = cardState.details;
  var extensionKey = (0, _helpers.getExtensionKey)(details);
  var isProductIntegrationSupported = (0, _helpers.hasAuthScopeOverrides)(details);
  var _useControlDataExport = (0, _useControlDataExportConfig.useControlDataExportConfig)(),
    _useControlDataExport2 = _useControlDataExport.shouldControlDataExport,
    shouldControlDataExport = _useControlDataExport2 === void 0 ? false : _useControlDataExport2;
  switch (status) {
    case 'pending':
    case 'resolving':
      return /*#__PURE__*/_react.default.createElement(_ResolvedView.default, {
        url: url,
        cardState: cardState,
        onClick: handleFrameClick,
        onError: onError,
        onResolve: onResolve,
        renderers: renderers,
        actionOptions: actionOptions,
        testId: testId ? "".concat(testId, "-resolving-view") : 'embed-card-resolving-view'
      });
    case 'resolved':
      var resolvedViewProps = (0, _embed.extractEmbedProps)(details, platform, iframeUrlType);
      if (onResolve) {
        var _resolvedViewProps$pr;
        onResolve({
          title: resolvedViewProps.title,
          url: url,
          aspectRatio: (_resolvedViewProps$pr = resolvedViewProps.preview) === null || _resolvedViewProps$pr === void 0 ? void 0 : _resolvedViewProps$pr.aspectRatio
        });
      }
      if ((0, _platformFeatureFlags.fg)('platform_smart_links_controlled_dsp_export_view')) {
        if ((0, _shouldDataExport.getIsDataExportEnabled)(shouldControlDataExport, cardState.details)) {
          var unauthViewProps = (0, _embed.extractEmbedProps)(details, platform);
          return /*#__PURE__*/_react.default.createElement(_unauthorizedView.default, {
            context: unauthViewProps.context,
            extensionKey: extensionKey,
            frameStyle: frameStyle,
            isProductIntegrationSupported: isProductIntegrationSupported,
            inheritDimensions: inheritDimensions,
            isSelected: isSelected,
            onAuthorize: handleAuthorize,
            onClick: handleFrameClick,
            testId: testId,
            url: unauthViewProps.link
          });
        }
      }
      if (resolvedViewProps.preview) {
        return /*#__PURE__*/_react.default.createElement(_ResolvedView3.EmbedCardResolvedView, (0, _extends2.default)({}, resolvedViewProps, {
          isSelected: isSelected,
          frameStyle: frameStyle,
          inheritDimensions: inheritDimensions,
          onClick: handleFrameClick,
          ref: iframeRef,
          onIframeDwell: onIframeDwell,
          onIframeFocus: onIframeFocus,
          testId: testId
        }));
      } else {
        if (platform === 'mobile') {
          var resolvedInlineViewProps = (0, _inline.extractInlineProps)(details);
          return /*#__PURE__*/_react.default.createElement(_ResolvedView2.InlineCardResolvedView, (0, _extends2.default)({}, resolvedInlineViewProps, {
            isSelected: isSelected,
            testId: testId,
            onClick: handleFrameClick
          }));
        }
        return /*#__PURE__*/_react.default.createElement(_ResolvedView.default, {
          url: url,
          cardState: cardState,
          onClick: handleFrameClick,
          onError: onError,
          onResolve: onResolve,
          renderers: renderers,
          actionOptions: actionOptions,
          testId: testId
        });
      }
    case 'unauthorized':
      if (onError) {
        onError({
          url: url,
          status: status
        });
      }
      var unauthorisedViewProps = (0, _embed.extractEmbedProps)(details, platform);
      return /*#__PURE__*/_react.default.createElement(_unauthorizedView.default, {
        context: unauthorisedViewProps.context,
        extensionKey: extensionKey,
        frameStyle: frameStyle,
        isProductIntegrationSupported: isProductIntegrationSupported,
        inheritDimensions: inheritDimensions,
        isSelected: isSelected,
        onAuthorize: handleAuthorize,
        onClick: handleFrameClick,
        testId: testId,
        url: unauthorisedViewProps.link
      });
    case 'forbidden':
      if (onError) {
        onError({
          url: url,
          status: status
        });
      }
      var forbiddenViewProps = (0, _embed.extractEmbedProps)(details, platform);
      var cardMetadata = (_details$meta = details === null || details === void 0 ? void 0 : details.meta) !== null && _details$meta !== void 0 ? _details$meta : (0, _jsonld.getForbiddenJsonLd)().meta;
      if (forbiddenViewProps.preview) {
        return /*#__PURE__*/_react.default.createElement(_ResolvedView3.EmbedCardResolvedView, (0, _extends2.default)({}, forbiddenViewProps, {
          title: forbiddenViewProps.link,
          frameStyle: frameStyle,
          isSelected: isSelected,
          inheritDimensions: inheritDimensions,
          onClick: handleFrameClick,
          ref: iframeRef
        }));
      }
      var forbiddenAccessContext = (0, _extractAccessContext.extractRequestAccessContextImproved)({
        jsonLd: cardMetadata,
        url: url,
        product: (_forbiddenViewProps$c = (_forbiddenViewProps$c2 = forbiddenViewProps.context) === null || _forbiddenViewProps$c2 === void 0 ? void 0 : _forbiddenViewProps$c2.text) !== null && _forbiddenViewProps$c !== void 0 ? _forbiddenViewProps$c : '',
        createAnalyticsEvent: createAnalyticsEvent
      });
      return /*#__PURE__*/_react.default.createElement(_forbiddenView.default, {
        context: forbiddenViewProps.context,
        frameStyle: frameStyle,
        inheritDimensions: inheritDimensions,
        isSelected: isSelected,
        onAuthorize: handleAuthorize,
        onClick: handleFrameClick,
        accessContext: forbiddenAccessContext,
        url: forbiddenViewProps.link
      });
    case 'not_found':
      if (onError) {
        onError({
          url: url,
          status: status
        });
      }
      var notFoundViewProps = (0, _embed.extractEmbedProps)(details, platform);
      var notFoundAccessContext = details !== null && details !== void 0 && details.meta ? (0, _extractAccessContext.extractRequestAccessContextImproved)({
        jsonLd: details === null || details === void 0 ? void 0 : details.meta,
        url: url,
        product: (_notFoundViewProps$co = (_notFoundViewProps$co2 = notFoundViewProps.context) === null || _notFoundViewProps$co2 === void 0 ? void 0 : _notFoundViewProps$co2.text) !== null && _notFoundViewProps$co !== void 0 ? _notFoundViewProps$co : '',
        createAnalyticsEvent: createAnalyticsEvent
      }) : undefined;
      return /*#__PURE__*/_react.default.createElement(_notFoundView.default, {
        context: notFoundViewProps.context,
        frameStyle: frameStyle,
        inheritDimensions: inheritDimensions,
        isSelected: isSelected,
        onClick: handleFrameClick,
        accessContext: notFoundAccessContext,
        url: notFoundViewProps.link
      });
    case 'fallback':
    case 'errored':
    default:
      if (onError) {
        onError({
          url: url,
          status: status
        });
      }
      return /*#__PURE__*/_react.default.createElement(_ErroredView.EmbedCardErroredView, {
        onRetry: handleErrorRetry,
        inheritDimensions: inheritDimensions,
        isSelected: isSelected
      });
  }
});
