// @atlaskit/user-picker/dist/cjs/components/CustomOption/main.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CustomOption = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _colors = require("@atlaskit/theme/colors");
var _react = require("@emotion/react");
var _react2 = _interopRequireDefault(require("react"));
var _AvatarItemOption = require("../AvatarItemOption");
var _HighlightText = require("../HighlightText");
var _SizeableAvatar = require("../SizeableAvatar");
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); } /**
 * @jsxRuntime classic
 * @jsx jsx
 */ // eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
var CustomOption = exports.CustomOption = /*#__PURE__*/function (_React$PureComponent) {
  function CustomOption() {
    var _this;
    (0, _classCallCheck2.default)(this, CustomOption);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, CustomOption, [].concat(args));
    (0, _defineProperty2.default)(_this, "getPrimaryText", function () {
      var _this$props$data = _this.props.data,
        name = _this$props$data.name,
        highlight = _this$props$data.highlight;
      return [(0, _react.jsx)("span", {
        key: "name"
        // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
        ,
        css: (0, _AvatarItemOption.textWrapper)(_this.props.isSelected ? "var(--ds-text-selected, ".concat(_colors.B400, ")") : "var(--ds-text, ".concat(_colors.N800, ")"))
      }, (0, _react.jsx)(_HighlightText.HighlightText, {
        highlights: highlight && highlight.name
      }, name))];
    });
    (0, _defineProperty2.default)(_this, "getBylineComponent", function (isSelected, message) {
      return (0, _react.jsx)("span", {
        // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
        css: (0, _AvatarItemOption.textWrapper)(isSelected ? "var(--ds-text-selected, ".concat(_colors.B400, ")") : "var(--ds-text-subtlest, ".concat(_colors.N200, ")")),
        "data-testid": "user-picker-custom-secondary-text"
      }, message);
    });
    (0, _defineProperty2.default)(_this, "renderByline", function () {
      var _this$props$data2;
      if (!((_this$props$data2 = _this.props.data) !== null && _this$props$data2 !== void 0 && _this$props$data2.byline)) {
        return undefined;
      }
      return _this.getBylineComponent(_this.props.isSelected, _this.props.data.byline);
    });
    (0, _defineProperty2.default)(_this, "renderAvatar", function () {
      var avatarUrl = _this.props.data.avatarUrl;
      return (0, _react.jsx)(_SizeableAvatar.SizeableAvatar, {
        appearance: "big",
        src: avatarUrl
      });
    });
    (0, _defineProperty2.default)(_this, "getLozengeProps", function () {
      return typeof _this.props.data.lozenge === 'string' ? {
        text: _this.props.data.lozenge
      } : _this.props.data.lozenge;
    });
    return _this;
  }
  (0, _inherits2.default)(CustomOption, _React$PureComponent);
  return (0, _createClass2.default)(CustomOption, [{
    key: "render",
    value: function render() {
      return (0, _react.jsx)(_AvatarItemOption.AvatarItemOption, {
        avatar: this.renderAvatar(),
        isDisabled: this.props.data.isDisabled,
        lozenge: this.getLozengeProps(),
        primaryText: this.getPrimaryText(),
        secondaryText: this.renderByline()
      });
    }
  }]);
}(_react2.default.PureComponent);
