// @atlaskit/react-select/dist/cjs/emotion/components/group.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.groupHeadingCSS = exports.groupCSS = exports.default = exports.GroupHeading = void 0;
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectDestructuringEmpty2 = _interopRequireDefault(require("@babel/runtime/helpers/objectDestructuringEmpty"));
var _react = require("@emotion/react");
var _utils = require("../../utils");
var _excluded = ["data"];
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled
var groupCSS = exports.groupCSS = function groupCSS(_ref) {
  (0, _objectDestructuringEmpty2.default)(_ref);
  return {
    paddingBottom: "var(--ds-space-100, 8px)",
    paddingTop: "var(--ds-space-100, 8px)"
  };
};
var Group = function Group(props) {
  var children = props.children,
    cx = props.cx,
    getStyles = props.getStyles,
    getClassNames = props.getClassNames,
    Heading = props.Heading,
    headingProps = props.headingProps,
    innerProps = props.innerProps,
    label = props.label,
    selectProps = props.selectProps;
  return (0, _react.jsx)("div", (0, _extends2.default)({}, (0, _utils.getStyleProps)(props, 'group', {
    group: true
  }), innerProps), label && (0, _react.jsx)(Heading, (0, _extends2.default)({}, headingProps, {
    selectProps: selectProps,
    getStyles: getStyles,
    getClassNames: getClassNames,
    cx: cx
  }), label), (0, _react.jsx)("div", null, children));
};
var groupHeadingCSS = exports.groupHeadingCSS = function groupHeadingCSS(_ref2) {
  (0, _objectDestructuringEmpty2.default)(_ref2);
  return {
    label: 'group',
    cursor: 'default',
    display: 'block',
    marginBottom: '0.25em',
    paddingLeft: "var(--ds-space-150, 12px)",
    paddingRight: "var(--ds-space-150, 12px)",
    font: "var(--ds-font-body-small, normal 400 11px/16px ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
    color: "var(--ds-text-subtle, #44546F)",
    fontWeight: "var(--ds-font-weight-bold, 700)",
    textTransform: 'none'
  };
};

// eslint-disable-next-line @repo/internal/react/require-jsdoc
var GroupHeading = exports.GroupHeading = function GroupHeading(props) {
  var _cleanCommonProps = (0, _utils.cleanCommonProps)(props),
    data = _cleanCommonProps.data,
    innerProps = (0, _objectWithoutProperties2.default)(_cleanCommonProps, _excluded);
  return (0, _react.jsx)("div", (0, _extends2.default)({}, (0, _utils.getStyleProps)(props, 'groupHeading', {
    'group-heading': true
  }), innerProps));
};

// eslint-disable-next-line @repo/internal/react/require-jsdoc
var _default = exports.default = Group;
