// @atlaskit/react-select/dist/cjs/emotion/components/placeholder.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.placeholderCSS = exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = require("@emotion/react");
var _utils = require("../../utils");
/**
 * @jsxRuntime classic
 * @jsx jsx
 */

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled

var placeholderCSS = exports.placeholderCSS = function placeholderCSS(_ref) {
  var isDisabled = _ref.isDisabled;
  return {
    label: 'placeholder',
    gridArea: '1 / 1 / 2 / 3',
    margin: "0 ".concat("var(--ds-space-025, 2px)"),
    color: isDisabled ? "var(--ds-text-disabled, #091E424F)" : "var(--ds-text-subtlest, #626F86)"
  };
};
var Placeholder = function Placeholder(props) {
  var children = props.children,
    innerProps = props.innerProps;
  return (0, _react.jsx)("div", (0, _extends2.default)({}, (0, _utils.getStyleProps)(props, 'placeholder', {
    placeholder: true
  }), innerProps), children);
};

// eslint-disable-next-line @repo/internal/react/require-jsdoc
var _default = exports.default = Placeholder;
