// @atlaskit/link-datasource/dist/esm/ui/issue-like-table/empty-state/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["as", "style"];
import "./index.compiled.css";
import { forwardRef } from 'react';
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
/* eslint-disable @atlaskit/design-system/use-tokens-typography */

import { Skeleton } from '@atlaskit/linking-common';
import { N40 } from '@atlaskit/theme/colors';
var SkeletonComponent = function SkeletonComponent(_ref) {
  var width = _ref.width,
    itemName = _ref.itemName;
  return /*#__PURE__*/React.createElement(Skeleton, {
    borderRadius: "var(--ds-border-radius-100, 8px)",
    testId: "".concat(itemName, "-empty-state-skeleton"),
    height: 14,
    width: width
  });
};
var tableBodyStyles = null;
var tableStyles = null;
var padding = "var(--ds-space-100, 8px)".concat(" ", "var(--ds-space-100, 8px)");
var cellStyles = null;
var baseColumns = [{
  key: 'type',
  width: 35
}, {
  key: 'issue',
  width: 50
}, {
  key: 'summary',
  width: 100
}, {
  key: 'assignee',
  width: 70
}, {
  key: 'priority',
  width: 60
}, {
  key: 'status',
  width: 60
}, {
  key: 'resolution',
  width: 55
}, {
  key: 'created',
  width: 50
}, {
  key: 'due',
  width: 50
}];
var longColumnWidths = [141, 208, 186, 212, 212, 151, 212, 182, 180, 163, 172, 211, 145, 190];
var shortColumnWidths = [75, 54, 66, 73, 52, 73, 55, 80, 67, 76, 58, 65, 56, 76];
var renderItem = function renderItem(_ref2, _ref3) {
  var key = _ref2.key,
    width = _ref2.width;
  var longWidth = _ref3.longWidth,
    shortWidth = _ref3.shortWidth;
  switch (key) {
    case 'status':
      return /*#__PURE__*/React.createElement(SkeletonComponent, {
        width: shortWidth,
        itemName: key
      });
    case 'summary':
      return /*#__PURE__*/React.createElement(SkeletonComponent, {
        width: longWidth,
        itemName: key
      });
    default:
      return /*#__PURE__*/React.createElement(SkeletonComponent, {
        width: width,
        itemName: key
      });
  }
};
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled
var TableHeading = forwardRef(function (_ref4, __cmplr) {
  var _ref4$as = _ref4.as,
    C = _ref4$as === void 0 ? "th" : _ref4$as,
    __cmpls = _ref4.style,
    __cmplp = _objectWithoutProperties(_ref4, _excluded);
  return /*#__PURE__*/React.createElement(C, _extends({}, __cmplp, {
    style: __cmpls,
    ref: __cmplr,
    className: ax(["_19itidpf _aks5idpf _1u3bidpf _1s7zl0fh _1o9ml0fh _1oupl0fh _n7zl1on0 _1fqk1on0 _1vvv1on0 _kqswh2mm _rc4eh2mm _q46wh2mm _vchhusvi _1wwnusvi _1gbausvi _vwz41tcg _s99e1tcg _3mvc1tcg _ca0qv77o _7cz2v77o _pcwyv77o _u5f31b66 _1pmc1b66 _11ep1b66 _n3tdv77o _6dd7v77o _191qv77o _19bv1b66 _9iya1b66 _15hu1b66 _4t3iqagn _qccgqagn _xo89qagn _s7n4yfq0 _ntwryfq0 _3eneyfq0 _bfhkchd4 _i7olchd4 _1oe6chd4 _12pn15vq _198v15vq _lura1osq _11681tcg _ab8av77o _11j11b66 _1xanv77o _8lvj1b66 _iscccj1k _1nd0cs5v _1t0u1e54 _1eqz4jg8 _95ex1hna _1hkzidpf _1r3pidpf _10i2idpf _1ls2idpf _ocbzidpf _3tfiidpf _1lhc1b66 _n2z81b66 _3ixo1b66", __cmplp.className])
  }));
});
if (process.env.NODE_ENV !== 'production') {
  TableHeading.displayName = 'TableHeading';
}
var EmptyState = function EmptyState(_ref5) {
  var isCompact = _ref5.isCompact,
    testId = _ref5.testId;
  var columnsToRender = isCompact ? baseColumns.slice(0, 6) : baseColumns;
  // if it is compact (non-modal), there is room for 14 rows
  // if it is modal (not compact), there is only room for 10 rows
  var rowsNumber = isCompact ? 14 : 10;
  var rows = new Array(rowsNumber).fill(null).map(function (_, index) {
    return {
      id: index,
      longWidth: longColumnWidths[index],
      shortWidth: shortColumnWidths[index]
    };
  });
  return /*#__PURE__*/React.createElement("div", {
    style: {
      // the IssueLikeDataTableView wraps the table in a container with the styling below while modal doesn't
      // this maxHeight comes from scrollableContainerHeight
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
      maxHeight: 590,
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
      padding: "var(--ds-space-0, 0px)",
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
      boxSizing: 'border-box'
    }
  }, /*#__PURE__*/React.createElement("table", {
    "data-testid": testId,
    className: ax(["_bfhkchd4"])
  }, /*#__PURE__*/React.createElement("thead", {
    style: {
      borderBottom: 0
    }
  }, /*#__PURE__*/React.createElement("tr", null, columnsToRender.map(function (_ref6) {
    var key = _ref6.key,
      width = _ref6.width;
    return (
      /*#__PURE__*/
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
      React.createElement(TableHeading, {
        key: key,
        style: {
          width: width,
          padding: padding
        }
      }, /*#__PURE__*/React.createElement(Skeleton, {
        appearance: "darkGray",
        borderRadius: 8,
        testId: "empty-state-skeleton",
        height: 12,
        width: width
      }))
    );
  }))), /*#__PURE__*/React.createElement("tbody", {
    className: ax(["_n7zlidpf"])
  }, rows.map(function (row) {
    return /*#__PURE__*/React.createElement("tr", {
      key: row.id
    }, columnsToRender.map(function (column) {
      return /*#__PURE__*/React.createElement("td", {
        key: column.key,
        className: ax(["_1yt41uce _1s7z1oij _n7zl1oij _1ls2idpf _1o59u2gc _1lhcu2gc"]),
        style: {
          "--_1n7euq6": ix("0.5px solid ".concat("var(--ds-border, ".concat(N40, ")")))
        }
      }, renderItem(column, row));
    }));
  }))));
};
export default EmptyState;
