// @atlaskit/link-datasource/dist/esm/ui/issue-like-table/empty-state/priority-old.js
/**
 * TODO - I believe this file is not used (I could not find any reference to it in the codebase)
 * If not used we should remove it
 * Delete when cleaning bandicoots-update-sllv-icons
 */
import React from 'react';
import { B300, R300, Y300 } from '@atlaskit/theme/colors';
import IconType from '../render-type/icon';
// https://stackoverflow.com/questions/61099149/svg-fill-color-not-working-with-hex-colors
var replaceHashInHex = function replaceHashInHex(hex) {
  return hex.replace('#', '%23');
};
var priorityToSVGPathMap = {
  low: " <path\n  d=\"M12.5 6.1c.5-.3 1.1-.1 1.4.4.3.5.1 1.1-.3 1.3l-5 3c-.3.2-.7.2-1 0l-5-3c-.6-.2-.7-.9-.4-1.3.2-.5.9-.7 1.3-.4L8 8.8l4.5-2.7z\"\n  fill=\"".concat(replaceHashInHex(B300), "\"\n/>"),
  medium: "<path\n  fill=\"".concat(replaceHashInHex(Y300), "\"\n  d=\"M3,4h10c0.6,0,1,0.4,1,1s-0.4,1-1,1H3C2.4,6,2,5.6,2,5S2.4,4,3,4z M3,10h10c0.6,0,1,0.4,1,1s-0.4,1-1,1H3\n c-0.6,0-1-0.4-1-1S2.4,10,3,10z\"\n/>"),
  high: "<path\n  d=\"M3.5 9.9c-.5.3-1.1.1-1.4-.3s-.1-1.1.4-1.4l5-3c.3-.2.7-.2 1 0l5 3c.5.3.6.9.3 1.4-.3.5-.9.6-1.4.3L8 7.2 3.5 9.9z\"\n  fill=\"".concat(replaceHashInHex(R300), "\"\n/>"),
  blocker: "<path\n  d=\"M8 15c-3.9 0-7-3.1-7-7s3.1-7 7-7 7 3.1 7 7-3.1 7-7 7zM4 7c-.6 0-1 .4-1 1s.4 1 1 1h8c.6 0 1-.4 1-1s-.4-1-1-1H4z\"\n  fill=\"".concat(replaceHashInHex(R300), "\"\n/>")
};
export default (function (_ref) {
  var priority = _ref.priority;
  var image = "<svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 16 16\">".concat(priorityToSVGPathMap[priority], "</svg>");
  return /*#__PURE__*/React.createElement(IconType, {
    source: "data:image/svg+xml;utf8,".concat(image)
  });
});
