// @atlaskit/link-datasource/dist/cjs/hooks/useDatasourceTableFlag.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useDatasourceTableFlag = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _reactIntlNext = require("react-intl-next");
var _uuid = _interopRequireDefault(require("uuid"));
var _flag = require("@atlaskit/flag");
var _crossCircle = _interopRequireDefault(require("@atlaskit/icon/core/migration/cross-circle"));
var _analytics = require("../analytics");
var _messages = require("../ui/issue-like-table/messages");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var getErrorReason = function getErrorReason(status) {
  switch (status) {
    case 403:
      return 'access_denied';
    default:
      return 'request_failed';
  }
};
var getExecuteActionErrorMessage = function getExecuteActionErrorMessage(status) {
  switch (status) {
    case 403:
      return {
        title: /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, _messages.issueLikeTableMessages.updateError403Title),
        description: /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, _messages.issueLikeTableMessages.updateError403Description)
      };
    default:
      return {
        title: /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, _messages.issueLikeTableMessages.updateErrorGenericTitle),
        description: /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, _messages.issueLikeTableMessages.updateErrorGenericDescription)
      };
  }
};
var getFetchActionErrorMessage = function getFetchActionErrorMessage(status) {
  switch (status) {
    default:
      return {
        title: /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, _messages.issueLikeTableMessages.fetchActionErrorGenericTitle),
        description: /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, _messages.issueLikeTableMessages.fetchActionErrorGenericDescription)
      };
  }
};
var useDatasourceTableFlag = exports.useDatasourceTableFlag = function useDatasourceTableFlag(options) {
  var _useFlags = (0, _flag.useFlags)(),
    showFlag = _useFlags.showFlag;
  var _useDatasourceAnalyti = (0, _analytics.useDatasourceAnalyticsEvents)(),
    fireEvent = _useDatasourceAnalyti.fireEvent;
  var showErrorFlag = (0, _react.useCallback)(function (args) {
    showFlag(_objectSpread(_objectSpread({
      icon: /*#__PURE__*/_react.default.createElement(_crossCircle.default, {
        spacing: "spacious",
        label: "Error",
        color: "var(--ds-icon-danger, #C9372C)"
      }),
      id: (0, _uuid.default)(),
      isAutoDismiss: true
    }, options !== null && options !== void 0 && options.isFetchAction ? getFetchActionErrorMessage(args === null || args === void 0 ? void 0 : args.status) : getExecuteActionErrorMessage(args === null || args === void 0 ? void 0 : args.status)), args));
    fireEvent('ui.error.shown.inlineEdit', {
      reason: getErrorReason(args === null || args === void 0 ? void 0 : args.status)
    });
  }, [fireEvent, options === null || options === void 0 ? void 0 : options.isFetchAction, showFlag]);
  return {
    showErrorFlag: showErrorFlag
  };
};
