// @atlaskit/mention/dist/cjs/components/Scrollable/styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ScrollableStyle = void 0;
var _styled = _interopRequireDefault(require("@emotion/styled"));
var _constants = require("@atlaskit/theme/constants");
var _sharedStyles = require("../../shared-styles");
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var ScrollableStyle = exports.ScrollableStyle = _styled.default.div({
  display: 'block',
  overflowX: 'hidden',
  overflowY: 'auto',
  padding: "var(--ds-space-050, 4px)".concat(" 0"),
  margin: 0,
  background: "var(--ds-surface, white)",
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  maxHeight: _sharedStyles.scrollableMaxHeight,
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  borderRadius: "".concat((0, _constants.borderRadius)(), "px")
});
