// @atlaskit/jql-editor/dist/cjs/analytics/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ActionSubjectId = exports.ActionSubject = exports.Action = void 0;
var Action = exports.Action = /*#__PURE__*/function (Action) {
  Action["CLICKED"] = "clicked";
  Action["RETRIEVE_FAILED"] = "retrieveFailed";
  Action["RETRIEVED"] = "retrieved";
  Action["SELECTED"] = "selected";
  Action["VIEWED"] = "viewed";
  return Action;
}({});
var ActionSubject = exports.ActionSubject = /*#__PURE__*/function (ActionSubject) {
  ActionSubject["BUTTON"] = "button";
  ActionSubject["AUTOCOMPLETE_OPTION"] = "autocompleteOption";
  ActionSubject["ERROR_MESSAGE"] = "errorMessage";
  return ActionSubject;
}({});
var ActionSubjectId = exports.ActionSubjectId = /*#__PURE__*/function (ActionSubjectId) {
  ActionSubjectId["EDITOR_EXPAND"] = "jqlEditorExpand";
  ActionSubjectId["EDITOR_SEARCH"] = "jqlEditorSearch";
  ActionSubjectId["EDITOR_HELP"] = "jqlEditorHelp";
  ActionSubjectId["JQL_RESULT"] = "jqlResult";
  return ActionSubjectId;
}({});
