// @atlaskit/jql-editor/dist/cjs/analytics/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ANALYTICS_CHANNEL", {
  enumerable: true,
  get: function get() {
    return _jqlEditorCommon.ANALYTICS_CHANNEL;
  }
});
Object.defineProperty(exports, "Action", {
  enumerable: true,
  get: function get() {
    return _constants.Action;
  }
});
Object.defineProperty(exports, "ActionSubject", {
  enumerable: true,
  get: function get() {
    return _constants.ActionSubject;
  }
});
Object.defineProperty(exports, "ActionSubjectId", {
  enumerable: true,
  get: function get() {
    return _constants.ActionSubjectId;
  }
});
Object.defineProperty(exports, "EventType", {
  enumerable: true,
  get: function get() {
    return _jqlEditorCommon.EventType;
  }
});
Object.defineProperty(exports, "JQLEditorAnalyticsListener", {
  enumerable: true,
  get: function get() {
    return _listener.JQLEditorAnalyticsListener;
  }
});
Object.defineProperty(exports, "useJqlEditorAnalytics", {
  enumerable: true,
  get: function get() {
    return _util.useJqlEditorAnalytics;
  }
});
Object.defineProperty(exports, "useJqlPackageAnalytics", {
  enumerable: true,
  get: function get() {
    return _jqlEditorCommon.useJqlPackageAnalytics;
  }
});
var _jqlEditorCommon = require("@atlaskit/jql-editor-common");
var _constants = require("./constants");
var _listener = require("./listener");
var _util = require("./util");
