// @atlaskit/editor-plugin-text-color/dist/cjs/ui/FloatingToolbarComponent.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FloatingToolbarComponent = FloatingToolbarComponent;
var _react = _interopRequireDefault(require("react"));
var _hooks = require("@atlaskit/editor-common/hooks");
var _types = require("../types");
var _ToolbarTextColor = _interopRequireDefault(require("./ToolbarTextColor"));
// Ignored via go/ees005
// eslint-disable-next-line import/no-named-as-default

var FloatingToolbarSettings = {
  disabled: false,
  isReducedSpacing: true
};
function FloatingToolbarComponent(_ref) {
  var api = _ref.api,
    editorView = _ref.editorView,
    dispatchAnalyticsEvent = _ref.dispatchAnalyticsEvent;
  var _useSharedPluginState = (0, _hooks.useSharedPluginState)(api, ['textColor']),
    textColorState = _useSharedPluginState.textColorState;
  return /*#__PURE__*/_react.default.createElement(_ToolbarTextColor.default
  // Ignored via go/ees005
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  , {
    pluginState: textColorState,
    isReducedSpacing: FloatingToolbarSettings.isReducedSpacing,
    editorView: editorView,
    dispatchAnalyticsEvent: dispatchAnalyticsEvent,
    disabled: FloatingToolbarSettings.disabled,
    pluginInjectionApi: api,
    toolbarType: _types.ToolbarType.FLOATING
  });
}
