// @atlaskit/editor-core/dist/esm/plugins/rule/index.js
import * as React from 'react';
import { rule } from '@atlaskit/adf-schema';
import { messages } from '../insert-block/ui/ToolbarInsertBlock';
import keymapPlugin from './pm-plugins/keymap';
import inputRulePlugin from './pm-plugins/input-rule';
import { addAnalytics, ACTION, ACTION_SUBJECT, ACTION_SUBJECT_ID, INPUT_METHOD, EVENT_TYPE, } from '../analytics';
import { IconDivider } from '../quick-insert/assets';
var rulePlugin = function () { return ({
    nodes: function () {
        return [{ name: 'rule', node: rule }];
    },
    pmPlugins: function () {
        return [
            {
                name: 'ruleInputRule',
                plugin: function (_a) {
                    var schema = _a.schema;
                    return inputRulePlugin(schema);
                },
            },
            {
                name: 'ruleKeymap',
                plugin: function () { return keymapPlugin(); },
            },
        ];
    },
    pluginsOptions: {
        quickInsert: function (_a) {
            var formatMessage = _a.formatMessage;
            return [
                {
                    title: formatMessage(messages.horizontalRule),
                    description: formatMessage(messages.horizontalRuleDescription),
                    keywords: ['horizontal rule', 'rule', 'line'],
                    priority: 1200,
                    keyshortcut: '---',
                    icon: function () { return (React.createElement(IconDivider, { label: formatMessage(messages.horizontalRule) })); },
                    action: function (insert, state) {
                        var tr = insert(state.schema.nodes.rule.createChecked());
                        return addAnalytics(tr, {
                            action: ACTION.INSERTED,
                            actionSubject: ACTION_SUBJECT.DOCUMENT,
                            actionSubjectId: ACTION_SUBJECT_ID.DIVIDER,
                            attributes: { inputMethod: INPUT_METHOD.QUICK_INSERT },
                            eventType: EVENT_TYPE.TRACK,
                        });
                    },
                },
            ];
        },
    },
}); };
export default rulePlugin;
