// @atlaskit/link-picker/dist/cjs/ui/link-picker/search-results/scrolling-tabs/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ScrollingTabList = void 0;
require("./index.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _new = require("@atlaskit/button/new");
var _chevronLeft = _interopRequireDefault(require("@atlaskit/icon/utility/migration/chevron-left"));
var _chevronRight = _interopRequireDefault(require("@atlaskit/icon/utility/migration/chevron-right"));
var _colors = require("@atlaskit/theme/colors");
var _scrollingTabs = require("./scrolling-tabs");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t.return || t.return(); } finally { if (u) throw o; } } }; }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function isTouchDevice() {
  return 'ontouchstart' in window ||
  // eslint-disable-next-line compat/compat
  navigator.maxTouchPoints > 0;
}
var initialConditionalButtonsState = {
  back: false,
  forward: false
};
var scrollingContainerStyles = null;
var containerStyles = null;
var backButtonStyles = null;
var nextButtonStyles = null;

/* eslint-disable @repo/internal/dom-events/no-unsafe-event-listeners */
var ScrollingTabList = exports.ScrollingTabList = function ScrollingTabList(props) {
  var ref = (0, _react.useRef)(null);
  var _useState = (0, _react.useState)(initialConditionalButtonsState),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    conditionalButtons = _useState2[0],
    setConditionalButtons = _useState2[1];
  var ghost = (0, _react.useMemo)(function () {
    return (0, _scrollingTabs.createGhost)();
  }, []);
  var onTabClick = (0, _react.useCallback)(function (e) {
    var target = e.currentTarget;
    if (target) {
      target.scrollIntoView({
        behavior: 'smooth',
        block: 'nearest',
        inline: 'center'
      });
    }
  }, []);
  _react.default.useLayoutEffect(function () {
    var container = ref.current;
    var scrollingContainer;
    var tabs;
    var handleConditionalButtonsChange = function handleConditionalButtonsChange() {
      var buttons = (0, _scrollingTabs.calculateConditionalButtons)(scrollingContainer, isTouchDevice());
      setConditionalButtons(buttons);
    };
    var observerCallback = function observerCallback(mutationList) {
      var tablist = (0, _scrollingTabs.getTabList)(ref);
      var _iterator = _createForOfIteratorHelper(mutationList),
        _step;
      try {
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          var mutation = _step.value;
          if (mutation.type === 'childList') {
            var addedNodes = Array.from(mutation.addedNodes);
            var found = addedNodes.find(function (node) {
              return node.getAttribute('role') === 'tab';
            });
            if (found && tablist) {
              ghost.remove();
              var _tabs = Array.from(tablist.children);
              _tabs.forEach(function (tab) {
                tab.removeEventListener('click', onTabClick);
                tab.addEventListener('click', onTabClick);
              });
              tablist.appendChild(ghost);
              handleConditionalButtonsChange();
            }
          }
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }
    };
    var observer = new MutationObserver(observerCallback);
    if (container) {
      var tablist = (0, _scrollingTabs.getTabList)(ref);
      scrollingContainer = tablist === null || tablist === void 0 ? void 0 : tablist.parentElement;
      observer.observe(container, {
        attributes: false,
        childList: true,
        subtree: true
      });
      if (scrollingContainer instanceof HTMLElement && tablist) {
        tablist.appendChild(ghost);
        tabs = Array.from(tablist.children);
        tabs.forEach(function (tab) {
          return tab.addEventListener('click', onTabClick);
        });
        handleConditionalButtonsChange();
        scrollingContainer.addEventListener('scroll', handleConditionalButtonsChange);
      }
      return function () {
        if (scrollingContainer) {
          // eslint-disable-next-line @repo/internal/dom-events/no-unsafe-event-listeners
          scrollingContainer.removeEventListener('scroll', handleConditionalButtonsChange);
        }
        if (tabs.length) {
          tabs.forEach(function (tab) {
            return tab.removeEventListener('click', onTabClick);
          });
        }
      };
    }
  }, [onTabClick, ghost, ref]);
  var buttonProps = {
    appearance: 'subtle',
    spacing: 'compact'
  };
  return /*#__PURE__*/_react.default.createElement("div", {
    ref: ref,
    "data-testid": "scrolling-tabs",
    className: (0, _runtime.ax)(["_kqswh2mm _18q9glyw _7hip261p _1fud261p _1dze261p"])
  }, conditionalButtons.back && /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _runtime.ax)(["_kqswstnw _154i1b66 _1pby18h8 _bfhkppbu _1ltvidpf", "back"]),
    style: {
      "--_1t3jkli": (0, _runtime.ix)("var(--ds-surface, ".concat(_colors.N0, ")"))
    }
  }, /*#__PURE__*/_react.default.createElement(_new.IconButton, (0, _extends2.default)({
    "data-test-id": "back",
    onClick: function onClick() {
      return (0, _scrollingTabs.scrollBack)(ref);
    },
    label: "back",
    icon: _chevronLeft.default
  }, buttonProps))), /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _runtime.ax)(["_1bh4vqrj _1reo1wug _1dm8q5uf _1fjgglyw _rfx3yh40 _pascidpf _eid3idpf _zr3eidpf _fntnidpf _1kt9b3bt _19woidpf _1cs8stnw _1mp41kw7 _qnecu2gc _oounu2gc _kfgtyh40 _1russudh _1cicglyw"]),
    style: {
      "--_lvpq93": (0, _runtime.ix)("var(--ds-border, ".concat(_colors.N30, ")"))
    }
  }, props.children), conditionalButtons.forward && /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _runtime.ax)(["_kqswstnw _154i1b66 _1pby18h8 _bfhkppbu _1xi2idpf"]),
    style: {
      "--_1t3jkli": (0, _runtime.ix)("var(--ds-surface, ".concat(_colors.N0, ")"))
    }
  }, /*#__PURE__*/_react.default.createElement(_new.IconButton, (0, _extends2.default)({
    "data-test-id": "forward",
    onClick: function onClick() {
      return (0, _scrollingTabs.scrollForward)(ref);
    }
  }, buttonProps, {
    label: "forward",
    icon: _chevronRight.default
  }))));
};
