// @atlaskit/jql-editor/dist/cjs/ui/jql-editor-portal-provider/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.JQLEditorPortalRenderer = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _reactDom = require("react-dom");
var _context = require("./context");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var initialState = {
  components: {},
  containers: {}
};
var reducer = function reducer(state, action) {
  switch (action.type) {
    case 'createPortal':
      {
        var _action$payload = action.payload,
          key = _action$payload.key,
          portalComponent = _action$payload.portalComponent,
          container = _action$payload.container;
        return _objectSpread(_objectSpread({}, state), {}, {
          components: _objectSpread(_objectSpread({}, state.components), {}, (0, _defineProperty2.default)({}, key, {
            portalComponent: portalComponent,
            container: container
          }))
        });
      }
    case 'destroyPortal':
      {
        var _state$components = state.components,
          _action$payload2 = action.payload,
          toDestroy = _state$components[_action$payload2],
          components = (0, _objectWithoutProperties2.default)(_state$components, [_action$payload2].map(_toPropertyKey));
        return _objectSpread(_objectSpread({}, state), {}, {
          components: components
        });
      }
    case 'registerPluginContainer':
      {
        var _action$payload3 = action.payload,
          element = _action$payload3.element,
          containerKey = _action$payload3.containerKey;
        if (!element) {
          return state;
        }
        return _objectSpread(_objectSpread({}, state), {}, {
          containers: _objectSpread(_objectSpread({}, state.containers), {}, (0, _defineProperty2.default)({}, containerKey, element))
        });
      }
    default:
      {
        return state;
      }
  }
};

/**
 * Provides actions to manage portals via context and renders portalled components into the React tree.
 */
var JQLEditorPortalRenderer = exports.JQLEditorPortalRenderer = function JQLEditorPortalRenderer(_ref) {
  var children = _ref.children;
  var _useReducer = (0, _react.useReducer)(reducer, initialState),
    _useReducer2 = (0, _slicedToArray2.default)(_useReducer, 2),
    state = _useReducer2[0],
    dispatch = _useReducer2[1];
  var actionsContext = (0, _react.useMemo)(function () {
    return {
      onCreatePortal: function onCreatePortal(key, portalComponent, container) {
        return dispatch({
          type: 'createPortal',
          payload: {
            key: key,
            portalComponent: portalComponent,
            container: container
          }
        });
      },
      onDestroyPortal: function onDestroyPortal(key) {
        return dispatch({
          type: 'destroyPortal',
          payload: key
        });
      },
      onRegisterPluginContainer: function onRegisterPluginContainer(containerKey, element) {
        return dispatch({
          type: 'registerPluginContainer',
          payload: {
            containerKey: containerKey,
            element: element
          }
        });
      }
    };
  }, [dispatch]);
  return /*#__PURE__*/_react.default.createElement(_context.PortalActionsContext.Provider, {
    value: actionsContext
  }, children, Object.entries(state.components).map(function (_ref2) {
    var _ref3 = (0, _slicedToArray2.default)(_ref2, 2),
      key = _ref3[0],
      value = _ref3[1];
    var container = value.container instanceof Element ? value.container : state.containers[value.container];
    return container !== undefined ? /*#__PURE__*/(0, _reactDom.createPortal)(value.portalComponent, container, key) : null;
  }));
};
