// @atlaskit/media-ui/dist/cjs/abuseModal/abuseModal.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AbuseModal = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactIntlNext = require("react-intl-next");
var _new = _interopRequireWildcard(require("@atlaskit/button/new"));
var _closeCross = _interopRequireDefault(require("@atlaskit/icon/core/migration/close--cross"));
var _modalDialog = _interopRequireWildcard(require("@atlaskit/modal-dialog"));
var _primitives = require("@atlaskit/primitives");
var _messages = require("../messages");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var gridStyles = (0, _primitives.xcss)({
  width: '100%'
});
var closeContainerStyles = (0, _primitives.xcss)({
  gridArea: 'close'
});
var titleContainerStyles = (0, _primitives.xcss)({
  gridArea: 'title'
});
var AbuseModal = exports.AbuseModal = (0, _reactIntlNext.injectIntl)(function (_ref) {
  var isOpen = _ref.isOpen,
    onConfirm = _ref.onConfirm,
    onClose = _ref.onClose,
    formatMessage = _ref.intl.formatMessage;
  return /*#__PURE__*/_react.default.createElement(_modalDialog.ModalTransition, null, isOpen && /*#__PURE__*/_react.default.createElement(_modalDialog.default, {
    onClose: onClose,
    testId: "mediaAbuseModal"
  }, /*#__PURE__*/_react.default.createElement(_modalDialog.ModalHeader, null, /*#__PURE__*/_react.default.createElement(_primitives.Grid, {
    gap: "space.200",
    templateAreas: ['title close'],
    xcss: gridStyles
  }, /*#__PURE__*/_react.default.createElement(_primitives.Flex, {
    xcss: closeContainerStyles,
    justifyContent: "end"
  }, /*#__PURE__*/_react.default.createElement(_new.IconButton, {
    appearance: "subtle",
    icon: _closeCross.default,
    label: "Close Modal",
    onClick: onClose
  })), /*#__PURE__*/_react.default.createElement(_primitives.Flex, {
    xcss: titleContainerStyles,
    justifyContent: "start"
  }, /*#__PURE__*/_react.default.createElement(_modalDialog.ModalTitle, {
    appearance: "warning"
  }, formatMessage(_messages.messages.abuse_modal_title))))), /*#__PURE__*/_react.default.createElement(_modalDialog.ModalBody, null, formatMessage(_messages.messages.abuse_modal_body)), /*#__PURE__*/_react.default.createElement(_modalDialog.ModalFooter, null, /*#__PURE__*/_react.default.createElement(_new.default, {
    appearance: "subtle",
    onClick: onClose
  }, formatMessage(_messages.messages.cancel)), /*#__PURE__*/_react.default.createElement(_new.default, {
    appearance: "warning",
    onClick: onConfirm
  }, formatMessage(_messages.messages.abuse_modal_submit)))));
});
