// @atlaskit/media-card/dist/es2019/card/ui/progressBar/styledBar-compiled.js
/* styledBar-compiled.tsx generated by @compiled/babel-plugin v0.36.1 */
import "./styledBar-compiled.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { Breakpoint, getTitleBoxHeight, responsiveSettings } from '../common';
const height = 3;
const padding = 1;
const width = 95; // %

const smallSizeSettings = {
  marginBottom: 4
};
const largeSizeSettings = {
  marginBottom: 12
};
export function generateResponsiveStyles(breakpoint, positionBottom, showOnTop, multiplier = 1) {
  const setting = breakpoint === Breakpoint.SMALL ? smallSizeSettings : largeSizeSettings;
  const marginPositionBottom = responsiveSettings[breakpoint].titleBox.verticalPadding;
  const marginBottom = setting.marginBottom * multiplier + (positionBottom ? marginPositionBottom : getTitleBoxHeight(breakpoint));
  const spacing = `${marginBottom}px`;
  return showOnTop ? {
    top: spacing
  } : {
    bottom: spacing
  };
}
const styledBarStyles = null;
const styleChild = null;
export const StyledBar = props => {
  const {
    progress,
    breakpoint,
    positionBottom,
    showOnTop
  } = props;
  return /*#__PURE__*/React.createElement("div", {
    id: "styledBar",
    role: "progressbar",
    "aria-valuenow": progress,
    style: {
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
      ...generateResponsiveStyles(breakpoint, positionBottom, showOnTop)
    },
    className: ax(["_1reo15vq _18m915vq _ca0qt94y _u5f3t94y _n3tdt94y _19bvt94y _2rko1l7b _kqswstnw _1bsb1lk2 _1ltv1ghs _bfhkrcrw _4t3i14y2 _vchhusvi"])
  }, /*#__PURE__*/React.createElement("div", {
    style: {
      width: `${progress}%`
    },
    className: ax(["_2rko1l7b _bfhk1p0d _1e0c1ule _4t3i1osq"])
  }));
};
StyledBar.displayName = 'StyledProgressBar';
