// @atlaskit/smart-card/dist/cjs/view/LinkUrl/LinkWarningModal/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./index.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _reactIntlNext = require("react-intl-next");
var _standardButton = _interopRequireDefault(require("@atlaskit/button/standard-button"));
var _modalDialog = _interopRequireWildcard(require("@atlaskit/modal-dialog"));
var _messages = require("../../../messages");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var breakWordsCss = null;
var WarningModal = function WarningModal(props) {
  var isOpen = props.isOpen,
    unsafeLinkText = props.unsafeLinkText,
    url = props.url,
    onClose = props.onClose,
    onContinue = props.onContinue,
    intl = props.intl;
  var content = /*#__PURE__*/React.createElement(_modalDialog.ModalTransition, null, isOpen && /*#__PURE__*/React.createElement(_modalDialog.default, {
    onClose: onClose,
    testId: "link-with-safety-warning"
  }, /*#__PURE__*/React.createElement(_modalDialog.ModalHeader, null, /*#__PURE__*/React.createElement(_modalDialog.ModalTitle, {
    appearance: "warning"
  }, /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, _messages.messages.check_this_link))), /*#__PURE__*/React.createElement(_modalDialog.ModalBody, null, /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)(["_o5726vc6 _1nmz9jpi"])
  }, url && unsafeLinkText && /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, (0, _extends2.default)({}, _messages.messages.link_safety_warning_message, {
    values: {
      unsafeLinkText: unsafeLinkText,
      a: function a() {
        return /*#__PURE__*/React.createElement("a", {
          href: url,
          target: "_blank",
          rel: "noopener noreferrer"
        }, url);
      }
    }
  })))), /*#__PURE__*/React.createElement(_modalDialog.ModalFooter, null, /*#__PURE__*/React.createElement(_standardButton.default, {
    appearance: "subtle",
    onClick: onClose
  }, /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, _messages.messages.go_back)), /*#__PURE__*/React.createElement(_standardButton.default, {
    appearance: "warning",
    onClick: onContinue,
    autoFocus: true
  }, /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, _messages.messages.continue)))));
  return intl ? content : /*#__PURE__*/React.createElement(_reactIntlNext.IntlProvider, {
    locale: "en"
  }, content);
};
var _default = exports.default = (0, _reactIntlNext.injectIntl)(WarningModal, {
  enforceContext: false
});
