// @atlaskit/media-picker/dist/cjs/components/errorFlagGroup/ErrorFlagGroup.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireDefault(require("react"));
var _flag = require("@atlaskit/flag");
var _crossCircle = _interopRequireDefault(require("@atlaskit/icon/core/migration/cross-circle"));
var _colors = require("@atlaskit/theme/colors");
var _messages = require("./messages");
var _reactIntlNext = require("react-intl-next");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var defaultOptions = {
  style: 'unit',
  unitDisplay: 'short',
  minimumFractionDigits: 2
};
var fileSizeFormatters = {
  kilobyte: new Intl.NumberFormat('en', _objectSpread(_objectSpread({}, defaultOptions), {}, {
    unit: 'kilobyte'
  })),
  megabyte: new Intl.NumberFormat('en', _objectSpread(_objectSpread({}, defaultOptions), {}, {
    unit: 'megabyte'
  })),
  gigabyte: new Intl.NumberFormat('en', _objectSpread(_objectSpread({}, defaultOptions), {}, {
    unit: 'gigabyte'
  }))
};
var formatFileSize = function formatFileSize(size) {
  var formattedSize;
  if (size < 1000000) {
    formattedSize = size / 1000;
    return fileSizeFormatters.kilobyte.format(formattedSize);
  } else if (size >= 1000000 && size < 1000000000) {
    formattedSize = size / 1000000;
    return fileSizeFormatters.megabyte.format(formattedSize);
  } else {
    formattedSize = size / 1000000000;
    return fileSizeFormatters.gigabyte.format(formattedSize);
  }
};
var FlagGroupContent = function FlagGroupContent(_ref) {
  var flagData = _ref.flagData,
    onFlagDismissed = _ref.onFlagDismissed;
  var intl = (0, _reactIntlNext.useIntl)();
  return /*#__PURE__*/_react.default.createElement(_flag.FlagGroup, {
    onDismissed: onFlagDismissed
  }, flagData.map(function (data, i) {
    return /*#__PURE__*/_react.default.createElement(_flag.AutoDismissFlag, {
      id: i,
      icon: /*#__PURE__*/_react.default.createElement(_crossCircle.default, {
        color: "var(--ds-icon-danger, ".concat(_colors.R300, ")"),
        label: "Fail",
        LEGACY_size: "medium",
        spacing: "spacious"
      }),
      key: i,
      title: intl.formatMessage(_messages.errorFlagMessages.errorTitle),
      description: data.reason === 'fileEmpty' ? intl.formatMessage(_messages.errorFlagMessages.fileEmptyDescription, {
        fileName: data.fileName
      }) : intl.formatMessage(_messages.errorFlagMessages.uploadRejectionDescription, {
        fileName: data.fileName,
        limit: formatFileSize(data.limit)
      })
    });
  }));
};
var ErrorFlagGroup = function ErrorFlagGroup(_ref2) {
  var flagData = _ref2.flagData,
    onFlagDismissed = _ref2.onFlagDismissed,
    intl = _ref2.intl;
  return intl ? /*#__PURE__*/_react.default.createElement(FlagGroupContent, {
    flagData: flagData,
    onFlagDismissed: onFlagDismissed
  }) : /*#__PURE__*/_react.default.createElement(_reactIntlNext.IntlProvider, {
    locale: "en"
  }, /*#__PURE__*/_react.default.createElement(FlagGroupContent, {
    flagData: flagData,
    onFlagDismissed: onFlagDismissed
  }));
};
var _default = exports.default = (0, _reactIntlNext.injectIntl)(ErrorFlagGroup, {
  enforceContext: false
});
