// @atlaskit/link-picker/dist/esm/ui/link-picker/search-results/link-search-list/list-item/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import "./index.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { forwardRef, Fragment } from 'react';
import { useIntl } from 'react-intl-next';
import { fg } from '@atlaskit/platform-feature-flags';
import { Text } from '@atlaskit/primitives/compiled';
import { B100, B400, B50, N20, N200, N300, N400, N40A } from '@atlaskit/theme/colors';
/* eslint-disable-next-line @atlassian/tangerine/import/no-parent-imports */
import { transformTimeStamp } from '../../../transformTimeStamp';
export var testIds = {
  searchResultItem: 'link-search-list-item',
  searchResultIcon: 'link-search-list-item-icon'
};
var isSVG = function isSVG(icon) {
  return icon.startsWith('<svg') && icon.endsWith('</svg>');
};
var base64SVG = function base64SVG(icon) {
  return "data:image/svg+xml;base64,".concat(Buffer.from(icon).toString('base64'));
};
var itemIconStyles = null;
var listItemContextStyles = null;
var listItemContainerStyles = null;
var listItemContainerInnerStyles = null;
var listItemContainerInnerStylesA11yRefresh = null;
var listItemContainerInnerStylesSeparatorLinkPickerRefresh = null;
var itemNameStyles = null;
var imgStyles = null;
var listItemBaseStyles = null;
var ListItemIcon = function ListItemIcon(props) {
  var item = props.item,
    intl = props.intl;
  var icon = item.icon,
    iconAlt = item.iconAlt;
  if (!icon) {
    return null;
  }
  var alt = typeof iconAlt === 'string' ? iconAlt : intl.formatMessage(iconAlt);
  if (typeof icon !== 'string') {
    var Glyph = icon;
    return /*#__PURE__*/React.createElement("span", {
      className: ax(["_1ul9pxbi _19pk1b66 _2hwxutpp"])
    }, /*#__PURE__*/React.createElement(Glyph, {
      alt: alt,
      "data-testid": testIds.searchResultIcon
    }));
  }
  return /*#__PURE__*/React.createElement("span", {
    className: ax(["_1ul9pxbi _19pk1b66 _2hwxutpp"])
  }, /*#__PURE__*/React.createElement("img", {
    "data-testid": testIds.searchResultIcon,
    src: isSVG(icon) ? base64SVG(icon) : icon,
    alt: alt,
    className: ax(["_p12fpxbi"])
  }));
};
var ListItemSubtitle = function ListItemSubtitle(_ref) {
  var _ref$items = _slicedToArray(_ref.items, 2),
    firstItem = _ref$items[0],
    secondItem = _ref$items[1];
  return /*#__PURE__*/React.createElement("div", {
    "data-testid": "".concat(testIds.searchResultItem, "-subtitle"),
    className: ax(["_11c81o8v _syazct01 _1e0c1txw"]),
    style: {
      "--_khvm95": ix("var(--ds-text, ".concat(N300, ")"))
    }
  }, /*#__PURE__*/React.createElement("div", {
    className: ax(["_1reo15vq _18m915vq _1bto1l2s"])
  }, /*#__PURE__*/React.createElement("span", {
    className: ax([fg('bandicoots-a11y-link-picker-styling') ? "_syaz121t _o5721q9c" : "_syazpsu8 _o5721q9c"]),
    style: {
      "--_1pseksa": ix("var(--ds-text-subtle, ".concat(N400, ")")),
      "--_vrm7dk": ix("var(--ds-text-subtlest, ".concat(N200, ")"))
    }
  }, firstItem)), secondItem && /*#__PURE__*/React.createElement("div", {
    className: ax([fg('bandicoots-a11y-link-picker-styling') ? "_syaz121t _o5721q9c" : "_syazpsu8 _o5721q9c"]),
    style: {
      "--_1pseksa": ix("var(--ds-text-subtle, ".concat(N400, ")")),
      "--_vrm7dk": ix("var(--ds-text-subtlest, ".concat(N200, ")"))
    }
  }, fg('platform-linking-visual-refresh-link-picker') ? /*#__PURE__*/React.createElement("span", {
    className: ax(["_syaz1svm"]),
    style: {
      "--_1trgtlr": ix("var(--ds-border, ".concat(N40A, ")"))
    }
  }, /*#__PURE__*/React.createElement(Fragment, null, "\xA0 \u2022\xA0 ")) : /*#__PURE__*/React.createElement(Fragment, null, "\xA0 \u2022\xA0 "), /*#__PURE__*/React.createElement(Fragment, null, secondItem)));
};
var getDefaultSubtitleItems = function getDefaultSubtitleItems(item, intl) {
  var container = item.container;
  var date = transformTimeStamp(intl, item.lastViewedDate, item.lastUpdatedDate);
  if (container) {
    if (date) {
      return [container, date];
    }
    return [container];
  }
  if (date) {
    return [date];
  }
  return undefined;
};
var listItemActive = null;
var listItemSelected = null;
var listItemFocusStyles = null;
export var LinkSearchListItem = /*#__PURE__*/forwardRef(function (_ref2, ref) {
  var item = _ref2.item,
    selected = _ref2.selected,
    id = _ref2.id,
    role = _ref2.role,
    onSelect = _ref2.onSelect,
    tabIndex = _ref2.tabIndex,
    onKeyDown = _ref2.onKeyDown,
    onFocus = _ref2.onFocus,
    _ref2$nameMaxLines = _ref2.nameMaxLines,
    nameMaxLines = _ref2$nameMaxLines === void 0 ? 1 : _ref2$nameMaxLines;
  var intl = useIntl();
  var handleSelect = function handleSelect() {
    return onSelect(item.objectId);
  };
  var subtitleItems = item.subtitleItems || getDefaultSubtitleItems(item, intl);
  return /*#__PURE__*/React.createElement("div", {
    role: role,
    id: id,
    "aria-selected": selected,
    "data-testid": testIds.searchResultItem,
    onKeyDown: onKeyDown,
    onClick: handleSelect,
    onFocus: onFocus,
    tabIndex: tabIndex,
    ref: ref,
    className: ax(["_19pkidpf _2hwxidpf _otyridpf _18u0idpf _1e0c1txw _ca0qu2gc _n3tdu2gc _19bvrvlv _u5f3kq6n _80omtlke", !selected && "_irr38iyy _1dbz1f4r", selected && "_bfhk1kxl _16qs1f4r", "_nt751r31 _49pcglyw _1hvw1o36 _1a3b1r31 _4fprglyw _5goinqa1 _7ehif7x9"]),
    style: {
      "--_u7ouzj": ix("clamp( ".concat("var(--ds-space-100, 8px)", ", var(--link-picker-padding-left), 100% )")),
      "--_1kucder": ix("clamp( ".concat("var(--ds-space-100, 8px)", ", var(--link-picker-padding-right), 100% )")),
      "--_mmrjro": ix("var(--ds-background-neutral-subtle-hovered, ".concat(N20, ")")),
      "--_1mo126v": ix("inset 2px 0px 0px ".concat("var(--ds-border-selected, ".concat(B400, ")"))),
      "--_rryx0a": ix("var(--ds-background-selected, ".concat(B50, ")")),
      "--_ttguoj": ix("0 0 0 2px ".concat("var(--ds-border-focused, ".concat(B100, ")"), " inset"))
    }
  }, /*#__PURE__*/React.createElement(ListItemIcon, {
    item: item,
    intl: intl
  }), /*#__PURE__*/React.createElement("div", {
    className: ax(["_1reo15vq _18m915vq _ae4v1h6o _1bsb1osq"])
  }, /*#__PURE__*/React.createElement(Text, {
    maxLines: nameMaxLines
  }, /*#__PURE__*/React.createElement("span", {
    "data-testid": "".concat(testIds.searchResultItem, "-title"),
    title: item.name
  }, item.name)), subtitleItems && /*#__PURE__*/React.createElement(ListItemSubtitle, {
    items: subtitleItems
  })));
});
