// @atlaskit/smart-card/dist/cjs/view/BlockCard/views/unresolved-view/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./index.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _FlexibleCard = _interopRequireDefault(require("../../../FlexibleCard"));
var _blocks = require("../../../FlexibleCard/components/blocks");
var _utils = require("../utils");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var customBlockStylesOld = null;
var customBlockStyles = null;
var titleBlockCssOld = null;
var titleBlockCss = null;
var footerBlockCssOld = null;
var footerBlockCss = null;
var UnresolvedView = function UnresolvedView(_ref) {
  var actions = _ref.actions,
    cardState = _ref.cardState,
    children = _ref.children,
    onAuthorize = _ref.onAuthorize,
    onClick = _ref.onClick,
    onError = _ref.onError,
    _ref$showPreview = _ref.showPreview,
    showPreview = _ref$showPreview === void 0 ? false : _ref$showPreview,
    testId = _ref.testId,
    title = _ref.title,
    url = _ref.url;
  return /*#__PURE__*/React.createElement(_FlexibleCard.default, {
    appearance: "block",
    cardState: cardState,
    onAuthorize: onAuthorize,
    onClick: onClick,
    onError: onError,
    origin: "smartLinkCard",
    testId: testId,
    ui: _utils.FlexibleCardUiOptions,
    url: url
  }, /*#__PURE__*/React.createElement(_blocks.TitleBlock, (0, _extends2.default)({}, _utils.titleBlockOptions, {
    hideIcon: !!title,
    text: title,
    status: cardState.status,
    className: (0, _runtime.ax)([(0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v1') ? "_zulpu2gc _13mh1pd9 _2yhz14q2" : "_zulp1f3t _13mh1pd9"])
  })), /*#__PURE__*/React.createElement(_blocks.CustomBlock, {
    status: cardState.status,
    className: (0, _runtime.ax)([(0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v1') ? "_zulp1b66 _1e0c1txw _4cvr1y6m _1wpz1fhb _7yjtidpf _ngc01fxt" : "_4cvr1y6m"])
  }, children), showPreview && /*#__PURE__*/React.createElement(_blocks.PreviewBlock, (0, _extends2.default)({}, _utils.PreviewBlockOptions, {
    status: cardState.status
  })), /*#__PURE__*/React.createElement(_blocks.InternalFooterBlock, {
    actions: actions,
    testId: "smart-block-card-footer",
    status: cardState.status,
    className: (0, _runtime.ax)([(0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v1') ? "_4t3i1k8s _1e0c1txw _1bah1yb4 _4cvresu3 _1wpz1fhb" : "_4t3i1k8s _g0eioxwj _1qryoxwj _sxbroxwj _1gyioxwj"])
  }));
};
var _default = exports.default = UnresolvedView;
