// @atlaskit/editor-plugin-context-panel/dist/cjs/contextPanelPlugin.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.pluginKey = exports.contextPanelPlugin = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _safePlugin = require("@atlaskit/editor-common/safe-plugin");
var _state = require("@atlaskit/editor-prosemirror/state");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _transforms = require("./pm-plugins/transforms");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var pluginKey = exports.pluginKey = new _state.PluginKey('contextPanelPluginKey');
function contextPanelPluginFactory(contextPanels, dispatch) {
  return new _safePlugin.SafePlugin({
    key: pluginKey,
    state: {
      init: function init(_config, state) {
        return {
          handlers: contextPanels,
          contents: contextPanels.map(function (panelContent) {
            return panelContent(state);
          })
        };
      },
      apply: function apply(tr, pluginState, _oldState, newState) {
        var newPluginState = pluginState;
        var meta = tr.getMeta(pluginKey);
        if (tr.docChanged || tr.selectionSet || meta && meta.changed) {
          var newContents = pluginState.handlers.map(function (panelContent) {
            return panelContent(newState);
          });
          var contentsLengthChanged = newContents.length !== newPluginState.contents.length;
          var contentChanged = newContents.some(function (node, index) {
            return newPluginState.contents[index] !== node;
          });
          if (contentsLengthChanged || contentChanged) {
            newPluginState = _objectSpread(_objectSpread({}, newPluginState), {}, {
              contents: newContents
            });
          }
        }
        if (newPluginState !== pluginState) {
          dispatch(pluginKey, newPluginState);
        }
        return newPluginState;
      }
    }
  });
}

/**
 * Context panel plugin to be added to an `EditorPresetBuilder` and used with `ComposableEditor`
 * from `@atlaskit/editor-core`.
 */
var contextPanelPlugin = exports.contextPanelPlugin = function contextPanelPlugin(_ref) {
  var config = _ref.config;
  return {
    name: 'contextPanel',
    actions: {
      applyChange: _transforms.applyChange,
      showPanel: config !== null && config !== void 0 && config.objectSideBar.showPanel && (0, _platformFeatureFlags.fg)('platform_editor_ai_object_sidebar_injection') ? function (panel, behavior, panelWidth) {
        return config.objectSideBar.showPanel(panel, behavior, panelWidth);
      } : undefined,
      closePanel: config !== null && config !== void 0 && config.objectSideBar.closePanel && (0, _platformFeatureFlags.fg)('platform_editor_ai_object_sidebar_injection') ? function () {
        return config.objectSideBar.closePanel();
      } : undefined,
      closePanelById: config !== null && config !== void 0 && config.objectSideBar.closePanelById && (0, _platformFeatureFlags.fg)('platform_editor_ai_object_sidebar_injection') ? function (id) {
        return config.objectSideBar.closePanelById(id);
      } : undefined
    },
    getSharedState: function getSharedState(state) {
      var _pluginKey$getState;
      if (!state) {
        return undefined;
      }
      var _ref2 = (_pluginKey$getState = pluginKey.getState(state)) !== null && _pluginKey$getState !== void 0 ? _pluginKey$getState : {},
        contents = _ref2.contents;
      return {
        contents: contents
      };
    },
    pmPlugins: function pmPlugins() {
      var contextPanels = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
      return [{
        name: 'contextPanel',
        plugin: function plugin(_ref3) {
          var dispatch = _ref3.dispatch;
          return contextPanelPluginFactory(contextPanels.filter(Boolean), dispatch);
        }
      }];
    }
  };
};
