// @atlaskit/media-card/dist/cjs/card/ui/actionsBar/cardActions/styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.cardActionButtonStyles = exports.CardActionIconButtonVariant = void 0;
var _react = require("@emotion/react");
var _colors = require("@atlaskit/theme/colors");
var _mediaUi = require("@atlaskit/media-ui");
var _styles = require("../../styles");
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
var CardActionIconButtonVariant = exports.CardActionIconButtonVariant = /*#__PURE__*/function (CardActionIconButtonVariant) {
  CardActionIconButtonVariant["default"] = "default";
  CardActionIconButtonVariant["filled"] = "filled";
  return CardActionIconButtonVariant;
}({});
var getVariantStyles = function getVariantStyles(variant) {
  return variant === 'filled' ? "\n    background-color: ".concat("var(--ds-surface-overlay, ".concat((0, _styles.rgba)(_colors.N0, 0.8), ")"), ";\n\n    &:hover {\n      background-color: ", "var(--ds-surface-overlay-hovered, ".concat((0, _styles.rgba)(_colors.N0, 0.6), ")"), "\n    }\n  ") : "\n    &:hover {\n      background-color: ".concat("var(--ds-background-neutral-subtle-hovered, rgba(9, 30, 66, 0.06))", ";\n    }\n  ");
};
var cardActionButtonStyles = exports.cardActionButtonStyles = function cardActionButtonStyles(_ref) {
  var variant = _ref.variant;
  return (0, _react.css)({
    appearance: 'none',
    border: 'none'
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  _mediaUi.center,
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  _mediaUi.borderRadius,
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  (0, _mediaUi.size)(26), {
    color: "var(--ds-icon, ".concat(_colors.N500, ")"),
    '&:hover': {
      cursor: 'pointer'
    }
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  getVariantStyles(variant));
};
