// @atlaskit/react-select/dist/cjs/emotion/components/internal/a11y-text.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = require("@emotion/react");
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled

// Assistive text to describe visual elements. Hidden for sighted users.
var styles = (0, _react.css)({
  width: 1,
  height: 1,
  padding: 0,
  position: 'absolute',
  zIndex: 9999,
  border: 0,
  clip: 'rect(1px, 1px, 1px, 1px)',
  label: 'a11yText',
  overflow: 'hidden',
  userSelect: 'none',
  // while hidden text is sitting in the DOM, it should not be selectable
  whiteSpace: 'nowrap'
});
var A11yText = function A11yText(props) {
  return (0, _react.jsx)("span", (0, _extends2.default)({
    css: styles
    // eslint-disable-next-line @repo/internal/react/no-unsafe-spread-props
  }, props));
};
var _default = exports.default = A11yText;
