// @atlaskit/react-select/dist/cjs/emotion/components/internal/dummy-input.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = DummyInput;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = require("@emotion/react");
var _utils = require("../../../utils");
var _excluded = ["innerRef"];
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled
var dummyInputStyles = (0, _react.css)({
  width: 1,
  padding: 0,
  position: 'relative',
  background: 0,
  border: 0,
  caretColor: 'transparent',
  color: 'transparent',
  // eslint-disable-next-line @atlaskit/design-system/use-tokens-typography
  fontSize: 'inherit',
  gridArea: '1 / 1 / 2 / 3',
  insetInlineStart: -100,
  label: 'dummyInput',
  opacity: 0,
  outline: 0,
  transform: 'scale(.01)'
});
function DummyInput(_ref) {
  var innerRef = _ref.innerRef,
    props = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  // Remove animation props not meant for HTML elements
  var filteredProps = (0, _utils.removeProps)(props, 'onExited', 'in', 'enter', 'exit', 'appear');
  return (0, _react.jsx)("input", (0, _extends2.default)({
    ref: innerRef
  }, filteredProps, {
    css: dummyInputStyles
  }));
}
