// @atlaskit/react-select/dist/cjs/emotion/components/internal/required-input.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("@emotion/react");
var _noop = _interopRequireDefault(require("@atlaskit/ds-lib/noop"));
/**
 * @jsxRuntime classic
 * @jsx jsx
 */

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled

var styles = (0, _react.css)({
  width: '100%',
  position: 'absolute',
  insetBlockEnd: 0,
  insetInlineEnd: 0,
  insetInlineStart: 0,
  label: 'requiredInput',
  opacity: 0,
  pointerEvents: 'none'
});
var RequiredInput = function RequiredInput(_ref) {
  var name = _ref.name,
    onFocus = _ref.onFocus;
  return (0, _react.jsx)("input", {
    required: true,
    name: name,
    tabIndex: -1,
    "aria-hidden": "true",
    onFocus: onFocus,
    css: styles
    // Prevent `Switching from uncontrolled to controlled` error
    ,
    value: "",
    onChange: _noop.default
  });
};

// eslint-disable-next-line @repo/internal/react/require-jsdoc
var _default = exports.default = RequiredInput;
