// @atlaskit/react-select/dist/cjs/emotion/components/internal/scroll-manager.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ScrollManager;
var _react = require("react");
var _react2 = require("@emotion/react");
var _useScrollCapture = _interopRequireDefault(require("./use-scroll-capture"));
var _useScrollLock = _interopRequireDefault(require("./use-scroll-lock"));
/**
 * @jsxRuntime classic
 * @jsx jsx
 */

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled

var styles = (0, _react2.css)({
  position: 'fixed',
  insetBlockEnd: 0,
  insetBlockStart: 0,
  insetInlineEnd: 0,
  insetInlineStart: 0
});
var blurSelectInput = function blurSelectInput(event) {
  var element = event.target;
  return element.ownerDocument.activeElement && element.ownerDocument.activeElement.blur();
};
function ScrollManager(_ref) {
  var children = _ref.children,
    lockEnabled = _ref.lockEnabled,
    _ref$captureEnabled = _ref.captureEnabled,
    captureEnabled = _ref$captureEnabled === void 0 ? true : _ref$captureEnabled,
    onBottomArrive = _ref.onBottomArrive,
    onBottomLeave = _ref.onBottomLeave,
    onTopArrive = _ref.onTopArrive,
    onTopLeave = _ref.onTopLeave;
  var setScrollCaptureTarget = (0, _useScrollCapture.default)({
    isEnabled: captureEnabled,
    onBottomArrive: onBottomArrive,
    onBottomLeave: onBottomLeave,
    onTopArrive: onTopArrive,
    onTopLeave: onTopLeave
  });
  var setScrollLockTarget = (0, _useScrollLock.default)({
    isEnabled: lockEnabled
  });
  var targetRef = function targetRef(element) {
    setScrollCaptureTarget(element);
    setScrollLockTarget(element);
  };
  return (0, _react2.jsx)(_react.Fragment, null, lockEnabled &&
  // eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions
  (0, _react2.jsx)("div", {
    onClick: blurSelectInput,
    css: styles
  }), children(targetRef));
}
