// @atlaskit/media-picker/dist/cjs/util/getPreviewFromImage.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SCALE_FACTOR_DEFAULT = void 0;
exports.getPreviewFromImage = getPreviewFromImage;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _mediaUi = require("@atlaskit/media-ui");
function getPreviewFromImage(_x, _x2) {
  return _getPreviewFromImage.apply(this, arguments);
}
function _getPreviewFromImage() {
  _getPreviewFromImage = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(file, devicePixelRatio) {
    var fileInfo, imageInfo, width, height, scaleFactor, preview;
    return _regenerator.default.wrap(function _callee$(_context) {
      while (1) switch (_context.prev = _context.next) {
        case 0:
          _context.next = 2;
          return (0, _mediaUi.getFileInfo)(file);
        case 2:
          fileInfo = _context.sent;
          _context.next = 5;
          return (0, _mediaUi.getImageInfo)(fileInfo);
        case 5:
          imageInfo = _context.sent;
          if (!(imageInfo === null)) {
            _context.next = 10;
            break;
          }
          return _context.abrupt("return", {
            file: file
          });
        case 10:
          width = imageInfo.width, height = imageInfo.height, scaleFactor = imageInfo.scaleFactor;
          preview = {
            file: file,
            dimensions: {
              width: width,
              height: height
            },
            scaleFactor: devicePixelRatio || scaleFactor
          };
          return _context.abrupt("return", preview);
        case 13:
        case "end":
          return _context.stop();
      }
    }, _callee);
  }));
  return _getPreviewFromImage.apply(this, arguments);
}
var SCALE_FACTOR_DEFAULT = exports.SCALE_FACTOR_DEFAULT = 1;
