// @atlaskit/editor-plugin-insert-block/dist/cjs/ui/ToolbarInsertBlock/styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.triggerWrapper = exports.expandWrapperStyle = void 0;
var _react = require("@emotion/react");
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var triggerWrapper = exports.triggerWrapper = (0, _react.css)({
  width: '42px',
  display: 'flex',
  alignItems: 'center',
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '> div, > span': {
    display: 'flex'
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '> div > div': {
    display: 'flex'
  }
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var expandWrapperStyle = exports.expandWrapperStyle = (0, _react.css)({
  marginLeft: "var(--ds-space-050, 4px)"
});
