// @atlaskit/editor-plugin-width/dist/cjs/widthPlugin.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.widthPlugin = void 0;
var _main = require("./pm-plugins/main");
var _pluginKey = require("./pm-plugins/plugin-key");
var _useResizeWidthObserver = require("./ui/hooks/useResizeWidthObserver");
/**
 * Width plugin to be added to an `EditorPresetBuilder` and used with `ComposableEditor`
 * from `@atlaskit/editor-core`.
 */
var widthPlugin = exports.widthPlugin = function widthPlugin() {
  return {
    name: 'width',
    pmPlugins: function pmPlugins() {
      return [{
        name: 'width',
        plugin: function plugin(_ref) {
          var dispatch = _ref.dispatch;
          return (0, _main.createPlugin)(dispatch);
        }
      }];
    },
    getSharedState: function getSharedState(editorState) {
      if (!editorState) {
        return undefined;
      }
      return _pluginKey.pluginKey.getState(editorState);
    },
    usePluginHook: function usePluginHook(_ref2) {
      var editorView = _ref2.editorView,
        containerElement = _ref2.containerElement;
      return (0, _useResizeWidthObserver.useResizeWidthObserver)({
        editorView: editorView,
        containerElement: containerElement
      });
    }
  };
};
