// @atlaskit/editor-element-browser/dist/cjs/ui/ExpandablePanel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ExpandablePanel = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _menu = require("@atlaskit/menu");
var _primitives = require("@atlaskit/primitives");
var _ListButtonItem = require("./ListButtonItem");
var _NavigationButton = require("./NavigationButton");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var headingContainerStyles = (0, _primitives.xcss)({
  paddingTop: 'space.100',
  paddingBottom: 'space.075',
  paddingLeft: 'space.300',
  paddingRight: 'space.300'
});
var itemsContainerStyles = (0, _primitives.xcss)({
  paddingTop: 'space.0',
  paddingBottom: 'space.150',
  paddingLeft: 'space.200',
  paddingRight: 'space.200'
});
var ExpandablePanel = exports.ExpandablePanel = /*#__PURE__*/(0, _react.memo)(function (_ref) {
  var id = _ref.id,
    label = _ref.label,
    items = _ref.items,
    hasSeparator = _ref.hasSeparator,
    attributes = _ref.attributes,
    _onItemSelected = _ref.onItemSelected,
    onViewAllSelected = _ref.onViewAllSelected;
  var _useState = (0, _react.useState)(true),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    isExpanded = _useState2[0],
    setIsExpanded = _useState2[1];

  // Dirty fix for AI label
  var shortAIlabel = label === 'Atlassian Intelligence';
  return /*#__PURE__*/_react.default.createElement(_menu.Section, {
    hasSeparator: hasSeparator
  }, label && /*#__PURE__*/_react.default.createElement(_primitives.Box, {
    xcss: [headingContainerStyles]
  }, /*#__PURE__*/_react.default.createElement(_NavigationButton.ExpandableNavButton, {
    id: id,
    label: label,
    isExpanded: isExpanded,
    attributes: attributes,
    onClick: function onClick() {
      return setIsExpanded(!isExpanded);
    }
  })), isExpanded && /*#__PURE__*/_react.default.createElement(_primitives.Stack, {
    space: "space.025",
    xcss: [itemsContainerStyles]
  }, items.map(function (item) {
    return /*#__PURE__*/_react.default.createElement(_ListButtonItem.ListButtonItem, {
      key: item.index,
      index: item.index,
      title: item.title,
      description: item.description,
      showDescription: item.showDescription,
      attributes: item.attributes,
      keyshortcut: item.keyshortcut,
      renderIcon: item.renderIcon,
      onItemSelected: function onItemSelected(index) {
        return _onItemSelected === null || _onItemSelected === void 0 ? void 0 : _onItemSelected(index, id);
      }
    });
  }), /*#__PURE__*/_react.default.createElement(_ListButtonItem.ViewAllButtonItem, {
    label: "View all ".concat(shortAIlabel ? 'AI' : label.toLocaleLowerCase(), " options"),
    onClick: function onClick() {
      return onViewAllSelected === null || onViewAllSelected === void 0 ? void 0 : onViewAllSelected(id);
    }
  })));
});
