// @atlaskit/editor-element-browser/dist/cjs/ui/QuickInsertPanel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.QuickInsertPanel = void 0;
var _react = _interopRequireWildcard(require("react"));
var _typeAhead = require("@atlaskit/editor-common/type-ahead");
var _section = _interopRequireDefault(require("@atlaskit/menu/section"));
var _primitives = require("@atlaskit/primitives");
var _useItems2 = require("../utils/use-items");
var _ListButtonGroup = require("./ListButtonGroup");
var _ListButtonItem = require("./ListButtonItem");
var _NavigationButton = require("./NavigationButton");
var _SubPanel = require("./SubPanel");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var navButtonContainerStyles = (0, _primitives.xcss)({
  paddingTop: 'space.0',
  paddingBottom: 'space.0',
  paddingLeft: 'space.200',
  paddingRight: 'space.200'
});
var DefaultView = function DefaultView(_ref) {
  var suggested = _ref.suggested,
    itemsRegistry = _ref.itemsRegistry,
    categoryRegistry = _ref.categoryRegistry,
    onItemSelected = _ref.onItemSelected,
    onCategorySelected = _ref.onCategorySelected,
    setSelectedItem = _ref.setSelectedItem;
  return /*#__PURE__*/_react.default.createElement(_primitives.Stack, null, suggested && /*#__PURE__*/_react.default.createElement(_ListButtonGroup.ListButtonGroupWithHeading, {
    id: suggested.id,
    items: suggested.items.slice(0, 6),
    label: suggested.label,
    onItemSelected: onItemSelected,
    setSelectedItem: setSelectedItem,
    startingIndex: 0
  }), _useItems2.CategoriesStructure.map(function (category) {
    var defautltItems = category.defautltItems;
    var items = defautltItems.length > 0 ? defautltItems.map(function (itemIdOrTitle) {
      return itemsRegistry[itemIdOrTitle];
    }).filter(function (item) {
      return item;
    }) : categoryRegistry[category.id] ? categoryRegistry[category.id].slice(0, 5) : [];
    return items.length > 0 && /*#__PURE__*/_react.default.createElement(_section.default, {
      key: category.id,
      hasSeparator: true
    }, /*#__PURE__*/_react.default.createElement(_primitives.Box, {
      xcss: navButtonContainerStyles
    }, /*#__PURE__*/_react.default.createElement(_NavigationButton.LinkNavButton, {
      id: category.id,
      label: category.title,
      onClick: onCategorySelected
    })));
  }));
};
var CategoryView = function CategoryView(_ref2) {
  var categoryRegistry = _ref2.categoryRegistry,
    selectedCategory = _ref2.selectedCategory,
    onItemSelected = _ref2.onItemSelected,
    onBackButtonClicked = _ref2.onBackButtonClicked;
  var categoryData = _useItems2.PredefinedCategories.get(selectedCategory);
  var categoryLevelItems = categoryRegistry[selectedCategory] || [];
  var startingIndex = categoryLevelItems.length;
  return /*#__PURE__*/_react.default.createElement(_primitives.Stack, null, /*#__PURE__*/_react.default.createElement(_SubPanel.SubPanelWithBackButton, {
    label: categoryData.title,
    buttonLabel: "Back to all items",
    onClick: onBackButtonClicked
  }, categoryLevelItems.length > 0 && /*#__PURE__*/_react.default.createElement(_ListButtonGroup.ListButtonGroup, {
    id: categoryData.id,
    items: categoryLevelItems,
    onItemSelected: onItemSelected,
    startingIndex: 0
  }), categoryData && categoryData.subcategories.map(function (_ref3) {
    var id = _ref3.id,
      title = _ref3.title;
    if (categoryRegistry[id]) {
      var list = /*#__PURE__*/_react.default.createElement(_ListButtonGroup.ListButtonGroupWithHeading, {
        id: id,
        label: title,
        items: categoryRegistry[id],
        onItemSelected: onItemSelected,
        startingIndex: startingIndex
      });
      startingIndex += categoryRegistry[id].length;
      return list;
    }
  })));
};
var SearchView = function SearchView(_ref4) {
  var items = _ref4.items,
    onItemSelected = _ref4.onItemSelected,
    setSelectedItem = _ref4.setSelectedItem;
  return /*#__PURE__*/_react.default.createElement(_ListButtonGroup.ListButtonGroup, {
    id: 'search',
    items: items,
    onItemSelected: onItemSelected,
    setSelectedItem: setSelectedItem,
    startingIndex: 0
  });
};
var EmptySearchView = function EmptySearchView() {
  return /*#__PURE__*/_react.default.createElement(_primitives.Box, {
    padding: "space.300"
  }, /*#__PURE__*/_react.default.createElement(_primitives.Stack, {
    space: "space.0"
  }, /*#__PURE__*/_react.default.createElement(_primitives.Text, {
    align: "center",
    as: "p"
  }, "We couldn't find any results."), /*#__PURE__*/_react.default.createElement(_primitives.Text, {
    align: "center",
    as: "p"
  }, "Select ", /*#__PURE__*/_react.default.createElement(_primitives.Text, {
    weight: "medium"
  }, "View all"), " to browser inserts.")));
};
var QuickInsertPanel = exports.QuickInsertPanel = /*#__PURE__*/(0, _react.memo)(function (_ref5) {
  var items = _ref5.items,
    onItemInsert = _ref5.onItemInsert,
    onViewAllItemsClick = _ref5.onViewAllItemsClick,
    query = _ref5.query,
    setSelectedItem = _ref5.setSelectedItem;
  var onItemSelect = (0, _react.useCallback)(function (index) {
    onItemInsert(_typeAhead.SelectItemMode.SELECTED, index);
  }, [onItemInsert]);
  var _useItems = (0, _useItems2.useItems)(items, query),
    suggested = _useItems.suggested,
    categoryRegistry = _useItems.categoryRegistry,
    itemsRegistry = _useItems.itemsRegistry,
    selectedCategory = _useItems.selectedCategory,
    searchItems = _useItems.searchItems,
    setSelectedCategory = _useItems.setSelectedCategory;
  var showViewAllItems = Boolean(onViewAllItemsClick);
  var view = searchItems ? searchItems.length > 0 ? /*#__PURE__*/_react.default.createElement(SearchView, {
    items: searchItems,
    onItemSelected: onItemSelect,
    setSelectedItem: setSelectedItem
  }) : /*#__PURE__*/_react.default.createElement(EmptySearchView, null) : selectedCategory ? /*#__PURE__*/_react.default.createElement(CategoryView, {
    categoryRegistry: categoryRegistry,
    selectedCategory: selectedCategory,
    onItemSelected: onItemSelect,
    onBackButtonClicked: function onBackButtonClicked() {
      return setSelectedCategory(undefined);
    }
  }) : /*#__PURE__*/_react.default.createElement(DefaultView, {
    suggested: suggested,
    itemsRegistry: itemsRegistry,
    categoryRegistry: categoryRegistry,
    onItemSelected: onItemSelect,
    onCategorySelected: function onCategorySelected(categoryId) {
      setSelectedCategory(categoryId);
    },
    setSelectedItem: setSelectedItem
  });
  if (showViewAllItems) {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, view, showViewAllItems && /*#__PURE__*/_react.default.createElement(_ListButtonItem.ViewAllButtonItem, {
      isViewAllInserts: true,
      label: 'View all inserts' // TODO: ED-26961 - add i18n
      ,
      onClick: onViewAllItemsClick ? onViewAllItemsClick : function () {}
    }));
  } else {
    return view;
  }
});
