// @atlaskit/editor-element-browser/dist/cjs/ui/SearchPanel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SearchPanel = void 0;
var _react = _interopRequireWildcard(require("react"));
var _search = _interopRequireDefault(require("@atlaskit/icon/core/search"));
var _primitives = require("@atlaskit/primitives");
var _textfield = _interopRequireDefault(require("@atlaskit/textfield"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var searchPanelStyles = (0, _primitives.xcss)({
  paddingTop: 'space.025',
  paddingBottom: 'space.150',
  paddingLeft: 'space.300',
  paddingRight: 'space.300'
});
var searchIconStyles = (0, _primitives.xcss)({
  paddingTop: 'space.050',
  paddingBottom: 'space.0',
  paddingLeft: 'space.100',
  paddingRight: 'space.0'
});
var SearchPanel = exports.SearchPanel = /*#__PURE__*/(0, _react.memo)(function (_ref) {
  var onChange = _ref.onChange;
  return /*#__PURE__*/_react.default.createElement(_primitives.Box, {
    xcss: [searchPanelStyles]
  }, /*#__PURE__*/_react.default.createElement(_textfield.default, {
    name: "search",
    id: "search-textfield",
    placeholder: "Find or describe...",
    elemBeforeInput: /*#__PURE__*/_react.default.createElement(_primitives.Box, {
      xcss: searchIconStyles
    }, /*#__PURE__*/_react.default.createElement(_search.default, {
      label: "search",
      color: "var(--ds-text-subtlest, #626F86)"
    })),
    onChange: onChange
  }));
});
