// @atlaskit/editor-element-browser/dist/cjs/ui/SideInsertPanel.js
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SideInsertPanel = void 0;
var _react = _interopRequireWildcard(require("react"));
var _typeAhead = require("@atlaskit/editor-common/type-ahead");
var _compiled = require("@atlaskit/primitives/compiled");
var _useItems2 = require("../utils/use-items");
var _ExpandablePanel = require("./ExpandablePanel");
var _IconButtonGroup = require("./IconButtonGroup");
var _ListButtonGroup = require("./ListButtonGroup");
var _SearchPanel = require("./SearchPanel");
var _SubPanel = require("./SubPanel");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var DefaultView = function DefaultView(_ref) {
  var suggested = _ref.suggested,
    itemsRegistry = _ref.itemsRegistry,
    categoryRegistry = _ref.categoryRegistry,
    onItemSelected = _ref.onItemSelected,
    onCategorySelected = _ref.onCategorySelected;
  return /*#__PURE__*/_react.default.createElement(_compiled.Stack, null, suggested && /*#__PURE__*/_react.default.createElement(_IconButtonGroup.IconButtonGroup, {
    id: suggested.id,
    items: suggested.items,
    label: suggested.label,
    onItemSelected: onItemSelected
  }), _useItems2.CategoriesStructure.map(function (category, i) {
    var defautltItems = category.defautltItems;
    var items = defautltItems.length > 0 ? defautltItems.map(function (itemIdOrTitle) {
      return itemsRegistry[itemIdOrTitle];
    }).filter(function (item) {
      return item;
    }) : categoryRegistry[category.id] ? categoryRegistry[category.id].slice(0, 5) : [];
    var shoudldShowSeparator = i === 0 && !suggested ? false : true;
    return items.length > 0 && /*#__PURE__*/_react.default.createElement(_ExpandablePanel.ExpandablePanel, {
      key: category.id,
      id: category.id,
      label: category.title,
      items: items,
      hasSeparator: shoudldShowSeparator,
      onItemSelected: onItemSelected,
      onViewAllSelected: onCategorySelected
    });
  }));
};
var CategoryView = function CategoryView(_ref2) {
  var categoryRegistry = _ref2.categoryRegistry,
    selectedCategory = _ref2.selectedCategory,
    onItemSelected = _ref2.onItemSelected,
    onBackButtonClicked = _ref2.onBackButtonClicked;
  var categoryData = _useItems2.PredefinedCategories.get(selectedCategory);
  var categoryLevelItems = categoryRegistry[selectedCategory] || [];
  return /*#__PURE__*/_react.default.createElement(_compiled.Stack, null, /*#__PURE__*/_react.default.createElement(_SubPanel.SubPanelWithBackButton, {
    label: categoryData.title,
    buttonLabel: "Back to all items",
    onClick: onBackButtonClicked
  }, categoryLevelItems.length > 0 && /*#__PURE__*/_react.default.createElement(_ListButtonGroup.ListButtonGroup, {
    id: categoryData.id,
    items: categoryLevelItems,
    onItemSelected: onItemSelected
  }), categoryData && categoryData.subcategories.map(function (_ref3) {
    var id = _ref3.id,
      title = _ref3.title;
    return categoryRegistry[id] && /*#__PURE__*/_react.default.createElement(_ListButtonGroup.ListButtonGroupWithHeading, {
      id: id,
      label: title,
      items: categoryRegistry[id],
      onItemSelected: onItemSelected
    });
  })));
};
var SearchView = function SearchView(_ref4) {
  var items = _ref4.items,
    onItemSelected = _ref4.onItemSelected;
  return /*#__PURE__*/_react.default.createElement(_ListButtonGroup.ListButtonGroup, {
    id: 'search',
    items: items,
    onItemSelected: onItemSelected
  });
};
var SideInsertPanel = exports.SideInsertPanel = function SideInsertPanel(_ref5) {
  var items = _ref5.items,
    onItemInsert = _ref5.onItemInsert;
  var onItemSelect = (0, _react.useCallback)(function (index) {
    onItemInsert(_typeAhead.SelectItemMode.SELECTED, index);
  }, [onItemInsert]);
  var _useItems = (0, _useItems2.useItems)(items),
    categoryRegistry = _useItems.categoryRegistry,
    itemsRegistry = _useItems.itemsRegistry,
    selectedCategory = _useItems.selectedCategory,
    searchItems = _useItems.searchItems,
    setSearchText = _useItems.setSearchText,
    setSelectedCategory = _useItems.setSelectedCategory;
  var view = searchItems ? /*#__PURE__*/_react.default.createElement(SearchView, {
    items: searchItems,
    onItemSelected: onItemSelect
  }) : selectedCategory ? /*#__PURE__*/_react.default.createElement(CategoryView, {
    categoryRegistry: categoryRegistry,
    selectedCategory: selectedCategory,
    onItemSelected: onItemSelect,
    onBackButtonClicked: function onBackButtonClicked() {
      return setSelectedCategory(undefined);
    }
  }) : /*#__PURE__*/_react.default.createElement(DefaultView, {
    suggested: undefined,
    itemsRegistry: itemsRegistry,
    categoryRegistry: categoryRegistry,
    onItemSelected: onItemSelect,
    onCategorySelected: function onCategorySelected(categoryId) {
      setSelectedCategory(categoryId);
    }
  });
  return /*#__PURE__*/_react.default.createElement(_compiled.Stack, null, /*#__PURE__*/_react.default.createElement(_SearchPanel.SearchPanel, {
    onChange: function onChange(event) {
      setSearchText(event.target.value);
    }
  }), view);
};
