// @atlaskit/link-datasource/dist/cjs/ui/confluence-search-modal/basic-filters/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./index.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireDefault(require("react"));
var _compiled = require("@atlaskit/primitives/compiled");
var _dateRangePicker = require("./filters/date-range-picker");
var _editedOrCreatedBy = _interopRequireDefault(require("./filters/edited-or-created-by"));
var _types = require("./types");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var styles = {
  basicFilterContainerStyles: "_19bvu2gc"
};
var BasicFilterContainer = function BasicFilterContainer(_ref) {
  var cloudId = _ref.cloudId,
    onChange = _ref.onChange,
    _ref$selections = _ref.selections,
    selections = _ref$selections === void 0 ? {} : _ref$selections,
    isHydrating = _ref.isHydrating;
  var lastModified = selections.lastModified,
    editedOrCreatedBy = selections.editedOrCreatedBy;
  var _ref2 = lastModified || [],
    _ref3 = (0, _slicedToArray2.default)(_ref2, 1),
    lastModifiedValue = _ref3[0];
  var lastModifiedSelection = (lastModifiedValue === null || lastModifiedValue === void 0 ? void 0 : lastModifiedValue.optionType) === 'dateRange' ? lastModifiedValue : undefined;
  var onDateRangePickerChange = function onDateRangePickerChange(updatedOption) {
    onChange(_types.CLOLBasicFilters.lastModified, _objectSpread({
      optionType: 'dateRange',
      label: updatedOption.value
    }, updatedOption));
  };
  return /*#__PURE__*/_react.default.createElement(_compiled.Flex, {
    xcss: styles.basicFilterContainerStyles,
    gap: "space.100",
    testId: "clol-basic-filter-container"
  }, /*#__PURE__*/_react.default.createElement(_editedOrCreatedBy.default, {
    cloudId: cloudId,
    onSelectionChange: onChange,
    selection: editedOrCreatedBy || [],
    isHydrating: isHydrating
  }), /*#__PURE__*/_react.default.createElement(_dateRangePicker.DateRangePicker, {
    selection: lastModifiedSelection,
    onSelectionChange: onDateRangePickerChange,
    filterName: "clol-basic-filter-".concat(_types.CLOLBasicFilters.lastModified)
  }));
};
var _default = exports.default = BasicFilterContainer;
