// @atlaskit/editor-plugin-media/dist/cjs/ui/MediaPicker/ClipboardWrapper.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ClipboardWrapper = void 0;
var _react = _interopRequireDefault(require("react"));
var _mediaPicker = require("@atlaskit/media-picker");
var _PickerFacadeProvider = _interopRequireDefault(require("./PickerFacadeProvider"));
var ClipboardWrapper = exports.ClipboardWrapper = function ClipboardWrapper(_ref) {
  var mediaState = _ref.mediaState,
    featureFlags = _ref.featureFlags,
    container = _ref.container;
  return /*#__PURE__*/_react.default.createElement(_PickerFacadeProvider.default, {
    mediaState: mediaState,
    analyticsName: "clipboard"
  }, function (_ref2) {
    var mediaClientConfig = _ref2.mediaClientConfig,
      config = _ref2.config,
      pickerFacadeInstance = _ref2.pickerFacadeInstance;
    var clipboardConfig = Object.assign({}, config);
    clipboardConfig.container = container;
    clipboardConfig.onPaste = function (event) {
      event.stopPropagation();
      return false;
    };
    return /*#__PURE__*/_react.default.createElement(_mediaPicker.Clipboard, {
      mediaClientConfig: mediaClientConfig,
      config: clipboardConfig,
      onError: pickerFacadeInstance.handleUploadError,
      onPreviewUpdate: pickerFacadeInstance.handleUploadPreviewUpdate,
      onEnd: pickerFacadeInstance.handleReady,
      featureFlags: featureFlags
    });
  });
};
