// @atlaskit/editor-plugin-media/dist/cjs/ui/MediaPicker/DropzoneWrapper.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DropzoneWrapper = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireDefault(require("react"));
var _ui = require("@atlaskit/editor-common/ui");
var _mediaPicker = require("@atlaskit/media-picker");
var _PickerFacadeProvider = _interopRequireDefault(require("./PickerFacadeProvider"));
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var DropzoneWrapper = exports.DropzoneWrapper = function DropzoneWrapper(_ref) {
  var mediaState = _ref.mediaState,
    isActive = _ref.isActive,
    featureFlags = _ref.featureFlags,
    editorDomElement = _ref.editorDomElement,
    appearance = _ref.appearance;
  return /*#__PURE__*/_react.default.createElement(_PickerFacadeProvider.default, {
    mediaState: mediaState,
    analyticsName: "dropzone"
  }, function (_ref2) {
    var mediaClientConfig = _ref2.mediaClientConfig,
      config = _ref2.config,
      pickerFacadeInstance = _ref2.pickerFacadeInstance;
    var customDropzoneContainer = mediaState.options.customDropzoneContainer,
      handleDrag = mediaState.handleDrag;
    // Ignored via go/ees005
    // eslint-disable-next-line @atlaskit/editor/no-as-casting
    var editorHtmlElement = editorDomElement;
    var scrollParent = appearance === 'full-page' && (0, _ui.findOverflowScrollParent)(editorHtmlElement);
    var container = customDropzoneContainer || (scrollParent ? scrollParent : editorHtmlElement);
    var dropzoneConfig = _objectSpread(_objectSpread({}, config), {}, {
      container: container
    });
    return isActive ? /*#__PURE__*/_react.default.createElement(_mediaPicker.Dropzone, {
      mediaClientConfig: mediaClientConfig,
      config: dropzoneConfig,
      onError: pickerFacadeInstance.handleUploadError,
      onPreviewUpdate: pickerFacadeInstance.handleUploadPreviewUpdate,
      onEnd: pickerFacadeInstance.handleReady,
      onDragEnter: function onDragEnter() {
        return handleDrag('enter');
      },
      onDragLeave: function onDragLeave() {
        return handleDrag('leave');
      },
      featureFlags: featureFlags
    }) : null;
  });
};
