// @atlaskit/media-ui/dist/cjs/BlockCard/components/Provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/** @jsx jsx */
var core_1 = require("@emotion/core");
var colors_1 = require("@atlaskit/theme/colors");
var utils_1 = require("../utils");
exports.Provider = function (_a) {
    var name = _a.name, icon = _a.icon;
    var iconToRender = icon || null;
    if (typeof icon === 'string') {
        iconToRender = (core_1.jsx("img", { className: "smart-link-icon", css: { height: utils_1.gs(1.5), width: utils_1.gs(1.5) }, src: icon }));
    }
    return (core_1.jsx("div", { css: { display: 'flex', alignItems: 'center' } },
        iconToRender,
        core_1.jsx("span", { css: {
                fontSize: utils_1.gs(1.5),
                color: colors_1.N300,
                margin: 0,
                marginLeft: utils_1.gs(0.5),
                // EDM-713: fixes copy-paste from renderer to editor for Firefox
                // due to HTML its unwrapping behaviour on paste.
                MozUserSelect: 'none',
            } }, name)));
};
