// @atlaskit/jql-ast/dist/cjs/constants/operator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WAS_OPERATORS = exports.WAS_IN_OPERATORS = exports.OPERATOR_WAS_NOT_IN = exports.OPERATOR_WAS_NOT = exports.OPERATOR_WAS_IN = exports.OPERATOR_WAS = exports.OPERATOR_NOT_LIKE = exports.OPERATOR_NOT_IN = exports.OPERATOR_NOT_EQUALS = exports.OPERATOR_LT_EQUALS = exports.OPERATOR_LT = exports.OPERATOR_LIKE = exports.OPERATOR_IS_NOT = exports.OPERATOR_IS = exports.OPERATOR_IN = exports.OPERATOR_GT_EQUALS = exports.OPERATOR_GT = exports.OPERATOR_EQUALS = exports.OPERATOR_CHANGED = exports.OPERATORS = exports.LIST_OPERATORS = exports.LIKE_OPERATORS = exports.IS_OPERATORS = exports.IN_OPERATORS = exports.EQUALS_OPERATORS = exports.COMPARISON_OPERATORS = exports.CLAUSE_OPERATOR_NOT = exports.CHANGED_OPERATORS = void 0;
var CLAUSE_OPERATOR_NOT = exports.CLAUSE_OPERATOR_NOT = 'not';
var OPERATOR_EQUALS = exports.OPERATOR_EQUALS = '=';
var OPERATOR_NOT_EQUALS = exports.OPERATOR_NOT_EQUALS = '!=';
var OPERATOR_GT = exports.OPERATOR_GT = '>';
var OPERATOR_LT = exports.OPERATOR_LT = '<';
var OPERATOR_GT_EQUALS = exports.OPERATOR_GT_EQUALS = '>=';
var OPERATOR_LT_EQUALS = exports.OPERATOR_LT_EQUALS = '<=';
var OPERATOR_IN = exports.OPERATOR_IN = 'in';
var OPERATOR_NOT_IN = exports.OPERATOR_NOT_IN = 'not in';
var OPERATOR_LIKE = exports.OPERATOR_LIKE = '~';
var OPERATOR_NOT_LIKE = exports.OPERATOR_NOT_LIKE = '!~';
var OPERATOR_IS = exports.OPERATOR_IS = 'is';
var OPERATOR_IS_NOT = exports.OPERATOR_IS_NOT = 'is not';
var OPERATOR_WAS = exports.OPERATOR_WAS = 'was';
var OPERATOR_WAS_IN = exports.OPERATOR_WAS_IN = 'was in';
var OPERATOR_WAS_NOT_IN = exports.OPERATOR_WAS_NOT_IN = 'was not in';
var OPERATOR_WAS_NOT = exports.OPERATOR_WAS_NOT = 'was not';
var OPERATOR_CHANGED = exports.OPERATOR_CHANGED = 'changed';

// Operator groups

var EQUALS_OPERATORS = exports.EQUALS_OPERATORS = [OPERATOR_EQUALS, OPERATOR_NOT_EQUALS];
var LIKE_OPERATORS = exports.LIKE_OPERATORS = [OPERATOR_LIKE, OPERATOR_NOT_LIKE];
var COMPARISON_OPERATORS = exports.COMPARISON_OPERATORS = [OPERATOR_LT, OPERATOR_GT, OPERATOR_LT_EQUALS, OPERATOR_GT_EQUALS];
var IN_OPERATORS = exports.IN_OPERATORS = [OPERATOR_IN, OPERATOR_NOT_IN];
var IS_OPERATORS = exports.IS_OPERATORS = [OPERATOR_IS, OPERATOR_IS_NOT];
var WAS_OPERATORS = exports.WAS_OPERATORS = [OPERATOR_WAS, OPERATOR_WAS_NOT];
var WAS_IN_OPERATORS = exports.WAS_IN_OPERATORS = [OPERATOR_WAS_IN, OPERATOR_WAS_NOT_IN];
var CHANGED_OPERATORS = exports.CHANGED_OPERATORS = [OPERATOR_CHANGED];
var OPERATORS = exports.OPERATORS = [].concat(EQUALS_OPERATORS, LIKE_OPERATORS, COMPARISON_OPERATORS, IN_OPERATORS, IS_OPERATORS, WAS_OPERATORS, WAS_IN_OPERATORS, CHANGED_OPERATORS);
var LIST_OPERATORS = exports.LIST_OPERATORS = [OPERATOR_IN, OPERATOR_NOT_IN, OPERATOR_WAS_IN, OPERATOR_WAS_NOT_IN];
