// @atlaskit/jql-ast/dist/cjs/constants/predicate.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PREDICATE_OPERATOR_TO = exports.PREDICATE_OPERATOR_ON = exports.PREDICATE_OPERATOR_FROM = exports.PREDICATE_OPERATOR_DURING = exports.PREDICATE_OPERATOR_BY = exports.PREDICATE_OPERATOR_BEFORE = exports.PREDICATE_OPERATOR_AFTER = exports.PREDICATE_OPERATORS = void 0;
var PREDICATE_OPERATOR_AFTER = exports.PREDICATE_OPERATOR_AFTER = 'after';
var PREDICATE_OPERATOR_BEFORE = exports.PREDICATE_OPERATOR_BEFORE = 'before';
var PREDICATE_OPERATOR_BY = exports.PREDICATE_OPERATOR_BY = 'by';
var PREDICATE_OPERATOR_DURING = exports.PREDICATE_OPERATOR_DURING = 'during';
var PREDICATE_OPERATOR_FROM = exports.PREDICATE_OPERATOR_FROM = 'from';
var PREDICATE_OPERATOR_ON = exports.PREDICATE_OPERATOR_ON = 'on';
var PREDICATE_OPERATOR_TO = exports.PREDICATE_OPERATOR_TO = 'to';
var PREDICATE_OPERATORS = exports.PREDICATE_OPERATORS = [PREDICATE_OPERATOR_AFTER, PREDICATE_OPERATOR_BEFORE, PREDICATE_OPERATOR_BY, PREDICATE_OPERATOR_DURING, PREDICATE_OPERATOR_FROM, PREDICATE_OPERATOR_ON, PREDICATE_OPERATOR_TO];
