// @atlaskit/jql-ast/dist/cjs/constants/privacy-safe-fields.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WATCHER_PROPERTY = exports.VOTER_PROPERTY = exports.UPDATED_FIELD = exports.TYPE_FIELD = exports.TEXT_FIELD = exports.TEXFIELDS_FIELD = exports.SUMMARY_FIELD = exports.STATUS_FIELD = exports.STATUS_CATEGORY_FIELD = exports.SPRINT_FIELD = exports.RESOLVED_FIELD = exports.RESOLUTION_FIELD = exports.RESOLUTION_DATE_FIELD = exports.REPORTER_PROPERTY = exports.REPORTER_FIELD = exports.PROJECT_PROPERTY = exports.PROJECT_FIELD = exports.PRIVACY_SAFE_FIELDS = exports.PRIORITY_FIELD = exports.PARENT_FIELD = exports.LAST_VIEWED_FIELD = exports.LABELS_FIELD = exports.ISSUE_TYPE_FIELD = exports.ISSUE_PROPERTY = exports.ISSUE_KEY_FIELD = exports.FIX_VERSION_FIELD = exports.FILTER_FIELD = exports.EPIC_LINK_FIELD = exports.ENVIRONMENT_FIELD = exports.DUE_DATE_FIELD = exports.DEVELOPMENT_FIELD = exports.DESCRIPTION_FIELD = exports.CREATOR_PROPERTY = exports.CREATED_FIELD = exports.COMPONENT_FIELD = exports.CATEGORY_FIELD = exports.ASSIGNEE_PROPERTY = exports.ASSIGNEE_FIELD = exports.APPROVALS_FIELD = exports.AFFECTED_VERSION_FIELD = void 0;
var ASSIGNEE_FIELD = exports.ASSIGNEE_FIELD = 'assignee';
var CREATED_FIELD = exports.CREATED_FIELD = 'created';
var DUE_DATE_FIELD = exports.DUE_DATE_FIELD = 'duedate';
var ISSUE_KEY_FIELD = exports.ISSUE_KEY_FIELD = 'issuekey';
var ISSUE_TYPE_FIELD = exports.ISSUE_TYPE_FIELD = 'issuetype';
var LAST_VIEWED_FIELD = exports.LAST_VIEWED_FIELD = 'lastviewed';
var PRIORITY_FIELD = exports.PRIORITY_FIELD = 'priority';
var PROJECT_FIELD = exports.PROJECT_FIELD = 'project';
var REPORTER_FIELD = exports.REPORTER_FIELD = 'reporter';
var RESOLVED_FIELD = exports.RESOLVED_FIELD = 'resolved';
var RESOLUTION_DATE_FIELD = exports.RESOLUTION_DATE_FIELD = 'resolutiondate';
var STATUS_FIELD = exports.STATUS_FIELD = 'status';
var STATUS_CATEGORY_FIELD = exports.STATUS_CATEGORY_FIELD = 'statuscategory';
var SUMMARY_FIELD = exports.SUMMARY_FIELD = 'summary';
var TEXT_FIELD = exports.TEXT_FIELD = 'text';
var TEXFIELDS_FIELD = exports.TEXFIELDS_FIELD = 'textfields';
var TYPE_FIELD = exports.TYPE_FIELD = 'type';
var UPDATED_FIELD = exports.UPDATED_FIELD = 'updated';
var AFFECTED_VERSION_FIELD = exports.AFFECTED_VERSION_FIELD = 'affectedversion';
var APPROVALS_FIELD = exports.APPROVALS_FIELD = 'approvals';
var CATEGORY_FIELD = exports.CATEGORY_FIELD = 'category';
var COMPONENT_FIELD = exports.COMPONENT_FIELD = 'component';
var DESCRIPTION_FIELD = exports.DESCRIPTION_FIELD = 'description';
var DEVELOPMENT_FIELD = exports.DEVELOPMENT_FIELD = 'development';
var ENVIRONMENT_FIELD = exports.ENVIRONMENT_FIELD = 'environment';
var EPIC_LINK_FIELD = exports.EPIC_LINK_FIELD = 'epic link';
var FILTER_FIELD = exports.FILTER_FIELD = 'filter';
var FIX_VERSION_FIELD = exports.FIX_VERSION_FIELD = 'fixversion';
var LABELS_FIELD = exports.LABELS_FIELD = 'labels';
var PARENT_FIELD = exports.PARENT_FIELD = 'parent';
var RESOLUTION_FIELD = exports.RESOLUTION_FIELD = 'resolution';
var SPRINT_FIELD = exports.SPRINT_FIELD = 'sprint';
var ASSIGNEE_PROPERTY = exports.ASSIGNEE_PROPERTY = 'assignee.property';
var CREATOR_PROPERTY = exports.CREATOR_PROPERTY = 'creator.property';
var ISSUE_PROPERTY = exports.ISSUE_PROPERTY = 'issue.property';
var PROJECT_PROPERTY = exports.PROJECT_PROPERTY = 'project.property';
var REPORTER_PROPERTY = exports.REPORTER_PROPERTY = 'reporter.property';
var VOTER_PROPERTY = exports.VOTER_PROPERTY = 'voter.property';
var WATCHER_PROPERTY = exports.WATCHER_PROPERTY = 'watcher.property';

// List of Jira system fields that don't need to be anonymized. This list may not be exhaustive and is only used to
// provide extra insight when analysing JQL.
// All the field names here must be in lowercase
var PRIVACY_SAFE_FIELDS = exports.PRIVACY_SAFE_FIELDS = [ASSIGNEE_FIELD, CREATED_FIELD, DUE_DATE_FIELD, ISSUE_KEY_FIELD, ISSUE_TYPE_FIELD, LAST_VIEWED_FIELD, PRIORITY_FIELD, PROJECT_FIELD, REPORTER_FIELD, RESOLVED_FIELD, RESOLUTION_DATE_FIELD, STATUS_FIELD, STATUS_CATEGORY_FIELD, SUMMARY_FIELD, TEXT_FIELD, TEXFIELDS_FIELD, TYPE_FIELD, UPDATED_FIELD, AFFECTED_VERSION_FIELD, APPROVALS_FIELD, CATEGORY_FIELD, COMPONENT_FIELD, DESCRIPTION_FIELD, DEVELOPMENT_FIELD, ENVIRONMENT_FIELD, EPIC_LINK_FIELD, FILTER_FIELD, FIX_VERSION_FIELD, LABELS_FIELD, PARENT_FIELD, RESOLUTION_FIELD, SPRINT_FIELD,
// entity properties
ASSIGNEE_PROPERTY, CREATOR_PROPERTY, ISSUE_PROPERTY, PROJECT_PROPERTY, REPORTER_PROPERTY, VOTER_PROPERTY, WATCHER_PROPERTY];
