// @atlaskit/editor-plugin-media/dist/cjs/pm-plugins/linking/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mediaLinkingPluginKey = exports.getMediaLinkingState = exports.default = exports.createMediaLinkingCommand = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _safePlugin = require("@atlaskit/editor-common/safe-plugin");
var _utils = require("@atlaskit/editor-common/utils");
var _state = require("@atlaskit/editor-prosemirror/state");
var _reducer = _interopRequireDefault(require("./reducer"));
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var mediaLinkingPluginKey = exports.mediaLinkingPluginKey = new _state.PluginKey('mediaLinking');
var initialState = {
  visible: false,
  editable: false,
  mediaPos: null,
  link: ''
};
function mapping(tr, pluginState) {
  if (pluginState && pluginState.mediaPos !== null) {
    return _objectSpread(_objectSpread({}, pluginState), {}, {
      mediaPos: tr.mapping.map(pluginState.mediaPos)
    });
  }
  return pluginState;
}
function onSelectionChanged(tr) {
  var isNodeSelection = tr.selection instanceof _state.NodeSelection;
  if (!isNodeSelection) {
    return initialState;
  }
  var pos = tr.selection.$from.pos;
  var mediaPos = tr.selection.node.type.name === 'mediaInline' ? pos : pos + 1;
  var node = tr.doc.nodeAt(mediaPos);
  if (!node || !['media', 'mediaInline'].includes(node.type.name)) {
    return initialState;
  }
  var mark = node.marks.find(function (mark) {
    return mark.type.name === 'link';
  });
  if (mark) {
    return _objectSpread(_objectSpread({}, initialState), {}, {
      mediaPos: mediaPos,
      editable: true,
      link: mark.attrs.href
    });
  }
  return _objectSpread(_objectSpread({}, initialState), {}, {
    mediaPos: mediaPos
  });
}
var mediaLinkingPluginFactory = (0, _utils.pluginFactory)(mediaLinkingPluginKey, _reducer.default, {
  mapping: mapping,
  onSelectionChanged: onSelectionChanged
});
var createMediaLinkingCommand = exports.createMediaLinkingCommand = mediaLinkingPluginFactory.createCommand,
  getMediaLinkingState = exports.getMediaLinkingState = mediaLinkingPluginFactory.getPluginState;
var _default = exports.default = function _default(dispatch) {
  return new _safePlugin.SafePlugin({
    key: mediaLinkingPluginKey,
    state: mediaLinkingPluginFactory.createPluginState(dispatch, initialState)
  });
};
