// @atlaskit/editor-plugin-highlight/dist/cjs/ui/shared/PaletteDropdown.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PaletteDropdown = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireDefault(require("react"));
var _uiColor = require("@atlaskit/editor-common/ui-color");
var _uiMenu = require("@atlaskit/editor-common/ui-menu");
var _editorPalette = require("@atlaskit/editor-palette");
var _editorSharedStyles = require("@atlaskit/editor-shared-styles");
var _experiments = require("@atlaskit/tmp-editor-statsig/experiments");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var PaletteDropdown = exports.PaletteDropdown = function PaletteDropdown(props) {
  var popupsMountPoint = props.popupsMountPoint,
    popupsBoundariesElement = props.popupsBoundariesElement,
    popupsScrollableElement = props.popupsScrollableElement,
    isOpen = props.isOpen,
    activeColor = props.activeColor,
    trigger = props.trigger,
    onColorChange = props.onColorChange,
    isOpenedByKeyboard = props.isOpenedByKeyboard,
    handleClickOutside = props.handleClickOutside,
    handleEscapeKeydown = props.handleEscapeKeydown;

  // pixels, used to determine where to horizontally position the dropdown when space is limited
  // this should reflect the width of the dropdown when fully populated with colors, including translations due to layering
  var fitWidth = 242;
  var colorPalette = _uiColor.highlightColorPalette;
  if ((0, _experiments.editorExperiment)('editor_text_highlight_orange_to_yellow', 'test', {
    exposure: true
  })) {
    colorPalette = _uiColor.highlightColorPalette.map(function (item) {
      if (item.label === 'Orange') {
        return _objectSpread(_objectSpread({}, item), {}, {
          label: 'Yellow',
          message: _uiColor.colorPaletteMessages.yellow
        });
      }
      return item;
    });
  }
  var _getSelectedRowAndCol = (0, _uiColor.getSelectedRowAndColumnFromPalette)(colorPalette, activeColor),
    selectedRowIndex = _getSelectedRowAndCol.selectedRowIndex,
    selectedColumnIndex = _getSelectedRowAndCol.selectedColumnIndex;
  return /*#__PURE__*/_react.default.createElement(_uiMenu.DropdownContainer, {
    mountTo: popupsMountPoint,
    boundariesElement: popupsBoundariesElement,
    scrollableElement: popupsScrollableElement,
    isOpen: isOpen,
    handleClickOutside: handleClickOutside,
    handleEscapeKeydown: handleEscapeKeydown,
    zIndex: _editorSharedStyles.akEditorMenuZIndex,
    fitWidth: fitWidth,
    closeOnTab: true,
    arrowKeyNavigationProviderOptions: {
      type: _uiMenu.ArrowKeyNavigationType.COLOR,
      selectedRowIndex: selectedRowIndex,
      selectedColumnIndex: selectedColumnIndex,
      isOpenedByKeyboard: isOpenedByKeyboard,
      isPopupPositioned: true
    },
    trigger: trigger
  }, /*#__PURE__*/_react.default.createElement("div", {
    "data-testid": "highlight-color-palette"
  }, /*#__PURE__*/_react.default.createElement(_uiColor.ColorPalette, {
    onClick: onColorChange,
    selectedColor: activeColor,
    paletteOptions: {
      palette: colorPalette,
      hexToPaletteColor: _editorPalette.hexToEditorTextBackgroundPaletteColor
    }
  })));
};
