// @atlaskit/editor-core/dist/esm/plugins/table/utils/dom.js
import { tableResizeHandleWidth } from '@atlaskit/editor-common';
import { TableCssClassName as ClassName } from '../types';
import { closestElement } from '../../../utils';
import { tableToolbarSize } from '../ui/styles';
export var isCell = function (node) {
    return node && ['TH', 'TD'].indexOf(node.tagName) > -1;
};
export var isCornerButton = function (node) {
    var cl = node.classList;
    return cl.contains(ClassName.CONTROLS_CORNER_BUTTON);
};
export var isInsertRowButton = function (node) {
    var cl = node.classList;
    return (cl.contains(ClassName.CONTROLS_INSERT_ROW) ||
        closestElement(node, "." + ClassName.CONTROLS_INSERT_ROW) ||
        (cl.contains(ClassName.CONTROLS_BUTTON_OVERLAY) &&
            closestElement(node, "." + ClassName.ROW_CONTROLS)));
};
export var getColumnOrRowIndex = function (target) { return [
    parseInt(target.getAttribute('data-start-index') || '-1', 10),
    parseInt(target.getAttribute('data-end-index') || '-1', 10),
]; };
export var isColumnControlsDecorations = function (node) {
    var cl = node.classList;
    return cl.contains(ClassName.COLUMN_CONTROLS_DECORATIONS);
};
export var isRowControlsButton = function (node) {
    var cl = node.classList;
    return (cl.contains(ClassName.ROW_CONTROLS_BUTTON) ||
        cl.contains(ClassName.NUMBERED_COLUMN_BUTTON));
};
export var isTableControlsButton = function (node) {
    var cl = node.classList;
    return (cl.contains(ClassName.CONTROLS_BUTTON) ||
        cl.contains(ClassName.ROW_CONTROLS_BUTTON_WRAP));
};
export var getMousePositionHorizontalRelativeByElement = function (mouseEvent) {
    var element = mouseEvent.target;
    if (element instanceof HTMLElement) {
        var elementRect = element.getBoundingClientRect();
        if (elementRect.width <= 0) {
            return null;
        }
        var x = mouseEvent.clientX - elementRect.left;
        return x / elementRect.width > 0.5 ? 'right' : 'left';
    }
    return null;
};
export var getMousePositionVerticalRelativeByElement = function (mouseEvent) {
    var element = mouseEvent.target;
    if (element instanceof HTMLElement) {
        var elementRect = element.getBoundingClientRect();
        if (elementRect.height <= 0) {
            return null;
        }
        var y = mouseEvent.clientY - elementRect.top;
        return y / elementRect.height > 0.5 ? 'bottom' : 'top';
    }
    return null;
};
export var updateResizeHandles = function (tableRef, resizeState, index) {
    if (!tableRef) {
        return;
    }
    var height = tableRef.offsetHeight + tableToolbarSize;
    // see ED-7600
    var nodes = Array.from(tableRef.querySelectorAll("." + ClassName.RESIZE_HANDLE));
    if (!nodes || !nodes.length) {
        return;
    }
    if (resizeState && typeof index === 'number' && nodes[index]) {
        nodes.forEach(function (node, idx) {
            if (index !== idx) {
                node.style.display = "none";
            }
        });
        var colSpanIndex = parseInt(nodes[index].getAttribute('data-colspan-index') || '-1', 10);
        var left = colSpanIndex > 0
            ? resizeState.cols
                .slice(index - colSpanIndex, index + 1)
                .reduce(function (acc, col) { return acc + col.width; }, 0)
            : resizeState.cols[index].width;
        var offset = tableResizeHandleWidth / 2 + 4;
        nodes[index].style.left = left - offset + "px";
        nodes[index].style.height = height + "px";
    }
    else {
        nodes.forEach(function (node) {
            node.style.height = height + "px";
            node.style.display = "block";
        });
    }
};
