// @atlaskit/drawer/dist/es2019/components/index.js
/* eslint-disable @repo/internal/dom-events/no-unsafe-event-listeners */
import React, { useCallback, useEffect } from 'react';
import { canUseDOM } from 'exenv';
import { usePlatformLeafEventHandler } from '@atlaskit/analytics-next';
import { Layering, useCloseOnEscapePress } from '@atlaskit/layering';
import Portal from '@atlaskit/portal';
import Blanket from './blanket';
import DrawerPrimitive from './primitives';
// escape close manager for layering
const EscapeCloseManager = ({
  onClose
}) => {
  // wrap so that we can cast the event to a React.KeyboardEvent
  const handleClose = useCallback(evt => {
    onClose && onClose(evt);
  }, [onClose]);
  useCloseOnEscapePress({
    onClose: handleClose
  });
  return /*#__PURE__*/React.createElement("span", null);
};

/**
 * __Drawer__
 *
 * A drawer is a panel that slides in from the left side of the screen.
 *
 * - [Examples](https://atlassian.design/components/drawer/examples)
 * - [Code](https://atlassian.design/components/drawer/code)
 * - [Usage](https://atlassian.design/components/drawer/usage)
 */
const Drawer = ({
  width = 'narrow',
  isOpen,
  isFocusLockEnabled = true,
  shouldReturnFocus = true,
  autoFocusFirstElem = true,
  onKeyDown,
  onClose,
  testId,
  children,
  icon,
  closeLabel,
  scrollContentLabel,
  onCloseComplete,
  onOpenComplete,
  // eslint-disable-next-line @repo/internal/react/consistent-props-definitions
  overrides,
  zIndex = 'unset',
  label,
  titleId,
  enterFrom
}) => {
  const body = canUseDOM ? document.querySelector('body') : undefined;
  const handleClose = usePlatformLeafEventHandler({
    fn: (evt, analyticsEvent) => onClose && onClose(evt, analyticsEvent),
    action: 'dismissed',
    componentName: 'drawer',
    packageName: "@atlaskit/drawer",
    packageVersion: "10.1.0",
    analyticsData: {
      trigger: 'escKey'
    }
  });
  const handleKeyDown = useCallback(evt => {
    onKeyDown && onKeyDown(evt);
  }, [onKeyDown]);
  useEffect(() => {
    if (isOpen) {
      window.addEventListener('keydown', handleKeyDown);
    }
    return () => {
      window.removeEventListener('keydown', handleKeyDown);
    };
  }, [handleKeyDown, isOpen]);
  const handleBlanketClick = usePlatformLeafEventHandler({
    fn: (evt, analyticsEvent) => onClose && onClose(evt, analyticsEvent),
    action: 'dismissed',
    componentName: 'drawer',
    packageName: "@atlaskit/drawer",
    packageVersion: "10.1.0",
    analyticsData: {
      trigger: 'blanket'
    }
  });
  const handleBackButtonClick = usePlatformLeafEventHandler({
    fn: (evt, analyticsEvent) => onClose && onClose(evt, analyticsEvent),
    action: 'dismissed',
    componentName: 'drawer',
    packageName: "@atlaskit/drawer",
    packageVersion: "10.1.0",
    analyticsData: {
      trigger: 'backButton'
    }
  });
  if (!body) {
    return null;
  }
  return /*#__PURE__*/React.createElement(Portal, {
    zIndex: zIndex
  }, /*#__PURE__*/React.createElement(Blanket, {
    isOpen: isOpen,
    onBlanketClicked: handleBlanketClick,
    testId: testId && `${testId}--blanket`
  }), /*#__PURE__*/React.createElement(DrawerPrimitive, {
    testId: testId,
    icon: icon,
    closeLabel: closeLabel,
    in: isOpen,
    onClose: handleBackButtonClick,
    onCloseComplete: onCloseComplete,
    onOpenComplete: onOpenComplete,
    width: width,
    enterFrom: enterFrom,
    label: label,
    titleId: titleId
    // eslint-disable-next-line @repo/internal/react/no-unsafe-overrides
    ,
    overrides: overrides,
    autoFocusFirstElem: autoFocusFirstElem,
    isFocusLockEnabled: isFocusLockEnabled,
    shouldReturnFocus: shouldReturnFocus,
    scrollContentLabel: scrollContentLabel
  }, isOpen ? /*#__PURE__*/React.createElement(Layering, {
    isDisabled: false
  }, children, /*#__PURE__*/React.createElement(EscapeCloseManager, {
    onClose: handleClose
  })) : children));
};
export default Drawer;
