// @atlaskit/editor-plugin-card/dist/cjs/pm-plugins/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.showLinkToolbar = exports.showDatasourceModal = exports.setProvider = exports.setDatasourceTableRef = exports.setDatasourceStash = exports.setCardLayoutAndDatasourceTableRef = exports.setCardLayout = exports.resolveCard = exports.removeDatasourceStash = exports.removeCard = exports.registerSmartCardEvents = exports.registerRemoveOverlay = exports.registerCard = exports.queueCards = exports.hideLinkToolbar = exports.hideDatasourceModal = exports.clearOverlayCandidate = exports.cardAction = void 0;
var _pluginKey = require("./plugin-key");
var cardAction = exports.cardAction = function cardAction(tr, action) {
  return tr.setMeta(_pluginKey.pluginKey, action);
};
var resolveCard = exports.resolveCard = function resolveCard(url) {
  return function (tr) {
    return cardAction(tr, {
      type: 'RESOLVE',
      url: url
    });
  };
};
var queueCards = exports.queueCards = function queueCards(requests) {
  return function (tr) {
    return cardAction(tr, {
      type: 'QUEUE',
      requests: requests
    });
  };
};
var registerCard = exports.registerCard = function registerCard(info) {
  return function (tr) {
    return cardAction(tr, {
      type: 'REGISTER',
      info: info
    });
  };
};
var removeCard = exports.removeCard = function removeCard(info) {
  return function (tr) {
    return cardAction(tr, {
      type: 'REMOVE_CARD',
      info: info
    });
  };
};
var registerRemoveOverlay = exports.registerRemoveOverlay = function registerRemoveOverlay(callback, info) {
  return function (tr) {
    return cardAction(tr, {
      type: 'REGISTER_REMOVE_OVERLAY_ON_INSERTED_LINK',
      callback: callback,
      info: info
    });
  };
};
var registerSmartCardEvents = exports.registerSmartCardEvents = function registerSmartCardEvents(smartLinkEvents) {
  return function (tr) {
    return cardAction(tr, {
      type: 'REGISTER_EVENTS',
      smartLinkEvents: smartLinkEvents
    });
  };
};
var setProvider = exports.setProvider = function setProvider(cardProvider) {
  return function (tr) {
    return cardAction(tr, {
      type: 'SET_PROVIDER',
      provider: cardProvider
    });
  };
};
var setDatasourceTableRef = exports.setDatasourceTableRef = function setDatasourceTableRef(datasourceTableRef) {
  return function (tr) {
    return cardAction(tr, {
      type: 'SET_DATASOURCE_TABLE_REF',
      datasourceTableRef: datasourceTableRef
    });
  };
};
var setCardLayout = exports.setCardLayout = function setCardLayout(layout) {
  return function (tr) {
    return cardAction(tr, {
      type: 'SET_CARD_LAYOUT',
      layout: layout
    });
  };
};
var setCardLayoutAndDatasourceTableRef = exports.setCardLayoutAndDatasourceTableRef = function setCardLayoutAndDatasourceTableRef(_ref) {
  var layout = _ref.layout,
    datasourceTableRef = _ref.datasourceTableRef;
  return function (tr) {
    return cardAction(tr, {
      type: 'SET_CARD_LAYOUT_AND_DATASOURCE_TABLE_REF',
      layout: layout,
      datasourceTableRef: datasourceTableRef
    });
  };
};
var showLinkToolbar = exports.showLinkToolbar = function showLinkToolbar(tr) {
  return cardAction(tr, {
    type: 'SHOW_LINK_TOOLBAR'
  });
};
var hideLinkToolbar = exports.hideLinkToolbar = function hideLinkToolbar(tr) {
  return cardAction(tr, {
    type: 'HIDE_LINK_TOOLBAR'
  });
};
var showDatasourceModal = exports.showDatasourceModal = function showDatasourceModal(modalType) {
  return function (tr) {
    return cardAction(tr, {
      type: 'SHOW_DATASOURCE_MODAL',
      modalType: modalType
    });
  };
};
var hideDatasourceModal = exports.hideDatasourceModal = function hideDatasourceModal(tr) {
  return cardAction(tr, {
    type: 'HIDE_DATASOURCE_MODAL'
  });
};
var clearOverlayCandidate = exports.clearOverlayCandidate = function clearOverlayCandidate(tr) {
  return cardAction(tr, {
    type: 'CLEAR_OVERLAY_CANDIDATE'
  });
};
var setDatasourceStash = exports.setDatasourceStash = function setDatasourceStash(tr, datasourceStash) {
  return cardAction(tr, {
    type: 'SET_DATASOURCE_STASH',
    datasourceStash: datasourceStash
  });
};
var removeDatasourceStash = exports.removeDatasourceStash = function removeDatasourceStash(tr, url) {
  return cardAction(tr, {
    type: 'REMOVE_DATASOURCE_STASH',
    url: url
  });
};
