// @atlaskit/smart-card/dist/cjs/view/FlexibleCard/components/common/loading-skeleton/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LoadingSkeletonOld = exports.LoadingSkeletonNew = void 0;
require("./index.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var animationNameStyles = null;
var loadingSkeletonStyle = null;
var LoadingSkeletonNew = exports.LoadingSkeletonNew = function LoadingSkeletonNew(_ref) {
  var testId = _ref.testId,
    width = _ref.width,
    height = _ref.height;
  return /*#__PURE__*/React.createElement("span", {
    "data-testid": testId,
    className: (0, _runtime.ax)(["_2rkoyh40 _uiztglyw _bfhk178v _1itkx3i8 _12vemgnk _1lrw6rms _1e0c1o8l _5sag9cwz _1o51q7pw _tip812c5 _j7hqqkar _1pglp3kn"]),
    style: {
      width: width,
      height: height,
      "--_16116mi": (0, _runtime.ix)("linear-gradient( to right, transparent 0%, ".concat("var(--ds-skeleton, #edeef1)", " 20%, transparent 40%, transparent 100% )"))
    }
  });
};
var LoadingSkeletonOld = exports.LoadingSkeletonOld = function LoadingSkeletonOld(props) {
  return /*#__PURE__*/React.createElement(LoadingSkeletonNew, (0, _extends2.default)({}, props, {
    width: "".concat(props.width, "rem"),
    height: "".concat(props.height, "rem")
  }));
};
