// @atlaskit/ufo/dist/esm/platform-client/core/experience/global-page-load-experience.js
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _get from "@babel/runtime/helpers/get";
import _inherits from "@babel/runtime/helpers/inherits";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
import _regeneratorRuntime from "@babel/runtime/regenerator";
function _callSuper(t, o, e) { return o = _getPrototypeOf(o), _possibleConstructorReturn(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], _getPrototypeOf(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function _superPropGet(t, o, e, r) { var p = _get(_getPrototypeOf(1 & r ? t.prototype : t), o, e); return 2 & r && "function" == typeof p ? function (t) { return p.apply(e, t); } : p; }
import { untilAll } from '../../utils/until-helpers';
import { perfNowOrTimestamp, UFOAbstractExperience } from './abstract-experience';
import { ExperiencePerformanceTypes, ExperienceTypes } from './experience-types';
import { PageSegmentExperienceTypes } from './page-segment-experience-types';
var UFOGlobalPageLoadExperience = /*#__PURE__*/function (_UFOAbstractExperienc) {
  function UFOGlobalPageLoadExperience() {
    var _this;
    _classCallCheck(this, UFOGlobalPageLoadExperience);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, UFOGlobalPageLoadExperience, [].concat(args));
    _defineProperty(_this, "loadingPageLoadId", 'UNKNOWN');
    _defineProperty(_this, "type", ExperienceTypes.Load);
    _defineProperty(_this, "performanceType", ExperiencePerformanceTypes.PageLoad);
    _defineProperty(_this, "initial", true);
    return _this;
  }
  _inherits(UFOGlobalPageLoadExperience, _UFOAbstractExperienc);
  return _createClass(UFOGlobalPageLoadExperience, [{
    key: "startPageLoad",
    value: function startPageLoad(id) {
      var initial = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
      var startTime = arguments.length > 2 ? arguments[2] : undefined;
      var result = _superPropGet(UFOGlobalPageLoadExperience, "start", this, 3)([initial ? 0 : perfNowOrTimestamp(startTime)]);
      this.setPageLoadId(id);
      this.initial = initial;
      return result;
    }
  }, {
    key: "updateConfig",
    value: function updateConfig(config) {
      Object.assign(this.config, config);
    }
  }, {
    key: "setPageLoadId",
    value: function setPageLoadId(id) {
      this.loadingPageLoadId = id;
    }
  }, {
    key: "getId",
    value: function () {
      var _getId = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              return _context.abrupt("return", this.loadingPageLoadId);
            case 1:
            case "end":
              return _context.stop();
          }
        }, _callee, this);
      }));
      function getId() {
        return _getId.apply(this, arguments);
      }
      return getId;
    }()
  }, {
    key: "exportData",
    value: function () {
      var _exportData = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee2() {
        var data;
        return _regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              _context2.next = 2;
              return _superPropGet(UFOGlobalPageLoadExperience, "exportData", this, 3)([]);
            case 2:
              data = _context2.sent;
              return _context2.abrupt("return", _objectSpread(_objectSpread({}, data), {}, {
                initial: this.initial
              }));
            case 4:
            case "end":
              return _context2.stop();
          }
        }, _callee2, this);
      }));
      function exportData() {
        return _exportData.apply(this, arguments);
      }
      return exportData;
    }()
  }]);
}(UFOAbstractExperience);
export var GlobalPageLoadExperience = new UFOGlobalPageLoadExperience('UFO_GLOBAL_PAGE_LOAD', {
  type: ExperienceTypes.Load,
  performanceType: ExperiencePerformanceTypes.PageLoad,
  // this should come from global config
  until: untilAll([{
    category: PageSegmentExperienceTypes.PRODUCT
  }])
});
