// @atlaskit/media-ui/dist/es2019/customMediaPlayer/styled-compiled.js
/* styled-compiled.tsx generated by @compiled/babel-plugin v0.36.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./styled-compiled.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
/* eslint-disable @atlaskit/design-system/ensure-design-token-usage */

import React, { forwardRef } from 'react';
import { R300 } from '@atlaskit/theme/colors';
import { getControlsWrapperClassName } from './getControlsWrapperClassName';
const volumeStyles = null;
const showSliderVolumeStyles = null;
export const VolumeWrapper = ({
  showSlider,
  children,
  ...props
}) => /*#__PURE__*/React.createElement("div", _extends({}, props, {
  className: ax(["_1reo15vq _18m915vq _v5641rb3 _1e0c1txw _1bsbldb6 _4cvr1h6o _94n5ze3t _1ltv1jfw", showSlider && "_13y6o43e _195p1w81 _o7kso43e _1chp1w81 _1gglo43e _m7bd1w81"])
}), children);
const currentTimeStyles = null;
export const CurrentTime = ({
  children,
  ...props
}) => /*#__PURE__*/React.createElement("div", _extends({}, props, {
  className: ax(["_syaz1poe _uiztglyw _2hwxu2gc _o5721q9c"])
}), children);
const timelineStyles = null;
export const TimeLine = /*#__PURE__*/forwardRef(({
  children,
  ...props
}, ref) => {
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: ref
  }, props, {
    className: ax(["_v5641eho _2rko14y2 _1bsb1osq _4t3iyh40 _1lww9cwz _bfhk9ucv _kqswh2mm"])
  }), children);
});
const currentTimeLineStyles = null;
export const CurrentTimeLine = ({
  children,
  ...props
}) => /*#__PURE__*/React.createElement("div", _extends({}, props, {
  className: ax(["_2rko1kw7 _bfhk7g9i _4t3i1kw7 _kqswstnw _154ize3t _p12f1osq"])
}), children);
const thumbStyles = null;
export const Thumb = ({
  children,
  ...props
}) => /*#__PURE__*/React.createElement("div", _extends({}, props, {
  className: ax(["_2rko1osq _v5641eho _189et94y _1dqonqa1 _1h6d16fc _lcxvglyw _1bsbdlk8 _4t3idlk8 _bfhku67f _kqswstnw _1xi2idpf _154iv77o _t9ec17tb _1lww9cwz"])
}), children);
const currentTimeLineThumbStyles = null;
export const CurrentTimeLineThumb = /*#__PURE__*/forwardRef(({
  children,
  ...props
}, ref) => {
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: ref
  }, props, {
    className: ax(["_19itglyw _2rko1osq _1a371rek _kqswstnw _1e0c1ule _1xi2idpf _154i1ssb _t9ec1tul _bfhk7g9i _4t3ik7ql _1bsbk7ql _lcxvglyw _tzy4idpf _1cwgv77o _1h7hkb7n"])
  }), children);
});
const bufferedTimeStyles = null;
export const BufferedTime = ({
  children,
  ...props
}) => /*#__PURE__*/React.createElement("div", _extends({}, props, {
  className: ax(["_2rko1kw7 _bfhk1ncv _4t3i1kw7 _1bsbidpf"])
}), children);
const leftControlsStyles = null;
export const LeftControls = ({
  children,
  ...props
}) => /*#__PURE__*/React.createElement("div", _extends({}, props, {
  className: ax(["_1e0c1txw _18u0utpp"])
}), children);
const rightControlStyles = null;
export const RightControls = ({
  children,
  ...props
}) => /*#__PURE__*/React.createElement("div", _extends({}, props, {
  className: ax(["_1e0c1txw _4cvr1h6o _2hwxutpp"])
}), children);
const controlsWrapperStyles = null;
const hiddenControlsWrapperStyles = null;
export const ControlsWrapper = /*#__PURE__*/forwardRef(({
  children,
  controlsHidden,
  ...props
}, ref) => {
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: ref
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop
    ,
    className: ax(["_11q7acks _94n5idpf _1ltvidpf _1bsb1osq _4t3i1wug _kqswstnw", controlsHidden && "_1mygkb7n", getControlsWrapperClassName(controlsHidden)])
  }, props), children);
});
const volumeToggleWrapperStyles = null;
const mutedVolumeToggleWrapperStyles = null;
export const VolumeToggleWrapper = ({
  isMuted,
  children,
  ...props
}) => /*#__PURE__*/React.createElement("div", _extends({}, props, {
  className: ax(["_kqswh2mm _2hwxw1gv _14hp1j2z", isMuted && "_d32vqb75"])
}), children);
const volumeTimeRangeWrapperStyles = null;
export const VolumeTimeRangeWrapper = ({
  children,
  ...props
}) => /*#__PURE__*/React.createElement("div", _extends({}, props, {
  className: ax(["_1bsb1osq _2hwxv47k"])
}), children);
const mutedIndicatorStyles = null;
const isMutedStyles = null;
export const MutedIndicator = ({
  isMuted,
  children,
  ...props
}) => /*#__PURE__*/React.createElement("div", _extends({}, props, {
  className: ax(["_bfhk18rw _1bsb1pa9 _4t3iyh40 _kqswstnw _154iu2gc _1ltvu2gc _1pbycs5v _t9ec17wz _tzy4idpf _lcxvglyw", isMuted && "_tzy4kb7n"])
}), children);
const currentTimeTooltipStyles = null;
const currentTimeToolTipDraggingStyles = null;
const currentTimeToolTipNotDraggingStyles = null;
const currentTimeTooltipThumbHoveredStyles = null;
const currentTimeTooltipThumbFocusedStyles = null;
export const CurrentTimeTooltip = ({
  isDragging,
  timeLineThumbIsHover,
  timeLineThumbIsFocus,
  children,
  ...props
}) => /*#__PURE__*/React.createElement("div", _extends({
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
  className: ax(["_11c8dcr7 _2rko1b66 _v564g17y _kqswstnw _uiztglyw _154iukr8 _bfhk39qu _syaz1pcz _ca0q1b66 _u5f3u2gc _n3td1b66 _19bvu2gc _1ltv1ssb _t9eczr6s _1nmz16rz", isDragging && "_tzy4kb7n", !isDragging && "_tzy4idpf", timeLineThumbIsHover && "_tzy4kb7n", timeLineThumbIsFocus && "_tzy4kb7n", "current-time-tooltip"])
}, props), children);
const timeRangeWrapperStyles = null;
export const TimeRangeWrapper = /*#__PURE__*/forwardRef(({
  children,
  ...props
}, ref) => {
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: ref
  }, props, {
    className: ax(["_1e0c1txw _4cvr1h6o _4t3iqslr _80omtlke _1bsb1osq _1qcea6sd _1az8glyw _1838glyw _4i4mdlk8 _9cyhdlk8 _14lz1lpd _u2q4a6sd _17dfdlk8 _rs6wdlk8 _1w9uglyw _7cve9ucv _iofsa6sd _1a96a6sd _dey49ucv"])
  }), children);
});
