// @atlaskit/editor-plugin-help-dialog/dist/es2019/ui/styles.js
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import { css } from '@emotion/react';
import { akEditorUnitZIndex, relativeFontSizeToBase16 } from '@atlaskit/editor-shared-styles';
// Ignored via go/ees005
// eslint-disable-next-line import/no-namespace
import * as colors from '@atlaskit/theme/colors';
import { B300, N400 } from '@atlaskit/theme/colors';
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
export const header = css({
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  zIndex: akEditorUnitZIndex,
  minHeight: "var(--ds-space-300, 24px)",
  padding: `${"var(--ds-space-250, 20px)"} ${"var(--ds-space-500, 40px)"}`,
  display: 'flex',
  justifyContent: 'space-between',
  alignItems: 'center',
  boxShadow: "'none'",
  font: "var(--ds-font-heading-large, normal 500 24px/28px ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
  color: `var(--ds-text, ${colors.N400})`,
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
  backgroundColor: `var(--ds-background-neutral-subtle, ${colors.N0})`,
  borderRadius: "var(--ds-border-radius, 3px)"
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
export const footer = css({
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  zIndex: akEditorUnitZIndex,
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
  color: `var(--ds-text-subtlest, ${colors.N300})`,
  font: "var(--ds-font-body, normal 400 14px/20px ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  padding: "var(--ds-space-300, 24px)",
  textAlign: 'right',
  boxShadow: "'none'"
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
export const contentWrapper = css({
  padding: `${"var(--ds-space-250, 20px)"} ${"var(--ds-space-500, 40px)"}`,
  borderBottomRightRadius: "var(--ds-border-radius, 3px)",
  overflow: 'auto',
  position: 'relative',
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
  color: `var(--ds-text-subtle, ${colors.N400})`,
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
  backgroundColor: `var(--ds-background-neutral-subtle, ${colors.N0})`
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
export const line = css({
  background: "var(--ds-background-neutral-subtle, #fff)",
  content: "''",
  display: 'block',
  height: "var(--ds-space-025, 2px)",
  left: 0,
  position: 'absolute',
  top: 0,
  right: 0,
  width: '100%',
  minWidth: '604px'
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
export const content = css({
  minWidth: '524px',
  width: '100%',
  position: 'relative',
  display: 'flex',
  justifyContent: 'space-between'
});
export const column = {
  width: '44%',
  '& > ul': {
    padding: 0
  }
};

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
export const row = css({
  margin: `${"var(--ds-space-250, 20px)"} 0`,
  display: 'flex',
  justifyContent: 'space-between'
});
export const dialogHeader = {
  '&': {
    fontSize: relativeFontSizeToBase16(24),
    fontWeight: "var(--ds-font-weight-regular, 400)",
    color: `var(--ds-text-subtle, ${N400})`,
    letterSpacing: 'normal',
    lineHeight: 1.42857142857143
  }
};

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
export const shortcutsArray = css({
  display: 'flex',
  flexDirection: 'column',
  flexShrink: 0,
  gap: "var(--ds-space-150, 12px)"
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
export const componentFromKeymapWrapperStyles = css({
  flexShrink: 0
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
export const toolbarButton = css({
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-selectors
  '&:focus:not(:focus-visible)': {
    outline: `2px solid ${`var(--ds-border-focused, ${B300})`}`,
    outlineOffset: "var(--ds-space-025, 2px)"
  }
});
