// @atlaskit/jql-editor/dist/cjs/ui/jql-editor-footer-content/jql-editor-help/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.JQLEditorHelp = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireDefault(require("react"));
var _form = require("@atlaskit/form");
var _compiled = require("@atlaskit/primitives/compiled");
var _constants = require("../../../common/constants");
var _useEditorTheme = require("../../../hooks/use-editor-theme");
var _state = require("../../../state");
var _messages = require("./messages");
var _styled = require("./styled");
var JQLEditorHelp = exports.JQLEditorHelp = function JQLEditorHelp() {
  var _useIntl = (0, _state.useIntl)(),
    _useIntl2 = (0, _slicedToArray2.default)(_useIntl, 1),
    formatMessage = _useIntl2[0].formatMessage;
  var _useEditorViewHasFocu = (0, _state.useEditorViewHasFocus)(),
    _useEditorViewHasFocu2 = (0, _slicedToArray2.default)(_useEditorViewHasFocu, 1),
    editorViewHasFocus = _useEditorViewHasFocu2[0];
  var _useEditorThemeContex = (0, _useEditorTheme.useEditorThemeContext)(),
    isSearch = _useEditorThemeContex.isSearch;
  var _useScopedId = (0, _state.useScopedId)(_constants.JQL_EDITOR_HELP_CONTENT_ID),
    _useScopedId2 = (0, _slicedToArray2.default)(_useScopedId, 1),
    helpContentId = _useScopedId2[0];
  var SearchMessage = /*#__PURE__*/_react.default.createElement(_compiled.Text, {
    size: "UNSAFE_small",
    color: "inherit"
  }, formatMessage(_messages.messages.searchCommand, {
    b: function b(text) {
      return /*#__PURE__*/_react.default.createElement("b", null, text);
    }
  }));
  var NewLineMessage = /*#__PURE__*/_react.default.createElement(_compiled.Text, {
    size: "UNSAFE_small",
    color: "inherit"
  }, formatMessage(isSearch ? _messages.messages.newLineCommand : _messages.messages.fieldNewLineCommand, {
    b: function b(text) {
      return /*#__PURE__*/_react.default.createElement("b", null, text);
    }
  }));
  return /*#__PURE__*/_react.default.createElement(_styled.HelpContainer, {
    id: helpContentId,
    isVisible: editorViewHasFocus
  }, isSearch && /*#__PURE__*/_react.default.createElement(_form.HelperMessage, null, SearchMessage), /*#__PURE__*/_react.default.createElement(_form.HelperMessage, null, NewLineMessage));
};
