// @atlaskit/editor-plugin-media/dist/cjs/pm-plugins/commands/helpers.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.updateMediaNodeAttrs = exports.updateCurrentMediaNodeAttrs = exports.updateAllMediaNodesAttrs = exports.replaceExternalMedia = exports.isMediaNode = exports.findMediaNode = exports.findAllMediaNodes = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _steps = require("@atlaskit/adf-schema/steps");
var _collab = require("@atlaskit/editor-common/collab");
var _pluginKey = require("../plugin-key");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
/**
 * Note that Media Inline is inserted like a media single node into the media plugin state.
 * Though it is not of type mediaSingle, it shares the same `findMediaSingleNode` method
 *
 */
var findMediaNode = exports.findMediaNode = function findMediaNode(mediaPluginState, id) {
  var mediaNodes = mediaPluginState.mediaNodes;

  // Array#find... no IE support
  return mediaNodes.reduce(function (memo, nodeWithPos) {
    if (memo) {
      return memo;
    }
    var node = nodeWithPos.node;
    if (node.attrs.id === id) {
      return nodeWithPos;
    }
    return memo;
  }, null);
};
var findAllMediaNodes = exports.findAllMediaNodes = function findAllMediaNodes(mediaPluginState, id) {
  var mediaNodes = mediaPluginState.mediaNodes;
  return mediaNodes.filter(function (nodeWithHandler) {
    return nodeWithHandler.node.attrs.id === id;
  });
};
var isMediaNode = exports.isMediaNode = function isMediaNode(pos, state) {
  var node = state.doc.nodeAt(pos);
  return node && ['media', 'mediaInline'].includes(node.type.name);
};
var updateAllMediaNodesAttrs = exports.updateAllMediaNodesAttrs = function updateAllMediaNodesAttrs(id, attrs) {
  return function (state, dispatch) {
    var mediaPluginState = _pluginKey.stateKey.getState(state);
    if (!mediaPluginState) {
      return false;
    }
    var mediaNodes = findAllMediaNodes(mediaPluginState, id);
    var validMediaNodePositions = mediaNodes.reduce(function (acc, _ref) {
      var getPos = _ref.getPos;
      var pos = getPos();
      if (typeof pos !== 'number' || isNaN(pos) || typeof pos === 'number' && !isMediaNode(pos, state)) {
        return acc;
      }
      acc.push(pos);
      return acc;
    }, []);
    if (validMediaNodePositions.length === 0) {
      return false;
    }
    var tr = state.tr;
    validMediaNodePositions.forEach(function (pos) {
      return tr.step(new _steps.SetAttrsStep(pos, attrs));
    });
    tr.setMeta('addToHistory', false);
    (0, _collab.tintDirtyTransaction)(tr);
    if (dispatch) {
      dispatch(tr);
    }
    return true;
  };
};
var updateCurrentMediaNodeAttrs = exports.updateCurrentMediaNodeAttrs = function updateCurrentMediaNodeAttrs(attrs, mediaNode, tintDirtyTr) {
  return function (state, dispatch) {
    var pos = mediaNode.getPos();
    if (typeof pos !== 'number' || isNaN(pos) || typeof pos === 'number' && !isMediaNode(pos, state)) {
      return false;
    }
    var tr = state.tr;
    tr.step(new _steps.SetAttrsStep(pos, attrs));
    tr.setMeta('addToHistory', false);
    if (tintDirtyTr) {
      (0, _collab.tintDirtyTransaction)(tr);
    }
    if (dispatch) {
      dispatch(tr);
    }
    return true;
  };
};
var updateMediaNodeAttrs = exports.updateMediaNodeAttrs = function updateMediaNodeAttrs(id, attrs) {
  return function (state, dispatch) {
    var mediaPluginState = _pluginKey.stateKey.getState(state);
    if (!mediaPluginState) {
      return false;
    }
    var mediaNodeWithPos = findMediaNode(mediaPluginState, id);
    if (!mediaNodeWithPos) {
      return false;
    }
    var tr = state.tr;
    var pos = mediaNodeWithPos.getPos();
    if (typeof pos !== 'number' || !isMediaNode(pos, state)) {
      return false;
    }
    tr.step(new _steps.SetAttrsStep(pos, attrs)).setMeta('addToHistory', false);
    if (dispatch) {
      dispatch(tr);
    }
    return true;
  };
};
var replaceExternalMedia = exports.replaceExternalMedia = function replaceExternalMedia(pos, attrs) {
  return function (state, dispatch) {
    var tr = state.tr;
    var node = tr.doc.nodeAt(pos);
    if (!node || node.type.name !== 'media') {
      return false;
    }
    tr.step(new _steps.SetAttrsStep(pos, _objectSpread({
      type: 'file',
      url: null
    }, attrs))).setMeta('addToHistory', false);
    if (dispatch) {
      dispatch(tr);
    }
    return true;
  };
};
