// @atlaskit/editor-plugin-paste-options-toolbar/dist/es2019/pm-plugins/reducer.js
import { PastePluginActionTypes as ActionTypes } from '../editor-actions/actions';
export const reducer = (state, action) => {
  switch (action.type) {
    case ActionTypes.SHOW_PASTE_OPTIONS:
      {
        return {
          ...state,
          showToolbar: true,
          highlightContent: false,
          isPlainText: action.data.isPlainText,
          plaintext: action.data.plaintext,
          selectedOption: action.data.selectedOption,
          richTextSlice: action.data.richTextSlice,
          pasteStartPos: action.data.pasteStartPos,
          pasteEndPos: action.data.pasteEndPos
        };
      }
    case ActionTypes.HIDE_PASTE_OPTIONS:
      {
        return {
          ...state,
          highlightContent: false,
          showToolbar: false
        };
      }
    case ActionTypes.HIGHLIGHT_CONTENT:
      {
        return {
          ...state,
          highlightContent: true
        };
      }
    case ActionTypes.CHANGE_FORMAT:
      {
        return {
          ...state,
          highlightContent: true,
          selectedOption: action.data.selectedOption
        };
      }
    default:
      return state;
  }
};
