// @atlaskit/link-datasource/dist/es2019/ui/table-footer/provider-link/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import "./index.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
import React, { useMemo } from 'react';
import { useIntl } from 'react-intl-next';
import { JiraServiceManagementIcon } from '@atlaskit/logo';
import { fg } from '@atlaskit/platform-feature-flags';
import { Anchor, Inline } from '@atlaskit/primitives/compiled';
import { AssetsIcon } from '@atlaskit/temp-nav-app-icons/assets';
import { useDatasourceAnalyticsEvents } from '../../../analytics';
import { ASSETS_LIST_OF_LINKS_DATASOURCE_ID } from '../../assets-modal';
import { footerMessages } from '../messages';
const styles = {
  anchor: "_4bfu1r31 _1hmsglyw _ajmmnqa1 _syaz131l _9oik1r31 _1bnxglyw _jf4cnqa1 _30l3131l"
};
export const ProviderLink = ({
  datasourceId
}) => {
  const intl = useIntl();
  const {
    fireEvent
  } = useDatasourceAnalyticsEvents();
  const anchor = useMemo(() => {
    if (datasourceId === ASSETS_LIST_OF_LINKS_DATASOURCE_ID) {
      return fg('assets_as_an_app_v2') ? {
        content: intl.formatMessage(footerMessages.poweredByAssets),
        extensionKey: 'jsm-cmdb-gateway',
        icon: /*#__PURE__*/React.createElement(AssetsIcon, {
          size: "20"
        }),
        interactionName: 'atlas-link',
        url: '/jira/assets'
      } : {
        content: intl.formatMessage(footerMessages.powerByJSM),
        extensionKey: 'jsm-cmdb-gateway',
        icon: /*#__PURE__*/React.createElement(JiraServiceManagementIcon, {
          appearance: "neutral",
          size: "small"
        }),
        interactionName: 'atlas-link',
        url: '/jira/servicedesk/assets'
      };
    }
  }, [datasourceId, intl]);
  return anchor ? /*#__PURE__*/React.createElement(Anchor, {
    href: anchor.url,
    interactionName: anchor.interactionName,
    onClick: () => fireEvent('ui.link.clicked.poweredBy', {
      componentHierarchy: 'datasourceTable',
      extensionKey: anchor.extensionKey
    }),
    rel: "noopener noreferrer",
    target: "_blank",
    testId: "powered-by-jsm-assets-link",
    xcss: styles.anchor
  }, /*#__PURE__*/React.createElement(Inline, {
    alignBlock: "center",
    space: "space.075"
  }, anchor.icon, anchor.content)) : null;
};
