// @atlaskit/avatar-group/dist/cjs/components/internal/components/focus-manager.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.FocusManagerContext = void 0;
var _react = _interopRequireWildcard(require("react"));
var _bindEventListener = require("bind-event-listener");
var _noop = _interopRequireDefault(require("@atlaskit/ds-lib/noop"));
var _handleFocus = _interopRequireDefault(require("../utiles/handle-focus"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 *
 *
 * Context provider which maintains the list of focusable elements and a method to
 * register new menu items.
 * This list drives the keyboard navgation of the menu.
 *
 */
var FocusManagerContext = exports.FocusManagerContext = /*#__PURE__*/(0, _react.createContext)({
  menuItemRefs: [],
  registerRef: _noop.default
});

/**
 * Focus manager logic
 */
var FocusManager = function FocusManager(_ref) {
  var children = _ref.children;
  var menuItemRefs = (0, _react.useRef)([]);
  var registerRef = (0, _react.useCallback)(function (ref) {
    if (ref && !menuItemRefs.current.includes(ref)) {
      menuItemRefs.current.push(ref);
    }
  }, []);

  // set focus and intentionally rebinding listener and clean up listener on each render
  (0, _react.useEffect)(function () {
    (0, _bindEventListener.bind)(window, {
      type: 'keydown',
      listener: (0, _handleFocus.default)(menuItemRefs.current)
    });
    var unbind = function unbind() {
      (0, _bindEventListener.bind)(window, {
        type: 'keydown',
        listener: (0, _handleFocus.default)(menuItemRefs.current)
      });
    };
    return unbind;
  }, []);
  var contextValue = {
    menuItemRefs: menuItemRefs.current,
    registerRef: registerRef
  };
  return /*#__PURE__*/_react.default.createElement(FocusManagerContext.Provider, {
    value: contextValue
  }, children);
};
var _default = exports.default = FocusManager;
