// @atlaskit/editor-plugin-alignment/dist/es2019/ui/FloatingToolbarComponent.js
import React, { useCallback } from 'react';
import { INPUT_METHOD } from '@atlaskit/editor-common/analytics';
import { useSharedPluginState } from '@atlaskit/editor-common/hooks';
import { editorExperiment } from '@atlaskit/tmp-editor-statsig/experiments';
import { changeAlignment } from '../editor-commands';
import { ToolbarType } from '../pm-plugins/types';
import ToolbarAlignment from './ToolbarAlignment';
const FloatingToolbarSettings = {
  disabled: false,
  isToolbarReducedSpacing: true
};
export function FloatingToolbarComponent({
  api,
  editorView
}) {
  const {
    alignmentState
  } = useSharedPluginState(api, ['alignment']);
  const changeAlignmentCallback = useCallback(align => {
    return changeAlignment(align, api, INPUT_METHOD.FLOATING_TB)(editorView.state, editorView.dispatch);
  }, [editorView, api]);
  return /*#__PURE__*/React.createElement(ToolbarAlignment, {
    pluginState: alignmentState,
    isReducedSpacing: editorExperiment('platform_editor_controls', 'variant1') ? false : FloatingToolbarSettings.isToolbarReducedSpacing,
    changeAlignment: changeAlignmentCallback
    // Ignored via go/ees005
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    ,
    disabled: FloatingToolbarSettings.disabled || !alignmentState.isEnabled,
    toolbarType: ToolbarType.FLOATING,
    api: api
  });
}
