// @atlaskit/editor-plugin-alignment/dist/es2019/ui/PrimaryToolbarComponent.js
import React, { useCallback } from 'react';
import { INPUT_METHOD } from '@atlaskit/editor-common/analytics';
import { useSharedPluginState } from '@atlaskit/editor-common/hooks';
import { ToolbarSize } from '@atlaskit/editor-common/types';
import { changeAlignment } from '../editor-commands';
import { ToolbarType } from '../pm-plugins/types';
import ToolbarAlignment from './ToolbarAlignment';
export function PrimaryToolbarComponent({
  api,
  editorView,
  disabled,
  popupsMountPoint,
  popupsBoundariesElement,
  popupsScrollableElement,
  isToolbarReducedSpacing,
  toolbarSize = ToolbarSize.XXL
}) {
  const {
    alignmentState
  } = useSharedPluginState(api, ['alignment']);
  const changeAlignmentCallback = useCallback(align => changeAlignment(align, api, INPUT_METHOD.TOOLBAR)(editorView.state, editorView.dispatch), [editorView, api]);
  return /*#__PURE__*/React.createElement(ToolbarAlignment, {
    pluginState: alignmentState,
    isReducedSpacing: isToolbarReducedSpacing,
    changeAlignment: changeAlignmentCallback
    // Ignored via go/ees005
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    ,
    disabled: disabled || !alignmentState.isEnabled,
    popupsMountPoint: popupsMountPoint,
    popupsBoundariesElement: popupsBoundariesElement,
    popupsScrollableElement: popupsScrollableElement,
    api: api,
    toolbarType: ToolbarType.PRIMARY,
    toolbarSize: toolbarSize
  });
}
