// @atlaskit/smart-card/dist/es2019/view/FlexibleCard/components/elements/lozenge/lozenge-action/lozenge-action-trigger/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./index.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
/* eslint-disable @atlaskit/design-system/use-tokens-typography */

import { useCallback, useMemo, useState } from 'react';
import ChevronDownIcon from '@atlaskit/icon/utility/migration/chevron-down';
import Lozenge from '@atlaskit/lozenge';
import { fg } from '@atlaskit/platform-feature-flags';
import { Box } from '@atlaskit/primitives/compiled';
const styles = {
  chevronDown: "_18u012x7 _1e0c1txw"
};
const triggerLozengeStyles = null;
const triggerButtonStyles = null;
const LozengeActionTrigger = ({
  appearance,
  isOpen,
  testId,
  text,
  triggerRef,
  ...props
}) => {
  const [isBold, setIsBold] = useState(false);
  const onMouseEnter = useCallback(() => setIsBold(true), []);
  const onMouseLeave = useCallback(() => setIsBold(false), []);
  const lozenge = useMemo(() => /*#__PURE__*/React.createElement(Lozenge, {
    appearance: appearance,
    isBold: fg('platform-component-visual-refresh') ? true : isBold
  }, /*#__PURE__*/React.createElement("span", {
    className: ax(["_4cvr1h6o _1e0c1txw"])
  }, /*#__PURE__*/React.createElement("span", null, text), /*#__PURE__*/React.createElement(Box, {
    as: "span",
    xcss: styles.chevronDown
  }, /*#__PURE__*/React.createElement(ChevronDownIcon, {
    color: "currentColor",
    label: "options",
    LEGACY_size: "medium",
    LEGACY_margin: "-4px -8px -4px -7px"
  })))), [appearance, isBold, text]);
  return /*#__PURE__*/React.createElement("button", _extends({
    type: "button"
  }, props, {
    "data-action-open": isOpen,
    "data-testid": `${testId}--trigger`,
    onMouseEnter: onMouseEnter,
    onMouseLeave: onMouseLeave,
    ref: triggerRef,
    className: ax(["_kkk2n7od _ca0qidpf _u5f3idpf _n3tdidpf _19bvidpf _15vhn7od _1dqonqa1 _189eyh40 _1h6d1j28 _bfhk1j28 _syazn7od _80omtlke _ect4n7od _1wybn7od _zg8ln7od _k48pn7od _vwz4idpf _1p1dn7od _19pkv77o _2hwxv77o _otyrv77o _18u0v77o _mizu1r31 _1guj1r31 _p8tl1r31 _ra3xglyw _pbhaglyw _1q2vglyw _128m1o36 _11ow1o36 _j8b81o36 _1bas14y2 _19h614y2 _1d3914y2 _1ejj1vm4 _1u6l1vm4 _6ssh1vm4"])
  }), lozenge);
};
export default LozengeActionTrigger;
