// @atlaskit/user-picker/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CustomType", {
  enumerable: true,
  get: function get() {
    return _types.CustomType;
  }
});
Object.defineProperty(exports, "EmailType", {
  enumerable: true,
  get: function get() {
    return _types.EmailType;
  }
});
Object.defineProperty(exports, "ExternalUserType", {
  enumerable: true,
  get: function get() {
    return _types.ExternalUserType;
  }
});
Object.defineProperty(exports, "GroupType", {
  enumerable: true,
  get: function get() {
    return _types.GroupType;
  }
});
Object.defineProperty(exports, "PopupUserPicker", {
  enumerable: true,
  get: function get() {
    return _PopupUserPicker.PopupUserPicker;
  }
});
Object.defineProperty(exports, "TeamType", {
  enumerable: true,
  get: function get() {
    return _types.TeamType;
  }
});
Object.defineProperty(exports, "UserType", {
  enumerable: true,
  get: function get() {
    return _types.UserType;
  }
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _UserPicker.UserPicker;
  }
});
Object.defineProperty(exports, "isEmail", {
  enumerable: true,
  get: function get() {
    return _utils.isEmail;
  }
});
Object.defineProperty(exports, "isExternalUser", {
  enumerable: true,
  get: function get() {
    return _utils.isExternalUser;
  }
});
Object.defineProperty(exports, "isGroup", {
  enumerable: true,
  get: function get() {
    return _utils.isGroup;
  }
});
Object.defineProperty(exports, "isTeam", {
  enumerable: true,
  get: function get() {
    return _utils.isTeam;
  }
});
Object.defineProperty(exports, "isUser", {
  enumerable: true,
  get: function get() {
    return _utils.isUser;
  }
});
Object.defineProperty(exports, "isValidEmail", {
  enumerable: true,
  get: function get() {
    return _emailValidation.isValidEmail;
  }
});
var _emailValidation = require("./components/emailValidation");
var _UserPicker = require("./components/UserPicker");
var _PopupUserPicker = require("./components/PopupUserPicker");
var _utils = require("./components/utils");
var _types = require("./types");
