// @atlaskit/user-picker/dist/cjs/option.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CustomOption", {
  enumerable: true,
  get: function get() {
    return _CustomOption.default;
  }
});
Object.defineProperty(exports, "EmailOption", {
  enumerable: true,
  get: function get() {
    return _EmailOption.default;
  }
});
Object.defineProperty(exports, "GroupOption", {
  enumerable: true,
  get: function get() {
    return _GroupOption.default;
  }
});
Object.defineProperty(exports, "Option", {
  enumerable: true,
  get: function get() {
    return _Option.Option;
  }
});
Object.defineProperty(exports, "TeamOption", {
  enumerable: true,
  get: function get() {
    return _TeamOption.default;
  }
});
Object.defineProperty(exports, "UserOption", {
  enumerable: true,
  get: function get() {
    return _UserOption.UserOption;
  }
});
var _Option = require("./components/Option");
var _CustomOption = _interopRequireDefault(require("./components/CustomOption"));
var _EmailOption = _interopRequireDefault(require("./components/EmailOption"));
var _GroupOption = _interopRequireDefault(require("./components/GroupOption"));
var _UserOption = require("./components/UserOption");
var _TeamOption = _interopRequireDefault(require("./components/TeamOption"));
