// @atlaskit/media-client-react/dist/es2019/copyIntent/copyIntentAnalytics.js
import { isRequestError } from '@atlaskit/media-client';
const getRequestErrorDetails = err => {
  if (isRequestError(err)) {
    var _err$metadata, _err$metadata2, _err$metadata3, _err$metadata4;
    return {
      request: {
        statusCode: (_err$metadata = err.metadata) === null || _err$metadata === void 0 ? void 0 : _err$metadata.statusCode,
        traceContext: (_err$metadata2 = err.metadata) === null || _err$metadata2 === void 0 ? void 0 : _err$metadata2.traceContext,
        mediaEnv: (_err$metadata3 = err.metadata) === null || _err$metadata3 === void 0 ? void 0 : _err$metadata3.mediaEnv,
        mediaRegion: (_err$metadata4 = err.metadata) === null || _err$metadata4 === void 0 ? void 0 : _err$metadata4.mediaRegion
      }
    };
  }
};
export const getCopyIntentErrorPayload = (error, fileId = '') => {
  return {
    eventType: 'operational',
    action: 'failed',
    actionSubject: 'mediaCopyIntent',
    failReason: isRequestError(error) ? error.reason : 'nativeError',
    attributes: {
      status: 'fail',
      fileAttributes: {
        fileId
      },
      ...getRequestErrorDetails(error)
    }
  };
};
export const getCopyIntentSuccessPayload = (fileId = '') => {
  return {
    eventType: 'operational',
    action: 'succeeded',
    actionSubject: 'mediaCopyIntent',
    attributes: {
      fileAttributes: {
        fileId
      }
    }
  };
};
