// @atlaskit/user-picker/dist/cjs/components/EmailOption/main.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EmailOption = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireDefault(require("react"));
var _reactIntlNext = require("react-intl-next");
var _colors = require("@atlaskit/theme/colors");
var _react2 = require("@emotion/react");
var _AddOptionAvatar = require("../AddOptionAvatar");
var _AvatarItemOption = require("../AvatarItemOption");
var _i18n = require("../i18n");
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); } /**
 * @jsxRuntime classic
 * @jsx jsx
 */ // eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
var getAddEmailMessage = function getAddEmailMessage(validity) {
  return validity === 'POTENTIAL' ? _i18n.messages.continueToAddEmail : _i18n.messages.selectToAddEmail;
};
var EmailOption = exports.EmailOption = /*#__PURE__*/function (_React$PureComponent) {
  function EmailOption() {
    var _this;
    (0, _classCallCheck2.default)(this, EmailOption);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, EmailOption, [].concat(args));
    (0, _defineProperty2.default)(_this, "getLozengeProps", function () {
      return typeof _this.props.email.lozenge === 'string' ? {
        text: _this.props.email.lozenge
      } : _this.props.email.lozenge;
    });
    (0, _defineProperty2.default)(_this, "renderPrimaryText", function () {
      var id = _this.props.email.id;
      return (0, _react2.jsx)("span", {
        key: "name"
        // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
        ,
        css: (0, _AvatarItemOption.textWrapper)(_this.props.isSelected ? "var(--ds-text-selected, ".concat(_colors.B400, ")") : "var(--ds-text, ".concat(_colors.N800, ")"))
      }, id);
    });
    (0, _defineProperty2.default)(_this, "renderSecondaryText", function (label) {
      return (0, _react2.jsx)("span", {
        // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
        css: (0, _AvatarItemOption.textWrapper)(_this.props.isSelected ? "var(--ds-text-selected, ".concat(_colors.B400, ")") : "var(--ds-text-subtlest, ".concat(_colors.N200, ")")),
        "data-testid": "user-picker-email-secondary-text"
      }, label);
    });
    (0, _defineProperty2.default)(_this, "renderOption", function (label) {
      return (0, _react2.jsx)(_AvatarItemOption.AvatarItemOption, {
        avatar: (0, _react2.jsx)(_AddOptionAvatar.AddOptionAvatar, {
          label: label
        }),
        lozenge: _this.getLozengeProps(),
        isDisabled: _this.props.email.isDisabled,
        primaryText: _this.renderPrimaryText(),
        secondaryText: _this.renderSecondaryText(label)
      });
    });
    return _this;
  }
  (0, _inherits2.default)(EmailOption, _React$PureComponent);
  return (0, _createClass2.default)(EmailOption, [{
    key: "render",
    value: function render() {
      var _this2 = this;
      var _this$props = this.props,
        label = _this$props.label,
        emailValidity = _this$props.emailValidity;
      return label !== undefined ? this.renderOption(label) : (0, _react2.jsx)(_reactIntlNext.FormattedMessage, getAddEmailMessage(emailValidity), function (label) {
        return _this2.renderOption(label);
      });
    }
  }]);
}(_react.default.PureComponent);
