// @atlaskit/link-datasource/dist/cjs/ui/issue-like-table/render-type/tag/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TAG_TYPE_TEST_ID = void 0;
var _react = _interopRequireDefault(require("react"));
var _tag = require("@atlaskit/tag");
var TAG_TYPE_TEST_ID = exports.TAG_TYPE_TEST_ID = 'link-datasource-render-type--tag';
var TagRenderType = function TagRenderType(_ref) {
  var _tag$color;
  var tag = _ref.tag,
    _ref$testId = _ref.testId,
    testId = _ref$testId === void 0 ? TAG_TYPE_TEST_ID : _ref$testId;
  var text = tag === null || tag === void 0 ? void 0 : tag.text;
  if (!(text && typeof text === 'string')) {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null);
  }
  return /*#__PURE__*/_react.default.createElement(_tag.SimpleTag, {
    text: text,
    testId: testId,
    appearance: 'default',
    color: (_tag$color = tag === null || tag === void 0 ? void 0 : tag.color) !== null && _tag$color !== void 0 ? _tag$color : 'standard',
    href: tag === null || tag === void 0 ? void 0 : tag.url
  });
};
var _default = exports.default = TagRenderType;
