// @atlaskit/editor-common/dist/cjs/media-single/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.wrappedLayouts = exports.MEDIA_SINGLE_VIDEO_MIN_PIXEL_WIDTH = exports.MEDIA_SINGLE_SNAP_GAP = exports.MEDIA_SINGLE_RESIZE_THROTTLE_TIME = exports.MEDIA_SINGLE_HIGHLIGHT_GAP = exports.MEDIA_SINGLE_HANDLE_MARGIN = exports.MEDIA_SINGLE_GUTTER_SIZE = exports.MEDIA_SINGLE_DEFAULT_MIN_PIXEL_WIDTH = exports.MEDIA_SINGLE_ADJACENT_HANDLE_MARGIN = exports.Layout = exports.DEFAULT_ROUNDING_INTERVAL = exports.DEFAULT_IMAGE_WIDTH = exports.DEFAULT_IMAGE_HEIGHT = exports.CAPTION_PLACEHOLDER_ID = void 0;
var MEDIA_SINGLE_DEFAULT_MIN_PIXEL_WIDTH = exports.MEDIA_SINGLE_DEFAULT_MIN_PIXEL_WIDTH = 24;
var MEDIA_SINGLE_VIDEO_MIN_PIXEL_WIDTH = exports.MEDIA_SINGLE_VIDEO_MIN_PIXEL_WIDTH = 320;
var MEDIA_SINGLE_SNAP_GAP = exports.MEDIA_SINGLE_SNAP_GAP = 3;
var MEDIA_SINGLE_HIGHLIGHT_GAP = exports.MEDIA_SINGLE_HIGHLIGHT_GAP = 10;
var MEDIA_SINGLE_HANDLE_MARGIN = exports.MEDIA_SINGLE_HANDLE_MARGIN = 12;
var MEDIA_SINGLE_ADJACENT_HANDLE_MARGIN = exports.MEDIA_SINGLE_ADJACENT_HANDLE_MARGIN = 8;
var MEDIA_SINGLE_GUTTER_SIZE = exports.MEDIA_SINGLE_GUTTER_SIZE = MEDIA_SINGLE_HANDLE_MARGIN * 2;
var DEFAULT_IMAGE_WIDTH = exports.DEFAULT_IMAGE_WIDTH = 250;
var DEFAULT_IMAGE_HEIGHT = exports.DEFAULT_IMAGE_HEIGHT = 200;
var MEDIA_SINGLE_RESIZE_THROTTLE_TIME = exports.MEDIA_SINGLE_RESIZE_THROTTLE_TIME = 100;
var Layout = exports.Layout = /*#__PURE__*/function (Layout) {
  Layout["FULL_WIDTH"] = "full-width";
  Layout["WIDE"] = "wide";
  Layout["CENTER"] = "center";
  Layout["ALIGN_START"] = "align-start";
  Layout["ALIGN_END"] = "align-end";
  Layout["WRAP_RIGHT"] = "wrap-right";
  Layout["WRAP_LEFT"] = "wrap-left";
  return Layout;
}({});
var wrappedLayouts = exports.wrappedLayouts = ['wrap-left', 'wrap-right', 'align-end', 'align-start'];
var DEFAULT_ROUNDING_INTERVAL = exports.DEFAULT_ROUNDING_INTERVAL = 0.5;
var CAPTION_PLACEHOLDER_ID = exports.CAPTION_PLACEHOLDER_ID = 'caption-placeholder';
