// @atlaskit/editor-common/dist/cjs/media-single/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CAPTION_PLACEHOLDER_ID", {
  enumerable: true,
  get: function get() {
    return _constants.CAPTION_PLACEHOLDER_ID;
  }
});
Object.defineProperty(exports, "CommentBadge", {
  enumerable: true,
  get: function get() {
    return _CommentBadge.CommentBadge;
  }
});
Object.defineProperty(exports, "CommentBadgeNext", {
  enumerable: true,
  get: function get() {
    return _CommentBadgeNext.CommentBadgeNext;
  }
});
Object.defineProperty(exports, "DEFAULT_IMAGE_HEIGHT", {
  enumerable: true,
  get: function get() {
    return _constants.DEFAULT_IMAGE_HEIGHT;
  }
});
Object.defineProperty(exports, "DEFAULT_IMAGE_WIDTH", {
  enumerable: true,
  get: function get() {
    return _constants.DEFAULT_IMAGE_WIDTH;
  }
});
Object.defineProperty(exports, "ExternalImageBadge", {
  enumerable: true,
  get: function get() {
    return _ExternalImageBadge.ExternalImageBadge;
  }
});
Object.defineProperty(exports, "MEDIA_SINGLE_ADJACENT_HANDLE_MARGIN", {
  enumerable: true,
  get: function get() {
    return _constants.MEDIA_SINGLE_ADJACENT_HANDLE_MARGIN;
  }
});
Object.defineProperty(exports, "MEDIA_SINGLE_DEFAULT_MIN_PIXEL_WIDTH", {
  enumerable: true,
  get: function get() {
    return _constants.MEDIA_SINGLE_DEFAULT_MIN_PIXEL_WIDTH;
  }
});
Object.defineProperty(exports, "MEDIA_SINGLE_GUTTER_SIZE", {
  enumerable: true,
  get: function get() {
    return _constants.MEDIA_SINGLE_GUTTER_SIZE;
  }
});
Object.defineProperty(exports, "MEDIA_SINGLE_HIGHLIGHT_GAP", {
  enumerable: true,
  get: function get() {
    return _constants.MEDIA_SINGLE_HIGHLIGHT_GAP;
  }
});
Object.defineProperty(exports, "MEDIA_SINGLE_RESIZE_THROTTLE_TIME", {
  enumerable: true,
  get: function get() {
    return _constants.MEDIA_SINGLE_RESIZE_THROTTLE_TIME;
  }
});
Object.defineProperty(exports, "MEDIA_SINGLE_SNAP_GAP", {
  enumerable: true,
  get: function get() {
    return _constants.MEDIA_SINGLE_SNAP_GAP;
  }
});
Object.defineProperty(exports, "MEDIA_SINGLE_VIDEO_MIN_PIXEL_WIDTH", {
  enumerable: true,
  get: function get() {
    return _constants.MEDIA_SINGLE_VIDEO_MIN_PIXEL_WIDTH;
  }
});
Object.defineProperty(exports, "MediaBadges", {
  enumerable: true,
  get: function get() {
    return _MediaBadges.MediaBadges;
  }
});
Object.defineProperty(exports, "MediaSingleLayout", {
  enumerable: true,
  get: function get() {
    return _constants.Layout;
  }
});
Object.defineProperty(exports, "calcMediaSingleMaxWidth", {
  enumerable: true,
  get: function get() {
    return _utils.calcMediaSingleMaxWidth;
  }
});
Object.defineProperty(exports, "calcMediaSinglePixelWidth", {
  enumerable: true,
  get: function get() {
    return _utils.calcMediaSinglePixelWidth;
  }
});
Object.defineProperty(exports, "calcMinWidth", {
  enumerable: true,
  get: function get() {
    return _utils.calcMinWidth;
  }
});
Object.defineProperty(exports, "calculateOffsetLeft", {
  enumerable: true,
  get: function get() {
    return _utils.calculateOffsetLeft;
  }
});
Object.defineProperty(exports, "currentMediaNodeWithPos", {
  enumerable: true,
  get: function get() {
    return _utils.currentMediaNodeWithPos;
  }
});
Object.defineProperty(exports, "getMaxWidthForNestedNode", {
  enumerable: true,
  get: function get() {
    return _utils.getMaxWidthForNestedNode;
  }
});
Object.defineProperty(exports, "getMaxWidthForNestedNodeNext", {
  enumerable: true,
  get: function get() {
    return _utils.getMaxWidthForNestedNodeNext;
  }
});
Object.defineProperty(exports, "getMediaSingleInitialWidth", {
  enumerable: true,
  get: function get() {
    return _utils.getMediaSingleInitialWidth;
  }
});
Object.defineProperty(exports, "getMediaSinglePixelWidth", {
  enumerable: true,
  get: function get() {
    return _utils.getMediaSinglePixelWidth;
  }
});
Object.defineProperty(exports, "roundToNearest", {
  enumerable: true,
  get: function get() {
    return _utils.roundToNearest;
  }
});
Object.defineProperty(exports, "wrappedLayouts", {
  enumerable: true,
  get: function get() {
    return _constants.wrappedLayouts;
  }
});
var _constants = require("./constants");
var _utils = require("./utils");
var _CommentBadge = require("./CommentBadge");
var _CommentBadgeNext = require("./CommentBadgeNext");
var _ExternalImageBadge = require("./ExternalImageBadge");
var _MediaBadges = require("./MediaBadges");
