// @atlaskit/media-ui/dist/cjs/MediaInlineCard/LoadingView/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MediaInlineCardLoadingView = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _react = _interopRequireDefault(require("react"));
var _Frame = require("../Frame");
var _spinner = _interopRequireDefault(require("@atlaskit/spinner"));
var _IconAndTitleLayout = require("../IconAndTitleLayout");
var _styled = require("./styled");
var _styled2 = require("../IconAndTitleLayout/styled");
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var MediaInlineCardLoadingView = exports.MediaInlineCardLoadingView = /*#__PURE__*/function (_React$Component) {
  function MediaInlineCardLoadingView() {
    (0, _classCallCheck2.default)(this, MediaInlineCardLoadingView);
    return _callSuper(this, MediaInlineCardLoadingView, arguments);
  }
  (0, _inherits2.default)(MediaInlineCardLoadingView, _React$Component);
  return (0, _createClass2.default)(MediaInlineCardLoadingView, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
        message = _this$props.message,
        onClick = _this$props.onClick,
        isSelected = _this$props.isSelected,
        inlinePreloaderStyle = _this$props.inlinePreloaderStyle,
        innerRef = _this$props.innerRef,
        _this$props$testId = _this$props.testId,
        testId = _this$props$testId === void 0 ? 'media-inline-card-loading-view' : _this$props$testId;
      if (inlinePreloaderStyle === 'on-right-without-skeleton') {
        return /*#__PURE__*/_react.default.createElement(_Frame.Frame, {
          innerRef: innerRef,
          testId: testId,
          onClick: onClick,
          isSelected: isSelected
        }, /*#__PURE__*/_react.default.createElement(_styled2.IconTitleWrapper, null, message, /*#__PURE__*/_react.default.createElement(_styled2.RightIconPositionWrapper, null, /*#__PURE__*/_react.default.createElement(_styled.SpinnerWrapper, {
          className: "inline-loading-spinner"
        }, /*#__PURE__*/_react.default.createElement(_spinner.default, {
          size: 14
        })))));
      } else {
        return /*#__PURE__*/_react.default.createElement(_Frame.Frame, {
          innerRef: innerRef,
          testId: testId,
          onClick: onClick,
          isSelected: isSelected
        }, /*#__PURE__*/_react.default.createElement(_IconAndTitleLayout.IconAndTitleLayout, {
          title: message
        }, /*#__PURE__*/_react.default.createElement(_styled.SpinnerWrapper, {
          className: "inline-loading-spinner"
        }, /*#__PURE__*/_react.default.createElement(_spinner.default, {
          size: 14
        }))));
      }
    }
  }]);
}(_react.default.Component);
