// @atlaskit/react-ufo/dist/cjs/assets/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.round = exports.getTypeOfRequest = exports.checkIfTimingsAvailable = exports.calculateTransferType = exports.cacheableTypes = exports.NETWORK_KEY = exports.MEMORY_KEY = exports.DISK_KEY = void 0;
var _selfMeasurements = require("../self-measurements");
/* Borrowed from https://bitbucket.org/atlassian/atlassian-frontend/src/master/packages/performance/browser-metrics/src/plugins/timings/resource.ts */
var cacheableTypes = exports.cacheableTypes = ['script', 'link', 'other'];
var MEMORY_KEY = exports.MEMORY_KEY = 'mem';
var DISK_KEY = exports.DISK_KEY = 'disk';
var NETWORK_KEY = exports.NETWORK_KEY = 'net';
var calculateTransferType = exports.calculateTransferType = (0, _selfMeasurements.withProfiling)(function calculateTransferType(name, type, duration, size) {
  if (!cacheableTypes.includes(type) && !(type === 'other' && name.includes('.js'))) {
    return null;
  }
  if ((size === undefined || size === 0) && duration === 0) {
    return MEMORY_KEY;
  }
  if (size === 0 && duration > 0) {
    return DISK_KEY;
  }
  if (size === undefined) {
    return null;
  }
  return NETWORK_KEY;
});
var getTypeOfRequest = exports.getTypeOfRequest = (0, _selfMeasurements.withProfiling)(function getTypeOfRequest(_ref) {
  var name = _ref.name,
    type = _ref.initiatorType;
  var category = 'other';
  var urlWithoutQuery = name.split('?')[0];
  switch (type) {
    case 'script':
      category = 'js';
      break;
    case 'link':
      if (urlWithoutQuery.endsWith('.css')) {
        category = 'css';
      }
      if (urlWithoutQuery.endsWith('.js')) {
        category = 'js';
      }
      break;
    case 'img':
      category = 'image';
      break;
    case 'font':
      category = 'font';
      break;
    default:
      if (urlWithoutQuery.endsWith('.js')) {
        category = 'js';
      } else if (urlWithoutQuery.endsWith('.css')) {
        category = 'css';
      } else if (urlWithoutQuery.match(/\.(woff|woff2|ttf|otf)$/)) {
        category = 'font';
      } else if (urlWithoutQuery.match(/\.(png|jpg|jpeg|gif|svg)$/)) {
        category = 'image';
      }
      break;
  }
  return category;
});
var checkIfTimingsAvailable = exports.checkIfTimingsAvailable = (0, _selfMeasurements.withProfiling)(function checkIfTimingsAvailable(entry) {
  if (entry.decodedSize === 0 && entry.encodedSize === 0 && entry.requestStart === 0 && entry.responseStart === 0) {
    return false;
  }
  return true;
});
var round = exports.round = (0, _selfMeasurements.withProfiling)(function round(n) {
  if (isNaN(n)) {
    return 0;
  }
  return Math.round(n * 10000) / 10000;
});
