// @atlaskit/editor-plugin-hyperlink/dist/cjs/pm-plugins/decorations.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ButtonWrapper = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactDom = require("react-dom");
var _reactIntlNext = require("react-intl-next");
var _link = require("@atlaskit/editor-common/link");
var ButtonWrapper = exports.ButtonWrapper = function ButtonWrapper(_ref) {
  var editorView = _ref.editorView,
    pos = _ref.pos,
    stateKey = _ref.stateKey,
    intl = _ref.intl,
    onOpenLinkClick = _ref.onOpenLinkClick;
  var wrapper = document.createElement('span');
  wrapper.style.position = 'relative';
  var nonBreakingCharacter = "\u2060";
  var onDropdownChange = function onDropdownChange(isOpen) {
    editorView.dispatch(editorView.state.tr.setMeta(stateKey, {
      type: _link.LinkAction.SET_CONFIGURE_DROPDOWN_OPEN,
      isOpen: isOpen
    }));
  };
  (0, _reactDom.render)( /*#__PURE__*/_react.default.createElement(_reactIntlNext.IntlProvider, {
    locale: intl.locale || 'en',
    messages: intl.messages,
    formats: intl.formats
  }, /*#__PURE__*/_react.default.createElement(_link.OverlayButton, {
    targetElementPos: pos,
    editorView: editorView,
    onDropdownChange: onDropdownChange,
    onOpenLinkClick: onOpenLinkClick
  }), nonBreakingCharacter), wrapper);
  return wrapper;
};
