// @atlaskit/editor-plugin-hyperlink/dist/cjs/pm-plugins/main.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.stateKey = exports.plugin = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _uuid = _interopRequireDefault(require("uuid"));
var _link3 = require("@atlaskit/editor-common/link");
var _safePlugin = require("@atlaskit/editor-common/safe-plugin");
var _utils = require("@atlaskit/editor-common/utils");
var _model = require("@atlaskit/editor-prosemirror/model");
var _state = require("@atlaskit/editor-prosemirror/state");
var _view = require("@atlaskit/editor-prosemirror/view");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _decorations2 = require("./decorations");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var mapTransactionToState = function mapTransactionToState(state, tr) {
  if (!state) {
    return undefined;
  } else if (state.type === _link3.InsertStatus.EDIT_LINK_TOOLBAR || state.type === _link3.InsertStatus.EDIT_INSERTED_TOOLBAR) {
    var _tr$mapping$mapResult = tr.mapping.mapResult(state.pos, 1),
      pos = _tr$mapping$mapResult.pos,
      deleted = _tr$mapping$mapResult.deleted;
    var node = tr.doc.nodeAt(pos);
    // If the position was not deleted & it is still a link
    if (!deleted && !!node.type.schema.marks.link.isInSet(node.marks)) {
      if (node === state.node && pos === state.pos) {
        return state;
      }
      return _objectSpread(_objectSpread({}, state), {}, {
        pos: pos,
        node: node
      });
    }
    // If the position has been deleted, then require a navigation to show the toolbar again
    return;
  } else if (state.type === _link3.InsertStatus.INSERT_LINK_TOOLBAR) {
    return _objectSpread(_objectSpread({}, state), {}, {
      from: tr.mapping.map(state.from),
      to: tr.mapping.map(state.to)
    });
  }
  return;
};
var toState = function toState(state, action, editorState) {
  // Show insert or edit toolbar
  if (!state) {
    switch (action) {
      case _link3.LinkAction.SHOW_INSERT_TOOLBAR:
        {
          var _editorState$selectio = editorState.selection,
            from = _editorState$selectio.from,
            to = _editorState$selectio.to;
          if ((0, _utils.canLinkBeCreatedInRange)(from, to)(editorState)) {
            return {
              type: _link3.InsertStatus.INSERT_LINK_TOOLBAR,
              from: from,
              to: to
            };
          }
          return undefined;
        }
      case _link3.LinkAction.SELECTION_CHANGE:
        // If the user has moved their cursor, see if they're in a link
        var link = (0, _link3.getActiveLinkMark)(editorState);
        if (link) {
          return _objectSpread(_objectSpread({}, link), {}, {
            type: _link3.InsertStatus.EDIT_LINK_TOOLBAR
          });
        }
        return undefined;
      default:
        return undefined;
    }
  }
  // Update toolbar state if selection changes, or if toolbar is hidden
  if (state.type === _link3.InsertStatus.EDIT_LINK_TOOLBAR) {
    switch (action) {
      case _link3.LinkAction.EDIT_INSERTED_TOOLBAR:
        {
          var _link = (0, _link3.getActiveLinkMark)(editorState);
          if (_link) {
            if (_link.pos === state.pos && _link.node === state.node) {
              return _objectSpread(_objectSpread({}, state), {}, {
                type: _link3.InsertStatus.EDIT_INSERTED_TOOLBAR
              });
            }
            return _objectSpread(_objectSpread({}, _link), {}, {
              type: _link3.InsertStatus.EDIT_INSERTED_TOOLBAR
            });
          }
          return undefined;
        }
      case _link3.LinkAction.SELECTION_CHANGE:
        var _link2 = (0, _link3.getActiveLinkMark)(editorState);
        if (_link2) {
          if (_link2.pos === state.pos && _link2.node === state.node) {
            // Make sure we return the same object, if it's the same link
            return state;
          }
          return _objectSpread(_objectSpread({}, _link2), {}, {
            type: _link3.InsertStatus.EDIT_LINK_TOOLBAR
          });
        }
        return undefined;
      case _link3.LinkAction.HIDE_TOOLBAR:
        return undefined;
      default:
        return state;
    }
  }

  // Remove toolbar if user changes selection or toolbar is hidden
  if (state.type === _link3.InsertStatus.INSERT_LINK_TOOLBAR) {
    switch (action) {
      case _link3.LinkAction.SELECTION_CHANGE:
      case _link3.LinkAction.HIDE_TOOLBAR:
        return undefined;
      default:
        return state;
    }
  }
  return;
};
var getActiveText = function getActiveText(selection) {
  var currentSlice = selection.content();
  if (currentSlice.size === 0) {
    return;
  }
  if (currentSlice.content.childCount === 1 && currentSlice.content.firstChild && selection instanceof _state.TextSelection) {
    return currentSlice.content.firstChild.textContent;
  }
  return;
};
var stateKey = exports.stateKey = new _state.PluginKey('hyperlinkPlugin');
var plugin = exports.plugin = function plugin(dispatch, intl, editorAppearance, pluginInjectionApi, onClickCallback, __livePage) {
  return new _safePlugin.SafePlugin({
    state: {
      init: function init(_, state) {
        var canInsertLink = (0, _utils.canLinkBeCreatedInRange)(state.selection.from, state.selection.to)(state);
        return {
          activeText: getActiveText(state.selection),
          canInsertLink: canInsertLink,
          timesViewed: 0,
          activeLinkMark: toState(undefined, _link3.LinkAction.SELECTION_CHANGE, state),
          editorAppearance: editorAppearance
        };
      },
      apply: function apply(tr, pluginState, oldState, newState) {
        var state = pluginState;
        var action = tr.getMeta(stateKey) && tr.getMeta(stateKey).type;
        var inputMethod = tr.getMeta(stateKey) && tr.getMeta(stateKey).inputMethod;
        if (tr.docChanged) {
          state = {
            activeText: state.activeText,
            canInsertLink: (0, _utils.canLinkBeCreatedInRange)(newState.selection.from, newState.selection.to)(newState),
            timesViewed: state.timesViewed,
            inputMethod: inputMethod,
            activeLinkMark: mapTransactionToState(state.activeLinkMark, tr),
            editorAppearance: editorAppearance
          };
        }
        if (action) {
          var stateForAnalytics = [_link3.LinkAction.SHOW_INSERT_TOOLBAR, _link3.LinkAction.EDIT_INSERTED_TOOLBAR].includes(action) ? {
            timesViewed: ++state.timesViewed,
            searchSessionId: (0, _uuid.default)()
          } : {
            timesViewed: state.timesViewed,
            searchSessionId: state.searchSessionId
          };
          state = _objectSpread(_objectSpread({}, __livePage && (0, _platformFeatureFlags.fg)('linking_platform_smart_links_in_live_pages') && state), {}, {
            activeText: state.activeText,
            canInsertLink: state.canInsertLink,
            inputMethod: inputMethod,
            activeLinkMark: toState(state.activeLinkMark, action, newState),
            editorAppearance: editorAppearance
          }, stateForAnalytics);
          if ((0, _platformFeatureFlags.fg)('linking_platform_smart_links_in_live_pages')) {
            var _pluginInjectionApi$e;
            var isViewMode = (pluginInjectionApi === null || pluginInjectionApi === void 0 || (_pluginInjectionApi$e = pluginInjectionApi.editorViewMode) === null || _pluginInjectionApi$e === void 0 || (_pluginInjectionApi$e = _pluginInjectionApi$e.sharedState.currentState()) === null || _pluginInjectionApi$e === void 0 ? void 0 : _pluginInjectionApi$e.mode) === 'view';
            if (__livePage && !isViewMode) {
              if (action === _link3.LinkAction.SET_CONFIGURE_DROPDOWN_OPEN) {
                var configureDropdownOpen = tr.getMeta(stateKey).isOpen;
                // Hide overlay when the dropdown is closed (state is updated to false)
                var decorations = configureDropdownOpen ? {} : {
                  decorations: _view.DecorationSet.empty
                };
                state = _objectSpread(_objectSpread(_objectSpread({}, state), decorations), {}, {
                  configureDropdownOpen: configureDropdownOpen
                });
              }
              if (action === _link3.LinkAction.SET_CONFIGURE_BUTTON_TARGET_POS) {
                var configureButtonTargetPos = tr.getMeta(stateKey).pos;
                var targetPosHasChanged = pluginState.configureButtonTargetPos !== configureButtonTargetPos;
                var _decorations = pluginState.decorations;
                if (targetPosHasChanged && state.configureDropdownOpen !== true) {
                  if (configureButtonTargetPos === undefined) {
                    _decorations = _view.DecorationSet.empty;
                  } else {
                    var decoration = _view.Decoration.widget(configureButtonTargetPos, function (view) {
                      return (0, _decorations2.ButtonWrapper)({
                        editorView: view,
                        pos: configureButtonTargetPos,
                        stateKey: stateKey,
                        intl: intl,
                        onOpenLinkClick: function onOpenLinkClick(event) {
                          var _node$marks$find, _pluginInjectionApi$a;
                          if (configureButtonTargetPos === undefined || typeof configureButtonTargetPos !== 'number') {
                            return;
                          }
                          var node = view.state.tr.doc.nodeAt(configureButtonTargetPos);
                          if (node === null) {
                            return;
                          }
                          var url = (_node$marks$find = node.marks.find(function (mark) {
                            return mark.type.name === 'link';
                          })) === null || _node$marks$find === void 0 ? void 0 : _node$marks$find.attrs.href;
                          if (!url) {
                            return;
                          }
                          (0, _link3.handleNavigation)({
                            fireAnalyticsEvent: pluginInjectionApi === null || pluginInjectionApi === void 0 || (_pluginInjectionApi$a = pluginInjectionApi.analytics) === null || _pluginInjectionApi$a === void 0 ? void 0 : _pluginInjectionApi$a.actions.fireAnalyticsEvent,
                            onClickCallback: onClickCallback,
                            url: url,
                            event: event
                          });
                        }
                      });
                    });
                    _decorations = _view.DecorationSet.create(newState.doc, [decoration]);
                  }
                }
                state = _objectSpread(_objectSpread({}, state), {}, {
                  configureButtonTargetPos: configureButtonTargetPos,
                  decorations: _decorations
                });
              }
            }
          }
        }
        var hasPositionChanged = oldState.selection.from !== newState.selection.from || oldState.selection.to !== newState.selection.to;
        if (tr.selectionSet && hasPositionChanged) {
          state = {
            activeText: getActiveText(newState.selection),
            canInsertLink: (0, _utils.canLinkBeCreatedInRange)(newState.selection.from, newState.selection.to)(newState),
            activeLinkMark: toState(state.activeLinkMark, _link3.LinkAction.SELECTION_CHANGE, newState),
            timesViewed: state.timesViewed,
            searchSessionId: state.searchSessionId,
            inputMethod: inputMethod,
            editorAppearance: editorAppearance
          };
        }
        if (!(0, _utils.shallowEqual)(state, pluginState)) {
          dispatch(stateKey, state);
        }
        return state;
      }
    },
    key: stateKey,
    props: _objectSpread({
      decorations: function decorations(state) {
        if (__livePage && (0, _platformFeatureFlags.fg)('linking_platform_smart_links_in_live_pages')) {
          var _stateKey$getState;
          var _ref = (_stateKey$getState = stateKey.getState(state)) !== null && _stateKey$getState !== void 0 ? _stateKey$getState : {},
            decorations = _ref.decorations;
          return decorations;
        } else {
          return _view.DecorationSet.empty;
        }
      },
      handleDOMEvents: {
        // Ignored via go/ees005
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        mouseup: function mouseup(_, event) {
          // this prevents redundant selection transaction when clicking on link
          // link state will be update on slection change which happens on mousedown
          if (isLinkDirectTarget(event)) {
            event.preventDefault();
            return true;
          }
          return false;
        },
        // Ignored via go/ees005
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        mousedown: function mousedown(view, event) {
          // since link clicks are disallowed by browsers inside contenteditable
          // so we need to handle shift+click selection ourselves in this case
          if (!event.shiftKey || !isLinkDirectTarget(event)) {
            return false;
          }
          var state = view.state;
          var $anchor = state.selection.$anchor;
          var newPosition = view.posAtCoords({
            left: event.clientX,
            top: event.clientY
          });
          if ((newPosition === null || newPosition === void 0 ? void 0 : newPosition.pos) != null && newPosition.pos !== $anchor.pos) {
            var tr = state.tr.setSelection(_state.TextSelection.create(state.doc, $anchor.pos, newPosition.pos));
            view.dispatch(tr);
            return true;
          }
          return false;
        }
      }
    }, __livePage && (0, _platformFeatureFlags.fg)('linking_platform_smart_links_in_live_pages') && {
      markViews: {
        link: function link(mark, view, inline) {
          var toDOM = mark.type.spec.toDOM;
          if (!toDOM) {
            throw new Error('toDom method missing');
          }
          var dom = _model.DOMSerializer.renderSpec(document, toDOM(mark, inline)).dom;
          if (!(dom instanceof HTMLElement)) {
            throw new Error('Error rendering hyperlink spec to dom');
          }
          var setTargetElementPos = function setTargetElementPos(val) {
            var tr = view.state.tr;
            tr.setMeta(stateKey, {
              type: _link3.LinkAction.SET_CONFIGURE_BUTTON_TARGET_POS,
              pos: val
            });
            view.dispatch(tr);
          };
          dom.onmouseenter = function () {
            var _stateKey$getState2;
            var _ref2 = (_stateKey$getState2 = stateKey.getState(view.state)) !== null && _stateKey$getState2 !== void 0 ? _stateKey$getState2 : {},
              activeLinkMark = _ref2.activeLinkMark,
              configureButtonTargetPos = _ref2.configureButtonTargetPos;
            if (!activeLinkMark) {
              var nodePos = view.posAtDOM(dom, -1);
              if (nodePos !== configureButtonTargetPos) {
                setTargetElementPos(nodePos);
              }
            }
          };
          dom.onmouseleave = function () {
            var _stateKey$getState3;
            var _ref3 = (_stateKey$getState3 = stateKey.getState(view.state)) !== null && _stateKey$getState3 !== void 0 ? _stateKey$getState3 : {},
              configureButtonTargetPos = _ref3.configureButtonTargetPos;
            if (configureButtonTargetPos !== undefined) {
              setTargetElementPos(undefined);
            }
          };
          dom.onclick = function (event) {
            if (isDirectTarget(event, dom)) {
              var _pluginInjectionApi$a2;
              var url = mark.attrs.href;
              // event is globalThis.MouseEvent, while handleNavigation
              // (and editor-common OnClickCallback) require React.MouseEvent
              var reactMouseEvent = event;
              (0, _link3.handleNavigation)({
                fireAnalyticsEvent: pluginInjectionApi === null || pluginInjectionApi === void 0 || (_pluginInjectionApi$a2 = pluginInjectionApi.analytics) === null || _pluginInjectionApi$a2 === void 0 ? void 0 : _pluginInjectionApi$a2.actions.fireAnalyticsEvent,
                onClickCallback: onClickCallback,
                url: url,
                event: reactMouseEvent
              });
            }
          };
          return {
            dom: dom
          };
        }
      }
    })
  });
};
function isLinkDirectTarget(event) {
  return (event === null || event === void 0 ? void 0 : event.target) instanceof HTMLElement && event.target.tagName === 'A';
}
function isDirectTarget(event, element) {
  return (event === null || event === void 0 ? void 0 : event.target) instanceof HTMLElement && event.target === element;
}
