// @atlaskit/media-ui/dist/cjs/customMediaPlayer/analytics/utils/playbackAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createPlaybackAttributes = createPlaybackAttributes;
var _mediaCommon = require("@atlaskit/media-common");
function createPlaybackAttributes(playbackState) {
  var isAutoPlay = playbackState.isAutoPlay,
    isHDAvailable = playbackState.isHDAvailable,
    isHDActive = playbackState.isHDActive,
    playerSize = playbackState.playerSize,
    isFullScreenEnabled = playbackState.isFullScreenEnabled,
    playbackSpeed = playbackState.playbackSpeed,
    status = playbackState.status,
    currentTime = playbackState.currentTime,
    volume = playbackState.volume,
    duration = playbackState.duration,
    isMuted = playbackState.isMuted;
  return (0, _mediaCommon.omitBy)({
    isAutoPlay: isAutoPlay,
    isHDAvailable: isHDAvailable,
    isHDActive: isHDActive,
    playbackSpeed: playbackSpeed,
    playerSize: playerSize,
    isFullScreenEnabled: isFullScreenEnabled,
    durationInSec: duration && parseInt("".concat(duration), 10),
    absoluteTimeInSec: currentTime && parseInt("".concat(currentTime), 10),
    relativeTime: duration && currentTime && duration > 0 ? Math.round(currentTime / duration * 100) / 100 : undefined,
    status: status,
    volume: volume,
    isMuted: isMuted
  }, _mediaCommon.isUndefined);
}
