// @atlaskit/teams-public/dist/esm/ui/team-containers/no-product-access-empty-state/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import "./index.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
import React from 'react';
import { defineMessages, FormattedMessage } from 'react-intl-next';
import Heading from '@atlaskit/heading';
import { Box, Inline, Stack } from '@atlaskit/primitives/compiled';
import { NoProductAccessIcon } from './no-product-access-icon';
var styles = {
  emptyStateContainer: "_2rkoiti9 _189ee4h9 _1h6dmuej _18zrxy5q _12y3e4h9 _12jimuej _1qu2nqa1",
  contentWrapper: "_bozgxy5q"
};
export var NoProductAccessState = function NoProductAccessState() {
  return /*#__PURE__*/React.createElement(Stack, {
    space: "space.100",
    xcss: styles.emptyStateContainer,
    testId: "team-containers-no-product-access-state"
  }, /*#__PURE__*/React.createElement(Inline, {
    alignBlock: "center"
  }, /*#__PURE__*/React.createElement(NoProductAccessIcon, null), /*#__PURE__*/React.createElement(Box, {
    xcss: styles.contentWrapper
  }, /*#__PURE__*/React.createElement(Stack, {
    space: "space.100"
  }, /*#__PURE__*/React.createElement(Heading, {
    size: "small"
  }, /*#__PURE__*/React.createElement(FormattedMessage, messages.teamContainerEmptyStateTitle)), /*#__PURE__*/React.createElement(FormattedMessage, messages.teamContainerEmptyStateDescription)))));
};
var messages = defineMessages({
  teamContainerEmptyStateTitle: {
    id: 'ptc.team-profile-page.team-containers.empty-state.title',
    defaultMessage: 'Add the places your team works',
    description: 'Empty state title when user has no product access'
  },
  teamContainerEmptyStateDescription: {
    id: 'ptc.team-profile-page.team-containers.empty-state.description',
    defaultMessage: 'Your team can link their Jira projects, Confluence spaces here to show where they work',
    description: 'Empty state description when user has no product access'
  }
});
