// @atlaskit/editor-plugin-table/dist/cjs/pm-plugins/decorations/utils/column-resizing.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.clearColumnResizingDecorations = exports.buildColumnResizingDecorations = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _types = require("../../../types");
var _decoration = require("../../utils/decoration");
var _composeDecorations = require("./compose-decorations");
var emptyDecorations = [[], []];
var updateColumnResizeHandle = function updateColumnResizeHandle(columnResizesDecorations) {
  return function (_ref) {
    var decorationSet = _ref.decorationSet,
      tr = _ref.tr;
    return (0, _decoration.updateDecorations)(tr.doc, decorationSet, columnResizesDecorations, _types.TableDecorations.COLUMN_RESIZING_HANDLE_WIDGET);
  };
};
var updateLastCellElement = function updateLastCellElement(lastCellElementsDecorations) {
  return function (_ref2) {
    var decorationSet = _ref2.decorationSet,
      tr = _ref2.tr;
    return (0, _decoration.updateDecorations)(tr.doc, decorationSet, lastCellElementsDecorations, _types.TableDecorations.LAST_CELL_ELEMENT);
  };
};
var buildColumnResizingDecorations = exports.buildColumnResizingDecorations = function buildColumnResizingDecorations(rowEndIndex, columnEndIndex, includeTooltip, getIntl, nodeViewPortalProviderAPI) {
  return function (_ref3) {
    var tr = _ref3.tr,
      decorationSet = _ref3.decorationSet;
    var _ref4 = columnEndIndex < 0 ? emptyDecorations : (0, _decoration.createResizeHandleDecoration)(tr, rowEndIndex, {
        right: columnEndIndex
      }, includeTooltip, getIntl, nodeViewPortalProviderAPI),
      _ref5 = (0, _slicedToArray2.default)(_ref4, 2),
      columnResizesDecorations = _ref5[0],
      lastCellElementsDecorations = _ref5[1];
    return (0, _composeDecorations.composeDecorations)([updateColumnResizeHandle(columnResizesDecorations), updateLastCellElement(lastCellElementsDecorations)])({
      decorationSet: decorationSet,
      tr: tr
    });
  };
};
var clearColumnResizingDecorations = exports.clearColumnResizingDecorations = function clearColumnResizingDecorations() {
  return function (_ref6) {
    var tr = _ref6.tr,
      decorationSet = _ref6.decorationSet;
    var columnResizesDecorations = emptyDecorations[0],
      lastCellElementsDecorations = emptyDecorations[1];
    return (0, _composeDecorations.composeDecorations)([updateColumnResizeHandle(columnResizesDecorations), updateLastCellElement(lastCellElementsDecorations)])({
      decorationSet: decorationSet,
      tr: tr
    });
  };
};
