// @atlaskit/editor-common/dist/cjs/link/ConfigureLinkOverlay/handleNavigation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.handleNavigation = handleNavigation;
var _analytics = require("../../analytics");
/**
 * Function that opens a new page and fires the relevant analytics events
 */
function handleNavigation(_ref) {
  var url = _ref.url,
    event = _ref.event,
    onClickCallback = _ref.onClickCallback,
    fireAnalyticsEvent = _ref.fireAnalyticsEvent;
  if (fireAnalyticsEvent) {
    fireAnalyticsEvent({
      action: _analytics.ACTION.VISITED,
      actionSubject: _analytics.ACTION_SUBJECT.LINK,
      eventType: _analytics.EVENT_TYPE.TRACK,
      attributes: {
        platform: _analytics.PLATFORMS.WEB,
        mode: _analytics.MODE.EDITOR
      }
    });
  }
  if (url) {
    try {
      onClickCallback === null || onClickCallback === void 0 || onClickCallback({
        event: event,
        url: url
      });
    } catch (_unused) {}
    /**
     * Links should navigate by default in live pages if:
     * - the link is the direct target of the click event
     * - default handling wasn't prevented with `event.preventDefault()`
     */
    if (!event.defaultPrevented) {
      window.location.href = url;
    }
  }
}
