// @atlaskit/editor-common/dist/cjs/link/ConfigureLinkOverlay/StyledButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StyledButton = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = require("@emotion/react");
var _button = _interopRequireDefault(require("@atlaskit/button"));
var _excluded = ["innerRef"];
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
var buttonStyles = (0, _react.css)({
  display: 'flex',
  background: "var(--ds-background-neutral, #091E420F)",
  color: "var(--ds-icon, #44546F)",
  '&:hover': {
    background: "var(--ds-background-neutral-hovered, #091E4224)"
  },
  '&:active': {
    background: "var(--ds-background-neutral-pressed, #091E424F)"
  },
  width: '1.375rem',
  height: '1.25rem'
});
var StyledButton = exports.StyledButton = function StyledButton(_ref) {
  var innerRef = _ref.innerRef,
    props = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  return (0, _react.jsx)(_button.default
  // Ignored via go/ees005
  // eslint-disable-next-line react/jsx-props-no-spreading
  , (0, _extends2.default)({}, props, {
    ref: innerRef
    // eslint-disable-next-line @atlaskit/design-system/no-unsafe-style-overrides
    ,
    css: buttonStyles
  }));
};
