// @atlaskit/link-datasource/dist/cjs/ui/assets-modal/modal/render-assets-content/initial-state-view/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InitialStateView = void 0;
require("./index.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _reactIntlNext = require("react-intl-next");
var _link = _interopRequireDefault(require("@atlaskit/link"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _typography = require("@atlaskit/theme/typography");
var _search = require("../../../../../common/ui/rich-icon/search");
var _crystalBallSvgOld = require("./assets/crystal-ball-svg-old");
var _messages = require("./messages");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var initialStateViewContainerStyles = null;
var svgAndTextsWrapperStylesOld = null;
var svgAndTextsWrapperStyles = null;
var searchTitleStylesOld = null;
var searchTitleStyles = null;
var mainTextStyles = null;
var AQLSupportDocumentLink = 'https://support.atlassian.com/jira-service-management-cloud/docs/use-assets-query-language-aql/';
var InitialStateView = exports.InitialStateView = function InitialStateView() {
  var _useIntl = (0, _reactIntlNext.useIntl)(),
    formatMessage = _useIntl.formatMessage;
  return /*#__PURE__*/React.createElement("div", {
    "data-testid": "assets-aql-datasource-modal--initial-state-view",
    className: (0, _runtime.ax)(["_1e0c1txw _1bah1h6o _4t3i1osq"])
  }, /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)([(0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv') ? "_y3gn1h6o _1wpz1h6o _ca0qoahv _n3tdoahv _19bvoahv _u5f3oahv _p12f1kdj" : "_y3gn1h6o _1wpz1h6o"])
  }, (0, _platformFeatureFlags.fg)('bandicoots-update-sllv-icons') ? /*#__PURE__*/React.createElement(_search.RichIconSearch, {
    size: "xlarge",
    alt: formatMessage((0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv') ? _messages.initialStateViewMessages.searchTitle : _messages.initialStateViewMessages.searchTitleOld)
  }) : /*#__PURE__*/React.createElement(_crystalBallSvgOld.CrystalBallSVGOld, null), /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)([(0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv') ? "_11c8lodh _syaz1fxt _ca0q1ejb _n3tdpxbi" : "_ca0qpxbi _n3tdu2gc _wtt4ykg7 _pfzw85h9 _zzcc1umn _x3fx1hgc _v2gl1yp8 _1y0bdipb _z7oy1jq3"])
  }, formatMessage((0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv') ? _messages.initialStateViewMessages.searchTitle : _messages.initialStateViewMessages.searchTitleOld)), /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)([(0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv') && "_syaz1gjq"])
  }, formatMessage(_messages.initialStateViewMessages.searchDescription)), (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv') ? /*#__PURE__*/React.createElement(_link.default, {
    href: AQLSupportDocumentLink
  }, formatMessage(_messages.initialStateViewMessages.learnMoreLink)) : /*#__PURE__*/React.createElement("a", {
    href: AQLSupportDocumentLink,
    target: "_blank"
  }, formatMessage(_messages.initialStateViewMessages.learnMoreLinkOld))));
};
