// @atlaskit/smart-card/dist/es2019/view/common/ai-prism/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import "./index.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { useThemeObserver } from '@atlaskit/tokens';
import AIGlowingBorder from './ai-glowing-border';
import { AI_BORDER_PALETTE } from './constants';
const contentStyles = null;
const contentStylesPrismVisible = null;
const animatedSvgContainerStyles = {
  true: "_v564fs5d _tzy4kb7n",
  false: "_v564fs5d _tzy4idpf"
};
const popupContainerStyles = null;
const AIPrism = ({
  children,
  isGlowing = true,
  isMoving = true,
  isVisible,
  testId
}) => {
  var _AI_BORDER_PALETTE$co;
  const {
    colorMode = 'light'
  } = useThemeObserver();
  return /*#__PURE__*/React.createElement(AIGlowingBorder, {
    palette: (_AI_BORDER_PALETTE$co = AI_BORDER_PALETTE[colorMode]) !== null && _AI_BORDER_PALETTE$co !== void 0 ? _AI_BORDER_PALETTE$co : AI_BORDER_PALETTE.light,
    isGlowing: isGlowing,
    isMoving: isMoving,
    testId: testId,
    className: ax([animatedSvgContainerStyles[isVisible ? 'true' : 'false']])
  }, /*#__PURE__*/React.createElement("div", {
    className: ax(["_2rkopd34 _bfhk1bhr _16qsie2t", "_v5649spz", isVisible && "_16qs1aef"])
  }, children));
};
export default AIPrism;
