// @atlaskit/editor-core/dist/cjs/plugins/base/pm-plugins/filter-steps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var prosemirror_state_1 = require("prosemirror-state");
var sendLogs_1 = require("../../../utils/sendLogs");
var hasInvalidSteps = function (tr) {
    return (tr.steps || []).some(function (step) { return step.from > step.to; });
};
exports.default = (function () {
    return new prosemirror_state_1.Plugin({
        filterTransaction: function (tr) {
            if (hasInvalidSteps(tr)) {
                // eslint-disable-next-line no-console
                console.warn('The transaction was blocked because it contains invalid steps', tr.steps);
                sendLogs_1.sendLogs({
                    events: [
                        {
                            name: 'atlaskit.fabric.editor.invalidstep',
                            product: 'atlaskit',
                            properties: {
                                message: 'Blocked transaction with invalid steps',
                            },
                            serverTime: new Date().getTime(),
                            server: 'local',
                            user: '-',
                        },
                    ],
                });
                return false;
            }
            return true;
        },
    });
});
