// @atlaskit/editor-plugin-block-type/dist/cjs/ui/consts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "BLOCK_QUOTE", {
  enumerable: true,
  get: function get() {
    return _blockTypes.BLOCK_QUOTE;
  }
});
Object.defineProperty(exports, "CODE_BLOCK", {
  enumerable: true,
  get: function get() {
    return _blockTypes.CODE_BLOCK;
  }
});
Object.defineProperty(exports, "HEADING_1", {
  enumerable: true,
  get: function get() {
    return _blockTypes.HEADING_1;
  }
});
Object.defineProperty(exports, "HEADING_2", {
  enumerable: true,
  get: function get() {
    return _blockTypes.HEADING_2;
  }
});
Object.defineProperty(exports, "HEADING_3", {
  enumerable: true,
  get: function get() {
    return _blockTypes.HEADING_3;
  }
});
Object.defineProperty(exports, "HEADING_4", {
  enumerable: true,
  get: function get() {
    return _blockTypes.HEADING_4;
  }
});
Object.defineProperty(exports, "HEADING_5", {
  enumerable: true,
  get: function get() {
    return _blockTypes.HEADING_5;
  }
});
Object.defineProperty(exports, "HEADING_6", {
  enumerable: true,
  get: function get() {
    return _blockTypes.HEADING_6;
  }
});
Object.defineProperty(exports, "NORMAL_TEXT", {
  enumerable: true,
  get: function get() {
    return _blockTypes.NORMAL_TEXT;
  }
});
Object.defineProperty(exports, "PANEL", {
  enumerable: true,
  get: function get() {
    return _blockTypes.PANEL;
  }
});
var _blockTypes = require("../pm-plugins/block-types");
