// @atlaskit/link-datasource/dist/es2019/hooks/useObjectSchemas.js
import { useCallback, useState } from 'react';
import { fetchObjectSchemas } from '../services/cmdbService';
export const useObjectSchemas = workspaceId => {
  const [loading, setLoading] = useState(false);
  const [objectSchemas, setObjectSchemas] = useState();
  const [totalObjectSchemas, setTotalObjectSchemas] = useState();
  const [error, setError] = useState();
  const request = useCallback(async query => {
    setLoading(true);
    setError(undefined);
    let fetchedObjectSchemas;
    let fetchTotalObjectSchemas;
    try {
      const fetchedObjectSchemasResponse = await fetchObjectSchemas(workspaceId, query);
      setObjectSchemas(fetchedObjectSchemasResponse.values);
      setTotalObjectSchemas(fetchedObjectSchemasResponse.total);
      fetchedObjectSchemas = fetchedObjectSchemasResponse.values;
      fetchTotalObjectSchemas = fetchedObjectSchemasResponse.total;
    } catch (err) {
      if (err instanceof Error) {
        setError(err);
      } else {
        setError(new Error('Unexpected error occured'));
      }
    } finally {
      setLoading(false);
    }
    return {
      objectSchemas: fetchedObjectSchemas,
      totalObjectSchemas: fetchTotalObjectSchemas
    };
  }, [workspaceId]);
  return {
    objectSchemasLoading: loading,
    objectSchemasError: error,
    objectSchemas,
    totalObjectSchemas,
    fetchObjectSchemas: request
  };
};
