// @atlaskit/media-ui/dist/cjs/test-helpers/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "atlassianLogoUrl", {
  enumerable: true,
  get: function get() {
    return _atlassianLogoUrl.atlassianLogoUrl;
  }
});
Object.defineProperty(exports, "disableMockGlobalImage", {
  enumerable: true,
  get: function get() {
    return _MockGlobalImage.disableMockGlobalImage;
  }
});
Object.defineProperty(exports, "enableMockGlobalImage", {
  enumerable: true,
  get: function get() {
    return _MockGlobalImage.enableMockGlobalImage;
  }
});
Object.defineProperty(exports, "fakeIntl", {
  enumerable: true,
  get: function get() {
    return _fakeI18n.fakeIntl;
  }
});
Object.defineProperty(exports, "mockFileReader", {
  enumerable: true,
  get: function get() {
    return _fileReader.mockFileReader;
  }
});
Object.defineProperty(exports, "mockFileReaderError", {
  enumerable: true,
  get: function get() {
    return _fileReader.mockFileReaderError;
  }
});
Object.defineProperty(exports, "mockFileReaderWithError", {
  enumerable: true,
  get: function get() {
    return _fileReader.mockFileReaderWithError;
  }
});
Object.defineProperty(exports, "mountWithIntlContext", {
  enumerable: true,
  get: function get() {
    return _mountWithIntlContext.mountWithIntlContext;
  }
});
Object.defineProperty(exports, "mountWithIntlWrapper", {
  enumerable: true,
  get: function get() {
    return _mountWithIntlContext.mountWithIntlWrapper;
  }
});
Object.defineProperty(exports, "renderWithIntl", {
  enumerable: true,
  get: function get() {
    return _renderWithIntl.renderWithIntl;
  }
});
Object.defineProperty(exports, "shallowWithIntlContext", {
  enumerable: true,
  get: function get() {
    return _mountWithIntlContext.shallowWithIntlContext;
  }
});
Object.defineProperty(exports, "unmockFileReader", {
  enumerable: true,
  get: function get() {
    return _fileReader.unmockFileReader;
  }
});
var _atlassianLogoUrl = require("./atlassianLogoUrl");
var _fileReader = require("./fileReader");
var _MockGlobalImage = require("./MockGlobalImage");
var _mountWithIntlContext = require("./mountWithIntlContext");
var _fakeI18n = require("./fakeI18n");
var _renderWithIntl = require("./renderWithIntl");
