// @atlaskit/media-ui/dist/cjs/test-helpers/mountWithIntlContext.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.shallowWithIntlContext = exports.mountWithIntlWrapper = exports.mountWithIntlContext = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireDefault(require("react"));
var _reactIntlNext = require("react-intl-next");
var _enzyme = require("enzyme");
var _excluded = ["context", "childContextTypes"],
  _excluded2 = ["context"]; // eslint-disable-next-line import/no-extraneous-dependencies
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var mockIntl = (0, _reactIntlNext.createIntl)({
  locale: 'en'
});

/**
 * When using React-Intl `injectIntl` on components, props.intl is required.
 */
function nodeWithIntlProp(node) {
  var intl = !!node.props.intl ? node.props.intl : mockIntl;
  return /*#__PURE__*/_react.default.createElement(_reactIntlNext.RawIntlProvider, {
    value: intl
  }, /*#__PURE__*/_react.default.cloneElement(node, {
    intl: intl
  }));
}

/* TODO: We are explicitly using the third arg of ReactWrapper to work around the following TS issue which prevents a d.ts from being generated
 * and therefore fails the build:
 * error TS2742: The inferred type of 'mountWithIntlContext' cannot be named without a reference to 'react-transition-group/node_modules/@types/react'. This is likely not portable. A type annotation is necessary.
 * TS is resolving enzyme's usage of react to react-transition-group???
 */
var mountWithIntlContext = exports.mountWithIntlContext = function mountWithIntlContext(node) {
  var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var _ref$context = _ref.context,
    context = _ref$context === void 0 ? {} : _ref$context,
    _ref$childContextType = _ref.childContextTypes,
    childContextTypes = _ref$childContextType === void 0 ? {} : _ref$childContextType,
    additionalOptions = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  var intl = !!node.props.intl ? node.props.intl : mockIntl;
  return (0, _enzyme.mount)(nodeWithIntlProp(node), _objectSpread({
    context: _objectSpread({
      intl: intl
    }, context)
  }, additionalOptions));
};
var shallowWithIntlContext = exports.shallowWithIntlContext = function shallowWithIntlContext(node) {
  var _ref2 = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var _ref2$context = _ref2.context,
    context = _ref2$context === void 0 ? {} : _ref2$context,
    additionalOptions = (0, _objectWithoutProperties2.default)(_ref2, _excluded2);
  var intl = !!node.props.intl ? node.props.intl : mockIntl;
  return (0, _enzyme.shallow)(nodeWithIntlProp(node), _objectSpread({
    context: _objectSpread({
      intl: intl
    }, context)
  }, additionalOptions));
};
var mountWithIntlWrapper = exports.mountWithIntlWrapper = function mountWithIntlWrapper(node) {
  return (0, _enzyme.mount)( /*#__PURE__*/_react.default.createElement(function (props) {
    return /*#__PURE__*/_react.default.createElement(_reactIntlNext.IntlProvider, {
      locale: "en"
    }, /*#__PURE__*/_react.default.cloneElement(node, _objectSpread({}, props)));
  }));
};
