// @atlaskit/link-datasource/dist/cjs/ui/issue-like-table/render-type/richtext/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.parseRichText = exports.default = void 0;
require("./index.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _schemaDefault = require("@atlaskit/adf-schema/schema-default");
var _model = require("@atlaskit/editor-prosemirror/model");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var rootStyles = null;
var smartLinkNodeToUrl = function smartLinkNodeToUrl(node) {
  var _node$attrs$url;
  return (_node$attrs$url = node.attrs.url) !== null && _node$attrs$url !== void 0 ? _node$attrs$url : '';
};
var schemaSmartLinkOverride = new _model.Schema({
  nodes: _schemaDefault.defaultSchema.spec.nodes.update('inlineCard', _objectSpread(_objectSpread({}, _schemaDefault.defaultSchema.spec.nodes.get('inlineCard')), {}, {
    leafText: smartLinkNodeToUrl
  })).update('blockCard', _objectSpread(_objectSpread({}, _schemaDefault.defaultSchema.spec.nodes.get('blockCard')), {}, {
    leafText: smartLinkNodeToUrl
  })).update('embedCard', _objectSpread(_objectSpread({}, _schemaDefault.defaultSchema.spec.nodes.get('embedCard')), {}, {
    leafText: smartLinkNodeToUrl
  })),
  marks: _schemaDefault.defaultSchema.spec.marks
});
var parseRichText = exports.parseRichText = function parseRichText(value) {
  try {
    if (value.type === 'adf') {
      var adf = JSON.parse(value.text);
      return _model.Node.fromJSON(schemaSmartLinkOverride, _objectSpread(_objectSpread({}, adf), {}, {
        content: (0, _toConsumableArray2.default)(adf.content.slice(0, 2))
      })).textContent;
    }
    return null;
  } catch (e) {
    // eslint-disable-next-line no-console
    console.error('error parsing adf', e);
    return null;
  }
};
var RichTextType = function RichTextType(_ref) {
  var value = _ref.value;
  var adfPlainText = (0, _react.useMemo)(function () {
    return parseRichText(value);
  }, [value]);
  if (adfPlainText) {
    return /*#__PURE__*/React.createElement("span", {
      "data-testid": "richtext-plaintext",
      className: (0, _runtime.ax)(["_kqswh2mm"])
    }, adfPlainText);
  } else {
    return /*#__PURE__*/React.createElement("span", {
      "data-testid": "richtext-unsupported"
    });
  }
};
var _default = exports.default = RichTextType;
