// @atlaskit/editor-wikimarkup-transformer/dist/cjs/parser/abstract-tree.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _text = require("./text");
var _normalize = require("./utils/normalize");
var AbstractTree = exports.default = /*#__PURE__*/function () {
  function AbstractTree(schema, wikiMarkup) {
    (0, _classCallCheck2.default)(this, AbstractTree);
    this.schema = schema;
    this.wikiMarkup = wikiMarkup;
  }

  /**
   * Convert reduced macros tree into prosemirror model tree
   */
  return (0, _createClass2.default)(AbstractTree, [{
    key: "getProseMirrorModel",
    value: function getProseMirrorModel(context) {
      var content = (0, _text.parseString)({
        context: context,
        ignoreTokenTypes: [],
        input: this.wikiMarkup,
        schema: this.schema
      });
      return this.schema.nodes.doc.createChecked({}, (0, _normalize.normalizePMNodes)(content, this.schema, 'doc'));
    }
  }]);
}();
