// @atlaskit/editor-plugin-media/dist/cjs/mediaPlugin.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mediaPlugin = void 0;
var _react = _interopRequireWildcard(require("react"));
var _analytics = require("@atlaskit/editor-common/analytics");
var _hooks = require("@atlaskit/editor-common/hooks");
var _messages = require("@atlaskit/editor-common/messages");
var _quickInsert = require("@atlaskit/editor-common/quick-insert");
var _safePlugin = require("@atlaskit/editor-common/safe-plugin");
var _state = require("@atlaskit/editor-prosemirror/state");
var _mediaCommon = require("@atlaskit/media-common");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _experiments = require("@atlaskit/tmp-editor-statsig/experiments");
var _lazyMedia = require("./nodeviews/lazy-media");
var _lazyMediaGroup = require("./nodeviews/lazy-media-group");
var _lazyMediaInline = require("./nodeviews/lazy-media-inline");
var _lazyMediaSingle = require("./nodeviews/lazy-media-single");
var _media = require("./nodeviews/toDOM-fixes/media");
var _mediaGroup = require("./nodeviews/toDOM-fixes/mediaGroup");
var _mediaInline = require("./nodeviews/toDOM-fixes/mediaInline");
var _mediaSingle = require("./nodeviews/toDOM-fixes/mediaSingle");
var _altText = require("./pm-plugins/alt-text");
var _keymap = _interopRequireDefault(require("./pm-plugins/alt-text/keymap"));
var _commands = require("./pm-plugins/commands");
var _keymap2 = _interopRequireDefault(require("./pm-plugins/keymap"));
var _keymapMedia = _interopRequireDefault(require("./pm-plugins/keymap-media"));
var _linking = _interopRequireDefault(require("./pm-plugins/linking"));
var _keymap3 = _interopRequireDefault(require("./pm-plugins/linking/keymap"));
var _main = require("./pm-plugins/main");
var _pixelResizing = require("./pm-plugins/pixel-resizing");
var _pluginKey = require("./pm-plugins/plugin-key");
var _mediaCommon2 = require("./pm-plugins/utils/media-common");
var _mediaSingle2 = require("./pm-plugins/utils/media-single");
var _MediaPicker = require("./ui/MediaPicker");
var _PortalWrapper = require("./ui/MediaViewer/PortalWrapper");
var _toolbar = require("./ui/toolbar");
var _ToolbarMedia = _interopRequireDefault(require("./ui/ToolbarMedia"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var MediaPickerFunctionalComponent = function MediaPickerFunctionalComponent(_ref) {
  var api = _ref.api,
    editorDomElement = _ref.editorDomElement,
    appearance = _ref.appearance;
  var _useSharedPluginState = (0, _hooks.useSharedPluginState)(api, ['media']),
    mediaState = _useSharedPluginState.mediaState;
  if (!mediaState) {
    return null;
  }
  return /*#__PURE__*/_react.default.createElement(_MediaPicker.MediaPickerComponents, {
    editorDomElement: editorDomElement,
    mediaState: mediaState,
    appearance: appearance,
    api: api
  });
};
var MediaViewerFunctionalComponent = function MediaViewerFunctionalComponent(_ref2) {
  var api = _ref2.api,
    editorView = _ref2.editorView;
  var _useSharedPluginState2 = (0, _hooks.useSharedPluginState)(api, ['media']),
    mediaState = _useSharedPluginState2.mediaState;

  // Only traverse document once when media viewer is visible, media viewer items will not update
  // when document changes are made while media viewer is open

  var mediaItems = (0, _react.useMemo)(function () {
    if (mediaState !== null && mediaState !== void 0 && mediaState.isMediaViewerVisible) {
      var mediaNodes = (0, _mediaCommon2.extractMediaNodes)(editorView.state.doc);
      return (0, _mediaCommon2.createMediaIdentifierArray)(mediaNodes);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps -- only update mediaItems once when media viewer is visible
  }, [mediaState === null || mediaState === void 0 ? void 0 : mediaState.isMediaViewerVisible]);

  // Viewer does not have required attributes to render the media viewer
  if (!(mediaState !== null && mediaState !== void 0 && mediaState.isMediaViewerVisible) || !(mediaState !== null && mediaState !== void 0 && mediaState.mediaViewerSelectedMedia) || !(mediaState !== null && mediaState !== void 0 && mediaState.mediaClientConfig)) {
    return null;
  }
  var handleOnClose = function handleOnClose() {
    // Run Command to hide the media viewer
    api === null || api === void 0 || api.core.actions.execute(api === null || api === void 0 ? void 0 : api.media.commands.hideMediaViewer);
  };
  return /*#__PURE__*/_react.default.createElement(_PortalWrapper.RenderMediaViewer, {
    mediaClientConfig: mediaState === null || mediaState === void 0 ? void 0 : mediaState.mediaClientConfig,
    onClose: handleOnClose,
    selectedNodeAttrs: mediaState.mediaViewerSelectedMedia,
    items: mediaItems
  });
};
var mediaPlugin = exports.mediaPlugin = function mediaPlugin(_ref3) {
  var _ref3$config = _ref3.config,
    options = _ref3$config === void 0 ? {} : _ref3$config,
    api = _ref3.api;
  var previousMediaProvider;
  return {
    name: 'media',
    getSharedState: function getSharedState(editorState) {
      if (!editorState) {
        return null;
      }
      return _pluginKey.stateKey.getState(editorState) || null;
    },
    actions: {
      insertMediaAsMediaSingle: function insertMediaAsMediaSingle(view, node, inputMethod, insertMediaVia) {
        var _api$analytics;
        return (0, _mediaSingle2.insertMediaAsMediaSingle)(view, node, inputMethod, api === null || api === void 0 || (_api$analytics = api.analytics) === null || _api$analytics === void 0 ? void 0 : _api$analytics.actions, insertMediaVia, options === null || options === void 0 ? void 0 : options.allowPixelResizing);
      },
      setProvider: function setProvider(provider) {
        var _api$core$actions$exe;
        // Prevent someone trying to set the exact same provider twice for performance reasons
        if (previousMediaProvider === provider) {
          return false;
        }
        previousMediaProvider = provider;
        return (_api$core$actions$exe = api === null || api === void 0 ? void 0 : api.core.actions.execute(function (_ref4) {
          var tr = _ref4.tr;
          return tr.setMeta(_pluginKey.stateKey, {
            mediaProvider: provider
          });
        })) !== null && _api$core$actions$exe !== void 0 ? _api$core$actions$exe : false;
      }
    },
    commands: {
      showMediaViewer: _commands.showMediaViewer,
      hideMediaViewer: _commands.hideMediaViewer,
      trackMediaPaste: _commands.trackMediaPaste
    },
    nodes: function nodes() {
      var _ref5 = options || {},
        _ref5$allowMediaGroup = _ref5.allowMediaGroup,
        allowMediaGroup = _ref5$allowMediaGroup === void 0 ? true : _ref5$allowMediaGroup,
        _ref5$allowMediaSingl = _ref5.allowMediaSingle,
        allowMediaSingle = _ref5$allowMediaSingl === void 0 ? false : _ref5$allowMediaSingl,
        _ref5$allowPixelResiz = _ref5.allowPixelResizing,
        allowPixelResizing = _ref5$allowPixelResiz === void 0 ? false : _ref5$allowPixelResiz,
        allowCaptions = _ref5.allowCaptions,
        allowMediaInlineImages = _ref5.allowMediaInlineImages,
        mediaFeatureFlags = _ref5.featureFlags;
      var allowMediaInline = (0, _platformFeatureFlags.fg)('platform_editor_remove_media_inline_feature_flag') ? allowMediaInlineImages : (0, _mediaCommon.getMediaFeatureFlag)('mediaInline', mediaFeatureFlags);
      var mediaSingleOption = {
        withCaption: allowCaptions,
        withExtendedWidthTypes: allowPixelResizing
      };
      return [{
        name: 'mediaGroup',
        node: (0, _mediaGroup.mediaGroupSpecWithFixedToDOM)()
      }, {
        name: 'mediaSingle',
        node: (0, _mediaSingle.mediaSingleSpecWithFixedToDOM)(mediaSingleOption)
      }, {
        name: 'media',
        node: (0, _media.mediaSpecWithFixedToDOM)()
      }, {
        name: 'mediaInline',
        node: (0, _mediaInline.mediaInlineSpecWithFixedToDOM)()
      }].filter(function (node) {
        if (node.name === 'mediaGroup') {
          return allowMediaGroup;
        }
        if (node.name === 'mediaSingle') {
          return allowMediaSingle;
        }
        if (node.name === 'mediaInline') {
          return allowMediaInline;
        }
        return true;
      });
    },
    pmPlugins: function pmPlugins() {
      var pmPlugins = [{
        name: 'media',
        plugin: function plugin(_ref6) {
          var schema = _ref6.schema,
            dispatch = _ref6.dispatch,
            getIntl = _ref6.getIntl,
            eventDispatcher = _ref6.eventDispatcher,
            providerFactory = _ref6.providerFactory,
            errorReporter = _ref6.errorReporter,
            portalProviderAPI = _ref6.portalProviderAPI,
            dispatchAnalyticsEvent = _ref6.dispatchAnalyticsEvent,
            nodeViewPortalProviderAPI = _ref6.nodeViewPortalProviderAPI;
          return (0, _main.createPlugin)(schema, {
            providerFactory: providerFactory,
            nodeViews: {
              mediaGroup: (0, _lazyMediaGroup.lazyMediaGroupView)(portalProviderAPI, eventDispatcher, providerFactory, options, api),
              mediaSingle: (0, _lazyMediaSingle.lazyMediaSingleView)(portalProviderAPI, eventDispatcher, providerFactory, api, dispatchAnalyticsEvent, options),
              media: (0, _lazyMedia.lazyMediaView)(portalProviderAPI, eventDispatcher, providerFactory, options, api),
              mediaInline: (0, _lazyMediaInline.lazyMediaInlineView)(portalProviderAPI, eventDispatcher, providerFactory, api)
            },
            errorReporter: errorReporter,
            uploadErrorHandler: options && options.uploadErrorHandler,
            waitForMediaUpload: options && options.waitForMediaUpload,
            customDropzoneContainer: options && options.customDropzoneContainer,
            customMediaPicker: options && options.customMediaPicker,
            allowResizing: !!(options && options.allowResizing)
          }, getIntl, api, nodeViewPortalProviderAPI, dispatch, options);
        }
      }, {
        name: 'mediaKeymap',
        plugin: function plugin(_ref7) {
          var _api$analytics2, _api$selection;
          var getIntl = _ref7.getIntl;
          return (0, _keymap2.default)(options, api === null || api === void 0 || (_api$analytics2 = api.analytics) === null || _api$analytics2 === void 0 ? void 0 : _api$analytics2.actions, api === null || api === void 0 || (_api$selection = api.selection) === null || _api$selection === void 0 ? void 0 : _api$selection.actions, api === null || api === void 0 ? void 0 : api.width, getIntl);
        }
      }];
      if (options && options.allowMediaSingle) {
        pmPlugins.push({
          name: 'mediaSingleKeymap',
          plugin: function plugin(_ref8) {
            var schema = _ref8.schema;
            return (0, _keymapMedia.default)(schema);
          }
        });
      }
      if (options && options.allowAltTextOnImages) {
        pmPlugins.push({
          name: 'mediaAltText',
          plugin: _altText.createPlugin
        });
        pmPlugins.push({
          name: 'mediaAltTextKeymap',
          plugin: function plugin(_ref9) {
            var _api$analytics3;
            var schema = _ref9.schema;
            return (0, _keymap.default)(schema, api === null || api === void 0 || (_api$analytics3 = api.analytics) === null || _api$analytics3 === void 0 ? void 0 : _api$analytics3.actions);
          }
        });
      }
      if (options && options.allowLinking) {
        pmPlugins.push({
          name: 'mediaLinking',
          plugin: function plugin(_ref10) {
            var dispatch = _ref10.dispatch;
            return (0, _linking.default)(dispatch);
          }
        });
        pmPlugins.push({
          name: 'mediaLinkingKeymap',
          plugin: function plugin(_ref11) {
            var schema = _ref11.schema;
            return (0, _keymap3.default)(schema);
          }
        });
      }
      if (options && options.allowAdvancedToolBarOptions && options.allowResizing && (0, _experiments.editorExperiment)('platform_editor_controls', 'variant1') && options.allowPixelResizing) {
        pmPlugins.push({
          name: 'mediaPixelResizing',
          plugin: _pixelResizing.createPlugin
        });
      }
      pmPlugins.push({
        name: 'mediaSelectionHandler',
        plugin: function plugin() {
          var mediaSelectionHandlerPlugin = new _safePlugin.SafePlugin({
            key: new _state.PluginKey('mediaSelectionHandlerPlugin'),
            props: {
              handleScrollToSelection: function handleScrollToSelection(view) {
                var selection = view.state.selection;
                if (!(selection instanceof _state.NodeSelection) || selection.node.type.name !== 'media') {
                  return false;
                }
                var _view$domAtPos = view.domAtPos(selection.from),
                  node = _view$domAtPos.node,
                  offset = _view$domAtPos.offset;
                if (
                // Is the media element mounted already?
                offset === node.childNodes.length) {
                  // Media is not ready, so stop the scroll request
                  return true;
                }

                // Media is ready, keep the scrolling request
                return false;
              }
            }
          });
          return mediaSelectionHandlerPlugin;
        }
      });
      return pmPlugins;
    },
    contentComponent: function contentComponent(_ref12) {
      var editorView = _ref12.editorView,
        appearance = _ref12.appearance;
      return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(MediaViewerFunctionalComponent, {
        api: api,
        editorView: editorView
      }), /*#__PURE__*/_react.default.createElement(MediaPickerFunctionalComponent, {
        editorDomElement: editorView.dom,
        appearance: appearance,
        api: api
      }));
    },
    secondaryToolbarComponent: function secondaryToolbarComponent(_ref13) {
      var editorView = _ref13.editorView,
        eventDispatcher = _ref13.eventDispatcher,
        disabled = _ref13.disabled;
      return /*#__PURE__*/_react.default.createElement(_ToolbarMedia.default, {
        isDisabled: disabled,
        isReducedSpacing: true,
        api: api
      });
    },
    pluginsOptions: {
      quickInsert: function quickInsert(_ref14) {
        var formatMessage = _ref14.formatMessage;
        return (0, _platformFeatureFlags.fg)('platform_editor_add_media_from_url_rollout') ? [] : [{
          id: 'media',
          title: formatMessage(_messages.toolbarInsertBlockMessages.mediaFiles),
          description: formatMessage(_messages.toolbarInsertBlockMessages.mediaFilesDescription),
          priority: 400,
          keywords: ['attachment', 'gif', 'media', 'picture', 'image', 'video', 'file'],
          icon: function icon() {
            return /*#__PURE__*/_react.default.createElement(_quickInsert.IconImages, null);
          },
          isDisabledOffline: true,
          action: function action(insert, state) {
            var _api$analytics4;
            var pluginState = _pluginKey.stateKey.getState(state);
            pluginState === null || pluginState === void 0 || pluginState.showMediaPicker();
            var tr = insert('');
            api === null || api === void 0 || (_api$analytics4 = api.analytics) === null || _api$analytics4 === void 0 || _api$analytics4.actions.attachAnalyticsEvent({
              action: _analytics.ACTION.OPENED,
              actionSubject: _analytics.ACTION_SUBJECT.PICKER,
              actionSubjectId: _analytics.ACTION_SUBJECT_ID.PICKER_CLOUD,
              attributes: {
                inputMethod: _analytics.INPUT_METHOD.QUICK_INSERT
              },
              eventType: _analytics.EVENT_TYPE.UI
            })(tr);
            return tr;
          }
        }];
      },
      floatingToolbar: function floatingToolbar(state, intl, providerFactory) {
        var _api$editorViewMode;
        return (0, _toolbar.floatingToolbar)(state, intl, {
          providerFactory: providerFactory,
          allowMediaInline: options && (0, _mediaCommon.getMediaFeatureFlag)('mediaInline', options.featureFlags),
          allowResizing: options && options.allowResizing,
          allowResizingInTables: options && options.allowResizingInTables,
          allowCommentsOnMedia: options && options.allowCommentsOnMedia,
          allowLinking: options && options.allowLinking,
          allowAdvancedToolBarOptions: options && options.allowAdvancedToolBarOptions,
          allowAltTextOnImages: options && options.allowAltTextOnImages,
          allowImagePreview: options && options.allowImagePreview,
          altTextValidator: options && options.altTextValidator,
          fullWidthEnabled: options && options.fullWidthEnabled,
          allowMediaInlineImages: options && options.allowMediaInlineImages,
          isViewOnly: (api === null || api === void 0 || (_api$editorViewMode = api.editorViewMode) === null || _api$editorViewMode === void 0 || (_api$editorViewMode = _api$editorViewMode.sharedState.currentState()) === null || _api$editorViewMode === void 0 ? void 0 : _api$editorViewMode.mode) === 'view',
          allowPixelResizing: options === null || options === void 0 ? void 0 : options.allowPixelResizing
        }, api);
      }
    }
  };
};
