// @atlaskit/mention/dist/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AbstractMentionResource", {
  enumerable: true,
  get: function get() {
    return _MentionResource.AbstractMentionResource;
  }
});
Object.defineProperty(exports, "AbstractPresenceResource", {
  enumerable: true,
  get: function get() {
    return _PresenceResource.AbstractPresenceResource;
  }
});
Object.defineProperty(exports, "Actions", {
  enumerable: true,
  get: function get() {
    return _types.Actions;
  }
});
Object.defineProperty(exports, "ComponentNames", {
  enumerable: true,
  get: function get() {
    return _types.ComponentNames;
  }
});
Object.defineProperty(exports, "ContextMentionResource", {
  enumerable: true,
  get: function get() {
    return _ContextMentionResource.default;
  }
});
Object.defineProperty(exports, "DefaultMentionNameResolver", {
  enumerable: true,
  get: function get() {
    return _MentionNameResolver.DefaultMentionNameResolver;
  }
});
Object.defineProperty(exports, "ELEMENTS_CHANNEL", {
  enumerable: true,
  get: function get() {
    return _constants.ELEMENTS_CHANNEL;
  }
});
Object.defineProperty(exports, "Mention", {
  enumerable: true,
  get: function get() {
    return _Mention.default;
  }
});
Object.defineProperty(exports, "MentionItem", {
  enumerable: true,
  get: function get() {
    return _MentionItem.default;
  }
});
Object.defineProperty(exports, "MentionList", {
  enumerable: true,
  get: function get() {
    return _MentionList.default;
  }
});
Object.defineProperty(exports, "MentionNameStatus", {
  enumerable: true,
  get: function get() {
    return _types.MentionNameStatus;
  }
});
Object.defineProperty(exports, "MentionPicker", {
  enumerable: true,
  get: function get() {
    return _MentionPicker.MentionPickerWithAnalytics;
  }
});
Object.defineProperty(exports, "MentionResource", {
  enumerable: true,
  get: function get() {
    return _MentionResource.default;
  }
});
Object.defineProperty(exports, "PresenceResource", {
  enumerable: true,
  get: function get() {
    return _PresenceResource.default;
  }
});
Object.defineProperty(exports, "ResourcedMention", {
  enumerable: true,
  get: function get() {
    return _ResourcedMention.default;
  }
});
Object.defineProperty(exports, "ResourcedMentionList", {
  enumerable: true,
  get: function get() {
    return _ResourcedMentionList.default;
  }
});
Object.defineProperty(exports, "SliNames", {
  enumerable: true,
  get: function get() {
    return _types.SliNames;
  }
});
Object.defineProperty(exports, "TeamMentionResource", {
  enumerable: true,
  get: function get() {
    return _TeamMentionResource.default;
  }
});
Object.defineProperty(exports, "UserAccessLevel", {
  enumerable: true,
  get: function get() {
    return _types.UserAccessLevel;
  }
});
Object.defineProperty(exports, "UserType", {
  enumerable: true,
  get: function get() {
    return _types.UserType;
  }
});
exports.default = void 0;
Object.defineProperty(exports, "isResolvingMentionProvider", {
  enumerable: true,
  get: function get() {
    return _MentionResource.isResolvingMentionProvider;
  }
});
Object.defineProperty(exports, "isSpecialMention", {
  enumerable: true,
  get: function get() {
    return _types.isSpecialMention;
  }
});
var _MentionResource = _interopRequireWildcard(require("./api/MentionResource"));
var _TeamMentionResource = _interopRequireDefault(require("./api/TeamMentionResource"));
var _PresenceResource = _interopRequireWildcard(require("./api/PresenceResource"));
var _MentionNameResolver = require("./api/MentionNameResolver");
var _MentionItem = _interopRequireDefault(require("./components/MentionItem"));
var _MentionList = _interopRequireDefault(require("./components/MentionList"));
var _ResourcedMentionList = _interopRequireDefault(require("./components/ResourcedMentionList"));
var _MentionPicker = require("./components/MentionPicker");
var _Mention = _interopRequireDefault(require("./components/Mention"));
var _ResourcedMention = _interopRequireDefault(require("./components/Mention/ResourcedMention"));
var _types = require("./types");
var _constants = require("./_constants");
var _ContextMentionResource = _interopRequireDefault(require("./api/ContextMentionResource"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var _default = exports.default = _MentionPicker.MentionPickerWithAnalytics;
