// @atlaskit/mention/dist/cjs/resource.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AbstractMentionResource", {
  enumerable: true,
  get: function get() {
    return _MentionResource.AbstractMentionResource;
  }
});
Object.defineProperty(exports, "AbstractPresenceResource", {
  enumerable: true,
  get: function get() {
    return _PresenceResource.AbstractPresenceResource;
  }
});
Object.defineProperty(exports, "ContextMentionResource", {
  enumerable: true,
  get: function get() {
    return _ContextMentionResource.default;
  }
});
Object.defineProperty(exports, "DefaultMentionNameResolver", {
  enumerable: true,
  get: function get() {
    return _MentionNameResolver.DefaultMentionNameResolver;
  }
});
Object.defineProperty(exports, "ELEMENTS_CHANNEL", {
  enumerable: true,
  get: function get() {
    return _constants.ELEMENTS_CHANNEL;
  }
});
Object.defineProperty(exports, "MentionNameStatus", {
  enumerable: true,
  get: function get() {
    return _types.MentionNameStatus;
  }
});
Object.defineProperty(exports, "MentionResource", {
  enumerable: true,
  get: function get() {
    return _MentionResource.default;
  }
});
Object.defineProperty(exports, "PresenceResource", {
  enumerable: true,
  get: function get() {
    return _PresenceResource.default;
  }
});
Object.defineProperty(exports, "SLI_EVENT_TYPE", {
  enumerable: true,
  get: function get() {
    return _analytics.SLI_EVENT_TYPE;
  }
});
Object.defineProperty(exports, "SMART_EVENT_TYPE", {
  enumerable: true,
  get: function get() {
    return _analytics.SMART_EVENT_TYPE;
  }
});
Object.defineProperty(exports, "buildSliPayload", {
  enumerable: true,
  get: function get() {
    return _analytics.buildSliPayload;
  }
});
Object.defineProperty(exports, "isResolvingMentionProvider", {
  enumerable: true,
  get: function get() {
    return _MentionResource.isResolvingMentionProvider;
  }
});
Object.defineProperty(exports, "isSpecialMention", {
  enumerable: true,
  get: function get() {
    return _types.isSpecialMention;
  }
});
var _MentionResource = _interopRequireWildcard(require("./api/MentionResource"));
var _PresenceResource = _interopRequireWildcard(require("./api/PresenceResource"));
var _MentionNameResolver = require("./api/MentionNameResolver");
var _types = require("./types");
var _constants = require("./_constants");
var _ContextMentionResource = _interopRequireDefault(require("./api/ContextMentionResource"));
var _analytics = require("./util/analytics");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
