// @atlaskit/editor-core/dist/esm/plugins/table/pm-plugins/table-resizing/plugin.js
import { Plugin, PluginKey } from 'prosemirror-state';
import classnames from 'classnames';
import { getResizeCellPos } from './utils';
import { TableCssClassName as ClassName, } from '../../types';
import { handleMouseDown } from './event-handlers';
import { pluginFactory } from '../../../../utils/plugin-state-factory';
import reducer from './reducer';
export var pluginKey = new PluginKey('tableFlexiColumnResizing');
var _a = pluginFactory(pluginKey, reducer), createPluginState = _a.createPluginState, createCommand = _a.createCommand, getPluginState = _a.getPluginState;
export function createPlugin(dispatch, _a) {
    var _b = _a.lastColumnResizable, lastColumnResizable = _b === void 0 ? true : _b, _c = _a.dynamicTextSizing, dynamicTextSizing = _c === void 0 ? false : _c;
    return new Plugin({
        key: pluginKey,
        state: createPluginState(dispatch, {
            lastColumnResizable: lastColumnResizable,
            dynamicTextSizing: dynamicTextSizing,
            resizeHandlePos: null,
            dragging: null,
            lastClick: null,
        }),
        props: {
            attributes: function (state) {
                var _a;
                var pluginState = getPluginState(state);
                return {
                    class: classnames(ClassName.RESIZING_PLUGIN, (_a = {},
                        _a[ClassName.RESIZE_CURSOR] = pluginState.resizeHandlePos !== null,
                        _a[ClassName.IS_RESIZING] = !!pluginState.dragging,
                        _a)),
                };
            },
            handleDOMEvents: {
                mousedown: function (view, event) {
                    var state = view.state;
                    var resizeHandlePos = 
                    // we're setting `resizeHandlePos` via command in integration tests
                    getPluginState(state).resizeHandlePos ||
                        getResizeCellPos(view, event, lastColumnResizable);
                    var dragging = getPluginState(state).dragging;
                    if (resizeHandlePos !== null && !dragging) {
                        return handleMouseDown(view, event, resizeHandlePos, dynamicTextSizing);
                    }
                    return false;
                },
            },
        },
    });
}
export { createCommand, getPluginState };
