// @atlaskit/media-ui/dist/cjs/customMediaPlayer/index-emotion.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CustomMediaPlayerBase = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _videoPlayVidPlay = _interopRequireDefault(require("@atlaskit/icon/core/migration/video-play--vid-play"));
var _videoPauseVidPause = _interopRequireDefault(require("@atlaskit/icon/core/migration/video-pause--vid-pause"));
var _fullscreenEnterVidFullScreenOn = _interopRequireDefault(require("@atlaskit/icon/core/migration/fullscreen-enter--vid-full-screen-on"));
var _shrinkDiagonalVidFullScreenOff = _interopRequireDefault(require("@atlaskit/icon/core/migration/shrink-diagonal--vid-full-screen-off"));
var _volumeHighHipchatOutgoingSound = _interopRequireDefault(require("@atlaskit/icon/core/migration/volume-high--hipchat-outgoing-sound"));
var _vidHdCircle = _interopRequireDefault(require("@atlaskit/icon/glyph/vid-hd-circle"));
var _download = _interopRequireDefault(require("@atlaskit/icon/core/migration/download"));
var _primitives = require("@atlaskit/primitives");
var _MediaButton = _interopRequireDefault(require("../MediaButton"));
var _spinner = _interopRequireDefault(require("@atlaskit/spinner"));
var _widthDetector = require("@atlaskit/width-detector");
var _reactVideoRenderer = _interopRequireDefault(require("react-video-renderer"));
var _colors = require("@atlaskit/theme/colors");
var _timeRange = require("./timeRange");
var _volumeRange = _interopRequireDefault(require("./volumeRange"));
var _styled = require("./styled");
var _styledEmotion = require("./styled-emotion");
var _analytics = require("./analytics");
var _formatDuration = require("../formatDuration");
var _shortcut = require("../shortcut");
var _fullscreen = require("./fullscreen");
var _messages = require("../messages");
var _simultaneousPlayManager = _interopRequireDefault(require("./simultaneousPlayManager"));
var _timeSaver = require("./timeSaver");
var _playbackSpeedControls = _interopRequireDefault(require("./playbackSpeedControls"));
var _playPauseBlanket = require("./playPauseBlanket");
var _tooltip = _interopRequireDefault(require("@atlaskit/tooltip"));
var _legacyCustomIcons = require("@atlaskit/legacy-custom-icons");
var _getControlsWrapperClassName = require("./getControlsWrapperClassName");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _videoSkipForwardTen = _interopRequireDefault(require("@atlaskit/icon/core/video-skip-forward-ten"));
var _videoSkipBackwardTen = _interopRequireDefault(require("@atlaskit/icon/core/video-skip-backward-ten"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); } /* eslint-disable @atlaskit/design-system/ensure-design-token-usage */ // Keep media player components used in media-viewer to use static colors from the new color palette to
// support the hybrid theming in media viewer https://product-fabric.atlassian.net/browse/DSP-6067
var MEDIUM_VIDEO_MAX_WIDTH = 400;
var SMALL_VIDEO_MAX_WIDTH = 160;
var MINIMUM_DURATION_BEFORE_SAVING_TIME = 60;
var VIEWED_TRACKING_SECS = 2;

/* Styles */

var timebarWrapperStyles = (0, _primitives.xcss)({
  position: 'absolute',
  width: '100%',
  bottom: 'space.100'
});
var customVideoWrapperStyles = (0, _primitives.xcss)({
  width: '100%',
  height: '100%',
  userSelect: 'none'
});
var videoWrapperStyles = (0, _primitives.xcss)({
  width: '100%',
  height: '100%'
});
var timeWrapperStyles = (0, _primitives.xcss)({
  marginTop: 'space.0',
  marginInline: 'space.250',
  marginBottom: 'space.500'
});
var spinnerWrapperStyles = (0, _primitives.xcss)({
  position: 'absolute',
  top: 'space.0',
  left: 'space.0',
  width: '100%',
  height: '100%'
});
var CustomMediaPlayerBase = exports.CustomMediaPlayerBase = /*#__PURE__*/function (_Component) {
  function CustomMediaPlayerBase() {
    var _this;
    (0, _classCallCheck2.default)(this, CustomMediaPlayerBase);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, CustomMediaPlayerBase, [].concat(args));
    (0, _defineProperty2.default)(_this, "videoWrapperRef", /*#__PURE__*/_react.default.createRef());
    (0, _defineProperty2.default)(_this, "videoState", {
      isLoading: true,
      buffered: 0,
      currentTime: 0,
      volume: 1,
      status: 'paused',
      duration: 0,
      isMuted: false
    });
    (0, _defineProperty2.default)(_this, "wasPlayedOnce", false);
    (0, _defineProperty2.default)(_this, "lastCurrentTime", 0);
    (0, _defineProperty2.default)(_this, "timeSaver", new _timeSaver.TimeSaver(_this.props.lastWatchTimeConfig));
    (0, _defineProperty2.default)(_this, "state", {
      isFullScreenEnabled: false,
      playerSize: 'large',
      playbackSpeed: 1
    });
    (0, _defineProperty2.default)(_this, "fireFirstPlayedTrackEvent", function () {
      var _this$props = _this.props,
        type = _this$props.type,
        fileId = _this$props.fileId,
        isHDActive = _this$props.isHDActive,
        isHDAvailable = _this$props.isHDAvailable,
        isAutoPlay = _this$props.isAutoPlay,
        createAnalyticsEvent = _this$props.createAnalyticsEvent;
      var _this$state = _this.state,
        isFullScreenEnabled = _this$state.isFullScreenEnabled,
        playerSize = _this$state.playerSize,
        playbackSpeed = _this$state.playbackSpeed;
      (0, _analytics.fireAnalyticsEvent)((0, _analytics.createFirstPlayedTrackEvent)(type, {
        isAutoPlay: isAutoPlay,
        isHDAvailable: isHDAvailable,
        isHDActive: isHDActive,
        isFullScreenEnabled: isFullScreenEnabled,
        playerSize: playerSize,
        playbackSpeed: playbackSpeed
      }, fileId), createAnalyticsEvent);
    });
    (0, _defineProperty2.default)(_this, "onFullScreenChange", function (e) {
      if (e.target !== _this.videoWrapperRef.current) {
        return;
      }
      var currentFullScreenMode = _this.state.isFullScreenEnabled;
      var isFullScreenEnabled = !!(0, _fullscreen.getFullscreenElement)();
      if (currentFullScreenMode !== isFullScreenEnabled) {
        var _this$props$onFullscr, _this$props2;
        (_this$props$onFullscr = (_this$props2 = _this.props).onFullscreenChange) === null || _this$props$onFullscr === void 0 || _this$props$onFullscr.call(_this$props2, isFullScreenEnabled);
        _this.setState({
          isFullScreenEnabled: isFullScreenEnabled
        });
      }
    });
    (0, _defineProperty2.default)(_this, "onTimeChanged", function () {
      _this.createAndFireUIEvent('timeRangeNavigate', 'time');
    });
    (0, _defineProperty2.default)(_this, "onVolumeChanged", function () {
      _this.createAndFireUIEvent('volumeRangeNavigate', 'volume');
    });
    (0, _defineProperty2.default)(_this, "onCurrentTimeChange", function (currentTime, duration) {
      if (duration - currentTime > MINIMUM_DURATION_BEFORE_SAVING_TIME) {
        _this.timeSaver.defaultTime = currentTime;
      } else {
        _this.timeSaver.defaultTime = 0;
      }
    });
    (0, _defineProperty2.default)(_this, "renderCurrentTime", function (_ref) {
      var currentTime = _ref.currentTime,
        duration = _ref.duration;
      return /*#__PURE__*/_react.default.createElement(_styled.CurrentTime, {
        draggable: false
      }, (0, _formatDuration.formatDuration)(currentTime), " / ", (0, _formatDuration.formatDuration)(duration));
    });
    (0, _defineProperty2.default)(_this, "renderHDButton", function () {
      var _this$props3 = _this.props,
        type = _this$props3.type,
        isHDAvailable = _this$props3.isHDAvailable,
        isHDActive = _this$props3.isHDActive,
        onHDToggleClick = _this$props3.onHDToggleClick;
      if (type === 'audio' || !isHDAvailable) {
        return;
      }
      var primaryColor = isHDActive ? '#579DFF' : '#c7d1db';
      var secondaryColor = isHDActive ? _colors.N0 : _colors.DN60;
      return /*#__PURE__*/_react.default.createElement(_MediaButton.default, {
        testId: "custom-media-player-hd-button",
        onClick: !!onHDToggleClick ? _this.getMediaButtonClickHandler(onHDToggleClick, 'HDButton') : undefined,
        iconBefore:
        /*#__PURE__*/
        //TODO: https://product-fabric.atlassian.net/browse/DSP-20900
        // eslint-disable-next-line @atlaskit/design-system/no-legacy-icons
        _react.default.createElement(_vidHdCircle.default, {
          primaryColor: primaryColor,
          secondaryColor: secondaryColor,
          label: "hd"
        })
      });
    });
    (0, _defineProperty2.default)(_this, "onPlaybackSpeedChange", function (playbackSpeed) {
      if (!_this.actions) {
        return;
      }
      _this.actions.setPlaybackSpeed(playbackSpeed);
      _this.setState({
        playbackSpeed: playbackSpeed
      });
      _this.createAndFireUIEvent('playbackSpeedChange');
    });
    (0, _defineProperty2.default)(_this, "renderSpeedControls", function () {
      var playbackSpeed = _this.state.playbackSpeed;
      var originalDimensions = _this.props.originalDimensions;
      return /*#__PURE__*/_react.default.createElement(_playbackSpeedControls.default, {
        originalDimensions: originalDimensions,
        playbackSpeed: playbackSpeed,
        onPlaybackSpeedChange: _this.onPlaybackSpeedChange,
        onClick: function onClick() {
          return _this.createAndFireUIEvent('mediaButtonClick', 'playbackSpeedButton');
        }
      });
    });
    (0, _defineProperty2.default)(_this, "renderVolume", function (videoState, actions, showSlider) {
      return /*#__PURE__*/_react.default.createElement(_styled.VolumeWrapper, {
        showSlider: showSlider
      }, /*#__PURE__*/_react.default.createElement(_styled.VolumeToggleWrapper, {
        isMuted: videoState.isMuted
      }, /*#__PURE__*/_react.default.createElement(_styled.MutedIndicator, {
        isMuted: videoState.isMuted
      }), /*#__PURE__*/_react.default.createElement(_MediaButton.default, {
        testId: "custom-media-player-volume-toggle-button",
        onClick: _this.getMediaButtonClickHandler(actions.toggleMute, 'muteButton'),
        iconBefore: /*#__PURE__*/_react.default.createElement(_volumeHighHipchatOutgoingSound.default, {
          color: "currentColor",
          label: _this.props.intl.formatMessage(_messages.messages.volumeMuteButtonAria)
        }),
        "aria-pressed": videoState.isMuted
      })), showSlider && /*#__PURE__*/_react.default.createElement(_styled.VolumeTimeRangeWrapper, null, /*#__PURE__*/_react.default.createElement(_volumeRange.default, {
        onChange: actions.setVolume,
        currentVolume: videoState.volume,
        isAlwaysActive: true,
        onChanged: _this.onVolumeChanged,
        ariaLabel: _this.props.intl.formatMessage(_messages.messages.volumeLevelControlAria)
      })));
    });
    (0, _defineProperty2.default)(_this, "toggleFullscreen", function () {
      return _this.videoWrapperRef.current && (0, _fullscreen.toggleFullscreen)(_this.videoWrapperRef.current);
    });
    (0, _defineProperty2.default)(_this, "onFullScreenButtonClick", function () {
      _this.toggleFullscreen();
      _this.createAndFireUIEvent('mediaButtonClick', 'fullScreenButton');
    });
    (0, _defineProperty2.default)(_this, "onResize", function (width) {
      if (width > MEDIUM_VIDEO_MAX_WIDTH) {
        _this.setState({
          playerSize: 'large'
        });
      } else if (width > SMALL_VIDEO_MAX_WIDTH) {
        _this.setState({
          playerSize: 'medium'
        });
      } else {
        _this.setState({
          playerSize: 'small'
        });
      }
    });
    (0, _defineProperty2.default)(_this, "renderFullScreenButton", function () {
      var _this$props4 = _this.props,
        formatMessage = _this$props4.intl.formatMessage,
        type = _this$props4.type;
      if (type === 'audio') {
        return;
      }
      var isFullScreenEnabled = _this.state.isFullScreenEnabled;
      var icon = isFullScreenEnabled ? /*#__PURE__*/_react.default.createElement(_shrinkDiagonalVidFullScreenOff.default, {
        color: "currentColor",
        label: formatMessage(_messages.messages.disable_fullscreen)
      }) : /*#__PURE__*/_react.default.createElement(_fullscreenEnterVidFullScreenOn.default, {
        color: "currentColor",
        label: formatMessage(_messages.messages.enable_fullscreen)
      });
      return /*#__PURE__*/_react.default.createElement(_MediaButton.default, {
        testId: "custom-media-player-fullscreen-button",
        onClick: _this.onFullScreenButtonClick,
        iconBefore: icon
      });
    });
    (0, _defineProperty2.default)(_this, "renderDownloadButton", function () {
      var onDownloadClick = _this.props.onDownloadClick;
      if (!onDownloadClick) {
        return;
      }
      return /*#__PURE__*/_react.default.createElement(_MediaButton.default, {
        testId: "custom-media-player-download-button",
        onClick: _this.getMediaButtonClickHandler(onDownloadClick, 'downloadButton'),
        iconBefore: /*#__PURE__*/_react.default.createElement(_download.default, {
          color: "currentColor",
          label: "download"
        })
      });
    });
    (0, _defineProperty2.default)(_this, "renderShortcuts", function (_ref2) {
      var togglePlayPauseAction = _ref2.togglePlayPauseAction,
        toggleMute = _ref2.toggleMute,
        skipBackward = _ref2.skipBackward,
        skipForward = _ref2.skipForward;
      var isShortcutEnabled = _this.props.isShortcutEnabled;
      var isFullScreenEnabled = _this.state.isFullScreenEnabled;
      var shortcuts = (isShortcutEnabled || isFullScreenEnabled) && [/*#__PURE__*/_react.default.createElement(_shortcut.Shortcut, {
        key: "space-shortcut",
        code: _shortcut.keyCodes.space,
        handler: _this.getKeyboardShortcutHandler(togglePlayPauseAction, 'space')
      }), /*#__PURE__*/_react.default.createElement(_shortcut.Shortcut, {
        key: "m-shortcut",
        code: _shortcut.keyCodes.m,
        handler: _this.getKeyboardShortcutHandler(toggleMute, 'mute')
      })];
      if (shortcuts && isFullScreenEnabled) {
        // Fullscreen shortcuts only. We don't want to override left/right keys in media-viewer settings
        shortcuts.push( /*#__PURE__*/_react.default.createElement(_shortcut.Shortcut, {
          key: "skip-backward-shortcut",
          code: _shortcut.keyCodes.leftArrow,
          handler: _this.getKeyboardShortcutHandler(skipBackward, 'leftArrow')
        }));
        shortcuts.push( /*#__PURE__*/_react.default.createElement(_shortcut.Shortcut, {
          key: "skip-forward-shortcut",
          code: _shortcut.keyCodes.rightArrow,
          handler: _this.getKeyboardShortcutHandler(skipForward, 'rightArrow')
        }));
      }
      return shortcuts;
    });
    (0, _defineProperty2.default)(_this, "renderPlayPauseButton", function (isPlaying) {
      var formatMessage = _this.props.intl.formatMessage;
      var toggleButtonIcon = isPlaying ? /*#__PURE__*/_react.default.createElement(_videoPauseVidPause.default, {
        spacing: "spacious",
        color: "currentColor",
        label: formatMessage(_messages.messages.pause)
      }) : /*#__PURE__*/_react.default.createElement(_videoPlayVidPlay.default, {
        spacing: "spacious",
        color: "currentColor",
        label: formatMessage(_messages.messages.play)
      });
      return /*#__PURE__*/_react.default.createElement(_tooltip.default, {
        content: formatMessage(isPlaying ? _messages.messages.pause : _messages.messages.play),
        position: "top"
      }, /*#__PURE__*/_react.default.createElement(_MediaButton.default, {
        testId: "custom-media-player-play-toggle-button",
        "data-test-is-playing": isPlaying,
        iconBefore: toggleButtonIcon,
        onClick: isPlaying ? _this.pausePlayByButtonClick : _this.startPlayByButtonClick
      }));
    });
    (0, _defineProperty2.default)(_this, "renderSkipBackwardButton", function (skipBackward) {
      var formatMessage = _this.props.intl.formatMessage;
      return /*#__PURE__*/_react.default.createElement(_tooltip.default, {
        content: formatMessage(_messages.messages.skipBackward),
        position: "top"
      }, /*#__PURE__*/_react.default.createElement(_MediaButton.default, {
        testId: "custom-media-player-skip-backward-button",
        iconBefore: /*#__PURE__*/_react.default.createElement(_videoSkipBackwardTen.default, {
          spacing: "spacious",
          LEGACY_fallbackIcon: _legacyCustomIcons.SkipTenBackwardIcon,
          label: formatMessage(_messages.messages.skipBackward)
        }),
        onClick: _this.getMediaButtonClickHandler(skipBackward, 'skipBackwardButton')
      }));
    });
    (0, _defineProperty2.default)(_this, "renderSkipForwardButton", function (skipForward) {
      var formatMessage = _this.props.intl.formatMessage;
      return /*#__PURE__*/_react.default.createElement(_tooltip.default, {
        content: formatMessage(_messages.messages.skipForward),
        position: "top"
      }, /*#__PURE__*/_react.default.createElement(_MediaButton.default, {
        testId: "custom-media-player-skip-forward-button",
        iconBefore: /*#__PURE__*/_react.default.createElement(_videoSkipForwardTen.default, {
          spacing: "spacious",
          LEGACY_fallbackIcon: _legacyCustomIcons.SkipTenForwardIcon,
          label: formatMessage(_messages.messages.skipForward)
        }),
        onClick: _this.getMediaButtonClickHandler(skipForward, 'skipForwardButton')
      }));
    });
    (0, _defineProperty2.default)(_this, "renderSpinner", function () {
      return /*#__PURE__*/_react.default.createElement(_primitives.Flex, {
        direction: "column",
        alignItems: "center",
        justifyContent: "center",
        xcss: spinnerWrapperStyles
      }, /*#__PURE__*/_react.default.createElement(_spinner.default, {
        appearance: "invert",
        size: "large"
      }));
    });
    (0, _defineProperty2.default)(_this, "pause", function () {
      if (_this.actions) {
        _this.actions.pause();
      }
    });
    (0, _defineProperty2.default)(_this, "play", function () {
      var onFirstPlay = _this.props.onFirstPlay;
      if (_this.actions) {
        _this.actions.play();
      }
      _simultaneousPlayManager.default.pauseOthers(_this);
      if (!_this.wasPlayedOnce && onFirstPlay) {
        _this.fireFirstPlayedTrackEvent();
        _this.wasPlayedOnce = true;
        onFirstPlay();
      }
    });
    (0, _defineProperty2.default)(_this, "getMediaButtonClickHandler", function (action, buttonType) {
      return function () {
        action();
        _this.createAndFireUIEvent('mediaButtonClick', buttonType);
      };
    });
    (0, _defineProperty2.default)(_this, "getKeyboardShortcutHandler", function (action, shortcutType) {
      return function () {
        var showControls = _this.props.showControls;
        action();
        if (showControls) {
          showControls();
        }
        _this.createAndFireUIEvent('shortcutPress', shortcutType);
      };
    });
    (0, _defineProperty2.default)(_this, "onViewed", function (videoState) {
      var _this$props5 = _this.props,
        createAnalyticsEvent = _this$props5.createAnalyticsEvent,
        fileId = _this$props5.fileId,
        isAutoPlay = _this$props5.isAutoPlay,
        isHDAvailable = _this$props5.isHDAvailable,
        isHDActive = _this$props5.isHDActive,
        type = _this$props5.type;
      var _this$state2 = _this.state,
        isFullScreenEnabled = _this$state2.isFullScreenEnabled,
        playerSize = _this$state2.playerSize,
        playbackSpeed = _this$state2.playbackSpeed;
      var status = videoState.status,
        currentTime = videoState.currentTime;
      if (status === 'playing' && (currentTime < _this.lastCurrentTime || currentTime >= _this.lastCurrentTime + VIEWED_TRACKING_SECS)) {
        (0, _analytics.fireAnalyticsEvent)((0, _analytics.createPlayedTrackEvent)(type, _objectSpread(_objectSpread({}, videoState), {}, {
          isAutoPlay: isAutoPlay,
          isHDAvailable: isHDAvailable,
          isHDActive: isHDActive,
          isFullScreenEnabled: isFullScreenEnabled,
          playerSize: playerSize,
          playbackSpeed: playbackSpeed
        }), fileId), createAnalyticsEvent);
        _this.lastCurrentTime = currentTime;
      }
    });
    (0, _defineProperty2.default)(_this, "resetPendingPlayPauseToggleTimer", function () {
      if (_this.clickToTogglePlayTimeoutId !== undefined) {
        clearTimeout(_this.clickToTogglePlayTimeoutId);
      }
    });
    (0, _defineProperty2.default)(_this, "doubleClickToFullscreen", function () {
      _this.resetPendingPlayPauseToggleTimer();
      _this.toggleFullscreen();
      // TODO Add an event similar to "playPauseBlanketClick" but for fullscreen trigger
    });
    (0, _defineProperty2.default)(_this, "togglePlayByBlanketClick", function (action) {
      _this.resetPendingPlayPauseToggleTimer();
      _this.clickToTogglePlayTimeoutId = setTimeout(function () {
        action();
        _this.createAndFireUIEvent('playPauseBlanketClick');
      }, 200);
    });
    (0, _defineProperty2.default)(_this, "startPlayByBlanketClick", function () {
      _this.togglePlayByBlanketClick(_this.play);
    });
    (0, _defineProperty2.default)(_this, "pausePlayByBlanketClick", function () {
      _this.togglePlayByBlanketClick(_this.pause);
    });
    (0, _defineProperty2.default)(_this, "startPlayByButtonClick", _this.getMediaButtonClickHandler(_this.play, 'playButton'));
    (0, _defineProperty2.default)(_this, "pausePlayByButtonClick", _this.getMediaButtonClickHandler(_this.pause, 'pauseButton'));
    return _this;
  }
  (0, _inherits2.default)(CustomMediaPlayerBase, _Component);
  return (0, _createClass2.default)(CustomMediaPlayerBase, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this$props6 = this.props,
        type = _this$props6.type,
        fileId = _this$props6.fileId,
        isAutoPlay = _this$props6.isAutoPlay,
        isHDAvailable = _this$props6.isHDAvailable,
        isHDActive = _this$props6.isHDActive,
        onFirstPlay = _this$props6.onFirstPlay,
        createAnalyticsEvent = _this$props6.createAnalyticsEvent;
      var _this$state3 = this.state,
        isFullScreenEnabled = _this$state3.isFullScreenEnabled,
        playerSize = _this$state3.playerSize,
        playbackSpeed = _this$state3.playbackSpeed;
      (0, _analytics.fireAnalyticsEvent)((0, _analytics.createCustomMediaPlayerScreenEvent)(type, {
        isAutoPlay: isAutoPlay,
        isHDAvailable: isHDAvailable,
        isHDActive: isHDActive,
        isFullScreenEnabled: isFullScreenEnabled,
        playerSize: playerSize,
        playbackSpeed: playbackSpeed
      }, fileId), createAnalyticsEvent);
      if (this.videoWrapperRef.current) {
        this.videoWrapperRef.current.addEventListener('fullscreenchange', this.onFullScreenChange);
      }
      _simultaneousPlayManager.default.subscribe(this);
      if (isAutoPlay) {
        _simultaneousPlayManager.default.pauseOthers(this);
        if (onFirstPlay) {
          this.fireFirstPlayedTrackEvent();
          this.wasPlayedOnce = true;
          onFirstPlay();
        }
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (this.videoWrapperRef.current) {
        this.videoWrapperRef.current.removeEventListener('fullscreenchange', this.onFullScreenChange);
      }
      if (this.state.isFullScreenEnabled) {
        var _this$props$onFullscr2, _this$props7;
        (_this$props$onFullscr2 = (_this$props7 = this.props).onFullscreenChange) === null || _this$props$onFullscr2 === void 0 || _this$props$onFullscr2.call(_this$props7, false);
      }
      _simultaneousPlayManager.default.unsubscribe(this);
    }
  }, {
    key: "setActions",
    value: function setActions(actions) {
      // Actions are being sent constantly while the video is playing,
      // though play and pause functions are always the same objects
      if (!this.actions) {
        this.actions = actions;
      }
    }
  }, {
    key: "createAndFireUIEvent",
    value: function createAndFireUIEvent(eventType, actionSubjectId) {
      var _this$props8 = this.props,
        type = _this$props8.type,
        fileId = _this$props8.fileId,
        isHDActive = _this$props8.isHDActive,
        isHDAvailable = _this$props8.isHDAvailable,
        isAutoPlay = _this$props8.isAutoPlay,
        createAnalyticsEvent = _this$props8.createAnalyticsEvent;
      var _this$state4 = this.state,
        isFullScreenEnabled = _this$state4.isFullScreenEnabled,
        playerSize = _this$state4.playerSize,
        playbackSpeed = _this$state4.playbackSpeed;
      var playbackState = _objectSpread(_objectSpread({}, this.videoState), {}, {
        isAutoPlay: isAutoPlay,
        isHDAvailable: isHDAvailable,
        isHDActive: isHDActive,
        isFullScreenEnabled: isFullScreenEnabled,
        playerSize: playerSize,
        playbackSpeed: playbackSpeed
      });
      var analyticsEvent;
      switch (eventType) {
        case 'mediaButtonClick':
          analyticsEvent = (0, _analytics.createMediaButtonClickedEvent)(type, playbackState, actionSubjectId, fileId);
          break;
        case 'shortcutPress':
          analyticsEvent = (0, _analytics.createMediaShortcutPressedEvent)(type, playbackState, actionSubjectId, fileId);
          break;
        case 'playPauseBlanketClick':
          analyticsEvent = (0, _analytics.createPlayPauseBlanketClickedEvent)(type, playbackState, fileId);
          break;
        case 'timeRangeNavigate':
        case 'volumeRangeNavigate':
          analyticsEvent = (0, _analytics.createTimeRangeNavigatedEvent)(type, playbackState, actionSubjectId, fileId);
          break;
        case 'playbackSpeedChange':
          analyticsEvent = (0, _analytics.createPlaybackSpeedChangedEvent)(type, playbackState, fileId);
          break;
        default:
          analyticsEvent = {
            eventType: 'ui',
            action: 'default',
            actionSubject: 'customMediaPlayer',
            attributes: {
              type: type
            }
          };
      }
      (0, _analytics.fireAnalyticsEvent)(analyticsEvent, createAnalyticsEvent);
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;
      var _this$props9 = this.props,
        type = _this$props9.type,
        src = _this$props9.src,
        isAutoPlay = _this$props9.isAutoPlay,
        onCanPlay = _this$props9.onCanPlay,
        onError = _this$props9.onError,
        poster = _this$props9.poster,
        videoControlsWrapperRef = _this$props9.videoControlsWrapperRef;
      return /*#__PURE__*/_react.default.createElement(_primitives.Box, {
        xcss: customVideoWrapperStyles,
        ref: this.videoWrapperRef,
        testId: "custom-media-player"
      }, /*#__PURE__*/_react.default.createElement(_reactVideoRenderer.default, {
        sourceType: type,
        src: src,
        autoPlay: isAutoPlay,
        onCanPlay: onCanPlay,
        defaultTime: this.timeSaver.defaultTime,
        onTimeChange: this.onCurrentTimeChange,
        onError: onError,
        poster: poster
      }, function (video, videoState, actions) {
        _this2.onViewed(videoState);
        _this2.setActions(actions);
        //Video State(either prop or variable) is ReadOnly
        _this2.videoState = videoState;
        var status = videoState.status,
          currentTime = videoState.currentTime,
          buffered = videoState.buffered,
          duration = videoState.duration,
          isLoading = videoState.isLoading;
        var playerSize = _this2.state.playerSize;
        var isPlaying = status === 'playing';
        var isLargePlayer = playerSize === 'large';
        var isMediumPlayer = playerSize === 'medium';
        var defaultSkipAmount = 10;
        var skipBackward = function skipBackward() {
          var skipAmount = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : defaultSkipAmount;
          var newTime = videoState.currentTime - skipAmount;
          actions.navigate(Math.max(newTime, 0));
        };
        var skipForward = function skipForward() {
          var skipAmount = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : defaultSkipAmount;
          var newTime = videoState.currentTime + skipAmount;
          actions.navigate(Math.min(newTime, videoState.duration));
        };
        var shortcuts = _this2.renderShortcuts({
          togglePlayPauseAction: isPlaying ? _this2.pause : _this2.play,
          toggleMute: actions.toggleMute,
          skipBackward: skipBackward,
          skipForward: skipForward
        });
        return /*#__PURE__*/_react.default.createElement(_primitives.Flex, {
          direction: "column",
          xcss: videoWrapperStyles
        }, /*#__PURE__*/_react.default.createElement(_widthDetector.WidthObserver, {
          setWidth: _this2.onResize
        }), shortcuts, isLoading && _this2.renderSpinner(), /*#__PURE__*/_react.default.createElement(_playPauseBlanket.PlayPauseBlanket, {
          onDoubleClick: _this2.doubleClickToFullscreen,
          onClick: isPlaying ? _this2.pausePlayByBlanketClick : _this2.startPlayByBlanketClick,
          "data-testid": "play-pause-blanket"
        }, video), /*#__PURE__*/_react.default.createElement(_styledEmotion.ControlsWrapper, {
          ref: videoControlsWrapperRef
          // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
          ,
          className: (0, _getControlsWrapperClassName.getControlsWrapperClassName)(_this2.wasPlayedOnce)
        }, /*#__PURE__*/_react.default.createElement(_primitives.Box, {
          xcss: timeWrapperStyles
        }, /*#__PURE__*/_react.default.createElement(_timeRange.TimeRange, {
          currentTime: currentTime,
          bufferedTime: buffered,
          duration: duration,
          onChange: actions.navigate,
          onChanged: _this2.onTimeChanged,
          disableThumbTooltip: true,
          skipBackward: skipBackward,
          skipForward: skipForward,
          isAlwaysActive: false
        })), /*#__PURE__*/_react.default.createElement(_primitives.Flex, {
          alignItems: "center",
          justifyContent: "space-between",
          xcss: timebarWrapperStyles
        }, /*#__PURE__*/_react.default.createElement(_styled.LeftControls, null, _this2.renderPlayPauseButton(isPlaying), isLargePlayer && _this2.renderSkipBackwardButton(skipBackward), isLargePlayer && _this2.renderSkipForwardButton(skipForward), _this2.renderVolume(videoState, actions, isLargePlayer)), /*#__PURE__*/_react.default.createElement(_styled.RightControls, null, (isMediumPlayer || isLargePlayer) && _this2.renderCurrentTime(videoState), isLargePlayer && !(0, _platformFeatureFlags.fg)('platform_media_disable_video_640p_artifact_usage') && _this2.renderHDButton(), isLargePlayer && _this2.renderSpeedControls(), _this2.renderFullScreenButton(), isLargePlayer && _this2.renderDownloadButton()))));
      }));
    }
  }]);
}(_react.Component);
