// @atlaskit/media-ui/dist/cjs/customMediaPlayer/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CustomMediaPlayerBase = exports.CustomMediaPlayer = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactIntlNext = require("react-intl-next");
var _analyticsNext = require("@atlaskit/analytics-next");
var _indexEmotion = require("./index-emotion");
var _indexCompiled = require("./index-compiled");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _mediaCommon = require("@atlaskit/media-common");
var CustomMediaPlayerBase = exports.CustomMediaPlayerBase = function CustomMediaPlayerBase(props) {
  return (0, _platformFeatureFlags.fg)('platform_media_compiled') ? /*#__PURE__*/_react.default.createElement(_indexCompiled.CustomMediaPlayerBase, props) : /*#__PURE__*/_react.default.createElement(_indexEmotion.CustomMediaPlayerBase, props);
};
var packageName = "@atlaskit/media-ui";
var packageVersion = "28.1.7";

// @ts-ignore: [PIT-1685] Fails in post-office due to backwards incompatibility issue with React 18
var CustomMediaPlayer = exports.CustomMediaPlayer = (0, _mediaCommon.withMediaAnalyticsContext)({
  packageVersion: packageVersion,
  packageName: packageName,
  componentName: 'customMediaPlayer',
  component: 'customMediaPlayer'
})((0, _analyticsNext.withAnalyticsEvents)()((0, _reactIntlNext.injectIntl)(CustomMediaPlayerBase)));
