// @atlaskit/media-ui/dist/cjs/customMediaPlayer/styled-compiled.js
/* styled-compiled.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VolumeWrapper = exports.VolumeToggleWrapper = exports.VolumeTimeRangeWrapper = exports.TimeRangeWrapper = exports.TimeLine = exports.Thumb = exports.RightControls = exports.MutedIndicator = exports.LeftControls = exports.CurrentTimeTooltip = exports.CurrentTimeLineThumb = exports.CurrentTimeLine = exports.CurrentTime = exports.ControlsWrapper = exports.BufferedTime = void 0;
require("./styled-compiled.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireWildcard(require("react"));
var _colors = require("@atlaskit/theme/colors");
var _getControlsWrapperClassName = require("./getControlsWrapperClassName");
var _excluded = ["showSlider", "children"],
  _excluded2 = ["children"],
  _excluded3 = ["children"],
  _excluded4 = ["children"],
  _excluded5 = ["children"],
  _excluded6 = ["children"],
  _excluded7 = ["children"],
  _excluded8 = ["children"],
  _excluded9 = ["children"],
  _excluded10 = ["children", "controlsHidden"],
  _excluded11 = ["isMuted", "children"],
  _excluded12 = ["children"],
  _excluded13 = ["isMuted", "children"],
  _excluded14 = ["isDragging", "timeLineThumbIsHover", "timeLineThumbIsFocus", "children"],
  _excluded15 = ["children"];
/* eslint-disable @atlaskit/design-system/ensure-design-token-usage */
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var volumeStyles = null;
var showSliderVolumeStyles = null;
var VolumeWrapper = exports.VolumeWrapper = function VolumeWrapper(_ref) {
  var showSlider = _ref.showSlider,
    children = _ref.children,
    props = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, props, {
    className: (0, _runtime.ax)(["_1reo15vq _18m915vq _v5641rb3 _1e0c1txw _1bsbldb6 _4cvr1h6o _94n5ze3t _1ltv1jfw", showSlider && "_13y6o43e _195p1w81 _o7kso43e _1chp1w81 _1gglo43e _m7bd1w81"])
  }), children);
};
var currentTimeStyles = null;
var CurrentTime = exports.CurrentTime = function CurrentTime(_ref2) {
  var children = _ref2.children,
    props = (0, _objectWithoutProperties2.default)(_ref2, _excluded2);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, props, {
    className: (0, _runtime.ax)(["_syaz1poe _uiztglyw _2hwxu2gc _o5721q9c"])
  }), children);
};
var timelineStyles = null;
var TimeLine = exports.TimeLine = /*#__PURE__*/(0, _react.forwardRef)(function (_ref3, ref) {
  var children = _ref3.children,
    props = (0, _objectWithoutProperties2.default)(_ref3, _excluded3);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    ref: ref
  }, props, {
    className: (0, _runtime.ax)(["_v5641eho _2rko14y2 _1bsb1osq _4t3iyh40 _1lww9cwz _bfhk9ucv _kqswh2mm"])
  }), children);
});
var currentTimeLineStyles = null;
var CurrentTimeLine = exports.CurrentTimeLine = function CurrentTimeLine(_ref4) {
  var children = _ref4.children,
    props = (0, _objectWithoutProperties2.default)(_ref4, _excluded4);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, props, {
    className: (0, _runtime.ax)(["_2rko1kw7 _bfhk7g9i _4t3i1kw7 _kqswstnw _154ize3t _p12f1osq"])
  }), children);
};
var thumbStyles = null;
var Thumb = exports.Thumb = function Thumb(_ref5) {
  var children = _ref5.children,
    props = (0, _objectWithoutProperties2.default)(_ref5, _excluded5);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, props, {
    className: (0, _runtime.ax)(["_2rko1osq _v5641eho _189et94y _1dqonqa1 _1h6d16fc _lcxvglyw _1bsbdlk8 _4t3idlk8 _bfhku67f _kqswstnw _1xi2idpf _154iv77o _t9ec17tb _1lww9cwz"])
  }), children);
};
var currentTimeLineThumbStyles = null;
var CurrentTimeLineThumb = exports.CurrentTimeLineThumb = /*#__PURE__*/(0, _react.forwardRef)(function (_ref6, ref) {
  var children = _ref6.children,
    props = (0, _objectWithoutProperties2.default)(_ref6, _excluded6);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    ref: ref
  }, props, {
    className: (0, _runtime.ax)(["_19itglyw _2rko1osq _1a37q36z _kqswstnw _1e0c1ule _1xi2idpf _154i1ssb _t9ec1tul _bfhk7g9i _4t3ik7ql _1bsbk7ql _lcxvglyw _tzy4idpf _1cwgv77o _1h7hkb7n"]),
    style: {
      "--_k4vj5w": (0, _runtime.ix)("2px solid ".concat("var(--ds-border-focused, #85B8FF)"))
    }
  }), children);
});
var bufferedTimeStyles = null;
var BufferedTime = exports.BufferedTime = function BufferedTime(_ref7) {
  var children = _ref7.children,
    props = (0, _objectWithoutProperties2.default)(_ref7, _excluded7);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, props, {
    className: (0, _runtime.ax)(["_2rko1kw7 _bfhk1ncv _4t3i1kw7 _1bsbidpf"])
  }), children);
};
var leftControlsStyles = null;
var LeftControls = exports.LeftControls = function LeftControls(_ref8) {
  var children = _ref8.children,
    props = (0, _objectWithoutProperties2.default)(_ref8, _excluded8);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, props, {
    className: (0, _runtime.ax)(["_1e0c1txw _18u0utpp"])
  }), children);
};
var rightControlStyles = null;
var RightControls = exports.RightControls = function RightControls(_ref9) {
  var children = _ref9.children,
    props = (0, _objectWithoutProperties2.default)(_ref9, _excluded9);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, props, {
    className: (0, _runtime.ax)(["_1e0c1txw _4cvr1h6o _2hwxutpp"])
  }), children);
};
var controlsWrapperStyles = null;
var hiddenControlsWrapperStyles = null;
var ControlsWrapper = exports.ControlsWrapper = /*#__PURE__*/(0, _react.forwardRef)(function (_ref10, ref) {
  var children = _ref10.children,
    controlsHidden = _ref10.controlsHidden,
    props = (0, _objectWithoutProperties2.default)(_ref10, _excluded10);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    ref: ref
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop
    ,
    className: (0, _runtime.ax)(["_11q7acks _94n5idpf _1ltvidpf _1bsb1osq _4t3i1wug _kqswstnw", controlsHidden && "_1mygkb7n", (0, _getControlsWrapperClassName.getControlsWrapperClassName)(controlsHidden)])
  }, props), children);
});
var volumeToggleWrapperStyles = null;
var mutedVolumeToggleWrapperStyles = null;
var VolumeToggleWrapper = exports.VolumeToggleWrapper = function VolumeToggleWrapper(_ref11) {
  var isMuted = _ref11.isMuted,
    children = _ref11.children,
    props = (0, _objectWithoutProperties2.default)(_ref11, _excluded11);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, props, {
    className: (0, _runtime.ax)(["_kqswh2mm _2hwxw1gv _14hp1j2z", isMuted && "_d32vqb75"])
  }), children);
};
var volumeTimeRangeWrapperStyles = null;
var VolumeTimeRangeWrapper = exports.VolumeTimeRangeWrapper = function VolumeTimeRangeWrapper(_ref12) {
  var children = _ref12.children,
    props = (0, _objectWithoutProperties2.default)(_ref12, _excluded12);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, props, {
    className: (0, _runtime.ax)(["_1bsb1osq _2hwxv47k"])
  }), children);
};
var mutedIndicatorStyles = null;
var isMutedStyles = null;
var MutedIndicator = exports.MutedIndicator = function MutedIndicator(_ref13) {
  var isMuted = _ref13.isMuted,
    children = _ref13.children,
    props = (0, _objectWithoutProperties2.default)(_ref13, _excluded13);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, props, {
    className: (0, _runtime.ax)(["_bfhk18rw _1bsb1pa9 _4t3iyh40 _kqswstnw _154iu2gc _1ltvu2gc _1pbycs5v _t9ec17wz _tzy4idpf _lcxvglyw", isMuted && "_tzy4kb7n"])
  }), children);
};
var currentTimeTooltipStyles = null;
var currentTimeToolTipDraggingStyles = null;
var currentTimeToolTipNotDraggingStyles = null;
var currentTimeTooltipThumbHoveredStyles = null;
var currentTimeTooltipThumbFocusedStyles = null;
var CurrentTimeTooltip = exports.CurrentTimeTooltip = function CurrentTimeTooltip(_ref14) {
  var isDragging = _ref14.isDragging,
    timeLineThumbIsHover = _ref14.timeLineThumbIsHover,
    timeLineThumbIsFocus = _ref14.timeLineThumbIsFocus,
    children = _ref14.children,
    props = (0, _objectWithoutProperties2.default)(_ref14, _excluded14);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    className: (0, _runtime.ax)(["_11c8dcr7 _2rko1b66 _v564g17y _kqswstnw _uiztglyw _154iukr8 _bfhk39qu _syaz1pcz _ca0q1b66 _u5f3u2gc _n3td1b66 _19bvu2gc _1ltv1ssb _t9eczr6s _1nmz16rz", isDragging && "_tzy4kb7n", !isDragging && "_tzy4idpf", timeLineThumbIsHover && "_tzy4kb7n", timeLineThumbIsFocus && "_tzy4kb7n", "current-time-tooltip"])
  }, props), children);
};
var timeRangeWrapperStyles = null;
var TimeRangeWrapper = exports.TimeRangeWrapper = /*#__PURE__*/(0, _react.forwardRef)(function (_ref15, ref) {
  var children = _ref15.children,
    props = (0, _objectWithoutProperties2.default)(_ref15, _excluded15);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    ref: ref
  }, props, {
    className: (0, _runtime.ax)(["_1e0c1txw _4cvr1h6o _4t3iqslr _80omtlke _1bsb1osq _1qcea6sd _1az8glyw _1838glyw _4i4mdlk8 _9cyhdlk8 _14lz1lpd _u2q4a6sd _17dfdlk8 _rs6wdlk8 _1w9uglyw _7cve9ucv _iofsa6sd _1a96a6sd _dey49ucv"])
  }), children);
});
