// @atlaskit/media-ui/dist/cjs/customMediaPlayer/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VolumeWrapper = exports.VolumeToggleWrapper = exports.VolumeTimeRangeWrapper = exports.TimeLine = exports.Thumb = exports.RightControls = exports.MutedIndicator = exports.LeftControls = exports.CurrentTimeLineThumb = exports.CurrentTimeLine = exports.CurrentTime = exports.BufferedTime = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _styledEmotion = require("./styled-emotion");
var _styledCompiled = require("./styled-compiled");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var VolumeWrapper = exports.VolumeWrapper = function VolumeWrapper(props) {
  return (0, _platformFeatureFlags.fg)('platform_media_compiled') ? /*#__PURE__*/_react.default.createElement(_styledCompiled.VolumeWrapper, props) : /*#__PURE__*/_react.default.createElement(_styledEmotion.VolumeWrapper, props);
};
var CurrentTime = exports.CurrentTime = function CurrentTime(props) {
  return (0, _platformFeatureFlags.fg)('platform_media_compiled') ? /*#__PURE__*/_react.default.createElement(_styledCompiled.CurrentTime, props) : /*#__PURE__*/_react.default.createElement(_styledEmotion.CurrentTime, props);
};
var TimeLine = exports.TimeLine = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  return (0, _platformFeatureFlags.fg)('platform_media_compiled') ? /*#__PURE__*/_react.default.createElement(_styledCompiled.TimeLine, (0, _extends2.default)({}, props, {
    ref: ref
  })) : /*#__PURE__*/_react.default.createElement(_styledEmotion.TimeLine, (0, _extends2.default)({}, props, {
    ref: ref
  }));
});
var CurrentTimeLine = exports.CurrentTimeLine = function CurrentTimeLine(props) {
  return (0, _platformFeatureFlags.fg)('platform_media_compiled') ? /*#__PURE__*/_react.default.createElement(_styledCompiled.CurrentTimeLine, props) : /*#__PURE__*/_react.default.createElement(_styledEmotion.CurrentTimeLine, props);
};
var Thumb = exports.Thumb = function Thumb(props) {
  return (0, _platformFeatureFlags.fg)('platform_media_compiled') ? /*#__PURE__*/_react.default.createElement(_styledCompiled.Thumb, props) : /*#__PURE__*/_react.default.createElement(_styledEmotion.Thumb, props);
};
var CurrentTimeLineThumb = exports.CurrentTimeLineThumb = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  return (0, _platformFeatureFlags.fg)('platform_media_compiled') ? /*#__PURE__*/_react.default.createElement(_styledCompiled.CurrentTimeLineThumb, (0, _extends2.default)({}, props, {
    ref: ref
  })) : /*#__PURE__*/_react.default.createElement(_styledEmotion.CurrentTimeLineThumb, (0, _extends2.default)({}, props, {
    ref: ref
  }));
});
var BufferedTime = exports.BufferedTime = function BufferedTime(props) {
  return (0, _platformFeatureFlags.fg)('platform_media_compiled') ? /*#__PURE__*/_react.default.createElement(_styledCompiled.BufferedTime, props) : /*#__PURE__*/_react.default.createElement(_styledEmotion.BufferedTime, props);
};
var LeftControls = exports.LeftControls = function LeftControls(props) {
  return (0, _platformFeatureFlags.fg)('platform_media_compiled') ? /*#__PURE__*/_react.default.createElement(_styledCompiled.LeftControls, props) : /*#__PURE__*/_react.default.createElement(_styledEmotion.LeftControls, props);
};
var RightControls = exports.RightControls = function RightControls(props) {
  return (0, _platformFeatureFlags.fg)('platform_media_compiled') ? /*#__PURE__*/_react.default.createElement(_styledCompiled.RightControls, props) : /*#__PURE__*/_react.default.createElement(_styledEmotion.RightControls, props);
};
var VolumeToggleWrapper = exports.VolumeToggleWrapper = function VolumeToggleWrapper(props) {
  return (0, _platformFeatureFlags.fg)('platform_media_compiled') ? /*#__PURE__*/_react.default.createElement(_styledCompiled.VolumeToggleWrapper, props) : /*#__PURE__*/_react.default.createElement(_styledEmotion.VolumeToggleWrapper, props);
};
var VolumeTimeRangeWrapper = exports.VolumeTimeRangeWrapper = function VolumeTimeRangeWrapper(props) {
  return (0, _platformFeatureFlags.fg)('platform_media_compiled') ? /*#__PURE__*/_react.default.createElement(_styledCompiled.VolumeTimeRangeWrapper, props) : /*#__PURE__*/_react.default.createElement(_styledEmotion.VolumeTimeRangeWrapper, props);
};
var MutedIndicator = exports.MutedIndicator = function MutedIndicator(props) {
  return (0, _platformFeatureFlags.fg)('platform_media_compiled') ? /*#__PURE__*/_react.default.createElement(_styledCompiled.MutedIndicator, props) : /*#__PURE__*/_react.default.createElement(_styledEmotion.MutedIndicator, props);
};
