// @atlaskit/editor-plugin-selection-extension/dist/cjs/ui/toolbar/SelectionExtensionDropdownMenuButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SelectionExtensionDropdownMenuButton = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactIntlNext = require("react-intl-next");
var _messages = require("@atlaskit/editor-common/messages");
var _uiMenu = require("@atlaskit/editor-common/ui-menu");
var _apps = _interopRequireDefault(require("@atlaskit/icon/core/apps"));
var _chevronDown = _interopRequireDefault(require("@atlaskit/icon/utility/migration/chevron-down"));
var SelectionExtensionDropdownMenuButtonComponent = function SelectionExtensionDropdownMenuButtonComponent(_ref) {
  var onClick = _ref.onClick,
    selected = _ref.selected,
    ariaExpanded = _ref['aria-expanded'],
    intl = _ref.intl;
  return /*#__PURE__*/_react.default.createElement(_uiMenu.ToolbarButton, {
    testId: "selection-extension-dropdown-button",
    "aria-label": intl.formatMessage(_messages.selectionExtensionMessages.selectionExtensionDropdownButtonLabel),
    "aria-haspopup": "true",
    spacing: "compact",
    title: intl.formatMessage(_messages.selectionExtensionMessages.selectionExtensionDropdownButtonLabel),
    onClick: onClick,
    "aria-expanded": ariaExpanded,
    selected: selected,
    iconAfter: /*#__PURE__*/_react.default.createElement(_chevronDown.default, {
      spacing: "none",
      label: intl.formatMessage(_messages.selectionExtensionMessages.selectionExtensionDropdownButtonLabel)
    })
  }, /*#__PURE__*/_react.default.createElement(_apps.default, {
    label: intl.formatMessage(_messages.selectionExtensionMessages.selectionExtensionDropdownButtonLabel),
    spacing: "none"
  }));
};
var SelectionExtensionDropdownMenuButton = exports.SelectionExtensionDropdownMenuButton = (0, _reactIntlNext.injectIntl)(SelectionExtensionDropdownMenuButtonComponent);
