// @atlaskit/editor-plugin-selection-extension/dist/cjs/ui/toolbar/SelectionExtensionItems.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SelectionExtensionItemsComponent = exports.SelectionExtensionItems = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _reactIntlNext = require("react-intl-next");
var _uuid = require("uuid");
var _analytics = require("@atlaskit/editor-common/analytics");
var _SelectionExtensionDropdownMenu = require("./SelectionExtensionDropdownMenu");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; } /**
 * @jsxRuntime classic
 * @jsx
 */
var transformExtensionsToItems = function transformExtensionsToItems(extensions) {
  var extensionToItems = extensions.map(function (extension) {
    return {
      key: extension.id,
      content: extension.name,
      value: {
        name: extension.id
      }
    };
  });
  return [{
    items: extensionToItems
  }];
};
var SelectionExtensionItemsComponent = exports.SelectionExtensionItemsComponent = function SelectionExtensionItemsComponent(_ref) {
  var extensions = _ref.extensions,
    onExtensionClick = _ref.onExtensionClick,
    editorAnalyticsAPI = _ref.editorAnalyticsAPI;
  var extensionsWithIdentifier = (0, _react.useMemo)(function () {
    return extensions.map(function (extension) {
      return _objectSpread(_objectSpread({}, extension), {}, {
        id: (0, _uuid.v4)()
      });
    });
  }, [extensions]);
  var items = (0, _react.useMemo)(function () {
    return transformExtensionsToItems(extensionsWithIdentifier);
  }, [extensionsWithIdentifier]);
  var handleOnItemActivated = function handleOnItemActivated(_ref2) {
    var item = _ref2.item;
    var extension = extensionsWithIdentifier.find(function (ext) {
      return ext.id === item.value.name;
    });
    if (extension) {
      onExtensionClick(extension);
      if (editorAnalyticsAPI) {
        editorAnalyticsAPI.fireAnalyticsEvent({
          action: _analytics.ACTION.CLICKED,
          actionSubject: _analytics.ACTION_SUBJECT.BUTTON,
          actionSubjectId: _analytics.ACTION_SUBJECT_ID.EDITOR_PLUGIN_SELECTION_EXTENSION_ITEM,
          eventType: _analytics.EVENT_TYPE.TRACK
        });
      }
    }
  };
  return /*#__PURE__*/_react.default.createElement(_SelectionExtensionDropdownMenu.SelectionExtensionDropdownMenu, {
    editorAnalyticsAPI: editorAnalyticsAPI,
    items: items,
    onItemActivated: handleOnItemActivated
  });
};
var SelectionExtensionItems = exports.SelectionExtensionItems = (0, _reactIntlNext.injectIntl)(SelectionExtensionItemsComponent);
