// @atlaskit/jql-ast/dist/cjs/types/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "isChangedOperator", {
  enumerable: true,
  get: function get() {
    return _ast.isChangedOperator;
  }
});
Object.defineProperty(exports, "isComparisonOperator", {
  enumerable: true,
  get: function get() {
    return _ast.isComparisonOperator;
  }
});
Object.defineProperty(exports, "isEqualsOperator", {
  enumerable: true,
  get: function get() {
    return _ast.isEqualsOperator;
  }
});
Object.defineProperty(exports, "isInOperator", {
  enumerable: true,
  get: function get() {
    return _ast.isInOperator;
  }
});
Object.defineProperty(exports, "isIsOperator", {
  enumerable: true,
  get: function get() {
    return _ast.isIsOperator;
  }
});
Object.defineProperty(exports, "isLikeOperator", {
  enumerable: true,
  get: function get() {
    return _ast.isLikeOperator;
  }
});
Object.defineProperty(exports, "isListOperator", {
  enumerable: true,
  get: function get() {
    return _ast.isListOperator;
  }
});
Object.defineProperty(exports, "isOperandNode", {
  enumerable: true,
  get: function get() {
    return _ast.isOperandNode;
  }
});
Object.defineProperty(exports, "isOperator", {
  enumerable: true,
  get: function get() {
    return _ast.isOperator;
  }
});
Object.defineProperty(exports, "isPredicateOperator", {
  enumerable: true,
  get: function get() {
    return _ast.isPredicateOperator;
  }
});
Object.defineProperty(exports, "isWasInOperator", {
  enumerable: true,
  get: function get() {
    return _ast.isWasInOperator;
  }
});
Object.defineProperty(exports, "isWasOperator", {
  enumerable: true,
  get: function get() {
    return _ast.isWasOperator;
  }
});
var _ast = require("./ast");
