// @atlaskit/editor-plugin-block-controls/dist/es2019/pm-plugins/interaction-tracking/pm-plugin.js
import { SafePlugin } from '@atlaskit/editor-common/safe-plugin';
import { PluginKey } from '@atlaskit/editor-prosemirror/state';
import { handleKeyDown } from './handle-key-down';
import { handleMouseMove } from './handle-mouse-move';
export const interactionTrackingPluginKey = new PluginKey('interactionTrackingPlugin');
export const createInteractionTrackingPlugin = () => {
  return new SafePlugin({
    key: interactionTrackingPluginKey,
    state: {
      init() {
        return {
          isEditing: false
        };
      },
      apply(tr, pluginState) {
        const meta = tr.getMeta(interactionTrackingPluginKey);
        switch (meta === null || meta === void 0 ? void 0 : meta.type) {
          case 'startEditing':
            return {
              isEditing: true
            };
          case 'stopEditing':
            return {
              isEditing: false
            };
        }
        return pluginState;
      }
    },
    props: {
      handleKeyDown,
      handleDOMEvents: {
        mousemove: handleMouseMove
      }
    }
  });
};
export const getInteractionTrackingState = state => {
  return interactionTrackingPluginKey.getState(state);
};
