// @atlaskit/teams-public/dist/cjs/ui/team-containers/no-product-access-empty-state/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NoProductAccessState = void 0;
require("./index.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _react = _interopRequireDefault(require("react"));
var _reactIntlNext = require("react-intl-next");
var _heading = _interopRequireDefault(require("@atlaskit/heading"));
var _compiled = require("@atlaskit/primitives/compiled");
var _noProductAccessIcon = require("./no-product-access-icon");
var styles = {
  emptyStateContainer: "_2rkoiti9 _189ee4h9 _1h6dmuej _18zrxy5q _12y3e4h9 _12jimuej _1qu2nqa1",
  contentWrapper: "_bozgxy5q"
};
var NoProductAccessState = exports.NoProductAccessState = function NoProductAccessState() {
  return /*#__PURE__*/_react.default.createElement(_compiled.Stack, {
    space: "space.100",
    xcss: styles.emptyStateContainer,
    testId: "team-containers-no-product-access-state"
  }, /*#__PURE__*/_react.default.createElement(_compiled.Inline, {
    alignBlock: "center"
  }, /*#__PURE__*/_react.default.createElement(_noProductAccessIcon.NoProductAccessIcon, null), /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    xcss: styles.contentWrapper
  }, /*#__PURE__*/_react.default.createElement(_compiled.Stack, {
    space: "space.100"
  }, /*#__PURE__*/_react.default.createElement(_heading.default, {
    size: "small"
  }, /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, messages.teamContainerEmptyStateTitle)), /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, messages.teamContainerEmptyStateDescription)))));
};
var messages = (0, _reactIntlNext.defineMessages)({
  teamContainerEmptyStateTitle: {
    id: 'ptc.team-profile-page.team-containers.empty-state.title',
    defaultMessage: 'Add the places your team works',
    description: 'Empty state title when user has no product access'
  },
  teamContainerEmptyStateDescription: {
    id: 'ptc.team-profile-page.team-containers.empty-state.description',
    defaultMessage: 'Your team can link their Jira projects, Confluence spaces here to show where they work',
    description: 'Empty state description when user has no product access'
  }
});
