// @atlaskit/media-ui/dist/cjs/MediaInlineCard/ErroredView/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MediaInlineCardErroredView = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireDefault(require("react"));
var _colors = require("@atlaskit/theme/colors");
var _warning = _interopRequireDefault(require("@atlaskit/icon/utility/migration/warning"));
var _Frame = require("../Frame");
var _IconAndTitleLayout = require("../IconAndTitleLayout");
var _Icon = require("../Icon");
var _styled = require("../styled");
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var MediaInlineCardErroredView = exports.MediaInlineCardErroredView = /*#__PURE__*/function (_React$Component) {
  function MediaInlineCardErroredView() {
    var _this;
    (0, _classCallCheck2.default)(this, MediaInlineCardErroredView);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, MediaInlineCardErroredView, [].concat(args));
    (0, _defineProperty2.default)(_this, "renderMessage", function () {
      var message = _this.props.message;
      var errorMessage = /*#__PURE__*/_react.default.createElement(_styled.NoLinkAppearance, null, message);
      return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, errorMessage);
    });
    return _this;
  }
  (0, _inherits2.default)(MediaInlineCardErroredView, _React$Component);
  return (0, _createClass2.default)(MediaInlineCardErroredView, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
        onClick = _this$props.onClick,
        isSelected = _this$props.isSelected,
        innerRef = _this$props.innerRef,
        _this$props$testId = _this$props.testId,
        testId = _this$props$testId === void 0 ? 'media-inline-card-errored-view' : _this$props$testId,
        icon = _this$props.icon;
      return /*#__PURE__*/_react.default.createElement(_Frame.Frame, {
        testId: testId,
        innerRef: innerRef,
        onClick: onClick,
        isSelected: isSelected
      }, /*#__PURE__*/_react.default.createElement(_IconAndTitleLayout.IconAndTitleLayout, {
        icon: icon || /*#__PURE__*/_react.default.createElement(_Icon.AKIconWrapper, null, /*#__PURE__*/_react.default.createElement(_warning.default, {
          label: "error",
          LEGACY_size: "small",
          color: "var(--ds-icon-danger, ".concat(_colors.R300, ")")
        })),
        title: this.renderMessage()
      }));
    }
  }]);
}(_react.default.Component);
