// @atlaskit/media-ui/dist/esm/ellipsify-compiled.js
/* ellipsify-compiled.tsx generated by @compiled/babel-plugin v0.36.1 */
import "./ellipsify-compiled.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { useRef, useEffect } from 'react';
import * as exenv from 'exenv';
var styleMap = {
  inline: "_1e0c1nu9",
  notInline: ""
};
export var Ellipsify = function Ellipsify(_ref) {
  var text = _ref.text,
    lines = _ref.lines,
    endLength = _ref.endLength,
    inline = _ref.inline,
    testId = _ref.testId;
  var element = useRef(null);
  useEffect(function () {
    if (!element.current) {
      return;
    }
    setEllipsis(element.current, {
      lines: lines,
      endLength: endLength
    });
  }, [element, lines, endLength]);
  return /*#__PURE__*/React.createElement("div", {
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    className: ax([styleMap[inline ? 'inline' : 'notInline'], "ellipsed-text"]),
    ref: element,
    "aria-label": text,
    "data-testid": testId
  }, text);
};
var setEllipsis = function setEllipsis(element, props) {
  var maximumLines = props.lines;
  var height = element.getBoundingClientRect().height;
  var text = element.textContent;
  element.textContent = 'a';
  var lineHeight = element.getBoundingClientRect().height;
  var lineCount = height / lineHeight;
  var maximumHeight = lineHeight * maximumLines;
  if (lineCount <= maximumLines) {
    element.textContent = text;
    return;
  }
  var textContent = text;
  var endLength = typeof props.endLength === 'number' && props.endLength >= 0 ? props.endLength : 8;
  var beginningText = text.substr(0, text.length * maximumLines / lineCount - endLength);
  var endText = text.substr(text.length - endLength, endLength);
  element.textContent = textContent = "".concat(beginningText, "...").concat(endText);
  var finalHeight = element.getBoundingClientRect().height;
  if (finalHeight > maximumHeight) {
    var adjustedBeginningText = beginningText.substr(0, beginningText.length - beginningText.length / maximumLines * 0.25);
    textContent = "".concat(adjustedBeginningText, "...").concat(endText);
  }
  delayRun(function () {
    return element.textContent = textContent;
  });
};
var timeout = function timeout(fn) {
  return setTimeout(fn, 1);
};
var delayRun = exenv.canUseDOM && window.requestAnimationFrame ? window.requestAnimationFrame : timeout;
export default Ellipsify;
