// @atlaskit/media-ui/dist/esm/ellipsify-emotion.js
import React, { useRef, useEffect } from 'react';
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import styled from '@emotion/styled';
import * as exenv from 'exenv';
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-dynamic-styles -- Ignored via go/DSP-18766
var Wrapper = styled.div(function (_ref) {
  var inline = _ref.inline;
  return {
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    display: inline ? 'inline' : undefined
  };
});
Wrapper.displayName = 'Ellipsify';
export var Ellipsify = function Ellipsify(_ref2) {
  var text = _ref2.text,
    lines = _ref2.lines,
    endLength = _ref2.endLength,
    inline = _ref2.inline,
    testId = _ref2.testId;
  var element = useRef(null);
  useEffect(function () {
    if (!element.current) {
      return;
    }
    setEllipsis(element.current, {
      lines: lines,
      endLength: endLength
    });
  }, [element, lines, endLength]);
  return /*#__PURE__*/React.createElement(Wrapper
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
  , {
    className: "ellipsed-text",
    ref: element,
    "aria-label": text,
    inline: inline,
    "data-testid": testId
  }, text);
};
var setEllipsis = function setEllipsis(element, props) {
  var maximumLines = props.lines;
  var height = element.getBoundingClientRect().height;
  var text = element.textContent;
  element.textContent = 'a';
  var lineHeight = element.getBoundingClientRect().height;
  var lineCount = height / lineHeight;
  var maximumHeight = lineHeight * maximumLines;
  if (lineCount <= maximumLines) {
    element.textContent = text;
    return;
  }
  var textContent = text;
  var endLength = typeof props.endLength === 'number' && props.endLength >= 0 ? props.endLength : 8;
  var beginningText = text.substr(0, text.length * maximumLines / lineCount - endLength);
  var endText = text.substr(text.length - endLength, endLength);
  element.textContent = textContent = "".concat(beginningText, "...").concat(endText);
  var finalHeight = element.getBoundingClientRect().height;
  if (finalHeight > maximumHeight) {
    var adjustedBeginningText = beginningText.substr(0, beginningText.length - beginningText.length / maximumLines * 0.25);
    textContent = "".concat(adjustedBeginningText, "...").concat(endText);
  }
  delayRun(function () {
    return element.textContent = textContent;
  });
};
var timeout = function timeout(fn) {
  return setTimeout(fn, 1);
};
var delayRun = exenv.canUseDOM && window.requestAnimationFrame ? window.requestAnimationFrame : timeout;
export default Ellipsify;
