// @atlaskit/date/dist/cjs/components/Date.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Date = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _DateLozenge = require("./DateLozenge");
var _react = _interopRequireDefault(require("react"));
var _format = _interopRequireDefault(require("date-fns/format"));
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var isClickable = function isClickable(props) {
  return !!props.onClick;
};
var Date = exports.Date = /*#__PURE__*/function (_React$Component) {
  function Date() {
    var _this;
    (0, _classCallCheck2.default)(this, Date);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, Date, [].concat(args));
    (0, _defineProperty2.default)(_this, "handleOnClick", function (event) {
      if (isClickable(_this.props)) {
        _this.props.onClick(_this.props.value, event);
      }
    });
    (0, _defineProperty2.default)(_this, "renderContent", function () {
      if (_this.props.children) {
        if (typeof _this.props.children === 'function') {
          return (
            // prettier-ignore
            // eslint-disable-next-line @typescript-eslint/ban-types
            _this.props.children(_this.props)
          );
        }
        return _this.props.children;
      }
      return (0, _format.default)(_this.props.value, _this.props.format || '');
    });
    return _this;
  }
  (0, _inherits2.default)(Date, _React$Component);
  return (0, _createClass2.default)(Date, [{
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement(_DateLozenge.DateLozenge
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
      , {
        className: this.props.className,
        onClick: isClickable(this.props) ? this.handleOnClick : undefined,
        color: this.props.color
      }, this.renderContent());
    }
  }]);
}(_react.default.Component);
(0, _defineProperty2.default)(Date, "displayName", 'Date');
(0, _defineProperty2.default)(Date, "defaultProps", {
  format: 'dd/MM/yyyy',
  color: 'grey'
});
