// @atlaskit/date/dist/cjs/components/DateLozenge.js
/* DateLozenge.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DateLozenge = void 0;
require("./DateLozenge.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireDefault(require("react"));
var _excluded = ["className", "color", "children"];
var baseStyles = null;
var onClickStyles = null;
var colorStyles = {
  red: "_bfhk1vk5 _syaz9lu1 _irr31rtt",
  yellow: "_bfhk1ozg _syazfb2s _irr33fv2",
  blue: "_bfhkavxe _syazpnps _irr34v9p",
  green: "_bfhkn729 _syazh55r _irr31wnw",
  purple: "_bfhkp7j4 _syaz10m5 _irr31jkz",
  grey: "_bfhkm7j4 _syaz1fxt _irr315ej"
};
var DateLozenge = exports.DateLozenge = function DateLozenge(props) {
  var className = props.className,
    _props$color = props.color,
    color = _props$color === void 0 ? 'grey' : _props$color,
    children = props.children,
    rest = (0, _objectWithoutProperties2.default)(props, _excluded);
  return /*#__PURE__*/_react.default.createElement("span", (0, _extends2.default)({}, rest, {
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop
    className: (0, _runtime.ax)(["_2rko1sit _19pkidpf _2hwxt94y _otyridpf _18u0t94y _v5641y8x _ca0qv77o _u5f31b66 _n3tdv77o _19bv1b66 _kqswh2mm _o5721q9c _80omn7od", props.onClick && "_80omtlke", colorStyles[color], className]),
    role: "button"
  }), children);
};
