// @atlaskit/smart-card/dist/es2019/view/FlexibleCard/components/actions/automation-action/automation-manual-triggers/manual-triggers-modal/sub-components/rule/main.js
/* main.tsx generated by @compiled/babel-plugin v0.36.1 */
import "./main.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { useAnalyticsEvents } from '@atlaskit/analytics-next';
import { ButtonItem } from '@atlaskit/menu';
import { Box, Stack } from '@atlaskit/primitives/compiled';
import Spinner from '@atlaskit/spinner';
import { useAutomationMenu } from '../../menu-context';
const styles = {
  ruleButtonStyle: "_2rkoglpi _1h6dmuej _1dqonqa1 _189ee4h9",
  selectedRuleButtonStyle: "_2rkoglpi _1h6d15qp _1dqonqa1 _189e1bk5",
  ruleNameStyle: "_4t3i1f4h"
};
export const AutomationModalRule = ({
  rule,
  selectedRule,
  setSelectedRule
}) => {
  const {
    createAnalyticsEvent
  } = useAnalyticsEvents();
  const {
    analyticsSource,
    invokingRuleId
  } = useAutomationMenu();
  const isSelectedRule = (selectedRule === null || selectedRule === void 0 ? void 0 : selectedRule.id) === rule.id;
  const isExecuting = rule.id === invokingRuleId;
  const ruleOnClick = () => {
    createAnalyticsEvent({
      type: 'sendUIEvent',
      data: {
        action: 'clicked',
        actionSubject: 'button',
        actionSubjectId: 'selectManualTriggerAutomation',
        source: analyticsSource,
        attributes: {
          ruleId: rule.id.toString()
        }
      }
    }).fire();
    setSelectedRule(rule);
  };
  return /*#__PURE__*/React.createElement(Box, {
    xcss: isSelectedRule ? styles.selectedRuleButtonStyle : styles.ruleButtonStyle,
    backgroundColor: isSelectedRule ? 'color.background.accent.blue.subtlest' : undefined
  }, /*#__PURE__*/React.createElement(ButtonItem, {
    isDisabled: !!invokingRuleId,
    iconBefore: isExecuting ? /*#__PURE__*/React.createElement(Spinner, {
      size: "small"
    }) : null,
    onClick: ruleOnClick
  }, /*#__PURE__*/React.createElement(Stack, {
    xcss: styles.ruleNameStyle,
    alignBlock: "center"
  }, rule.name)));
};
