// @atlaskit/side-navigation/dist/cjs/components/Header/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Container = void 0;
require("./index.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireWildcard(require("react"));
var _colors = require("@atlaskit/theme/colors");
var _Item = require("../Item");
var _excluded = ["children", "data-testid"],
  _excluded2 = ["children"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var styles = {
  header: "_vrjt140y _1cmd10s3 _1cg71wug _11kjqtfy _jl2n73ad _1et6qtfy _1v6j10s3"
};

/**
 * __Container__
 *
 * A container for Header and Footer that safely handles props to the child component
 */
var Container = exports.Container = function Container(_ref) {
  var children = _ref.children,
    testId = _ref['data-testid'],
    props = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  // https://stackoverflow.com/a/39333479
  var safeProps = function (_ref2) {
    var className = _ref2.className,
      onClick = _ref2.onClick,
      onMouseDown = _ref2.onMouseDown,
      onDragStart = _ref2.onDragStart,
      draggable = _ref2.draggable,
      ref = _ref2.ref,
      tabIndex = _ref2.tabIndex,
      disabled = _ref2.disabled;
    return {
      className: className,
      onClick: onClick,
      onMouseDown: onMouseDown,
      onDragStart: onDragStart,
      draggable: draggable,
      ref: ref,
      tabIndex: tabIndex,
      disabled: disabled
    };
  }(props);
  return (
    /*#__PURE__*/
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
    _react.default.createElement("div", (0, _extends2.default)({
      "data-testid": testId,
      style: {
        position: 'relative'
      }
    }, safeProps), children)
  );
};
/**
 * __Header__
 *
 * - [Examples](https://atlassian.design/components/side-navigation/examples#header-and-footer)
 * - [Code](https://atlassian.design/components/side-navigation/code)
 */
var Header = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  return /*#__PURE__*/_react.default.createElement(_Item.CustomItem, (0, _extends2.default)({}, props, {
    ref: ref,
    component: props.component || Container,
    // eslint-disable-next-line @repo/internal/react/no-unsafe-overrides
    overrides: {
      Title: {
        // eslint-disable-next-line @atlaskit/design-system/use-heading
        render: function render(_, _ref3) {
          var children = _ref3.children,
            props = (0, _objectWithoutProperties2.default)(_ref3, _excluded2);
          return /*#__PURE__*/_react.default.createElement("h2", props, children);
        }
      }
    },
    className: (0, _runtime.ax)([styles.header])
  }));
});
var _default = exports.default = Header;
