// react-loosely-lazy/dist/esm/lazy/components/server.js
import { getAssetUrlsFromId } from '@react-loosely-lazy/manifest';
import React, { useContext } from 'react';
import { getConfig } from '../../config';
import { PHASE } from '../../constants';
import { LazySuspenseContext } from '../../suspense';
import { getExport } from '../../utils';
import { createLoaderError } from '../errors';
function load(moduleId, loader) {
  try {
    return getExport(loader());
  } catch (err) {
    throw createLoaderError(err);
  }
}
export function createComponentServer({
  dataLazyId,
  defer,
  loader,
  moduleId,
  ssr
}) {
  return props => {
    var _getAssetUrlsFromId;
    const Resolved = ssr ? load(moduleId, loader) : null;
    const {
      fallback
    } = useContext(LazySuspenseContext);
    const {
      crossOrigin,
      manifest
    } = getConfig();
    return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("input", {
      type: "hidden",
      "data-lazy-begin": dataLazyId
    }), defer !== PHASE.LAZY && ((_getAssetUrlsFromId = getAssetUrlsFromId(manifest, moduleId)) == null ? void 0 : _getAssetUrlsFromId.map(url => /*#__PURE__*/React.createElement("link", {
      key: url,
      rel: defer === PHASE.PAINT ? 'preload' : 'prefetch',
      href: url,
      crossOrigin: crossOrigin,
      as: "script"
    }))), Resolved ? /*#__PURE__*/React.createElement(Resolved, props) : fallback, /*#__PURE__*/React.createElement("input", {
      type: "hidden",
      "data-lazy-end": dataLazyId
    }));
  };
}
