// react-loosely-lazy/dist/cjs/phase/context.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useLazyPhase = exports.setCurrent = exports.LazyPhaseContext = void 0;
var _react = require("react");
var _constants = require("../constants");
var _listeners = require("./listeners");
var _utils = require("./utils");
let CURRENT_PHASE = _constants.PHASE.PAINT;
const setCurrent = phase => {
  CURRENT_PHASE = phase;
  _listeners.LISTENERS.slice(0).forEach(listener => listener(phase));
};
exports.setCurrent = setCurrent;
const LazyPhaseContext = exports.LazyPhaseContext = /*#__PURE__*/(0, _react.createContext)({
  subscribe: (0, _utils.createSubscribe)(_listeners.LISTENERS),
  currentValue: () => CURRENT_PHASE,
  api: {
    startNextPhase: () => {
      setCurrent(_constants.PHASE.AFTER_PAINT);
    },
    resetPhase: () => setCurrent(_constants.PHASE.PAINT)
  }
});
const useLazyPhase = () => {
  const v = (0, _react.useContext)(LazyPhaseContext);
  return v.api;
};
exports.useLazyPhase = useLazyPhase;
