// @atlaskit/editor-common/dist/es2019/types/annotation.js
import EventEmitter from 'events';
import { AnnotationTypes, } from '@atlaskit/adf-schema';
export var AnnotationUpdateEvent;
(function (AnnotationUpdateEvent) {
    AnnotationUpdateEvent["SET_ANNOTATION_FOCUS"] = "SET_ANNOTATION_FOCUS";
    AnnotationUpdateEvent["SET_ANNOTATION_STATE"] = "SET_ANNOTATION_STATE";
    AnnotationUpdateEvent["REMOVE_ANNOTATION_FOCUS"] = "REMOVE_ANNOTATION_FOCUS";
})(AnnotationUpdateEvent || (AnnotationUpdateEvent = {}));
export class AnnotationUpdateEmitter {
    constructor() {
        this.emitter = new EventEmitter();
    }
    emit(event, ...payload) {
        if (typeof payload === 'undefined') {
            return this.emitter.emit(event);
        }
        else {
            return this.emitter.emit(event, ...payload);
        }
    }
    on(event, listener) {
        return this.emitter.on(event, listener);
    }
    off(event, listener) {
        return this.emitter.off(event, listener);
    }
    listeners(event) {
        return this.emitter.listeners(event);
    }
}
