// @atlaskit/jql-ast/dist/cjs/jast-listeners/jql-insights-listener.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.splitTextByNewLine = exports.computeJqlInsights = exports.collapsedFieldType = void 0;
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _api = require("../api");
var _constants = require("../constants");
/**
 * Splits the provided text by new line characters.
 */
var splitTextByNewLine = exports.splitTextByNewLine = function splitTextByNewLine(text) {
  return text.split(/(?:\r\n?|\n)/);
};

/**
 * Finds the field type in the collapsed field syntax. eg. "field[xyz]" -> "[xyz]"
 */
var collapsedFieldType = exports.collapsedFieldType = function collapsedFieldType(text) {
  var match = _constants.COLLAPSED_CUSTOM_FIELD_PATTERN_NO_QUOTES.exec(text.toLowerCase());
  if (match) {
    return "[".concat(match[1], "]");
  }
};

/**
 * Analytics computed for a JQL query.
 */
var JastAnalyticsListener = /*#__PURE__*/(0, _createClass2.default)(function JastAnalyticsListener(jast) {
  var _this = this;
  (0, _classCallCheck2.default)(this, JastAnalyticsListener);
  // Default attributes
  (0, _defineProperty2.default)(this, "attributes", {
    jqlFieldValueCount: {
      issueType: 0,
      project: 0,
      assignee: 0,
      reporter: 0,
      priority: 0,
      status: 0,
      resolution: 0,
      team: 0
    },
    jqlFieldIsUsed: {
      summary: false,
      due: false,
      resolutionDate: false,
      created: false,
      lastviewed: false,
      updated: false,
      team: false
    },
    jqlUsedFields: [],
    jqlUsedFieldsCount: 0,
    jqlUsedFieldsOrderBy: [],
    jqlLineCount: 0,
    jqlErrorCount: 0,
    jqlClauseCount: {
      orderBy: 0,
      leaf: 0,
      and: 0,
      not: 0,
      or: 0
    },
    jqlMaxCompoundClauseDepth: 0
  });
  (0, _defineProperty2.default)(this, "usedFields", new Set());
  (0, _defineProperty2.default)(this, "usedFieldsOrderBy", new Set());
  (0, _defineProperty2.default)(this, "fieldValueCount", 0);
  // The current number of nested compound clauses for an AST node
  (0, _defineProperty2.default)(this, "compoundClauseDepth", 0);
  (0, _defineProperty2.default)(this, "getFieldName", function (field) {
    return collapsedFieldType(field.value) || field.value.toLowerCase();
  });
  (0, _defineProperty2.default)(this, "incrementFieldValueCount", function () {
    _this.fieldValueCount += 1;
  });
  (0, _defineProperty2.default)(this, "exitQuery", function () {
    // Track all fields used in the query using the term 'other' for non-privacy safe fields
    var privacySafeFields = new Set();
    _this.usedFields.forEach(function (field) {
      return privacySafeFields.add(_constants.PRIVACY_SAFE_FIELDS.includes(field) ? field : 'other');
    });
    _this.attributes.jqlUsedFields = Array.from(privacySafeFields).sort();
    _this.attributes.jqlUsedFieldsCount = _this.usedFields.size;
    _this.attributes.jqlUsedFieldsOrderBy = Array.from(_this.usedFieldsOrderBy);
  });
  (0, _defineProperty2.default)(this, "enterCompoundClause", function (compoundClause) {
    if (compoundClause.operator.value === _constants.COMPOUND_OPERATOR_AND) {
      _this.attributes.jqlClauseCount.and += 1;
    } else if (compoundClause.operator.value === _constants.COMPOUND_OPERATOR_OR) {
      _this.attributes.jqlClauseCount.or += 1;
    }
    _this.compoundClauseDepth += 1;
    if (_this.compoundClauseDepth > _this.attributes.jqlMaxCompoundClauseDepth) {
      _this.attributes.jqlMaxCompoundClauseDepth = _this.compoundClauseDepth;
    }
  });
  (0, _defineProperty2.default)(this, "exitCompoundClause", function () {
    _this.compoundClauseDepth -= 1;
  });
  (0, _defineProperty2.default)(this, "enterTerminalClause", function (terminalClause) {
    var field = _this.getFieldName(terminalClause.field);
    switch (field) {
      case _constants.SUMMARY_FIELD:
        _this.attributes.jqlFieldIsUsed.summary = true;
        break;
      case _constants.DUE_DATE_FIELD:
        _this.attributes.jqlFieldIsUsed.due = true;
        break;
      case _constants.RESOLUTION_DATE_FIELD:
        _this.attributes.jqlFieldIsUsed.resolutionDate = true;
        break;
      case _constants.CREATED_FIELD:
        _this.attributes.jqlFieldIsUsed.created = true;
        break;
      case _constants.LAST_VIEWED_FIELD:
        _this.attributes.jqlFieldIsUsed.lastviewed = true;
        break;
      case _constants.UPDATED_FIELD:
        _this.attributes.jqlFieldIsUsed.updated = true;
        break;
      case _constants.TEAM_CUSTOM_FIELD_TYPE:
        _this.attributes.jqlFieldIsUsed.team = true;
        break;
    }
    _this.attributes.jqlClauseCount.leaf += 1;

    // Track all fields used in the query
    _this.usedFields.add(field);

    // Reset our count of RHS values for the field
    _this.fieldValueCount = 0;
  });
  (0, _defineProperty2.default)(this, "exitTerminalClause", function (terminalClause) {
    switch (_this.getFieldName(terminalClause.field)) {
      case _constants.ISSUE_TYPE_FIELD:
      case _constants.TYPE_FIELD:
        _this.attributes.jqlFieldValueCount.issueType += _this.fieldValueCount;
        break;
      case _constants.PROJECT_FIELD:
        _this.attributes.jqlFieldValueCount.project += _this.fieldValueCount;
        break;
      case _constants.ASSIGNEE_FIELD:
        _this.attributes.jqlFieldValueCount.assignee += _this.fieldValueCount;
        break;
      case _constants.REPORTER_FIELD:
        _this.attributes.jqlFieldValueCount.reporter += _this.fieldValueCount;
        break;
      case _constants.PRIORITY_FIELD:
        _this.attributes.jqlFieldValueCount.priority += _this.fieldValueCount;
        break;
      case _constants.STATUS_FIELD:
        _this.attributes.jqlFieldValueCount.status += _this.fieldValueCount;
        break;
      case _constants.RESOLUTION_FIELD:
        _this.attributes.jqlFieldValueCount.resolution += _this.fieldValueCount;
        break;
      case _constants.TEAM_CUSTOM_FIELD_TYPE:
        _this.attributes.jqlFieldValueCount.team += _this.fieldValueCount;
        break;
      default:
        break;
    }
  });
  (0, _defineProperty2.default)(this, "enterNotClause", function () {
    _this.attributes.jqlClauseCount.not += 1;
  });
  (0, _defineProperty2.default)(this, "enterFunctionOperand", this.incrementFieldValueCount);
  (0, _defineProperty2.default)(this, "enterKeywordOperand", this.incrementFieldValueCount);
  (0, _defineProperty2.default)(this, "enterValueOperand", this.incrementFieldValueCount);
  (0, _defineProperty2.default)(this, "enterOrderByField", function (orderByField) {
    _this.usedFieldsOrderBy.add(_this.getFieldName(orderByField.field));
    _this.attributes.jqlClauseCount.orderBy += 1;
  });
  this.attributes.jqlLineCount = splitTextByNewLine(jast.represents).length;
  this.attributes.jqlErrorCount = jast.errors.length;
});
var computeJqlInsights = exports.computeJqlInsights = function computeJqlInsights(jast) {
  var listener = new JastAnalyticsListener(jast);
  (0, _api.walkAST)(listener, jast);
  return listener.attributes;
};
