// @atlaskit/editor-common/dist/cjs/quick-insert/assets/layout.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = IconLayout;
var _react = _interopRequireDefault(require("react"));
var _useIconThemed2 = require("../use-icon-themed");
function IconLayout() {
  var _useIconThemed = (0, _useIconThemed2.useIconThemed)(),
    iconThemed = _useIconThemed.iconThemed;
  return /*#__PURE__*/_react.default.createElement("svg", {
    focusable: "false",
    "aria-hidden": true,
    width: 40,
    height: 40,
    viewBox: "0 0 40 40"
  }, /*#__PURE__*/_react.default.createElement("g", {
    fill: "none",
    fillRule: "evenodd"
  }, /*#__PURE__*/_react.default.createElement("path", {
    fill: iconThemed({
      light: '#FFF',
      dark: '#161A1D'
    }),
    d: "M0 0h40v40H0z"
  }), /*#__PURE__*/_react.default.createElement("rect", {
    fill: iconThemed({
      light: '#A5ADBA',
      dark: '#738496'
    }),
    x: 6,
    y: 6,
    width: 28,
    height: 1,
    rx: 0.5
  }), /*#__PURE__*/_react.default.createElement("rect", {
    fill: iconThemed({
      light: '#A5ADBA',
      dark: '#738496'
    }),
    x: 6,
    y: 10,
    width: 28,
    height: 1,
    rx: 0.5
  }), /*#__PURE__*/_react.default.createElement("rect", {
    fill: iconThemed({
      light: '#A5ADBA',
      dark: '#738496'
    }),
    x: 6,
    y: 29,
    width: 28,
    height: 1,
    rx: 0.5
  }), /*#__PURE__*/_react.default.createElement("rect", {
    fill: iconThemed({
      light: '#A5ADBA',
      dark: '#738496'
    }),
    x: 6,
    y: 33,
    width: 16,
    height: 1,
    rx: 0.5
  }), /*#__PURE__*/_react.default.createElement("rect", {
    stroke: iconThemed({
      light: '#4C9AFF',
      dark: '#0C66E4'
    }),
    strokeWidth: 0.5,
    fill: iconThemed({
      light: '#DEEBFF',
      dark: '#09326C'
    }),
    x: 6.25,
    y: 14.25,
    width: 12.5,
    height: 11.5,
    rx: 1
  }), /*#__PURE__*/_react.default.createElement("rect", {
    stroke: iconThemed({
      light: '#4C9AFF',
      dark: '#0C66E4'
    }),
    strokeWidth: 0.5,
    fill: iconThemed({
      light: '#DEEBFF',
      dark: '#09326C'
    }),
    x: 21.25,
    y: 14.25,
    width: 12.5,
    height: 11.5,
    rx: 1
  })));
}
