// @atlaskit/editor-common/dist/cjs/quick-insert/assets/list.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = IconList;
var _react = _interopRequireDefault(require("react"));
var _useIconThemed2 = require("../use-icon-themed");
function IconList() {
  var _useIconThemed = (0, _useIconThemed2.useIconThemed)(),
    iconThemed = _useIconThemed.iconThemed;
  var dotColour = iconThemed({
    light: '#6C798F',
    dark: '#5A6977'
  });
  var lineColour = iconThemed({
    light: '#C1C7D0',
    dark: '#454F59'
  });
  return /*#__PURE__*/_react.default.createElement("svg", {
    focusable: "false",
    "aria-hidden": true,
    width: 40,
    height: 40,
    viewBox: "0 0 40 40"
  }, /*#__PURE__*/_react.default.createElement("g", {
    fill: "none",
    fillRule: "evenodd"
  }, /*#__PURE__*/_react.default.createElement("path", {
    fill: iconThemed({
      light: '#FFF',
      dark: '#161A1D'
    }),
    d: "M0 0h40v40H0z"
  }), /*#__PURE__*/_react.default.createElement("g", {
    transform: "translate(8 10)"
  }, /*#__PURE__*/_react.default.createElement("rect", {
    fill: dotColour,
    width: 4,
    height: 4,
    rx: 2
  }), /*#__PURE__*/_react.default.createElement("rect", {
    fill: lineColour,
    x: 7,
    y: 9,
    width: 17,
    height: 2,
    rx: 1
  }), /*#__PURE__*/_react.default.createElement("rect", {
    fill: lineColour,
    x: 7,
    y: 1,
    width: 17,
    height: 2,
    rx: 1
  }), /*#__PURE__*/_react.default.createElement("rect", {
    fill: lineColour,
    x: 7,
    y: 17,
    width: 17,
    height: 2,
    rx: 1
  }), /*#__PURE__*/_react.default.createElement("rect", {
    fill: dotColour,
    y: 8,
    width: 4,
    height: 4,
    rx: 2
  }), /*#__PURE__*/_react.default.createElement("rect", {
    fill: dotColour,
    y: 16,
    width: 4,
    height: 4,
    rx: 2
  }))));
}
