// @atlaskit/link-datasource/dist/es2019/ui/issue-like-table/empty-state/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./index.compiled.css";
import { forwardRef } from 'react';
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
/* eslint-disable @atlaskit/design-system/use-tokens-typography */

import { Skeleton } from '@atlaskit/linking-common';
import { N40 } from '@atlaskit/theme/colors';
const SkeletonComponent = ({
  width,
  itemName
}) => /*#__PURE__*/React.createElement(Skeleton, {
  borderRadius: "var(--ds-border-radius-100, 8px)",
  testId: `${itemName}-empty-state-skeleton`,
  height: 14,
  width: width
});
const tableBodyStyles = null;
const tableStyles = null;
const padding = `${"var(--ds-space-100, 8px)"} ${"var(--ds-space-100, 8px)"}`;
const cellStyles = null;
const baseColumns = [{
  key: 'type',
  width: 35
}, {
  key: 'issue',
  width: 50
}, {
  key: 'summary',
  width: 100
}, {
  key: 'assignee',
  width: 70
}, {
  key: 'priority',
  width: 60
}, {
  key: 'status',
  width: 60
}, {
  key: 'resolution',
  width: 55
}, {
  key: 'created',
  width: 50
}, {
  key: 'due',
  width: 50
}];
const longColumnWidths = [141, 208, 186, 212, 212, 151, 212, 182, 180, 163, 172, 211, 145, 190];
const shortColumnWidths = [75, 54, 66, 73, 52, 73, 55, 80, 67, 76, 58, 65, 56, 76];
const renderItem = ({
  key,
  width
}, {
  longWidth,
  shortWidth
}) => {
  switch (key) {
    case 'status':
      return /*#__PURE__*/React.createElement(SkeletonComponent, {
        width: shortWidth,
        itemName: key
      });
    case 'summary':
      return /*#__PURE__*/React.createElement(SkeletonComponent, {
        width: longWidth,
        itemName: key
      });
    default:
      return /*#__PURE__*/React.createElement(SkeletonComponent, {
        width: width,
        itemName: key
      });
  }
};
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled
const TableHeading = forwardRef(({
  as: C = "th",
  style: __cmpls,
  ...__cmplp
}, __cmplr) => {
  return /*#__PURE__*/React.createElement(C, _extends({}, __cmplp, {
    style: __cmpls,
    ref: __cmplr,
    className: ax(["_19itidpf _aks5idpf _1u3bidpf _1s7zl0fh _1o9ml0fh _1oupl0fh _n7zl1on0 _1fqk1on0 _1vvv1on0 _kqswh2mm _rc4eh2mm _q46wh2mm _vchhusvi _1wwnusvi _1gbausvi _vwz41tcg _s99e1tcg _3mvc1tcg _ca0qv77o _7cz2v77o _pcwyv77o _u5f31b66 _1pmc1b66 _11ep1b66 _n3tdv77o _6dd7v77o _191qv77o _19bv1b66 _9iya1b66 _15hu1b66 _4t3iqagn _qccgqagn _xo89qagn _s7n4yfq0 _ntwryfq0 _3eneyfq0 _bfhkchd4 _i7olchd4 _1oe6chd4 _12pn15vq _198v15vq _lura1osq _11681tcg _ab8av77o _11j11b66 _1xanv77o _8lvj1b66 _iscccj1k _1nd0cs5v _1t0u1e54 _1eqz4jg8 _95ex1hna _1hkzidpf _1r3pidpf _10i2idpf _1ls2idpf _ocbzidpf _3tfiidpf _1lhc1b66 _n2z81b66 _3ixo1b66", __cmplp.className])
  }));
});
if (process.env.NODE_ENV !== 'production') {
  TableHeading.displayName = 'TableHeading';
}
const EmptyState = ({
  isCompact,
  testId
}) => {
  const columnsToRender = isCompact ? baseColumns.slice(0, 6) : baseColumns;
  // if it is compact (non-modal), there is room for 14 rows
  // if it is modal (not compact), there is only room for 10 rows
  const rowsNumber = isCompact ? 14 : 10;
  const rows = new Array(rowsNumber).fill(null).map((_, index) => ({
    id: index,
    longWidth: longColumnWidths[index],
    shortWidth: shortColumnWidths[index]
  }));
  return /*#__PURE__*/React.createElement("div", {
    style: {
      // the IssueLikeDataTableView wraps the table in a container with the styling below while modal doesn't
      // this maxHeight comes from scrollableContainerHeight
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
      maxHeight: 590,
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
      padding: "var(--ds-space-0, 0px)",
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
      boxSizing: 'border-box'
    }
  }, /*#__PURE__*/React.createElement("table", {
    "data-testid": testId,
    className: ax(["_bfhkchd4"])
  }, /*#__PURE__*/React.createElement("thead", {
    style: {
      borderBottom: 0
    }
  }, /*#__PURE__*/React.createElement("tr", null, columnsToRender.map(({
    key,
    width
  }) =>
  /*#__PURE__*/
  // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
  React.createElement(TableHeading, {
    key: key,
    style: {
      width,
      padding
    }
  }, /*#__PURE__*/React.createElement(Skeleton, {
    appearance: "darkGray",
    borderRadius: 8,
    testId: "empty-state-skeleton",
    height: 12,
    width: width
  }))))), /*#__PURE__*/React.createElement("tbody", {
    className: ax(["_n7zlidpf"])
  }, rows.map(row => /*#__PURE__*/React.createElement("tr", {
    key: row.id
  }, columnsToRender.map(column => /*#__PURE__*/React.createElement("td", {
    key: column.key,
    className: ax(["_1yt41uce _1s7zl0fh _n7zll0fh _1ls2idpf _1o59u2gc _1lhcu2gc"])
  }, renderItem(column, row))))))));
};
export default EmptyState;
