// @atlaskit/media-card/dist/es2019/card/ui/actionsBar/cardActions/cardActionButton-compiled.js
/* cardActionButton-compiled.tsx generated by @compiled/babel-plugin v0.36.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./cardActionButton-compiled.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { forwardRef } from 'react';
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

import { N500 } from '@atlaskit/theme/colors';
const variantStyleMap = {
  filled: "_bfhk1pja _irr3tl7t",
  default: "_irr3x5l4"
};
const cardActionButtonStyles = null;
export const CardActionButton = /*#__PURE__*/forwardRef((props, ref) => {
  var _props$variant;
  return /*#__PURE__*/React.createElement("button", _extends({}, props, {
    id: "cardActionButton",
    "data-testid": "media-card-primary-action",
    "aria-label": props.label,
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
    style: {
      ...props.style,
      cursor: props.disabled ? 'not-allowed' : 'default'
    },
    onClick: props.onClick,
    onMouseDown: props.onMouseDown,
    ref: ref,
    className: ax(["_19itglyw _2rko1sit _r06hglyw _1e0c1txw _4cvr1h6o _1bah1h6o _1bsb1ou6 _4t3i1ou6 _syazilmm _d0altlke", variantStyleMap[(_props$variant = props.variant) !== null && _props$variant !== void 0 ? _props$variant : 'default']])
  }), props.children);
});
