// @atlaskit/media-card/dist/es2019/card/ui/actionsBar/cardActions/styles.js
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import { css } from '@emotion/react';
import { N500, N0 } from '@atlaskit/theme/colors';
import { borderRadius, size, center } from '@atlaskit/media-ui';
import { rgba } from '../../styles';
export let CardActionIconButtonVariant = /*#__PURE__*/function (CardActionIconButtonVariant) {
  CardActionIconButtonVariant["default"] = "default";
  CardActionIconButtonVariant["filled"] = "filled";
  return CardActionIconButtonVariant;
}({});
const getVariantStyles = variant => {
  return variant === 'filled' ? `
    background-color: ${`var(--ds-surface-overlay, ${rgba(N0, 0.8)})`};

    &:hover {
      background-color: ${`var(--ds-surface-overlay-hovered, ${rgba(N0, 0.6)})`}
    }
  ` : `
    &:hover {
      background-color: ${"var(--ds-background-neutral-subtle-hovered, rgba(9, 30, 66, 0.06))"};
    }
  `;
};
export const cardActionButtonStyles = ({
  variant
}) => css({
  appearance: 'none',
  border: 'none'
},
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
center,
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
borderRadius,
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
size(26), {
  color: `var(--ds-icon, ${N500})`,
  '&:hover': {
    cursor: 'pointer'
  }
},
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
getVariantStyles(variant));
