// @atlaskit/link-datasource/dist/cjs/ui/common/modal/popup-select/control.js
/* control.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.CustomControlOld = exports.CustomControl = void 0;
require("./control.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireDefault(require("react"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _compiled = require("@atlaskit/primitives/compiled");
var _select = require("@atlaskit/select");
var _colors = require("@atlaskit/theme/colors");
var _withFeatureFlaggedComponent = require("../../../../common/utils/withFeatureFlaggedComponent");
var _excluded = ["children"],
  _excluded2 = ["children"];
var popupCustomControlStylesOld = function popupCustomControlStylesOld() {
  return {
    display: 'flex',
    padding: "var(--ds-space-050, 4px)",
    border: 'none',
    borderBottom: "solid 1px ".concat("var(--ds-border, ".concat(_colors.N40, ")")),
    minHeight: 'auto'
  };
};
var popupCustomControlStyles = function popupCustomControlStyles() {
  return {
    display: 'flex',
    borderRadius: '3px',
    border: "solid 1px ".concat("var(--ds-border-input, #8C8F97)")
  };
};
var popupCustomControlVisualRefreshStyles = {
  container: "_ca0q1b66 _n3td1b66 _19bvh9hh _u5f3h9hh"
};
var CustomControlOld = exports.CustomControlOld = function CustomControlOld(_ref) {
  var children = _ref.children,
    innerProps = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  return /*#__PURE__*/_react.default.createElement(_select.components.Control, (0, _extends2.default)({}, innerProps, {
    getStyles: popupCustomControlStylesOld
  }), children);
};
var CustomControl = exports.CustomControl = function CustomControl(_ref2) {
  var children = _ref2.children,
    innerProps = (0, _objectWithoutProperties2.default)(_ref2, _excluded2);
  return /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    xcss: popupCustomControlVisualRefreshStyles.container
  }, /*#__PURE__*/_react.default.createElement(_select.components.Control, (0, _extends2.default)({}, innerProps, {
    getStyles: popupCustomControlStyles
  }), children));
};
var _default = exports.default = (0, _withFeatureFlaggedComponent.withFeatureFlaggedComponent)(CustomControlOld, CustomControl, function () {
  return (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv');
});
