// @atlaskit/link-datasource/dist/cjs/ui/common/modal/popup-select/footer.js
/* footer.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./footer.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _react2 = require("@compiled/react");
var _reactIntlNext = require("react-intl-next");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _compiled = require("@atlaskit/primitives/compiled");
var _colors = require("@atlaskit/theme/colors");
var _messages = require("./messages");
var styles = {
  footerContainerStylesOld: "_ca0q1b66 _n3td1b66 _uwhke4h9 _15a5nqa1 _1i53b5v6",
  footerContainerStyles: "_ca0qutpp _n3tdutpp _19bvpxbi _u5f3pxbi _uwhke4h9 _15a5nqa1 _1i53b5v6",
  footerPaginationInfoStylesOld: "_1mouu2gc _195gutpp _syaz131l",
  footerPaginationInfoStyles: "_syaz131l"
};
var PopupFooter = function PopupFooter(_ref) {
  var currentDisplayCount = _ref.currentDisplayCount,
    totalCount = _ref.totalCount,
    filterName = _ref.filterName;
  return /*#__PURE__*/_react.default.createElement(_compiled.Flex, {
    testId: "".concat(filterName, "--footer"),
    direction: "row",
    alignItems: "center",
    justifyContent: "end",
    xcss: (0, _react2.cx)((0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv') ? styles.footerContainerStyles : styles.footerContainerStylesOld)
  }, /*#__PURE__*/_react.default.createElement(_compiled.Inline, {
    xcss: (0, _react2.cx)((0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv') ? styles.footerPaginationInfoStyles : styles.footerPaginationInfoStylesOld)
  }, /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, (0, _extends2.default)({}, _messages.asyncPopupSelectMessages.paginationDetails, {
    values: {
      currentDisplayCount: currentDisplayCount,
      totalCount: totalCount
    }
  }))));
};
var _default = exports.default = PopupFooter;
