// @atlaskit/link-datasource/dist/cjs/ui/common/modal/popup-select/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FilterPopupSelect = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _isEqual = _interopRequireDefault(require("lodash/isEqual"));
var _reactIntlNext = require("react-intl-next");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _select = require("@atlaskit/select");
var _analytics = require("../../../../analytics");
var _checkboxOptionVisualRefreshSllv = require("./checkbox-option-visual-refresh-sllv");
var _control = _interopRequireDefault(require("./control"));
var _dropdownIndicator = _interopRequireDefault(require("./dropdownIndicator"));
var _footer = _interopRequireDefault(require("./footer"));
var _formatOptionLabel = _interopRequireDefault(require("./formatOptionLabel"));
var _menuList = _interopRequireDefault(require("./menu-list"));
var _messages = require("./messages");
var _trigger = _interopRequireDefault(require("./trigger"));
var _excluded = ["isOpen"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
// Needed to disable filtering from react-select
var noFilterOptions = function noFilterOptions() {
  return true;
};
var FilterPopupSelect = exports.FilterPopupSelect = function FilterPopupSelect(_ref) {
  var filterName = _ref.filterName,
    _ref$totalCount = _ref.totalCount,
    totalCount = _ref$totalCount === void 0 ? 0 : _ref$totalCount,
    status = _ref.status,
    buttonLabel = _ref.buttonLabel,
    searchPlaceholder = _ref.searchPlaceholder,
    _ref$showLoading = _ref.showLoading,
    showLoading = _ref$showLoading === void 0 ? false : _ref$showLoading,
    _ref$isDisabled = _ref.isDisabled,
    isDisabled = _ref$isDisabled === void 0 ? false : _ref$isDisabled,
    _ref$showHydrating = _ref.showHydrating,
    showHydrating = _ref$showHydrating === void 0 ? false : _ref$showHydrating,
    _ref$shouldShowFooter = _ref.shouldShowFooter,
    shouldShowFooter = _ref$shouldShowFooter === void 0 ? true : _ref$shouldShowFooter,
    menuListProps = _ref.menuListProps,
    options = _ref.options,
    selectedOptions = _ref.selectedOptions,
    onSelectionChange = _ref.onSelectionChange,
    onInputChange = _ref.onInputChange,
    onMenuClose = _ref.onMenuClose,
    onMenuOpen = _ref.onMenuOpen;
  var _useIntl = (0, _reactIntlNext.useIntl)(),
    formatMessage = _useIntl.formatMessage;
  var _useDatasourceAnalyti = (0, _analytics.useDatasourceAnalyticsEvents)(),
    fireEvent = _useDatasourceAnalyti.fireEvent;
  var _useState = (0, _react.useState)(''),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    searchTerm = _useState2[0],
    setSearchTerm = _useState2[1];
  var _useState3 = (0, _react.useState)([]),
    _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
    sortedOptions = _useState4[0],
    setSortedOptions = _useState4[1];
  var sortPaginatedResults = (0, _react.useRef)(false); // this is to track pagination for sorting purpose

  var sortOptionsOnPopupOpen = (0, _react.useCallback)(function () {
    if (selectedOptions.length === 0) {
      return setSortedOptions(options);
    }
    var nonSelectedOptions = options.filter(function (option) {
      return !selectedOptions.find(function (selectedOption) {
        return selectedOption.value === option.value;
      });
    });
    var newOptions = [].concat((0, _toConsumableArray2.default)(selectedOptions), (0, _toConsumableArray2.default)(nonSelectedOptions));
    if (!(0, _isEqual.default)(newOptions, sortedOptions)) {
      setSortedOptions(newOptions);
    }
  }, [selectedOptions, options, sortedOptions]);
  var sortOptionsOnResolve = (0, _react.useCallback)(function () {
    // when the user is searching, we want the search result to be displayed as it is, and the select component will take care of marking the selected items
    if (searchTerm) {
      sortPaginatedResults.current = false; // set to false to indicate pagination resolve action is completed from the sorting perspective
      return setSortedOptions(options);
    }

    // sortedOptions is empty initially, this will take care of setting the initial value and bring the selected items to the top
    if (sortedOptions.length === 0) {
      return sortOptionsOnPopupOpen();
    }

    // this block handles the pagination, where on pagination, we will just append newOptions to the current list
    if (sortPaginatedResults.current) {
      var newOptions = options.filter(function (option) {
        return !sortedOptions.find(function (sortedOption) {
          return sortedOption.value === option.value;
        });
      });
      if (newOptions.length > 0) {
        setSortedOptions([].concat((0, _toConsumableArray2.default)(sortedOptions), (0, _toConsumableArray2.default)(newOptions)));
      }
      sortPaginatedResults.current = false; // set to false to indicate pagination resolve action is completed from the sorting perspective
      return;
    }
    sortPaginatedResults.current = false; // set to false to indicate pagination resolve action is completed from the sorting perspective
    sortOptionsOnPopupOpen();
  }, [options, searchTerm, sortOptionsOnPopupOpen, sortedOptions]);
  var handleMenuOpen = (0, _react.useCallback)(function () {
    if (status === 'resolved') {
      sortOptionsOnPopupOpen();
    }
    fireEvent('ui.dropdown.opened.basicSearchDropdown', {
      filterName: filterName,
      selectionCount: selectedOptions.length
    });
    onMenuOpen === null || onMenuOpen === void 0 || onMenuOpen();
  }, [filterName, fireEvent, onMenuOpen, selectedOptions.length, sortOptionsOnPopupOpen, status]);
  var handleMenuClose = (0, _react.useCallback)(function () {
    /**
     * Clearing the search is to ensure that the sortOptionsOnPopupOpen logic does not mess up.
     * Without this, when the user opens, sortOptionsOnPopupOpen will inject the selected options to the list and the list count and values will be off
     */
    if (searchTerm) {
      setSearchTerm('');
      onInputChange('', {
        action: 'input-change',
        prevInputValue: searchTerm
      });
    }
    onMenuClose === null || onMenuClose === void 0 || onMenuClose();
    fireEvent('ui.dropdown.closed.basicSearchDropdown', {
      filterName: filterName,
      selectionCount: selectedOptions.length
    });
  }, [filterName, fireEvent, onInputChange, onMenuClose, searchTerm, selectedOptions.length]);
  var handleInputChange = (0, _react.useCallback)( /*#__PURE__*/function () {
    var _ref2 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(newSearchTerm, actionMeta) {
      return _regenerator.default.wrap(function _callee$(_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            if (actionMeta.action === 'input-change' && newSearchTerm !== searchTerm) {
              setSearchTerm(newSearchTerm);
              onInputChange(newSearchTerm, actionMeta);
            }
          case 1:
          case "end":
            return _context.stop();
        }
      }, _callee);
    }));
    return function (_x, _x2) {
      return _ref2.apply(this, arguments);
    };
  }(), [onInputChange, searchTerm]);
  (0, _react.useEffect)(function () {
    if (status === 'resolved') {
      sortOptionsOnResolve();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [status]); // we only want the sortOptionsOnResolve to run when there is a status change

  (0, _react.useEffect)(function () {
    if (status === 'loadingMore') {
      sortPaginatedResults.current = true;
    }
  }, [status]);
  return /*#__PURE__*/_react.default.createElement(_select.PopupSelect, {
    isMulti: true,
    maxMenuWidth: 300,
    minMenuWidth: 300,
    testId: "".concat(filterName, "-popup-select"),
    inputId: "".concat(filterName, "-popup-select--input")
    /*
         this threshold controls the display of the search control (input field for search)
         if this threshold is less than 0, when typing a search string that returns no results it will not remove the search control
         if this threshold is 0 or higher, it will remove the search control when there are no results, the user will be unable to clear the search to see more results
       */,
    searchThreshold: -1,
    inputValue: searchTerm,
    closeMenuOnSelect: false,
    shouldCloseMenuOnTab: false,
    hideSelectedOptions: false,
    isLoading: showLoading,
    placeholder: searchPlaceholder && (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv') ? searchPlaceholder : formatMessage(_messages.asyncPopupSelectMessages.selectPlaceholder)
    // @ts-ignore - https://product-fabric.atlassian.net/browse/DSP-21000
    ,
    menuListProps: menuListProps,
    components: {
      Option: (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv') ? _checkboxOptionVisualRefreshSllv.CheckboxOptionVisualRefreshSllv : _select.CheckboxOption,
      Control: _control.default,
      MenuList: _menuList.default,
      DropdownIndicator: _dropdownIndicator.default,
      LoadingIndicator: undefined // disables the three ... indicator in the searchbox when picker is loading
    }
    // eslint-disable-next-line @atlaskit/design-system/no-unsafe-style-overrides
    ,
    styles: (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv') ? {
      container: function container(base) {
        return _objectSpread(_objectSpread({}, base), {}, {
          paddingTop: "var(--ds-space-075, 6px)",
          paddingBottom: 0
        });
      },
      menuList: function menuList(base) {
        return _objectSpread(_objectSpread({}, base), {}, {
          paddingTop: "var(--ds-space-050, 4px)",
          paddingBottom: 0
        });
      }
    } : undefined,
    options: sortedOptions,
    value: selectedOptions,
    filterOption: noFilterOptions,
    formatOptionLabel: _formatOptionLabel.default,
    onChange: onSelectionChange,
    onInputChange: handleInputChange,
    onOpen: handleMenuOpen,
    onClose: handleMenuClose,
    target: function target(_ref3) {
      var isOpen = _ref3.isOpen,
        triggerProps = (0, _objectWithoutProperties2.default)(_ref3, _excluded);
      return /*#__PURE__*/_react.default.createElement(_trigger.default, (0, _extends2.default)({}, triggerProps, {
        label: buttonLabel,
        selectedOptions: selectedOptions,
        isSelected: isOpen,
        isDisabled: isDisabled,
        isLoading: showHydrating,
        testId: filterName
      }));
    },
    footer: shouldShowFooter && /*#__PURE__*/_react.default.createElement(_footer.default, {
      currentDisplayCount: options.length,
      totalCount: totalCount,
      filterName: filterName
    })
  });
};
