// @atlaskit/link-datasource/dist/esm/ui/assets-modal/modal/render-assets-content/initial-state-view/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import "./index.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { useIntl } from 'react-intl-next';
import Link from '@atlaskit/link';
import { fg } from '@atlaskit/platform-feature-flags';
import { fontFallback } from '@atlaskit/theme/typography';
import { RichIconSearch } from '../../../../../common/ui/rich-icon/search';
import { CrystalBallSVGOld } from './assets/crystal-ball-svg-old';
import { initialStateViewMessages } from './messages';
var initialStateViewContainerStyles = null;
var svgAndTextsWrapperStylesOld = null;
var svgAndTextsWrapperStyles = null;
var searchTitleStylesOld = null;
var searchTitleStyles = null;
var mainTextStyles = null;
var AQLSupportDocumentLink = 'https://support.atlassian.com/jira-service-management-cloud/docs/use-assets-query-language-aql/';
export var InitialStateView = function InitialStateView() {
  var _useIntl = useIntl(),
    formatMessage = _useIntl.formatMessage;
  return /*#__PURE__*/React.createElement("div", {
    "data-testid": "assets-aql-datasource-modal--initial-state-view",
    className: ax(["_1e0c1txw _1bah1h6o _4t3i1osq"])
  }, /*#__PURE__*/React.createElement("div", {
    className: ax([fg('platform-linking-visual-refresh-sllv') ? "_y3gn1h6o _1wpz1h6o _ca0qoahv _n3tdoahv _19bvoahv _u5f3oahv _p12f1kdj" : "_y3gn1h6o _1wpz1h6o"])
  }, fg('bandicoots-update-sllv-icons') ? /*#__PURE__*/React.createElement(RichIconSearch, {
    size: "xlarge",
    alt: formatMessage(fg('platform-linking-visual-refresh-sllv') ? initialStateViewMessages.searchTitle : initialStateViewMessages.searchTitleOld)
  }) : /*#__PURE__*/React.createElement(CrystalBallSVGOld, null), /*#__PURE__*/React.createElement("div", {
    className: ax([fg('platform-linking-visual-refresh-sllv') ? "_11c8lodh _syaz1fxt _ca0q1ejb _n3tdpxbi" : "_ca0qpxbi _n3tdu2gc _wtt4ykg7 _pfzw85h9 _zzcc1umn _x3fx1hgc _v2gl1yp8 _1y0bdipb _z7oy1jq3"])
  }, formatMessage(fg('platform-linking-visual-refresh-sllv') ? initialStateViewMessages.searchTitle : initialStateViewMessages.searchTitleOld)), /*#__PURE__*/React.createElement("div", {
    className: ax([fg('platform-linking-visual-refresh-sllv') && "_syaz1gjq"])
  }, formatMessage(initialStateViewMessages.searchDescription)), fg('platform-linking-visual-refresh-sllv') ? /*#__PURE__*/React.createElement(Link, {
    href: AQLSupportDocumentLink
  }, formatMessage(initialStateViewMessages.learnMoreLink)) : /*#__PURE__*/React.createElement("a", {
    href: AQLSupportDocumentLink,
    target: "_blank"
  }, formatMessage(initialStateViewMessages.learnMoreLinkOld))));
};
