// @atlaskit/media-viewer/dist/cjs/analytics/events/ui/failedPreviewDownloadButtonClicked.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createFailedPreviewDownloadButtonClickedEvent = void 0;
var _ = require("../..");
var _errors = require("../../../errors");
var createFailedPreviewDownloadButtonClickedEvent = exports.createFailedPreviewDownloadButtonClickedEvent = function createFailedPreviewDownloadButtonClickedEvent(fileState, error) {
  var _getFileAttributes = (0, _.getFileAttributes)(fileState),
    fileId = _getFileAttributes.fileId,
    fileMediatype = _getFileAttributes.fileMediatype,
    fileMimetype = _getFileAttributes.fileMimetype,
    fileSize = _getFileAttributes.fileSize;
  return {
    eventType: 'ui',
    action: 'clicked',
    actionSubject: 'button',
    actionSubjectId: 'failedPreviewDownloadButton',
    attributes: {
      failReason: (0, _errors.getPrimaryErrorReason)(error),
      fileAttributes: {
        fileId: fileId,
        fileMediatype: fileMediatype,
        fileMimetype: fileMimetype,
        fileSize: fileSize
      },
      fileProcessingStatus: fileState.status
    }
  };
};
