// @atlaskit/media-test-helpers/dist/esm/mocks/media-mock.js
import { __read, __spread } from "tslib";
import { Server } from 'kakapo';
import * as exenv from 'exenv';
import uuid from 'uuid/v4';
import { createApiRouter, createMediaPlaygroundRouter } from './routers';
import { createDatabase } from './database';
import { mapDataUriToBlob } from '../utils';
var MediaMock = /** @class */ (function () {
    function MediaMock(collections) {
        this.collections = collections;
        this.server = new Server();
        this.routers = [];
        this.dbs = [];
    }
    MediaMock.prototype.enable = function (config) {
        if (config === void 0) { config = {}; }
        var isSlowServer = config.isSlowServer;
        if (!exenv.canUseDOM) {
            return;
        }
        this.routers = [
            createMediaPlaygroundRouter(),
            createApiRouter(isSlowServer),
        ];
        this.dbs = [createDatabase(this.collections)];
        __spread(this.routers, this.dbs).forEach(this.server.use.bind(this.server));
    };
    MediaMock.prototype.disable = function () {
        __spread(this.routers, this.dbs).forEach(this.server.remove.bind(this.server));
        this.routers.forEach(function (router) { return router.reset(); });
        this.dbs.forEach(function (db) { return db.reset(); });
        this.routers = [];
        this.dbs = [];
    };
    return MediaMock;
}());
export { MediaMock };
export function generateFilesFromTestData(files) {
    return files.map(function (file) {
        var blob = mapDataUriToBlob(file.dataUri);
        var id = file.id || uuid();
        var name = file.name || "test-file-" + id;
        return {
            id: id,
            blob: blob,
            mimeType: blob.type,
            mediaType: 'image',
            name: name,
            size: blob.size,
            artifacts: {},
            representations: {
                image: {},
            },
        };
    });
}
export var mediaMock = new MediaMock();
