// @atlaskit/editor-common/dist/cjs/element-browser/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SIDEBAR_WIDTH = exports.SIDEBAR_HEADING_WRAPPER_HEIGHT = exports.SIDEBAR_HEADING_PADDING_LEFT = exports.SEARCH_ITEM_HEIGHT_WIDTH = exports.SCROLLBAR_WIDTH = exports.INLINE_SIDEBAR_HEIGHT = exports.GRID_SIZE = exports.FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS = exports.ELEMENT_LIST_PADDING = exports.ELEMENT_ITEM_PADDING = exports.ELEMENT_ITEM_HEIGHT = exports.ELEMENT_BROWSER_ID = exports.DEVICE_BREAKPOINT_NUMBERS = void 0;
var GRID_SIZE = exports.GRID_SIZE = 8;
var DEVICE_BREAKPOINT_NUMBERS = exports.DEVICE_BREAKPOINT_NUMBERS = {
  small: GRID_SIZE * 40,
  medium: GRID_SIZE * 75,
  large: GRID_SIZE * 128
};
var FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS = exports.FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS = {
  small: GRID_SIZE * 50,
  medium: DEVICE_BREAKPOINT_NUMBERS.medium,
  large: DEVICE_BREAKPOINT_NUMBERS.large
};
var SIDEBAR_WIDTH = exports.SIDEBAR_WIDTH = "".concat(GRID_SIZE * 25, "px");
var SIDEBAR_HEADING_WRAPPER_HEIGHT = exports.SIDEBAR_HEADING_WRAPPER_HEIGHT = "".concat(GRID_SIZE * 6, "px");
var SIDEBAR_HEADING_PADDING_LEFT = exports.SIDEBAR_HEADING_PADDING_LEFT = '12px';
var INLINE_SIDEBAR_HEIGHT = exports.INLINE_SIDEBAR_HEIGHT = '54px';
var SEARCH_ITEM_HEIGHT_WIDTH = exports.SEARCH_ITEM_HEIGHT_WIDTH = '20px';
var SCROLLBAR_WIDTH = exports.SCROLLBAR_WIDTH = 15;
var ELEMENT_LIST_PADDING = exports.ELEMENT_LIST_PADDING = 2;
var ELEMENT_ITEM_HEIGHT = exports.ELEMENT_ITEM_HEIGHT = 60;
var ELEMENT_ITEM_PADDING = exports.ELEMENT_ITEM_PADDING = 10;
var ELEMENT_BROWSER_ID = exports.ELEMENT_BROWSER_ID = 'editor-element-browser';
