// @atlaskit/analytics-listeners/dist/cjs/media/process-event.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.processEvent = processEvent;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _analyticsGasTypes = require("@atlaskit/analytics-gas-types");
var _analyticsNamespacedContext = require("@atlaskit/analytics-namespaced-context");
var _merge = _interopRequireDefault(require("lodash/merge"));
var _last = _interopRequireDefault(require("lodash/last"));
var _extractDataFromEvent = require("../atlaskit/extract-data-from-event");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var listenerVersion = "9.0.1";
function getMediaContexts(event) {
  return (0, _extractDataFromEvent.extractFromEventContext)(_analyticsNamespacedContext.MEDIA_CONTEXT, event);
}
function getMediaRegion() {
  try {
    return window && window.sessionStorage && window.sessionStorage.getItem('media-api-region');
  } catch (e) {
    return;
  }
}
function processEvent(event) {
  var sources = (0, _extractDataFromEvent.getSources)(event);
  var _ref = (0, _last.default)((0, _extractDataFromEvent.getPackageInfo)(event)) || {},
    packageName = _ref.packageName,
    packageVersion = _ref.packageVersion;
  var extraAttributes = (0, _extractDataFromEvent.getExtraAttributes)(event);
  var mediaContexts = getMediaContexts(event) || [];
  var mediaAttributes = _merge.default.apply(void 0, [{}].concat((0, _toConsumableArray2.default)(mediaContexts)));
  var tags = new Set(event.payload.tags || []);
  tags.add('media');
  var components = (0, _extractDataFromEvent.getComponents)(event);
  var mediaRegion = getMediaRegion();
  var payload = {
    source: (0, _last.default)(sources) || _analyticsGasTypes.DEFAULT_SOURCE,
    actionSubject: event.payload.actionSubject,
    action: event.payload.action,
    eventType: event.payload.eventType,
    actionSubjectId: event.payload.actionSubjectId,
    // for backward compatibilty
    name: event.payload.name,
    tags: Array.from(tags),
    attributes: _objectSpread(_objectSpread({
      packageName: packageName,
      packageVersion: packageVersion
    }, (0, _merge.default)({}, extraAttributes, mediaAttributes, event.payload.attributes)), {}, {
      componentHierarchy: components.join('.') || undefined,
      packageHierarchy: (0, _extractDataFromEvent.getPackageHierarchy)(event),
      sourceHierarchy: sources.join('.') || undefined,
      listenerVersion: listenerVersion
    }, mediaRegion ? {
      mediaRegion: mediaRegion
    } : undefined),
    nonPrivacySafeAttributes: event.payload.nonPrivacySafeAttributes
  };
  return payload;
}
