// @atlaskit/media-viewer/dist/cjs/newgen/media-viewer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var react_intl_1 = require("react-intl");
var media_ui_1 = require("@atlaskit/media-ui");
var analytics_next_1 = require("@atlaskit/analytics-next");
var media_viewer_1 = require("./analytics/media-viewer");
var closed_1 = require("./analytics/closed");
var index_1 = require("./analytics/index");
var list_1 = require("./list");
var collection_1 = require("./collection");
var content_1 = require("./content");
var styled_1 = require("./styled");
var perf_marks_1 = require("perf-marks");
var MediaViewerComponent = /** @class */ (function (_super) {
    tslib_1.__extends(MediaViewerComponent, _super);
    function MediaViewerComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            isSidebarVisible: false,
        };
        _this.fireAnalytics = function (payload) {
            var createAnalyticsEvent = _this.props.createAnalyticsEvent;
            if (createAnalyticsEvent) {
                var ev = createAnalyticsEvent(payload);
                ev.fire(index_1.channel);
            }
        };
        _this.onShortcutClosed = function () {
            _this.sendClosedEvent('escKey');
            var onClose = _this.props.onClose;
            if (onClose) {
                onClose();
            }
        };
        _this.onContentClose = function (_e, analyticsEvent) {
            var onClose = _this.props.onClose;
            if (analyticsEvent &&
                analyticsEvent.payload &&
                analyticsEvent.payload.actionSubject === 'button') {
                _this.sendClosedEvent('button');
            }
            if (onClose) {
                onClose();
            }
        };
        _this.toggleSidebar = function () {
            _this.setState({
                isSidebarVisible: !_this.state.isSidebarVisible,
            });
        };
        _this.renderSidebar = function () {
            var extensions = _this.props.extensions;
            var _a = _this.state, isSidebarVisible = _a.isSidebarVisible, selectedIdentifier = _a.selectedIdentifier;
            var sidebardSelectedIdentifier = selectedIdentifier || _this.defaultSelectedItem;
            if (sidebardSelectedIdentifier &&
                isSidebarVisible &&
                extensions &&
                extensions.sidebar) {
                return (react_1.default.createElement(styled_1.SidebarWrapper, { "data-testid": "media-viewer-sidebar-content" }, extensions.sidebar.renderer(sidebardSelectedIdentifier, {
                    close: _this.toggleSidebar,
                })));
            }
        };
        _this.onNavigationChange = function (selectedIdentifier) {
            _this.setState({ selectedIdentifier: selectedIdentifier });
        };
        return _this;
    }
    MediaViewerComponent.prototype.UNSAFE_componentWillMount = function () {
        this.fireAnalytics(media_viewer_1.mediaViewerModalEvent());
        perf_marks_1.start('MediaViewer.SessionDuration');
    };
    MediaViewerComponent.prototype.sendClosedEvent = function (input) {
        this.fireAnalytics(closed_1.closedEvent(input));
    };
    Object.defineProperty(MediaViewerComponent.prototype, "defaultSelectedItem", {
        get: function () {
            var _a = this.props, itemSource = _a.itemSource, selectedItem = _a.selectedItem;
            if (itemSource.kind === 'COLLECTION') {
                return selectedItem;
            }
            var items = itemSource.items;
            var firstItem = items[0];
            return selectedItem || firstItem;
        },
        enumerable: true,
        configurable: true
    });
    MediaViewerComponent.prototype.render = function () {
        var content = (react_1.default.createElement(styled_1.Blanket, { "data-testid": "media-viewer-popup" },
            react_1.default.createElement(media_ui_1.Shortcut, { keyCode: 27, handler: this.onShortcutClosed }),
            react_1.default.createElement(content_1.Content, { onClose: this.onContentClose }, this.renderContent()),
            this.renderSidebar()));
        return this.context.intl ? (content) : (react_1.default.createElement(react_intl_1.IntlProvider, { locale: "en" }, content));
    };
    MediaViewerComponent.prototype.renderContent = function () {
        var _a = this.props, mediaClient = _a.mediaClient, onClose = _a.onClose, itemSource = _a.itemSource, extensions = _a.extensions, contextId = _a.contextId;
        var isSidebarVisible = this.state.isSidebarVisible;
        if (itemSource.kind === 'COLLECTION') {
            return (react_1.default.createElement(collection_1.Collection, { pageSize: itemSource.pageSize, defaultSelectedItem: this.defaultSelectedItem, collectionName: itemSource.collectionName, mediaClient: mediaClient, onClose: onClose, extensions: extensions, onNavigationChange: this.onNavigationChange, onSidebarButtonClick: this.toggleSidebar }));
        }
        else if (itemSource.kind === 'ARRAY') {
            var items = itemSource.items;
            return (react_1.default.createElement(list_1.List, { defaultSelectedItem: this.defaultSelectedItem || items[0], items: items, mediaClient: mediaClient, onClose: onClose, extensions: extensions, onNavigationChange: this.onNavigationChange, onSidebarButtonClick: this.toggleSidebar, isSidebarVisible: isSidebarVisible, contextId: contextId }));
        }
        else {
            return null;
        }
    };
    MediaViewerComponent.contextTypes = {
        intl: react_intl_1.intlShape,
    };
    return MediaViewerComponent;
}(react_1.default.Component));
exports.MediaViewerComponent = MediaViewerComponent;
exports.MediaViewer = analytics_next_1.withAnalyticsEvents()(MediaViewerComponent);
