// @atlaskit/media-client/dist/cjs/client/mobile-upload.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MobileUploadImpl = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _mediaCommon = require("@atlaskit/media-common");
var _mediaState = require("@atlaskit/media-state");
var _fileStreamsCache = require("../file-streams-cache");
var _createFileDataLoader = require("../utils/createFileDataLoader");
var _mobileUpload = require("../utils/mobileUpload");
var MobileUploadImpl = exports.MobileUploadImpl = /*#__PURE__*/function () {
  function MobileUploadImpl(mediaApi) {
    var _this = this;
    var store = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : _mediaState.mediaStore;
    (0, _classCallCheck2.default)(this, MobileUploadImpl);
    (0, _defineProperty2.default)(this, "getErrorFileState", function (error, id, occurrenceKey) {
      return typeof error === 'string' ? {
        status: 'error',
        id: id,
        reason: error,
        occurrenceKey: occurrenceKey,
        message: error
      } : {
        status: 'error',
        id: id,
        reason: error === null || error === void 0 ? void 0 : error.reason,
        details: error === null || error === void 0 ? void 0 : error.attributes,
        occurrenceKey: occurrenceKey,
        message: error === null || error === void 0 ? void 0 : error.message
      };
    });
    (0, _defineProperty2.default)(this, "setFileState", function (id, fileState) {
      _this.store.setState(function (state) {
        state.files[id] = fileState;
      });
    });
    this.store = store;
    this.dataloader = (0, _createFileDataLoader.createFileDataloader)(mediaApi);
    this.servicesCache = (0, _mobileUpload.createServicesCache)();
  }
  return (0, _createClass2.default)(MobileUploadImpl, [{
    key: "notifyUploadStart",
    value: function notifyUploadStart(event) {
      var _this2 = this;
      var fileId = event.fileId,
        collectionName = event.collectionName,
        occurrenceKey = event.occurrenceKey,
        fileName = event.fileName,
        fileSize = event.fileSize,
        fileMimetype = event.fileMimetype,
        preview = event.preview,
        createdAt = event.createdAt;
      var mediaType = (0, _mediaCommon.getMediaTypeFromMimeType)(fileMimetype);
      var initialState = {
        status: 'uploading',
        id: fileId,
        occurrenceKey: occurrenceKey,
        name: fileName,
        size: fileSize,
        progress: 0,
        mediaType: mediaType,
        mimeType: fileMimetype,
        preview: preview,
        createdAt: createdAt
      };
      var service = (0, _mobileUpload.createMobileUploadService)((0, _mobileUpload.createMobileUploadStateMachine)(this.dataloader, initialState, collectionName));
      var subject = (0, _mobileUpload.createMobileFileStateSubject)(service);
      subject.subscribe({
        next: function next(fileState) {
          _this2.setFileState(fileId, fileState);
        },
        error: function error(err) {
          var errorFileState = _this2.getErrorFileState(err, fileId, occurrenceKey);
          _this2.setFileState(fileId, errorFileState);
        }
      });
      this.servicesCache.set(fileId, service);
      (0, _fileStreamsCache.getFileStreamsCache)().set(fileId, subject);
    }
  }, {
    key: "notifyUploadProgress",
    value: function notifyUploadProgress(event) {
      var fileId = event.fileId,
        progress = event.progress;
      var service = this.servicesCache.get(fileId);
      if (service) {
        service.send('UPLOAD_PROGRESS', {
          progress: progress
        });
      }
    }
  }, {
    key: "notifyUploadEnd",
    value: function notifyUploadEnd(event) {
      var fileId = event.fileId;
      var service = this.servicesCache.get(fileId);
      if (service) {
        service.send('UPLOAD_END');
      }
    }
  }, {
    key: "notifyUploadError",
    value: function notifyUploadError(event) {
      var fileId = event.fileId,
        message = event.message;
      var service = this.servicesCache.get(fileId);
      if (service) {
        service.send('UPLOAD_ERROR', {
          message: message
        });
      }
    }
  }]);
}();
