// @atlaskit/media-client/dist/cjs/client/stargate-client.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StargateClient = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var StargateClient = exports.StargateClient = /*#__PURE__*/function () {
  function StargateClient(baseUrl) {
    (0, _classCallCheck2.default)(this, StargateClient);
    this.baseUrl = baseUrl || '/gateway/api';
  }
  return (0, _createClass2.default)(StargateClient, [{
    key: "fetchToken",
    value: function () {
      var _fetchToken = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(clientId) {
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              _context.next = 2;
              return fetch("".concat(this.baseUrl, "/media/auth/smartedge"), {
                credentials: 'include',
                headers: {
                  'x-client-id': clientId
                }
              });
            case 2:
              return _context.abrupt("return", _context.sent.json());
            case 3:
            case "end":
              return _context.stop();
          }
        }, _callee, this);
      }));
      function fetchToken(_x) {
        return _fetchToken.apply(this, arguments);
      }
      return fetchToken;
    }()
  }, {
    key: "isTokenExpired",
    value: function isTokenExpired(token) {
      var currentTimeInSeconds = new Date().getTime() / 1000;
      return token.data.iat + token.data.expiresIn - currentTimeInSeconds < 0;
    }
  }]);
}();
