// @atlaskit/media-card/dist/cjs/card/ui/blanket/styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.fixedBlanketStyles = exports.blanketStyles = exports.blanketClassName = void 0;
var _react = require("@emotion/react");
var _styles = require("../styles");
var _colors = require("@atlaskit/theme/colors");
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var blanketClassName = exports.blanketClassName = 'media-card-blanket';
var fixedBlanketStyles = exports.fixedBlanketStyles = "background-color: ".concat("var(--ds-blanket, ".concat(_colors.N90A, ")"), ";");
var blanketStyles = exports.blanketStyles = function blanketStyles(isFixed) {
  return (0, _react.css)(
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  (0, _styles.transition)(), {
    position: 'absolute',
    width: '100%',
    height: '100%',
    maxHeight: '100%',
    maxWidth: '100%',
    left: 0,
    top: 0
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  isFixed ? fixedBlanketStyles : '');
};
blanketStyles.displayName = 'Blanket';
