// @atlaskit/editor-plugin-media/dist/es2019/ui/Media/DropPlaceholder.js
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import { css, jsx } from '@emotion/react';
import { injectIntl } from 'react-intl-next';
import { hexToRgba } from '@atlaskit/adf-schema';
import { dropPlaceholderMessages } from '@atlaskit/editor-common/media';
import DocumentFilledIcon from '@atlaskit/icon/core/migration/file--document-filled';
import { B200, B300, B400 } from '@atlaskit/theme/colors';
import { borderRadius } from '@atlaskit/theme/constants';
import { FILE_WIDTH, MEDIA_HEIGHT } from '../../nodeviews/mediaNodeView/media';
const iconWrapperStyles = css({
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
  color: `var(--ds-icon-accent-blue, ${hexToRgba(B400, 0.4) || B400})`,
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
  background: `var(--ds-background-accent-blue-subtle, ${hexToRgba(B300, 0.6) || B300})`,
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  borderRadius: `${borderRadius()}px`,
  margin: `${"var(--ds-space-075, 6px)"} ${"var(--ds-space-050, 4px)"} ${"var(--ds-space-300, 24px)"}`,
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  width: `${FILE_WIDTH}px`,
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  minHeight: `${MEDIA_HEIGHT}px`,
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center'
});
const dropLineStyles = css({
  background: `var(--ds-border-focused, ${B200})`,
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  borderRadius: `${borderRadius()}px`,
  margin: `${"var(--ds-space-025, 2px)"} 0`,
  width: '100%',
  height: '2px'
});
const IconWrapperComponent = props => {
  const {
    intl
  } = props;
  const {
    dropPlaceholderLabel
  } = dropPlaceholderMessages;
  return jsx("div", {
    css: iconWrapperStyles
  }, jsx(DocumentFilledIcon, {
    label: intl.formatMessage(dropPlaceholderLabel),
    LEGACY_size: "medium"
  }));
};
const IntlIconWrapper = injectIntl(IconWrapperComponent);
export default (({
  type = 'group'
}) => type === 'single' ? jsx("div", {
  css: dropLineStyles
}) : jsx(IntlIconWrapper, null));
