// @atlaskit/teams-public/dist/cjs/ui/team-containers/linked-container-card/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LinkedContainerCard = void 0;
require("./index.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _reactIntlNext = require("react-intl-next");
var _analyticsNext = require("@atlaskit/analytics-next");
var _avatar = _interopRequireDefault(require("@atlaskit/avatar"));
var _new = require("@atlaskit/button/new");
var _cross = _interopRequireDefault(require("@atlaskit/icon/utility/cross"));
var _link = _interopRequireDefault(require("@atlaskit/link"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _compiled = require("@atlaskit/primitives/compiled");
var _tooltip = _interopRequireDefault(require("@atlaskit/tooltip"));
var _analytics = require("../../../common/utils/analytics");
var _getContainerProperties = require("../../../common/utils/get-container-properties");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var styles = {
  container: "_2rkoop52 _1h6dz9xs _12y3e4h9 _12jimuej _1qu2nqa1 _ca0qutpp _u5f3utpp _n3tdutpp _19bvutpp _syaz1fxt",
  card: "_4cvr1h6o _1bsb1osq",
  iconWrapper: "_1bsbzwfg _4t3izwfg _1ul9zwfg _1tkezwfg",
  crossIconWrapper: "_1e0c1txw _4cvr1h6o _1bahesu3 _18u01wug"
};
var LinkedCardWrapper = function LinkedCardWrapper(_ref) {
  var children = _ref.children,
    href = _ref.href,
    handleMouseEnter = _ref.handleMouseEnter,
    handleMouseLeave = _ref.handleMouseLeave,
    containerType = _ref.containerType,
    containerId = _ref.containerId;
  var _useState = (0, _react.useState)(false),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    hovered = _useState2[0],
    setHovered = _useState2[1];
  var onMouseEnter = function onMouseEnter() {
    handleMouseEnter();
    setHovered(true);
  };
  var _usePeopleAndTeamAnal = (0, _analytics.usePeopleAndTeamAnalytics)(),
    fireUIEvent = _usePeopleAndTeamAnal.fireUIEvent;
  var _useAnalyticsEvents = (0, _analyticsNext.useAnalyticsEvents)(),
    createAnalyticsEvent = _useAnalyticsEvents.createAnalyticsEvent;
  var onMouseLeave = function onMouseLeave() {
    handleMouseLeave();
    setHovered(false);
  };
  return /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    backgroundColor: hovered ? 'color.background.input.hovered' : 'color.background.input',
    xcss: styles.container,
    onMouseEnter: onMouseEnter,
    onMouseLeave: onMouseLeave,
    testId: "linked-container-card-inner"
  }, /*#__PURE__*/_react.default.createElement(_link.default, {
    href: href,
    appearance: "subtle",
    onClick: function onClick() {
      fireUIEvent(createAnalyticsEvent, {
        action: _analytics.AnalyticsAction.CLICKED,
        actionSubject: 'container',
        actionSubjectId: 'teamContainer',
        attributes: {
          containerSelected: {
            container: containerType,
            containerId: containerId
          }
        }
      });
    }
  }, children));
};
var LinkedContainerCard = exports.LinkedContainerCard = function LinkedContainerCard(_ref2) {
  var containerType = _ref2.containerType,
    title = _ref2.title,
    containerIcon = _ref2.containerIcon,
    link = _ref2.link,
    containerId = _ref2.containerId,
    containerTypeProperties = _ref2.containerTypeProperties,
    onDisconnectButtonClick = _ref2.onDisconnectButtonClick;
  var _useAnalyticsEvents2 = (0, _analyticsNext.useAnalyticsEvents)(),
    createAnalyticsEvent = _useAnalyticsEvents2.createAnalyticsEvent;
  var _getContainerProperti = (0, _getContainerProperties.getContainerProperties)(containerType, 'small', containerTypeProperties),
    description = _getContainerProperti.description,
    icon = _getContainerProperti.icon,
    containerTypeText = _getContainerProperti.containerTypeText;
  var _useState3 = (0, _react.useState)(false),
    _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
    showCloseIcon = _useState4[0],
    setShowCloseIcon = _useState4[1];
  var _useIntl = (0, _reactIntlNext.useIntl)(),
    formatMessage = _useIntl.formatMessage;
  var _usePeopleAndTeamAnal2 = (0, _analytics.usePeopleAndTeamAnalytics)(),
    fireUIEvent = _usePeopleAndTeamAnal2.fireUIEvent;
  return /*#__PURE__*/_react.default.createElement(LinkedCardWrapper, {
    href: link || '#',
    handleMouseEnter: function handleMouseEnter() {
      return setShowCloseIcon(true);
    },
    handleMouseLeave: function handleMouseLeave() {
      return setShowCloseIcon(false);
    },
    containerType: containerType,
    containerId: containerId
  }, /*#__PURE__*/_react.default.createElement(_compiled.Inline, {
    space: "space.100",
    xcss: styles.card
  }, /*#__PURE__*/_react.default.createElement(_avatar.default, {
    appearance: "square",
    size: "medium",
    src: containerIcon,
    testId: "linked-container-icon"
  }), /*#__PURE__*/_react.default.createElement(_compiled.Stack, null, /*#__PURE__*/_react.default.createElement(_compiled.Text, {
    maxLines: 1,
    weight: "medium",
    color: "color.text"
  }, title), /*#__PURE__*/_react.default.createElement(_compiled.Flex, (0, _extends2.default)({
    gap: "space.050"
  }, (0, _platformFeatureFlags.fg)('enable_card_alignment_fix') ? {
    alignItems: 'center'
  } : {}), icon, /*#__PURE__*/_react.default.createElement(_compiled.Inline, {
    space: "space.050"
  }, /*#__PURE__*/_react.default.createElement(_compiled.Text, {
    size: "small",
    color: "color.text.subtle"
  }, description), /*#__PURE__*/_react.default.createElement(_compiled.Text, {
    size: "small",
    color: "color.text.subtle"
  }, containerTypeText)))), showCloseIcon && /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    xcss: styles.crossIconWrapper
  }, /*#__PURE__*/_react.default.createElement(_tooltip.default, {
    content: formatMessage(messages.disconnectTooltip),
    position: "top"
  }, /*#__PURE__*/_react.default.createElement(_new.IconButton, {
    label: "disconnect the container ".concat(title),
    appearance: "subtle",
    icon: _cross.default,
    spacing: "compact",
    onClick: function onClick(e) {
      e.preventDefault();
      e.stopPropagation();
      onDisconnectButtonClick();
      fireUIEvent(createAnalyticsEvent, {
        action: _analytics.AnalyticsAction.CLICKED,
        actionSubject: 'button',
        actionSubjectId: 'containerUnlinkButton'
      });
    }
  })))));
};
var messages = (0, _reactIntlNext.defineMessages)({
  disconnectTooltip: {
    id: 'ptc-directory.team-containers.disconnect-button.tooltip',
    defaultMessage: 'Disconnect',
    description: 'Tooltip for the disconnect button'
  }
});
