// @atlaskit/editor-core/dist/cjs/nodeviews/getInlineNodeViewProducer.styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InlineNodeViewSharedStyles = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = require("@emotion/react");
var _reactNodeView = require("@atlaskit/editor-common/react-node-view");
var _whitespace = require("@atlaskit/editor-common/whitespace");
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

// For reasoning behind styles, see comments in:
// ./getInlineNodeViewProducer -> portalChildren()

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var InlineNodeViewSharedStyles = exports.InlineNodeViewSharedStyles = (0, _react.css)((0, _defineProperty2.default)((0, _defineProperty2.default)((0, _defineProperty2.default)((0, _defineProperty2.default)({}, ".".concat(_reactNodeView.inlineNodeViewClassname), {
  display: 'inline',
  userSelect: 'all',
  whiteSpace: 'nowrap',
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors, @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
  '& > *:not(.zeroWidthSpaceContainer)': {
    whiteSpace: 'pre-wrap'
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '& > .assistive': {
    userSelect: 'none'
  }
}), '&.ua-safari', (0, _defineProperty2.default)({}, ".".concat(_reactNodeView.inlineNodeViewClassname), {
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors, @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
  '::selection, *::selection': {
    background: 'transparent'
  }
})), "&.ua-chrome .".concat(_reactNodeView.inlineNodeViewClassname, " > span"), {
  userSelect: 'none'
}), ".".concat(_reactNodeView.inlineNodeViewClassname, "AddZeroWidthSpace"), {
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
  '::after': {
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    content: "'".concat(_whitespace.ZERO_WIDTH_SPACE, "'")
  }
}));
