// @atlaskit/media-card/dist/esm/utils/mockIntersectionObserver.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
var noop = function noop() {};
export var MockIntersectionObserver = /*#__PURE__*/function () {
  function MockIntersectionObserver() {
    _classCallCheck(this, MockIntersectionObserver);
    _defineProperty(this, "callback", noop);
  }
  return _createClass(MockIntersectionObserver, [{
    key: "storeMockObserver",
    value: function storeMockObserver(_ref) {
      var observe = _ref.observe,
        disconnect = _ref.disconnect;
      // eslint-disable-next-line @typescript-eslint/no-this-alias
      var outerThis = this;
      this.mockObserver = /*#__PURE__*/_createClass(function _class(callback) {
        _classCallCheck(this, _class);
        _defineProperty(this, "observe", observe !== null && observe !== void 0 ? observe : noop);
        _defineProperty(this, "unobserve", noop);
        _defineProperty(this, "disconnect", disconnect !== null && disconnect !== void 0 ? disconnect : noop);
        outerThis.callback = callback;
        outerThis.mockObserverInstance = this;
      });
    }
  }, {
    key: "addMockObserverToWindow",
    value: function addMockObserverToWindow() {
      this.originalObserver = window.IntersectionObserver;
      window.IntersectionObserver = this.mockObserver;
    }
  }, {
    key: "removeMockObserverFromWindow",
    value: function removeMockObserverFromWindow() {
      window.IntersectionObserver = this.originalObserver;
    }
  }, {
    key: "setup",
    value: function setup() {
      var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      this.storeMockObserver(options);
      this.addMockObserverToWindow();
    }
  }, {
    key: "cleanup",
    value: function cleanup() {
      this.removeMockObserverFromWindow();
    }
  }, {
    key: "triggerIntersect",
    value: function triggerIntersect(entry) {
      var entries = [entry];
      this.callback(entries, this.mockObserverInstance);
    }
  }]);
}();
