// @atlaskit/link-datasource/dist/cjs/ui/common/modal/display-view-dropdown/display-view-drop-down.js
/* display-view-drop-down.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DisplayViewDropDown = void 0;
require("./display-view-drop-down.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireDefault(require("react"));
var _reactIntlNext = require("react-intl-next");
var _new = _interopRequireDefault(require("@atlaskit/button/new"));
var _dropdownMenu = _interopRequireWildcard(require("@atlaskit/dropdown-menu"));
var _chevronDown = _interopRequireDefault(require("@atlaskit/icon/utility/chevron-down"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _compiled = require("@atlaskit/primitives/compiled");
var _messages = require("./messages");
var _excluded = ["triggerRef"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var styles = {
  dropDownItemGroupStylesOld: "_2rkoglpi _1bsbnklw _4t3i11ux _ca0q1b66 _n3td1b66",
  dropDownItemGroupStyles: "_2rkoglpi _1bsb1pna _4t3i7iqz"
};
var InlineIconOld = /*#__PURE__*/_react.default.createElement("svg", {
  xmlns: "http://www.w3.org/2000/svg",
  width: "24",
  height: "24",
  viewBox: "0 0 24 24",
  fill: "none"
}, /*#__PURE__*/_react.default.createElement("path", {
  fillRule: "evenodd",
  clipRule: "evenodd",
  d: "M4 9C2.89543 9 2 9.67155 2 10.5V13.5C2 14.3285 2.89543 15 4 15H20C21.1046 15 22 14.3285 22 13.5V10.5C22 9.67155 21.1046 9 20 9H4ZM4.22222 11.25C4.22222 10.5 4.22222 10.5 5 10.5H7C7.55556 10.5 7.55556 10.5 7.55556 11.25V13C7.55556 13.5 7.55556 13.5 7 13.5H5C4.22222 13.5 4.22222 13.5 4.22222 13V11.25ZM9 12C9 11.7929 9.2239 11.625 9.5 11.625H19.5C19.7761 11.625 20 11.7929 20 12C20 12.2071 19.7761 12.375 19.5 12.375H9.5C9.2239 12.375 9 12.2071 9 12Z",
  fill: 'currentColor'
}));
var InlineIcon = /*#__PURE__*/_react.default.createElement("svg", {
  xmlns: "http://www.w3.org/2000/svg",
  width: "24",
  height: "24",
  fill: "none",
  viewBox: "0 0 24 24"
}, /*#__PURE__*/_react.default.createElement("path", {
  fill: "currentColor",
  fillRule: "evenodd",
  d: "M6 10.5a.5.5 0 0 0-.5.5v2a.5.5 0 0 0 .5.5h2.5v-3H6Zm4 0v3h8a.5.5 0 0 0 .5-.5v-2a.5.5 0 0 0-.5-.5h-8ZM4 11a2 2 0 0 1 2-2h12a2 2 0 0 1 2 2v2a2 2 0 0 1-2 2H6a2 2 0 0 1-2-2v-2Z",
  clipRule: "evenodd"
}));
var ListIconOld = /*#__PURE__*/_react.default.createElement("svg", {
  xmlns: "http://www.w3.org/2000/svg",
  width: "24",
  height: "24",
  viewBox: "0 0 15 15",
  fill: "none"
}, /*#__PURE__*/_react.default.createElement("path", {
  fillRule: "evenodd",
  clipRule: "evenodd",
  d: "M0 2C0 0.89543 0.895431 0 2 0H13C14.1046 0 15 0.895431 15 2V13C15 14.1046 14.1046 15 13 15H2C0.89543 15 0 14.1046 0 13V2ZM5 6C5 5.44772 5.44772 5 6 5L12 5C12.5523 5 13 5.44772 13 6C13 6.55229 12.5523 7 12 7L6 7C5.44772 7 5 6.55228 5 6ZM6 8C5.44772 8 5 8.44771 5 9C5 9.55228 5.44772 10 6 10L12 10C12.5523 10 13 9.55229 13 9C13 8.44772 12.5523 8 12 8L6 8ZM5 12C5 11.4477 5.44772 11 6 11L12 11C12.5523 11 13 11.4477 13 12C13 12.5523 12.5523 13 12 13L6 13C5.44772 13 5 12.5523 5 12ZM3 7C3.55228 7 4 6.55228 4 6C4 5.44772 3.55228 5 3 5C2.44772 5 2 5.44772 2 6C2 6.55228 2.44772 7 3 7ZM5 3C5 2.44772 5.44772 2 6 2L12 2C12.5523 2 13 2.44772 13 3C13 3.55229 12.5523 4 12 4L6 4C5.44772 4 5 3.55228 5 3ZM3 4C3.55228 4 4 3.55228 4 3C4 2.44772 3.55228 2 3 2C2.44772 2 2 2.44772 2 3C2 3.55228 2.44772 4 3 4ZM4 9C4 9.55229 3.55228 10 3 10C2.44772 10 2 9.55229 2 9C2 8.44771 2.44772 8 3 8C3.55228 8 4 8.44771 4 9ZM3 13C3.55228 13 4 12.5523 4 12C4 11.4477 3.55228 11 3 11C2.44772 11 2 11.4477 2 12C2 12.5523 2.44772 13 3 13Z",
  fill: 'currentColor'
}));
var ListIcon = /*#__PURE__*/_react.default.createElement("svg", {
  xmlns: "http://www.w3.org/2000/svg",
  width: "24",
  height: "24",
  fill: "none",
  viewBox: "0 0 24 24"
}, /*#__PURE__*/_react.default.createElement("path", {
  fill: "currentColor",
  fillRule: "evenodd",
  d: "M5 7a2 2 0 0 1 2-2h10a2 2 0 0 1 2 2v10a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V7Zm2-.5a.5.5 0 0 0-.5.5v10a.5.5 0 0 0 .5.5h10a.5.5 0 0 0 .5-.5V7a.5.5 0 0 0-.5-.5H7Zm2.5 3H8V8h1.5v1.5Zm6.5 0h-5V8h5v1.5Zm-6.5 3.25H8v-1.5h1.5v1.5Zm6.5 0h-5v-1.5h5v1.5ZM9.5 16H8v-1.5h1.5V16Zm6.5 0h-5v-1.5h5V16Z",
  clipRule: "evenodd"
}));
var DisplayViewDropDown = exports.DisplayViewDropDown = function DisplayViewDropDown(_ref) {
  var onViewModeChange = _ref.onViewModeChange,
    viewMode = _ref.viewMode;
  var _useIntl = (0, _reactIntlNext.useIntl)(),
    formatMessage = _useIntl.formatMessage;
  var isTable = viewMode === 'table';
  var triggerText = isTable ? formatMessage(_messages.displayViewDropDownMessages.viewModeListLabel) : formatMessage(_messages.displayViewDropDownMessages.viewModeInlineLinkLabel);
  if ((0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv')) {
    return /*#__PURE__*/_react.default.createElement(_dropdownMenu.default, {
      trigger: function trigger(_ref2) {
        var triggerRef = _ref2.triggerRef,
          triggerProps = (0, _objectWithoutProperties2.default)(_ref2, _excluded);
        return /*#__PURE__*/_react.default.createElement(_new.default, (0, _extends2.default)({}, triggerProps, {
          ref: triggerRef,
          iconAfter: function iconAfter() {
            return /*#__PURE__*/_react.default.createElement(_chevronDown.default, {
              label: "",
              color: "currentColor"
            });
          }
        }), triggerText);
      },
      testId: "datasource-modal--view-drop-down"
    }, /*#__PURE__*/_react.default.createElement(_compiled.Box, {
      xcss: styles.dropDownItemGroupStyles
    }, /*#__PURE__*/_react.default.createElement(_dropdownMenu.DropdownItemGroup, null, /*#__PURE__*/_react.default.createElement(_dropdownMenu.DropdownItem, {
      testId: "dropdown-item-table",
      onClick: function onClick() {
        return onViewModeChange('table');
      },
      isSelected: isTable,
      description: formatMessage(_messages.displayViewDropDownMessages.viewModeListDescription),
      elemBefore: ListIcon
    }, /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, _messages.displayViewDropDownMessages.viewModeListLabel)), /*#__PURE__*/_react.default.createElement(_dropdownMenu.DropdownItem, {
      testId: "dropdown-item-inline-link",
      onClick: function onClick() {
        return onViewModeChange('inline');
      },
      isSelected: !isTable,
      description: formatMessage(_messages.displayViewDropDownMessages.viewModeInlineLinkDescription),
      elemBefore: InlineIcon
    }, /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, _messages.displayViewDropDownMessages.viewModeInlineLinkLabel)))));
  }
  return /*#__PURE__*/_react.default.createElement(_dropdownMenu.default, {
    trigger: triggerText,
    testId: "datasource-modal--view-drop-down"
  }, /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    xcss: styles.dropDownItemGroupStylesOld
  }, /*#__PURE__*/_react.default.createElement(_dropdownMenu.DropdownItemGroup, null, /*#__PURE__*/_react.default.createElement(_dropdownMenu.DropdownItem, {
    testId: "dropdown-item-table",
    onClick: function onClick() {
      return onViewModeChange('table');
    },
    isSelected: isTable,
    description: formatMessage(_messages.displayViewDropDownMessages.viewModeListDescriptionOld),
    elemBefore: ListIconOld
  }, /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, _messages.displayViewDropDownMessages.viewModeListLabel)), /*#__PURE__*/_react.default.createElement(_dropdownMenu.DropdownItem, {
    testId: "dropdown-item-inline-link",
    onClick: function onClick() {
      return onViewModeChange('inline');
    },
    isSelected: !isTable,
    description: formatMessage(_messages.displayViewDropDownMessages.viewModeInlineLinkDescriptionOld),
    elemBefore: InlineIconOld
  }, /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, _messages.displayViewDropDownMessages.viewModeInlineLinkLabel)))));
};
