// @atlaskit/link-datasource/dist/cjs/ui/common/modal/display-view-dropdown/messages.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.displayViewDropDownMessages = void 0;
var _reactIntlNext = require("react-intl-next");
var displayViewDropDownMessages = exports.displayViewDropDownMessages = (0, _reactIntlNext.defineMessages)({
  viewModeListLabel: {
    id: 'linkDataSource.jira-issues.configmodal.viewModeListLabel',
    description: 'Display search results as a list',
    defaultMessage: 'List'
  },
  viewModeListDescription: {
    id: 'linkDataSource.jira-issues.configmodal.viewModeListDescription',
    description: 'Description for list view mode',
    defaultMessage: 'Display the number of search results as a list'
  },
  viewModeListDescriptionOld: {
    id: 'linkDataSource.jira-issues.configmodal.viewModeListDescriptionOld',
    description: 'Description for list view mode',
    defaultMessage: 'Display search results as a list'
  },
  viewModeInlineLinkLabel: {
    id: 'linkDataSource.jira-issues.configmodal.viewModeInlineLinkLabel',
    description: 'Display the number of search results as an inline smart link',
    defaultMessage: 'Inline link'
  },
  viewModeInlineLinkDescription: {
    id: 'linkDataSource.jira-issues.configmodal.viewModeInlineLinkDescription',
    description: 'Description for inline link view mode',
    defaultMessage: 'Display the number of search results or as an inline Smart Link'
  },
  viewModeInlineLinkDescriptionOld: {
    id: 'linkDataSource.jira-issues.configmodal.viewModeInlineLinkDescriptionOld',
    description: 'Description for inline link view mode',
    defaultMessage: 'Display the number of search results as an inline Smart Link'
  }
});
