// @atlaskit/editor-core/dist/cjs/create-editor/create-universal-preset.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createUniversalPreset = createUniversalPreset;
var _universal = _interopRequireDefault(require("../presets/universal"));
var _featureFlagsFromProps = require("../utils/feature-flags-from-props");
var _createPluginsList = require("./create-plugins-list");
// Separate file, we should not accidentally import this into the `ComposableEditor`
// otherwise it will blow up the bundle size.
function createUniversalPreset(_ref) {
  var props = _ref.props,
    prevProps = _ref.prevProps,
    initialPluginConfiguration = _ref.initialPluginConfiguration;
  return (0, _universal.default)({
    appearance: props.appearance,
    props: (0, _createPluginsList.getDefaultPresetOptionsFromEditorProps)(props),
    initialPluginConfiguration: initialPluginConfiguration,
    featureFlags: (0, _featureFlagsFromProps.createFeatureFlagsFromProps)(props.featureFlags),
    prevAppearance: prevProps === null || prevProps === void 0 ? void 0 : prevProps.appearance
  });
}
