// @atlaskit/editor-core/dist/cjs/create-editor/editorStateNotificationPlugin.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createEditorStateNotificationPlugin = void 0;
var _safePlugin = require("@atlaskit/editor-common/safe-plugin");
var _state = require("@atlaskit/editor-prosemirror/state");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var key = new _state.PluginKey('editorStateNotificationPlugin');
var createEditorStateNotificationPlugin = exports.createEditorStateNotificationPlugin = function createEditorStateNotificationPlugin(onEditorStateUpdated, onEditorViewStateUpdatedCallbacks) {
  var transactions = [];
  return new _safePlugin.SafePlugin({
    key: key,
    state: {
      init: function init() {
        return {
          latestTransaction: undefined
        };
      },
      apply: function apply(tr) {
        transactions.push(tr);
        return {
          latestTransaction: tr
        };
      }
    },
    view: function view() {
      return {
        update: function update(view, oldEditorState) {
          var _key$getState;
          var originalTransaction = (_key$getState = key.getState(view.state)) === null || _key$getState === void 0 ? void 0 : _key$getState.latestTransaction;
          if (originalTransaction && (0, _platformFeatureFlags.fg)('platform_editor_migrate_state_updates')) {
            onEditorViewStateUpdatedCallbacks.forEach(function (entry) {
              entry.callback({
                originalTransaction: originalTransaction,
                transactions: transactions,
                oldEditorState: oldEditorState,
                newEditorState: view.state
              });
            });
            transactions = [];
          }
          onEditorStateUpdated({
            oldEditorState: oldEditorState,
            newEditorState: view.state
          });
        }
      };
    }
  });
};
