// @atlaskit/media-ui/dist/es2019/mediaImage/styled.js
import _extends from "@babel/runtime/helpers/extends";
/* eslint-disable @atlaskit/design-system/ensure-design-token-usage/preview */
import React from 'react';

/*
  TODO: https://product-fabric.atlassian.net/browse/CXP-3132

  We need to revisit omitting percentage values for Design Token since the parent applies transform(-50%, -50%) on it (packages/media/media-ui/src/mediaImage/index.tsx). This styling is complex and may potentially cause a bug if not treaded carefully.
*/

const imgStyle = {
  position: 'absolute',
  left: '50%',
  top: '50%',
  objectFit: 'contain',
  imageOrientation: 'none'
};
export const ImageComponent = props => {
  const {
    style,
    imageRef,
    ...otherProps
  } = props;
  return (
    /*#__PURE__*/
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop, jsx-a11y/alt-text -- Ignored via go/DSP-18766
    React.createElement("img", _extends({}, otherProps, {
      ref: imageRef,
      style: {
        ...imgStyle,
        ...style
      }
    }))
  );
};
