// @atlaskit/linking-common/dist/cjs/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.cardAction = exports.ACTION_UPDATE_METADATA_STATUS = exports.ACTION_RESOLVING = exports.ACTION_RESOLVED = exports.ACTION_RELOADING = exports.ACTION_PRELOAD = exports.ACTION_PENDING = exports.ACTION_ERROR_FALLBACK = exports.ACTION_ERROR = void 0;
// TODO: eventually move this file to @atlaskit/link-actions

var ACTION_PENDING = exports.ACTION_PENDING = 'pending';
var ACTION_RESOLVING = exports.ACTION_RESOLVING = 'resolving';
var ACTION_RESOLVED = exports.ACTION_RESOLVED = 'resolved';
var ACTION_RELOADING = exports.ACTION_RELOADING = 'reloading';
var ACTION_ERROR = exports.ACTION_ERROR = 'errored';
var ACTION_ERROR_FALLBACK = exports.ACTION_ERROR_FALLBACK = 'fallback';
var ACTION_PRELOAD = exports.ACTION_PRELOAD = 'preload';
var ACTION_UPDATE_METADATA_STATUS = exports.ACTION_UPDATE_METADATA_STATUS = 'metadata';

// export const ANALYTICS_RESOLVING = 'resolving';
// export const ANALYTICS_ERROR = 'errored';
// export const ANALYTICS_FALLBACK = 'fallback';

var cardAction = exports.cardAction = function cardAction(type, _ref, payload, error, metadataStatus, ignoreStatusCheck) {
  var url = _ref.url;
  return {
    type: type,
    url: url,
    payload: payload,
    error: error,
    metadataStatus: metadataStatus,
    ignoreStatusCheck: ignoreStatusCheck
  };
};
