// @atlaskit/linking-common/dist/cjs/environments.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getResolverUrl = exports.getBaseUrl = exports.default = exports.BaseUrls = void 0;
var devBaseUrl = 'https://api-private.dev.atlassian.com';
var stgBaseUrl = 'https://pug.jira-dev.com/gateway/api';
var prodBaseUrl = 'https://api-private.atlassian.com';
var BaseUrls = exports.BaseUrls = {
  dev: devBaseUrl,
  development: devBaseUrl,
  stg: stgBaseUrl,
  staging: stgBaseUrl,
  prd: prodBaseUrl,
  prod: prodBaseUrl,
  production: prodBaseUrl
};
var getBaseUrl = exports.getBaseUrl = function getBaseUrl(envKey) {
  // If an environment is provided, then use Stargate.
  if (envKey) {
    return envKey in BaseUrls ? BaseUrls[envKey] : prodBaseUrl;
  }

  // Otherwise, use the current origin of the page.
  return typeof window.location !== 'undefined' ? window.location.origin : '';
};
var getResolverUrl = exports.getResolverUrl = function getResolverUrl(envKey, baseUrlOverride) {
  // If an environment is provided, then use Stargate directly for requests.
  if (envKey || baseUrlOverride) {
    var baseUrl = baseUrlOverride || getBaseUrl(envKey);
    return "".concat(baseUrl, "/object-resolver");
  } else {
    // Otherwise, we fallback to using the Edge Proxy to access Stargate,
    // which fixes some cookie issues with strict Browser policies.
    return '/gateway/api/object-resolver';
  }
};
var _default = exports.default = BaseUrls;
