// @atlaskit/linking-common/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ACTION_ERROR", {
  enumerable: true,
  get: function get() {
    return _actions.ACTION_ERROR;
  }
});
Object.defineProperty(exports, "ACTION_ERROR_FALLBACK", {
  enumerable: true,
  get: function get() {
    return _actions.ACTION_ERROR_FALLBACK;
  }
});
Object.defineProperty(exports, "ACTION_PENDING", {
  enumerable: true,
  get: function get() {
    return _actions.ACTION_PENDING;
  }
});
Object.defineProperty(exports, "ACTION_PRELOAD", {
  enumerable: true,
  get: function get() {
    return _actions.ACTION_PRELOAD;
  }
});
Object.defineProperty(exports, "ACTION_RELOADING", {
  enumerable: true,
  get: function get() {
    return _actions.ACTION_RELOADING;
  }
});
Object.defineProperty(exports, "ACTION_RESOLVED", {
  enumerable: true,
  get: function get() {
    return _actions.ACTION_RESOLVED;
  }
});
Object.defineProperty(exports, "ACTION_RESOLVING", {
  enumerable: true,
  get: function get() {
    return _actions.ACTION_RESOLVING;
  }
});
Object.defineProperty(exports, "ACTION_UPDATE_METADATA_STATUS", {
  enumerable: true,
  get: function get() {
    return _actions.ACTION_UPDATE_METADATA_STATUS;
  }
});
Object.defineProperty(exports, "APIError", {
  enumerable: true,
  get: function get() {
    return _errors.APIError;
  }
});
Object.defineProperty(exports, "BaseUrls", {
  enumerable: true,
  get: function get() {
    return _environments.BaseUrls;
  }
});
Object.defineProperty(exports, "DATASOURCE_DEFAULT_LAYOUT", {
  enumerable: true,
  get: function get() {
    return _constants.DATASOURCE_DEFAULT_LAYOUT;
  }
});
Object.defineProperty(exports, "NetworkError", {
  enumerable: true,
  get: function get() {
    return _api.NetworkError;
  }
});
Object.defineProperty(exports, "Pulse", {
  enumerable: true,
  get: function get() {
    return _Pulse.Pulse;
  }
});
Object.defineProperty(exports, "Skeleton", {
  enumerable: true,
  get: function get() {
    return _Skeleton.Skeleton;
  }
});
Object.defineProperty(exports, "SpanSkeleton", {
  enumerable: true,
  get: function get() {
    return _Skeleton.SpanSkeleton;
  }
});
Object.defineProperty(exports, "cardAction", {
  enumerable: true,
  get: function get() {
    return _actions.cardAction;
  }
});
Object.defineProperty(exports, "filterSiteProducts", {
  enumerable: true,
  get: function get() {
    return _filterSiteProducts.filterSiteProducts;
  }
});
Object.defineProperty(exports, "getBaseUrl", {
  enumerable: true,
  get: function get() {
    return _environments.getBaseUrl;
  }
});
Object.defineProperty(exports, "getResolverUrl", {
  enumerable: true,
  get: function get() {
    return _environments.getResolverUrl;
  }
});
Object.defineProperty(exports, "getStatus", {
  enumerable: true,
  get: function get() {
    return _getStatus.getStatus;
  }
});
Object.defineProperty(exports, "getUrl", {
  enumerable: true,
  get: function get() {
    return _store.getUrl;
  }
});
Object.defineProperty(exports, "promiseDebounce", {
  enumerable: true,
  get: function get() {
    return _promiseDebounce.promiseDebounce;
  }
});
Object.defineProperty(exports, "request", {
  enumerable: true,
  get: function get() {
    return _api.request;
  }
});
var _constants = require("./common/utils/constants");
var _actions = require("./actions");
var _errors = require("./errors");
var _store = require("./store");
var _promiseDebounce = require("./utils/promise-debounce");
var _getStatus = require("./utils/get-status");
var _filterSiteProducts = require("./utils/filter-site-products");
var _environments = require("./environments");
var _api = require("./api");
var _Pulse = require("./components/Pulse");
var _Skeleton = require("./components/Skeleton");
