// @atlaskit/mention/dist/cjs/components/MentionItem/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "MENTION_ITEM_HEIGHT", {
  enumerable: true,
  get: function get() {
    return _styles.MENTION_ITEM_HEIGHT;
  }
});
exports.default = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));
var _avatar = _interopRequireDefault(require("@atlaskit/avatar"));
var _lozenge = _interopRequireDefault(require("@atlaskit/lozenge"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _colors = require("@atlaskit/theme/colors");
var _react = _interopRequireDefault(require("react"));
var _informationEditorPanel = _interopRequireDefault(require("@atlaskit/icon/core/migration/information--editor-panel"));
var _types = require("../../types");
var _i18n = require("../../util/i18n");
var _mouse = require("../../util/mouse");
var _NoAccessTooltip = _interopRequireDefault(require("../NoAccessTooltip"));
var _LockCircleIcon = _interopRequireDefault(require("../LockCircleIcon"));
var _styles = require("./styles");
var _MentionHighlightHelpers = require("./MentionHighlightHelpers");
var _MentionDescriptionByline = _interopRequireDefault(require("../MentionDescriptionByline"));
var _MessagesIntlProvider = _interopRequireDefault(require("../MessagesIntlProvider"));
var _MentionAvatar = require("../MentionAvatar");
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function renderLozenge(lozenge) {
  if (typeof lozenge === 'string') {
    return /*#__PURE__*/_react.default.createElement(_lozenge.default, null, lozenge);
  }
  if ((0, _typeof2.default)(lozenge) === 'object') {
    var appearance = lozenge.appearance,
      text = lozenge.text;
    return /*#__PURE__*/_react.default.createElement(_lozenge.default, {
      appearance: appearance
    }, text);
  }
  return null;
}
function renderTime(time) {
  if (time) {
    return /*#__PURE__*/_react.default.createElement(_styles.TimeStyle, null, time);
  }
  return null;
}
var MentionItem = exports.default = /*#__PURE__*/function (_React$PureComponent) {
  function MentionItem() {
    var _this;
    (0, _classCallCheck2.default)(this, MentionItem);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, MentionItem, [].concat(args));
    // internal, used for callbacks
    (0, _defineProperty2.default)(_this, "onMentionSelected", function (event) {
      if ((0, _mouse.leftClick)(event) && _this.props.onSelection) {
        event.preventDefault();
        _this.props.onSelection(_this.props.mention, event);
      }
    });
    (0, _defineProperty2.default)(_this, "onMentionMenuItemMouseMove", function (event) {
      if (_this.props.onMouseMove) {
        _this.props.onMouseMove(_this.props.mention, event);
      }
    });
    (0, _defineProperty2.default)(_this, "onMentionMenuItemMouseEnter", function (event) {
      if (_this.props.onMouseEnter) {
        _this.props.onMouseEnter(_this.props.mention, event);
      }
    });
    return _this;
  }
  (0, _inherits2.default)(MentionItem, _React$PureComponent);
  return (0, _createClass2.default)(MentionItem, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
        mention = _this$props.mention,
        selected = _this$props.selected;
      var id = mention.id,
        highlight = mention.highlight,
        avatarUrl = mention.avatarUrl,
        presence = mention.presence,
        name = mention.name,
        mentionName = mention.mentionName,
        lozenge = mention.lozenge,
        accessLevel = mention.accessLevel,
        isXProductUser = mention.isXProductUser;
      var _ref = presence || {},
        status = _ref.status,
        time = _ref.time;
      var restricted = (0, _types.isRestricted)(accessLevel);
      var nameHighlights = highlight && highlight.name;
      var borderColor = selected ? "var(--ds-border, ".concat(_colors.N30, ")") : undefined;
      var xProductUserInfoIconColor = selected ? "var(--ds-icon-selected, #0C66E4)" : "var(--ds-icon, #44546F)";
      return /*#__PURE__*/_react.default.createElement(_MessagesIntlProvider.default, null, /*#__PURE__*/_react.default.createElement(_styles.MentionItemStyle, {
        selected: selected,
        onMouseDown: this.onMentionSelected,
        onMouseMove: this.onMentionMenuItemMouseMove,
        onMouseEnter: this.onMentionMenuItemMouseEnter,
        "data-mention-item": true,
        "data-testid": "mention-item-".concat(id),
        "data-mention-id": id,
        "data-mention-name": mentionName,
        "data-selected": selected
      }, /*#__PURE__*/_react.default.createElement(_styles.RowStyle, null, /*#__PURE__*/_react.default.createElement(_styles.AvatarStyle, {
        restricted: restricted
      }, (0, _platformFeatureFlags.fg)('team-avatar-in-mention-picker') ? /*#__PURE__*/_react.default.createElement(_MentionAvatar.MentionAvatar, {
        selected: selected,
        mention: mention
      }) : /*#__PURE__*/_react.default.createElement(_avatar.default, {
        src: avatarUrl,
        size: "medium",
        presence: status,
        borderColor: borderColor
      })), /*#__PURE__*/_react.default.createElement(_styles.NameSectionStyle, {
        restricted: restricted
      }, (0, _MentionHighlightHelpers.renderHighlight)(_styles.FullNameStyle, name, nameHighlights), /*#__PURE__*/_react.default.createElement(_MentionDescriptionByline.default, {
        mention: mention
      })), /*#__PURE__*/_react.default.createElement(_styles.InfoSectionStyle, {
        restricted: restricted
      }, renderLozenge(lozenge), renderTime(time)), restricted ? /*#__PURE__*/_react.default.createElement(_react.default.Suspense, {
        fallback: null
      }, /*#__PURE__*/_react.default.createElement(_NoAccessTooltip.default, {
        name: name
      }, /*#__PURE__*/_react.default.createElement(_styles.AccessSectionStyle, null, /*#__PURE__*/_react.default.createElement(_i18n.NoAccessLabel, null, function (text) {
        return /*#__PURE__*/_react.default.createElement(_LockCircleIcon.default, {
          label: text
        });
      } /* safe to cast to string given there is no value binding */)))) : null, isXProductUser && /*#__PURE__*/_react.default.createElement(_informationEditorPanel.default, {
        color: xProductUserInfoIconColor,
        label: ''
      }))));
    }
  }]);
}(_react.default.PureComponent);
