// @atlaskit/jql-editor/dist/cjs/ui/jql-editor/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactIntlNext = require("react-intl-next");
var _reactMagneticDi = require("react-magnetic-di");
var _analytics = require("../../analytics");
var _useEditorTheme = require("../../hooks/use-editor-theme");
var _state = require("../../state");
var _jqlEditorPortalProvider = require("../jql-editor-portal-provider");
var _jqlEditorView = _interopRequireDefault(require("../jql-editor-view"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
// eslint-disable-next-line @atlassian/tangerine/import/no-parent-imports

// eslint-disable-next-line @atlassian/tangerine/import/no-parent-imports

var JQLEditorInner = function JQLEditorInner(_ref) {
  var analyticsSource = _ref.analyticsSource,
    query = _ref.query,
    _ref$messages = _ref.messages,
    messages = _ref$messages === void 0 ? [] : _ref$messages,
    inputRef = _ref.inputRef,
    isSearching = _ref.isSearching,
    onEditorMounted = _ref.onEditorMounted,
    onDebugUnsafeMessage = _ref.onDebugUnsafeMessage,
    onHydrate = _ref.onHydrate,
    onUpdate = _ref.onUpdate,
    onSearch = _ref.onSearch,
    autocompleteProvider = _ref.autocompleteProvider,
    intl = _ref.intl,
    _ref$enableRichInline = _ref.enableRichInlineNodes,
    enableRichInlineNodes = _ref$enableRichInline === void 0 ? false : _ref$enableRichInline,
    isCompact = _ref.isCompact,
    onSyntaxHelp = _ref.onSyntaxHelp,
    onFocus = _ref.onFocus,
    customComponents = _ref.customComponents;
  var editorTheme = (0, _useEditorTheme.useEditorTheme)({
    isSearch: !!onSearch,
    isCompact: isCompact
  });
  var _useJqlEditorAnalytic = (0, _analytics.useJqlEditorAnalytics)(analyticsSource),
    createAndFireAnalyticsEvent = _useJqlEditorAnalytic.createAndFireAnalyticsEvent;

  // Create and update a mutable ref for our intl object so it can be used by Prosemirror plugins.
  var intlRef = (0, _react.useRef)(intl);
  (0, _react.useEffect)(function () {
    intlRef.current = intl;
  }, [intl]);
  return /*#__PURE__*/_react.default.createElement(_state.EditorStateContainer, {
    query: query,
    externalMessages: messages,
    isSearching: isSearching,
    intlRef: intlRef,
    onEditorMounted: onEditorMounted,
    onDebugUnsafeMessage: onDebugUnsafeMessage,
    onHydrate: onHydrate,
    onUpdate: onUpdate,
    onSearch: onSearch,
    createAndFireAnalyticsEvent: createAndFireAnalyticsEvent,
    autocompleteProvider: autocompleteProvider,
    enableRichInlineNodes: enableRichInlineNodes,
    onSyntaxHelp: onSyntaxHelp,
    onFocus: onFocus,
    customComponents: customComponents
  }, /*#__PURE__*/_react.default.createElement(_useEditorTheme.EditorThemeContext.Provider, {
    value: editorTheme
  }, /*#__PURE__*/_react.default.createElement(_jqlEditorPortalProvider.JQLEditorPortalRenderer, null, /*#__PURE__*/_react.default.createElement(_jqlEditorView.default, {
    inputRef: inputRef
  }))));
};
var _default = exports.default = (0, _reactIntlNext.injectIntl)(JQLEditorInner);
