// @atlaskit/editor-core/dist/cjs/ui/WithEditorActions/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _react = _interopRequireDefault(require("react"));
var _UNSAFE_do_not_use_editor_context = require("@atlaskit/editor-common/UNSAFE_do_not_use_editor_context");
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
// Ignored via go/ees005
// eslint-disable-next-line @repo/internal/react/no-class-components, react/prefer-stateless-function
var WithEditorActions = exports.default = /*#__PURE__*/function (_React$Component) {
  function WithEditorActions(props) {
    (0, _classCallCheck2.default)(this, WithEditorActions);
    return _callSuper(this, WithEditorActions, [props]);
  }
  (0, _inherits2.default)(WithEditorActions, _React$Component);
  return (0, _createClass2.default)(WithEditorActions, [{
    key: "render",
    value: function render() {
      // Ignored via go/ees005
      // eslint-disable-next-line react/jsx-props-no-spreading
      return /*#__PURE__*/_react.default.createElement(WithEditorActionsNew, this.props);
    }
  }]);
}(_react.default.Component);
function WithEditorActionsNew(props) {
  var render = props.render;
  var context = _react.default.useContext(_UNSAFE_do_not_use_editor_context.EditorContext);
  return /*#__PURE__*/_react.default.createElement(WithEditorActionsInner, {
    render: render,
    editorActions: context === null || context === void 0 ? void 0 : context.editorActions
  });
}

// Ignored via go/ees005
// eslint-disable-next-line @repo/internal/react/no-class-components
var WithEditorActionsInner = /*#__PURE__*/function (_React$Component2) {
  function WithEditorActionsInner() {
    (0, _classCallCheck2.default)(this, WithEditorActionsInner);
    return _callSuper(this, WithEditorActionsInner, arguments);
  }
  (0, _inherits2.default)(WithEditorActionsInner, _React$Component2);
  return (0, _createClass2.default)(WithEditorActionsInner, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this = this;
      this.props.editorActions._privateSubscribe(function () {
        return _this.forceUpdate();
      });
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      var _this2 = this;
      this.props.editorActions._privateUnsubscribe(function () {
        return _this2.forceUpdate();
      });
    }
  }, {
    key: "render",
    value: function render() {
      return this.props.render(this.props.editorActions);
    }
  }]);
}(_react.default.Component);
