// @atlaskit/smart-card/dist/es2019/view/FlexibleCard/components/blocks/footer-block/resolved/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import _extends from "@babel/runtime/helpers/extends";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { useCallback, useMemo } from 'react';
import { fg } from '@atlaskit/platform-feature-flags';
import { SmartLinkAlignment, SmartLinkDirection, SmartLinkSize, SmartLinkWidth } from '../../../../../../constants';
import { useFlexibleUiContext } from '../../../../../../state/flexible-ui-context';
import { Provider } from '../../../elements';
import ActionGroup from '../../action-group';
import Block from '../../block';
import ElementGroup from '../../element-group';
import { filterActionItems } from '../../utils';
const actionGroupStylesOld = null;
const actionGroupStyles = null;
const FooterBlockResolvedView = props => {
  const {
    actions,
    testId,
    onActionMenuOpenChange,
    size = SmartLinkSize.Medium,
    hideProvider
  } = props;
  const context = useFlexibleUiContext();
  const hasActions = useMemo(() => {
    var _filterActionItems;
    return ((_filterActionItems = filterActionItems(actions, context)) === null || _filterActionItems === void 0 ? void 0 : _filterActionItems.length) > 0;
  }, [actions, context]);
  const onDropdownOpenChange = useCallback(isOpen => {
    if (onActionMenuOpenChange) {
      onActionMenuOpenChange({
        isOpen
      });
    }
  }, [onActionMenuOpenChange]);
  return /*#__PURE__*/React.createElement(Block, _extends({}, props, {
    testId: `${testId}-resolved-view`
  }), !hideProvider && /*#__PURE__*/React.createElement(Provider, _extends({}, fg('platform-linking-visual-refresh-v1') ? {
    appearance: 'subtle'
  } : {}, {
    testId: `${testId}-provider`
  })), actions && hasActions ? /*#__PURE__*/React.createElement(ElementGroup, _extends({
    testId: "smart-element-group-actions",
    align: SmartLinkAlignment.Right,
    direction: SmartLinkDirection.Horizontal
  }, fg('platform-linking-visual-refresh-v1') ? {
    css: size === SmartLinkSize.XLarge && actionGroupStyles
  } : {
    css: size === SmartLinkSize.XLarge && actionGroupStylesOld
  }, {
    width: SmartLinkWidth.Flexible
  }), /*#__PURE__*/React.createElement(ActionGroup, {
    onDropdownOpenChange: onDropdownOpenChange,
    items: actions,
    appearance: "default",
    size: size
  })) : null);
};
export default FooterBlockResolvedView;
