// @atlaskit/jql-editor/dist/es2019/analytics/listener/jql-editor-analytics-listener.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
/**
 * Adapted from https://bitbucket.org/atlassian/atlassian-frontend/src/master/packages/analytics/analytics-listeners/src/fabric/FabricEditorListener.tsx
 * In future if this package is migrated into the Atlassian Frontend repo, then this code and related logic should be
 * moved into @atlaskit/analytics-listeners.
 */

import React from 'react';
import { AnalyticsListener } from '@atlaskit/analytics-next';
import { ANALYTICS_CHANNEL } from '@atlaskit/jql-editor-common';
import { handleEvent } from './handle-event';
import Logger from './helpers/logger';
// eslint-disable-next-line @repo/internal/react/no-class-components
export default class JQLEditorAnalyticsListener extends React.Component {
  constructor(props) {
    super(props);
    _defineProperty(this, "handleEventWrapper", event => {
      handleEvent(event, this.logger, this.props.client);
    });
    this.logger = new Logger({
      logLevel: props.logLevel
    });
  }
  render() {
    return /*#__PURE__*/React.createElement(AnalyticsListener, {
      onEvent: this.handleEventWrapper,
      channel: ANALYTICS_CHANNEL
    }, this.props.children);
  }
}
