// @atlaskit/teams-public/dist/cjs/common/utils/error/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StatusCode = exports.SLOIgnoreError = exports.HttpError = exports.GraphQLError = exports.DefaultError = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _wrapNativeSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/wrapNativeSuper"));
var _GraphQLError;
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var StatusCode = exports.StatusCode = /*#__PURE__*/function (StatusCode) {
  StatusCode[StatusCode["UNAUTHORIZED"] = 401] = "UNAUTHORIZED";
  StatusCode[StatusCode["FORBIDDEN"] = 403] = "FORBIDDEN";
  StatusCode[StatusCode["TIMEOUT"] = 408] = "TIMEOUT";
  StatusCode[StatusCode["GONE"] = 410] = "GONE";
  StatusCode[StatusCode["PAYLOAD_TOO_LARGE"] = 413] = "PAYLOAD_TOO_LARGE";
  return StatusCode;
}({});
var CommonError = /*#__PURE__*/function (_Error) {
  function CommonError(message) {
    var _this;
    (0, _classCallCheck2.default)(this, CommonError);
    _this = _callSuper(this, CommonError, [message]);
    _this.name = _this.constructor.name;
    _this.message = message || 'UnknownError';
    _this.stack = new Error(message).stack || '';
    return _this;
  }
  (0, _inherits2.default)(CommonError, _Error);
  return (0, _createClass2.default)(CommonError);
}( /*#__PURE__*/(0, _wrapNativeSuper2.default)(Error));
var DefaultError = exports.DefaultError = /*#__PURE__*/function (_CommonError) {
  function DefaultError(_ref) {
    var _this2;
    var message = _ref.message;
    (0, _classCallCheck2.default)(this, DefaultError);
    _this2 = _callSuper(this, DefaultError, [message || 'UnknownError']);
    Object.setPrototypeOf(_this2, DefaultError.prototype);
    return _this2;
  }
  (0, _inherits2.default)(DefaultError, _CommonError);
  return (0, _createClass2.default)(DefaultError);
}(CommonError);
/**
 * These errors will not fail UFO experiences
 */
var SLOIgnoreError = exports.SLOIgnoreError = /*#__PURE__*/function (_CommonError2) {
  function SLOIgnoreError(_ref2) {
    var _this3;
    var message = _ref2.message;
    (0, _classCallCheck2.default)(this, SLOIgnoreError);
    _this3 = _callSuper(this, SLOIgnoreError, ["SentryIgnore: ".concat(message || 'UnknownError')]);
    Object.setPrototypeOf(_this3, SLOIgnoreError.prototype);
    return _this3;
  }
  (0, _inherits2.default)(SLOIgnoreError, _CommonError2);
  return (0, _createClass2.default)(SLOIgnoreError);
}(CommonError);
// Http Errors
var HttpError = exports.HttpError = /*#__PURE__*/function (_CommonError3) {
  function HttpError(_ref3) {
    var _this4;
    var message = _ref3.message,
      status = _ref3.status,
      traceId = _ref3.traceId,
      path = _ref3.path;
    (0, _classCallCheck2.default)(this, HttpError);
    _this4 = _callSuper(this, HttpError, [message]);
    Object.setPrototypeOf(_this4, HttpError.prototype);
    _this4.status = status;
    _this4.traceId = traceId;
    _this4.path = path;
    return _this4;
  }
  (0, _inherits2.default)(HttpError, _CommonError3);
  return (0, _createClass2.default)(HttpError);
}(CommonError); // Graphql Errors
var GraphQLError = exports.GraphQLError = /*#__PURE__*/function (_CommonError4) {
  // tslint:disable-line no-any

  function GraphQLError(_ref4) {
    var _this5;
    var message = _ref4.message,
      _ref4$category = _ref4.category,
      category = _ref4$category === void 0 ? 'default' : _ref4$category,
      fields = _ref4.fields;
    (0, _classCallCheck2.default)(this, GraphQLError);
    _this5 = _callSuper(this, GraphQLError, [message]);
    Object.setPrototypeOf(_this5, GraphQLError.prototype);
    _this5.category = category;
    if (fields) {
      _this5.fields = fields;
    }
    return _this5;
  }
  (0, _inherits2.default)(GraphQLError, _CommonError4);
  return (0, _createClass2.default)(GraphQLError);
}(CommonError);
_GraphQLError = GraphQLError;
(0, _defineProperty2.default)(GraphQLError, "from", function (rawErrors) {
  var firstError = rawErrors[0];
  var errorData = {
    category: firstError.category,
    message: firstError.message
  };
  if (firstError.fields) {
    errorData.fields = firstError.fields.reduce(function (obj, item) {
      obj[item.field] = item.message;
      return obj;
    }, {});
  }
  return new _GraphQLError(errorData);
});
