// @atlaskit/smart-card/dist/cjs/view/FlexibleCard/components/actions/edit-action/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _reactIntlNext = require("react-intl-next");
var _edit = _interopRequireDefault(require("@atlaskit/icon/core/migration/edit"));
var _messages = require("../../../../../messages");
var _action = _interopRequireDefault(require("../action"));
var EditAction = function EditAction(props) {
  return /*#__PURE__*/_react.default.createElement(_action.default, (0, _extends2.default)({
    content: /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, _messages.messages.edit),
    icon: /*#__PURE__*/_react.default.createElement(_edit.default, {
      color: "currentColor",
      spacing: "spacious",
      label: "Edit"
    }),
    testId: "smart-action-edit-action",
    tooltipMessage: /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, _messages.messages.edit)
  }, props));
};
var _default = exports.default = EditAction;
