// @atlaskit/jql-editor/dist/cjs/plugins/rich-inline-nodes/nodes/user/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.user = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireDefault(require("react"));
var _avatar = _interopRequireDefault(require("@atlaskit/avatar"));
var _state = require("../../../../state");
var _styled = require("./styled");
var User = function User(_ref) {
  var id = _ref.id,
    name = _ref.name,
    fieldName = _ref.fieldName,
    selected = _ref.selected,
    error = _ref.error;
  var _useHydratedUser = (0, _state.useHydratedUser)({
      id: id,
      fieldName: fieldName
    }),
    _useHydratedUser2 = (0, _slicedToArray2.default)(_useHydratedUser, 1),
    user = _useHydratedUser2[0];
  return /*#__PURE__*/_react.default.createElement(_styled.UserContainer, {
    selected: selected,
    error: error
  }, /*#__PURE__*/_react.default.createElement(_styled.AvatarWrapper, null, /*#__PURE__*/_react.default.createElement(_avatar.default, {
    src: user === null || user === void 0 ? void 0 : user.avatarUrl,
    borderColor: "transparent",
    size: "xsmall"
  })), /*#__PURE__*/_react.default.createElement(_styled.NameContainer, null, name));
};
var user = exports.user = {
  component: User,
  attrs: {
    id: {},
    name: {},
    fieldName: {}
  }
};
