// @atlaskit/jql-editor/dist/cjs/plugins/rich-inline-nodes/nodes/user/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UserContainer = exports.NameContainer = exports.AvatarWrapper = void 0;
var _react = require("@emotion/react");
var _styled = _interopRequireDefault(require("@emotion/styled"));
var _colors = require("@atlaskit/theme/colors");
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var UserContainer = exports.UserContainer = _styled.default.span({
  display: 'inline-flex',
  alignItems: 'baseline',
  paddingLeft: "var(--ds-space-025, 2px)",
  borderRadius: '10px',
  cursor: 'pointer',
  userSelect: 'none'
},
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-dynamic-styles -- Ignored via go/DSP-18766
function (_ref) {
  var selected = _ref.selected,
    error = _ref.error;
  if (selected) {
    if (error) {
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
      return (0, _react.css)({
        color: "var(--ds-text-inverse, ".concat(_colors.N0, ")"),
        backgroundColor: "var(--ds-background-danger-bold, ".concat(_colors.R400, ")"),
        textDecoration: 'wavy underline',
        textDecorationThickness: '1px',
        textDecorationSkipInk: 'none',
        textDecorationColor: "var(--ds-text-inverse, ".concat(_colors.N0, ")")
      });
    } else {
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
      return (0, _react.css)({
        color: "var(--ds-text, ".concat(_colors.N0, ")"),
        backgroundColor: "var(--ds-background-selected, ".concat(_colors.N500, ")"),
        boxShadow: "0 0 0 1px ".concat("var(--ds-border-selected, transparent)")
      });
    }
  } else {
    if (error) {
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
      return (0, _react.css)({
        color: "var(--ds-text-subtle, ".concat(_colors.N500, ")"),
        backgroundColor: "var(--ds-background-neutral, ".concat(_colors.N40, ")"),
        textDecoration: 'wavy underline',
        textDecorationThickness: '1px',
        textDecorationSkipInk: 'none',
        textDecorationColor: "var(--ds-text-danger, ".concat(_colors.R400, ")"),
        '&:hover': {
          backgroundColor: "var(--ds-background-neutral-hovered, ".concat(_colors.N50, ")")
        }
      });
    } else {
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
      return (0, _react.css)({
        color: "var(--ds-text-subtle, ".concat(_colors.N500, ")"),
        backgroundColor: "var(--ds-background-neutral, ".concat(_colors.N40, ")"),
        '&:hover': {
          backgroundColor: "var(--ds-background-neutral-hovered, ".concat(_colors.N50, ")")
        }
      });
    }
  }
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var NameContainer = exports.NameContainer = _styled.default.span({
  marginLeft: "var(--ds-space-075, 6px)",
  marginRight: "var(--ds-space-100, 8px)",
  // eslint-disable-next-line @atlaskit/design-system/use-tokens-typography
  lineHeight: "var(--ds-space-250, 20px)"
});

/* Override Avatar styles to match design spec */
// eslint-disable-next-line @atlaskit/design-system/ensure-design-token-usage, @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var AvatarWrapper = exports.AvatarWrapper = _styled.default.div({
  height: "var(--ds-space-200, 16px)",
  width: "var(--ds-space-200, 16px)",
  alignSelf: 'center',
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '> div span': {
    margin: "var(--ds-space-0, 0px)"
  }
});
