// @atlaskit/editor-plugin-caption/dist/cjs/captionPlugin.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _adfSchema = require("@atlaskit/adf-schema");
var _keymap = require("./pm-plugins/keymap");
var _main = _interopRequireDefault(require("./pm-plugins/main"));
var captionPlugin = function captionPlugin(_ref) {
  var api = _ref.api;
  return {
    name: 'caption',
    nodes: function nodes() {
      return [{
        name: 'caption',
        node: _adfSchema.caption
      }];
    },
    pmPlugins: function pmPlugins() {
      return [{
        name: 'caption',
        plugin: function plugin(_ref2) {
          var portalProviderAPI = _ref2.portalProviderAPI,
            providerFactory = _ref2.providerFactory,
            eventDispatcher = _ref2.eventDispatcher,
            dispatch = _ref2.dispatch;
          return (0, _main.default)(portalProviderAPI, eventDispatcher, providerFactory, dispatch, api);
        }
      }, {
        name: 'captionKeymap',
        plugin: _keymap.captionKeymap
      }];
    }
  };
};
var _default = exports.default = captionPlugin;
