// @atlaskit/link-datasource/dist/cjs/ui/jira-issues-modal/basic-filters/ui/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.availableBasicFilterTypes = void 0;
require("./index.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _react = _interopRequireWildcard(require("react"));
var _compiled = require("@atlaskit/primitives/compiled");
var _extractValuesFromNonComplexJQL = require("../utils/extractValuesFromNonComplexJQL");
var _asyncPopupSelect = _interopRequireDefault(require("./async-popup-select"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var availableBasicFilterTypes = exports.availableBasicFilterTypes = ['project', 'type', 'status', 'assignee'];
var styles = {
  basicFilterContainerStyles: "_19bvu2gc"
};
var BasicFilterContainer = function BasicFilterContainer(_ref) {
  var jql = _ref.jql,
    site = _ref.site,
    onChange = _ref.onChange,
    selections = _ref.selections,
    isJQLHydrating = _ref.isJQLHydrating;
  var extractedFilterValues = (0, _react.useMemo)(function () {
    return isJQLHydrating ? (0, _extractValuesFromNonComplexJQL.extractValuesFromNonComplexJQL)(jql) : {};
  }, [isJQLHydrating, jql]);
  var _ref2 = site || {},
    cloudId = _ref2.cloudId;
  return /*#__PURE__*/_react.default.createElement(_compiled.Flex, {
    xcss: styles.basicFilterContainerStyles,
    gap: "space.100",
    testId: "jlol-basic-filter-container"
  }, availableBasicFilterTypes.map(function (filter) {
    var _extractedFilterValue;
    var shouldShowHydrationLoader = isJQLHydrating && ((_extractedFilterValue = extractedFilterValues[filter]) === null || _extractedFilterValue === void 0 ? void 0 : _extractedFilterValue.length) > 0;
    return /*#__PURE__*/_react.default.createElement(_asyncPopupSelect.default, {
      site: site,
      filterType: filter,
      key: filter,
      selection: selections[filter] || [],
      isJQLHydrating: shouldShowHydrationLoader,
      isDisabled: !cloudId,
      onSelectionChange: onChange
    });
  }));
};
var _default = exports.default = BasicFilterContainer;
