// @atlaskit/task-decision/dist/cjs/components/DecisionItem.js
/* DecisionItem.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./DecisionItem.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _decisionEditorDecision = _interopRequireDefault(require("@atlaskit/icon/core/migration/decision--editor-decision"));
var _Item = _interopRequireDefault(require("./Item"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var iconStyles = null;
var iconStylesWithPlaceholder = null;
var DecisionItem = function DecisionItem(_ref) {
  var appearance = _ref.appearance,
    children = _ref.children,
    contentRef = _ref.contentRef,
    placeholder = _ref.placeholder,
    showPlaceholder = _ref.showPlaceholder,
    dataAttributes = _ref.dataAttributes;
  var icon = /*#__PURE__*/React.createElement("span", {
    contentEditable: false,
    className: (0, _runtime.ax)(["_16jlidpf _1o9zidpf _i0dlexct _4t3i7vkz _1bsb7vkz _19pk1b66 _2hwxutpp _otyridpf _18u0idpf _syaz1yjp _1e0c1txw _4cvr1h6o _1bah1h6o", showPlaceholder && "_syaz1oju"])
  }, /*#__PURE__*/React.createElement(_decisionEditorDecision.default, {
    label: "Decision",
    spacing: "spacious",
    color: "currentColor",
    LEGACY_size: "large"
  }));
  return /*#__PURE__*/React.createElement(_Item.default, {
    appearance: appearance,
    contentRef: contentRef,
    icon: icon,
    placeholder: placeholder,
    showPlaceholder: showPlaceholder,
    itemType: "DECISION",
    dataAttributes: dataAttributes
  }, children);
};
var _default = exports.default = DecisionItem;
