// @atlaskit/task-decision/dist/cjs/components/Item.js
/* Item.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./Item.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var placeholderStyles = null;
var placeholderTaskStyles = null;
var decisionStyles = null;
var contentStyles = null;
var taskStyles = null;
var Item = exports.default = /*#__PURE__*/function (_PureComponent) {
  function Item() {
    (0, _classCallCheck2.default)(this, Item);
    return _callSuper(this, Item, arguments);
  }
  (0, _inherits2.default)(Item, _PureComponent);
  return (0, _createClass2.default)(Item, [{
    key: "renderPlaceholder",
    value: function renderPlaceholder() {
      var _this$props = this.props,
        children = _this$props.children,
        placeholder = _this$props.placeholder,
        showPlaceholder = _this$props.showPlaceholder,
        itemType = _this$props.itemType;
      if (!showPlaceholder || !placeholder || children) {
        return null;
      }
      return /*#__PURE__*/React.createElement("span", {
        "data-testid": "task-decision-item-placeholder",
        "data-component": "placeholder",
        contentEditable: false,
        className: (0, _runtime.ax)(["_18s81hzh _1reo15vq _18m915vq _kqswstnw _syaz131l _lcxvglyw _1bto1l2s _o5721q9c _p12f1dza", itemType === 'TASK' && "_18s815t0"]),
        style: {
          "--_tptct2": (0, _runtime.ix)("0 0 0 calc(".concat("var(--ds-space-100, 8px)", " * 3.5)")),
          "--_1yfnyhr": (0, _runtime.ix)("0 0 0 calc(".concat("var(--ds-space-100, 8px)", " * 3)"))
        }
      }, placeholder);
    }
  }, {
    key: "renderMessageAppearance",
    value: function renderMessageAppearance() {
      var _this$props2 = this.props,
        contentRef = _this$props2.contentRef,
        children = _this$props2.children,
        icon = _this$props2.icon,
        itemType = _this$props2.itemType,
        checkBoxId = _this$props2.checkBoxId,
        dataAttributes = _this$props2.dataAttributes;
      if (itemType === 'TASK') {
        return /*#__PURE__*/React.createElement("div", {
          id: "".concat(checkBoxId, "-wrapper"),
          className: (0, _runtime.ax)(["_1e0c1txw _2lx2vrvc _kqswh2mm"])
        }, icon, this.renderPlaceholder(), /*#__PURE__*/React.createElement("div", (0, _extends2.default)({
          "data-component": "content",
          ref: contentRef
        }, dataAttributes, {
          className: (0, _runtime.ax)(["_19pkidpf _2hwxidpf _otyridpf _18u0idpf _16jlkb7n _1o9zkb7n _i0dl1wug _slp31hna _1ul9idpf"])
        }), children));
      } else if (itemType === 'DECISION') {
        return /*#__PURE__*/React.createElement("div", {
          "data-testid": "elements-decision-item",
          "data-decision-wrapper": "true",
          className: (0, _runtime.ax)(["_2rkosqtm _1e0c1txw _2lx2vrvc _19pku2gc _2hwxidpf _otyridpf _18u0idpf _ca0qu2gc _u5f3u2gc _n3tdu2gc _19bvutpp _bfhkm7j4 _kqswh2mm _14hd1wug"])
        }, icon, this.renderPlaceholder(), /*#__PURE__*/React.createElement("div", (0, _extends2.default)({
          "data-component": "content",
          ref: contentRef
        }, dataAttributes, {
          className: (0, _runtime.ax)(["_19pkidpf _2hwxidpf _otyridpf _18u0idpf _16jlkb7n _1o9zkb7n _i0dl1wug _slp31hna _1ul9idpf"])
        }), children));
      }
      return null;
    }
  }, {
    key: "render",
    value: function render() {
      return this.renderMessageAppearance();
    }
  }]);
}(_react.PureComponent);
(0, _defineProperty2.default)(Item, "defaultProps", {
  appearance: 'inline'
});
