// @atlaskit/task-decision/dist/cjs/components/TaskList.js
/* TaskList.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./TaskList.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _react = _interopRequireWildcard(require("react"));
var _analyticsNamespacedContext = require("@atlaskit/analytics-namespaced-context");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var listStyles = null;
var taskListStyles = null;
var TaskList = exports.default = /*#__PURE__*/function (_PureComponent) {
  function TaskList() {
    (0, _classCallCheck2.default)(this, TaskList);
    return _callSuper(this, TaskList, arguments);
  }
  (0, _inherits2.default)(TaskList, _PureComponent);
  return (0, _createClass2.default)(TaskList, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
        listId = _this$props.listId,
        children = _this$props.children;
      var listSize = _react.default.Children.count(children);
      if (!children) {
        return null;
      }

      // Data attributes are required for copy and paste from rendered content
      // to the editor to preserve the task.
      // This allows the editor to differentiate between numbered and ordered lists,
      // and action items, which all share the common `<li>` element.
      // The value of `data-task-local-id` should be discarded upon paste, with a
      // a new uuid generated by the editor for the cloned content.
      return /*#__PURE__*/_react.default.createElement("div", {
        "data-task-list-local-id": "",
        className: (0, _runtime.ax)(["_2mzuglyw _19bvidpf"])
      }, _react.default.Children.map(children, function (child, idx) {
        var _ref = child.props,
          localId = _ref.localId;
        return /*#__PURE__*/_react.default.createElement(_analyticsNamespacedContext.FabricElementsAnalyticsContext, {
          data: {
            listLocalId: listId,
            listSize: listSize,
            position: idx
          }
        }, /*#__PURE__*/_react.default.createElement("div", {
          key: idx,
          "data-task-local-id": localId || '',
          className: (0, _runtime.ax)(["_19pk1b66"])
        }, child));
      }));
    }
  }]);
}(_react.PureComponent);
