// @atlaskit/smart-card/dist/cjs/view/FlexibleCard/components/blocks/footer-block/resolved/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _constants = require("../../../../../../constants");
var _flexibleUiContext = require("../../../../../../state/flexible-ui-context");
var _elements = require("../../../elements");
var _actionGroup = _interopRequireDefault(require("../../action-group"));
var _block = _interopRequireDefault(require("../../block"));
var _elementGroup = _interopRequireDefault(require("../../element-group"));
var _utils = require("../../utils");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var actionGroupStylesOld = null;
var actionGroupStyles = null;
var FooterBlockResolvedView = function FooterBlockResolvedView(props) {
  var actions = props.actions,
    testId = props.testId,
    onActionMenuOpenChange = props.onActionMenuOpenChange,
    _props$size = props.size,
    size = _props$size === void 0 ? _constants.SmartLinkSize.Medium : _props$size,
    hideProvider = props.hideProvider;
  var context = (0, _flexibleUiContext.useFlexibleUiContext)();
  var hasActions = (0, _react.useMemo)(function () {
    var _filterActionItems;
    return ((_filterActionItems = (0, _utils.filterActionItems)(actions, context)) === null || _filterActionItems === void 0 ? void 0 : _filterActionItems.length) > 0;
  }, [actions, context]);
  var onDropdownOpenChange = (0, _react.useCallback)(function (isOpen) {
    if (onActionMenuOpenChange) {
      onActionMenuOpenChange({
        isOpen: isOpen
      });
    }
  }, [onActionMenuOpenChange]);
  return /*#__PURE__*/React.createElement(_block.default, (0, _extends2.default)({}, props, {
    testId: "".concat(testId, "-resolved-view")
  }), !hideProvider && /*#__PURE__*/React.createElement(_elements.Provider, (0, _extends2.default)({}, (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v1') ? {
    appearance: 'subtle'
  } : {}, {
    testId: "".concat(testId, "-provider")
  })), actions && hasActions ? /*#__PURE__*/React.createElement(_elementGroup.default, (0, _extends2.default)({
    testId: "smart-element-group-actions",
    align: _constants.SmartLinkAlignment.Right,
    direction: _constants.SmartLinkDirection.Horizontal
  }, (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v1') ? {
    css: size === _constants.SmartLinkSize.XLarge && actionGroupStyles
  } : {
    css: size === _constants.SmartLinkSize.XLarge && actionGroupStylesOld
  }, {
    width: _constants.SmartLinkWidth.Flexible
  }), /*#__PURE__*/React.createElement(_actionGroup.default, {
    onDropdownOpenChange: onDropdownOpenChange,
    items: actions,
    appearance: "default",
    size: size
  })) : null);
};
var _default = exports.default = FooterBlockResolvedView;
