// @atlaskit/link-extractors/dist/cjs/common/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.JIRA_TASK = exports.JIRA_SUB_TASK = exports.JIRA_STORY = exports.JIRA_SERVICE_REQUEST = exports.JIRA_PROBLEM = exports.JIRA_INCIDENT = exports.JIRA_GENERATOR_ID = exports.JIRA_EPIC = exports.JIRA_CUSTOM_TASK_TYPE = exports.JIRA_CHANGE = exports.JIRA_BUG = exports.CONFLUENCE_GENERATOR_ID = void 0;
var CONFLUENCE_GENERATOR_ID = exports.CONFLUENCE_GENERATOR_ID = 'https://www.atlassian.com/#Confluence';
var JIRA_GENERATOR_ID = exports.JIRA_GENERATOR_ID = 'https://www.atlassian.com/#Jira';
// To support icons for all Jira task types.
var JIRA_TASK = exports.JIRA_TASK = 'JiraTask';
var JIRA_SUB_TASK = exports.JIRA_SUB_TASK = 'JiraSubTask';
var JIRA_STORY = exports.JIRA_STORY = 'JiraStory';
var JIRA_BUG = exports.JIRA_BUG = 'JiraBug';
var JIRA_EPIC = exports.JIRA_EPIC = 'JiraEpic';
var JIRA_INCIDENT = exports.JIRA_INCIDENT = 'JiraIncident';
var JIRA_SERVICE_REQUEST = exports.JIRA_SERVICE_REQUEST = 'JiraServiceRequest';
var JIRA_CHANGE = exports.JIRA_CHANGE = 'JiraChange';
var JIRA_PROBLEM = exports.JIRA_PROBLEM = 'JiraProblem';
var JIRA_CUSTOM_TASK_TYPE = exports.JIRA_CUSTOM_TASK_TYPE = 'JiraCustomTaskType';
