// @atlaskit/editor-plugin-find-replace/dist/es2019/ui/FindReplaceDropdown.js
import _extends from "@babel/runtime/helpers/extends";
import React from 'react';
import { injectIntl } from 'react-intl-next';
import { TRIGGER_METHOD } from '@atlaskit/editor-common/analytics';
import { ArrowKeyNavigationType, Dropdown } from '@atlaskit/editor-common/ui-menu';
import { akEditorFloatingPanelZIndex } from '@atlaskit/editor-shared-styles';
import { Box, xcss } from '@atlaskit/primitives';
import FindReplace from './FindReplace';
const findReplaceWrapperStyles = xcss({
  display: 'flex',
  flexDirection: 'column'
});

// Magic number taken from ../FindReplaceToolbarButton.tsx
const dropdownWidthNewDesign = 382;
const FindReplaceDropdown = props => {
  const {
    findText,
    replaceText,
    isActive,
    index,
    numMatches,
    popupsMountPoint,
    onCancel
  } = props;
  if (!popupsMountPoint) {
    return null;
  }
  return /*#__PURE__*/React.createElement(Dropdown, {
    target: popupsMountPoint,
    mountTo: popupsMountPoint,
    forcePlacement: true,
    alignX: 'right',
    alignY: 'start',
    offset: [4, 0],
    isOpen: isActive,
    handleEscapeKeydown: () => {
      if (isActive) {
        onCancel({
          triggerMethod: TRIGGER_METHOD.KEYBOARD
        });
      }
    },
    fitWidth: dropdownWidthNewDesign,
    zIndex: akEditorFloatingPanelZIndex,
    arrowKeyNavigationProviderOptions: {
      type: ArrowKeyNavigationType.MENU,
      disableArrowKeyNavigation: true
    }
  }, /*#__PURE__*/React.createElement(Box, {
    xcss: findReplaceWrapperStyles
  }, /*#__PURE__*/React.createElement(FindReplace, _extends({
    findText: findText,
    replaceText: replaceText,
    count: {
      index,
      total: numMatches
    }
    // Ignored via go/ees005
    // eslint-disable-next-line react/jsx-props-no-spreading
  }, props))));
};
export default injectIntl(FindReplaceDropdown);
