// @atlaskit/react-ufo/dist/cjs/vc/vc-observer-new/metric-calculator/abstract-base-vc-calculator.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _selfMeasurements = require("../../../self-measurements");
var _percentileCalc = _interopRequireDefault(require("./percentile-calc"));
var _getViewportHeight = _interopRequireDefault(require("./utils/get-viewport-height"));
var _getViewportWidth = _interopRequireDefault(require("./utils/get-viewport-width"));
var AbstractVCCalculatorBase = exports.default = /*#__PURE__*/function () {
  function AbstractVCCalculatorBase(revisionNo) {
    (0, _classCallCheck2.default)(this, AbstractVCCalculatorBase);
    this.revisionNo = revisionNo;
    this.calculate = (0, _selfMeasurements.withProfiling)(this.calculate.bind(this), ['vc']);
  }
  return (0, _createClass2.default)(AbstractVCCalculatorBase, [{
    key: "calculate",
    value: function () {
      var _calculate = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(_ref) {
        var _this = this,
          _vcDetails$90$t,
          _vcDetails$;
        var startTime, stopTime, orderedEntries, filteredEntries, isVCClean, vcDetails;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              startTime = _ref.startTime, stopTime = _ref.stopTime, orderedEntries = _ref.orderedEntries;
              filteredEntries = orderedEntries.filter(function (entry) {
                return _this.isEntryIncluded(entry);
              });
              isVCClean = this.isVCClean(filteredEntries);
              if (isVCClean) {
                _context.next = 5;
                break;
              }
              return _context.abrupt("return", {
                revision: this.revisionNo,
                'metric:vc90': null,
                clean: false
              });
            case 5:
              _context.next = 7;
              return (0, _percentileCalc.default)({
                viewport: {
                  width: (0, _getViewportWidth.default)(),
                  height: (0, _getViewportHeight.default)()
                },
                startTime: startTime,
                stopTime: stopTime,
                orderedEntries: filteredEntries,
                percentiles: [25, 50, 75, 80, 85, 90, 95, 98, 99]
              });
            case 7:
              vcDetails = _context.sent;
              return _context.abrupt("return", {
                revision: this.revisionNo,
                vcDetails: vcDetails !== null && vcDetails !== void 0 ? vcDetails : undefined,
                clean: isVCClean,
                'metric:vc90': (_vcDetails$90$t = vcDetails === null || vcDetails === void 0 || (_vcDetails$ = vcDetails['90']) === null || _vcDetails$ === void 0 ? void 0 : _vcDetails$.t) !== null && _vcDetails$90$t !== void 0 ? _vcDetails$90$t : null
              });
            case 9:
            case "end":
              return _context.stop();
          }
        }, _callee, this);
      }));
      function calculate(_x) {
        return _calculate.apply(this, arguments);
      }
      return calculate;
    }()
  }]);
}();
