// @atlaskit/code/dist/es2019/code.js
/* code.tsx generated by @compiled/babel-plugin v0.36.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./code.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
import React, { forwardRef, memo } from 'react';
import CodeBidiWarning from './bidi-warning';
import codeBidiWarningDecorator from './bidi-warning/bidi-warning-decorator';
import { getCodeStyles } from './internal/theme/styles';
const styles = {
  base: "_ca0qyh40 _u5f3m5ip _n3tdyh40 _19bvm5ip _2rko1sit _11c81u0j _1reo1wug _18m91wug _1dqoglyw _1e0c1nu9 _bfhktkvp _16d9qvcn _syaz1fxt _vwz41kw7 _1i4q1hna _o5721jtm"
};

/**
 * __Code__
 *
 * Code highlights short strings of code snippets inline with body text.
 *
 * - [Examples](https://atlassian.design/components/code/examples)
 * - [Code](https://atlassian.design/components/code/code)
 * - [Usage](https://atlassian.design/components/code/usage)
 */
const Code = /*#__PURE__*/memo( /*#__PURE__*/forwardRef(function Code({
  testId,
  ...props
}, ref) {
  // eslint-disable-next-line @atlaskit/platform/ensure-feature-flag-prefix

  const {
    children,
    codeBidiWarnings = true,
    codeBidiWarningLabel,
    codeBidiWarningTooltipEnabled = true,
    ...otherProps
  } = props;
  const decoratedChildren = codeBidiWarnings ? /*#__PURE__*/React.createElement(RenderCodeChildrenWithBidiWarnings, {
    codeBidiWarningLabel: codeBidiWarningLabel,
    codeBidiWarningTooltipEnabled: codeBidiWarningTooltipEnabled
  }, children) : children;
  return /*#__PURE__*/React.createElement("code", _extends({
    ref: ref,
    "data-testid": testId
  }, otherProps, {
    className: ax([styles.base])
  }), decoratedChildren);
}));
function RenderCodeChildrenWithBidiWarnings({
  children,
  codeBidiWarningLabel,
  codeBidiWarningTooltipEnabled
}) {
  const replacedChildren = React.Children.map(children, childNode => {
    if (typeof childNode === 'string') {
      const decorated = codeBidiWarningDecorator(childNode, ({
        bidiCharacter,
        index
      }) => /*#__PURE__*/React.createElement(CodeBidiWarning, {
        bidiCharacter: bidiCharacter,
        key: index,
        label: codeBidiWarningLabel,
        tooltipEnabled: codeBidiWarningTooltipEnabled
      }));
      return decorated;
    }
    if (isReactElement(childNode) && childNode.props.children) {
      // eslint-disable-next-line @repo/internal/react/no-clone-element
      const newChildNode = /*#__PURE__*/React.cloneElement(childNode, {
        children: /*#__PURE__*/React.createElement(RenderCodeChildrenWithBidiWarnings, {
          codeBidiWarningLabel: codeBidiWarningLabel,
          codeBidiWarningTooltipEnabled: codeBidiWarningTooltipEnabled
        }, childNode.props.children)
      });
      return newChildNode;
    }
    return childNode;
  });
  return /*#__PURE__*/React.createElement(React.Fragment, null, replacedChildren);
}
function isReactElement(child) {
  return !!child.type;
}
Code.displayName = 'Code';
export { getCodeStyles };
export default Code;
