// @atlaskit/editor-core/dist/es2019/ui/WithFlash/index.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
import React from 'react';

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import { css, jsx, keyframes } from '@emotion/react';
const pulseBackground = keyframes({
  '50%': {
    backgroundColor: "var(--ds-blanket-danger, #EF5C4814)"
  }
});
const pulseBackgroundReverse = keyframes({
  '0%': {
    backgroundColor: "var(--ds-blanket-danger, #EF5C4814)"
  },
  '50%': {
    backgroundColor: 'auto'
  },
  '100%': {
    backgroundColor: "var(--ds-blanket-danger, #EF5C4814)"
  }
});
const flashWrapper = css({
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '&.-flash > div': {
    animation: `0.25s ease-in-out ${pulseBackgroundReverse}`
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '& > div': {
    animation: "'none'"
  }
});
const flashWrapperAnimated = css({
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  [`${flashWrapper} & > div`]: {
    animation: `0.25s ease-in-out ${pulseBackground}`
  }
});
// Ignored via go/ees005
// eslint-disable-next-line @repo/internal/react/no-class-components
export default class WithFlash extends React.Component {
  constructor(...args) {
    super(...args);
    _defineProperty(this, "toggle", false);
  }
  render() {
    const {
      animate,
      children
    } = this.props;
    this.toggle = animate && !this.toggle;
    return (
      // eslint-disable-next-line @atlaskit/design-system/prefer-primitives
      jsx("div", {
        css: animate ? flashWrapperAnimated : flashWrapper
        // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
        ,
        className: this.toggle ? '-flash' : ''
      }, children)
    );
  }
}
