// @atlaskit/editor-plugin-media/dist/es2019/ui/ToolbarMedia/index.js
import React from 'react';
import { injectIntl } from 'react-intl-next';
import { useSharedPluginState } from '@atlaskit/editor-common/hooks';
import { toolbarMediaMessages } from '@atlaskit/editor-common/media';
import { TOOLBAR_BUTTON, ToolbarButton } from '@atlaskit/editor-common/ui-menu';
import AttachmentIcon from '@atlaskit/icon/core/migration/attachment--editor-attachment';
const onClickMediaButton = pluginState => () => {
  pluginState.showMediaPicker();
  return true;
};
const ToolbarMedia = ({
  isDisabled,
  isReducedSpacing,
  intl,
  api
}) => {
  const {
    mediaState
  } = useSharedPluginState(api, ['media']);
  if (!(mediaState !== null && mediaState !== void 0 && mediaState.allowsUploads)) {
    return null;
  }
  const {
    toolbarMediaTitle
  } = toolbarMediaMessages;
  return /*#__PURE__*/React.createElement(ToolbarButton, {
    buttonId: TOOLBAR_BUTTON.MEDIA,
    onClick: onClickMediaButton(mediaState),
    disabled: isDisabled,
    title: intl.formatMessage(toolbarMediaTitle),
    spacing: isReducedSpacing ? 'none' : 'default',
    iconBefore: /*#__PURE__*/React.createElement(AttachmentIcon, {
      label: intl.formatMessage(toolbarMediaTitle)
    })
  });
};
export default injectIntl(ToolbarMedia);
