// @atlaskit/media-viewer/dist/es2019/viewers/archiveSidebar/archive-sidebar-header.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import React from 'react';
import ArrowLeftIcon from '@atlaskit/icon/core/migration/arrow-left';
import { ButtonItem } from '@atlaskit/side-navigation';
import HomeIcon from '@atlaskit/icon/core/migration/home';
import { getFormattedFolderName } from '../../utils';
import { SidebarHeaderEntry, SidebarHeaderIcon, SidebarHeaderWrapper } from './styleWrappers';
import { itemStyle } from './styles';
export class ArchiveSidebarHeader extends React.Component {
  constructor(...args) {
    super(...args);
    _defineProperty(this, "getHeaderIcon", () => !!this.props.folderName ? /*#__PURE__*/React.createElement(ArrowLeftIcon, {
      color: "currentColor",
      spacing: "spacious",
      label: "Back"
    }) : /*#__PURE__*/React.createElement(HomeIcon, {
      color: "currentColor",
      spacing: "spacious",
      label: "Home"
    }));
  }
  render() {
    const {
      folderName,
      onHeaderClick
    } = this.props;
    return (
      /*#__PURE__*/
      // eslint-disable-next-line @atlaskit/design-system/no-deprecated-apis
      React.createElement(ButtonItem, {
        onClick: onHeaderClick,
        cssFn: () => itemStyle
      }, /*#__PURE__*/React.createElement(SidebarHeaderWrapper, null, /*#__PURE__*/React.createElement(SidebarHeaderIcon, null, this.getHeaderIcon()), /*#__PURE__*/React.createElement(SidebarHeaderEntry, null, getFormattedFolderName(folderName))))
    );
  }
}
