// @atlaskit/media-viewer/dist/es2019/viewers/archiveSidebar/styleWrappers-compiled.js
/* styleWrappers-compiled.tsx generated by @compiled/babel-plugin v0.36.1 */
import "./styleWrappers-compiled.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { forwardRef } from 'react';
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

import { DN500 } from '@atlaskit/theme/colors';
export const ARCHIVE_SIDE_BAR_WIDTH = 300;
const archiveItemViewerWrapperStyles = null;
const archiveSideBarStyles = null;
const slideDown = null;
const archiveDownloadButtonWrapperStyles = null;
const disabledArchiveDownloadButtonWrapperStyles = null;
const archiveSidebarFolderWrapperStyles = null;
const sidebarItemWrapperStyles = null;
const archiveSidebarFileEntryWrapperStyles = null;
const archiveLayoutStyles = null;
const archiveViewerWrapperStyles = null;
const separatorStyles = null;
const sidebarHeaderWrapperStyles = null;
const sidebarHeaderIconStyles = null;
const sidebarHeaderEntryStyles = null;
export const ArchiveItemViewerWrapper = ({
  children
}) => {
  return /*#__PURE__*/React.createElement("div", {
    className: ax(["_1bsb1osq _1e0c1txw _1bah1h6o"])
  }, children);
};
export const ArchiveSideBar = /*#__PURE__*/forwardRef(({
  children
}, ref) => {
  return /*#__PURE__*/React.createElement("div", {
    ref: ref,
    className: ax(["_ca0qqslr _u5f3v47k _19bvv47k _n3tdv47k _bfhky0hh _kqswstnw _1ltvidpf _154iidpf _1bsb1mjv _94n5idpf _vchhusvi _18m915zi"])
  }, children);
});
export const ArchiveSidebarFolderWrapper = ({
  children
}) => {
  return /*#__PURE__*/React.createElement("div", {
    "data-testid": "archive-sidebar-folder-wrapper",
    className: ax(["_v564t24z _y44vau1n _t9ecz6y5 _tzy4idpf"])
  }, children);
};
export const ArchiveDownloadButtonWrapper = ({
  children,
  onClick
}) => {
  return /*#__PURE__*/React.createElement("div", {
    onClick: onClick,
    "data-testid": "media-archiveDownloadButton",
    className: ax(["_19itglyw _2rko1l7b _ca0qu2gc _u5f31v6z _n3td14y2 _19bvu2gc _bfhk1j28 _syazwydc _d0altlke _irr3ebx2 _1qamtlke _1di6j4ot"])
  }, children);
};
export const DisabledArchiveDownloadButtonWrapper = ({
  children
}) => {
  return /*#__PURE__*/React.createElement("div", {
    "data-testid": "media-disabledArchiveDownloadButton",
    className: ax(["_19itglyw _2rko1l7b _ca0qu2gc _u5f31v6z _n3td14y2 _19bvu2gc _bfhk1j28 _syazwydc _80om13gf"])
  }, children);
};
export const SidebarItemWrapper = ({
  children
}) => {
  return /*#__PURE__*/React.createElement("div", {
    className: ax(["_1bsb13sl"])
  }, children);
};
export const ArchiveSidebarFileEntryWrapper = ({
  children,
  index
}) => {
  return /*#__PURE__*/React.createElement("div", {
    key: index,
    className: ax(["_v564rca0 _n3td1vik _1e0c1txw _4cvr1h6o _80omtlke"])
  }, children);
};
export const ArchiveLayout = ({
  children
}) => {
  return /*#__PURE__*/React.createElement("div", {
    "data-testid": "archive-layout",
    className: ax(["_1e0c1txw _1bsb1osq _4t3i1osq"])
  }, children);
};
export const ArchiveViewerWrapper = ({
  children
}) => {
  return /*#__PURE__*/React.createElement("div", {
    className: ax(["_kqswstnw _154iidpf _1ltv1mjv _1xi2idpf _94n5idpf _4cvr1h6o _1e0c1txw"])
  }, children);
};
export const Separator = () => {
  return /*#__PURE__*/React.createElement("div", {
    className: ax(["_2rkot94y _4t3iyh40 _19pk1te7 _2hwxidpf _otyr1te7 _18u0idpf _bfhk21we _1o9zidpf"])
  });
};
export const SidebarHeaderWrapper = ({
  children
}) => {
  return /*#__PURE__*/React.createElement("span", {
    className: ax(["_1e0c1txw _4cvr1h6o _1o9zidpf"])
  }, children);
};
export const SidebarHeaderIcon = ({
  children
}) => {
  return /*#__PURE__*/React.createElement("div", {
    className: ax(["_1e0c1txw _4cvr1h6o _2hwx1yov _1o9zidpf"])
  }, children);
};
export const SidebarHeaderEntry = ({
  children
}) => {
  return /*#__PURE__*/React.createElement("div", {
    className: ax(["_16jlkb7n _1o9zkb7n _i0dl1wug _1reo15vq _18m915vq _1bto1l2s _o5721q9c _vwz41xde _syazg8nh"])
  }, children);
};
