// @atlaskit/media-viewer/dist/es2019/viewers/archiveSidebar/styleWrappers-emotion.js
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
import { forwardRef } from 'react';
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import { jsx, css, keyframes } from '@emotion/react';
import { ArchiveSideBarWidth } from './styles';
import { DN500 } from '@atlaskit/theme/colors';
const archiveItemViewerWrapperStyles = css({
  width: '100%',
  display: 'flex',
  justifyContent: 'center'
});
const archiveSideBarStyles = css({
  padding: `22px ${"var(--ds-space-250, 20px)"} ${"var(--ds-space-250, 20px)"} ${"var(--ds-space-250, 20px)"}`,
  backgroundColor: "var(--ds-surface, #101214)",
  position: 'absolute',
  left: 0,
  top: 0,
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  width: `${ArchiveSideBarWidth}px`,
  bottom: 0,
  boxSizing: 'border-box',
  overflowY: 'scroll'
});
const slideDown = keyframes({
  '0%': {
    opacity: 0,
    transform: 'translateY(-100%)'
  },
  '100%': {
    transform: 'translateY(0)',
    opacity: 1
  }
});
const archiveDownloadButtonWrapperStyles = css({
  padding: `${"var(--ds-space-100, 8px)"} 7px 5px ${"var(--ds-space-100, 8px)"}`,
  border: 'none',
  borderRadius: '3px',
  backgroundColor: 'transparent',
  color: "var(--ds-icon, #9FADBC)",
  '&:hover': {
    cursor: 'pointer',
    backgroundColor: "var(--ds-background-neutral-subtle-hovered, #A1BDD914)"
  },
  '&:active': {
    cursor: 'pointer',
    backgroundColor: "var(--ds-background-neutral-subtle-pressed, #A6C5E229)"
  }
});
const disabledArchiveDownloadButtonWrapperStyles = css({
  padding: `${"var(--ds-space-100, 8px)"} 7px 5px ${"var(--ds-space-100, 8px)"}`,
  border: 'none',
  borderRadius: '3px',
  backgroundColor: 'transparent',
  color: "var(--ds-icon, #9FADBC)",
  cursor: 'not-allowed'
});
const archiveSidebarFolderWrapperStyles = css({
  transform: 'translateY(-100%)',
  transition: 'all 1s',
  opacity: 0,
  animation: `${slideDown} 0.3s forwards`
});
const sidebarItemWrapperStyles = css({
  width: '85%'
});
const archiveSidebarFileEntryWrapperStyles = css({
  paddingBottom: "var(--ds-space-075, 5px)",
  display: 'flex',
  alignItems: 'center',
  cursor: 'pointer',
  transition: 'background-color 0.3s'
});
const archiveLayoutStyles = css({
  display: 'flex',
  width: '100%',
  height: '100%'
});
const archiveViewerWrapperStyles = css({
  position: 'absolute',
  top: 0,
  // eslint-disable-next-line @atlaskit/design-system/ensure-design-token-usage, @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  left: `${ArchiveSideBarWidth}px`,
  right: 0,
  bottom: 0,
  alignItems: 'center',
  display: 'flex'
});
const separatorStyles = css({
  borderRadius: '1px',
  height: '2px',
  margin: `${"var(--ds-space-200, 19px)"} 0`,
  backgroundColor: "var(--ds-border, #A6C5E229)",
  flexShrink: 0
});
const sidebarHeaderWrapperStyles = css({
  display: 'flex',
  alignItems: 'center',
  flexShrink: 0
});
const sidebarHeaderIconStyles = css({
  display: 'flex',
  alignItems: 'center',
  marginRight: "var(--ds-space-100, 10px)",
  flexShrink: 0
});
const sidebarHeaderEntryStyles = css({
  flex: '1 1 auto',
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap',
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values
  ['lineHeight']: 1.14286,
  color: `var(--ds-text, ${DN500})`
});
export const ArchiveItemViewerWrapper = ({
  children
}) => {
  return jsx("div", {
    css: archiveItemViewerWrapperStyles
  }, children);
};
export const ArchiveSideBar = /*#__PURE__*/forwardRef(({
  children
}, ref) => {
  return jsx("div", {
    css: archiveSideBarStyles,
    ref: ref
  }, children);
});
export const ArchiveSidebarFolderWrapper = ({
  children
}) => {
  return jsx("div", {
    css: archiveSidebarFolderWrapperStyles,
    "data-testid": "archive-sidebar-folder-wrapper"
  }, children);
};
export const ArchiveDownloadButtonWrapper = ({
  children,
  onClick
}) => {
  return (
    // eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions
    jsx("div", {
      css: archiveDownloadButtonWrapperStyles,
      onClick: onClick,
      "data-testid": "media-archiveDownloadButton"
    }, children)
  );
};
export const DisabledArchiveDownloadButtonWrapper = ({
  children
}) => {
  return (
    // eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions
    jsx("div", {
      css: disabledArchiveDownloadButtonWrapperStyles,
      "data-testid": "media-disabledArchiveDownloadButton"
    }, children)
  );
};
export const SidebarItemWrapper = ({
  children
}) => {
  return jsx("div", {
    css: sidebarItemWrapperStyles
  }, children);
};
export const ArchiveSidebarFileEntryWrapper = ({
  children,
  index
}) => {
  return jsx("div", {
    css: archiveSidebarFileEntryWrapperStyles,
    key: index
  }, children);
};
export const ArchiveLayout = ({
  children
}) => {
  return jsx("div", {
    css: archiveLayoutStyles,
    "data-testid": "archive-layout"
  }, children);
};
export const ArchiveViewerWrapper = ({
  children
}) => {
  return jsx("div", {
    css: archiveViewerWrapperStyles
  }, children);
};
export const Separator = () => {
  return jsx("div", {
    css: separatorStyles
  });
};
export const SidebarHeaderWrapper = ({
  children
}) => {
  return jsx("span", {
    css: sidebarHeaderWrapperStyles
  }, children);
};
export const SidebarHeaderIcon = ({
  children
}) => {
  return jsx("div", {
    css: sidebarHeaderIconStyles
  }, children);
};
export const SidebarHeaderEntry = ({
  children
}) => {
  return jsx("div", {
    css: sidebarHeaderEntryStyles
  }, children);
};
