// @atlaskit/media-card/dist/esm/root/cardView.js
import { __extends } from "tslib";
import React from 'react';
import { withAnalyticsEvents, } from '@atlaskit/analytics-next';
import { FileCardImageView } from '../files';
import { breakpointSize } from '../utils/breakpoint';
import { defaultImageCardDimensions, getDefaultCardDimensions, } from '../utils/cardDimensions';
import { isValidPercentageUnit } from '../utils/isValidPercentageUnit';
import { getCSSUnitValue } from '../utils/getCSSUnitValue';
import { getElementDimension } from '../utils/getElementDimension';
import { Wrapper } from './styled';
import { createAndFireMediaEvent } from '../utils/analytics';
import { attachDetailsToActions } from '../actions';
import { getErrorMessage } from '../utils/getErrorMessage';
import { toHumanReadableMediaSize } from '@atlaskit/media-ui';
/**
 * This is classic vanilla CardView class. To create an instance of class one would need to supply
 * `createAnalyticsEvent` prop to satisfy it's Analytics Events needs.
 */
var CardViewBase = /** @class */ (function (_super) {
    __extends(CardViewBase, _super);
    function CardViewBase() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {};
        _this.divRef = React.createRef();
        _this.renderFileNewExperience = function () {
            // Let's replace this with the new experience
            return null;
        };
        _this.renderFile = function () {
            var _a = _this.props, status = _a.status, metadata = _a.metadata, dataURI = _a.dataURI, progress = _a.progress, onRetry = _a.onRetry, resizeMode = _a.resizeMode, dimensions = _a.dimensions, selectable = _a.selectable, selected = _a.selected, disableOverlay = _a.disableOverlay, previewOrientation = _a.previewOrientation, alt = _a.alt, onDisplayImage = _a.onDisplayImage, actions = _a.actions;
            var _b = metadata || {}, name = _b.name, mediaType = _b.mediaType, size = _b.size;
            var actionsWithDetails = metadata && actions ? attachDetailsToActions(actions, metadata) : [];
            var errorMessage = getErrorMessage(status);
            var fileSize = !size ? '' : toHumanReadableMediaSize(size);
            return (React.createElement(FileCardImageView, { error: errorMessage, dimensions: dimensions, selectable: selectable, selected: selected, dataURI: dataURI, mediaName: name, mediaType: mediaType, fileSize: fileSize, status: status, progress: progress, resizeMode: resizeMode, onRetry: onRetry, onDisplayImage: onDisplayImage, actions: actionsWithDetails, disableOverlay: disableOverlay, previewOrientation: previewOrientation, alt: alt }));
        };
        return _this;
    }
    CardViewBase.prototype.componentDidMount = function () {
        this.saveElementWidth();
    };
    Object.defineProperty(CardViewBase.prototype, "width", {
        // This width is only used to calculate breakpoints, dimensions are passed down as
        // integrator pass it to the root component
        get: function () {
            var elementWidth = this.state.elementWidth;
            if (elementWidth) {
                return elementWidth;
            }
            var width = (this.props.dimensions || { width: undefined }).width;
            if (!width) {
                return defaultImageCardDimensions.width;
            }
            return getCSSUnitValue(width);
        },
        enumerable: true,
        configurable: true
    });
    // If the dimensions.width is a percentage, we need to transform it
    // into a pixel value in order to get the right breakpoints applied.
    CardViewBase.prototype.saveElementWidth = function () {
        var dimensions = this.props.dimensions;
        if (!dimensions) {
            return;
        }
        var width = dimensions.width;
        if (width && isValidPercentageUnit(width) && this.divRef.current) {
            var elementWidth = getElementDimension(this.divRef.current, 'width');
            this.setState({ elementWidth: elementWidth });
        }
    };
    CardViewBase.prototype.render = function () {
        var _a = this.props, dimensions = _a.dimensions, appearance = _a.appearance, onClick = _a.onClick, onMouseEnter = _a.onMouseEnter, testId = _a.testId, featureFlags = _a.featureFlags;
        var wrapperDimensions = dimensions
            ? dimensions
            : getDefaultCardDimensions(appearance);
        var isNewExpFeatureFlagOn = featureFlags
            ? featureFlags.enableNewExperience
            : false;
        return (React.createElement(Wrapper, { "data-testid": testId || 'media-card-view', shouldUsePointerCursor: true, breakpointSize: breakpointSize(this.width), dimensions: wrapperDimensions, onClick: onClick, onMouseEnter: onMouseEnter, innerRef: this.divRef }, isNewExpFeatureFlagOn
            ? this.renderFileNewExperience()
            : this.renderFile()));
    };
    CardViewBase.defaultProps = {
        appearance: 'auto',
    };
    return CardViewBase;
}(React.Component));
export { CardViewBase };
export var CardView = withAnalyticsEvents({
    onClick: createAndFireMediaEvent({
        eventType: 'ui',
        action: 'clicked',
        actionSubject: 'mediaCard',
    }),
})(CardViewBase);
