// @atlaskit/page/dist/cjs/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.varGridSpacing = exports.varColumnsNum = exports.varColumnSpan = exports.spacingMapping = exports.defaultSpacing = exports.defaultMedium = exports.defaultLayout = exports.defaultGridColumns = exports.defaultGridColumnWidth = exports.defaultBannerHeight = void 0;
var defaultGridSize = 8;
var defaultGridColumns = exports.defaultGridColumns = 12;
var defaultGridColumnWidth = exports.defaultGridColumnWidth = defaultGridSize * 10;
var spacingMapping = exports.spacingMapping = {
  comfortable: defaultGridSize * 5,
  cosy: defaultGridSize * 2,
  compact: defaultGridSize * 0.5
};
var defaultSpacing = exports.defaultSpacing = 'cosy';
var defaultBannerHeight = exports.defaultBannerHeight = 52;

// Corresponds to an `auto` width/flex-basis
var defaultMedium = exports.defaultMedium = 0;
var defaultLayout = exports.defaultLayout = 'fixed';

/**
 * The number of available columns in each row.
 */
var varColumnsNum = exports.varColumnsNum = '--ds-columns-num';

/**
 * The number of columns that a `GridColumn` covers.
 */
var varColumnSpan = exports.varColumnSpan = '--ds-column-span';

/**
 * The spacing (in `px`) between each column.
 */
var varGridSpacing = exports.varGridSpacing = '--ds-grid-spacing';
