// @atlaskit/link-datasource/dist/cjs/ui/common/error-state/access-required.js
/* access-required.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AccessRequired = void 0;
require("./access-required.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _reactIntlNext = require("react-intl-next");
var _emptyState = _interopRequireDefault(require("@atlaskit/empty-state"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _compiled = require("@atlaskit/primitives/compiled");
var _colors = require("@atlaskit/theme/colors");
var _typography = require("@atlaskit/theme/typography");
var _analytics = require("../../../analytics");
var _padlockKey = require("../../../common/ui/spot/basics/padlock-key");
var _accessRequiredSvg = require("./access-required-svg");
var _messages = require("./messages");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var styles = {
  iconContainerStyles: "_otyrpxbi",
  descriptionMessageStyles: "_zulppxbi _1e0c1txw _2lx21bp4",
  urlStylesOld: "_11c82smr _syaznoz8",
  urlStyles: "_11c8lodh _syaz1fxt _19pkmgjw"
};
var Description = function Description(_ref) {
  var message = _ref.message,
    url = _ref.url;
  return /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)([styles.descriptionMessageStyles])
  }, /*#__PURE__*/React.createElement("span", {
    className: (0, _runtime.ax)([(0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv') ? styles.urlStyles : styles.urlStylesOld])
  }, url), /*#__PURE__*/React.createElement("span", null, message));
};
var noop = function noop() {
  return '';
};
var IconContainer = function IconContainer() {
  var _ref2 = (0, _platformFeatureFlags.fg)('bandicoots-update-sllv-icons') ?
    // eslint-disable-next-line react-hooks/rules-of-hooks
    (0, _reactIntlNext.useIntl)() : {
      formatMessage: noop
    },
    formatMessage = _ref2.formatMessage;
  return /*#__PURE__*/React.createElement(_compiled.Box, {
    xcss: styles.iconContainerStyles
  }, (0, _platformFeatureFlags.fg)('bandicoots-update-sllv-icons') ? /*#__PURE__*/React.createElement(_padlockKey.SpotPadlockKey, {
    size: 'xlarge',
    alt: formatMessage(_messages.loadingErrorMessages.accessRequired)
  }) : /*#__PURE__*/React.createElement(_accessRequiredSvg.AccessRequiredSVGOld, null));
};
var AccessRequired = exports.AccessRequired = function AccessRequired(_ref3) {
  var url = _ref3.url;
  var _useIntl = (0, _reactIntlNext.useIntl)(),
    formatMessage = _useIntl.formatMessage;
  var _useDatasourceAnalyti = (0, _analytics.useDatasourceAnalyticsEvents)(),
    fireEvent = _useDatasourceAnalyti.fireEvent;
  (0, _react.useEffect)(function () {
    fireEvent('ui.error.shown', {
      reason: 'access'
    });
  }, [fireEvent]);
  if (url) {
    return /*#__PURE__*/React.createElement(_emptyState.default, {
      testId: "datasource--access-required-with-url",
      header: formatMessage((0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv') ? _messages.loadingErrorMessages.accessRequiredWithSite : _messages.loadingErrorMessages.accessRequiredWithSiteOld),
      headingLevel: (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv') ? 2 : undefined,
      description: /*#__PURE__*/React.createElement(Description, {
        message: formatMessage(_messages.loadingErrorMessages.accessInstructions),
        url: url
      }),
      renderImage: IconContainer
    });
  }
  return /*#__PURE__*/React.createElement(_emptyState.default, {
    testId: "datasource--access-required",
    header: formatMessage(_messages.loadingErrorMessages.accessRequired),
    description: formatMessage(_messages.loadingErrorMessages.accessInstructions),
    renderImage: IconContainer
  });
};
