// @atlaskit/link-datasource/dist/cjs/ui/common/error-state/modal-loading-error.js
/* modal-loading-error.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ModalLoadingError = void 0;
require("./modal-loading-error.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _react2 = require("@compiled/react");
var _reactIntlNext = require("react-intl-next");
var _new = _interopRequireDefault(require("@atlaskit/button/new"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _compiled = require("@atlaskit/primitives/compiled");
var _analytics = require("../../../analytics");
var _search = require("../../../common/ui/spot/error-state/search");
var _loadingErrorSvgOld = require("./loading-error-svg-old");
var _messages = require("./messages");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var styles = {
  errorContainerStyles: "_zulp1ejb _4cvr1h6o _6a6z1h6o _1wpz1h6o _10fp1h6o _1e0c11p5",
  errorContainerStylesNew: "_zulp1ejb _4cvr1h6o _6a6z1h6o _1wpz1h6o _10fp1h6o _1e0c11p5 _ca0qoahv _u5f3oahv _n3tdoahv _19bvoahv _p12f1kdj",
  errorMessageContainerStylesOld: "_zulpu2gc _4cvr1h6o _6a6z1h6o _1e0c11p5",
  errorMessageContainerStyles: "_zulppxbi _4cvr1h6o _6a6z1h6o _1e0c11p5 _y3gn1h6o"
};
var ModalLoadingError = exports.ModalLoadingError = function ModalLoadingError(_ref) {
  var _ref$errorMessage = _ref.errorMessage,
    errorMessage = _ref$errorMessage === void 0 ? /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, _messages.loadingErrorMessages.checkConnection) : _ref$errorMessage,
    onRefresh = _ref.onRefresh;
  var _useDatasourceAnalyti = (0, _analytics.useDatasourceAnalyticsEvents)(),
    fireEvent = _useDatasourceAnalyti.fireEvent;
  (0, _react.useEffect)(function () {
    fireEvent('ui.error.shown', {
      reason: 'network'
    });
  }, [fireEvent]);
  return /*#__PURE__*/React.createElement(_compiled.Box, {
    xcss: (0, _react2.cx)((0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv') ? styles.errorContainerStylesNew : styles.errorContainerStyles),
    testId: "datasource-modal--loading-error"
  }, (0, _platformFeatureFlags.fg)('bandicoots-update-sllv-icons') ? /*#__PURE__*/React.createElement(_search.SpotErrorSearch, {
    size: 'xlarge',
    alt: ""
  }) : /*#__PURE__*/React.createElement(_loadingErrorSvgOld.LoadingErrorSVGOld, null), /*#__PURE__*/React.createElement(_compiled.Box, {
    xcss: (0, _react2.cx)((0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv') ? styles.errorMessageContainerStyles : styles.errorMessageContainerStylesOld)
  }, (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv') ? /*#__PURE__*/React.createElement(_compiled.Text, {
    as: "span",
    size: "large",
    weight: "bold"
  }, /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, _messages.loadingErrorMessages.unableToLoadResults)) : /*#__PURE__*/React.createElement(_compiled.Text, {
    size: "small"
  }, /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, _messages.loadingErrorMessages.unableToLoadResultsOld)), (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv') ? /*#__PURE__*/React.createElement(_compiled.Text, {
    as: "span",
    size: "medium"
  }, errorMessage) : /*#__PURE__*/React.createElement(_compiled.Text, {
    as: "p"
  }, errorMessage), onRefresh && (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv') && /*#__PURE__*/React.createElement(_new.default, {
    appearance: "primary",
    onClick: onRefresh,
    testId: "datasource-modal--loading-error-refresh"
  }, /*#__PURE__*/React.createElement(_reactIntlNext.FormattedMessage, _messages.loadingErrorMessages.refresh))));
};
