// @atlaskit/editor-common/dist/cjs/types/annotation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var events_1 = tslib_1.__importDefault(require("events"));
var adf_schema_1 = require("@atlaskit/adf-schema");
var AnnotationUpdateEvent;
(function (AnnotationUpdateEvent) {
    AnnotationUpdateEvent["SET_ANNOTATION_FOCUS"] = "SET_ANNOTATION_FOCUS";
    AnnotationUpdateEvent["SET_ANNOTATION_STATE"] = "SET_ANNOTATION_STATE";
    AnnotationUpdateEvent["REMOVE_ANNOTATION_FOCUS"] = "REMOVE_ANNOTATION_FOCUS";
})(AnnotationUpdateEvent = exports.AnnotationUpdateEvent || (exports.AnnotationUpdateEvent = {}));
var AnnotationUpdateEmitter = /** @class */ (function () {
    function AnnotationUpdateEmitter() {
        this.emitter = new events_1.default();
    }
    AnnotationUpdateEmitter.prototype.emit = function (event) {
        var _a;
        var payload = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            payload[_i - 1] = arguments[_i];
        }
        if (typeof payload === 'undefined') {
            return this.emitter.emit(event);
        }
        else {
            return (_a = this.emitter).emit.apply(_a, tslib_1.__spread([event], payload));
        }
    };
    AnnotationUpdateEmitter.prototype.on = function (event, listener) {
        return this.emitter.on(event, listener);
    };
    AnnotationUpdateEmitter.prototype.off = function (event, listener) {
        return this.emitter.off(event, listener);
    };
    AnnotationUpdateEmitter.prototype.listeners = function (event) {
        return this.emitter.listeners(event);
    };
    return AnnotationUpdateEmitter;
}());
exports.AnnotationUpdateEmitter = AnnotationUpdateEmitter;
