// @atlaskit/link-picker/dist/cjs/ui/link-picker/search-results/link-search-list/link-search-no-results/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.testIds = exports.messages = exports.NoResults = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactIntlNext = require("react-intl-next");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _emptyState = require("../../../../../common/ui/empty-state");
var _noResultsSvg = require("./no-results-svg");
var _noResultsSvgV = require("./no-results-svg-v2");
var messages = exports.messages = (0, _reactIntlNext.defineMessages)({
  noResults: {
    id: 'fabric.linkPicker.search.noResults.heading',
    defaultMessage: 'We couldn’t find anything matching your search.',
    description: 'Heading message shown when a search has no results'
  },
  noResultsDescription: {
    id: 'fabric.linkPicker.search.noResults.description',
    defaultMessage: 'Try again with a different term.',
    description: 'Describes possible action when a search returns no results'
  }
});
var testIds = exports.testIds = {
  emptyResultPage: 'link-search-no-results'
};
var NoResults = exports.NoResults = function NoResults() {
  var intl = (0, _reactIntlNext.useIntl)();
  return /*#__PURE__*/_react.default.createElement(_emptyState.EmptyState, {
    testId: testIds.emptyResultPage,
    header: intl.formatMessage(messages.noResults),
    description: intl.formatMessage(messages.noResultsDescription),
    renderImage: (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-link-picker') ? function () {
      return /*#__PURE__*/_react.default.createElement(_noResultsSvgV.NoResultsV2SVG, null);
    } : function () {
      return /*#__PURE__*/_react.default.createElement(_noResultsSvg.NoResultsSVG, null);
    }
  });
};
