// @atlaskit/ufo/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ConcurrentExperience", {
  enumerable: true,
  get: function get() {
    return _platformClient.ConcurrentExperience;
  }
});
Object.defineProperty(exports, "ExperiencePerformanceTypes", {
  enumerable: true,
  get: function get() {
    return _platformClient.ExperiencePerformanceTypes;
  }
});
Object.defineProperty(exports, "ExperienceTypes", {
  enumerable: true,
  get: function get() {
    return _platformClient.ExperienceTypes;
  }
});
Object.defineProperty(exports, "GlobalPageLoadExperience", {
  enumerable: true,
  get: function get() {
    return _platformClient.GlobalPageLoadExperience;
  }
});
Object.defineProperty(exports, "UFOExperience", {
  enumerable: true,
  get: function get() {
    return _platformClient.UFOExperience;
  }
});
Object.defineProperty(exports, "UFOExperienceState", {
  enumerable: true,
  get: function get() {
    return _platformClient.UFOExperienceState;
  }
});
Object.defineProperty(exports, "UFO_EXPERIMENTAL_BUILD_VERSION", {
  enumerable: true,
  get: function get() {
    return _buildVersion.UFO_EXPERIMENTAL_BUILD_VERSION;
  }
});
Object.defineProperty(exports, "experiencePayloadEvent", {
  enumerable: true,
  get: function get() {
    return _globalStreamBuffer.experiencePayloadEvent;
  }
});
Object.defineProperty(exports, "getGlobalEventStream", {
  enumerable: true,
  get: function get() {
    return _globalStreamBuffer.getGlobalEventStream;
  }
});
Object.defineProperty(exports, "setGlobalEventStream", {
  enumerable: true,
  get: function get() {
    return _globalStreamBuffer.setGlobalEventStream;
  }
});
Object.defineProperty(exports, "subscribeEvent", {
  enumerable: true,
  get: function get() {
    return _globalStreamBuffer.subscribeEvent;
  }
});
Object.defineProperty(exports, "ufolog", {
  enumerable: true,
  get: function get() {
    return _logger.ufolog;
  }
});
Object.defineProperty(exports, "ufologger", {
  enumerable: true,
  get: function get() {
    return _logger.ufologger;
  }
});
Object.defineProperty(exports, "ufowarn", {
  enumerable: true,
  get: function get() {
    return _logger.ufowarn;
  }
});
Object.defineProperty(exports, "unsubscribeEvent", {
  enumerable: true,
  get: function get() {
    return _globalStreamBuffer.unsubscribeEvent;
  }
});
Object.defineProperty(exports, "untilAll", {
  enumerable: true,
  get: function get() {
    return _untilHelpers.untilAll;
  }
});
Object.defineProperty(exports, "visibilityChangeObserver", {
  enumerable: true,
  get: function get() {
    return _visibilityChangeObserver.visibilityChangeObserver;
  }
});
var _platformClient = require("./platform-client");
var _logger = require("./logger");
var _globalStreamBuffer = require("./global-stream-buffer");
var _buildVersion = require("./buildVersion");
var _visibilityChangeObserver = require("./observer/visibility-change-observer");
var _untilHelpers = require("./platform-client/utils/until-helpers");
