// @atlaskit/ufo/dist/cjs/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UFOGlobalEventStreamEventType = exports.SUBSCRIBE_ALL = exports.PageVisibleState = exports.PageSegmentLoadMetrics = exports.PageLoadMetrics = exports.InteractionMetrics = exports.INLINE_RESPONSE_MARK = exports.FMP_MARK = exports.CustomMetrics = void 0;
var UFOGlobalEventStreamEventType = exports.UFOGlobalEventStreamEventType = /*#__PURE__*/function (UFOGlobalEventStreamEventType) {
  UFOGlobalEventStreamEventType[UFOGlobalEventStreamEventType["EXPERIENCE_PAYLOAD"] = 0] = "EXPERIENCE_PAYLOAD";
  UFOGlobalEventStreamEventType[UFOGlobalEventStreamEventType["SUBSCRIBE"] = 1] = "SUBSCRIBE";
  UFOGlobalEventStreamEventType[UFOGlobalEventStreamEventType["UNSUBSCRIBE"] = 2] = "UNSUBSCRIBE";
  return UFOGlobalEventStreamEventType;
}({});
var PageVisibleState = exports.PageVisibleState = /*#__PURE__*/function (PageVisibleState) {
  PageVisibleState["VISIBLE"] = "visible";
  PageVisibleState["HIDDEN"] = "hidden";
  PageVisibleState["MIXED"] = "mixed";
  return PageVisibleState;
}({});
var PageLoadMetrics = exports.PageLoadMetrics = /*#__PURE__*/function (PageLoadMetrics) {
  PageLoadMetrics["fmp"] = "fmp";
  PageLoadMetrics["tti"] = "tti";
  return PageLoadMetrics;
}({});
var PageSegmentLoadMetrics = exports.PageSegmentLoadMetrics = /*#__PURE__*/function (PageSegmentLoadMetrics) {
  PageSegmentLoadMetrics["fmp"] = "fmp";
  PageSegmentLoadMetrics["tti"] = "tti";
  return PageSegmentLoadMetrics;
}({});
var InteractionMetrics = exports.InteractionMetrics = /*#__PURE__*/function (InteractionMetrics) {
  InteractionMetrics["response"] = "response";
  InteractionMetrics["result"] = "result";
  return InteractionMetrics;
}({});
var CustomMetrics = exports.CustomMetrics = /*#__PURE__*/function (CustomMetrics) {
  CustomMetrics["duration"] = "duration";
  return CustomMetrics;
}({});
var SUBSCRIBE_ALL = exports.SUBSCRIBE_ALL = '__SUBSCRIBE_ALL__';
var FMP_MARK = exports.FMP_MARK = 'fmp';
var INLINE_RESPONSE_MARK = exports.INLINE_RESPONSE_MARK = 'inline-response';
