// @atlaskit/editor-plugin-text-formatting/dist/es2019/ui/Toolbar/constants.js
import { ToolbarSize } from '@atlaskit/editor-common/types';
import { IconTypes } from './types';
export const DefaultButtonsToolbar = [IconTypes.strong, IconTypes.em];
export const DefaultFloatingToolbarButtonsNext = [IconTypes.strong];
export const DefaultButtonsToolbarNext = [IconTypes.strong, IconTypes.em, IconTypes.underline];
export const DefaultButtonsMenu = [IconTypes.underline, IconTypes.strike, IconTypes.code, IconTypes.subscript, IconTypes.superscript];
export const DefaultButtonsMenuNext = [IconTypes.strike, IconTypes.code, IconTypes.subscript, IconTypes.superscript];

// eslint-disable-next-line @repo/internal/deprecations/deprecation-ticket-required -- Ignored via go/ED-25883
/** @deprecated
 * When tidying platform_editor_controls this is to be removed in favor of ResponsiveCustomButtonToolbarNext
 * along with references to platform_editor_controls gate
 */
export const ResponsiveCustomButtonToolbar = {
  [ToolbarSize.XXL]: DefaultButtonsToolbar,
  [ToolbarSize.XL]: DefaultButtonsToolbar,
  [ToolbarSize.L]: DefaultButtonsToolbar,
  [ToolbarSize.M]: DefaultButtonsToolbar,
  [ToolbarSize.S]: DefaultButtonsToolbar,
  [ToolbarSize.XXXS]: []
};
export const ResponsiveCustomButtonToolbarNext = {
  [ToolbarSize.XXL]: DefaultButtonsToolbarNext,
  [ToolbarSize.XL]: DefaultButtonsToolbarNext,
  [ToolbarSize.L]: DefaultButtonsToolbarNext,
  [ToolbarSize.M]: [IconTypes.strong, IconTypes.em],
  [ToolbarSize.S]: [IconTypes.strong, IconTypes.em],
  [ToolbarSize.XXXS]: [IconTypes.strong]
};

// eslint-disable-next-line @repo/internal/deprecations/deprecation-ticket-required -- Ignored via go/ED-25883
/** @deprecated
 * When tidying platform_editor_controls this is to be removed in favor of ResponsiveCustomMenuNext
 * along with references to platform_editor_controls gate
 */
export const ResponsiveCustomMenu = {
  [ToolbarSize.XXL]: DefaultButtonsMenu,
  [ToolbarSize.XL]: DefaultButtonsMenu,
  [ToolbarSize.L]: DefaultButtonsMenu,
  [ToolbarSize.M]: DefaultButtonsMenu,
  [ToolbarSize.S]: DefaultButtonsMenu,
  [ToolbarSize.XXXS]: [IconTypes.strong, IconTypes.em, ...DefaultButtonsMenu]
};
export const ResponsiveCustomMenuNext = {
  [ToolbarSize.XXL]: DefaultButtonsMenuNext,
  [ToolbarSize.XL]: DefaultButtonsMenuNext,
  [ToolbarSize.L]: DefaultButtonsMenuNext,
  [ToolbarSize.M]: [IconTypes.underline, ...DefaultButtonsMenuNext],
  [ToolbarSize.S]: [IconTypes.underline, ...DefaultButtonsMenuNext],
  [ToolbarSize.XXXS]: [IconTypes.em, IconTypes.underline, ...DefaultButtonsMenuNext]
};
