// @atlaskit/media-card/dist/cjs/card/ui/openMediaViewerButton/openMediaViewerButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireDefault(require("react"));
var _reactIntlNext = require("react-intl-next");
var _mediaUi = require("@atlaskit/media-ui");
var _primitives = require("@atlaskit/primitives");
var _visuallyHidden = _interopRequireDefault(require("@atlaskit/visually-hidden"));
var _excluded = ["fileName", "innerRef"];
var OpenMediaViewerButton = function OpenMediaViewerButton(_ref) {
  var fileName = _ref.fileName,
    innerRef = _ref.innerRef,
    props = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  var intl = (0, _reactIntlNext.useIntl)();
  return /*#__PURE__*/_react.default.createElement(_visuallyHidden.default, null, /*#__PURE__*/_react.default.createElement(_primitives.Pressable, (0, _extends2.default)({
    ref: innerRef
  }, props), intl.formatMessage(_mediaUi.messages.open_file_in_viewer, {
    name: fileName
  })));
};
var OpenMediaViewerButtonWrapper = function OpenMediaViewerButtonWrapper(props) {
  return props.intl ? /*#__PURE__*/_react.default.createElement(OpenMediaViewerButton, props) : /*#__PURE__*/_react.default.createElement(_reactIntlNext.IntlProvider, {
    locale: "en"
  }, /*#__PURE__*/_react.default.createElement(OpenMediaViewerButton, props));
};
var _default = exports.default = (0, _reactIntlNext.injectIntl)(OpenMediaViewerButtonWrapper, {
  enforceContext: false
});
