// @atlaskit/smart-card/dist/esm/view/RelatedLinksModal/components/related-link-item/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import "./index.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { fg } from '@atlaskit/platform-feature-flags';
import { SmartLinkStatus } from '../../../../constants';
import { Card, ElementName, SmartLinkPosition, SmartLinkSize, SmartLinkTheme, TitleBlock } from '../../../../index';
var hoverStyleOld = null;
var hoverStyle = null;
var baseStyle = null;
var selectedStyle = null;
var relatedLinkItemStyles = null;
var RelatedLinkItem = function RelatedLinkItem(_ref) {
  var url = _ref.url,
    testId = _ref.testId,
    isSelected = _ref.isSelected,
    onFocus = _ref.onFocus;
  var subtitle = [{
    name: ElementName.Provider,
    hideIcon: true
  }];
  var ui = {
    hideElevation: true,
    hidePadding: true,
    hideBackground: true,
    theme: SmartLinkTheme.Black,
    clickableContainer: true,
    size: fg('platform-linking-visual-refresh-v2') ? SmartLinkSize.Medium : SmartLinkSize.Large
  };
  return /*#__PURE__*/React.createElement("div", {
    "data-testId": testId,
    onFocus: onFocus,
    className: ax([fg('platform-linking-visual-refresh-v2') && "_195g10v4 _18zr1ejb", fg('platform-linking-visual-refresh-v2') ? "_irr3166n _1di61dty" : "_8vu4iti9 _1mxox0bf _10swu2gc _irr31d5g", isSelected && fg('platform-linking-visual-refresh-v2') && "_bfhkfg4m _16qsd1ld _irr3i1yw _1di619ru"]),
    style: {
      "--_g94mg3": ix("inset 2px 0 0 0 ".concat("var(--ds-border-selected, #0C66E4)"))
    }
  }, /*#__PURE__*/React.createElement(Card, {
    appearance: "block",
    ui: ui,
    url: url
  }, /*#__PURE__*/React.createElement(TitleBlock, {
    maxLines: 1,
    hideTitleTooltip: true,
    position: fg('platform-linking-visual-refresh-v2') ? SmartLinkPosition.Top : SmartLinkPosition.Center,
    subtitle: subtitle,
    anchorTarget: "_blank",
    size: ui.size,
    theme: ui.theme,
    status: SmartLinkStatus.Resolved,
    className: ax(["_zulputpp _11c81o8v _ca0qu2gc _n3tdu2gc"])
  })));
};
export default RelatedLinkItem;
