// @atlaskit/link-datasource/dist/es2019/ui/jira-issues-modal/jql-editor/index.js
import React, { useEffect, useRef } from 'react';
import { JQLEditor } from '@atlaskit/jql-editor';
import { useAutocompleteProvider } from '@atlaskit/jql-editor-autocomplete-rest';
import { makeGetJqlAutocompleteData } from '../../../services/makeGetJqlAutocompleteData';
import { makeGetJqlSuggestionsData } from '../../../services/makeGetJqlSuggestionsData';
export const JiraJQLEditor = ({
  cloudId,
  isSearching,
  onChange,
  onSearch,
  query
}) => {
  const autocompleteProvider = useAutocompleteProvider('link-datasource', makeGetJqlAutocompleteData(cloudId), makeGetJqlSuggestionsData(cloudId));

  // This is an expected (pretty strange imo) way of making sure text field is in focus when rendered
  const inputRef = useRef({
    focus: () => {}
  });
  useEffect(() => {
    requestAnimationFrame(() => {
      var _inputRef$current;
      (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 ? void 0 : _inputRef$current.focus();
    });
  }, []);
  const searchIfValidJql = (_, jast) => {
    if (jast.errors.length === 0) {
      onSearch();
    }
  };
  return /*#__PURE__*/React.createElement(JQLEditor, {
    analyticsSource: "link-datasource",
    autocompleteProvider: autocompleteProvider,
    onSearch: searchIfValidJql,
    onUpdate: onChange,
    isSearching: isSearching,
    inputRef: inputRef,
    query: query,
    "aria-label": "JQL Query Editor"
  });
};
