// @atlaskit/editor-plugin-block-controls/dist/esm/blockControlsPlugin.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
import React from 'react';
import { expandSelectionBounds } from '@atlaskit/editor-common/selection';
import { TextSelection } from '@atlaskit/editor-prosemirror/state';
import { fg } from '@atlaskit/platform-feature-flags';
import { editorExperiment } from '@atlaskit/tmp-editor-statsig/experiments';
import { moveNode } from './editor-commands/move-node';
import { moveToLayout } from './editor-commands/move-to-layout';
import { firstNodeDecPlugin } from './pm-plugins/first-node-dec-plugin';
import { createInteractionTrackingPlugin, interactionTrackingPluginKey } from './pm-plugins/interaction-tracking/pm-plugin';
import { createPlugin, key } from './pm-plugins/main';
import { selectNode } from './pm-plugins/utils/getSelection';
import BlockMenu from './ui/block-menu';
import { DragHandleMenu } from './ui/drag-handle-menu';
import { GlobalStylesWrapper } from './ui/global-styles';
export var blockControlsPlugin = function blockControlsPlugin(_ref) {
  var api = _ref.api;
  return {
    name: 'blockControls',
    pmPlugins: function pmPlugins() {
      var pmPlugins = [{
        name: 'blockControlsPmPlugin',
        plugin: function plugin(_ref2) {
          var getIntl = _ref2.getIntl,
            nodeViewPortalProviderAPI = _ref2.nodeViewPortalProviderAPI;
          return createPlugin(api, getIntl, nodeViewPortalProviderAPI);
        }
      }];
      if (editorExperiment('platform_editor_controls', 'variant1')) {
        if (fg('platform_editor_controls_widget_visibility')) {
          pmPlugins.push({
            name: 'blockControlsInteractionTrackingPlugin',
            plugin: createInteractionTrackingPlugin
          });
        }
        pmPlugins.push({
          name: 'firstNodeDec',
          plugin: firstNodeDecPlugin
        });
      }
      return pmPlugins;
    },
    commands: {
      moveNode: moveNode(api),
      moveToLayout: moveToLayout(api),
      showDragHandleAt: function showDragHandleAt(pos, anchorName, nodeType, handleOptions, rootPos, rootAnchorName, rootNodeType) {
        return function (_ref3) {
          var tr = _ref3.tr;
          var currMeta = tr.getMeta(key);
          tr.setMeta(key, _objectSpread(_objectSpread({}, currMeta), {}, {
            activeNode: {
              pos: pos,
              anchorName: anchorName,
              nodeType: nodeType,
              handleOptions: handleOptions,
              rootPos: rootPos,
              rootAnchorName: rootAnchorName,
              rootNodeType: rootNodeType
            },
            closeMenu: editorExperiment('platform_editor_controls', 'variant1') ? true : undefined
          }));
          return tr;
        };
      },
      toggleBlockMenu: function toggleBlockMenu(options) {
        return function (_ref4) {
          var tr = _ref4.tr;
          var currMeta = tr.getMeta(key);
          if (options !== null && options !== void 0 && options.closeMenu) {
            tr.setMeta(key, _objectSpread(_objectSpread({}, currMeta), {}, {
              closeMenu: true
            }));
            return tr;
          }
          tr.setMeta(key, _objectSpread(_objectSpread({}, currMeta), {}, {
            toggleMenu: {
              anchorName: options === null || options === void 0 ? void 0 : options.anchorName
            }
          }));
          return tr;
        };
      },
      setNodeDragged: function setNodeDragged(getPos, anchorName, nodeType) {
        return function (_ref5) {
          var tr = _ref5.tr;
          var pos = getPos();
          if (pos === undefined) {
            return tr;
          }
          var currMeta = tr.getMeta(key);
          tr.setMeta(key, _objectSpread(_objectSpread({}, currMeta), {}, {
            isDragging: true,
            activeNode: {
              pos: pos,
              anchorName: anchorName,
              nodeType: nodeType
            }
          }));
          if (fg('platform_editor_ease_of_use_metrics')) {
            var _api$metrics;
            api === null || api === void 0 || (_api$metrics = api.metrics) === null || _api$metrics === void 0 || _api$metrics.commands.handleIntentToStartEdit({
              shouldStartTimer: false,
              shouldPersistActiveSession: true
            })({
              tr: tr
            });
          }
          if (fg('platform_editor_user_intent_plugin')) {
            var _api$userIntent;
            api === null || api === void 0 || (_api$userIntent = api.userIntent) === null || _api$userIntent === void 0 || _api$userIntent.commands.setCurrentUserIntent('dragging')({
              tr: tr
            });
          }
          return tr;
        };
      },
      setMultiSelectPositions: function setMultiSelectPositions(anchor, head) {
        return function (_ref6) {
          var _api$selection, _$to$nodeBefore, _$from$nodeAfter;
          var tr = _ref6.tr;
          var _tr$selection = tr.selection,
            userAnchor = _tr$selection.anchor,
            userHead = _tr$selection.head;
          var $expandedAnchor, $expandedHead;
          if (anchor !== undefined && head !== undefined) {
            $expandedAnchor = tr.doc.resolve(anchor);
            $expandedHead = tr.doc.resolve(head);
          } else {
            var expandedSelection = expandSelectionBounds(tr.selection.$anchor, tr.selection.$head);
            $expandedAnchor = expandedSelection.$anchor;
            $expandedHead = expandedSelection.$head;
          }
          api === null || api === void 0 || (_api$selection = api.selection) === null || _api$selection === void 0 || _api$selection.commands.setManualSelection($expandedAnchor.pos, $expandedHead.pos)({
            tr: tr
          });
          var $from = $expandedAnchor.min($expandedHead);
          var $to = $expandedAnchor.max($expandedHead);
          var expandedNormalisedSel;
          if ($from.nodeAfter === $to.nodeBefore) {
            selectNode(tr, $from.pos, $expandedAnchor.node().type.name);
            expandedNormalisedSel = tr.selection;
          } else if (((_$to$nodeBefore = $to.nodeBefore) === null || _$to$nodeBefore === void 0 ? void 0 : _$to$nodeBefore.type.name) === 'mediaSingle' || ((_$from$nodeAfter = $from.nodeAfter) === null || _$from$nodeAfter === void 0 ? void 0 : _$from$nodeAfter.type.name) === 'mediaSingle') {
            expandedNormalisedSel = new TextSelection($expandedAnchor, $expandedHead);
            tr.setSelection(expandedNormalisedSel);
          } else {
            // this is to normalise the selection's boundaries to inline positions, preventing it from collapsing
            expandedNormalisedSel = TextSelection.between($expandedAnchor, $expandedHead);
            tr.setSelection(expandedNormalisedSel);
          }
          var multiSelectDnD = {
            anchor: $expandedAnchor.pos,
            head: $expandedHead.pos,
            textAnchor: expandedNormalisedSel.anchor,
            textHead: expandedNormalisedSel.head,
            userAnchor: userAnchor,
            userHead: userHead
          };
          var currMeta = tr.getMeta(key);
          tr.setMeta(key, _objectSpread(_objectSpread({}, currMeta), {}, {
            multiSelectDnD: multiSelectDnD
          }));
          return tr;
        };
      },
      setSelectedViaDragHandle: function setSelectedViaDragHandle(isSelectedViaDragHandle) {
        return function (_ref7) {
          var tr = _ref7.tr;
          var currMeta = tr.getMeta(key);
          return tr.setMeta(key, _objectSpread(_objectSpread({}, currMeta), {}, {
            isSelectedViaDragHandle: isSelectedViaDragHandle
          }));
        };
      }
    },
    getSharedState: function getSharedState(editorState) {
      var _key$getState$isMenuO, _key$getState, _key$getState$menuTri, _key$getState2, _key$getState$activeN, _key$getState3, _key$getState$isDragg, _key$getState4, _key$getState$isPMDra, _key$getState5, _key$getState$multiSe, _key$getState6, _key$getState$isShift, _key$getState7, _key$getState$lastDra, _key$getState8, _interactionTrackingP, _key$getState$isSelec, _key$getState9;
      if (!editorState) {
        return undefined;
      }
      return {
        isMenuOpen: (_key$getState$isMenuO = (_key$getState = key.getState(editorState)) === null || _key$getState === void 0 ? void 0 : _key$getState.isMenuOpen) !== null && _key$getState$isMenuO !== void 0 ? _key$getState$isMenuO : false,
        menuTriggerBy: (_key$getState$menuTri = (_key$getState2 = key.getState(editorState)) === null || _key$getState2 === void 0 ? void 0 : _key$getState2.menuTriggerBy) !== null && _key$getState$menuTri !== void 0 ? _key$getState$menuTri : undefined,
        activeNode: (_key$getState$activeN = (_key$getState3 = key.getState(editorState)) === null || _key$getState3 === void 0 ? void 0 : _key$getState3.activeNode) !== null && _key$getState$activeN !== void 0 ? _key$getState$activeN : undefined,
        isDragging: (_key$getState$isDragg = (_key$getState4 = key.getState(editorState)) === null || _key$getState4 === void 0 ? void 0 : _key$getState4.isDragging) !== null && _key$getState$isDragg !== void 0 ? _key$getState$isDragg : false,
        isPMDragging: (_key$getState$isPMDra = (_key$getState5 = key.getState(editorState)) === null || _key$getState5 === void 0 ? void 0 : _key$getState5.isPMDragging) !== null && _key$getState$isPMDra !== void 0 ? _key$getState$isPMDra : false,
        multiSelectDnD: (_key$getState$multiSe = (_key$getState6 = key.getState(editorState)) === null || _key$getState6 === void 0 ? void 0 : _key$getState6.multiSelectDnD) !== null && _key$getState$multiSe !== void 0 ? _key$getState$multiSe : undefined,
        isShiftDown: (_key$getState$isShift = (_key$getState7 = key.getState(editorState)) === null || _key$getState7 === void 0 ? void 0 : _key$getState7.isShiftDown) !== null && _key$getState$isShift !== void 0 ? _key$getState$isShift : undefined,
        lastDragCancelled: (_key$getState$lastDra = (_key$getState8 = key.getState(editorState)) === null || _key$getState8 === void 0 ? void 0 : _key$getState8.lastDragCancelled) !== null && _key$getState$lastDra !== void 0 ? _key$getState$lastDra : false,
        isEditing: (_interactionTrackingP = interactionTrackingPluginKey.getState(editorState)) === null || _interactionTrackingP === void 0 ? void 0 : _interactionTrackingP.isEditing,
        isSelectedViaDragHandle: (_key$getState$isSelec = (_key$getState9 = key.getState(editorState)) === null || _key$getState9 === void 0 ? void 0 : _key$getState9.isSelectedViaDragHandle) !== null && _key$getState$isSelec !== void 0 ? _key$getState$isSelec : false
      };
    },
    contentComponent: function contentComponent(_ref8) {
      var editorView = _ref8.editorView,
        popupsMountPoint = _ref8.popupsMountPoint,
        popupsBoundariesElement = _ref8.popupsBoundariesElement,
        popupsScrollableElement = _ref8.popupsScrollableElement;
      return /*#__PURE__*/React.createElement(React.Fragment, null, editorExperiment('platform_editor_controls', 'variant1') ? /*#__PURE__*/React.createElement(BlockMenu, {
        editorView: editorView,
        mountPoint: popupsMountPoint,
        boundariesElement: popupsBoundariesElement,
        scrollableElement: popupsScrollableElement,
        api: api
      }) : /*#__PURE__*/React.createElement(DragHandleMenu, {
        api: api
      }), /*#__PURE__*/React.createElement(GlobalStylesWrapper, null));
    }
  };
};
