// @atlaskit/page/dist/es2019/Grid.js
var _a;
import React, { Component } from 'react';
import { ThemeProvider, withTheme } from 'styled-components';
import Grid from './internal/GridElement';
import { defaultGridColumns } from './internal/vars';
export default withTheme((_a = class AkGrid extends Component {
        constructor() {
            super(...arguments);
            this.getTheme = (props) => ({
                columns: props.theme && props.theme.columns
                    ? props.theme.columns
                    : defaultGridColumns,
                spacing: props.theme && props.theme.spacing
                    ? props.theme.spacing
                    : props.spacing,
                isNestedGrid: props.theme && props.theme.isNestedGrid
                    ? props.theme.isNestedGrid
                    : false,
            });
        }
        render() {
            const { layout, spacing, testId } = this.props;
            return (React.createElement(ThemeProvider, { theme: this.getTheme(this.props) },
                React.createElement(Grid, { "data-testid": testId, spacing: spacing, layout: layout }, this.props.children)));
        }
    },
    _a.defaultProps = {
        spacing: 'cosy',
        layout: 'fixed',
    },
    _a));
