// @atlaskit/page/dist/es2019/Page.js
import React, { Component } from 'react';
import styled, { ThemeProvider } from 'styled-components';
const Wrapper = styled.div `
  display: flex;
  flex-direction: column;
  min-height: 100%;
  width: 100%;
`;
const NavigationAndContent = styled.div `
  display: flex;
  flex: 1 1 auto;
`;
const BannerContainer = styled.div `
  flex: 1 0 auto;
  transition: height 0.25s ease-in-out;
  height: ${props => (props.isBannerOpen ? props.bannerHeight : 0)}px;
  position: relative;
  width: 100%;
  z-index: 3;
`;
const Banner = styled.div `
  position: fixed;
  width: 100%;
`;
const Navigation = styled.div `
  position: relative;
  z-index: 2;
`;
const PageContent = styled.div `
  flex: 1 1 auto;
  position: relative;
  z-index: 1;
  min-width: 0;
`;
const emptyTheme = {};
export default class Page extends Component {
    render() {
        const { isBannerOpen, banner, navigation, children, bannerHeight, } = this.props;
        return (React.createElement(ThemeProvider, { theme: emptyTheme },
            React.createElement(Wrapper, null,
                this.props.banner ? (React.createElement(BannerContainer, { "aria-hidden": !isBannerOpen, isBannerOpen: isBannerOpen, bannerHeight: bannerHeight },
                    React.createElement(Banner, null, banner))) : null,
                React.createElement(NavigationAndContent, null,
                    React.createElement(Navigation, null, navigation),
                    React.createElement(PageContent, null, children)))));
    }
}
Page.displayName = 'AkPage';
Page.defaultProps = {
    isBannerOpen: false,
    bannerHeight: 52,
};
