// @atlaskit/editor-plugin-help-dialog/dist/cjs/pm-plugins/keymap.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.keymapPlugin = void 0;
var _analytics = require("@atlaskit/editor-common/analytics");
var _keymaps = require("@atlaskit/editor-common/keymaps");
var _keymap = require("@atlaskit/editor-prosemirror/keymap");
var _commands = require("./commands");
var _pluginKey = require("./plugin-key");
var keymapPlugin = exports.keymapPlugin = function keymapPlugin(editorAnalyticsAPI) {
  var list = {};
  (0, _keymaps.bindKeymapWithCommand)(
  // Ignored via go/ees005
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  _keymaps.openHelp.common, function (state, dispatch) {
    var tr = state.tr;
    var isVisible = tr.getMeta(_pluginKey.pluginKey);
    if (!isVisible) {
      editorAnalyticsAPI === null || editorAnalyticsAPI === void 0 || editorAnalyticsAPI.attachAnalyticsEvent({
        action: _analytics.ACTION.CLICKED,
        actionSubject: _analytics.ACTION_SUBJECT.BUTTON,
        actionSubjectId: _analytics.ACTION_SUBJECT_ID.BUTTON_HELP,
        attributes: {
          inputMethod: _analytics.INPUT_METHOD.SHORTCUT
        },
        eventType: _analytics.EVENT_TYPE.UI
      })(tr);
      (0, _commands.openHelpCommand)(tr, dispatch);
    }
    return true;
  }, list);
  return (0, _keymap.keymap)(list);
};
