// @atlaskit/editor-plugin-highlight/dist/cjs/highlightPlugin.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.highlightPlugin = void 0;
var _react = _interopRequireDefault(require("react"));
var _adfSchema = require("@atlaskit/adf-schema");
var _experiments = require("@atlaskit/tmp-editor-statsig/experiments");
var _changeColor = require("./editor-commands/change-color");
var _keymap = require("./pm-plugins/keymap");
var _main = require("./pm-plugins/main");
var _FloatingToolbarHighlightColor = require("./ui/FloatingToolbarHighlightColor");
var _PrimaryToolbarHighlightColor = require("./ui/PrimaryToolbarHighlightColor");
var highlightPlugin = exports.highlightPlugin = function highlightPlugin(_ref) {
  var _api$analytics, _api$primaryToolbar;
  var api = _ref.api,
    options = _ref.config;
  var editorAnalyticsAPI = api === null || api === void 0 || (_api$analytics = api.analytics) === null || _api$analytics === void 0 ? void 0 : _api$analytics.actions;
  var primaryToolbarComponent = function primaryToolbarComponent(_ref2) {
    var popupsMountPoint = _ref2.popupsMountPoint,
      popupsBoundariesElement = _ref2.popupsBoundariesElement,
      popupsScrollableElement = _ref2.popupsScrollableElement,
      disabled = _ref2.disabled,
      isToolbarReducedSpacing = _ref2.isToolbarReducedSpacing,
      editorView = _ref2.editorView;
    return /*#__PURE__*/_react.default.createElement(_PrimaryToolbarHighlightColor.PrimaryToolbarHighlightColorWithIntl, {
      popupsMountPoint: popupsMountPoint,
      popupsBoundariesElement: popupsBoundariesElement,
      popupsScrollableElement: popupsScrollableElement,
      disabled: disabled,
      isToolbarReducedSpacing: isToolbarReducedSpacing,
      pluginInjectionApi: api,
      editorView: editorView
    });
  };
  api === null || api === void 0 || (_api$primaryToolbar = api.primaryToolbar) === null || _api$primaryToolbar === void 0 || _api$primaryToolbar.actions.registerComponent({
    name: 'highlight',
    component: primaryToolbarComponent
  });
  return {
    name: 'highlight',
    marks: function marks() {
      return [{
        name: 'backgroundColor',
        mark: _adfSchema.backgroundColor
      }];
    },
    commands: {
      changeColor: (0, _changeColor.changeColor)(editorAnalyticsAPI)
    },
    pmPlugins: function pmPlugins() {
      return [{
        name: 'highlight',
        plugin: function plugin() {
          return (0, _main.createPlugin)({
            api: api
          });
        }
      }, {
        name: 'highlightKeymap',
        plugin: function plugin() {
          return (0, _keymap.keymapPlugin)({
            api: api
          });
        }
      }];
    },
    getSharedState: function getSharedState(editorState) {
      if (!editorState) {
        return;
      }
      return _main.highlightPluginKey.getState(editorState);
    },
    pluginsOptions: {
      selectionToolbar: function selectionToolbar() {
        var _api$selectionToolbar;
        if ((api === null || api === void 0 || (_api$selectionToolbar = api.selectionToolbar) === null || _api$selectionToolbar === void 0 || (_api$selectionToolbar = _api$selectionToolbar.sharedState) === null || _api$selectionToolbar === void 0 || (_api$selectionToolbar = _api$selectionToolbar.currentState()) === null || _api$selectionToolbar === void 0 ? void 0 : _api$selectionToolbar.toolbarDocking) === 'none' && (0, _experiments.editorExperiment)('platform_editor_controls', 'variant1', {
          exposure: true
        })) {
          var toolbarCustom = {
            type: 'custom',
            render: function render(_view, _idx, dispatchAnalyticsEvent) {
              return /*#__PURE__*/_react.default.createElement(_FloatingToolbarHighlightColor.FloatingToolbarHighlightColorWithIntl, {
                dispatchAnalyticsEvent: dispatchAnalyticsEvent,
                pluginInjectionApi: api
              });
            },
            fallback: []
          };
          var rank = (0, _experiments.editorExperiment)('platform_editor_controls', 'variant1') ? 5 : -9;
          return {
            rank: rank,
            isToolbarAbove: true,
            items: [toolbarCustom],
            pluginName: 'highlight'
          };
        } else {
          return undefined;
        }
      }
    },
    primaryToolbarComponent: !(api !== null && api !== void 0 && api.primaryToolbar) ? primaryToolbarComponent : undefined
  };
};
