// @atlaskit/media-card/dist/cjs/utils/metadata.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getFileDetails = void 0;
var _mediaClient = require("@atlaskit/media-client");
var getProcessingStatusFromFileState = function getProcessingStatusFromFileState(status) {
  switch (status) {
    case 'processed':
      return 'succeeded';
    case 'processing':
      return 'running';
    case 'failed-processing':
      return 'failed';
  }
};
var getFileDetailsFromFileState = function getFileDetailsFromFileState(state) {
  return {
    id: state.id,
    name: state.name,
    size: state.size,
    mimeType: state.mimeType,
    createdAt: state.createdAt,
    mediaType: state.mediaType,
    processingStatus: getProcessingStatusFromFileState(state.status)
  };
};
var getFileDetails = exports.getFileDetails = function getFileDetails(identifier, fileState) {
  return (0, _mediaClient.isFileIdentifier)(identifier) ? fileState && !(0, _mediaClient.isErrorFileState)(fileState) ? getFileDetailsFromFileState(fileState) : {
    id: identifier.id
  } : {
    id: identifier.mediaItemType,
    name: identifier.name || identifier.dataURI,
    mediaType: 'image'
  };
};
