// @atlaskit/drawer/dist/esm/compiled/drawer-panel/drawer-close-button.js
import _extends from "@babel/runtime/helpers/extends";
import React from 'react';
import { IconButton } from '@atlaskit/button/new';
import ArrowLeft from '@atlaskit/icon/core/migration/arrow-left';
import { useEnsureIsInsideDrawer, useOnClose } from '../../context';
/**
 * __Drawer close button
 *
 * A button that closes the drawer.
 */
export var DrawerCloseButton = function DrawerCloseButton(_ref) {
  var Icon = _ref.icon,
    _ref$label = _ref.label,
    label = _ref$label === void 0 ? 'Close drawer' : _ref$label,
    _ref$testId = _ref.testId,
    testId = _ref$testId === void 0 ? 'DrawerCloseButton' : _ref$testId;
  useEnsureIsInsideDrawer();
  var onClose = useOnClose();
  return /*#__PURE__*/React.createElement(IconButton, {
    onClick: onClose,
    testId: testId,
    icon: Icon ? function (iconProps) {
      return /*#__PURE__*/React.createElement(Icon, _extends({}, iconProps, {
        size: "large",
        LEGACY_size: "large"
      }));
    } : ArrowLeft,
    label: label,
    shape: "circle",
    appearance: "subtle"
  });
};
