// @atlaskit/editor-plugin-text-color/dist/es2019/pm-plugins/main.js
import { SafePlugin } from '@atlaskit/editor-common/safe-plugin';
import { textColorPalette } from '@atlaskit/editor-common/ui-color';
import { PluginKey } from '@atlaskit/editor-prosemirror/state';
import { getActiveColor } from './utils/color';
import { DEFAULT_COLOR } from './utils/constants';
import { getDisabledState } from './utils/disabled';
function createInitialPluginState(editorState, pluginConfig) {
  const defaultColor = (pluginConfig === null || pluginConfig === void 0 ? void 0 : pluginConfig.defaultColor) || DEFAULT_COLOR;
  const palette = [{
    value: defaultColor.color,
    label: defaultColor.label,
    border: "var(--ds-border, #091E4224)"
  }, ...textColorPalette];
  const state = {
    color: getActiveColor(editorState),
    disabled: getDisabledState(editorState),
    palette,
    defaultColor: defaultColor.color
  };
  return state;
}
export let ACTIONS = /*#__PURE__*/function (ACTIONS) {
  ACTIONS[ACTIONS["RESET_COLOR"] = 0] = "RESET_COLOR";
  ACTIONS[ACTIONS["SET_COLOR"] = 1] = "SET_COLOR";
  ACTIONS[ACTIONS["DISABLE"] = 2] = "DISABLE";
  return ACTIONS;
}({});
export const pluginKey = new PluginKey('textColorPlugin');
export function createPlugin(dispatch, pluginConfig) {
  return new SafePlugin({
    key: pluginKey,
    state: {
      init(_config, editorState) {
        return createInitialPluginState(editorState, pluginConfig);
      },
      apply(tr, pluginState, _, newState) {
        const meta = tr.getMeta(pluginKey) || {};
        let nextState;
        switch (meta.action) {
          case ACTIONS.RESET_COLOR:
            nextState = {
              ...pluginState,
              color: pluginState.defaultColor
            };
            break;
          case ACTIONS.SET_COLOR:
            nextState = {
              ...pluginState,
              color: meta.color,
              disabled: false
            };
            break;
          case ACTIONS.DISABLE:
            nextState = {
              ...pluginState,
              disabled: true
            };
            break;
          default:
            nextState = {
              ...pluginState,
              color: getActiveColor(newState),
              disabled: getDisabledState(newState)
            };
        }
        if (pluginState && pluginState.color !== nextState.color || pluginState && pluginState.disabled !== nextState.disabled) {
          dispatch(pluginKey, nextState);
          return nextState;
        }
        return pluginState;
      }
    }
  });
}
