// @atlaskit/teams-public/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "TeamContainers", {
  enumerable: true,
  get: function get() {
    return _teamContainers.TeamContainers;
  }
});
Object.defineProperty(exports, "getContainerProperties", {
  enumerable: true,
  get: function get() {
    return _getContainerProperties.getContainerProperties;
  }
});
Object.defineProperty(exports, "hasProductPermission", {
  enumerable: true,
  get: function get() {
    return _utils.hasProductPermission;
  }
});
Object.defineProperty(exports, "useConnectedTeams", {
  enumerable: true,
  get: function get() {
    return _useTeamContainers.useConnectedTeams;
  }
});
Object.defineProperty(exports, "useProductPermissions", {
  enumerable: true,
  get: function get() {
    return _useProductPermission.useProductPermissions;
  }
});
Object.defineProperty(exports, "useTeamContainers", {
  enumerable: true,
  get: function get() {
    return _useTeamContainers.useTeamContainers;
  }
});
var _teamContainers = require("./ui/team-containers");
var _useTeamContainers = require("./controllers/hooks/use-team-containers");
var _useProductPermission = require("./controllers/hooks/use-product-permission");
var _utils = require("./controllers/product-permission/utils");
var _getContainerProperties = require("./common/utils/get-container-properties");
