// @atlaskit/jql-editor/dist/es2019/ui/error-boundary/index.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import React, { Component } from 'react';

// eslint-disable-next-line @atlassian/tangerine/import/no-parent-imports
import { JQLEditorReadOnly } from '../jql-editor-layout';
// eslint-disable-next-line @atlassian/tangerine/import/no-parent-imports

// eslint-disable-next-line @repo/internal/react/no-class-components
class ErrorBoundary extends Component {
  constructor(...args) {
    super(...args);
    _defineProperty(this, "state", {
      error: null
    });
  }
  componentDidCatch(error) {
    this.setState({
      error
    });
    this.props.onError && this.props.onError(error);
  }
  render() {
    const {
      children,
      render: renderFallback
    } = this.props;
    const {
      error
    } = this.state;
    if (error === null) {
      return children;
    }
    return renderFallback({
      error
    });
  }
}
export const withErrorBoundary = WrappedComponent => {
  return props => {
    const {
      query,
      onRenderError,
      onSearch,
      isCompact
    } = props;
    return /*#__PURE__*/React.createElement(ErrorBoundary, {
      onError: onRenderError,
      render: () => /*#__PURE__*/React.createElement(JQLEditorReadOnly, {
        query: query,
        isSearch: !!onSearch,
        isCompact: isCompact
      })
    }, /*#__PURE__*/React.createElement(WrappedComponent, props));
  };
};
