// @atlaskit/smart-card/dist/cjs/extractors/flexible/extract-priority.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _linkExtractors = require("@atlaskit/link-extractors");
var _constants = require("../../constants");
var priorityMapper = {
  blocker: _constants.IconType.PriorityBlocker,
  critical: _constants.IconType.PriorityCritical,
  high: _constants.IconType.PriorityHigh,
  highest: _constants.IconType.PriorityHighest,
  low: _constants.IconType.PriorityLow,
  lowest: _constants.IconType.PriorityLowest,
  major: _constants.IconType.PriorityMajor,
  medium: _constants.IconType.PriorityMedium,
  minor: _constants.IconType.PriorityMinor,
  trivial: _constants.IconType.PriorityTrivial,
  undefined: _constants.IconType.PriorityUndefined
};
var extractPriority = function extractPriority(data) {
  var priority = data['atlassian:priority'];
  if (priority) {
    if (typeof priority === 'string') {
      var icon = priorityMapper[priority.toLowerCase()];
      if (icon) {
        return {
          icon: icon,
          label: priority
        };
      }
    } else if (priority['@type'] === 'Object') {
      return {
        label: priority.name,
        url: priority.icon && (0, _linkExtractors.extractUrlFromIconJsonLd)(priority.icon)
      };
    }
  }
};
var _default = exports.default = extractPriority;
