// @atlaskit/smart-card/dist/cjs/extractors/flexible/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _linkExtractors = require("@atlaskit/link-extractors");
var _primitives = require("../common/primitives");
var _actions = require("./actions");
var _collaboratorGroup = require("./collaboratorGroup");
var _extractPreview = _interopRequireDefault(require("./extract-preview"));
var _extractPriority = _interopRequireDefault(require("./extract-priority"));
var _extractState = _interopRequireDefault(require("./extract-state"));
var _icon = require("./icon");
var _extractProviderIcon = _interopRequireDefault(require("./icon/extract-provider-icon"));
var _latestCommit = require("./latest-commit");
var _utils = require("./utils");
var _excluded = ["appearance", "fireEvent", "id", "origin", "renderers", "resolve", "actionOptions", "response", "aiSummaryConfig"];
var extractFlexibleUiContext = function extractFlexibleUiContext() {
  var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var appearance = _ref.appearance,
    fireEvent = _ref.fireEvent,
    id = _ref.id,
    origin = _ref.origin,
    renderers = _ref.renderers,
    resolve = _ref.resolve,
    actionOptions = _ref.actionOptions,
    response = _ref.response,
    aiSummaryConfig = _ref.aiSummaryConfig,
    props = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  if (!response) {
    return undefined;
  }
  var data = response.data;
  var url = (0, _linkExtractors.extractLink)(data);
  return {
    actions: (0, _actions.extractFlexibleCardActions)({
      actionOptions: actionOptions,
      aiSummaryConfig: aiSummaryConfig,
      appearance: appearance,
      fireEvent: fireEvent,
      id: id,
      origin: origin,
      response: response,
      url: props.url // Use the original URL in edge cases, such as short links for AI summary and copy link actions.
    }),
    appliedToComponentsCount: (0, _utils.extractAppliedToComponentsCount)(data),
    assignedToGroup: (0, _utils.extractPersonAssignedToAsArray)(data),
    attachmentCount: (0, _utils.extractAttachmentCount)(data),
    authorGroup: (0, _linkExtractors.extractPersonCreatedBy)(data),
    ownedByGroup: (0, _linkExtractors.extractPersonOwnedBy)(data),
    collaboratorGroup: (0, _collaboratorGroup.extractPersonsUpdatedBy)(data),
    commentCount: (0, _utils.extractCommentCount)(data),
    viewCount: (0, _utils.extractViewCount)(data),
    reactCount: (0, _utils.extractReactCount)(data),
    voteCount: (0, _utils.extractVoteCount)(data),
    checklistProgress: (0, _utils.extractChecklistProgress)(data),
    createdBy: (0, _utils.extractCreatedBy)(data),
    ownedBy: (0, _utils.extractOwnedBy)(data),
    assignedTo: (0, _utils.extractAssignedTo)(data),
    createdOn: (0, _linkExtractors.extractDateCreated)(data),
    dueOn: (0, _utils.extractDueOn)(data),
    latestCommit: (0, _latestCommit.extractLatestCommit)(data),
    linkIcon: (0, _icon.extractLinkIcon)(response, renderers),
    location: (0, _utils.extractLocation)(data),
    modifiedBy: (0, _utils.extractModifiedBy)(data),
    modifiedOn: (0, _linkExtractors.extractDateUpdated)(data),
    preview: (0, _extractPreview.default)(data),
    priority: (0, _extractPriority.default)(data),
    provider: (0, _extractProviderIcon.default)(data),
    programmingLanguage: (0, _utils.extractProgrammingLanguage)(data),
    readTime: (0, _utils.extractReadTime)(data),
    sentOn: (0, _utils.extractSentOn)(data),
    snippet: (0, _primitives.extractSummary)(data) || undefined,
    // Explicitly set here to remove an empty string
    sourceBranch: (0, _utils.extractSourceBranch)(data),
    state: (0, _extractState.default)(response, actionOptions, id, appearance, origin, fireEvent, resolve),
    subscriberCount: (0, _utils.extractSubscriberCount)(data),
    subTasksProgress: (0, _utils.extractSubTasksProgress)(data),
    storyPoints: (0, _utils.extractStoryPoints)(data),
    targetBranch: (0, _utils.extractTargetBranch)(data),
    title: (0, _linkExtractors.extractTitle)(data) || url,
    url: url,
    ari: (0, _linkExtractors.extractAri)(data)
  };
};
var _default = exports.default = extractFlexibleUiContext;
