// @atlaskit/editor-plugin-card/dist/es2019/ui/ConfigureOverlay/index.js
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
import { useCallback, useState } from 'react';

// not permitted to migrate atlaskit packages to compiled yet, see https://hello.atlassian.net/wiki/spaces/UAF/pages/3006969423/Migrating+AFM+platform+components+to+Compiled+a+guide
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled
import { css, jsx } from '@emotion/react';
import { OverlayButton } from '@atlaskit/editor-common/link';
const ConfigureOverlayWrapperStyles = css({
  position: 'relative',
  left: "var(--ds-space-025, 2px)"
});
const OverlayWrapper = ({
  view,
  targetElementPos,
  children,
  isHoveredCallback: hoverCallback,
  onOpenLinkClick
}) => {
  const [showConfigureButton, setShowConfigureButton] = useState(false);
  const [dropdownOpen, setDropdownOpen] = useState(false);
  const onDropdownChange = useCallback(isOpen => {
    setDropdownOpen(isOpen);
    if (!isOpen) {
      setShowConfigureButton(false);
    }
  }, []);
  return (
    // eslint-disable-next-line jsx-a11y/no-static-element-interactions
    jsx("span", {
      onMouseEnter: () => {
        setShowConfigureButton(true);
        hoverCallback(true);
      },
      onMouseLeave: () => {
        if (!dropdownOpen) {
          setShowConfigureButton(false);
          hoverCallback(false);
        }
      },
      "data-testid": "inline-card-overlay-wrapper"
    }, jsx("span", {
      css: ConfigureOverlayWrapperStyles
    }, showConfigureButton && jsx(OverlayButton, {
      editorView: view,
      targetElementPos: targetElementPos,
      onDropdownChange: onDropdownChange,
      onOpenLinkClick: onOpenLinkClick
    })), children)
  );
};
export default OverlayWrapper;
