// @atlaskit/frontend-utilities/dist/esm/hooks/use-local-storage/index.js
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import { useEffect, useMemo, useState } from 'react';
import { StorageClient } from '../../storage/storage-client';
export var useLocalStorage = function useLocalStorage(key, defaultValue) {
  var storageClient = useMemo(function () {
    return new StorageClient(key);
  }, [key]);
  var _useState = useState(function () {
      var _storageClient$getIte;
      return (_storageClient$getIte = storageClient.getItem(key)) !== null && _storageClient$getIte !== void 0 ? _storageClient$getIte : defaultValue;
    }),
    _useState2 = _slicedToArray(_useState, 2),
    value = _useState2[0],
    setValue = _useState2[1];
  useEffect(function () {
    var _ref;
    var loadedValue = (_ref = storageClient.getItem(key)) !== null && _ref !== void 0 ? _ref : undefined;
    if (loadedValue !== undefined) {
      setValue(loadedValue);
    }
  }, [key, storageClient]);
  useEffect(function () {
    storageClient.setItemWithExpiry(key, value);
  }, [key, storageClient, value]);
  return [value, setValue];
};
export function useLocalStorageRecord(key) {
  var initialValue = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
  var maxLength = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 100;
  var _useLocalStorage = useLocalStorage(key, initialValue),
    _useLocalStorage2 = _slicedToArray(_useLocalStorage, 2),
    records = _useLocalStorage2[0],
    setRecords = _useLocalStorage2[1];
  var putRecord = function putRecord(record) {
    if (!records) {
      setRecords([record]);
      return;
    }
    //just to keep storage limited somehow
    if (maxLength > 0 && records.length > maxLength - 1) {
      records.shift();
    }
    if (records.find(function (elem) {
      return JSON.stringify(elem) === JSON.stringify(record);
    }) === undefined) {
      setRecords([].concat(_toConsumableArray(records), [record]));
    }
  };
  var removeRecord = function removeRecord(query) {
    if (!records) {
      return;
    }
    var filtered = records.filter(function (elem) {
      return JSON.stringify(elem).indexOf("\"".concat(query, "\"")) === -1;
    });
    setRecords(filtered);
  };
  var actions = {
    putRecord: putRecord,
    removeRecord: removeRecord
  };
  return {
    records: records,
    actions: actions
  };
}
