// @atlaskit/editor-core/dist/cjs/ui/CollapsedEditor/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireDefault(require("react"));
var _ChromeCollapsed = _interopRequireDefault(require("../ChromeCollapsed"));
var _IntlProviderIfMissingWrapper = require("../IntlProviderIfMissingWrapper/IntlProviderIfMissingWrapper");
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
// Ignored via go/ees005
// eslint-disable-next-line @repo/internal/react/no-class-components
var CollapsedEditor = exports.default = /*#__PURE__*/function (_React$Component) {
  function CollapsedEditor() {
    var _this;
    (0, _classCallCheck2.default)(this, CollapsedEditor);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, CollapsedEditor, [].concat(args));
    (0, _defineProperty2.default)(_this, "handleEditorRef", function (editorRef, editorRefCallback) {
      if (editorRefCallback && typeof editorRefCallback === 'function') {
        editorRefCallback(editorRef);
      }
      _this.editorComponent = editorRef;
    });
    return _this;
  }
  (0, _inherits2.default)(CollapsedEditor, _React$Component);
  return (0, _createClass2.default)(CollapsedEditor, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      if (this.props.isExpanded && (this.editorComponent || this.functionalEditor) && (!this.previouslyExpanded || this.previouslyExpanded === undefined)) {
        var _this$props$onExpand, _this$props;
        (_this$props$onExpand = (_this$props = this.props).onExpand) === null || _this$props$onExpand === void 0 || _this$props$onExpand.call(_this$props);
      }
      this.previouslyExpanded = this.props.isExpanded;
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;
      var child = _react.default.Children.only(this.props.children);
      this.functionalEditor = typeof child.type === 'function';
      if (!this.props.isExpanded) {
        return /*#__PURE__*/_react.default.createElement(_IntlProviderIfMissingWrapper.IntlProviderIfMissingWrapper, null, /*#__PURE__*/_react.default.createElement(_ChromeCollapsed.default, {
          onFocus: this.props.onFocus,
          text: this.props.placeholder,
          label: this.props.assistiveLabel
        }));
      }

      // Let's avoid ref logic for functional Editor
      if (this.functionalEditor) {
        return child;
      }
      return /*#__PURE__*/_react.default.cloneElement(child, {
        ref: function ref(editorComponent) {
          return _this2.handleEditorRef(editorComponent, child.ref);
        }
      });
    }
  }]);
}(_react.default.Component);
