// @atlaskit/editor-common/dist/cjs/ui/FloatingToolbar/Separator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("@emotion/react");
var _experiments = require("@atlaskit/tmp-editor-statsig/experiments");
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var separator = (0, _react.css)({
  background: "var(--ds-border, #091E4224)",
  width: '1px',
  height: '20px',
  margin: "0 ".concat("var(--ds-space-050, 4px)"),
  alignSelf: 'center'
});
var separatorNew = (0, _react.css)({
  background: "var(--ds-border, #091E4224)",
  width: '1px',
  height: '20px',
  alignSelf: 'center'
});
var separatorFullHeight = (0, _react.css)({
  height: '40px'
});
var _default = exports.default = function _default(_ref) {
  var fullHeight = _ref.fullHeight;
  return (0, _react.jsx)("div", {
    css: [(0, _experiments.editorExperiment)('platform_editor_controls', 'variant1') ? separatorNew : separator, fullHeight && separatorFullHeight]
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    ,
    className: "separator"
  });
};
