// @atlaskit/link-picker/dist/es2019/ui/link-picker/form-footer/link-picker-submit-button/index.js
import React from 'react';
import { defineMessages, useIntl } from 'react-intl-next';
import LoadingButton from '@atlaskit/button/loading-button';
import Button from '@atlaskit/button/new';
import { fg } from '@atlaskit/platform-feature-flags';
import { checkSubmitDisabled } from './utils';
export const messages = defineMessages({
  saveButton: {
    id: 'fabric.linkPicker.button.save',
    defaultMessage: 'Save',
    description: 'Button to save edited link'
  },
  insertButton: {
    id: 'fabric.linkPicker.button.insert',
    defaultMessage: 'Insert',
    description: 'Button to insert searched or selected link'
  }
});
export const LinkPickerSubmitButton = ({
  isEditing,
  isLoading,
  isSubmitting,
  customSubmitButtonLabel,
  error,
  items,
  queryState,
  submitMessageId,
  testId,
  url
}) => {
  const intl = useIntl();
  const insertButtonMsg = isEditing ? messages.saveButton : messages.insertButton;
  const isSubmitDisabled = checkSubmitDisabled(isLoading, isSubmitting, error, url, queryState, items);
  const ButtonComponent = fg('platform-link-picker-remove-legacy-button') ? Button : LoadingButton;
  return /*#__PURE__*/React.createElement(ButtonComponent, {
    type: "submit",
    appearance: "primary",
    testId: testId,
    isDisabled: isSubmitDisabled,
    "aria-labelledby": isSubmitting ? submitMessageId : undefined,
    isLoading: isSubmitting
  }, customSubmitButtonLabel ? intl.formatMessage(customSubmitButtonLabel) : intl.formatMessage(insertButtonMsg));
};
