// @atlaskit/link-datasource/dist/cjs/ui/issue-like-table/render-type/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.stringifyType = exports.renderType = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _boolean = _interopRequireDefault(require("./boolean"));
var _dateTime = _interopRequireWildcard(require("./date-time"));
var _icon = _interopRequireDefault(require("./icon"));
var _link = _interopRequireDefault(require("./link"));
var _newIcon = _interopRequireDefault(require("./new-icon"));
var _number = _interopRequireDefault(require("./number"));
var _richtext = _interopRequireWildcard(require("./richtext"));
var _status = _interopRequireDefault(require("./status"));
var _tag = _interopRequireDefault(require("./tag"));
var _text = _interopRequireDefault(require("./text"));
var _user = _interopRequireDefault(require("./user"));
var _messages = require("./user/messages");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var stringifyType = exports.stringifyType = function stringifyType(_ref, formatMessage, formatDate) {
  var type = _ref.type,
    value = _ref.value;
  switch (type) {
    case 'boolean':
    case 'number':
      return (value === null || value === void 0 ? void 0 : value.toString()) || '';
    case 'date':
      return (0, _dateTime.getFormattedDate)(value, 'date', formatDate);
    case 'datetime':
      return (0, _dateTime.getFormattedDate)(value, 'datetime', formatDate);
    case 'time':
      return (0, _dateTime.getFormattedDate)(value, 'time', formatDate);
    case 'icon':
      return (value === null || value === void 0 ? void 0 : value.label) || '';
    case 'status':
      return (value === null || value === void 0 ? void 0 : value.text.toString()) || '';
    case 'string':
      return value;
    case 'tag':
      return (value === null || value === void 0 ? void 0 : value.text) || '';
    case 'user':
      return (value === null || value === void 0 ? void 0 : value.displayName) || formatMessage(_messages.userTypeMessages.userDefaultdisplayNameValue);
    case 'richtext':
      var adfPlainText = (0, _richtext.parseRichText)(value);
      return adfPlainText || '';
    case 'link':
    default:
      return '';
  }
};
var renderType = exports.renderType = function renderType(item) {
  switch (item.type) {
    case 'boolean':
      return item.values.map(function (booleanValue) {
        return /*#__PURE__*/_react.default.createElement(_boolean.default, {
          value: booleanValue
        });
      });
    case 'date':
      return item.values.map(function (dateValue) {
        return /*#__PURE__*/_react.default.createElement(_dateTime.default, {
          value: dateValue,
          display: "date"
        });
      });
    case 'datetime':
      return item.values.map(function (datTimeValue) {
        return /*#__PURE__*/_react.default.createElement(_dateTime.default, {
          value: datTimeValue,
          display: "datetime"
        });
      });
    case 'icon':
      return (0, _platformFeatureFlags.fg)('platform-datasources-enable-two-way-sync-priority') ? item.values.map(function (iconValue) {
        return /*#__PURE__*/_react.default.createElement(_newIcon.default, iconValue);
      }) : item.values.map(function (iconValue) {
        return /*#__PURE__*/_react.default.createElement(_icon.default, iconValue);
      });
    case 'link':
      return item.values.map(function (linkValue) {
        return /*#__PURE__*/_react.default.createElement(_link.default, (0, _extends2.default)({
          key: linkValue === null || linkValue === void 0 ? void 0 : linkValue.url
        }, linkValue));
      });
    case 'number':
      return item.values.map(function (numberValue) {
        return /*#__PURE__*/_react.default.createElement(_number.default, {
          number: numberValue
        });
      });
    case 'status':
      return item.values.map(function (statusValue) {
        return /*#__PURE__*/_react.default.createElement(_status.default, statusValue);
      });
    case 'string':
      return item.values.map(function (stringValue) {
        return /*#__PURE__*/_react.default.createElement(_text.default, {
          key: stringValue,
          text: stringValue
        });
      });
    case 'tag':
      return item.values.map(function (tagValue) {
        return /*#__PURE__*/_react.default.createElement(_tag.default, {
          key: (tagValue === null || tagValue === void 0 ? void 0 : tagValue.id) || (tagValue === null || tagValue === void 0 ? void 0 : tagValue.text),
          tag: tagValue
        });
      });
    case 'time':
      return item.values.map(function (timeValue) {
        return /*#__PURE__*/_react.default.createElement(_dateTime.default, {
          value: timeValue,
          display: "time"
        });
      });
    case 'user':
      return /*#__PURE__*/_react.default.createElement(_user.default, {
        users: item.values
      });
    case 'richtext':
      return item.values.map(function (richValue) {
        return /*#__PURE__*/_react.default.createElement(_richtext.default, {
          value: richValue
        });
      });
    default:
      return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null);
  }
};
