// @atlaskit/media-card/dist/es2019/root/styled.js
import styled from 'styled-components';
import { borderRadius } from '@atlaskit/theme/constants';
import { getCSSUnitValue } from '../utils/getCSSUnitValue';
import { breakpointStyles } from '../utils/breakpoint';
import { getSelectedBorderStyle } from '../styles/getSelectedBorderStyle';
const getWrapperHeight = (dimensions) => dimensions && dimensions.height
    ? `height: ${getCSSUnitValue(dimensions.height)}; max-height: 100%;`
    : '';
const getWrapperWidth = (dimensions) => dimensions && dimensions.width
    ? `width: ${getCSSUnitValue(dimensions.width)}; max-width: 100%;`
    : '';
export const Wrapper = styled.div `
  ${({ dimensions, breakpointSize = 'medium', shouldUsePointerCursor, }) => {
    return `
      ${breakpointStyles({ breakpointSize })}
      ${getWrapperHeight(dimensions)}
      ${getWrapperWidth(dimensions)}
      cursor: ${shouldUsePointerCursor ? 'pointer' : 'default'};
    `;
}};
`;
Wrapper.displayName = 'CardViewWrapper';
export const InlinePlayerWrapper = styled.div `
  overflow: hidden;
  border-radius: ${borderRadius()};
  position: relative;

  max-width: 100%;
  max-height: 100%;

  ${getSelectedBorderStyle}

  video {
    width: 100%;
    height: 100%;
  }
`;
InlinePlayerWrapper.displayName = 'InlinePlayerWrapper';
