// @atlaskit/smart-card/dist/esm/extractors/inline/extractPropsFromSourceCodePullRequest.js
import { __assign } from "tslib";
import React from 'react';
import PullRequestIcon from '@atlaskit/icon-object/glyph/pull-request/16';
import { extractInlineViewPropsFromObject } from './extractPropsFromObject';
import { buildName } from './extractPropsFromSourceCodeCommon';
export var pullRequestStateToLozengeMap = {
    open: 'inprogress',
    merged: 'success',
    declined: 'removed',
    closed: 'success',
};
var buildInlineSourceCodePullRequestTag = function (json) {
    if (json['atlassian:state']) {
        var state = json['atlassian:state'].toLowerCase();
        return {
            lozenge: {
                appearance: pullRequestStateToLozengeMap[state] || 'default',
                text: state,
            },
        };
    }
    return {};
};
export var buildIcon = function (json) {
    var name = json.name;
    return { icon: React.createElement(PullRequestIcon, { label: name }) };
};
export var extractInlineViewPropsFromSourceCodePullRequest = function (json) {
    var props = extractInlineViewPropsFromObject(json);
    return __assign(__assign(__assign(__assign({}, props), buildIcon(json)), buildName(props, json)), buildInlineSourceCodePullRequestTag(json));
};
