// @atlaskit/profilecard/dist/cjs/components/team-profile-card/team-containers-skeleton/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TeamContainersSkeleton = void 0;
require("./index.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _react = _interopRequireDefault(require("react"));
var _compiled = require("@atlaskit/primitives/compiled");
var _linkedContainerCardSkeleton = require("./linked-container-card-skeleton");
var styles = {
  showMoreButtonSkeleton: "_2rko1q77 _1rjc1b66 _1bsbvbgk"
};
var TeamContainersSkeleton = exports.TeamContainersSkeleton = function TeamContainersSkeleton(_ref) {
  var numberOfContainers = _ref.numberOfContainers;
  return /*#__PURE__*/_react.default.createElement(_compiled.Stack, {
    space: "space.200",
    testId: "profile-card-team-containers-skeleton"
  }, /*#__PURE__*/_react.default.createElement(_compiled.Stack, {
    space: "space.100"
  }, (0, _toConsumableArray2.default)(Array(numberOfContainers)).map(function (_, index) {
    return /*#__PURE__*/_react.default.createElement(_linkedContainerCardSkeleton.LinkedContainerCardSkeleton, {
      key: index
    });
  })), /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    backgroundColor: "color.background.neutral",
    xcss: styles.showMoreButtonSkeleton
  }));
};
