// @atlaskit/atlassian-navigation/dist/cjs/components/BadgeContainer/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BadgeContainer = void 0;
var _react = require("@emotion/react");
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var badgeContainerStyles = (0, _react.css)({
  position: 'absolute',
  insetBlockStart: "var(--ds-space-negative-050, -4px)",
  insetInlineEnd: "var(--ds-space-negative-025, -2px)",
  pointerEvents: 'none'
});
var containerStyles = (0, _react.css)({
  position: 'relative'
});

// Not exported to consumers
// eslint-disable-next-line @repo/internal/react/require-jsdoc
var BadgeContainer = exports.BadgeContainer = function BadgeContainer(props) {
  var badgeId = props.id,
    Badge = props.badge,
    children = props.children,
    role = props.role;
  return (0, _react.jsx)("div", {
    css: containerStyles,
    role: role
  }, children, (0, _react.jsx)("div", {
    css: badgeContainerStyles,
    id: badgeId,
    "aria-hidden": true
  }, (0, _react.jsx)(Badge, null)));
};
