// @atlaskit/editor-plugin-mentions/dist/es2019/ui/ToolbarMention/index.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import React, { PureComponent } from 'react';
import { injectIntl } from 'react-intl-next';
import { mentionMessages as messages } from '@atlaskit/editor-common/messages';
import { TOOLBAR_BUTTON, ToolbarButton } from '@atlaskit/editor-common/ui-menu';
import MentionIcon from '@atlaskit/icon/core/migration/mention--editor-mention';
// Ignored via go/ees005
// eslint-disable-next-line @repo/internal/react/no-class-components
class ToolbarMention extends PureComponent {
  constructor(...args) {
    super(...args);
    _defineProperty(this, "handleInsertMention", () => {
      if (!this.props.editorView) {
        return false;
      }
      this.props.onInsertMention();
      return true;
    });
  }
  render() {
    const mentionStringTranslated = this.props.intl.formatMessage(messages.mentionsIconLabel);
    return /*#__PURE__*/React.createElement(ToolbarButton, {
      testId: this.props.testId,
      buttonId: TOOLBAR_BUTTON.MENTION,
      spacing: "none",
      onClick: this.handleInsertMention,
      disabled: this.props.isDisabled,
      title: mentionStringTranslated + '@',
      iconBefore: /*#__PURE__*/React.createElement(MentionIcon, {
        label: mentionStringTranslated
      })
    });
  }
}
export default injectIntl(ToolbarMention);
