// @atlaskit/jql-editor/dist/cjs/ui/jql-editor-controls-content/read-only-controls-content/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ReadOnlyControlsContent = void 0;
var _react = _interopRequireDefault(require("react"));
var _noop = _interopRequireDefault(require("lodash/noop"));
var _useEditorTheme = require("../../../hooks/use-editor-theme");
var _baseExpandToggle = require("../base-expand-toggle");
var _baseSearch = require("../base-search");
var _baseSyntaxHelp = require("../base-syntax-help");
// eslint-disable-next-line @atlassian/tangerine/import/no-parent-imports

// eslint-disable-next-line @atlassian/tangerine/import/no-parent-imports

// eslint-disable-next-line @atlassian/tangerine/import/no-parent-imports

var ReadOnlyControlsContent = exports.ReadOnlyControlsContent = function ReadOnlyControlsContent() {
  var _useEditorThemeContex = (0, _useEditorTheme.useEditorThemeContext)(),
    isSearch = _useEditorThemeContex.isSearch;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_baseExpandToggle.BaseExpandToggle
  // The label here is for avoiding a11y violations only
  // This component ReadOnlyControlsContent is only used as a fallback for a loading state
  , {
    label: 'Editor',
    editorId: '',
    expanded: false,
    onClick: _noop.default,
    isDisabled: true
  }), /*#__PURE__*/_react.default.createElement(_baseSyntaxHelp.BaseSyntaxHelp, {
    label: '',
    onClick: _noop.default,
    isDisabled: true
  }), isSearch && /*#__PURE__*/_react.default.createElement(_baseSearch.BaseSearch
  // The label here is for avoiding a11y violations only
  // This component ReadOnlyControlsContent is only used as a fallback for a loading state
  , {
    label: 'Search',
    onSearch: _noop.default,
    isDisabled: true
  }));
};
