// @atlaskit/editor-shared-styles/dist/cjs/consts/consts.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.editorFontSize = exports.breakoutWideScaleRatio = exports.blockNodesVerticalMargin = exports.avatarColors = exports.akRichMediaResizeZIndex = exports.akLayoutGutterOffset = exports.akEditorWrappedNodeZIndex = exports.akEditorWideLayoutWidth = exports.akEditorUnitZIndex = exports.akEditorToolbarKeylineHeight = exports.akEditorTableToolbarSize = exports.akEditorTableToolbar = exports.akEditorTableNumberColumnWidth = exports.akEditorTableLegacyCellMinWidth = exports.akEditorTableHeaderCellBackgroundDark = exports.akEditorTableHeaderCellBackground = exports.akEditorTableCellOnStickyHeaderZIndex = exports.akEditorTableCellMinWidth = exports.akEditorTableCellBlanketSelected = exports.akEditorTableCellBlanketDeleted = exports.akEditorTableCellBackgroundOpacity = exports.akEditorTableBorderSelected = exports.akEditorTableBorderRadius = exports.akEditorTableBorder = exports.akEditorSwoopCubicBezier = exports.akEditorSubtleAccent = exports.akEditorStickyHeaderZIndex = exports.akEditorSmallZIndex = exports.akEditorShadowZIndex = exports.akEditorSelectedNodeClassName = exports.akEditorSelectedBoxShadow = exports.akEditorSelectedBorderSize = exports.akEditorSelectedBorderColor = exports.akEditorSelectedBorderBoldSize = exports.akEditorSelectedBorder = exports.akEditorSelectedBoldBoxShadow = exports.akEditorSelectedBlanketOpacity = exports.akEditorRuleBorderRadius = exports.akEditorMobileMaxWidth = exports.akEditorMobileBreakoutPoint = exports.akEditorMenuZIndex = exports.akEditorMediaResizeHandlerPaddingWide = exports.akEditorMediaResizeHandlerPadding = exports.akEditorLineHeight = exports.akEditorGutterPaddingDynamic = exports.akEditorGutterPadding = exports.akEditorGridLineZIndex = exports.akEditorFullWidthLayoutWidth = exports.akEditorFullWidthLayoutLineLength = exports.akEditorFullPageMaxWidth = exports.akEditorFullPageDefaultFontSize = exports.akEditorFloatingPanelZIndex = exports.akEditorFloatingOverlapPanelZIndex = exports.akEditorFloatingDialogZIndex = exports.akEditorDeleteIconColor = exports.akEditorDeleteBorder = exports.akEditorDeleteBackgroundWithOpacity = exports.akEditorDeleteBackground = exports.akEditorDefaultLayoutWidth = exports.akEditorCustomIconSize = exports.akEditorContextPanelWidth = exports.akEditorCodeInlinePadding = exports.akEditorCodeFontFamily = exports.akEditorCodeBlockPadding = exports.akEditorCodeBackground = exports.akEditorCalculatedWideLayoutWidthSmallViewport = exports.akEditorCalculatedWideLayoutWidth = exports.akEditorBreakoutPadding = exports.akEditorBlockquoteBorderColor = exports.VIEWPORT_SIZES = exports.MAX_BROWSER_SCROLLBAR_HEIGHT = exports.FULL_PAGE_EDITOR_TOOLBAR_HEIGHT = exports.DEFAULT_EMBED_CARD_WIDTH = exports.DEFAULT_EMBED_CARD_HEIGHT = exports.ATLASSIAN_NAVIGATION_HEIGHT = void 0;
exports.getTableCellBackgroundDarkModeColorCSS = getTableCellBackgroundDarkModeColorCSS;
exports.relativeSize = exports.relativeFontSizeToBase16 = exports.participantColors = exports.layoutBreakpointWidth = exports.gridMediumMaxWidth = exports.getTableCellBackgroundDarkModeColors = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _experiments = require("@atlaskit/tmp-editor-statsig/experiments");
/* eslint-disable @atlaskit/design-system/ensure-design-token-usage */

var akEditorFullPageDefaultFontSize = exports.akEditorFullPageDefaultFontSize = 16;
var akEditorCodeFontFamily = exports.akEditorCodeFontFamily = "var(--ds-font-family-code, ui-monospace, Menlo, \"Segoe UI Mono\", \"Ubuntu Mono\", monospace)";
var akEditorSubtleAccent = exports.akEditorSubtleAccent = "var(--ds-background-accent-gray-subtler, #DCDFE4)";
var akEditorBlockquoteBorderColor = exports.akEditorBlockquoteBorderColor = "var(--ds-border, #091E4224)";
var akEditorCodeBackground = exports.akEditorCodeBackground = "var(--ds-background-accent-gray-subtlest, #F1F2F4)";
var akEditorCodeBlockPadding = exports.akEditorCodeBlockPadding = '12px';
var akEditorCodeInlinePadding = exports.akEditorCodeInlinePadding = '2px 4px';
var akEditorDeleteBackground = exports.akEditorDeleteBackground = "var(--ds-background-danger, #FFECEB)";
var akEditorDeleteBackgroundWithOpacity = exports.akEditorDeleteBackgroundWithOpacity = 'rgba(255, 189, 173, 0.5)'; // R75 with 50% opacity
var akEditorDeleteBorder = exports.akEditorDeleteBorder = "var(--ds-border-danger, #E2483D)";
var akEditorDeleteIconColor = exports.akEditorDeleteIconColor = "var(--ds-icon-danger, #C9372C)";
var akEditorCustomIconSize = exports.akEditorCustomIconSize = 20;
var akEditorSelectedBorderColor = exports.akEditorSelectedBorderColor = "var(--ds-border-selected, #0C66E4)";
var akEditorSelectedBorderSize = exports.akEditorSelectedBorderSize = 1;
var akEditorSelectedBorderBoldSize = exports.akEditorSelectedBorderBoldSize = 2;
var akEditorSelectedBorder = exports.akEditorSelectedBorder = "".concat(akEditorSelectedBorderSize, "px solid ", "var(--ds-border-selected, #0C66E4)");
var akEditorSelectedBoxShadow = exports.akEditorSelectedBoxShadow = "0 0 0 ".concat(akEditorSelectedBorderSize, "px ", "var(--ds-border-selected, #0C66E4)");
var akEditorSelectedBoldBoxShadow = exports.akEditorSelectedBoldBoxShadow = "0 0 0 ".concat(akEditorSelectedBorderBoldSize, "px ", "var(--ds-border-selected, #0C66E4)");
var akEditorSelectedBlanketOpacity = exports.akEditorSelectedBlanketOpacity = 0.3;
var akEditorUnitZIndex = exports.akEditorUnitZIndex = 1;
var akEditorShadowZIndex = exports.akEditorShadowZIndex = 2;
var akEditorStickyHeaderZIndex = exports.akEditorStickyHeaderZIndex = 11; // filmstrip uses 10 for its shadow
var akEditorSmallZIndex = exports.akEditorSmallZIndex = akEditorStickyHeaderZIndex + 1;
var akEditorGridLineZIndex = exports.akEditorGridLineZIndex = 2;

// z-index for main menu bar -
// this is highest as it should be above anything else in editor below.
var akEditorMenuZIndex = exports.akEditorMenuZIndex = 500;
// z-index used for floating toolbars like code block, table etc
var akEditorFloatingPanelZIndex = exports.akEditorFloatingPanelZIndex = 400;
// z-index used for pickers (date, emoji, mentions) and type-aheads, hyperlinks
var akEditorFloatingDialogZIndex = exports.akEditorFloatingDialogZIndex = akEditorMenuZIndex + 10;
// z-index used for table cell menu options button on a sticky header
var akEditorTableCellOnStickyHeaderZIndex = exports.akEditorTableCellOnStickyHeaderZIndex = akEditorFloatingDialogZIndex - 5;
// z-index used for floating toolbars table cell menu which are above block toolbars
var akEditorFloatingOverlapPanelZIndex = exports.akEditorFloatingOverlapPanelZIndex = akEditorFloatingPanelZIndex + 5;
var akEditorTableToolbarSize = exports.akEditorTableToolbarSize = 11;
var akEditorTableBorder = exports.akEditorTableBorder = "var(--ds-background-accent-gray-subtler, #DCDFE4)";
var akEditorTableToolbar = exports.akEditorTableToolbar = "var(--ds-background-accent-gray-subtlest, #F1F2F4)";
var akEditorTableHeaderCellBackground = exports.akEditorTableHeaderCellBackground = '#F1F2F4';
var akEditorTableHeaderCellBackgroundDark = exports.akEditorTableHeaderCellBackgroundDark = '#2C333A';
var akEditorTableBorderSelected = exports.akEditorTableBorderSelected = "var(--ds-border-focused, #388BFF)";
var akEditorTableBorderRadius = exports.akEditorTableBorderRadius = '3px';
var akEditorTableCellBackgroundOpacity = exports.akEditorTableCellBackgroundOpacity = 0.5;
var akEditorFullPageMaxWidth = exports.akEditorFullPageMaxWidth = 680;
var akEditorDefaultLayoutWidth = exports.akEditorDefaultLayoutWidth = 760;
var akEditorWideLayoutWidth = exports.akEditorWideLayoutWidth = 960; // This value is used for tables with breakouts and is used to calculate `breakoutWidthPx`
var akEditorFullWidthLayoutWidth = exports.akEditorFullWidthLayoutWidth = 1800;
var akEditorFullWidthLayoutLineLength = exports.akEditorFullWidthLayoutLineLength = 1792;
var akEditorTableNumberColumnWidth = exports.akEditorTableNumberColumnWidth = 42;
var akEditorBreakoutPadding = exports.akEditorBreakoutPadding = 96;
var akEditorGutterPadding = exports.akEditorGutterPadding = 32;
var akEditorGutterPaddingDynamic = exports.akEditorGutterPaddingDynamic = function akEditorGutterPaddingDynamic() {
  return (
    // eslint-disable-next-line @atlaskit/platform/no-preconditioning
    (0, _platformFeatureFlags.fg)('platform_editor_controls_increase_full_page_gutter') && (0, _experiments.editorExperiment)('platform_editor_controls', 'variant1') ? 72 : 52
  );
};
var akEditorMobileBreakoutPoint = exports.akEditorMobileBreakoutPoint = 720;
var akEditorTableCellMinWidth = exports.akEditorTableCellMinWidth = 48;
var akEditorTableLegacyCellMinWidth = exports.akEditorTableLegacyCellMinWidth = 128;
var akEditorMediaResizeHandlerPaddingWide = exports.akEditorMediaResizeHandlerPaddingWide = 12;
var akEditorMediaResizeHandlerPadding = exports.akEditorMediaResizeHandlerPadding = 4;
var akEditorSwoopCubicBezier = exports.akEditorSwoopCubicBezier = "cubic-bezier(0.15, 1, 0.3, 1)";
var gridMediumMaxWidth = exports.gridMediumMaxWidth = 1024;
var breakoutWideScaleRatio = exports.breakoutWideScaleRatio = 1.33;
var akEditorCalculatedWideLayoutWidthSmallViewport = exports.akEditorCalculatedWideLayoutWidthSmallViewport = 905; // from breakoutConsts.calcWideWidth, layoutMaxWidth * breakoutConsts.wideScaleRatio = 904.8 ~ 905 This is a resulting width value that is applied to nodes that currently use breakouts (except table) and are set to `wide` when the viewport's width is <= 1266px.
var akEditorCalculatedWideLayoutWidth = exports.akEditorCalculatedWideLayoutWidth = 1011; // = akEditorDefaultLayoutWidth * breakoutWideScaleRatio = 1010.8 ~ 1011 This is a resulting width value that is applied to nodes that currently use breakouts (except table) and are set to `wide` when the viewport's width is > 1329px.
var akRichMediaResizeZIndex = exports.akRichMediaResizeZIndex = akEditorUnitZIndex * 99;
var akLayoutGutterOffset = exports.akLayoutGutterOffset = 12;
var akEditorLineHeight = exports.akEditorLineHeight = 1.714;
var akEditorRuleBorderRadius = exports.akEditorRuleBorderRadius = '1px';
var akEditorToolbarKeylineHeight = exports.akEditorToolbarKeylineHeight = 2;
var akEditorContextPanelWidth = exports.akEditorContextPanelWidth = 320;
var akEditorTableCellBlanketSelected = exports.akEditorTableCellBlanketSelected = 'rgba(179, 212, 255, 0.3)';
var akEditorTableCellBlanketDeleted = exports.akEditorTableCellBlanketDeleted = 'rgba(255, 235, 230, 0.3)';
var akEditorWrappedNodeZIndex = exports.akEditorWrappedNodeZIndex = akEditorUnitZIndex * 2; //prevent paragraph nodes from rendering above wrapped nodes

var layoutBreakpointWidth = exports.layoutBreakpointWidth = {
  SMALL: 550,
  MEDIUM: 760
};

// @see typography spreadsheet: https://docs.google.com/spreadsheets/d/1iYusRGCT4PoPfvxbJ8NrgjtfFgXLm5lpDWXzjua1W2E/edit#gid=93913128
var blockNodesVerticalMargin = exports.blockNodesVerticalMargin = '0.75rem';
var DEFAULT_EMBED_CARD_WIDTH = exports.DEFAULT_EMBED_CARD_WIDTH = 680;
var DEFAULT_EMBED_CARD_HEIGHT = exports.DEFAULT_EMBED_CARD_HEIGHT = 480;
var MAX_BROWSER_SCROLLBAR_HEIGHT = exports.MAX_BROWSER_SCROLLBAR_HEIGHT = 20;

// eslint-disable-next-line @repo/internal/deprecations/deprecation-ticket-required -- Ignored via go/ED-25883
// @deprecated
var ATLASSIAN_NAVIGATION_HEIGHT = exports.ATLASSIAN_NAVIGATION_HEIGHT = '56px';
var DEFAULT_FONT_SIZE = 14;
var FULL_PAGE_EDITOR_TOOLBAR_HEIGHT = exports.FULL_PAGE_EDITOR_TOOLBAR_HEIGHT = function FULL_PAGE_EDITOR_TOOLBAR_HEIGHT() {
  if ((0, _experiments.editorExperiment)('platform_editor_controls', 'variant1', {
    exposure: true
  })) {
    return "var(--ds-space-500, 40px)";
  }
  if ((0, _platformFeatureFlags.fg)('live_pages_content_jump_mitigation')) {
    return '2.188rem';
  }
  return "var(--ds-space-500, 40px)";
};
var akEditorSelectedNodeClassName = exports.akEditorSelectedNodeClassName = 'ak-editor-selected-node';
var editorFontSize = exports.editorFontSize = function editorFontSize(_ref) {
  var theme = _ref.theme;
  return theme && theme.baseFontSize ? theme.baseFontSize : DEFAULT_FONT_SIZE;
};
var relativeSize = exports.relativeSize = function relativeSize(multiplier) {
  return function (_ref2) {
    var theme = _ref2.theme;
    return editorFontSize({
      theme: theme
    }) * multiplier;
  };
};
var relativeFontSizeToBase16 = exports.relativeFontSizeToBase16 = function relativeFontSizeToBase16(px) {
  if (typeof px === 'string') {
    px = parseInt(px);
  }
  if (isNaN(px)) {
    throw new Error("Invalid font size: '".concat(px, "'"));
  }
  return "".concat(px / 16, "rem");
};
var VIEWPORT_SIZES = exports.VIEWPORT_SIZES = {
  laptopHiDPI: {
    width: 1440,
    height: 900
  },
  laptopMDPI: {
    width: 1280,
    height: 800
  },
  tabletL: {
    width: 1024,
    height: 1366
  },
  tabletS: {
    width: 768,
    height: 1024
  },
  mobileM: {
    width: 414,
    height: 736
  },
  mobileS: {
    width: 375,
    height: 667
  },
  mobileXS: {
    width: 320,
    height: 568
  }
};

// to be updated in ED-10790: this should be variable threshold based on how many buttons enabled on main toolbar
var akEditorMobileMaxWidth = exports.akEditorMobileMaxWidth = 0;
var getTableCellBackgroundDarkModeColors = exports.getTableCellBackgroundDarkModeColors = [['White', '#000000'], ['Light blue', '#0C294F'], ['Light teal', '#0C343B'], ['Light green', '#052E21'], ['Light yellow', '#484123'], ['Light red', '#441C13'], ['Light purple', '#282249'], ['Light gray', '#202328'], ['Blue', '#0B3165'], ['Teal', '#044853'], ['Green', '#053927'], ['Yellow', '#6F5C25'], ['Red', '#582013'], ['Purple', '#3E327B'], ['Gray', '#475262'], ['Dark blue', '#003884'], ['Dark teal', '#055866'], ['Dark green', '#044932'], ['Dark yellow', '#82641C'], ['Dark red', '#6B2A19'], ['Dark purple', '#4D38B2']].map(function (_ref3) {
  var _ref4 = (0, _slicedToArray2.default)(_ref3, 2),
    colorName = _ref4[0],
    colorValue = _ref4[1];
  return getTableCellBackgroundDarkModeColorCSS(colorName, colorValue);
}).join('\n');
function getTableCellBackgroundDarkModeColorCSS(colorName, colorValue) {
  return "\n  &[colorname=\"".concat(colorName, "\"] {\n    background-color: ").concat(colorValue, " !important; // !important to override default style color\n  }\n  ");
}
var avatarColors = exports.avatarColors = ["var(--ds-icon-accent-red, #C9372C)", "var(--ds-background-accent-red-bolder-hovered, #AE2E24)", "var(--ds-icon-accent-magenta, #CD519D)", "var(--ds-background-accent-magenta-bolder-hovered, #943D73)", "var(--ds-icon-accent-orange, #E56910)", "var(--ds-background-accent-orange-bolder-hovered, #A54800)", "var(--ds-icon-accent-green, #22A06B)", "var(--ds-background-accent-green-bolder-hovered, #216E4E)", "var(--ds-icon-accent-blue, #1D7AFC)", "var(--ds-background-accent-blue-bolder-hovered, #0055CC)", "var(--ds-icon-accent-teal, #2898BD)", "var(--ds-background-accent-teal-bolder-hovered, #206A83)", "var(--ds-icon-accent-lime, #6A9A23)", "var(--ds-background-accent-lime-bolder-hovered, #4C6B1F)", "var(--ds-icon-accent-purple, #8270DB)", "var(--ds-background-accent-purple-bolder-hovered, #5E4DB2)", "var(--ds-icon-accent-gray, #758195)", "var(--ds-background-accent-gray-bolder-hovered, #44546F)", "var(--ds-icon-accent-yellow, #B38600)", "var(--ds-background-accent-yellow-bolder-hovered, #7F5F01)", "var(--ds-background-accent-yellow-bolder-pressed, #533F04)"];

// MDS-569: Will update references to avatarColors to use participantColors instead, we will then deprecate avatarColors
var participantColors = exports.participantColors = [{
  textColor: "var(--ds-text-inverse, #FFFFFF)",
  backgroundColor: "var(--ds-background-accent-red-bolder, #C9372C)"
}, {
  textColor: "var(--ds-text-inverse, #FFFFFF)",
  backgroundColor: "var(--ds-background-accent-blue-bolder, #0C66E4)"
}, {
  textColor: "var(--ds-text-inverse, #FFFFFF)",
  backgroundColor: "var(--ds-background-accent-green-bolder, #1F845A)"
}, {
  textColor: "var(--ds-text-inverse, #FFFFFF)",
  backgroundColor: "var(--ds-background-accent-yellow-bolder, #946F00)"
}, {
  textColor: "var(--ds-text-inverse, #FFFFFF)",
  backgroundColor: "var(--ds-background-accent-purple-bolder, #6E5DC6)"
}, {
  textColor: "var(--ds-text-inverse, #FFFFFF)",
  backgroundColor: "var(--ds-background-accent-magenta-bolder, #AE4787)"
}, {
  textColor: "var(--ds-text-inverse, #FFFFFF)",
  backgroundColor: "var(--ds-background-accent-teal-bolder, #227D9B)"
}, {
  textColor: "var(--ds-text-inverse, #FFFFFF)",
  backgroundColor: "var(--ds-background-accent-orange-bolder, #C25100)"
}, {
  textColor: "var(--ds-text-inverse, #FFFFFF)",
  backgroundColor: "var(--ds-background-accent-lime-bolder, #5B7F24)"
}, {
  textColor: "var(--ds-text-inverse, #FFFFFF)",
  backgroundColor: "var(--ds-background-accent-gray-bolder, #626F86)"
}, {
  textColor: "var(--ds-text-accent-gray-bolder, #091E42)",
  backgroundColor: "var(--ds-background-accent-blue-subtle, #579DFF)"
}, {
  textColor: "var(--ds-text-accent-gray-bolder, #091E42)",
  backgroundColor: "var(--ds-background-accent-red-subtle, #F87168)"
}, {
  textColor: "var(--ds-text-accent-gray-bolder, #091E42)",
  backgroundColor: "var(--ds-background-accent-orange-subtle, #FEA362)"
}, {
  textColor: "var(--ds-text-accent-gray-bolder, #091E42)",
  backgroundColor: "var(--ds-background-accent-yellow-subtle, #F5CD47)"
}, {
  textColor: "var(--ds-text-accent-gray-bolder, #091E42)",
  backgroundColor: "var(--ds-background-accent-green-subtle, #4BCE97)"
}, {
  textColor: "var(--ds-text-accent-gray-bolder, #091E42)",
  backgroundColor: "var(--ds-background-accent-teal-subtle, #6CC3E0)"
}, {
  textColor: "var(--ds-text-accent-gray-bolder, #091E42)",
  backgroundColor: "var(--ds-background-accent-purple-subtle, #9F8FEF)"
}, {
  textColor: "var(--ds-text-accent-gray-bolder, #091E42)",
  backgroundColor: "var(--ds-background-accent-magenta-subtle, #E774BB)"
}];
