// @atlaskit/editor-shared-styles/dist/cjs/consts/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ATLASSIAN_NAVIGATION_HEIGHT", {
  enumerable: true,
  get: function get() {
    return _consts.ATLASSIAN_NAVIGATION_HEIGHT;
  }
});
Object.defineProperty(exports, "DEFAULT_EMBED_CARD_HEIGHT", {
  enumerable: true,
  get: function get() {
    return _consts.DEFAULT_EMBED_CARD_HEIGHT;
  }
});
Object.defineProperty(exports, "DEFAULT_EMBED_CARD_WIDTH", {
  enumerable: true,
  get: function get() {
    return _consts.DEFAULT_EMBED_CARD_WIDTH;
  }
});
Object.defineProperty(exports, "FULL_PAGE_EDITOR_TOOLBAR_HEIGHT", {
  enumerable: true,
  get: function get() {
    return _consts.FULL_PAGE_EDITOR_TOOLBAR_HEIGHT;
  }
});
Object.defineProperty(exports, "MAX_BROWSER_SCROLLBAR_HEIGHT", {
  enumerable: true,
  get: function get() {
    return _consts.MAX_BROWSER_SCROLLBAR_HEIGHT;
  }
});
Object.defineProperty(exports, "VIEWPORT_SIZES", {
  enumerable: true,
  get: function get() {
    return _consts.VIEWPORT_SIZES;
  }
});
Object.defineProperty(exports, "akEditorBlockquoteBorderColor", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorBlockquoteBorderColor;
  }
});
Object.defineProperty(exports, "akEditorBreakoutPadding", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorBreakoutPadding;
  }
});
Object.defineProperty(exports, "akEditorCalculatedWideLayoutWidth", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorCalculatedWideLayoutWidth;
  }
});
Object.defineProperty(exports, "akEditorCalculatedWideLayoutWidthSmallViewport", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorCalculatedWideLayoutWidthSmallViewport;
  }
});
Object.defineProperty(exports, "akEditorCodeBackground", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorCodeBackground;
  }
});
Object.defineProperty(exports, "akEditorCodeBlockPadding", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorCodeBlockPadding;
  }
});
Object.defineProperty(exports, "akEditorCodeFontFamily", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorCodeFontFamily;
  }
});
Object.defineProperty(exports, "akEditorCodeInlinePadding", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorCodeInlinePadding;
  }
});
Object.defineProperty(exports, "akEditorContextPanelWidth", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorContextPanelWidth;
  }
});
Object.defineProperty(exports, "akEditorCustomIconSize", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorCustomIconSize;
  }
});
Object.defineProperty(exports, "akEditorDefaultLayoutWidth", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorDefaultLayoutWidth;
  }
});
Object.defineProperty(exports, "akEditorDeleteBackground", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorDeleteBackground;
  }
});
Object.defineProperty(exports, "akEditorDeleteBackgroundWithOpacity", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorDeleteBackgroundWithOpacity;
  }
});
Object.defineProperty(exports, "akEditorDeleteBorder", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorDeleteBorder;
  }
});
Object.defineProperty(exports, "akEditorDeleteIconColor", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorDeleteIconColor;
  }
});
Object.defineProperty(exports, "akEditorFloatingDialogZIndex", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorFloatingDialogZIndex;
  }
});
Object.defineProperty(exports, "akEditorFloatingOverlapPanelZIndex", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorFloatingOverlapPanelZIndex;
  }
});
Object.defineProperty(exports, "akEditorFloatingPanelZIndex", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorFloatingPanelZIndex;
  }
});
Object.defineProperty(exports, "akEditorFullPageDefaultFontSize", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorFullPageDefaultFontSize;
  }
});
Object.defineProperty(exports, "akEditorFullPageMaxWidth", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorFullPageMaxWidth;
  }
});
Object.defineProperty(exports, "akEditorFullWidthLayoutLineLength", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorFullWidthLayoutLineLength;
  }
});
Object.defineProperty(exports, "akEditorFullWidthLayoutWidth", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorFullWidthLayoutWidth;
  }
});
Object.defineProperty(exports, "akEditorGridLineZIndex", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorGridLineZIndex;
  }
});
Object.defineProperty(exports, "akEditorGutterPadding", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorGutterPadding;
  }
});
Object.defineProperty(exports, "akEditorGutterPaddingDynamic", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorGutterPaddingDynamic;
  }
});
Object.defineProperty(exports, "akEditorLineHeight", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorLineHeight;
  }
});
Object.defineProperty(exports, "akEditorMediaResizeHandlerPadding", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorMediaResizeHandlerPadding;
  }
});
Object.defineProperty(exports, "akEditorMediaResizeHandlerPaddingWide", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorMediaResizeHandlerPaddingWide;
  }
});
Object.defineProperty(exports, "akEditorMenuZIndex", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorMenuZIndex;
  }
});
Object.defineProperty(exports, "akEditorMobileBreakoutPoint", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorMobileBreakoutPoint;
  }
});
Object.defineProperty(exports, "akEditorMobileMaxWidth", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorMobileMaxWidth;
  }
});
Object.defineProperty(exports, "akEditorRuleBorderRadius", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorRuleBorderRadius;
  }
});
Object.defineProperty(exports, "akEditorSelectedBlanketOpacity", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorSelectedBlanketOpacity;
  }
});
Object.defineProperty(exports, "akEditorSelectedBoldBoxShadow", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorSelectedBoldBoxShadow;
  }
});
Object.defineProperty(exports, "akEditorSelectedBorder", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorSelectedBorder;
  }
});
Object.defineProperty(exports, "akEditorSelectedBorderBoldSize", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorSelectedBorderBoldSize;
  }
});
Object.defineProperty(exports, "akEditorSelectedBorderColor", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorSelectedBorderColor;
  }
});
Object.defineProperty(exports, "akEditorSelectedBorderSize", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorSelectedBorderSize;
  }
});
Object.defineProperty(exports, "akEditorSelectedBoxShadow", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorSelectedBoxShadow;
  }
});
Object.defineProperty(exports, "akEditorSelectedNodeClassName", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorSelectedNodeClassName;
  }
});
Object.defineProperty(exports, "akEditorShadowZIndex", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorShadowZIndex;
  }
});
Object.defineProperty(exports, "akEditorSmallZIndex", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorSmallZIndex;
  }
});
Object.defineProperty(exports, "akEditorStickyHeaderZIndex", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorStickyHeaderZIndex;
  }
});
Object.defineProperty(exports, "akEditorSubtleAccent", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorSubtleAccent;
  }
});
Object.defineProperty(exports, "akEditorSwoopCubicBezier", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorSwoopCubicBezier;
  }
});
Object.defineProperty(exports, "akEditorTableBorder", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorTableBorder;
  }
});
Object.defineProperty(exports, "akEditorTableBorderRadius", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorTableBorderRadius;
  }
});
Object.defineProperty(exports, "akEditorTableBorderSelected", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorTableBorderSelected;
  }
});
Object.defineProperty(exports, "akEditorTableCellBackgroundOpacity", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorTableCellBackgroundOpacity;
  }
});
Object.defineProperty(exports, "akEditorTableCellBlanketDeleted", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorTableCellBlanketDeleted;
  }
});
Object.defineProperty(exports, "akEditorTableCellBlanketSelected", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorTableCellBlanketSelected;
  }
});
Object.defineProperty(exports, "akEditorTableCellMinWidth", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorTableCellMinWidth;
  }
});
Object.defineProperty(exports, "akEditorTableCellOnStickyHeaderZIndex", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorTableCellOnStickyHeaderZIndex;
  }
});
Object.defineProperty(exports, "akEditorTableHeaderCellBackground", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorTableHeaderCellBackground;
  }
});
Object.defineProperty(exports, "akEditorTableHeaderCellBackgroundDark", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorTableHeaderCellBackgroundDark;
  }
});
Object.defineProperty(exports, "akEditorTableLegacyCellMinWidth", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorTableLegacyCellMinWidth;
  }
});
Object.defineProperty(exports, "akEditorTableNumberColumnWidth", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorTableNumberColumnWidth;
  }
});
Object.defineProperty(exports, "akEditorTableToolbar", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorTableToolbar;
  }
});
Object.defineProperty(exports, "akEditorTableToolbarSize", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorTableToolbarSize;
  }
});
Object.defineProperty(exports, "akEditorToolbarKeylineHeight", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorToolbarKeylineHeight;
  }
});
Object.defineProperty(exports, "akEditorUnitZIndex", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorUnitZIndex;
  }
});
Object.defineProperty(exports, "akEditorWideLayoutWidth", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorWideLayoutWidth;
  }
});
Object.defineProperty(exports, "akEditorWrappedNodeZIndex", {
  enumerable: true,
  get: function get() {
    return _consts.akEditorWrappedNodeZIndex;
  }
});
Object.defineProperty(exports, "akLayoutGutterOffset", {
  enumerable: true,
  get: function get() {
    return _consts.akLayoutGutterOffset;
  }
});
Object.defineProperty(exports, "akRichMediaResizeZIndex", {
  enumerable: true,
  get: function get() {
    return _consts.akRichMediaResizeZIndex;
  }
});
Object.defineProperty(exports, "avatarColors", {
  enumerable: true,
  get: function get() {
    return _consts.avatarColors;
  }
});
Object.defineProperty(exports, "blockNodesVerticalMargin", {
  enumerable: true,
  get: function get() {
    return _consts.blockNodesVerticalMargin;
  }
});
Object.defineProperty(exports, "breakoutWideScaleRatio", {
  enumerable: true,
  get: function get() {
    return _consts.breakoutWideScaleRatio;
  }
});
Object.defineProperty(exports, "editorFontSize", {
  enumerable: true,
  get: function get() {
    return _consts.editorFontSize;
  }
});
Object.defineProperty(exports, "getTableCellBackgroundDarkModeColors", {
  enumerable: true,
  get: function get() {
    return _consts.getTableCellBackgroundDarkModeColors;
  }
});
Object.defineProperty(exports, "gridMediumMaxWidth", {
  enumerable: true,
  get: function get() {
    return _consts.gridMediumMaxWidth;
  }
});
Object.defineProperty(exports, "layoutBreakpointWidth", {
  enumerable: true,
  get: function get() {
    return _consts.layoutBreakpointWidth;
  }
});
Object.defineProperty(exports, "participantColors", {
  enumerable: true,
  get: function get() {
    return _consts.participantColors;
  }
});
Object.defineProperty(exports, "relativeFontSizeToBase16", {
  enumerable: true,
  get: function get() {
    return _consts.relativeFontSizeToBase16;
  }
});
Object.defineProperty(exports, "relativeSize", {
  enumerable: true,
  get: function get() {
    return _consts.relativeSize;
  }
});
var _consts = require("./consts");
