// @atlaskit/media-card/dist/cjs/card/ui/titleBox/titleBoxComponents-emotion.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TitleBoxWrapper = exports.TitleBoxIcon = exports.TitleBoxHeader = exports.TitleBoxFooter = exports.ErrorMessageWrapper = void 0;
var _react = require("@emotion/react");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _styles = require("./styles");
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var TitleBoxWrapper = exports.TitleBoxWrapper = function TitleBoxWrapper(props) {
  var breakpoint = props.breakpoint,
    titleBoxBgColor = props.titleBoxBgColor,
    hidden = props.hidden;
  return (0, _react.jsx)("div", {
    id: "titleBoxWrapper",
    "data-testid": "media-title-box"
    // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
    ,
    css: (0, _styles.titleBoxWrapperStyles)({
      breakpoint: breakpoint,
      display: hidden ? 'none' : 'flex',
      titleBoxBgColor: titleBoxBgColor
    })
  }, props.children);
};
var TitleBoxHeader = exports.TitleBoxHeader = function TitleBoxHeader(props) {
  var hasIconOverlap = props.hasIconOverlap;
  return (0, _react.jsx)("div", {
    id: "titleBoxHeader",
    "data-testid": "title-box-header"
    // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
    ,
    css: (0, _styles.titleBoxHeaderStyles)({
      hasIconOverlap: hasIconOverlap
    })
  }, props.children);
};
var TitleBoxFooter = exports.TitleBoxFooter = function TitleBoxFooter(props) {
  var hasIconOverlap = props.hasIconOverlap;
  return (0, _react.jsx)("div", {
    id: "titleBoxFooter",
    "data-testid": "title-box-footer"
    // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
    ,
    css: (0, _styles.titleBoxFooterStyles)({
      hasIconOverlap: hasIconOverlap
    })
  }, props.children);
};
var TitleBoxIcon = exports.TitleBoxIcon = function TitleBoxIcon(props) {
  return (0, _react.jsx)("div", {
    id: "titleBoxIcon",
    "data-testid": "title-box-icon",
    css: (0, _platformFeatureFlags.fg)('platform-visual-refresh-icons') ?
    // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
    _styles.newTitleBoxIconStyles :
    // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
    _styles.titleBoxIconStyles
  }, props.children);
};
var ErrorMessageWrapper = exports.ErrorMessageWrapper = function ErrorMessageWrapper(props) {
  // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
  return (0, _react.jsx)("div", {
    css: _styles.errorMessageWrapperStyles
  }, props.children);
};
