// @atlaskit/editor-plugin-card/dist/es2019/nodeviews/toDOM-fixes/blockCard.js
import { blockCard } from '@atlaskit/adf-schema';
import { convertToInlineCss } from '@atlaskit/editor-common/lazy-node-view';
import { B400 } from '@atlaskit/theme/colors';
// @nodeSpecException:toDOM patch
export const blockCardSpecWithFixedToDOM = () => {
  return {
    ...blockCard,
    toDOM: node => {
      var _node$attrs;
      const {
        url
      } = node.attrs;
      const {
        data
      } = node.attrs;
      const {
        layout,
        width,
        datasource
      } = node.attrs;
      const attrs = {
        'data-block-card': '',
        'data-card-url': url || '',
        'data-card-data': data ? JSON.stringify(data) : '',
        'data-datasource': datasource ? JSON.stringify(datasource) : '',
        'data-layout': layout,
        'data-width': `${width}`,
        class: 'blockCardView-content-wrap'
      };
      return ['div', attrs, ['a', {
        // To match `packages/linking-platform/smart-card/src/view/CardWithUrl/component-lazy/LoadingCardLink.tsx`
        // Which uses frame styling `packages/linking-platform/smart-card/src/view/InlineCard/Frame/styled.ts`
        style: convertToInlineCss({
          padding: `${"var(--ds-space-025, 2px)"} 0px`,
          marginLeft: "var(--ds-space-negative-025, -2px)",
          display: 'inline',
          boxDecorationBreak: 'clone',
          WebkitBoxDecorationBreak: 'clone',
          borderRadius: "var(--ds-border-radius-100, 4px)",
          color: `var(--ds-link, ${B400})`,
          lineHeight: '22px',
          WebkitTransition: '0.1s all ease-in-out',
          transition: '0.1s all ease-in-out',
          userSelect: 'text',
          WebkitUserSelect: 'text',
          msUserSelect: 'text',
          MozUserSelect: 'none' // -moz-user-select
        }),
        href: url || ''
      }, (node === null || node === void 0 ? void 0 : (_node$attrs = node.attrs) === null || _node$attrs === void 0 ? void 0 : _node$attrs.url) || ' ']];
    }
  };
};
