// @atlaskit/smart-card/dist/es2019/utils/analytics/analytics.js
export const ANALYTICS_CHANNEL = 'media';
export const context = {
  componentName: 'smart-cards',
  packageName: "@atlaskit/smart-card",
  packageVersion: "36.9.0"
};
export let TrackQuickActionType = /*#__PURE__*/function (TrackQuickActionType) {
  TrackQuickActionType["StatusUpdate"] = "StatusUpdate";
  return TrackQuickActionType;
}({});
export let TrackQuickActionFailureReason = /*#__PURE__*/function (TrackQuickActionFailureReason) {
  TrackQuickActionFailureReason["PermissionError"] = "PermissionError";
  TrackQuickActionFailureReason["ValidationError"] = "ValidationError";
  TrackQuickActionFailureReason["UnknownError"] = "UnknownError";
  return TrackQuickActionFailureReason;
}({});
export class SmartLinkEvents {
  insertSmartLink(url, type, createAnalyticsEvent) {
    fireSmartLinkEvent({
      action: 'inserted',
      actionSubject: 'smartLink',
      eventType: 'track',
      attributes: {
        type
      },
      nonPrivacySafeAttributes: {
        domainName: url
      }
    }, createAnalyticsEvent);
  }
}
export const fireSmartLinkEvent = (payload, createAnalyticsEvent) => {
  if (createAnalyticsEvent) {
    createAnalyticsEvent(payload).fire(ANALYTICS_CHANNEL);
  }
};
