// @atlaskit/link-datasource/dist/cjs/ui/common/modal/popup-select/menu-list/errorMessage.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _reactIntlNext = require("react-intl-next");
var _useDebounce = require("use-debounce");
var _error = _interopRequireDefault(require("@atlaskit/icon/glyph/error"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _colors = require("@atlaskit/theme/colors");
var _analytics = require("../../../../../analytics");
var _error2 = require("../../../../../common/ui/spot/error-state/error");
var _errorOld = require("../../../../../common/ui/spot/error-state/error-old");
var _constants = require("../constants");
var _messages = require("./messages");
var _selectMessage = _interopRequireDefault(require("./selectMessage"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var getErrorReasonType = function getErrorReasonType(errors) {
  var _ref = errors || [],
    _ref2 = (0, _slicedToArray2.default)(_ref, 1),
    error = _ref2[0];
  if (error instanceof Error) {
    return 'network';
  }
  if (errors && errors.length > 0) {
    return 'agg';
  }
  return 'unknown';
};
var noop = function noop() {
  return '';
};
var CustomErrorMessage = function CustomErrorMessage(_ref3) {
  var filterName = _ref3.filterName,
    errors = _ref3.errors;
  var _useDatasourceAnalyti = (0, _analytics.useDatasourceAnalyticsEvents)(),
    fireEvent = _useDatasourceAnalyti.fireEvent;

  /**
   * Debounce is required because our search is debounced
   * ref: ./noOptionsMessage.tsx
   */
  var _useDebouncedCallback = (0, _useDebounce.useDebouncedCallback)(function () {
      fireEvent('ui.error.shown.basicSearchDropdown', {
        filterName: filterName,
        reason: getErrorReasonType(errors)
      });
    }, _constants.SEARCH_DEBOUNCE_MS),
    _useDebouncedCallback2 = (0, _slicedToArray2.default)(_useDebouncedCallback, 1),
    debouncedAnalyticsCallback = _useDebouncedCallback2[0];
  (0, _react.useEffect)(debouncedAnalyticsCallback, [debouncedAnalyticsCallback]);
  var _ref4 = (0, _platformFeatureFlags.fg)('bandicoots-update-sllv-icons') ?
    // eslint-disable-next-line react-hooks/rules-of-hooks
    (0, _reactIntlNext.useIntl)() : {
      formatMessage: noop
    },
    formatMessage = _ref4.formatMessage;
  return /*#__PURE__*/_react.default.createElement(_selectMessage.default, {
    icon: (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv') ? /*#__PURE__*/_react.default.createElement(_error2.SpotError, {
      size: "large",
      alt: formatMessage(_messages.asyncPopupSelectMessages.errorMessage)
    }) : (0, _platformFeatureFlags.fg)('bandicoots-update-sllv-icons') ? /*#__PURE__*/_react.default.createElement(_errorOld.SpotErrorOld, {
      size: 'medium',
      alt: formatMessage(_messages.asyncPopupSelectMessages.errorMessageOld)
    }) : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_error.default, {
      primaryColor: "var(--ds-icon, ".concat(_colors.N500, ")"),
      label: "",
      size: "xlarge"
    })),
    message: (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv') ? _messages.asyncPopupSelectMessages.errorMessage : _messages.asyncPopupSelectMessages.errorMessageOld,
    description: (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv') ? _messages.asyncPopupSelectMessages.errorDescription : undefined,
    testId: "".concat(filterName, "--error-message")
  });
};
var _default = exports.default = CustomErrorMessage;
