// @atlaskit/link-datasource/dist/cjs/ui/common/modal/popup-select/menu-list/selectMessage.js
/* selectMessage.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./selectMessage.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _react = _interopRequireDefault(require("react"));
var _reactIntlNext = require("react-intl-next");
var _heading = _interopRequireDefault(require("@atlaskit/heading"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _compiled = require("@atlaskit/primitives/compiled");
var styles = {
  boxStylesOld: "_4t3ivbgk _otyrpxbi",
  boxStyles: "_4t3i16xz",
  stackStylesOld: "_ca0qu2gc _n3tdu2gc _19bvutpp _u5f3utpp",
  stackStyles: "_zulppxbi _ca0q1jfw _n3td1jfw _19bvoahv _u5f3oahv _y3gn1h6o"
};
var CustomSelectMessage = function CustomSelectMessage(_ref) {
  var icon = _ref.icon,
    message = _ref.message,
    description = _ref.description,
    testId = _ref.testId;
  if ((0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv')) {
    return /*#__PURE__*/_react.default.createElement(_compiled.Stack, {
      xcss: styles.stackStyles,
      testId: testId,
      alignInline: "center"
    }, /*#__PURE__*/_react.default.createElement(_compiled.Flex, {
      xcss: styles.boxStyles,
      alignItems: "center",
      justifyContent: "center"
    }, icon), /*#__PURE__*/_react.default.createElement(_heading.default, {
      size: "small"
    }, /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, message)), description && /*#__PURE__*/_react.default.createElement(_compiled.Text, {
      size: "medium"
    }, /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, description)));
  }
  return /*#__PURE__*/_react.default.createElement(_compiled.Stack, {
    xcss: styles.stackStylesOld,
    testId: testId,
    alignInline: "center"
  }, /*#__PURE__*/_react.default.createElement(_compiled.Flex, {
    xcss: styles.boxStylesOld,
    alignItems: "center",
    justifyContent: "center"
  }, icon), /*#__PURE__*/_react.default.createElement(_heading.default, {
    size: "xsmall"
  }, /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, message)));
};
var _default = exports.default = CustomSelectMessage;
