// @atlaskit/link-datasource/dist/cjs/ui/common/modal/popup-select/menu-list/showMoreButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactIntlNext = require("react-intl-next");
var _button = _interopRequireDefault(require("@atlaskit/button"));
var _new = _interopRequireDefault(require("@atlaskit/button/new"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _analytics = require("../../../../../analytics");
var _messages = require("./messages");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var ShowMoreButton = function ShowMoreButton(_ref) {
  var onShowMore = _ref.onShowMore,
    filterName = _ref.filterName;
  var _useIntl = (0, _reactIntlNext.useIntl)(),
    formatMessage = _useIntl.formatMessage;
  var _useDatasourceAnalyti = (0, _analytics.useDatasourceAnalyticsEvents)(),
    fireEvent = _useDatasourceAnalyti.fireEvent;
  var handleShowMore = (0, _react.useCallback)(function () {
    fireEvent('ui.button.clicked.basicSearchDropdown', {
      filterName: filterName,
      type: 'showMore'
    });
    onShowMore();
  }, [filterName, fireEvent, onShowMore]);
  if ((0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv')) {
    return /*#__PURE__*/_react.default.createElement(_new.default, {
      onClick: handleShowMore,
      appearance: "subtle",
      testId: "".concat(filterName, "--show-more-button"),
      spacing: "compact"
    }, formatMessage(_messages.asyncPopupSelectMessages.showMoreMessage));
  }
  return /*#__PURE__*/_react.default.createElement(_button.default, {
    onClick: handleShowMore,
    appearance: "link",
    testId: "".concat(filterName, "--show-more-button")
  }, formatMessage(_messages.asyncPopupSelectMessages.showMoreMessage));
};
var _default = exports.default = ShowMoreButton;
