// @atlaskit/media-card/dist/cjs/card/ui/tickBox/tickBoxWrapper-emotion.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TickBoxWrapper = void 0;
var _react = require("@emotion/react");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _styles = require("./styles");
var _styles2 = require("../styles");
var _colors = require("@atlaskit/theme/colors");
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var getSelectedStyles = function getSelectedStyles(selected) {
  return selected ? "background-color: ".concat("var(--ds-icon-information, ".concat(_colors.B200, ")"), ";\n      color: ", "var(--ds-icon-inverse, white)", ";") : "";
};
var oldWrapperStyes = (0, _react.css)({
  width: '14px',
  height: '14px',
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  span: {
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
    svg: {
      height: '14px'
    }
  }
});
var wrapperStyles = function wrapperStyles(selected) {
  return (0, _react.css)(
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  _styles2.transition && (0, _styles2.transition)(), {
    font: "var(--ds-font-body, normal 400 14px/20px ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
    width: "var(--ds-space-200, 16px)",
    height: "var(--ds-space-200, 16px)",
    position: 'absolute',
    top: "var(--ds-space-075, 7px)",
    left: "var(--ds-space-075, 7px)",
    borderRadius: "var(--ds-space-250, 20px)",
    color: 'transparent',
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
    span: {
      display: 'block'
    },
    display: 'grid',
    placeItems: 'center'
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  getSelectedStyles(selected));
};
wrapperStyles.displayName = 'TickBoxWrapper';
var TickBoxWrapper = exports.TickBoxWrapper = function TickBoxWrapper(props) {
  return (0, _react.jsx)("div", {
    id: "tickBoxWrapper",
    css: [
    // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage -- Ignored via go/DSP-18766
    wrapperStyles(props.selected), !(0, _platformFeatureFlags.fg)('platform-visual-refresh-icons') && oldWrapperStyes]
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    ,
    className: _styles.tickBoxClassName
  }, props.children);
};
