// @atlaskit/editor-core/dist/cjs/ui/WithFlash/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireDefault(require("react"));
var _react2 = require("@emotion/react");
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); } /**
 * @jsxRuntime classic
 * @jsx jsx
 */ // eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
var pulseBackground = (0, _react2.keyframes)({
  '50%': {
    backgroundColor: "var(--ds-blanket-danger, #EF5C4814)"
  }
});
var pulseBackgroundReverse = (0, _react2.keyframes)({
  '0%': {
    backgroundColor: "var(--ds-blanket-danger, #EF5C4814)"
  },
  '50%': {
    backgroundColor: 'auto'
  },
  '100%': {
    backgroundColor: "var(--ds-blanket-danger, #EF5C4814)"
  }
});
var flashWrapper = (0, _react2.css)({
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '&.-flash > div': {
    animation: "0.25s ease-in-out ".concat(pulseBackgroundReverse)
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '& > div': {
    animation: "'none'"
  }
});
var flashWrapperAnimated = (0, _react2.css)((0, _defineProperty2.default)({}, "".concat(flashWrapper, " & > div"), {
  animation: "0.25s ease-in-out ".concat(pulseBackground)
}));
// Ignored via go/ees005
// eslint-disable-next-line @repo/internal/react/no-class-components
var WithFlash = exports.default = /*#__PURE__*/function (_React$Component) {
  function WithFlash() {
    var _this;
    (0, _classCallCheck2.default)(this, WithFlash);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, WithFlash, [].concat(args));
    (0, _defineProperty2.default)(_this, "toggle", false);
    return _this;
  }
  (0, _inherits2.default)(WithFlash, _React$Component);
  return (0, _createClass2.default)(WithFlash, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
        animate = _this$props.animate,
        children = _this$props.children;
      this.toggle = animate && !this.toggle;
      return (
        // eslint-disable-next-line @atlaskit/design-system/prefer-primitives
        (0, _react2.jsx)("div", {
          css: animate ? flashWrapperAnimated : flashWrapper
          // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
          ,
          className: this.toggle ? '-flash' : ''
        }, children)
      );
    }
  }]);
}(_react.default.Component);
