// @atlaskit/media-card/dist/cjs/card/ui/actionsBar/styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.wrapperStyles = exports.fixedActionBarStyles = exports.actionsBarClassName = void 0;
var _react = require("@emotion/react");
var _styles = require("../styles");
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var actionsBarClassName = exports.actionsBarClassName = 'media-card-actions-bar';
var fixedActionBarStyles = exports.fixedActionBarStyles = "opacity: 1;";
var wrapperStyles = exports.wrapperStyles = function wrapperStyles(isFixed) {
  return (
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
    (0, _react.css)(isFixed ? fixedActionBarStyles : 'opacity: 0;', (0, _styles.transition)(), {
      position: 'absolute',
      top: 0,
      right: 0,
      display: 'flex',
      flexFlow: 'row nowrap',
      justifyContent: 'right',
      padding: "var(--ds-space-100, 8px)",
      gap: "var(--ds-space-100, 8px)"
    })
  );
};
wrapperStyles.displayName = 'ActionsBarWrapper';
