// @atlaskit/media-filmstrip/dist/cjs/filmstripView/wrappers-compiled.js
/* wrappers-compiled.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ShadowRight = exports.RightArrow = exports.LeftArrow = exports.FilmStripViewWrapper = exports.FilmStripListWrapper = exports.FilmStripListItem = exports.FilmStripList = exports.ArrowRightWrapper = exports.ArrowLeftWrapper = void 0;
require("./wrappers-compiled.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _react = _interopRequireWildcard(require("react"));
var _arrowLeft = _interopRequireDefault(require("@atlaskit/icon/core/migration/arrow-left"));
var _arrowRight = _interopRequireDefault(require("@atlaskit/icon/core/migration/arrow-right"));
var _63 = require(".");
var _colors = require("@atlaskit/theme/colors");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var filmStripViewStyles = null;
var filmStripListWrapperStyles = null;
var filmStripListStyles = null;
var filmStripListItemStyles = null;
var arrowWrapperStyles = null;
var arrowLeftWrapperStyles = null;
var arrowRightWrapperStyles = null;
var shadowStyles = null;
var shadowLeftStyles = null;
var shadowRightStyles = null;
var ShadowLeft = function ShadowLeft(_ref) {
  var children = _ref.children;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _runtime.ax)(["_kqswstnw _1pbyzo7u _4t3i1osq _154iidpf _1bsbyh40 _bfhk1tuc", "_1ltvidpf"])
  }, children);
};
var ArrowLeftWrapper = exports.ArrowLeftWrapper = function ArrowLeftWrapper(_ref2) {
  var children = _ref2.children,
    onClick = _ref2.onClick;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _runtime.ax)(["_2rko1osq _v564g17y _kqswstnw _154i1ssb _t9ec1i34 _bfhk1rd0 _1e0c1txw _80omtlke _16qsztl5 _syazgpis _1bsb1ul9 _4t3i1ul9 _1bah1h6o _tzy4idpf _17521ul9 _w8l5gktf _30l3lyvy _irr3if84 _9h8hln51 _1di61kxl", "_1ltvu2gc _uhasv77o", "arrow"]),
    onClick: onClick,
    style: {
      "--_14yez9t": (0, _runtime.ix)("var(--ds-surface-overlay, ".concat(_colors.N20, ")")),
      "--_snhgxu": (0, _runtime.ix)("var(--ds-surface-overlay-hovered, ".concat(_colors.N40, ")")),
      "--_1hfkvbo": (0, _runtime.ix)("var(--ds-text-selected, ".concat(_colors.B400, ")")),
      "--_rryx0a": (0, _runtime.ix)("var(--ds-background-selected, ".concat(_colors.B50, ")"))
    }
  }, children);
};
var ShadowRight = exports.ShadowRight = function ShadowRight(_ref3) {
  var children = _ref3.children;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _runtime.ax)(["_kqswstnw _1pbyzo7u _4t3i1osq _154iidpf _1bsbyh40 _bfhk1tuc", "_1xi2idpf"])
  }, children);
};
var ArrowRightWrapper = exports.ArrowRightWrapper = function ArrowRightWrapper(_ref4) {
  var children = _ref4.children,
    onClick = _ref4.onClick;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _runtime.ax)(["_2rko1osq _v564g17y _kqswstnw _154i1ssb _t9ec1i34 _bfhk1rd0 _1e0c1txw _80omtlke _16qsztl5 _syazgpis _1bsb1ul9 _4t3i1ul9 _1bah1h6o _tzy4idpf _17521ul9 _w8l5gktf _30l3lyvy _irr3if84 _9h8hln51 _1di61kxl", "_1xi2u2gc _7wy9v77o", "arrow"]),
    onClick: onClick,
    style: {
      "--_14yez9t": (0, _runtime.ix)("var(--ds-surface-overlay, ".concat(_colors.N20, ")")),
      "--_snhgxu": (0, _runtime.ix)("var(--ds-surface-overlay-hovered, ".concat(_colors.N40, ")")),
      "--_1hfkvbo": (0, _runtime.ix)("var(--ds-text-selected, ".concat(_colors.B400, ")")),
      "--_rryx0a": (0, _runtime.ix)("var(--ds-background-selected, ".concat(_colors.B50, ")"))
    }
  }, children);
};
var LeftArrow = exports.LeftArrow = function LeftArrow(_ref5) {
  var onClick = _ref5.onClick;
  return /*#__PURE__*/_react.default.createElement(ShadowLeft, null, /*#__PURE__*/_react.default.createElement(ArrowLeftWrapper, {
    onClick: onClick
  }, /*#__PURE__*/_react.default.createElement(_arrowLeft.default, {
    color: "currentColor",
    spacing: "spacious",
    label: "left"
  })));
};
var RightArrow = exports.RightArrow = function RightArrow(_ref6) {
  var onClick = _ref6.onClick;
  return /*#__PURE__*/_react.default.createElement(ShadowRight, null, /*#__PURE__*/_react.default.createElement(ArrowRightWrapper, {
    onClick: onClick
  }, /*#__PURE__*/_react.default.createElement(_arrowRight.default, {
    color: "currentColor",
    spacing: "spacious",
    label: "right"
  })));
};
var FilmStripViewWrapper = exports.FilmStripViewWrapper = function FilmStripViewWrapper(_ref7) {
  var children = _ref7.children,
    dataTestId = _ref7['data-testid'];
  return /*#__PURE__*/_react.default.createElement("div", {
    "data-testid": dataTestId,
    className: (0, _runtime.ax)(["_2rko1l7b _kqswh2mm _ca0q1ysw _u5f3idpf _n3td1ysw _19bvidpf _1yqxkb7n _1ajs4jg8"])
  }, children);
};
var FilmStripListWrapper = exports.FilmStripListWrapper = /*#__PURE__*/(0, _react.forwardRef)(function (_ref8, ref) {
  var children = _ref8.children,
    onWheel = _ref8.onWheel,
    onTouchStart = _ref8.onTouchStart,
    onTouchMove = _ref8.onTouchMove,
    onTouchEnd = _ref8.onTouchEnd,
    dataTestId = _ref8['data-testid'];
  return /*#__PURE__*/_react.default.createElement("div", {
    ref: ref,
    onWheel: onWheel,
    onTouchStart: onTouchStart,
    onTouchMove: onTouchMove,
    onTouchEnd: onTouchEnd,
    "data-testid": dataTestId,
    className: (0, _runtime.ax)(["_1reo15vq _18m915vq _1bsb1kw7 _ca0qv77o _u5f31ysw _n3tdv77o _19bv1ysw"])
  }, children);
});
var FilmStripList = exports.FilmStripList = /*#__PURE__*/_react.default.forwardRef(function (_ref9, ref) {
  var children = _ref9.children,
    style = _ref9.style;
  return /*#__PURE__*/_react.default.createElement("ul", {
    ref: ref
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
    ,
    style: style,
    className: (0, _runtime.ax)(["_19pkidpf _2hwxidpf _otyridpf _18u0idpf _ca0qidpf _u5f3idpf _n3tdidpf _19bvidpf _k8m01e03 _6fl41kli _o5721q9c _1e0c1o8l"])
  }, children);
});
var FilmStripListItem = exports.FilmStripListItem = function FilmStripListItem(_ref10) {
  var children = _ref10.children,
    index = _ref10.index;
  return /*#__PURE__*/_react.default.createElement("li", {
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    className: (0, _runtime.ax)(["_19pkidpf _2hwxidpf _otyridpf _18u0idpf _2mzuglyw _ca0qidpf _u5f31b66 _n3tdidpf _19bv1b66 _1e0c1o8l _s7n4nkob _1wybidpf _1lhcidpf", _63.MediaFilmStripListItemSelector]),
    "data-testid": "media-filmstrip-list-item",
    key: index
  }, children);
};
