// @atlaskit/media-filmstrip/dist/cjs/filmstripView/wrappers.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ShadowRight = exports.RightArrow = exports.LeftArrow = exports.FilmStripViewWrapper = exports.FilmStripListWrapper = exports.FilmStripListItem = exports.FilmStripList = exports.ArrowRightWrapper = exports.ArrowLeftWrapper = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _wrappersEmotion = require("./wrappers-emotion");
var _wrappersCompiled = require("./wrappers-compiled");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var ArrowLeftWrapper = exports.ArrowLeftWrapper = function ArrowLeftWrapper(props) {
  return (0, _platformFeatureFlags.fg)('platform_media_compiled') ? /*#__PURE__*/_react.default.createElement(_wrappersCompiled.ArrowLeftWrapper, props) : /*#__PURE__*/_react.default.createElement(_wrappersEmotion.ArrowLeftWrapper, props);
};
var ShadowRight = exports.ShadowRight = function ShadowRight(props) {
  return (0, _platformFeatureFlags.fg)('platform_media_compiled') ? /*#__PURE__*/_react.default.createElement(_wrappersCompiled.ShadowRight, props) : /*#__PURE__*/_react.default.createElement(_wrappersEmotion.ShadowRight, props);
};
var ArrowRightWrapper = exports.ArrowRightWrapper = function ArrowRightWrapper(props) {
  return (0, _platformFeatureFlags.fg)('platform_media_compiled') ? /*#__PURE__*/_react.default.createElement(_wrappersCompiled.ArrowRightWrapper, props) : /*#__PURE__*/_react.default.createElement(_wrappersEmotion.ArrowRightWrapper, props);
};
var LeftArrow = exports.LeftArrow = function LeftArrow(props) {
  return (0, _platformFeatureFlags.fg)('platform_media_compiled') ? /*#__PURE__*/_react.default.createElement(_wrappersCompiled.LeftArrow, props) : /*#__PURE__*/_react.default.createElement(_wrappersEmotion.LeftArrow, props);
};
var RightArrow = exports.RightArrow = function RightArrow(props) {
  return (0, _platformFeatureFlags.fg)('platform_media_compiled') ? /*#__PURE__*/_react.default.createElement(_wrappersCompiled.RightArrow, props) : /*#__PURE__*/_react.default.createElement(_wrappersEmotion.RightArrow, props);
};
var FilmStripViewWrapper = exports.FilmStripViewWrapper = function FilmStripViewWrapper(props) {
  return (0, _platformFeatureFlags.fg)('platform_media_compiled') ? /*#__PURE__*/_react.default.createElement(_wrappersCompiled.FilmStripViewWrapper, props) : /*#__PURE__*/_react.default.createElement(_wrappersEmotion.FilmStripViewWrapper, props);
};
var FilmStripListWrapper = exports.FilmStripListWrapper = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  return (0, _platformFeatureFlags.fg)('platform_media_compiled') ? /*#__PURE__*/_react.default.createElement(_wrappersCompiled.FilmStripListWrapper, (0, _extends2.default)({}, props, {
    ref: ref
  })) : /*#__PURE__*/_react.default.createElement(_wrappersEmotion.FilmStripListWrapper, (0, _extends2.default)({}, props, {
    ref: ref
  }));
});
var FilmStripList = exports.FilmStripList = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  return (0, _platformFeatureFlags.fg)('platform_media_compiled') ? /*#__PURE__*/_react.default.createElement(_wrappersCompiled.FilmStripList, (0, _extends2.default)({}, props, {
    ref: ref
  })) : /*#__PURE__*/_react.default.createElement(_wrappersEmotion.FilmStripList, (0, _extends2.default)({}, props, {
    ref: ref
  }));
});
var FilmStripListItem = exports.FilmStripListItem = function FilmStripListItem(props) {
  return (0, _platformFeatureFlags.fg)('platform_media_compiled') ? /*#__PURE__*/_react.default.createElement(_wrappersCompiled.FilmStripListItem, props) : /*#__PURE__*/_react.default.createElement(_wrappersEmotion.FilmStripListItem, props);
};
