// @atlaskit/link-datasource/dist/es2019/ui/common/error-state/provider-auth-required.js
/* provider-auth-required.tsx generated by @compiled/babel-plugin v0.36.1 */
import "./provider-auth-required.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
import React, { useEffect } from 'react';
import { useIntl } from 'react-intl-next';
import Button from '@atlaskit/button';
import EmptyState from '@atlaskit/empty-state';
import { AuthError, auth as outboundAuth } from '@atlaskit/outbound-auth-flow-client';
import { Anchor } from '@atlaskit/primitives/compiled';
import { useDatasourceAnalyticsEvents } from '../../../analytics';
import useErrorLogger from '../../../hooks/useErrorLogger';
import { loadingErrorMessages } from './messages';
import { ProviderAuthRequiredSVG } from './provider-auth-required-svg';
const buttonStyles = null;
const learnMoreAboutSmartLinksUrl = 'https://support.atlassian.com/confluence-cloud/docs/insert-links-and-anchors/#Smart-Links-from-Jira-and-other-products';
export const ProviderAuthRequired = ({
  auth = [],
  onAuthSuccess,
  onAuthError,
  extensionKey,
  providerName,
  datasourceId
}) => {
  const {
    formatMessage
  } = useIntl();
  const {
    captureError
  } = useErrorLogger({
    datasourceId
  });
  const {
    fireEvent
  } = useDatasourceAnalyticsEvents();
  const [authInfo] = auth;
  useEffect(() => {
    fireEvent('ui.error.shown', {
      reason: 'access'
    });
  }, [fireEvent]);
  const onAuthRequest = async () => {
    try {
      await outboundAuth(authInfo.url);
      fireEvent('operational.provider.authSuccess', {
        extensionKey,
        experience: 'datasource'
      });
      onAuthSuccess === null || onAuthSuccess === void 0 ? void 0 : onAuthSuccess();
    } catch (error) {
      fireEvent('operational.provider.authFailure', {
        reason: error instanceof AuthError && error.type ? error.type : null,
        extensionKey,
        experience: 'datasource'
      });
      captureError('ProviderOnAuthRequest', error);
      onAuthError === null || onAuthError === void 0 ? void 0 : onAuthError();
    }
  };
  const renderAuthDescription = () => /*#__PURE__*/React.createElement(React.Fragment, null, formatMessage(loadingErrorMessages.authScreenDescriptionText, {
    providerName
  }), ' ', /*#__PURE__*/React.createElement(Anchor, {
    href: learnMoreAboutSmartLinksUrl,
    target: "_blank",
    rel: "noreferrer noopener"
  }, formatMessage(loadingErrorMessages.learnMoreAboutSmartLinks)));
  const renderAuthConnectButton = () => /*#__PURE__*/React.createElement(Button, {
    onClick: onAuthRequest,
    appearance: "primary",
    className: ax(["_19pkpxbi"])
  }, formatMessage(loadingErrorMessages.authConnectButtonText));
  return /*#__PURE__*/React.createElement(EmptyState, {
    testId: "datasource--access-required-with-auth",
    header: formatMessage(loadingErrorMessages.authScreenHeaderText, {
      providerName
    }),
    description: renderAuthDescription(),
    renderImage: ProviderAuthRequiredSVG,
    primaryAction: renderAuthConnectButton()
  });
};
