// @atlaskit/link-datasource/dist/cjs/ui/table-footer/provider-link/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ProviderLink = void 0;
require("./index.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _react = _interopRequireWildcard(require("react"));
var _reactIntlNext = require("react-intl-next");
var _logo = require("@atlaskit/logo");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _compiled = require("@atlaskit/primitives/compiled");
var _assets = require("@atlaskit/temp-nav-app-icons/assets");
var _analytics = require("../../../analytics");
var _assetsModal = require("../../assets-modal");
var _messages = require("../messages");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var styles = {
  anchor: "_4bfu1r31 _1hmsglyw _ajmmnqa1 _syaz131l _9oik1r31 _1bnxglyw _jf4cnqa1 _30l3131l"
};
var ProviderLink = exports.ProviderLink = function ProviderLink(_ref) {
  var datasourceId = _ref.datasourceId;
  var intl = (0, _reactIntlNext.useIntl)();
  var _useDatasourceAnalyti = (0, _analytics.useDatasourceAnalyticsEvents)(),
    fireEvent = _useDatasourceAnalyti.fireEvent;
  var anchor = (0, _react.useMemo)(function () {
    if (datasourceId === _assetsModal.ASSETS_LIST_OF_LINKS_DATASOURCE_ID) {
      return (0, _platformFeatureFlags.fg)('assets_as_an_app_v2') ? {
        content: intl.formatMessage(_messages.footerMessages.poweredByAssets),
        extensionKey: 'jsm-cmdb-gateway',
        icon: /*#__PURE__*/_react.default.createElement(_assets.AssetsIcon, {
          size: "20"
        }),
        interactionName: 'atlas-link',
        url: '/jira/assets'
      } : {
        content: intl.formatMessage(_messages.footerMessages.powerByJSM),
        extensionKey: 'jsm-cmdb-gateway',
        icon: /*#__PURE__*/_react.default.createElement(_logo.JiraServiceManagementIcon, {
          appearance: "neutral",
          size: "small"
        }),
        interactionName: 'atlas-link',
        url: '/jira/servicedesk/assets'
      };
    }
  }, [datasourceId, intl]);
  return anchor ? /*#__PURE__*/_react.default.createElement(_compiled.Anchor, {
    href: anchor.url,
    interactionName: anchor.interactionName,
    onClick: function onClick() {
      return fireEvent('ui.link.clicked.poweredBy', {
        componentHierarchy: 'datasourceTable',
        extensionKey: anchor.extensionKey
      });
    },
    rel: "noopener noreferrer",
    target: "_blank",
    testId: "powered-by-jsm-assets-link",
    xcss: styles.anchor
  }, /*#__PURE__*/_react.default.createElement(_compiled.Inline, {
    alignBlock: "center",
    space: "space.075"
  }, anchor.icon, anchor.content)) : null;
};
