// @atlaskit/smart-card/dist/esm/view/FlexibleCard/components/actions/automation-action/automation-manual-triggers/manual-triggers-container/common/mocks.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
import { UserInputType } from './types';
var mockScope = {
  resources: []
};
export var mockRules = {
  data: [{
    id: 0,
    name: 'Test rule 0 (no inputs)',
    ruleScope: mockScope,
    userInputs: []
  }, {
    id: 1,
    name: 'Test rule 1 (has inputs)',
    ruleScope: mockScope,
    userInputs: [{
      defaultValue: '',
      displayName: 'Test user input',
      inputType: UserInputType.TEXT,
      required: false,
      variableName: 'testUserInput'
    }]
  }, {
    id: 2,
    name: 'Test rule 2 (has multiple inputs)',
    ruleScope: mockScope,
    userInputs: [{
      defaultValue: '',
      displayName: 'Text user input',
      inputType: UserInputType.TEXT,
      required: false,
      variableName: 'textUserInput'
    }, {
      defaultValue: '',
      displayName: 'Number user input',
      inputType: UserInputType.NUMBER,
      required: false,
      variableName: 'numberUserInput'
    }, {
      defaultValue: true,
      displayName: 'Boolean user input',
      inputType: UserInputType.BOOLEAN,
      required: false,
      variableName: 'booleanUserInput'
    }, {
      defaultValue: [''],
      displayName: 'Dropdown user input',
      inputType: UserInputType.DROPDOWN,
      required: false,
      variableName: 'dropdownUserInput'
    }, {
      defaultValue: '',
      displayName: 'Paragraph user input',
      inputType: UserInputType.PARAGRAPH,
      required: false,
      variableName: 'paragraphUserInput'
    }]
  }]
};
export var mockTransformedRules = [_objectSpread(_objectSpread({}, mockRules.data[0]), {}, {
  userInputPrompts: mockRules.data[0].userInputs
}), _objectSpread(_objectSpread({}, mockRules.data[1]), {}, {
  userInputPrompts: mockRules.data[1].userInputs
}), _objectSpread(_objectSpread({}, mockRules.data[2]), {}, {
  userInputPrompts: mockRules.data[2].userInputs
})];
export var mockGetUrl = 'https://test-api/get/rules';
export var mockInvokeUrl = 'https://test-api/invoke';
