// @atlaskit/onboarding/dist/cjs/components/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Modal", {
  enumerable: true,
  get: function get() {
    return _modal.default;
  }
});
Object.defineProperty(exports, "Spotlight", {
  enumerable: true,
  get: function get() {
    return _spotlight.default;
  }
});
Object.defineProperty(exports, "SpotlightCard", {
  enumerable: true,
  get: function get() {
    return _spotlightCard.default;
  }
});
Object.defineProperty(exports, "SpotlightManager", {
  enumerable: true,
  get: function get() {
    return _spotlightManager.default;
  }
});
Object.defineProperty(exports, "SpotlightTarget", {
  enumerable: true,
  get: function get() {
    return _spotlightTarget.default;
  }
});
Object.defineProperty(exports, "SpotlightTransition", {
  enumerable: true,
  get: function get() {
    return _spotlightTransition.default;
  }
});
Object.defineProperty(exports, "modalButtonTheme", {
  enumerable: true,
  get: function get() {
    return _theme.modalButtonTheme;
  }
});
Object.defineProperty(exports, "spotlightButtonTheme", {
  enumerable: true,
  get: function get() {
    return _theme.spotlightButtonTheme;
  }
});
Object.defineProperty(exports, "useSpotlight", {
  enumerable: true,
  get: function get() {
    return _useSpotlight.default;
  }
});
var _modal = _interopRequireDefault(require("./modal"));
var _spotlight = _interopRequireDefault(require("./spotlight"));
var _spotlightCard = _interopRequireDefault(require("./spotlight-card"));
var _spotlightManager = _interopRequireDefault(require("./spotlight-manager"));
var _spotlightTarget = _interopRequireDefault(require("./spotlight-target"));
var _spotlightTransition = _interopRequireDefault(require("./spotlight-transition"));
var _useSpotlight = _interopRequireDefault(require("./use-spotlight"));
var _theme = require("./theme");
