// @atlaskit/media-picker/dist/esm/popup/tools/websocket/upload/remoteUploadActivity.js
import { __assign } from "tslib";
import { EventEmitter2 } from 'eventemitter2';
import { isRemoteUploadStartData, isRemoteUploadProgressData, isRemoteUploadEndData, isRemoteUploadErrorData, isNotifyMetadata, } from '../wsMessageData';
var RemoteUploadActivity = /** @class */ (function () {
    function RemoteUploadActivity(uploadId, serviceName, dispatchEvent) {
        this.uploadId = uploadId;
        this.serviceName = serviceName;
        this.dispatchEvent = dispatchEvent;
        this.eventEmitter = new EventEmitter2();
    }
    RemoteUploadActivity.prototype.processWebSocketData = function (data) {
        if (!this.shouldProcessWsData(data)) {
            return;
        }
        var basePayload = {
            // First try to use alternative response shape
            // Will be removed after backend unifies response schema
            uploadId: (data.data && data.data.uploadId) || data.uploadId,
            serviceName: this.serviceName,
        };
        if (isRemoteUploadStartData(data)) {
            this.dispatchEvent('RemoteUploadStart', __assign({}, basePayload));
            this.notifyActivityStarted();
        }
        else if (isRemoteUploadProgressData(data)) {
            this.dispatchEvent('RemoteUploadProgress', __assign({ bytes: data.currentAmount, fileSize: data.totalAmount }, basePayload));
        }
        else if (isRemoteUploadEndData(data)) {
            this.dispatchEvent('RemoteUploadEnd', __assign({ fileId: data.fileId }, basePayload));
            this.notifyActivityCompleted();
        }
        else if (isRemoteUploadErrorData(data)) {
            this.dispatchEvent('RemoteUploadFail', __assign({ description: (data.data && data.data.reason) || data.reason }, basePayload));
            this.notifyActivityCompleted();
        }
        else if (isNotifyMetadata(data)) {
            this.dispatchEvent('NotifyMetadata', __assign({ metadata: data.metadata }, basePayload));
        }
    };
    RemoteUploadActivity.prototype.connectionLost = function () {
        if (this.uploadId) {
            this.dispatchEvent('RemoteUploadFail', {
                uploadId: this.uploadId,
                description: 'Websocket connection lost',
                serviceName: this.serviceName,
            });
        }
    };
    RemoteUploadActivity.prototype.on = function (event, handler) {
        this.eventEmitter.on(event, handler);
    };
    RemoteUploadActivity.prototype.off = function (event, handler) {
        this.eventEmitter.off(event, handler);
    };
    RemoteUploadActivity.prototype.shouldProcessWsData = function (data) {
        var shouldProcess = !!(data.uploadId &&
            this.uploadId &&
            data.uploadId === this.uploadId);
        // Try to use alternative response shape
        // Will be removed after backend unifies response schema
        var shouldProcessAlt = !!(data.data &&
            data.data.uploadId &&
            this.uploadId &&
            data.data.uploadId === this.uploadId);
        return shouldProcess || shouldProcessAlt;
    };
    RemoteUploadActivity.prototype.notifyActivityStarted = function () {
        this.eventEmitter.emit('Started', this);
    };
    RemoteUploadActivity.prototype.notifyActivityCompleted = function () {
        this.eventEmitter.emit('Completed', this);
    };
    return RemoteUploadActivity;
}());
export { RemoteUploadActivity };
