// @atlaskit/logo/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var AtlassianLogo_1 = require("./AtlassianLogo");
exports.AtlassianIcon = AtlassianLogo_1.AtlassianIcon;
exports.AtlassianLogo = AtlassianLogo_1.AtlassianLogo;
exports.AtlassianWordmark = AtlassianLogo_1.AtlassianWordmark;
var BitbucketLogo_1 = require("./BitbucketLogo");
exports.BitbucketIcon = BitbucketLogo_1.BitbucketIcon;
exports.BitbucketLogo = BitbucketLogo_1.BitbucketLogo;
exports.BitbucketWordmark = BitbucketLogo_1.BitbucketWordmark;
var ConfluenceLogo_1 = require("./ConfluenceLogo");
exports.ConfluenceIcon = ConfluenceLogo_1.ConfluenceIcon;
exports.ConfluenceLogo = ConfluenceLogo_1.ConfluenceLogo;
exports.ConfluenceWordmark = ConfluenceLogo_1.ConfluenceWordmark;
var HipchatLogo_1 = require("./HipchatLogo");
exports.HipchatIcon = HipchatLogo_1.HipchatIcon;
exports.HipchatLogo = HipchatLogo_1.HipchatLogo;
exports.HipchatWordmark = HipchatLogo_1.HipchatWordmark;
var JiraCoreLogo_1 = require("./JiraCoreLogo");
exports.JiraCoreIcon = JiraCoreLogo_1.JiraCoreIcon;
exports.JiraCoreLogo = JiraCoreLogo_1.JiraCoreLogo;
exports.JiraCoreWordmark = JiraCoreLogo_1.JiraCoreWordmark;
var JiraLogo_1 = require("./JiraLogo");
exports.JiraIcon = JiraLogo_1.JiraIcon;
exports.JiraLogo = JiraLogo_1.JiraLogo;
exports.JiraWordmark = JiraLogo_1.JiraWordmark;
var JiraServiceDeskLogo_1 = require("./JiraServiceDeskLogo");
exports.JiraServiceDeskIcon = JiraServiceDeskLogo_1.JiraServiceDeskIcon;
exports.JiraServiceDeskLogo = JiraServiceDeskLogo_1.JiraServiceDeskLogo;
exports.JiraServiceDeskWordmark = JiraServiceDeskLogo_1.JiraServiceDeskWordmark;
var JiraSoftwareLogo_1 = require("./JiraSoftwareLogo");
exports.JiraSoftwareIcon = JiraSoftwareLogo_1.JiraSoftwareIcon;
exports.JiraSoftwareLogo = JiraSoftwareLogo_1.JiraSoftwareLogo;
exports.JiraSoftwareWordmark = JiraSoftwareLogo_1.JiraSoftwareWordmark;
var OpsGenieLogo_1 = require("./OpsGenieLogo");
exports.OpsGenieIcon = OpsGenieLogo_1.OpsGenieIcon;
exports.OpsGenieLogo = OpsGenieLogo_1.OpsGenieLogo;
exports.OpsGenieWordmark = OpsGenieLogo_1.OpsGenieWordmark;
var StatuspageLogo_1 = require("./StatuspageLogo");
exports.StatuspageIcon = StatuspageLogo_1.StatuspageIcon;
exports.StatuspageLogo = StatuspageLogo_1.StatuspageLogo;
exports.StatuspageWordmark = StatuspageLogo_1.StatuspageWordmark;
var StrideLogo_1 = require("./StrideLogo");
exports.StrideIcon = StrideLogo_1.StrideIcon;
exports.StrideLogo = StrideLogo_1.StrideLogo;
exports.StrideWordmark = StrideLogo_1.StrideWordmark;
var TrelloLogo_1 = require("./TrelloLogo");
exports.TrelloIcon = TrelloLogo_1.TrelloIcon;
exports.TrelloLogo = TrelloLogo_1.TrelloLogo;
exports.TrelloWordmark = TrelloLogo_1.TrelloWordmark;
