// @atlaskit/link-datasource/dist/cjs/ui/assets-modal/search-container/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AssetsSearchContainer = void 0;
require("./index.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _form = _interopRequireDefault(require("@atlaskit/form"));
var _analytics = require("../../../analytics");
var _aqlSearchInput = require("./aql-search-input");
var _objectSchemaSelect = require("./object-schema-select");
var _styled = require("./styled");
var _excluded = ["as", "style"],
  _excluded2 = ["as", "style"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var DEFAULT_AQL_QUERY = '';
var SEARCH_FORM_ID = 'linkDataSource.assets.configModal.searchContainer-form';

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled
var SchemaSelectContainer = (0, _react.forwardRef)(function (_ref, __cmplr) {
  var _ref$as = _ref.as,
    C = _ref$as === void 0 ? "div" : _ref$as,
    __cmpls = _ref.style,
    __cmplp = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  return /*#__PURE__*/React.createElement(C, (0, _extends2.default)({}, __cmplp, {
    style: __cmpls,
    ref: __cmplr,
    className: (0, _runtime.ax)(["_1bsb1osq _p12ffkat", __cmplp.className])
  }));
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled
if (process.env.NODE_ENV !== 'production') {
  SchemaSelectContainer.displayName = 'SchemaSelectContainer';
}
var FormContainer = (0, _react.forwardRef)(function (_ref2, __cmplr) {
  var _ref2$as = _ref2.as,
    C = _ref2$as === void 0 ? "form" : _ref2$as,
    __cmpls = _ref2.style,
    __cmplp = (0, _objectWithoutProperties2.default)(_ref2, _excluded2);
  return /*#__PURE__*/React.createElement(C, (0, _extends2.default)({}, __cmplp, {
    style: __cmpls,
    ref: __cmplr,
    className: (0, _runtime.ax)(["_1e0c11p5 _1p57pxbi _1bsb1osq", __cmplp.className])
  }));
});
if (process.env.NODE_ENV !== 'production') {
  FormContainer.displayName = 'FormContainer';
}
var AssetsSearchContainer = exports.AssetsSearchContainer = function AssetsSearchContainer(props) {
  var onSearch = props.onSearch,
    workspaceId = props.workspaceId,
    initialSearchData = props.initialSearchData,
    modalTitle = props.modalTitle,
    isSearching = props.isSearching;
  var _useDatasourceAnalyti = (0, _analytics.useDatasourceAnalyticsEvents)(),
    fireEvent = _useDatasourceAnalyti.fireEvent;
  var onFormSubmit = function onFormSubmit(searchFormValues) {
    var aql = searchFormValues.aql,
      objectSchema = searchFormValues.objectSchema;
    if (aql && objectSchema) {
      fireEvent('ui.aqlEditor.searched', {});
      // Pass the validated aql and object schema back to modal
      onSearch(aql, objectSchema.value);
    }
  };
  return /*#__PURE__*/React.createElement(_form.default, {
    onSubmit: onFormSubmit
  }, function (_ref3) {
    var _initialSearchData$ob, _initialSearchData$ob2, _initialSearchData$aq;
    var formProps = _ref3.formProps;
    return /*#__PURE__*/React.createElement(FormContainer, (0, _extends2.default)({}, formProps, {
      id: SEARCH_FORM_ID
    }), /*#__PURE__*/React.createElement(_styled.FormRowContainer, {
      isNarrowGap: true
    }, modalTitle, /*#__PURE__*/React.createElement(SchemaSelectContainer, null, /*#__PURE__*/React.createElement(_objectSchemaSelect.AssetsObjectSchemaSelect, {
      value: (_initialSearchData$ob = initialSearchData.objectSchema) !== null && _initialSearchData$ob !== void 0 ? _initialSearchData$ob : undefined,
      workspaceId: workspaceId,
      initialObjectSchemas: (_initialSearchData$ob2 = initialSearchData.objectSchemas) !== null && _initialSearchData$ob2 !== void 0 ? _initialSearchData$ob2 : undefined,
      classNamePrefix: "assets-datasource-modal--object-schema-select"
    }))), /*#__PURE__*/React.createElement(_styled.FormRowContainer, null, /*#__PURE__*/React.createElement(_aqlSearchInput.AqlSearchInput, {
      value: (_initialSearchData$aq = initialSearchData.aql) !== null && _initialSearchData$aq !== void 0 ? _initialSearchData$aq : DEFAULT_AQL_QUERY,
      workspaceId: workspaceId,
      isSearching: isSearching
    })));
  });
};
