// @atlaskit/link-picker/dist/cjs/ui/error-boundary/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ErrorBoundary = ErrorBoundary;
var _react = _interopRequireWildcard(require("react"));
var _analyticsNext = require("@atlaskit/analytics-next");
var _experiences = require("../../common/analytics/experiences");
var _constants = require("../../common/constants");
var _analytics = _interopRequireDefault(require("../../common/utils/analytics/analytics.codegen"));
var _sessionProvider = require("../../controllers/session-provider");
var _errorBoundaryBase = require("./error-boundary-base");
var _errorBoundaryFallback = require("./error-boundary-fallback");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ErrorBoundary(_ref) {
  var children = _ref.children;
  var _useAnalyticsEvents = (0, _analyticsNext.useAnalyticsEvents)(),
    createAnalyticsEvent = _useAnalyticsEvents.createAnalyticsEvent;
  var linkPickerSessionId = (0, _sessionProvider.useLinkPickerSessionId)();
  var handleError = (0, _react.useCallback)(function (error, info) {
    var _window, _info$componentStack;
    // Fire Analytics event
    createAnalyticsEvent((0, _analytics.default)('ui.linkPicker.unhandledErrorCaught', {
      browserInfo: ((_window = window) === null || _window === void 0 || (_window = _window.navigator) === null || _window === void 0 ? void 0 : _window.userAgent) || 'unknown',
      error: error.toString(),
      componentStack: (_info$componentStack = info === null || info === void 0 ? void 0 : info.componentStack) !== null && _info$componentStack !== void 0 ? _info$componentStack : ''
    })).fire(_constants.ANALYTICS_CHANNEL);

    // Fire UFO failed experience
    (0, _experiences.failUfoExperience)(_experiences.ufoExperience.mounted, linkPickerSessionId);
  }, [createAnalyticsEvent, linkPickerSessionId]);
  return /*#__PURE__*/_react.default.createElement(_errorBoundaryBase.BaseErrorBoundary, {
    onError: handleError,
    ErrorComponent: _errorBoundaryFallback.ErrorBoundaryFallback
  }, children);
}
