// @atlaskit/drawer/dist/cjs/compiled/drawer-panel/drawer-close-button.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DrawerCloseButton = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _new = require("@atlaskit/button/new");
var _arrowLeft = _interopRequireDefault(require("@atlaskit/icon/core/migration/arrow-left"));
var _context = require("../../context");
/**
 * __Drawer close button
 *
 * A button that closes the drawer.
 */
var DrawerCloseButton = exports.DrawerCloseButton = function DrawerCloseButton(_ref) {
  var Icon = _ref.icon,
    _ref$label = _ref.label,
    label = _ref$label === void 0 ? 'Close drawer' : _ref$label,
    _ref$testId = _ref.testId,
    testId = _ref$testId === void 0 ? 'DrawerCloseButton' : _ref$testId;
  (0, _context.useEnsureIsInsideDrawer)();
  var onClose = (0, _context.useOnClose)();
  return /*#__PURE__*/_react.default.createElement(_new.IconButton, {
    onClick: onClose,
    testId: testId,
    icon: Icon ? function (iconProps) {
      return /*#__PURE__*/_react.default.createElement(Icon, (0, _extends2.default)({}, iconProps, {
        size: "large",
        LEGACY_size: "large"
      }));
    } : _arrowLeft.default,
    label: label,
    shape: "circle",
    appearance: "subtle"
  });
};
