// @atlaskit/media-viewer/dist/es2019/analytics/events/operational/download.js
import { getFileAttributes } from '../..';
import { getPrimaryErrorReason, getSecondaryErrorReason, getErrorDetail, getRequestMetadata } from '../../../errors';
export const createDownloadSucceededEventPayload = (fileState, traceContext) => {
  const {
    fileId,
    fileMediatype,
    fileMimetype,
    fileSize
  } = getFileAttributes(fileState);
  return {
    eventType: 'operational',
    actionSubject: 'mediaFile',
    action: 'downloadSucceeded',
    attributes: {
      status: 'success',
      fileMediatype,
      fileMimetype,
      fileAttributes: {
        fileId,
        fileMediatype,
        fileMimetype,
        fileSize
      },
      traceContext
    }
  };
};
export const createDownloadFailedEventPayload = (fileId, error, fileState, traceContext) => {
  const {
    fileMediatype,
    fileMimetype,
    fileSize
  } = getFileAttributes(fileState);
  return {
    eventType: 'operational',
    actionSubject: 'mediaFile',
    action: 'downloadFailed',
    attributes: {
      status: 'fail',
      failReason: getPrimaryErrorReason(error),
      error: getSecondaryErrorReason(error),
      errorDetail: getErrorDetail(error),
      request: getRequestMetadata(error),
      fileMimetype,
      fileMediatype,
      fileAttributes: {
        fileId,
        fileMediatype,
        fileMimetype,
        fileSize
      },
      traceContext
    }
  };
};
