// @atlaskit/media-viewer/dist/es2019/analytics/events/operational/zipEntryLoadFailed.js
import { getFileAttributes } from '../..';
import { getMimeTypeFromFilename } from '../../../utils';
import { getPrimaryErrorReason, getSecondaryErrorReason, getErrorDetail } from '../../../errors';
export const createZipEntryLoadFailedEvent = (fileState, error, zipEntry) => {
  const {
    fileId,
    fileMediatype,
    fileMimetype,
    fileSize,
    fileStatus
  } = getFileAttributes(fileState);
  return {
    eventType: 'operational',
    actionSubject: 'mediaFile',
    action: 'zipEntryLoadFailed',
    attributes: {
      status: 'fail',
      failReason: getPrimaryErrorReason(error),
      error: getSecondaryErrorReason(error),
      errorDetail: getErrorDetail(error),
      fileAttributes: {
        fileId,
        fileMediatype,
        fileMimetype,
        fileSize,
        fileStatus
      },
      size: zipEntry ? zipEntry.size : -1,
      encrypted: zipEntry ? zipEntry.encrypted : false,
      compressedSize: zipEntry ? zipEntry.compressedSize : -1,
      mimeType: zipEntry ? getMimeTypeFromFilename(zipEntry.name) : 'undefined'
    }
  };
};
