// @atlaskit/analytics-listeners/dist/cjs/linkingPlatform/process-event.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _last = _interopRequireDefault(require("lodash/last"));
var _merge = _interopRequireDefault(require("lodash/merge"));
var _analyticsGasTypes = require("@atlaskit/analytics-gas-types");
var _analyticsNamespacedContext = require("@atlaskit/analytics-namespaced-context");
var _extractDataFromEvent = require("../helpers/extract-data-from-event");
var _types = require("../types");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var listenerVersion = "9.0.1";
var _default = exports.default = function _default(event) {
  var sources = (0, _extractDataFromEvent.getSources)(event, _analyticsNamespacedContext.LINKING_PLATFORM_CONTEXT);
  var packages = (0, _extractDataFromEvent.getPackageInfo)(event, _analyticsNamespacedContext.LINKING_PLATFORM_CONTEXT);
  var components = (0, _extractDataFromEvent.getComponents)(event, _analyticsNamespacedContext.LINKING_PLATFORM_CONTEXT);
  var extraAttributes = (0, _extractDataFromEvent.getExtraAttributes)(event, _analyticsNamespacedContext.LINKING_PLATFORM_CONTEXT);
  var _ref = (0, _last.default)(packages) || {},
    packageName = _ref.packageName,
    packageVersion = _ref.packageVersion;
  var packageHierarchy = packages.map(function (p) {
    return p.packageVersion ? "".concat(p.packageName, "@").concat(p.packageVersion) : p.packageName;
  });
  var tags = new Set(event.payload.tags || []);
  tags.add(_types.FabricChannel.linkingPlatform);
  var payload = {
    name: event.payload.name,
    action: event.payload.action,
    actionSubject: event.payload.actionSubject,
    actionSubjectId: event.payload.actionSubjectId,
    eventType: event.payload.eventType,
    attributes: _objectSpread({
      packageName: packageName,
      packageVersion: packageVersion,
      listenerVersion: listenerVersion,
      sourceHierarchy: sources.join('.') || undefined,
      componentHierarchy: components.join('.') || undefined,
      packageHierarchy: packageHierarchy.join(',') || undefined
    }, (0, _merge.default)(extraAttributes, event.payload.attributes)),
    tags: Array.from(tags),
    source: (0, _last.default)(sources) || _analyticsGasTypes.DEFAULT_SOURCE,
    nonPrivacySafeAttributes: event.payload.nonPrivacySafeAttributes
  };
  return payload;
};
