// @atlaskit/page/dist/es2019/GridColumn.js
var _a;
import React, { Component } from 'react';
import { ThemeProvider, withTheme } from 'styled-components';
import GridColumn from './internal/GridColumnElement';
import { defaultGridColumns } from './internal/vars';
const defaultSpacing = 'cosy';
export default withTheme((_a = class AkGridColumn extends Component {
        constructor() {
            super(...arguments);
            this.getTheme = (props) => ({
                columns: props.theme && props.theme.columns
                    ? props.theme.columns
                    : defaultGridColumns,
                spacing: props.theme && props.theme.spacing
                    ? props.theme.spacing
                    : defaultSpacing,
                isNestedGrid: false,
            });
            this.getNestedTheme = (props) => ({
                columns: props.medium,
                spacing: props.theme && props.theme.spacing
                    ? props.theme.spacing
                    : defaultSpacing,
                isNestedGrid: true,
            });
        }
        render() {
            return (React.createElement(ThemeProvider, { theme: this.getTheme(this.props) },
                React.createElement(GridColumn, { medium: this.props.medium },
                    React.createElement(ThemeProvider, { theme: this.getNestedTheme(this.props) },
                        React.createElement("div", null, this.props.children)))));
        }
    },
    _a.defaultProps = {
        medium: 0,
    },
    _a));
