// @atlaskit/smart-user-picker/dist/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "EmailType", {
  enumerable: true,
  get: function get() {
    return _userPicker.EmailType;
  }
});
Object.defineProperty(exports, "GroupType", {
  enumerable: true,
  get: function get() {
    return _userPicker.GroupType;
  }
});
Object.defineProperty(exports, "TeamType", {
  enumerable: true,
  get: function get() {
    return _userPicker.TeamType;
  }
});
Object.defineProperty(exports, "UserType", {
  enumerable: true,
  get: function get() {
    return _userPicker.UserType;
  }
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _components.default;
  }
});
Object.defineProperty(exports, "getUserRecommendations", {
  enumerable: true,
  get: function get() {
    return _service.getUserRecommendations;
  }
});
Object.defineProperty(exports, "hydrateDefaultValues", {
  enumerable: true,
  get: function get() {
    return _service.hydrateDefaultValues;
  }
});
Object.defineProperty(exports, "isEmail", {
  enumerable: true,
  get: function get() {
    return _userPicker.isEmail;
  }
});
Object.defineProperty(exports, "isExternalUser", {
  enumerable: true,
  get: function get() {
    return _userPicker.isExternalUser;
  }
});
Object.defineProperty(exports, "isGroup", {
  enumerable: true,
  get: function get() {
    return _userPicker.isGroup;
  }
});
Object.defineProperty(exports, "isTeam", {
  enumerable: true,
  get: function get() {
    return _userPicker.isTeam;
  }
});
Object.defineProperty(exports, "isUser", {
  enumerable: true,
  get: function get() {
    return _userPicker.isUser;
  }
});
Object.defineProperty(exports, "isValidEmail", {
  enumerable: true,
  get: function get() {
    return _userPicker.isValidEmail;
  }
});
var _components = _interopRequireDefault(require("./components"));
var _service = require("./service");
var _userPicker = require("@atlaskit/user-picker");
