// @atlaskit/media-viewer/dist/cjs/viewers/archiveSidebar/archive-download-button.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ArchiveDownloadButton = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _download = _interopRequireDefault(require("@atlaskit/icon/core/migration/download"));
var _styleWrappers = require("./styleWrappers");
var _abuseModal = require("@atlaskit/media-ui/abuseModal");
var _tooltip = _interopRequireDefault(require("@atlaskit/tooltip"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var ArchiveDownloadButton = exports.ArchiveDownloadButton = function ArchiveDownloadButton(_ref) {
  var downloadFn = _ref.downloadFn,
    shouldRenderAbuseModal = _ref.shouldRenderAbuseModal,
    isDisabled = _ref.isDisabled,
    tooltip = _ref.tooltip;
  var _useState = (0, _react.useState)(false),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    isAbuseModalOpen = _useState2[0],
    setIsAbuseModalOpen = _useState2[1];
  var downloadOrWarn = function downloadOrWarn() {
    if (isDisabled) {
      return;
    }
    if (shouldRenderAbuseModal) {
      setIsAbuseModalOpen(true);
    } else {
      downloadFn();
    }
  };
  var icon = /*#__PURE__*/_react.default.createElement(_download.default, {
    color: isDisabled ? "var(--ds-icon-disabled, #091E424F)" : 'currentColor',
    spacing: "spacious",
    label: "Download"
  });
  var downloadButton = isDisabled ? /*#__PURE__*/_react.default.createElement(_styleWrappers.DisabledArchiveDownloadButtonWrapper, null, icon) : /*#__PURE__*/_react.default.createElement(_styleWrappers.ArchiveDownloadButtonWrapper, {
    onClick: downloadOrWarn
  }, icon);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, shouldRenderAbuseModal && /*#__PURE__*/_react.default.createElement(_abuseModal.AbuseModal, {
    isOpen: isAbuseModalOpen,
    onConfirm: downloadFn,
    onClose: function onClose() {
      return setIsAbuseModalOpen(false);
    }
  }), tooltip ? /*#__PURE__*/_react.default.createElement(_tooltip.default, {
    content: tooltip,
    position: "bottom",
    tag: "span"
  }, downloadButton) : downloadButton);
};
