// @atlaskit/media-viewer/dist/cjs/viewers/archiveSidebar/styleWrappers-compiled.js
/* styleWrappers-compiled.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SidebarItemWrapper = exports.SidebarHeaderWrapper = exports.SidebarHeaderIcon = exports.SidebarHeaderEntry = exports.Separator = exports.DisabledArchiveDownloadButtonWrapper = exports.ArchiveViewerWrapper = exports.ArchiveSidebarFolderWrapper = exports.ArchiveSidebarFileEntryWrapper = exports.ArchiveSideBar = exports.ArchiveLayout = exports.ArchiveItemViewerWrapper = exports.ArchiveDownloadButtonWrapper = exports.ARCHIVE_SIDE_BAR_WIDTH = void 0;
require("./styleWrappers-compiled.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _colors = require("@atlaskit/theme/colors");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
var ARCHIVE_SIDE_BAR_WIDTH = exports.ARCHIVE_SIDE_BAR_WIDTH = 300;
var archiveItemViewerWrapperStyles = null;
var archiveSideBarStyles = null;
var slideDown = null;
var archiveDownloadButtonWrapperStyles = null;
var disabledArchiveDownloadButtonWrapperStyles = null;
var archiveSidebarFolderWrapperStyles = null;
var sidebarItemWrapperStyles = null;
var archiveSidebarFileEntryWrapperStyles = null;
var archiveLayoutStyles = null;
var archiveViewerWrapperStyles = null;
var separatorStyles = null;
var sidebarHeaderWrapperStyles = null;
var sidebarHeaderIconStyles = null;
var sidebarHeaderEntryStyles = null;
var ArchiveItemViewerWrapper = exports.ArchiveItemViewerWrapper = function ArchiveItemViewerWrapper(_ref) {
  var children = _ref.children;
  return /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)(["_1bsb1osq _1e0c1txw _1bah1h6o"])
  }, children);
};
var ArchiveSideBar = exports.ArchiveSideBar = /*#__PURE__*/(0, _react.forwardRef)(function (_ref2, ref) {
  var children = _ref2.children;
  return /*#__PURE__*/React.createElement("div", {
    ref: ref,
    className: (0, _runtime.ax)(["_ca0qqslr _u5f3v47k _19bvv47k _n3tdv47k _bfhky0hh _kqswstnw _1ltvidpf _154iidpf _1bsb15ov _94n5idpf _vchhusvi _18m915zi"]),
    style: {
      "--_53yrec": (0, _runtime.ix)("".concat(ARCHIVE_SIDE_BAR_WIDTH, "px"))
    }
  }, children);
});
var ArchiveSidebarFolderWrapper = exports.ArchiveSidebarFolderWrapper = function ArchiveSidebarFolderWrapper(_ref3) {
  var children = _ref3.children;
  return /*#__PURE__*/React.createElement("div", {
    "data-testid": "archive-sidebar-folder-wrapper",
    className: (0, _runtime.ax)(["_v564t24z _y44v1nhz _t9ecz6y5 _tzy4idpf"]),
    style: {
      "--_1yttpsa": (0, _runtime.ix)("".concat(slideDown, " 0.3s forwards"))
    }
  }, children);
};
var ArchiveDownloadButtonWrapper = exports.ArchiveDownloadButtonWrapper = function ArchiveDownloadButtonWrapper(_ref4) {
  var children = _ref4.children,
    onClick = _ref4.onClick;
  return /*#__PURE__*/React.createElement("div", {
    onClick: onClick,
    "data-testid": "media-archiveDownloadButton",
    className: (0, _runtime.ax)(["_19itglyw _2rko1l7b _ca0qu2gc _u5f31v6z _n3td14y2 _19bvu2gc _bfhk1j28 _syazwydc _d0altlke _irr3ebx2 _1qamtlke _1di6j4ot"])
  }, children);
};
var DisabledArchiveDownloadButtonWrapper = exports.DisabledArchiveDownloadButtonWrapper = function DisabledArchiveDownloadButtonWrapper(_ref5) {
  var children = _ref5.children;
  return /*#__PURE__*/React.createElement("div", {
    "data-testid": "media-disabledArchiveDownloadButton",
    className: (0, _runtime.ax)(["_19itglyw _2rko1l7b _ca0qu2gc _u5f31v6z _n3td14y2 _19bvu2gc _bfhk1j28 _syazwydc _80om13gf"])
  }, children);
};
var SidebarItemWrapper = exports.SidebarItemWrapper = function SidebarItemWrapper(_ref6) {
  var children = _ref6.children;
  return /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)(["_1bsb13sl"])
  }, children);
};
var ArchiveSidebarFileEntryWrapper = exports.ArchiveSidebarFileEntryWrapper = function ArchiveSidebarFileEntryWrapper(_ref7) {
  var children = _ref7.children,
    index = _ref7.index;
  return /*#__PURE__*/React.createElement("div", {
    key: index,
    className: (0, _runtime.ax)(["_v564rca0 _n3td1vik _1e0c1txw _4cvr1h6o _80omtlke"])
  }, children);
};
var ArchiveLayout = exports.ArchiveLayout = function ArchiveLayout(_ref8) {
  var children = _ref8.children;
  return /*#__PURE__*/React.createElement("div", {
    "data-testid": "archive-layout",
    className: (0, _runtime.ax)(["_1e0c1txw _1bsb1osq _4t3i1osq"])
  }, children);
};
var ArchiveViewerWrapper = exports.ArchiveViewerWrapper = function ArchiveViewerWrapper(_ref9) {
  var children = _ref9.children;
  return /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)(["_kqswstnw _154iidpf _1ltv15ov _1xi2idpf _94n5idpf _4cvr1h6o _1e0c1txw"]),
    style: {
      "--_53yrec": (0, _runtime.ix)("".concat(ARCHIVE_SIDE_BAR_WIDTH, "px"))
    }
  }, children);
};
var Separator = exports.Separator = function Separator() {
  return /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)(["_2rkot94y _4t3iyh40 _19pk1te7 _2hwxidpf _otyr1te7 _18u0idpf _bfhk21we _1o9zidpf"])
  });
};
var SidebarHeaderWrapper = exports.SidebarHeaderWrapper = function SidebarHeaderWrapper(_ref10) {
  var children = _ref10.children;
  return /*#__PURE__*/React.createElement("span", {
    className: (0, _runtime.ax)(["_1e0c1txw _4cvr1h6o _1o9zidpf"])
  }, children);
};
var SidebarHeaderIcon = exports.SidebarHeaderIcon = function SidebarHeaderIcon(_ref11) {
  var children = _ref11.children;
  return /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)(["_1e0c1txw _4cvr1h6o _2hwx1yov _1o9zidpf"])
  }, children);
};
var SidebarHeaderEntry = exports.SidebarHeaderEntry = function SidebarHeaderEntry(_ref12) {
  var children = _ref12.children;
  return /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)(["_16jlkb7n _1o9zkb7n _i0dl1wug _1reo15vq _18m915vq _1bto1l2s _o5721q9c _vwz41xde _syazrsv4"]),
    style: {
      "--_1k41sl": (0, _runtime.ix)("var(--ds-text, ".concat(_colors.DN500, ")"))
    }
  }, children);
};
