// @atlaskit/media-viewer/dist/cjs/viewers/archiveSidebar/styleWrappers-emotion.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SidebarItemWrapper = exports.SidebarHeaderWrapper = exports.SidebarHeaderIcon = exports.SidebarHeaderEntry = exports.Separator = exports.DisabledArchiveDownloadButtonWrapper = exports.ArchiveViewerWrapper = exports.ArchiveSidebarFolderWrapper = exports.ArchiveSidebarFileEntryWrapper = exports.ArchiveSideBar = exports.ArchiveLayout = exports.ArchiveItemViewerWrapper = exports.ArchiveDownloadButtonWrapper = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = require("react");
var _react2 = require("@emotion/react");
var _styles = require("./styles");
var _colors = require("@atlaskit/theme/colors");
/**
 * @jsxRuntime classic
 * @jsx jsx
 */

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var archiveItemViewerWrapperStyles = (0, _react2.css)({
  width: '100%',
  display: 'flex',
  justifyContent: 'center'
});
var archiveSideBarStyles = (0, _react2.css)({
  padding: "22px ".concat("var(--ds-space-250, 20px)", " ", "var(--ds-space-250, 20px)", " ", "var(--ds-space-250, 20px)"),
  backgroundColor: "var(--ds-surface, #101214)",
  position: 'absolute',
  left: 0,
  top: 0,
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  width: "".concat(_styles.ArchiveSideBarWidth, "px"),
  bottom: 0,
  boxSizing: 'border-box',
  overflowY: 'scroll'
});
var slideDown = (0, _react2.keyframes)({
  '0%': {
    opacity: 0,
    transform: 'translateY(-100%)'
  },
  '100%': {
    transform: 'translateY(0)',
    opacity: 1
  }
});
var archiveDownloadButtonWrapperStyles = (0, _react2.css)({
  padding: "var(--ds-space-100, 8px)".concat(" 7px 5px ", "var(--ds-space-100, 8px)"),
  border: 'none',
  borderRadius: '3px',
  backgroundColor: 'transparent',
  color: "var(--ds-icon, #9FADBC)",
  '&:hover': {
    cursor: 'pointer',
    backgroundColor: "var(--ds-background-neutral-subtle-hovered, #A1BDD914)"
  },
  '&:active': {
    cursor: 'pointer',
    backgroundColor: "var(--ds-background-neutral-subtle-pressed, #A6C5E229)"
  }
});
var disabledArchiveDownloadButtonWrapperStyles = (0, _react2.css)({
  padding: "var(--ds-space-100, 8px)".concat(" 7px 5px ", "var(--ds-space-100, 8px)"),
  border: 'none',
  borderRadius: '3px',
  backgroundColor: 'transparent',
  color: "var(--ds-icon, #9FADBC)",
  cursor: 'not-allowed'
});
var archiveSidebarFolderWrapperStyles = (0, _react2.css)({
  transform: 'translateY(-100%)',
  transition: 'all 1s',
  opacity: 0,
  animation: "".concat(slideDown, " 0.3s forwards")
});
var sidebarItemWrapperStyles = (0, _react2.css)({
  width: '85%'
});
var archiveSidebarFileEntryWrapperStyles = (0, _react2.css)({
  paddingBottom: "var(--ds-space-075, 5px)",
  display: 'flex',
  alignItems: 'center',
  cursor: 'pointer',
  transition: 'background-color 0.3s'
});
var archiveLayoutStyles = (0, _react2.css)({
  display: 'flex',
  width: '100%',
  height: '100%'
});
var archiveViewerWrapperStyles = (0, _react2.css)({
  position: 'absolute',
  top: 0,
  // eslint-disable-next-line @atlaskit/design-system/ensure-design-token-usage, @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  left: "".concat(_styles.ArchiveSideBarWidth, "px"),
  right: 0,
  bottom: 0,
  alignItems: 'center',
  display: 'flex'
});
var separatorStyles = (0, _react2.css)({
  borderRadius: '1px',
  height: '2px',
  margin: "var(--ds-space-200, 19px)".concat(" 0"),
  backgroundColor: "var(--ds-border, #A6C5E229)",
  flexShrink: 0
});
var sidebarHeaderWrapperStyles = (0, _react2.css)({
  display: 'flex',
  alignItems: 'center',
  flexShrink: 0
});
var sidebarHeaderIconStyles = (0, _react2.css)({
  display: 'flex',
  alignItems: 'center',
  marginRight: "var(--ds-space-100, 10px)",
  flexShrink: 0
});
var sidebarHeaderEntryStyles = (0, _react2.css)((0, _defineProperty2.default)((0, _defineProperty2.default)({
  flex: '1 1 auto',
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap'
}, 'lineHeight', 1.14286), "color", "var(--ds-text, ".concat(_colors.DN500, ")")));
var ArchiveItemViewerWrapper = exports.ArchiveItemViewerWrapper = function ArchiveItemViewerWrapper(_ref) {
  var children = _ref.children;
  return (0, _react2.jsx)("div", {
    css: archiveItemViewerWrapperStyles
  }, children);
};
var ArchiveSideBar = exports.ArchiveSideBar = /*#__PURE__*/(0, _react.forwardRef)(function (_ref2, ref) {
  var children = _ref2.children;
  return (0, _react2.jsx)("div", {
    css: archiveSideBarStyles,
    ref: ref
  }, children);
});
var ArchiveSidebarFolderWrapper = exports.ArchiveSidebarFolderWrapper = function ArchiveSidebarFolderWrapper(_ref3) {
  var children = _ref3.children;
  return (0, _react2.jsx)("div", {
    css: archiveSidebarFolderWrapperStyles,
    "data-testid": "archive-sidebar-folder-wrapper"
  }, children);
};
var ArchiveDownloadButtonWrapper = exports.ArchiveDownloadButtonWrapper = function ArchiveDownloadButtonWrapper(_ref4) {
  var children = _ref4.children,
    onClick = _ref4.onClick;
  return (
    // eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions
    (0, _react2.jsx)("div", {
      css: archiveDownloadButtonWrapperStyles,
      onClick: onClick,
      "data-testid": "media-archiveDownloadButton"
    }, children)
  );
};
var DisabledArchiveDownloadButtonWrapper = exports.DisabledArchiveDownloadButtonWrapper = function DisabledArchiveDownloadButtonWrapper(_ref5) {
  var children = _ref5.children;
  return (
    // eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions
    (0, _react2.jsx)("div", {
      css: disabledArchiveDownloadButtonWrapperStyles,
      "data-testid": "media-disabledArchiveDownloadButton"
    }, children)
  );
};
var SidebarItemWrapper = exports.SidebarItemWrapper = function SidebarItemWrapper(_ref6) {
  var children = _ref6.children;
  return (0, _react2.jsx)("div", {
    css: sidebarItemWrapperStyles
  }, children);
};
var ArchiveSidebarFileEntryWrapper = exports.ArchiveSidebarFileEntryWrapper = function ArchiveSidebarFileEntryWrapper(_ref7) {
  var children = _ref7.children,
    index = _ref7.index;
  return (0, _react2.jsx)("div", {
    css: archiveSidebarFileEntryWrapperStyles,
    key: index
  }, children);
};
var ArchiveLayout = exports.ArchiveLayout = function ArchiveLayout(_ref8) {
  var children = _ref8.children;
  return (0, _react2.jsx)("div", {
    css: archiveLayoutStyles,
    "data-testid": "archive-layout"
  }, children);
};
var ArchiveViewerWrapper = exports.ArchiveViewerWrapper = function ArchiveViewerWrapper(_ref9) {
  var children = _ref9.children;
  return (0, _react2.jsx)("div", {
    css: archiveViewerWrapperStyles
  }, children);
};
var Separator = exports.Separator = function Separator() {
  return (0, _react2.jsx)("div", {
    css: separatorStyles
  });
};
var SidebarHeaderWrapper = exports.SidebarHeaderWrapper = function SidebarHeaderWrapper(_ref10) {
  var children = _ref10.children;
  return (0, _react2.jsx)("span", {
    css: sidebarHeaderWrapperStyles
  }, children);
};
var SidebarHeaderIcon = exports.SidebarHeaderIcon = function SidebarHeaderIcon(_ref11) {
  var children = _ref11.children;
  return (0, _react2.jsx)("div", {
    css: sidebarHeaderIconStyles
  }, children);
};
var SidebarHeaderEntry = exports.SidebarHeaderEntry = function SidebarHeaderEntry(_ref12) {
  var children = _ref12.children;
  return (0, _react2.jsx)("div", {
    css: sidebarHeaderEntryStyles
  }, children);
};
