// @atlaskit/media-file-preview/dist/es2019/errors.js
/**
 * Primary reason is logged through Data Portal.
 * Make sure all the values are whitelisted in Measure -> Event Regitry -> "mediaCardRender failed" event
 */

export class MediaFilePreviewError extends Error {
  constructor(primaryReason, secondaryError) {
    super(primaryReason);
    // https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-2.html#support-for-newtarget
    this.primaryReason = primaryReason;
    this.secondaryError = secondaryError;
    Object.setPrototypeOf(this, new.target.prototype);

    // https://v8.dev/docs/stack-trace-api
    if ('captureStackTrace' in Error) {
      Error.captureStackTrace(this, new.target);
    }
  }
}
export class LocalPreviewError extends MediaFilePreviewError {
  constructor(primaryReason, secondaryError) {
    super(primaryReason, secondaryError);
    this.primaryReason = primaryReason;
    this.secondaryError = secondaryError;
  }
}
export class RemotePreviewError extends MediaFilePreviewError {
  constructor(primaryReason, secondaryError) {
    super(primaryReason, secondaryError);
    this.primaryReason = primaryReason;
    this.secondaryError = secondaryError;
  }
}
export class SsrPreviewError extends MediaFilePreviewError {
  constructor(primaryReason, secondaryError) {
    super(primaryReason, secondaryError);
    this.primaryReason = primaryReason;
    this.secondaryError = secondaryError;
  }
}
const getImageLoadPrimaryReason = source => {
  switch (source) {
    case 'cache-remote':
      return 'cache-remote-uri';
    case 'cache-local':
      return 'cache-local-uri';
    case 'external':
      return 'external-uri';
    case 'local':
      return 'local-uri';
    case 'remote':
      return 'remote-uri';
    case 'ssr-client':
      return 'ssr-client-uri';
    case 'ssr-server':
    case 'ssr-data':
      return 'ssr-server-uri';
    // This fail reason will come from a bug, most likely.
    default:
      return `unknown-uri`;
  }
};
export class ImageLoadError extends MediaFilePreviewError {
  constructor(source) {
    super(getImageLoadPrimaryReason(source));
  }
}
export function isMediaFilePreviewError(err) {
  return err instanceof MediaFilePreviewError;
}
export const isLocalPreviewError = err => err instanceof LocalPreviewError;
export const isRemotePreviewError = err => err instanceof RemotePreviewError;
export const isUnsupportedLocalPreviewError = err => isMediaFilePreviewError(err) && err.primaryReason === 'local-preview-unsupported';

// In a try/catch statement, the error caught is the type of unknown.
// We can use this helper to ensure that the error handled is the type of MediaFilePreviewError if unsure
// If updatePrimaryReason is true, if it's a MediaFilePreviewError already, it will update it's primary reason
export const ensureMediaFilePreviewError = (primaryReason, error, updatePrimaryReason) => {
  if (isMediaFilePreviewError(error)) {
    if (updatePrimaryReason && error.primaryReason !== primaryReason) {
      return new MediaFilePreviewError(primaryReason, error.secondaryError);
    }
    return error;
  }
  return new MediaFilePreviewError(primaryReason, error);
};
