// @atlaskit/media-viewer/dist/esm/analytics/events/operational/loadFailed.js
import { getFileAttributes } from '../..';
import { getPrimaryErrorReason, getSecondaryErrorReason, getErrorDetail, getRequestMetadata } from '../../../errors';
export var createLoadFailedEvent = function createLoadFailedEvent(fileId, error, fileState, traceContext) {
  var _getFileAttributes = getFileAttributes(fileState),
    fileMediatype = _getFileAttributes.fileMediatype,
    fileMimetype = _getFileAttributes.fileMimetype,
    fileSize = _getFileAttributes.fileSize;
  return {
    eventType: 'operational',
    actionSubject: 'mediaFile',
    action: 'loadFailed',
    attributes: {
      status: 'fail',
      failReason: getPrimaryErrorReason(error),
      error: getSecondaryErrorReason(error),
      errorDetail: getErrorDetail(error),
      request: getRequestMetadata(error),
      fileMimetype: fileMimetype,
      fileAttributes: {
        fileId: fileId,
        fileMediatype: fileMediatype,
        fileMimetype: fileMimetype,
        fileSize: fileSize
      },
      traceContext: fileMediatype === 'image' ? traceContext : undefined
    }
  };
};
