// @atlaskit/jql-editor-autocomplete-rest/dist/cjs/hooks/use-fetch-field-values/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useFetchFieldValues = exports.getAutocompleteSuggestionsUrl = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _react = require("react");
var _jqlEditorCommon = require("@atlaskit/jql-editor-common");
var _analytics = require("../../analytics");
var getAutocompleteSuggestionsUrl = exports.getAutocompleteSuggestionsUrl = function getAutocompleteSuggestionsUrl(field, query) {
  return "/rest/api/latest/jql/autocompletedata/suggestions?fieldName=".concat(field, "&fieldValue=").concat(query !== null && query !== void 0 ? query : '');
};
var useFetchFieldValues = exports.useFetchFieldValues = function useFetchFieldValues(getSuggestions, createAndFireAnalyticsEvent) {
  return (0, _react.useCallback)( /*#__PURE__*/function () {
    var _ref = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(field, query) {
      var data, div, results;
      return _regenerator.default.wrap(function _callee$(_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            _context.prev = 0;
            _context.next = 3;
            return getSuggestions(getAutocompleteSuggestionsUrl(field, query));
          case 3:
            data = _context.sent;
            div = document.createElement('div');
            results = data.results.map(function (operand) {
              var _div$textContent;
              // The displayName from the API includes HTML tags to bold matching text e.g. `<b>o</b>pen`
              // We set this as innerHTML so we can get the raw textContent
              div.innerHTML = operand.displayName;
              return {
                name: (_div$textContent = div.textContent) !== null && _div$textContent !== void 0 ? _div$textContent : '',
                value: operand.value
              };
            });
            createAndFireAnalyticsEvent({
              action: _analytics.Action.SUCCESS,
              actionSubject: _analytics.ActionSubject.AUTOCOMPLETE_SUGGESTIONS,
              eventType: _jqlEditorCommon.EventType.OPERATIONAL
            });
            return _context.abrupt("return", results);
          case 10:
            _context.prev = 10;
            _context.t0 = _context["catch"](0);
            createAndFireAnalyticsEvent({
              action: _analytics.Action.FAILED,
              actionSubject: _analytics.ActionSubject.AUTOCOMPLETE_SUGGESTIONS,
              eventType: _jqlEditorCommon.EventType.OPERATIONAL
            });
            throw _context.t0;
          case 14:
          case "end":
            return _context.stop();
        }
      }, _callee, null, [[0, 10]]);
    }));
    return function (_x, _x2) {
      return _ref.apply(this, arguments);
    };
  }(), [createAndFireAnalyticsEvent, getSuggestions]);
};
