// @atlaskit/teams-public/dist/esm/ui/team-containers/main.js
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _regeneratorRuntime from "@babel/runtime/regenerator";
import React, { useCallback, useEffect, useState } from 'react';
import { defineMessages, FormattedMessage } from 'react-intl-next';
import { useAnalyticsEvents } from '@atlaskit/analytics-next';
import Button from '@atlaskit/button/new';
import ModalTransition from '@atlaskit/modal-dialog/modal-transition';
import { Grid, Inline, Stack } from '@atlaskit/primitives';
import { N0, N90 } from '@atlaskit/theme/colors';
import { AnalyticsAction, usePeopleAndTeamAnalytics } from '../../common/utils/analytics';
import { hasProductPermission } from '../../controllers';
import { useProductPermissions } from '../../controllers/hooks/use-product-permission';
import { useTeamContainers, useTeamContainersHook } from '../../controllers/hooks/use-team-containers';
import { AddContainerCard } from './add-container-card';
import { DisconnectDialogLazy } from './disconnect-dialog/async';
import { LinkedContainerCard } from './linked-container-card';
import { NoProductAccessState } from './no-product-access-empty-state';
import { TeamContainersSkeleton } from './team-containers-skeleton';
export var ICON_BACKGROUND = "var(--ds-icon-inverse, ".concat(N0, ")");
export var ICON_COLOR = "var(--ds-icon-subtle, ".concat(N90, ")");
export var MAX_NUMBER_OF_CONTAINERS_TO_SHOW = 4;
export var TeamContainers = function TeamContainers(_ref) {
  var teamId = _ref.teamId,
    _onAddAContainerClick = _ref.onAddAContainerClick,
    components = _ref.components,
    userId = _ref.userId,
    cloudId = _ref.cloudId,
    filterContainerId = _ref.filterContainerId,
    isDisplayedOnProfileCard = _ref.isDisplayedOnProfileCard;
  var _useAnalyticsEvents = useAnalyticsEvents(),
    createAnalyticsEvent = _useAnalyticsEvents.createAnalyticsEvent;
  var _useTeamContainers = useTeamContainers(teamId),
    teamContainers = _useTeamContainers.teamContainers,
    loading = _useTeamContainers.loading,
    unlinkError = _useTeamContainers.unlinkError;
  var _useTeamContainersHoo = useTeamContainersHook(),
    _useTeamContainersHoo2 = _slicedToArray(_useTeamContainersHoo, 2),
    _ = _useTeamContainersHoo2[0],
    actions = _useTeamContainersHoo2[1];
  var _useState = useState(false),
    _useState2 = _slicedToArray(_useState, 2),
    showAddJiraContainer = _useState2[0],
    setShowAddJiraContainer = _useState2[1];
  var _useState3 = useState(false),
    _useState4 = _slicedToArray(_useState3, 2),
    showAddConfluenceContainer = _useState4[0],
    setShowAddConfluenceContainer = _useState4[1];
  var _useState5 = useState(false),
    _useState6 = _slicedToArray(_useState5, 2),
    showMore = _useState6[0],
    setShowMore = _useState6[1];
  var _useState7 = useState(false),
    _useState8 = _slicedToArray(_useState7, 2),
    isDisconnectDialogOpen = _useState8[0],
    setIsDisconnectDialogOpen = _useState8[1];
  var _useState9 = useState(),
    _useState10 = _slicedToArray(_useState9, 2),
    selectedContainerDetails = _useState10[0],
    setSelectedContainerDetails = _useState10[1];
  var _useState11 = useState(teamContainers),
    _useState12 = _slicedToArray(_useState11, 2),
    filteredTeamContainers = _useState12[0],
    setFilteredTeamContainers = _useState12[1];
  var _usePeopleAndTeamAnal = usePeopleAndTeamAnalytics(),
    fireOperationalEvent = _usePeopleAndTeamAnal.fireOperationalEvent,
    fireTrackEvent = _usePeopleAndTeamAnal.fireTrackEvent;
  var _useProductPermission = useProductPermissions({
      userId: userId,
      cloudId: cloudId
    }),
    productPermissions = _useProductPermission.data,
    productPermissionIsLoading = _useProductPermission.loading;
  useEffect(function () {
    if (isDisplayedOnProfileCard && filterContainerId) {
      setFilteredTeamContainers(teamContainers.filter(function (container) {
        return container.id !== filterContainerId;
      }));
    } else {
      setFilteredTeamContainers(teamContainers);
    }
  }, [isDisplayedOnProfileCard, filterContainerId, teamContainers]);
  useEffect(function () {
    if (filteredTeamContainers.length > MAX_NUMBER_OF_CONTAINERS_TO_SHOW || isDisplayedOnProfileCard) {
      setShowAddJiraContainer(false);
      setShowAddConfluenceContainer(false);
    } else {
      var hasJiraProject = filteredTeamContainers.some(function (container) {
        return container.type === 'JiraProject';
      });
      var hasConfluenceSpace = filteredTeamContainers.some(function (container) {
        return container.type === 'ConfluenceSpace';
      });
      setShowAddJiraContainer(!hasJiraProject && !!productPermissions && !!hasProductPermission(productPermissions, 'jira'));
      setShowAddConfluenceContainer(!hasConfluenceSpace && !!productPermissions && !!hasProductPermission(productPermissions, 'confluence'));
    }
  }, [isDisplayedOnProfileCard, productPermissions, filteredTeamContainers]);
  var handleShowMore = function handleShowMore() {
    setShowMore(!showMore);
  };
  var handleOpenDisconnectDialog = useCallback(function (containerDetails) {
    setSelectedContainerDetails(containerDetails);
    setIsDisconnectDialogOpen(true);
    fireTrackEvent(createAnalyticsEvent, {
      action: AnalyticsAction.OPENED,
      actionSubject: 'unlinkContainerDialog',
      attributes: {
        teamId: teamId
      }
    });
  }, [createAnalyticsEvent, fireTrackEvent, teamId]);
  var LinkedContainerCardComponent = (components === null || components === void 0 ? void 0 : components.ContainerCard) || LinkedContainerCard;
  var handleDisconnect = useCallback( /*#__PURE__*/function () {
    var _ref2 = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee(containerId) {
      var removedContainer;
      return _regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            removedContainer = filteredTeamContainers.find(function (container) {
              return container.id === containerId;
            });
            _context.next = 3;
            return actions.unlinkTeamContainers(teamId, containerId);
          case 3:
            setIsDisconnectDialogOpen(false);
            if (unlinkError) {
              fireOperationalEvent(createAnalyticsEvent, {
                action: AnalyticsAction.FAILED,
                actionSubject: 'teamContainerUnlinked'
              });
            } else {
              fireOperationalEvent(createAnalyticsEvent, {
                action: AnalyticsAction.SUCCEEDED,
                actionSubject: 'teamContainerUnlinked',
                attributes: {
                  containerRemoved: {
                    containerId: removedContainer === null || removedContainer === void 0 ? void 0 : removedContainer.id,
                    container: removedContainer === null || removedContainer === void 0 ? void 0 : removedContainer.type
                  },
                  teamId: teamId
                }
              });
            }
          case 5:
          case "end":
            return _context.stop();
        }
      }, _callee);
    }));
    return function (_x) {
      return _ref2.apply(this, arguments);
    };
  }(), [actions, createAnalyticsEvent, fireOperationalEvent, filteredTeamContainers, teamId, unlinkError]);
  var TeamContainersSkeletonComponent = (components === null || components === void 0 ? void 0 : components.TeamContainersSkeleton) || TeamContainersSkeleton;
  if (loading || productPermissionIsLoading) {
    return /*#__PURE__*/React.createElement(TeamContainersSkeletonComponent, {
      numberOfContainers: MAX_NUMBER_OF_CONTAINERS_TO_SHOW
    });
  }
  if (filteredTeamContainers.length === 0 && !isDisplayedOnProfileCard && (!productPermissions || !(productPermissions && (hasProductPermission(productPermissions, 'jira') || hasProductPermission(productPermissions, 'confluence'))))) {
    return /*#__PURE__*/React.createElement(NoProductAccessState, null);
  }
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Stack, {
    space: "space.200"
  }, /*#__PURE__*/React.createElement(Grid, {
    templateColumns: "repeat(auto-fill, minmax(300px, 1fr))",
    gap: isDisplayedOnProfileCard ? 'space.0' : 'space.100'
  }, filteredTeamContainers.slice(0, MAX_NUMBER_OF_CONTAINERS_TO_SHOW).map(function (container) {
    return /*#__PURE__*/React.createElement(LinkedContainerCardComponent, {
      key: container.id,
      containerType: container.type,
      containerTypeProperties: container.containerTypeProperties,
      title: container.name,
      containerIcon: container.icon || undefined,
      link: container.link || undefined,
      containerId: container.id,
      onDisconnectButtonClick: function onDisconnectButtonClick() {
        return handleOpenDisconnectDialog({
          containerId: container.id,
          containerType: container.type,
          containerName: container.name
        });
      }
    });
  }), showAddJiraContainer && /*#__PURE__*/React.createElement(AddContainerCard, {
    onAddAContainerClick: function onAddAContainerClick(e) {
      return _onAddAContainerClick(e, 'Jira');
    },
    containerType: "JiraProject"
  }), showAddConfluenceContainer && /*#__PURE__*/React.createElement(AddContainerCard, {
    onAddAContainerClick: function onAddAContainerClick(e) {
      return _onAddAContainerClick(e, 'Confluence');
    },
    containerType: "ConfluenceSpace"
  }), showMore && filteredTeamContainers.slice(MAX_NUMBER_OF_CONTAINERS_TO_SHOW).map(function (container) {
    return /*#__PURE__*/React.createElement(LinkedContainerCardComponent, {
      key: container.id,
      containerType: container.type,
      containerTypeProperties: container.containerTypeProperties,
      title: container.name,
      containerId: container.id,
      containerIcon: container.icon || undefined,
      link: container.link || undefined,
      onDisconnectButtonClick: function onDisconnectButtonClick() {
        return handleOpenDisconnectDialog({
          containerId: container.id,
          containerType: container.type,
          containerName: container.name
        });
      }
    });
  })), filteredTeamContainers.length > MAX_NUMBER_OF_CONTAINERS_TO_SHOW && /*#__PURE__*/React.createElement(Inline, null, /*#__PURE__*/React.createElement(Button, {
    appearance: "subtle",
    onClick: handleShowMore
  }, showMore ? /*#__PURE__*/React.createElement(FormattedMessage, messages.showLess) : /*#__PURE__*/React.createElement(FormattedMessage, messages.showMore)))), /*#__PURE__*/React.createElement(ModalTransition, null, isDisconnectDialogOpen && selectedContainerDetails && /*#__PURE__*/React.createElement(DisconnectDialogLazy, {
    containerName: selectedContainerDetails.containerName,
    containerType: selectedContainerDetails.containerType,
    onClose: function onClose() {
      return setIsDisconnectDialogOpen(false);
    },
    onDisconnect: function onDisconnect() {
      return handleDisconnect(selectedContainerDetails.containerId);
    }
  })));
};
var messages = defineMessages({
  showMore: {
    id: 'ptc-directory.team-profile-page.team-containers.show-more.non-final',
    defaultMessage: 'Show more',
    description: 'Button to show more containers'
  },
  showLess: {
    id: 'ptc-directory.team-profile-page.team-containers.show-less.non-final',
    defaultMessage: 'Show less',
    description: 'Button to show less containers'
  }
});
