// @atlaskit/react-ufo/dist/cjs/additional-payload/utils/lighthouse-metrics/utils/observer/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.startLTObserver = exports.startLSObserver = void 0;
var _selfMeasurements = require("../../../../../self-measurements");
var _const = require("../../const");
var _buffer = require("../buffer");
var pe = null;
var getObserver = (0, _selfMeasurements.withProfiling)(function getObserver() {
  if (typeof PerformanceObserver !== 'function') {
    // Only instantiate the IntersectionObserver if it's supported
    return null;
  }
  if (pe !== null) {
    return pe;
  }
  var performanceObserverCallback = (0, _selfMeasurements.withProfiling)(function performanceObserverCallback(list) {
    list.getEntries().forEach(function (entry) {
      if (entry.entryType === _const.PerformanceObserverEntryTypes.LayoutShift) {
        _buffer.EntriesBuffer[_const.PerformanceObserverEntryTypes.LayoutShift].push(entry);
      }
      if (entry.entryType === _const.PerformanceObserverEntryTypes.LongTask) {
        _buffer.EntriesBuffer[_const.PerformanceObserverEntryTypes.LongTask].push(entry);
      }
    });
  });
  pe = new PerformanceObserver(performanceObserverCallback);
  return pe;
});
var startLSObserver = exports.startLSObserver = (0, _selfMeasurements.withProfiling)(function startLSObserver() {
  var _getObserver;
  (_getObserver = getObserver()) === null || _getObserver === void 0 || _getObserver.observe({
    type: _const.PerformanceObserverEntryTypes.LayoutShift,
    buffered: true
  });
});
var startLTObserver = exports.startLTObserver = (0, _selfMeasurements.withProfiling)(function startLTObserver() {
  var _getObserver2;
  (_getObserver2 = getObserver()) === null || _getObserver2 === void 0 || _getObserver2.observe({
    type: _const.PerformanceObserverEntryTypes.LongTask,
    buffered: true
  });
});
