// @atlaskit/media-client/dist/cjs/utils/request/errors.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RequestError = void 0;
exports.isRequestError = isRequestError;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _errors = require("../../models/errors");
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var RequestError = exports.RequestError = /*#__PURE__*/function (_BaseMediaClientError) {
  function RequestError(reason, metadata, innerError) {
    (0, _classCallCheck2.default)(this, RequestError);
    return _callSuper(this, RequestError, [reason, metadata, innerError]);
  }

  // TODO: Deprecate this getter https://product-fabric.atlassian.net/browse/CXP-4665
  /** Will be deprecated. Use the properties `reason` and `metadata` instead */
  (0, _inherits2.default)(RequestError, _BaseMediaClientError);
  return (0, _createClass2.default)(RequestError, [{
    key: "attributes",
    get: function get() {
      var reason = this.reason,
        _this$metadata = this.metadata,
        _this$metadata2 = _this$metadata === void 0 ? {} : _this$metadata,
        method = _this$metadata2.method,
        endpoint = _this$metadata2.endpoint,
        mediaRegion = _this$metadata2.mediaRegion,
        mediaEnv = _this$metadata2.mediaEnv,
        attempts = _this$metadata2.attempts,
        clientExhaustedRetries = _this$metadata2.clientExhaustedRetries,
        statusCode = _this$metadata2.statusCode,
        traceContext = _this$metadata2.traceContext,
        innerError = this.innerError;
      return {
        reason: reason,
        method: method,
        endpoint: endpoint,
        mediaRegion: mediaRegion,
        mediaEnv: mediaEnv,
        attempts: attempts,
        clientExhaustedRetries: clientExhaustedRetries,
        statusCode: statusCode,
        metadata: {
          traceContext: traceContext
        },
        innerError: innerError
      };
    }
  }]);
}(_errors.BaseMediaClientError);
function isRequestError(err) {
  return err instanceof RequestError;
}
