// @atlaskit/smart-card/dist/cjs/view/EmbedModal/components/link-info/link-info-button/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _button = _interopRequireDefault(require("@atlaskit/button"));
var _tooltip = _interopRequireDefault(require("@atlaskit/tooltip"));
var _useLinkClicked = require("../../../../../state/analytics/useLinkClicked");
var LinkInfoButton = function LinkInfoButton(_ref) {
  var content = _ref.content,
    href = _ref.href,
    icon = _ref.icon,
    onClick = _ref.onClick,
    target = _ref.target,
    testId = _ref.testId;
  var onMouseDown = (0, _useLinkClicked.useMouseDownEvent)();
  return /*#__PURE__*/_react.default.createElement(_tooltip.default, {
    content: content,
    hideTooltipOnClick: true,
    tag: "span",
    testId: "".concat(testId, "-tooltip")
  }, /*#__PURE__*/_react.default.createElement(_button.default, {
    appearance: "subtle",
    href: href,
    iconBefore: icon,
    onClick: onClick,
    onMouseDown: onMouseDown,
    spacing: "none",
    target: target,
    testId: "".concat(testId, "-button")
  }));
};
var _default = exports.default = LinkInfoButton;
