// @atlaskit/editor-plugin-text-formatting/dist/es2019/ui/PrimaryToolbarComponent.js
import React from 'react';
import { useSharedPluginState } from '@atlaskit/editor-common/hooks';
import Toolbar from './Toolbar';
import { ToolbarType } from './Toolbar/types';
export function PrimaryToolbarComponent({
  api,
  popupsMountPoint,
  popupsScrollableElement,
  toolbarSize,
  editorView,
  disabled,
  isReducedSpacing,
  shouldUseResponsiveToolbar
}) {
  var _api$analytics;
  const {
    textFormattingState
  } = useSharedPluginState(api, ['textFormatting']);
  return /*#__PURE__*/React.createElement(Toolbar, {
    textFormattingState: textFormattingState,
    popupsMountPoint: popupsMountPoint,
    popupsScrollableElement: popupsScrollableElement,
    toolbarSize: toolbarSize,
    isReducedSpacing: isReducedSpacing,
    editorView: editorView,
    isToolbarDisabled: disabled,
    shouldUseResponsiveToolbar: shouldUseResponsiveToolbar,
    editorAnalyticsAPI: api === null || api === void 0 ? void 0 : (_api$analytics = api.analytics) === null || _api$analytics === void 0 ? void 0 : _api$analytics.actions,
    api: api,
    toolbarType: ToolbarType.PRIMARY
  });
}
