// @atlaskit/select/dist/cjs/PopupSelect/components.js
/* components.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultComponents = exports.MenuDialog = exports.DummyControl = void 0;
require("./components.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _searchEditorSearch = _interopRequireDefault(require("@atlaskit/icon/core/migration/search--editor-search"));
var _reactSelect = require("@atlaskit/react-select");
var _colors = require("@atlaskit/theme/colors");
var _constants = require("@atlaskit/theme/constants");
var _visuallyHidden = _interopRequireDefault(require("@atlaskit/visually-hidden"));
var _excluded = ["innerRef", "innerProps"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
// ==============================
// Styled Components
// ==============================

var menuDialogStyles = null;

/**
 * __Menu dialog__
 * Wrapper for PopupSelect component.
 */
var MenuDialog = exports.MenuDialog = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var children = _ref.children,
    id = _ref.id,
    style = _ref.style,
    testId = _ref.testId;
  return /*#__PURE__*/React.createElement("div", {
    ref: ref,
    id: id,
    "data-testid": testId && "".concat(testId, "--menu"),
    className: (0, _runtime.ax)(["_2rkoiti9 _1pby16oo _bfhk1bhr _16qsu838"]),
    style: _objectSpread(_objectSpread({}, style), {}, {
      "--_z6sjud": (0, _runtime.ix)("var(--ds-shadow-overlay, ".concat("0 0 0 1px ".concat(_colors.N40A, ", 0 4px 11px ").concat(_colors.N40A), ")"))
    })
  }, children);
});

// ==============================
// Custom Components
// ==============================

var dropdownStyles = null;
var DropdownIndicator = function DropdownIndicator() {
  return /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)(["_1bsbzwfg _12l2v77o _y3gn1h6o"])
  }, /*#__PURE__*/React.createElement(_searchEditorSearch.default, {
    color: "currentColor",
    label: ""
  }));
};
var controlStyles = null;
var Control = function Control(_ref2) {
  var innerRef = _ref2.innerRef,
    innerProps = _ref2.innerProps,
    props = (0, _objectWithoutProperties2.default)(_ref2, _excluded);
  return /*#__PURE__*/React.createElement("div", {
    ref: innerRef,
    className: (0, _runtime.ax)(["_85i51b66 _1q51u2gc _y4tiu2gc _bozgu2gc"])
  }, /*#__PURE__*/React.createElement(_reactSelect.components.Control, (0, _extends2.default)({}, props, {
    innerProps: innerProps
  })));
};

/**
 * __Dummy control__
 * Overrides the default DummyControl component in Select.
 */
var DummyControl = exports.DummyControl = function DummyControl(props) {
  return /*#__PURE__*/React.createElement(_visuallyHidden.default, null, /*#__PURE__*/React.createElement(_reactSelect.components.Control, props));
};

// NOTE `props` intentionally omitted from `Fragment`
var Menu = function Menu(_ref3) {
  var children = _ref3.children,
    innerProps = _ref3.innerProps;
  return /*#__PURE__*/React.createElement("div", innerProps, children);
};

// eslint-disable-next-line @repo/internal/react/no-unsafe-spread-props
var ClearIndicator = function ClearIndicator(props) {
  return /*#__PURE__*/React.createElement(_reactSelect.components.ClearIndicator, props);
};
var MultiValueRemove = function MultiValueRemove(props) {
  return (
    /*#__PURE__*/
    // eslint-disable-next-line @repo/internal/react/no-unsafe-spread-props
    React.createElement(_reactSelect.components.MultiValueRemove, props)
  );
};
var defaultComponents = exports.defaultComponents = {
  Control: Control,
  DropdownIndicator: DropdownIndicator,
  Menu: Menu,
  ClearIndicator: ClearIndicator,
  MultiValueRemove: MultiValueRemove
};
