// @atlaskit/link-datasource/dist/cjs/ui/jira-issues-modal/jira-search-container/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.JiraSearchContainer = exports.DEFAULT_JQL_QUERY = exports.ALLOWED_ORDER_BY_KEYS = void 0;
require("./index.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _reactIntlNext = require("react-intl-next");
var _useDebounce = require("use-debounce");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _compiled = require("@atlaskit/primitives/compiled");
var _analytics = require("../../../analytics");
var _basicSearchInput = require("../../common/modal/basic-search-input");
var _messages = require("../../common/modal/basic-search-input/messages");
var _modeSwitcher = require("../../common/modal/mode-switcher");
var _constants = require("../../common/modal/popup-select/constants");
var _basicFilters = require("../basic-filters");
var _useHydrateJqlQuery2 = require("../basic-filters/hooks/useHydrateJqlQuery");
var _isQueryTooComplex = require("../basic-filters/utils/isQueryTooComplex");
var _jqlEditor = require("../jql-editor");
var _buildJQL = require("./buildJQL");
var _messages2 = require("./messages");
var _excluded = ["basicInputTextValue"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var styles = {
  basicSearchInputBoxStyles: "_1bsb1osq",
  basicSearchInputContainerStyles: "_16jlkb7n",
  inputContainerStylesOld: "_4cvr1q9y _1e0c1txw _1tket9kd",
  inputContainerStyles: "_4cvr1y6m _1e0c1txw _1tket9kd",
  modeSwitcherContainerStyles: "_19pk1b66"
};
var DEFAULT_JQL_QUERY = exports.DEFAULT_JQL_QUERY = 'ORDER BY created DESC';
var ALLOWED_ORDER_BY_KEYS = exports.ALLOWED_ORDER_BY_KEYS = ['key', 'summary', 'assignee', 'status', 'created'];
var JiraSearchContainer = exports.JiraSearchContainer = function JiraSearchContainer(props) {
  var isSearching = props.isSearching,
    parameters = props.parameters,
    onSearch = props.onSearch,
    onSearchMethodChangeCallback = props.onSearchMethodChange,
    initialSearchMethod = props.initialSearchMethod,
    setSearchBarJql = props.setSearchBarJql,
    _props$searchBarJql = props.searchBarJql,
    searchBarJql = _props$searchBarJql === void 0 ? DEFAULT_JQL_QUERY : _props$searchBarJql,
    site = props.site;
  var _ref = parameters || {},
    currentCloudId = _ref.cloudId;
  var _useIntl = (0, _reactIntlNext.useIntl)(),
    formatMessage = _useIntl.formatMessage;
  var _useDatasourceAnalyti = (0, _analytics.useDatasourceAnalyticsEvents)(),
    fireEvent = _useDatasourceAnalyti.fireEvent;
  var _useState = (0, _react.useState)(''),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    basicSearchTerm = _useState2[0],
    setBasicSearchTerm = _useState2[1];
  var _useState3 = (0, _react.useState)(initialSearchMethod),
    _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
    currentSearchMethod = _useState4[0],
    setCurrentSearchMethod = _useState4[1];
  var _useState5 = (0, _react.useState)(currentCloudId),
    _useState6 = (0, _slicedToArray2.default)(_useState5, 2),
    cloudId = _useState6[0],
    setCloudId = _useState6[1];
  var _useState7 = (0, _react.useState)(false),
    _useState8 = (0, _slicedToArray2.default)(_useState7, 2),
    isComplexQuery = _useState8[0],
    setIsComplexQuery = _useState8[1];
  var _useState9 = (0, _react.useState)(),
    _useState10 = (0, _slicedToArray2.default)(_useState9, 2),
    orderKey = _useState10[0],
    setOrderKey = _useState10[1];
  var _useState11 = (0, _react.useState)(),
    _useState12 = (0, _slicedToArray2.default)(_useState11, 2),
    orderDirection = _useState12[0],
    setOrderDirection = _useState12[1];
  var _useState13 = (0, _react.useState)({}),
    _useState14 = (0, _slicedToArray2.default)(_useState13, 2),
    filterSelections = _useState14[0],
    setFilterSelections = _useState14[1];
  var modeSwitcherOptionsMap = (0, _react.useMemo)(function () {
    return {
      jql: {
        label: 'JQL',
        value: 'jql'
      },
      basic: {
        label: formatMessage(_messages2.modeSwitcherMessages.basicTextSearchLabel),
        value: 'basic',
        disabled: isComplexQuery,
        tooltipText: isComplexQuery ? formatMessage(_messages2.modeSwitcherMessages.basicModeSwitchDisabledTooltipText) : ''
      }
    };
  }, [formatMessage, isComplexQuery]);
  var modeSwitcherOptions = (0, _react.useMemo)(function () {
    return [modeSwitcherOptionsMap.basic, modeSwitcherOptionsMap.jql];
  }, [modeSwitcherOptionsMap]);
  var _useHydrateJqlQuery = (0, _useHydrateJqlQuery2.useHydrateJqlQuery)(cloudId || '', searchBarJql),
    hydratedOptions = _useHydrateJqlQuery.hydratedOptions,
    fetchHydratedJqlOptions = _useHydrateJqlQuery.fetchHydratedJqlOptions,
    basicFilterHydrationStatus = _useHydrateJqlQuery.status;
  var onSearchMethodChange = (0, _react.useCallback)(function (searchMethod) {
    onSearchMethodChangeCallback(searchMethod);
    setCurrentSearchMethod(searchMethod);
  }, [onSearchMethodChangeCallback]);
  var handleBasicSearchChange = (0, _react.useCallback)(function (e) {
    var rawSearch = e.currentTarget.value;
    setBasicSearchTerm(rawSearch);
    setSearchBarJql((0, _buildJQL.buildJQL)({
      rawSearch: rawSearch,
      filterValues: filterSelections,
      orderDirection: orderDirection,
      orderKey: orderKey
    }));
  }, [setSearchBarJql, filterSelections, orderDirection, orderKey]);
  var onQueryChange = (0, _react.useCallback)(function (query) {
    var _query$split$map$filt, _fragments$at, _fragments$at2, _fragments$at3;
    // determine if order keys have been set so they can be saved and persisted when changes occur in basic search
    var fragments = (_query$split$map$filt = query === null || query === void 0 ? void 0 : query.split(/(^| )(order by)( |$)/i).map(function (item) {
      return item.trim();
    }).filter(Boolean)) !== null && _query$split$map$filt !== void 0 ? _query$split$map$filt : [];
    var hasOrder = ((_fragments$at = fragments.at(-2)) === null || _fragments$at === void 0 ? void 0 : _fragments$at.toLowerCase()) === 'order by';
    var key = hasOrder ? (_fragments$at2 = fragments.at(-1)) === null || _fragments$at2 === void 0 ? void 0 : _fragments$at2.split(' ').at(-2) : undefined;
    var order = hasOrder ? (_fragments$at3 = fragments.at(-1)) === null || _fragments$at3 === void 0 ? void 0 : _fragments$at3.split(' ').at(-1) : undefined;

    // TODO: confirm if these are the only order keys we want to preserve - existing whiteboard logic
    if (key && ALLOWED_ORDER_BY_KEYS.includes(key)) {
      setOrderKey(key);
      setOrderDirection(order);
    }
    setSearchBarJql(query);
  }, [setSearchBarJql]);
  var handleSearch = (0, _react.useCallback)(function () {
    var isCurrentQueryComplex = (0, _isQueryTooComplex.isQueryTooComplex)(searchBarJql);
    onSearch({
      jql: searchBarJql
    }, {
      searchMethod: currentSearchMethod,
      basicFilterSelections: filterSelections,
      isQueryComplex: isCurrentQueryComplex
    });
    if (currentSearchMethod === 'jql') {
      fireEvent('ui.jqlEditor.searched', {
        isQueryComplex: isCurrentQueryComplex
      });
      setIsComplexQuery(isCurrentQueryComplex);
      if (!isCurrentQueryComplex) {
        fetchHydratedJqlOptions();
      }
    }
  }, [currentSearchMethod, fetchHydratedJqlOptions, filterSelections, fireEvent, searchBarJql, onSearch]);
  var _useDebouncedCallback = (0, _useDebounce.useDebouncedCallback)(function (filterValues) {
      var jqlWithFilterValues = (0, _buildJQL.buildJQL)({
        rawSearch: basicSearchTerm,
        filterValues: filterValues,
        orderDirection: orderDirection,
        orderKey: orderKey
      });
      setSearchBarJql(jqlWithFilterValues);
      var isCurrentQueryComplex = (0, _isQueryTooComplex.isQueryTooComplex)(jqlWithFilterValues);
      onSearch({
        jql: jqlWithFilterValues
      }, {
        searchMethod: currentSearchMethod,
        basicFilterSelections: filterSelections,
        isQueryComplex: isCurrentQueryComplex
      });
    }, _constants.FILTER_SELECTION_DEBOUNCE_MS),
    _useDebouncedCallback2 = (0, _slicedToArray2.default)(_useDebouncedCallback, 1),
    debouncedBasicFilterSelectionChange = _useDebouncedCallback2[0];
  var handleBasicFilterSelectionChange = (0, _react.useCallback)(function (filterType, options) {
    var updatedSelection = _objectSpread(_objectSpread({}, filterSelections), {}, (0, _defineProperty2.default)({}, filterType, options));
    setFilterSelections(updatedSelection);
    debouncedBasicFilterSelectionChange(updatedSelection);
  }, [debouncedBasicFilterSelectionChange, filterSelections]);
  (0, _react.useEffect)(function () {
    var isCurrentQueryComplex = (0, _isQueryTooComplex.isQueryTooComplex)(searchBarJql);
    setIsComplexQuery(isCurrentQueryComplex);
    if (!isCurrentQueryComplex && searchBarJql !== DEFAULT_JQL_QUERY) {
      fetchHydratedJqlOptions();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  (0, _react.useEffect)(function () {
    if (basicFilterHydrationStatus === 'resolved') {
      var basicInputTextValue = hydratedOptions.basicInputTextValue,
        hydratedFilterOptions = (0, _objectWithoutProperties2.default)(hydratedOptions, _excluded);
      setFilterSelections(hydratedFilterOptions);
      if (basicInputTextValue) {
        setBasicSearchTerm(basicInputTextValue);
      }
    }
  }, [hydratedOptions, basicFilterHydrationStatus]);
  (0, _react.useEffect)(function () {
    if (currentCloudId !== cloudId) {
      setBasicSearchTerm('');
      setSearchBarJql(DEFAULT_JQL_QUERY);
      setIsComplexQuery(false);
      setOrderKey(undefined);
      setOrderDirection(undefined);
      setFilterSelections({});
      setCloudId(currentCloudId);
    }
  }, [currentCloudId, cloudId, setSearchBarJql]);
  if ((0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-sllv')) {
    return /*#__PURE__*/_react.default.createElement("div", {
      "data-testid": "jira-search-container",
      className: (0, _runtime.ax)([styles.inputContainerStyles])
    }, currentSearchMethod === 'basic' && /*#__PURE__*/_react.default.createElement(_compiled.Box, {
      xcss: styles.basicSearchInputBoxStyles
    }, /*#__PURE__*/_react.default.createElement(_compiled.Flex, {
      alignItems: "center",
      xcss: styles.basicSearchInputContainerStyles
    }, /*#__PURE__*/_react.default.createElement(_basicSearchInput.BasicSearchInput, {
      isSearching: isSearching,
      onChange: handleBasicSearchChange,
      onSearch: handleSearch,
      searchTerm: basicSearchTerm,
      placeholder: _messages.basicSearchInputMessages.basicTextSearchLabel,
      ariaLabel: _messages.basicSearchInputMessages.basicTextSearchLabel,
      testId: "jira-datasource-modal",
      fullWidth: false
    }), /*#__PURE__*/_react.default.createElement(_basicFilters.BasicFilters, {
      jql: searchBarJql,
      site: site,
      onChange: handleBasicFilterSelectionChange,
      selections: filterSelections,
      isJQLHydrating: basicFilterHydrationStatus === 'loading'
    })), currentSearchMethod === 'basic' && /*#__PURE__*/_react.default.createElement(_compiled.Text, {
      size: "small",
      color: "color.text.subtlest",
      testId: "jira-search-placeholder"
    }, formatMessage(_messages.basicSearchInputMessages.basicTextSearchLabel))), currentSearchMethod === 'jql' && /*#__PURE__*/_react.default.createElement(_jqlEditor.JiraJQLEditor, {
      cloudId: cloudId || '',
      isSearching: isSearching,
      onChange: onQueryChange,
      onSearch: handleSearch,
      query: searchBarJql
    }), /*#__PURE__*/_react.default.createElement(_compiled.Box, {
      xcss: styles.modeSwitcherContainerStyles
    }, /*#__PURE__*/_react.default.createElement(_modeSwitcher.ModeSwitcher, {
      onOptionValueChange: onSearchMethodChange,
      selectedOptionValue: currentSearchMethod,
      options: modeSwitcherOptions
    })));
  }
  return /*#__PURE__*/_react.default.createElement("div", {
    "data-testid": "jira-search-container",
    className: (0, _runtime.ax)([styles.inputContainerStylesOld])
  }, currentSearchMethod === 'basic' && /*#__PURE__*/_react.default.createElement(_compiled.Flex, {
    alignItems: "center",
    xcss: styles.basicSearchInputContainerStyles
  }, /*#__PURE__*/_react.default.createElement(_basicSearchInput.BasicSearchInput, {
    isSearching: isSearching,
    onChange: handleBasicSearchChange,
    onSearch: handleSearch,
    searchTerm: basicSearchTerm,
    placeholder: _messages.basicSearchInputMessages.basicTextSearchLabel,
    ariaLabel: _messages.basicSearchInputMessages.basicTextSearchLabel,
    testId: "jira-datasource-modal",
    fullWidth: false
  }), /*#__PURE__*/_react.default.createElement(_basicFilters.BasicFilters, {
    jql: searchBarJql,
    site: site,
    onChange: handleBasicFilterSelectionChange,
    selections: filterSelections,
    isJQLHydrating: basicFilterHydrationStatus === 'loading'
  })), currentSearchMethod === 'jql' && /*#__PURE__*/_react.default.createElement(_jqlEditor.JiraJQLEditor, {
    cloudId: cloudId || '',
    isSearching: isSearching,
    onChange: onQueryChange,
    onSearch: handleSearch,
    query: searchBarJql
  }), /*#__PURE__*/_react.default.createElement(_modeSwitcher.ModeSwitcher, {
    onOptionValueChange: onSearchMethodChange,
    selectedOptionValue: currentSearchMethod,
    options: modeSwitcherOptions
  }));
};
