// @atlaskit/flag/dist/cjs/theme.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.flagTextColorToken = exports.flagTextColor = exports.flagIconGlyph = exports.flagIconColor = exports.flagBackgroundColor = exports.actionTextColor = exports.actionBackgroundColor = void 0;
var _react = _interopRequireDefault(require("react"));
var _error = _interopRequireDefault(require("@atlaskit/icon/core/migration/error"));
var _informationInfo = _interopRequireDefault(require("@atlaskit/icon/core/migration/information--info"));
var _successCheckCircle = _interopRequireDefault(require("@atlaskit/icon/core/migration/success--check-circle"));
var _warning = _interopRequireDefault(require("@atlaskit/icon/core/migration/warning"));
var _colors = require("@atlaskit/theme/colors");
/* eslint-disable @atlaskit/design-system/ensure-design-token-usage */

var flagBackgroundColor = exports.flagBackgroundColor = {
  error: 'color.background.danger.bold',
  info: 'color.background.neutral.bold',
  normal: 'elevation.surface.overlay',
  success: 'color.background.success.bold',
  warning: 'color.background.warning.bold'
};
var flagIconColor = exports.flagIconColor = {
  error: "var(--ds-icon-inverse, #FFFFFF)",
  info: "var(--ds-icon-inverse, #FFFFFF)",
  normal: "var(--ds-icon-subtle, #626F86)",
  success: "var(--ds-icon-inverse, #FFFFFF)",
  warning: "var(--ds-icon-warning-inverse, #172B4D)"
};
var flagIconGlyph = exports.flagIconGlyph = {
  error: /*#__PURE__*/_react.default.createElement(_error.default, {
    label: "",
    LEGACY_primaryColor: "var(--ds-icon-inverse, ".concat(_colors.N0, ")"),
    LEGACY_secondaryColor: "var(--ds-background-danger-bold, #C9372C)"
  }),
  info: /*#__PURE__*/_react.default.createElement(_informationInfo.default, {
    label: "",
    LEGACY_primaryColor: "var(--ds-icon-inverse, ".concat(_colors.N0, ")"),
    LEGACY_secondaryColor: "var(--ds-background-neutral-bold, #44546F)"
  }),
  normal: /*#__PURE__*/_react.default.createElement(_informationInfo.default, {
    label: "",
    LEGACY_primaryColor: "var(--ds-icon-subtle, ".concat(_colors.N500, ")"),
    LEGACY_secondaryColor: "var(--ds-surface-overlay, #FFFFFF)"
  }),
  success: /*#__PURE__*/_react.default.createElement(_successCheckCircle.default, {
    label: "",
    LEGACY_primaryColor: "var(--ds-icon-inverse, ".concat(_colors.N0, ")"),
    LEGACY_secondaryColor: "var(--ds-background-success-bold, #1F845A)"
  }),
  warning: /*#__PURE__*/_react.default.createElement(_warning.default, {
    label: "",
    LEGACY_primaryColor: "var(--ds-icon-warning-inverse, ".concat(_colors.N700, ")"),
    LEGACY_secondaryColor: "var(--ds-background-warning-bold, #F5CD47)"
  })
};

/* eslint-disable @atlaskit/design-system/ensure-design-token-usage */
var flagTextColor = exports.flagTextColor = {
  error: 'color.text.inverse',
  info: 'color.text.inverse',
  normal: 'color.text',
  success: 'color.text.inverse',
  warning: 'color.text.warning.inverse'
};
var flagTextColorToken = exports.flagTextColorToken = {
  error: "var(--ds-text-inverse, ".concat(_colors.N0, ")"),
  info: "var(--ds-text-inverse, ".concat(_colors.N0, ")"),
  normal: "var(--ds-text, ".concat(_colors.N800, ")"),
  success: "var(--ds-text-inverse, ".concat(_colors.N0, ")"),
  warning: "var(--ds-text-warning-inverse, ".concat(_colors.N700, ")")
};
// TODO: DSP-2519 Interaction tokens should be used for hovered and pressed states
// https://product-fabric.atlassian.net/browse/DSP-2519
var actionBackgroundColor = exports.actionBackgroundColor = {
  success: {
    default: "var(--ds-background-inverse-subtle, ".concat(_colors.N30A, ")"),
    active: "var(--ds-background-inverse-subtle-pressed, ".concat(_colors.N30A, ")"),
    pressed: "var(--ds-background-inverse-subtle-hovered, ".concat(_colors.N30A, ")")
  },
  info: {
    default: "var(--ds-background-inverse-subtle, ".concat(_colors.N30A, ")"),
    active: "var(--ds-background-inverse-subtle-pressed, ".concat(_colors.N30A, ")"),
    pressed: "var(--ds-background-inverse-subtle-hovered, ".concat(_colors.N30A, ")")
  },
  error: {
    default: "var(--ds-background-inverse-subtle, ".concat(_colors.N30A, ")"),
    active: "var(--ds-background-inverse-subtle-pressed, ".concat(_colors.N30A, ")"),
    pressed: "var(--ds-background-inverse-subtle-hovered, ".concat(_colors.N30A, ")")
  },
  warning: {
    default: "var(--ds-background-inverse-subtle, ".concat(_colors.N30A, ")"),
    active: "var(--ds-background-inverse-subtle-pressed, ".concat(_colors.N30A, ")"),
    pressed: "var(--ds-background-inverse-subtle-hovered, ".concat(_colors.N30A, ")")
  },
  normal: {
    default: 'none',
    active: 'none',
    pressed: 'none'
  }
};
var actionTextColor = exports.actionTextColor = {
  success: "var(--ds-text-inverse, ".concat(_colors.N0, ")"),
  info: "var(--ds-text-inverse, ".concat(_colors.N0, ")"),
  error: "var(--ds-text-inverse, ".concat(_colors.N0, ")"),
  warning: "var(--ds-text-warning-inverse, ".concat(_colors.N700, ")"),
  normal: "var(--ds-link, ".concat(_colors.B400, ")")
};
