// @atlaskit/smart-card/dist/es2019/state/analytics/ufoExperiences.js
import { ConcurrentExperience, ExperiencePerformanceTypes, ExperienceTypes } from '@atlaskit/ufo';
const inlineExperience = {
  platform: {
    component: 'smart-links'
  },
  type: ExperienceTypes.Experience,
  performanceType: ExperiencePerformanceTypes.InlineResult
};
const renderExperience = {
  platform: {
    component: 'smart-links'
  },
  type: ExperienceTypes.Load,
  performanceType: ExperiencePerformanceTypes.PageSegmentLoad
};
const aiExperience = {
  platform: {
    component: 'smart-links'
  },
  type: ExperienceTypes.Experience,
  performanceType: ExperiencePerformanceTypes.InlineResult
};
const ufoExperiences = {
  'smart-link-rendered': new ConcurrentExperience('smart-link-rendered', renderExperience),
  'smart-link-authenticated': new ConcurrentExperience('smart-link-authenticated', inlineExperience),
  'smart-link-action-invocation': new ConcurrentExperience('smart-link-action-invocation', inlineExperience),
  'smart-link-ai-summary': new ConcurrentExperience('smart-link-ai-summary', aiExperience)
};
export const startUfoExperience = (experienceName, id, properties) => {
  const experience = ufoExperiences[experienceName].getInstance(id);
  experience.start();
  if (properties) {
    experience.addMetadata(properties);
  }
};
export const succeedUfoExperience = (experienceName, id, properties) => {
  const experience = ufoExperiences[experienceName].getInstance(id);
  experience.success({
    metadata: properties
  });
};
export const failUfoExperience = (experienceName, id, properties) => {
  const experience = ufoExperiences[experienceName].getInstance(id);
  experience.failure({
    metadata: properties
  });
};
export const addMetadataToExperience = (experienceName, id, properties) => {
  ufoExperiences[experienceName].getInstance(id).addMetadata(properties);
};
