// @atlaskit/link-picker/dist/cjs/ui/link-picker/search-results/link-search-list/list-item/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.testIds = exports.LinkSearchListItem = void 0;
require("./index.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _reactIntlNext = require("react-intl-next");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _compiled = require("@atlaskit/primitives/compiled");
var _colors = require("@atlaskit/theme/colors");
var _transformTimeStamp = require("../../../transformTimeStamp");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/* eslint-disable-next-line @atlassian/tangerine/import/no-parent-imports */
var testIds = exports.testIds = {
  searchResultItem: 'link-search-list-item',
  searchResultIcon: 'link-search-list-item-icon'
};
var isSVG = function isSVG(icon) {
  return icon.startsWith('<svg') && icon.endsWith('</svg>');
};
var base64SVG = function base64SVG(icon) {
  return "data:image/svg+xml;base64,".concat(Buffer.from(icon).toString('base64'));
};
var itemIconStyles = null;
var listItemContextStyles = null;
var listItemContainerStyles = null;
var listItemContainerInnerStyles = null;
var listItemContainerInnerStylesA11yRefresh = null;
var listItemContainerInnerStylesSeparatorLinkPickerRefresh = null;
var itemNameStyles = null;
var imgStyles = null;
var listItemBaseStyles = null;
var ListItemIcon = function ListItemIcon(props) {
  var item = props.item,
    intl = props.intl;
  var icon = item.icon,
    iconAlt = item.iconAlt;
  if (!icon) {
    return null;
  }
  var alt = typeof iconAlt === 'string' ? iconAlt : intl.formatMessage(iconAlt);
  if (typeof icon !== 'string') {
    var Glyph = icon;
    return /*#__PURE__*/React.createElement("span", {
      className: (0, _runtime.ax)(["_1ul9pxbi _19pk1b66 _2hwxutpp"])
    }, /*#__PURE__*/React.createElement(Glyph, {
      alt: alt,
      "data-testid": testIds.searchResultIcon
    }));
  }
  return /*#__PURE__*/React.createElement("span", {
    className: (0, _runtime.ax)(["_1ul9pxbi _19pk1b66 _2hwxutpp"])
  }, /*#__PURE__*/React.createElement("img", {
    "data-testid": testIds.searchResultIcon,
    src: isSVG(icon) ? base64SVG(icon) : icon,
    alt: alt,
    className: (0, _runtime.ax)(["_p12fpxbi"])
  }));
};
var ListItemSubtitle = function ListItemSubtitle(_ref) {
  var _ref$items = (0, _slicedToArray2.default)(_ref.items, 2),
    firstItem = _ref$items[0],
    secondItem = _ref$items[1];
  return /*#__PURE__*/React.createElement("div", {
    "data-testid": "".concat(testIds.searchResultItem, "-subtitle"),
    className: (0, _runtime.ax)(["_11c81o8v _syazct01 _1e0c1txw"]),
    style: {
      "--_khvm95": (0, _runtime.ix)("var(--ds-text, ".concat(_colors.N300, ")"))
    }
  }, /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)(["_1reo15vq _18m915vq _1bto1l2s"])
  }, /*#__PURE__*/React.createElement("span", {
    className: (0, _runtime.ax)([(0, _platformFeatureFlags.fg)('bandicoots-a11y-link-picker-styling') ? "_syaz121t _o5721q9c" : "_syazpsu8 _o5721q9c"]),
    style: {
      "--_1pseksa": (0, _runtime.ix)("var(--ds-text-subtle, ".concat(_colors.N400, ")")),
      "--_vrm7dk": (0, _runtime.ix)("var(--ds-text-subtlest, ".concat(_colors.N200, ")"))
    }
  }, firstItem)), secondItem && /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)([(0, _platformFeatureFlags.fg)('bandicoots-a11y-link-picker-styling') ? "_syaz121t _o5721q9c" : "_syazpsu8 _o5721q9c"]),
    style: {
      "--_1pseksa": (0, _runtime.ix)("var(--ds-text-subtle, ".concat(_colors.N400, ")")),
      "--_vrm7dk": (0, _runtime.ix)("var(--ds-text-subtlest, ".concat(_colors.N200, ")"))
    }
  }, (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-link-picker') ? /*#__PURE__*/React.createElement("span", {
    className: (0, _runtime.ax)(["_syaz1svm"]),
    style: {
      "--_1trgtlr": (0, _runtime.ix)("var(--ds-border, ".concat(_colors.N40A, ")"))
    }
  }, /*#__PURE__*/React.createElement(_react.Fragment, null, "\xA0 \u2022\xA0 ")) : /*#__PURE__*/React.createElement(_react.Fragment, null, "\xA0 \u2022\xA0 "), /*#__PURE__*/React.createElement(_react.Fragment, null, secondItem)));
};
var getDefaultSubtitleItems = function getDefaultSubtitleItems(item, intl) {
  var container = item.container;
  var date = (0, _transformTimeStamp.transformTimeStamp)(intl, item.lastViewedDate, item.lastUpdatedDate);
  if (container) {
    if (date) {
      return [container, date];
    }
    return [container];
  }
  if (date) {
    return [date];
  }
  return undefined;
};
var listItemActive = null;
var listItemSelected = null;
var listItemFocusStyles = null;
var LinkSearchListItem = exports.LinkSearchListItem = /*#__PURE__*/(0, _react.forwardRef)(function (_ref2, ref) {
  var item = _ref2.item,
    selected = _ref2.selected,
    id = _ref2.id,
    role = _ref2.role,
    onSelect = _ref2.onSelect,
    tabIndex = _ref2.tabIndex,
    onKeyDown = _ref2.onKeyDown,
    onFocus = _ref2.onFocus,
    _ref2$nameMaxLines = _ref2.nameMaxLines,
    nameMaxLines = _ref2$nameMaxLines === void 0 ? 1 : _ref2$nameMaxLines;
  var intl = (0, _reactIntlNext.useIntl)();
  var handleSelect = function handleSelect() {
    return onSelect(item.objectId);
  };
  var subtitleItems = item.subtitleItems || getDefaultSubtitleItems(item, intl);
  return /*#__PURE__*/React.createElement("div", {
    role: role,
    id: id,
    "aria-selected": selected,
    "data-testid": testIds.searchResultItem,
    onKeyDown: onKeyDown,
    onClick: handleSelect,
    onFocus: onFocus,
    tabIndex: tabIndex,
    ref: ref,
    className: (0, _runtime.ax)(["_19pkidpf _2hwxidpf _otyridpf _18u0idpf _1e0c1txw _ca0qu2gc _n3tdu2gc _19bvrvlv _u5f3kq6n _80omtlke", !selected && "_irr38iyy _1dbz1f4r", selected && "_bfhk1kxl _16qs1f4r", "_nt751r31 _49pcglyw _1hvw1o36 _1a3b1r31 _4fprglyw _5goinqa1 _7ehif7x9"]),
    style: {
      "--_u7ouzj": (0, _runtime.ix)("clamp( ".concat("var(--ds-space-100, 8px)", ", var(--link-picker-padding-left), 100% )")),
      "--_1kucder": (0, _runtime.ix)("clamp( ".concat("var(--ds-space-100, 8px)", ", var(--link-picker-padding-right), 100% )")),
      "--_mmrjro": (0, _runtime.ix)("var(--ds-background-neutral-subtle-hovered, ".concat(_colors.N20, ")")),
      "--_1mo126v": (0, _runtime.ix)("inset 2px 0px 0px ".concat("var(--ds-border-selected, ".concat(_colors.B400, ")"))),
      "--_rryx0a": (0, _runtime.ix)("var(--ds-background-selected, ".concat(_colors.B50, ")")),
      "--_ttguoj": (0, _runtime.ix)("0 0 0 2px ".concat("var(--ds-border-focused, ".concat(_colors.B100, ")"), " inset"))
    }
  }, /*#__PURE__*/React.createElement(ListItemIcon, {
    item: item,
    intl: intl
  }), /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)(["_1reo15vq _18m915vq _ae4v1h6o _1bsb1osq"])
  }, /*#__PURE__*/React.createElement(_compiled.Text, {
    maxLines: nameMaxLines
  }, /*#__PURE__*/React.createElement("span", {
    "data-testid": "".concat(testIds.searchResultItem, "-title"),
    title: item.name
  }, item.name)), subtitleItems && /*#__PURE__*/React.createElement(ListItemSubtitle, {
    items: subtitleItems
  })));
});
