// @atlaskit/link-datasource/dist/cjs/analytics/ufoExperiences/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.succeedUfoExperience = exports.startUfoExperience = exports.failUfoExperience = exports.addMetadataToExperience = void 0;
var _ufo = require("@atlaskit/ufo");
var datasourcePageSegmentLoadExperienceConfig = {
  platform: {
    component: 'datasource'
  },
  type: _ufo.ExperienceTypes.Load,
  performanceType: _ufo.ExperiencePerformanceTypes.PageSegmentLoad
};
var columnPickerCustomExperienceConfig = {
  platform: {
    component: 'datasource'
  },
  type: _ufo.ExperienceTypes.Operation,
  performanceType: _ufo.ExperiencePerformanceTypes.Custom
};
var inlineEditCustomExperienceConfig = {
  platform: {
    component: 'datasource'
  },
  type: _ufo.ExperienceTypes.Load,
  performanceType: _ufo.ExperiencePerformanceTypes.InlineResult
};
var ufoExperiences = {
  'datasource-rendered': new _ufo.ConcurrentExperience('datasource-rendered', datasourcePageSegmentLoadExperienceConfig),
  'column-picker-rendered': new _ufo.ConcurrentExperience('column-picker-rendered', columnPickerCustomExperienceConfig),
  'inline-edit-rendered': new _ufo.ConcurrentExperience('inline-edit-rendered', inlineEditCustomExperienceConfig)
};
var startUfoExperience = exports.startUfoExperience = function startUfoExperience(_ref, id) {
  var name = _ref.name,
    metadata = _ref.metadata;
  var experienceInstance = ufoExperiences[name].getInstance(id);
  experienceInstance.start();
  if (metadata) {
    experienceInstance.addMetadata(metadata);
  }
};
var succeedUfoExperience = exports.succeedUfoExperience = function succeedUfoExperience(_ref2, id) {
  var name = _ref2.name,
    metadata = _ref2.metadata;
  var experienceInstance = ufoExperiences[name].getInstance(id);
  experienceInstance.success({
    metadata: metadata
  });
};
var failUfoExperience = exports.failUfoExperience = function failUfoExperience(_ref3, id) {
  var name = _ref3.name,
    metadata = _ref3.metadata;
  var experienceInstance = ufoExperiences[name].getInstance(id);
  experienceInstance.failure({
    metadata: metadata
  });
};
var addMetadataToExperience = exports.addMetadataToExperience = function addMetadataToExperience(_ref4, id) {
  var name = _ref4.name,
    metadata = _ref4.metadata;
  ufoExperiences[name].getInstance(id).addMetadata(metadata);
};
