// @atlaskit/media-viewer/dist/cjs/errorMessage.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getErrorMessageFromError = exports.errorReasonToMessages = exports.default = exports.ErrorMessage = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _react = _interopRequireDefault(require("react"));
var _reactIntlNext = require("react-intl-next");
var _mediaUi = require("@atlaskit/media-ui");
var _analyticsNext = require("@atlaskit/analytics-next");
var _styleWrappers = require("./styleWrappers");
var _errorImages = require("./error-images");
var _analytics = require("./analytics");
var _errors = require("./errors");
var _loadFailed = require("./analytics/events/operational/loadFailed");
var _previewUnsupported = require("./analytics/events/operational/previewUnsupported");
var _ufoExperiences = require("./analytics/ufoExperiences");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var errorLoadingFileImage = function errorLoadingFileImage(formatMessage) {
  return /*#__PURE__*/_react.default.createElement(_styleWrappers.ErrorImage, {
    src: _errorImages.errorLoadingFile,
    alt: formatMessage(_mediaUi.messages.error_loading_file)
  });
};
var errorReasonToMessages = exports.errorReasonToMessages = [['serverRateLimited', _mediaUi.messages.might_be_a_hiccup], ['invalidFileId', _mediaUi.messages.item_not_found_in_list], ['itemviewer-file-failed-processing-status', _mediaUi.messages.image_format_invalid_error], ['archiveviewer-read-binary', _mediaUi.messages.zip_entry_load_fail], ['archiveviewer-create-url', _mediaUi.messages.zip_entry_load_fail], ['archiveviewer-missing-name-src', _mediaUi.messages.zip_entry_load_fail], ['archiveviewer-encrypted-entry', _mediaUi.messages.couldnt_generate_encrypted_entry_preview], ['archiveviewer-codeviewer-file-size-exceeds', _mediaUi.messages.couldnt_load_file], ['codeviewer-file-size-exceeds', _mediaUi.messages.couldnt_load_file]];
var getErrorMessageFromError = exports.getErrorMessageFromError = function getErrorMessageFromError(error) {
  var matchingRow = errorReasonToMessages.find(function (row) {
    return row[0] === (0, _errors.getPrimaryErrorReason)(error) || row[0] === (0, _errors.getSecondaryErrorReason)(error);
  });
  return matchingRow ? matchingRow[1] : undefined;
};
var ErrorMessage = exports.ErrorMessage = /*#__PURE__*/function (_React$Component) {
  function ErrorMessage() {
    (0, _classCallCheck2.default)(this, ErrorMessage);
    return _callSuper(this, ErrorMessage, arguments);
  }
  (0, _inherits2.default)(ErrorMessage, _React$Component);
  return (0, _createClass2.default)(ErrorMessage, [{
    key: "getErrorInfo",
    value: function getErrorInfo() {
      var _this$props = this.props,
        formatMessage = _this$props.intl.formatMessage,
        error = _this$props.error;
      var errorInfo = {
        icon: errorLoadingFileImage(formatMessage),
        messages: [_mediaUi.messages.something_went_wrong, _mediaUi.messages.couldnt_generate_preview]
      };
      var message = getErrorMessageFromError(error);
      if (message) {
        errorInfo.messages.push(message);
      }
      return errorInfo;
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      var props = this.props;
      var supressAnalytics = props.supressAnalytics,
        error = props.error,
        fileState = props.fileState,
        fileId = props.fileId,
        traceContext = props.traceContext,
        createAnalyticsEvent = props.createAnalyticsEvent,
        fileStateFlags = props.fileStateFlags;
      if (supressAnalytics !== true) {
        var payload = ErrorMessage.getEventPayload(error, fileId, fileState, traceContext);
        (0, _analytics.fireAnalytics)(payload, createAnalyticsEvent);
        var rawPayload = _objectSpread(_objectSpread({}, payload === null || payload === void 0 ? void 0 : payload.attributes), {}, {
          fileStateFlags: fileStateFlags
        });
        if (Object.keys(rawPayload).includes('status')) {
          delete rawPayload['status'];
        }
        var failMediaFileUfoExperiencePayload = rawPayload;
        (0, _ufoExperiences.failMediaFileUfoExperience)(failMediaFileUfoExperiencePayload);
      }
    }
  }, {
    key: "render",
    value: function render() {
      var errorInfo = this.getErrorInfo();
      return /*#__PURE__*/_react.default.createElement(_styleWrappers.ErrorMessageWrapper, {
        "data-testid": "media-viewer-error"
      }, /*#__PURE__*/_react.default.createElement("div", null, errorInfo.icon, errorInfo.messages.map(function (formatMessage, i) {
        return (
          /*#__PURE__*/
          // eslint-disable-next-line @atlaskit/design-system/use-primitives-text
          _react.default.createElement("p", {
            key: "p".concat(i)
          }, /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, formatMessage))
        );
      })), this.props.children);
    }
  }], [{
    key: "getEventPayload",
    value: function getEventPayload(error, fileId, fileState, traceContext) {
      if (fileState && (0, _errors.getPrimaryErrorReason)(error) === 'unsupported') {
        // this is not an SLI, its just a useful metric for unsupported
        return (0, _previewUnsupported.createPreviewUnsupportedEvent)(fileState);
      } else {
        return (0, _loadFailed.createLoadFailedEvent)(fileId, error, fileState, traceContext);
      }
    }
  }]);
}(_react.default.Component); // @ts-ignore: [PIT-1685] Fails in post-office due to backwards incompatibility issue with React 18
var ErroMsg = (0, _analyticsNext.withAnalyticsEvents)()((0, _reactIntlNext.injectIntl)(ErrorMessage));
var _default = exports.default = ErroMsg;
