// @atlaskit/media-viewer/dist/cjs/styleWrappers-emotion.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ZoomWrapper = exports.ZoomRightControls = exports.ZoomLevelIndicator = exports.ZoomCenterControls = exports.Video = exports.SpinnerWrapper = exports.SidebarWrapper = exports.RightWrapper = exports.RightHeader = exports.PDFWrapper = exports.MetadataWrapper = exports.MetadataSubText = exports.MetadataIconWrapper = exports.MetadataFileName = exports.MedatadataTextWrapper = exports.ListWrapper = exports.LeftWrapper = exports.LeftHeader = exports.Img = exports.ImageWrapper = exports.HeaderWrapper = exports.Header = exports.HDIconGroupWrapper = exports.FormattedMessageWrapper = exports.ErrorMessageWrapper = exports.ErrorImage = exports.DownloadButtonWrapper = exports.DefaultCoverWrapper = exports.CustomVideoPlayerWrapper = exports.CustomAudioPlayerWrapper = exports.ContentWrapper = exports.CloseButtonWrapper = exports.Blanket = exports.BaselineExtend = exports.AudioPlayer = exports.AudioCover = exports.Audio = exports.ArrowsWrapper = exports.Arrow = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = require("@emotion/react");
var _react2 = require("react");
var _styles = require("./styles");
var _mediaUi = require("@atlaskit/media-ui");
var _reactScrolllock = require("react-scrolllock");
var _useCallbackRef = require("use-callback-ref");
var _modalSpinner = require("./viewers/modalSpinner");
var _styles2 = require("./viewers/archiveSidebar/styles");
var _constants = require("@atlaskit/theme/constants");
var _primitives = require("@atlaskit/primitives");
/* eslint-disable @atlaskit/design-system/use-tokens-typography */
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var SIDEBAR_WIDTH = 416;
var blanketStyles = (0, _react.css)({
  position: 'fixed',
  top: 0,
  left: 0,
  bottom: 0,
  right: 0,
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  backgroundColor: _styles.blanketColor,
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  zIndex: _styles.overlayZindex,
  display: 'flex'
});
var headerWrapperStyles = function headerWrapperStyles(_ref) {
  var isArchiveSideBarVisible = _ref.isArchiveSideBarVisible;
  return (0, _react.css)({
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    height: '98px',
    opacity: 0.85,
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    background: "linear-gradient( to bottom, ".concat(_modalSpinner.headerAndSidebarBackgroundColor, ", rgba(14, 22, 36, 0) ) no-repeat"),
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
    backgroundPosition: isArchiveSideBarVisible ? "".concat(_styles2.ArchiveSideBarWidth, "px 0") : '0',
    // eslint-disable-next-line @atlaskit/design-system/ensure-design-token-usage
    color: '#c7d1db',
    fontWeight: "var(--ds-font-weight-medium, 500)",
    padding: "var(--ds-space-300, 24px)",
    boxSizing: 'border-box',
    pointerEvents: 'none',
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    zIndex: _styles.overlayZindex + 1
  });
};
var listWrapperStyles = (0, _react.css)({
  width: '100%',
  height: '100%',
  position: 'relative',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center'
});
var closeButtonWrapperStyles = (0, _react.css)({
  position: 'absolute',
  top: "var(--ds-space-300, 24px)",
  right: "var(--ds-space-250, 20px)",
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  zIndex: _styles.overlayZindex + 2
});
var contentWrapperStyles = function contentWrapperStyles(_ref2) {
  var isSidebarVisible = _ref2.isSidebarVisible;
  return (0, _react.css)({
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    width: isSidebarVisible ? "calc(100% - ".concat(SIDEBAR_WIDTH, "px)") : '100%'
  });
};
var zoomWrapperStyles = (0, _react.css)({
  width: '100%',
  position: 'absolute',
  bottom: '0px',
  height: '98px',
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  backgroundImage: "linear-gradient( to top, ".concat(_modalSpinner.headerAndSidebarBackgroundColor, ", rgba(14, 22, 36, 0) )"),
  opacity: 0.85,
  pointerEvents: 'none',
  boxSizing: 'border-box',
  display: 'flex',
  alignItems: 'flex-end',
  padding: "var(--ds-space-100, 10px)".concat(" ", "var(--ds-space-300, 24px)")
});
var zoomCenterControlsStyles = (0, _react.css)({
  width: '100%',
  display: 'flex',
  justifyContent: 'center',
  gap: "var(--ds-space-100, 10px)",
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '> *': {
    pointerEvents: 'all'
  }
});
var zoomRightControlsStyles = (0, _react.css)({
  position: 'absolute',
  right: "var(--ds-space-300, 24px)",
  // eslint-disable-next-line @atlaskit/design-system/ensure-design-token-usage
  color: '#c7d1db',
  pointerEvents: 'all',
  display: 'flex',
  justifyContent: 'right',
  gap: "var(--ds-space-100, 10px)"
});
var zoomLevelIndicatorStyles = (0, _react.css)({
  height: '32px',
  lineHeight: '32px',
  verticalAlign: 'middle'
});
var hdIconGroupWrapperStyles = (0, _react.css)({
  display: 'flex',
  alignItems: 'center',
  gap: "var(--ds-space-100, 10px)",
  position: 'relative',
  width: '24px',
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '> *': {
    position: 'absolute'
  }
});
var errorMessageWrapperStyles = (0, _react.css)({
  textAlign: 'center',
  // eslint-disable-next-line @atlaskit/design-system/ensure-design-token-usage
  color: '#c7d1db',
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  p: {
    lineHeight: '100%'
  }
});
var errorImageStyles = (0, _react.css)({
  marginBottom: "var(--ds-space-100, 10px)",
  userSelect: 'none'
});
var videoStyles = (0, _react.css)({
  width: '100vw',
  height: '100vh'
});
var pdfWrapperStyles = (0, _react.css)((0, _defineProperty2.default)({
  overflow: 'auto',
  position: 'absolute',
  top: 0,
  left: 0,
  bottom: 0,
  right: 0
}, ".".concat(_mediaUi.hideControlsClassName), {
  position: 'fixed'
}));
var arrowStyles = (0, _react.css)({
  cursor: 'pointer',
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  svg: {
    filter: 'drop-shadow(0px 1px 1px rgb(9 30 66 / 25%)) drop-shadow(0px 0px 1px rgb(9 30 66 / 31%))'
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors, @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
  '&& button': {
    '&:hover': {
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
      svg: {
        // eslint-disable-next-line @atlaskit/design-system/ensure-design-token-usage
        color: '#b6c2cf'
      }
    },
    '&:active': {
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
      svg: {
        // eslint-disable-next-line @atlaskit/design-system/ensure-design-token-usage
        color: '#c7d1db'
      }
    }
  }
});
var arrowWrapperStyles = (0, _react.css)({
  position: 'absolute',
  top: '50%',
  transform: 'translateY(-50%)',
  padding: "var(--ds-space-250, 20px)"
});
var arrowsWrapperStyles = (0, _react.css)({
  display: 'flex',
  position: 'absolute',
  top: '50%',
  transform: 'translateY(-50%)',
  left: 0,
  width: '100%'
});
var leftWrapperStyles = function leftWrapperStyles(_ref3) {
  var isArchiveSideBarVisible = _ref3.isArchiveSideBarVisible;
  return (
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    (0, _react.css)(arrowWrapperStyles, {
      textAlign: 'left',
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
      left: isArchiveSideBarVisible ? "".concat(_styles2.ArchiveSideBarWidth, "px") : '0'
    })
  );
};

// eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
var rightWrapperStyles = (0, _react.css)(arrowWrapperStyles, {
  textAlign: 'right',
  right: 0
});
var headerStyles = function headerStyles(_ref4) {
  var isArchiveSideBarVisible = _ref4.isArchiveSideBarVisible;
  return (0, _react.css)({
    display: 'flex',
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
    paddingLeft: isArchiveSideBarVisible ? "".concat(_styles2.ArchiveSideBarWidth, "px") : '0'
  });
};
var leftHeaderStyles = (0, _react.css)({
  flex: 1,
  overflow: 'hidden',
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '> *': {
    pointerEvents: 'all'
  }
});
var imageWrapperStyles = (0, _react.css)({
  width: '100vw',
  height: '100vh',
  overflow: 'auto',
  textAlign: 'center',
  verticalAlign: 'middle',
  whiteSpace: 'nowrap'
});
var baselineExtendStyles = (0, _react.css)({
  height: '100%',
  display: 'inline-block',
  verticalAlign: 'middle'
});
var imgStyles = function imgStyles(_ref5) {
  var cursor = _ref5.cursor,
    shouldPixelate = _ref5.shouldPixelate;
  return (0, _react.css)({
    display: 'inline-block',
    verticalAlign: 'middle',
    position: 'relative',
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    cursor: cursor
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  shouldPixelate ? "/* Prevent images from being smoothed when scaled up */\n    image-rendering: optimizeSpeed; /* Legal fallback */\n    image-rendering: -moz-crisp-edges; /* Firefox        */\n    image-rendering: -o-crisp-edges; /* Opera          */\n    image-rendering: -webkit-optimize-contrast; /* Safari         */\n    image-rendering: optimize-contrast; /* CSS3 Proposed  */\n    image-rendering: crisp-edges; /* CSS4 Proposed  */\n    image-rendering: pixelated; /* CSS4 Proposed  */\n    -ms-interpolation-mode: nearest-neighbor; /* IE8+           */" : "");
};
var medatadataTextWrapperStyles = (0, _react.css)({
  overflow: 'hidden'
});
var metadataWrapperStyles = (0, _react.css)({
  display: 'flex'
});

// eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
var metadataFileNameStyles = (0, _react.css)((0, _mediaUi.ellipsis)());
var metadataSubTextStyles = (0, _react.css)({
  // eslint-disable-next-line @atlaskit/design-system/ensure-design-token-usage
  color: '#c7d1db'
},
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
(0, _mediaUi.ellipsis)());
var metadataIconWrapperStyles = (0, _primitives.xcss)({
  paddingTop: 'space.050',
  paddingRight: 'space.150'
});
var rightHeaderStyles = (0, _react.css)({
  textAlign: 'right',
  marginRight: "var(--ds-space-500, 40px)",
  minWidth: '200px',
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '> *': {
    pointerEvents: 'all'
  }
});
var customAudioPlayerWrapperStyles = (0, _react.css)({
  position: 'absolute',
  bottom: 0,
  left: 0,
  width: '100%'
});
var audioPlayerStyles = (0, _react.css)({
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  backgroundColor: _styles.blanketColor,
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  borderRadius: (0, _constants.borderRadius)(),
  alignItems: 'center',
  justifyContent: 'center',
  width: '400px',
  height: '400px',
  overflow: 'hidden',
  display: 'flex',
  flexDirection: 'column',
  position: 'relative'
});
var audioStyles = (0, _react.css)({
  width: '100%',
  position: 'absolute',
  bottom: 0,
  left: 0
});
var audioCoverStyles = (0, _react.css)({
  width: '100%',
  height: '100%',
  objectFit: 'scale-down',
  // eslint-disable-next-line @atlaskit/design-system/ensure-design-token-usage
  backgroundColor: '#000'
});
var defaultCoverWrapperStyles = (0, _react.css)({
  width: '100%',
  height: '100%',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  color: "var(--ds-text, #9FADBC)",
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '> *': {
    transform: 'scale(2)'
  }
});
var downloadButtonWrapperStyles = (0, _react.css)({
  marginTop: "var(--ds-space-300, 28px)",
  textAlign: 'center',
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  button: {
    '&:hover, &:active': {
      // eslint-disable-next-line @atlaskit/design-system/ensure-design-token-usage, @atlaskit/ui-styling-standard/no-important-styles -- Ignored via go/DSP-18766
      color: '#161a1d !important'
    }
  }
});
var customVideoPlayerWrapperStyles = (0, _react.css)({
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  video: {
    flex: 1,
    width: '100vw',
    height: '100vh',
    maxHeight: '100vh'
  }
});
var sidebarWrapperStyles = (0, _react.css)({
  top: 0,
  right: 0,
  width: "".concat(SIDEBAR_WIDTH, "px"),
  height: '100vh',
  overflow: 'hidden auto',
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
  backgroundColor: "var(--ds-surface, ".concat(_modalSpinner.headerAndSidebarBackgroundColor, ")"),
  color: "var(--ds-text, #c7d1db)"
});
var spinnerWrapperStyles = (0, _react.css)({
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  height: '100%'
});
var formattedMessageWrapperStyles = (0, _react.css)({});
// We are keeping this data-testid since JIRA is still using it in their codebase to perform checks. Before removing this, we need to ensure this 'media-viewer-popup' test id is not being used anywhere else in other codebases
var Blanket = exports.Blanket = function Blanket(_ref6) {
  var datatestId = _ref6['data-testid'],
    className = _ref6.className,
    children = _ref6.children;
  return (
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    (0, _react.jsx)("div", {
      css: blanketStyles,
      "data-testid": datatestId,
      className: className
    }, children)
  );
};
var HeaderWrapper = exports.HeaderWrapper = function HeaderWrapper(_ref7) {
  var className = _ref7.className,
    children = _ref7.children,
    isArchiveSideBarVisible = _ref7.isArchiveSideBarVisible;
  return (0, _react.jsx)("div", {
    // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage -- Ignored via go/DSP-18766
    css: headerWrapperStyles({
      isArchiveSideBarVisible: isArchiveSideBarVisible
    })
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    ,
    className: className
  }, children);
};
HeaderWrapper.displayName = 'HeaderWrapper';
var ListWrapper = exports.ListWrapper = function ListWrapper(_ref8) {
  var children = _ref8.children;
  return (0, _react.jsx)("div", {
    css: listWrapperStyles
  }, children);
};
ListWrapper.displayName = 'ListWrapper';
var ArrowsWrapper = exports.ArrowsWrapper = function ArrowsWrapper(_ref9) {
  var children = _ref9.children;
  return (0, _react.jsx)("div", {
    id: "media-viewer-navigation",
    css: arrowsWrapperStyles
  }, children);
};
var CloseButtonWrapper = exports.CloseButtonWrapper = function CloseButtonWrapper(_ref10) {
  var className = _ref10.className,
    children = _ref10.children;
  return (
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    (0, _react.jsx)("div", {
      css: closeButtonWrapperStyles,
      className: className
    }, children)
  );
};
var ContentWrapper = exports.ContentWrapper = function ContentWrapper(_ref11) {
  var isSidebarVisible = _ref11.isSidebarVisible,
    children = _ref11.children;
  return (
    // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage -- Ignored via go/DSP-18766
    (0, _react.jsx)("div", {
      css: contentWrapperStyles({
        isSidebarVisible: isSidebarVisible
      })
    }, children)
  );
};
var ZoomWrapper = exports.ZoomWrapper = function ZoomWrapper(_ref12) {
  var className = _ref12.className,
    children = _ref12.children;
  return (
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    (0, _react.jsx)("div", {
      css: zoomWrapperStyles,
      className: className
    }, children)
  );
};
var ZoomCenterControls = exports.ZoomCenterControls = function ZoomCenterControls(_ref13) {
  var children = _ref13.children;
  return (0, _react.jsx)("div", {
    css: zoomCenterControlsStyles
  }, children);
};
var ZoomRightControls = exports.ZoomRightControls = function ZoomRightControls(_ref14) {
  var children = _ref14.children;
  return (0, _react.jsx)("div", {
    css: zoomRightControlsStyles
  }, children);
};
var ZoomLevelIndicator = exports.ZoomLevelIndicator = function ZoomLevelIndicator(_ref15) {
  var children = _ref15.children;
  return (0, _react.jsx)("span", {
    css: zoomLevelIndicatorStyles,
    "data-testid": "zoom-level-indicator"
  }, children);
};
var HDIconGroupWrapper = exports.HDIconGroupWrapper = function HDIconGroupWrapper(_ref16) {
  var className = _ref16.className,
    children = _ref16.children;
  return (
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    (0, _react.jsx)("div", {
      css: hdIconGroupWrapperStyles,
      className: className
    }, children)
  );
};
var ErrorMessageWrapper = exports.ErrorMessageWrapper = function ErrorMessageWrapper(_ref17) {
  var datatestId = _ref17['data-testid'],
    children = _ref17.children;
  return (0, _react.jsx)("div", {
    css: errorMessageWrapperStyles,
    "data-testid": datatestId
  }, children);
};
var ErrorImage = exports.ErrorImage = function ErrorImage(_ref18) {
  var src = _ref18.src,
    alt = _ref18.alt;
  return (0, _react.jsx)("img", {
    css: errorImageStyles,
    alt: alt,
    src: src
  });
};
var Video = exports.Video = function Video(_ref19) {
  var autoPlay = _ref19.autoPlay,
    controls = _ref19.controls,
    src = _ref19.src;
  return (
    // eslint-disable-next-line jsx-a11y/media-has-caption
    (0, _react.jsx)("video", {
      css: videoStyles,
      autoPlay: autoPlay,
      controls: controls,
      src: src
    })
  );
};
var PDFWrapperBody = /*#__PURE__*/(0, _react2.forwardRef)(function (_ref20, ref) {
  var innerRef = _ref20.innerRef,
    datatestId = _ref20['data-testid'],
    children = _ref20.children;
  var bodyRef = (0, _useCallbackRef.useMergeRefs)([ref, innerRef]);
  return (0, _react.jsx)("div", {
    css: pdfWrapperStyles,
    ref: bodyRef,
    "data-testid": datatestId
  }, children);
});
var PDFWrapper = exports.PDFWrapper = /*#__PURE__*/(0, _react2.forwardRef)(function (props, ref) {
  return (0, _react.jsx)(_reactScrolllock.TouchScrollable, null, (0, _react.jsx)(PDFWrapperBody, (0, _extends2.default)({
    innerRef: ref
  }, props)));
});
var Arrow = exports.Arrow = function Arrow(_ref21) {
  var className = _ref21.className,
    children = _ref21.children;
  return (
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    (0, _react.jsx)("span", {
      css: arrowStyles,
      className: className
    }, children)
  );
};
var LeftWrapper = exports.LeftWrapper = function LeftWrapper(_ref22) {
  var children = _ref22.children,
    isArchiveSideBarVisible = _ref22.isArchiveSideBarVisible;
  return (
    // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage -- Ignored via go/DSP-18766
    (0, _react.jsx)("div", {
      css: leftWrapperStyles({
        isArchiveSideBarVisible: isArchiveSideBarVisible
      })
    }, children)
  );
};
var RightWrapper = exports.RightWrapper = function RightWrapper(_ref23) {
  var children = _ref23.children;
  return (0, _react.jsx)("div", {
    css: rightWrapperStyles
  }, children);
};

// header.tsx

var Header = exports.Header = function Header(_ref24) {
  var children = _ref24.children,
    isArchiveSideBarVisible = _ref24.isArchiveSideBarVisible,
    className = _ref24.className;
  return (
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop, @atlaskit/design-system/consistent-css-prop-usage -- Ignored via go/DSP-18766
    (0, _react.jsx)("div", {
      css: headerStyles({
        isArchiveSideBarVisible: isArchiveSideBarVisible
      }),
      className: className
    }, children)
  );
};
var LeftHeader = exports.LeftHeader = function LeftHeader(_ref25) {
  var children = _ref25.children;
  return (0, _react.jsx)("div", {
    css: leftHeaderStyles
  }, children);
};
var ImageWrapper = exports.ImageWrapper = /*#__PURE__*/(0, _react2.forwardRef)(function (_ref26, ref) {
  var children = _ref26.children,
    datatestId = _ref26['data-testid'],
    onClick = _ref26.onClick,
    style = _ref26.style,
    className = _ref26.className;
  return (
    // eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions
    (0, _react.jsx)("div", {
      "data-testid": datatestId,
      onClick: onClick,
      ref: ref,
      css: imageWrapperStyles
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
      ,
      style: style
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
      ,
      className: className
    }, children)
  );
});
var BaselineExtend = exports.BaselineExtend = function BaselineExtend() {
  return (0, _react.jsx)("div", {
    css: baselineExtendStyles
  });
};
var Img = exports.Img = function Img(_ref27) {
  var canDrag = _ref27.canDrag,
    isDragging = _ref27.isDragging,
    shouldPixelate = _ref27.shouldPixelate,
    datatestId = _ref27['data-testid'],
    src = _ref27.src,
    style = _ref27.style,
    onLoad = _ref27.onLoad,
    onError = _ref27.onError,
    onMouseDown = _ref27.onMouseDown,
    className = _ref27.className;
  var cursor = (0, _react2.useMemo)(function () {
    if (canDrag && isDragging) {
      return 'grabbing';
    } else if (canDrag) {
      return 'grab';
    } else {
      return 'auto';
    }
  }, [canDrag, isDragging]);
  return (
    // eslint-disable-next-line jsx-a11y/alt-text, jsx-a11y/no-noninteractive-element-interactions
    (0, _react.jsx)("img", {
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
      className: className
      // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage -- Ignored via go/DSP-18766
      ,
      css: imgStyles({
        cursor: cursor,
        shouldPixelate: shouldPixelate
      }),
      "data-testid": datatestId,
      src: src
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
      ,
      style: style,
      onLoad: onLoad,
      onError: onError,
      onMouseDown: onMouseDown
    })
  );
};
var MedatadataTextWrapper = exports.MedatadataTextWrapper = function MedatadataTextWrapper(_ref28) {
  var children = _ref28.children;
  return (0, _react.jsx)("div", {
    css: medatadataTextWrapperStyles
  }, children);
};
var MetadataWrapper = exports.MetadataWrapper = function MetadataWrapper(_ref29) {
  var children = _ref29.children;
  return (0, _react.jsx)("div", {
    css: metadataWrapperStyles
  }, children);
};
var MetadataFileName = exports.MetadataFileName = function MetadataFileName(_ref30) {
  var datatestId = _ref30['data-testid'],
    children = _ref30.children;
  return (0, _react.jsx)("div", {
    css: metadataFileNameStyles,
    "data-testid": datatestId
  }, children);
};
var MetadataSubText = exports.MetadataSubText = function MetadataSubText(_ref31) {
  var datatestId = _ref31['data-testid'],
    children = _ref31.children;
  return (0, _react.jsx)("div", {
    css: metadataSubTextStyles,
    "data-testid": datatestId
  }, children);
};
var MetadataIconWrapper = exports.MetadataIconWrapper = function MetadataIconWrapper(_ref32) {
  var children = _ref32.children;
  return (0, _react.jsx)(_primitives.Box, {
    xcss: metadataIconWrapperStyles
  }, children);
};
var RightHeader = exports.RightHeader = function RightHeader(_ref33) {
  var children = _ref33.children;
  return (0, _react.jsx)("div", {
    css: rightHeaderStyles
  }, children);
};
var CustomAudioPlayerWrapper = exports.CustomAudioPlayerWrapper = function CustomAudioPlayerWrapper(_ref34) {
  var children = _ref34.children;
  return (0, _react.jsx)("div", {
    css: customAudioPlayerWrapperStyles
  }, children);
};
var AudioPlayer = exports.AudioPlayer = function AudioPlayer(_ref35) {
  var datatestId = _ref35['data-testid'],
    children = _ref35.children;
  return (0, _react.jsx)("div", {
    css: audioPlayerStyles,
    "data-testid": datatestId
  }, children);
};
AudioPlayer.displayName = 'AudioPlayer';
var Audio = exports.Audio = /*#__PURE__*/(0, _react2.forwardRef)(function (_ref36, ref) {
  var autoPlay = _ref36.autoPlay,
    controls = _ref36.controls,
    src = _ref36.src,
    preload = _ref36.preload;
  return (
    // eslint-disable-next-line jsx-a11y/media-has-caption
    (0, _react.jsx)("audio", {
      css: audioStyles,
      ref: ref,
      autoPlay: autoPlay,
      controls: controls,
      src: src,
      preload: preload
    })
  );
});
var AudioCover = exports.AudioCover = function AudioCover(_ref37) {
  var src = _ref37.src,
    alt = _ref37.alt;
  return (0, _react.jsx)("img", {
    css: audioCoverStyles,
    alt: alt,
    src: src
  });
};
var DefaultCoverWrapper = exports.DefaultCoverWrapper = function DefaultCoverWrapper(_ref38) {
  var children = _ref38.children;
  return (0, _react.jsx)("div", {
    css: defaultCoverWrapperStyles
  }, children);
};
var DownloadButtonWrapper = exports.DownloadButtonWrapper = function DownloadButtonWrapper(_ref39) {
  var children = _ref39.children;
  return (0, _react.jsx)("div", {
    css: downloadButtonWrapperStyles
  }, children);
};
var CustomVideoPlayerWrapper = exports.CustomVideoPlayerWrapper = function CustomVideoPlayerWrapper(_ref40) {
  var datatestId = _ref40['data-testid'],
    children = _ref40.children;
  return (0, _react.jsx)("div", {
    css: customVideoPlayerWrapperStyles,
    "data-testid": datatestId
  }, children);
};
var SidebarWrapper = exports.SidebarWrapper = function SidebarWrapper(_ref41) {
  var datatestId = _ref41['data-testid'],
    children = _ref41.children;
  return (0, _react.jsx)("div", {
    css: sidebarWrapperStyles,
    "data-testid": datatestId
  }, children);
};
var SpinnerWrapper = exports.SpinnerWrapper = function SpinnerWrapper(_ref42) {
  var children = _ref42.children;
  return (0, _react.jsx)("div", {
    css: spinnerWrapperStyles
  }, children);
};
var FormattedMessageWrapper = exports.FormattedMessageWrapper = function FormattedMessageWrapper(_ref43) {
  var children = _ref43.children;
  return (0, _react.jsx)("span", {
    css: formattedMessageWrapperStyles
  }, children);
};
