// @atlaskit/editor-plugin-text-color/dist/es2019/ui/PrimaryToolbarComponent.js
import React from 'react';
import { useSharedPluginState } from '@atlaskit/editor-common/hooks';
import { ToolbarType } from '../types';

// Ignored via go/ees005
// eslint-disable-next-line import/no-named-as-default
import ToolbarTextColor from './ToolbarTextColor';
export function PrimaryToolbarComponent({
  api,
  isReducedSpacing,
  editorView,
  popupsMountPoint,
  popupsScrollableElement,
  popupsBoundariesElement,
  dispatchAnalyticsEvent,
  disabled
}) {
  const {
    textColorState
  } = useSharedPluginState(api, ['textColor']);
  return /*#__PURE__*/React.createElement(ToolbarTextColor
  // Ignored via go/ees005
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  , {
    pluginState: textColorState,
    isReducedSpacing: isReducedSpacing,
    editorView: editorView,
    popupsMountPoint: popupsMountPoint,
    popupsBoundariesElement: popupsBoundariesElement,
    popupsScrollableElement: popupsScrollableElement,
    dispatchAnalyticsEvent: dispatchAnalyticsEvent,
    disabled: disabled,
    pluginInjectionApi: api,
    toolbarType: ToolbarType.PRIMARY
  });
}
