// @atlaskit/media-ui/dist/cjs/BlockCard/components/Byline.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/** @jsx jsx */
var core_1 = require("@emotion/core");
var colors_1 = require("@atlaskit/theme/colors");
var utils_1 = require("../utils");
exports.Byline = function (_a) {
    var text = _a.text, children = _a.children, testId = _a.testId;
    return (core_1.jsx("span", { css: utils_1.mq({
            fontSize: utils_1.gs(1.5),
            lineHeight: utils_1.gs(2.5),
            color: "" + colors_1.N300,
            fontWeight: 'normal',
            marginTop: utils_1.gs(0.5),
            // Spec: only allow two lines MAX to be shown.
            display: '-webkit-box',
            overflow: 'hidden',
            textOverflow: 'ellipsis',
            WebkitLineClamp: 2,
            WebkitBoxOrient: 'vertical',
            wordBreak: 'break-word',
            // Fallback options.
            maxHeight: utils_1.gs(5),
            whiteSpace: 'pre-line',
            // EDM-713: fixes copy-paste from renderer to editor for Firefox
            // due to HTML its unwrapping behaviour on paste.
            MozUserSelect: 'none',
        }), "data-testid": testId }, text || children));
};
