// @atlaskit/editor-plugin-card/dist/cjs/ui/OpenButtonOverlay/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireWildcard(require("react"));
var _react2 = require("@emotion/react");
var _linkExternal = _interopRequireDefault(require("@atlaskit/icon/core/link-external"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _primitives = require("@atlaskit/primitives");
var _excluded = ["children", "isVisible", "url", "editorAppearance"];
/* eslint-disable @atlaskit/design-system/no-nested-styles */
/* eslint-disable @atlaskit/design-system/prefer-primitives */
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-line @atlaskit/ui-styling-standard/use-compiled
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var containerStyles = (0, _react2.css)({
  position: 'relative'
});
var iconWrapperStyles = (0, _primitives.xcss)({
  display: 'inline-flex',
  justifyContent: 'center',
  alignItems: 'center',
  height: '17px',
  width: '17px'
});
var hiddenTextStyle = (0, _react2.css)({
  overflow: 'hidden',
  whiteSpace: 'nowrap',
  position: 'absolute',
  visibility: 'hidden'
});
var linkStyles = (0, _primitives.xcss)({
  position: 'absolute',
  left: 'space.025',
  top: '-1px',
  display: 'inline-flex',
  alignItems: 'center',
  verticalAlign: 'middle',
  paddingBlock: 'space.025',
  paddingInline: 'space.050',
  gap: 'space.025',
  overflow: 'hidden',
  zIndex: 'card',
  backgroundColor: 'color.background.accent.gray.subtlest',
  borderRadius: "var(--ds-border-radius, 3px)",
  color: 'color.text.subtle',
  textDecoration: 'none',
  ':hover': {
    backgroundColor: 'elevation.surface.hovered',
    color: 'color.text.subtle',
    textDecoration: 'none'
  }
});
var linkStylesFix = (0, _primitives.xcss)({
  whiteSpace: 'nowrap'
});
var MIN_AVAILABLE_SPACE_WITH_LABEL_OVERLAY = 45;
var ICON_WIDTH = 16;
var DEFAULT_OPEN_TEXT_WIDTH = 28; // Default open text width in English

var OpenButtonOverlay = function OpenButtonOverlay(_ref) {
  var children = _ref.children,
    _ref$isVisible = _ref.isVisible,
    isVisible = _ref$isVisible === void 0 ? false : _ref$isVisible,
    url = _ref.url,
    editorAppearance = _ref.editorAppearance,
    props = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  // TODO: ED-26961 - add translation
  var label = 'Open';
  var containerRef = (0, _react.useRef)(null);
  var openButtonRef = (0, _react.useRef)(null);
  var hiddenTextRef = (0, _react.useRef)(null);
  var _useState = (0, _react.useState)(true),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    showLabel = _useState2[0],
    setShowLabel = _useState2[1];
  var _useState3 = (0, _react.useState)(false),
    _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
    isHovered = _useState4[0],
    setHovered = _useState4[1];
  var openTextWidthRef = (0, _react.useRef)(DEFAULT_OPEN_TEXT_WIDTH);
  var handleDoubleClick = function handleDoubleClick() {
    // Double click opens the link in a new tab
    window.open(url, '_blank');
  };
  (0, _react.useLayoutEffect)(function () {
    var hiddenText = hiddenTextRef.current;
    if (!hiddenText) {
      return;
    }
    // Measure the width of the hidden text
    // Temporarily make the element visible to measure its width
    hiddenText.style.visibility = 'hidden';
    hiddenText.style.display = 'inline';
    openTextWidthRef.current = hiddenText.offsetWidth;

    // Reset the hiddenText's display property
    hiddenText.style.display = 'none';
    hiddenText.style.visibility = 'inherit';
  }, []);
  (0, _react.useLayoutEffect)(function () {
    var _containerRef$current, _openButtonRef$curren;
    if (!isVisible || !isHovered) {
      return;
    }
    var cardWidth = (_containerRef$current = containerRef.current) === null || _containerRef$current === void 0 ? void 0 : _containerRef$current.offsetWidth;
    var openButtonWidth = (_openButtonRef$curren = openButtonRef.current) === null || _openButtonRef$curren === void 0 ? void 0 : _openButtonRef$curren.offsetWidth;
    if (!cardWidth || !openButtonWidth) {
      return;
    }
    var canShowLabel = true;
    if ((0, _platformFeatureFlags.fg)('platform_editor_controls_patch_2')) {
      canShowLabel = cardWidth - openTextWidthRef.current > MIN_AVAILABLE_SPACE_WITH_LABEL_OVERLAY + ICON_WIDTH;
    } else {
      canShowLabel = cardWidth - openButtonWidth > MIN_AVAILABLE_SPACE_WITH_LABEL_OVERLAY;
    }
    setShowLabel(canShowLabel);
  }, [isVisible, isHovered]);
  var handleOverlayChange = function handleOverlayChange(isHovered) {
    setHovered(isHovered);
  };
  return (
    // eslint-disable-next-line jsx-a11y/no-static-element-interactions
    (0, _react2.jsx)("span", {
      ref: containerRef,
      css: containerStyles,
      onDoubleClick: handleDoubleClick,
      onMouseEnter: function onMouseEnter() {
        return handleOverlayChange(true);
      },
      onMouseLeave: function onMouseLeave() {
        return handleOverlayChange(false);
      }
    }, children, (0, _platformFeatureFlags.fg)('platform_editor_controls_patch_2') && (0, _react2.jsx)("span", {
      css: hiddenTextStyle,
      "aria-hidden": "true"
    }, (0, _react2.jsx)(_primitives.Text, {
      ref: hiddenTextRef,
      size: "small",
      maxLines: 1
    }, label)), isHovered && (0, _react2.jsx)(_primitives.Anchor, {
      ref: openButtonRef,
      xcss: [linkStyles, (0, _platformFeatureFlags.fg)('platform_editor_controls_patch_5') && linkStylesFix],
      href: url,
      target: "_blank",
      style: {
        paddingBlock: editorAppearance === 'comment' && (0, _platformFeatureFlags.fg)('platform_editor_controls_patch_6') ? '1px' : "var(--ds-space-025, 2px)"
      }
    }, (0, _react2.jsx)(_primitives.Box, {
      xcss: iconWrapperStyles,
      "data-inlinecard-button-overlay": "icon-wrapper-line-height"
    }, (0, _react2.jsx)(_linkExternal.default, {
      label: ""
    })), showLabel && (0, _react2.jsx)(_primitives.Text, {
      size: "small",
      color: "color.text.subtle",
      maxLines: 1
    }, label)))
  );
};
var _default = exports.default = OpenButtonOverlay;
