// @atlaskit/tag/dist/cjs/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.textLinkCssVar = exports.textHoverCssVar = exports.textDefaultCssVar = exports.textActiveCssVar = exports.removeButtonHoverCssVar = exports.removeButtonDefaultCssVar = exports.focusRingCssVar = exports.borderRadiusCssVar = exports.borderColorCssVar = exports.backgroundHoverCssVar = exports.backgroundDefaultCssVar = exports.backgroundActiveCssVar = void 0;
var borderColorCssVar = exports.borderColorCssVar = '--ds-bc';
var focusRingCssVar = exports.focusRingCssVar = '--ds-cfr';
var textLinkCssVar = exports.textLinkCssVar = '--ds-ctl';
var textDefaultCssVar = exports.textDefaultCssVar = '--ds-ct';
var textHoverCssVar = exports.textHoverCssVar = '--ds-cth';
var textActiveCssVar = exports.textActiveCssVar = '--ds-ctp';
var backgroundDefaultCssVar = exports.backgroundDefaultCssVar = '--ds-cb';
var backgroundHoverCssVar = exports.backgroundHoverCssVar = '--ds-cbh';
var backgroundActiveCssVar = exports.backgroundActiveCssVar = '--ds-cba';
var removeButtonDefaultCssVar = exports.removeButtonDefaultCssVar = '--ds-rb';
var removeButtonHoverCssVar = exports.removeButtonHoverCssVar = '--ds-rbh';
var borderRadiusCssVar = exports.borderRadiusCssVar = '--ds-br';
