// @atlaskit/smart-card/dist/cjs/view/FlexibleCard/components/blocks/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AIFooterBlock", {
  enumerable: true,
  get: function get() {
    return _aiFooterBlock.default;
  }
});
Object.defineProperty(exports, "AISummaryBlock", {
  enumerable: true,
  get: function get() {
    return _aiSummaryBlock.default;
  }
});
Object.defineProperty(exports, "ActionBlock", {
  enumerable: true,
  get: function get() {
    return _actionBlock.default;
  }
});
Object.defineProperty(exports, "CustomBlock", {
  enumerable: true,
  get: function get() {
    return _block.default;
  }
});
Object.defineProperty(exports, "FooterBlock", {
  enumerable: true,
  get: function get() {
    return _footerBlock.default;
  }
});
Object.defineProperty(exports, "InternalFooterBlock", {
  enumerable: true,
  get: function get() {
    return _resolved.default;
  }
});
Object.defineProperty(exports, "MetadataBlock", {
  enumerable: true,
  get: function get() {
    return _metadataBlock.default;
  }
});
Object.defineProperty(exports, "PreviewBlock", {
  enumerable: true,
  get: function get() {
    return _previewBlock.default;
  }
});
Object.defineProperty(exports, "SnippetBlock", {
  enumerable: true,
  get: function get() {
    return _snippetBlock.default;
  }
});
Object.defineProperty(exports, "TitleBlock", {
  enumerable: true,
  get: function get() {
    return _titleBlock.default;
  }
});
var _footerBlock = _interopRequireDefault(require("./footer-block"));
var _metadataBlock = _interopRequireDefault(require("./metadata-block"));
var _previewBlock = _interopRequireDefault(require("./preview-block"));
var _snippetBlock = _interopRequireDefault(require("./snippet-block"));
var _titleBlock = _interopRequireDefault(require("./title-block"));
var _block = _interopRequireDefault(require("./block"));
var _resolved = _interopRequireDefault(require("./footer-block/resolved"));
var _aiSummaryBlock = _interopRequireDefault(require("./ai-summary-block"));
var _aiFooterBlock = _interopRequireDefault(require("./ai-footer-block"));
var _actionBlock = _interopRequireDefault(require("./action-block"));
