// @atlaskit/editor-common/dist/cjs/card/link-toolbar-dropdown-options.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getDropdownOption = void 0;
var _react = _interopRequireDefault(require("react"));
var _linkToolbarButtonGroupOptions = require("./link-toolbar-button-group-options");
var getDropdownOption = exports.getDropdownOption = function getDropdownOption(intl, dispatchCommand, _ref) {
  var disabled = _ref.disabled,
    _onClick = _ref.onClick,
    selected = _ref.selected,
    appearance = _ref.appearance,
    testId = _ref.testId,
    tooltip = _ref.tooltip;
  var _appearancePropsMap = _linkToolbarButtonGroupOptions.appearancePropsMap[appearance !== null && appearance !== void 0 ? appearance : 'url'],
    title = _appearancePropsMap.title,
    Icon = _appearancePropsMap.icon;
  return {
    title: intl.formatMessage(title),
    icon: /*#__PURE__*/_react.default.createElement(Icon, {
      label: intl.formatMessage(title)
    }),
    onClick: function onClick() {
      return dispatchCommand(_onClick);
    },
    disabled: Boolean(disabled),
    testId: testId,
    selected: selected,
    tooltip: tooltip !== null && tooltip !== void 0 ? tooltip : ''
  };
};
