// @atlaskit/profilecard/dist/cjs/components/team-profile-card/team-containers-skeleton/linked-container-card-skeleton/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LinkedContainerCardSkeleton = void 0;
require("./index.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _react = _interopRequireDefault(require("react"));
var _compiled = require("@atlaskit/primitives/compiled");
var styles = {
  containerWrapperSkeleton: "_2rkoiti9 _1h6dz9xs _12y3e4h9 _12jimuej _1qu2nqa1 _syaz1fxt _ca0qu2gc _u5f3u2gc _n3tdu2gc _19bvu2gc",
  containerIconSkeleton: "_2rkoiti9 _4t3i1tcg _1bsb1tcg _ca0qu2gc _u5f3u2gc _n3tdu2gc _19bvu2gc",
  firstLineSkeleton: "_2rko1q77 _1rjc1b66",
  secondLineSkeleton: "_2rko1q77 _1rjc1b66 _1bsbqkoa",
  containerTypeIconSkeleton: "_2rkoiti9 _4t3i7vkz _1bsb7vkz _ca0qu2gc _u5f3u2gc _n3tdu2gc _19bvu2gc"
};
var LinkedContainerCardSkeleton = exports.LinkedContainerCardSkeleton = function LinkedContainerCardSkeleton() {
  return /*#__PURE__*/_react.default.createElement(_compiled.Inline, {
    space: "space.100",
    xcss: styles.containerWrapperSkeleton
  }, /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    xcss: styles.containerIconSkeleton,
    backgroundColor: "color.background.neutral"
  }), /*#__PURE__*/_react.default.createElement(_compiled.Stack, {
    grow: "fill",
    space: "space.050"
  }, /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    xcss: styles.firstLineSkeleton,
    backgroundColor: "color.background.neutral"
  }), /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    xcss: styles.secondLineSkeleton,
    backgroundColor: "color.background.neutral"
  })), /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    xcss: styles.containerTypeIconSkeleton,
    backgroundColor: "color.background.neutral"
  }));
};
