// @atlaskit/media-card/dist/es2019/utils/mockIntersectionObserver.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
const noop = () => {};
export class MockIntersectionObserver {
  constructor() {
    _defineProperty(this, "callback", noop);
  }
  storeMockObserver({
    observe,
    disconnect
  }) {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const outerThis = this;
    this.mockObserver = class {
      constructor(callback) {
        _defineProperty(this, "observe", observe !== null && observe !== void 0 ? observe : noop);
        _defineProperty(this, "unobserve", noop);
        _defineProperty(this, "disconnect", disconnect !== null && disconnect !== void 0 ? disconnect : noop);
        outerThis.callback = callback;
        outerThis.mockObserverInstance = this;
      }
    };
  }
  addMockObserverToWindow() {
    this.originalObserver = window.IntersectionObserver;
    window.IntersectionObserver = this.mockObserver;
  }
  removeMockObserverFromWindow() {
    window.IntersectionObserver = this.originalObserver;
  }
  setup(options = {}) {
    this.storeMockObserver(options);
    this.addMockObserverToWindow();
  }
  cleanup() {
    this.removeMockObserverFromWindow();
  }
  triggerIntersect(entry) {
    const entries = [entry];
    this.callback(entries, this.mockObserverInstance);
  }
}
