// @atlaskit/media-viewer/dist/es2019/newgen/analytics/item-viewer.js
import { packageAttributes, fileStateToFileGasPayload } from './index';
export const mediaFileCommencedEvent = (id) => {
    return {
        eventType: 'operational',
        action: 'commenced',
        actionSubject: 'mediaFile',
        actionSubjectId: id,
        attributes: {
            fileId: id,
            ...packageAttributes,
        },
    };
};
export const mediaFileLoadSucceededEvent = (file) => {
    return {
        eventType: 'operational',
        actionSubject: 'mediaFile',
        action: 'loadSucceeded',
        actionSubjectId: file.id,
        attributes: {
            status: 'success',
            ...fileStateToFileGasPayload(file),
            ...packageAttributes,
        },
    };
};
export const mediaFileLoadFailedEvent = (id, failReason, file) => {
    const fileAttributes = file
        ? fileStateToFileGasPayload(file)
        : {
            fileId: id,
        };
    return {
        eventType: 'operational',
        actionSubject: 'mediaFile',
        action: 'loadFailed',
        actionSubjectId: id,
        attributes: {
            status: 'fail',
            ...fileAttributes,
            failReason,
            ...packageAttributes,
        },
    };
};
export const mediaPreviewFailedEvent = (failReason, fileState) => {
    const fileId = fileState ? fileState.id : undefined;
    const fileAttributes = fileState && fileStateToFileGasPayload(fileState);
    return {
        eventType: 'operational',
        actionSubject: 'mediaFile',
        action: 'previewFailed',
        actionSubjectId: fileId,
        attributes: {
            status: 'fail',
            ...fileAttributes,
            failReason,
            ...packageAttributes,
        },
    };
};
