// @atlaskit/editor-core/dist/cjs/plugins/rule/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = tslib_1.__importStar(require("react"));
var adf_schema_1 = require("@atlaskit/adf-schema");
var ToolbarInsertBlock_1 = require("../insert-block/ui/ToolbarInsertBlock");
var keymap_1 = tslib_1.__importDefault(require("./pm-plugins/keymap"));
var input_rule_1 = tslib_1.__importDefault(require("./pm-plugins/input-rule"));
var analytics_1 = require("../analytics");
var assets_1 = require("../quick-insert/assets");
var rulePlugin = function () { return ({
    nodes: function () {
        return [{ name: 'rule', node: adf_schema_1.rule }];
    },
    pmPlugins: function () {
        return [
            {
                name: 'ruleInputRule',
                plugin: function (_a) {
                    var schema = _a.schema;
                    return input_rule_1.default(schema);
                },
            },
            {
                name: 'ruleKeymap',
                plugin: function () { return keymap_1.default(); },
            },
        ];
    },
    pluginsOptions: {
        quickInsert: function (_a) {
            var formatMessage = _a.formatMessage;
            return [
                {
                    title: formatMessage(ToolbarInsertBlock_1.messages.horizontalRule),
                    description: formatMessage(ToolbarInsertBlock_1.messages.horizontalRuleDescription),
                    keywords: ['horizontal rule', 'rule', 'line'],
                    priority: 1200,
                    keyshortcut: '---',
                    icon: function () { return (React.createElement(assets_1.IconDivider, { label: formatMessage(ToolbarInsertBlock_1.messages.horizontalRule) })); },
                    action: function (insert, state) {
                        var tr = insert(state.schema.nodes.rule.createChecked());
                        return analytics_1.addAnalytics(tr, {
                            action: analytics_1.ACTION.INSERTED,
                            actionSubject: analytics_1.ACTION_SUBJECT.DOCUMENT,
                            actionSubjectId: analytics_1.ACTION_SUBJECT_ID.DIVIDER,
                            attributes: { inputMethod: analytics_1.INPUT_METHOD.QUICK_INSERT },
                            eventType: analytics_1.EVENT_TYPE.TRACK,
                        });
                    },
                },
            ];
        },
    },
}); };
exports.default = rulePlugin;
