// @atlaskit/editor-core/dist/esm/plugins/media/nodeviews/mediaNodeUpdater.js
import { __awaiter, __generator } from "tslib";
import uuidV4 from 'uuid/v4';
import { updateMediaNodeAttrs, replaceExternalMedia } from '../commands';
import { DEFAULT_IMAGE_HEIGHT, DEFAULT_IMAGE_WIDTH, } from '@atlaskit/editor-common';
import { getViewMediaClientConfigFromMediaProvider, getUploadMediaClientConfigFromMediaProvider, } from '../utils/media-common';
import { getMediaClient } from '@atlaskit/media-client';
import { ACTION, ACTION_SUBJECT, EVENT_TYPE, } from '../../analytics';
var MediaNodeUpdater = /** @class */ (function () {
    function MediaNodeUpdater(props) {
        var _this = this;
        // Updates the node with contextId if it doesn't have one already
        this.updateContextId = function () { return __awaiter(_this, void 0, void 0, function () {
            var attrs, id, objectId;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        attrs = this.getAttrs();
                        if (!attrs) {
                            return [2 /*return*/];
                        }
                        id = attrs.id;
                        return [4 /*yield*/, this.getObjectId()];
                    case 1:
                        objectId = _a.sent();
                        updateMediaNodeAttrs(id, {
                            __contextId: objectId,
                            contextId: objectId,
                        }, this.props.isMediaSingle)(this.props.view.state, this.props.view.dispatch);
                        return [2 /*return*/];
                }
            });
        }); };
        this.hasFileAttributesDefined = function () {
            var attrs = _this.getAttrs();
            return (attrs && attrs.__fileName && attrs.__fileMimeType && attrs.__fileSize);
        };
        this.updateFileAttrs = function () { return __awaiter(_this, void 0, void 0, function () {
            var attrs, mediaProvider, mediaClientConfig, mediaClient, options, fileState, name, mimeType, size;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        attrs = this.getAttrs();
                        return [4 /*yield*/, this.props.mediaProvider];
                    case 1:
                        mediaProvider = _a.sent();
                        if (!mediaProvider ||
                            !mediaProvider.uploadParams ||
                            !attrs ||
                            !attrs.id ||
                            this.hasFileAttributesDefined()) {
                            return [2 /*return*/];
                        }
                        return [4 /*yield*/, getViewMediaClientConfigFromMediaProvider(mediaProvider)];
                    case 2:
                        mediaClientConfig = _a.sent();
                        mediaClient = getMediaClient({
                            mediaClientConfig: mediaClientConfig,
                        });
                        options = {
                            collectionName: attrs.collection,
                        };
                        return [4 /*yield*/, mediaClient.file.getCurrentState(attrs.id, options)];
                    case 3:
                        fileState = _a.sent();
                        if (fileState.status === 'error') {
                            return [2 /*return*/];
                        }
                        name = fileState.name, mimeType = fileState.mimeType, size = fileState.size;
                        updateMediaNodeAttrs(attrs.id, {
                            __fileName: name,
                            __fileMimeType: mimeType,
                            __fileSize: size,
                        }, true)(this.props.view.state, this.props.view.dispatch);
                        return [2 /*return*/];
                }
            });
        }); };
        this.getAttrs = function () {
            var attrs = _this.props.node.attrs;
            if (attrs) {
                return attrs;
            }
            return undefined;
        };
        this.getObjectId = function () { return __awaiter(_this, void 0, void 0, function () {
            var contextIdentifierProvider;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.props
                            .contextIdentifierProvider];
                    case 1:
                        contextIdentifierProvider = _a.sent();
                        return [2 /*return*/, contextIdentifierProvider && contextIdentifierProvider.objectId];
                }
            });
        }); };
        this.uploadExternalMedia = function (pos) { return __awaiter(_this, void 0, void 0, function () {
            var node, mediaProvider, uploadMediaClientConfig, mediaClient, collection, uploader, uploadableFileUpfrontIds, dimensions, e_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        node = this.props.node;
                        return [4 /*yield*/, this.props.mediaProvider];
                    case 1:
                        mediaProvider = _a.sent();
                        if (!(node && mediaProvider)) return [3 /*break*/, 6];
                        return [4 /*yield*/, getUploadMediaClientConfigFromMediaProvider(mediaProvider)];
                    case 2:
                        uploadMediaClientConfig = _a.sent();
                        if (!uploadMediaClientConfig || !node.attrs.url) {
                            return [2 /*return*/];
                        }
                        mediaClient = getMediaClient({
                            mediaClientConfig: uploadMediaClientConfig,
                        });
                        collection = mediaProvider.uploadParams && mediaProvider.uploadParams.collection;
                        _a.label = 3;
                    case 3:
                        _a.trys.push([3, 5, , 6]);
                        return [4 /*yield*/, mediaClient.file.uploadExternal(node.attrs.url, collection)];
                    case 4:
                        uploader = _a.sent();
                        uploadableFileUpfrontIds = uploader.uploadableFileUpfrontIds, dimensions = uploader.dimensions;
                        replaceExternalMedia(pos + 1, {
                            id: uploadableFileUpfrontIds.id,
                            collection: collection,
                            height: dimensions.height,
                            width: dimensions.width,
                        })(this.props.view.state, this.props.view.dispatch);
                        return [3 /*break*/, 6];
                    case 5:
                        e_1 = _a.sent();
                        //keep it as external media
                        if (this.props.dispatchAnalyticsEvent) {
                            this.props.dispatchAnalyticsEvent({
                                action: ACTION.UPLOAD_EXTERNAL_FAIL,
                                actionSubject: ACTION_SUBJECT.EDITOR,
                                eventType: EVENT_TYPE.OPERATIONAL,
                            });
                        }
                        return [3 /*break*/, 6];
                    case 6: return [2 /*return*/];
                }
            });
        }); };
        this.getCurrentContextId = function () {
            var attrs = _this.getAttrs();
            if (!attrs) {
                return undefined;
            }
            return attrs.__contextId;
        };
        this.updateDimensions = function (dimensions) {
            updateMediaNodeAttrs(dimensions.id, {
                height: dimensions.height,
                width: dimensions.width,
            }, true)(_this.props.view.state, _this.props.view.dispatch);
        };
        this.isNodeFromDifferentCollection = function () { return __awaiter(_this, void 0, void 0, function () {
            var mediaProvider, currentCollectionName, attrs, nodeCollection, __contextId, contextId, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.props.mediaProvider];
                    case 1:
                        mediaProvider = _b.sent();
                        if (!mediaProvider || !mediaProvider.uploadParams) {
                            return [2 /*return*/, false];
                        }
                        currentCollectionName = mediaProvider.uploadParams.collection;
                        attrs = this.getAttrs();
                        if (!attrs) {
                            return [2 /*return*/, false];
                        }
                        nodeCollection = attrs.collection, __contextId = attrs.__contextId;
                        _a = __contextId;
                        if (_a) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.getObjectId()];
                    case 2:
                        _a = (_b.sent());
                        _b.label = 3;
                    case 3:
                        contextId = _a;
                        if (contextId && currentCollectionName !== nodeCollection) {
                            return [2 /*return*/, true];
                        }
                        return [2 /*return*/, false];
                }
            });
        }); };
        this.copyNode = function () { return __awaiter(_this, void 0, void 0, function () {
            var mediaProvider, _a, isMediaSingle, view, attrs, currentCollectionName, contextId, _b, uploadMediaClientConfig, mediaClient, auth_1, id, collection, source, destination, mediaFile;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0: return [4 /*yield*/, this.props.mediaProvider];
                    case 1:
                        mediaProvider = _c.sent();
                        _a = this.props, isMediaSingle = _a.isMediaSingle, view = _a.view;
                        attrs = this.getAttrs();
                        if (!mediaProvider || !mediaProvider.uploadParams || !attrs) {
                            return [2 /*return*/];
                        }
                        currentCollectionName = mediaProvider.uploadParams.collection;
                        _b = this.getCurrentContextId();
                        if (_b) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.getObjectId()];
                    case 2:
                        _b = (_c.sent());
                        _c.label = 3;
                    case 3:
                        contextId = _b;
                        return [4 /*yield*/, getUploadMediaClientConfigFromMediaProvider(mediaProvider)];
                    case 4:
                        uploadMediaClientConfig = _c.sent();
                        if (!uploadMediaClientConfig) {
                            return [2 /*return*/];
                        }
                        mediaClient = getMediaClient({
                            mediaClientConfig: uploadMediaClientConfig,
                        });
                        if (!(uploadMediaClientConfig.getAuthFromContext && contextId)) return [3 /*break*/, 7];
                        return [4 /*yield*/, uploadMediaClientConfig.getAuthFromContext(contextId)];
                    case 5:
                        auth_1 = _c.sent();
                        id = attrs.id, collection = attrs.collection;
                        source = {
                            id: id,
                            collection: collection,
                            authProvider: function () { return Promise.resolve(auth_1); },
                        };
                        destination = {
                            collection: currentCollectionName,
                            authProvider: uploadMediaClientConfig.authProvider,
                            occurrenceKey: uuidV4(),
                        };
                        return [4 /*yield*/, mediaClient.file.copyFile(source, destination)];
                    case 6:
                        mediaFile = _c.sent();
                        updateMediaNodeAttrs(source.id, {
                            id: mediaFile.id,
                            collection: currentCollectionName,
                        }, isMediaSingle)(view.state, view.dispatch);
                        _c.label = 7;
                    case 7: return [2 /*return*/];
                }
            });
        }); };
        this.props = props;
    }
    MediaNodeUpdater.prototype.getRemoteDimensions = function () {
        return __awaiter(this, void 0, void 0, function () {
            var mediaProvider, _a, node, mediaPluginOptions, attrs, _b, height, type, width, _c, id, collection, viewMediaClientConfig, mediaClient, state;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0: return [4 /*yield*/, this.props.mediaProvider];
                    case 1:
                        mediaProvider = _d.sent();
                        _a = this.props, node = _a.node, mediaPluginOptions = _a.mediaPluginOptions;
                        attrs = node.attrs;
                        if (!mediaProvider || !attrs) {
                            return [2 /*return*/, false];
                        }
                        _b = attrs, height = _b.height, type = _b.type, width = _b.width;
                        if (type === 'external') {
                            return [2 /*return*/, false];
                        }
                        _c = attrs, id = _c.id, collection = _c.collection;
                        if (height && width) {
                            return [2 /*return*/, false];
                        }
                        // can't fetch remote dimensions on mobile, so we'll default them
                        if (mediaPluginOptions && !mediaPluginOptions.allowRemoteDimensionsFetch) {
                            return [2 /*return*/, {
                                    id: id,
                                    height: DEFAULT_IMAGE_HEIGHT,
                                    width: DEFAULT_IMAGE_WIDTH,
                                }];
                        }
                        return [4 /*yield*/, getViewMediaClientConfigFromMediaProvider(mediaProvider)];
                    case 2:
                        viewMediaClientConfig = _d.sent();
                        mediaClient = getMediaClient({
                            mediaClientConfig: viewMediaClientConfig,
                        });
                        return [4 /*yield*/, mediaClient.getImageMetadata(id, {
                                collection: collection,
                            })];
                    case 3:
                        state = _d.sent();
                        if (!state || !state.original) {
                            return [2 /*return*/, false];
                        }
                        return [2 /*return*/, {
                                id: id,
                                height: state.original.height || DEFAULT_IMAGE_HEIGHT,
                                width: state.original.width || DEFAULT_IMAGE_WIDTH,
                            }];
                }
            });
        });
    };
    return MediaNodeUpdater;
}());
export { MediaNodeUpdater };
