// @atlaskit/side-navigation/dist/cjs/components/Footer/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./index.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _warnOnce = _interopRequireDefault(require("@atlaskit/ds-lib/warn-once"));
var _compiled = require("@atlaskit/primitives/compiled");
var _colors = require("@atlaskit/theme/colors");
var _Header = require("../Header");
var _Item = require("../Item");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/* eslint-disable @repo/internal/react/consistent-props-definitions */
var styles = {
  iconContainer: "_1e0c1o8l _1bsb1k8s _4t3i1k8s",
  oldFooter: "_vrjtdcr7 _16vx1h6o _1cg71wug _if291ule _14741h6o _oiqe1tcg _1kba1h6o _w1ap1osq _1mt2ze3t _159iu2gc _76zj1o8l _13e21h6o _14bw1o8l _eq7s12x7 _1jkv12x7 _1rla12x7 _z2a212x7 _11kjqtfy _jl2n73ad _1et6qtfy _1v6jjjyb"
};
/**
 * __Footer__
 *
 * - [Examples](https://atlassian.design/components/side-navigation/examples#header-and-footer)
 * - [Code](https://atlassian.design/components/side-navigation/code)
 */
var OldFooter = function OldFooter(props) {
  // https://stackoverflow.com/a/39333479
  var safeProps = function (_ref) {
    var iconBefore = _ref.iconBefore,
      onClick = _ref.onClick,
      description = _ref.description,
      children = _ref.children,
      testId = _ref.testId;
    return {
      iconBefore: iconBefore,
      onClick: onClick,
      description: description,
      children: children,
      testId: testId
    };
  }(props);
  return /*#__PURE__*/React.createElement(_Item.CustomItem, (0, _extends2.default)({}, safeProps, {
    component: props.component || _Header.Container,
    className: (0, _runtime.ax)([styles.oldFooter])
  }));
};

/**
 * __Footer__
 *
 * - [Examples](https://atlassian.design/components/side-navigation/examples#header-and-footer)
 * - [Code](https://atlassian.design/components/side-navigation/code)
 */
var Footer = function Footer(_ref2) {
  var description = _ref2.description,
    children = _ref2.children,
    iconBefore = _ref2.iconBefore,
    testId = _ref2.testId;
  return /*#__PURE__*/React.createElement(_compiled.Box, {
    padding: "space.100",
    testId: testId
  }, /*#__PURE__*/React.createElement(_compiled.Stack, {
    space: "space.100",
    alignInline: "center"
  }, /*#__PURE__*/React.createElement(_compiled.Box, {
    xcss: styles.iconContainer
  }, iconBefore), /*#__PURE__*/React.createElement(_compiled.Stack, {
    space: "space.075"
  }, /*#__PURE__*/React.createElement(_compiled.Text, {
    size: "UNSAFE_small",
    align: "center",
    color: "inherit"
  }, children), description && /*#__PURE__*/React.createElement(_compiled.Text, {
    size: "UNSAFE_small",
    as: "p",
    align: "center",
    color: "color.text.subtlest"
  }, description))));
};

/**
 * __Footer__
 *
 * - [Examples](https://atlassian.design/components/side-navigation/examples#header-and-footer)
 * - [Code](https://atlassian.design/components/side-navigation/code)
 */
var FooterFacade = function FooterFacade(_ref3) {
  var _ref3$useDeprecatedAp = _ref3.useDeprecatedApi,
    useDeprecatedApi = _ref3$useDeprecatedAp === void 0 ? true : _ref3$useDeprecatedAp,
    description = _ref3.description,
    iconBefore = _ref3.iconBefore,
    testId = _ref3.testId,
    children = _ref3.children,
    component = _ref3.component,
    onClick = _ref3.onClick;
  if (!useDeprecatedApi) {
    return /*#__PURE__*/React.createElement(Footer, {
      iconBefore: iconBefore,
      description: description,
      testId: testId
    }, children);
  }
  if (typeof process !== 'undefined' && process.env.NODE_ENV !== 'production' && process.env.NODE_ENV !== 'CI') {
    (0, _warnOnce.default)('The `@atlaskit/side-navigation` Footer has simplified its API, removing support for the `component` and `onClick` props.\n\nThese props will stop working in a future major release. Reach out to #help-design-system on slack for help.');
  }
  return /*#__PURE__*/React.createElement(OldFooter, {
    onClick: onClick,
    component: component,
    iconBefore: iconBefore,
    description: description,
    testId: testId
  }, children);
};
var _default = exports.default = FooterFacade;
