// @atlaskit/editor-plugin-help-dialog/dist/cjs/ui/ModalHeader.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("@emotion/react");
var _reactIntlNext = require("react-intl-next");
var _messages = require("@atlaskit/editor-common/messages");
var _uiMenu = require("@atlaskit/editor-common/ui-menu");
var _heading = _interopRequireDefault(require("@atlaskit/heading"));
var _closeCross = _interopRequireDefault(require("@atlaskit/icon/core/migration/close--cross"));
var _styles = require("./styles");
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var ModalHeader = (0, _reactIntlNext.injectIntl)(function (_ref) {
  var formatMessage = _ref.intl.formatMessage,
    onClose = _ref.onClose;
  return (
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/design-system/consistent-css-prop-usage -- Ignored via go/DSP-18766
    (0, _react.jsx)("div", {
      css: _styles.header
    }, (0, _react.jsx)(_heading.default, {
      size: "large"
    }, (0, _react.jsx)(_reactIntlNext.FormattedMessage
    // Ignored via go/ees005
    // eslint-disable-next-line react/jsx-props-no-spreading
    , _messages.helpDialogMessages.editorHelp)), (0, _react.jsx)("div", null, (0, _react.jsx)(_uiMenu.ToolbarButton
    // @ts-expect-error modal onClose handler requires second parameter of UIAnalyticsEvent, which we don't want to pass
    , {
      onClick: onClose,
      title: formatMessage(_messages.helpDialogMessages.closeHelpDialog),
      spacing: "compact",
      iconBefore: (0, _react.jsx)(_closeCross.default, {
        label: formatMessage(_messages.helpDialogMessages.closeHelpDialog),
        color: "currentColor",
        spacing: "spacious"
      })
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/design-system/no-unsafe-style-overrides -- Ignored via go/DSP-18766
      ,
      css: _styles.toolbarButton
    })))
  );
});
var _default = exports.default = ModalHeader;
