// @atlaskit/editor-plugin-help-dialog/dist/cjs/ui/styles.js
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.toolbarButton = exports.shortcutsArray = exports.row = exports.line = exports.header = exports.footer = exports.dialogHeader = exports.contentWrapper = exports.content = exports.componentFromKeymapWrapperStyles = exports.column = void 0;
var _react = require("@emotion/react");
var _editorSharedStyles = require("@atlaskit/editor-shared-styles");
var _colors = _interopRequireWildcard(require("@atlaskit/theme/colors"));
var colors = _colors;
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

// Ignored via go/ees005
// eslint-disable-next-line import/no-namespace

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
var header = exports.header = (0, _react.css)({
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  zIndex: _editorSharedStyles.akEditorUnitZIndex,
  minHeight: "var(--ds-space-300, 24px)",
  padding: "var(--ds-space-250, 20px)".concat(" ", "var(--ds-space-500, 40px)"),
  display: 'flex',
  justifyContent: 'space-between',
  alignItems: 'center',
  boxShadow: "'none'",
  font: "var(--ds-font-heading-large, normal 500 24px/28px ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
  color: "var(--ds-text, ".concat(colors.N400, ")"),
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
  backgroundColor: "var(--ds-background-neutral-subtle, ".concat(colors.N0, ")"),
  borderRadius: "var(--ds-border-radius, 3px)"
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
var footer = exports.footer = (0, _react.css)({
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  zIndex: _editorSharedStyles.akEditorUnitZIndex,
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
  color: "var(--ds-text-subtlest, ".concat(colors.N300, ")"),
  font: "var(--ds-font-body, normal 400 14px/20px ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  padding: "var(--ds-space-300, 24px)",
  textAlign: 'right',
  boxShadow: "'none'"
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var contentWrapper = exports.contentWrapper = (0, _react.css)({
  padding: "var(--ds-space-250, 20px)".concat(" ", "var(--ds-space-500, 40px)"),
  borderBottomRightRadius: "var(--ds-border-radius, 3px)",
  overflow: 'auto',
  position: 'relative',
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
  color: "var(--ds-text-subtle, ".concat(colors.N400, ")"),
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
  backgroundColor: "var(--ds-background-neutral-subtle, ".concat(colors.N0, ")")
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var line = exports.line = (0, _react.css)({
  background: "var(--ds-background-neutral-subtle, #fff)",
  content: "''",
  display: 'block',
  height: "var(--ds-space-025, 2px)",
  left: 0,
  position: 'absolute',
  top: 0,
  right: 0,
  width: '100%',
  minWidth: '604px'
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var content = exports.content = (0, _react.css)({
  minWidth: '524px',
  width: '100%',
  position: 'relative',
  display: 'flex',
  justifyContent: 'space-between'
});
var column = exports.column = {
  width: '44%',
  '& > ul': {
    padding: 0
  }
};

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var row = exports.row = (0, _react.css)({
  margin: "var(--ds-space-250, 20px)".concat(" 0"),
  display: 'flex',
  justifyContent: 'space-between'
});
var dialogHeader = exports.dialogHeader = {
  '&': {
    fontSize: (0, _editorSharedStyles.relativeFontSizeToBase16)(24),
    fontWeight: "var(--ds-font-weight-regular, 400)",
    color: "var(--ds-text-subtle, ".concat(_colors.N400, ")"),
    letterSpacing: 'normal',
    lineHeight: 1.42857142857143
  }
};

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var shortcutsArray = exports.shortcutsArray = (0, _react.css)({
  display: 'flex',
  flexDirection: 'column',
  flexShrink: 0,
  gap: "var(--ds-space-150, 12px)"
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var componentFromKeymapWrapperStyles = exports.componentFromKeymapWrapperStyles = (0, _react.css)({
  flexShrink: 0
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var toolbarButton = exports.toolbarButton = (0, _react.css)({
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-selectors
  '&:focus:not(:focus-visible)': {
    outline: "2px solid ".concat("var(--ds-border-focused, ".concat(_colors.B300, ")")),
    outlineOffset: "var(--ds-space-025, 2px)"
  }
});
