// @atlaskit/user-picker/dist/es2019/components/ClearIndicator.js
/* ClearIndicator.tsx generated by @compiled/babel-plugin v0.36.1 */
import _extends from "@babel/runtime/helpers/extends";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import "./ClearIndicator.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
import React from 'react';
import { components } from '@atlaskit/select';
const AsyncTooltip = /*#__PURE__*/React.lazy(() => import( /* webpackChunkName: "@atlaskit-internal_@atlaskit/tooltip" */'@atlaskit/tooltip').then(module => {
  return {
    default: module.default
  };
}));
const clearIndicatorStyles = {
  root: "_v5641ppt _tzy4kb7n _ca0qze3t _u5f3ze3t _n3tdze3t _19bvze3t _30l3s4qr _1mhjidpf"
};
export class ClearIndicator extends React.PureComponent {
  constructor(...args) {
    super(...args);
    _defineProperty(this, "handleMouseDown", event => {
      if (event && event.type === 'mousedown' && event.button !== 0) {
        return;
      }
      // Prevent focus when clear on blurred state
      const {
        clearValue,
        selectProps
      } = this.props;
      clearValue();
      //@ts-ignore react-select unsupported props
      if (selectProps && !selectProps.isFocused) {
        event.stopPropagation();
      }
    });
  }
  render() {
    const {
      //@ts-ignore react-select unsupported props
      selectProps: {
        clearValueLabel
      }
    } = this.props;
    const Indicator = /*#__PURE__*/React.createElement(components.ClearIndicator, _extends({}, this.props, {
      innerProps: {
        ...this.props.innerProps,
        onMouseDown: this.handleMouseDown
      },
      xcss: clearIndicatorStyles.root
    }));
    return clearValueLabel ? /*#__PURE__*/React.createElement(React.Suspense, {
      fallback: Indicator
    }, /*#__PURE__*/React.createElement(AsyncTooltip, {
      content: clearValueLabel
    }, Indicator)) : Indicator;
  }
}
