// @atlaskit/user-picker/dist/es2019/components/PopupControl.js
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
import React from 'react';
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import { css, jsx } from '@emotion/react';
import { N200 } from '@atlaskit/theme/colors';
import Control from './Control';
const controlWrapper = css({
  display: 'flex',
  flexDirection: 'column',
  padding: `0px ${"var(--ds-space-100, 8px)"} ${"var(--ds-space-100, 8px)"}`
});
const labelStyle = css({
  color: `var(--ds-text-subtlest, ${N200})`,
  font: "var(--ds-font-body-UNSAFE_small, normal 400 12px/16px ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  fontWeight: "var(--ds-font-weight-semibold, 600)",
  paddingBottom: "var(--ds-space-050, 4px)",
  paddingLeft: "var(--ds-space-0, 0px)",
  paddingRight: "var(--ds-space-0, 0px)",
  paddingTop: "var(--ds-space-250, 20px)"
});
export class PopupControl extends React.PureComponent {
  render() {
    const {
      //@ts-ignore react-select unsupported props
      selectProps: {
        popupTitle
      }
    } = this.props;
    return jsx("div", {
      css: controlWrapper
    }, jsx("div", {
      css: labelStyle
    }, popupTitle), jsx(Control, this.props));
  }
}
