// @atlaskit/teams-public/dist/esm/common/utils/error/index.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _createClass from "@babel/runtime/helpers/createClass";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import _wrapNativeSuper from "@babel/runtime/helpers/wrapNativeSuper";
var _GraphQLError;
function _callSuper(t, o, e) { return o = _getPrototypeOf(o), _possibleConstructorReturn(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], _getPrototypeOf(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
export var StatusCode = /*#__PURE__*/function (StatusCode) {
  StatusCode[StatusCode["UNAUTHORIZED"] = 401] = "UNAUTHORIZED";
  StatusCode[StatusCode["FORBIDDEN"] = 403] = "FORBIDDEN";
  StatusCode[StatusCode["TIMEOUT"] = 408] = "TIMEOUT";
  StatusCode[StatusCode["GONE"] = 410] = "GONE";
  StatusCode[StatusCode["PAYLOAD_TOO_LARGE"] = 413] = "PAYLOAD_TOO_LARGE";
  return StatusCode;
}({});
var CommonError = /*#__PURE__*/function (_Error) {
  function CommonError(message) {
    var _this;
    _classCallCheck(this, CommonError);
    _this = _callSuper(this, CommonError, [message]);
    _this.name = _this.constructor.name;
    _this.message = message || 'UnknownError';
    _this.stack = new Error(message).stack || '';
    return _this;
  }
  _inherits(CommonError, _Error);
  return _createClass(CommonError);
}( /*#__PURE__*/_wrapNativeSuper(Error));
export var DefaultError = /*#__PURE__*/function (_CommonError) {
  function DefaultError(_ref) {
    var _this2;
    var message = _ref.message;
    _classCallCheck(this, DefaultError);
    _this2 = _callSuper(this, DefaultError, [message || 'UnknownError']);
    Object.setPrototypeOf(_this2, DefaultError.prototype);
    return _this2;
  }
  _inherits(DefaultError, _CommonError);
  return _createClass(DefaultError);
}(CommonError);

/**
 * These errors will not fail UFO experiences
 */
export var SLOIgnoreError = /*#__PURE__*/function (_CommonError2) {
  function SLOIgnoreError(_ref2) {
    var _this3;
    var message = _ref2.message;
    _classCallCheck(this, SLOIgnoreError);
    _this3 = _callSuper(this, SLOIgnoreError, ["SentryIgnore: ".concat(message || 'UnknownError')]);
    Object.setPrototypeOf(_this3, SLOIgnoreError.prototype);
    return _this3;
  }
  _inherits(SLOIgnoreError, _CommonError2);
  return _createClass(SLOIgnoreError);
}(CommonError);
// Http Errors
export var HttpError = /*#__PURE__*/function (_CommonError3) {
  function HttpError(_ref3) {
    var _this4;
    var message = _ref3.message,
      status = _ref3.status,
      traceId = _ref3.traceId,
      path = _ref3.path;
    _classCallCheck(this, HttpError);
    _this4 = _callSuper(this, HttpError, [message]);
    Object.setPrototypeOf(_this4, HttpError.prototype);
    _this4.status = status;
    _this4.traceId = traceId;
    _this4.path = path;
    return _this4;
  }
  _inherits(HttpError, _CommonError3);
  return _createClass(HttpError);
}(CommonError);

// Graphql Errors

export var GraphQLError = /*#__PURE__*/function (_CommonError4) {
  // tslint:disable-line no-any

  function GraphQLError(_ref4) {
    var _this5;
    var message = _ref4.message,
      _ref4$category = _ref4.category,
      category = _ref4$category === void 0 ? 'default' : _ref4$category,
      fields = _ref4.fields;
    _classCallCheck(this, GraphQLError);
    _this5 = _callSuper(this, GraphQLError, [message]);
    Object.setPrototypeOf(_this5, GraphQLError.prototype);
    _this5.category = category;
    if (fields) {
      _this5.fields = fields;
    }
    return _this5;
  }
  _inherits(GraphQLError, _CommonError4);
  return _createClass(GraphQLError);
}(CommonError);
_GraphQLError = GraphQLError;
_defineProperty(GraphQLError, "from", function (rawErrors) {
  var firstError = rawErrors[0];
  var errorData = {
    category: firstError.category,
    message: firstError.message
  };
  if (firstError.fields) {
    errorData.fields = firstError.fields.reduce(function (obj, item) {
      obj[item.field] = item.message;
      return obj;
    }, {});
  }
  return new _GraphQLError(errorData);
});
