// @atlaskit/link-extractors/dist/cjs/smart-link/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.extractSmartLinkUrl = exports.extractSmartLinkTitle = exports.extractSmartLinkProvider = exports.extractSmartLinkIcon = exports.extractSmartLinkEmbed = void 0;
var _common = require("../common");
var _extractPreview = require("../common/preview/extractPreview");
var _entity = require("../entity");
var extractSmartLinkEmbed = exports.extractSmartLinkEmbed = function extractSmartLinkEmbed(response, iframeUrlType) {
  if ((0, _entity.isEntityPresent)(response)) {
    var embedUrl = (0, _entity.extractEntityEmbedUrl)(response);
    return embedUrl ? {
      src: embedUrl
    } : undefined;
  }
  return (0, _extractPreview.extractPreview)(response === null || response === void 0 ? void 0 : response.data, 'web', iframeUrlType);
};
var extractSmartLinkTitle = exports.extractSmartLinkTitle = function extractSmartLinkTitle(response, removeTextHighlightingFromTitle) {
  if ((0, _entity.isEntityPresent)(response)) {
    return (0, _entity.extractEntityTitle)(response);
  }
  return (0, _common.extractTitle)(response === null || response === void 0 ? void 0 : response.data, removeTextHighlightingFromTitle);
};
var extractSmartLinkUrl = exports.extractSmartLinkUrl = function extractSmartLinkUrl(response) {
  if ((0, _entity.isEntityPresent)(response)) {
    var _extractEntity;
    return (_extractEntity = (0, _entity.extractEntity)(response)) === null || _extractEntity === void 0 ? void 0 : _extractEntity.url;
  }
  return (0, _common.extractLink)(response === null || response === void 0 ? void 0 : response.data);
};

/**
 * This funtions assumes that the response contains nounData.
 */
var extractSmartLinkIcon = exports.extractSmartLinkIcon = function extractSmartLinkIcon(response) {
  var entity = (0, _entity.extractEntity)(response);
  var url;
  if ((0, _entity.instanceOfDesignEntity)(entity)) {
    url = entity === null || entity === void 0 ? void 0 : entity['atlassian:design'].iconUrl;
  }
  var label = (0, _entity.extractEntityTitle)(response);
  return {
    url: url,
    label: label
  };
};
var extractSmartLinkProvider = exports.extractSmartLinkProvider = function extractSmartLinkProvider(response) {
  if ((0, _entity.isEntityPresent)(response)) {
    return (0, _entity.extractEntityProvider)(response);
  }
  return (0, _common.extractProvider)(response === null || response === void 0 ? void 0 : response.data);
};
