// @atlaskit/editor-plugin-layout/dist/cjs/pm-plugins/resizing.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.pluginKey = exports.default = void 0;
var _safePlugin = require("@atlaskit/editor-common/safe-plugin");
var _state = require("@atlaskit/editor-prosemirror/state");
var _nodeviews = require("../nodeviews");
var pluginKey = exports.pluginKey = new _state.PluginKey('layoutResizingPlugin');
var _default = exports.default = function _default(options, pluginInjectionApi, portalProviderAPI, eventDispatcher) {
  return new _safePlugin.SafePlugin({
    key: pluginKey,
    props: {
      nodeViews: {
        layoutSection: function layoutSection(node, view, getPos) {
          return new _nodeviews.LayoutSectionView({
            node: node,
            view: view,
            getPos: getPos,
            portalProviderAPI: portalProviderAPI,
            eventDispatcher: eventDispatcher,
            pluginInjectionApi: pluginInjectionApi,
            options: options
          }).init();
        }
      }
    }
  });
};
