// @atlaskit/editor-plugin-selection-extension/dist/cjs/pm-plugins/main.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.selectionExtensionPluginKey = exports.createPlugin = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _safePlugin = require("@atlaskit/editor-common/safe-plugin");
var _state = require("@atlaskit/editor-prosemirror/state");
var _types = require("../types");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var selectionExtensionPluginKey = exports.selectionExtensionPluginKey = new _state.PluginKey('selectionExtensionPlugin');
var createPlugin = exports.createPlugin = function createPlugin() {
  return new _safePlugin.SafePlugin({
    key: selectionExtensionPluginKey,
    state: {
      init: function init() {
        return {
          activeExtension: undefined
        };
      },
      apply: function apply(tr, pluginState) {
        var meta = tr.getMeta(selectionExtensionPluginKey);
        switch (meta === null || meta === void 0 ? void 0 : meta.type) {
          case _types.SelectionExtensionActionTypes.SET_ACTIVE_EXTENSION:
            return _objectSpread(_objectSpread({}, pluginState), {}, {
              activeExtension: meta.extension
            });
          case _types.SelectionExtensionActionTypes.CLEAR_ACTIVE_EXTENSION:
            return _objectSpread(_objectSpread({}, pluginState), {}, {
              activeExtension: undefined
            });
        }
        return pluginState;
      }
    }
  });
};
