// @atlaskit/jql-editor/dist/es2019/ui/jql-editor-footer-content/jql-messages/infos/index.js
import React from 'react';
import { di } from 'react-magnetic-di';
import { HelperMessage } from '@atlaskit/form';
import LegacyInfoIcon from '@atlaskit/icon/glyph/info';
import InfoIcon from '@atlaskit/icon/utility/information';
import { Box } from '@atlaskit/primitives/compiled';
import * as colors from '@atlaskit/theme/colors';
import { useEditorViewHasInfos } from '../../../../hooks/use-editor-view-has-infos';
import { useExternalMessages } from '../../../../state';
import { FormatMessages, MessageContainer } from '../format';
export const useFormattedInfoMessage = () => {
  const [{
    infos: externalInfos
  }] = useExternalMessages();
  const hasInfos = useEditorViewHasInfos();
  if (!hasInfos) {
    return null;
  }
  return /*#__PURE__*/React.createElement(FormatMessages, {
    messages: externalInfos
  });
};
export const InfoMessages = () => {
  const infoMessage = useFormattedInfoMessage();
  return infoMessage != null ? /*#__PURE__*/React.createElement(MessageContainer, null, /*#__PURE__*/React.createElement(HelperMessage, {
    testId: "jql-editor-info-message"
  }, /*#__PURE__*/React.createElement(Box, {
    as: "span",
    paddingInlineEnd: "space.050"
  }, /*#__PURE__*/React.createElement(InfoIcon, {
    label: "",
    color: `var(--ds-icon-information, ${colors.B500})`,
    LEGACY_size: "small",
    LEGACY_fallbackIcon: LegacyInfoIcon,
    LEGACY_margin: "0 -4px 0 0"
  })), infoMessage)) : null;
};
