// @atlaskit/analytics-listeners/dist/cjs/helpers/extract-data-from-event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
/**
 * Largely taken from analytics-web-react
 */
var lodash_merge_1 = tslib_1.__importDefault(require("lodash.merge"));
var extractFromEventContext = function (propertyNames, event, namespacedContextOnly, contextName) {
    if (namespacedContextOnly === void 0) { namespacedContextOnly = true; }
    return event.context.reduce(function (acc, contextItem) {
        propertyNames.forEach(function (propertyName) {
            var navContext = contextItem[contextName];
            var navContextProp = navContext ? navContext[propertyName] : null;
            var value = namespacedContextOnly
                ? navContextProp
                : navContextProp || contextItem[propertyName];
            if (value) {
                acc.push(value);
            }
        });
        return acc;
    }, []);
};
exports.getSources = function (event, contextName) {
    return extractFromEventContext(['source'], event, false, contextName);
};
exports.getComponents = function (event, contextName) {
    return extractFromEventContext(['component', 'componentName'], event, false, contextName);
};
exports.getExtraAttributes = function (event, contextName) {
    return extractFromEventContext(['attributes'], event, true, contextName).reduce(function (result, extraAttributes) { return lodash_merge_1.default(result, extraAttributes); }, {});
};
exports.getPackageInfo = function (event, contextName) {
    return event.context
        .map(function (contextItem) {
        var navContext = contextItem[contextName];
        var item = navContext ? navContext : contextItem;
        return {
            packageName: item.packageName,
            packageVersion: item.packageVersion,
        };
    })
        .filter(function (p) { return p.packageName; });
};
exports.getPackageVersion = function (event, contextName) { return extractFromEventContext(['packageVersion'], event, true, contextName); };
