// @atlaskit/side-navigation/dist/es2019/components/Footer/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./index.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
/* eslint-disable @repo/internal/react/consistent-props-definitions */

import warnOnce from '@atlaskit/ds-lib/warn-once';
import { Box, Stack, Text } from '@atlaskit/primitives/compiled';
import { N500 } from '@atlaskit/theme/colors';
import { Container } from '../Header';
import { CustomItem } from '../Item';
const styles = {
  iconContainer: "_1e0c1o8l _1bsb1k8s _4t3i1k8s",
  oldFooter: "_vrjtdcr7 _16vx1h6o _1cg71wug _if291ule _14741h6o _oiqe1tcg _1kba1h6o _w1ap1osq _1mt2ze3t _159iu2gc _76zj1o8l _13e21h6o _14bw1o8l _eq7s12x7 _1jkv12x7 _1rla12x7 _z2a212x7 _11kjqtfy _jl2n73ad _1et6qtfy _1v6jjjyb"
};
/**
 * __Footer__
 *
 * - [Examples](https://atlassian.design/components/side-navigation/examples#header-and-footer)
 * - [Code](https://atlassian.design/components/side-navigation/code)
 */
const OldFooter = props => {
  // https://stackoverflow.com/a/39333479
  const safeProps = (({
    iconBefore,
    onClick,
    description,
    children,
    testId
  }) => ({
    iconBefore,
    onClick,
    description,
    children,
    testId
  }))(props);
  return /*#__PURE__*/React.createElement(CustomItem, _extends({}, safeProps, {
    component: props.component || Container,
    className: ax([styles.oldFooter])
  }));
};

/**
 * __Footer__
 *
 * - [Examples](https://atlassian.design/components/side-navigation/examples#header-and-footer)
 * - [Code](https://atlassian.design/components/side-navigation/code)
 */
const Footer = ({
  description,
  children,
  iconBefore,
  testId
}) => {
  return /*#__PURE__*/React.createElement(Box, {
    padding: "space.100",
    testId: testId
  }, /*#__PURE__*/React.createElement(Stack, {
    space: "space.100",
    alignInline: "center"
  }, /*#__PURE__*/React.createElement(Box, {
    xcss: styles.iconContainer
  }, iconBefore), /*#__PURE__*/React.createElement(Stack, {
    space: "space.075"
  }, /*#__PURE__*/React.createElement(Text, {
    size: "UNSAFE_small",
    align: "center",
    color: "inherit"
  }, children), description && /*#__PURE__*/React.createElement(Text, {
    size: "UNSAFE_small",
    as: "p",
    align: "center",
    color: "color.text.subtlest"
  }, description))));
};

/**
 * __Footer__
 *
 * - [Examples](https://atlassian.design/components/side-navigation/examples#header-and-footer)
 * - [Code](https://atlassian.design/components/side-navigation/code)
 */
const FooterFacade = ({
  useDeprecatedApi = true,
  description,
  iconBefore,
  testId,
  children,
  component,
  onClick
}) => {
  if (!useDeprecatedApi) {
    return /*#__PURE__*/React.createElement(Footer, {
      iconBefore: iconBefore,
      description: description,
      testId: testId
    }, children);
  }
  if (typeof process !== 'undefined' && process.env.NODE_ENV !== 'production' && process.env.NODE_ENV !== 'CI') {
    warnOnce('The `@atlaskit/side-navigation` Footer has simplified its API, removing support for the `component` and `onClick` props.\n\nThese props will stop working in a future major release. Reach out to #help-design-system on slack for help.');
  }
  return /*#__PURE__*/React.createElement(OldFooter, {
    onClick: onClick,
    component: component,
    iconBefore: iconBefore,
    description: description,
    testId: testId
  }, children);
};
export default FooterFacade;
