// @atlaskit/width-detector/dist/cjs/WidthObserver/iframe-fallbacks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var hooks_1 = require("./hooks");
var utils_1 = require("./utils");
var emptySubscribe = function () { return function () { }; };
var _a = react_1.default.createContext({
    subscribe: null,
}), IframeWrapperConsumer = _a.Consumer, IframeWrapperProvider = _a.Provider;
exports.IframeWrapperConsumer = IframeWrapperConsumer;
exports.IframeWrapperProvider = IframeWrapperProvider;
function ObjectIframe(props) {
    var onResize = props.onResize;
    var ref = react_1.useRef(null);
    react_1.useEffect(function () {
        if (!ref || !ref.current) {
            return;
        }
        var iframe = ref.current;
        if (!iframe.contentDocument || !iframe.contentDocument.defaultView) {
            return;
        }
        var iframeWindow = iframe.contentDocument.defaultView;
        iframeWindow.addEventListener('resize', onResize);
        return function () {
            iframeWindow.removeEventListener('resize', onResize);
        };
    }, [ref, onResize]);
    return (react_1.default.createElement("object", { ref: ref, data: "about:blank", type: "text/html", style: { position: 'absolute', height: '0', width: '100%' }, "aria-hidden": true, tabIndex: -1 }));
}
/**
 * IE11 requires IFrame with src to avoid Access Denied error.
 *
 * https://stackoverflow.com/questions/1886547/access-is-denied-javascript-error-when-trying-to-access-the-document-object-of
 **/
function NativeIframe(props) {
    var onResize = props.onResize;
    var ref = react_1.useRef(null);
    var onLoad = react_1.default.useCallback(function () {
        if (!ref || !ref.current) {
            return;
        }
        var iframe = ref.current;
        if (!iframe.contentDocument || !iframe.contentDocument.defaultView) {
            return;
        }
        var iframeWindow = iframe.contentDocument.defaultView;
        iframeWindow.addEventListener('resize', onResize);
    }, [onResize]);
    return (react_1.default.createElement("iframe", { ref: ref, src: "about:blank", frameBorder: "0", style: { position: 'absolute', height: '0', width: '100%' }, onLoad: onLoad, "aria-hidden": true, tabIndex: -1 }));
}
function Iframe(props) {
    if (utils_1.getIEVersion() === 11) {
        return react_1.default.createElement(NativeIframe, tslib_1.__assign({}, props));
    }
    return react_1.default.createElement(ObjectIframe, tslib_1.__assign({}, props));
}
var IframeWrapper = function (_a) {
    var children = _a.children;
    return (react_1.default.createElement("div", { style: {
            position: 'absolute',
            height: '0',
            width: '100%',
        } }, children));
};
var emptySubscription = function () { return function () { }; };
var SubscribeIframeResize = react_1.default.memo(function (_a) {
    var subscribe = _a.subscribe, setWidth = _a.setWidth;
    var ref = react_1.default.useRef(null);
    react_1.useEffect(function () {
        if (ref && ref.current) {
            var target = ref.current;
            var width = target.offsetWidth;
            setWidth(width);
        }
        var unsubscribe = subscribe(function () {
            if (ref && ref.current) {
                var target = ref.current;
                var width = target.offsetWidth;
                setWidth(width);
            }
        });
        return unsubscribe;
    }, [ref, subscribe, setWidth]);
    return react_1.default.createElement("div", { ref: ref });
});
exports.SubscribeIframeResize = SubscribeIframeResize;
var SubscribeIframeResizeWhenVisible = react_1.default.memo(function (_a) {
    var subscribe = _a.subscribe, setWidth = _a.setWidth;
    var _b = tslib_1.__read(hooks_1.useInView({
        /* Optional options */
        threshold: 0,
    }), 4), inViewRef = _b[0], inView = _b[1], target = _b[2], entry = _b[3];
    react_1.useEffect(function () {
        if (inView && entry && entry.target instanceof HTMLElement) {
            var boundingClientRect = entry.boundingClientRect;
            var width = boundingClientRect.width;
            setWidth(width);
        }
        var unsubscribe = subscribe(function () {
            // The first time the component is mounted will need to
            // get this information from the HTMLElement
            if (inView && target instanceof HTMLElement) {
                var width = target.offsetWidth;
                setWidth(width);
            }
        });
        return unsubscribe;
    }, [inView, entry, setWidth, subscribe, target]);
    return (react_1.default.createElement("div", { style: {
            position: 'absolute',
            height: '0',
            width: '100%',
        }, ref: inViewRef }));
});
exports.SubscribeIframeResizeWhenVisible = SubscribeIframeResizeWhenVisible;
function getSubscribeIframe(useIntersectionObserver) {
    if (useIntersectionObserver) {
        return SubscribeIframeResizeWhenVisible;
    }
    return SubscribeIframeResize;
}
/**
 * This component is responsible for creating an iframe using HTMLObjectElement.
 * It will provide a context object with a `subscribe` function as a callback,
 * so other components can subscribe to know when the iframe was resized.
 */
var IframeWidthObserverFallback = react_1.default.memo(function (props) {
    var _a = tslib_1.__read(react_1.useState(new Map()), 1), listeners = _a[0];
    var subscribe = react_1.default.useCallback(function (cb) {
        listeners.set(cb, null);
        return function () {
            listeners.delete(cb);
        };
    }, [listeners]);
    var onResize = react_1.default.useCallback(function () {
        listeners.forEach(function (_, cb) { return cb(); });
    }, [listeners]);
    return (react_1.default.createElement(react_1.default.Fragment, null,
        react_1.default.createElement(IframeWrapper, null,
            react_1.default.createElement(Iframe, { onResize: onResize })),
        react_1.default.createElement(IframeWrapperProvider, { value: { subscribe: subscribe } }, props.children)));
});
exports.IframeWidthObserverFallback = IframeWidthObserverFallback;
var NonIframeWidthObserverFallback = react_1.default.memo(function (props) {
    return (react_1.default.createElement(IframeWrapperProvider, { value: { subscribe: emptySubscribe } }, props.children));
});
exports.IframeWidthObserver = react_1.default.memo(function (_a) {
    var setWidth = _a.setWidth, useIntersectionObserver = _a.useIntersectionObserver;
    var Component = getSubscribeIframe(useIntersectionObserver);
    return (react_1.default.createElement(IframeWrapperConsumer, null, function (_a) {
        var subscribe = _a.subscribe;
        return (react_1.default.createElement(Component, { setWidth: setWidth, subscribe: subscribe || emptySubscription }));
    }));
});
exports.IframeWidthObserverFallbackWrapper = react_1.default.memo(function (props) {
    var supportsResizeObserver = utils_1.browser.supportsResizeObserver, supportsIntersectionObserver = utils_1.browser.supportsIntersectionObserver;
    if (supportsResizeObserver && supportsIntersectionObserver) {
        return (react_1.default.createElement(NonIframeWidthObserverFallback, null, props.children));
    }
    return (react_1.default.createElement(IframeWidthObserverFallback, null, props.children));
});
