// @atlaskit/react-ufo/dist/cjs/vc/vc-observer/revisions/ViewportUpdateClassifier.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ViewportUpdateClassifier = void 0;
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var ViewportUpdateClassifier = exports.ViewportUpdateClassifier = /*#__PURE__*/function () {
  function ViewportUpdateClassifier() {
    (0, _classCallCheck2.default)(this, ViewportUpdateClassifier);
    (0, _defineProperty2.default)(this, "types", []);
    (0, _defineProperty2.default)(this, "filters", []);
    (0, _defineProperty2.default)(this, "removedFilters", []);
    (0, _defineProperty2.default)(this, "__combinedTypes", []);
    (0, _defineProperty2.default)(this, "__combinedFilters", []);
  }
  return (0, _createClass2.default)(ViewportUpdateClassifier, [{
    key: "VCCalculationMethod",
    value: function VCCalculationMethod(_) {
      return {
        VC: {},
        VCBox: {}
      };
    }

    // @todo remove it once fixed as described: https://product-fabric.atlassian.net/browse/AFO-3443
  }, {
    key: "filterComponentsLog",
    value: function filterComponentsLog(_ref) {
      var componentsLog = _ref.componentsLog;
      return componentsLog;
    }
  }, {
    key: "mergeConfig",
    value: function mergeConfig() {
      var _this = this;
      this.__combinedTypes = [].concat((0, _toConsumableArray2.default)(this.types), (0, _toConsumableArray2.default)((this === null || this === void 0 ? void 0 : this.__combinedTypes) || []));
      var previousFilters = this.removedFilters.length === 0 ? this.__combinedFilters : this.__combinedFilters.filter(function (filter) {
        return !_this.removedFilters.includes(filter.name);
      });
      this.__combinedFilters = [].concat((0, _toConsumableArray2.default)(this.filters), (0, _toConsumableArray2.default)(previousFilters));
    }
  }, {
    key: "classifyUpdate",
    value: function classifyUpdate(_ref2) {
      var element = _ref2.element,
        type = _ref2.type,
        tags = _ref2.tags,
        ignoreReason = _ref2.ignoreReason;
      if (!this.__combinedTypes.includes(type)) {
        return false;
      }
      return this.__combinedFilters.every(function (_ref3) {
        var filter = _ref3.filter,
          name = _ref3.name;
        return filter({
          type: type,
          tags: tags,
          ignoreReason: ignoreReason
        });
      });
    }
  }]);
}();
