// @atlaskit/editor-plugin-selection/dist/es2019/pm-plugins/mark-boundary-cursor/ui/mark-boundary-cursor-decoration.js
import { Decoration } from '@atlaskit/editor-prosemirror/view';
const containerStyle = `
	position: relative;
`;
const caretStyle = `
	position: absolute;
	height: 100%;
	width: 1px;
	left: -0.5px;
	bottom: 0;
	background-color: ${"var(--ds-text, #172B4D)"};
`;
const toDOM = () => {
  var _caret$animate;
  const container = document.createElement('span');
  container.className = 'ProseMirror-mark-boundary-cursor';
  container.setAttribute('style', containerStyle);
  const caret = document.createElement('div');
  caret.setAttribute('style', caretStyle);
  (_caret$animate = caret.animate) === null || _caret$animate === void 0 ? void 0 : _caret$animate.call(caret, [{
    opacity: 1
  }, {
    opacity: 1,
    offset: 0.5
  }, {
    opacity: 0,
    offset: 0.5
  }, {
    opacity: 0
  }], {
    duration: 1000,
    iterations: Infinity
  });
  container.appendChild(caret);
  return container;
};
export const createMarkBoundaryCursorDecoration = (pos, side) => {
  return Decoration.widget(pos, toDOM, {
    key: 'mark-boundary-cursor',
    side
  });
};
