// react-loosely-lazy/dist/cjs/lazy/components/server.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createComponentServer = createComponentServer;
var _manifest = require("@react-loosely-lazy/manifest");
var _react = _interopRequireWildcard(require("react"));
var _config = require("../../config");
var _constants = require("../../constants");
var _suspense = require("../../suspense");
var _utils = require("../../utils");
var _errors = require("../errors");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function load(moduleId, loader) {
  try {
    return (0, _utils.getExport)(loader());
  } catch (err) {
    throw (0, _errors.createLoaderError)(err);
  }
}
function createComponentServer({
  dataLazyId,
  defer,
  loader,
  moduleId,
  ssr
}) {
  return props => {
    var _getAssetUrlsFromId;
    const Resolved = ssr ? load(moduleId, loader) : null;
    const {
      fallback
    } = (0, _react.useContext)(_suspense.LazySuspenseContext);
    const {
      crossOrigin,
      manifest
    } = (0, _config.getConfig)();
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("input", {
      type: "hidden",
      "data-lazy-begin": dataLazyId
    }), defer !== _constants.PHASE.LAZY && ((_getAssetUrlsFromId = (0, _manifest.getAssetUrlsFromId)(manifest, moduleId)) === null || _getAssetUrlsFromId === void 0 ? void 0 : _getAssetUrlsFromId.map(url => /*#__PURE__*/_react.default.createElement("link", {
      key: url,
      rel: defer === _constants.PHASE.PAINT ? 'preload' : 'prefetch',
      href: url,
      crossOrigin: crossOrigin,
      as: "script"
    }))), Resolved ? /*#__PURE__*/_react.default.createElement(Resolved, props) : fallback, /*#__PURE__*/_react.default.createElement("input", {
      type: "hidden",
      "data-lazy-end": dataLazyId
    }));
  };
}
