// @atlaskit/jql-ast/dist/cjs/creators/compoundOperator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.compoundOperatorInternal = exports.compoundOperator = void 0;
var _common = require("./common");
function acceptCompoundOperator(visitor) {
  return visitor.visitCompoundOperator ? visitor.visitCompoundOperator(this) : visitor.visitChildren(this);
}
function enterNode(listener) {
  listener.enterCompoundOperator && listener.enterCompoundOperator(this);
}
function exitNode(listener) {
  listener.exitCompoundOperator && listener.exitCompoundOperator(this);
}
var compoundOperator = exports.compoundOperator = function compoundOperator(value) {
  return compoundOperatorInternal(value, []);
};
var compoundOperatorInternal = exports.compoundOperatorInternal = function compoundOperatorInternal(value, operatorPositions) {
  return {
    value: value,
    positions: operatorPositions,
    position: null,
    accept: acceptCompoundOperator,
    enterNode: enterNode,
    exitNode: exitNode,
    getChildren: _common.noChildren,
    parent: null
  };
};
