// @atlaskit/jql-ast/dist/cjs/creators/field.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.fieldInternal = exports.field = void 0;
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _sanitiseJqlString = require("../utils/sanitise-jql-string");
var _common = require("./common");
function acceptField(visitor) {
  return visitor.visitField ? visitor.visitField(this) : visitor.visitChildren(this);
}
function enterNode(listener) {
  listener.enterField && listener.enterField(this);
}
function exitNode(listener) {
  listener.exitField && listener.exitField(this);
}
function getChildren() {
  return this.properties ? (0, _toConsumableArray2.default)(this.properties) : [];
}
var field = exports.field = function field(value, properties) {
  return fieldInternal(value, (0, _sanitiseJqlString.sanitiseJqlString)(value), properties);
};
var fieldInternal = exports.fieldInternal = function fieldInternal(value, text, properties) {
  var position = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : null;
  var node = {
    text: text,
    value: value,
    properties: properties,
    position: position,
    accept: acceptField,
    enterNode: enterNode,
    exitNode: exitNode,
    getChildren: getChildren,
    parent: null
  };
  (0, _common.assignParent)(node);
  return node;
};
