// @atlaskit/jql-ast/dist/cjs/creators/functionString.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.functionStringInternal = exports.functionString = void 0;
var _sanitiseJqlString = require("../utils/sanitise-jql-string");
var _common = require("./common");
function acceptFunctionString(visitor) {
  return visitor.visitFunction ? visitor.visitFunction(this) : visitor.visitChildren(this);
}
function enterNode(listener) {
  listener.enterFunction && listener.enterFunction(this);
}
function exitNode(listener) {
  listener.exitFunction && listener.exitFunction(this);
}
var functionString = exports.functionString = function functionString(value) {
  return functionStringInternal(value, (0, _sanitiseJqlString.sanitiseJqlString)(value));
};
var functionStringInternal = exports.functionStringInternal = function functionStringInternal(value, text) {
  var position = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
  return {
    text: text,
    value: value,
    position: position,
    accept: acceptFunctionString,
    enterNode: enterNode,
    exitNode: exitNode,
    getChildren: _common.noChildren,
    parent: null
  };
};
