// @atlaskit/jql-ast/dist/cjs/creators/keywordOperand.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.keywordOperandInternal = exports.keywordOperand = void 0;
var _constants = require("../constants");
var _common = require("./common");
function acceptKeywordOperand(visitor) {
  return visitor.visitKeywordOperand ? visitor.visitKeywordOperand(this) : visitor.visitChildren(this);
}
function enterNode(listener) {
  listener.enterKeywordOperand && listener.enterKeywordOperand(this);
}
function exitNode(listener) {
  listener.exitKeywordOperand && listener.exitKeywordOperand(this);
}
var keywordOperand = exports.keywordOperand = function keywordOperand() {
  return keywordOperandInternal(_constants.OPERAND_EMPTY);
};
var keywordOperandInternal = exports.keywordOperandInternal = function keywordOperandInternal(value) {
  var position = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
  return {
    type: _constants.NODE_TYPE_OPERAND,
    operandType: _constants.OPERAND_TYPE_KEYWORD,
    value: value,
    position: position,
    accept: acceptKeywordOperand,
    enterNode: enterNode,
    exitNode: exitNode,
    getChildren: _common.noChildren,
    parent: null
  };
};
