// @atlaskit/jql-ast/dist/cjs/creators/listOperand.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.listOperandInternal = exports.listOperand = void 0;
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _constants = require("../constants");
var _listOperand = require("../transformers/listOperand");
var _common = require("./common");
function acceptListOperand(visitor) {
  return visitor.visitListOperand ? visitor.visitListOperand(this) : visitor.visitChildren(this);
}
function enterNode(listener) {
  listener.enterListOperand && listener.enterListOperand(this);
}
function exitNode(listener) {
  listener.exitListOperand && listener.exitListOperand(this);
}
function getChildren() {
  return (0, _toConsumableArray2.default)(this.values);
}
var listOperand = exports.listOperand = function listOperand(values) {
  return listOperandInternal(values);
};
var listOperandInternal = exports.listOperandInternal = function listOperandInternal(values) {
  var position = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
  var node = {
    type: _constants.NODE_TYPE_OPERAND,
    operandType: _constants.OPERAND_TYPE_LIST,
    values: values,
    position: position,
    accept: acceptListOperand,
    enterNode: enterNode,
    exitNode: exitNode,
    getChildren: getChildren,
    parent: null,
    appendOperand: _listOperand.appendOperand
  };
  (0, _common.assignParent)(node);
  return node;
};
