// @atlaskit/jql-ast/dist/cjs/creators/notClauseOperator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.notClauseOperatorInternal = exports.notClauseOperator = void 0;
var _constants = require("../constants");
var _common = require("./common");
function acceptNotClauseOperator(visitor) {
  return visitor.visitNotClauseOperator ? visitor.visitNotClauseOperator(this) : visitor.visitChildren(this);
}
function enterNode(listener) {
  listener.enterNotClauseOperator && listener.enterNotClauseOperator(this);
}
function exitNode(listener) {
  listener.exitNotClauseOperator && listener.exitNotClauseOperator(this);
}
var notClauseOperator = exports.notClauseOperator = function notClauseOperator() {
  return notClauseOperatorInternal();
};
var notClauseOperatorInternal = exports.notClauseOperatorInternal = function notClauseOperatorInternal() {
  var position = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
  return {
    value: _constants.CLAUSE_OPERATOR_NOT,
    position: position,
    accept: acceptNotClauseOperator,
    enterNode: enterNode,
    exitNode: exitNode,
    getChildren: _common.noChildren,
    parent: null
  };
};
