// @atlaskit/jql-ast/dist/cjs/creators/orderBy.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.orderByInternal = exports.orderBy = void 0;
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _constants = require("../constants");
var _orderBy = require("../transformers/orderBy");
var _common = require("./common");
var _orderByOperator = require("./orderByOperator");
function acceptOrderBy(visitor) {
  return visitor.visitOrderBy ? visitor.visitOrderBy(this) : visitor.visitChildren(this);
}
function enterNode(listener) {
  listener.enterOrderBy && listener.enterOrderBy(this);
}
function exitNode(listener) {
  listener.exitOrderBy && listener.exitOrderBy(this);
}
function getChildren() {
  return [this.operator].concat((0, _toConsumableArray2.default)(this.fields));
}
var orderBy = exports.orderBy = function orderBy(fields) {
  return orderByInternal(fields, (0, _orderByOperator.orderByOperator)());
};
var orderByInternal = exports.orderByInternal = function orderByInternal(fields, operator) {
  var position = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
  var node = {
    type: _constants.NODE_TYPE_ORDER_BY,
    operator: operator,
    fields: fields,
    position: position,
    accept: acceptOrderBy,
    enterNode: enterNode,
    exitNode: exitNode,
    getChildren: getChildren,
    parent: null,
    prependOrderField: _orderBy.prependOrderField,
    setOrderDirection: _orderBy.setOrderDirection,
    replace: _orderBy.replace,
    replaceOrderField: _orderBy.replaceOrderField,
    remove: _orderBy.remove,
    removeOrderField: _orderBy.removeOrderField
  };
  (0, _common.assignParent)(node);
  return node;
};
