// @atlaskit/jql-ast/dist/cjs/creators/orderByOperator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.orderByOperatorInternal = exports.orderByOperator = void 0;
var _constants = require("../constants");
var _common = require("./common");
function acceptOrderByOperator(visitor) {
  return visitor.visitOrderByOperator ? visitor.visitOrderByOperator(this) : visitor.visitChildren(this);
}
function enterNode(listener) {
  listener.enterOrderByOperator && listener.enterOrderByOperator(this);
}
function exitNode(listener) {
  listener.exitOrderByOperator && listener.exitOrderByOperator(this);
}
var orderByOperator = exports.orderByOperator = function orderByOperator() {
  return orderByOperatorInternal();
};
var orderByOperatorInternal = exports.orderByOperatorInternal = function orderByOperatorInternal() {
  var position = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
  return {
    value: _constants.ORDER_BY_OPERATOR_ORDER_BY,
    position: position,
    accept: acceptOrderByOperator,
    enterNode: enterNode,
    exitNode: exitNode,
    getChildren: _common.noChildren,
    parent: null
  };
};
