// @atlaskit/jql-ast/dist/cjs/creators/orderByDirection.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.orderByDirectionInternal = exports.orderByDirection = void 0;
var _common = require("./common");
function acceptOrderByDirection(visitor) {
  return visitor.visitOrderByDirection ? visitor.visitOrderByDirection(this) : visitor.visitChildren(this);
}
function enterNode(listener) {
  listener.enterOrderByDirection && listener.enterOrderByDirection(this);
}
function exitNode(listener) {
  listener.exitOrderByDirection && listener.exitOrderByDirection(this);
}
var orderByDirection = exports.orderByDirection = function orderByDirection(value) {
  return orderByDirectionInternal(value);
};
var orderByDirectionInternal = exports.orderByDirectionInternal = function orderByDirectionInternal(value) {
  var position = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
  return {
    value: value,
    position: position,
    accept: acceptOrderByDirection,
    enterNode: enterNode,
    exitNode: exitNode,
    getChildren: _common.noChildren,
    parent: null
  };
};
