// @atlaskit/jql-ast/dist/cjs/creators/orderByField.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.orderByFieldInternal = exports.orderByField = void 0;
var _orderByField = require("../transformers/orderByField");
var _common = require("./common");
function acceptOrderByField(visitor) {
  return visitor.visitOrderByField ? visitor.visitOrderByField(this) : visitor.visitChildren(this);
}
function enterNode(listener) {
  listener.enterOrderByField && listener.enterOrderByField(this);
}
function exitNode(listener) {
  listener.exitOrderByField && listener.exitOrderByField(this);
}
function getChildren() {
  var children = [this.field];
  if (this.direction) {
    children.push(this.direction);
  }
  return children;
}
var orderByField = exports.orderByField = function orderByField(field, direction) {
  return orderByFieldInternal(field, direction);
};
var orderByFieldInternal = exports.orderByFieldInternal = function orderByFieldInternal(field, direction) {
  var position = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
  var node = {
    field: field,
    direction: direction,
    position: position,
    accept: acceptOrderByField,
    enterNode: enterNode,
    exitNode: exitNode,
    getChildren: getChildren,
    parent: null,
    setOrderDirection: _orderByField.setOrderDirection,
    replace: _orderByField.replace,
    remove: _orderByField.remove
  };
  (0, _common.assignParent)(node);
  return node;
};
