// @atlaskit/jql-ast/dist/cjs/creators/predicate.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.predicateInternal = exports.predicate = void 0;
var _common = require("./common");
function acceptPredicate(visitor) {
  return visitor.visitPredicate ? visitor.visitPredicate(this) : visitor.visitChildren(this);
}
function enterNode(listener) {
  listener.enterPredicate && listener.enterPredicate(this);
}
function exitNode(listener) {
  listener.exitPredicate && listener.exitPredicate(this);
}
function getChildren() {
  var children = [this.operator];
  if (this.operand) {
    children.push(this.operand);
  }
  return children;
}
var predicate = exports.predicate = function predicate(operator, operand) {
  return predicateInternal(operator, operand);
};
var predicateInternal = exports.predicateInternal = function predicateInternal(operator, operand) {
  var position = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
  var node = {
    operator: operator,
    operand: operand,
    position: position,
    accept: acceptPredicate,
    enterNode: enterNode,
    exitNode: exitNode,
    getChildren: getChildren,
    parent: null
  };
  (0, _common.assignParent)(node);
  return node;
};
