// @atlaskit/jql-ast/dist/cjs/creators/predicateOperator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.predicateOperatorInternal = exports.predicateOperator = void 0;
var _common = require("./common");
function acceptPredicateOperator(visitor) {
  return visitor.visitPredicateOperator ? visitor.visitPredicateOperator(this) : visitor.visitChildren(this);
}
function enterNode(listener) {
  listener.enterPredicateOperator && listener.enterPredicateOperator(this);
}
function exitNode(listener) {
  listener.exitPredicateOperator && listener.exitPredicateOperator(this);
}
var predicateOperator = exports.predicateOperator = function predicateOperator(value) {
  return predicateOperatorInternal(value, value);
};
var predicateOperatorInternal = exports.predicateOperatorInternal = function predicateOperatorInternal(value, text) {
  var position = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
  return {
    text: text,
    value: value,
    position: position,
    accept: acceptPredicateOperator,
    enterNode: enterNode,
    exitNode: exitNode,
    getChildren: _common.noChildren,
    parent: null
  };
};
