// @atlaskit/editor-common/dist/cjs/mark/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "anyMarkActive", {
  enumerable: true,
  get: function get() {
    return _textFormatting.anyMarkActive;
  }
});
Object.defineProperty(exports, "applyMarkOnRange", {
  enumerable: true,
  get: function get() {
    return _commands.applyMarkOnRange;
  }
});
Object.defineProperty(exports, "entireSelectionContainsMark", {
  enumerable: true,
  get: function get() {
    return _commands.entireSelectionContainsMark;
  }
});
Object.defineProperty(exports, "filterChildrenBetween", {
  enumerable: true,
  get: function get() {
    return _commands.filterChildrenBetween;
  }
});
Object.defineProperty(exports, "isMarkAllowedInRange", {
  enumerable: true,
  get: function get() {
    return _textFormatting.isMarkAllowedInRange;
  }
});
Object.defineProperty(exports, "isMarkExcluded", {
  enumerable: true,
  get: function get() {
    return _textFormatting.isMarkExcluded;
  }
});
Object.defineProperty(exports, "removeMark", {
  enumerable: true,
  get: function get() {
    return _commands.removeMark;
  }
});
Object.defineProperty(exports, "toggleMark", {
  enumerable: true,
  get: function get() {
    return _commands.toggleMark;
  }
});
Object.defineProperty(exports, "transformNonTextNodesToText", {
  enumerable: true,
  get: function get() {
    return _commands.transformNonTextNodesToText;
  }
});
var _commands = require("./commands");
var _textFormatting = require("./text-formatting");
