// @atlaskit/editor-plugin-block-controls/dist/esm/ui/block-menu-items.js
/* eslint-disable @atlaskit/design-system/prefer-primitives */
/**
 * @jsxRuntime classic
 * @jsx jsx
 */

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import { jsx } from '@emotion/react';
import { dragToMoveDown, dragToMoveLeft, dragToMoveRight, dragToMoveUp, tooltip } from '@atlaskit/editor-common/keymaps';
import { blockControlsMessages } from '@atlaskit/editor-common/messages';
import { shortcutStyle } from '@atlaskit/editor-shared-styles/shortcut';
import ArrowDownIcon from '@atlaskit/icon/core/arrow-down';
import ArrowLeftIcon from '@atlaskit/icon/core/arrow-left';
import ArrowRightnIcon from '@atlaskit/icon/core/arrow-right';
import ArrowUpIcon from '@atlaskit/icon/core/arrow-up';
import { moveNodeViaShortcut } from '../editor-commands/move-node';
import { DIRECTION } from '../pm-plugins/utils/consts';
export var getBlockMenuItems = function getBlockMenuItems(formatMessage) {
  return [{
    items: [{
      content: formatMessage(blockControlsMessages.moveUp),
      elemBefore: jsx(ArrowUpIcon, {
        label: ""
      }),
      value: {
        name: 'moveUp'
      },
      key: 'move_up',
      // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values
      elemAfter: jsx("div", {
        css: shortcutStyle
      }, tooltip(dragToMoveUp))
    }, {
      content: formatMessage(blockControlsMessages.moveDown),
      elemBefore: jsx(ArrowDownIcon, {
        label: ""
      }),
      value: {
        name: 'moveDown'
      },
      key: 'move_down',
      // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values
      elemAfter: jsx("div", {
        css: shortcutStyle
      }, tooltip(dragToMoveDown))
    }, {
      content: formatMessage(blockControlsMessages.moveLeft),
      elemBefore: jsx(ArrowLeftIcon, {
        label: ""
      }),
      value: {
        name: 'moveLeft'
      },
      key: 'move_left',
      // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values
      elemAfter: jsx("div", {
        css: shortcutStyle
      }, tooltip(dragToMoveLeft))
    }, {
      content: formatMessage(blockControlsMessages.moveRight),
      elemBefore: jsx(ArrowRightnIcon, {
        label: ""
      }),
      value: {
        name: 'moveRight'
      },
      key: 'move_right',
      // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values
      elemAfter: jsx("div", {
        css: shortcutStyle
      }, tooltip(dragToMoveRight))
    }]
  }];
};
export var menuItemsCallback = {
  moveUp: function moveUp(api, formatMessage) {
    return moveNodeViaShortcut(api, DIRECTION.UP, formatMessage);
  },
  moveDown: function moveDown(api, formatMessage) {
    return moveNodeViaShortcut(api, DIRECTION.DOWN, formatMessage);
  }
};
