// @atlaskit/link-picker/dist/cjs/ui/error-boundary/error-boundary-fallback/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ErrorBoundaryFallback = void 0;
require("./index.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _reactIntlNext = require("react-intl-next");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _constants = require("../../../common/constants");
var _genericErrorSvg = require("../../../common/generic-error-svg");
var _genericErrorSvgV = require("../../../common/generic-error-svg-v2");
var _emptyState = require("../../../common/ui/empty-state");
var _minHeightContainer = require("../../../common/ui/min-height-container");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var errorBoundaryFallbackStyles = null;
var oldErrorBoundaryFallbackStyles = null;
var messages = (0, _reactIntlNext.defineMessages)({
  heading: {
    defaultMessage: 'Something went wrong!',
    description: 'Heading displayed when an unhandled error occurs.',
    id: 'fabric.linkPicker.unhandledError.heading'
  },
  description: {
    defaultMessage: 'Try reloading the page.',
    description: 'Body message shown underneath the heading when an unhandled error occurs.',
    id: 'fabric.linkPicker.unhandledError.description'
  }
});
var ErrorBoundaryFallback = exports.ErrorBoundaryFallback = function ErrorBoundaryFallback() {
  var intl = (0, _reactIntlNext.useIntl)();
  var header = intl.formatMessage(messages.heading);
  var description = intl.formatMessage(messages.description);
  return /*#__PURE__*/React.createElement(_minHeightContainer.MinHeightContainer, {
    minHeight: _constants.LINK_PICKER_MIN_HEIGHT_IN_PX_FALLBACK,
    "data-testid": "link-picker-root-error-boundary-ui",
    className: (0, _runtime.ax)([(0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v1') ? "_11c81vhk" : "_vwz44jg8"])
  }, /*#__PURE__*/React.createElement(_emptyState.EmptyState, {
    header: header,
    renderImage: (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-link-picker') ? function () {
      return /*#__PURE__*/React.createElement(_genericErrorSvgV.GenericErrorSVGV2, null);
    } : function () {
      return /*#__PURE__*/React.createElement(_genericErrorSvg.GenericErrorSVG, null);
    },
    description: description
  }));
};
