// @atlaskit/link-datasource/dist/cjs/ui/common/modal/count-view-smart-link/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SmartLink = exports.SmartCardPlaceholder = void 0;
require("./index.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _react = _interopRequireDefault(require("react"));
var _reactIntlNext = require("react-intl-next");
var _compiled = require("@atlaskit/primitives/compiled");
var _link = _interopRequireDefault(require("../../../issue-like-table/render-type/link"));
var styles = {
  placeholderSmartLinkStyles: "_2rko16jk _syaz1e6v _ca0qze3t _n3tdze3t _19bvv77o _u5f3v77o _16qszucj",
  smartLinkContainerStyles: "_19bvv77o"
};
var SmartCardPlaceholder = exports.SmartCardPlaceholder = function SmartCardPlaceholder(_ref) {
  var placeholderText = _ref.placeholderText;
  return /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    xcss: styles.smartLinkContainerStyles
  }, /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    as: "span",
    testId: "datasource-modal--smart-card-placeholder",
    xcss: styles.placeholderSmartLinkStyles
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
    ,
    style: {
      backgroundColor: "var(--ds-surface-raised, #FFFFFF)"
    }
  }, /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, placeholderText)));
};
var SmartLink = exports.SmartLink = function SmartLink(_ref2) {
  var url = _ref2.url;
  return /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    xcss: styles.smartLinkContainerStyles
  }, /*#__PURE__*/_react.default.createElement(_link.default, {
    url: url
  }));
};
