// @atlaskit/smart-user-picker/dist/cjs/service/teams-client.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _config = require("../config");
var _constants = require("./constants");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var transformTeam = function transformTeam(team, id) {
  return {
    id: id,
    name: team.displayName,
    type: 'team',
    avatarUrl: team.smallAvatarImageUrl
  };
};
var hydrateTeamFromLegion = function hydrateTeamFromLegion(request) {
  var url = _config.config.getTeamsUrl(request.id, request.baseUrl, request.siteId);
  return fetch(url, {
    method: 'GET',
    credentials: 'include',
    headers: {
      'content-type': 'application/json'
    }
  }).then(function (response) {
    if (response.status === 200) {
      return response.json();
    }
    return Promise.reject({
      message: "error calling Legion, statusCode=".concat(response.status, ", statusText=").concat(response.statusText)
    });
  }).then(function (response) {
    return transformTeam(response, request.id);
  }).catch(function () {
    return _objectSpread(_objectSpread({}, _constants.UNKNOWN_TEAM), {}, {
      // on network error, return original team with label 'Unknown'
      id: request.id
    });
  });
};
var _default = exports.default = hydrateTeamFromLegion;
