// @atlaskit/editor-common/dist/cjs/ui/TableSelector/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TableSelectorPopup = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = require("react");
var _react2 = require("@emotion/react");
var _bindEventListener = require("bind-event-listener");
var _editorSharedStyles = require("@atlaskit/editor-shared-styles");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _colors = require("@atlaskit/theme/colors");
var _experiments = require("@atlaskit/tmp-editor-statsig/experiments");
var _uiReact = require("../../ui-react");
var _Popup = _interopRequireDefault(require("../Popup"));
var _tableSelector = _interopRequireWildcard(require("./table-selector"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; } /**
 * @jsxRuntime classic
 * @jsx jsx
 */ // eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
var TABLE_SELECTOR_PADDING_TOP = 8;
var TABLE_SELECTOR_PADDING_SIDE = 10;
var DEFAULT_TABLE_SELECTOR_ROWS = 5;
var DEFAULT_TABLE_SELECTOR_COLS = 10;
var DEFAULT_TABLE_SELECTOR_SELECTION_SIZE = 1;
var DEFAULT_MAX_TABLE_SELECTOR_ROWS = 10;
var TableSelectorWithListeners = (0, _uiReact.withReactEditorViewOuterListeners)(_tableSelector.default);
var initialSizeState = {
  col: DEFAULT_TABLE_SELECTOR_SELECTION_SIZE,
  row: DEFAULT_TABLE_SELECTOR_SELECTION_SIZE,
  maxCol: DEFAULT_TABLE_SELECTOR_COLS,
  maxRow: DEFAULT_TABLE_SELECTOR_ROWS
};
var tableSelectorPopupWrapperStyles = (0, _react2.css)({
  borderRadius: "var(--ds-border-radius, 3px)",
  backgroundColor: "var(--ds-surface-overlay, ".concat(_colors.N0, ")"),
  boxShadow: "var(--ds-shadow-overlay, ".concat("0 0 0 1px ".concat(_colors.N30A, ", 0 2px 1px ").concat(_colors.N30A, ", 0 0 20px -6px ").concat(_colors.N60A), ")"),
  padding: "".concat("var(--ds-space-100, ".concat("".concat(TABLE_SELECTOR_PADDING_TOP, "px"), ")"), " ", TABLE_SELECTOR_PADDING_SIDE, "px")
});
var TableSelectorPopup = exports.TableSelectorPopup = function TableSelectorPopup(props) {
  var _useState = (0, _react.useState)(_objectSpread(_objectSpread({}, initialSizeState), props.defaultSize)),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    size = _useState2[0],
    setSize = _useState2[1];
  var tablePopupRef = (0, _react.useRef)(null);
  // If popup opened by keyboard enable keyboard mode
  var isKeyboardMode = (0, _react.useRef)(props.isOpenedByKeyboard);
  var enableKeyboardMode = (0, _react.useCallback)(function () {
    if (!isKeyboardMode.current) {
      isKeyboardMode.current = true;
    }
  }, [isKeyboardMode]);
  var disableKeyboardMode = (0, _react.useCallback)(function () {
    if (isKeyboardMode.current) {
      isKeyboardMode.current = false;
    }
  }, [isKeyboardMode]);

  // Mouse move is used to allow selection changes outside of the popup and is more reactive to changes.
  var handleMouseMove = (0, _react.useCallback)(function (e) {
    if (!tablePopupRef.current) {
      return;
    }
    disableKeyboardMode();
    var tablePopup = tablePopupRef.current;
    var _tablePopup$getBoundi = tablePopup.getBoundingClientRect(),
      left = _tablePopup$getBoundi.left,
      top = _tablePopup$getBoundi.top;

    // Mouse position on popup
    var selectedWidth = e.clientX - left;
    var selectedHeight = e.clientY - top;

    // Calculate number grid cells selected
    var selectedGridCols = Math.ceil((selectedWidth - TABLE_SELECTOR_PADDING_SIDE + _tableSelector.TABLE_SELECTOR_BUTTON_GAP) / (_tableSelector.TABLE_SELECTOR_BUTTON_GAP + _tableSelector.TABLE_SELECTOR_BUTTON_SIZE));
    var selectedGridRows = Math.ceil((selectedHeight - TABLE_SELECTOR_PADDING_TOP + _tableSelector.TABLE_SELECTOR_BUTTON_GAP) / (_tableSelector.TABLE_SELECTOR_BUTTON_GAP + _tableSelector.TABLE_SELECTOR_BUTTON_SIZE));
    // Keep the selected rows and columns within the defined bounds
    var gridRows = DEFAULT_TABLE_SELECTOR_ROWS;
    if (selectedGridCols < 1) {
      selectedGridCols = 1;
    }
    if (selectedGridCols > size.maxCol) {
      selectedGridCols = size.maxCol;
    }
    if (selectedGridRows < 1) {
      selectedGridRows = 1;
    }

    // Expand grid when row selection is greater than the default grid size
    if (selectedGridRows >= DEFAULT_TABLE_SELECTOR_ROWS && selectedGridRows < DEFAULT_MAX_TABLE_SELECTOR_ROWS) {
      gridRows = selectedGridRows + 1;
    }
    if (selectedGridRows >= DEFAULT_MAX_TABLE_SELECTOR_ROWS) {
      selectedGridRows = DEFAULT_MAX_TABLE_SELECTOR_ROWS;
      gridRows = DEFAULT_MAX_TABLE_SELECTOR_ROWS;
    }
    if (selectedGridCols !== size.col || selectedGridRows !== size.row) {
      setSize({
        col: selectedGridCols,
        row: selectedGridRows,
        maxCol: DEFAULT_TABLE_SELECTOR_COLS,
        maxRow: gridRows
      });
    }
  }, [disableKeyboardMode, size, setSize]);
  var decreasingSequence = function decreasingSequence(maxNumber, prevNumber) {
    var nextNumber = prevNumber - 1;
    if (prevNumber === 1) {
      nextNumber = maxNumber;
    }
    return nextNumber;
  };
  var getMaxRow = function getMaxRow(prevSize, eventKey) {
    switch (eventKey) {
      case 'ArrowDown':
        // Expand the grid size when last row selected
        if (prevSize.maxRow < DEFAULT_MAX_TABLE_SELECTOR_ROWS && prevSize.row >= DEFAULT_TABLE_SELECTOR_ROWS - 1) {
          return prevSize.maxRow + 1;
        }
        if (prevSize.row === DEFAULT_MAX_TABLE_SELECTOR_ROWS) {
          return DEFAULT_TABLE_SELECTOR_ROWS;
        }
        return prevSize.maxRow;
      case 'ArrowLeft':
        var moveToPrevRow = prevSize.col === 1 && prevSize.row > 1;
        var moveToLastRow = prevSize.row === 1 && prevSize.col === 1;
        // Expand the popup to max size when selected row wraps around to last row
        if (moveToLastRow) {
          return DEFAULT_MAX_TABLE_SELECTOR_ROWS;
        }
        // Decrease the popup when decreased row selection
        if (prevSize.maxRow > DEFAULT_TABLE_SELECTOR_ROWS && moveToPrevRow) {
          return prevSize.row;
        }
        return prevSize.maxRow;
      case 'ArrowUp':
        if (prevSize.row === 1) {
          return DEFAULT_MAX_TABLE_SELECTOR_ROWS;
          // Decrease the popup size when decreased row selection
        } else if (prevSize.maxRow > DEFAULT_TABLE_SELECTOR_ROWS) {
          return prevSize.row;
        }
        return prevSize.maxRow;
      case 'ArrowRight':
        var moveToNextRow = prevSize.col === DEFAULT_TABLE_SELECTOR_COLS;
        var increaseMaxRow = prevSize.maxRow < DEFAULT_MAX_TABLE_SELECTOR_ROWS && moveToNextRow && prevSize.row + 1 === prevSize.maxRow;

        // Decrease popup size for wrap around to selection 1 x 1
        if (prevSize.row === DEFAULT_MAX_TABLE_SELECTOR_ROWS && prevSize.col === DEFAULT_TABLE_SELECTOR_COLS) {
          return DEFAULT_TABLE_SELECTOR_ROWS;
          // Decrease the popup size when decreased row selection
        } else if (increaseMaxRow) {
          return prevSize.maxRow + 1;
        }
        return prevSize.maxRow;
      default:
        return prevSize.maxRow;
    }
  };
  var handleInitialButtonFocus = (0, _react.useCallback)(function () {
    if (isKeyboardMode.current !== true) {
      enableKeyboardMode();
      setSize(initialSizeState);
    }
  }, [enableKeyboardMode, setSize]);
  var handleKeyDown = (0, _react.useCallback)(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  function (event) {
    if (event.key === 'ArrowDown') {
      enableKeyboardMode();
      setSize(function (prevSize) {
        return _objectSpread(_objectSpread({}, prevSize), {}, {
          row: prevSize.row % prevSize.maxRow + 1,
          maxRow: getMaxRow(prevSize, event.key)
        });
      });
      if ((0, _experiments.editorExperiment)('platform_editor_controls', 'variant1') && (0, _platformFeatureFlags.fg)('platform_editor_controls_table_picker')) {
        event.preventDefault();
      }
    }
    if (event.key === 'ArrowRight') {
      enableKeyboardMode();
      setSize(function (prevSize) {
        var moveToNextRow = prevSize.col === DEFAULT_TABLE_SELECTOR_COLS;
        return _objectSpread(_objectSpread({}, prevSize), {}, {
          col: prevSize.col % DEFAULT_TABLE_SELECTOR_COLS + 1,
          row: moveToNextRow ? prevSize.row % prevSize.maxRow + 1 : prevSize.row,
          maxRow: getMaxRow(prevSize, event.key)
        });
      });
    }
    if (event.key === 'ArrowLeft') {
      enableKeyboardMode();
      setSize(function (prevSize) {
        var getRow = function getRow(prevRow, prevCol) {
          var row = prevRow;
          // Move to previous row for wrap around
          if (prevSize.col === 1 && prevSize.row > 1) {
            return prevRow - 1;
            // Increase the selection to max size when selected row and column wraps around
          } else if (prevRow === 1 && prevCol === 1) {
            return DEFAULT_MAX_TABLE_SELECTOR_ROWS;
          }
          return row;
        };
        return _objectSpread(_objectSpread({}, prevSize), {}, {
          col: decreasingSequence(prevSize.maxCol, prevSize.col),
          row: getRow(prevSize.row, prevSize.col),
          maxRow: getMaxRow(prevSize, event.key)
        });
      });
    }
    if (event.key === 'ArrowUp') {
      enableKeyboardMode();
      setSize(function (prevSize) {
        var moveToLastRow = prevSize.row === 1;
        return _objectSpread(_objectSpread({}, prevSize), {}, {
          row: moveToLastRow ? DEFAULT_MAX_TABLE_SELECTOR_ROWS : decreasingSequence(prevSize.maxRow, prevSize.row),
          maxRow: getMaxRow(prevSize, event.key)
        });
      });
      if ((0, _experiments.editorExperiment)('platform_editor_controls', 'variant1') && (0, _platformFeatureFlags.fg)('platform_editor_controls_table_picker')) {
        event.preventDefault();
      }
    }
  }, [enableKeyboardMode, setSize]);
  (0, _react.useEffect)(function () {
    var unbind;
    var target = props.allowOutsideSelection ? window : tablePopupRef.current;
    if (target) {
      unbind = (0, _bindEventListener.bind)(target, {
        type: 'mousemove',
        listener: handleMouseMove
      });
    }
    return unbind;
  }, [handleMouseMove, props.allowOutsideSelection, tablePopupRef]);
  return (0, _react2.jsx)(_Popup.default, {
    target: props.target,
    offset: [0, 3],
    mountTo: props.popupsMountPoint,
    boundariesElement: props.popupsBoundariesElement,
    scrollableElement: props.popupsScrollableElement,
    focusTrap: true,
    onUnmount: props.onUnmount,
    zIndex: _editorSharedStyles.akEditorMenuZIndex
  }, (0, _react2.jsx)("div", {
    css: tableSelectorPopupWrapperStyles,
    ref: tablePopupRef
  }, (0, _react2.jsx)(TableSelectorWithListeners, {
    handleClickOutside: props.handleClickOutside,
    handleEscapeKeydown: props.handleEscapeKeydown,
    maxCols: size.maxCol,
    maxRows: size.maxRow,
    onSelection: props.onSelection,
    selectedCol: size.col,
    selectedRow: size.row,
    onKeyDown: handleKeyDown,
    isFocused: isKeyboardMode.current,
    handleInitialButtonFocus: handleInitialButtonFocus
  })));
};
var _default = exports.default = TableSelectorPopup;
