// @atlaskit/datetime-picker/dist/cjs/internal/fixed-layer-menu.js
/* fixed-layer-menu.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FixedLayerMenu = void 0;
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _select = require("@atlaskit/select");
var _fixedLayer = _interopRequireDefault(require("../internal/fixed-layer"));
var _excluded = ["className", "clearValue", "cx", "getStyles", "getValue", "hasValue", "innerProps", "innerRef", "isLoading", "isMulti", "isRtl", "maxMenuHeight", "menuPlacement", "menuPosition", "menuShouldScrollIntoView", "minMenuHeight", "options", "placement", "selectOption", "selectProps", "setValue", "children"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * This is the fixed layer menu used in the time picker.
 */
var FixedLayerMenu = exports.FixedLayerMenu = function FixedLayerMenu(_ref) {
  var className = _ref.className,
    clearValue = _ref.clearValue,
    cx = _ref.cx,
    getStyles = _ref.getStyles,
    getValue = _ref.getValue,
    hasValue = _ref.hasValue,
    innerProps = _ref.innerProps,
    innerRef = _ref.innerRef,
    isLoading = _ref.isLoading,
    isMulti = _ref.isMulti,
    isRtl = _ref.isRtl,
    maxMenuHeight = _ref.maxMenuHeight,
    menuPlacement = _ref.menuPlacement,
    menuPosition = _ref.menuPosition,
    menuShouldScrollIntoView = _ref.menuShouldScrollIntoView,
    minMenuHeight = _ref.minMenuHeight,
    options = _ref.options,
    placement = _ref.placement,
    selectOption = _ref.selectOption,
    selectProps = _ref.selectProps,
    setValue = _ref.setValue,
    children = _ref.children,
    rest = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  return /*#__PURE__*/React.createElement(_fixedLayer.default, {
    inputValue: selectProps.inputValue
    //@ts-ignore react-select unsupported props
    ,
    containerRef: selectProps.fixedLayerRef,
    content: /*#__PURE__*/React.createElement(_select.components.Menu, (0, _extends2.default)({}, rest, {
      // eslint-disable-next-line @atlaskit/design-system/no-unsafe-style-overrides, @atlaskit/ui-styling-standard/no-classname-prop
      className: className,
      clearValue: clearValue,
      cx: cx,
      getStyles: getStyles,
      getValue: getValue,
      hasValue: hasValue,
      innerProps: innerProps,
      innerRef: innerRef,
      isLoading: isLoading,
      isMulti: isMulti,
      isRtl: isRtl,
      maxMenuHeight: maxMenuHeight,
      menuPlacement: menuPlacement,
      menuPosition: menuPosition,
      menuShouldScrollIntoView: false || menuShouldScrollIntoView,
      minMenuHeight: minMenuHeight,
      options: options,
      placement: placement,
      selectOption: selectOption,
      selectProps: selectProps,
      setValue: setValue
    }), children)
    //@ts-ignore react-select unsupported props
    ,
    testId: selectProps.testId
  });
};
