// @atlaskit/media-test-helpers/dist/cjs/mocks/media-mock.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var kakapo_1 = require("kakapo");
var exenv = tslib_1.__importStar(require("exenv"));
var v4_1 = tslib_1.__importDefault(require("uuid/v4"));
var routers_1 = require("./routers");
var database_1 = require("./database");
var utils_1 = require("../utils");
var MediaMock = /** @class */ (function () {
    function MediaMock(collections) {
        this.collections = collections;
        this.server = new kakapo_1.Server();
        this.routers = [];
        this.dbs = [];
    }
    MediaMock.prototype.enable = function (config) {
        if (config === void 0) { config = {}; }
        var isSlowServer = config.isSlowServer;
        if (!exenv.canUseDOM) {
            return;
        }
        this.routers = [
            routers_1.createMediaPlaygroundRouter(),
            routers_1.createApiRouter(isSlowServer),
        ];
        this.dbs = [database_1.createDatabase(this.collections)];
        tslib_1.__spread(this.routers, this.dbs).forEach(this.server.use.bind(this.server));
    };
    MediaMock.prototype.disable = function () {
        tslib_1.__spread(this.routers, this.dbs).forEach(this.server.remove.bind(this.server));
        this.routers.forEach(function (router) { return router.reset(); });
        this.dbs.forEach(function (db) { return db.reset(); });
        this.routers = [];
        this.dbs = [];
    };
    return MediaMock;
}());
exports.MediaMock = MediaMock;
function generateFilesFromTestData(files) {
    return files.map(function (file) {
        var blob = utils_1.mapDataUriToBlob(file.dataUri);
        var id = file.id || v4_1.default();
        var name = file.name || "test-file-" + id;
        return {
            id: id,
            blob: blob,
            mimeType: blob.type,
            mediaType: 'image',
            name: name,
            size: blob.size,
            artifacts: {},
            representations: {
                image: {},
            },
        };
    });
}
exports.generateFilesFromTestData = generateFilesFromTestData;
exports.mediaMock = new MediaMock();
