// @atlaskit/profilecard/dist/cjs/styled/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var components_1 = require("@atlaskit/theme/components");
var colors = tslib_1.__importStar(require("@atlaskit/theme/colors"));
exports.bgColor = components_1.themed({
    light: colors.N0,
    dark: colors.DN50,
});
exports.headerBgColor = components_1.themed({
    light: colors.B500,
    dark: colors.B100,
});
exports.headerBgColorDisabledUser = components_1.themed({
    light: colors.N30,
    dark: colors.B100,
});
exports.headerTextColor = components_1.themed({
    light: colors.N0,
    dark: colors.N0,
});
exports.headerTextColorInactive = components_1.themed({
    light: colors.N800,
    dark: colors.N0,
});
exports.appLabelBgColor = components_1.themed({
    light: colors.N20,
    dark: colors.N20,
});
exports.appLabelTextColor = components_1.themed({
    light: colors.N500,
    dark: colors.N500,
});
exports.labelTextColor = components_1.themed({
    light: colors.N800,
    dark: colors.DN900,
});
exports.labelIconColor = components_1.themed({
    light: colors.N60,
    dark: colors.DN100,
});
exports.errorIconColor = components_1.themed({
    light: colors.N90,
    dark: colors.DN90,
});
exports.errorTitleColor = components_1.themed({
    light: colors.N800,
    dark: colors.DN800,
});
exports.errorTextColor = components_1.themed({
    light: colors.N90,
    dark: colors.DN90,
});
