// @atlaskit/link-extractors/dist/es2019/common/context/extractProvider.js
import React from 'react';
import { ConfluenceIcon } from '@atlaskit/logo/confluence-icon';
import { JiraIcon } from '@atlaskit/logo/jira-icon';
import { CONFLUENCE_GENERATOR_ID, JIRA_GENERATOR_ID } from '../constants';
import { extractUrlFromIconJsonLd, extractUrlFromLinkJsonLd } from '../url';
/**
 * @deprecated Use extractSmartLinkProvider instead
 */
export const extractProvider = jsonLd => {
  const generator = jsonLd.generator;
  if (generator) {
    if (typeof generator === 'string') {
      throw Error('Link.generator requires a name and icon.');
    } else if (generator['@type'] === 'Link') {
      if (generator.name) {
        return {
          text: generator.name
        };
      }
    } else {
      if (generator.name) {
        const id = generator['@id'];
        return {
          text: generator.name,
          icon: extractProviderIcon(generator.icon, id),
          id,
          image: extractProviderImage(generator.image)
        };
      }
    }
  }
};
export const extractProviderIcon = (icon, id) => {
  if (id) {
    if (id === CONFLUENCE_GENERATOR_ID) {
      return /*#__PURE__*/React.createElement(ConfluenceIcon, {
        appearance: "brand",
        size: "xsmall"
      });
    } else if (id === JIRA_GENERATOR_ID) {
      return /*#__PURE__*/React.createElement(JiraIcon, {
        appearance: "brand",
        size: "xsmall"
      });
    }
  }
  if (icon) {
    return extractUrlFromIconJsonLd(icon);
  }
};
const extractProviderImage = image => {
  if (image) {
    if (typeof image === 'string') {
      return image;
    } else if (image['@type'] === 'Link') {
      return extractUrlFromLinkJsonLd(image);
    } else if (image['@type'] === 'Image') {
      if (image.url) {
        return extractUrlFromLinkJsonLd(image.url);
      }
    }
  }
};
export const isConfluenceGenerator = id => {
  return id === CONFLUENCE_GENERATOR_ID;
};
