// @atlaskit/media-card/dist/esm/styles/index.js
import { __makeTemplateObject } from "tslib";
import styled from 'styled-components';
import { fontFamily } from '@atlaskit/theme/constants';
import { fadeIn } from '@atlaskit/media-ui';
export { defaultTransitionDuration } from './config';
export { antialiased, borderRadiusLeft, capitalize, centerSelf, centerSelfX, centerSelfY, centerX, hexToRgb, rgba, spaceAround, transition, withAppearance, } from './mixins';
export { easeOutCubic, easeOutExpo } from './easing';
export { spin } from './animations';
export var Root = styled.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  box-sizing: border-box;\n  font-family: ", ";\n\n  * {\n    box-sizing: border-box;\n  }\n"], ["\n  box-sizing: border-box;\n  font-family: ", ";\n\n  * {\n    box-sizing: border-box;\n  }\n"])), fontFamily());
export var cardShadow = "\n  box-shadow: 0 1px 1px rgba(9, 30, 66, 0.2), 0 0 1px 0 rgba(9, 30, 66, 0.24);\n";
export var FadeinImage = styled.div(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  ", ";\n"], ["\n  ", ";\n"])), fadeIn);
export default Root;
var templateObject_1, templateObject_2;
