// @atlaskit/media-filmstrip/dist/es2019/filmstripView/wrappers-compiled.js
/* wrappers-compiled.tsx generated by @compiled/babel-plugin v0.36.1 */
import "./wrappers-compiled.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
import React, { forwardRef } from 'react';
import ArrowLeft from '@atlaskit/icon/core/migration/arrow-left';
import ArrowRight from '@atlaskit/icon/core/migration/arrow-right';
import { MediaFilmStripListItemSelector } from '.';
import { N20, N40, B400, B50 } from '@atlaskit/theme/colors';
const filmStripViewStyles = null;
const filmStripListWrapperStyles = null;
const filmStripListStyles = null;
const filmStripListItemStyles = null;
const arrowWrapperStyles = null;
const arrowLeftWrapperStyles = null;
const arrowRightWrapperStyles = null;
const shadowStyles = null;
const shadowLeftStyles = null;
const shadowRightStyles = null;
const ShadowLeft = ({
  children
}) => /*#__PURE__*/React.createElement("div", {
  className: ax(["_kqswstnw _1pbyzo7u _4t3i1osq _154iidpf _1bsbyh40 _bfhk1tuc", "_1ltvidpf"])
}, children);
export const ArrowLeftWrapper = ({
  children,
  onClick
}) => /*#__PURE__*/React.createElement("div", {
  className: ax(["_2rko1osq _v564g17y _kqswstnw _154i1ssb _t9ec1i34 _bfhkexu1 _1e0c1txw _80omtlke _16qsztl5 _syazgpis _1bsb1ul9 _4t3i1ul9 _1bah1h6o _tzy4idpf _17521ul9 _w8l5gktf _30l3lyvy _irr3376n _9h8h1ldt _1di629zg", "_1ltvu2gc _uhasv77o", "arrow"]),
  onClick: onClick
}, children);
export const ShadowRight = ({
  children
}) => /*#__PURE__*/React.createElement("div", {
  className: ax(["_kqswstnw _1pbyzo7u _4t3i1osq _154iidpf _1bsbyh40 _bfhk1tuc", "_1xi2idpf"])
}, children);
export const ArrowRightWrapper = ({
  children,
  onClick
}) => /*#__PURE__*/React.createElement("div", {
  className: ax(["_2rko1osq _v564g17y _kqswstnw _154i1ssb _t9ec1i34 _bfhkexu1 _1e0c1txw _80omtlke _16qsztl5 _syazgpis _1bsb1ul9 _4t3i1ul9 _1bah1h6o _tzy4idpf _17521ul9 _w8l5gktf _30l3lyvy _irr3376n _9h8h1ldt _1di629zg", "_1xi2u2gc _7wy9v77o", "arrow"]),
  onClick: onClick
}, children);
export const LeftArrow = ({
  onClick
}) => /*#__PURE__*/React.createElement(ShadowLeft, null, /*#__PURE__*/React.createElement(ArrowLeftWrapper, {
  onClick: onClick
}, /*#__PURE__*/React.createElement(ArrowLeft, {
  color: "currentColor",
  spacing: "spacious",
  label: "left"
})));
export const RightArrow = ({
  onClick
}) => /*#__PURE__*/React.createElement(ShadowRight, null, /*#__PURE__*/React.createElement(ArrowRightWrapper, {
  onClick: onClick
}, /*#__PURE__*/React.createElement(ArrowRight, {
  color: "currentColor",
  spacing: "spacious",
  label: "right"
})));
export const FilmStripViewWrapper = ({
  children,
  'data-testid': dataTestId
}) => /*#__PURE__*/React.createElement("div", {
  "data-testid": dataTestId,
  className: ax(["_2rko1l7b _kqswh2mm _ca0q1ysw _u5f3idpf _n3td1ysw _19bvidpf _1yqxkb7n _1ajs4jg8"])
}, children);
export const FilmStripListWrapper = /*#__PURE__*/forwardRef(({
  children,
  onWheel,
  onTouchStart,
  onTouchMove,
  onTouchEnd,
  'data-testid': dataTestId
}, ref) => /*#__PURE__*/React.createElement("div", {
  ref: ref,
  onWheel: onWheel,
  onTouchStart: onTouchStart,
  onTouchMove: onTouchMove,
  onTouchEnd: onTouchEnd,
  "data-testid": dataTestId,
  className: ax(["_1reo15vq _18m915vq _1bsb1kw7 _ca0qv77o _u5f31ysw _n3tdv77o _19bv1ysw"])
}, children));
export const FilmStripList = /*#__PURE__*/React.forwardRef(({
  children,
  style
}, ref) => /*#__PURE__*/React.createElement("ul", {
  ref: ref
  // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
  ,
  style: style,
  className: ax(["_19pkidpf _2hwxidpf _otyridpf _18u0idpf _ca0qidpf _u5f3idpf _n3tdidpf _19bvidpf _k8m01e03 _6fl41kli _o5721q9c _1e0c1o8l"])
}, children));
export const FilmStripListItem = ({
  children,
  index
}) => /*#__PURE__*/React.createElement("li", {
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
  className: ax(["_19pkidpf _2hwxidpf _otyridpf _18u0idpf _2mzuglyw _ca0qidpf _u5f31b66 _n3tdidpf _19bv1b66 _1e0c1o8l _s7n4nkob _1wybidpf _1lhcidpf", MediaFilmStripListItemSelector]),
  "data-testid": "media-filmstrip-list-item",
  key: index
}, children);
