// @atlaskit/smart-card/dist/es2019/view/EmbedCard/components/ErrorFrame.js
/* ErrorFrame.tsx generated by @compiled/babel-plugin v0.36.1 */
import "./ErrorFrame.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
import React from 'react';
import { B200, N20A, N30A, N40A, N50A } from '@atlaskit/theme/colors';
export const Frame = (props = {
  isSelected: false,
  isHoverable: false,
  isFluidHeight: false
}) => props.compact ? /*#__PURE__*/React.createElement(CompactFrame, props) : /*#__PURE__*/React.createElement(ExpandedFrame, props);
const sharedBaseFrameStyles = null;
const sharedFrameStyles = null;
const expandedFrameHoverStyles = null;
const expandedFrameFluidHeightTrueStyles = null;
const expandedFrameFluidHeightFalseStyles = null;
const expandedFrameSelectedStyles = null;
const expandedFrameNotSelectedStyles = null;
const expandedFrameStyles = null;
export const ExpandedFrame = ({
  children,
  isSelected,
  isHoverable,
  testId,
  className,
  isFluidHeight
}) => {
  return /*#__PURE__*/React.createElement("div", {
    "data-testid": testId
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    ,
    className: ax(["_1bsb1osq _1e0c1txw", "_p12fukw8 _bfhkhp5a", isHoverable && "_irr3h3i3 _d0altlke", isFluidHeight ? "_1tkeidpf" : "_1tke68cl", isSelected ? "_2rko1l7b _1h6d1can _1dqonqa1 _189eyh40" : "_2rko1lya _1h6d1j28 _1dqonqa1 _189eyh40", "_1reo15vq _18m915vq _1bah1yb4 _16qsy59s", className]),
    "data-trello-do-not-use-override": testId
  }, children);
};
const compactFrameHoverStyles = null;
const compactFrameInheritDimensionsTrueStyles = null;
const compactFrameInheritDimensionsFalseStyles = null;
const compactFrameAlignStyles = null;
const compactFrameStyles = null;
const compactFrameNotSelectedStyles = null;
export const CompactFrame = ({
  children,
  isHoverable,
  isSelected,
  testId,
  className,
  inheritDimensions
}) => {
  return /*#__PURE__*/React.createElement("div", {
    "data-testid": testId
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    ,
    className: ax(["_1bsb1osq _1e0c1txw", "_p12fukw8 _bfhkhp5a", isHoverable && "_irr317ks", isSelected && "_2rko1l7b _1h6d1can _1dqonqa1 _189eyh40", !isSelected && "_2rko1lya", "_1bah1h6o _4cvr1h6o _bfhk1j54", inheritDimensions ? "_4t3i1osq" : "_4t3i1ylp", "_ca0qidpf _u5f3idpf _n3tdidpf _19bvidpf _1bsb1osq", className])
  }, children);
};
