// @atlaskit/smart-card/dist/es2019/view/EmbedCard/components/styled.js
/* styled.tsx generated by @compiled/babel-plugin v0.36.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./styled.compiled.css";
import { forwardRef } from 'react';
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
/* eslint-disable @atlaskit/ui-styling-standard/no-dynamic-styles */

import { B100, B50, N30, N300, N40 } from '@atlaskit/theme/colors';
export const className = 'media-card-frame';
export const embedHeaderHeight = 32;
// TODO Delete when cleaning platform-linking-visual-refresh-v1
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled,@atlaskit/ui-styling-standard/no-exported-styles
export const LinkWrapperOldVisualRefresh = forwardRef(({
  as: C = "div",
  style: __cmpls,
  ...__cmplp
}, __cmplr) => {
  const {
    isSelected,
    frameStyle,
    minWidth,
    maxWidth,
    isInteractive,
    inheritDimensions,
    ...__cmpldp
  } = __cmplp;
  return /*#__PURE__*/React.createElement(C, _extends({}, __cmpldp, {
    style: {
      ...__cmpls,
      "--_10q0298": ix(__cmplp.minWidth, "px"),
      "--_1et2rhb": ix(__cmplp.maxWidth, "px")
    },
    ref: __cmplr,
    className: ax(["_19pkidpf _2hwx1wug _otyridpf _18u01wug _v5641y8x _1e0c116y _2lx21bp4 _vchhusvi _ect41gqc _1bsb1osq _uiztglyw _vwz44jg8 _kqswh2mm _qc5o94zs _14mjpd34 _aetrb3bt _18postnw _1qdg1rgy _9oik1r31 _1bnxglyw _jf4cnqa1", __cmplp.isSelected && __cmplp.frameStyle === 'hide' ? "_2rkoiti9" : "_2rkopd34", __cmplp.minWidth && "_1ul91pd6", __cmplp.maxWidth && "_p12f1ayu", __cmplp.isInteractive && __cmplp.frameStyle !== 'hide' && "_1u3rh6cr", __cmplp.isInteractive && __cmplp.frameStyle !== 'hide' && "_4tpuhp5a", __cmplp.isInteractive && __cmplp.frameStyle !== 'hide' ? "_x148kb7n" : "_x148idpf", __cmplp.isInteractive && "_1di629zg", __cmplp.frameStyle === 'show' || __cmplp.isSelected && __cmplp.frameStyle !== 'hide' ? "_1ifikb7n" : "_1ifiidpf", __cmplp.inheritDimensions ? "_4t3i1osq" : "_4t3ik5xz", __cmplp.frameStyle === 'show' && "_631ousvi", __cmplp.frameStyle === 'show' ? "_z0ai1osq" : "_z0aie9lw", __cmplp.frameStyle !== 'show' ? "_19dox0bf" : __cmplp.isSelected ? "_19doidpf" : "_19do1kw7", __cmplp.frameStyle === 'show' || __cmplp.isSelected && __cmplp.frameStyle !== 'hide' ? "_eq43h6cr" : "_eq431kw7", __cmplp.frameStyle === 'show' || __cmplp.isSelected && __cmplp.frameStyle !== 'hide' ? "_1hfkhp5a" : "_1hfk1kw7", __cmplp.isSelected && __cmplp.frameStyle !== 'hide' ? "_6syztlke" : "_6syz1kw7", __cmplp.isSelected && __cmplp.frameStyle !== 'hide' ? "_8x3umqxb" : "_8x3u1kw7", __cmplp.className])
  }));
});

// TODO Delete when cleaning platform-linking-visual-refresh-v1
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled,@atlaskit/ui-styling-standard/no-exported-styles
if (process.env.NODE_ENV !== 'production') {
  LinkWrapperOldVisualRefresh.displayName = 'LinkWrapperOldVisualRefresh';
}
export const WrapperOldVisualRefresh = forwardRef(({
  as: C = "div",
  style: __cmpls,
  ...__cmplp
}, __cmplr) => {
  const {
    isSelected,
    frameStyle,
    minWidth,
    maxWidth,
    isInteractive,
    inheritDimensions,
    ...__cmpldp
  } = __cmplp;
  return /*#__PURE__*/React.createElement(C, _extends({}, __cmpldp, {
    style: {
      ...__cmpls,
      "--_10q0298": ix(__cmplp.minWidth, "px"),
      "--_1et2rhb": ix(__cmplp.maxWidth, "px")
    },
    ref: __cmplr,
    className: ax(["_19pkidpf _2hwx1wug _otyridpf _18u01wug _v5641y8x _1e0c116y _2lx21bp4 _vchhusvi _ect41gqc _1bsb1osq _uiztglyw _vwz44jg8 _kqswh2mm _19pk19bv _qc5o94zs _14mjpd34 _aetrb3bt _18postnw _1qdg1rgy", __cmplp.isSelected && __cmplp.frameStyle === 'hide' ? "_2rkoiti9" : "_2rkopd34", __cmplp.minWidth && "_1ul91pd6", __cmplp.maxWidth && "_p12f1ayu", __cmplp.isInteractive && __cmplp.frameStyle !== 'hide' && "_1u3rh6cr", __cmplp.isInteractive && __cmplp.frameStyle !== 'hide' && "_4tpuhp5a", __cmplp.isInteractive && __cmplp.frameStyle !== 'hide' ? "_x148kb7n" : "_x148idpf", __cmplp.isInteractive && "_1di629zg", __cmplp.frameStyle === 'show' || __cmplp.isSelected && __cmplp.frameStyle !== 'hide' ? "_1ifikb7n" : "_1ifiidpf", __cmplp.inheritDimensions ? "_4t3i1osq" : "_4t3ik5xz", __cmplp.frameStyle === 'show' && "_631ousvi", __cmplp.frameStyle === 'show' ? "_z0ai1osq" : "_z0aie9lw", __cmplp.frameStyle !== 'show' ? "_19dox0bf" : __cmplp.isSelected ? "_19doidpf" : "_19do1kw7", __cmplp.frameStyle === 'show' || __cmplp.isSelected && __cmplp.frameStyle !== 'hide' ? "_eq43h6cr" : "_eq431kw7", __cmplp.frameStyle === 'show' ? "_1hfkhp5a" : "_1hfk1kw7", __cmplp.isSelected && __cmplp.frameStyle !== 'hide' ? "_6syztlke" : "_6syz1kw7", __cmplp.isSelected && __cmplp.frameStyle !== 'hide' ? "_8x3umqxb" : "_8x3u1kw7", __cmplp.className])
  }));
});

// TODO Delete when cleaning platform-linking-visual-refresh-v1
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled,@atlaskit/ui-styling-standard/no-exported-styles
if (process.env.NODE_ENV !== 'production') {
  WrapperOldVisualRefresh.displayName = 'WrapperOldVisualRefresh';
}
export const HeaderOldVisualRefresh = forwardRef(({
  as: C = "div",
  style: __cmpls,
  ...__cmplp
}, __cmplr) => {
  const {
    frameStyle,
    ...__cmpldp
  } = __cmplp;
  return /*#__PURE__*/React.createElement(C, _extends({}, __cmpldp, {
    style: __cmpls,
    ref: __cmplr,
    className: ax(["_v564ojwh _4t3izwfg _kqswstnw _1pbykb7n _1bsb1osq _1e0c1txw _4cvr1h6o _syazqjbk _tzy4idpf", __cmplp.frameStyle === 'show' && "_vchhusvi", __cmplp.frameStyle === 'show' && "_1yt4h8dc", __cmplp.className])
  }));
});

// TODO Delete when cleaning platform-linking-visual-refresh-v1
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled,@atlaskit/ui-styling-standard/no-exported-styles
if (process.env.NODE_ENV !== 'production') {
  HeaderOldVisualRefresh.displayName = 'HeaderOldVisualRefresh';
}
export const IconWrapperOldVisualRefresh = forwardRef(({
  as: C = "div",
  style: __cmpls,
  ...__cmplp
}, __cmplr) => {
  const {
    isPlaceholder,
    ...__cmpldp
  } = __cmplp;
  return /*#__PURE__*/React.createElement(C, _extends({}, __cmpldp, {
    style: __cmpls,
    ref: __cmplr,
    className: ax(["_2rko1sit _1bsb7vkz _4t3i7vkz _2hwx1b66", __cmplp.isPlaceholder && "_bfhkqrzy", __cmplp.className])
  }));
});

// TODO Delete when cleaning platform-linking-visual-refresh-v1
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled,@atlaskit/ui-styling-standard/no-exported-styles
if (process.env.NODE_ENV !== 'production') {
  IconWrapperOldVisualRefresh.displayName = 'IconWrapperOldVisualRefresh';
}
export const TextWrapperOldVisualRefresh = forwardRef(({
  as: C = "div",
  style: __cmpls,
  ...__cmplp
}, __cmplr) => {
  const {
    isPlaceholder,
    ...__cmpldp
  } = __cmplp;
  return /*#__PURE__*/React.createElement(C, _extends({}, __cmpldp, {
    style: __cmpls,
    ref: __cmplr,
    className: ax(["_11c8dcr7 _1reo15vq _18m915vq _syaz1rc1 _1bto1l2s _o5721q9c", __cmplp.isPlaceholder && "_2rko1sit", __cmplp.isPlaceholder && "_1bsb176v", __cmplp.isPlaceholder && "_4t3i1crf", __cmplp.isPlaceholder && "_bfhkqrzy", __cmplp.className])
  }));
});

// TODO Delete when cleaning platform-linking-visual-refresh-v1
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled,@atlaskit/ui-styling-standard/no-exported-styles
if (process.env.NODE_ENV !== 'production') {
  TextWrapperOldVisualRefresh.displayName = 'TextWrapperOldVisualRefresh';
}
export const TooltipWrapperOldVisualRefresh = forwardRef(({
  as: C = "div",
  style: __cmpls,
  ...__cmplp
}, __cmplr) => {
  return /*#__PURE__*/React.createElement(C, _extends({}, __cmplp, {
    style: __cmpls,
    ref: __cmplr,
    className: ax(["_1reo15vq _18m915vq", __cmplp.className])
  }));
});

// TODO Delete when cleaning platform-linking-visual-refresh-v1
// NB: `overflow` is kept as `hidden` since
// the internal contents of the `iframe` should
// manage scrolling behaviour.
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled,@atlaskit/ui-styling-standard/no-exported-styles
if (process.env.NODE_ENV !== 'production') {
  TooltipWrapperOldVisualRefresh.displayName = 'TooltipWrapperOldVisualRefresh';
}
export const ContentOldVisualRefresh = forwardRef(({
  as: C = "div",
  style: __cmpls,
  ...__cmplp
}, __cmplr) => {
  const {
    removeOverflow,
    allowScrollBar,
    frameStyle,
    ...__cmpldp
  } = __cmplp;
  return /*#__PURE__*/React.createElement(C, _extends({}, __cmpldp, {
    style: __cmpls,
    ref: __cmplr,
    className: ax(["_2rkoiti9 _19ith6cr _v56414ba _bfhkhp5a _kqswstnw _1pbykb7n _hlq8idpf _rd9b1wug _1g4jidpf _jk3p1wug", !__cmplp.removeOverflow && (__cmplp.allowScrollBar ? "_1reo1wug _18m91wug" : "_1reo15vq _18m915vq"), __cmplp.frameStyle === 'show' ? "_1bsbwgg2" : "_1bsb1osq", __cmplp.frameStyle === 'show' && "_18s81xf9", __cmplp.frameStyle === 'hide' ? "_4t3i1osq" : "_4t3i1j2a", __cmplp.frameStyle !== 'hide' && "_154ixy5q", __cmplp.className])
  }));
});
if (process.env.NODE_ENV !== 'production') {
  ContentOldVisualRefresh.displayName = 'ContentOldVisualRefresh';
}
const getSizeWithUnit = (value = '100%') => {
  const unit = typeof value === 'number' ? 'px' : '';
  return `${value}${unit}`;
};

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled,@atlaskit/ui-styling-standard/no-exported-styles
export const Image = forwardRef(({
  as: C = "img",
  style: __cmpls,
  ...__cmplp
}, __cmplr) => {
  return /*#__PURE__*/React.createElement(C, _extends({}, __cmplp, {
    style: {
      ...__cmpls,
      "--_rcxkve": ix(getSizeWithUnit(__cmplp.size))
    },
    ref: __cmplr,
    className: ax(["_2rko1sit _1reo15vq _18m915vq _1bsb17ym _4t3i17ym", __cmplp.className])
  }));
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled,@atlaskit/ui-styling-standard/no-exported-styles
if (process.env.NODE_ENV !== 'production') {
  Image.displayName = 'Image';
}
export const Thumbnail = forwardRef(({
  as: C = "div",
  style: __cmpls,
  ...__cmplp
}, __cmplr) => {
  return /*#__PURE__*/React.createElement(C, _extends({}, __cmplp, {
    style: {
      ...__cmpls,
      "--_13orr8u": ix(`url(${__cmplp.src})`)
    },
    ref: __cmplr,
    className: ax(["_2rko1sit _1bsbckbl _4t3ickbl _1aaxusic _19pk1b66 _2hwxidpf _otyrutpp _18u0utpp _bfhkqrzy _1lrw1dfr _1itk2pcs", __cmplp.className])
  }));
});
if (process.env.NODE_ENV !== 'production') {
  Thumbnail.displayName = 'Thumbnail';
}
