// @atlaskit/link-datasource/dist/cjs/ui/assets-modal/modal/render-assets-content/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RenderAssetsContent = exports.MODAL_HEIGHT = void 0;
require("./index.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _colors = require("@atlaskit/theme/colors");
var _accessRequired = require("../../../common/error-state/access-required");
var _modalLoadingError = require("../../../common/error-state/modal-loading-error");
var _noResults = require("../../../common/error-state/no-results");
var _issueLikeTable = require("../../../issue-like-table");
var _initialStateView = require("./initial-state-view");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var MODAL_HEIGHT = exports.MODAL_HEIGHT = 420;

// This is to prevent y scrollbar when showing table loading state
var disableOverflowStyles = null;
var contentContainerStyles = null;
var tableBordersStyles = null;
var RejectedView = function RejectedView() {
  return /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)(["_1reo1wug _18m91wug _4t3i1pna _1e0c11p5"])
  }, /*#__PURE__*/React.createElement(_modalLoadingError.ModalLoadingError, null));
};
var UnauthorizedView = function UnauthorizedView() {
  return /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)(["_1reo1wug _18m91wug _4t3i1pna _1e0c11p5"])
  }, /*#__PURE__*/React.createElement(_accessRequired.AccessRequired, null));
};
var EmptyView = function EmptyView() {
  return /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)(["_1reo1wug _18m91wug _4t3i1pna _1e0c11p5"])
  }, /*#__PURE__*/React.createElement(_initialStateView.InitialStateView, null));
};
var NoResultsView = function NoResultsView() {
  return /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)(["_1reo1wug _18m91wug _4t3i1pna _1e0c11p5"])
  }, /*#__PURE__*/React.createElement(_noResults.NoResults, null));
};
var LoadingView = function LoadingView() {
  return /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)(["_1reo1wug _18m91wug _4t3i1pna _1e0c11p5", "_1reo1wug _18m91wug _4t3i1pna _1e0c11p5", "_19itn8v0 _n7zl1oe8 _13lipd34 _qrwqpd34 _1itk1bg1 _12vemgnk _1lrw1sd4 _qmxj1c00", "_1reo15vq _18m915vq"]),
    style: {
      "--_vzvmm2": (0, _runtime.ix)("1px solid ".concat("var(--ds-border, ".concat(_colors.N40, ")"))),
      "--_cfwo9c": (0, _runtime.ix)("2px solid ".concat("var(--ds-background-accent-gray-subtler, ".concat(_colors.N40, ")"))),
      "--_17x9dh1": (0, _runtime.ix)("\n\t\tlinear-gradient(90deg, ".concat("var(--ds-elevation-surface-current, #FFF)", " 30%, rgba(255, 255, 255, 0)),\n\t\tlinear-gradient(90deg, ", "var(--ds-shadow-overflow-perimeter, rgba(0, 0, 0, 0.1))", ", rgba(0, 0, 0, 0)),\n\t\tlinear-gradient(90deg, rgba(255, 255, 255, 0), ", "var(--ds-elevation-surface-current, #FFF)", " 70%),\n\t\tlinear-gradient(90deg, rgba(0, 0, 0, 0), ", "var(--ds-shadow-overflow-perimeter, rgba(0, 0, 0, 0.1))", "])\n\t\tlinear-gradient(0deg, rgba(255, 255, 255, 0),  ", "var(--ds-elevation-surface-current, #FFF)", " 30%),\n\t\tlinear-gradient(0deg, rgba(0, 0, 0, 0), ", "var(--ds-shadow-overflow-perimeter, rgba(0, 0, 0, 0.05))", "),\n\t\tlinear-gradient(0deg, ", "var(--ds-elevation-surface-current, #FFF)", " 30%, rgba(255, 255, 255, 0)),\n\t\tlinear-gradient(0deg, ", "var(--ds-shadow-overflow-perimeter, rgba(0, 0, 0, 0.05))", ", rgba(0, 0, 0, 0))\n\t\t"))
    }
  }, /*#__PURE__*/React.createElement(_issueLikeTable.EmptyState, {
    testId: "assets-aql-datasource-modal--loading-state"
  }));
};
var RenderAssetsContent = exports.RenderAssetsContent = function RenderAssetsContent(props) {
  var status = props.status,
    responseItems = props.responseItems,
    responseItemIds = props.responseItemIds,
    visibleColumnKeys = props.visibleColumnKeys,
    onNextPage = props.onNextPage,
    hasNextPage = props.hasNextPage,
    loadDatasourceDetails = props.loadDatasourceDetails,
    columns = props.columns,
    defaultVisibleColumnKeys = props.defaultVisibleColumnKeys,
    onVisibleColumnKeysChange = props.onVisibleColumnKeysChange,
    isFetchingInitialData = props.isFetchingInitialData;
  var resolvedWithNoResults = status === 'resolved' && !responseItems.length;
  var issueLikeDataTableView = (0, _react.useMemo)(function () {
    return /*#__PURE__*/React.createElement("div", {
      className: (0, _runtime.ax)(["_1reo1wug _18m91wug _4t3i1pna _1e0c11p5", "_19itn8v0 _n7zl1oe8 _13lipd34 _qrwqpd34 _1itk1bg1 _12vemgnk _1lrw1sd4 _qmxj1c00"]),
      style: {
        "--_vzvmm2": (0, _runtime.ix)("1px solid ".concat("var(--ds-border, ".concat(_colors.N40, ")"))),
        "--_cfwo9c": (0, _runtime.ix)("2px solid ".concat("var(--ds-background-accent-gray-subtler, ".concat(_colors.N40, ")"))),
        "--_17x9dh1": (0, _runtime.ix)("\n\t\tlinear-gradient(90deg, ".concat("var(--ds-elevation-surface-current, #FFF)", " 30%, rgba(255, 255, 255, 0)),\n\t\tlinear-gradient(90deg, ", "var(--ds-shadow-overflow-perimeter, rgba(0, 0, 0, 0.1))", ", rgba(0, 0, 0, 0)),\n\t\tlinear-gradient(90deg, rgba(255, 255, 255, 0), ", "var(--ds-elevation-surface-current, #FFF)", " 70%),\n\t\tlinear-gradient(90deg, rgba(0, 0, 0, 0), ", "var(--ds-shadow-overflow-perimeter, rgba(0, 0, 0, 0.1))", "])\n\t\tlinear-gradient(0deg, rgba(255, 255, 255, 0),  ", "var(--ds-elevation-surface-current, #FFF)", " 30%),\n\t\tlinear-gradient(0deg, rgba(0, 0, 0, 0), ", "var(--ds-shadow-overflow-perimeter, rgba(0, 0, 0, 0.05))", "),\n\t\tlinear-gradient(0deg, ", "var(--ds-elevation-surface-current, #FFF)", " 30%, rgba(255, 255, 255, 0)),\n\t\tlinear-gradient(0deg, ", "var(--ds-shadow-overflow-perimeter, rgba(0, 0, 0, 0.05))", ", rgba(0, 0, 0, 0))\n\t\t"))
      }
    }, /*#__PURE__*/React.createElement(_issueLikeTable.IssueLikeDataTableView, {
      testId: "asset-datasource-table",
      status: status,
      columns: columns,
      items: responseItems,
      itemIds: responseItemIds,
      hasNextPage: hasNextPage,
      visibleColumnKeys: visibleColumnKeys || defaultVisibleColumnKeys,
      onNextPage: onNextPage,
      onLoadDatasourceDetails: loadDatasourceDetails,
      onVisibleColumnKeysChange: onVisibleColumnKeysChange
    }));
  }, [columns, defaultVisibleColumnKeys, hasNextPage, loadDatasourceDetails, onNextPage, onVisibleColumnKeysChange, responseItems, responseItemIds, status, visibleColumnKeys]);
  var renderAssetsContentView = (0, _react.useCallback)(function () {
    if (isFetchingInitialData) {
      // Placing this check first as it's a priority before all others
      return /*#__PURE__*/React.createElement(LoadingView, null);
    } else if (status === 'rejected') {
      return /*#__PURE__*/React.createElement(RejectedView, null);
    } else if (status === 'unauthorized') {
      return /*#__PURE__*/React.createElement(UnauthorizedView, null);
    } else if (status === 'empty') {
      return /*#__PURE__*/React.createElement(EmptyView, null);
    } else if (resolvedWithNoResults) {
      return /*#__PURE__*/React.createElement(NoResultsView, null);
    } else if (status === 'loading' && !columns.length) {
      return /*#__PURE__*/React.createElement(LoadingView, null);
    }
    return issueLikeDataTableView;
  }, [columns.length, isFetchingInitialData, issueLikeDataTableView, resolvedWithNoResults, status]);
  return renderAssetsContentView();
};
