// @atlaskit/media-client-react/dist/cjs/test-helpers/MockedMediaClientProvider.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mockedMediaClientConfig = exports.MockedMediaClientProvider = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _react = _interopRequireWildcard(require("react"));
var _mediaClient = require("@atlaskit/media-client");
var _mediaState = require("@atlaskit/media-state");
var _MediaClientProvider = require("../MediaClientProvider");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var mockedMediaClientConfig = exports.mockedMediaClientConfig = {
  authProvider: function () {
    var _authProvider = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee() {
      return _regenerator.default.wrap(function _callee$(_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            return _context.abrupt("return", {
              clientId: 'MockedMediaClientProvider-client-id',
              token: 'MockedMediaClientProvider-token',
              baseUrl: 'MockedMediaClientProvider-service-host'
            });
          case 1:
          case "end":
            return _context.stop();
        }
      }, _callee);
    }));
    function authProvider() {
      return _authProvider.apply(this, arguments);
    }
    return authProvider;
  }()
};
var MockedMediaClientProvider = exports.MockedMediaClientProvider = function MockedMediaClientProvider(_ref) {
  var children = _ref.children,
    mediaStore = _ref.mediaStore,
    mockedMediaApi = _ref.mockedMediaApi,
    mediaClientConfig = _ref.mediaClientConfig;
  // WARNING: when mediaStore is updated externally, it gets out of sync with FileStreamCache. This resutls in unexpected behaviour.
  var currentStore = (0, _react.useMemo)(function () {
    return mediaStore || (0, _mediaState.createMediaStore)();
  }, [mediaStore]);
  var resolvedMediaClientConfig = (0, _react.useMemo)(function () {
    return _objectSpread(_objectSpread({}, mockedMediaClientConfig), mediaClientConfig);
  }, [mediaClientConfig]);
  var mediaClient = (0, _react.useMemo)(function () {
    return new _mediaClient.MediaClient(resolvedMediaClientConfig, currentStore, mockedMediaApi);
  }, [mockedMediaApi, currentStore, resolvedMediaClientConfig]);
  return /*#__PURE__*/_react.default.createElement(_MediaClientProvider.MediaClientContext.Provider, {
    value: mediaClient
  }, children);
};
