// @atlaskit/media-common/dist/cjs/test-helpers/getJest.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getJest = void 0;
// Mock the bare minimum of jest used within `@atlaskit/media-client/test-helpers`.
// Ideally this is not required in the future when functions are no longer used in Examples
// as they should only be used in unit tests.

var mockStub = function mockStub() {
  return Object.assign(function () {}, {
    mockReturnValue: function mockReturnValue(value) {},
    mockImplementation: function mockImplementation(value) {},
    mockResolvedValue: function mockResolvedValue(value) {}
  });
};
var mockJest = function mockJest() {
  // eslint-disable-next-line no-console
  console.warn('Using mock jest. Please remove references to @atlaskit/media-client/test-helpers in non-test code.');
  return {
    fn: function fn(value) {
      return mockStub();
    },
    spyOn: function spyOn() {
      return mockStub();
    },
    genMockFromModule: undefined,
    requireActual: function requireActual() {},
    doMock: function doMock() {}
  };
};

// ED-15806 Required as some examples currently use test modules and complain about not having `jest`.
var getJest = exports.getJest = function getJest() {
  return typeof jest === 'undefined' ? mockJest() : jest;
};
