// @atlaskit/media-common/dist/cjs/test-helpers/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "VRTestSmallImage", {
  enumerable: true,
  get: function get() {
    return _vr_test_small_image.VRTestSmallImage;
  }
});
Object.defineProperty(exports, "asMock", {
  enumerable: true,
  get: function get() {
    return _jestHelpers.asMock;
  }
});
Object.defineProperty(exports, "asMockFunction", {
  enumerable: true,
  get: function get() {
    return _jestHelpers.asMockFunction;
  }
});
Object.defineProperty(exports, "asMockFunctionResolvedValue", {
  enumerable: true,
  get: function get() {
    return _jestHelpers.asMockFunctionResolvedValue;
  }
});
Object.defineProperty(exports, "asMockFunctionReturnValue", {
  enumerable: true,
  get: function get() {
    return _jestHelpers.asMockFunctionReturnValue;
  }
});
Object.defineProperty(exports, "asMockReturnValue", {
  enumerable: true,
  get: function get() {
    return _jestHelpers.asMockReturnValue;
  }
});
Object.defineProperty(exports, "awaitError", {
  enumerable: true,
  get: function get() {
    return _awaitError.awaitError;
  }
});
Object.defineProperty(exports, "expectConstructorToHaveBeenCalledWith", {
  enumerable: true,
  get: function get() {
    return _jestHelpers.expectConstructorToHaveBeenCalledWith;
  }
});
Object.defineProperty(exports, "expectFunctionToHaveBeenCalledWith", {
  enumerable: true,
  get: function get() {
    return _jestHelpers.expectFunctionToHaveBeenCalledWith;
  }
});
Object.defineProperty(exports, "expectToEqual", {
  enumerable: true,
  get: function get() {
    return _jestHelpers.expectToEqual;
  }
});
Object.defineProperty(exports, "flushPromises", {
  enumerable: true,
  get: function get() {
    return _flushPromises.flushPromises;
  }
});
Object.defineProperty(exports, "getJest", {
  enumerable: true,
  get: function get() {
    return _getJest.getJest;
  }
});
Object.defineProperty(exports, "nextTick", {
  enumerable: true,
  get: function get() {
    return _nextTick.nextTick;
  }
});
Object.defineProperty(exports, "sleep", {
  enumerable: true,
  get: function get() {
    return _nextTick.sleep;
  }
});
Object.defineProperty(exports, "smallImage", {
  enumerable: true,
  get: function get() {
    return _smallImageURI.smallImage;
  }
});
Object.defineProperty(exports, "smallTransparentImage", {
  enumerable: true,
  get: function get() {
    return _smallTransparentImageURI.smallTransparentImage;
  }
});
Object.defineProperty(exports, "tallImage", {
  enumerable: true,
  get: function get() {
    return _tallImageURI.tallImage;
  }
});
Object.defineProperty(exports, "timeoutPromise", {
  enumerable: true,
  get: function get() {
    return _timeoutPromise.timeoutPromise;
  }
});
Object.defineProperty(exports, "videoPreviewURI", {
  enumerable: true,
  get: function get() {
    return _videoPreviewURI.videoPreviewURI;
  }
});
Object.defineProperty(exports, "videoURI", {
  enumerable: true,
  get: function get() {
    return _videoURI.videoURI;
  }
});
Object.defineProperty(exports, "wideImage", {
  enumerable: true,
  get: function get() {
    return _wideImageURI.wideImage;
  }
});
Object.defineProperty(exports, "wideTransparentImage", {
  enumerable: true,
  get: function get() {
    return _wideTransparentImageURI.wideTransparentImage;
  }
});
var _jestHelpers = require("./jestHelpers");
var _flushPromises = require("./flushPromises");
var _awaitError = require("./await-error");
var _nextTick = require("./nextTick");
var _timeoutPromise = require("./timeoutPromise");
var _smallImageURI = require("./dataURIs/smallImageURI");
var _smallTransparentImageURI = require("./dataURIs/smallTransparentImageURI");
var _tallImageURI = require("./dataURIs/tallImageURI");
var _videoPreviewURI = require("./dataURIs/videoPreviewURI");
var _videoURI = require("./dataURIs/videoURI");
var _vr_test_small_image = require("./dataURIs/vr_test_small_image");
var _wideImageURI = require("./dataURIs/wideImageURI");
var _wideTransparentImageURI = require("./dataURIs/wideTransparentImageURI");
var _getJest = require("./getJest");
