// @atlaskit/mention/dist/es2019/util/i18n.js
import React from 'react';
import { FormattedMessage } from 'react-intl';
import { messages } from '../components/i18n';
export const noPropFormatter = (messageDescriptor) => props => (React.createElement(FormattedMessage, Object.assign({}, props, messageDescriptor)));
export const NoAccessWarning = ({ name, ...props }) => (React.createElement(FormattedMessage, Object.assign({}, props, { values: { name } }, messages.noAccessWarning)));
export const NoAccessLabel = noPropFormatter(messages.noAccessLabel);
export const DefaultHeadline = noPropFormatter(messages.defaultHeadline);
export const DefaultAdvisedAction = noPropFormatter(messages.defaultAdvisedAction);
export const LoginAgain = noPropFormatter(messages.loginAgain);
export const DifferentText = noPropFormatter(messages.differentText);
export const TeamMentionHighlightTitle = noPropFormatter(messages.TeamMentionHighlightTitle);
export const TeamMentionHighlightCloseTooltip = noPropFormatter(messages.TeamMentionHighlightCloseButtonToolTip);
export const TeamMentionHighlightDescription = noPropFormatter(messages.TeamMentionHighlightDescription);
export const TeamMentionHighlightDescriptionLink = noPropFormatter(messages.TeamMentionHighlightDescriptionLink);
