// @atlaskit/editor-plugin-text-formatting/dist/cjs/ui/Toolbar/more-button.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MoreButton = void 0;
var _react = _interopRequireDefault(require("react"));
var _react2 = require("@emotion/react");
var _styles = require("@atlaskit/editor-common/styles");
var _uiMenu = require("@atlaskit/editor-common/ui-menu");
var _showMoreHorizontalEditorMore = _interopRequireDefault(require("@atlaskit/icon/utility/migration/show-more-horizontal--editor-more"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
/**
 * @jsxRuntime classic
 * @jsx jsx
 */

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var MoreIconStyle = (0, _react2.css)({
  display: 'flex',
  alignItems: 'center',
  height: "var(--ds-space-300, 24px)"
});
var MoreButton = exports.MoreButton = /*#__PURE__*/_react.default.memo(function (_ref) {
  var label = _ref.label,
    ariaExpanded = _ref['aria-expanded'],
    isReducedSpacing = _ref.isReducedSpacing,
    isSelected = _ref.isSelected,
    isDisabled = _ref.isDisabled,
    onClick = _ref.onClick,
    onKeyDown = _ref.onKeyDown;
  return (0, _react2.jsx)(_uiMenu.ToolbarButton, {
    disabled: isDisabled,
    selected: isSelected,
    onClick: onClick,
    onKeyDown: onKeyDown,
    spacing: isReducedSpacing ? 'none' : 'default',
    title: label,
    iconBefore:
    // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/platform/ensure-feature-flag-prefix -- Ignored via go/DSP-18766
    (0, _react2.jsx)("div", {
      css: (0, _platformFeatureFlags.fg)('platform-visual-refresh-icons') ? MoreIconStyle : _styles.triggerWrapperStyles
    }, (0, _react2.jsx)(_showMoreHorizontalEditorMore.default, {
      label: "",
      color: "currentColor",
      spacing: "spacious"
    })),
    "aria-expanded": ariaExpanded,
    "aria-label": label,
    "aria-haspopup": true
  });
});
