// @atlaskit/linking-common/dist/es2019/hooks/useAvailableSites/types.js
export let AvailableSitesProductType = /*#__PURE__*/function (AvailableSitesProductType) {
  AvailableSitesProductType["WHITEBOARD"] = "atlassian-whiteboard";
  AvailableSitesProductType["BEACON"] = "beacon";
  AvailableSitesProductType["COMPASS"] = "compass";
  AvailableSitesProductType["CONFLUENCE"] = "confluence.ondemand";
  AvailableSitesProductType["JIRA_BUSINESS"] = "jira-core.ondemand";
  AvailableSitesProductType["JIRA_INCIDENT_MANAGER"] = "jira-incident-manager.ondemand";
  AvailableSitesProductType["JIRA_PRODUCT_DISCOVERY"] = "jira-product-discovery";
  AvailableSitesProductType["JIRA_SERVICE_DESK"] = "jira-servicedesk.ondemand";
  AvailableSitesProductType["JIRA_SOFTWARE"] = "jira-software.ondemand";
  AvailableSitesProductType["MERCURY"] = "mercury";
  AvailableSitesProductType["OPSGENIE"] = "opsgenie";
  AvailableSitesProductType["STATUS_PAGE"] = "statuspage";
  AvailableSitesProductType["ATLAS"] = "townsquare";
  AvailableSitesProductType["LOOM"] = "loom";
  return AvailableSitesProductType;
}({});
