// @atlaskit/ufo/dist/cjs/global-stream-buffer/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.unsubscribeEvent = exports.subscribeEvent = exports.setGlobalEventStream = exports.getGlobalEventStream = exports.experiencePayloadEvent = void 0;
var _logger = require("../logger");
var _types = require("../types");
var MAX_EARLY_QUEUE_LENGTH = 250;
var getGlobalEventStream = exports.getGlobalEventStream = function getGlobalEventStream() {
  if (!globalThis.__UFO_GLOBAL_EVENT_STREAM__) {
    var earlyStream = function () {
      var stream = [];
      return {
        __buffer_only__: true,
        push: function push(event) {
          if (stream.length < MAX_EARLY_QUEUE_LENGTH) {
            stream.push(event);
          }
        },
        getStream: function getStream() {
          return stream;
        }
      };
    }();
    globalThis.__UFO_GLOBAL_EVENT_STREAM__ = earlyStream;
  }
  return globalThis.__UFO_GLOBAL_EVENT_STREAM__;
};
var setGlobalEventStream = exports.setGlobalEventStream = function setGlobalEventStream(eventStream) {
  var _globalThis$__UFO_GLO, _globalThis$__UFO_GLO2;
  if (globalThis.__UFO_GLOBAL_EVENT_STREAM__ && !((_globalThis$__UFO_GLO = globalThis.__UFO_GLOBAL_EVENT_STREAM__) !== null && _globalThis$__UFO_GLO !== void 0 && _globalThis$__UFO_GLO.__buffer_only__)) {
    (0, _logger.ufowarn)('Global Event Stream already initialised. Ignoring new event stream setup');
    return;
  }
  if ((_globalThis$__UFO_GLO2 = globalThis.__UFO_GLOBAL_EVENT_STREAM__) !== null && _globalThis$__UFO_GLO2 !== void 0 && _globalThis$__UFO_GLO2.__buffer_only__) {
    globalThis.__UFO_GLOBAL_EVENT_STREAM__.getStream().forEach(eventStream.push.bind(eventStream));
  }
  globalThis.__UFO_GLOBAL_EVENT_STREAM__ = eventStream;
};
var subscribeEvent = exports.subscribeEvent = function subscribeEvent(experienceId, callback) {
  return {
    type: _types.UFOGlobalEventStreamEventType.SUBSCRIBE,
    payload: {
      experienceId: experienceId,
      callback: callback
    }
  };
};
var unsubscribeEvent = exports.unsubscribeEvent = function unsubscribeEvent(experienceId, callback) {
  return {
    type: _types.UFOGlobalEventStreamEventType.UNSUBSCRIBE,
    payload: {
      experienceId: experienceId,
      callback: callback
    }
  };
};
var experiencePayloadEvent = exports.experiencePayloadEvent = function experiencePayloadEvent(data) {
  return {
    type: _types.UFOGlobalEventStreamEventType.EXPERIENCE_PAYLOAD,
    payload: data
  };
};
