// @atlaskit/jql-ast/dist/cjs/types/ast/operator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isWasOperator = exports.isWasInOperator = exports.isOperator = exports.isListOperator = exports.isLikeOperator = exports.isIsOperator = exports.isInOperator = exports.isEqualsOperator = exports.isComparisonOperator = exports.isChangedOperator = void 0;
var _constants = require("../../constants");
// Guards

var isOperator = exports.isOperator = function isOperator(operator) {
  return _constants.OPERATORS.includes(operator);
};
var isEqualsOperator = exports.isEqualsOperator = function isEqualsOperator(operator) {
  return _constants.EQUALS_OPERATORS.includes(operator);
};
var isLikeOperator = exports.isLikeOperator = function isLikeOperator(operator) {
  return _constants.LIKE_OPERATORS.includes(operator);
};
var isComparisonOperator = exports.isComparisonOperator = function isComparisonOperator(operator) {
  return _constants.COMPARISON_OPERATORS.includes(operator);
};
var isInOperator = exports.isInOperator = function isInOperator(operator) {
  return _constants.IN_OPERATORS.includes(operator);
};
var isIsOperator = exports.isIsOperator = function isIsOperator(operator) {
  return _constants.IS_OPERATORS.includes(operator);
};
var isWasOperator = exports.isWasOperator = function isWasOperator(operator) {
  return _constants.WAS_OPERATORS.includes(operator);
};
var isWasInOperator = exports.isWasInOperator = function isWasInOperator(operator) {
  return _constants.WAS_IN_OPERATORS.includes(operator);
};
var isChangedOperator = exports.isChangedOperator = function isChangedOperator(operator) {
  return _constants.CHANGED_OPERATORS.includes(operator);
};
var isListOperator = exports.isListOperator = function isListOperator(operator) {
  return _constants.LIST_OPERATORS.includes(operator);
};
