// @atlaskit/editor-plugin-extension/dist/cjs/ui/ConfigPanel/ErrorMessage/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactIntlNext = require("react-intl-next");
var _extensions = require("@atlaskit/editor-common/extensions");
var _emptyState = _interopRequireDefault(require("@atlaskit/empty-state"));
var _ErrorImage = _interopRequireDefault(require("./ErrorImage"));
var ConfigPanelErrorMessage = function ConfigPanelErrorMessage(_ref) {
  var errorMessage = _ref.errorMessage,
    intl = _ref.intl;
  return /*#__PURE__*/_react.default.createElement(_emptyState.default, {
    header: intl.formatMessage(_extensions.configPanelMessages.configFailedToLoad),
    description: errorMessage,
    renderImage: function renderImage() {
      return /*#__PURE__*/_react.default.createElement(_ErrorImage.default, null);
    },
    size: "narrow",
    imageHeight: 80,
    testId: "config-panel-error-message"
  });
};
var _default = exports.default = (0, _reactIntlNext.injectIntl)(ConfigPanelErrorMessage);
