// @atlaskit/media-ui/dist/es2019/shortcut.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import { Component } from 'react';
export const keyCodes = {
  space: 'Space',
  m: 'KeyM',
  rightArrow: 'ArrowRight',
  leftArrow: 'ArrowLeft'
};
export class Shortcut extends Component {
  constructor(...args) {
    super(...args);
    _defineProperty(this, "eventType", 'keydown');
    _defineProperty(this, "keyHandler", e => {
      const {
        handler,
        code
      } = this.props;
      if (code !== undefined && e.code === code) {
        handler();
      }
    });
    _defineProperty(this, "init", () => {
      const {
        eventType = 'keydown'
      } = this.props;
      this.eventType = eventType;
      document.addEventListener(this.eventType, this.keyHandler);
    });
    _defineProperty(this, "release", () => {
      document.removeEventListener(this.eventType, this.keyHandler);
    });
  }
  componentDidMount() {
    this.init();
  }
  componentWillUnmount() {
    this.release();
  }
  render() {
    return null;
  }
}
