// @atlaskit/tag/dist/cjs/tag/internal/simple/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _base = _interopRequireDefault(require("../shared/base"));
var _before = _interopRequireDefault(require("../shared/before"));
var _content = _interopRequireDefault(require("../shared/content"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var SimpleTagComponent = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var appearance = _ref.appearance,
    _ref$elemBefore = _ref.elemBefore,
    elemBefore = _ref$elemBefore === void 0 ? null : _ref$elemBefore,
    _ref$color = _ref.color,
    color = _ref$color === void 0 ? 'standard' : _ref$color,
    href = _ref.href,
    testId = _ref.testId,
    _ref$text = _ref.text,
    text = _ref$text === void 0 ? '' : _ref$text,
    linkComponent = _ref.linkComponent;
  var content = /*#__PURE__*/React.createElement(_content.default, {
    elemBefore: elemBefore,
    isRemovable: false,
    text: text,
    color: color,
    href: href,
    linkComponent: linkComponent
  });
  return /*#__PURE__*/React.createElement(_base.default, {
    ref: ref,
    testId: testId,
    href: href,
    appearance: appearance,
    color: color,
    before: /*#__PURE__*/React.createElement(_before.default, {
      elemBefore: elemBefore
    }),
    contentElement: content
  });
});

/**
 * __Simple tag__
 *
 * A tag labels UI objects for quick recognition and navigation.
 *
 * `SimpleTag` is the default form of a tag, where text is required. Tags with static text can be used as a flag or as a reference to an object or attribute.
 *
 * - [Examples](https://atlassian.design/components/tag/examples)
 * - [Code](https://atlassian.design/components/tag/code)
 * - [Usage](https://atlassian.design/components/tag/usage)
 */
var SimpleTag = /*#__PURE__*/(0, _react.memo)(SimpleTagComponent);
var _default = exports.default = SimpleTag;
