// @atlaskit/editor-plugin-annotation/dist/cjs/types/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AnnotationTestIds = exports.AnnotationSelectionType = void 0;
/**
 * type of target that annotation apply to.
 * This is used to apply correct decoration to a draft comment
 */
/**
 * The source of draft comment being created
 */
var AnnotationSelectionType = exports.AnnotationSelectionType = /*#__PURE__*/function (AnnotationSelectionType) {
  AnnotationSelectionType["INVALID"] = "invalid";
  // Annotation should not be created, toolbar should not be shown
  AnnotationSelectionType["DISABLED"] = "disabled";
  // Annotation should not be created, toolbar should be shown, but disabled
  AnnotationSelectionType["VALID"] = "valid"; // Annotation can be created
  return AnnotationSelectionType;
}({});
var prefix = 'ak-editor-annotation';
var AnnotationTestIds = exports.AnnotationTestIds = {
  prefix: prefix,
  floatingComponent: "".concat(prefix, "-floating-component"),
  floatingToolbarCreateButton: "".concat(prefix, "-toolbar-create-button"),
  componentSave: "".concat(prefix, "-dummy-save-button"),
  componentClose: "".concat(prefix, "-dummy-close-button")
};
