// @atlaskit/media-card/dist/cjs/card/ui/common.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.responsiveSettings = exports.getTitleBoxHeight = exports.Breakpoint = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var Breakpoint = exports.Breakpoint = /*#__PURE__*/function (Breakpoint) {
  Breakpoint["SMALL"] = "small";
  Breakpoint["LARGE"] = "large";
  return Breakpoint;
}({});
var responsiveSettings = exports.responsiveSettings = (0, _defineProperty2.default)((0, _defineProperty2.default)({}, Breakpoint.SMALL, {
  fontSize: 11,
  lineHeight: 14,
  titleBox: {
    verticalPadding: 4,
    horizontalPadding: 8
  }
}), Breakpoint.LARGE, {
  fontSize: 14,
  lineHeight: 22,
  titleBox: {
    verticalPadding: 8,
    horizontalPadding: 12
  }
});
var getTitleBoxHeight = exports.getTitleBoxHeight = function getTitleBoxHeight(breakpoint) {
  return (responsiveSettings[breakpoint].lineHeight + responsiveSettings[breakpoint].titleBox.verticalPadding) * 2;
};
