// @atlaskit/jql-editor/dist/cjs/ui/jql-editor-footer-content/jql-messages/format/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.extractMessageNodes = exports.MessageContainer = exports.FormatMessages = void 0;
var _react = _interopRequireDefault(require("react"));
var _useEditorTheme = require("../../../../hooks/use-editor-theme");
var _styled = require("./styled");
// Max number messages we want to show
var MAX_MESSAGES = 10;
var MessageContainer = exports.MessageContainer = function MessageContainer(_ref) {
  var children = _ref.children;
  var _useEditorThemeContex = (0, _useEditorTheme.useEditorThemeContext)(),
    isSearch = _useEditorThemeContex.isSearch;
  return /*#__PURE__*/_react.default.createElement(_styled.MessageContainer, {
    isSearch: isSearch
  }, children);
};
var FormatMessages = exports.FormatMessages = function FormatMessages(_ref2) {
  var messages = _ref2.messages;
  var messageNodes = extractMessageNodes(messages);
  if (messageNodes.length === 0) {
    return null;
  }
  if (messageNodes.length === 1) {
    return /*#__PURE__*/_react.default.createElement("span", null, messageNodes[0]);
  }
  return /*#__PURE__*/_react.default.createElement(_styled.MessageList, null, messageNodes.map(function (m, i) {
    return /*#__PURE__*/_react.default.createElement("li", {
      key: i
    }, m);
  }));
};

/**
 * This function was extracted from FormatMessages, so that the rendering is decoupled from the logic
 * This is so that the extractMessageNodes can used elsewhere where rendering is delegated to a different renderer
 *
 * Simply put, this function only handles getting m.message, and limiting to MAX_MESSAGES
 */
var extractMessageNodes = exports.extractMessageNodes = function extractMessageNodes(messages) {
  return messages.slice(0, MAX_MESSAGES).map(function (m) {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, m.message);
  });
};
