// @atlaskit/editor-plugins/dist/cjs/table/ui/common-styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "baseTableStyles", {
  enumerable: true,
  get: function get() {
    return _commonStyles.baseTableStyles;
  }
});
Object.defineProperty(exports, "insertColumnButtonOffset", {
  enumerable: true,
  get: function get() {
    return _commonStyles.insertColumnButtonOffset;
  }
});
Object.defineProperty(exports, "tableCommentEditorStyles", {
  enumerable: true,
  get: function get() {
    return _commonStyles.tableCommentEditorStyles;
  }
});
Object.defineProperty(exports, "tableFullPageEditorStyles", {
  enumerable: true,
  get: function get() {
    return _commonStyles.tableFullPageEditorStyles;
  }
});
Object.defineProperty(exports, "tableRowHeight", {
  enumerable: true,
  get: function get() {
    return _commonStyles.tableRowHeight;
  }
});
Object.defineProperty(exports, "tableStyles", {
  enumerable: true,
  get: function get() {
    return _commonStyles.tableStyles;
  }
});
var _commonStyles = require("@atlaskit/editor-plugin-table/ui/common-styles");
