// @atlaskit/editor-plugins/dist/cjs/table/ui/consts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "STICKY_HEADER_TOGGLE_TOLERANCE_MS", {
  enumerable: true,
  get: function get() {
    return _consts.STICKY_HEADER_TOGGLE_TOLERANCE_MS;
  }
});
Object.defineProperty(exports, "TABLE_DRAG_MENU_MENU_GROUP_BEFORE_HEIGHT", {
  enumerable: true,
  get: function get() {
    return _consts.TABLE_DRAG_MENU_MENU_GROUP_BEFORE_HEIGHT;
  }
});
Object.defineProperty(exports, "TABLE_DRAG_MENU_PADDING_TOP", {
  enumerable: true,
  get: function get() {
    return _consts.TABLE_DRAG_MENU_PADDING_TOP;
  }
});
Object.defineProperty(exports, "TABLE_DRAG_MENU_SORT_GROUP_HEIGHT", {
  enumerable: true,
  get: function get() {
    return _consts.TABLE_DRAG_MENU_SORT_GROUP_HEIGHT;
  }
});
Object.defineProperty(exports, "TABLE_GUIDELINE_VISIBLE_ADJUSTMENT", {
  enumerable: true,
  get: function get() {
    return _consts.TABLE_GUIDELINE_VISIBLE_ADJUSTMENT;
  }
});
Object.defineProperty(exports, "TABLE_HIGHLIGHT_GAP", {
  enumerable: true,
  get: function get() {
    return _consts.TABLE_HIGHLIGHT_GAP;
  }
});
Object.defineProperty(exports, "TABLE_HIGHLIGHT_TOLERANCE", {
  enumerable: true,
  get: function get() {
    return _consts.TABLE_HIGHLIGHT_TOLERANCE;
  }
});
Object.defineProperty(exports, "TABLE_SNAP_GAP", {
  enumerable: true,
  get: function get() {
    return _consts.TABLE_SNAP_GAP;
  }
});
Object.defineProperty(exports, "colorPalletteColumns", {
  enumerable: true,
  get: function get() {
    return _consts.colorPalletteColumns;
  }
});
Object.defineProperty(exports, "columnControlsDecorationHeight", {
  enumerable: true,
  get: function get() {
    return _consts.columnControlsDecorationHeight;
  }
});
Object.defineProperty(exports, "columnControlsSelectedZIndex", {
  enumerable: true,
  get: function get() {
    return _consts.columnControlsSelectedZIndex;
  }
});
Object.defineProperty(exports, "columnControlsZIndex", {
  enumerable: true,
  get: function get() {
    return _consts.columnControlsZIndex;
  }
});
Object.defineProperty(exports, "columnResizeHandleZIndex", {
  enumerable: true,
  get: function get() {
    return _consts.columnResizeHandleZIndex;
  }
});
Object.defineProperty(exports, "contextualMenuDropdownWidth", {
  enumerable: true,
  get: function get() {
    return _consts.contextualMenuDropdownWidth;
  }
});
Object.defineProperty(exports, "contextualMenuDropdownWidthDnD", {
  enumerable: true,
  get: function get() {
    return _consts.contextualMenuDropdownWidthDnD;
  }
});
Object.defineProperty(exports, "contextualMenuTriggerSize", {
  enumerable: true,
  get: function get() {
    return _consts.contextualMenuTriggerSize;
  }
});
Object.defineProperty(exports, "dragMenuDropdownWidth", {
  enumerable: true,
  get: function get() {
    return _consts.dragMenuDropdownWidth;
  }
});
Object.defineProperty(exports, "dragTableInsertColumnButtonSize", {
  enumerable: true,
  get: function get() {
    return _consts.dragTableInsertColumnButtonSize;
  }
});
Object.defineProperty(exports, "dropTargetExtendedWidth", {
  enumerable: true,
  get: function get() {
    return _consts.dropTargetExtendedWidth;
  }
});
Object.defineProperty(exports, "dropTargetsZIndex", {
  enumerable: true,
  get: function get() {
    return _consts.dropTargetsZIndex;
  }
});
Object.defineProperty(exports, "insertLineWidth", {
  enumerable: true,
  get: function get() {
    return _consts.insertLineWidth;
  }
});
Object.defineProperty(exports, "layoutButtonSize", {
  enumerable: true,
  get: function get() {
    return _consts.layoutButtonSize;
  }
});
Object.defineProperty(exports, "lineMarkerOffsetFromColumnControls", {
  enumerable: true,
  get: function get() {
    return _consts.lineMarkerOffsetFromColumnControls;
  }
});
Object.defineProperty(exports, "lineMarkerSize", {
  enumerable: true,
  get: function get() {
    return _consts.lineMarkerSize;
  }
});
Object.defineProperty(exports, "resizeHandlerAreaWidth", {
  enumerable: true,
  get: function get() {
    return _consts.resizeHandlerAreaWidth;
  }
});
Object.defineProperty(exports, "resizeHandlerZIndex", {
  enumerable: true,
  get: function get() {
    return _consts.resizeHandlerZIndex;
  }
});
Object.defineProperty(exports, "resizeLineWidth", {
  enumerable: true,
  get: function get() {
    return _consts.resizeLineWidth;
  }
});
Object.defineProperty(exports, "rowControlsZIndex", {
  enumerable: true,
  get: function get() {
    return _consts.rowControlsZIndex;
  }
});
Object.defineProperty(exports, "stickyHeaderBorderBottomWidth", {
  enumerable: true,
  get: function get() {
    return _consts.stickyHeaderBorderBottomWidth;
  }
});
Object.defineProperty(exports, "stickyRowOffsetTop", {
  enumerable: true,
  get: function get() {
    return _consts.stickyRowOffsetTop;
  }
});
Object.defineProperty(exports, "stickyRowZIndex", {
  enumerable: true,
  get: function get() {
    return _consts.stickyRowZIndex;
  }
});
Object.defineProperty(exports, "tableBorderColor", {
  enumerable: true,
  get: function get() {
    return _consts.tableBorderColor;
  }
});
Object.defineProperty(exports, "tableBorderDeleteColor", {
  enumerable: true,
  get: function get() {
    return _consts.tableBorderDeleteColor;
  }
});
Object.defineProperty(exports, "tableBorderRadiusSize", {
  enumerable: true,
  get: function get() {
    return _consts.tableBorderRadiusSize;
  }
});
Object.defineProperty(exports, "tableBorderSelectedColor", {
  enumerable: true,
  get: function get() {
    return _consts.tableBorderSelectedColor;
  }
});
Object.defineProperty(exports, "tableCellBackgroundColor", {
  enumerable: true,
  get: function get() {
    return _consts.tableCellBackgroundColor;
  }
});
Object.defineProperty(exports, "tableCellDeleteColor", {
  enumerable: true,
  get: function get() {
    return _consts.tableCellDeleteColor;
  }
});
Object.defineProperty(exports, "tableCellHoverDeleteIconBackground", {
  enumerable: true,
  get: function get() {
    return _consts.tableCellHoverDeleteIconBackground;
  }
});
Object.defineProperty(exports, "tableCellHoverDeleteIconColor", {
  enumerable: true,
  get: function get() {
    return _consts.tableCellHoverDeleteIconColor;
  }
});
Object.defineProperty(exports, "tableCellSelectedColor", {
  enumerable: true,
  get: function get() {
    return _consts.tableCellSelectedColor;
  }
});
Object.defineProperty(exports, "tableCellSelectedDeleteIconBackground", {
  enumerable: true,
  get: function get() {
    return _consts.tableCellSelectedDeleteIconBackground;
  }
});
Object.defineProperty(exports, "tableCellSelectedDeleteIconColor", {
  enumerable: true,
  get: function get() {
    return _consts.tableCellSelectedDeleteIconColor;
  }
});
Object.defineProperty(exports, "tableControlsSpacing", {
  enumerable: true,
  get: function get() {
    return _consts.tableControlsSpacing;
  }
});
Object.defineProperty(exports, "tableDeleteButtonOffset", {
  enumerable: true,
  get: function get() {
    return _consts.tableDeleteButtonOffset;
  }
});
Object.defineProperty(exports, "tableDeleteButtonSize", {
  enumerable: true,
  get: function get() {
    return _consts.tableDeleteButtonSize;
  }
});
Object.defineProperty(exports, "tableFloatingControlsColor", {
  enumerable: true,
  get: function get() {
    return _consts.tableFloatingControlsColor;
  }
});
Object.defineProperty(exports, "tableHeaderCellBackgroundColor", {
  enumerable: true,
  get: function get() {
    return _consts.tableHeaderCellBackgroundColor;
  }
});
Object.defineProperty(exports, "tableInsertColumnButtonOffset", {
  enumerable: true,
  get: function get() {
    return _consts.tableInsertColumnButtonOffset;
  }
});
Object.defineProperty(exports, "tableInsertColumnButtonSize", {
  enumerable: true,
  get: function get() {
    return _consts.tableInsertColumnButtonSize;
  }
});
Object.defineProperty(exports, "tableMarginFullWidthMode", {
  enumerable: true,
  get: function get() {
    return _consts.tableMarginFullWidthMode;
  }
});
Object.defineProperty(exports, "tableOverflowShadowWidth", {
  enumerable: true,
  get: function get() {
    return _consts.tableOverflowShadowWidth;
  }
});
Object.defineProperty(exports, "tableOverflowShadowWidthWide", {
  enumerable: true,
  get: function get() {
    return _consts.tableOverflowShadowWidthWide;
  }
});
Object.defineProperty(exports, "tablePadding", {
  enumerable: true,
  get: function get() {
    return _consts.tablePadding;
  }
});
Object.defineProperty(exports, "tablePopupMenuFitHeight", {
  enumerable: true,
  get: function get() {
    return _consts.tablePopupMenuFitHeight;
  }
});
Object.defineProperty(exports, "tableResizerWidth", {
  enumerable: true,
  get: function get() {
    return _consts.tableResizerWidth;
  }
});
Object.defineProperty(exports, "tableScrollbarOffset", {
  enumerable: true,
  get: function get() {
    return _consts.tableScrollbarOffset;
  }
});
Object.defineProperty(exports, "tableTextColor", {
  enumerable: true,
  get: function get() {
    return _consts.tableTextColor;
  }
});
Object.defineProperty(exports, "tableToolbarColor", {
  enumerable: true,
  get: function get() {
    return _consts.tableToolbarColor;
  }
});
Object.defineProperty(exports, "tableToolbarDeleteColor", {
  enumerable: true,
  get: function get() {
    return _consts.tableToolbarDeleteColor;
  }
});
Object.defineProperty(exports, "tableToolbarSelectedColor", {
  enumerable: true,
  get: function get() {
    return _consts.tableToolbarSelectedColor;
  }
});
Object.defineProperty(exports, "tableToolbarSize", {
  enumerable: true,
  get: function get() {
    return _consts.tableToolbarSize;
  }
});
var _consts = require("@atlaskit/editor-plugin-table/ui/consts");
