// @atlaskit/editor-element-browser/dist/cjs/utils/get-mapped-items.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getMappedItems = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireDefault(require("react"));
var _dividerElement = _interopRequireDefault(require("@atlaskit/icon-lab/core/divider-element"));
var _expandElement = _interopRequireDefault(require("@atlaskit/icon-lab/core/expand-element"));
var _lozenge = _interopRequireDefault(require("@atlaskit/icon-lab/core/lozenge"));
var _angleBrackets = _interopRequireDefault(require("@atlaskit/icon/core/angle-brackets"));
var _calendar = _interopRequireDefault(require("@atlaskit/icon/core/calendar"));
var _decision = _interopRequireDefault(require("@atlaskit/icon/core/decision"));
var _emoji = _interopRequireDefault(require("@atlaskit/icon/core/emoji"));
var _grid = _interopRequireDefault(require("@atlaskit/icon/core/grid"));
var _image = _interopRequireDefault(require("@atlaskit/icon/core/image"));
var _informationCircle = _interopRequireDefault(require("@atlaskit/icon/core/information-circle"));
var _layoutThreeColumns = _interopRequireDefault(require("@atlaskit/icon/core/layout-three-columns"));
var _link = _interopRequireDefault(require("@atlaskit/icon/core/link"));
var _mention = _interopRequireDefault(require("@atlaskit/icon/core/mention"));
var _quotationMark = _interopRequireDefault(require("@atlaskit/icon/core/quotation-mark"));
var _task = _interopRequireDefault(require("@atlaskit/icon/core/task"));
var _video = _interopRequireDefault(require("@atlaskit/icon/core/video"));
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; } // Status
// Codeblock
// Decision
// Emoji
// Table
// Media, Media-insert
// Temporaty Panel icon
// Layout 2-cols
// import LayoutTwoColumnsIcon from '@atlaskit/icon/core/layout-two-columns';
// Layout 3-cols
// Link
// Mention
// Placehoder icon for native elements:
// Date, Divider (rule), Expand, Layout (1, 4, 5 cols), Panel (all), Status
// import MinusIcon from '@atlaskit/icon/core/minus';
// Blockquote
// Action
// Loom
var getItemData = function getItemData(item, tempKey) {
  var itemId = item.id;
  var itemTitle = item.title;
  if (itemId && itemId !== 'datasource') {
    switch (itemId) {
      case 'action':
        return _objectSpread(_objectSpread({}, item), {}, {
          icon: function icon() {
            return /*#__PURE__*/_react.default.createElement(_task.default, {
              label: ""
            });
          },
          categories: ['text structure'],
          shouldDisplayAtTop: true,
          tempKey: tempKey
        });
      case 'media-insert':
        return _objectSpread(_objectSpread({}, item), {}, {
          icon: function icon() {
            return /*#__PURE__*/_react.default.createElement(_image.default, {
              label: ""
            });
          },
          categories: ['media'],
          shouldDisplayAtTop: true,
          tempKey: tempKey
        });
      case 'mention':
        return _objectSpread(_objectSpread({}, item), {}, {
          icon: function icon() {
            return /*#__PURE__*/_react.default.createElement(_mention.default, {
              label: ""
            });
          },
          categories: ['collaborate'],
          shouldDisplayAtTop: true,
          tempKey: tempKey
        });
      case 'emoji':
        return _objectSpread(_objectSpread({}, item), {}, {
          icon: function icon() {
            return /*#__PURE__*/_react.default.createElement(_emoji.default, {
              label: ""
            });
          },
          categories: ['media'],
          shouldDisplayAtTop: true,
          tempKey: tempKey
        });
      case 'expand':
        return _objectSpread(_objectSpread({}, item), {}, {
          icon: function icon() {
            return /*#__PURE__*/_react.default.createElement(_expandElement.default, {
              label: ""
            });
          },
          categories: ['page structure'],
          shouldDisplayAtTop: true,
          tempKey: tempKey
        });
      case 'table':
        return _objectSpread(_objectSpread({}, item), {}, {
          icon: function icon() {
            return /*#__PURE__*/_react.default.createElement(_grid.default, {
              label: ""
            });
          },
          categories: ['page structure'],
          shouldDisplayAtTop: true,
          tempKey: tempKey
        });
      case 'codeblock':
        return _objectSpread(_objectSpread({}, item), {}, {
          icon: function icon() {
            return /*#__PURE__*/_react.default.createElement(_angleBrackets.default, {
              label: ""
            });
          },
          categories: ['text structure'],
          shouldDisplayAtTop: true,
          tempKey: tempKey
        });
      case 'status':
        return _objectSpread(_objectSpread({}, item), {}, {
          icon: function icon() {
            return /*#__PURE__*/_react.default.createElement(_lozenge.default, {
              label: ""
            });
          },
          categories: ['text structure'],
          shouldDisplayAtTop: true,
          tempKey: tempKey
        });
      case 'infopanel':
        return _objectSpread(_objectSpread({}, item), {}, {
          icon: function icon() {
            return /*#__PURE__*/_react.default.createElement(_informationCircle.default, {
              label: ""
            });
          },
          categories: ['page structure'],
          shouldDisplayAtTop: true,
          title: 'Panel',
          tempKey: tempKey
        });
      case 'date':
        return _objectSpread(_objectSpread({}, item), {}, {
          icon: function icon() {
            return /*#__PURE__*/_react.default.createElement(_calendar.default, {
              label: ""
            });
          },
          categories: ['date'],
          shouldDisplayAtTop: true,
          tempKey: tempKey
        });
      case 'loom':
        return _objectSpread(_objectSpread({}, item), {}, {
          icon: function icon() {
            return /*#__PURE__*/_react.default.createElement(_video.default, {
              label: ""
            });
          },
          categories: ['media'],
          shouldDisplayAtTop: true,
          tempKey: tempKey
        });
      case 'decision':
        return _objectSpread(_objectSpread({}, item), {}, {
          icon: function icon() {
            return /*#__PURE__*/_react.default.createElement(_decision.default, {
              label: ""
            });
          },
          categories: ['text structure'],
          shouldDisplayAtTop: true,
          tempKey: tempKey
        });
      case 'threecolumnslayout':
        return _objectSpread(_objectSpread({}, item), {}, {
          icon: function icon() {
            return /*#__PURE__*/_react.default.createElement(_layoutThreeColumns.default, {
              label: ""
            });
          },
          categories: ['page structure'],
          shouldDisplayAtTop: true,
          title: 'Layout',
          tempKey: tempKey
        });
      case 'hyperlink':
        return _objectSpread(_objectSpread({}, item), {}, {
          icon: function icon() {
            return /*#__PURE__*/_react.default.createElement(_link.default, {
              label: ""
            });
          },
          categories: ['media'],
          shouldDisplayAtTop: true,
          tempKey: tempKey
        });
      case 'rule':
        return _objectSpread(_objectSpread({}, item), {}, {
          icon: function icon() {
            return /*#__PURE__*/_react.default.createElement(_dividerElement.default, {
              label: ""
            });
          },
          categories: ['page structure'],
          shouldDisplayAtTop: true,
          tempKey: tempKey
        });
      case 'blockquote':
        return _objectSpread(_objectSpread({}, item), {}, {
          icon: function icon() {
            return /*#__PURE__*/_react.default.createElement(_quotationMark.default, {
              label: ""
            });
          },
          categories: ['text structure'],
          shouldDisplayAtTop: true,
          tempKey: tempKey
        });
    }
  } else if (itemTitle) {
    switch (itemTitle) {
      case 'Assets':
        return _objectSpread(_objectSpread({}, item), {}, {
          categories: ['data'],
          tempKey: tempKey
        });
      case 'Jira Issues':
      case 'Jira Road Map':
        return _objectSpread(_objectSpread({}, item), {}, {
          categories: ['jira'],
          tempKey: tempKey
        });
      case 'Labels List':
      case 'Related Labels':
      case 'Labels Gadget':
      case 'Popular Labels':
        return _objectSpread(_objectSpread({}, item), {}, {
          categories: ['labels'],
          tempKey: tempKey
        });
      case 'Version Report':
      case 'Time to First Response':
      case 'Decision report':
      case 'Two Dimensional Filter Statistics':
      case 'Task report':
      case 'Issue Statistics':
      case 'Page Properties Report':
        return _objectSpread(_objectSpread({}, item), {}, {
          categories: ['reports'],
          tempKey: tempKey
        });
      case 'Opsgenie Incident Timeline EU':
      case 'Opsgenie Incident Timeline':
      case 'Jira timeline':
      case 'JSM Incident Timeline EU':
      case 'JSM Incident Timeline':
      case 'Resolution Time':
        return _objectSpread(_objectSpread({}, item), {}, {
          categories: ['timelines'],
          tempKey: tempKey
        });
      case 'Create database':
      case 'Filter by label':
      case 'Chart':
      case 'Create Jira issue':
      case 'Content Report Table':
        return _objectSpread(_objectSpread({}, item), {}, {
          categories: ['data'],
          shouldDisplayAtTop: true,
          tempKey: tempKey
        });
      case 'Heat Map':
      case 'Pie Chart':
      case 'Time Since Chart':
      case 'Recently Created Chart':
      case 'Created vs. Resolved Chart':
      case 'Average Age Chart':
        return _objectSpread(_objectSpread({}, item), {}, {
          categories: ['charts'],
          tempKey: tempKey
        });
      case 'Agile Wallboard Gadget':
      case 'Sprint Burndown Gadget':
        return _objectSpread(_objectSpread({}, item), {}, {
          categories: ['gadgets'],
          tempKey: tempKey
        });
      case 'Anchor link':
      case 'Child pages':
      case 'Page Index':
        return _objectSpread(_objectSpread({}, item), {}, {
          categories: ['navigation'],
          tempKey: tempKey
        });
      case 'Table of contents':
        return _objectSpread(_objectSpread({}, item), {}, {
          categories: ['page structure'],
          tempKey: tempKey
        });
      case 'Insert Confluence list':
      case 'Live Search':
      case 'Page Tree Search':
      case 'Page Tree':
        return _objectSpread(_objectSpread({}, item), {}, {
          categories: ['search'],
          tempKey: tempKey
        });
      case 'Excerpt':
      case 'Recently Updated Dashboard':
      case 'Create from Template':
        return _objectSpread(_objectSpread({}, item), {}, {
          categories: ['connect pages'],
          tempKey: tempKey
        });
      case 'Powerpoint':
      case 'Excel':
      case 'Iframe':
      case 'Profile Picture':
      case 'Blog Posts':
      case 'Gallery':
      case 'Word':
      case 'PDF':
      case 'Widget Connector':
        return _objectSpread(_objectSpread({}, item), {}, {
          categories: ['media'],
          tempKey: tempKey
        });
      case 'Create whiteboard':
        return _objectSpread(_objectSpread({}, item), {}, {
          categories: ['media'],
          shouldDisplayAtTop: true,
          tempKey: tempKey
        });
      case 'Contributors':
      case 'Contributors Summary':
      case 'User List':
      case 'Voted Issues':
      case 'User Profile':
      case 'Spaces List':
        return _objectSpread(_objectSpread({}, item), {}, {
          categories: ['collaborate'],
          tempKey: tempKey
        });
      case 'Gliffy Diagram':
        return _objectSpread(_objectSpread({}, item), {}, {
          categories: ['apps'],
          tempKey: tempKey
        });
      default:
        return _objectSpread(_objectSpread({}, item), {}, {
          tempKey: tempKey
        });
    }
  }
  return _objectSpread(_objectSpread({}, item), {}, {
    categories: ['apps'],
    // category: 'apps',
    // subCategory: undefined,
    tempKey: tempKey
  });
};
var getMappedItems = exports.getMappedItems = function getMappedItems(items) {
  return items.map(function (item, index) {
    return getItemData(item, index);
  });
};
