// @atlaskit/editor-plugin-selection/dist/cjs/selectionPlugin.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.selectionPlugin = exports.default = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _autoExpandSelectionRangeOnInlineNodeMain = require("./pm-plugins/auto-expand-selection-range-on-inline-node-main");
var _commands = require("./pm-plugins/commands");
var _gapCursorKeymap = _interopRequireDefault(require("./pm-plugins/gap-cursor-keymap"));
var _gapCursorMain = _interopRequireDefault(require("./pm-plugins/gap-cursor-main"));
var _gapCursorPluginKey = require("./pm-plugins/gap-cursor-plugin-key");
var _keymap = _interopRequireDefault(require("./pm-plugins/keymap"));
var _markBoundaryCursorMain = require("./pm-plugins/mark-boundary-cursor-main");
var _selectionMain = require("./pm-plugins/selection-main");
var _types = require("./types");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var displayGapCursor = function displayGapCursor(toggle) {
  return function (_ref) {
    var tr = _ref.tr;
    return tr.setMeta(_gapCursorPluginKey.gapCursorPluginKey, {
      displayGapCursor: toggle
    });
  };
};
var clearManualSelection = function clearManualSelection() {
  return function (_ref2) {
    var tr = _ref2.tr;
    var currMeta = tr.getMeta(_types.selectionPluginKey);
    return tr.setMeta(_types.selectionPluginKey, _objectSpread(_objectSpread({}, currMeta), {}, {
      manualSelection: {}
    }));
  };
};
var setManualSelection = function setManualSelection(anchor, head) {
  return function (_ref3) {
    var tr = _ref3.tr;
    var currMeta = tr.getMeta(_types.selectionPluginKey);
    return tr.setMeta(_types.selectionPluginKey, _objectSpread(_objectSpread({}, currMeta), {}, {
      manualSelection: {
        anchor: anchor,
        head: head
      }
    }));
  };
};
var selectionPlugin = exports.selectionPlugin = function selectionPlugin(_ref4) {
  var options = _ref4.config;
  return {
    name: 'selection',
    commands: {
      displayGapCursor: displayGapCursor,
      clearManualSelection: clearManualSelection,
      setManualSelection: setManualSelection
    },
    actions: {
      selectNearNode: function selectNearNode(_ref5) {
        var selectionRelativeToNode = _ref5.selectionRelativeToNode,
          selection = _ref5.selection;
        return function (state) {
          return (0, _commands.selectNearNode)(selectionRelativeToNode, selection)({
            tr: state.tr
          }) || state.tr;
        };
      }
    },
    getSharedState: function getSharedState(editorState) {
      if (!editorState) {
        return undefined;
      }
      var pluginState = _types.selectionPluginKey.getState(editorState);
      return {
        selectionRelativeToNode: pluginState === null || pluginState === void 0 ? void 0 : pluginState.selectionRelativeToNode,
        selection: pluginState === null || pluginState === void 0 ? void 0 : pluginState.selection
      };
    },
    pmPlugins: function pmPlugins() {
      return [{
        name: 'selection',
        plugin: function plugin(_ref6) {
          var dispatch = _ref6.dispatch,
            dispatchAnalyticsEvent = _ref6.dispatchAnalyticsEvent;
          return (0, _selectionMain.createPlugin)(dispatch, dispatchAnalyticsEvent, options);
        }
      }, {
        name: 'selectionKeymap',
        plugin: _keymap.default
      }, {
        name: 'gapCursorKeymap',
        plugin: function plugin() {
          return (0, _gapCursorKeymap.default)();
        }
      }, {
        name: 'gapCursor',
        plugin: function plugin() {
          return _gapCursorMain.default;
        }
      }, {
        name: 'markBoundaryCursor',
        plugin: function plugin() {
          return (0, _markBoundaryCursorMain.createMarkBoundaryCursorPlugin)();
        }
      }, {
        name: 'autoExpandSelectionRangeOnInlineNode',
        plugin: function plugin() {
          return (0, _autoExpandSelectionRangeOnInlineNodeMain.createAutoExpandSelectionRangeOnInlineNodePlugin)();
        }
      }];
    }
  };
};
var _default = exports.default = selectionPlugin;
