// @atlaskit/media-ui/dist/cjs/EmbedCard/views/UnresolvedView.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
/** @jsx jsx */
var core_1 = require("@emotion/core");
var button_1 = tslib_1.__importDefault(require("@atlaskit/button"));
var theme_1 = require("@atlaskit/theme");
var react_intl_1 = require("react-intl");
var messages_1 = require("../../messages");
var utils_1 = require("../../BlockCard/utils");
exports.EmbedCardUnresolvedView = function (_a) {
    var image = _a.image, title = _a.title, description = _a.description, button = _a.button, context = _a.context, onClick = _a.onClick, testId = _a.testId;
    return (core_1.jsx("div", { css: {
            display: 'flex',
            flexDirection: 'column',
            alignItems: 'center',
            justifyContent: 'center',
            paddingTop: utils_1.gs(4.5),
            paddingBottom: utils_1.gs(6),
        }, "data-testid": testId + "-unresolved-container" },
        core_1.jsx("img", { src: image, css: { height: utils_1.gs(14), marginBottom: utils_1.gs(4) } }),
        core_1.jsx("span", { css: {
                fontSize: utils_1.gs(2.5),
                marginBottom: utils_1.gs(1.5),
                width: utils_1.gs(50),
                textAlign: 'center',
            } },
            core_1.jsx(react_intl_1.FormattedMessage, tslib_1.__assign({}, messages_1.messages[title], { values: { context: context } }))),
        core_1.jsx("span", { css: {
                fontSize: theme_1.fontSize(),
                marginBottom: utils_1.gs(2.5),
                textAlign: 'center',
                width: utils_1.gs(50),
                lineHeight: utils_1.gs(3),
            } },
            core_1.jsx(react_intl_1.FormattedMessage, tslib_1.__assign({}, messages_1.messages[description], { values: { context: context } }))),
        button && (core_1.jsx(button_1.default, { testId: testId + "-button", appearance: button.appearance, onClick: onClick },
            core_1.jsx(react_intl_1.FormattedMessage, tslib_1.__assign({}, messages_1.messages[button.text]))))));
};
