// @atlaskit/profilecard/dist/esm/components/team-profile-card/team-containers-skeleton/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import "./index.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
import React from 'react';
import { Box, Stack } from '@atlaskit/primitives/compiled';
import { LinkedContainerCardSkeleton } from './linked-container-card-skeleton';
var styles = {
  showMoreButtonSkeleton: "_2rko1q77 _1rjc1b66 _1bsbvbgk"
};
export var TeamContainersSkeleton = function TeamContainersSkeleton(_ref) {
  var numberOfContainers = _ref.numberOfContainers;
  return /*#__PURE__*/React.createElement(Stack, {
    space: "space.200",
    testId: "profile-card-team-containers-skeleton"
  }, /*#__PURE__*/React.createElement(Stack, {
    space: "space.100"
  }, _toConsumableArray(Array(numberOfContainers)).map(function (_, index) {
    return /*#__PURE__*/React.createElement(LinkedContainerCardSkeleton, {
      key: index
    });
  })), /*#__PURE__*/React.createElement(Box, {
    backgroundColor: "color.background.neutral",
    xcss: styles.showMoreButtonSkeleton
  }));
};
