// @atlaskit/editor-plugin-undo-redo/dist/es2019/undoRedoPlugin.js
import React from 'react';
import { SafePlugin } from '@atlaskit/editor-common/safe-plugin';
import { redo, undo } from '@atlaskit/editor-prosemirror/history';
import { editorExperiment } from '@atlaskit/tmp-editor-statsig/experiments';
import { attachInputMeta } from './pm-plugins/attach-input-meta';
import { InputSource } from './pm-plugins/enums';
import { keymapPlugin } from './pm-plugins/keymaps';
import { createPlugin } from './pm-plugins/main';
import { forceFocus } from './pm-plugins/utils';
// Ignored via go/ees005
// eslint-disable-next-line import/no-named-as-default
import ToolbarUndoRedo from './ui/ToolbarUndoRedo';
export const undoRedoPlugin = ({
  api
}) => {
  var _api$primaryToolbar;
  const editorViewRef = {
    current: null
  };
  const primaryToolbarComponent = ({
    editorView,
    disabled,
    isToolbarReducedSpacing
  }) => {
    return /*#__PURE__*/React.createElement(ToolbarUndoRedo, {
      isReducedSpacing: isToolbarReducedSpacing,
      disabled: disabled,
      editorView: editorView,
      api: api
    });
  };
  api === null || api === void 0 ? void 0 : (_api$primaryToolbar = api.primaryToolbar) === null || _api$primaryToolbar === void 0 ? void 0 : _api$primaryToolbar.actions.registerComponent({
    name: 'undoRedoPlugin',
    component: primaryToolbarComponent
  });
  const handleUndo = inputSource => {
    if (!editorViewRef.current) {
      return false;
    }
    return forceFocus(editorViewRef.current, api)(attachInputMeta(inputSource || InputSource.EXTERNAL)(undo));
  };
  const handleRedo = inputSource => {
    if (!editorViewRef.current) {
      return false;
    }
    return forceFocus(editorViewRef.current, api)(attachInputMeta(inputSource || InputSource.EXTERNAL)(redo));
  };
  return {
    name: 'undoRedoPlugin',
    actions: {
      undo: handleUndo,
      redo: handleRedo
    },
    pmPlugins() {
      const plugins = [{
        name: 'undoRedoKeyMap',
        plugin: () => keymapPlugin()
      }, {
        name: 'undoRedoPlugin',
        plugin: options => createPlugin(options)
      }];
      if (editorExperiment('platform_editor_controls', 'variant1', {
        exposure: false
      })) {
        plugins.push({
          name: 'undoRedoGetEditorViewReferencePlugin',
          plugin: () => {
            return new SafePlugin({
              view: editorView => {
                editorViewRef.current = editorView;
                return {
                  destroy: () => {
                    editorViewRef.current = null;
                  }
                };
              }
            });
          }
        });
      }
      return plugins;
    },
    primaryToolbarComponent: !(api !== null && api !== void 0 && api.primaryToolbar) ? primaryToolbarComponent : undefined
  };
};
