// @atlaskit/editor-plugin-avatar-group/dist/es2019/ui/avatars-with-plugin-state.js
import React from 'react';
import { injectIntl } from 'react-intl-next';
import { useSharedPluginState } from '@atlaskit/editor-common/hooks';
import messages from '@atlaskit/editor-common/messages';
import { Avatars } from './avatars';
import { InviteToEditButton } from './invite-to-edit';
const AvatarsWithPluginState = props => {
  const title = props.intl.formatMessage(messages.inviteToEditButtonTitle);
  const {
    isInviteToEditButtonSelected: selected,
    inviteToEditHandler: onClick,
    inviteToEditComponent: Component,
    editorView,
    featureFlags,
    editorAnalyticsAPI,
    editorAPI
  } = props;
  const {
    collabEditState
  } = useSharedPluginState(editorAPI, ['collabEdit']);
  if (!collabEditState) {
    return null;
  }
  return /*#__PURE__*/React.createElement(Avatars, {
    sessionId: collabEditState.sessionId,
    participants: collabEditState.activeParticipants,
    editorView: editorView,
    featureFlags: featureFlags,
    editorAnalyticsAPI: editorAnalyticsAPI,
    editorAPI: editorAPI
  }, /*#__PURE__*/React.createElement(InviteToEditButton, {
    title: title,
    selected: selected,
    onClick: onClick,
    Component: Component
  }));
};
export default injectIntl(AvatarsWithPluginState);
