// @atlaskit/editor-plugin-avatar-group/dist/es2019/ui/invite-to-edit.js
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
import React, { Fragment } from 'react';

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import { jsx } from '@emotion/react';
import { ToolbarButton } from '@atlaskit/editor-common/ui-menu';
import AddIcon from '@atlaskit/icon/core/migration/add--editor-add';
import { inviteTeamWrapperStyles } from './styles';
const ID = props => jsx(Fragment, null, props.children);
export const InviteToEditButton = props => {
  const {
    Component,
    onClick,
    selected,
    title
  } = props;
  const iconBefore = React.useMemo(() => jsx(AddIcon, {
    label: title
  }), [title]);
  if (!Component && !onClick) {
    return null;
  }
  const Wrapper = Component ? Component : ID;
  return (
    // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
    jsx("div", {
      css: inviteTeamWrapperStyles
    }, jsx(Wrapper, null, jsx(ToolbarButton
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    , {
      className: "invite-to-edit",
      onClick: onClick,
      selected: selected,
      title: title,
      titlePosition: "bottom",
      iconBefore: iconBefore
    })))
  );
};
