// @atlaskit/ufo/dist/cjs/logger/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ufowarn = exports.ufologger = exports.ufolog = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var UFOLogger = /*#__PURE__*/function () {
  function UFOLogger() {
    (0, _classCallCheck2.default)(this, UFOLogger);
    (0, _defineProperty2.default)(this, "enabled", false);
  }
  return (0, _createClass2.default)(UFOLogger, [{
    key: "log",
    value: function log() {
      var _console;
      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }
      // eslint-disable-next-line no-console
      this.enabled && (_console = console).log.apply(_console, [UFOLogger.UFOprefix].concat(args));
    }
  }, {
    key: "warn",
    value: function warn() {
      var _console2;
      for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
        args[_key2] = arguments[_key2];
      }
      // eslint-disable-next-line no-console
      this.enabled && (_console2 = console).warn.apply(_console2, [UFOLogger.UFOprefix].concat(args));
    }
  }, {
    key: "enable",
    value: function enable() {
      this.enabled = true;
    }
  }, {
    key: "disable",
    value: function disable() {
      this.enabled = false;
    }
  }]);
}();
(0, _defineProperty2.default)(UFOLogger, "UFOprefix", '[ufo🛸]');
var ufologger = exports.ufologger = new UFOLogger();
var ufolog = exports.ufolog = function ufolog() {
  ufologger.log.apply(ufologger, arguments);
};
var ufowarn = exports.ufowarn = function ufowarn() {
  ufologger.warn.apply(ufologger, arguments);
};
