// @atlaskit/editor-plugin-table/dist/cjs/pm-plugins/commands/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "addBoldInEmptyHeaderCells", {
  enumerable: true,
  get: function get() {
    return _misc.addBoldInEmptyHeaderCells;
  }
});
Object.defineProperty(exports, "addResizeHandleDecorations", {
  enumerable: true,
  get: function get() {
    return _misc.addResizeHandleDecorations;
  }
});
Object.defineProperty(exports, "autoSizeTable", {
  enumerable: true,
  get: function get() {
    return _misc.autoSizeTable;
  }
});
Object.defineProperty(exports, "clearHoverSelection", {
  enumerable: true,
  get: function get() {
    return _hover.clearHoverSelection;
  }
});
Object.defineProperty(exports, "clearMultipleCells", {
  enumerable: true,
  get: function get() {
    return _clear.clearMultipleCells;
  }
});
Object.defineProperty(exports, "convertFirstRowToHeader", {
  enumerable: true,
  get: function get() {
    return _misc.convertFirstRowToHeader;
  }
});
Object.defineProperty(exports, "createTable", {
  enumerable: true,
  get: function get() {
    return _insert.createTable;
  }
});
Object.defineProperty(exports, "deleteTable", {
  enumerable: true,
  get: function get() {
    return _misc.deleteTable;
  }
});
Object.defineProperty(exports, "getNextLayout", {
  enumerable: true,
  get: function get() {
    return _toggle.getNextLayout;
  }
});
Object.defineProperty(exports, "goToNextCell", {
  enumerable: true,
  get: function get() {
    return _goToNextCell.goToNextCell;
  }
});
Object.defineProperty(exports, "hideInsertColumnOrRowButton", {
  enumerable: true,
  get: function get() {
    return _misc.hideInsertColumnOrRowButton;
  }
});
Object.defineProperty(exports, "hideResizeHandleLine", {
  enumerable: true,
  get: function get() {
    return _hover.hideResizeHandleLine;
  }
});
Object.defineProperty(exports, "hoverCell", {
  enumerable: true,
  get: function get() {
    return _hover.hoverCell;
  }
});
Object.defineProperty(exports, "hoverColumns", {
  enumerable: true,
  get: function get() {
    return _hover.hoverColumns;
  }
});
Object.defineProperty(exports, "hoverMergedCells", {
  enumerable: true,
  get: function get() {
    return _hover.hoverMergedCells;
  }
});
Object.defineProperty(exports, "hoverRows", {
  enumerable: true,
  get: function get() {
    return _hover.hoverRows;
  }
});
Object.defineProperty(exports, "hoverTable", {
  enumerable: true,
  get: function get() {
    return _hover.hoverTable;
  }
});
Object.defineProperty(exports, "insertColumn", {
  enumerable: true,
  get: function get() {
    return _insert.insertColumn;
  }
});
Object.defineProperty(exports, "insertRow", {
  enumerable: true,
  get: function get() {
    return _insert.insertRow;
  }
});
Object.defineProperty(exports, "moveCursorBackward", {
  enumerable: true,
  get: function get() {
    return _misc.moveCursorBackward;
  }
});
Object.defineProperty(exports, "removeDescendantNodes", {
  enumerable: true,
  get: function get() {
    return _referentiality.removeDescendantNodes;
  }
});
Object.defineProperty(exports, "selectColumn", {
  enumerable: true,
  get: function get() {
    return _misc.selectColumn;
  }
});
Object.defineProperty(exports, "selectColumns", {
  enumerable: true,
  get: function get() {
    return _misc.selectColumns;
  }
});
Object.defineProperty(exports, "selectRow", {
  enumerable: true,
  get: function get() {
    return _misc.selectRow;
  }
});
Object.defineProperty(exports, "selectRows", {
  enumerable: true,
  get: function get() {
    return _misc.selectRows;
  }
});
Object.defineProperty(exports, "setCellAttr", {
  enumerable: true,
  get: function get() {
    return _misc.setCellAttr;
  }
});
Object.defineProperty(exports, "setEditorFocus", {
  enumerable: true,
  get: function get() {
    return _misc.setEditorFocus;
  }
});
Object.defineProperty(exports, "setFocusToCellMenu", {
  enumerable: true,
  get: function get() {
    return _misc.setFocusToCellMenu;
  }
});
Object.defineProperty(exports, "setMultipleCellAttrs", {
  enumerable: true,
  get: function get() {
    return _misc.setMultipleCellAttrs;
  }
});
Object.defineProperty(exports, "setTableDisplayMode", {
  enumerable: true,
  get: function get() {
    return _displayMode.setTableDisplayMode;
  }
});
Object.defineProperty(exports, "setTableHovered", {
  enumerable: true,
  get: function get() {
    return _hover.setTableHovered;
  }
});
Object.defineProperty(exports, "setTableRef", {
  enumerable: true,
  get: function get() {
    return _misc.setTableRef;
  }
});
Object.defineProperty(exports, "showInsertColumnButton", {
  enumerable: true,
  get: function get() {
    return _misc.showInsertColumnButton;
  }
});
Object.defineProperty(exports, "showInsertRowButton", {
  enumerable: true,
  get: function get() {
    return _misc.showInsertRowButton;
  }
});
Object.defineProperty(exports, "showResizeHandleLine", {
  enumerable: true,
  get: function get() {
    return _hover.showResizeHandleLine;
  }
});
Object.defineProperty(exports, "sortByColumn", {
  enumerable: true,
  get: function get() {
    return _sort.sortByColumn;
  }
});
Object.defineProperty(exports, "toggleContextualMenu", {
  enumerable: true,
  get: function get() {
    return _toggle.toggleContextualMenu;
  }
});
Object.defineProperty(exports, "toggleHeaderColumn", {
  enumerable: true,
  get: function get() {
    return _toggle.toggleHeaderColumn;
  }
});
Object.defineProperty(exports, "toggleHeaderRow", {
  enumerable: true,
  get: function get() {
    return _toggle.toggleHeaderRow;
  }
});
Object.defineProperty(exports, "toggleNumberColumn", {
  enumerable: true,
  get: function get() {
    return _toggle.toggleNumberColumn;
  }
});
Object.defineProperty(exports, "toggleTableLayout", {
  enumerable: true,
  get: function get() {
    return _toggle.toggleTableLayout;
  }
});
Object.defineProperty(exports, "transformSliceToAddTableHeaders", {
  enumerable: true,
  get: function get() {
    return _misc.transformSliceToAddTableHeaders;
  }
});
Object.defineProperty(exports, "triggerUnlessTableHeader", {
  enumerable: true,
  get: function get() {
    return _misc.triggerUnlessTableHeader;
  }
});
Object.defineProperty(exports, "updateWidthToWidest", {
  enumerable: true,
  get: function get() {
    return _misc.updateWidthToWidest;
  }
});
var _hover = require("./hover");
var _insert = require("./insert");
var _toggle = require("./toggle");
var _clear = require("./clear");
var _misc = require("./misc");
var _sort = require("./sort");
var _goToNextCell = require("./go-to-next-cell");
var _referentiality = require("./referentiality");
var _displayMode = require("./display-mode");
