// @atlaskit/editor-palette/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "hexToEditorBackgroundPaletteColor", {
  enumerable: true,
  get: function get() {
    return _background.hexToEditorBackgroundPaletteColor;
  }
});
Object.defineProperty(exports, "hexToEditorBackgroundPaletteRawValue", {
  enumerable: true,
  get: function get() {
    return _background.hexToEditorBackgroundPaletteRawValue;
  }
});
Object.defineProperty(exports, "hexToEditorBorderPaletteColor", {
  enumerable: true,
  get: function get() {
    return _border.hexToEditorBorderPaletteColor;
  }
});
Object.defineProperty(exports, "hexToEditorTableChartsPaletteColor", {
  enumerable: true,
  get: function get() {
    return _tableCharts.hexToEditorTableChartsPaletteColor;
  }
});
Object.defineProperty(exports, "hexToEditorTableChartsPaletteRawValue", {
  enumerable: true,
  get: function get() {
    return _tableCharts.hexToEditorTableChartsPaletteRawValue;
  }
});
Object.defineProperty(exports, "hexToEditorTextBackgroundPaletteColor", {
  enumerable: true,
  get: function get() {
    return _textBackgroundColor.hexToEditorTextBackgroundPaletteColor;
  }
});
Object.defineProperty(exports, "hexToEditorTextPaletteColor", {
  enumerable: true,
  get: function get() {
    return _text.hexToEditorTextPaletteColor;
  }
});
var _background = require("./background");
var _border = require("./border");
var _text = require("./text");
var _tableCharts = require("./table-charts");
var _textBackgroundColor = require("./text-background-color");
