// @atlaskit/media-viewer/dist/es2019/analytics/ufoExperiences.js
import { UFOExperience, ExperiencePerformanceTypes, ExperienceTypes } from '@atlaskit/ufo';
import { getMediaEnvironment, getMediaRegion } from '@atlaskit/media-client';
import { getFeatureFlagKeysAllProducts } from '@atlaskit/media-common';
const packageName = "@atlaskit/media-viewer";
const packageVersion = "52.0.9";
let ufoExperience;
const getExperience = () => {
  if (!ufoExperience) {
    const inlineExperience = {
      platform: {
        component: 'media-viewer'
      },
      type: ExperienceTypes.Experience,
      performanceType: ExperiencePerformanceTypes.InlineResult,
      featureFlags: getFeatureFlagKeysAllProducts()
    };
    ufoExperience = new UFOExperience('media-file', inlineExperience);
  }
  return ufoExperience;
};
export const startMediaFileUfoExperience = () => {
  getExperience().start();
};
export const succeedMediaFileUfoExperience = properties => {
  getExperience().success({
    metadata: {
      ...properties,
      packageName,
      packageVersion,
      mediaEnvironment: getMediaEnvironment(),
      mediaRegion: getMediaRegion()
    }
  });
};
export const failMediaFileUfoExperience = properties => {
  const refinedMetadata = {
    ...properties,
    packageName,
    packageVersion,
    mediaEnvironment: getMediaEnvironment(),
    mediaRegion: getMediaRegion()
  };
  getExperience().failure({
    metadata: refinedMetadata
  });
};
