// @atlaskit/editor-common/dist/cjs/guideline/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "INNER_GRID_GUIDELINE_PREFIX", {
  enumerable: true,
  get: function get() {
    return _constants.INNER_GRID_GUIDELINE_PREFIX;
  }
});
Object.defineProperty(exports, "MEDIA_DYNAMIC_GUIDELINE_PREFIX", {
  enumerable: true,
  get: function get() {
    return _constants.MEDIA_DYNAMIC_GUIDELINE_PREFIX;
  }
});
Object.defineProperty(exports, "createFixedGuidelinesFromLengths", {
  enumerable: true,
  get: function get() {
    return _fixedGuideline.createFixedGuidelinesFromLengths;
  }
});
Object.defineProperty(exports, "createGuidesFromLengths", {
  enumerable: true,
  get: function get() {
    return _fixedGuideline.createGuidesFromLengths;
  }
});
Object.defineProperty(exports, "findClosestSnap", {
  enumerable: true,
  get: function get() {
    return _snapping.findClosestSnap;
  }
});
Object.defineProperty(exports, "generateDefaultGuidelines", {
  enumerable: true,
  get: function get() {
    return _defaultGuideline.generateDefaultGuidelines;
  }
});
Object.defineProperty(exports, "generateDynamicGuidelines", {
  enumerable: true,
  get: function get() {
    return _dynamicGuideline.generateDynamicGuidelines;
  }
});
Object.defineProperty(exports, "getContainerWidthOrFullEditorWidth", {
  enumerable: true,
  get: function get() {
    return _utils.getContainerWidthOrFullEditorWidth;
  }
});
Object.defineProperty(exports, "getGuidelineSnaps", {
  enumerable: true,
  get: function get() {
    return _snapping.getGuidelineSnaps;
  }
});
Object.defineProperty(exports, "getGuidelineTypeFromKey", {
  enumerable: true,
  get: function get() {
    return _utils.getGuidelineTypeFromKey;
  }
});
Object.defineProperty(exports, "getGuidelinesWithHighlights", {
  enumerable: true,
  get: function get() {
    return _updateGuideline.getGuidelinesWithHighlights;
  }
});
Object.defineProperty(exports, "getMediaSingleDimensions", {
  enumerable: true,
  get: function get() {
    return _utils.getMediaSingleDimensions;
  }
});
Object.defineProperty(exports, "getRelativeGuideSnaps", {
  enumerable: true,
  get: function get() {
    return _relativeGuideline.getRelativeGuideSnaps;
  }
});
Object.defineProperty(exports, "getRelativeGuidelines", {
  enumerable: true,
  get: function get() {
    return _relativeGuideline.getRelativeGuidelines;
  }
});
Object.defineProperty(exports, "isVerticalPosition", {
  enumerable: true,
  get: function get() {
    return _utils.isVerticalPosition;
  }
});
var _dynamicGuideline = require("./dynamicGuideline");
var _fixedGuideline = require("./fixedGuideline");
var _defaultGuideline = require("./defaultGuideline");
var _updateGuideline = require("./updateGuideline");
var _constants = require("./constants");
var _snapping = require("./snapping");
var _utils = require("./utils");
var _relativeGuideline = require("./relativeGuideline");
