// @atlaskit/editor-plugin-media/dist/cjs/ui/ResizableMediaSingle/ResizableMediaSingleNext.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.resizerNextTestId = exports.default = exports.ResizableMediaSingleNextFunctional = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _react2 = require("@emotion/react");
var _classnames = _interopRequireDefault(require("classnames"));
var _throttle = _interopRequireDefault(require("lodash/throttle"));
var _memoizeOne = _interopRequireDefault(require("memoize-one"));
var _guideline = require("@atlaskit/editor-common/guideline");
var _mediaSingle = require("@atlaskit/editor-common/media-single");
var _resizer = require("@atlaskit/editor-common/resizer");
var _styles = require("@atlaskit/editor-common/styles");
var _ui = require("@atlaskit/editor-common/ui");
var _utils = require("@atlaskit/editor-common/utils");
var _state = require("@atlaskit/editor-prosemirror/state");
var _utils2 = require("@atlaskit/editor-prosemirror/utils");
var _editorSharedStyles = require("@atlaskit/editor-shared-styles");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _main = require("../../pm-plugins/main");
var _analytics = require("../../pm-plugins/utils/analytics");
var _checkMediaType = require("../../pm-plugins/utils/check-media-type");
var _ResizableMediaMigrationNotification = require("./ResizableMediaMigrationNotification");
var _styled = require("./styled");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); } /**
 * @jsxRuntime classic
 * @jsx jsx
 */ // eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
var resizerNextTestId = exports.resizerNextTestId = 'mediaSingle.resizerNext.testid';
// eslint-disable-next-line @repo/internal/react/no-class-components
var ResizableMediaSingleNext = /*#__PURE__*/function (_React$Component) {
  function ResizableMediaSingleNext(props) {
    var _this;
    (0, _classCallCheck2.default)(this, ResizableMediaSingleNext);
    _this = _callSuper(this, ResizableMediaSingleNext, [props]);
    (0, _defineProperty2.default)(_this, "lastSnappedGuidelineKeys", []);
    (0, _defineProperty2.default)(_this, "updateGuidelines", function () {
      var _this$props = _this.props,
        view = _this$props.view,
        lineLength = _this$props.lineLength;
      var defaultGuidelines = _this.getDefaultGuidelines();
      var _generateDynamicGuide = (0, _guideline.generateDynamicGuidelines)(view.state, lineLength, {
          styles: {
            lineStyle: 'dashed'
          },
          show: false
        }),
        relativeGuides = _generateDynamicGuide.relativeGuides,
        dynamicGuides = _generateDynamicGuide.dynamicGuides;

      // disable guidelines for nested media single node
      var dynamicGuidelines = _this.isAdjacentMode() ? [] : dynamicGuides;
      _this.setState({
        relativeGuides: relativeGuides,
        guidelines: [].concat((0, _toConsumableArray2.default)(defaultGuidelines), (0, _toConsumableArray2.default)(dynamicGuidelines))
      });
    });
    (0, _defineProperty2.default)(_this, "calcNewLayout", function (newWidth, stop) {
      var _this$props2 = _this.props,
        layout = _this$props2.layout,
        containerWidth = _this$props2.containerWidth,
        lineLength = _this$props2.lineLength,
        fullWidthMode = _this$props2.fullWidthMode;
      var newPct = (0, _ui.calcPctFromPx)(newWidth, lineLength) * 100;
      if (newPct <= 100 && _this.wrappedLayout) {
        if (!stop || newPct !== 100) {
          return layout;
        }
      }
      return _this.calcUnwrappedLayout(newWidth, containerWidth || 0, lineLength, fullWidthMode, _this.isAdjacentMode());
    });
    (0, _defineProperty2.default)(_this, "calcUnwrappedLayout", function (width, containerWidth, contentWidth, fullWidthMode, isNestedNode) {
      if (isNestedNode) {
        return 'center';
      }
      if (fullWidthMode) {
        if (width < contentWidth) {
          return 'center';
        }
        return 'full-width';
      }

      // handle top-level node in fixed-width editor
      if (width <= contentWidth) {
        return 'center';
      }
      if (width < Math.min(containerWidth - (0, _editorSharedStyles.akEditorGutterPaddingDynamic)() * 2, _editorSharedStyles.akEditorFullWidthLayoutWidth)) {
        return 'wide';
      }

      // set full width to be containerWidth - akEditorGutterPaddingDynamic() * 2
      // instead of containerWidth - akEditorBreakoutPadding,
      // so that we have image aligned with text
      return 'full-width';
    });
    (0, _defineProperty2.default)(_this, "calcPxHeight", function (newWidth) {
      var _this$props3 = _this.props,
        _this$props3$width = _this$props3.width,
        width = _this$props3$width === void 0 ? newWidth : _this$props3$width,
        height = _this$props3.height;
      return Math.round(height / width * newWidth);
    });
    (0, _defineProperty2.default)(_this, "displayGuideline", function (guidelines) {
      var _this$props$pluginInj;
      return (_this$props$pluginInj = _this.props.pluginInjectionApi) === null || _this$props$pluginInj === void 0 || (_this$props$pluginInj = _this$props$pluginInj.guideline) === null || _this$props$pluginInj === void 0 || (_this$props$pluginInj = _this$props$pluginInj.actions) === null || _this$props$pluginInj === void 0 ? void 0 : _this$props$pluginInj.displayGuideline(_this.props.view)({
        guidelines: guidelines
      });
    });
    (0, _defineProperty2.default)(_this, "setIsResizing", function (isResizing) {
      var _this$props$view = _this.props.view,
        state = _this$props$view.state,
        dispatch = _this$props$view.dispatch;
      var tr = state.tr;
      tr.setMeta(_main.MEDIA_PLUGIN_IS_RESIZING_KEY, isResizing);
      tr.setMeta('is-resizer-resizing', isResizing);
      return dispatch(tr);
    });
    (0, _defineProperty2.default)(_this, "updateSizeInPluginState", (0, _throttle.default)(function (width) {
      var _this$props$view2 = _this.props.view,
        state = _this$props$view2.state,
        dispatch = _this$props$view2.dispatch;
      var tr = state.tr;
      tr.setMeta(_main.MEDIA_PLUGIN_RESIZING_WIDTH_KEY, width);
      return dispatch(tr);
    }, _mediaSingle.MEDIA_SINGLE_RESIZE_THROTTLE_TIME));
    (0, _defineProperty2.default)(_this, "calcMaxWidth", (0, _memoizeOne.default)(function (contentWidth, containerWidth, fullWidthMode) {
      if (_this.isAdjacentMode() || fullWidthMode) {
        return contentWidth;
      }
      return (0, _mediaSingle.calcMediaSingleMaxWidth)(containerWidth, _this.props.editorAppearance);
    }));
    (0, _defineProperty2.default)(_this, "calcMinWidth", (0, _memoizeOne.default)(function (isVideoFile, contentWidth) {
      return Math.min(contentWidth || _editorSharedStyles.akEditorDefaultLayoutWidth, isVideoFile ? _mediaSingle.MEDIA_SINGLE_VIDEO_MIN_PIXEL_WIDTH : _mediaSingle.MEDIA_SINGLE_DEFAULT_MIN_PIXEL_WIDTH);
    }));
    (0, _defineProperty2.default)(_this, "getRelativeGuides", function () {
      var _this$props$pluginInj2;
      var guidelinePluginState = (_this$props$pluginInj2 = _this.props.pluginInjectionApi) === null || _this$props$pluginInj2 === void 0 || (_this$props$pluginInj2 = _this$props$pluginInj2.guideline) === null || _this$props$pluginInj2 === void 0 || (_this$props$pluginInj2 = _this$props$pluginInj2.sharedState) === null || _this$props$pluginInj2 === void 0 ? void 0 : _this$props$pluginInj2.currentState();
      var _ref = (guidelinePluginState === null || guidelinePluginState === void 0 ? void 0 : guidelinePluginState.rect) || {
          top: 0,
          left: 0
        },
        topOffset = _ref.top;
      var $pos = _this.$pos;
      var relativeGuides = $pos && $pos.nodeAfter && _this.state.size.width ? (0, _guideline.getRelativeGuidelines)(_this.state.relativeGuides, {
        node: $pos.nodeAfter,
        pos: $pos.pos
      }, _this.props.view, _this.props.lineLength, topOffset, _this.state.size) : [];
      return relativeGuides;
    });
    (0, _defineProperty2.default)(_this, "updateActiveGuidelines", function () {
      var width = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0;
      var guidelines = arguments.length > 1 ? arguments[1] : undefined;
      var guidelineSnapsReference = arguments.length > 2 ? arguments[2] : undefined;
      if (guidelineSnapsReference.snaps.x) {
        var _findClosestSnap = (0, _guideline.findClosestSnap)(width, guidelineSnapsReference.snaps.x, guidelineSnapsReference.guidelineReference, _mediaSingle.MEDIA_SINGLE_SNAP_GAP),
          gap = _findClosestSnap.gap,
          activeGuidelineKeys = _findClosestSnap.keys;
        var relativeGuidelines = activeGuidelineKeys.length ? [] : _this.getRelativeGuides();
        _this.lastSnappedGuidelineKeys = activeGuidelineKeys.length ? activeGuidelineKeys : relativeGuidelines.map(function (rg) {
          return rg.key;
        });
        _this.displayGuideline([].concat((0, _toConsumableArray2.default)((0, _guideline.getGuidelinesWithHighlights)(gap, _mediaSingle.MEDIA_SINGLE_SNAP_GAP, activeGuidelineKeys, guidelines)), (0, _toConsumableArray2.default)(relativeGuidelines)));
      }
    });
    (0, _defineProperty2.default)(_this, "calculateSizeState", function (size, delta) {
      var onResizeStop = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
      var calculatedWidth = Math.round(size.width + delta.width);
      var calculatedWidthWithLayout = _this.calcNewLayout(calculatedWidth, onResizeStop);
      return {
        width: calculatedWidth,
        height: calculatedWidth / _this.aspectRatio,
        layout: calculatedWidthWithLayout
      };
    });
    (0, _defineProperty2.default)(_this, "selectCurrentMediaNode", function () {
      // TODO: ED-26962 - if adding !this.props.selected, it doesn't work if media single node is at top postion
      if (_this.pos === null) {
        return;
      }
      (0, _utils.setNodeSelection)(_this.props.view, _this.pos);
    });
    (0, _defineProperty2.default)(_this, "handleResizeStart", function () {
      _this.setState({
        isResizing: true
      });
      _this.selectCurrentMediaNode();
      _this.setIsResizing(true);
      _this.updateSizeInPluginState(_this.state.size.width);
      // re-calculate guidelines
      if (_this.isGuidelineEnabled) {
        _this.updateGuidelines();
      }
    });
    (0, _defineProperty2.default)(_this, "handleResize", function (size, delta) {
      var _this$props4 = _this.props,
        layout = _this$props4.layout,
        updateSize = _this$props4.updateSize,
        lineLength = _this$props4.lineLength;
      var _this$calculateSizeSt = _this.calculateSizeState(size, delta),
        width = _this$calculateSizeSt.width,
        height = _this$calculateSizeSt.height,
        newLayout = _this$calculateSizeSt.layout;
      if (_this.isGuidelineEnabled) {
        var guidelineSnaps = (0, _guideline.getGuidelineSnaps)(_this.state.guidelines, lineLength, layout);
        _this.updateActiveGuidelines(width, _this.state.guidelines, guidelineSnaps);
        var relativeSnaps = (0, _guideline.getRelativeGuideSnaps)(_this.state.relativeGuides, _this.aspectRatio);
        _this.setState({
          size: {
            width: width,
            height: height
          },
          snaps: {
            x: [].concat((0, _toConsumableArray2.default)(guidelineSnaps.snaps.x || []), (0, _toConsumableArray2.default)(relativeSnaps))
          }
        });
      } else {
        _this.setState({
          size: {
            width: width,
            height: height
          }
        });
      }
      _this.updateSizeInPluginState(width);
      if (newLayout !== layout) {
        updateSize(width, newLayout);
      }
    });
    (0, _defineProperty2.default)(_this, "handleResizeStop", function (size, delta) {
      var _this$props5 = _this.props,
        updateSize = _this$props5.updateSize,
        dispatchAnalyticsEvent = _this$props5.dispatchAnalyticsEvent,
        nodeType = _this$props5.nodeType;
      var _this$calculateSizeSt2 = _this.calculateSizeState(size, delta, true),
        width = _this$calculateSizeSt2.width,
        height = _this$calculateSizeSt2.height,
        newLayout = _this$calculateSizeSt2.layout;
      if (dispatchAnalyticsEvent) {
        var $pos = _this.$pos;
        var event = (0, _analytics.getMediaResizeAnalyticsEvent)(nodeType || 'mediaSingle', {
          width: width,
          layout: newLayout,
          widthType: 'pixel',
          snapType: (0, _guideline.getGuidelineTypeFromKey)(_this.lastSnappedGuidelineKeys, _this.state.guidelines),
          parentNode: $pos ? $pos.parent.type.name : undefined,
          inputMethod: 'mouse'
        });
        if (event) {
          dispatchAnalyticsEvent(event);
        }
      }
      _this.setIsResizing(false);
      _this.displayGuideline([]);
      var newWidth = width;
      if (newLayout === 'full-width') {
        // When a node reaches full width in current viewport,
        // update its width with 1800 to align with pixel entry
        newWidth = _editorSharedStyles.akEditorFullWidthLayoutWidth;
      }
      _this.setState({
        isResizing: false,
        size: {
          width: newWidth,
          height: height
        }
      }, function () {
        updateSize(newWidth, newLayout);
      });
    });
    (0, _defineProperty2.default)(_this, "getMaxWidth", function () {
      var _this$props6 = _this.props,
        lineLength = _this$props6.lineLength,
        containerWidth = _this$props6.containerWidth,
        fullWidthMode = _this$props6.fullWidthMode,
        editorAppearance = _this$props6.editorAppearance,
        forceHandlePositioning = _this$props6.forceHandlePositioning;
      var isResizing = _this.state.isResizing;
      if (editorAppearance === 'chromeless' && forceHandlePositioning === 'adjacent' && (0, _platformFeatureFlags.fg)('platform_editor_inline_resize_media_to_edge')) {
        return containerWidth - _mediaSingle.MEDIA_SINGLE_ADJACENT_HANDLE_MARGIN * 2;
      }
      return !isResizing && _this.isAdjacentMode() ?
      // set undefined to fall back to 100%
      undefined : _this.calcMaxWidth(lineLength, containerWidth, fullWidthMode);
    });
    var initialWidth = props.mediaSingleWidth || _mediaSingle.DEFAULT_IMAGE_WIDTH;
    _this.state = {
      isVideoFile: true,
      isResizing: false,
      size: {
        width: initialWidth,
        height: _this.calcPxHeight(initialWidth)
      },
      snaps: {},
      relativeGuides: {},
      guidelines: []
    };
    return _this;
  }
  (0, _inherits2.default)(ResizableMediaSingleNext, _React$Component);
  return (0, _createClass2.default)(ResizableMediaSingleNext, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (prevProps.mediaSingleWidth !== this.props.mediaSingleWidth && this.props.mediaSingleWidth) {
        // update size when lineLength becomes defined later
        // ensures extended experience renders legacy image with the same size as the legacy experience
        var initialWidth = this.props.mediaSingleWidth;
        this.setState({
          size: {
            width: initialWidth,
            height: this.calcPxHeight(initialWidth)
          }
        });
      }
      return true;
    }
  }, {
    key: "componentDidMount",
    value: function () {
      var _componentDidMount = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee() {
        var viewMediaClientConfig;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              viewMediaClientConfig = this.props.viewMediaClientConfig;
              if (!viewMediaClientConfig) {
                _context.next = 4;
                break;
              }
              _context.next = 4;
              return this.checkVideoFile(viewMediaClientConfig);
            case 4:
            case "end":
              return _context.stop();
          }
        }, _callee, this);
      }));
      function componentDidMount() {
        return _componentDidMount.apply(this, arguments);
      }
      return componentDidMount;
    }()
  }, {
    key: "UNSAFE_componentWillReceiveProps",
    value: function UNSAFE_componentWillReceiveProps(nextProps) {
      if (this.props.viewMediaClientConfig !== nextProps.viewMediaClientConfig) {
        this.checkVideoFile(nextProps.viewMediaClientConfig);
      }
    }
  }, {
    key: "wrappedLayout",
    get: function get() {
      return _ui.wrappedLayouts.indexOf(this.props.layout) > -1;
    }
  }, {
    key: "pos",
    get: function get() {
      if (typeof this.props.getPos !== 'function') {
        return null;
      }
      var pos = this.props.getPos();
      if (Number.isNaN(pos) || typeof pos !== 'number') {
        return null;
      }
      return pos;
    }
  }, {
    key: "$pos",
    get: function get() {
      var pos = this.pos;
      // need to pass view because we may not get updated props in time
      return pos === null ? pos : this.props.view.state.doc.resolve(pos);
    }
  }, {
    key: "aspectRatio",
    get: function get() {
      var _this$props7 = this.props,
        width = _this$props7.width,
        height = _this$props7.height;
      if (width) {
        return width / height;
      }

      // TODO: ED-26962 - handle this case
      return 1;
    }
  }, {
    key: "insideInlineLike",
    get: function get() {
      var $pos = this.$pos;
      if (!$pos) {
        return false;
      }
      var listItem = this.props.view.state.schema.nodes.listItem;
      return !!(0, _utils2.findParentNodeOfTypeClosestToPos)($pos, [listItem]);
    }
  }, {
    key: "insideLayout",
    get: function get() {
      var $pos = this.$pos;
      if (!$pos) {
        return false;
      }
      var layoutColumn = this.props.view.state.schema.nodes.layoutColumn;
      return !!(0, _utils2.findParentNodeOfTypeClosestToPos)($pos, [layoutColumn]);
    }
  }, {
    key: "isGuidelineEnabled",
    get: function get() {
      var _this$props$pluginInj3;
      return !!((_this$props$pluginInj3 = this.props.pluginInjectionApi) !== null && _this$props$pluginInj3 !== void 0 && _this$props$pluginInj3.guideline);
    }

    // check if is inside of layout, table, expand, nestedExpand and list item
  }, {
    key: "isNestedNode",
    value: function isNestedNode() {
      var $pos = this.$pos;
      return !!($pos && $pos.depth !== 0);
    }

    // Check if adjacement mode should be activated;
  }, {
    key: "isAdjacentMode",
    value: function isAdjacentMode() {
      var _this$isNestedNode;
      if ((0, _platformFeatureFlags.fg)('platform_editor_inline_resize_media_to_edge')) {
        if (this.props.forceHandlePositioning === 'adjacent') {
          return true;
        }
      }
      return (_this$isNestedNode = this.isNestedNode()) !== null && _this$isNestedNode !== void 0 ? _this$isNestedNode : false;
    }
  }, {
    key: "getHandlePositioning",
    value: function getHandlePositioning() {
      return this.isAdjacentMode() ? 'adjacent' : undefined;
    }
  }, {
    key: "getDefaultGuidelines",
    value: function getDefaultGuidelines() {
      var _this$props8 = this.props,
        lineLength = _this$props8.lineLength,
        containerWidth = _this$props8.containerWidth,
        fullWidthMode = _this$props8.fullWidthMode;

      // disable guidelines for nested media single node
      return this.isAdjacentMode() ? [] : (0, _guideline.generateDefaultGuidelines)(lineLength, containerWidth || 0, fullWidthMode);
    }
  }, {
    key: "checkVideoFile",
    value: function () {
      var _checkVideoFile = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee2(viewMediaClientConfig) {
        var mediaNode, mediaType, isVideoFile;
        return _regenerator.default.wrap(function _callee2$(_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              if (!(this.pos === null || !viewMediaClientConfig)) {
                _context2.next = 2;
                break;
              }
              return _context2.abrupt("return");
            case 2:
              mediaNode = this.props.view.state.doc.nodeAt(this.pos + 1);
              if (!mediaNode) {
                _context2.next = 9;
                break;
              }
              _context2.next = 6;
              return (0, _checkMediaType.checkMediaType)(mediaNode, viewMediaClientConfig);
            case 6:
              _context2.t0 = _context2.sent;
              _context2.next = 10;
              break;
            case 9:
              _context2.t0 = undefined;
            case 10:
              mediaType = _context2.t0;
              isVideoFile = mediaType !== 'external' && mediaType !== 'image';
              if (this.state.isVideoFile !== isVideoFile) {
                this.setState({
                  isVideoFile: isVideoFile
                });
              }
            case 13:
            case "end":
              return _context2.stop();
          }
        }, _callee2, this);
      }));
      function checkVideoFile(_x) {
        return _checkVideoFile.apply(this, arguments);
      }
      return checkVideoFile;
    }()
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;
      var _this$props9 = this.props,
        origWidth = _this$props9.width,
        layout = _this$props9.layout,
        containerWidth = _this$props9.containerWidth,
        fullWidthMode = _this$props9.fullWidthMode,
        selected = _this$props9.selected,
        children = _this$props9.children,
        lineLength = _this$props9.lineLength,
        showLegacyNotification = _this$props9.showLegacyNotification;
      var _this$state = this.state,
        isResizing = _this$state.isResizing,
        size = _this$state.size,
        isVideoFile = _this$state.isVideoFile;
      var enable = {};
      _ui.handleSides.forEach(function (side) {
        var oppositeSide = side === 'left' ? 'right' : 'left';
        enable[side] = _utils.nonWrappedLayouts.concat("wrap-".concat(oppositeSide)).concat("align-".concat(_ui.imageAlignmentMap[oppositeSide])).indexOf(layout) > -1;
        if (side === 'left' && _this2.insideInlineLike) {
          enable[side] = false;
        }
      });

      // TODO: ED-26962 - Clean up where this lives and how it gets generated
      var className = (0, _classnames.default)(_styles.richMediaClassName, "image-".concat(layout), isResizing ? 'is-resizing' : 'not-resizing', this.props.className, {
        'richMedia-selected': selected,
        'rich-media-wrapped': layout === 'wrap-left' || layout === 'wrap-right'
      });
      var resizerNextClassName = (0, _classnames.default)(className, _styles.resizerStyles);
      var isNestedNode = this.isAdjacentMode();
      var handlePositioning = this.getHandlePositioning();
      var maxWidth = this.getMaxWidth();
      var minWidth = this.calcMinWidth(isVideoFile, lineLength);

      // while is not resizing, we take 100% as min-width if the container width is less than the min-width
      var minViewWidth = isResizing ? minWidth : "min(".concat(minWidth, "px, 100%)");
      return (0, _react2.jsx)("div", {
        // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/design-system/consistent-css-prop-usage -- Ignored via go/DSP-18766
        css: (0, _styled.wrapperStyle)({
          layout: layout,
          containerWidth: containerWidth || origWidth,
          fullWidthMode: fullWidthMode,
          mediaSingleWidth: this.state.size.width,
          isNestedNode: isNestedNode,
          isExtendedResizeExperienceOn: true
        })
      }, (0, _react2.jsx)(_resizer.ResizerNext, {
        minWidth: minViewWidth,
        maxWidth: maxWidth
        // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
        ,
        className: resizerNextClassName,
        snapGap: _mediaSingle.MEDIA_SINGLE_SNAP_GAP,
        enable: enable,
        width: size.width,
        handleResizeStart: this.handleResizeStart,
        handleResize: this.handleResize,
        handleResizeStop: this.handleResizeStop,
        snap: this.state.snaps,
        resizeRatio: _utils.nonWrappedLayouts.includes(layout) ? 2 : 1,
        "data-testid": resizerNextTestId,
        isHandleVisible: selected,
        handlePositioning: handlePositioning,
        handleHighlight: "full-height"
      }, children, showLegacyNotification && (0, _react2.jsx)(_ResizableMediaMigrationNotification.ResizableMediaMigrationNotification, null)));
    }
  }]);
}(_react.default.Component);
var calcPxHeight = function calcPxHeight(props) {
  var newWidth = props.newWidth,
    previousWidth = props.previousWidth,
    previousHeight = props.previousHeight;
  return Math.round(previousHeight / previousWidth * newWidth);
};
var calcMinWidth = function calcMinWidth(_ref2) {
  var isVideoFile = _ref2.isVideoFile,
    contentWidth = _ref2.contentWidth;
  return Math.min(contentWidth || _editorSharedStyles.akEditorDefaultLayoutWidth, isVideoFile ? _mediaSingle.MEDIA_SINGLE_VIDEO_MIN_PIXEL_WIDTH : _mediaSingle.MEDIA_SINGLE_DEFAULT_MIN_PIXEL_WIDTH);
};
var calcMaxWidth = function calcMaxWidth(_ref3) {
  var containerWidth = _ref3.containerWidth,
    editorAppearance = _ref3.editorAppearance;
  return (0, _mediaSingle.calcMediaSingleMaxWidth)(containerWidth, editorAppearance);
};
var setIsResizingPluginState = function setIsResizingPluginState(_ref4) {
  var isResizing = _ref4.isResizing,
    nodePosition = _ref4.nodePosition,
    initialWidth = _ref4.initialWidth;
  return function (state, dispatch) {
    var tr = state.tr;
    tr.setMeta(_main.MEDIA_PLUGIN_IS_RESIZING_KEY, isResizing);
    tr.setMeta('is-resizer-resizing', isResizing);
    if (isResizing && typeof nodePosition === 'number') {
      tr.setSelection(_state.NodeSelection.create(state.doc, nodePosition));
    }
    if (isResizing && typeof initialWidth === 'number') {
      tr.setMeta(_main.MEDIA_PLUGIN_RESIZING_WIDTH_KEY, initialWidth);
    }
    if (dispatch) {
      dispatch(tr);
    }
    return true;
  };
};
var calcUnwrappedLayout = function calcUnwrappedLayout(width, containerWidth, contentWidth, fullWidthMode, isNestedNode) {
  if (isNestedNode) {
    return 'center';
  }
  if (fullWidthMode) {
    if (width < contentWidth) {
      return 'center';
    }
    return 'full-width';
  }

  // handle top-level node in fixed-width editor
  if (width <= contentWidth) {
    return 'center';
  }
  if (width < Math.min(containerWidth - (0, _editorSharedStyles.akEditorGutterPaddingDynamic)() * 2, _editorSharedStyles.akEditorFullWidthLayoutWidth)) {
    return 'wide';
  }

  // set full width to be containerWidth - akEditorGutterPaddingDynamic() * 2
  // instead of containerWidth - akEditorBreakoutPadding,
  // so that we have image aligned with text
  return 'full-width';
};
var calcNewLayout = function calcNewLayout(_ref5) {
  var layout = _ref5.layout,
    containerWidth = _ref5.containerWidth,
    lineLength = _ref5.lineLength,
    fullWidthMode = _ref5.fullWidthMode,
    isNestedNode = _ref5.isNestedNode;
  return function (newWidth, stop) {
    var newPct = (0, _ui.calcPctFromPx)(newWidth, lineLength) * 100;
    var wrappedLayout = _ui.wrappedLayouts.indexOf(layout) > -1;
    if (newPct <= 100 && wrappedLayout) {
      if (!stop || newPct !== 100) {
        return layout;
      }
    }
    return calcUnwrappedLayout(newWidth, containerWidth, lineLength, fullWidthMode, isNestedNode);
  };
};
var calculateSizeState = function calculateSizeState(props) {
  return function (size, delta) {
    var onResizeStop = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
    var aspectRatio = arguments.length > 3 ? arguments[3] : undefined;
    var calculatedWidth = Math.round(size.width + delta.width);
    var calculatedWidthWithLayout = calcNewLayout(props)(calculatedWidth, onResizeStop);
    return {
      width: calculatedWidth,
      height: calculatedWidth / aspectRatio,
      layout: calculatedWidthWithLayout
    };
  };
};
var getAspectRatio = function getAspectRatio(_ref6) {
  var width = _ref6.width,
    height = _ref6.height;
  if (width && height > 0) {
    return width / height;
  }

  // TODO: ED-26962 - handle this case
  return 1;
};
var updateSizeInPluginState = (0, _throttle.default)(function (_ref7) {
  var width = _ref7.width,
    view = _ref7.view;
  var state = view.state,
    dispatch = view.dispatch;
  var tr = state.tr;
  tr.setMeta(_main.MEDIA_PLUGIN_RESIZING_WIDTH_KEY, width);
  return dispatch(tr);
}, _mediaSingle.MEDIA_SINGLE_RESIZE_THROTTLE_TIME);
var ResizableMediaSingleNextFunctional = exports.ResizableMediaSingleNextFunctional = function ResizableMediaSingleNextFunctional(props) {
  var origWidth = props.width,
    children = props.children,
    containerWidth = props.containerWidth,
    fullWidthMode = props.fullWidthMode,
    layout = props.layout,
    selected = props.selected,
    showLegacyNotification = props.showLegacyNotification,
    className = props.className,
    dispatchAnalyticsEvent = props.dispatchAnalyticsEvent,
    editorAppearance = props.editorAppearance,
    getPos = props.getPos,
    lineLength = props.lineLength,
    mediaSingleWidth = props.mediaSingleWidth,
    height = props.height,
    nodeType = props.nodeType,
    pluginInjectionApi = props.pluginInjectionApi,
    updateSize = props.updateSize,
    view = props.view,
    viewMediaClientConfig = props.viewMediaClientConfig,
    forceHandlePositioning = props.forceHandlePositioning;
  var initialWidth = (0, _react.useMemo)(function () {
    return mediaSingleWidth || _mediaSingle.DEFAULT_IMAGE_WIDTH;
  }, [mediaSingleWidth]);
  var _useState = (0, _react.useState)({
      width: initialWidth,
      height: calcPxHeight({
        newWidth: initialWidth,
        previousWidth: initialWidth,
        previousHeight: height
      })
    }),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    dimensions = _useState2[0],
    setDimensions = _useState2[1];
  var dimensionsRef = (0, _react.useRef)(dimensions);
  var lastSnappedGuidelineKeysRef = (0, _react.useRef)([]);
  var _useState3 = (0, _react.useState)({}),
    _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
    snaps = _useState4[0],
    setSnaps = _useState4[1];
  var _useState5 = (0, _react.useState)(false),
    _useState6 = (0, _slicedToArray2.default)(_useState5, 2),
    isResizing = _useState6[0],
    setIsResizing = _useState6[1];
  var _useState7 = (0, _react.useState)(!(0, _platformFeatureFlags.fg)('platform_editor_ssr_media')),
    _useState8 = (0, _slicedToArray2.default)(_useState7, 2),
    isVideoFile = _useState8[0],
    setIsVideoFile = _useState8[1];
  var nodePosition = (0, _react.useMemo)(function () {
    if (typeof getPos !== 'function') {
      return null;
    }
    var pos = getPos();
    if (Number.isNaN(pos) || typeof pos !== 'number') {
      return null;
    }
    return pos;
  }, [getPos]);
  var isNestedNode = (0, _react.useMemo)(function () {
    if (nodePosition === null) {
      return false;
    }
    var $pos = view.state.doc.resolve(nodePosition);
    return !!($pos && $pos.depth !== 0);
  }, [nodePosition, view]);
  var isAdjacentMode = (0, _react.useMemo)(function () {
    if ((0, _platformFeatureFlags.fg)('platform_editor_inline_resize_media_to_edge')) {
      if (forceHandlePositioning === 'adjacent') {
        return true;
      }
    }
    return isNestedNode;
  }, [isNestedNode, forceHandlePositioning]);
  var maybeContainerWidth = containerWidth || origWidth;
  var memoizedCss = (0, _react.useMemo)(function () {
    return (0, _styled.wrapperStyle)({
      layout: layout,
      containerWidth: maybeContainerWidth,
      fullWidthMode: fullWidthMode,
      mediaSingleWidth: dimensions.width,
      isNestedNode: isAdjacentMode,
      isExtendedResizeExperienceOn: true
    });
  }, [layout, maybeContainerWidth, fullWidthMode, dimensions.width, isAdjacentMode]);
  var maxWidth = (0, _react.useMemo)(function () {
    if (editorAppearance === 'chromeless' && forceHandlePositioning === 'adjacent' && (0, _platformFeatureFlags.fg)('platform_editor_inline_resize_media_to_edge')) {
      return containerWidth - _mediaSingle.MEDIA_SINGLE_ADJACENT_HANDLE_MARGIN * 2;
    }
    if (!isResizing && isAdjacentMode) {
      return undefined;
    }
    if (isAdjacentMode || fullWidthMode) {
      return lineLength;
    }
    return calcMaxWidth({
      containerWidth: containerWidth,
      editorAppearance: editorAppearance
    });
  }, [isAdjacentMode, fullWidthMode, lineLength, editorAppearance, containerWidth, isResizing, forceHandlePositioning]);
  var minWidth = calcMinWidth({
    isVideoFile: isVideoFile,
    contentWidth: lineLength
  });

  // while is not resizing, we take 100% as min-width if the container width is less than the min-width
  var minViewWidth = isResizing ? minWidth : "min(".concat(minWidth, "px, 100%)");
  var resizerNextClassName = (0, _react.useMemo)(function () {
    // TODO: ED-26962 - Clean up where this lives and how it gets generated
    var classNameNext = (0, _classnames.default)(_styles.richMediaClassName, "image-".concat(layout), isResizing ? 'is-resizing' : 'not-resizing', className, {
      'richMedia-selected': selected,
      'rich-media-wrapped': layout === 'wrap-left' || layout === 'wrap-right'
    });
    return (0, _classnames.default)(classNameNext, _styles.resizerStyles);
  }, [className, isResizing, layout, selected]);
  var isInsideInlineLike = (0, _react.useMemo)(function () {
    if (nodePosition === null) {
      return false;
    }
    var $pos = view.state.doc.resolve(nodePosition);
    var listItem = view.state.schema.nodes.listItem;
    return !!(0, _utils2.findParentNodeOfTypeClosestToPos)($pos, [listItem]);
  }, [nodePosition, view]);
  var enable = (0, _react.useMemo)(function () {
    return _ui.handleSides.reduce(function (acc, side) {
      var oppositeSide = side === 'left' ? 'right' : 'left';
      acc[side] = _utils.nonWrappedLayouts.concat("wrap-".concat(oppositeSide)).concat("align-".concat(_ui.imageAlignmentMap[oppositeSide])).indexOf(layout) > -1;
      if (side === 'left' && isInsideInlineLike) {
        acc[side] = false;
      }
      return acc;
    }, {});
  }, [layout, isInsideInlineLike]);
  var defaultGuidelines = (0, _react.useMemo)(function () {
    if (isAdjacentMode) {
      return [];
    }
    return (0, _guideline.generateDefaultGuidelines)(lineLength, containerWidth, fullWidthMode);
  }, [isAdjacentMode, lineLength, containerWidth, fullWidthMode]);
  var relativeGuidesRef = (0, _react.useRef)({});
  var guidelinesRef = (0, _react.useRef)([]);
  var updateGuidelines = (0, _react.useCallback)(function () {
    var _generateDynamicGuide2 = (0, _guideline.generateDynamicGuidelines)(view.state, lineLength, {
        styles: {
          lineStyle: 'dashed'
        },
        show: false
      }),
      relativeGuides = _generateDynamicGuide2.relativeGuides,
      dynamicGuides = _generateDynamicGuide2.dynamicGuides;

    // disable guidelines for nested media single node
    var dynamicGuidelines = isAdjacentMode ? [] : dynamicGuides;
    relativeGuidesRef.current = relativeGuides;
    guidelinesRef.current = [].concat((0, _toConsumableArray2.default)(defaultGuidelines), (0, _toConsumableArray2.default)(dynamicGuidelines));
  }, [view, lineLength, defaultGuidelines, isAdjacentMode]);
  var isGuidelineEnabled = (0, _react.useMemo)(function () {
    return !!(pluginInjectionApi !== null && pluginInjectionApi !== void 0 && pluginInjectionApi.guideline);
  }, [pluginInjectionApi]);
  var handleResizeStart = (0, _react.useCallback)(function () {
    setIsResizing(true);
    setIsResizingPluginState({
      isResizing: true,
      nodePosition: nodePosition,
      initialWidth: dimensionsRef.current.width
    })(view.state, view.dispatch);
    if (isGuidelineEnabled) {
      updateGuidelines();
    }
  }, [view, nodePosition, updateGuidelines, isGuidelineEnabled]);
  var getRelativeGuides = (0, _react.useCallback)(function () {
    var _pluginInjectionApi$g;
    if (typeof nodePosition !== 'number') {
      return [];
    }
    var guidelinePluginState = pluginInjectionApi === null || pluginInjectionApi === void 0 || (_pluginInjectionApi$g = pluginInjectionApi.guideline) === null || _pluginInjectionApi$g === void 0 || (_pluginInjectionApi$g = _pluginInjectionApi$g.sharedState) === null || _pluginInjectionApi$g === void 0 ? void 0 : _pluginInjectionApi$g.currentState();
    var _ref8 = (guidelinePluginState === null || guidelinePluginState === void 0 ? void 0 : guidelinePluginState.rect) || {
        top: 0,
        left: 0
      },
      topOffset = _ref8.top;
    var $pos = view.state.doc.resolve(nodePosition);
    var relativeGuides = $pos && $pos.nodeAfter && dimensionsRef.current.width ? (0, _guideline.getRelativeGuidelines)(relativeGuidesRef.current, {
      node: $pos.nodeAfter,
      pos: $pos.pos
    }, view, lineLength, topOffset, dimensionsRef.current) : [];
    return relativeGuides;
  }, [pluginInjectionApi, nodePosition, view, lineLength]);
  var updateActiveGuidelines = (0, _react.useCallback)(function () {
    var _pluginInjectionApi$g2;
    var width = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0;
    var guidelines = arguments.length > 1 ? arguments[1] : undefined;
    var guidelineSnapsReference = arguments.length > 2 ? arguments[2] : undefined;
    if (!guidelineSnapsReference.snaps.x) {
      return;
    }
    var _findClosestSnap2 = (0, _guideline.findClosestSnap)(width, guidelineSnapsReference.snaps.x, guidelineSnapsReference.guidelineReference, _mediaSingle.MEDIA_SINGLE_SNAP_GAP),
      gap = _findClosestSnap2.gap,
      activeGuidelineKeys = _findClosestSnap2.keys;
    var relativeGuidelines = activeGuidelineKeys.length ? [] : getRelativeGuides();
    lastSnappedGuidelineKeysRef.current = activeGuidelineKeys.length ? activeGuidelineKeys : relativeGuidelines.map(function (rg) {
      return rg.key;
    });
    var nextGuideLines = [].concat((0, _toConsumableArray2.default)((0, _guideline.getGuidelinesWithHighlights)(gap, _mediaSingle.MEDIA_SINGLE_SNAP_GAP, activeGuidelineKeys, guidelines)), (0, _toConsumableArray2.default)(relativeGuidelines));
    pluginInjectionApi === null || pluginInjectionApi === void 0 || (_pluginInjectionApi$g2 = pluginInjectionApi.guideline) === null || _pluginInjectionApi$g2 === void 0 || (_pluginInjectionApi$g2 = _pluginInjectionApi$g2.actions) === null || _pluginInjectionApi$g2 === void 0 || _pluginInjectionApi$g2.displayGuideline(view)({
      guidelines: nextGuideLines
    });
  }, [getRelativeGuides, pluginInjectionApi, view]);
  var aspectRatioRef = (0, _react.useRef)(getAspectRatio({
    width: props.width,
    height: props.height
  }));
  var handleResize = (0, _react.useCallback)(function (size, delta) {
    var _calculateSizeState = calculateSizeState({
        layout: layout,
        containerWidth: containerWidth,
        lineLength: lineLength,
        fullWidthMode: fullWidthMode,
        isNestedNode: isAdjacentMode
      })(size, delta, false, aspectRatioRef.current),
      width = _calculateSizeState.width,
      height = _calculateSizeState.height,
      newLayout = _calculateSizeState.layout;
    if (isGuidelineEnabled) {
      var guidelineSnaps = (0, _guideline.getGuidelineSnaps)(guidelinesRef.current, lineLength, layout);
      updateActiveGuidelines(width, guidelinesRef.current, guidelineSnaps);
      var relativeSnaps = (0, _guideline.getRelativeGuideSnaps)(relativeGuidesRef.current, aspectRatioRef.current);
      setSnaps({
        x: [].concat((0, _toConsumableArray2.default)(guidelineSnaps.snaps.x || []), (0, _toConsumableArray2.default)(relativeSnaps))
      });
    }
    setDimensions({
      width: width,
      height: height
    });
    updateSizeInPluginState({
      width: width,
      view: view
    });
    if (newLayout !== layout) {
      updateSize(width, newLayout);
    }
  }, [view, updateSize, layout, isGuidelineEnabled, containerWidth, lineLength, fullWidthMode, isAdjacentMode, updateActiveGuidelines]);
  var handleResizeStop = (0, _react.useCallback)(function (size, delta) {
    var _pluginInjectionApi$g3;
    if (typeof nodePosition !== 'number') {
      return;
    }
    var _calculateSizeState2 = calculateSizeState({
        layout: layout,
        containerWidth: containerWidth,
        lineLength: lineLength,
        fullWidthMode: fullWidthMode,
        isNestedNode: isAdjacentMode
      })(size, delta, false, aspectRatioRef.current),
      width = _calculateSizeState2.width,
      height = _calculateSizeState2.height,
      newLayout = _calculateSizeState2.layout;
    if (dispatchAnalyticsEvent) {
      var $pos = view.state.doc.resolve(nodePosition);
      var event = (0, _analytics.getMediaResizeAnalyticsEvent)(nodeType || 'mediaSingle', {
        width: width,
        layout: newLayout,
        widthType: 'pixel',
        snapType: (0, _guideline.getGuidelineTypeFromKey)(lastSnappedGuidelineKeysRef.current, guidelinesRef.current),
        parentNode: $pos ? $pos.parent.type.name : undefined,
        inputMethod: 'mouse'
      });
      if (event) {
        dispatchAnalyticsEvent(event);
      }
    }
    setIsResizing(false);
    setIsResizingPluginState({
      isResizing: false
    })(view.state, view.dispatch);
    pluginInjectionApi === null || pluginInjectionApi === void 0 || (_pluginInjectionApi$g3 = pluginInjectionApi.guideline) === null || _pluginInjectionApi$g3 === void 0 || (_pluginInjectionApi$g3 = _pluginInjectionApi$g3.actions) === null || _pluginInjectionApi$g3 === void 0 || _pluginInjectionApi$g3.displayGuideline(view)({
      guidelines: []
    });
    var newWidth = width;
    if (newLayout === 'full-width') {
      // When a node reaches full width in current viewport,
      // update its width with 1800 to align with pixel entry
      newWidth = _editorSharedStyles.akEditorFullWidthLayoutWidth;
    }
    setDimensions({
      width: newWidth,
      height: height
    });
    updateSize(newWidth, newLayout);
  }, [nodeType, dispatchAnalyticsEvent, containerWidth, fullWidthMode, isAdjacentMode, layout, lineLength, view, nodePosition, pluginInjectionApi, updateSize]);
  var mountedRef = _react.default.useRef(true);
  (0, _react.useLayoutEffect)(function () {
    mountedRef.current = true;
    return function () {
      mountedRef.current = false;
    };
  }, []);
  (0, _react.useLayoutEffect)(function () {
    setDimensions({
      width: initialWidth,
      height: calcPxHeight({
        newWidth: initialWidth,
        previousWidth: initialWidth,
        previousHeight: height
      })
    });
  }, [initialWidth, height]);
  (0, _react.useEffect)(function () {
    dimensionsRef.current = dimensions;
  }, [dimensions]);
  (0, _react.useEffect)(function () {
    if (!viewMediaClientConfig || typeof nodePosition !== 'number') {
      return;
    }
    var mediaNode = view.state.doc.nodeAt(nodePosition + 1);
    if (!mediaNode) {
      return;
    }
    (0, _checkMediaType.checkMediaType)(mediaNode, viewMediaClientConfig).then(function (mediaType) {
      if (mountedRef.current) {
        var _isVideoFile = mediaType !== 'external' && mediaType !== 'image';
        setIsVideoFile(_isVideoFile);
      }
    });
  }, [view, viewMediaClientConfig, nodePosition]);
  var handlePositioning = (0, _react.useMemo)(function () {
    if (forceHandlePositioning) {
      return forceHandlePositioning;
    }
    return isAdjacentMode ? 'adjacent' : undefined;
  }, [forceHandlePositioning, isAdjacentMode]);
  return (0, _react2.jsx)("div", {
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/design-system/consistent-css-prop-usage -- Ignored via go/DSP-18766
    css: memoizedCss
  }, (0, _react2.jsx)(_resizer.ResizerNext, {
    minWidth: minViewWidth,
    maxWidth: maxWidth
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    ,
    className: resizerNextClassName,
    snapGap: _mediaSingle.MEDIA_SINGLE_SNAP_GAP,
    enable: enable,
    width: dimensions.width,
    handleResizeStart: handleResizeStart,
    handleResize: handleResize,
    handleResizeStop: handleResizeStop,
    snap: snaps,
    resizeRatio: _utils.nonWrappedLayouts.includes(layout) ? 2 : 1,
    "data-testid": resizerNextTestId,
    isHandleVisible: selected,
    handlePositioning: handlePositioning,
    handleHighlight: "full-height"
  }, children, showLegacyNotification && (0, _react2.jsx)(_ResizableMediaMigrationNotification.ResizableMediaMigrationNotification, null)));
};
var ResizableMediaSingleToggle = function ResizableMediaSingleToggle(_ref9) {
  var allowBreakoutSnapPoints = _ref9.allowBreakoutSnapPoints,
    children = _ref9.children,
    className = _ref9.className,
    containerWidth = _ref9.containerWidth,
    dataAttributes = _ref9.dataAttributes,
    disableHandles = _ref9.disableHandles,
    dispatchAnalyticsEvent = _ref9.dispatchAnalyticsEvent,
    editorAppearance = _ref9.editorAppearance,
    fullWidthMode = _ref9.fullWidthMode,
    getPos = _ref9.getPos,
    gridSize = _ref9.gridSize,
    handleMediaSingleRef = _ref9.handleMediaSingleRef,
    hasFallbackContainer = _ref9.hasFallbackContainer,
    height = _ref9.height,
    isInsideOfInlineExtension = _ref9.isInsideOfInlineExtension,
    isLoading = _ref9.isLoading,
    layout = _ref9.layout,
    lineLength = _ref9.lineLength,
    mediaSingleWidth = _ref9.mediaSingleWidth,
    nodeType = _ref9.nodeType,
    pctWidth = _ref9.pctWidth,
    pluginInjectionApi = _ref9.pluginInjectionApi,
    selected = _ref9.selected,
    showLegacyNotification = _ref9.showLegacyNotification,
    size = _ref9.size,
    updateSize = _ref9.updateSize,
    view = _ref9.view,
    viewMediaClientConfig = _ref9.viewMediaClientConfig,
    width = _ref9.width,
    forceHandlePositioning = _ref9.forceHandlePositioning;
  if ((0, _platformFeatureFlags.fg)('platform_editor_react18_phase2__media_single') || (0, _platformFeatureFlags.fg)('platform_editor_react18_phase2__media_single_jira')) {
    return (0, _react2.jsx)(ResizableMediaSingleNextFunctional, {
      allowBreakoutSnapPoints: allowBreakoutSnapPoints
      // eslint-disable-next-line react/no-children-prop
      ,
      children: children
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
      ,
      className: className,
      containerWidth: containerWidth,
      dataAttributes: dataAttributes,
      disableHandles: disableHandles,
      dispatchAnalyticsEvent: dispatchAnalyticsEvent,
      editorAppearance: editorAppearance,
      fullWidthMode: fullWidthMode,
      getPos: getPos,
      gridSize: gridSize,
      handleMediaSingleRef: handleMediaSingleRef,
      hasFallbackContainer: hasFallbackContainer,
      height: height,
      isInsideOfInlineExtension: isInsideOfInlineExtension,
      isLoading: isLoading,
      layout: layout,
      lineLength: lineLength,
      mediaSingleWidth: mediaSingleWidth,
      nodeType: nodeType,
      pctWidth: pctWidth,
      pluginInjectionApi: pluginInjectionApi,
      selected: selected,
      showLegacyNotification: showLegacyNotification,
      size: size,
      updateSize: updateSize,
      view: view,
      viewMediaClientConfig: viewMediaClientConfig,
      width: width,
      forceHandlePositioning: forceHandlePositioning
    });
  }
  return (0, _react2.jsx)(ResizableMediaSingleNext, {
    allowBreakoutSnapPoints: allowBreakoutSnapPoints
    // eslint-disable-next-line react/no-children-prop
    ,
    children: children
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    ,
    className: className,
    containerWidth: containerWidth,
    dataAttributes: dataAttributes,
    disableHandles: disableHandles,
    dispatchAnalyticsEvent: dispatchAnalyticsEvent,
    editorAppearance: editorAppearance,
    fullWidthMode: fullWidthMode,
    getPos: getPos,
    gridSize: gridSize,
    handleMediaSingleRef: handleMediaSingleRef,
    hasFallbackContainer: hasFallbackContainer,
    height: height,
    isInsideOfInlineExtension: isInsideOfInlineExtension,
    isLoading: isLoading,
    layout: layout,
    lineLength: lineLength,
    mediaSingleWidth: mediaSingleWidth,
    nodeType: nodeType,
    pctWidth: pctWidth,
    pluginInjectionApi: pluginInjectionApi,
    selected: selected,
    showLegacyNotification: showLegacyNotification,
    size: size,
    updateSize: updateSize,
    view: view,
    viewMediaClientConfig: viewMediaClientConfig,
    width: width,
    forceHandlePositioning: forceHandlePositioning
  });
};
var _default = exports.default = ResizableMediaSingleToggle;
