// @atlaskit/editor-plugin-media/dist/cjs/ui/ResizableMediaSingle/styles.js
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.resizableMediaMigrationNotificationStyle = void 0;
var _react = require("@emotion/react");
var colors = _interopRequireWildcard(require("@atlaskit/theme/colors"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

// Ignored via go/ees005
// eslint-disable-next-line import/no-namespace

var NOTIFICATION_SIZE = 8;
var SPACE_FROM_EDGE = "var(--ds-space-100, 8px)";

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var resizableMediaMigrationNotificationStyle = exports.resizableMediaMigrationNotificationStyle = (0, _react.css)({
  position: 'absolute',
  top: SPACE_FROM_EDGE,
  right: "calc(".concat("var(--ds-space-150, 12px)", " + ", SPACE_FROM_EDGE, ")"),
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
  backgroundColor: "var(--ds-background-warning-bold, ".concat(colors.Y300, ")"),
  borderRadius: "var(--ds-border-radius-circle, 50%)",
  width: "".concat(NOTIFICATION_SIZE, "px"),
  height: "".concat(NOTIFICATION_SIZE, "px"),
  pointerEvents: 'none'
});
