// @atlaskit/editor-plugin-card/dist/es2019/ui/ToolbarViewedEvent.js
import React, { useEffect } from 'react';
import { useAnalyticsEvents } from '@atlaskit/analytics-next/useAnalyticsEvents';
import { useSmartLinkContext } from '@atlaskit/link-provider';
import { getResolvedAttributesFromStore } from '../pm-plugins/utils';
import { EditorAnalyticsContext } from './EditorAnalyticsContext';
/**
 * Handles firing the toolbar viewed event
 */
const ToolbarViewedEventBase = ({
  url,
  display,
  cardContext
}) => {
  const {
    createAnalyticsEvent
  } = useAnalyticsEvents();
  const store = cardContext === null || cardContext === void 0 ? void 0 : cardContext.store;
  useEffect(() => {
    createAnalyticsEvent({
      action: 'viewed',
      actionSubject: 'inlineDialog',
      actionSubjectId: 'editLinkToolbar',
      eventType: 'ui',
      attributes: {
        ...getResolvedAttributesFromStore(url, display, store),
        display
      }
    }).fire('media');
  }, [createAnalyticsEvent, display, url, store]);
  return null;
};

/**
 * Provides analytics context and card context
 */
export const ToolbarViewedEvent = ({
  url,
  display,
  editorView
}) => {
  const cardContext = useSmartLinkContext();
  return /*#__PURE__*/React.createElement(EditorAnalyticsContext, {
    editorView: editorView
  }, url ? /*#__PURE__*/React.createElement(ToolbarViewedEventBase, {
    url: url,
    display: display,
    cardContext: cardContext
  }) : null);
};
