// @atlaskit/user-picker/dist/es2019/components/TeamOption/main.js
import _extends from "@babel/runtime/helpers/extends";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
import { B400, N800, N200 } from '@atlaskit/theme/colors';
import { fg } from '@atlaskit/platform-feature-flags';
import { Inline } from '@atlaskit/primitives/compiled';
import { VerifiedTeamIcon } from '@atlaskit/people-teams-ui-public/verified-team-icon';
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import { jsx } from '@emotion/react';
import React from 'react';
import { FormattedMessage } from 'react-intl-next';
import { AvatarItemOption, textWrapper } from '../AvatarItemOption';
import { HighlightText } from '../HighlightText';
import { messages } from '../i18n';
import { SizeableAvatar } from '../SizeableAvatar';
export class TeamOption extends React.PureComponent {
  constructor(...args) {
    super(...args);
    _defineProperty(this, "getPrimaryText", () => {
      const {
        team: {
          name,
          highlight,
          verified
        }
      } = this.props;
      return [jsx("span", {
        key: "name"
        // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
        ,
        css: textWrapper(this.props.isSelected ? `var(--ds-text-selected, ${B400})` : `var(--ds-text, ${N800})`)
      }, fg('verified-team-in-user-picker') ? jsx(Inline, {
        alignBlock: "center"
      }, jsx(HighlightText, {
        highlights: highlight && highlight.name
      }, name), verified && jsx(VerifiedTeamIcon, null)) : jsx(HighlightText, {
        highlights: highlight && highlight.name
      }, name))];
    });
    _defineProperty(this, "renderByline", () => {
      const {
        isSelected,
        team: {
          memberCount,
          includesYou
        }
      } = this.props;

      // if Member count is missing, do not show the byline, regardless of the availability of includesYou
      if (memberCount === null || typeof memberCount === 'undefined') {
        return undefined;
      } else {
        if (includesYou === true) {
          if (memberCount > 50) {
            return this.getBylineComponent(isSelected, jsx(FormattedMessage, messages.plus50MembersWithYou));
          } else {
            return this.getBylineComponent(isSelected, jsx(FormattedMessage, _extends({}, messages.memberCountWithYou, {
              values: {
                count: memberCount
              }
            })));
          }
        } else {
          if (memberCount > 50) {
            return this.getBylineComponent(isSelected, jsx(FormattedMessage, messages.plus50MembersWithoutYou));
          } else {
            return this.getBylineComponent(isSelected, jsx(FormattedMessage, _extends({}, messages.memberCountWithoutYou, {
              values: {
                count: memberCount
              }
            })));
          }
        }
      }
    });
    _defineProperty(this, "getBylineComponent", (isSelected, message) => jsx("span", {
      // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
      css: textWrapper(isSelected ? `var(--ds-text-selected, ${B400})` : `var(--ds-text-subtlest, ${N200})`),
      "data-testid": "user-picker-team-secondary-text"
    }, message));
    _defineProperty(this, "renderAvatar", () => {
      const {
        team: {
          avatarUrl
        }
      } = this.props;
      return jsx(SizeableAvatar, {
        appearance: "big",
        src: avatarUrl,
        type: "team"
      });
    });
    _defineProperty(this, "getLozengeProps", () => typeof this.props.team.lozenge === 'string' ? {
      text: this.props.team.lozenge
    } : this.props.team.lozenge);
    _defineProperty(this, "renderCustomByLine", () => {
      var _this$props$team;
      if (!((_this$props$team = this.props.team) !== null && _this$props$team !== void 0 && _this$props$team.byline)) {
        return undefined;
      }
      return jsx("span", {
        // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
        css: textWrapper(this.props.isSelected ? `var(--ds-text-selected, ${B400})` : `var(--ds-text-subtlest, ${N200})`)
      }, this.props.team.byline);
    });
  }
  render() {
    return jsx(AvatarItemOption, {
      avatar: this.renderAvatar(),
      isDisabled: this.props.team.isDisabled,
      lozenge: this.getLozengeProps(),
      primaryText: this.getPrimaryText(),
      secondaryText: this.renderCustomByLine() || this.renderByline()
    });
  }
}
