// @atlaskit/link-datasource/dist/cjs/ui/common/modal/datasources-table-in-modal-preview/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _types = require("../../../../analytics/types");
var _userInteractions = require("../../../../contexts/user-interactions");
var _issueLikeTable = require("../../../issue-like-table");
var _datasourceContext = require("../datasource-context");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var Table = function Table(props) {
  var _useDatasourceContext = (0, _datasourceContext.useDatasourceContext)(),
    columnCustomSizes = _useDatasourceContext.columnCustomSizes,
    onColumnResize = _useDatasourceContext.onColumnResize,
    wrappedColumnKeys = _useDatasourceContext.wrappedColumnKeys,
    onWrappedColumnChange = _useDatasourceContext.onWrappedColumnChange,
    visibleColumnKeys = _useDatasourceContext.visibleColumnKeys,
    onVisibleColumnKeysChange = _useDatasourceContext.onVisibleColumnKeysChange,
    _useDatasourceContext2 = _useDatasourceContext.tableState,
    status = _useDatasourceContext2.status,
    onNextPage = _useDatasourceContext2.onNextPage,
    responseItems = _useDatasourceContext2.responseItems,
    responseItemIds = _useDatasourceContext2.responseItemIds,
    hasNextPage = _useDatasourceContext2.hasNextPage,
    columns = _useDatasourceContext2.columns,
    defaultVisibleColumnKeys = _useDatasourceContext2.defaultVisibleColumnKeys,
    loadDatasourceDetails = _useDatasourceContext2.loadDatasourceDetails,
    _useDatasourceContext3 = _useDatasourceContext2.extensionKey,
    extensionKey = _useDatasourceContext3 === void 0 ? null : _useDatasourceContext3;
  var userInteractions = (0, _userInteractions.useUserInteractions)();
  var handleOnNextPage = (0, _react.useCallback)(function () {
    var onNextPageProps = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    userInteractions.add(_types.DatasourceAction.NEXT_PAGE_SCROLLED);
    onNextPage(onNextPageProps);
  }, [onNextPage, userInteractions]);
  return /*#__PURE__*/_react.default.createElement(_issueLikeTable.IssueLikeDataTableView, (0, _extends2.default)({}, props, {
    status: status,
    columns: columns,
    items: responseItems,
    itemIds: responseItemIds,
    hasNextPage: hasNextPage,
    visibleColumnKeys: visibleColumnKeys || defaultVisibleColumnKeys,
    onNextPage: handleOnNextPage,
    onLoadDatasourceDetails: loadDatasourceDetails,
    onVisibleColumnKeysChange: onVisibleColumnKeysChange,
    extensionKey: extensionKey,
    columnCustomSizes: columnCustomSizes,
    onColumnResize: onColumnResize,
    wrappedColumnKeys: wrappedColumnKeys,
    onWrappedColumnChange: onWrappedColumnChange
  }));
};
var _default = exports.default = Table;
