// @atlaskit/link-datasource/dist/cjs/ui/common/modal/search-count/messages.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.searchCountMessages = void 0;
var _reactIntlNext = require("react-intl-next");
var searchCountMessages = exports.searchCountMessages = (0, _reactIntlNext.defineMessages)({
  resultCountText: {
    id: 'linkDataSource.search.configmodal.resultCountText',
    description: 'Text that indicates the number of search results.',
    defaultMessage: '{searchCount, plural, one {# result} other {# results}}'
  },
  issueCountText: {
    id: 'linkDataSource.search.configmodal.issueCountText',
    description: 'Text that indicates the number of jira issues in the search result.',
    defaultMessage: '{searchCount, plural, one {# issue} other {# issues}}'
  },
  itemCountText: {
    id: 'linkDataSource.search.configmodal.itemCountText',
    description: 'Text that indicates the number of items in the search result.',
    defaultMessage: '{searchCount, plural, one {# item} other {# items}}'
  },
  issueCountTextIssueTermRefresh: {
    id: 'linkDataSource.search.configmodal.issueCountText-issue-term-refresh',
    description: 'Text that indicates the number of jira issues in the search result.',
    defaultMessage: '{searchCount, plural, one {# work item} other {# work items}}'
  }
});
