// @atlaskit/media-card/dist/cjs/utils/cardActions/cardActionsDropdownMenu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var react_2 = require("react");
var more_1 = tslib_1.__importDefault(require("@atlaskit/icon/glyph/more"));
var dropdown_menu_1 = tslib_1.__importStar(require("@atlaskit/dropdown-menu"));
var styled_1 = require("./styled");
var analytics_next_1 = require("@atlaskit/analytics-next");
var analytics_1 = require("../analytics");
var CardActionButtonWithProps = function (props) { return (react_1.default.createElement(styled_1.CardActionButton, tslib_1.__assign({}, props))); };
var CardActionButtonWithAnalytics = analytics_next_1.withAnalyticsEvents({
    onClick: analytics_1.createAndFireMediaEvent({
        eventType: 'ui',
        action: 'clicked',
        actionSubject: 'button',
        actionSubjectId: 'mediaCardDropDownMenu',
    }),
})(CardActionButtonWithProps);
var DropdownItemWithProps = function (props) { return (react_1.default.createElement(dropdown_menu_1.DropdownItem, tslib_1.__assign({ "data-testid": "media-card-actions-menu-item" }, props))); };
var createDropdownItemWithAnalytics = function (action, index) {
    var label = action.label, handler = action.handler;
    var DropdownItemWithAnalytics = analytics_next_1.withAnalyticsEvents({
        onClick: analytics_1.createAndFireMediaEvent({
            eventType: 'ui',
            action: 'clicked',
            actionSubject: 'button',
            actionSubjectId: 'mediaCardDropDownMenuItem',
            attributes: {
                label: label,
            },
        }),
    })(DropdownItemWithProps);
    return (react_1.default.createElement(DropdownItemWithAnalytics, { key: index, onClick: handler }, label));
};
var CardActionsDropdownMenu = /** @class */ (function (_super) {
    tslib_1.__extends(CardActionsDropdownMenu, _super);
    function CardActionsDropdownMenu() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CardActionsDropdownMenu.prototype.render = function () {
        var _a = this.props, actions = _a.actions, triggerColor = _a.triggerColor, onOpenChange = _a.onOpenChange;
        if (actions.length > 0) {
            return (react_1.default.createElement(dropdown_menu_1.default, { "data-testid": "media-card-actions-menu", onOpenChange: onOpenChange, trigger: react_1.default.createElement(CardActionButtonWithAnalytics, { style: { color: triggerColor } },
                    react_1.default.createElement(more_1.default, { label: "more" })) },
                react_1.default.createElement(dropdown_menu_1.DropdownItemGroup, null, actions.map(createDropdownItemWithAnalytics))));
        }
        else {
            return null;
        }
    };
    return CardActionsDropdownMenu;
}(react_2.Component));
exports.CardActionsDropdownMenu = CardActionsDropdownMenu;
