// @atlaskit/editor-plugin-table/dist/cjs/nodeviews/TableNodeViewBase.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _model = require("@atlaskit/editor-prosemirror/model");
var TableNodeView = exports.default = /*#__PURE__*/(0, _createClass2.default)(
/**
 * @constructor
 */
function TableNodeView(node, view, getPos, eventDispatcher) {
  (0, _classCallCheck2.default)(this, TableNodeView);
  this.node = node;
  this.view = view;
  this.getPos = getPos;
  this.eventDispatcher = eventDispatcher;
  // Ignored via go/ees005
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  var _DOMSerializer$render = _model.DOMSerializer.renderSpec(document, node.type.spec.toDOM(node)),
    dom = _DOMSerializer$render.dom,
    contentDOM = _DOMSerializer$render.contentDOM;
  this.dom = dom;
  this.contentDOM = contentDOM;
}

/**
 * Variables
 */);
