// @atlaskit/link-datasource/dist/es2019/ui/issue-like-table/shared-components/avatar/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import "./index.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
import React from 'react';
import { FormattedMessage, useIntl } from 'react-intl-next';
import Avatar from '@atlaskit/avatar';
import { Box } from '@atlaskit/primitives/compiled';
import Tooltip from '@atlaskit/tooltip';
import { USER_TYPE_TEST_ID } from '../../render-type/user';
import { userTypeMessages } from '../../render-type/user/messages';
const styles = {
  avatarWrapperStyles: "_2hwxu2gc",
  emptyAvatarWrapperStyles: "_11c82smr _1rjc1b66 _18zru2gc _1e0c1txw _4cvr1h6o _tzy4idpf _16l4kb7n"
};
export const EmptyAvatar = () => {
  const intl = useIntl();
  return /*#__PURE__*/React.createElement(Tooltip, {
    content: intl.formatMessage(userTypeMessages.userDefaultdisplayNameValue)
  }, /*#__PURE__*/React.createElement(Box, {
    xcss: styles.emptyAvatarWrapperStyles,
    testId: USER_TYPE_TEST_ID
  }, /*#__PURE__*/React.createElement(Box, {
    xcss: styles.avatarWrapperStyles
  }, /*#__PURE__*/React.createElement(Avatar, {
    appearance: "circle",
    size: 'small',
    testId: `${USER_TYPE_TEST_ID}--avatar`
  })), /*#__PURE__*/React.createElement(FormattedMessage, userTypeMessages.userDefaultdisplayNameValue)));
};
