// @atlaskit/media-ui/dist/cjs/BlockCard/ResolvedView/PreviewView/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var react_render_image_1 = tslib_1.__importDefault(require("react-render-image"));
var image_1 = tslib_1.__importDefault(require("@atlaskit/icon/glyph/image"));
var NoImageIcon_1 = tslib_1.__importDefault(require("./NoImageIcon"));
var styled_1 = require("./styled");
exports.LoadingView = function () { return (react_1.default.createElement(styled_1.IconWrapper, null,
    react_1.default.createElement(image_1.default, { size: "xlarge", label: "loading" }))); };
exports.NoImageView = function () { return (react_1.default.createElement(styled_1.IconWrapper, null,
    react_1.default.createElement(NoImageIcon_1.default, null))); };
exports.LoadedView = function (_a) {
    var url = _a.url;
    return (react_1.default.createElement(styled_1.ImageWrapper, { url: url }));
};
var PreviewView = /** @class */ (function (_super) {
    tslib_1.__extends(PreviewView, _super);
    function PreviewView() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PreviewView.prototype.renderContent = function () {
        var url = this.props.url;
        if (!url) {
            return react_1.default.createElement(exports.NoImageView, null);
        }
        return (react_1.default.createElement(react_render_image_1.default, { src: url, loading: react_1.default.createElement(exports.LoadingView, null), loaded: react_1.default.createElement(exports.LoadedView, { url: url }), errored: react_1.default.createElement(exports.NoImageView, null) }));
    };
    PreviewView.prototype.render = function () {
        return react_1.default.createElement(styled_1.Wrapper, null, this.renderContent());
    };
    return PreviewView;
}(react_1.default.Component));
exports.PreviewView = PreviewView;
