// @atlaskit/editor-performance-metrics/dist/cjs/internals/editorPerformanceObserver.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EditorPerformanceObserver = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _dom = require("./dom");
var _DOMObservers = require("./DOMObservers");
var _firstInteractionObserver = require("./firstInteractionObserver");
var _getElementName = require("./getElementName");
var _userEventsObserver = require("./userEventsObserver");
var _wrapperFetch = require("./wrapperFetch");
var _wrapperTimers = require("./wrapperTimers");
/* eslint-disable compat/compat */

var defaultOptions = {
  timers: {
    setTimeout: {
      maxTimeoutAllowedToTrack: 2000 // two seconds
    }
  }
};
var EditorPerformanceObserver = exports.EditorPerformanceObserver = /*#__PURE__*/function () {
  function EditorPerformanceObserver(timeline, givenOptions) {
    var _this = this;
    (0, _classCallCheck2.default)(this, EditorPerformanceObserver);
    (0, _defineProperty2.default)(this, "startTime", null);
    (0, _defineProperty2.default)(this, "isStarted", false);
    (0, _defineProperty2.default)(this, "observedTargetRef", null);
    (0, _defineProperty2.default)(this, "wrapperApplied", false);
    (0, _defineProperty2.default)(this, "wrapperCleanupFunctions", []);
    this.options = Object.assign(defaultOptions, givenOptions || {});
    this.timeline = timeline;
    this.vcSections = new WeakSet();
    // TODO: ED-26959 - make sure to get it from config
    var selectorConfig = {
      id: false,
      testId: true,
      role: false,
      className: false,
      dataVC: true
    };
    this.userEventsObserver = new _userEventsObserver.UserEventsObserver({
      onEventEntries: function onEventEntries(entries) {
        entries.forEach(function (event) {
          var elementName = 'not found';
          var element = event.elementRef.deref();
          if (element) {
            elementName = (0, _getElementName.getElementName)(selectorConfig, element);
          }
          var eventCategory = (0, _dom.getEventCategory)(event.eventName);
          timeline.markEvent({
            type: "user-event:".concat(eventCategory),
            startTime: event.startTime,
            data: {
              category: eventCategory,
              eventName: event.eventName,
              duration: event.duration,
              elementName: elementName
            }
          });
        });
      }
    });
    this.firstInteraction = new _firstInteractionObserver.FirstInteractionObserver(this.onFirstInteraction.bind(this));
    this.domObservers = new _DOMObservers.DOMObservers({
      timeline: this.timeline,
      onChange: function onChange(_ref) {
        var _this$observedTargetR;
        var startTime = _ref.startTime,
          elementRef = _ref.elementRef,
          visible = _ref.visible,
          rect = _ref.rect,
          previousRect = _ref.previousRect,
          source = _ref.source;
        if (!visible) {
          // Ignored via go/ees007
          // eslint-disable-next-line @atlaskit/editor/enforce-todo-comment-format
          // TODO: when element is not visible we probably wants to add something
          return;
        }
        var element = elementRef.deref();
        if (!element) {
          return;
        }
        var elementName = (0, _getElementName.getElementName)(selectorConfig, element);
        var observedTarget = (_this$observedTargetR = _this.observedTargetRef) === null || _this$observedTargetR === void 0 ? void 0 : _this$observedTargetR.deref();
        var wrapperSection = (0, _dom.searchAncestors)(element, function (e) {
          var hasSection = _this.vcSections.has(e);
          if (hasSection) {
            return 'found';
          }
          var isRootElement = e.parentElement === observedTarget;
          if (isRootElement) {
            return 'abort';
          }
          return 'continue';
        });
        if (!wrapperSection && element) {
          _this.vcSections.add(element);
        }
        var wrapperSectionName = (0, _getElementName.getElementName)(selectorConfig, wrapperSection || element);
        _this.timeline.markEvent({
          type: 'element:changed',
          startTime: startTime,
          data: {
            wrapperSectionName: wrapperSectionName,
            elementName: elementName,
            rect: rect,
            previousRect: previousRect,
            source: source
          }
        });
      },
      onDOMContentChange: function onDOMContentChange(_ref2) {
        var _this$observedTargetR2;
        var targetRef = _ref2.targetRef;
        var mutationTarget = targetRef.deref();
        if (!mutationTarget) {
          return;
        }
        var observedTarget = (_this$observedTargetR2 = _this.observedTargetRef) === null || _this$observedTargetR2 === void 0 ? void 0 : _this$observedTargetR2.deref();
        var isRootChange = mutationTarget.parentElement === observedTarget;
        if (!isRootChange) {
          _this.vcSections.add(mutationTarget);
        }
      }
    });
    this.timeline.onceAllSubscribersCleaned(function () {
      _this.cleanupWrappers();
    });
  }
  return (0, _createClass2.default)(EditorPerformanceObserver, [{
    key: "onFirstInteraction",
    value: function onFirstInteraction(_ref3) {
      var event = _ref3.event,
        time = _ref3.time;
      this.timeline.markEvent({
        type: 'abort:user-interaction',
        startTime: time,
        data: {
          source: event
        }
      });
      this.domObservers.disconnect();
    }
  }, {
    key: "start",
    value: function start(_ref4) {
      var startTime = _ref4.startTime;
      if (this.isStarted) {
        return;
      }
      var target = document.body;
      this.observedTargetRef = new WeakRef(target);
      this.startTime = startTime;
      this.domObservers.observe(target);
      this.isStarted = true;
    }
  }, {
    key: "onIdleBuffer",
    value: function onIdleBuffer(cb) {
      this.applyWrappersOnce();
      return this.timeline.onIdleBufferFlush(cb);
    }
  }, {
    key: "onceNextIdle",
    value: function onceNextIdle(cb) {
      this.applyWrappersOnce();
      return this.timeline.onNextIdle(cb);
    }
  }, {
    key: "stop",
    value: function stop() {
      if (!this.isStarted) {
        return;
      }
      this.isStarted = false;
      this.domObservers.disconnect();
      this.firstInteraction.disconnect();
      this.userEventsObserver.disconnect();
      this.cleanupWrappers();
      this.timeline.cleanupSubscribers();
    }
  }, {
    key: "applyWrappersOnce",
    value: function applyWrappersOnce() {
      if (this.wrapperApplied) {
        return;
      }
      var timersCleanup = (0, _wrapperTimers.wrapperTimers)({
        globalContext: globalThis || window,
        timelineHoldable: this.timeline,
        maxTimeoutAllowed: this.options.timers.setTimeout.maxTimeoutAllowedToTrack
      });
      var fetchCleanup = (0, _wrapperFetch.wrapperFetch)({
        globalContext: globalThis || window,
        timelineHoldable: this.timeline
      });
      this.wrapperCleanupFunctions.push(timersCleanup, fetchCleanup);
      this.wrapperApplied = true;
    }
  }, {
    key: "cleanupWrappers",
    value: function cleanupWrappers() {
      this.wrapperCleanupFunctions.forEach(function (cleanup) {
        return cleanup();
      });
      this.wrapperCleanupFunctions = [];
      this.wrapperApplied = false;
    }
  }]);
}();
