// @atlaskit/editor-performance-metrics/dist/cjs/internals/userEventsObserver.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.processEntry = exports.createPerformanceObserver = exports.UserEventsObserver = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _backgroundTasks = require("./backgroundTasks");
var ALLOWED_DURATION = 40;
var processEntry = exports.processEntry = function processEntry(entry) {
  var startTime = entry.startTime,
    processingEnd = entry.processingEnd,
    processingStart = entry.processingStart,
    target = entry.target,
    duration = entry.duration,
    name = entry.name;
  if (typeof processingEnd !== 'number' || typeof processingStart !== 'number') {
    return null;
  }
  var timeToProcessEvent = processingEnd - processingStart;
  if (timeToProcessEvent <= 0) {
    return null;
  }
  if (!(target instanceof HTMLElement)) {
    return null;
  }
  var eventProcessingDuration = processingEnd - startTime;
  if (duration <= ALLOWED_DURATION || eventProcessingDuration <= ALLOWED_DURATION) {
    return null;
  }
  return {
    eventName: name,
    startTime: startTime,
    duration: duration,
    elementRef: new WeakRef(target)
  };
};
var createPerformanceObserver = exports.createPerformanceObserver = function createPerformanceObserver(cb) {
  return new PerformanceObserver(cb);
};
var UserEventsObserver = exports.UserEventsObserver = /*#__PURE__*/function () {
  function UserEventsObserver(_ref) {
    var onEventEntries = _ref.onEventEntries;
    (0, _classCallCheck2.default)(this, UserEventsObserver);
    if (typeof globalThis.PerformanceObserver !== 'function') {
      return;
    }
    this.observer = createPerformanceObserver(function (list) {
      (0, _backgroundTasks.backgroundTask)(function () {
        var mappedEvents = list.getEntries().reduce(function (acc, value) {
          var transformed = processEntry(value);
          if (transformed) {
            acc.push(transformed);
          }
          return acc;
        }, []);
        onEventEntries(mappedEvents);
      });
    });
    this.observer.observe({
      type: 'event',
      buffered: true,
      // @ts-expect-error
      durationThreshold: 16
    });
  }
  return (0, _createClass2.default)(UserEventsObserver, [{
    key: "disconnect",
    value: function disconnect() {
      var _this$observer;
      (_this$observer = this.observer) === null || _this$observer === void 0 || _this$observer.disconnect();
    }
  }]);
}();
