// @atlaskit/frontend-utilities/dist/cjs/network/retry-operation/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DEFAULT_RETRIES", {
  enumerable: true,
  get: function get() {
    return _constants.DEFAULT_RETRIES;
  }
});
Object.defineProperty(exports, "FailedFetchError", {
  enumerable: true,
  get: function get() {
    return _main.FailedFetchError;
  }
});
Object.defineProperty(exports, "LAZY_LOAD_RETRIES", {
  enumerable: true,
  get: function get() {
    return _constants.LAZY_LOAD_RETRIES;
  }
});
Object.defineProperty(exports, "NO_RETRIES", {
  enumerable: true,
  get: function get() {
    return _constants.NO_RETRIES;
  }
});
Object.defineProperty(exports, "UP_TO_TWO_INSTANT_RETRIES", {
  enumerable: true,
  get: function get() {
    return _constants.UP_TO_TWO_INSTANT_RETRIES;
  }
});
Object.defineProperty(exports, "retryOnException", {
  enumerable: true,
  get: function get() {
    return _main.retryOnException;
  }
});
var _main = require("./main");
var _constants = require("./constants");
