// @atlaskit/smart-card/dist/cjs/view/FlexibleCard/components/blocks/element-group/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
/* eslint-disable @atlaskit/design-system/use-tokens-typography */
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./index.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _constants = require("../../../../../constants");
var _utils = require("../utils");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var alignmentStyleMap = {
  right: "_1dthh9n0 _r291h9n0 _1bahesu3 _y3gnusic",
  left: "_1dthv2br _r291v2br _1bah1y6m _y3gn1e5h"
};

// TODO: Remove on fg cleanup: platform-linking-visual-refresh-v1
var baseStyleBySizeOld = {
  xlarge: "_zulp147b _1reo15vq _18m915vq _4cvr1h6o _1e0c1txw _1ul9idpf _vwz41j6v",
  large: "_zulp1e9d _1reo15vq _18m915vq _4cvr1h6o _1e0c1txw _1ul9idpf _vwz41j6v",
  medium: "_zulpcxkx _1reo15vq _18m915vq _4cvr1h6o _1e0c1txw _1ul9idpf _vwz41j6v",
  small: "_zulppdf9 _1reo15vq _18m915vq _4cvr1h6o _1e0c1txw _1ul9idpf _vwz41j6v"
};
var baseStyleBySize = {
  xlarge: "_zulpv47k _1reo15vq _18m915vq _4cvr1h6o _1e0c1txw _1ul9idpf",
  large: "_zulppxbi _1reo15vq _18m915vq _4cvr1h6o _1e0c1txw _1ul9idpf",
  medium: "_zulpu2gc _1reo15vq _18m915vq _4cvr1h6o _1e0c1txw _1ul9idpf",
  small: "_zulp1b66 _1reo15vq _18m915vq _4cvr1h6o _1e0c1txw _1ul9idpf"
};
var baseStyleByDirection = {
  horizontal: "_2lx2vrvc _4cvr1h6o",
  vertical: "_2lx21bp4 _4cvr1y6m"
};
var baseStyleOld = null;
var baseStyles = null;
var widthStyle = {
  flexible: "_16jlkb7n _1o9z11wp _i0dlf1ug",
  'fit-to-content': ""
};
var positionStyle = {
  top: "_1wpz1y6m",
  center: ""
};
var horizontalStyleBase = null;

// TODO: Remove on fg cleanup: platform-linking-visual-refresh-v1
var horizontalStyleByHeightOld = {
  xlarge: "_1e0ccj1k _102k1m1q",
  large: "_1e0ccj1k _102k1m1q",
  medium: "_1e0ccj1k _102k1k8s",
  small: "_1e0ccj1k _102k1k8s"
};
var horizontalStyleByHeight = {
  xlarge: "_1e0ccj1k _102k125z",
  large: "_1e0ccj1k _102k125z",
  medium: "_1e0ccj1k _102kz42d",
  small: "_1e0ccj1k _102kz42d"
};
var gapStylesLeft = {
  xlarge: "_7yhbv47k _m6ukidpf",
  large: "_7yhbpxbi _m6ukidpf",
  medium: "_7yhbu2gc _m6ukidpf",
  small: "_7yhb1b66 _m6ukidpf"
};
var gapStylesRight = {
  xlarge: "_24rcv47k _3dveidpf",
  large: "_24rcpxbi _3dveidpf",
  medium: "_24rcu2gc _3dveidpf",
  small: "_24rc1b66 _3dveidpf"
};
var minWidthStyle = null;

/**
 * Creates a group of Action components. Accepts an array of Actions, in addition to some styling
 * preferences.
 * @internal
 * @param {ActionGroupProps} ActionGroupProps
 * @see Action
 */
var ElementGroup = function ElementGroup(_ref) {
  var _ref$align = _ref.align,
    align = _ref$align === void 0 ? _constants.SmartLinkAlignment.Left : _ref$align,
    children = _ref.children,
    _ref$direction = _ref.direction,
    direction = _ref$direction === void 0 ? _constants.SmartLinkDirection.Horizontal : _ref$direction,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? _constants.SmartLinkSize.Medium : _ref$size,
    _ref$testId = _ref.testId,
    testId = _ref$testId === void 0 ? 'smart-element-group' : _ref$testId,
    _ref$width = _ref.width,
    width = _ref$width === void 0 ? _constants.SmartLinkWidth.FitToContent : _ref$width,
    _ref$position = _ref.position,
    position = _ref$position === void 0 ? _constants.SmartLinkPosition.Center : _ref$position,
    className = _ref.className;
  var isHorizontal = direction === _constants.SmartLinkDirection.Horizontal;
  return /*#__PURE__*/React.createElement("div", {
    "data-smart-element-group": true,
    "data-testid": testId
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop
    ,
    className: (0, _runtime.ax)([!(0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v1') && baseStyleBySizeOld[size], (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v1') && baseStyleBySize[size], baseStyleByDirection[direction], !(0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v1') && "_fiawglyw _1ouwidpf _f6ju1ns9", (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v1') && "_fiawglyw", alignmentStyleMap[align], "_1ul99by0", widthStyle[width], isHorizontal && "_1reo15vq _18m915vq _1e0c1ule _s7n4nkob _1bto1l2s _1nmz1hna _1yyjkb7n _sudp1e54 _1aaynkob _15tynkob _1iu61nu9 _166k1nu9 _1div1nu9 _bmfb1nu9", isHorizontal && !(0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v1') && horizontalStyleByHeightOld[size], isHorizontal && (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v1') && horizontalStyleByHeight[size], isHorizontal && align === _constants.SmartLinkAlignment.Left && gapStylesLeft[size], isHorizontal && align === _constants.SmartLinkAlignment.Right && gapStylesRight[size], positionStyle[position], className])
  }, (0, _utils.renderChildren)(children, size));
};
var _default = exports.default = ElementGroup;
