// @atlaskit/editor-plugin-media/dist/cjs/pm-plugins/alt-text/commands.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.updateAltText = exports.openMediaAltTextMenu = exports.closeMediaAltTextMenuAndSave = exports.closeMediaAltTextMenu = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _analytics = require("@atlaskit/editor-common/analytics");
var _editorAnalytics = require("@atlaskit/editor-common/editor-analytics");
var _state = require("@atlaskit/editor-prosemirror/state");
var _mediaCommon = require("../../pm-plugins/utils/media-common");
var _commands = require("../../ui/toolbar/commands");
var _index = require("./index");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var createCommandWithAnalytics = function createCommandWithAnalytics(actionType, action, transform) {
  return function (editorAnalyticsAPI) {
    return (0, _editorAnalytics.withAnalytics)(editorAnalyticsAPI, function (state) {
      var mediaNode = (0, _mediaCommon.getMediaSingleOrInlineNodeFromSelection)(state);
      var type = (0, _commands.getNodeType)(state);
      return {
        action: actionType,
        actionSubject: _analytics.ACTION_SUBJECT.MEDIA,
        actionSubjectId: _analytics.ACTION_SUBJECT_ID.ALT_TEXT,
        eventType: _analytics.EVENT_TYPE.TRACK,
        attributes: {
          type: type,
          // Ignored via go/ees005
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
          mediaType: mediaNode.attrs.type
        }
      };
    })((0, _index.createCommand)(action, transform));
  };
};

// pass in undefined to close the alt text menu without saving
var closeMediaAltTextMenuAndSave = exports.closeMediaAltTextMenuAndSave = function closeMediaAltTextMenuAndSave(altText) {
  var commandTransform = typeof altText === 'string' ? updateAltTextTransform(altText) : undefined;
  return (0, _index.createCommand)(function (state) {
    if ((0, _mediaCommon.isMediaSingleOrInlineNodeSelected)(state)) {
      return {
        type: 'closeMediaAltTextMenu'
      };
    }
    return false;
  }, commandTransform);
};
var closeMediaAltTextMenu = exports.closeMediaAltTextMenu = closeMediaAltTextMenuAndSave();
var openMediaAltTextMenu = exports.openMediaAltTextMenu = createCommandWithAnalytics(_analytics.ACTION.OPENED, function (state) {
  if ((0, _mediaCommon.isMediaSingleOrInlineNodeSelected)(state)) {
    return {
      type: 'openMediaAltTextMenu'
    };
  }
  return false;
}, function (tr) {
  return tr.setMeta('scrollIntoView', false);
});
var updateAltTextTransform = function updateAltTextTransform(newAltText) {
  return function (tr, state) {
    var mediaNode = (0, _mediaCommon.getMediaSingleOrInlineNodeFromSelection)(state);
    if (mediaNode) {
      // mediaSingle or mediaInline
      var originalSelectionPos = tr.selection.from;
      var mediaPos = mediaNode.type === state.schema.nodes.media ? originalSelectionPos + 1 // media inside mediaSingle
      : originalSelectionPos; // mediaInline

      /**
       * Any changes to attributes of a node count the node as "recreated" in Prosemirror[1]
       * This makes it so Prosemirror resets the selection to the child i.e. "media" instead of "media-single"
       * The recommended fix is to reset the selection.[2]
       *
       * [1] https://discuss.prosemirror.net/t/setnodemarkup-loses-current-nodeselection/976
       * [2] https://discuss.prosemirror.net/t/setnodemarkup-and-deselect/3673
       */
      tr.setMeta('scrollIntoView', false).setNodeMarkup(mediaPos, undefined, _objectSpread(_objectSpread({}, mediaNode.attrs), {}, {
        alt: newAltText
      })).setSelection(_state.NodeSelection.create(tr.doc, originalSelectionPos));
    }
    return tr;
  };
};
var updateAltText = exports.updateAltText = function updateAltText(newAltText) {
  return (0, _index.createCommand)(function (state) {
    if ((0, _mediaCommon.isMediaSingleOrInlineNodeSelected)(state)) {
      return {
        type: 'updateAltText'
      };
    }
    return false;
  }, updateAltTextTransform(newAltText));
};
