// @atlaskit/smart-card/dist/cjs/extractors/flexible/actions/extract-automation-action.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.extractAutomationAction = void 0;
var _constants = require("../../../constants");
var extractAutomationAction = exports.extractAutomationAction = function extractAutomationAction(response) {
  var _response$meta;
  if (!response.data) {
    return undefined;
  }
  if (!(response !== null && response !== void 0 && (_response$meta = response.meta) !== null && _response$meta !== void 0 && (_response$meta = _response$meta.supportedFeature) !== null && _response$meta !== void 0 && _response$meta.includes(_constants.ActionName.AutomationAction))) {
    return undefined;
  }
  var data = response.data;
  var meta = response.meta;
  var name = data.name,
    objectAri = data['atlassian:ari'];
  var product = meta.product,
    tenantId = meta.tenantId,
    resourceType = meta.resourceType;
  return {
    product: product,
    resourceType: resourceType,
    siteAri: "ari:cloud:".concat(product, "::site/").concat(tenantId),
    objectName: name,
    objectAri: objectAri,
    // TODO: The admin experience will be a follow up
    canManageAutomation: false,
    baseAutomationUrl: '',
    analyticsSource: 'smartCardAutomationAction'
  };
};
