// @atlaskit/analytics-listeners/dist/cjs/cross-flow/process-event.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _analyticsGasTypes = require("@atlaskit/analytics-gas-types");
var _last = _interopRequireDefault(require("lodash/last"));
var _analyticsNamespacedContext = require("@atlaskit/analytics-namespaced-context");
var _extractDataFromEvent = require("../atlaskit/extract-data-from-event");
var _extractDataFromEvent2 = require("../helpers/extract-data-from-event");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; } /**
 * Inspired by analytics-web-react
 */
var CROSS_FLOW_TAG = 'crossFlow';
var _default = exports.default = function _default(event, logger) {
  var _event$payload$source;
  var _event$payload = event.payload,
    _event$payload$eventT = _event$payload.eventType,
    eventType = _event$payload$eventT === void 0 ? _analyticsGasTypes.UI_EVENT_TYPE : _event$payload$eventT,
    action = _event$payload.action,
    actionSubject = _event$payload.actionSubject,
    actionSubjectId = _event$payload.actionSubjectId,
    name = _event$payload.name;
  var sources = (0, _extractDataFromEvent.getSources)(event);
  var lastSourceFromContext = (0, _last.default)(sources);
  var source = (_event$payload$source = event.payload.source) !== null && _event$payload$source !== void 0 ? _event$payload$source : lastSourceFromContext;
  if (lastSourceFromContext !== source) {
    sources.push(source);
  }
  var attributes = _objectSpread(_objectSpread(_objectSpread(_objectSpread({}, (0, _extractDataFromEvent.getExtraAttributes)(event)), (0, _extractDataFromEvent2.getExtraAttributes)(event, _analyticsNamespacedContext.NAVIGATION_CONTEXT)), event.payload.attributes), {}, {
    namespaces: sources.join('.')
  });

  // Ensure navigation tag is not duplicated by using Set
  var tags = new Set(event.payload.tags || []);
  tags.add(CROSS_FLOW_TAG);
  if (event.payload) {
    switch (eventType) {
      case _analyticsGasTypes.UI_EVENT_TYPE:
      case _analyticsGasTypes.OPERATIONAL_EVENT_TYPE:
      case _analyticsGasTypes.TRACK_EVENT_TYPE:
        return {
          eventType: eventType,
          source: source,
          actionSubject: actionSubject,
          action: action,
          actionSubjectId: actionSubjectId,
          attributes: attributes,
          tags: Array.from(tags)
        };
      case _analyticsGasTypes.SCREEN_EVENT_TYPE:
        return {
          eventType: eventType,
          name: name,
          attributes: attributes,
          tags: Array.from(tags)
        };
      default:
        logger.error('Invalid event type', eventType);
        break;
    }
  }
  return null;
};
