// @atlaskit/link-datasource/dist/es2019/ui/table-footer/powered-by-jsm-assets/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import "./index.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
import React from 'react';
import { JiraServiceManagementIcon } from '@atlaskit/logo';
import { fg } from '@atlaskit/platform-feature-flags';
import { Box } from '@atlaskit/primitives/compiled';
import { AssetsIcon } from '@atlaskit/temp-nav-app-icons/assets';
import { useDatasourceAnalyticsEvents } from '../../../analytics';
const styles = {
  jsmTextStyles: "_syaz1gjq _18u012x7 _30l316c2",
  jsmContainerStyles: "_1e0c1txw _2lx21sbv"
};
export const PoweredByJSMAssets = props => {
  const ASSETS_LINK = '/jira/servicedesk/assets';
  const ASSETS_APP_LINK = '/jira/assets';
  const {
    fireEvent
  } = useDatasourceAnalyticsEvents();
  return /*#__PURE__*/React.createElement(Box, {
    xcss: styles.jsmContainerStyles,
    padding: "space.150"
  }, /*#__PURE__*/React.createElement("a", {
    "data-testid": 'powered-by-jsm-assets-link',
    onClick: () => fireEvent('ui.link.clicked.poweredBy', {
      componentHierarchy: 'datasourceTable',
      extensionKey: 'jsm-cmdb-gateway'
    }),
    href: fg('assets_as_an_app_v2') ? ASSETS_APP_LINK : ASSETS_LINK,
    rel: "noreferrer",
    target: "_blank",
    style: {
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
      display: 'inline-flex',
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- Ignored via go/DSP-18766
      alignItems: 'center'
    }
  }, fg('assets_as_an_app_v2') ? /*#__PURE__*/React.createElement(AssetsIcon, {
    size: "24"
  }) : /*#__PURE__*/React.createElement(JiraServiceManagementIcon, {
    size: "xsmall",
    appearance: "brand",
    label: props.text
  }), /*#__PURE__*/React.createElement(Box, {
    xcss: styles.jsmTextStyles
    // eslint-disable-next-line @atlaskit/design-system/use-tokens-typography, @atlaskit/ui-styling-standard/enforce-style-prop
    ,
    style: {
      fontSize: 'small'
    }
  }, props.text)));
};
