// @atlaskit/media-client/dist/cjs/models/file-state.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mapMediaItemToFileState = exports.mapMediaFileToFileState = exports.isUploadingFileState = exports.isProcessingFileState = exports.isProcessedFileState = exports.isPreviewableFileState = exports.isNonErrorFinalFileState = exports.isImageRepresentationReady = exports.isFinalFileState = exports.isErrorFileState = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var isUploadingFileState = exports.isUploadingFileState = function isUploadingFileState(fileState) {
  return fileState.status === 'uploading';
};
var isProcessingFileState = exports.isProcessingFileState = function isProcessingFileState(fileState) {
  return fileState.status === 'processing';
};
var isProcessedFileState = exports.isProcessedFileState = function isProcessedFileState(fileState) {
  return fileState.status === 'processed';
};
var isErrorFileState = exports.isErrorFileState = function isErrorFileState(fileState) {
  return fileState.status === 'error';
};
var isPreviewableFileState = exports.isPreviewableFileState = function isPreviewableFileState(fileState) {
  return !isErrorFileState(fileState) && !!fileState.preview;
};
var isFinalFileState = exports.isFinalFileState = function isFinalFileState(fileState) {
  return ['processed', 'failed-processing', 'error'].includes(fileState.status);
};
var isNonErrorFinalFileState = exports.isNonErrorFinalFileState = function isNonErrorFinalFileState(fileState) {
  return ['processed', 'failed-processing'].includes(fileState.status);
};
var isImageRepresentationReady = exports.isImageRepresentationReady = function isImageRepresentationReady(fileState) {
  switch (fileState.status) {
    case 'processing':
    case 'processed':
    case 'failed-processing':
      return !!(fileState.representations && fileState.representations.image);
    default:
      return false;
  }
};
var mapMediaFileToFileState = exports.mapMediaFileToFileState = function mapMediaFileToFileState(mediaFile) {
  var _mediaFile$data = mediaFile.data,
    id = _mediaFile$data.id,
    name = _mediaFile$data.name,
    size = _mediaFile$data.size,
    processingStatus = _mediaFile$data.processingStatus,
    artifacts = _mediaFile$data.artifacts,
    mediaType = _mediaFile$data.mediaType,
    mimeType = _mediaFile$data.mimeType,
    representations = _mediaFile$data.representations,
    createdAt = _mediaFile$data.createdAt,
    metadataTraceContext = _mediaFile$data.metadataTraceContext,
    hash = _mediaFile$data.hash,
    abuseClassification = _mediaFile$data.abuseClassification;
  var baseState = {
    id: id,
    name: name,
    size: size,
    mediaType: mediaType,
    mimeType: mimeType,
    artifacts: artifacts,
    representations: representations,
    createdAt: createdAt,
    hash: hash,
    metadataTraceContext: metadataTraceContext,
    abuseClassification: abuseClassification
  };
  switch (processingStatus) {
    case 'pending':
    case undefined:
      return _objectSpread(_objectSpread({}, baseState), {}, {
        status: 'processing'
      });
    case 'succeeded':
      return _objectSpread(_objectSpread({}, baseState), {}, {
        status: 'processed'
      });
    case 'failed':
      return _objectSpread(_objectSpread({}, baseState), {}, {
        status: 'failed-processing'
      });
  }
};
var mapMediaItemToFileState = exports.mapMediaItemToFileState = function mapMediaItemToFileState(id, item) {
  return mapMediaFileToFileState({
    data: _objectSpread({
      id: id
    }, item)
  });
};
