// @atlaskit/editor-plugin-card/dist/cjs/ui/LayoutButton/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.LayoutButton = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = require("react");
var _react2 = require("@emotion/react");
var _reactIntlNext = require("react-intl-next");
var _hooks = require("@atlaskit/editor-common/hooks");
var _ui = require("@atlaskit/editor-common/ui");
var _uiMenu = require("@atlaskit/editor-common/ui-menu");
var _utils = require("@atlaskit/editor-common/utils");
var _growHorizontalEditorExpand = _interopRequireDefault(require("@atlaskit/icon/core/migration/grow-horizontal--editor-expand"));
var _shrinkHorizontalEditorCollapse = _interopRequireDefault(require("@atlaskit/icon/core/migration/shrink-horizontal--editor-collapse"));
var _collapse = _interopRequireDefault(require("@atlaskit/icon/glyph/editor/collapse"));
var _expand = _interopRequireDefault(require("@atlaskit/icon/glyph/editor/expand"));
var _linkingCommon = require("@atlaskit/linking-common");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _colors = require("@atlaskit/theme/colors");
var _actions = require("../../pm-plugins/actions");
var _utils2 = require("../../pm-plugins/utils");
var _utils3 = require("./utils");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; } /**
 * @jsxRuntime classic
 * @jsx jsx
 */ // eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
var toolbarButtonWrapperStyles = (0, _react2.css)({
  background: "".concat("var(--ds-background-neutral, ".concat(_colors.N20A, ")")),
  color: "".concat("var(--ds-icon, ".concat(_colors.N300, ")")),
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
  ':hover': {
    background: "".concat("var(--ds-background-neutral-hovered, ".concat(_colors.B300, ")")),
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-important-styles -- Ignored via go/DSP-18766
    color: "var(--ds-icon, white)".concat(" !important")
  }
});
var LayoutButton = exports.LayoutButton = function LayoutButton(_ref) {
  var onLayoutChange = _ref.onLayoutChange,
    _ref$layout = _ref.layout,
    layout = _ref$layout === void 0 ? _linkingCommon.DATASOURCE_DEFAULT_LAYOUT : _ref$layout,
    formatMessage = _ref.intl.formatMessage,
    mountPoint = _ref.mountPoint,
    boundariesElement = _ref.boundariesElement,
    scrollableElement = _ref.scrollableElement,
    targetElement = _ref.targetElement,
    _ref$testId = _ref.testId,
    testId = _ref$testId === void 0 ? 'datasource-table-layout-button' : _ref$testId;
  var handleClick = (0, _react.useCallback)(function () {
    onLayoutChange && onLayoutChange((0, _utils.getNextBreakoutMode)(layout));
  }, [layout, onLayoutChange]);
  var title = (0, _react.useMemo)(function () {
    return formatMessage((0, _utils.getTitle)(layout));
  }, [formatMessage, layout]);
  if (!targetElement) {
    return null;
  }
  var collapseIcon = (0, _platformFeatureFlags.fg)('platform-editor-plugin-card-icon-migration') ? (0, _react2.jsx)(_shrinkHorizontalEditorCollapse.default, {
    label: title
  }) : (0, _react2.jsx)(_collapse.default, {
    label: title
  });
  var expandIcon = (0, _platformFeatureFlags.fg)('platform-editor-plugin-card-icon-migration') ? (0, _react2.jsx)(_growHorizontalEditorExpand.default, {
    label: title
  }) : (0, _react2.jsx)(_expand.default, {
    label: title
  });
  return (0, _react2.jsx)(_ui.Popup, {
    mountTo: mountPoint,
    boundariesElement: boundariesElement,
    scrollableElement: scrollableElement,
    target: targetElement,
    alignY: "start",
    alignX: "end",
    forcePlacement: true,
    stick: true,
    ariaLabel: title
  }, (0, _react2.jsx)(_uiMenu.ToolbarButton, {
    testId: testId,
    css: toolbarButtonWrapperStyles,
    title: title,
    onClick: handleClick,
    iconBefore: layout === 'full-width' ? collapseIcon : expandIcon
  }));
};
var LayoutButtonWrapper = function LayoutButtonWrapper(_ref2) {
  var _node$attrs;
  var editorView = _ref2.editorView,
    mountPoint = _ref2.mountPoint,
    scrollableElement = _ref2.scrollableElement,
    boundariesElement = _ref2.boundariesElement,
    intl = _ref2.intl,
    api = _ref2.api;
  var _useSharedPluginState = (0, _hooks.useSharedPluginState)(api, ['card']),
    cardState = _useSharedPluginState.cardState;
  var _getDatasource = (0, _utils3.getDatasource)(editorView),
    node = _getDatasource.node,
    pos = _getDatasource.pos;
  var isDatasource = (0, _utils2.isDatasourceNode)(node);
  if (!isDatasource) {
    return null;
  }

  //  If layout doesn't exist in ADF it returns null, we want to change to undefined
  //  which results in default parameter value being used in LayoutButton.
  var _ref3 = cardState !== null && cardState !== void 0 ? cardState : {},
    datasourceTableRef = _ref3.datasourceTableRef,
    _ref3$layout = _ref3.layout,
    layout = _ref3$layout === void 0 ? (node === null || node === void 0 || (_node$attrs = node.attrs) === null || _node$attrs === void 0 ? void 0 : _node$attrs.layout) || undefined : _ref3$layout;
  var onLayoutChange = function onLayoutChange(layout) {
    var _getDatasource$node;
    if (pos === undefined) {
      return;
    }
    var state = editorView.state,
      dispatch = editorView.dispatch;
    // If the button does not re-render due to no card state change, node reference will be stale
    var datasourceNode = (_getDatasource$node = (0, _utils3.getDatasource)(editorView).node) !== null && _getDatasource$node !== void 0 ? _getDatasource$node : node;
    var tr = state.tr.setNodeMarkup(pos, undefined, _objectSpread(_objectSpread({}, datasourceNode === null || datasourceNode === void 0 ? void 0 : datasourceNode.attrs), {}, {
      layout: layout
    }));
    tr.setMeta('scrollIntoView', false);
    dispatch((0, _actions.setCardLayout)(layout)(tr));
  };
  return (0, _react2.jsx)(LayoutButton, {
    mountPoint: mountPoint,
    scrollableElement: scrollableElement,
    boundariesElement: boundariesElement
    // Ignored via go/ees005
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    ,
    targetElement: datasourceTableRef,
    layout: (0, _utils3.isDatasourceTableLayout)(layout) ? layout : undefined,
    onLayoutChange: onLayoutChange,
    intl: intl
  });
};
var _default = exports.default = (0, _reactIntlNext.injectIntl)(LayoutButtonWrapper);
