// @atlaskit/media-client/dist/cjs/utils/with-media-client-hoc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = tslib_1.__importStar(require("react"));
var media_client_1 = require("../client/media-client");
var mediaClientsMap = new Map();
exports.getMediaClient = function (mediaClientConfig) {
    var mediaClient = mediaClientsMap.get(mediaClientConfig);
    if (!mediaClient) {
        mediaClient = new media_client_1.MediaClient(mediaClientConfig);
        mediaClientsMap.set(mediaClientConfig, mediaClient);
    }
    return mediaClient;
};
var createEmptyMediaClient = function () {
    var emptyConfig = {
        authProvider: function () {
            return Promise.resolve({
                clientId: '',
                token: '',
                baseUrl: '',
            });
        },
    };
    return new media_client_1.MediaClient(emptyConfig);
};
exports.withMediaClient = function (Component) {
    return /** @class */ (function (_super) {
        tslib_1.__extends(class_1, _super);
        function class_1() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        class_1.prototype.render = function () {
            // TODO MS-1552: clean up after we move mediaClientConfig into FileIdentifier
            var props = this.props;
            var mediaClientConfig = props.mediaClientConfig, identifier = props.identifier;
            var isExternalIdentifier = identifier && identifier.mediaItemType === 'external-image';
            var mediaClient = isExternalIdentifier
                ? createEmptyMediaClient()
                : exports.getMediaClient(mediaClientConfig);
            return React.createElement(Component, tslib_1.__assign({}, props, { mediaClient: mediaClient }));
        };
        return class_1;
    }(React.Component));
};
