// @atlaskit/editor-plugin-extension/dist/es2019/ui/ConfigPanel/Fields/Expand.js
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
import React, { useState } from 'react';

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import { css, jsx } from '@emotion/react';
import { injectIntl } from 'react-intl-next';
import { IconButton } from '@atlaskit/button/new';
import { configPanelMessages as messages } from '@atlaskit/editor-common/extensions';
import ChevronDownIcon from '@atlaskit/icon/utility/chevron-down';
import ChevronRightIcon from '@atlaskit/icon/utility/chevron-right';
import { N40 } from '@atlaskit/theme/colors';
const expandContainerStyles = css({
  borderBottom: `1px solid ${`var(--ds-border, ${N40})`}`
});
const expandControlStyles = css({
  display: 'flex',
  height: "var(--ds-space-600, 48px)",
  justifyContent: 'center',
  paddingRight: "var(--ds-space-100, 8px)"
});
const chevronContainerStyles = css({
  display: 'flex',
  alignItems: 'center',
  // eslint-disable-next-line @atlaskit/design-system/no-nested-styles, @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  '& > button': {
    width: "var(--ds-space-300, 24px)",
    height: "var(--ds-space-300, 24px)"
  }
});
const labelContainerStyles = css({
  width: '100%',
  alignItems: 'center',
  display: 'flex',
  fontWeight: "var(--ds-font-weight-medium, 500)"
});
const expandContentContainerHiddenStyles = css({
  display: 'none',
  marginTop: "var(--ds-space-negative-100, -8px)"
});
const expandContentContainerVisibleStyles = css({
  display: 'block',
  marginTop: "var(--ds-space-negative-100, -8px)"
});
function Expand({
  field,
  children,
  isExpanded = false,
  intl
}) {
  const [expanded, setExpanded] = useState(isExpanded);
  return jsx("div", {
    "data-testid": "expand-config-field",
    css: expandContainerStyles
  }, jsx("div", {
    css: expandControlStyles
  }, jsx("div", {
    css: labelContainerStyles
  }, field.label), jsx("div", {
    css: chevronContainerStyles
  }, jsx(IconButton, {
    onClick: () => {
      setExpanded(!expanded);
    },
    label: intl.formatMessage(expanded ? messages.collapse : messages.expand),
    testId: "form-expand-toggle",
    icon: expanded ? ChevronDownIcon : ChevronRightIcon
  }))), jsx("div", {
    "data-testid": "expand-content-container",
    css: expanded ? expandContentContainerVisibleStyles : expandContentContainerHiddenStyles
  }, children));
}
export default injectIntl(Expand);
