// @atlaskit/media-card/dist/cjs/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isSSRServerPreview = exports.isSSRPreview = exports.isSSRDataPreview = exports.isSSRClientPreview = exports.MediaCardCursor = void 0;
/**
 * Entry Point: @atlaskit/media-card/types
 * tsconfig.entry-points.json
 */

var isSSRPreview = exports.isSSRPreview = function isSSRPreview(preview) {
  return isSSRClientPreview(preview) || isSSRServerPreview(preview) || isSSRDataPreview(preview);
};
var isSSRServerPreview = exports.isSSRServerPreview = function isSSRServerPreview(preview) {
  var ssrClientSources = ['ssr-server', 'cache-ssr-server'];
  return ssrClientSources.includes(preview.source);
};
var isSSRClientPreview = exports.isSSRClientPreview = function isSSRClientPreview(preview) {
  var ssrClientSources = ['ssr-client', 'cache-ssr-client'];
  return ssrClientSources.includes(preview.source);
};
var isSSRDataPreview = exports.isSSRDataPreview = function isSSRDataPreview(preview) {
  return preview.source === 'ssr-data';
};
var MediaCardCursor = exports.MediaCardCursor = /*#__PURE__*/function (MediaCardCursor) {
  MediaCardCursor["Action"] = "pointer";
  MediaCardCursor["NotReady"] = "wait";
  return MediaCardCursor;
}({});
