// @atlaskit/linking-common/dist/esm/hooks/useAvailableSites/utils.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _typeof from "@babel/runtime/helpers/typeof";
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
import { getTraceId } from '../../utils/get-trace-id';
var getUrlPath = function getUrlPath(url) {
  try {
    return new URL(url).pathname;
  } catch (_unused) {
    return 'Failed to parse pathname from url';
  }
};
export var getNetworkFields = function getNetworkFields(error) {
  if (error instanceof Response) {
    return {
      traceId: getTraceId(error),
      status: error.status,
      path: getUrlPath(error.url)
    };
  }
  return {
    traceId: null,
    status: null,
    path: null
  };
};
export var getErrorType = function getErrorType(error) {
  if (error instanceof Response) {
    return 'NetworkError';
  }
  if (error instanceof Error) {
    return error.name;
  }
  return _typeof(error);
};
export var getOperationFailedAttributes = function getOperationFailedAttributes(err) {
  if (err instanceof Response) {
    return _objectSpread({
      error: 'NetworkError',
      errorType: 'NetworkError'
    }, getNetworkFields(err));
  }
  var error = err instanceof Error ? err : new Error('unknown error');
  return {
    error: error.name,
    errorType: error.name,
    traceId: null,
    status: null,
    path: null
  };
};
