// @atlaskit/media-card/dist/cjs/styles/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var styled_components_1 = tslib_1.__importDefault(require("styled-components"));
var constants_1 = require("@atlaskit/theme/constants");
var media_ui_1 = require("@atlaskit/media-ui");
var config_1 = require("./config");
exports.defaultTransitionDuration = config_1.defaultTransitionDuration;
var mixins_1 = require("./mixins");
exports.antialiased = mixins_1.antialiased;
exports.borderRadiusLeft = mixins_1.borderRadiusLeft;
exports.capitalize = mixins_1.capitalize;
exports.centerSelf = mixins_1.centerSelf;
exports.centerSelfX = mixins_1.centerSelfX;
exports.centerSelfY = mixins_1.centerSelfY;
exports.centerX = mixins_1.centerX;
exports.hexToRgb = mixins_1.hexToRgb;
exports.rgba = mixins_1.rgba;
exports.spaceAround = mixins_1.spaceAround;
exports.transition = mixins_1.transition;
exports.withAppearance = mixins_1.withAppearance;
var easing_1 = require("./easing");
exports.easeOutCubic = easing_1.easeOutCubic;
exports.easeOutExpo = easing_1.easeOutExpo;
var animations_1 = require("./animations");
exports.spin = animations_1.spin;
exports.Root = styled_components_1.default.div(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n  box-sizing: border-box;\n  font-family: ", ";\n\n  * {\n    box-sizing: border-box;\n  }\n"], ["\n  box-sizing: border-box;\n  font-family: ", ";\n\n  * {\n    box-sizing: border-box;\n  }\n"])), constants_1.fontFamily());
exports.cardShadow = "\n  box-shadow: 0 1px 1px rgba(9, 30, 66, 0.2), 0 0 1px 0 rgba(9, 30, 66, 0.24);\n";
exports.FadeinImage = styled_components_1.default.div(templateObject_2 || (templateObject_2 = tslib_1.__makeTemplateObject(["\n  ", ";\n"], ["\n  ", ";\n"])), media_ui_1.fadeIn);
exports.default = exports.Root;
var templateObject_1, templateObject_2;
