// @atlaskit/media-viewer/dist/cjs/analytics/events/operational/loadFailed.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createLoadFailedEvent = void 0;
var _ = require("../..");
var _errors = require("../../../errors");
var createLoadFailedEvent = exports.createLoadFailedEvent = function createLoadFailedEvent(fileId, error, fileState, traceContext) {
  var _getFileAttributes = (0, _.getFileAttributes)(fileState),
    fileMediatype = _getFileAttributes.fileMediatype,
    fileMimetype = _getFileAttributes.fileMimetype,
    fileSize = _getFileAttributes.fileSize;
  return {
    eventType: 'operational',
    actionSubject: 'mediaFile',
    action: 'loadFailed',
    attributes: {
      status: 'fail',
      failReason: (0, _errors.getPrimaryErrorReason)(error),
      error: (0, _errors.getSecondaryErrorReason)(error),
      errorDetail: (0, _errors.getErrorDetail)(error),
      request: (0, _errors.getRequestMetadata)(error),
      fileMimetype: fileMimetype,
      fileAttributes: {
        fileId: fileId,
        fileMediatype: fileMediatype,
        fileMimetype: fileMimetype,
        fileSize: fileSize
      },
      traceContext: fileMediatype === 'image' ? traceContext : undefined
    }
  };
};
