// @atlaskit/media-viewer/dist/cjs/analytics/events/operational/zipEntryLoadSucceeded.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createZipEntryLoadSucceededEvent = void 0;
var _utils = require("../../../utils");
var _ = require("../..");
var createZipEntryLoadSucceededEvent = exports.createZipEntryLoadSucceededEvent = function createZipEntryLoadSucceededEvent(fileState, zipEntry) {
  var _getFileAttributes = (0, _.getFileAttributes)(fileState),
    fileId = _getFileAttributes.fileId,
    fileMediatype = _getFileAttributes.fileMediatype,
    fileMimetype = _getFileAttributes.fileMimetype,
    fileSize = _getFileAttributes.fileSize,
    fileStatus = _getFileAttributes.fileStatus;
  return {
    eventType: 'operational',
    actionSubject: 'mediaFile',
    action: 'zipEntryLoadSucceeded',
    attributes: {
      status: 'success',
      fileAttributes: {
        fileId: fileId,
        fileMediatype: fileMediatype,
        fileMimetype: fileMimetype,
        fileSize: fileSize,
        fileStatus: fileStatus
      },
      size: zipEntry.size,
      encrypted: zipEntry.encrypted,
      compressedSize: zipEntry.compressedSize,
      mimeType: (0, _utils.getMimeTypeFromFilename)(zipEntry.name)
    }
  };
};
