// @atlaskit/media-viewer/dist/cjs/analytics/events/operational/zipEntryLoadFailed.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createZipEntryLoadFailedEvent = void 0;
var _ = require("../..");
var _utils = require("../../../utils");
var _errors = require("../../../errors");
var createZipEntryLoadFailedEvent = exports.createZipEntryLoadFailedEvent = function createZipEntryLoadFailedEvent(fileState, error, zipEntry) {
  var _getFileAttributes = (0, _.getFileAttributes)(fileState),
    fileId = _getFileAttributes.fileId,
    fileMediatype = _getFileAttributes.fileMediatype,
    fileMimetype = _getFileAttributes.fileMimetype,
    fileSize = _getFileAttributes.fileSize,
    fileStatus = _getFileAttributes.fileStatus;
  return {
    eventType: 'operational',
    actionSubject: 'mediaFile',
    action: 'zipEntryLoadFailed',
    attributes: {
      status: 'fail',
      failReason: (0, _errors.getPrimaryErrorReason)(error),
      error: (0, _errors.getSecondaryErrorReason)(error),
      errorDetail: (0, _errors.getErrorDetail)(error),
      fileAttributes: {
        fileId: fileId,
        fileMediatype: fileMediatype,
        fileMimetype: fileMimetype,
        fileSize: fileSize,
        fileStatus: fileStatus
      },
      size: zipEntry ? zipEntry.size : -1,
      encrypted: zipEntry ? zipEntry.encrypted : false,
      compressedSize: zipEntry ? zipEntry.compressedSize : -1,
      mimeType: zipEntry ? (0, _utils.getMimeTypeFromFilename)(zipEntry.name) : 'undefined'
    }
  };
};
