// @atlaskit/media-card/dist/cjs/root/card/objectURLCache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var lru_fast_1 = require("lru-fast");
exports.PREVIEW_CACHE_LRU_SIZE = 20;
var ExtendedLRUCache = /** @class */ (function (_super) {
    tslib_1.__extends(ExtendedLRUCache, _super);
    function ExtendedLRUCache() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ExtendedLRUCache.prototype.shift = function () {
        var entry = _super.prototype.shift.call(this);
        if (entry) {
            URL.revokeObjectURL(entry.value);
        }
        return entry;
    };
    return ExtendedLRUCache;
}(lru_fast_1.LRUCache));
var ObjectURLCache = /** @class */ (function () {
    function ObjectURLCache(size) {
        this.cache = new ExtendedLRUCache(size);
    }
    ObjectURLCache.prototype.has = function (key) {
        return !!this.cache.get(key);
    };
    ObjectURLCache.prototype.get = function (key) {
        return this.cache.get(key);
    };
    ObjectURLCache.prototype.set = function (key, value) {
        this.cache.set(key, value);
    };
    ObjectURLCache.prototype.reset = function () {
        this.cache.removeAll();
    };
    return ObjectURLCache;
}());
exports.ObjectURLCache = ObjectURLCache;
exports.objectURLCache = new ObjectURLCache(exports.PREVIEW_CACHE_LRU_SIZE);
