// @atlaskit/editor-plugin-find-replace/dist/cjs/ui/FindReplace.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireDefault(require("react"));
var _react2 = require("@emotion/react");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _Find = _interopRequireDefault(require("./Find"));
var _Replace = _interopRequireDefault(require("./Replace"));
var _ReplaceNext = _interopRequireDefault(require("./ReplaceNext"));
var _uiStyles = require("./ui-styles");
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); } /* eslint-disable @atlaskit/design-system/consistent-css-prop-usage */ /* eslint-disable @atlaskit/design-system/prefer-primitives */ /**
 * @jsxRuntime classic
 * @jsx jsx
 */ // eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
// eslint-disable-next-line @repo/internal/react/no-class-components
var FindReplace = /*#__PURE__*/function (_React$PureComponent) {
  function FindReplace(props) {
    var _this;
    (0, _classCallCheck2.default)(this, FindReplace);
    _this = _callSuper(this, FindReplace, [props]);
    (0, _defineProperty2.default)(_this, "findTextfield", null);
    (0, _defineProperty2.default)(_this, "replaceTextfield", null);
    /**
     * Delete this function on cleanup of
     * editor_a11y_refactor_find_replace_style
     */
    (0, _defineProperty2.default)(_this, "handleTabNavigation", function (event) {
      if ((0, _platformFeatureFlags.fg)('editor_a11y_refactor_find_replace_style')) {
        return;
      }
      if (event.key === 'Tab') {
        event.preventDefault();
        var modalFindReplace = _this.modalRef.current;
        if (!modalFindReplace || !modalFindReplace.contains(document.activeElement)) {
          return;
        }
        var focusableElements = Array.from(modalFindReplace.querySelectorAll('[tabindex]:not([tabindex="-1"]), input, button')).filter(function (el) {
          return el.getAttribute('tabindex') !== '-1';
        });
        var currentIndex = focusableElements.findIndex(function (el) {
          return el === document.activeElement;
        });
        var isShiftPressed = event.shiftKey;
        if (isShiftPressed) {
          var prevIndex = (currentIndex - 1 + focusableElements.length) % focusableElements.length;
          focusableElements[prevIndex].focus();
        } else {
          var nextIndex = (currentIndex + 1) % focusableElements.length;
          focusableElements[nextIndex].focus();
        }
      }
    });
    (0, _defineProperty2.default)(_this, "state", {
      findTyped: false
    });
    (0, _defineProperty2.default)(_this, "setFindTyped", function (value) {
      _this.setState({
        findTyped: value
      });
    });
    (0, _defineProperty2.default)(_this, "setFindTextfieldRef", function (findTextfieldRef) {
      _this.findTextfield = findTextfieldRef.current;
    });
    (0, _defineProperty2.default)(_this, "setReplaceTextfieldRef", function (replaceTextfieldRef) {
      _this.replaceTextfield = replaceTextfieldRef.current;
    });
    (0, _defineProperty2.default)(_this, "setFocusToFind", function () {
      if (_this.findTextfield) {
        _this.findTextfield.focus();
      }
    });
    (0, _defineProperty2.default)(_this, "setFocusToReplace", function () {
      if (_this.replaceTextfield) {
        _this.replaceTextfield.focus();
      }
    });
    _this.modalRef = /*#__PURE__*/_react.default.createRef();
    return _this;
  }
  (0, _inherits2.default)(FindReplace, _React$PureComponent);
  return (0, _createClass2.default)(FindReplace, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      // eslint-disable-next-line
      window.addEventListener('keydown', this.handleTabNavigation);
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      // eslint-disable-next-line
      window.removeEventListener('keydown', this.handleTabNavigation);
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
        findText = _this$props.findText,
        count = _this$props.count,
        shouldFocus = _this$props.shouldFocus,
        onFind = _this$props.onFind,
        onFindBlur = _this$props.onFindBlur,
        onFindNext = _this$props.onFindNext,
        onFindPrev = _this$props.onFindPrev,
        onCancel = _this$props.onCancel,
        replaceText = _this$props.replaceText,
        onReplace = _this$props.onReplace,
        onReplaceAll = _this$props.onReplaceAll,
        dispatchAnalyticsEvent = _this$props.dispatchAnalyticsEvent,
        allowMatchCase = _this$props.allowMatchCase,
        shouldMatchCase = _this$props.shouldMatchCase,
        onToggleMatchCase = _this$props.onToggleMatchCase;
      var focusToolbarButton = this.props.focusToolbarButton || function () {};
      return (
        // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
        (0, _react2.jsx)("div", {
          ref: this.modalRef,
          css: [_uiStyles.wrapperStyles, _uiStyles.wrapperPaddingStyles]
        }, (0, _react2.jsx)(_Find.default, {
          allowMatchCase: allowMatchCase,
          shouldMatchCase: shouldMatchCase,
          onToggleMatchCase: onToggleMatchCase,
          findText: findText,
          count: count,
          shouldFocus: shouldFocus,
          onFind: onFind,
          onFindBlur: onFindBlur,
          onFindPrev: onFindPrev,
          onFindNext: onFindNext,
          onFindTextfieldRefSet: this.setFindTextfieldRef,
          onCancel: onCancel,
          onArrowDown: this.setFocusToReplace,
          findTyped: this.state.findTyped,
          setFindTyped: this.setFindTyped
        }), (0, _react2.jsx)("hr", {
          role: "presentation",
          css: _uiStyles.ruleStyles,
          id: "replace-hr-element"
        }), (0, _platformFeatureFlags.fg)('editor_a11y_refactor_find_replace_style') ? (0, _react2.jsx)(_ReplaceNext.default, {
          canReplace: count.total > 0,
          replaceText: replaceText,
          onReplace: onReplace,
          onReplaceAll: onReplaceAll,
          onReplaceTextfieldRefSet: this.setReplaceTextfieldRef,
          onArrowUp: this.setFocusToFind,
          onCancel: onCancel,
          count: count,
          onFindPrev: onFindPrev,
          onFindNext: onFindNext,
          dispatchAnalyticsEvent: dispatchAnalyticsEvent,
          findTyped: this.state.findTyped,
          setFindTyped: this.setFindTyped,
          focusToolbarButton: focusToolbarButton
        }) : (0, _react2.jsx)(_Replace.default, {
          canReplace: count.total > 0,
          replaceText: replaceText,
          onReplace: onReplace,
          onReplaceAll: onReplaceAll,
          onReplaceTextfieldRefSet: this.setReplaceTextfieldRef,
          onArrowUp: this.setFocusToFind,
          onCancel: onCancel,
          count: count,
          onFindPrev: onFindPrev,
          onFindNext: onFindNext,
          dispatchAnalyticsEvent: dispatchAnalyticsEvent,
          findTyped: this.state.findTyped,
          setFindTyped: this.setFindTyped,
          focusToolbarButton: focusToolbarButton
        }))
      );
    }
  }]);
}(_react.default.PureComponent);
var _default = exports.default = FindReplace;
