// @atlaskit/editor-plugin-find-replace/dist/cjs/ui/Replace.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _react2 = require("@emotion/react");
var _reactIntlNext = require("react-intl-next");
var _new = _interopRequireDefault(require("@atlaskit/button/new"));
var _analytics = require("@atlaskit/editor-common/analytics");
var _messages = require("@atlaskit/editor-common/messages");
var _form = require("@atlaskit/form");
var _chevronDownHipchatChevronDown = _interopRequireDefault(require("@atlaskit/icon/utility/migration/chevron-down--hipchat-chevron-down"));
var _chevronUpHipchatChevronUp = _interopRequireDefault(require("@atlaskit/icon/utility/migration/chevron-up--hipchat-chevron-up"));
var _primitives = require("@atlaskit/primitives");
var _textfield = _interopRequireDefault(require("@atlaskit/textfield"));
var _FindReplaceTooltipButton = require("./FindReplaceTooltipButton");
var _uiStyles = require("./ui-styles");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); } /* eslint-disable @atlaskit/design-system/consistent-css-prop-usage */ /**
 * @jsxRuntime classic
 * @jsx jsx
 */ // eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
// eslint-disable-next-line @repo/internal/react/no-class-components
var Replace = /*#__PURE__*/function (_React$PureComponent) {
  function Replace(props) {
    var _this;
    (0, _classCallCheck2.default)(this, Replace);
    _this = _callSuper(this, Replace, [props]);
    (0, _defineProperty2.default)(_this, "replaceTextfieldRef", /*#__PURE__*/_react.default.createRef());
    (0, _defineProperty2.default)(_this, "successReplacementMessageRef", /*#__PURE__*/_react.default.createRef());
    (0, _defineProperty2.default)(_this, "isComposing", false);
    (0, _defineProperty2.default)(_this, "skipWhileComposing", function (fn) {
      if (_this.state.isComposing) {
        return;
      }
      fn();
    });
    (0, _defineProperty2.default)(_this, "handleReplaceClick", function () {
      return _this.skipWhileComposing(function () {
        _this.props.onReplace({
          triggerMethod: _analytics.TRIGGER_METHOD.BUTTON,
          replaceText: _this.state.replaceText
        });
        // for replace button replaceCount always 1;
        var replaceCount = 1;
        _this.triggerSuccessReplacementMessageUpdate(replaceCount);
        _this.setState({
          isHelperMessageVisible: true,
          replaceCount: replaceCount
        });
        _this.props.setFindTyped(false);
      });
    });
    (0, _defineProperty2.default)(_this, "handleReplaceChange", function (event) {
      return _this.skipWhileComposing(function () {
        _this.updateReplaceValue(event.target.value);
      });
    });
    (0, _defineProperty2.default)(_this, "updateReplaceValue", function (replaceText) {
      var dispatchAnalyticsEvent = _this.props.dispatchAnalyticsEvent;
      if (dispatchAnalyticsEvent) {
        dispatchAnalyticsEvent({
          eventType: _analytics.EVENT_TYPE.TRACK,
          action: _analytics.ACTION.CHANGED_REPLACEMENT_TEXT,
          actionSubject: _analytics.ACTION_SUBJECT.FIND_REPLACE_DIALOG
        });
      }
      _this.setState({
        replaceText: replaceText
      });
    });
    (0, _defineProperty2.default)(_this, "handleReplaceKeyDown", function (event) {
      return _this.skipWhileComposing(function () {
        if (event.key === 'Enter') {
          _this.props.onReplace({
            triggerMethod: _analytics.TRIGGER_METHOD.KEYBOARD,
            replaceText: _this.state.replaceText
          });
        } else if (event.key === 'ArrowUp') {
          // we want to move focus between find & replace texfields when user hits up/down arrows
          _this.props.onArrowUp();
        }
      });
    });
    (0, _defineProperty2.default)(_this, "handleReplaceAllClick", function () {
      return _this.skipWhileComposing(function () {
        _this.props.onReplaceAll({
          replaceText: _this.state.replaceText
        });
        _this.setState({
          isHelperMessageVisible: true
        });
        _this.triggerSuccessReplacementMessageUpdate(_this.props.count.total);
        _this.setState({
          replaceCount: _this.props.count.total
        });
        _this.props.setFindTyped(false);
      });
    });
    (0, _defineProperty2.default)(_this, "handleCompositionStart", function () {
      _this.setState({
        isComposing: true
      });
    });
    (0, _defineProperty2.default)(_this, "handleCompositionEnd", function (event) {
      _this.setState({
        isComposing: false
      });
      // type for React.CompositionEvent doesn't set type for target correctly
      _this.updateReplaceValue(event.target.value);
    });
    (0, _defineProperty2.default)(_this, "clearSearch", function () {
      _this.props.onCancel({
        triggerMethod: _analytics.TRIGGER_METHOD.BUTTON
      });
      _this.props.focusToolbarButton && _this.props.focusToolbarButton();
    });
    (0, _defineProperty2.default)(_this, "handleFindNextClick", function () {
      if (_this.isComposing) {
        return;
      }
      _this.props.onFindNext({
        triggerMethod: _analytics.TRIGGER_METHOD.BUTTON
      });
    });
    (0, _defineProperty2.default)(_this, "handleFindPrevClick", function () {
      if (_this.isComposing) {
        return;
      }
      _this.props.onFindPrev({
        triggerMethod: _analytics.TRIGGER_METHOD.BUTTON
      });
    });
    var _replaceText = props.replaceText,
      formatMessage = props.intl.formatMessage;
    _this.state = {
      replaceText: _replaceText || '',
      isComposing: false,
      isHelperMessageVisible: false,
      fakeSuccessReplacementMessageUpdate: false,
      replaceCount: 0
    };
    _this.replace = formatMessage(_messages.findReplaceMessages.replace);
    _this.replaceWith = formatMessage(_messages.findReplaceMessages.replaceWith);
    _this.replaceAll = formatMessage(_messages.findReplaceMessages.replaceAll);
    _this.findNext = formatMessage(_messages.findReplaceMessages.findNext);
    _this.findPrevious = formatMessage(_messages.findReplaceMessages.findPrevious);
    _this.closeFindReplaceDialog = formatMessage(_messages.findReplaceMessages.closeFindReplaceDialog);
    return _this;
  }
  (0, _inherits2.default)(Replace, _React$PureComponent);
  return (0, _createClass2.default)(Replace, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.props.onReplaceTextfieldRefSet(this.replaceTextfieldRef);
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(_ref) {
      var prevReplaceText = _ref.replaceText;
      var replaceText = this.props.replaceText;
      if (replaceText && replaceText !== prevReplaceText) {
        this.setState({
          replaceText: replaceText,
          isComposing: false
        });
      }
      var findTextField = document.getElementById('find-text-field');
      var replaceButton = document.getElementById('replace-button');
      var replaceAllButton = document.getElementById('replaceAll-button');
      if (((replaceButton === null || replaceButton === void 0 ? void 0 : replaceButton.tabIndex) === -1 || (replaceAllButton === null || replaceAllButton === void 0 ? void 0 : replaceAllButton.tabIndex) === -1) && findTextField) {
        findTextField.focus();
      }
    }
  }, {
    key: "triggerSuccessReplacementMessageUpdate",
    value: function triggerSuccessReplacementMessageUpdate(currentReplaceCount) {
      var _this$state;
      if (((_this$state = this.state) === null || _this$state === void 0 ? void 0 : _this$state.replaceCount) === currentReplaceCount) {
        this.setState({
          fakeSuccessReplacementMessageUpdate: !this.state.fakeSuccessReplacementMessageUpdate
        });
      }
      if (this.successReplacementMessageRef && this.successReplacementMessageRef.current) {
        var ariaLiveRegion = this.successReplacementMessageRef.current.querySelector('[aria-live="polite"]');
        ariaLiveRegion === null || ariaLiveRegion === void 0 || ariaLiveRegion.removeAttribute('aria-live');
        ariaLiveRegion === null || ariaLiveRegion === void 0 || ariaLiveRegion.setAttribute('aria-live', 'polite');
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$state2 = this.state,
        replaceText = _this$state2.replaceText,
        isHelperMessageVisible = _this$state2.isHelperMessageVisible,
        replaceCount = _this$state2.replaceCount;
      var _this$props = this.props,
        canReplace = _this$props.canReplace,
        count = _this$props.count,
        formatMessage = _this$props.intl.formatMessage;
      var resultsReplace = formatMessage(_messages.findReplaceMessages.replaceSuccess, {
        numberOfMatches: replaceCount
      });
      return (0, _react2.jsx)(_react.Fragment, null, (0, _react2.jsx)("div", {
        css: [_uiStyles.sectionWrapperStyles, _uiStyles.sectionWrapperStylesAlternate]
      }, (0, _react2.jsx)("div", {
        css: _uiStyles.textFieldWrapper
      }, (0, _react2.jsx)(_form.Label, {
        htmlFor: "replace-text-field"
      }, this.replaceWith), (0, _react2.jsx)(_textfield.default, {
        name: "replace",
        id: "replace-text-field",
        testId: "replace-field",
        appearance: "standard",
        defaultValue: replaceText,
        ref: this.replaceTextfieldRef,
        autoComplete: "off",
        onChange: this.handleReplaceChange,
        onKeyDown: this.handleReplaceKeyDown,
        onCompositionStart: this.handleCompositionStart,
        onCompositionEnd: this.handleCompositionEnd
      }), isHelperMessageVisible && this.props.findTyped === false && (0, _react2.jsx)("div", {
        ref: this.successReplacementMessageRef
      }, (0, _react2.jsx)(_form.ValidMessage, {
        testId: "message-success-replacement"
      },
      /*
      Replacement needed to trigger the SR announcement if message hasn't changed. e.g Replace button clicked twice.
      '\u00a0' is value for &nbsp
      */
      this.state.fakeSuccessReplacementMessageUpdate ?
      // Ignored via go/ees005
      // eslint-disable-next-line require-unicode-regexp
      resultsReplace.replace(/ /, "\xA0") : resultsReplace)))), (0, _react2.jsx)("div", {
        css: [_uiStyles.sectionWrapperStyles, _uiStyles.sectionWrapperStylesAlternate, _uiStyles.sectionWrapperJustified]
      }, (0, _react2.jsx)("div", {
        css: _uiStyles.orderOneStyles
      }, (0, _react2.jsx)("div", {
        css: _uiStyles.nextPreviousItemStyles
      }, (0, _react2.jsx)(_FindReplaceTooltipButton.FindReplaceTooltipButton, {
        title: this.findNext,
        icon: _chevronDownHipchatChevronDown.default,
        iconLabel: this.findNext,
        keymapDescription: 'Enter',
        onClick: this.handleFindNextClick,
        disabled: count.total <= 1
      })), (0, _react2.jsx)("div", {
        css: _uiStyles.nextPreviousItemStyles
      }, (0, _react2.jsx)(_FindReplaceTooltipButton.FindReplaceTooltipButton, {
        title: this.findPrevious,
        icon: _chevronUpHipchatChevronUp.default,
        iconLabel: this.findPrevious,
        keymapDescription: 'Shift Enter',
        onClick: this.handleFindPrevClick,
        disabled: count.total <= 1
      })), (0, _react2.jsx)(_primitives.Inline, {
        space: "space.075"
        // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
        ,
        xcss: (0, _primitives.xcss)({
          // eslint-disable-next-line @atlaskit/design-system/ensure-design-token-usage/preview
          paddingInlineStart: 'space.050',
          // eslint-disable-next-line @atlaskit/design-system/ensure-design-token-usage/preview
          paddingInlineEnd: 'space.025'
        })
      }, (0, _react2.jsx)(_new.default, {
        testId: this.replace,
        id: "replace-button",
        onClick: this.handleReplaceClick,
        isDisabled: !canReplace
      }, this.replace), (0, _react2.jsx)(_new.default, {
        appearance: "primary",
        testId: this.replaceAll,
        id: "replaceAll-button",
        onClick: this.handleReplaceAllClick,
        isDisabled: !canReplace
      }, this.replaceAll))), (0, _react2.jsx)("div", {
        css: _uiStyles.orderZeroStyles
      }, (0, _react2.jsx)(_new.default, {
        appearance: "subtle",
        testId: this.closeFindReplaceDialog,
        onClick: this.clearSearch
      }, this.closeFindReplaceDialog))));
    }
  }]);
}(_react.default.PureComponent);
var _default = exports.default = (0, _reactIntlNext.injectIntl)(Replace);
