// @atlaskit/editor-plugin-find-replace/dist/cjs/ui/styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.selectedSearchMatchClass = exports.searchMatchClass = exports.findReplaceStyles = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = require("@emotion/react");
var _colors = require("@atlaskit/theme/colors");
/* eslint-disable @atlaskit/editor/no-re-export */
// Entry file in package.json

/* eslint-disable @atlaskit/design-system/ensure-design-token-usage/preview */
/* eslint-disable @atlaskit/design-system/ensure-design-token-usage */
/* eslint-disable */

var searchMatchClass = exports.searchMatchClass = 'search-match';
var selectedSearchMatchClass = exports.selectedSearchMatchClass = 'selected-search-match';
var findReplaceStyles = exports.findReplaceStyles = (0, _react.css)((0, _defineProperty2.default)((0, _defineProperty2.default)({}, ".".concat(searchMatchClass), {
  borderRadius: '3px',
  backgroundColor: "var(--ds-background-accent-teal-subtlest, #E7F9FF)",
  boxShadow: "var(--ds-shadow-raised, ".concat("0 1px 1px 0 ".concat(_colors.N50A, ", 0 0 1px 0 ").concat(_colors.N60A), ")") + ', inset 0 0 0 1px ' + "var(--ds-border-input, ".concat("".concat(_colors.N40A), ")")
}), ".".concat(selectedSearchMatchClass), {
  backgroundColor: "var(--ds-background-accent-teal-subtle, #6CC3E0)"
}));
