// @atlaskit/media-ui/dist/es2019/BlockCard/components/Thumbnail.js
/** @jsx jsx */
import { jsx } from '@emotion/core';
import { gs, mq } from '../utils';
export const Thumbnail = (props) => {
    return props.color ? (jsx(ThumbnailWithBackground, Object.assign({}, props))) : (jsx(ThumbnailDefault, Object.assign({}, props)));
};
const sharedStyles = {
    // The dimensions of the image are set in this manner
    // in order for `flex` to respect this value.
    minWidth: [gs(13), gs(20)],
    maxWidth: [gs(13), gs(20)],
};
export const ThumbnailDefault = ({ src, testId }) => {
    return (jsx("div", { css: mq({
            ...sharedStyles,
            backgroundSize: 'cover',
            backgroundRepeat: 'no-repeat',
            backgroundPosition: '50% 50%',
            backgroundImage: `url(${src})`,
        }), "data-testid": testId }));
};
export const ThumbnailWithBackground = ({ src, color, testId }) => {
    return (jsx("div", { css: mq({
            ...sharedStyles,
            backgroundColor: color,
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center',
        }), "data-testid": testId },
        jsx("img", { src: src, css: { height: '90px', width: '90px' } })));
};
