// @atlaskit/media-card/dist/cjs/utils/lightCards/styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.wrapperStyles = void 0;
var _react = require("@emotion/react");
var _mediaUi = require("@atlaskit/media-ui");
var _colors = require("@atlaskit/theme/colors");
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var wrapperStyles = exports.wrapperStyles = function wrapperStyles(_ref) {
  var dimensions = _ref.dimensions;
  return (0, _react.css)(
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  _mediaUi.center, {
    background: "var(--ds-background-neutral, ".concat(_colors.N20, ")"),
    color: "var(--ds-icon, ".concat(_colors.N50, ")")
  },
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  _mediaUi.borderRadius, {
    maxHeight: '100%',
    maxWidth: '100%',
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    width: dimensions.width,
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
    height: dimensions.height
  });
};
