// @atlaskit/media-card/dist/cjs/card/ui/progressBar/progressBar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getNormalizedProgress = exports.ProgressBar = void 0;
var _react = _interopRequireDefault(require("react"));
var _styledBar = require("./styledBar");
var _common = require("../common");
var getNormalizedProgress = exports.getNormalizedProgress = function getNormalizedProgress(progress) {
  return Math.min(1, Math.max(0, progress || 0)) * 100;
};
var ProgressBar = exports.ProgressBar = function ProgressBar(_ref) {
  var progress = _ref.progress,
    _ref$breakpoint = _ref.breakpoint,
    breakpoint = _ref$breakpoint === void 0 ? _common.Breakpoint.SMALL : _ref$breakpoint,
    _ref$positionBottom = _ref.positionBottom,
    positionBottom = _ref$positionBottom === void 0 ? false : _ref$positionBottom,
    _ref$showOnTop = _ref.showOnTop,
    showOnTop = _ref$showOnTop === void 0 ? false : _ref$showOnTop;
  return /*#__PURE__*/_react.default.createElement(_styledBar.StyledBar, {
    progress: getNormalizedProgress(progress),
    breakpoint: breakpoint,
    positionBottom: positionBottom,
    showOnTop: showOnTop
  });
};
