// @atlaskit/media-card/dist/cjs/card/ui/progressBar/styledBar-compiled.js
/* styledBar-compiled.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StyledBar = void 0;
exports.generateResponsiveStyles = generateResponsiveStyles;
require("./styledBar-compiled.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _common = require("../common");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var height = 3;
var padding = 1;
var width = 95; // %

var smallSizeSettings = {
  marginBottom: 4
};
var largeSizeSettings = {
  marginBottom: 12
};
function generateResponsiveStyles(breakpoint, positionBottom, showOnTop) {
  var multiplier = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 1;
  var setting = breakpoint === _common.Breakpoint.SMALL ? smallSizeSettings : largeSizeSettings;
  var marginPositionBottom = _common.responsiveSettings[breakpoint].titleBox.verticalPadding;
  var marginBottom = setting.marginBottom * multiplier + (positionBottom ? marginPositionBottom : (0, _common.getTitleBoxHeight)(breakpoint));
  var spacing = "".concat(marginBottom, "px");
  return showOnTop ? {
    top: spacing
  } : {
    bottom: spacing
  };
}
var styledBarStyles = null;
var styleChild = null;
var StyledBar = exports.StyledBar = function StyledBar(props) {
  var progress = props.progress,
    breakpoint = props.breakpoint,
    positionBottom = props.positionBottom,
    showOnTop = props.showOnTop;
  return /*#__PURE__*/React.createElement("div", {
    id: "styledBar",
    role: "progressbar",
    "aria-valuenow": progress,
    className: (0, _runtime.ax)(["_1reo15vq _18m915vq _1yt4xh0m _2rko1l7b _kqswstnw _1bsb1b7y _1ltv1ghs _bfhkrcrw _4t3i12iw _vchhusvi"]),
    style: _objectSpread(_objectSpread({}, generateResponsiveStyles(breakpoint, positionBottom, showOnTop)), {}, {
      "--_1xe2l1p": (0, _runtime.ix)("".concat(width, "%")),
      "--_15ce6z9": (0, _runtime.ix)("".concat(height + padding * 2, "px")),
      "--_cw4itz": (0, _runtime.ix)("".concat(padding, "px"))
    })
  }, /*#__PURE__*/React.createElement("div", {
    style: {
      width: "".concat(progress, "%")
    },
    className: (0, _runtime.ax)(["_2rko1l7b _bfhk1p0d _1e0c1ule _4t3i1osq"])
  }));
};
StyledBar.displayName = 'StyledProgressBar';
