// @atlaskit/editor-element-browser/dist/es2019/ui/ListButtonItem.js
import _extends from "@babel/runtime/helpers/extends";
import React, { memo, useCallback, useMemo, useEffect } from 'react';
import { ToolTipContent, formatShortcut } from '@atlaskit/editor-common/keymaps';
import ChevronRightIcon from '@atlaskit/icon/utility/chevron-right';
import Lozenge from '@atlaskit/lozenge';
import { ButtonItem } from '@atlaskit/menu';
import { Box, Inline, Text, xcss } from '@atlaskit/primitives';
import Tooltip from '@atlaskit/tooltip';
import { IconRenderer } from './IconRenderer';
const shortcutStyles = xcss({
  color: 'color.text.subtle',
  font: 'font.body.small',
  backgroundColor: 'color.background.accent.gray.subtlest',
  borderRadius: 'border.radius',
  borderStyle: 'none'
});
const selectedColorStyles = xcss({
  color: 'color.text.selected'
});
const disabledShortcutStyles = xcss({
  color: 'color.text.disabled',
  backgroundColor: 'color.background.disabled'
});
const viewAllStyles = xcss({
  background: "var(--ds-surface-overlay, #FFFFFF)",
  position: 'sticky',
  bottom: '-4px',
  width: '100%',
  height: '40px',
  borderTop: `1px solid ${"var(--ds-border, #091E4224)"}`
});
const ListButtonItemBase = /*#__PURE__*/memo(({
  index,
  title,
  description,
  keyshortcut,
  isSelected,
  isDisabled,
  showDescription = false,
  renderContent,
  renderIcon,
  renderElementAfter,
  onItemSelected
}) => {
  const beforeElement = useMemo(() => {
    if (!renderIcon) {
      return null;
    }
    return /*#__PURE__*/React.createElement(IconRenderer, null, renderIcon());
  }, [renderIcon]);
  const afterElement = useMemo(() => renderElementAfter === null || renderElementAfter === void 0 ? void 0 : renderElementAfter(), [renderElementAfter]);
  const content = useMemo(() => renderContent === null || renderContent === void 0 ? void 0 : renderContent(), [renderContent]);
  return /*#__PURE__*/React.createElement(Tooltip, {
    content: /*#__PURE__*/React.createElement(ToolTipContent, {
      description: title,
      keymap: keyshortcut
    }),
    position: "top",
    ignoreTooltipPointerEvents: true
  }, tooltipProps => /*#__PURE__*/React.createElement(ButtonItem
  // eslint-disable-next-line react/jsx-props-no-spreading
  , _extends({}, tooltipProps, {
    iconBefore: beforeElement,
    iconAfter: afterElement,
    description: showDescription ? description : undefined,
    shouldDescriptionWrap: true,
    isSelected: isSelected,
    isDisabled: isDisabled,
    onClick: () => onItemSelected === null || onItemSelected === void 0 ? void 0 : onItemSelected(index)
  }), content));
});
export const ListButtonItem = /*#__PURE__*/memo(props => {
  const {
    keyshortcut,
    isSelected,
    isDisabled,
    attributes,
    title,
    setSelectedItem,
    index
  } = props;
  const shortcutComponent = useCallback(() => {
    const shortcut = keyshortcut && formatShortcut(keyshortcut);
    if (!shortcut) {
      return null;
    }
    return /*#__PURE__*/React.createElement(Box, {
      paddingInline: "space.025",
      paddingBlock: "space.025",
      xcss: [shortcutStyles, isSelected && selectedColorStyles, isDisabled && disabledShortcutStyles]
    }, shortcut);
  }, [isDisabled, isSelected, keyshortcut]);
  const contentComponent = useCallback(() => {
    return /*#__PURE__*/React.createElement(Inline, {
      space: "space.100",
      alignBlock: 'center'
    }, /*#__PURE__*/React.createElement(Text, null, title), (attributes === null || attributes === void 0 ? void 0 : attributes.new) && /*#__PURE__*/React.createElement(Lozenge, {
      appearance: "new"
    }, "New"));
  }, [attributes === null || attributes === void 0 ? void 0 : attributes.new, title]);
  useEffect(() => {
    if (isSelected) {
      setSelectedItem === null || setSelectedItem === void 0 ? void 0 : setSelectedItem({
        index
      });
    }
  }, [isSelected, setSelectedItem, index]);
  return /*#__PURE__*/React.createElement(ListButtonItemBase
  // eslint-disable-next-line react/jsx-props-no-spreading
  , _extends({}, props, {
    renderContent: contentComponent,
    renderElementAfter: shortcutComponent
  }));
});
export const ViewAllButtonItem = /*#__PURE__*/memo(({
  label,
  onClick,
  isViewAllInserts = false
}) => {
  const contentComponent = useCallback(() => {
    return /*#__PURE__*/React.createElement(Inline, {
      space: "space.100",
      alignBlock: 'center'
    }, /*#__PURE__*/React.createElement(Text, {
      color: 'color.text.selected'
    }, label), /*#__PURE__*/React.createElement(Box, {
      xcss: [selectedColorStyles]
    }, /*#__PURE__*/React.createElement(ChevronRightIcon, {
      label: label
    })));
  }, [label]);
  return isViewAllInserts ? /*#__PURE__*/React.createElement(Box, {
    xcss: viewAllStyles
  }, /*#__PURE__*/React.createElement(ListButtonItemBase, {
    index: -1,
    title: label,
    onItemSelected: onClick,
    renderContent: contentComponent,
    isViewAll: true
  })) : /*#__PURE__*/React.createElement(ListButtonItemBase, {
    index: -1,
    title: label,
    onItemSelected: onClick,
    renderContent: contentComponent
  });
});
