// @atlaskit/editor-plugin-find-replace/dist/cjs/pm-plugins/main.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.initialState = exports.createPlugin = void 0;
var _safePlugin = require("@atlaskit/editor-common/safe-plugin");
var _view = require("@atlaskit/editor-prosemirror/view");
var _pluginFactory = require("./plugin-factory");
var _pluginKey = require("./plugin-key");
var initialState = exports.initialState = {
  isActive: false,
  shouldFocus: false,
  findText: '',
  replaceText: '',
  index: 0,
  matches: [],
  decorationSet: _view.DecorationSet.empty,
  shouldMatchCase: false
};
var createPlugin = exports.createPlugin = function createPlugin(dispatch) {
  return new _safePlugin.SafePlugin({
    key: _pluginKey.findReplacePluginKey,
    state: (0, _pluginFactory.createPluginState)(dispatch, function () {
      return initialState;
    }),
    props: {
      decorations: function decorations(state) {
        var _getPluginState = (0, _pluginFactory.getPluginState)(state),
          isActive = _getPluginState.isActive,
          findText = _getPluginState.findText,
          decorationSet = _getPluginState.decorationSet;
        if (isActive && findText) {
          return decorationSet;
        }
      }
    }
  });
};
