// @atlaskit/media-picker/dist/cjs/components/dropzone/dropzone.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.DropzoneBase = exports.Dropzone = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireDefault(require("react"));
var _analyticsNext = require("@atlaskit/analytics-next");
var _mediaCommon = require("@atlaskit/media-common");
var _localUploadReact = require("../localUploadReact");
var _analytics = require("../../util/analytics");
var _ErrorFlagGroup = _interopRequireDefault(require("../errorFlagGroup/ErrorFlagGroup"));
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function dragContainsFiles(event) {
  if (!event.dataTransfer) {
    return false;
  }
  var types = event.dataTransfer.types;
  return Array.from(types).indexOf('Files') > -1;
}
var COMPONENT_NAME = 'dropzone';
var DropzoneBase = exports.DropzoneBase = /*#__PURE__*/function (_LocalUploadComponent) {
  function DropzoneBase(props) {
    var _this;
    (0, _classCallCheck2.default)(this, DropzoneBase);
    _this = _callSuper(this, DropzoneBase, [props, COMPONENT_NAME]);
    (0, _defineProperty2.default)(_this, "uiActive", false);
    (0, _defineProperty2.default)(_this, "addContainerListeners", function (container) {
      var target = container !== null && container !== void 0 ? container : _this.getContainer();
      // TODO: migrate this file to Pragmatic drag and drop
      /* eslint-disable @atlaskit/design-system/no-direct-use-of-web-platform-drag-and-drop */
      target.addEventListener('dragover', _this.onDragOver, false);
      target.addEventListener('dragleave', _this.onDragLeave, false);
      target.addEventListener('drop', _this.onFileDropped);
      /* eslint-enable @atlaskit/design-system/no-direct-use-of-web-platform-drag-and-drop */
    });
    (0, _defineProperty2.default)(_this, "removeContainerListeners", function (container) {
      var target = container !== null && container !== void 0 ? container : _this.getContainer();
      target.removeEventListener('dragover', _this.onDragOver, false);
      target.removeEventListener('dragleave', _this.onDragLeave, false);
      target.removeEventListener('drop', _this.onFileDropped);
    });
    (0, _defineProperty2.default)(_this, "onDragOver", function (event) {
      event.preventDefault();
      if (event.dataTransfer && dragContainsFiles(event)) {
        var dataTransfer = event.dataTransfer;
        var allowed;
        try {
          allowed = dataTransfer.effectAllowed;
        } catch (e) {} // the error is expected in IE11

        dataTransfer.dropEffect = 'move' === allowed || 'linkMove' === allowed ? 'move' : 'copy';
        var length = _this.getDraggedItemsLength(dataTransfer);
        // [EDM-1636]: needed in order to make multiple dropzones in the page to work
        event.stopPropagation();
        _this.emitDragOver({
          length: length
        });
      }
    });
    (0, _defineProperty2.default)(_this, "onDragLeave", function (e) {
      if (e.dataTransfer) {
        e.preventDefault();
        var length = 0;
        if (dragContainsFiles(e)) {
          var dataTransfer = e.dataTransfer;
          length = _this.getDraggedItemsLength(dataTransfer);
        }
        _this.emitDragLeave({
          length: length
        });
      }
    });
    (0, _defineProperty2.default)(_this, "onFileDropped", /*#__PURE__*/function () {
      var _ref = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(dragEvent) {
        var files;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              if (dragEvent.dataTransfer) {
                _context.next = 2;
                break;
              }
              return _context.abrupt("return");
            case 2:
              dragEvent.preventDefault();
              dragEvent.stopPropagation();
              // refreshes uploadParams as only set once in parent constructor
              _this.setUploadParams(_this.props.config.uploadParams);
              _this.onDrop(dragEvent);
              files = Array.from(dragEvent.dataTransfer.files);
              _this.uploadService.addFiles(files);
            case 8:
            case "end":
              return _context.stop();
          }
        }, _callee);
      }));
      return function (_x) {
        return _ref.apply(this, arguments);
      };
    }());
    (0, _defineProperty2.default)(_this, "onDrop", function (e) {
      if (e.dataTransfer && dragContainsFiles(e)) {
        var dataTransfer = e.dataTransfer;
        var fileCount = _this.getDraggedItemsLength(dataTransfer);
        _this.sendAnalyticsAndEmitDragLeave(fileCount);
      }
    });
    (0, _defineProperty2.default)(_this, "sendAnalyticsAndEmitDragLeave", function (fileCount) {
      _this.fireAnalyticsEvent('droppedInto', fileCount);
      if (_this.props.onDrop) {
        _this.props.onDrop();
      }
      _this.emitDragLeave({
        length: fileCount
      });
    });
    return _this;
  }
  (0, _inherits2.default)(DropzoneBase, _LocalUploadComponent);
  return (0, _createClass2.default)(DropzoneBase, [{
    key: "getContainer",
    value: function getContainer() {
      var container = this.props.config.container;
      return container || document.body;
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      var onCancelFn = this.props.onCancelFn;
      this.addContainerListeners(this.getContainer());
      if (onCancelFn) {
        onCancelFn(this.cancel);
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.removeContainerListeners(this.getContainer());
    }
  }, {
    key: "UNSAFE_componentWillReceiveProps",
    value: function UNSAFE_componentWillReceiveProps(nextProps) {
      var newContainer = nextProps.config.container;
      var oldContainer = this.props.config.container;
      if (newContainer !== oldContainer) {
        this.removeContainerListeners(oldContainer);
        this.addContainerListeners(newContainer);
      }
    }
  }, {
    key: "getDraggedItemsLength",
    value:
    // Cross-browser way of getting dragged items length, we prioritize "items" if present
    // https://www.w3.org/TR/html51/editing.html#the-datatransfer-interface
    // This method is used on 'dragover' and we have no way to retrieve FileSystemFileEntry,
    // which contains info about if the dropped item is a file or directory. That info is only
    // available on 'drop'
    function getDraggedItemsLength(dataTransfer) {
      if (dataTransfer.items) {
        var items = Array.from(dataTransfer.items);
        return items.filter(function (i) {
          return i.kind === 'file';
        }).length;
      }
      // This is required for IE11
      return dataTransfer.files.length;
    }
  }, {
    key: "emitDragOver",
    value: function emitDragOver(payload) {
      if (!this.uiActive) {
        var onDragEnter = this.props.onDragEnter;
        this.uiActive = true;
        this.fireAnalyticsEvent('draggedInto', payload.length);
        if (onDragEnter) {
          onDragEnter(payload);
        }
      }
    }
  }, {
    key: "emitDragLeave",
    value: function emitDragLeave(payload) {
      var _this2 = this;
      if (this.uiActive) {
        this.uiActive = false;
        /*
            when drag over child elements, container will issue dragleave and then dragover immediately.
            The 50ms timeout will prevent from issuing that "false" dragleave event
            */
        window.setTimeout(function () {
          if (!_this2.uiActive) {
            var onDragLeave = _this2.props.onDragLeave;
            _this2.fireAnalyticsEvent('draggedOut', payload.length);
            if (onDragLeave) {
              onDragLeave(payload);
            }
          }
        }, 50);
      }
    }
  }, {
    key: "fireAnalyticsEvent",
    value: function fireAnalyticsEvent(action, fileCount) {
      var createAnalyticsEvent = this.props.createAnalyticsEvent;
      if (createAnalyticsEvent) {
        var payload = {
          eventType: 'ui',
          actionSubject: 'dropzone',
          action: action,
          attributes: {
            fileCount: fileCount
          }
        };
        var analyticsEvent = createAnalyticsEvent(payload);
        analyticsEvent.fire(_mediaCommon.ANALYTICS_MEDIA_CHANNEL);
      }
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement(_ErrorFlagGroup.default, {
        flagData: this.state.errorFlags,
        onFlagDismissed: this.dismissErrorFlag
      });
    }
  }]);
}(_localUploadReact.LocalUploadComponentReact);
var _default = exports.default = DropzoneBase;
var Dropzone = exports.Dropzone = (0, _mediaCommon.withMediaAnalyticsContext)((0, _analytics.getPackageAttributes)(COMPONENT_NAME))((0, _analyticsNext.withAnalyticsEvents)()(DropzoneBase));
