// react-loosely-lazy/dist/cjs/config/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "MODE", {
  enumerable: true,
  get: function () {
    return _constants.MODE;
  }
});
exports.setConfig = exports.getConfig = void 0;
var _constants = require("./constants");
const DEFAULT_CROSS_ORIGIN = undefined;
const DEFAULT_MANIFEST = {
  publicPath: '/',
  assets: {}
};
const DEFAULT_MODE = _constants.MODE.HYDRATE;
const DEFAULT_RETRY = 2;
const config = {
  crossOrigin: DEFAULT_CROSS_ORIGIN,
  manifest: DEFAULT_MANIFEST,
  mode: DEFAULT_MODE,
  retry: DEFAULT_RETRY
};
const getConfig = () => config;
exports.getConfig = getConfig;
const setConfig = nextConfig => {
  var _nextConfig$crossOrig, _nextConfig$manifest, _nextConfig$mode, _nextConfig$retry;
  config.crossOrigin = (_nextConfig$crossOrig = nextConfig.crossOrigin) !== null && _nextConfig$crossOrig !== void 0 ? _nextConfig$crossOrig : DEFAULT_CROSS_ORIGIN;
  config.manifest = (_nextConfig$manifest = nextConfig.manifest) !== null && _nextConfig$manifest !== void 0 ? _nextConfig$manifest : DEFAULT_MANIFEST;
  config.mode = (_nextConfig$mode = nextConfig.mode) !== null && _nextConfig$mode !== void 0 ? _nextConfig$mode : DEFAULT_MODE;
  config.retry = (_nextConfig$retry = nextConfig.retry) !== null && _nextConfig$retry !== void 0 ? _nextConfig$retry : DEFAULT_RETRY;
};
exports.setConfig = setConfig;
