// @atlaskit/editor-core/dist/cjs/presets/context.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PresetContextProvider = exports.EditorAPIContext = void 0;
exports.usePresetContext = usePresetContext;
exports.useSetPresetContext = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
// eslint-disable-next-line @typescript-eslint/no-explicit-any

var EditorAPIContext = exports.EditorAPIContext = /*#__PURE__*/_react.default.createContext({});
var PresetContextProvider = exports.PresetContextProvider = function PresetContextProvider(_ref) {
  var children = _ref.children;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  var _useState = (0, _react.useState)(),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    editorApi = _useState2[0],
    setEditorApi = _useState2[1];
  var contextValue = (0, _react.useMemo)(function () {
    return {
      editorApi: editorApi,
      setEditorApi: setEditorApi
    };
  }, [editorApi, setEditorApi]);
  return /*#__PURE__*/_react.default.createElement(EditorAPIContext.Provider, {
    value: contextValue
  }, children);
};
function usePresetContext() {
  var _useContext = (0, _react.useContext)(EditorAPIContext),
    editorApi = _useContext.editorApi;
  return editorApi;
}
var useSetPresetContext = exports.useSetPresetContext = function useSetPresetContext() {
  var _useContext2 = (0, _react.useContext)(EditorAPIContext),
    setEditorApi = _useContext2.setEditorApi;
  return setEditorApi;
};
