// @atlaskit/editor-core/dist/es2019/create-editor/editorStateNotificationPlugin.js
import { SafePlugin } from '@atlaskit/editor-common/safe-plugin';
import { PluginKey } from '@atlaskit/editor-prosemirror/state';
import { fg } from '@atlaskit/platform-feature-flags';
const key = new PluginKey('editorStateNotificationPlugin');
export const createEditorStateNotificationPlugin = (onEditorStateUpdated, onEditorViewStateUpdatedCallbacks) => {
  let transactions = [];
  return new SafePlugin({
    key: key,
    state: {
      init() {
        return {
          latestTransaction: undefined
        };
      },
      apply(tr) {
        transactions.push(tr);
        return {
          latestTransaction: tr
        };
      }
    },
    view: () => {
      return {
        update: (view, oldEditorState) => {
          var _key$getState;
          const originalTransaction = (_key$getState = key.getState(view.state)) === null || _key$getState === void 0 ? void 0 : _key$getState.latestTransaction;
          if (originalTransaction && fg('platform_editor_migrate_state_updates')) {
            onEditorViewStateUpdatedCallbacks.forEach(entry => {
              entry.callback({
                originalTransaction,
                transactions: transactions,
                oldEditorState,
                newEditorState: view.state
              });
            });
            transactions = [];
          }
          onEditorStateUpdated({
            oldEditorState,
            newEditorState: view.state
          });
        }
      };
    }
  });
};
