// @atlaskit/link-datasource/dist/es2019/ui/common/modal/popup-select/control.js
/* control.tsx generated by @compiled/babel-plugin v0.36.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./control.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
import React from 'react';
import { fg } from '@atlaskit/platform-feature-flags';
import { Box } from '@atlaskit/primitives/compiled';
import { components } from '@atlaskit/select';
import { N40 } from '@atlaskit/theme/colors';
import { withFeatureFlaggedComponent } from '../../../../common/utils/withFeatureFlaggedComponent';
const popupCustomControlStylesOld = () => ({
  display: 'flex',
  padding: "var(--ds-space-050, 4px)",
  border: 'none',
  borderBottom: `solid 1px ${`var(--ds-border, ${N40})`}`,
  minHeight: 'auto'
});
const popupCustomControlStyles = () => ({
  display: 'flex',
  borderRadius: '3px',
  border: `solid 1px ${"var(--ds-border-input, #8C8F97)"}`
});
const popupCustomControlVisualRefreshStyles = {
  container: "_ca0q1b66 _n3td1b66 _19bvh9hh _u5f3h9hh"
};
export const CustomControlOld = ({
  children,
  ...innerProps
}) => /*#__PURE__*/React.createElement(components.Control, _extends({}, innerProps, {
  getStyles: popupCustomControlStylesOld
}), children);
export const CustomControl = ({
  children,
  ...innerProps
}) => /*#__PURE__*/React.createElement(Box, {
  xcss: popupCustomControlVisualRefreshStyles.container
}, /*#__PURE__*/React.createElement(components.Control, _extends({}, innerProps, {
  getStyles: popupCustomControlStyles
}), children));
export default withFeatureFlaggedComponent(CustomControlOld, CustomControl, () => fg('platform-linking-visual-refresh-sllv'));
