// @atlaskit/editor-plugin-selection-extension/dist/es2019/ui/toolbar/SelectionExtensionDropdownMenu.js
import React, { useState } from 'react';
import { injectIntl } from 'react-intl-next';
import { ACTION, ACTION_SUBJECT, ACTION_SUBJECT_ID, EVENT_TYPE } from '@atlaskit/editor-common/analytics';
import { DropdownMenuWithKeyboardNavigation as DropdownMenu } from '@atlaskit/editor-common/ui-menu';
import { SelectionExtensionDropdownMenuButton } from './SelectionExtensionDropdownMenuButton';
const SelectionExtensionDropdownMenuComponent = /*#__PURE__*/React.memo(({
  items,
  onItemActivated,
  editorAnalyticsAPI
}) => {
  const [isMenuOpen, setIsMenuOpen] = useState(false);
  return /*#__PURE__*/React.createElement(DropdownMenu, {
    section: {
      hasSeparator: true
    },
    isOpen: isMenuOpen,
    items: items,
    fitHeight: 188,
    fitWidth: 136,
    onItemActivated: onItemActivated,
    "data-testid": "selection-extension-dropdown-menu"
  }, /*#__PURE__*/React.createElement(SelectionExtensionDropdownMenuButton, {
    "aria-expanded": isMenuOpen,
    selected: isMenuOpen,
    onClick: () => setIsMenuOpen(prevIsMenuOpen => {
      const nextIsMenuOpen = !prevIsMenuOpen;
      if (editorAnalyticsAPI) {
        editorAnalyticsAPI.fireAnalyticsEvent({
          action: ACTION.CLICKED,
          actionSubject: ACTION_SUBJECT.BUTTON,
          actionSubjectId: ACTION_SUBJECT_ID.EDITOR_PLUGIN_SELECTION_EXTENSION_DROPDOWN,
          eventType: EVENT_TYPE.TRACK,
          attributes: {
            toggle: nextIsMenuOpen ? ACTION.OPENED : ACTION.CLOSED
          }
        });
      }
      return nextIsMenuOpen;
    })
  }));
});
export const SelectionExtensionDropdownMenu = injectIntl(SelectionExtensionDropdownMenuComponent);
