// @atlaskit/editor-plugin-table/dist/cjs/ui/TableFullWidthLabel/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FullWidthDisplay = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactIntlNext = require("react-intl-next");
var _messages = require("@atlaskit/editor-common/messages");
var _primitives = require("@atlaskit/primitives");
var tableFullWidthLabelWrapperStyles = (0, _primitives.xcss)({
  height: "var(--ds-space-400, 32px)",
  display: 'flex',
  backgroundColor: 'elevation.surface.overlay',
  borderRadius: 'border.radius',
  boxShadow: 'elevation.shadow.overlay',
  // eslint-disable-next-line @atlaskit/design-system/use-tokens-typography
  lineHeight: 1,
  boxSizing: 'border-box',
  alignItems: 'center'
});
var tableFullWidthLabelStyles = (0, _primitives.xcss)({
  marginLeft: 'space.100',
  marginRight: 'space.100',
  paddingLeft: 'space.075',
  paddingRight: 'space.075',
  paddingTop: 'space.050',
  paddingBottom: 'space.050'
});
var FullWidthDisplay = exports.FullWidthDisplay = function FullWidthDisplay() {
  var _useIntl = (0, _reactIntlNext.useIntl)(),
    formatMessage = _useIntl.formatMessage;
  return /*#__PURE__*/_react.default.createElement(_primitives.Box, {
    xcss: tableFullWidthLabelWrapperStyles
  }, /*#__PURE__*/_react.default.createElement(_primitives.Inline, {
    xcss: tableFullWidthLabelStyles
  }, formatMessage(_messages.tableMessages.fullWidthLabel)));
};
