// @atlaskit/link-picker/dist/cjs/ui/link-picker/form-footer/link-picker-submit-button/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.messages = exports.LinkPickerSubmitButton = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactIntlNext = require("react-intl-next");
var _loadingButton = _interopRequireDefault(require("@atlaskit/button/loading-button"));
var _new = _interopRequireDefault(require("@atlaskit/button/new"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _utils = require("./utils");
var messages = exports.messages = (0, _reactIntlNext.defineMessages)({
  saveButton: {
    id: 'fabric.linkPicker.button.save',
    defaultMessage: 'Save',
    description: 'Button to save edited link'
  },
  insertButton: {
    id: 'fabric.linkPicker.button.insert',
    defaultMessage: 'Insert',
    description: 'Button to insert searched or selected link'
  }
});
var LinkPickerSubmitButton = exports.LinkPickerSubmitButton = function LinkPickerSubmitButton(_ref) {
  var isEditing = _ref.isEditing,
    isLoading = _ref.isLoading,
    isSubmitting = _ref.isSubmitting,
    customSubmitButtonLabel = _ref.customSubmitButtonLabel,
    error = _ref.error,
    items = _ref.items,
    queryState = _ref.queryState,
    submitMessageId = _ref.submitMessageId,
    testId = _ref.testId,
    url = _ref.url;
  var intl = (0, _reactIntlNext.useIntl)();
  var insertButtonMsg = isEditing ? messages.saveButton : messages.insertButton;
  var isSubmitDisabled = (0, _utils.checkSubmitDisabled)(isLoading, isSubmitting, error, url, queryState, items);
  var ButtonComponent = (0, _platformFeatureFlags.fg)('platform-link-picker-remove-legacy-button') ? _new.default : _loadingButton.default;
  return /*#__PURE__*/_react.default.createElement(ButtonComponent, {
    type: "submit",
    appearance: "primary",
    testId: testId,
    isDisabled: isSubmitDisabled,
    "aria-labelledby": isSubmitting ? submitMessageId : undefined,
    isLoading: isSubmitting
  }, customSubmitButtonLabel ? intl.formatMessage(customSubmitButtonLabel) : intl.formatMessage(insertButtonMsg));
};
