// @atlaskit/analytics-namespaced-context/dist/cjs/FabricEditorAnalyticsContext.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FabricEditorAnalyticsContext = exports.EDITOR_CONTEXT = exports.EDITOR_APPEARANCE_CONTEXT = void 0;
var _createNamespaceContext = _interopRequireDefault(require("./helper/createNamespaceContext"));
var EDITOR_CONTEXT = exports.EDITOR_CONTEXT = 'fabricEditorCtx';
var EDITOR_APPEARANCE_CONTEXT = exports.EDITOR_APPEARANCE_CONTEXT = /*#__PURE__*/function (EDITOR_APPEARANCE_CONTEXT) {
  EDITOR_APPEARANCE_CONTEXT["FIXED_WIDTH"] = "fixedWidth";
  EDITOR_APPEARANCE_CONTEXT["FULL_WIDTH"] = "fullWidth";
  EDITOR_APPEARANCE_CONTEXT["COMMENT"] = "comment";
  EDITOR_APPEARANCE_CONTEXT["CHROMELESS"] = "chromeless";
  EDITOR_APPEARANCE_CONTEXT["MOBILE"] = "mobile";
  return EDITOR_APPEARANCE_CONTEXT;
}({});
var FabricEditorAnalyticsContext = exports.FabricEditorAnalyticsContext = (0, _createNamespaceContext.default)(EDITOR_CONTEXT, 'FabricEditorAnalyticsContext');
