// @atlaskit/analytics-namespaced-context/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AIMateAnalyticsContext", {
  enumerable: true,
  get: function get() {
    return _AIMateAnalyticsContext.AIMateAnalyticsContext;
  }
});
Object.defineProperty(exports, "AI_MATE_CONTEXT", {
  enumerable: true,
  get: function get() {
    return _AIMateAnalyticsContext.AI_MATE_CONTEXT;
  }
});
Object.defineProperty(exports, "ATLAS_CONTEXT", {
  enumerable: true,
  get: function get() {
    return _AtlasAnalyticsContext.ATLAS_CONTEXT;
  }
});
Object.defineProperty(exports, "AVPAnalyticsContext", {
  enumerable: true,
  get: function get() {
    return _AVPAnalyticsContext.AVPAnalyticsContext;
  }
});
Object.defineProperty(exports, "AVP_CONTEXT", {
  enumerable: true,
  get: function get() {
    return _AVPAnalyticsContext.AVP_CONTEXT;
  }
});
Object.defineProperty(exports, "AtlasAnalyticsContext", {
  enumerable: true,
  get: function get() {
    return _AtlasAnalyticsContext.AtlasAnalyticsContext;
  }
});
Object.defineProperty(exports, "EDITOR_APPEARANCE_CONTEXT", {
  enumerable: true,
  get: function get() {
    return _FabricEditorAnalyticsContext.EDITOR_APPEARANCE_CONTEXT;
  }
});
Object.defineProperty(exports, "EDITOR_CONTEXT", {
  enumerable: true,
  get: function get() {
    return _FabricEditorAnalyticsContext.EDITOR_CONTEXT;
  }
});
Object.defineProperty(exports, "ELEMENTS_CONTEXT", {
  enumerable: true,
  get: function get() {
    return _FabricElementsAnalyticsContext.ELEMENTS_CONTEXT;
  }
});
Object.defineProperty(exports, "FabricEditorAnalyticsContext", {
  enumerable: true,
  get: function get() {
    return _FabricEditorAnalyticsContext.FabricEditorAnalyticsContext;
  }
});
Object.defineProperty(exports, "FabricElementsAnalyticsContext", {
  enumerable: true,
  get: function get() {
    return _FabricElementsAnalyticsContext.FabricElementsAnalyticsContext;
  }
});
Object.defineProperty(exports, "GROWTH_CONTEXT", {
  enumerable: true,
  get: function get() {
    return _GrowthAnalyticsContext.GROWTH_CONTEXT;
  }
});
Object.defineProperty(exports, "GrowthAnalyticsContext", {
  enumerable: true,
  get: function get() {
    return _GrowthAnalyticsContext.GrowthAnalyticsContext;
  }
});
Object.defineProperty(exports, "LINKING_PLATFORM_CONTEXT", {
  enumerable: true,
  get: function get() {
    return _LinkingPlatformAnalyticsContext.LINKING_PLATFORM_CONTEXT;
  }
});
Object.defineProperty(exports, "LinkingPlatformAnalyticsContext", {
  enumerable: true,
  get: function get() {
    return _LinkingPlatformAnalyticsContext.LinkingPlatformAnalyticsContext;
  }
});
Object.defineProperty(exports, "MEDIA_CONTEXT", {
  enumerable: true,
  get: function get() {
    return _MediaAnalyticsContext.MEDIA_CONTEXT;
  }
});
Object.defineProperty(exports, "NAVIGATION_CONTEXT", {
  enumerable: true,
  get: function get() {
    return _NavigationAnalyticsContext.NAVIGATION_CONTEXT;
  }
});
Object.defineProperty(exports, "NOTIFICATIONS_CONTEXT", {
  enumerable: true,
  get: function get() {
    return _NotificationsAnalyticsContext.NOTIFICATIONS_CONTEXT;
  }
});
Object.defineProperty(exports, "NavigationAnalyticsContext", {
  enumerable: true,
  get: function get() {
    return _NavigationAnalyticsContext.NavigationAnalyticsContext;
  }
});
Object.defineProperty(exports, "NotificationsAnalyticsContext", {
  enumerable: true,
  get: function get() {
    return _NotificationsAnalyticsContext.NotificationsAnalyticsContext;
  }
});
Object.defineProperty(exports, "PEOPLE_TEAMS_CONTEXT", {
  enumerable: true,
  get: function get() {
    return _PeopleTeamsAnalyticsContext.PEOPLE_TEAMS_CONTEXT;
  }
});
Object.defineProperty(exports, "POST_OFFICE_CONTEXT", {
  enumerable: true,
  get: function get() {
    return _PostOfficeAnalyticsContext.POST_OFFICE_CONTEXT;
  }
});
Object.defineProperty(exports, "PeopleTeamsAnalyticsContext", {
  enumerable: true,
  get: function get() {
    return _PeopleTeamsAnalyticsContext.PeopleTeamsAnalyticsContext;
  }
});
Object.defineProperty(exports, "PostOfficeAnalyticsContext", {
  enumerable: true,
  get: function get() {
    return _PostOfficeAnalyticsContext.PostOfficeAnalyticsContext;
  }
});
Object.defineProperty(exports, "RECENT_WORK_CONTEXT", {
  enumerable: true,
  get: function get() {
    return _RecentWorkAnalyticsContext.RECENT_WORK_CONTEXT;
  }
});
Object.defineProperty(exports, "RecentWorkAnalyticsContext", {
  enumerable: true,
  get: function get() {
    return _RecentWorkAnalyticsContext.RecentWorkAnalyticsContext;
  }
});
var _FabricElementsAnalyticsContext = require("./FabricElementsAnalyticsContext");
var _NavigationAnalyticsContext = require("./NavigationAnalyticsContext");
var _FabricEditorAnalyticsContext = require("./FabricEditorAnalyticsContext");
var _MediaAnalyticsContext = require("./MediaAnalyticsContext");
var _PeopleTeamsAnalyticsContext = require("./PeopleTeamsAnalyticsContext");
var _NotificationsAnalyticsContext = require("./NotificationsAnalyticsContext");
var _RecentWorkAnalyticsContext = require("./RecentWorkAnalyticsContext");
var _AtlasAnalyticsContext = require("./AtlasAnalyticsContext");
var _LinkingPlatformAnalyticsContext = require("./LinkingPlatformAnalyticsContext");
var _PostOfficeAnalyticsContext = require("./PostOfficeAnalyticsContext");
var _AIMateAnalyticsContext = require("./AIMateAnalyticsContext");
var _AVPAnalyticsContext = require("./AVPAnalyticsContext");
var _GrowthAnalyticsContext = require("./GrowthAnalyticsContext");
