// @atlaskit/smart-card/dist/es2019/view/FlexibleCard/components/container/layered-link/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import "./index.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { useMouseDownEvent } from '../../../../../state/analytics/useLinkClicked';
const styles = null;

/**
 * An anchor component to cover the entire container creating a clickable area.
 * @internal
 * @see `clickableContainer`
 */
const LayeredLink = ({
  onClick,
  target,
  testId,
  text,
  url
}) => {
  const onMouseDown = useMouseDownEvent();
  return /*#__PURE__*/React.createElement("a", {
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
    className: ax(["_1reo15vq _18m915vq _154iidpf _1xi2idpf _94n5idpf _1ltvidpf _azhw1osq _o5721q9c _hboxstnw _obkmidpf _a52dkb7n", "layered-link"]),
    "data-testid": `${testId}-layered-link`,
    href: url,
    onClick: onClick,
    onMouseDown: onMouseDown,
    target: target,
    tabIndex: -1 // Hide tab index and let the title link be the link.
  }, text);
};
export default LayeredLink;
