// @atlaskit/smart-card/dist/cjs/state/context/provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var react_2 = require("react");
var redux_1 = require("redux");
var reducers_1 = require("../reducers");
var _1 = require(".");
var constants_1 = require("../actions/constants");
var client_1 = tslib_1.__importDefault(require("../../client"));
function SmartCardProvider(_a) {
    var storeOptions = _a.storeOptions, customClient = _a.client, customCacheOptions = _a.cacheOptions, customAuthFlow = _a.authFlow, children = _a.children;
    var parentContext = react_2.useContext(_1.SmartCardContext);
    var providerValue = react_2.useMemo(function () {
        var initialState = (storeOptions || { initialState: {} }).initialState;
        var client = customClient || new client_1.default();
        var store = redux_1.createStore(reducers_1.cardReducer, initialState);
        var authFlow = customAuthFlow || 'oauth2';
        var cacheOptions = customCacheOptions || {
            maxAge: constants_1.MAX_RELOAD_DELAY,
            maxLoadingDelay: constants_1.MAX_LOADING_DELAY,
        };
        return {
            store: store,
            connections: {
                client: client,
            },
            config: tslib_1.__assign(tslib_1.__assign({}, cacheOptions), { authFlow: authFlow }),
        };
    }, [customClient, customCacheOptions, customAuthFlow, storeOptions]);
    return (react_1.default.createElement(_1.SmartCardContext.Provider, { value: parentContext || providerValue }, children));
}
exports.SmartCardProvider = SmartCardProvider;
exports.default = SmartCardProvider;
