// @atlaskit/editor-common/dist/esm/types/annotation.js
import { __read, __spread } from "tslib";
import EventEmitter from 'events';
import { AnnotationTypes, } from '@atlaskit/adf-schema';
export var AnnotationUpdateEvent;
(function (AnnotationUpdateEvent) {
    AnnotationUpdateEvent["SET_ANNOTATION_FOCUS"] = "SET_ANNOTATION_FOCUS";
    AnnotationUpdateEvent["SET_ANNOTATION_STATE"] = "SET_ANNOTATION_STATE";
    AnnotationUpdateEvent["REMOVE_ANNOTATION_FOCUS"] = "REMOVE_ANNOTATION_FOCUS";
})(AnnotationUpdateEvent || (AnnotationUpdateEvent = {}));
var AnnotationUpdateEmitter = /** @class */ (function () {
    function AnnotationUpdateEmitter() {
        this.emitter = new EventEmitter();
    }
    AnnotationUpdateEmitter.prototype.emit = function (event) {
        var _a;
        var payload = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            payload[_i - 1] = arguments[_i];
        }
        if (typeof payload === 'undefined') {
            return this.emitter.emit(event);
        }
        else {
            return (_a = this.emitter).emit.apply(_a, __spread([event], payload));
        }
    };
    AnnotationUpdateEmitter.prototype.on = function (event, listener) {
        return this.emitter.on(event, listener);
    };
    AnnotationUpdateEmitter.prototype.off = function (event, listener) {
        return this.emitter.off(event, listener);
    };
    AnnotationUpdateEmitter.prototype.listeners = function (event) {
        return this.emitter.listeners(event);
    };
    return AnnotationUpdateEmitter;
}());
export { AnnotationUpdateEmitter };
