// @atlaskit/adf-utils/dist/cjs/sanitize/sanitize-node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var traverse_1 = require("../traverse/traverse");
var remove_marks_1 = require("./remove-marks");
function sanitizeNode(json) {
    var sanitizedJSON = traverse_1.traverse(json, {
        text: function (node) {
            if (!node || !Array.isArray(node.marks)) {
                return node;
            }
            return tslib_1.__assign(tslib_1.__assign({}, node), { marks: node.marks.filter(function (mark) { return mark.type !== 'typeAheadQuery'; }) });
        },
        status: function (node) {
            if (node.attrs && !!node.attrs.text) {
                return remove_marks_1.removeMarks(node);
            }
            return false; // empty status
        },
        emoji: remove_marks_1.removeMarks,
        mention: remove_marks_1.removeMarks,
        date: remove_marks_1.removeMarks,
        hardBreak: remove_marks_1.removeMarks,
        inlineCard: remove_marks_1.removeMarks,
    });
    return sanitizedJSON;
}
exports.sanitizeNode = sanitizeNode;
