// @atlaskit/user-picker/dist/esm/components/CustomOption/main.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
function _callSuper(t, o, e) { return o = _getPrototypeOf(o), _possibleConstructorReturn(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], _getPrototypeOf(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
import { B400, N800, N200 } from '@atlaskit/theme/colors';
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import { jsx } from '@emotion/react';
import React from 'react';
import { AvatarItemOption, textWrapper } from '../AvatarItemOption';
import { HighlightText } from '../HighlightText';
import { SizeableAvatar } from '../SizeableAvatar';
export var CustomOption = /*#__PURE__*/function (_React$PureComponent) {
  function CustomOption() {
    var _this;
    _classCallCheck(this, CustomOption);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, CustomOption, [].concat(args));
    _defineProperty(_this, "getPrimaryText", function () {
      var _this$props$data = _this.props.data,
        name = _this$props$data.name,
        highlight = _this$props$data.highlight;
      return [jsx("span", {
        key: "name"
        // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
        ,
        css: textWrapper(_this.props.isSelected ? "var(--ds-text-selected, ".concat(B400, ")") : "var(--ds-text, ".concat(N800, ")"))
      }, jsx(HighlightText, {
        highlights: highlight && highlight.name
      }, name))];
    });
    _defineProperty(_this, "getBylineComponent", function (isSelected, message) {
      return jsx("span", {
        // eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
        css: textWrapper(isSelected ? "var(--ds-text-selected, ".concat(B400, ")") : "var(--ds-text-subtlest, ".concat(N200, ")")),
        "data-testid": "user-picker-custom-secondary-text"
      }, message);
    });
    _defineProperty(_this, "renderByline", function () {
      var _this$props$data2;
      if (!((_this$props$data2 = _this.props.data) !== null && _this$props$data2 !== void 0 && _this$props$data2.byline)) {
        return undefined;
      }
      return _this.getBylineComponent(_this.props.isSelected, _this.props.data.byline);
    });
    _defineProperty(_this, "renderAvatar", function () {
      var avatarUrl = _this.props.data.avatarUrl;
      return jsx(SizeableAvatar, {
        appearance: "big",
        src: avatarUrl
      });
    });
    _defineProperty(_this, "getLozengeProps", function () {
      return typeof _this.props.data.lozenge === 'string' ? {
        text: _this.props.data.lozenge
      } : _this.props.data.lozenge;
    });
    return _this;
  }
  _inherits(CustomOption, _React$PureComponent);
  return _createClass(CustomOption, [{
    key: "render",
    value: function render() {
      return jsx(AvatarItemOption, {
        avatar: this.renderAvatar(),
        isDisabled: this.props.data.isDisabled,
        lozenge: this.getLozengeProps(),
        primaryText: this.getPrimaryText(),
        secondaryText: this.renderByline()
      });
    }
  }]);
}(React.PureComponent);
