// @atlaskit/profilecard/dist/cjs/components/common/PopupTrigger.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PopupTrigger = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireWildcard(require("react"));
var _excluded = ["children", "trigger", "showProfilecard", "hideProfilecard", "ariaLabelledBy"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var PopupTriggerInner = function PopupTriggerInner(_ref, ref) {
  var children = _ref.children,
    trigger = _ref.trigger,
    showProfilecard = _ref.showProfilecard,
    hideProfilecard = _ref.hideProfilecard,
    ariaLabelledBy = _ref.ariaLabelledBy,
    props = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  var onMouseEnter = (0, _react.useCallback)(function () {
    showProfilecard();
  }, [showProfilecard]);
  var onKeyPress = (0, _react.useCallback)(function (event) {
    if (event.key === 'Enter' || event.key === ' ') {
      event.preventDefault();
      showProfilecard();
    }
  }, [showProfilecard]);
  var onClick = (0, _react.useCallback)(function (event) {
    // Prevent the click event from propagating to parent containers.
    event.stopPropagation();
    showProfilecard();
  }, [showProfilecard]);
  var containerListeners = (0, _react.useMemo)(function () {
    return trigger === 'hover' ? {
      onMouseEnter: onMouseEnter,
      onMouseLeave: hideProfilecard,
      onBlur: hideProfilecard,
      onKeyPress: onKeyPress
    } : {
      onClick: onClick,
      onKeyPress: onKeyPress
    };
  }, [hideProfilecard, onClick, onKeyPress, onMouseEnter, trigger]);
  return /*#__PURE__*/_react.default.createElement("span", (0, _extends2.default)({}, props, containerListeners, {
    ref: ref,
    "aria-labelledby": ariaLabelledBy
  }), children);
};
var PopupTrigger = exports.PopupTrigger = /*#__PURE__*/_react.default.forwardRef(PopupTriggerInner);
