// @atlaskit/jql-editor/dist/cjs/hooks/use-editor-theme/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useEditorThemeContext = exports.useEditorTheme = exports.EditorThemeContext = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = require("react");
var _noop = _interopRequireDefault(require("lodash/noop"));
var defaultMaxRows = 3;
var expandedRows = 15;
var defaultEditorTheme = {
  defaultMaxRows: defaultMaxRows,
  expanded: false,
  expandedRows: expandedRows,
  toggleExpanded: _noop.default,
  isSearch: false,
  isCompact: false
};
var EditorThemeContext = exports.EditorThemeContext = /*#__PURE__*/(0, _react.createContext)(defaultEditorTheme);

/**
 * Hook to manage the theming state of the editor.
 */
var useEditorTheme = exports.useEditorTheme = function useEditorTheme(_ref) {
  var _ref$isSearch = _ref.isSearch,
    isSearch = _ref$isSearch === void 0 ? false : _ref$isSearch,
    _ref$isCompact = _ref.isCompact,
    isCompact = _ref$isCompact === void 0 ? false : _ref$isCompact;
  var _useState = (0, _react.useState)(false),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    expanded = _useState2[0],
    setExpanded = _useState2[1];
  var toggleExpanded = (0, _react.useCallback)(function () {
    return setExpanded(function (prevState) {
      return !prevState;
    });
  }, []);
  return (0, _react.useMemo)(function () {
    return {
      defaultMaxRows: defaultMaxRows,
      expanded: expanded,
      expandedRows: expandedRows,
      toggleExpanded: toggleExpanded,
      isSearch: isSearch,
      isCompact: isCompact
    };
  }, [expanded, toggleExpanded, isSearch, isCompact]);
};
var useEditorThemeContext = exports.useEditorThemeContext = function useEditorThemeContext() {
  return (0, _react.useContext)(EditorThemeContext);
};
