// @atlaskit/side-navigation/dist/esm/components/Header/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["children", "data-testid"],
  _excluded2 = ["children"];
import "./index.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
import React, { forwardRef } from 'react';
import { N500 } from '@atlaskit/theme/colors';
import { CustomItem } from '../Item';
var styles = {
  header: "_vrjt140y _1cmd10s3 _1cg71wug _11kjqtfy _jl2n73ad _1et6qtfy _1v6j10s3"
};

/**
 * __Container__
 *
 * A container for Header and Footer that safely handles props to the child component
 */
export var Container = function Container(_ref) {
  var children = _ref.children,
    testId = _ref['data-testid'],
    props = _objectWithoutProperties(_ref, _excluded);
  // https://stackoverflow.com/a/39333479
  var safeProps = function (_ref2) {
    var className = _ref2.className,
      onClick = _ref2.onClick,
      onMouseDown = _ref2.onMouseDown,
      onDragStart = _ref2.onDragStart,
      draggable = _ref2.draggable,
      ref = _ref2.ref,
      tabIndex = _ref2.tabIndex,
      disabled = _ref2.disabled;
    return {
      className: className,
      onClick: onClick,
      onMouseDown: onMouseDown,
      onDragStart: onDragStart,
      draggable: draggable,
      ref: ref,
      tabIndex: tabIndex,
      disabled: disabled
    };
  }(props);
  return (
    /*#__PURE__*/
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
    React.createElement("div", _extends({
      "data-testid": testId,
      style: {
        position: 'relative'
      }
    }, safeProps), children)
  );
};
/**
 * __Header__
 *
 * - [Examples](https://atlassian.design/components/side-navigation/examples#header-and-footer)
 * - [Code](https://atlassian.design/components/side-navigation/code)
 */
var Header = /*#__PURE__*/forwardRef(function (props, ref) {
  return /*#__PURE__*/React.createElement(CustomItem, _extends({}, props, {
    ref: ref,
    component: props.component || Container,
    // eslint-disable-next-line @repo/internal/react/no-unsafe-overrides
    overrides: {
      Title: {
        // eslint-disable-next-line @atlaskit/design-system/use-heading
        render: function render(_, _ref3) {
          var children = _ref3.children,
            props = _objectWithoutProperties(_ref3, _excluded2);
          return /*#__PURE__*/React.createElement("h2", props, children);
        }
      }
    },
    className: ax([styles.header])
  }));
});
export default Header;
