// @atlaskit/media-viewer/dist/cjs/viewers/codeViewer/styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.codeViewerHeaderBarStyles = exports.codeViewerHTMLStyles = exports.codeViewWrapperStyles = void 0;
var _react = require("@emotion/react");
var _theme = require("@atlaskit/theme");
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

// AFP-2532 TODO: Fix automatic suppressions below
// eslint-disable-next-line @atlassian/tangerine/import/entry-points

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var codeViewWrapperStyles = exports.codeViewWrapperStyles = (0, _react.css)({
  position: 'absolute',
  left: 0,
  top: 0,
  right: 0,
  bottom: 0,
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
  backgroundColor: "var(--ds-surface, ".concat(_theme.colors.N20, ")"),
  overflow: 'auto'
});

/* eslint-disable @atlaskit/design-system/ensure-design-token-usage */
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var codeViewerHeaderBarStyles = exports.codeViewerHeaderBarStyles = (0, _react.css)({
  height: '75px',
  backgroundColor: '#1d2125'
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var codeViewerHTMLStyles = exports.codeViewerHTMLStyles = (0, _react.css)({
  display: 'flex',
  overflowX: 'auto',
  whiteSpace: 'pre',
  font: "var(--ds-font-body-UNSAFE_small, normal 400 12px/16px ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  padding: "var(--ds-space-100, 8px)"
});
