// @atlaskit/editor-plugin-type-ahead/dist/cjs/ui/TypeAheadErrorFallback/EmptyState.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EmptyState = void 0;
var _react = _interopRequireDefault(require("react"));
var _heading = _interopRequireDefault(require("@atlaskit/heading"));
var _primitives = require("@atlaskit/primitives");
var containerStyles = (0, _primitives.xcss)({
  marginBlockStart: 'space.600',
  marginBlockEnd: 'space.600',
  textAlign: 'center'
});
var EmptyState = exports.EmptyState = function EmptyState(_ref) {
  var testId = _ref.testId,
    header = _ref.header,
    description = _ref.description,
    renderImage = _ref.renderImage;
  return /*#__PURE__*/_react.default.createElement(_primitives.Flex, {
    xcss: containerStyles,
    testId: testId,
    direction: "column",
    alignItems: "center",
    gap: "space.300"
  }, renderImage === null || renderImage === void 0 ? void 0 : renderImage(), /*#__PURE__*/_react.default.createElement(_primitives.Flex, {
    direction: "column",
    alignItems: "center",
    gap: "space.200"
  }, /*#__PURE__*/_react.default.createElement(_heading.default, {
    size: "medium",
    as: "h2"
  }, header), description && /*#__PURE__*/_react.default.createElement(_primitives.Text, {
    as: "p",
    color: "color.text"
  }, description)));
};
