// @atlaskit/editor-plugin-type-ahead/dist/cjs/ui/TypeAheadErrorFallback/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TypeAheadErrorFallback = void 0;
var _react = require("@emotion/react");
var _reactIntlNext = require("react-intl-next");
var _typeAhead = require("@atlaskit/editor-common/type-ahead");
var _EmptyState = require("./EmptyState");
var _GenericErrorSVG = require("./GenericErrorSVG");
/**
 * @jsxRuntime classic
 * @jsx jsx
 */
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var minHeightComponentStyles = (0, _react.css)({
  display: 'flex',
  alignItems: 'stretch',
  justifyContent: 'center',
  minHeight: '290px'
});
var TypeAheadErrorFallback = exports.TypeAheadErrorFallback = function TypeAheadErrorFallback() {
  var intl = (0, _reactIntlNext.useIntl)();
  var header = intl.formatMessage(_typeAhead.typeAheadListMessages.offlineErrorFallbackHeading);
  var description = intl.formatMessage(_typeAhead.typeAheadListMessages.offlineErrorFallbackDesc);
  return (0, _react.jsx)("div", {
    "data-testid": "mentions-typeahead-error-boundary-ui",
    css: minHeightComponentStyles
  }, (0, _react.jsx)(_EmptyState.EmptyState, {
    header: header,
    renderImage: function renderImage() {
      return (0, _react.jsx)(_GenericErrorSVG.GenericErrorSVG, null);
    },
    description: description
  }));
};
