// @atlaskit/editor-plugin-block-type/dist/cjs/pm-plugins/block-types.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getBlockTypesInDropdown = exports.WRAPPER_BLOCK_TYPES = exports.TEXT_BLOCK_TYPES = exports.PANEL = exports.OTHER = exports.NORMAL_TEXT = exports.HEADING_6 = exports.HEADING_5 = exports.HEADING_4 = exports.HEADING_3 = exports.HEADING_2 = exports.HEADING_1 = exports.HEADINGS_BY_NAME = exports.HEADINGS_BY_LEVEL = exports.FORMATTING_NODE_TYPES = exports.FORMATTING_MARK_TYPES = exports.CODE_BLOCK = exports.BLOCK_QUOTE = exports.ALL_BLOCK_TYPES = void 0;
var _react = _interopRequireDefault(require("react"));
var _messages = require("@atlaskit/editor-common/messages");
var _textHeadingFive = _interopRequireDefault(require("@atlaskit/icon-lab/core/text-heading-five"));
var _textHeadingFour = _interopRequireDefault(require("@atlaskit/icon-lab/core/text-heading-four"));
var _textHeadingOne = _interopRequireDefault(require("@atlaskit/icon-lab/core/text-heading-one"));
var _textHeadingSix = _interopRequireDefault(require("@atlaskit/icon-lab/core/text-heading-six"));
var _textHeadingThree = _interopRequireDefault(require("@atlaskit/icon-lab/core/text-heading-three"));
var _textHeadingTwo = _interopRequireDefault(require("@atlaskit/icon-lab/core/text-heading-two"));
var _text = _interopRequireDefault(require("@atlaskit/icon/core/text"));
var _experiments = require("@atlaskit/tmp-editor-statsig/experiments");
var _icons = require("./ui/ToolbarBlockType/icons");
var NORMAL_TEXT = exports.NORMAL_TEXT = {
  name: 'normal',
  title: _messages.blockTypeMessages.normal,
  nodeName: 'paragraph',
  tagName: 'p',
  icon: /*#__PURE__*/_react.default.createElement(_text.default, {
    label: ""
  }),
  LEGACY_icon: /*#__PURE__*/_react.default.createElement(_icons.Text, null)
};
var HEADING_1 = exports.HEADING_1 = {
  name: 'heading1',
  title: _messages.blockTypeMessages.heading1,
  nodeName: 'heading',
  tagName: 'h1',
  level: 1,
  icon: /*#__PURE__*/_react.default.createElement(_textHeadingOne.default, {
    label: ""
  }),
  LEGACY_icon: /*#__PURE__*/_react.default.createElement(_icons.H1, null)
};
var HEADING_2 = exports.HEADING_2 = {
  name: 'heading2',
  title: _messages.blockTypeMessages.heading2,
  nodeName: 'heading',
  tagName: 'h2',
  level: 2,
  icon: /*#__PURE__*/_react.default.createElement(_textHeadingTwo.default, {
    label: ""
  }),
  LEGACY_icon: /*#__PURE__*/_react.default.createElement(_icons.H2, null)
};
var HEADING_3 = exports.HEADING_3 = {
  name: 'heading3',
  title: _messages.blockTypeMessages.heading3,
  nodeName: 'heading',
  tagName: 'h3',
  level: 3,
  icon: /*#__PURE__*/_react.default.createElement(_textHeadingThree.default, {
    label: ""
  }),
  LEGACY_icon: /*#__PURE__*/_react.default.createElement(_icons.H3, null)
};
var HEADING_4 = exports.HEADING_4 = {
  name: 'heading4',
  title: _messages.blockTypeMessages.heading4,
  nodeName: 'heading',
  tagName: 'h4',
  level: 4,
  icon: /*#__PURE__*/_react.default.createElement(_textHeadingFour.default, {
    label: ""
  }),
  LEGACY_icon: /*#__PURE__*/_react.default.createElement(_icons.H4, null)
};
var HEADING_5 = exports.HEADING_5 = {
  name: 'heading5',
  title: _messages.blockTypeMessages.heading5,
  nodeName: 'heading',
  tagName: 'h5',
  level: 5,
  icon: /*#__PURE__*/_react.default.createElement(_textHeadingFive.default, {
    label: ""
  }),
  LEGACY_icon: /*#__PURE__*/_react.default.createElement(_icons.H5, null)
};
var HEADING_6 = exports.HEADING_6 = {
  name: 'heading6',
  title: _messages.blockTypeMessages.heading6,
  nodeName: 'heading',
  tagName: 'h6',
  level: 6,
  icon: /*#__PURE__*/_react.default.createElement(_textHeadingSix.default, {
    label: ""
  }),
  LEGACY_icon: /*#__PURE__*/_react.default.createElement(_icons.H6, null)
};
var BLOCK_QUOTE = exports.BLOCK_QUOTE = {
  name: 'blockquote',
  title: _messages.blockTypeMessages.blockquote,
  nodeName: 'blockquote',
  tagName: 'blockquote'
};
var CODE_BLOCK = exports.CODE_BLOCK = {
  name: 'codeblock',
  title: _messages.blockTypeMessages.codeblock,
  nodeName: 'codeBlock'
};
var PANEL = exports.PANEL = {
  name: 'panel',
  title: _messages.blockTypeMessages.infoPanel,
  nodeName: 'panel'
};
var OTHER = exports.OTHER = {
  name: 'other',
  title: _messages.blockTypeMessages.other,
  nodeName: ''
};
var TEXT_BLOCK_TYPES = exports.TEXT_BLOCK_TYPES = [NORMAL_TEXT, HEADING_1, HEADING_2, HEADING_3, HEADING_4, HEADING_5, HEADING_6];
var FORMATTING_NODE_TYPES = exports.FORMATTING_NODE_TYPES = ['heading', 'blockquote'];
var FORMATTING_MARK_TYPES = exports.FORMATTING_MARK_TYPES = ['em', 'code', 'strike', 'strong', 'underline', 'textColor', 'subsup', 'backgroundColor'];
var WRAPPER_BLOCK_TYPES = exports.WRAPPER_BLOCK_TYPES = [BLOCK_QUOTE, CODE_BLOCK, PANEL];
var ALL_BLOCK_TYPES = exports.ALL_BLOCK_TYPES = TEXT_BLOCK_TYPES.concat(WRAPPER_BLOCK_TYPES);
var getBlockTypesInDropdown = exports.getBlockTypesInDropdown = function getBlockTypesInDropdown(includeBlockQuoteAsTextstyleOption) {
  return (0, _experiments.editorExperiment)('platform_editor_blockquote_in_text_formatting_menu', true, {
    exposure: true
  }) && includeBlockQuoteAsTextstyleOption ? [].concat(TEXT_BLOCK_TYPES, [BLOCK_QUOTE]) : TEXT_BLOCK_TYPES;
};
var HEADINGS_BY_LEVEL = exports.HEADINGS_BY_LEVEL = TEXT_BLOCK_TYPES.reduce(function (acc, blockType) {
  if (blockType.level && blockType.nodeName === 'heading') {
    acc[blockType.level] = blockType;
  }
  return acc;
}, {});
var HEADINGS_BY_NAME = exports.HEADINGS_BY_NAME = TEXT_BLOCK_TYPES.reduce(function (acc, blockType) {
  if (blockType.level && blockType.nodeName === 'heading') {
    acc[blockType.name] = blockType;
  }
  return acc;
}, {});
