// @atlaskit/editor-tables/dist/cjs/utils/analytics-helpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getSelectedCellInfo = getSelectedCellInfo;
exports.getSelectedTableInfo = getSelectedTableInfo;
var _tableMap = require("../table-map");
var _find = require("./find");
var _getSelectionRect = require("./get-selection-rect");
function getSelectedTableInfo(selection) {
  var map;
  var totalRowCount = 0;
  var totalColumnCount = 0;
  var table = (0, _find.findTable)(selection);
  if (table) {
    map = _tableMap.TableMap.get(table.node);
    totalRowCount = map.height;
    totalColumnCount = map.width;
  }
  return {
    table: table,
    map: map,
    totalRowCount: totalRowCount,
    totalColumnCount: totalColumnCount
  };
}
function getSelectedCellInfo(selection) {
  var horizontalCells = 1;
  var verticalCells = 1;
  var totalCells = 1;
  var _getSelectedTableInfo = getSelectedTableInfo(selection),
    table = _getSelectedTableInfo.table,
    map = _getSelectedTableInfo.map,
    totalRowCount = _getSelectedTableInfo.totalRowCount,
    totalColumnCount = _getSelectedTableInfo.totalColumnCount;
  if (table && map) {
    var rect = (0, _getSelectionRect.getSelectionRect)(selection);
    if (rect) {
      totalCells = map.cellsInRect(rect).length;
      horizontalCells = rect.right - rect.left;
      verticalCells = rect.bottom - rect.top;
    }
  }
  return {
    totalRowCount: totalRowCount,
    totalColumnCount: totalColumnCount,
    horizontalCells: horizontalCells,
    verticalCells: verticalCells,
    totalCells: totalCells
  };
}
