// @atlaskit/editor-tables/dist/cjs/utils/is-selected.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isTableSelected = exports.isRowSelected = exports.isRectSelected = exports.isColumnSelected = void 0;
var _tableMap = require("../table-map");
var _isSelectionType = require("./is-selection-type");
// Checks if a given CellSelection rect is selected
var isRectSelected = exports.isRectSelected = function isRectSelected(rect) {
  return function (selection) {
    if (!(0, _isSelectionType.isSelectionType)(selection, 'cell')) {
      return false;
    }
    var map = _tableMap.TableMap.get(selection.$anchorCell.node(-1));
    var start = selection.$anchorCell.start(-1);
    var cells = map.cellsInRect(rect);
    var selectedCells = map.cellsInRect(map.rectBetween(selection.$anchorCell.pos - start, selection.$headCell.pos - start));
    for (var i = 0, count = cells.length; i < count; i++) {
      if (selectedCells.indexOf(cells[i]) === -1) {
        return false;
      }
    }
    return true;
  };
};

// Checks if entire column at index `columnIndex` is selected.
var isColumnSelected = exports.isColumnSelected = function isColumnSelected(columnIndex) {
  return function (selection) {
    if ((0, _isSelectionType.isSelectionType)(selection, 'cell')) {
      var map = _tableMap.TableMap.get(selection.$anchorCell.node(-1));
      return isRectSelected({
        left: columnIndex,
        right: columnIndex + 1,
        top: 0,
        bottom: map.height
      })(selection);
    }
    return false;
  };
};

// Checks if entire row at index `rowIndex` is selected.
var isRowSelected = exports.isRowSelected = function isRowSelected(rowIndex) {
  return function (selection) {
    if ((0, _isSelectionType.isSelectionType)(selection, 'cell')) {
      var map = _tableMap.TableMap.get(selection.$anchorCell.node(-1));
      return isRectSelected({
        left: 0,
        right: map.width,
        top: rowIndex,
        bottom: rowIndex + 1
      })(selection);
    }
    return false;
  };
};

// Checks if entire table is selected
var isTableSelected = exports.isTableSelected = function isTableSelected(selection) {
  if ((0, _isSelectionType.isSelectionType)(selection, 'cell')) {
    var map = _tableMap.TableMap.get(selection.$anchorCell.node(-1));
    return isRectSelected({
      left: 0,
      right: map.width,
      top: 0,
      bottom: map.height
    })(selection);
  }
  return false;
};
