// @atlaskit/link-datasource/dist/esm/ui/assets-modal/modal/render-assets-content/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import "./index.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { useCallback, useMemo } from 'react';
import { N40 } from '@atlaskit/theme/colors';
import { AccessRequired } from '../../../common/error-state/access-required';
import { ModalLoadingError } from '../../../common/error-state/modal-loading-error';
import { NoResults } from '../../../common/error-state/no-results';
import { EmptyState, IssueLikeDataTableView } from '../../../issue-like-table';
import { InitialStateView } from './initial-state-view';
export var MODAL_HEIGHT = 420;

// This is to prevent y scrollbar when showing table loading state
var disableOverflowStyles = null;
var contentContainerStyles = null;
var tableBordersStyles = null;
var RejectedView = function RejectedView() {
  return /*#__PURE__*/React.createElement("div", {
    className: ax(["_1reo1wug _18m91wug _4t3i1pna _1e0c11p5"])
  }, /*#__PURE__*/React.createElement(ModalLoadingError, null));
};
var UnauthorizedView = function UnauthorizedView() {
  return /*#__PURE__*/React.createElement("div", {
    className: ax(["_1reo1wug _18m91wug _4t3i1pna _1e0c11p5"])
  }, /*#__PURE__*/React.createElement(AccessRequired, null));
};
var EmptyView = function EmptyView() {
  return /*#__PURE__*/React.createElement("div", {
    className: ax(["_1reo1wug _18m91wug _4t3i1pna _1e0c11p5"])
  }, /*#__PURE__*/React.createElement(InitialStateView, null));
};
var NoResultsView = function NoResultsView() {
  return /*#__PURE__*/React.createElement("div", {
    className: ax(["_1reo1wug _18m91wug _4t3i1pna _1e0c11p5"])
  }, /*#__PURE__*/React.createElement(NoResults, null));
};
var LoadingView = function LoadingView() {
  return /*#__PURE__*/React.createElement("div", {
    className: ax(["_1reo1wug _18m91wug _4t3i1pna _1e0c11p5", "_1reo1wug _18m91wug _4t3i1pna _1e0c11p5", "_19itn8v0 _n7zl1oe8 _13lipd34 _qrwqpd34 _1itk1bg1 _12vemgnk _1lrw1sd4 _qmxj1c00", "_1reo15vq _18m915vq"]),
    style: {
      "--_vzvmm2": ix("1px solid ".concat("var(--ds-border, ".concat(N40, ")"))),
      "--_cfwo9c": ix("2px solid ".concat("var(--ds-background-accent-gray-subtler, ".concat(N40, ")"))),
      "--_17x9dh1": ix("\n\t\tlinear-gradient(90deg, ".concat("var(--ds-elevation-surface-current, #FFF)", " 30%, rgba(255, 255, 255, 0)),\n\t\tlinear-gradient(90deg, ", "var(--ds-shadow-overflow-perimeter, rgba(0, 0, 0, 0.1))", ", rgba(0, 0, 0, 0)),\n\t\tlinear-gradient(90deg, rgba(255, 255, 255, 0), ", "var(--ds-elevation-surface-current, #FFF)", " 70%),\n\t\tlinear-gradient(90deg, rgba(0, 0, 0, 0), ", "var(--ds-shadow-overflow-perimeter, rgba(0, 0, 0, 0.1))", "])\n\t\tlinear-gradient(0deg, rgba(255, 255, 255, 0),  ", "var(--ds-elevation-surface-current, #FFF)", " 30%),\n\t\tlinear-gradient(0deg, rgba(0, 0, 0, 0), ", "var(--ds-shadow-overflow-perimeter, rgba(0, 0, 0, 0.05))", "),\n\t\tlinear-gradient(0deg, ", "var(--ds-elevation-surface-current, #FFF)", " 30%, rgba(255, 255, 255, 0)),\n\t\tlinear-gradient(0deg, ", "var(--ds-shadow-overflow-perimeter, rgba(0, 0, 0, 0.05))", ", rgba(0, 0, 0, 0))\n\t\t"))
    }
  }, /*#__PURE__*/React.createElement(EmptyState, {
    testId: "assets-aql-datasource-modal--loading-state"
  }));
};
export var RenderAssetsContent = function RenderAssetsContent(props) {
  var status = props.status,
    responseItems = props.responseItems,
    responseItemIds = props.responseItemIds,
    visibleColumnKeys = props.visibleColumnKeys,
    onNextPage = props.onNextPage,
    hasNextPage = props.hasNextPage,
    loadDatasourceDetails = props.loadDatasourceDetails,
    columns = props.columns,
    defaultVisibleColumnKeys = props.defaultVisibleColumnKeys,
    onVisibleColumnKeysChange = props.onVisibleColumnKeysChange,
    isFetchingInitialData = props.isFetchingInitialData;
  var resolvedWithNoResults = status === 'resolved' && !responseItems.length;
  var issueLikeDataTableView = useMemo(function () {
    return /*#__PURE__*/React.createElement("div", {
      className: ax(["_1reo1wug _18m91wug _4t3i1pna _1e0c11p5", "_19itn8v0 _n7zl1oe8 _13lipd34 _qrwqpd34 _1itk1bg1 _12vemgnk _1lrw1sd4 _qmxj1c00"]),
      style: {
        "--_vzvmm2": ix("1px solid ".concat("var(--ds-border, ".concat(N40, ")"))),
        "--_cfwo9c": ix("2px solid ".concat("var(--ds-background-accent-gray-subtler, ".concat(N40, ")"))),
        "--_17x9dh1": ix("\n\t\tlinear-gradient(90deg, ".concat("var(--ds-elevation-surface-current, #FFF)", " 30%, rgba(255, 255, 255, 0)),\n\t\tlinear-gradient(90deg, ", "var(--ds-shadow-overflow-perimeter, rgba(0, 0, 0, 0.1))", ", rgba(0, 0, 0, 0)),\n\t\tlinear-gradient(90deg, rgba(255, 255, 255, 0), ", "var(--ds-elevation-surface-current, #FFF)", " 70%),\n\t\tlinear-gradient(90deg, rgba(0, 0, 0, 0), ", "var(--ds-shadow-overflow-perimeter, rgba(0, 0, 0, 0.1))", "])\n\t\tlinear-gradient(0deg, rgba(255, 255, 255, 0),  ", "var(--ds-elevation-surface-current, #FFF)", " 30%),\n\t\tlinear-gradient(0deg, rgba(0, 0, 0, 0), ", "var(--ds-shadow-overflow-perimeter, rgba(0, 0, 0, 0.05))", "),\n\t\tlinear-gradient(0deg, ", "var(--ds-elevation-surface-current, #FFF)", " 30%, rgba(255, 255, 255, 0)),\n\t\tlinear-gradient(0deg, ", "var(--ds-shadow-overflow-perimeter, rgba(0, 0, 0, 0.05))", ", rgba(0, 0, 0, 0))\n\t\t"))
      }
    }, /*#__PURE__*/React.createElement(IssueLikeDataTableView, {
      testId: "asset-datasource-table",
      status: status,
      columns: columns,
      items: responseItems,
      itemIds: responseItemIds,
      hasNextPage: hasNextPage,
      visibleColumnKeys: visibleColumnKeys || defaultVisibleColumnKeys,
      onNextPage: onNextPage,
      onLoadDatasourceDetails: loadDatasourceDetails,
      onVisibleColumnKeysChange: onVisibleColumnKeysChange
    }));
  }, [columns, defaultVisibleColumnKeys, hasNextPage, loadDatasourceDetails, onNextPage, onVisibleColumnKeysChange, responseItems, responseItemIds, status, visibleColumnKeys]);
  var renderAssetsContentView = useCallback(function () {
    if (isFetchingInitialData) {
      // Placing this check first as it's a priority before all others
      return /*#__PURE__*/React.createElement(LoadingView, null);
    } else if (status === 'rejected') {
      return /*#__PURE__*/React.createElement(RejectedView, null);
    } else if (status === 'unauthorized') {
      return /*#__PURE__*/React.createElement(UnauthorizedView, null);
    } else if (status === 'empty') {
      return /*#__PURE__*/React.createElement(EmptyView, null);
    } else if (resolvedWithNoResults) {
      return /*#__PURE__*/React.createElement(NoResultsView, null);
    } else if (status === 'loading' && !columns.length) {
      return /*#__PURE__*/React.createElement(LoadingView, null);
    }
    return issueLikeDataTableView;
  }, [columns.length, isFetchingInitialData, issueLikeDataTableView, resolvedWithNoResults, status]);
  return renderAssetsContentView();
};
