// @atlaskit/editor-common/dist/cjs/analytics/types/general-events.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PLATFORMS = exports.MODE = exports.FULL_WIDTH_MODE = exports.BROWSER_FREEZE_INTERACTION_TYPE = void 0;
var PLATFORMS = exports.PLATFORMS = /*#__PURE__*/function (PLATFORMS) {
  PLATFORMS["NATIVE"] = "mobileNative";
  PLATFORMS["HYBRID"] = "mobileHybrid";
  PLATFORMS["WEB"] = "web";
  return PLATFORMS;
}({});
var MODE = exports.MODE = /*#__PURE__*/function (MODE) {
  MODE["RENDERER"] = "renderer";
  MODE["EDITOR"] = "editor";
  return MODE;
}({});
var FULL_WIDTH_MODE = exports.FULL_WIDTH_MODE = /*#__PURE__*/function (FULL_WIDTH_MODE) {
  FULL_WIDTH_MODE["FIXED_WIDTH"] = "fixedWidth";
  FULL_WIDTH_MODE["FULL_WIDTH"] = "fullWidth";
  return FULL_WIDTH_MODE;
}({});
var BROWSER_FREEZE_INTERACTION_TYPE = exports.BROWSER_FREEZE_INTERACTION_TYPE = /*#__PURE__*/function (BROWSER_FREEZE_INTERACTION_TYPE) {
  BROWSER_FREEZE_INTERACTION_TYPE["LOADING"] = "loading";
  BROWSER_FREEZE_INTERACTION_TYPE["TYPING"] = "typing";
  BROWSER_FREEZE_INTERACTION_TYPE["CLICKING"] = "clicking";
  BROWSER_FREEZE_INTERACTION_TYPE["PASTING"] = "pasting";
  return BROWSER_FREEZE_INTERACTION_TYPE;
}({});
