// @atlaskit/editor-plugin-engagement-platform/dist/cjs/engagementPlatformPlugin.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.engagementPlatformPlugin = void 0;
var _startMessage = require("./editor-actions/startMessage");
var _stopMessage = require("./editor-actions/stopMessage");
var _engagementPlatformPmPlugin = require("./pm-plugins/engagementPlatformPmPlugin/engagementPlatformPmPlugin");
var _engagementPlatformPmPluginKey = require("./pm-plugins/engagementPlatformPmPlugin/engagementPlatformPmPluginKey");
var engagementPlatformPlugin = exports.engagementPlatformPlugin = function engagementPlatformPlugin(_ref) {
  var config = _ref.config,
    api = _ref.api;
  return {
    name: 'engagementPlatform',
    actions: {
      startMessage: (0, _startMessage.startMessage)(api, config.coordinationClient),
      stopMessage: (0, _stopMessage.stopMessage)(api, config.coordinationClient)
    },
    pmPlugins: function pmPlugins() {
      return [{
        name: 'engagementPlatformPmPlugin',
        plugin: function plugin() {
          return (0, _engagementPlatformPmPlugin.engagementPlatformPmPlugin)(config);
        }
      }];
    },
    getSharedState: function getSharedState(editorState) {
      if (!editorState) {
        return undefined;
      }
      return _engagementPlatformPmPluginKey.engagementPlatformPmPluginKey.getState(editorState);
    }
  };
};
