// @atlaskit/media-ui/dist/cjs/BlockCard/actions/DownloadAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var react_intl_1 = require("react-intl");
var media_common_1 = require("@atlaskit/media-common");
var messages_1 = require("../../messages");
function downloadFunction(_a) {
    var url = _a.url;
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_b) {
            if (!url) {
                return [2 /*return*/];
            }
            media_common_1.downloadUrl(url);
            return [2 /*return*/];
        });
    });
}
exports.downloadFunction = downloadFunction;
exports.DownloadAction = function (_a) {
    var url = _a.url;
    return ({
        id: 'download-content',
        text: react_1.default.createElement(react_intl_1.FormattedMessage, tslib_1.__assign({}, messages_1.messages.download)),
        promise: function () {
            return downloadFunction({
                url: url,
            });
        },
    });
};
