// @atlaskit/linking-types/dist/cjs/smart-link-action-types.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SmartLinkActionType = exports.InvokeError = void 0;
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _wrapNativeSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/wrapNativeSuper"));
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
/**
 * The enum that describes Smart Link action types
 */
var SmartLinkActionType = exports.SmartLinkActionType = /*#__PURE__*/function (SmartLinkActionType) {
  SmartLinkActionType["FollowEntityAction"] = "FollowEntityAction";
  SmartLinkActionType["GetStatusTransitionsAction"] = "GetStatusTransitionsAction";
  SmartLinkActionType["StatusUpdateAction"] = "StatusUpdateAction";
  SmartLinkActionType["UnfollowEntityAction"] = "UnfollowEntityAction";
  return SmartLinkActionType;
}({});
/**
 * Default type for an invoke request. Parametrized with the type of payload
 */
/**
 * Captures information about an action
 */
/**
 * payload needed by {@link SmartLinkActionType.StatusUpdateAction} action type
 */
/**
 * Default type for an Action Response
 */
/**
 * Type that signifies an error response from the product
 */
/**
 * Payload from GetStatusTransitions action that contains a list of available statuses
 */
var InvokeError = exports.InvokeError = /*#__PURE__*/function (_Error) {
  function InvokeError(message, errorCode) {
    var _this;
    (0, _classCallCheck2.default)(this, InvokeError);
    _this = _callSuper(this, InvokeError, [message]);
    _this.errorCode = errorCode;
    return _this;
  }
  (0, _inherits2.default)(InvokeError, _Error);
  return (0, _createClass2.default)(InvokeError);
}( /*#__PURE__*/(0, _wrapNativeSuper2.default)(Error));
