// @atlaskit/editor-wikimarkup-transformer/dist/cjs/parser/tokenize/media.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.media = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _mediaSingle = _interopRequireDefault(require("../nodes/mediaSingle"));
var _attrs = require("../utils/attrs");
var _commonFormatter = require("./common-formatter");
var media = exports.media = function media(_ref) {
  var input = _ref.input,
    position = _ref.position,
    schema = _ref.schema,
    context = _ref.context;
  var rawContentProcessor = function rawContentProcessor(raw, length) {
    /**
     * !image.gif|align=right, vspace=4|ignore-this!
     * If it splits into more than 2 items, we ignore the rest
     */
    var _raw$split = raw.split('|'),
      _raw$split2 = (0, _slicedToArray2.default)(_raw$split, 2),
      rawContent = _raw$split2[0],
      _raw$split2$ = _raw$split2[1],
      rawAttrs = _raw$split2$ === void 0 ? '' : _raw$split2$;
    var node = (0, _mediaSingle.default)(schema, rawContent, (0, _attrs.parseAttrs)(rawAttrs, ','), context);
    return {
      type: 'pmnode',
      nodes: [node],
      length: length
    };
  };
  return (0, _commonFormatter.commonFormatter)(input, position, schema, {
    opening: '!',
    closing: '!',
    context: context,
    rawContentProcessor: rawContentProcessor
  });
};
