// @atlaskit/editor-plugin-type-ahead/dist/cjs/pm-plugins/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TYPE_AHEAD_POPUP_CONTENT_CLASS = exports.TYPE_AHEAD_DECORATION_KEY = exports.TYPE_AHEAD_DECORATION_ELEMENT_ID = exports.TYPE_AHEAD_DECORATION_DATA_ATTRIBUTE = exports.CloseSelectionOptions = void 0;
var TYPE_AHEAD_DECORATION_KEY = exports.TYPE_AHEAD_DECORATION_KEY = 'typeahead_decoration_key';
var TYPE_AHEAD_DECORATION_DATA_ATTRIBUTE = exports.TYPE_AHEAD_DECORATION_DATA_ATTRIBUTE = 'typeaheadDecoration';
var TYPE_AHEAD_POPUP_CONTENT_CLASS = exports.TYPE_AHEAD_POPUP_CONTENT_CLASS = 'fabric-editor-typeahead';
var TYPE_AHEAD_DECORATION_ELEMENT_ID = exports.TYPE_AHEAD_DECORATION_ELEMENT_ID = 'typeahaed_decoration_element_id';
var CloseSelectionOptions = exports.CloseSelectionOptions = /*#__PURE__*/function (CloseSelectionOptions) {
  CloseSelectionOptions["BEFORE_TEXT_INSERTED"] = "BEFORE_TEXT_INSERTED";
  CloseSelectionOptions["AFTER_TEXT_INSERTED"] = "AFTER_TEXT_INSERTED";
  return CloseSelectionOptions;
}({});
