// @atlaskit/logo/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AtlasIcon", {
  enumerable: true,
  get: function get() {
    return _atlasLogo.AtlasIcon;
  }
});
Object.defineProperty(exports, "AtlasLogo", {
  enumerable: true,
  get: function get() {
    return _atlasLogo.AtlasLogo;
  }
});
Object.defineProperty(exports, "AtlassianAccessIcon", {
  enumerable: true,
  get: function get() {
    return _atlassianAccessLogo.AtlassianAccessIcon;
  }
});
Object.defineProperty(exports, "AtlassianAccessLogo", {
  enumerable: true,
  get: function get() {
    return _atlassianAccessLogo.AtlassianAccessLogo;
  }
});
Object.defineProperty(exports, "AtlassianAdminIcon", {
  enumerable: true,
  get: function get() {
    return _atlassianAdminLogo.AtlassianAdminIcon;
  }
});
Object.defineProperty(exports, "AtlassianAdminLogo", {
  enumerable: true,
  get: function get() {
    return _atlassianAdminLogo.AtlassianAdminLogo;
  }
});
Object.defineProperty(exports, "AtlassianAdministrationIcon", {
  enumerable: true,
  get: function get() {
    return _atlassianAdministrationLogo.AtlassianAdministrationIcon;
  }
});
Object.defineProperty(exports, "AtlassianAdministrationLogo", {
  enumerable: true,
  get: function get() {
    return _atlassianAdministrationLogo.AtlassianAdministrationLogo;
  }
});
Object.defineProperty(exports, "AtlassianAnalyticsIcon", {
  enumerable: true,
  get: function get() {
    return _atlassianAnalyticsLogo.AtlassianAnalyticsIcon;
  }
});
Object.defineProperty(exports, "AtlassianAnalyticsLogo", {
  enumerable: true,
  get: function get() {
    return _atlassianAnalyticsLogo.AtlassianAnalyticsLogo;
  }
});
Object.defineProperty(exports, "AtlassianIcon", {
  enumerable: true,
  get: function get() {
    return _atlassianLogo.AtlassianIcon;
  }
});
Object.defineProperty(exports, "AtlassianLogo", {
  enumerable: true,
  get: function get() {
    return _atlassianLogo.AtlassianLogo;
  }
});
Object.defineProperty(exports, "AtlassianMarketplaceIcon", {
  enumerable: true,
  get: function get() {
    return _atlassianMarketplaceLogo.AtlassianMarketplaceIcon;
  }
});
Object.defineProperty(exports, "AtlassianMarketplaceLogo", {
  enumerable: true,
  get: function get() {
    return _atlassianMarketplaceLogo.AtlassianMarketplaceLogo;
  }
});
Object.defineProperty(exports, "AtlassianStartLogo", {
  enumerable: true,
  get: function get() {
    return _atlassianStartLogo.AtlassianStartLogo;
  }
});
Object.defineProperty(exports, "BitbucketIcon", {
  enumerable: true,
  get: function get() {
    return _bitbucketLogo.BitbucketIcon;
  }
});
Object.defineProperty(exports, "BitbucketLogo", {
  enumerable: true,
  get: function get() {
    return _bitbucketLogo.BitbucketLogo;
  }
});
Object.defineProperty(exports, "CompassIcon", {
  enumerable: true,
  get: function get() {
    return _compassLogo.CompassIcon;
  }
});
Object.defineProperty(exports, "CompassLogo", {
  enumerable: true,
  get: function get() {
    return _compassLogo.CompassLogo;
  }
});
Object.defineProperty(exports, "ConfluenceIcon", {
  enumerable: true,
  get: function get() {
    return _confluenceLogo.ConfluenceIcon;
  }
});
Object.defineProperty(exports, "ConfluenceLogo", {
  enumerable: true,
  get: function get() {
    return _confluenceLogo.ConfluenceLogo;
  }
});
Object.defineProperty(exports, "FocusIcon", {
  enumerable: true,
  get: function get() {
    return _focusLogo.FocusIcon;
  }
});
Object.defineProperty(exports, "FocusLogo", {
  enumerable: true,
  get: function get() {
    return _focusLogo.FocusLogo;
  }
});
Object.defineProperty(exports, "GuardIcon", {
  enumerable: true,
  get: function get() {
    return _guardLogo.GuardIcon;
  }
});
Object.defineProperty(exports, "GuardLogo", {
  enumerable: true,
  get: function get() {
    return _guardLogo.GuardLogo;
  }
});
Object.defineProperty(exports, "HalpIcon", {
  enumerable: true,
  get: function get() {
    return _halpLogo.HalpIcon;
  }
});
Object.defineProperty(exports, "HalpLogo", {
  enumerable: true,
  get: function get() {
    return _halpLogo.HalpLogo;
  }
});
Object.defineProperty(exports, "JiraAlignIcon", {
  enumerable: true,
  get: function get() {
    return _jiraAlignLogo.JiraAlignIcon;
  }
});
Object.defineProperty(exports, "JiraAlignLogo", {
  enumerable: true,
  get: function get() {
    return _jiraAlignLogo.JiraAlignLogo;
  }
});
Object.defineProperty(exports, "JiraIcon", {
  enumerable: true,
  get: function get() {
    return _jiraLogo.JiraIcon;
  }
});
Object.defineProperty(exports, "JiraLogo", {
  enumerable: true,
  get: function get() {
    return _jiraLogo.JiraLogo;
  }
});
Object.defineProperty(exports, "JiraProductDiscoveryIcon", {
  enumerable: true,
  get: function get() {
    return _jiraProductDiscoveryLogo.JiraProductDiscoveryIcon;
  }
});
Object.defineProperty(exports, "JiraProductDiscoveryLogo", {
  enumerable: true,
  get: function get() {
    return _jiraProductDiscoveryLogo.JiraProductDiscoveryLogo;
  }
});
Object.defineProperty(exports, "JiraServiceManagementIcon", {
  enumerable: true,
  get: function get() {
    return _jiraServiceManagementLogo.JiraServiceManagementIcon;
  }
});
Object.defineProperty(exports, "JiraServiceManagementLogo", {
  enumerable: true,
  get: function get() {
    return _jiraServiceManagementLogo.JiraServiceManagementLogo;
  }
});
Object.defineProperty(exports, "JiraSoftwareIcon", {
  enumerable: true,
  get: function get() {
    return _jiraSoftwareLogo.JiraSoftwareIcon;
  }
});
Object.defineProperty(exports, "JiraSoftwareLogo", {
  enumerable: true,
  get: function get() {
    return _jiraSoftwareLogo.JiraSoftwareLogo;
  }
});
Object.defineProperty(exports, "JiraWorkManagementIcon", {
  enumerable: true,
  get: function get() {
    return _jiraWorkManagementLogo.JiraWorkManagementIcon;
  }
});
Object.defineProperty(exports, "JiraWorkManagementLogo", {
  enumerable: true,
  get: function get() {
    return _jiraWorkManagementLogo.JiraWorkManagementLogo;
  }
});
Object.defineProperty(exports, "LoomAttributionIcon", {
  enumerable: true,
  get: function get() {
    return _loomAttributionLogo.LoomAttributionIcon;
  }
});
Object.defineProperty(exports, "LoomAttributionLogo", {
  enumerable: true,
  get: function get() {
    return _loomAttributionLogo.LoomAttributionLogo;
  }
});
Object.defineProperty(exports, "LoomIcon", {
  enumerable: true,
  get: function get() {
    return _loomLogo.LoomIcon;
  }
});
Object.defineProperty(exports, "LoomLogo", {
  enumerable: true,
  get: function get() {
    return _loomLogo.LoomLogo;
  }
});
Object.defineProperty(exports, "OpsgenieIcon", {
  enumerable: true,
  get: function get() {
    return _opsgenieLogo.OpsgenieIcon;
  }
});
Object.defineProperty(exports, "OpsgenieLogo", {
  enumerable: true,
  get: function get() {
    return _opsgenieLogo.OpsgenieLogo;
  }
});
Object.defineProperty(exports, "RovoIcon", {
  enumerable: true,
  get: function get() {
    return _rovoLogo.RovoIcon;
  }
});
Object.defineProperty(exports, "RovoLogo", {
  enumerable: true,
  get: function get() {
    return _rovoLogo.RovoLogo;
  }
});
Object.defineProperty(exports, "StatuspageIcon", {
  enumerable: true,
  get: function get() {
    return _statuspageLogo.StatuspageIcon;
  }
});
Object.defineProperty(exports, "StatuspageLogo", {
  enumerable: true,
  get: function get() {
    return _statuspageLogo.StatuspageLogo;
  }
});
Object.defineProperty(exports, "TrelloIcon", {
  enumerable: true,
  get: function get() {
    return _trelloLogo.TrelloIcon;
  }
});
Object.defineProperty(exports, "TrelloLogo", {
  enumerable: true,
  get: function get() {
    return _trelloLogo.TrelloLogo;
  }
});
var _atlassianLogo = require("./atlassian-logo");
var _atlassianStartLogo = require("./atlassian-start-logo");
var _atlassianAnalyticsLogo = require("./atlassian-analytics-logo");
var _bitbucketLogo = require("./bitbucket-logo");
var _compassLogo = require("./compass-logo");
var _confluenceLogo = require("./confluence-logo");
var _halpLogo = require("./halp-logo");
var _jiraAlignLogo = require("./jira-align-logo");
var _jiraLogo = require("./jira-logo");
var _jiraProductDiscoveryLogo = require("./jira-product-discovery-logo");
var _jiraServiceManagementLogo = require("./jira-service-management-logo");
var _jiraSoftwareLogo = require("./jira-software-logo");
var _jiraWorkManagementLogo = require("./jira-work-management-logo");
var _loomAttributionLogo = require("./loom-attribution-logo");
var _loomLogo = require("./loom-logo");
var _opsgenieLogo = require("./opsgenie-logo");
var _statuspageLogo = require("./statuspage-logo");
var _trelloLogo = require("./trello-logo");
var _atlasLogo = require("./atlas-logo");
var _atlassianMarketplaceLogo = require("./atlassian-marketplace-logo");
var _rovoLogo = require("./rovo-logo");
var _guardLogo = require("./guard-logo");
var _focusLogo = require("./focus-logo");
var _atlassianAccessLogo = require("./atlassian-access-logo");
var _atlassianAdministrationLogo = require("./atlassian-administration-logo");
var _atlassianAdminLogo = require("./atlassian-admin-logo");
