// @atlaskit/jql-editor/dist/cjs/ui/jql-editor-controls-content/search/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Search = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireDefault(require("react"));
var _tooltip = _interopRequireDefault(require("@atlaskit/tooltip"));
var _styled = require("../../../common/styled");
var _state = require("../../../state");
var _tooltipTag = require("../../tooltip-tag");
var _baseSearch = require("../base-search");
var _messages = require("./messages");
// eslint-disable-next-line @atlassian/tangerine/import/no-parent-imports

// eslint-disable-next-line @atlassian/tangerine/import/no-parent-imports

var Search = exports.Search = function Search() {
  var _useIntl = (0, _state.useIntl)(),
    _useIntl2 = (0, _slicedToArray2.default)(_useIntl, 1),
    intl = _useIntl2[0];
  var _useStoreActions = (0, _state.useStoreActions)(),
    _useStoreActions2 = (0, _slicedToArray2.default)(_useStoreActions, 2),
    onSearch = _useStoreActions2[1].onSearch;
  var label = intl.formatMessage(_messages.messages.searchLabel);
  var _useIsSearching = (0, _state.useIsSearching)(),
    _useIsSearching2 = (0, _slicedToArray2.default)(_useIsSearching, 1),
    isSearching = _useIsSearching2[0];
  return /*#__PURE__*/_react.default.createElement(_tooltip.default, {
    position: 'bottom',
    content: /*#__PURE__*/_react.default.createElement(_styled.TooltipContent, null, label)
    // @ts-ignore: [PIT-1685] Fails in post-office due to backwards incompatibility issue with React 18
    ,
    tag: _tooltipTag.TooltipTag
  }, /*#__PURE__*/_react.default.createElement(_baseSearch.BaseSearch, {
    label: label,
    onSearch: onSearch,
    isSearching: isSearching
  }));
};
