// @atlaskit/media-client/dist/cjs/file-streams-cache.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getFileStreamsCache = exports.StreamsCache = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var StreamsCache = exports.StreamsCache = /*#__PURE__*/function () {
  function StreamsCache(streams) {
    (0, _classCallCheck2.default)(this, StreamsCache);
    this.streams = streams;
  }
  return (0, _createClass2.default)(StreamsCache, [{
    key: "has",
    value: function has(id) {
      return !!this.streams.find(id);
    }
  }, {
    key: "set",
    value: function set(id, stream) {
      this.streams.set(id, stream);
    }
  }, {
    key: "get",
    value: function get(id) {
      return this.streams.get(id);
    }
  }, {
    key: "getOrInsert",
    value: function getOrInsert(id, callback) {
      if (!this.has(id)) {
        this.set(id, callback());
      }
      return this.get(id);
    }
  }, {
    key: "removeAll",
    value: function removeAll() {
      this.streams.clear();
    }
  }, {
    key: "remove",
    value: function remove(id) {
      this.streams.delete(id);
    }
  }, {
    key: "size",
    get: function get() {
      return this.streams.size;
    }
  }]);
}();
var streamCache;
var getFileStreamsCache = exports.getFileStreamsCache = function getFileStreamsCache() {
  if (!streamCache) {
    // TODO: we can move this into a static import like
    // import {mediaState} from '@atlaskit/media-core'
    var mediaState = require('@atlaskit/media-core').mediaState;
    streamCache = new StreamsCache(mediaState.streams);
  }
  return streamCache;
};
