// @atlaskit/editor-plugin-paste-options-toolbar/dist/cjs/pm-plugins/reducer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.reducer = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _actions = require("../editor-actions/actions");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var reducer = exports.reducer = function reducer(state, action) {
  switch (action.type) {
    case _actions.PastePluginActionTypes.SHOW_PASTE_OPTIONS:
      {
        return _objectSpread(_objectSpread({}, state), {}, {
          showToolbar: true,
          highlightContent: false,
          isPlainText: action.data.isPlainText,
          plaintext: action.data.plaintext,
          selectedOption: action.data.selectedOption,
          richTextSlice: action.data.richTextSlice,
          pasteStartPos: action.data.pasteStartPos,
          pasteEndPos: action.data.pasteEndPos
        });
      }
    case _actions.PastePluginActionTypes.HIDE_PASTE_OPTIONS:
      {
        return _objectSpread(_objectSpread({}, state), {}, {
          highlightContent: false,
          showToolbar: false
        });
      }
    case _actions.PastePluginActionTypes.HIGHLIGHT_CONTENT:
      {
        return _objectSpread(_objectSpread({}, state), {}, {
          highlightContent: true
        });
      }
    case _actions.PastePluginActionTypes.CHANGE_FORMAT:
      {
        return _objectSpread(_objectSpread({}, state), {}, {
          highlightContent: true,
          selectedOption: action.data.selectedOption
        });
      }
    default:
      return state;
  }
};
