// @atlaskit/inline-edit/dist/es2019/internal/buttons.js
/* buttons.tsx generated by @compiled/babel-plugin v0.36.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./buttons.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
/* eslint-disable @atlaskit/design-system/no-nested-styles */
/* eslint-disable @atlaskit/ui-styling-standard/no-nested-selectors */

import { IconButton } from '@atlaskit/button/new';
import CancelIcon from '@atlaskit/icon/glyph/cross';
import CrossIcon from '@atlaskit/icon/utility/cross';
import CheckMarkIcon from '@atlaskit/icon/utility/migration/check-mark--check';
import { B400, N0, N20A, N30A, N50A, N60A } from '@atlaskit/theme/colors';
const buttonsContainerStyles = null;
const buttonWrapperBaseStyles = null;
const Buttons = ({
  confirmButtonLabel,
  cancelButtonLabel,
  onMouseDown,
  onCancelClick,
  testId
}) => {
  return /*#__PURE__*/React.createElement("div", {
    className: ax(["_1e0c1txw _kqswstnw _1o9zidpf _152t1osq _rjxpidpf _1pfh12x7"])
  }, /*#__PURE__*/React.createElement("div", {
    tabIndex: -1,
    className: ax(["_2rko1sit _11c82smr _vchhusvi _1bsbxy5q _1pby1kze _bfhk1bhr _16qs1kf5 _1mpy1b66 _1nlxs66q _l2uv88o1 _7mfvdecw _szlb1x4z"])
  }, /*#__PURE__*/React.createElement(IconButton, {
    type: "submit",
    icon: iconProps => /*#__PURE__*/React.createElement(CheckMarkIcon, _extends({}, iconProps, {
      LEGACY_size: "small"
    })),
    onMouseDown: onMouseDown,
    label: confirmButtonLabel,
    testId: testId && `${testId}--confirm`
  })), /*#__PURE__*/React.createElement("div", {
    tabIndex: -1,
    className: ax(["_2rko1sit _11c82smr _vchhusvi _1bsbxy5q _1pby1kze _bfhk1bhr _16qs1kf5 _1mpy1b66 _1nlxs66q _l2uv88o1 _7mfvdecw _szlb1x4z"])
  }, /*#__PURE__*/React.createElement(IconButton, {
    icon: iconProps => /*#__PURE__*/React.createElement(CrossIcon, _extends({}, iconProps, {
      LEGACY_size: "small",
      LEGACY_fallbackIcon: CancelIcon
    })),
    label: cancelButtonLabel,
    onClick: onCancelClick,
    onMouseDown: onMouseDown,
    testId: testId && `${testId}--cancel`
  })));
};

// eslint-disable-next-line @repo/internal/react/require-jsdoc
export default Buttons;
