// @atlaskit/jql-editor/dist/cjs/plugins/jql-syntax-highlighting/visitor.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SyntaxHighlightingVisitor = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _jqlAst = require("@atlaskit/jql-ast");
var _getDocumentPosition = _interopRequireDefault(require("../common/get-document-position"));
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var SyntaxHighlightingVisitor = exports.SyntaxHighlightingVisitor = /*#__PURE__*/function (_AbstractJastVisitor) {
  function SyntaxHighlightingVisitor(state) {
    var _this;
    (0, _classCallCheck2.default)(this, SyntaxHighlightingVisitor);
    _this = _callSuper(this, SyntaxHighlightingVisitor);
    (0, _defineProperty2.default)(_this, "visitCompoundOperator", function (compoundOperator) {
      return compoundOperator.positions.map(function (position) {
        return _this.createHighlight('keyword', position);
      });
    });
    (0, _defineProperty2.default)(_this, "visitNotClauseOperator", function (notClauseOperator) {
      return _this.getHighlightsIfPosition('keyword', notClauseOperator.position);
    });
    (0, _defineProperty2.default)(_this, "visitOrderByOperator", function (orderByOperator) {
      return _this.getHighlightsIfPosition('keyword', orderByOperator.position);
    });
    (0, _defineProperty2.default)(_this, "visitOrderByDirection", function (orderByDirection) {
      return _this.getHighlightsIfPosition('keyword', orderByDirection.position);
    });
    (0, _defineProperty2.default)(_this, "visitField", function (field) {
      return _this.getHighlightsIfPosition('field', field.position);
    });
    (0, _defineProperty2.default)(_this, "visitOperator", function (operator) {
      return _this.getHighlightsIfPosition('operator', operator.position);
    });
    (0, _defineProperty2.default)(_this, "visitPredicateOperator", function (predicateOperator) {
      return _this.getHighlightsIfPosition('operator', predicateOperator.position);
    });
    (0, _defineProperty2.default)(_this, "getHighlightsIfPosition", function (tokenType, maybePosition) {
      if (!maybePosition) {
        return [];
      }
      return [_this.createHighlight(tokenType, maybePosition)];
    });
    (0, _defineProperty2.default)(_this, "createHighlight", function (tokenType, _ref) {
      var _ref2 = (0, _slicedToArray2.default)(_ref, 2),
        start = _ref2[0],
        end = _ref2[1];
      var documentFrom = (0, _getDocumentPosition.default)(_this.state.doc, start);
      var documentTo = (0, _getDocumentPosition.default)(_this.state.doc, end);
      return {
        tokenType: tokenType,
        documentFrom: documentFrom,
        documentTo: documentTo
      };
    });
    _this.state = state;
    return _this;
  }
  (0, _inherits2.default)(SyntaxHighlightingVisitor, _AbstractJastVisitor);
  return (0, _createClass2.default)(SyntaxHighlightingVisitor, [{
    key: "shouldVisitNextChild",
    value: function shouldVisitNextChild(node) {
      // Skip visiting operand subtrees, which don't require syntax highlighting and can make a big impact on performance
      return !(0, _jqlAst.isOperandNode)(node);
    }
  }, {
    key: "aggregateResult",
    value: function aggregateResult(aggregate, nextResult) {
      return aggregate.concat(nextResult);
    }
  }, {
    key: "defaultResult",
    value: function defaultResult() {
      return [];
    }
  }]);
}(_jqlAst.AbstractJastVisitor);
