// @atlaskit/jql-editor/dist/cjs/plugins/rich-inline-nodes/util/create-node-spec.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createNodeSpec = void 0;
var _mapKeys = _interopRequireDefault(require("lodash/mapKeys"));
var _mapValues = _interopRequireDefault(require("lodash/mapValues"));
var _constants = require("../constants");
var createNodeSpec = exports.createNodeSpec = function createNodeSpec(name, attrs) {
  return {
    group: _constants.RICH_INLINE_NODE,
    content: 'text*',
    inline: true,
    atom: true,
    // treat node as a leaf
    attrs: attrs,
    toDOM: function toDOM(node) {
      var domAttrs = (0, _mapKeys.default)(node.attrs, function (_, attr) {
        return "data-".concat(attr);
      });
      return [name, domAttrs, 0];
    },
    parseDOM: [{
      tag: name,
      getAttrs: function getAttrs(node) {
        if (node instanceof Element) {
          return (0, _mapValues.default)(attrs, function (_, attr) {
            return node.getAttribute("data-".concat(attr));
          });
        }
        return null; // should never™ happen
      }
    }]
  };
};
