// @atlaskit/profilecard/dist/cjs/styled/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.teamHeaderBgColor = exports.labelTextColor = exports.labelIconColor = exports.headerTextColorInactive = exports.headerTextColor = exports.headerBgColorDisabledUser = exports.headerBgColor = exports.errorTitleColor = exports.errorTextColor = exports.errorIconColor = exports.bgColor = exports.appLabelTextColor = exports.appLabelBgColor = void 0;
var _colors = require("@atlaskit/theme/colors");
var bgColor = exports.bgColor = "var(--ds-surface-overlay, ".concat(_colors.N0, ")");
var headerBgColor = exports.headerBgColor = "var(--ds-background-brand-bold, ".concat(_colors.B500, ")");
var teamHeaderBgColor = exports.teamHeaderBgColor = "var(--ds-background-neutral, ".concat(_colors.N50, ")");
var headerBgColorDisabledUser = exports.headerBgColorDisabledUser = "var(--ds-background-disabled, ".concat(_colors.N30, ")");
var headerTextColor = exports.headerTextColor = "var(--ds-text-inverse, ".concat(_colors.N0, ")");
var headerTextColorInactive = exports.headerTextColorInactive = "var(--ds-text, ".concat(_colors.N800, ")");
var appLabelBgColor = exports.appLabelBgColor = "var(--ds-background-neutral, ".concat(_colors.N20, ")");
var appLabelTextColor = exports.appLabelTextColor = "var(--ds-text, ".concat(_colors.N500, ")");
var labelTextColor = exports.labelTextColor = "var(--ds-text, ".concat(_colors.N800, ")");
var labelIconColor = exports.labelIconColor = "var(--ds-text-subtlest, ".concat(_colors.N60, ")");
var errorIconColor = exports.errorIconColor = "var(--ds-text-disabled, ".concat(_colors.N90, ")");
var errorTitleColor = exports.errorTitleColor = "var(--ds-text, ".concat(_colors.N800, ")");
var errorTextColor = exports.errorTextColor = "var(--ds-text-subtlest, ".concat(_colors.N90, ")");
