// @atlaskit/profilecard/dist/cjs/styled/Error.js
/* Error.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TeamErrorText = exports.ErrorWrapper = exports.ErrorTitle = void 0;
require("./Error.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _styled = _interopRequireDefault(require("@emotion/styled"));
var _css = require("@atlaskit/css");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _compiled = require("@atlaskit/primitives/compiled");
var _colors = require("@atlaskit/theme/colors");
var _constants = require("./constants");
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
var styles = {
  errorWrapper: "_y3gn1h6o _ca0q1ejb _u5f31ejb _n3td1ejb _19bv1ejb",
  errorTitle: "_11c82smr _syaz1fxt _19pku2gc",
  teamErrorText: "_syaz1wmz _19pku2gc"
};
var ErrorWrapper = exports.ErrorWrapper = function ErrorWrapper(props) {
  return (0, _platformFeatureFlags.fg)('compiled-migration-profilecard') ? /*#__PURE__*/_react.default.createElement(_compiled.Box, (0, _extends2.default)({
    xcss: (0, _css.cx)(styles.errorWrapper)
  }, props)) : /*#__PURE__*/_react.default.createElement(ErrorWrapperLegacy, {
    "data-testId": props.testId
  }, props.children);
};
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var ErrorWrapperLegacy = _styled.default.div({
  textAlign: 'center',
  padding: "var(--ds-space-300, 24px)",
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  color: _constants.errorIconColor
});
var ErrorTitle = exports.ErrorTitle = function ErrorTitle(props) {
  return (0, _platformFeatureFlags.fg)('compiled-migration-profilecard') ? /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    xcss: (0, _css.cx)(styles.errorTitle)
  }, props.children) : /*#__PURE__*/_react.default.createElement(ErrorTitleLegacy, null, props.children);
};
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var ErrorTitleLegacy = _styled.default.p({
  font: "var(--ds-font-body, normal 400 14px/20px ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
  color: _constants.errorTitleColor,
  margin: "var(--ds-space-100, 8px)".concat(" 0")
});
var TeamErrorText = exports.TeamErrorText = function TeamErrorText(props) {
  return (0, _platformFeatureFlags.fg)('compiled-migration-profilecard') ? /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    xcss: (0, _css.cx)(styles.teamErrorText)
  }, props.children) : /*#__PURE__*/_react.default.createElement(TeamErrorTextLegacy, null, props.children);
};

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var TeamErrorTextLegacy = _styled.default.p({
  color: "var(--ds-text-subtlest, ".concat(_colors.N200, ")"),
  marginTop: "var(--ds-space-100, 8px)"
});
