// @atlaskit/profilecard/dist/cjs/styled/ReportingLines.js
/* ReportingLines.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ReportingLinesSection = exports.OffsetWrapper = exports.ManagerSection = exports.ManagerName = void 0;
require("./ReportingLines.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _react = _interopRequireDefault(require("react"));
var _styled = _interopRequireDefault(require("@emotion/styled"));
var _css = require("@atlaskit/css");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _compiled = require("@atlaskit/primitives/compiled");
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
var styles = {
  reportingLinesSection: "_18u01b66 _19pku2gc",
  managerSection: "_1e0c1txw _4cvr1h6o _18u01b66 _19pk1b66 _otyr1b66 _2hwx1b66",
  managerName: "_11c81o8v _18u0u2gc",
  offsetWrapper: "_19pk1b66 _18u0r5cr"
};

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles, @atlaskit/design-system/use-primitives -- Ignored via go/DSP-18766
var ReportingLinesSectionLegacy = _styled.default.div({
  // Minor left margin to align better with existing icon fields
  marginLeft: "var(--ds-space-050, 4px)",
  marginTop: "var(--ds-space-100, 8px)"
});
var ReportingLinesSection = exports.ReportingLinesSection = function ReportingLinesSection(_ref) {
  var children = _ref.children;
  return (0, _platformFeatureFlags.fg)('compiled-migration-profilecard') ? /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    xcss: (0, _css.cx)(styles.reportingLinesSection)
  }, children) : /*#__PURE__*/_react.default.createElement(ReportingLinesSectionLegacy, null, children);
};

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var ManagerSectionLegacy = _styled.default.div({
  display: 'flex',
  alignItems: 'center',
  margin: "var(--ds-space-050, 4px)".concat(" ", "var(--ds-space-050, 4px)")
});
var ManagerSection = exports.ManagerSection = function ManagerSection(_ref2) {
  var children = _ref2.children;
  return (0, _platformFeatureFlags.fg)('compiled-migration-profilecard') ? /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    xcss: (0, _css.cx)(styles.managerSection)
  }, children) : /*#__PURE__*/_react.default.createElement(ManagerSectionLegacy, null, children);
};

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var ManagerNameLegacy = _styled.default.span({
  font: "var(--ds-font-body-small, normal 400 11px/16px ui-sans-serif, -apple-system, BlinkMacSystemFont, \"Segoe UI\", Ubuntu, \"Helvetica Neue\", sans-serif)",
  marginLeft: "var(--ds-space-100, 8px)"
});
var ManagerName = exports.ManagerName = function ManagerName(_ref3) {
  var children = _ref3.children;
  return (0, _platformFeatureFlags.fg)('compiled-migration-profilecard') ? /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    xcss: (0, _css.cx)(styles.managerName)
  }, children) : /*#__PURE__*/_react.default.createElement(ManagerNameLegacy, null, children);
};

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled, @atlaskit/ui-styling-standard/no-exported-styles, @atlaskit/design-system/use-primitives -- Ignored via go/DSP-18766
var OffsetWrapperLegacy = _styled.default.div({
  marginTop: "var(--ds-space-050, 4px)",
  // Offset left margin so the avatar aligns with the heading
  marginLeft: "var(--ds-space-negative-050, -4px)"
});
var OffsetWrapper = exports.OffsetWrapper = function OffsetWrapper(_ref4) {
  var children = _ref4.children;
  return (0, _platformFeatureFlags.fg)('compiled-migration-profilecard') ? /*#__PURE__*/_react.default.createElement(_compiled.Box, {
    xcss: (0, _css.cx)(styles.offsetWrapper)
  }, children) : /*#__PURE__*/_react.default.createElement(OffsetWrapperLegacy, null, children);
};
