// @atlaskit/editor-plugin-extension/dist/cjs/ui/ConfigPanel/Fields/Select.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = SelectField;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _form = require("@atlaskit/form");
var _select = _interopRequireDefault(require("@atlaskit/select"));
var _FieldMessages = _interopRequireDefault(require("../FieldMessages"));
var _utils = require("../utils");
var _SelectItem = require("./SelectItem");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function SelectField(_ref) {
  var name = _ref.name,
    field = _ref.field,
    onFieldChange = _ref.onFieldChange,
    autoFocus = _ref.autoFocus,
    placeholder = _ref.placeholder,
    fieldDefaultValue = _ref.fieldDefaultValue;
  //ignore arrays as mutli-value select fields are always clearable
  var hasValidSingleDefaultValue = !Array.isArray(fieldDefaultValue) && fieldDefaultValue !== undefined;
  var isClearable = !hasValidSingleDefaultValue || field.isMultiple;
  return /*#__PURE__*/_react.default.createElement(_form.Field, {
    name: name,
    label: field.label,
    defaultValue: (0, _utils.getOptionFromValue)(field.items, field.defaultValue),
    testId: "config-panel-select-".concat(name),
    isRequired: field.isRequired,
    validate: function validate(value) {
      // Ignored via go/ees005
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      return (0, _utils.validate)(field, value);
    },
    isDisabled: field.isDisabled
  }, function (_ref2) {
    var fieldProps = _ref2.fieldProps,
      error = _ref2.error;
    return /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_select.default
    // Ignored via go/ees005
    // eslint-disable-next-line react/jsx-props-no-spreading
    , (0, _extends2.default)({}, _objectSpread(_objectSpread({}, fieldProps), {}, {
      // Pass `id` as `inputId` so that the input gets the correct id, and make sure there are no duplicate ids
      inputId: fieldProps.id,
      id: undefined
    }), {
      onChange: function onChange(value) {
        fieldProps.onChange(value);
        onFieldChange(name, true);
      }
      // add type cast to avoid adding a "IsMulti" generic prop (TODO: ED-12072)
      ,
      isMulti: field.isMultiple || false,
      options: field.items || [],
      isClearable: isClearable,
      validationState: error ? 'error' : 'default',
      formatOptionLabel: _SelectItem.formatOptionLabel,
      autoFocus: autoFocus,
      menuPlacement: "auto",
      placeholder: placeholder
    })), /*#__PURE__*/_react.default.createElement(_FieldMessages.default, {
      error: error,
      description: field.description
    }));
  });
}
