// @atlaskit/editor-plugin-extension/dist/cjs/ui/ConfigPanel/Fields/String.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = String;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireWildcard(require("react"));
var _form = require("@atlaskit/form");
var _textarea = _interopRequireDefault(require("@atlaskit/textarea"));
var _textfield = _interopRequireDefault(require("@atlaskit/textfield"));
var _FieldMessages = _interopRequireDefault(require("../FieldMessages"));
var _utils = require("../utils");
var _excluded = ["onChange"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function String(_ref) {
  var name = _ref.name,
    field = _ref.field,
    autoFocus = _ref.autoFocus,
    onFieldChange = _ref.onFieldChange,
    placeholder = _ref.placeholder;
  var label = field.label,
    description = field.description,
    defaultValue = field.defaultValue,
    isRequired = field.isRequired,
    isDisabled = field.isDisabled;
  return /*#__PURE__*/_react.default.createElement(_form.Field, {
    name: name,
    label: label,
    defaultValue: defaultValue || '',
    isRequired: isRequired,
    validate: function validate(value) {
      return (0, _utils.validate)(field, value || '');
    },
    testId: "config-panel-string-".concat(name),
    isDisabled: isDisabled
  }, function (_ref2) {
    var fieldProps = _ref2.fieldProps,
      error = _ref2.error,
      meta = _ref2.meta;
    if (field.style === 'multiline') {
      var _onChange = fieldProps.onChange,
        restFieldProps = (0, _objectWithoutProperties2.default)(fieldProps, _excluded);
      var options = field.options;
      return /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_textarea.default
      // Ignored via go/ees005
      // eslint-disable-next-line react/jsx-props-no-spreading
      , (0, _extends2.default)({}, restFieldProps, options, {
        onChange: function onChange(e) {
          return _onChange(e.currentTarget.value);
        },
        onBlur: function onBlur() {
          fieldProps.onBlur();
          onFieldChange(name, meta.dirty);
        },
        placeholder: placeholder,
        "data-testid": "config-panel-textarea-".concat(name)
      })), /*#__PURE__*/_react.default.createElement(_FieldMessages.default, {
        error: error,
        description: description
      }));
    }
    return /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_textfield.default
    // Ignored via go/ees005
    // eslint-disable-next-line react/jsx-props-no-spreading
    , (0, _extends2.default)({}, fieldProps, {
      type: "text",
      autoFocus: autoFocus,
      onBlur: function onBlur() {
        fieldProps.onBlur();
        onFieldChange(name, meta.dirty);
      },
      placeholder: placeholder
    })), /*#__PURE__*/_react.default.createElement(_FieldMessages.default, {
      error: error,
      description: description
    }));
  });
}
