// @atlaskit/smart-card/dist/cjs/view/FlexibleCard/components/container/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getFlexibleUiBlock = exports.getChildrenOptions = exports.default = void 0;
require("./index.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireWildcard(require("react"));
var _reactMagneticDi = require("react-magnetic-di");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _colors = require("@atlaskit/theme/colors");
var _constants = require("../../../../constants");
var _flexibleUiContext = require("../../../../state/flexible-ui-context");
var _utils = require("../../../../state/flexible-ui-context/utils");
var _flexible = require("../../../../utils/flexible");
var _blocks = require("../blocks");
var _hoverCardControl = _interopRequireDefault(require("./hover-card-control"));
var _layeredLink = _interopRequireDefault(require("./layered-link"));
var _excluded = ["containerSize"];
/* eslint-disable @atlaskit/design-system/consistent-css-prop-usage */
/* eslint-disable @atlaskit/ui-styling-standard/no-unsafe-values */
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var getChildrenOptions = exports.getChildrenOptions = function getChildrenOptions(children, context) {
  var options = {};
  if ((0, _utils.isFlexUiPreviewPresent)(context)) {
    _react.default.Children.map(children, function (child) {
      if ( /*#__PURE__*/_react.default.isValidElement(child)) {
        if ((0, _flexible.isFlexibleUiPreviewBlock)(child)) {
          var placement = child.props.placement;
          if (placement === _constants.MediaPlacement.Left) {
            options.previewOnLeft = true;
          }
          if (placement === _constants.MediaPlacement.Right) {
            options.previewOnRight = true;
          }
        }
      }
    });
  }
  return options;
};
var renderChildren = function renderChildren(children, containerSize, containerTheme, status, retry, onClick, removeBlockRestriction) {
  return _react.default.Children.map(children, function (child) {
    if ( /*#__PURE__*/_react.default.isValidElement(child) && (0, _flexible.isFlexibleUiBlock)(child)) {
      if ((0, _flexible.isFlexibleUiTitleBlock)(child)) {
        if ((0, _flexible.isStyleCacheProvider)(child)) {
          return updateChildrenTitleBlock(child, {
            onClick: onClick,
            retry: retry,
            containerSize: containerSize,
            status: status,
            theme: containerTheme
          });
        } else {
          var _blockSize = child.props.size;
          var _size = _blockSize || containerSize;
          return /*#__PURE__*/_react.default.cloneElement(child, {
            // @ts-expect-error
            onClick: onClick,
            retry: retry,
            size: _size,
            status: status,
            theme: containerTheme
          });
        }
      }
      var blockSize = child.props.size;
      var size = blockSize || containerSize;
      if ((0, _flexible.isStyleCacheProvider)(child)) {
        return updateChildrenBlock(child, size, status);
      }

      // @ts-expect-error
      return /*#__PURE__*/_react.default.cloneElement(child, {
        size: size,
        status: status
      });
    }
    if ((0, _platformFeatureFlags.fg)('platform-linking-flexible-card-openness')) {
      if (removeBlockRestriction) {
        return child;
      }
    }
  });
};
var updateChildrenBlock = function updateChildrenBlock(node, size, status) {
  var updatedChildren = _react.default.Children.map(node.props.children, function (child) {
    if ( /*#__PURE__*/_react.default.isValidElement(child) && (0, _flexible.isFlexibleUiBlock)(child)) {
      // @ts-expect-error
      return /*#__PURE__*/_react.default.cloneElement(child, {
        size: size,
        status: status
      });
    }
    return child;
  });
  return /*#__PURE__*/_react.default.cloneElement(node, {
    children: updatedChildren
  });
};
var updateChildrenTitleBlock = function updateChildrenTitleBlock(node, _ref) {
  var containerSize = _ref.containerSize,
    props = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  var updatedChildren = _react.default.Children.map(node.props.children, function (child) {
    if ( /*#__PURE__*/_react.default.isValidElement(child) && (0, _flexible.isFlexibleUiTitleBlock)(child)) {
      var _ref2 = child.props,
        blockSize = _ref2.size;
      var size = blockSize || containerSize;
      return /*#__PURE__*/_react.default.cloneElement(child, _objectSpread(_objectSpread({}, props), {}, {
        size: size
      }));
    } else {
      return child;
    }
  });
  return /*#__PURE__*/_react.default.cloneElement(node, {
    children: updatedChildren
  });
};

/**
 * Note: This function is only necessary for CompiledCSS within Jest tests due to the way it handles Styles.
 * CompiledCSS will inject a StyleCacheProvider around the component tree, which
 * causes the children to be wrapped in a StyleCacheProvider as well. This function recursively
 * searches for the first valid TitleBlock within the children of the StyleCacheProvider.
 */
var _getFlexibleUiBlock = exports.getFlexibleUiBlock = function getFlexibleUiBlock(node) {
  if (! /*#__PURE__*/_react.default.isValidElement(node)) {
    return undefined;
  }
  if (node.type === _blocks.TitleBlock) {
    return node;
  }
  if ((0, _flexible.isStyleCacheProvider)(node)) {
    // Component wrapped with compiled at runtime, check for children
    var isChildrenValid;
    _react.default.Children.map(node.props.children, function (child) {
      var _child$type;
      if (typeof child.type !== 'string' && ((_child$type = child.type) === null || _child$type === void 0 ? void 0 : _child$type.name) !== 'Style') {
        isChildrenValid = _getFlexibleUiBlock(child);
      }
    });
    return isChildrenValid;
  }
  return undefined;
};
var getTitleBlockProps = function getTitleBlockProps(children) {
  var block = _react.default.Children.toArray(children).map(function (child) {
    return _getFlexibleUiBlock(child);
  }).filter(function (x) {
    return x !== undefined;
  }).at(0);
  if ( /*#__PURE__*/_react.default.isValidElement(block)) {
    return block.props;
  }
};
var getLayeredLink = function getLayeredLink(testId, context, children, onClick) {
  var _ref3 = context || {},
    title = _ref3.title,
    _ref3$url = _ref3.url,
    url = _ref3$url === void 0 ? '' : _ref3$url;
  var _ref4 = getTitleBlockProps(children) || {},
    target = _ref4.anchorTarget,
    text = _ref4.text;
  return /*#__PURE__*/_react.default.createElement(_layeredLink.default, {
    onClick: onClick,
    target: target,
    testId: testId,
    text: text || title,
    url: url
  });
};
var baseStyleCommon = null;

// TODO: Remove on fg cleanup: platform-linking-visual-refresh-v1
var backgroundStyleOld = null;
var backgroundStyles = null;

// TODO: Remove on fg cleanup: platform-linking-visual-refresh-v1
var elevationStylesOld = null;
var elevationStyles = null;
var clickableContainerStyles = null;

// TODO: Remove on fg cleanup: platform-linking-visual-refresh-v1
var gapStyleMapOld = {
  xlarge: "_zulp1jvo",
  large: "_zulph3jc",
  medium: "_zulp1qos",
  small: "_zulpfzsm"
};
var gapStyleMap = {
  xlarge: "_zulpr2fv",
  large: "_zulpzu5n",
  medium: "_zulp1bpf",
  small: "_zulpswc3"
};

/**
 * Get container padding based on smart link size
 * Equivalent version for DS primitives space token is getPrimitivesPaddingSpaceBySize()
 * at view/FlexibleCard/components/utils.tsx
 */
var getPadding = function getPadding(size) {
  if ((0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v1')) {
    switch (size) {
      case _constants.SmartLinkSize.XLarge:
        return "var(--ds-space-300, 24px)";
      case _constants.SmartLinkSize.Large:
        return "var(--ds-space-250, 20px)";
      case _constants.SmartLinkSize.Medium:
        return "var(--ds-space-200, 16px)";
      case _constants.SmartLinkSize.Small:
      default:
        return "var(--ds-space-100, 8px)";
    }
  }
  switch (size) {
    case _constants.SmartLinkSize.XLarge:
      return '1.5rem';
    case _constants.SmartLinkSize.Large:
      return '1.25rem';
    case _constants.SmartLinkSize.Medium:
      return '1rem';
    case _constants.SmartLinkSize.Small:
    default:
      return '.5rem';
  }
};
var getGap = function getGap(size) {
  if ((0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v1')) {
    switch (size) {
      case _constants.SmartLinkSize.XLarge:
        return "var(--ds-space-250, 20px)";
      case _constants.SmartLinkSize.Large:
        return "var(--ds-space-200, 16px)";
      case _constants.SmartLinkSize.Medium:
        return "var(--ds-space-100, 8px)";
      case _constants.SmartLinkSize.Small:
      default:
        return "var(--ds-space-050, 4px)";
    }
  }
  switch (size) {
    case _constants.SmartLinkSize.XLarge:
      return '1.25rem';
    case _constants.SmartLinkSize.Large:
      return '1rem';
    case _constants.SmartLinkSize.Medium:
      return '.5rem';
    case _constants.SmartLinkSize.Small:
    default:
      return '.25rem';
  }
};

// TODO: Remove on fg cleanup: platform-linking-visual-refresh-v1
var previewOnLeftStyleMapOld = {
  xlarge: "_19bv15sf _y5vk1k92",
  large: "_19bvf557 _y5vk1j6v",
  medium: "_19bvqox2 _y5vk1kwk",
  small: "_19bv1ltm _y5vk6mu8"
};
var previewOnLeftStyleMap = {
  xlarge: "_19bver0k _y5vk1ejb",
  large: "_19bv16ru _y5vkv47k",
  medium: "_19bv1g7w _y5vk1eak",
  small: "_19bvlbes _y5vku2gc"
};

// TODO: Remove on fg cleanup: platform-linking-visual-refresh-v1
var previewOnRightStyleMapOld = {
  xlarge: "_u5f315sf _1yob1k92",
  large: "_u5f3f557 _1yob1j6v",
  medium: "_u5f3qox2 _1yob1kwk",
  small: "_u5f31ltm _1yob6mu8"
};
var previewOnRightStyleMap = {
  xlarge: "_u5f3er0k _1yob1ejb",
  large: "_u5f316ru _1yobv47k",
  medium: "_u5f31g7w _1yob1eak",
  small: "_u5f3lbes _1yobu2gc"
};

/**
 * A container is a hidden component that build the Flexible Smart Link.
 * All the Flexible UI components are wrapped inside the container.
 * It inherits the ui props from Card component and applies the custom styling
 * accordingly.
 * @internal
 * @see Block
 */
var Container = function Container(_ref5) {
  var children = _ref5.children,
    _ref5$clickableContai = _ref5.clickableContainer,
    clickableContainer = _ref5$clickableContai === void 0 ? false : _ref5$clickableContai,
    _ref5$hideBackground = _ref5.hideBackground,
    hideBackground = _ref5$hideBackground === void 0 ? false : _ref5$hideBackground,
    _ref5$hideElevation = _ref5.hideElevation,
    hideElevation = _ref5$hideElevation === void 0 ? false : _ref5$hideElevation,
    _ref5$hidePadding = _ref5.hidePadding,
    hidePadding = _ref5$hidePadding === void 0 ? false : _ref5$hidePadding,
    onClick = _ref5.onClick,
    retry = _ref5.retry,
    _ref5$showHoverPrevie = _ref5.showHoverPreview,
    showHoverPreview = _ref5$showHoverPrevie === void 0 ? false : _ref5$showHoverPrevie,
    hoverPreviewOptions = _ref5.hoverPreviewOptions,
    actionOptions = _ref5.actionOptions,
    _ref5$removeBlockRest = _ref5.removeBlockRestriction,
    removeBlockRestriction = _ref5$removeBlockRest === void 0 ? false : _ref5$removeBlockRest,
    _ref5$size = _ref5.size,
    size = _ref5$size === void 0 ? _constants.SmartLinkSize.Medium : _ref5$size,
    status = _ref5.status,
    _ref5$testId = _ref5.testId,
    testId = _ref5$testId === void 0 ? 'smart-links-container' : _ref5$testId,
    _ref5$theme = _ref5.theme,
    theme = _ref5$theme === void 0 ? _constants.SmartLinkTheme.Link : _ref5$theme;
  var padding = hidePadding ? (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v1') ? '0px' : '0rem' : getPadding(size);
  var gap = getGap(size);
  var context = (0, _react.useContext)(_flexibleUiContext.FlexibleUiContext);
  var _getChildrenOptions = getChildrenOptions(children, context),
    previewOnLeft = _getChildrenOptions.previewOnLeft,
    previewOnRight = _getChildrenOptions.previewOnRight;
  var canShowHoverPreview = showHoverPreview && status === 'resolved';
  // `retry` object contains action that can be performed on
  // unresolved link (unauthorized, forbidden, not found, etc.)
  var canShowAuthTooltip = showHoverPreview && status === 'unauthorized' && retry !== undefined;
  var containerPaddingStyles = null;
  var container = (0, _platformFeatureFlags.fg)('platform-linking-visual-refresh-v1') ? /*#__PURE__*/_react.default.createElement("div", {
    "data-smart-link-container": true,
    "data-testid": testId,
    className: (0, _runtime.ax)(["_1e0c1txw _2lx21bp4 _1ul9idpf _1reo15vq _kqswh2mm _1pcmkb7n _1v7y1i6y _ufn21i6y", gapStyleMap[size], !hideBackground && "_bfhkhp5a", "_1yt418y6 _jb7v18y6 _y5vk167t _1yob1wow _r37x1r5k", previewOnLeft && previewOnLeftStyleMap[size], previewOnRight && previewOnRightStyleMap[size], !hideElevation && "_19itpp12 _2rkob7nn _19pkv77o _2hwxv77o _otyrv77o _18u0v77o", clickableContainer && "_nqwih2mm _18ywh2mm _1nq3h2mm _1t4ckb7n _1nxdkb7n _xnbykb7n"]),
    style: {
      "--_1xumd0e": (0, _runtime.ix)(padding),
      "--_73mooq": (0, _runtime.ix)(previewOnLeft ? gap : padding),
      "--_12k13bg": (0, _runtime.ix)(previewOnRight ? gap : padding),
      "--_18x9sjc": (0, _runtime.ix)("1px solid ".concat("var(--ds-border, #091E4224)"))
    }
  }, clickableContainer ? getLayeredLink(testId, context, children, onClick) : null, renderChildren(children, size, theme, status, retry, onClick, removeBlockRestriction)) : /*#__PURE__*/_react.default.createElement("div", {
    "data-smart-link-container": true,
    "data-testid": testId,
    className: (0, _runtime.ax)(["_1e0c1txw _2lx21bp4 _1ul9idpf _1reo15vq _kqswh2mm _1pcmkb7n _1v7y1i6y _ufn21i6y", gapStyleMapOld[size], !hideBackground && "_bfhkhp5a", "_1yt418y6 _jb7v18y6 _y5vk167t _1yob1wow _r37x1r5k", previewOnLeft && previewOnLeftStyleMapOld[size], previewOnRight && previewOnRightStyleMapOld[size], !hideElevation && "_19itn8v0 _2rkopd34 _19pkv77o _2hwxv77o _otyrv77o _18u0v77o", clickableContainer && "_nqwih2mm _18ywh2mm _1nq3h2mm _1t4ckb7n _1nxdkb7n _xnbykb7n"]),
    style: {
      "--_1xumd0e": (0, _runtime.ix)(padding),
      "--_73mooq": (0, _runtime.ix)(previewOnLeft ? gap : padding),
      "--_12k13bg": (0, _runtime.ix)(previewOnRight ? gap : padding),
      "--_vzvmm2": (0, _runtime.ix)("1px solid ".concat("var(--ds-border, ".concat(_colors.N40, ")")))
    }
  }, clickableContainer ? getLayeredLink(testId, context, children, onClick) : null, renderChildren(children, size, theme, status, retry, onClick, removeBlockRestriction));
  if (context !== null && context !== void 0 && context.url && (canShowHoverPreview || canShowAuthTooltip)) {
    return /*#__PURE__*/_react.default.createElement(_hoverCardControl.default, {
      isHoverPreview: canShowHoverPreview,
      isAuthTooltip: canShowAuthTooltip,
      actionOptions: actionOptions,
      testId: testId,
      url: context.url,
      hoverPreviewOptions: hoverPreviewOptions
    }, container);
  }
  return container;
};
var _default = exports.default = Container;
