// @atlaskit/teams-public/dist/es2019/ui/team-containers/team-containers-skeleton/linked-container-card-skeleton/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
import "./index.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
import React from 'react';
import { Box, Inline, Stack } from '@atlaskit/primitives/compiled';
const styles = {
  container: "_2rkoiti9 _1h6dz9xs _12y3e4h9 _12jimuej _1qu2nqa1 _syaz1fxt _ca0qpxbi _u5f3pxbi _n3tdpxbi _19bvpxbi",
  avatarSkeleton: "_2rkoglpi _ca0qpxbi _u5f3pxbi _n3tdpxbi _19bvpxbi",
  firstLineSkeleton: "_2rko1q77 _1rjc1b66",
  secondLineSkeleton: "_2rko1q77 _1rjc1b66 _1bsbqkoa"
};
export const LinkedContainerCardSkeleton = () => {
  return /*#__PURE__*/React.createElement(Box, {
    xcss: styles.container
  }, /*#__PURE__*/React.createElement(Inline, {
    space: "space.100",
    alignBlock: "center"
  }, /*#__PURE__*/React.createElement(Box, {
    backgroundColor: "color.background.neutral",
    xcss: styles.avatarSkeleton
  }), /*#__PURE__*/React.createElement(Stack, {
    space: "space.100",
    grow: "fill"
  }, /*#__PURE__*/React.createElement(Box, {
    backgroundColor: "color.background.neutral",
    xcss: styles.firstLineSkeleton
  }), /*#__PURE__*/React.createElement(Box, {
    backgroundColor: "color.background.neutral",
    xcss: styles.secondLineSkeleton
  }))));
};
