// @atlaskit/media-card/dist/cjs/card/ui/playButton/styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.playButtonWrapperStyles = exports.playButtonClassName = exports.fixedPlayButtonStyles = exports.bkgClassName = exports.backgroundStyles = void 0;
var _react = require("@emotion/react");
var _colors = require("@atlaskit/theme/colors");
// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var playButtonClassName = exports.playButtonClassName = 'media-card-play-button';
var bkgClassName = exports.bkgClassName = 'play-icon-background';
var discSize = 48;
var discSizeHover = 56;
var fixedPlayButtonStyles = exports.fixedPlayButtonStyles = "\n  .".concat(bkgClassName, " {\n    width: ").concat(discSizeHover, "px;\n    height: ").concat(discSizeHover, "px;\n  }\n");

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var playButtonWrapperStyles = exports.playButtonWrapperStyles = (0, _react.css)({
  position: 'absolute',
  top: 0,
  left: 0,
  width: '100%',
  height: '100%',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  color: "var(--ds-icon-inverse, ".concat(_colors.N0, ")"),
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
  span: {
    position: 'absolute'
  }
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-exported-styles -- Ignored via go/DSP-18766
var backgroundStyles = exports.backgroundStyles = (0, _react.css)({
  transitionProperty: 'width, height',
  transitionDuration: '0.1s',
  position: 'absolute',
  width: "".concat(discSize, "px"),
  height: "".concat(discSize, "px"),
  background: "var(--ds-background-neutral-bold, ".concat(_colors.N90A, ")"),
  borderRadius: '100%'
});
