// @atlaskit/user-picker/dist/es2019/components/assets/microsoft.js
import _extends from "@babel/runtime/helpers/extends";
import React from 'react';
export function MicrosoftIcon(props) {
  return /*#__PURE__*/React.createElement("svg", _extends({
    width: "12",
    height: "12",
    viewBox: "0 0 12 12",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, props), /*#__PURE__*/React.createElement("g", {
    clipPath: "url(#clip0)"
  }, /*#__PURE__*/React.createElement("path", {
    d: "M0 0H5.70233V5.7023H0V0Z",
    fill: "var(--ds-text-inverse, white)"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M6.29785 0H12.0002V5.7023H6.29785V0Z",
    fill: "var(--ds-text-inverse, white)"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M0 6.29688H5.70233V11.9992H0V6.29688Z",
    fill: "var(--ds-text-inverse, white)"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M6.29785 6.29688H12.0002V11.9992H6.29785V6.29688Z",
    fill: "var(--ds-text-inverse, white)"
  })), /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("clipPath", {
    id: "clip0"
  }, /*#__PURE__*/React.createElement("rect", {
    width: "12",
    height: "11.9999",
    fill: "var(--ds-text-inverse, white)"
  }))));
}
