// @atlaskit/rovo-agent-components/dist/cjs/common/ui/chat-pill/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ConversationStarterPill = exports.ChatPill = exports.BrowseAgentsPill = void 0;
require("./index.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _reactIntlNext = require("react-intl-next");
var _css = require("@atlaskit/css");
var _aiAgent = _interopRequireDefault(require("@atlaskit/icon/core/ai-agent"));
var _compiled = require("@atlaskit/primitives/compiled");
var _chatIcon = require("../chat-icon");
var _messages = _interopRequireDefault(require("./messages"));
var _excluded = ["children", "whiteSpacePreWrap", "renderIcon"],
  _excluded2 = ["children", "whiteSpacePreWrap"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var stylesCompiled = {
  pillLineHeight: "_vwz47vkz"
};
var styles = {
  conversationStarterPill: "_2rkopd34 _11c82smr _189ee4h9 _1dqonqa1 _1h6dmuej _syaz1fxt _ca0qu2gc _u5f3u2gc _n3tdu2gc _19bvu2gc _1o0zgqwt _k48p1wq8 _1bto1l2s _o5721q9c _1o9zkb7n _bfhk1j28 _1bsb1osq _irr315ej _1di6ip91",
  button: "_2rkoiti9 _11c82smr _189ee4h9 _1dqonqa1 _1h6dmuej _syaz1gjq _ca0q12x7 _u5f3utpp _n3td12x7 _19bvutpp _k48p1wq8 _1bto1l2s _o5721q9c _1o9zkb7n _bfhksm61 _irr3166n _1di61dty",
  buttonInline: "_18zrv77o",
  queryText: "_1nmz1hna _y3gn1e5h",
  whiteSpacePreWrap: "_o5721jtm"
};
var ChatPill = exports.ChatPill = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var children = _ref.children,
    _ref$whiteSpacePreWra = _ref.whiteSpacePreWrap,
    whiteSpacePreWrap = _ref$whiteSpacePreWra === void 0 ? true : _ref$whiteSpacePreWra,
    _ref$renderIcon = _ref.renderIcon,
    renderIcon = _ref$renderIcon === void 0 ? true : _ref$renderIcon,
    props = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  return /*#__PURE__*/React.createElement(_compiled.Pressable, (0, _extends2.default)({
    ref: ref
  }, props, {
    xcss: styles.button
  }), /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)([stylesCompiled.pillLineHeight])
  }, /*#__PURE__*/React.createElement(_compiled.Inline, {
    space: "space.075",
    alignBlock: "baseline"
  }, renderIcon ? /*#__PURE__*/React.createElement(_chatIcon.ChatPillIcon, null) : null, /*#__PURE__*/React.createElement(_compiled.Box, {
    xcss: (0, _css.cx)(styles.queryText, whiteSpacePreWrap && styles.whiteSpacePreWrap)
  }, children))));
});
var BrowseAgentsPill = exports.BrowseAgentsPill = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var _useIntl = (0, _reactIntlNext.useIntl)(),
    formatMessage = _useIntl.formatMessage;
  return /*#__PURE__*/React.createElement(_compiled.Pressable, (0, _extends2.default)({
    ref: ref
  }, props, {
    xcss: styles.button
  }), /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)([stylesCompiled.pillLineHeight])
  }, /*#__PURE__*/React.createElement(_compiled.Inline, {
    space: "space.050",
    xcss: styles.buttonInline
  }, /*#__PURE__*/React.createElement(_aiAgent.default, {
    color: "currentColor",
    label: ""
  }), /*#__PURE__*/React.createElement(_compiled.Box, {
    xcss: styles.queryText
  }, formatMessage(_messages.default.browseAgentsPillLabel)))));
});
var ConversationStarterPill = exports.ConversationStarterPill = /*#__PURE__*/(0, _react.forwardRef)(function (_ref2, ref) {
  var children = _ref2.children,
    _ref2$whiteSpacePreWr = _ref2.whiteSpacePreWrap,
    whiteSpacePreWrap = _ref2$whiteSpacePreWr === void 0 ? true : _ref2$whiteSpacePreWr,
    props = (0, _objectWithoutProperties2.default)(_ref2, _excluded2);
  return /*#__PURE__*/React.createElement(_compiled.Pressable, (0, _extends2.default)({
    ref: ref
  }, props, {
    xcss: styles.conversationStarterPill
  }), /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)([stylesCompiled.pillLineHeight])
  }, /*#__PURE__*/React.createElement(_compiled.Inline, {
    space: "space.050",
    xcss: styles.buttonInline
  }, /*#__PURE__*/React.createElement(_chatIcon.ChatPillIcon, null), /*#__PURE__*/React.createElement(_compiled.Box, {
    xcss: (0, _css.cx)(styles.queryText, whiteSpacePreWrap && styles.whiteSpacePreWrap)
  }, children))));
});
