// @atlaskit/jql-editor/dist/cjs/ui/jql-editor-controls-content/base-syntax-help/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BaseSyntaxHelp = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _button = _interopRequireDefault(require("@atlaskit/button"));
var _new = require("@atlaskit/button/new");
var _questionCircle = _interopRequireDefault(require("@atlaskit/icon/core/question-circle"));
var _question = _interopRequireDefault(require("@atlaskit/icon/glyph/question"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _primitives = require("@atlaskit/primitives");
var _colors = require("@atlaskit/theme/colors");
var _styled = require("./styled");
// Fixes icon margin issus after new icon migration
var iconStyle = (0, _primitives.xcss)({
  margin: 'space.025',
  display: 'inline-flex'
});
var BaseSyntaxHelp = exports.BaseSyntaxHelp = function BaseSyntaxHelp(_ref) {
  var describedby = _ref.describedby,
    isDisabled = _ref.isDisabled,
    label = _ref.label,
    onClick = _ref.onClick;
  return (0, _platformFeatureFlags.fg)('platform-component-visual-refresh') ? /*#__PURE__*/_react.default.createElement(_new.LinkIconButton, {
    "aria-describedby": describedby,
    label: label,
    isDisabled: isDisabled,
    appearance: "subtle",
    spacing: "compact",
    onClick: onClick,
    target: "blank",
    href: "https://confluence.atlassian.com/display/SERVICEDESKCLOUD/Advanced+searching",
    icon: function icon(iconProps) {
      return /*#__PURE__*/_react.default.createElement(_questionCircle.default, (0, _extends2.default)({}, iconProps, {
        color: "var(--ds-icon-subtle, #626F86)"
      }));
    }
  }) : /*#__PURE__*/_react.default.createElement(_styled.SyntaxHelpContainer, null, /*#__PURE__*/_react.default.createElement(_button.default, {
    "aria-describedby": describedby,
    "aria-label": label,
    isDisabled: isDisabled,
    appearance: 'subtle',
    spacing: 'none',
    target: 'blank',
    href: "https://confluence.atlassian.com/display/SERVICEDESKCLOUD/Advanced+searching",
    iconBefore: /*#__PURE__*/_react.default.createElement(_primitives.Box, {
      xcss: iconStyle
    }, /*#__PURE__*/_react.default.createElement(_questionCircle.default, {
      label: '',
      color: isDisabled ? "var(--ds-icon-disabled, ".concat(_colors.N0, ")") : "var(--ds-icon-inverse, ".concat(_colors.N0, ")"),
      LEGACY_size: 'small',
      LEGACY_fallbackIcon: _question.default,
      LEGACY_margin: "-2px"
    })),
    onClick: onClick
  }));
};
