// @atlaskit/media-picker/dist/cjs/components/localUploadReact.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LocalUploadComponentReact = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = require("react");
var _perfMarks = require("perf-marks");
var _mediaClient = require("@atlaskit/media-client");
var _mediaCommon = require("@atlaskit/media-common");
var _component = require("./component");
var _uploadServiceImpl = require("../service/uploadServiceImpl");
var _analytics = require("../util/analytics");
var _ufoExperiences = require("../util/ufoExperiences");
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
var LocalUploadComponentReact = exports.LocalUploadComponentReact = /*#__PURE__*/function (_Component) {
  function LocalUploadComponentReact(props, componentName) {
    var _this;
    (0, _classCallCheck2.default)(this, LocalUploadComponentReact);
    _this = _callSuper(this, LocalUploadComponentReact, [props]);
    (0, _defineProperty2.default)(_this, "uploadComponent", new _component.UploadComponent());
    (0, _defineProperty2.default)(_this, "state", {
      errorFlags: []
    });
    (0, _defineProperty2.default)(_this, "addErrorFlag", function (flagData) {
      _this.setState({
        errorFlags: [].concat((0, _toConsumableArray2.default)(_this.state.errorFlags), [flagData])
      });
    });
    (0, _defineProperty2.default)(_this, "dismissErrorFlag", function () {
      _this.setState({
        errorFlags: _this.state.errorFlags.slice(1)
      });
    });
    (0, _defineProperty2.default)(_this, "fireCommencedEvent", function (payload) {
      var files = payload.files,
        traceContext = payload.traceContext;
      files.forEach(function (_ref) {
        var fileId = _ref.id,
          fileSize = _ref.size,
          fileMimetype = _ref.type;
        (0, _perfMarks.start)("MediaPicker.fireUpload.".concat(fileId));
        _this.createAndFireAnalyticsEvent({
          eventType: 'operational',
          action: 'commenced',
          actionSubject: 'mediaUpload',
          actionSubjectId: 'localMedia',
          attributes: {
            sourceType: 'local',
            serviceName: 'upload',
            fileAttributes: {
              fileId: fileId,
              fileSize: fileSize,
              fileMimetype: fileMimetype
            },
            traceContext: traceContext
          }
        });
        (0, _ufoExperiences.startMediaUploadUfoExperience)(fileId, _this.componentName);
      });
    });
    (0, _defineProperty2.default)(_this, "fireUploadSucceeded", function (payload) {
      var _payload$file = payload.file,
        fileId = _payload$file.id,
        fileSize = _payload$file.size,
        fileMimetype = _payload$file.type,
        traceContext = payload.traceContext;
      var _end = (0, _perfMarks.end)("MediaPicker.fireUpload.".concat(fileId)),
        _end$duration = _end.duration,
        uploadDurationMsec = _end$duration === void 0 ? -1 : _end$duration;
      _this.createAndFireAnalyticsEvent({
        eventType: 'operational',
        action: 'succeeded',
        actionSubject: 'mediaUpload',
        actionSubjectId: 'localMedia',
        attributes: {
          sourceType: 'local',
          serviceName: 'upload',
          status: 'success',
          fileAttributes: {
            fileId: fileId,
            fileSize: fileSize,
            fileMimetype: fileMimetype
          },
          uploadDurationMsec: uploadDurationMsec,
          traceContext: traceContext
        }
      });
      (0, _ufoExperiences.succeedMediaUploadUfoExperience)(fileId, {
        fileId: fileId,
        fileSize: fileSize,
        fileMimetype: fileMimetype
      });
    });
    (0, _defineProperty2.default)(_this, "fireUploadFailed", /*#__PURE__*/function () {
      var _ref2 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(payload) {
        var fileId, _payload$error, errorName, rawError, traceContext, _end2, _end2$duration, uploadDurationMsec;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              fileId = payload.fileId, _payload$error = payload.error, errorName = _payload$error.name, rawError = _payload$error.rawError, traceContext = payload.traceContext;
              _end2 = (0, _perfMarks.end)("MediaPicker.fireUpload.".concat(fileId)), _end2$duration = _end2.duration, uploadDurationMsec = _end2$duration === void 0 ? -1 : _end2$duration;
              _this.createAndFireAnalyticsEvent({
                eventType: 'operational',
                action: 'failed',
                actionSubject: 'mediaUpload',
                actionSubjectId: 'localMedia',
                attributes: {
                  sourceType: 'local',
                  serviceName: 'upload',
                  status: 'fail',
                  failReason: errorName,
                  error: !!rawError ? (0, _mediaClient.getMediaClientErrorReason)(rawError) : 'unknown',
                  request: !!rawError ? (0, _analytics.getRequestMetadata)(rawError) : undefined,
                  fileAttributes: {
                    fileId: fileId
                  },
                  uploadDurationMsec: uploadDurationMsec,
                  traceContext: traceContext
                }
              });
              (0, _ufoExperiences.failMediaUploadUfoExperience)(fileId, {
                failReason: errorName,
                error: !!rawError ? (0, _mediaClient.getMediaClientErrorReason)(rawError) : 'unknown',
                request: !!rawError ? (0, _analytics.getRequestMetadata)(rawError) : undefined,
                fileAttributes: {
                  fileId: fileId
                },
                uploadDurationMsec: uploadDurationMsec
              });
            case 4:
            case "end":
              return _context.stop();
          }
        }, _callee);
      }));
      return function (_x) {
        return _ref2.apply(this, arguments);
      };
    }());
    (0, _defineProperty2.default)(_this, "createAndFireAnalyticsEvent", function (payload) {
      var createAnalyticsEvent = _this.props.createAnalyticsEvent;
      if (createAnalyticsEvent) {
        createAnalyticsEvent(payload).fire(_mediaCommon.ANALYTICS_MEDIA_CHANNEL);
      }
    });
    (0, _defineProperty2.default)(_this, "cancel", function (uniqueIdentifier) {
      _this.uploadService.cancel(uniqueIdentifier);
    });
    (0, _defineProperty2.default)(_this, "onFilesAdded", function (_ref3) {
      var files = _ref3.files,
        traceContext = _ref3.traceContext;
      _this.uploadComponent.emitUploadsStart(files, traceContext);
    });
    (0, _defineProperty2.default)(_this, "onFilePreviewUpdate", function (_ref4) {
      var file = _ref4.file,
        preview = _ref4.preview;
      _this.uploadComponent.emitUploadPreviewUpdate(file, preview);
    });
    (0, _defineProperty2.default)(_this, "onFileConverting", function (_ref5) {
      var file = _ref5.file,
        traceContext = _ref5.traceContext;
      _this.uploadComponent.emitUploadEnd(file, traceContext);
    });
    (0, _defineProperty2.default)(_this, "onUploadError", function (_ref6) {
      var fileId = _ref6.fileId,
        error = _ref6.error,
        traceContext = _ref6.traceContext;
      _this.uploadComponent.emitUploadError(fileId, error, traceContext);
    });
    _this.componentName = componentName;
    var _this$props = _this.props,
      mediaClient = _this$props.mediaClient,
      config = _this$props.config,
      onUploadsStart = _this$props.onUploadsStart,
      onPreviewUpdate = _this$props.onPreviewUpdate,
      onEnd = _this$props.onEnd,
      onError = _this$props.onError;
    var tenantUploadParams = config.uploadParams;
    var _config$shouldCopyFil = config.shouldCopyFileToRecents,
      shouldCopyFileToRecents = _config$shouldCopyFil === void 0 ? true : _config$shouldCopyFil;
    _this.uploadComponent.on('uploads-start', _this.fireCommencedEvent);
    _this.uploadComponent.on('upload-end', _this.fireUploadSucceeded);
    _this.uploadComponent.on('upload-error', _this.fireUploadFailed);
    if (onUploadsStart) {
      _this.uploadComponent.on('uploads-start', onUploadsStart);
    }
    if (onPreviewUpdate) {
      _this.uploadComponent.on('upload-preview-update', onPreviewUpdate);
    }
    if (onEnd) {
      _this.uploadComponent.on('upload-end', onEnd);
    }
    if (onError) {
      _this.uploadComponent.on('upload-error', onError);
    }
    _this.uploadService = new _uploadServiceImpl.UploadServiceImpl(mediaClient, tenantUploadParams, shouldCopyFileToRecents);
    _this.uploadService.on('files-added', _this.onFilesAdded);
    _this.uploadService.on('file-preview-update', _this.onFilePreviewUpdate);
    _this.uploadService.on('file-converting', _this.onFileConverting);
    _this.uploadService.on('file-upload-error', _this.onUploadError);
    var onFileRejection = function onFileRejection(rejectionData) {
      var onUploadRejection = tenantUploadParams.onUploadRejection;
      var shouldOverride = onUploadRejection === null || onUploadRejection === void 0 ? void 0 : onUploadRejection(rejectionData);
      if (!shouldOverride) {
        _this.addErrorFlag(rejectionData);
      }
    };
    _this.uploadService.onFileRejection(onFileRejection);
    _this.uploadService.onFileEmpty(onFileRejection);
    return _this;
  }
  (0, _inherits2.default)(LocalUploadComponentReact, _Component);
  return (0, _createClass2.default)(LocalUploadComponentReact, [{
    key: "setUploadParams",
    value: function setUploadParams(uploadParams) {
      this.uploadService.setUploadParams(uploadParams);
    }
  }]);
}(_react.Component);
