// @atlaskit/editor-plugin-extension/dist/cjs/ui/ConfigPanel/DescriptionSummary.js
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DescriptionSummary = DescriptionSummary;
var _react = _interopRequireWildcard(require("react"));
var _reactIntlNext = require("react-intl-next");
var _extensions = require("@atlaskit/editor-common/extensions");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _primitives = require("@atlaskit/primitives");
var _HelpLink = require("./HelpLink");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var descriptionStyles = (0, _primitives.xcss)({
  marginBottom: 'space.300'
});
var helpLinkStyles = (0, _primitives.xcss)({
  paddingTop: 'space.150'
});
function DescriptionSummary(_ref) {
  var extensionManifest = _ref.extensionManifest;
  var _useIntl = (0, _reactIntlNext.useIntl)(),
    formatMessage = _useIntl.formatMessage;
  var description = extensionManifest.description,
    deprecation = extensionManifest.deprecation,
    documentationUrl = extensionManifest.documentationUrl;
  // Use a temporary allowlist of top 3 macros to test out a new "Documentation" CTA ("Need help?")
  // This will be removed when Top 5 Modernized Macros updates are rolled out
  var modernizedMacrosList = ['children', 'recently-updated', 'excerpt'];
  var enableHelpCTA = modernizedMacrosList.includes(extensionManifest.key);
  return /*#__PURE__*/_react.default.createElement(_react.Fragment, null, (description || documentationUrl) && /*#__PURE__*/_react.default.createElement(_primitives.Box, {
    xcss: descriptionStyles
  }, /*#__PURE__*/_react.default.createElement(_primitives.Text, {
    as: "p",
    testId: "config-panel-header-description"
  }, description && /*#__PURE__*/_react.default.createElement(_react.Fragment, null,
  // Ignored via go/ees005
  // eslint-disable-next-line require-unicode-regexp
  description.replace(/([^.])$/, '$1.'), ' '), (deprecation === null || deprecation === void 0 ? void 0 : deprecation.isDeprecated) && (deprecation === null || deprecation === void 0 ? void 0 : deprecation.message) && (0, _platformFeatureFlags.fg)('platform_editor_extension_deprecation_status') && /*#__PURE__*/_react.default.createElement(_primitives.Box, {
    paddingBlockStart: "space.150"
  }, deprecation.message), documentationUrl && (enableHelpCTA ? /*#__PURE__*/_react.default.createElement(_primitives.Box, {
    xcss: helpLinkStyles
  }, /*#__PURE__*/_react.default.createElement(_primitives.Text, {
    as: "p"
  }, /*#__PURE__*/_react.default.createElement(_HelpLink.HelpLink, {
    documentationUrl: documentationUrl,
    label: formatMessage(_extensions.configPanelMessages.help)
  }))) : /*#__PURE__*/_react.default.createElement(_HelpLink.HelpLink, {
    documentationUrl: documentationUrl,
    label: formatMessage(_extensions.configPanelMessages.documentation)
  })))));
}
