// @atlaskit/editor-plugin-table/dist/es2019/ui/TableFloatingControls/FloatingControlsWithSelection.js
import React from 'react';
import { useSharedPluginState } from '@atlaskit/editor-common/hooks';
import { CornerControls } from './CornerControls/ClassicCornerControls';
import { RowControls } from './RowControls/ClassicControls';
export const FloatingControlsWithSelection = ({
  editorView,
  tableRef,
  isInDanger,
  isResizing,
  isHeaderRowEnabled,
  isHeaderColumnEnabled,
  hoveredRows,
  stickyTop,
  hoverRows,
  selectRow,
  tableActive,
  api
}) => {
  const {
    selectionState
  } = useSharedPluginState(api, ['selection']);
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(CornerControls, {
    editorView: editorView,
    tableRef: tableRef,
    isInDanger: isInDanger,
    isResizing: isResizing,
    isHeaderRowEnabled: isHeaderRowEnabled,
    isHeaderColumnEnabled: isHeaderColumnEnabled,
    hoveredRows: hoveredRows,
    stickyTop: tableActive ? stickyTop : undefined
  }), /*#__PURE__*/React.createElement(RowControls, {
    selection: selectionState === null || selectionState === void 0 ? void 0 : selectionState.selection,
    editorView: editorView,
    tableRef: tableRef,
    hoverRows: hoverRows,
    hoveredRows: hoveredRows,
    isInDanger: isInDanger,
    isResizing: isResizing,
    selectRow: selectRow,
    stickyTop: tableActive ? stickyTop : undefined
  }));
};
