// @atlaskit/editor-plugin-mentions/dist/es2019/ui/SecondaryToolbarComponent.js
import React, { useCallback } from 'react';
import { INPUT_METHOD } from '@atlaskit/editor-common/analytics';
import { useSharedPluginState } from '@atlaskit/editor-common/hooks';
import ToolbarMention from './ToolbarMention';
export function SecondaryToolbarComponent({
  editorView,
  api,
  typeAhead,
  disabled
}) {
  var _api$typeAhead2;
  const {
    mentionState
  } = useSharedPluginState(api, ['mention']);
  const openMentionTypeAhead = useCallback(() => {
    var _api$typeAhead, _api$typeAhead$action;
    api === null || api === void 0 ? void 0 : (_api$typeAhead = api.typeAhead) === null || _api$typeAhead === void 0 ? void 0 : (_api$typeAhead$action = _api$typeAhead.actions) === null || _api$typeAhead$action === void 0 ? void 0 : _api$typeAhead$action.open({
      triggerHandler: typeAhead,
      inputMethod: INPUT_METHOD.INSERT_MENU
    });
  }, [api, typeAhead]);
  return !(mentionState !== null && mentionState !== void 0 && mentionState.mentionProvider) ? null : /*#__PURE__*/React.createElement(ToolbarMention, {
    editorView: editorView,
    onInsertMention: openMentionTypeAhead,
    isDisabled: disabled || (api === null || api === void 0 ? void 0 : (_api$typeAhead2 = api.typeAhead) === null || _api$typeAhead2 === void 0 ? void 0 : _api$typeAhead2.actions.isAllowed(editorView.state))
  });
}
