// @atlaskit/link-datasource/dist/cjs/ui/confluence-search-modal/basic-filters/filters/date-range-picker/index.js
/* index.tsx generated by @compiled/babel-plugin v0.36.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DateRangePicker = void 0;
require("./index.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _reactIntlNext = require("react-intl-next");
var _datetimePicker = require("@atlaskit/datetime-picker");
var _error = _interopRequireDefault(require("@atlaskit/icon/utility/migration/error"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _popup = _interopRequireDefault(require("@atlaskit/popup"));
var _colors = require("@atlaskit/theme/colors");
var _constants = require("@atlaskit/theme/constants");
var _analytics = require("../../../../../analytics");
var _messages = require("./messages");
var _PopupComponent = require("./PopupComponent");
var _styled = require("./styled");
var _trigger = require("./trigger");
var _utils = require("./utils");
var _excluded = ["as", "style"],
  _excluded2 = ["as", "style"],
  _excluded3 = ["as", "style"],
  _excluded4 = ["as", "style"],
  _excluded5 = ["as", "style"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled
var DateRangeErrorMessage = (0, _react.forwardRef)(function (_ref, __cmplr) {
  var _ref$as = _ref.as,
    C = _ref$as === void 0 ? "div" : _ref$as,
    __cmpls = _ref.style,
    __cmplp = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  return /*#__PURE__*/_react.default.createElement(C, (0, _extends2.default)({}, __cmplp, {
    style: __cmpls,
    ref: __cmplr,
    className: (0, _runtime.ax)(["_zulpv77o _11c8dcr7 _1e0c1txw _19pk57x3 _bozgv77o _syaz1ick _4cvr1h6o", __cmplp.className])
  }));
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled
if (process.env.NODE_ENV !== 'production') {
  DateRangeErrorMessage.displayName = 'DateRangeErrorMessage';
}
var CustomDropdown = (0, _react.forwardRef)(function (_ref2, __cmplr) {
  var _ref2$as = _ref2.as,
    C = _ref2$as === void 0 ? "div" : _ref2$as,
    __cmpls = _ref2.style,
    __cmplp = (0, _objectWithoutProperties2.default)(_ref2, _excluded2);
  return /*#__PURE__*/_react.default.createElement(C, (0, _extends2.default)({}, __cmplp, {
    style: __cmpls,
    ref: __cmplr,
    className: (0, _runtime.ax)(["_2rkoiti9 _1bsb5x59 _bfhk1j9a _16qs1sus _1pby16oo", __cmplp.className])
  }));
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled
if (process.env.NODE_ENV !== 'production') {
  CustomDropdown.displayName = 'CustomDropdown';
}
var SelectDateRangeButton = (0, _react.forwardRef)(function (_ref3, __cmplr) {
  var _ref3$as = _ref3.as,
    C = _ref3$as === void 0 ? "button" : _ref3$as,
    __cmpls = _ref3.style,
    __cmplp = (0, _objectWithoutProperties2.default)(_ref3, _excluded3);
  return /*#__PURE__*/_react.default.createElement(C, (0, _extends2.default)({}, __cmplp, {
    style: __cmpls,
    ref: __cmplr,
    className: (0, _runtime.ax)(["_19itglyw _11c81kw7 _2rkoiti9 _bfhk1y9p _1bsbymf6 _4t3i1ylp _19pkutpp _irr3134c _d0altlke", __cmplp.className])
  }));
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled
if (process.env.NODE_ENV !== 'production') {
  SelectDateRangeButton.displayName = 'SelectDateRangeButton';
}
var CustomDateWrapper = (0, _react.forwardRef)(function (_ref4, __cmplr) {
  var _ref4$as = _ref4.as,
    C = _ref4$as === void 0 ? "div" : _ref4$as,
    __cmpls = _ref4.style,
    __cmplp = (0, _objectWithoutProperties2.default)(_ref4, _excluded4);
  return /*#__PURE__*/_react.default.createElement(C, (0, _extends2.default)({}, __cmplp, {
    style: __cmpls,
    ref: __cmplr,
    className: (0, _runtime.ax)(["_ca0qutpp _u5f3utpp _n3tdutpp _19bvutpp _vchhusvi", __cmplp.className])
  }));
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-styled
if (process.env.NODE_ENV !== 'production') {
  CustomDateWrapper.displayName = 'CustomDateWrapper';
}
var DatePickersWrapper = (0, _react.forwardRef)(function (_ref5, __cmplr) {
  var _ref5$as = _ref5.as,
    C = _ref5$as === void 0 ? "div" : _ref5$as,
    __cmpls = _ref5.style,
    __cmplp = (0, _objectWithoutProperties2.default)(_ref5, _excluded5);
  return /*#__PURE__*/_react.default.createElement(C, (0, _extends2.default)({}, __cmplp, {
    style: __cmpls,
    ref: __cmplr,
    className: (0, _runtime.ax)(["_1e0c1txw _1bah1yb4 _4cvr1h6o _1bsb1osq", __cmplp.className])
  }));
});
if (process.env.NODE_ENV !== 'production') {
  DatePickersWrapper.displayName = 'DatePickersWrapper';
}
var dateRangeValues = ['anyTime', 'today', 'yesterday', 'past7Days', 'past30Days', 'pastYear', 'custom'];
var defaultOptionValue = 'anyTime';
var DateRangePicker = exports.DateRangePicker = function DateRangePicker(_ref6) {
  var onSelectionChange = _ref6.onSelectionChange,
    selection = _ref6.selection,
    _ref6$filterName = _ref6.filterName,
    filterName = _ref6$filterName === void 0 ? 'datasource-date-range-picker' : _ref6$filterName;
  var _ref7 = selection || {},
    selectedValue = _ref7.value,
    fromDate = _ref7.from,
    toDate = _ref7.to;
  var todayDate = new Date().toISOString();
  var _useIntl = (0, _reactIntlNext.useIntl)(),
    locale = _useIntl.locale,
    formatMessage = _useIntl.formatMessage,
    formatDate = _useIntl.formatDate;
  var _useDatasourceAnalyti = (0, _analytics.useDatasourceAnalyticsEvents)(),
    fireEvent = _useDatasourceAnalyti.fireEvent;
  var _useState = (0, _react.useState)(selectedValue),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    currentOption = _useState2[0],
    setCurrentOption = _useState2[1];
  var _useState3 = (0, _react.useState)(fromDate),
    _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
    customFromDate = _useState4[0],
    setCustomFromDate = _useState4[1];
  var _useState5 = (0, _react.useState)(toDate),
    _useState6 = (0, _slicedToArray2.default)(_useState5, 2),
    customToDate = _useState6[0],
    setCustomToDate = _useState6[1];
  var _useState7 = (0, _react.useState)(undefined),
    _useState8 = (0, _slicedToArray2.default)(_useState7, 2),
    isPickerOpen = _useState8[0],
    setIsPickerOpen = _useState8[1];
  var isCustomSelected = currentOption === 'custom';
  var analyticsPayload = (0, _react.useMemo)(function () {
    return {
      filterName: filterName,
      selectionCount: (currentOption !== null && currentOption !== void 0 ? currentOption : defaultOptionValue) === defaultOptionValue ? 0 : 1
    };
  }, [filterName, currentOption]);
  var invalidDateRange = (0, _utils.useInvalidDateRange)(customFromDate, customToDate);
  var handleClickUpdateDateRange = function handleClickUpdateDateRange() {
    onSelectionChange({
      value: 'custom',
      from: customFromDate,
      to: customToDate
    });
    setIsPickerOpen(false);
  };
  var handleClickFilterOption = (0, _react.useCallback)(function (option) {
    if (option === currentOption) {
      setIsPickerOpen(false);
      return;
    }
    setCurrentOption(option);
    onSelectionChange({
      value: option
    });
    if (option !== 'custom') {
      setIsPickerOpen(false);
    }
  }, [onSelectionChange, currentOption]);
  var handlePickerToggle = function handlePickerToggle() {
    if (isPickerOpen) {
      setCustomToDate(toDate);
      setCustomFromDate(fromDate);
      setIsPickerOpen(false);
      return;
    }
    setIsPickerOpen(true);
    fireEvent('ui.dropdown.opened.basicSearchDropdown', analyticsPayload);
  };
  (0, _react.useEffect)(function () {
    if (isPickerOpen === false) {
      fireEvent('ui.dropdown.closed.basicSearchDropdown', analyticsPayload);
    }
  }, [analyticsPayload, fireEvent, isPickerOpen]);
  var popupContent = function popupContent() {
    return /*#__PURE__*/_react.default.createElement(CustomDropdown, null, dateRangeValues.map(function (option) {
      return /*#__PURE__*/_react.default.createElement(_styled.CustomDropdownItem, {
        key: option
        // want to show Anytime as selected if none of the other options are selected
        ,
        isSelected: selectedValue ? option === selectedValue : option === defaultOptionValue,
        onClick: function onClick() {
          return handleClickFilterOption(option);
        }
      }, (0, _utils.getDropdownLabel)(option, formatMessage));
    }), isPickerOpen && isCustomSelected && /*#__PURE__*/_react.default.createElement(CustomDateWrapper, null, /*#__PURE__*/_react.default.createElement(DatePickersWrapper, null, /*#__PURE__*/_react.default.createElement(_datetimePicker.DatePicker, {
      maxDate: todayDate,
      innerProps: {
        style: {
          width: 140
        }
      },
      testId: "date-from-picker",
      dateFormat: "D MMM YYYY",
      onChange: setCustomFromDate,
      defaultValue: fromDate,
      placeholder: formatMessage(_messages.dateRangeMessages.dateRangeFrom),
      isInvalid: Boolean(invalidDateRange),
      locale: locale,
      selectProps: {
        styles: {
          placeholder: function placeholder(base) {
            return _objectSpread(_objectSpread({}, base), {}, {
              width: 'max-content'
            });
          }
        }
      }
    }), /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, _messages.dateRangeMessages.dateRangeToLabel), /*#__PURE__*/_react.default.createElement(_datetimePicker.DatePicker, {
      maxDate: todayDate,
      innerProps: {
        style: {
          width: 140
        }
      },
      testId: "date-to-picker",
      dateFormat: "D MMM YYYY",
      onChange: setCustomToDate,
      defaultValue: toDate,
      placeholder: formatMessage(_messages.dateRangeMessages.dateRangeTo),
      isInvalid: Boolean(invalidDateRange),
      locale: locale,
      selectProps: {
        styles: {
          placeholder: function placeholder(base) {
            return _objectSpread(_objectSpread({}, base), {}, {
              width: 'max-content'
            });
          }
        }
      }
    })), invalidDateRange && /*#__PURE__*/_react.default.createElement(DateRangeErrorMessage, null, /*#__PURE__*/_react.default.createElement(_error.default, {
      LEGACY_size: "small",
      color: "var(--ds-icon-danger, ".concat(_colors.R400, ")"),
      label: (0, _platformFeatureFlags.fg)('bandicoots-update-sllv-icons') ? formatMessage(_messages.dateRangeMessages.dateRangeError) : 'Date range error',
      LEGACY_margin: "0 ".concat("var(--ds-space-negative-025, -2px)")
    }), invalidDateRange), /*#__PURE__*/_react.default.createElement(SelectDateRangeButton, {
      "data-testId": "custom-date-range-update-button",
      disabled: Boolean(invalidDateRange),
      onClick: handleClickUpdateDateRange
    }, formatMessage(_messages.dateRangeMessages.dateRangeUpdateButton))));
  };
  return /*#__PURE__*/_react.default.createElement(_popup.default, {
    isOpen: !!isPickerOpen,
    onClose: handlePickerToggle
    // @ts-ignore: [PIT-1685] Fails in post-office due to backwards incompatibility issue with React 18
    ,
    popupComponent: _PopupComponent.PopupComponent,
    zIndex: _constants.layers.modal(),
    content: popupContent,
    placement: "bottom-start",
    trigger: function trigger(triggerProps) {
      var labelText = (0, _utils.getCurrentOptionLabel)(formatDate, formatMessage, selectedValue, toDate, fromDate);
      var isSelected = !!selectedValue || !!isPickerOpen;
      return /*#__PURE__*/_react.default.createElement(_trigger.PopupTrigger, {
        triggerProps: triggerProps,
        isSelected: isSelected,
        labelPrefix: formatMessage(_messages.dateRangeMessages.dateRangeTitle),
        selectedLabel: labelText,
        onClick: handlePickerToggle
      });
    }
  });
};
