// @atlaskit/link-datasource/dist/cjs/ui/confluence-search-modal/basic-filters/filters/date-range-picker/utils.js
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useInvalidDateRange = exports.getDropdownLabel = exports.getCurrentOptionLabel = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactIntlNext = require("react-intl-next");
var _dateTime = require("../../../../../ui/issue-like-table/render-type/date-time");
var _messages = require("./messages");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var getDropdownLabel = exports.getDropdownLabel = function getDropdownLabel() {
  var option = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'anyTime';
  var formatMessage = arguments.length > 1 ? arguments[1] : undefined;
  var mapping = {
    anyTime: 'dateRangeAnyTime',
    today: 'dateRangeToday',
    yesterday: 'dateRangeYesterday',
    past7Days: 'dateRangeLastWeek',
    past30Days: 'dateRangeLastMonth',
    pastYear: 'dateRangeLastYear',
    custom: 'dateRangeCustom'
  };
  return formatMessage(_messages.dateRangeMessages[mapping[option]]);
};
var getCurrentOptionLabel = exports.getCurrentOptionLabel = function getCurrentOptionLabel(formatDate, formatMessage, value, to, from) {
  var selectedOption = value;
  if (selectedOption === 'custom') {
    var hasFromDate = !!from;
    var hasToDate = !!to;
    var formattedFromDate = (0, _dateTime.getFormattedDate)(from || '', 'date', formatDate);
    var formattedToDate = (0, _dateTime.getFormattedDate)(to || '', 'date', formatDate);
    if (hasFromDate && !hasToDate) {
      return formatMessage(_messages.dateRangeMessages.dateRangeAfterLabel, {
        date: formattedFromDate
      });
    }
    if (!hasFromDate && hasToDate) {
      return formatMessage(_messages.dateRangeMessages.dateRangeBeforeLabel, {
        date: formattedToDate
      });
    }
    if (hasFromDate && hasToDate) {
      return "".concat(formattedFromDate, " - ").concat(formattedToDate);
    }
  }
  return getDropdownLabel(selectedOption, formatMessage);
};
var useInvalidDateRange = exports.useInvalidDateRange = function useInvalidDateRange(from, to) {
  return (0, _react.useMemo)(function () {
    if (!from && !to) {
      return null;
    }
    var dateFrom = new Date("".concat(from).concat(from ? 'T00:00:00' : ''));
    var dateTo = new Date("".concat(to).concat(to ? 'T00:00:00' : ''));
    var now = new Date();
    if (dateFrom > now) {
      return /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, _messages.dateRangeMessages.dateRangeCustomInvalidDateAfterToday);
    }
    if (dateFrom > dateTo) {
      return /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, _messages.dateRangeMessages.dateRangeCustomInvalidDateAfterEnd);
    }
    if (dateTo > now) {
      return /*#__PURE__*/_react.default.createElement(_reactIntlNext.FormattedMessage, _messages.dateRangeMessages.dateRangeCustomInvalidToDateAfterToday);
    }
    return null;
  }, [from, to]);
};
