// @atlaskit/profilecard/dist/cjs/components/ProfileCard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var react_intl_1 = require("react-intl");
var colors_1 = require("@atlaskit/theme/colors");
var avatar_1 = tslib_1.__importDefault(require("@atlaskit/avatar"));
var button_1 = tslib_1.__importDefault(require("@atlaskit/button"));
var lozenge_1 = tslib_1.__importDefault(require("@atlaskit/lozenge"));
var spinner_1 = tslib_1.__importDefault(require("@atlaskit/spinner"));
var IconLabel_1 = tslib_1.__importDefault(require("./IconLabel"));
var ErrorMessage_1 = tslib_1.__importDefault(require("./ErrorMessage"));
var relative_date_1 = tslib_1.__importDefault(require("../internal/relative-date"));
var messages_1 = tslib_1.__importDefault(require("../messages"));
var Card_1 = require("../styled/Card");
var analytics_1 = require("../internal/analytics");
var Profilecard = /** @class */ (function (_super) {
    tslib_1.__extends(Profilecard, _super);
    function Profilecard(props) {
        var _this = _super.call(this, props) || this;
        _this.durationSince = function (from) {
            var fromParsed = from || 0;
            return fromParsed > 0 ? Date.now() - fromParsed : null;
        };
        _this.callClientFetchProfile = function () {
            var _a;
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            if (_this.props.clientFetchProfile) {
                (_a = _this.props).clientFetchProfile.apply(_a, tslib_1.__spread(args));
            }
        };
        _this.callAnalytics = function (id, options) {
            if (_this.props.analytics) {
                _this.props.analytics(id, options);
            }
        };
        _this.timeOpen = null;
        _this.clientFetchProfile = function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            _this.callAnalytics(analytics_1.AnalyticsName.PROFILE_CARD_RELOAD, {});
            _this.callClientFetchProfile.apply(_this, tslib_1.__spread(args));
        };
        return _this;
    }
    Profilecard.prototype.componentDidMount = function () {
        this.timeOpen = Date.now();
        this.callAnalytics(analytics_1.AnalyticsName.PROFILE_CARD_VIEW, {});
    };
    Profilecard.prototype.renderErrorMessage = function () {
        return (react_1.default.createElement(ErrorMessage_1.default, { reload: this.props.clientFetchProfile && this.clientFetchProfile, errorType: this.props.errorType }));
    };
    Profilecard.prototype.renderActionsButtons = function () {
        var _this = this;
        if (this.props.actions && this.props.actions.length === 0) {
            return null;
        }
        return (react_1.default.createElement(Card_1.ActionButtonGroup, null, this.props.actions &&
            this.props.actions.map(function (action, idx) { return (react_1.default.createElement(button_1.default, { appearance: idx === 0 ? 'default' : 'subtle', key: action.label, onClick: function () {
                    var args = [];
                    for (var _i = 0; _i < arguments.length; _i++) {
                        args[_i] = arguments[_i];
                    }
                    _this.callAnalytics(analytics_1.AnalyticsName.PROFILE_CARD_CLICK, {
                        id: action.id || null,
                        duration: _this.durationSince(_this.timeOpen),
                    });
                    if (action.callback) {
                        args[0].preventDefault();
                        action.callback.apply(action, tslib_1.__spread(args));
                    }
                }, href: action.link }, action.label)); })));
    };
    Profilecard.prototype.renderCardDetailsDefault = function () {
        var _a = this.props, meta = _a.meta, location = _a.location, email = _a.email, timestring = _a.timestring, companyName = _a.companyName;
        return (react_1.default.createElement(Card_1.DetailsGroup, null,
            this.renderFullNameAndPublicName(meta),
            meta && react_1.default.createElement(Card_1.JobTitleLabel, null, meta),
            react_1.default.createElement(IconLabel_1.default, { icon: "email" }, email),
            react_1.default.createElement(IconLabel_1.default, { icon: "time" }, timestring),
            react_1.default.createElement(IconLabel_1.default, { icon: "companyName" }, companyName),
            react_1.default.createElement(IconLabel_1.default, { icon: "location" }, location)));
    };
    Profilecard.prototype.renderCardDetailsForDisabledAccount = function () {
        var _a = this.props, status = _a.status, companyName = _a.companyName, hasDisabledAccountLozenge = _a.hasDisabledAccountLozenge;
        return (react_1.default.createElement(Card_1.DetailsGroup, null,
            react_1.default.createElement(Card_1.FullNameLabel, { noMeta: true, isDisabledAccount: true }, this.getDisabledAccountName()),
            hasDisabledAccountLozenge && (react_1.default.createElement(Card_1.LozengeWrapper, null,
                react_1.default.createElement(lozenge_1.default, { appearance: "default", isBold: true },
                    status === 'inactive' && (react_1.default.createElement(react_intl_1.FormattedMessage, tslib_1.__assign({}, messages_1.default.inactiveAccountMsg))),
                    status === 'closed' && (react_1.default.createElement(react_intl_1.FormattedMessage, tslib_1.__assign({}, messages_1.default.closedAccountMsg)))))),
            react_1.default.createElement(Card_1.DisabledInfo, null, this.getDisabledAccountDesc()),
            status === 'inactive' && (react_1.default.createElement(IconLabel_1.default, { icon: "companyName" }, companyName))));
    };
    Profilecard.prototype.getDisabledAccountName = function () {
        var _a = this.props, nickname = _a.nickname, fullName = _a.fullName, status = _a.status;
        if (status === 'inactive') {
            return fullName || nickname;
        }
        else if (status === 'closed') {
            return (nickname || (react_1.default.createElement(react_intl_1.FormattedMessage, tslib_1.__assign({}, messages_1.default.disabledAccountDefaultName))));
        }
        return null;
    };
    Profilecard.prototype.getDisabledAccountDesc = function () {
        var _a = this.props, _b = _a.status, status = _b === void 0 ? 'closed' : _b, statusModifiedDate = _a.statusModifiedDate, disabledAccountMessage = _a.disabledAccountMessage;
        var date = statusModifiedDate
            ? new Date(statusModifiedDate * 1000)
            : null;
        var relativeDateKey = relative_date_1.default(date);
        // consumer does not want to use built-in message
        if (disabledAccountMessage) {
            return disabledAccountMessage;
        }
        var secondSentence = null;
        if (relativeDateKey) {
            secondSentence = (react_1.default.createElement(react_intl_1.FormattedMessage
            // @ts-ignore
            , tslib_1.__assign({}, messages_1.default[status + "AccountDescMsgHasDate" + relativeDateKey])));
        }
        else {
            secondSentence = (
            // @ts-ignore
            react_1.default.createElement(react_intl_1.FormattedMessage, tslib_1.__assign({}, messages_1.default[status + "AccountDescMsgNoDate"])));
        }
        return (react_1.default.createElement("p", null,
            react_1.default.createElement(react_intl_1.FormattedMessage, tslib_1.__assign({}, messages_1.default.generalDescMsgForDisabledUser)),
            ' ',
            secondSentence));
    };
    Profilecard.prototype.renderFullNameAndPublicName = function (meta) {
        var _a = this.props, nickname = _a.nickname, fullName = _a.fullName;
        if (!fullName && !nickname) {
            return null;
        }
        var displayName = fullName === nickname
            ? fullName
            : "" + fullName + (nickname ? " (" + nickname + ") " : '');
        return react_1.default.createElement(Card_1.FullNameLabel, { noMeta: !meta }, displayName);
    };
    Profilecard.prototype.renderCardDetailsApp = function () {
        return (react_1.default.createElement(Card_1.DetailsGroup, null,
            this.renderFullNameAndPublicName(),
            react_1.default.createElement(Card_1.AppTitleLabel, null, "App")));
    };
    Profilecard.prototype.renderCardDetails = function () {
        var _a = this.props, isBot = _a.isBot, status = _a.status;
        if (isBot) {
            return this.renderCardDetailsApp();
        }
        if (status === 'inactive' || status === 'closed') {
            return this.renderCardDetailsForDisabledAccount();
        }
        return this.renderCardDetailsDefault();
    };
    Profilecard.prototype.render = function () {
        var _a = this.props, fullName = _a.fullName, status = _a.status, customElevation = _a.customElevation;
        var cardContent = null;
        // @FIXME do closed users have empty fullName field?
        var canRender = fullName || status === 'closed';
        if (this.props.hasError) {
            this.callAnalytics(analytics_1.AnalyticsName.PROFILE_CARD_ERROR, {});
            cardContent = this.renderErrorMessage();
        }
        else if (this.props.isLoading) {
            cardContent = (react_1.default.createElement(Card_1.SpinnerContainer, null,
                react_1.default.createElement(spinner_1.default, null)));
        }
        else if (canRender) {
            var isDisabledUser = status === 'inactive' || status === 'closed';
            var actions = this.renderActionsButtons();
            this.callAnalytics(analytics_1.AnalyticsName.PROFILE_CARD_LOADED, {
                duration: this.durationSince(this.timeOpen),
            });
            cardContent = (react_1.default.createElement(Card_1.CardContainer, { isDisabledUser: isDisabledUser },
                react_1.default.createElement(Card_1.ProfileImage, null,
                    react_1.default.createElement(avatar_1.default, { size: "xlarge", src: this.props.status !== 'closed'
                            ? this.props.avatarUrl
                            : undefined, borderColor: colors_1.N0 })),
                react_1.default.createElement(Card_1.CardContent, null,
                    this.renderCardDetails(),
                    actions ? (react_1.default.createElement(react_1.default.Fragment, null,
                        react_1.default.createElement(Card_1.ActionsFlexSpacer, null),
                        actions)) : null)));
        }
        return (react_1.default.createElement(Card_1.CardElevationWrapper, { customElevation: customElevation }, cardContent));
    };
    Profilecard.defaultProps = {
        isLoading: false,
        hasError: false,
        errorType: null,
        status: 'active',
        isBot: false,
        isNotMentionable: false,
        actions: [],
        hasDisabledAccountLozenge: true,
        customElevation: 'e200',
        analytics: function () { return null; },
        clientFetchProfile: function () { return null; },
    };
    return Profilecard;
}(react_1.default.PureComponent));
exports.default = Profilecard;
