// @atlaskit/media-picker/dist/es2019/components/browser/index.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import React from 'react';
export class BrowserLoader extends React.PureComponent {
  constructor(...args) {
    super(...args);
    _defineProperty(this, "mounted", false);
    _defineProperty(this, "state", {
      Browser: BrowserLoader.Browser
    });
  }
  componentDidMount() {
    this.mounted = true;
  }
  componentWillUnmount() {
    this.mounted = false;
  }
  async UNSAFE_componentWillMount() {
    if (!this.state.Browser) {
      const [mediaClient, browserModule] = await Promise.all([import( /* webpackChunkName: "@atlaskit-internal_media-client-react" */'@atlaskit/media-client-react'), import( /* webpackChunkName: "@atlaskit-internal_media-browser" */'./browser')]);

      // @ts-ignore: [PIT-1685] Fails in post-office due to backwards incompatibility issue with React 18
      BrowserLoader.Browser = mediaClient.withMediaClient(browserModule.Browser);
      if (this.mounted) {
        this.setState({
          Browser: BrowserLoader.Browser
        });
      }
    }
  }
  render() {
    if (!this.state.Browser) {
      return null;
    }
    return /*#__PURE__*/React.createElement(this.state.Browser, this.props);
  }
}
_defineProperty(BrowserLoader, "displayName", 'AsyncBrowser');
